{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240617/sc-E2G/test/results/GM12878/GM12878.multiome_7features/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240617/sc-E2G/test/results/GM12878/GM12878.multiome_7features/Predictions/EnhancerPredictionsAllPutative.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240617/sc-E2G/test/results/GM12878/GM12878.multiome_7features/Kendall/atac_matrix.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/GM12878/1.prepare_data/1.seurat_pipeline.240615/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/GM12878/2.Genome_wide_prediction/FigR/FigR.250211/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2637980 ranges and 3 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100028533-100029101      * |         AGL\n",
       "        [2]     chr1 100028533-100029101      * |      CDC14A\n",
       "        [3]     chr1 100028533-100029101      * |         DBT\n",
       "        [4]     chr1 100028533-100029101      * |        DPH5\n",
       "        [5]     chr1 100028533-100029101      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2637976]     chrX     9995654-9996249      * |       TBL1X\n",
       "  [2637977]     chrX     9995654-9996249      * |        WWC3\n",
       "  [2637978]     chrX     9997430-9997980      * |       CLCN4\n",
       "  [2637979]     chrX     9997430-9997980      * |       TBL1X\n",
       "  [2637980]     chrX     9997430-9997980      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-100028533-10002.. chr1-100028533-10002..\n",
       "        [2] chr1-100028533-10002.. chr1-100028533-10002..\n",
       "        [3] chr1-100028533-10002.. chr1-100028533-10002..\n",
       "        [4] chr1-100028533-10002.. chr1-100028533-10002..\n",
       "        [5] chr1-100028533-10002.. chr1-100028533-10002..\n",
       "        ...                    ...                    ...\n",
       "  [2637976]   chrX-9995654-9996249 chrX-9995654-9996249..\n",
       "  [2637977]   chrX-9995654-9996249 chrX-9995654-9996249..\n",
       "  [2637978]   chrX-9997430-9997980 chrX-9997430-9997980..\n",
       "  [2637979]   chrX-9997430-9997980 chrX-9997430-9997980..\n",
       "  [2637980]   chrX-9997430-9997980 chrX-9997430-9997980..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  3050 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dd80eda-2392-4d59-8250-1ae62d7e1d1e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b9e26c70-11fd-47c4-8e05-9c96b17d3f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a76b1603-6b00-4ac0-9399-fd341c5a7b88",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 155270 3050 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(155270): chr1-100028533-100029101 chr1-100029136-100029435 ...\n",
       "  chrX-9995654-9996249 chrX-9997430-9997980\n",
       "rowData names(2): PeakName bias\n",
       "colnames(3050): GM12878_AAACAGCCACTAAGAA-1 GM12878_AAACAGCCATAATCAC-1\n",
       "  ... GM12878_TTTGTTGGTCACAAAT-1 GM12878_TTTGTTGGTTGGTGAC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna.rename),\n",
    "                       index = 1:nrow(matrix.rna.rename))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  18.7052869796753 secs \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Time Elapsed:  35.2174661636353 mins \n",
      "\n",
      "[1] \"chr10\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.98847460746765 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.6209397395452 mins \n",
      "\n",
      "[1] \"chr11\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.8279700279236 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  23.1116465449333 mins \n",
      "\n",
      "[1] \"chr12\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.1825973987579 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  17.9215569297473 mins \n",
      "\n",
      "[1] \"chr13\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  3.90281009674072 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  6.75855862696966 mins \n",
      "\n",
      "[1] \"chr14\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.07729315757751 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  12.0937140345573 mins \n",
      "\n",
      "[1] \"chr15\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.60479092597961 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  12.7588753064473 mins \n",
      "\n",
      "[1] \"chr16\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  11.5097787380219 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  20.4591219186783 mins \n",
      "\n",
      "[1] \"chr17\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  12.9027438163757 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  25.8688065807025 mins \n",
      "\n",
      "[1] \"chr18\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.01811146736145 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  6.89790953397751 mins \n",
      "\n",
      "[1] \"chr19\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  20.9073295593262 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  37.0333812355995 mins \n",
      "\n",
      "[1] \"chr2\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  9.41478300094604 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  14.7206319411596 mins \n",
      "\n",
      "[1] \"chr20\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.2183301448822 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.00493383804957 mins \n",
      "\n",
      "[1] \"chr21\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.69114398956299 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  4.28450955152512 mins \n",
      "\n",
      "[1] \"chr22\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.312664270401 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.18652712504069 mins \n",
      "\n",
      "[1] \"chr3\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.53203964233398 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  14.0365696748098 mins \n",
      "\n",
      "[1] \"chr4\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.01591610908508 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.48756671746572 mins \n",
      "\n",
      "[1] \"chr5\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.93576574325562 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  10.8665943304698 mins \n",
      "\n",
      "[1] \"chr6\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.38404822349548 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  17.0325606624285 mins \n",
      "\n",
      "[1] \"chr7\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.7531754970551 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  15.8161489486694 mins \n",
      "\n",
      "[1] \"chr8\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.05320453643799 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.5752641121546 mins \n",
      "\n",
      "[1] \"chr9\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  9.04970669746399 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  14.9808695912361 mins \n",
      "\n",
      "[1] \"chrX\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.86609983444214 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  10.2661388317744 mins \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna.rename,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2533481 ranges and 7 metadata columns:\n",
       "       seqnames              ranges strand |  TargetGene               PeakName\n",
       "          <Rle>           <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 100028533-100029101      * |         AGL chr1-100028533-10002..\n",
       "  chr1     chr1 100028533-100029101      * |      CDC14A chr1-100028533-10002..\n",
       "  chr1     chr1 100028533-100029101      * |         DBT chr1-100028533-10002..\n",
       "  chr1     chr1 100028533-100029101      * |        DPH5 chr1-100028533-10002..\n",
       "  chr1     chr1 100028533-100029101      * |       EXTL2 chr1-100028533-10002..\n",
       "   ...      ...                 ...    ... .         ...                    ...\n",
       "  chrX     chrX     9995654-9996249      * |       TBL1X   chrX-9995654-9996249\n",
       "  chrX     chrX     9995654-9996249      * |        WWC3   chrX-9995654-9996249\n",
       "  chrX     chrX     9997430-9997980      * |       CLCN4   chrX-9997430-9997980\n",
       "  chrX     chrX     9997430-9997980      * |       TBL1X   chrX-9997430-9997980\n",
       "  chrX     chrX     9997430-9997980      * |        WWC3   chrX-9997430-9997980\n",
       "                     PairName         rObs     rBgSD     rBgMean     pvalZ\n",
       "                  <character>    <numeric> <numeric>   <numeric> <numeric>\n",
       "  chr1 chr1-100028533-10002..    0.0133078 0.0172454  0.01140358 0.4560396\n",
       "  chr1 chr1-100028533-10002..    0.0104488 0.0208114  0.01930096 0.6647091\n",
       "  chr1 chr1-100028533-10002..    0.0224650 0.0168418  0.00767372 0.1899049\n",
       "  chr1 chr1-100028533-10002..    0.0450955 0.0181122  0.01482144 0.0473145\n",
       "  chr1 chr1-100028533-10002..    0.0284114 0.0271716  0.02840211 0.4998631\n",
       "   ...                    ...          ...       ...         ...       ...\n",
       "  chrX chrX-9995654-9996249.. -0.006495872 0.0221791 -0.00693199 0.4921560\n",
       "  chrX chrX-9995654-9996249.. -0.010247842 0.0218800  0.01017632 0.8247090\n",
       "  chrX chrX-9997430-9997980.. -0.012348587 0.0186806  0.00251965 0.7869608\n",
       "  chrX chrX-9997430-9997980..  0.000772615 0.0227595 -0.00529658 0.3948630\n",
       "  chrX chrX-9997430-9997980..  0.043322815 0.0196999  0.00478476 0.0252175\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2533481 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>AGL     </td><td>K562</td><td> 0.013307751</td><td>0.01724536</td><td> 0.0114035817</td><td>0.45603960</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>CDC14A  </td><td>K562</td><td> 0.010448850</td><td>0.02081137</td><td> 0.0193009567</td><td>0.66470909</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>DBT     </td><td>K562</td><td> 0.022465002</td><td>0.01684183</td><td> 0.0076737166</td><td>0.18990487</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>DPH5    </td><td>K562</td><td> 0.045095506</td><td>0.01811224</td><td> 0.0148214401</td><td>0.04731448</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>EXTL2   </td><td>K562</td><td> 0.028411431</td><td>0.02717164</td><td> 0.0284021086</td><td>0.49986313</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>FRRS1   </td><td>K562</td><td> 0.017150919</td><td>0.01844658</td><td> 0.0163196487</td><td>0.48202829</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>MFSD14A </td><td>K562</td><td>-0.008365932</td><td>0.01832878</td><td> 0.0016496906</td><td>0.70761904</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>RTCA    </td><td>K562</td><td> 0.006028360</td><td>0.01974283</td><td> 0.0050826383</td><td>0.48089716</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>RTCA-AS1</td><td>K562</td><td>-0.010989079</td><td>0.01701535</td><td> 0.0028612812</td><td>0.79217529</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>SASS6   </td><td>K562</td><td>-0.002251422</td><td>0.01832740</td><td> 0.0002148152</td><td>0.55352232</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>SLC30A7 </td><td>K562</td><td> 0.003329924</td><td>0.01804794</td><td> 0.0130543615</td><td>0.70499157</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>SLC35A3 </td><td>K562</td><td> 0.022953477</td><td>0.01917905</td><td> 0.0015917885</td><td>0.13268180</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>TRMT13  </td><td>K562</td><td> 0.018339888</td><td>0.01635313</td><td> 0.0055654495</td><td>0.21735367</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>VCAM1   </td><td>K562</td><td> 0.053572989</td><td>0.02846444</td><td> 0.0224872506</td><td>0.13739673</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>AGL     </td><td>K562</td><td>-0.004457632</td><td>0.01789879</td><td> 0.0062081841</td><td>0.72437765</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>CDC14A  </td><td>K562</td><td>-0.003466294</td><td>0.01900510</td><td> 0.0044877707</td><td>0.66221745</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>DBT     </td><td>K562</td><td> 0.019892904</td><td>0.01776676</td><td> 0.0043469431</td><td>0.19078630</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>DPH5    </td><td>K562</td><td> 0.013616953</td><td>0.01802728</td><td> 0.0029149053</td><td>0.27637029</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>EXTL2   </td><td>K562</td><td> 0.009816065</td><td>0.02036463</td><td> 0.0126175149</td><td>0.55470771</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>FRRS1   </td><td>K562</td><td> 0.008823450</td><td>0.01927072</td><td> 0.0073734844</td><td>0.47001111</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>MFSD14A </td><td>K562</td><td> 0.013446344</td><td>0.01865863</td><td> 0.0033302081</td><td>0.29385093</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>RTCA    </td><td>K562</td><td> 0.036738457</td><td>0.01687865</td><td> 0.0031420616</td><td>0.02326974</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>RTCA-AS1</td><td>K562</td><td>-0.021766504</td><td>0.01929994</td><td> 0.0033811497</td><td>0.90371113</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>SASS6   </td><td>K562</td><td> 0.016385121</td><td>0.01840021</td><td>-0.0038076795</td><td>0.13622836</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>SLC30A7 </td><td>K562</td><td> 0.009428935</td><td>0.01613040</td><td> 0.0053047373</td><td>0.39909951</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>SLC35A3 </td><td>K562</td><td> 0.010946416</td><td>0.01700453</td><td> 0.0005722155</td><td>0.27090291</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>TRMT13  </td><td>K562</td><td> 0.009483572</td><td>0.01758771</td><td> 0.0040817037</td><td>0.37936885</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>VCAM1   </td><td>K562</td><td> 0.004404926</td><td>0.02497601</td><td> 0.0130506060</td><td>0.63538853</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034578</td><td>100034758</td><td>AGL     </td><td>K562</td><td>-0.015791236</td><td>0.01845864</td><td> 0.0172414562</td><td>0.96323697</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034578</td><td>100034758</td><td>CDC14A  </td><td>K562</td><td> 0.032417317</td><td>0.02070616</td><td> 0.0201888888</td><td>0.27740438</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>ASMT     </td><td>K562</td><td> 3.125955e-02</td><td>0.01931998</td><td> 0.0040569059</td><td>0.079564733</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>ASMTL    </td><td>K562</td><td> 2.048386e-02</td><td>0.02209108</td><td> 0.0257922007</td><td>0.594948530</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>CSF2RA   </td><td>K562</td><td>-9.391297e-04</td><td>0.02021967</td><td> 0.0171326552</td><td>0.814278180</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>GTPBP6   </td><td>K562</td><td>-3.506227e-03</td><td>0.02064658</td><td> 0.0199335677</td><td>0.871872489</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>IL3RA    </td><td>K562</td><td> 4.332036e-03</td><td>0.03381729</td><td> 0.0324357718</td><td>0.797026302</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>LINC00106</td><td>K562</td><td>-1.813808e-02</td><td>0.02188647</td><td> 0.0293285566</td><td>0.984949788</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>P2RY8    </td><td>K562</td><td>-1.336805e-02</td><td>0.03367565</td><td>-0.0220174847</td><td>0.398649083</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>PLCXD1   </td><td>K562</td><td> 1.536749e-02</td><td>0.02003704</td><td> 0.0159153994</td><td>0.510907717</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>PPP2R3B  </td><td>K562</td><td> 1.767213e-02</td><td>0.01981576</td><td> 0.0254860972</td><td>0.653331660</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>SLC25A6  </td><td>K562</td><td> 2.899253e-02</td><td>0.02336210</td><td>-0.0125439343</td><td>0.037706664</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988858</td><td>9989079</td><td>CLCN4    </td><td>K562</td><td>-7.586260e-03</td><td>0.01855314</td><td> 0.0005051081</td><td>0.668624599</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988858</td><td>9989079</td><td>TBL1X    </td><td>K562</td><td> 8.118087e-05</td><td>0.01788273</td><td>-0.0024474925</td><td>0.443775769</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988858</td><td>9989079</td><td>WWC3     </td><td>K562</td><td> 5.408002e-02</td><td>0.01979240</td><td> 0.0060411130</td><td>0.007609207</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>AKAP17A  </td><td>K562</td><td>-4.641255e-03</td><td>0.02126370</td><td>-0.0001636529</td><td>0.583390526</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>ASMT     </td><td>K562</td><td>-1.834867e-02</td><td>0.01718280</td><td> 0.0015935674</td><td>0.877096301</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>ASMTL    </td><td>K562</td><td>-1.917408e-02</td><td>0.02357437</td><td> 0.0179116579</td><td>0.942156539</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>CSF2RA   </td><td>K562</td><td>-2.865688e-03</td><td>0.01800071</td><td> 0.0164488106</td><td>0.858361157</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>GTPBP6   </td><td>K562</td><td> 3.862025e-02</td><td>0.01755228</td><td> 0.0165311281</td><td>0.104109794</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>IL3RA    </td><td>K562</td><td> 9.658144e-04</td><td>0.03359537</td><td> 0.0279406956</td><td>0.788993671</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>LINC00106</td><td>K562</td><td> 5.215708e-03</td><td>0.02330403</td><td> 0.0274524535</td><td>0.830009209</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>P2RY8    </td><td>K562</td><td> 1.742263e-03</td><td>0.04022438</td><td>-0.0168141337</td><td>0.322283737</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>PLCXD1   </td><td>K562</td><td>-5.122071e-03</td><td>0.01983656</td><td> 0.0198253972</td><td>0.895741020</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>PPP2R3B  </td><td>K562</td><td> 3.685852e-02</td><td>0.02172516</td><td> 0.0283532622</td><td>0.347716329</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>SLC25A6  </td><td>K562</td><td>-7.182472e-03</td><td>0.02481795</td><td>-0.0182731899</td><td>0.327479797</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995654</td><td>9996249</td><td>CLCN4    </td><td>K562</td><td> 3.161529e-02</td><td>0.01976669</td><td> 0.0053854685</td><td>0.092259140</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995654</td><td>9996249</td><td>TBL1X    </td><td>K562</td><td>-6.495872e-03</td><td>0.02217910</td><td>-0.0069319861</td><td>0.492155991</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995654</td><td>9996249</td><td>WWC3     </td><td>K562</td><td>-1.024784e-02</td><td>0.02188004</td><td> 0.0101763166</td><td>0.824708976</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997430</td><td>9997980</td><td>CLCN4    </td><td>K562</td><td>-1.234859e-02</td><td>0.01868056</td><td> 0.0025196485</td><td>0.786960837</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997430</td><td>9997980</td><td>TBL1X    </td><td>K562</td><td> 7.726147e-04</td><td>0.02275951</td><td>-0.0052965816</td><td>0.394862997</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997430</td><td>9997980</td><td>WWC3     </td><td>K562</td><td> 4.332281e-02</td><td>0.01969990</td><td> 0.0047847627</td><td>0.025217504</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2533481 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100028533 & 100029101 & AGL      & K562 &  0.013307751 & 0.01724536 &  0.0114035817 & 0.45603960\\\\\n",
       "\t chr1 & 100028533 & 100029101 & CDC14A   & K562 &  0.010448850 & 0.02081137 &  0.0193009567 & 0.66470909\\\\\n",
       "\t chr1 & 100028533 & 100029101 & DBT      & K562 &  0.022465002 & 0.01684183 &  0.0076737166 & 0.18990487\\\\\n",
       "\t chr1 & 100028533 & 100029101 & DPH5     & K562 &  0.045095506 & 0.01811224 &  0.0148214401 & 0.04731448\\\\\n",
       "\t chr1 & 100028533 & 100029101 & EXTL2    & K562 &  0.028411431 & 0.02717164 &  0.0284021086 & 0.49986313\\\\\n",
       "\t chr1 & 100028533 & 100029101 & FRRS1    & K562 &  0.017150919 & 0.01844658 &  0.0163196487 & 0.48202829\\\\\n",
       "\t chr1 & 100028533 & 100029101 & MFSD14A  & K562 & -0.008365932 & 0.01832878 &  0.0016496906 & 0.70761904\\\\\n",
       "\t chr1 & 100028533 & 100029101 & RTCA     & K562 &  0.006028360 & 0.01974283 &  0.0050826383 & 0.48089716\\\\\n",
       "\t chr1 & 100028533 & 100029101 & RTCA-AS1 & K562 & -0.010989079 & 0.01701535 &  0.0028612812 & 0.79217529\\\\\n",
       "\t chr1 & 100028533 & 100029101 & SASS6    & K562 & -0.002251422 & 0.01832740 &  0.0002148152 & 0.55352232\\\\\n",
       "\t chr1 & 100028533 & 100029101 & SLC30A7  & K562 &  0.003329924 & 0.01804794 &  0.0130543615 & 0.70499157\\\\\n",
       "\t chr1 & 100028533 & 100029101 & SLC35A3  & K562 &  0.022953477 & 0.01917905 &  0.0015917885 & 0.13268180\\\\\n",
       "\t chr1 & 100028533 & 100029101 & TRMT13   & K562 &  0.018339888 & 0.01635313 &  0.0055654495 & 0.21735367\\\\\n",
       "\t chr1 & 100028533 & 100029101 & VCAM1    & K562 &  0.053572989 & 0.02846444 &  0.0224872506 & 0.13739673\\\\\n",
       "\t chr1 & 100029136 & 100029435 & AGL      & K562 & -0.004457632 & 0.01789879 &  0.0062081841 & 0.72437765\\\\\n",
       "\t chr1 & 100029136 & 100029435 & CDC14A   & K562 & -0.003466294 & 0.01900510 &  0.0044877707 & 0.66221745\\\\\n",
       "\t chr1 & 100029136 & 100029435 & DBT      & K562 &  0.019892904 & 0.01776676 &  0.0043469431 & 0.19078630\\\\\n",
       "\t chr1 & 100029136 & 100029435 & DPH5     & K562 &  0.013616953 & 0.01802728 &  0.0029149053 & 0.27637029\\\\\n",
       "\t chr1 & 100029136 & 100029435 & EXTL2    & K562 &  0.009816065 & 0.02036463 &  0.0126175149 & 0.55470771\\\\\n",
       "\t chr1 & 100029136 & 100029435 & FRRS1    & K562 &  0.008823450 & 0.01927072 &  0.0073734844 & 0.47001111\\\\\n",
       "\t chr1 & 100029136 & 100029435 & MFSD14A  & K562 &  0.013446344 & 0.01865863 &  0.0033302081 & 0.29385093\\\\\n",
       "\t chr1 & 100029136 & 100029435 & RTCA     & K562 &  0.036738457 & 0.01687865 &  0.0031420616 & 0.02326974\\\\\n",
       "\t chr1 & 100029136 & 100029435 & RTCA-AS1 & K562 & -0.021766504 & 0.01929994 &  0.0033811497 & 0.90371113\\\\\n",
       "\t chr1 & 100029136 & 100029435 & SASS6    & K562 &  0.016385121 & 0.01840021 & -0.0038076795 & 0.13622836\\\\\n",
       "\t chr1 & 100029136 & 100029435 & SLC30A7  & K562 &  0.009428935 & 0.01613040 &  0.0053047373 & 0.39909951\\\\\n",
       "\t chr1 & 100029136 & 100029435 & SLC35A3  & K562 &  0.010946416 & 0.01700453 &  0.0005722155 & 0.27090291\\\\\n",
       "\t chr1 & 100029136 & 100029435 & TRMT13   & K562 &  0.009483572 & 0.01758771 &  0.0040817037 & 0.37936885\\\\\n",
       "\t chr1 & 100029136 & 100029435 & VCAM1    & K562 &  0.004404926 & 0.02497601 &  0.0130506060 & 0.63538853\\\\\n",
       "\t chr1 & 100034578 & 100034758 & AGL      & K562 & -0.015791236 & 0.01845864 &  0.0172414562 & 0.96323697\\\\\n",
       "\t chr1 & 100034578 & 100034758 & CDC14A   & K562 &  0.032417317 & 0.02070616 &  0.0201888888 & 0.27740438\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  997541 &  998270 & ASMT      & K562 &  3.125955e-02 & 0.01931998 &  0.0040569059 & 0.079564733\\\\\n",
       "\t chrX &  997541 &  998270 & ASMTL     & K562 &  2.048386e-02 & 0.02209108 &  0.0257922007 & 0.594948530\\\\\n",
       "\t chrX &  997541 &  998270 & CSF2RA    & K562 & -9.391297e-04 & 0.02021967 &  0.0171326552 & 0.814278180\\\\\n",
       "\t chrX &  997541 &  998270 & GTPBP6    & K562 & -3.506227e-03 & 0.02064658 &  0.0199335677 & 0.871872489\\\\\n",
       "\t chrX &  997541 &  998270 & IL3RA     & K562 &  4.332036e-03 & 0.03381729 &  0.0324357718 & 0.797026302\\\\\n",
       "\t chrX &  997541 &  998270 & LINC00106 & K562 & -1.813808e-02 & 0.02188647 &  0.0293285566 & 0.984949788\\\\\n",
       "\t chrX &  997541 &  998270 & P2RY8     & K562 & -1.336805e-02 & 0.03367565 & -0.0220174847 & 0.398649083\\\\\n",
       "\t chrX &  997541 &  998270 & PLCXD1    & K562 &  1.536749e-02 & 0.02003704 &  0.0159153994 & 0.510907717\\\\\n",
       "\t chrX &  997541 &  998270 & PPP2R3B   & K562 &  1.767213e-02 & 0.01981576 &  0.0254860972 & 0.653331660\\\\\n",
       "\t chrX &  997541 &  998270 & SLC25A6   & K562 &  2.899253e-02 & 0.02336210 & -0.0125439343 & 0.037706664\\\\\n",
       "\t chrX & 9988858 & 9989079 & CLCN4     & K562 & -7.586260e-03 & 0.01855314 &  0.0005051081 & 0.668624599\\\\\n",
       "\t chrX & 9988858 & 9989079 & TBL1X     & K562 &  8.118087e-05 & 0.01788273 & -0.0024474925 & 0.443775769\\\\\n",
       "\t chrX & 9988858 & 9989079 & WWC3      & K562 &  5.408002e-02 & 0.01979240 &  0.0060411130 & 0.007609207\\\\\n",
       "\t chrX &  999293 &  999875 & AKAP17A   & K562 & -4.641255e-03 & 0.02126370 & -0.0001636529 & 0.583390526\\\\\n",
       "\t chrX &  999293 &  999875 & ASMT      & K562 & -1.834867e-02 & 0.01718280 &  0.0015935674 & 0.877096301\\\\\n",
       "\t chrX &  999293 &  999875 & ASMTL     & K562 & -1.917408e-02 & 0.02357437 &  0.0179116579 & 0.942156539\\\\\n",
       "\t chrX &  999293 &  999875 & CSF2RA    & K562 & -2.865688e-03 & 0.01800071 &  0.0164488106 & 0.858361157\\\\\n",
       "\t chrX &  999293 &  999875 & GTPBP6    & K562 &  3.862025e-02 & 0.01755228 &  0.0165311281 & 0.104109794\\\\\n",
       "\t chrX &  999293 &  999875 & IL3RA     & K562 &  9.658144e-04 & 0.03359537 &  0.0279406956 & 0.788993671\\\\\n",
       "\t chrX &  999293 &  999875 & LINC00106 & K562 &  5.215708e-03 & 0.02330403 &  0.0274524535 & 0.830009209\\\\\n",
       "\t chrX &  999293 &  999875 & P2RY8     & K562 &  1.742263e-03 & 0.04022438 & -0.0168141337 & 0.322283737\\\\\n",
       "\t chrX &  999293 &  999875 & PLCXD1    & K562 & -5.122071e-03 & 0.01983656 &  0.0198253972 & 0.895741020\\\\\n",
       "\t chrX &  999293 &  999875 & PPP2R3B   & K562 &  3.685852e-02 & 0.02172516 &  0.0283532622 & 0.347716329\\\\\n",
       "\t chrX &  999293 &  999875 & SLC25A6   & K562 & -7.182472e-03 & 0.02481795 & -0.0182731899 & 0.327479797\\\\\n",
       "\t chrX & 9995654 & 9996249 & CLCN4     & K562 &  3.161529e-02 & 0.01976669 &  0.0053854685 & 0.092259140\\\\\n",
       "\t chrX & 9995654 & 9996249 & TBL1X     & K562 & -6.495872e-03 & 0.02217910 & -0.0069319861 & 0.492155991\\\\\n",
       "\t chrX & 9995654 & 9996249 & WWC3      & K562 & -1.024784e-02 & 0.02188004 &  0.0101763166 & 0.824708976\\\\\n",
       "\t chrX & 9997430 & 9997980 & CLCN4     & K562 & -1.234859e-02 & 0.01868056 &  0.0025196485 & 0.786960837\\\\\n",
       "\t chrX & 9997430 & 9997980 & TBL1X     & K562 &  7.726147e-04 & 0.02275951 & -0.0052965816 & 0.394862997\\\\\n",
       "\t chrX & 9997430 & 9997980 & WWC3      & K562 &  4.332281e-02 & 0.01969990 &  0.0047847627 & 0.025217504\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2533481 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100028533 | 100029101 | AGL      | K562 |  0.013307751 | 0.01724536 |  0.0114035817 | 0.45603960 |\n",
       "| chr1 | 100028533 | 100029101 | CDC14A   | K562 |  0.010448850 | 0.02081137 |  0.0193009567 | 0.66470909 |\n",
       "| chr1 | 100028533 | 100029101 | DBT      | K562 |  0.022465002 | 0.01684183 |  0.0076737166 | 0.18990487 |\n",
       "| chr1 | 100028533 | 100029101 | DPH5     | K562 |  0.045095506 | 0.01811224 |  0.0148214401 | 0.04731448 |\n",
       "| chr1 | 100028533 | 100029101 | EXTL2    | K562 |  0.028411431 | 0.02717164 |  0.0284021086 | 0.49986313 |\n",
       "| chr1 | 100028533 | 100029101 | FRRS1    | K562 |  0.017150919 | 0.01844658 |  0.0163196487 | 0.48202829 |\n",
       "| chr1 | 100028533 | 100029101 | MFSD14A  | K562 | -0.008365932 | 0.01832878 |  0.0016496906 | 0.70761904 |\n",
       "| chr1 | 100028533 | 100029101 | RTCA     | K562 |  0.006028360 | 0.01974283 |  0.0050826383 | 0.48089716 |\n",
       "| chr1 | 100028533 | 100029101 | RTCA-AS1 | K562 | -0.010989079 | 0.01701535 |  0.0028612812 | 0.79217529 |\n",
       "| chr1 | 100028533 | 100029101 | SASS6    | K562 | -0.002251422 | 0.01832740 |  0.0002148152 | 0.55352232 |\n",
       "| chr1 | 100028533 | 100029101 | SLC30A7  | K562 |  0.003329924 | 0.01804794 |  0.0130543615 | 0.70499157 |\n",
       "| chr1 | 100028533 | 100029101 | SLC35A3  | K562 |  0.022953477 | 0.01917905 |  0.0015917885 | 0.13268180 |\n",
       "| chr1 | 100028533 | 100029101 | TRMT13   | K562 |  0.018339888 | 0.01635313 |  0.0055654495 | 0.21735367 |\n",
       "| chr1 | 100028533 | 100029101 | VCAM1    | K562 |  0.053572989 | 0.02846444 |  0.0224872506 | 0.13739673 |\n",
       "| chr1 | 100029136 | 100029435 | AGL      | K562 | -0.004457632 | 0.01789879 |  0.0062081841 | 0.72437765 |\n",
       "| chr1 | 100029136 | 100029435 | CDC14A   | K562 | -0.003466294 | 0.01900510 |  0.0044877707 | 0.66221745 |\n",
       "| chr1 | 100029136 | 100029435 | DBT      | K562 |  0.019892904 | 0.01776676 |  0.0043469431 | 0.19078630 |\n",
       "| chr1 | 100029136 | 100029435 | DPH5     | K562 |  0.013616953 | 0.01802728 |  0.0029149053 | 0.27637029 |\n",
       "| chr1 | 100029136 | 100029435 | EXTL2    | K562 |  0.009816065 | 0.02036463 |  0.0126175149 | 0.55470771 |\n",
       "| chr1 | 100029136 | 100029435 | FRRS1    | K562 |  0.008823450 | 0.01927072 |  0.0073734844 | 0.47001111 |\n",
       "| chr1 | 100029136 | 100029435 | MFSD14A  | K562 |  0.013446344 | 0.01865863 |  0.0033302081 | 0.29385093 |\n",
       "| chr1 | 100029136 | 100029435 | RTCA     | K562 |  0.036738457 | 0.01687865 |  0.0031420616 | 0.02326974 |\n",
       "| chr1 | 100029136 | 100029435 | RTCA-AS1 | K562 | -0.021766504 | 0.01929994 |  0.0033811497 | 0.90371113 |\n",
       "| chr1 | 100029136 | 100029435 | SASS6    | K562 |  0.016385121 | 0.01840021 | -0.0038076795 | 0.13622836 |\n",
       "| chr1 | 100029136 | 100029435 | SLC30A7  | K562 |  0.009428935 | 0.01613040 |  0.0053047373 | 0.39909951 |\n",
       "| chr1 | 100029136 | 100029435 | SLC35A3  | K562 |  0.010946416 | 0.01700453 |  0.0005722155 | 0.27090291 |\n",
       "| chr1 | 100029136 | 100029435 | TRMT13   | K562 |  0.009483572 | 0.01758771 |  0.0040817037 | 0.37936885 |\n",
       "| chr1 | 100029136 | 100029435 | VCAM1    | K562 |  0.004404926 | 0.02497601 |  0.0130506060 | 0.63538853 |\n",
       "| chr1 | 100034578 | 100034758 | AGL      | K562 | -0.015791236 | 0.01845864 |  0.0172414562 | 0.96323697 |\n",
       "| chr1 | 100034578 | 100034758 | CDC14A   | K562 |  0.032417317 | 0.02070616 |  0.0201888888 | 0.27740438 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  997541 |  998270 | ASMT      | K562 |  3.125955e-02 | 0.01931998 |  0.0040569059 | 0.079564733 |\n",
       "| chrX |  997541 |  998270 | ASMTL     | K562 |  2.048386e-02 | 0.02209108 |  0.0257922007 | 0.594948530 |\n",
       "| chrX |  997541 |  998270 | CSF2RA    | K562 | -9.391297e-04 | 0.02021967 |  0.0171326552 | 0.814278180 |\n",
       "| chrX |  997541 |  998270 | GTPBP6    | K562 | -3.506227e-03 | 0.02064658 |  0.0199335677 | 0.871872489 |\n",
       "| chrX |  997541 |  998270 | IL3RA     | K562 |  4.332036e-03 | 0.03381729 |  0.0324357718 | 0.797026302 |\n",
       "| chrX |  997541 |  998270 | LINC00106 | K562 | -1.813808e-02 | 0.02188647 |  0.0293285566 | 0.984949788 |\n",
       "| chrX |  997541 |  998270 | P2RY8     | K562 | -1.336805e-02 | 0.03367565 | -0.0220174847 | 0.398649083 |\n",
       "| chrX |  997541 |  998270 | PLCXD1    | K562 |  1.536749e-02 | 0.02003704 |  0.0159153994 | 0.510907717 |\n",
       "| chrX |  997541 |  998270 | PPP2R3B   | K562 |  1.767213e-02 | 0.01981576 |  0.0254860972 | 0.653331660 |\n",
       "| chrX |  997541 |  998270 | SLC25A6   | K562 |  2.899253e-02 | 0.02336210 | -0.0125439343 | 0.037706664 |\n",
       "| chrX | 9988858 | 9989079 | CLCN4     | K562 | -7.586260e-03 | 0.01855314 |  0.0005051081 | 0.668624599 |\n",
       "| chrX | 9988858 | 9989079 | TBL1X     | K562 |  8.118087e-05 | 0.01788273 | -0.0024474925 | 0.443775769 |\n",
       "| chrX | 9988858 | 9989079 | WWC3      | K562 |  5.408002e-02 | 0.01979240 |  0.0060411130 | 0.007609207 |\n",
       "| chrX |  999293 |  999875 | AKAP17A   | K562 | -4.641255e-03 | 0.02126370 | -0.0001636529 | 0.583390526 |\n",
       "| chrX |  999293 |  999875 | ASMT      | K562 | -1.834867e-02 | 0.01718280 |  0.0015935674 | 0.877096301 |\n",
       "| chrX |  999293 |  999875 | ASMTL     | K562 | -1.917408e-02 | 0.02357437 |  0.0179116579 | 0.942156539 |\n",
       "| chrX |  999293 |  999875 | CSF2RA    | K562 | -2.865688e-03 | 0.01800071 |  0.0164488106 | 0.858361157 |\n",
       "| chrX |  999293 |  999875 | GTPBP6    | K562 |  3.862025e-02 | 0.01755228 |  0.0165311281 | 0.104109794 |\n",
       "| chrX |  999293 |  999875 | IL3RA     | K562 |  9.658144e-04 | 0.03359537 |  0.0279406956 | 0.788993671 |\n",
       "| chrX |  999293 |  999875 | LINC00106 | K562 |  5.215708e-03 | 0.02330403 |  0.0274524535 | 0.830009209 |\n",
       "| chrX |  999293 |  999875 | P2RY8     | K562 |  1.742263e-03 | 0.04022438 | -0.0168141337 | 0.322283737 |\n",
       "| chrX |  999293 |  999875 | PLCXD1    | K562 | -5.122071e-03 | 0.01983656 |  0.0198253972 | 0.895741020 |\n",
       "| chrX |  999293 |  999875 | PPP2R3B   | K562 |  3.685852e-02 | 0.02172516 |  0.0283532622 | 0.347716329 |\n",
       "| chrX |  999293 |  999875 | SLC25A6   | K562 | -7.182472e-03 | 0.02481795 | -0.0182731899 | 0.327479797 |\n",
       "| chrX | 9995654 | 9996249 | CLCN4     | K562 |  3.161529e-02 | 0.01976669 |  0.0053854685 | 0.092259140 |\n",
       "| chrX | 9995654 | 9996249 | TBL1X     | K562 | -6.495872e-03 | 0.02217910 | -0.0069319861 | 0.492155991 |\n",
       "| chrX | 9995654 | 9996249 | WWC3      | K562 | -1.024784e-02 | 0.02188004 |  0.0101763166 | 0.824708976 |\n",
       "| chrX | 9997430 | 9997980 | CLCN4     | K562 | -1.234859e-02 | 0.01868056 |  0.0025196485 | 0.786960837 |\n",
       "| chrX | 9997430 | 9997980 | TBL1X     | K562 |  7.726147e-04 | 0.02275951 | -0.0052965816 | 0.394862997 |\n",
       "| chrX | 9997430 | 9997980 | WWC3      | K562 |  4.332281e-02 | 0.01969990 |  0.0047847627 | 0.025217504 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType rObs          rBgSD     \n",
       "1       chr1 100028533 100029101 AGL        K562      0.013307751  0.01724536\n",
       "2       chr1 100028533 100029101 CDC14A     K562      0.010448850  0.02081137\n",
       "3       chr1 100028533 100029101 DBT        K562      0.022465002  0.01684183\n",
       "4       chr1 100028533 100029101 DPH5       K562      0.045095506  0.01811224\n",
       "5       chr1 100028533 100029101 EXTL2      K562      0.028411431  0.02717164\n",
       "6       chr1 100028533 100029101 FRRS1      K562      0.017150919  0.01844658\n",
       "7       chr1 100028533 100029101 MFSD14A    K562     -0.008365932  0.01832878\n",
       "8       chr1 100028533 100029101 RTCA       K562      0.006028360  0.01974283\n",
       "9       chr1 100028533 100029101 RTCA-AS1   K562     -0.010989079  0.01701535\n",
       "10      chr1 100028533 100029101 SASS6      K562     -0.002251422  0.01832740\n",
       "11      chr1 100028533 100029101 SLC30A7    K562      0.003329924  0.01804794\n",
       "12      chr1 100028533 100029101 SLC35A3    K562      0.022953477  0.01917905\n",
       "13      chr1 100028533 100029101 TRMT13     K562      0.018339888  0.01635313\n",
       "14      chr1 100028533 100029101 VCAM1      K562      0.053572989  0.02846444\n",
       "15      chr1 100029136 100029435 AGL        K562     -0.004457632  0.01789879\n",
       "16      chr1 100029136 100029435 CDC14A     K562     -0.003466294  0.01900510\n",
       "17      chr1 100029136 100029435 DBT        K562      0.019892904  0.01776676\n",
       "18      chr1 100029136 100029435 DPH5       K562      0.013616953  0.01802728\n",
       "19      chr1 100029136 100029435 EXTL2      K562      0.009816065  0.02036463\n",
       "20      chr1 100029136 100029435 FRRS1      K562      0.008823450  0.01927072\n",
       "21      chr1 100029136 100029435 MFSD14A    K562      0.013446344  0.01865863\n",
       "22      chr1 100029136 100029435 RTCA       K562      0.036738457  0.01687865\n",
       "23      chr1 100029136 100029435 RTCA-AS1   K562     -0.021766504  0.01929994\n",
       "24      chr1 100029136 100029435 SASS6      K562      0.016385121  0.01840021\n",
       "25      chr1 100029136 100029435 SLC30A7    K562      0.009428935  0.01613040\n",
       "26      chr1 100029136 100029435 SLC35A3    K562      0.010946416  0.01700453\n",
       "27      chr1 100029136 100029435 TRMT13     K562      0.009483572  0.01758771\n",
       "28      chr1 100029136 100029435 VCAM1      K562      0.004404926  0.02497601\n",
       "29      chr1 100034578 100034758 AGL        K562     -0.015791236  0.01845864\n",
       "30      chr1 100034578 100034758 CDC14A     K562      0.032417317  0.02070616\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮         \n",
       "2533452 chrX  997541    998270   ASMT       K562      3.125955e-02 0.01931998\n",
       "2533453 chrX  997541    998270   ASMTL      K562      2.048386e-02 0.02209108\n",
       "2533454 chrX  997541    998270   CSF2RA     K562     -9.391297e-04 0.02021967\n",
       "2533455 chrX  997541    998270   GTPBP6     K562     -3.506227e-03 0.02064658\n",
       "2533456 chrX  997541    998270   IL3RA      K562      4.332036e-03 0.03381729\n",
       "2533457 chrX  997541    998270   LINC00106  K562     -1.813808e-02 0.02188647\n",
       "2533458 chrX  997541    998270   P2RY8      K562     -1.336805e-02 0.03367565\n",
       "2533459 chrX  997541    998270   PLCXD1     K562      1.536749e-02 0.02003704\n",
       "2533460 chrX  997541    998270   PPP2R3B    K562      1.767213e-02 0.01981576\n",
       "2533461 chrX  997541    998270   SLC25A6    K562      2.899253e-02 0.02336210\n",
       "2533462 chrX 9988858   9989079   CLCN4      K562     -7.586260e-03 0.01855314\n",
       "2533463 chrX 9988858   9989079   TBL1X      K562      8.118087e-05 0.01788273\n",
       "2533464 chrX 9988858   9989079   WWC3       K562      5.408002e-02 0.01979240\n",
       "2533465 chrX  999293    999875   AKAP17A    K562     -4.641255e-03 0.02126370\n",
       "2533466 chrX  999293    999875   ASMT       K562     -1.834867e-02 0.01718280\n",
       "2533467 chrX  999293    999875   ASMTL      K562     -1.917408e-02 0.02357437\n",
       "2533468 chrX  999293    999875   CSF2RA     K562     -2.865688e-03 0.01800071\n",
       "2533469 chrX  999293    999875   GTPBP6     K562      3.862025e-02 0.01755228\n",
       "2533470 chrX  999293    999875   IL3RA      K562      9.658144e-04 0.03359537\n",
       "2533471 chrX  999293    999875   LINC00106  K562      5.215708e-03 0.02330403\n",
       "2533472 chrX  999293    999875   P2RY8      K562      1.742263e-03 0.04022438\n",
       "2533473 chrX  999293    999875   PLCXD1     K562     -5.122071e-03 0.01983656\n",
       "2533474 chrX  999293    999875   PPP2R3B    K562      3.685852e-02 0.02172516\n",
       "2533475 chrX  999293    999875   SLC25A6    K562     -7.182472e-03 0.02481795\n",
       "2533476 chrX 9995654   9996249   CLCN4      K562      3.161529e-02 0.01976669\n",
       "2533477 chrX 9995654   9996249   TBL1X      K562     -6.495872e-03 0.02217910\n",
       "2533478 chrX 9995654   9996249   WWC3       K562     -1.024784e-02 0.02188004\n",
       "2533479 chrX 9997430   9997980   CLCN4      K562     -1.234859e-02 0.01868056\n",
       "2533480 chrX 9997430   9997980   TBL1X      K562      7.726147e-04 0.02275951\n",
       "2533481 chrX 9997430   9997980   WWC3       K562      4.332281e-02 0.01969990\n",
       "        rBgMean       pvalZ      \n",
       "1        0.0114035817 0.45603960 \n",
       "2        0.0193009567 0.66470909 \n",
       "3        0.0076737166 0.18990487 \n",
       "4        0.0148214401 0.04731448 \n",
       "5        0.0284021086 0.49986313 \n",
       "6        0.0163196487 0.48202829 \n",
       "7        0.0016496906 0.70761904 \n",
       "8        0.0050826383 0.48089716 \n",
       "9        0.0028612812 0.79217529 \n",
       "10       0.0002148152 0.55352232 \n",
       "11       0.0130543615 0.70499157 \n",
       "12       0.0015917885 0.13268180 \n",
       "13       0.0055654495 0.21735367 \n",
       "14       0.0224872506 0.13739673 \n",
       "15       0.0062081841 0.72437765 \n",
       "16       0.0044877707 0.66221745 \n",
       "17       0.0043469431 0.19078630 \n",
       "18       0.0029149053 0.27637029 \n",
       "19       0.0126175149 0.55470771 \n",
       "20       0.0073734844 0.47001111 \n",
       "21       0.0033302081 0.29385093 \n",
       "22       0.0031420616 0.02326974 \n",
       "23       0.0033811497 0.90371113 \n",
       "24      -0.0038076795 0.13622836 \n",
       "25       0.0053047373 0.39909951 \n",
       "26       0.0005722155 0.27090291 \n",
       "27       0.0040817037 0.37936885 \n",
       "28       0.0130506060 0.63538853 \n",
       "29       0.0172414562 0.96323697 \n",
       "30       0.0201888888 0.27740438 \n",
       "⋮       ⋮             ⋮          \n",
       "2533452  0.0040569059 0.079564733\n",
       "2533453  0.0257922007 0.594948530\n",
       "2533454  0.0171326552 0.814278180\n",
       "2533455  0.0199335677 0.871872489\n",
       "2533456  0.0324357718 0.797026302\n",
       "2533457  0.0293285566 0.984949788\n",
       "2533458 -0.0220174847 0.398649083\n",
       "2533459  0.0159153994 0.510907717\n",
       "2533460  0.0254860972 0.653331660\n",
       "2533461 -0.0125439343 0.037706664\n",
       "2533462  0.0005051081 0.668624599\n",
       "2533463 -0.0024474925 0.443775769\n",
       "2533464  0.0060411130 0.007609207\n",
       "2533465 -0.0001636529 0.583390526\n",
       "2533466  0.0015935674 0.877096301\n",
       "2533467  0.0179116579 0.942156539\n",
       "2533468  0.0164488106 0.858361157\n",
       "2533469  0.0165311281 0.104109794\n",
       "2533470  0.0279406956 0.788993671\n",
       "2533471  0.0274524535 0.830009209\n",
       "2533472 -0.0168141337 0.322283737\n",
       "2533473  0.0198253972 0.895741020\n",
       "2533474  0.0283532622 0.347716329\n",
       "2533475 -0.0182731899 0.327479797\n",
       "2533476  0.0053854685 0.092259140\n",
       "2533477 -0.0069319861 0.492155991\n",
       "2533478  0.0101763166 0.824708976\n",
       "2533479  0.0025196485 0.786960837\n",
       "2533480 -0.0052965816 0.394862997\n",
       "2533481  0.0047847627 0.025217504"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   data.table_1.15.2                \n",
       " [3] foreach_1.5.2                     BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] genomation_1.34.0                 Seurat_5.0.3                     \n",
       "[11] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[13] FigR_0.1.0                        motifmatchr_1.24.0               \n",
       "[15] chromVAR_1.24.0                   cowplot_1.1.3                    \n",
       "[17] ggplot2_3.5.1                     dplyr_1.1.4                      \n",
       "[19] SummarizedExperiment_1.30.2       Biobase_2.60.0                   \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               MatrixGenerics_1.12.3            \n",
       "[27] matrixStats_1.3.0                 Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             vroom_1.6.5                \n",
       " [15] globals_0.16.3              lattice_0.22-6             \n",
       " [17] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [19] plotly_4.10.4               plotrix_3.8-4              \n",
       " [21] yaml_2.3.8                  httpuv_1.6.15              \n",
       " [23] sctransform_0.4.1           spam_2.10-0                \n",
       " [25] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [27] pbapply_1.7-2               DBI_1.2.2                  \n",
       " [29] CNEr_1.38.0                 RColorBrewer_1.1-3         \n",
       " [31] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [33] Rtsne_0.17                  purrr_1.0.2                \n",
       " [35] R.utils_2.12.3              RCurl_1.98-1.14            \n",
       " [37] pracma_2.4.4                GenomeInfoDbData_1.2.11    \n",
       " [39] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [41] spatstat.utils_3.0-4        listenv_0.9.1              \n",
       " [43] seqLogo_1.68.0              goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             spatstat.random_3.2-3      \n",
       " [47] annotate_1.80.0             fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           leiden_0.4.3.1             \n",
       " [51] codetools_0.2-20            DelayedArray_0.26.7        \n",
       " [53] DT_0.33                     tidyselect_1.2.1           \n",
       " [55] spatstat.explore_3.2-6      base64enc_0.1-3            \n",
       " [57] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [59] progressr_0.14.0            iterators_1.0.14           \n",
       " [61] ggridges_0.5.6              survival_3.6-4             \n",
       " [63] tools_4.3.3                 TFMPvalue_0.0.9            \n",
       " [65] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [67] glue_1.7.0                  gridExtra_2.3              \n",
       " [69] IRdisplay_1.1               withr_3.0.0                \n",
       " [71] fastmap_1.1.1               fansi_1.0.6                \n",
       " [73] caTools_1.18.2              digest_0.6.35              \n",
       " [75] R6_2.5.1                    mime_0.12                  \n",
       " [77] nabor_0.5.0                 seqPattern_1.34.0          \n",
       " [79] colorspace_2.1-0            scattermore_1.2            \n",
       " [81] GO.db_3.18.0                tensor_1.5                 \n",
       " [83] gtools_3.9.5                poweRlaw_0.80.0            \n",
       " [85] spatstat.data_3.0-4         RSQLite_2.3.6              \n",
       " [87] R.methodsS3_1.8.2           utf8_1.2.4                 \n",
       " [89] tidyr_1.3.1                 generics_0.1.3             \n",
       " [91] httr_1.4.7                  htmlwidgets_1.6.4          \n",
       " [93] S4Arrays_1.2.1              TFBSTools_1.40.0           \n",
       " [95] uwot_0.1.16                 pkgconfig_2.0.3            \n",
       " [97] gtable_0.3.5                blob_1.2.4                 \n",
       " [99] impute_1.76.0               lmtest_0.9-40              \n",
       "[101] htmltools_0.5.8.1           dotCall64_1.1-1            \n",
       "[103] scales_1.3.0                png_0.1-8                  \n",
       "[105] tzdb_0.4.0                  reshape2_1.4.4             \n",
       "[107] rjson_0.2.21                uuid_1.2-0                 \n",
       "[109] nlme_3.1-164                repr_1.1.7                 \n",
       "[111] cachem_1.0.8                zoo_1.8-12                 \n",
       "[113] stringr_1.5.1               KernSmooth_2.23-22         \n",
       "[115] miniUI_0.1.1.1              AnnotationDbi_1.64.1       \n",
       "[117] restfulr_0.0.15             pillar_1.9.0               \n",
       "[119] vctrs_0.6.5                 RANN_2.6.1                 \n",
       "[121] promises_1.3.0              xtable_1.8-4               \n",
       "[123] cluster_2.1.6               evaluate_0.23              \n",
       "[125] readr_2.1.5                 cli_3.6.2                  \n",
       "[127] compiler_4.3.3              Rsamtools_2.18.0           \n",
       "[129] rlang_1.1.3                 crayon_1.5.2               \n",
       "[131] future.apply_1.11.2         plyr_1.8.9                 \n",
       "[133] stringi_1.8.3               gridBase_0.4-7             \n",
       "[135] deldir_2.0-4                viridisLite_0.4.2          \n",
       "[137] BiocParallel_1.36.0         munsell_0.5.1              \n",
       "[139] lazyeval_0.2.2              spatstat.geom_3.2-9        \n",
       "[141] IRkernel_1.3.2              RcppHNSW_0.6.0             \n",
       "[143] hms_1.1.3                   patchwork_1.2.0            \n",
       "[145] bit64_4.0.5                 future_1.33.2              \n",
       "[147] KEGGREST_1.40.1             shiny_1.8.1.1              \n",
       "[149] ROCR_1.0-11                 igraph_2.0.3               \n",
       "[151] memoise_2.0.1               bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
