{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240617/sc-E2G/test/results/GM12878/GM12878.multiome_7features/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240617/sc-E2G/test/results/GM12878/GM12878.multiome_7features/Predictions/EnhancerPredictionsAllPutative.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240617/sc-E2G/test/results/GM12878/GM12878.multiome_7features/Kendall/atac_matrix.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/GM12878/1.prepare_data/1.seurat_pipeline.240615/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/GM12878/2.Genome_wide_prediction/FigR/FigR.240625/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2637980 ranges and 3 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100028533-100029101      * |         AGL\n",
       "        [2]     chr1 100028533-100029101      * |      CDC14A\n",
       "        [3]     chr1 100028533-100029101      * |         DBT\n",
       "        [4]     chr1 100028533-100029101      * |        DPH5\n",
       "        [5]     chr1 100028533-100029101      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2637976]     chrX     9995654-9996249      * |       TBL1X\n",
       "  [2637977]     chrX     9995654-9996249      * |        WWC3\n",
       "  [2637978]     chrX     9997430-9997980      * |       CLCN4\n",
       "  [2637979]     chrX     9997430-9997980      * |       TBL1X\n",
       "  [2637980]     chrX     9997430-9997980      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-100028533-10002.. chr1-100028533-10002..\n",
       "        [2] chr1-100028533-10002.. chr1-100028533-10002..\n",
       "        [3] chr1-100028533-10002.. chr1-100028533-10002..\n",
       "        [4] chr1-100028533-10002.. chr1-100028533-10002..\n",
       "        [5] chr1-100028533-10002.. chr1-100028533-10002..\n",
       "        ...                    ...                    ...\n",
       "  [2637976]   chrX-9995654-9996249 chrX-9995654-9996249..\n",
       "  [2637977]   chrX-9995654-9996249 chrX-9995654-9996249..\n",
       "  [2637978]   chrX-9997430-9997980 chrX-9997430-9997980..\n",
       "  [2637979]   chrX-9997430-9997980 chrX-9997430-9997980..\n",
       "  [2637980]   chrX-9997430-9997980 chrX-9997430-9997980..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  3050 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 155270 3050 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(155270): chr1-100028533-100029101 chr1-100029136-100029435 ...\n",
       "  chrX-9995654-9996249 chrX-9997430-9997980\n",
       "rowData names(2): PeakName bias\n",
       "colnames(3050): GM12878_AAACAGCCACTAAGAA-1 GM12878_AAACAGCCATAATCAC-1\n",
       "  ... GM12878_TTTGTTGGTCACAAAT-1 GM12878_TTTGTTGGTTGGTGAC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna),\n",
    "                       index = 1:nrow(matrix.rna))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  17.5536782741547 secs \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Time Elapsed:  27.7377895712852 mins \n",
      "\n",
      "[1] \"chr10\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.88808465003967 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.2938919425011 mins \n",
      "\n",
      "[1] \"chr11\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  12.4728405475616 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  18.3455096006393 mins \n",
      "\n",
      "[1] \"chr12\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.80539536476135 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  15.2269680738449 mins \n",
      "\n",
      "[1] \"chr13\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.09579825401306 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.42576632102331 mins \n",
      "\n",
      "[1] \"chr14\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.3325788974762 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.8491372704506 mins \n",
      "\n",
      "[1] \"chr15\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.02812743186951 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  13.2402209560076 mins \n",
      "\n",
      "[1] \"chr16\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.1066808700562 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  17.7139662583669 mins \n",
      "\n",
      "[1] \"chr17\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  13.4322621822357 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  24.4517054518064 mins \n",
      "\n",
      "[1] \"chr18\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.71622252464294 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.10625516176224 mins \n",
      "\n",
      "[1] \"chr19\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  23.7513115406036 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  32.4285821318626 mins \n",
      "\n",
      "[1] \"chr2\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  9.00406455993652 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  15.9869062781334 mins \n",
      "\n",
      "[1] \"chr20\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.46392917633057 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.6163697520892 mins \n",
      "\n",
      "[1] \"chr21\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.49650168418884 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  7.56126488844554 mins \n",
      "\n",
      "[1] \"chr22\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.55377769470215 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.7717722137769 mins \n",
      "\n",
      "[1] \"chr3\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.1199960708618 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  16.7457268675168 mins \n",
      "\n",
      "[1] \"chr4\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.75624990463257 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.7422688126564 mins \n",
      "\n",
      "[1] \"chr5\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.78698515892029 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  13.5860894123713 mins \n",
      "\n",
      "[1] \"chr6\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  11.9202716350555 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  17.5389284412066 mins \n",
      "\n",
      "[1] \"chr7\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.29364323616028 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  14.2834753195445 mins \n",
      "\n",
      "[1] \"chr8\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.80882143974304 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.9941868106524 mins \n",
      "\n",
      "[1] \"chr9\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.15836596488953 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  15.2544211824735 mins \n",
      "\n",
      "[1] \"chrX\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.55316185951233 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.8135809381803 mins \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5a93ede-8b2c-4646-9a36-b27ff71e2958",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2423604 ranges and 7 metadata columns:\n",
       "       seqnames              ranges strand |  TargetGene               PeakName\n",
       "          <Rle>           <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 100028533-100029101      * |         AGL chr1-100028533-10002..\n",
       "  chr1     chr1 100028533-100029101      * |      CDC14A chr1-100028533-10002..\n",
       "  chr1     chr1 100028533-100029101      * |         DBT chr1-100028533-10002..\n",
       "  chr1     chr1 100028533-100029101      * |        DPH5 chr1-100028533-10002..\n",
       "  chr1     chr1 100028533-100029101      * |       EXTL2 chr1-100028533-10002..\n",
       "   ...      ...                 ...    ... .         ...                    ...\n",
       "  chrX     chrX     9995654-9996249      * |       TBL1X   chrX-9995654-9996249\n",
       "  chrX     chrX     9995654-9996249      * |        WWC3   chrX-9995654-9996249\n",
       "  chrX     chrX     9997430-9997980      * |       CLCN4   chrX-9997430-9997980\n",
       "  chrX     chrX     9997430-9997980      * |       TBL1X   chrX-9997430-9997980\n",
       "  chrX     chrX     9997430-9997980      * |        WWC3   chrX-9997430-9997980\n",
       "                     PairName         rObs     rBgSD     rBgMean     pvalZ\n",
       "                  <character>    <numeric> <numeric>   <numeric> <numeric>\n",
       "  chr1 chr1-100028533-10002..    0.0133078 0.0200686   0.0135133 0.5040857\n",
       "  chr1 chr1-100028533-10002..    0.0104488 0.0203293   0.0177005 0.6393445\n",
       "  chr1 chr1-100028533-10002..    0.0224650 0.0175763   0.0071305 0.1914800\n",
       "  chr1 chr1-100028533-10002..    0.0450955 0.0197678   0.0116089 0.0451328\n",
       "  chr1 chr1-100028533-10002..    0.0284114 0.0272882   0.0237448 0.4321070\n",
       "   ...                    ...          ...       ...         ...       ...\n",
       "  chrX chrX-9995654-9996249.. -0.006495872 0.0228668 -0.00764159 0.4800198\n",
       "  chrX chrX-9995654-9996249.. -0.010247842 0.0204972  0.01123688 0.8527216\n",
       "  chrX chrX-9997430-9997980.. -0.012348587 0.0204654  0.00407203 0.7888272\n",
       "  chrX chrX-9997430-9997980..  0.000772615 0.0229810 -0.00760477 0.3577294\n",
       "  chrX chrX-9997430-9997980..  0.043322815 0.0198778  0.00469899 0.0260044\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2423604 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>AGL     </td><td>K562</td><td> 0.013307751</td><td>0.02006862</td><td> 0.0135132819</td><td>0.50408565</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>CDC14A  </td><td>K562</td><td> 0.010448850</td><td>0.02032930</td><td> 0.0177004597</td><td>0.63934453</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>DBT     </td><td>K562</td><td> 0.022465002</td><td>0.01757626</td><td> 0.0071304952</td><td>0.19147998</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>DPH5    </td><td>K562</td><td> 0.045095506</td><td>0.01976778</td><td> 0.0116089305</td><td>0.04513284</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>EXTL2   </td><td>K562</td><td> 0.028411431</td><td>0.02728823</td><td> 0.0237448037</td><td>0.43210699</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>FRRS1   </td><td>K562</td><td> 0.017150919</td><td>0.02108533</td><td> 0.0121480184</td><td>0.40622398</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>MFSD14A </td><td>K562</td><td>-0.008365932</td><td>0.01839028</td><td> 0.0047155313</td><td>0.76155850</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>RTCA    </td><td>K562</td><td> 0.006028360</td><td>0.01520566</td><td> 0.0012128211</td><td>0.37573798</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>RTCA-AS1</td><td>K562</td><td>-0.010989079</td><td>0.01950211</td><td> 0.0059057048</td><td>0.80683860</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>SASS6   </td><td>K562</td><td>-0.002251422</td><td>0.01722381</td><td> 0.0001442482</td><td>0.55531073</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>SLC30A7 </td><td>K562</td><td> 0.003329924</td><td>0.01588208</td><td> 0.0115996754</td><td>0.69871106</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>SLC35A3 </td><td>K562</td><td> 0.022953477</td><td>0.01750425</td><td> 0.0041166975</td><td>0.14093538</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>TRMT13  </td><td>K562</td><td> 0.018339888</td><td>0.01571902</td><td> 0.0056048369</td><td>0.20892180</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>VCAM1   </td><td>K562</td><td> 0.053572989</td><td>0.02490281</td><td> 0.0182248750</td><td>0.07788503</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>AGL     </td><td>K562</td><td>-0.004457632</td><td>0.01661980</td><td> 0.0081718465</td><td>0.77634455</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>CDC14A  </td><td>K562</td><td>-0.003466294</td><td>0.01836982</td><td> 0.0111601011</td><td>0.78704757</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>DBT     </td><td>K562</td><td> 0.019892904</td><td>0.01759068</td><td> 0.0062062424</td><td>0.21826581</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>DPH5    </td><td>K562</td><td> 0.013616953</td><td>0.02146011</td><td> 0.0045038311</td><td>0.33554447</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>EXTL2   </td><td>K562</td><td> 0.009816065</td><td>0.01887563</td><td> 0.0110784261</td><td>0.52666052</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>FRRS1   </td><td>K562</td><td> 0.008823450</td><td>0.01689535</td><td> 0.0070475471</td><td>0.45814349</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>MFSD14A </td><td>K562</td><td> 0.013446344</td><td>0.01664453</td><td>-0.0005385988</td><td>0.20039465</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>RTCA    </td><td>K562</td><td> 0.036738457</td><td>0.01777343</td><td> 0.0040990462</td><td>0.03314810</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>RTCA-AS1</td><td>K562</td><td>-0.021766504</td><td>0.01689104</td><td>-0.0000324883</td><td>0.90090376</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>SASS6   </td><td>K562</td><td> 0.016385121</td><td>0.01543190</td><td> 0.0024023919</td><td>0.18244342</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>SLC30A7 </td><td>K562</td><td> 0.009428935</td><td>0.01856813</td><td> 0.0051188581</td><td>0.40822151</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>SLC35A3 </td><td>K562</td><td> 0.010946416</td><td>0.01679384</td><td> 0.0011642227</td><td>0.28011930</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>TRMT13  </td><td>K562</td><td> 0.009483572</td><td>0.01748185</td><td> 0.0044174089</td><td>0.38598629</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>VCAM1   </td><td>K562</td><td> 0.004404926</td><td>0.02085092</td><td> 0.0108685409</td><td>0.62171640</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034578</td><td>100034758</td><td>AGL     </td><td>K562</td><td>-0.015791236</td><td>0.01950572</td><td> 0.0146974280</td><td>0.94098109</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034578</td><td>100034758</td><td>CDC14A  </td><td>K562</td><td> 0.032417317</td><td>0.02385925</td><td> 0.0182303749</td><td>0.27605220</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>ASMT     </td><td>K562</td><td> 3.125955e-02</td><td>0.01854107</td><td> 0.003839971</td><td>0.069589355</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>ASMTL    </td><td>K562</td><td> 2.048386e-02</td><td>0.01882048</td><td> 0.019852217</td><td>0.486613332</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>CSF2RA   </td><td>K562</td><td>-9.391297e-04</td><td>0.02137257</td><td> 0.019743296</td><td>0.833405120</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>GTPBP6   </td><td>K562</td><td>-3.506227e-03</td><td>0.01854163</td><td> 0.018032664</td><td>0.877311277</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>IL3RA    </td><td>K562</td><td> 4.332036e-03</td><td>0.02928276</td><td> 0.024998192</td><td>0.759826577</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>LINC00106</td><td>K562</td><td>-1.813808e-02</td><td>0.02167384</td><td> 0.030851479</td><td>0.988098924</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>P2RY8    </td><td>K562</td><td>-1.336805e-02</td><td>0.03465214</td><td>-0.019192585</td><td>0.433257887</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>PLCXD1   </td><td>K562</td><td> 1.536749e-02</td><td>0.02088157</td><td> 0.018723511</td><td>0.563841877</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>PPP2R3B  </td><td>K562</td><td> 1.767213e-02</td><td>0.02006108</td><td> 0.027820936</td><td>0.693534960</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>SLC25A6  </td><td>K562</td><td> 2.899253e-02</td><td>0.02188817</td><td>-0.015786621</td><td>0.020387269</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988858</td><td>9989079</td><td>CLCN4    </td><td>K562</td><td>-7.586260e-03</td><td>0.02076492</td><td> 0.001176876</td><td>0.663493474</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988858</td><td>9989079</td><td>TBL1X    </td><td>K562</td><td> 8.118087e-05</td><td>0.01524722</td><td>-0.004853582</td><td>0.373101503</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988858</td><td>9989079</td><td>WWC3     </td><td>K562</td><td> 5.408002e-02</td><td>0.01930517</td><td> 0.005931187</td><td>0.006314035</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>AKAP17A  </td><td>K562</td><td>-4.641255e-03</td><td>0.01869073</td><td>-0.001703902</td><td>0.562438880</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>ASMT     </td><td>K562</td><td>-1.834867e-02</td><td>0.01824577</td><td> 0.004600310</td><td>0.895762501</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>ASMTL    </td><td>K562</td><td>-1.917408e-02</td><td>0.02075847</td><td> 0.018832733</td><td>0.966442719</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>CSF2RA   </td><td>K562</td><td>-2.865688e-03</td><td>0.02030021</td><td> 0.017138500</td><td>0.837790573</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>GTPBP6   </td><td>K562</td><td> 3.862025e-02</td><td>0.01965268</td><td> 0.021200369</td><td>0.187704451</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>IL3RA    </td><td>K562</td><td> 9.658144e-04</td><td>0.03377033</td><td> 0.030287787</td><td>0.807378387</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>LINC00106</td><td>K562</td><td> 5.215708e-03</td><td>0.02415971</td><td> 0.029132486</td><td>0.838899470</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>P2RY8    </td><td>K562</td><td> 1.742263e-03</td><td>0.03412070</td><td>-0.015944023</td><td>0.302108881</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>PLCXD1   </td><td>K562</td><td>-5.122071e-03</td><td>0.01845258</td><td> 0.018045145</td><td>0.895351323</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>PPP2R3B  </td><td>K562</td><td> 3.685852e-02</td><td>0.02134755</td><td> 0.028050325</td><td>0.339946537</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>SLC25A6  </td><td>K562</td><td>-7.182472e-03</td><td>0.02459538</td><td>-0.013688983</td><td>0.395681174</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995654</td><td>9996249</td><td>CLCN4    </td><td>K562</td><td> 3.161529e-02</td><td>0.02004924</td><td> 0.003544645</td><td>0.080743972</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995654</td><td>9996249</td><td>TBL1X    </td><td>K562</td><td>-6.495872e-03</td><td>0.02286679</td><td>-0.007641585</td><td>0.480019839</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995654</td><td>9996249</td><td>WWC3     </td><td>K562</td><td>-1.024784e-02</td><td>0.02049721</td><td> 0.011236877</td><td>0.852721598</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997430</td><td>9997980</td><td>CLCN4    </td><td>K562</td><td>-1.234859e-02</td><td>0.02046544</td><td> 0.004072029</td><td>0.788827155</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997430</td><td>9997980</td><td>TBL1X    </td><td>K562</td><td> 7.726147e-04</td><td>0.02298103</td><td>-0.007604769</td><td>0.357729383</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997430</td><td>9997980</td><td>WWC3     </td><td>K562</td><td> 4.332281e-02</td><td>0.01987783</td><td> 0.004698987</td><td>0.026004413</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2423604 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100028533 & 100029101 & AGL      & K562 &  0.013307751 & 0.02006862 &  0.0135132819 & 0.50408565\\\\\n",
       "\t chr1 & 100028533 & 100029101 & CDC14A   & K562 &  0.010448850 & 0.02032930 &  0.0177004597 & 0.63934453\\\\\n",
       "\t chr1 & 100028533 & 100029101 & DBT      & K562 &  0.022465002 & 0.01757626 &  0.0071304952 & 0.19147998\\\\\n",
       "\t chr1 & 100028533 & 100029101 & DPH5     & K562 &  0.045095506 & 0.01976778 &  0.0116089305 & 0.04513284\\\\\n",
       "\t chr1 & 100028533 & 100029101 & EXTL2    & K562 &  0.028411431 & 0.02728823 &  0.0237448037 & 0.43210699\\\\\n",
       "\t chr1 & 100028533 & 100029101 & FRRS1    & K562 &  0.017150919 & 0.02108533 &  0.0121480184 & 0.40622398\\\\\n",
       "\t chr1 & 100028533 & 100029101 & MFSD14A  & K562 & -0.008365932 & 0.01839028 &  0.0047155313 & 0.76155850\\\\\n",
       "\t chr1 & 100028533 & 100029101 & RTCA     & K562 &  0.006028360 & 0.01520566 &  0.0012128211 & 0.37573798\\\\\n",
       "\t chr1 & 100028533 & 100029101 & RTCA-AS1 & K562 & -0.010989079 & 0.01950211 &  0.0059057048 & 0.80683860\\\\\n",
       "\t chr1 & 100028533 & 100029101 & SASS6    & K562 & -0.002251422 & 0.01722381 &  0.0001442482 & 0.55531073\\\\\n",
       "\t chr1 & 100028533 & 100029101 & SLC30A7  & K562 &  0.003329924 & 0.01588208 &  0.0115996754 & 0.69871106\\\\\n",
       "\t chr1 & 100028533 & 100029101 & SLC35A3  & K562 &  0.022953477 & 0.01750425 &  0.0041166975 & 0.14093538\\\\\n",
       "\t chr1 & 100028533 & 100029101 & TRMT13   & K562 &  0.018339888 & 0.01571902 &  0.0056048369 & 0.20892180\\\\\n",
       "\t chr1 & 100028533 & 100029101 & VCAM1    & K562 &  0.053572989 & 0.02490281 &  0.0182248750 & 0.07788503\\\\\n",
       "\t chr1 & 100029136 & 100029435 & AGL      & K562 & -0.004457632 & 0.01661980 &  0.0081718465 & 0.77634455\\\\\n",
       "\t chr1 & 100029136 & 100029435 & CDC14A   & K562 & -0.003466294 & 0.01836982 &  0.0111601011 & 0.78704757\\\\\n",
       "\t chr1 & 100029136 & 100029435 & DBT      & K562 &  0.019892904 & 0.01759068 &  0.0062062424 & 0.21826581\\\\\n",
       "\t chr1 & 100029136 & 100029435 & DPH5     & K562 &  0.013616953 & 0.02146011 &  0.0045038311 & 0.33554447\\\\\n",
       "\t chr1 & 100029136 & 100029435 & EXTL2    & K562 &  0.009816065 & 0.01887563 &  0.0110784261 & 0.52666052\\\\\n",
       "\t chr1 & 100029136 & 100029435 & FRRS1    & K562 &  0.008823450 & 0.01689535 &  0.0070475471 & 0.45814349\\\\\n",
       "\t chr1 & 100029136 & 100029435 & MFSD14A  & K562 &  0.013446344 & 0.01664453 & -0.0005385988 & 0.20039465\\\\\n",
       "\t chr1 & 100029136 & 100029435 & RTCA     & K562 &  0.036738457 & 0.01777343 &  0.0040990462 & 0.03314810\\\\\n",
       "\t chr1 & 100029136 & 100029435 & RTCA-AS1 & K562 & -0.021766504 & 0.01689104 & -0.0000324883 & 0.90090376\\\\\n",
       "\t chr1 & 100029136 & 100029435 & SASS6    & K562 &  0.016385121 & 0.01543190 &  0.0024023919 & 0.18244342\\\\\n",
       "\t chr1 & 100029136 & 100029435 & SLC30A7  & K562 &  0.009428935 & 0.01856813 &  0.0051188581 & 0.40822151\\\\\n",
       "\t chr1 & 100029136 & 100029435 & SLC35A3  & K562 &  0.010946416 & 0.01679384 &  0.0011642227 & 0.28011930\\\\\n",
       "\t chr1 & 100029136 & 100029435 & TRMT13   & K562 &  0.009483572 & 0.01748185 &  0.0044174089 & 0.38598629\\\\\n",
       "\t chr1 & 100029136 & 100029435 & VCAM1    & K562 &  0.004404926 & 0.02085092 &  0.0108685409 & 0.62171640\\\\\n",
       "\t chr1 & 100034578 & 100034758 & AGL      & K562 & -0.015791236 & 0.01950572 &  0.0146974280 & 0.94098109\\\\\n",
       "\t chr1 & 100034578 & 100034758 & CDC14A   & K562 &  0.032417317 & 0.02385925 &  0.0182303749 & 0.27605220\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  997541 &  998270 & ASMT      & K562 &  3.125955e-02 & 0.01854107 &  0.003839971 & 0.069589355\\\\\n",
       "\t chrX &  997541 &  998270 & ASMTL     & K562 &  2.048386e-02 & 0.01882048 &  0.019852217 & 0.486613332\\\\\n",
       "\t chrX &  997541 &  998270 & CSF2RA    & K562 & -9.391297e-04 & 0.02137257 &  0.019743296 & 0.833405120\\\\\n",
       "\t chrX &  997541 &  998270 & GTPBP6    & K562 & -3.506227e-03 & 0.01854163 &  0.018032664 & 0.877311277\\\\\n",
       "\t chrX &  997541 &  998270 & IL3RA     & K562 &  4.332036e-03 & 0.02928276 &  0.024998192 & 0.759826577\\\\\n",
       "\t chrX &  997541 &  998270 & LINC00106 & K562 & -1.813808e-02 & 0.02167384 &  0.030851479 & 0.988098924\\\\\n",
       "\t chrX &  997541 &  998270 & P2RY8     & K562 & -1.336805e-02 & 0.03465214 & -0.019192585 & 0.433257887\\\\\n",
       "\t chrX &  997541 &  998270 & PLCXD1    & K562 &  1.536749e-02 & 0.02088157 &  0.018723511 & 0.563841877\\\\\n",
       "\t chrX &  997541 &  998270 & PPP2R3B   & K562 &  1.767213e-02 & 0.02006108 &  0.027820936 & 0.693534960\\\\\n",
       "\t chrX &  997541 &  998270 & SLC25A6   & K562 &  2.899253e-02 & 0.02188817 & -0.015786621 & 0.020387269\\\\\n",
       "\t chrX & 9988858 & 9989079 & CLCN4     & K562 & -7.586260e-03 & 0.02076492 &  0.001176876 & 0.663493474\\\\\n",
       "\t chrX & 9988858 & 9989079 & TBL1X     & K562 &  8.118087e-05 & 0.01524722 & -0.004853582 & 0.373101503\\\\\n",
       "\t chrX & 9988858 & 9989079 & WWC3      & K562 &  5.408002e-02 & 0.01930517 &  0.005931187 & 0.006314035\\\\\n",
       "\t chrX &  999293 &  999875 & AKAP17A   & K562 & -4.641255e-03 & 0.01869073 & -0.001703902 & 0.562438880\\\\\n",
       "\t chrX &  999293 &  999875 & ASMT      & K562 & -1.834867e-02 & 0.01824577 &  0.004600310 & 0.895762501\\\\\n",
       "\t chrX &  999293 &  999875 & ASMTL     & K562 & -1.917408e-02 & 0.02075847 &  0.018832733 & 0.966442719\\\\\n",
       "\t chrX &  999293 &  999875 & CSF2RA    & K562 & -2.865688e-03 & 0.02030021 &  0.017138500 & 0.837790573\\\\\n",
       "\t chrX &  999293 &  999875 & GTPBP6    & K562 &  3.862025e-02 & 0.01965268 &  0.021200369 & 0.187704451\\\\\n",
       "\t chrX &  999293 &  999875 & IL3RA     & K562 &  9.658144e-04 & 0.03377033 &  0.030287787 & 0.807378387\\\\\n",
       "\t chrX &  999293 &  999875 & LINC00106 & K562 &  5.215708e-03 & 0.02415971 &  0.029132486 & 0.838899470\\\\\n",
       "\t chrX &  999293 &  999875 & P2RY8     & K562 &  1.742263e-03 & 0.03412070 & -0.015944023 & 0.302108881\\\\\n",
       "\t chrX &  999293 &  999875 & PLCXD1    & K562 & -5.122071e-03 & 0.01845258 &  0.018045145 & 0.895351323\\\\\n",
       "\t chrX &  999293 &  999875 & PPP2R3B   & K562 &  3.685852e-02 & 0.02134755 &  0.028050325 & 0.339946537\\\\\n",
       "\t chrX &  999293 &  999875 & SLC25A6   & K562 & -7.182472e-03 & 0.02459538 & -0.013688983 & 0.395681174\\\\\n",
       "\t chrX & 9995654 & 9996249 & CLCN4     & K562 &  3.161529e-02 & 0.02004924 &  0.003544645 & 0.080743972\\\\\n",
       "\t chrX & 9995654 & 9996249 & TBL1X     & K562 & -6.495872e-03 & 0.02286679 & -0.007641585 & 0.480019839\\\\\n",
       "\t chrX & 9995654 & 9996249 & WWC3      & K562 & -1.024784e-02 & 0.02049721 &  0.011236877 & 0.852721598\\\\\n",
       "\t chrX & 9997430 & 9997980 & CLCN4     & K562 & -1.234859e-02 & 0.02046544 &  0.004072029 & 0.788827155\\\\\n",
       "\t chrX & 9997430 & 9997980 & TBL1X     & K562 &  7.726147e-04 & 0.02298103 & -0.007604769 & 0.357729383\\\\\n",
       "\t chrX & 9997430 & 9997980 & WWC3      & K562 &  4.332281e-02 & 0.01987783 &  0.004698987 & 0.026004413\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2423604 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100028533 | 100029101 | AGL      | K562 |  0.013307751 | 0.02006862 |  0.0135132819 | 0.50408565 |\n",
       "| chr1 | 100028533 | 100029101 | CDC14A   | K562 |  0.010448850 | 0.02032930 |  0.0177004597 | 0.63934453 |\n",
       "| chr1 | 100028533 | 100029101 | DBT      | K562 |  0.022465002 | 0.01757626 |  0.0071304952 | 0.19147998 |\n",
       "| chr1 | 100028533 | 100029101 | DPH5     | K562 |  0.045095506 | 0.01976778 |  0.0116089305 | 0.04513284 |\n",
       "| chr1 | 100028533 | 100029101 | EXTL2    | K562 |  0.028411431 | 0.02728823 |  0.0237448037 | 0.43210699 |\n",
       "| chr1 | 100028533 | 100029101 | FRRS1    | K562 |  0.017150919 | 0.02108533 |  0.0121480184 | 0.40622398 |\n",
       "| chr1 | 100028533 | 100029101 | MFSD14A  | K562 | -0.008365932 | 0.01839028 |  0.0047155313 | 0.76155850 |\n",
       "| chr1 | 100028533 | 100029101 | RTCA     | K562 |  0.006028360 | 0.01520566 |  0.0012128211 | 0.37573798 |\n",
       "| chr1 | 100028533 | 100029101 | RTCA-AS1 | K562 | -0.010989079 | 0.01950211 |  0.0059057048 | 0.80683860 |\n",
       "| chr1 | 100028533 | 100029101 | SASS6    | K562 | -0.002251422 | 0.01722381 |  0.0001442482 | 0.55531073 |\n",
       "| chr1 | 100028533 | 100029101 | SLC30A7  | K562 |  0.003329924 | 0.01588208 |  0.0115996754 | 0.69871106 |\n",
       "| chr1 | 100028533 | 100029101 | SLC35A3  | K562 |  0.022953477 | 0.01750425 |  0.0041166975 | 0.14093538 |\n",
       "| chr1 | 100028533 | 100029101 | TRMT13   | K562 |  0.018339888 | 0.01571902 |  0.0056048369 | 0.20892180 |\n",
       "| chr1 | 100028533 | 100029101 | VCAM1    | K562 |  0.053572989 | 0.02490281 |  0.0182248750 | 0.07788503 |\n",
       "| chr1 | 100029136 | 100029435 | AGL      | K562 | -0.004457632 | 0.01661980 |  0.0081718465 | 0.77634455 |\n",
       "| chr1 | 100029136 | 100029435 | CDC14A   | K562 | -0.003466294 | 0.01836982 |  0.0111601011 | 0.78704757 |\n",
       "| chr1 | 100029136 | 100029435 | DBT      | K562 |  0.019892904 | 0.01759068 |  0.0062062424 | 0.21826581 |\n",
       "| chr1 | 100029136 | 100029435 | DPH5     | K562 |  0.013616953 | 0.02146011 |  0.0045038311 | 0.33554447 |\n",
       "| chr1 | 100029136 | 100029435 | EXTL2    | K562 |  0.009816065 | 0.01887563 |  0.0110784261 | 0.52666052 |\n",
       "| chr1 | 100029136 | 100029435 | FRRS1    | K562 |  0.008823450 | 0.01689535 |  0.0070475471 | 0.45814349 |\n",
       "| chr1 | 100029136 | 100029435 | MFSD14A  | K562 |  0.013446344 | 0.01664453 | -0.0005385988 | 0.20039465 |\n",
       "| chr1 | 100029136 | 100029435 | RTCA     | K562 |  0.036738457 | 0.01777343 |  0.0040990462 | 0.03314810 |\n",
       "| chr1 | 100029136 | 100029435 | RTCA-AS1 | K562 | -0.021766504 | 0.01689104 | -0.0000324883 | 0.90090376 |\n",
       "| chr1 | 100029136 | 100029435 | SASS6    | K562 |  0.016385121 | 0.01543190 |  0.0024023919 | 0.18244342 |\n",
       "| chr1 | 100029136 | 100029435 | SLC30A7  | K562 |  0.009428935 | 0.01856813 |  0.0051188581 | 0.40822151 |\n",
       "| chr1 | 100029136 | 100029435 | SLC35A3  | K562 |  0.010946416 | 0.01679384 |  0.0011642227 | 0.28011930 |\n",
       "| chr1 | 100029136 | 100029435 | TRMT13   | K562 |  0.009483572 | 0.01748185 |  0.0044174089 | 0.38598629 |\n",
       "| chr1 | 100029136 | 100029435 | VCAM1    | K562 |  0.004404926 | 0.02085092 |  0.0108685409 | 0.62171640 |\n",
       "| chr1 | 100034578 | 100034758 | AGL      | K562 | -0.015791236 | 0.01950572 |  0.0146974280 | 0.94098109 |\n",
       "| chr1 | 100034578 | 100034758 | CDC14A   | K562 |  0.032417317 | 0.02385925 |  0.0182303749 | 0.27605220 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  997541 |  998270 | ASMT      | K562 |  3.125955e-02 | 0.01854107 |  0.003839971 | 0.069589355 |\n",
       "| chrX |  997541 |  998270 | ASMTL     | K562 |  2.048386e-02 | 0.01882048 |  0.019852217 | 0.486613332 |\n",
       "| chrX |  997541 |  998270 | CSF2RA    | K562 | -9.391297e-04 | 0.02137257 |  0.019743296 | 0.833405120 |\n",
       "| chrX |  997541 |  998270 | GTPBP6    | K562 | -3.506227e-03 | 0.01854163 |  0.018032664 | 0.877311277 |\n",
       "| chrX |  997541 |  998270 | IL3RA     | K562 |  4.332036e-03 | 0.02928276 |  0.024998192 | 0.759826577 |\n",
       "| chrX |  997541 |  998270 | LINC00106 | K562 | -1.813808e-02 | 0.02167384 |  0.030851479 | 0.988098924 |\n",
       "| chrX |  997541 |  998270 | P2RY8     | K562 | -1.336805e-02 | 0.03465214 | -0.019192585 | 0.433257887 |\n",
       "| chrX |  997541 |  998270 | PLCXD1    | K562 |  1.536749e-02 | 0.02088157 |  0.018723511 | 0.563841877 |\n",
       "| chrX |  997541 |  998270 | PPP2R3B   | K562 |  1.767213e-02 | 0.02006108 |  0.027820936 | 0.693534960 |\n",
       "| chrX |  997541 |  998270 | SLC25A6   | K562 |  2.899253e-02 | 0.02188817 | -0.015786621 | 0.020387269 |\n",
       "| chrX | 9988858 | 9989079 | CLCN4     | K562 | -7.586260e-03 | 0.02076492 |  0.001176876 | 0.663493474 |\n",
       "| chrX | 9988858 | 9989079 | TBL1X     | K562 |  8.118087e-05 | 0.01524722 | -0.004853582 | 0.373101503 |\n",
       "| chrX | 9988858 | 9989079 | WWC3      | K562 |  5.408002e-02 | 0.01930517 |  0.005931187 | 0.006314035 |\n",
       "| chrX |  999293 |  999875 | AKAP17A   | K562 | -4.641255e-03 | 0.01869073 | -0.001703902 | 0.562438880 |\n",
       "| chrX |  999293 |  999875 | ASMT      | K562 | -1.834867e-02 | 0.01824577 |  0.004600310 | 0.895762501 |\n",
       "| chrX |  999293 |  999875 | ASMTL     | K562 | -1.917408e-02 | 0.02075847 |  0.018832733 | 0.966442719 |\n",
       "| chrX |  999293 |  999875 | CSF2RA    | K562 | -2.865688e-03 | 0.02030021 |  0.017138500 | 0.837790573 |\n",
       "| chrX |  999293 |  999875 | GTPBP6    | K562 |  3.862025e-02 | 0.01965268 |  0.021200369 | 0.187704451 |\n",
       "| chrX |  999293 |  999875 | IL3RA     | K562 |  9.658144e-04 | 0.03377033 |  0.030287787 | 0.807378387 |\n",
       "| chrX |  999293 |  999875 | LINC00106 | K562 |  5.215708e-03 | 0.02415971 |  0.029132486 | 0.838899470 |\n",
       "| chrX |  999293 |  999875 | P2RY8     | K562 |  1.742263e-03 | 0.03412070 | -0.015944023 | 0.302108881 |\n",
       "| chrX |  999293 |  999875 | PLCXD1    | K562 | -5.122071e-03 | 0.01845258 |  0.018045145 | 0.895351323 |\n",
       "| chrX |  999293 |  999875 | PPP2R3B   | K562 |  3.685852e-02 | 0.02134755 |  0.028050325 | 0.339946537 |\n",
       "| chrX |  999293 |  999875 | SLC25A6   | K562 | -7.182472e-03 | 0.02459538 | -0.013688983 | 0.395681174 |\n",
       "| chrX | 9995654 | 9996249 | CLCN4     | K562 |  3.161529e-02 | 0.02004924 |  0.003544645 | 0.080743972 |\n",
       "| chrX | 9995654 | 9996249 | TBL1X     | K562 | -6.495872e-03 | 0.02286679 | -0.007641585 | 0.480019839 |\n",
       "| chrX | 9995654 | 9996249 | WWC3      | K562 | -1.024784e-02 | 0.02049721 |  0.011236877 | 0.852721598 |\n",
       "| chrX | 9997430 | 9997980 | CLCN4     | K562 | -1.234859e-02 | 0.02046544 |  0.004072029 | 0.788827155 |\n",
       "| chrX | 9997430 | 9997980 | TBL1X     | K562 |  7.726147e-04 | 0.02298103 | -0.007604769 | 0.357729383 |\n",
       "| chrX | 9997430 | 9997980 | WWC3      | K562 |  4.332281e-02 | 0.01987783 |  0.004698987 | 0.026004413 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType rObs          rBgSD     \n",
       "1       chr1 100028533 100029101 AGL        K562      0.013307751  0.02006862\n",
       "2       chr1 100028533 100029101 CDC14A     K562      0.010448850  0.02032930\n",
       "3       chr1 100028533 100029101 DBT        K562      0.022465002  0.01757626\n",
       "4       chr1 100028533 100029101 DPH5       K562      0.045095506  0.01976778\n",
       "5       chr1 100028533 100029101 EXTL2      K562      0.028411431  0.02728823\n",
       "6       chr1 100028533 100029101 FRRS1      K562      0.017150919  0.02108533\n",
       "7       chr1 100028533 100029101 MFSD14A    K562     -0.008365932  0.01839028\n",
       "8       chr1 100028533 100029101 RTCA       K562      0.006028360  0.01520566\n",
       "9       chr1 100028533 100029101 RTCA-AS1   K562     -0.010989079  0.01950211\n",
       "10      chr1 100028533 100029101 SASS6      K562     -0.002251422  0.01722381\n",
       "11      chr1 100028533 100029101 SLC30A7    K562      0.003329924  0.01588208\n",
       "12      chr1 100028533 100029101 SLC35A3    K562      0.022953477  0.01750425\n",
       "13      chr1 100028533 100029101 TRMT13     K562      0.018339888  0.01571902\n",
       "14      chr1 100028533 100029101 VCAM1      K562      0.053572989  0.02490281\n",
       "15      chr1 100029136 100029435 AGL        K562     -0.004457632  0.01661980\n",
       "16      chr1 100029136 100029435 CDC14A     K562     -0.003466294  0.01836982\n",
       "17      chr1 100029136 100029435 DBT        K562      0.019892904  0.01759068\n",
       "18      chr1 100029136 100029435 DPH5       K562      0.013616953  0.02146011\n",
       "19      chr1 100029136 100029435 EXTL2      K562      0.009816065  0.01887563\n",
       "20      chr1 100029136 100029435 FRRS1      K562      0.008823450  0.01689535\n",
       "21      chr1 100029136 100029435 MFSD14A    K562      0.013446344  0.01664453\n",
       "22      chr1 100029136 100029435 RTCA       K562      0.036738457  0.01777343\n",
       "23      chr1 100029136 100029435 RTCA-AS1   K562     -0.021766504  0.01689104\n",
       "24      chr1 100029136 100029435 SASS6      K562      0.016385121  0.01543190\n",
       "25      chr1 100029136 100029435 SLC30A7    K562      0.009428935  0.01856813\n",
       "26      chr1 100029136 100029435 SLC35A3    K562      0.010946416  0.01679384\n",
       "27      chr1 100029136 100029435 TRMT13     K562      0.009483572  0.01748185\n",
       "28      chr1 100029136 100029435 VCAM1      K562      0.004404926  0.02085092\n",
       "29      chr1 100034578 100034758 AGL        K562     -0.015791236  0.01950572\n",
       "30      chr1 100034578 100034758 CDC14A     K562      0.032417317  0.02385925\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮         \n",
       "2423575 chrX  997541    998270   ASMT       K562      3.125955e-02 0.01854107\n",
       "2423576 chrX  997541    998270   ASMTL      K562      2.048386e-02 0.01882048\n",
       "2423577 chrX  997541    998270   CSF2RA     K562     -9.391297e-04 0.02137257\n",
       "2423578 chrX  997541    998270   GTPBP6     K562     -3.506227e-03 0.01854163\n",
       "2423579 chrX  997541    998270   IL3RA      K562      4.332036e-03 0.02928276\n",
       "2423580 chrX  997541    998270   LINC00106  K562     -1.813808e-02 0.02167384\n",
       "2423581 chrX  997541    998270   P2RY8      K562     -1.336805e-02 0.03465214\n",
       "2423582 chrX  997541    998270   PLCXD1     K562      1.536749e-02 0.02088157\n",
       "2423583 chrX  997541    998270   PPP2R3B    K562      1.767213e-02 0.02006108\n",
       "2423584 chrX  997541    998270   SLC25A6    K562      2.899253e-02 0.02188817\n",
       "2423585 chrX 9988858   9989079   CLCN4      K562     -7.586260e-03 0.02076492\n",
       "2423586 chrX 9988858   9989079   TBL1X      K562      8.118087e-05 0.01524722\n",
       "2423587 chrX 9988858   9989079   WWC3       K562      5.408002e-02 0.01930517\n",
       "2423588 chrX  999293    999875   AKAP17A    K562     -4.641255e-03 0.01869073\n",
       "2423589 chrX  999293    999875   ASMT       K562     -1.834867e-02 0.01824577\n",
       "2423590 chrX  999293    999875   ASMTL      K562     -1.917408e-02 0.02075847\n",
       "2423591 chrX  999293    999875   CSF2RA     K562     -2.865688e-03 0.02030021\n",
       "2423592 chrX  999293    999875   GTPBP6     K562      3.862025e-02 0.01965268\n",
       "2423593 chrX  999293    999875   IL3RA      K562      9.658144e-04 0.03377033\n",
       "2423594 chrX  999293    999875   LINC00106  K562      5.215708e-03 0.02415971\n",
       "2423595 chrX  999293    999875   P2RY8      K562      1.742263e-03 0.03412070\n",
       "2423596 chrX  999293    999875   PLCXD1     K562     -5.122071e-03 0.01845258\n",
       "2423597 chrX  999293    999875   PPP2R3B    K562      3.685852e-02 0.02134755\n",
       "2423598 chrX  999293    999875   SLC25A6    K562     -7.182472e-03 0.02459538\n",
       "2423599 chrX 9995654   9996249   CLCN4      K562      3.161529e-02 0.02004924\n",
       "2423600 chrX 9995654   9996249   TBL1X      K562     -6.495872e-03 0.02286679\n",
       "2423601 chrX 9995654   9996249   WWC3       K562     -1.024784e-02 0.02049721\n",
       "2423602 chrX 9997430   9997980   CLCN4      K562     -1.234859e-02 0.02046544\n",
       "2423603 chrX 9997430   9997980   TBL1X      K562      7.726147e-04 0.02298103\n",
       "2423604 chrX 9997430   9997980   WWC3       K562      4.332281e-02 0.01987783\n",
       "        rBgMean       pvalZ      \n",
       "1        0.0135132819 0.50408565 \n",
       "2        0.0177004597 0.63934453 \n",
       "3        0.0071304952 0.19147998 \n",
       "4        0.0116089305 0.04513284 \n",
       "5        0.0237448037 0.43210699 \n",
       "6        0.0121480184 0.40622398 \n",
       "7        0.0047155313 0.76155850 \n",
       "8        0.0012128211 0.37573798 \n",
       "9        0.0059057048 0.80683860 \n",
       "10       0.0001442482 0.55531073 \n",
       "11       0.0115996754 0.69871106 \n",
       "12       0.0041166975 0.14093538 \n",
       "13       0.0056048369 0.20892180 \n",
       "14       0.0182248750 0.07788503 \n",
       "15       0.0081718465 0.77634455 \n",
       "16       0.0111601011 0.78704757 \n",
       "17       0.0062062424 0.21826581 \n",
       "18       0.0045038311 0.33554447 \n",
       "19       0.0110784261 0.52666052 \n",
       "20       0.0070475471 0.45814349 \n",
       "21      -0.0005385988 0.20039465 \n",
       "22       0.0040990462 0.03314810 \n",
       "23      -0.0000324883 0.90090376 \n",
       "24       0.0024023919 0.18244342 \n",
       "25       0.0051188581 0.40822151 \n",
       "26       0.0011642227 0.28011930 \n",
       "27       0.0044174089 0.38598629 \n",
       "28       0.0108685409 0.62171640 \n",
       "29       0.0146974280 0.94098109 \n",
       "30       0.0182303749 0.27605220 \n",
       "⋮       ⋮             ⋮          \n",
       "2423575  0.003839971  0.069589355\n",
       "2423576  0.019852217  0.486613332\n",
       "2423577  0.019743296  0.833405120\n",
       "2423578  0.018032664  0.877311277\n",
       "2423579  0.024998192  0.759826577\n",
       "2423580  0.030851479  0.988098924\n",
       "2423581 -0.019192585  0.433257887\n",
       "2423582  0.018723511  0.563841877\n",
       "2423583  0.027820936  0.693534960\n",
       "2423584 -0.015786621  0.020387269\n",
       "2423585  0.001176876  0.663493474\n",
       "2423586 -0.004853582  0.373101503\n",
       "2423587  0.005931187  0.006314035\n",
       "2423588 -0.001703902  0.562438880\n",
       "2423589  0.004600310  0.895762501\n",
       "2423590  0.018832733  0.966442719\n",
       "2423591  0.017138500  0.837790573\n",
       "2423592  0.021200369  0.187704451\n",
       "2423593  0.030287787  0.807378387\n",
       "2423594  0.029132486  0.838899470\n",
       "2423595 -0.015944023  0.302108881\n",
       "2423596  0.018045145  0.895351323\n",
       "2423597  0.028050325  0.339946537\n",
       "2423598 -0.013688983  0.395681174\n",
       "2423599  0.003544645  0.080743972\n",
       "2423600 -0.007641585  0.480019839\n",
       "2423601  0.011236877  0.852721598\n",
       "2423602  0.004072029  0.788827155\n",
       "2423603 -0.007604769  0.357729383\n",
       "2423604  0.004698987  0.026004413"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   foreach_1.5.2                    \n",
       " [3] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.68.0                  \n",
       " [5] rtracklayer_1.60.1                Biostrings_2.70.1                \n",
       " [7] XVector_0.42.0                    genomation_1.34.0                \n",
       " [9] Seurat_5.0.3                      SeuratObject_5.0.1               \n",
       "[11] sp_2.1-4                          FigR_0.1.0                       \n",
       "[13] motifmatchr_1.24.0                chromVAR_1.24.0                  \n",
       "[15] cowplot_1.1.3                     ggplot2_3.5.1                    \n",
       "[17] dplyr_1.1.4                       SummarizedExperiment_1.30.2      \n",
       "[19] Biobase_2.60.0                    GenomicRanges_1.54.1             \n",
       "[21] GenomeInfoDb_1.38.1               IRanges_2.36.0                   \n",
       "[23] S4Vectors_0.40.2                  BiocGenerics_0.48.1              \n",
       "[25] MatrixGenerics_1.12.3             matrixStats_1.3.0                \n",
       "[27] Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             vroom_1.6.5                \n",
       " [15] globals_0.16.3              lattice_0.22-6             \n",
       " [17] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [19] plotly_4.10.4               plotrix_3.8-4              \n",
       " [21] yaml_2.3.8                  httpuv_1.6.15              \n",
       " [23] sctransform_0.4.1           spam_2.10-0                \n",
       " [25] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [27] pbapply_1.7-2               DBI_1.2.2                  \n",
       " [29] CNEr_1.38.0                 RColorBrewer_1.1-3         \n",
       " [31] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [33] Rtsne_0.17                  purrr_1.0.2                \n",
       " [35] R.utils_2.12.3              RCurl_1.98-1.14            \n",
       " [37] pracma_2.4.4                GenomeInfoDbData_1.2.11    \n",
       " [39] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [41] spatstat.utils_3.0-4        listenv_0.9.1              \n",
       " [43] seqLogo_1.68.0              goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             spatstat.random_3.2-3      \n",
       " [47] annotate_1.80.0             fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           leiden_0.4.3.1             \n",
       " [51] codetools_0.2-20            DelayedArray_0.26.7        \n",
       " [53] DT_0.33                     tidyselect_1.2.1           \n",
       " [55] spatstat.explore_3.2-6      base64enc_0.1-3            \n",
       " [57] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [59] progressr_0.14.0            iterators_1.0.14           \n",
       " [61] ggridges_0.5.6              survival_3.6-4             \n",
       " [63] tools_4.3.3                 TFMPvalue_0.0.9            \n",
       " [65] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [67] glue_1.7.0                  gridExtra_2.3              \n",
       " [69] IRdisplay_1.1               withr_3.0.0                \n",
       " [71] fastmap_1.1.1               fansi_1.0.6                \n",
       " [73] caTools_1.18.2              digest_0.6.35              \n",
       " [75] R6_2.5.1                    mime_0.12                  \n",
       " [77] nabor_0.5.0                 seqPattern_1.34.0          \n",
       " [79] colorspace_2.1-0            scattermore_1.2            \n",
       " [81] GO.db_3.18.0                tensor_1.5                 \n",
       " [83] gtools_3.9.5                poweRlaw_0.80.0            \n",
       " [85] spatstat.data_3.0-4         RSQLite_2.3.6              \n",
       " [87] R.methodsS3_1.8.2           utf8_1.2.4                 \n",
       " [89] tidyr_1.3.1                 generics_0.1.3             \n",
       " [91] data.table_1.15.2           httr_1.4.7                 \n",
       " [93] htmlwidgets_1.6.4           S4Arrays_1.2.1             \n",
       " [95] TFBSTools_1.40.0            uwot_0.1.16                \n",
       " [97] pkgconfig_2.0.3             gtable_0.3.5               \n",
       " [99] blob_1.2.4                  impute_1.76.0              \n",
       "[101] lmtest_0.9-40               htmltools_0.5.8.1          \n",
       "[103] dotCall64_1.1-1             scales_1.3.0               \n",
       "[105] png_0.1-8                   tzdb_0.4.0                 \n",
       "[107] reshape2_1.4.4              rjson_0.2.21               \n",
       "[109] uuid_1.2-0                  nlme_3.1-164               \n",
       "[111] repr_1.1.7                  cachem_1.0.8               \n",
       "[113] zoo_1.8-12                  stringr_1.5.1              \n",
       "[115] KernSmooth_2.23-22          miniUI_0.1.1.1             \n",
       "[117] AnnotationDbi_1.64.1        restfulr_0.0.15            \n",
       "[119] pillar_1.9.0                vctrs_0.6.5                \n",
       "[121] RANN_2.6.1                  promises_1.3.0             \n",
       "[123] xtable_1.8-4                cluster_2.1.6              \n",
       "[125] evaluate_0.23               readr_2.1.5                \n",
       "[127] cli_3.6.2                   compiler_4.3.3             \n",
       "[129] Rsamtools_2.18.0            rlang_1.1.3                \n",
       "[131] crayon_1.5.2                future.apply_1.11.2        \n",
       "[133] plyr_1.8.9                  stringi_1.8.3              \n",
       "[135] gridBase_0.4-7              deldir_2.0-4               \n",
       "[137] viridisLite_0.4.2           BiocParallel_1.36.0        \n",
       "[139] munsell_0.5.1               lazyeval_0.2.2             \n",
       "[141] spatstat.geom_3.2-9         IRkernel_1.3.2             \n",
       "[143] RcppHNSW_0.6.0              hms_1.1.3                  \n",
       "[145] patchwork_1.2.0             bit64_4.0.5                \n",
       "[147] future_1.33.2               KEGGREST_1.40.1            \n",
       "[149] shiny_1.8.1.1               ROCR_1.0-11                \n",
       "[151] igraph_2.0.3                memoise_2.0.1              \n",
       "[153] bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
