{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: rhdf5\n",
      "\n",
      "Warning message:\n",
      "“no DISPLAY variable so Tk is not available”\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SnapATAC)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(parallel)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/HCT116/HCT116/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/HCT116/HCT116/multiome_powerlaw_v3/encode_e2g_predictions.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/HCT116/HCT116/Kendall/atac_matrix.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/scE2G_input_Maya/HCT116/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/HCT116/1.Genome_wide_prediction/SnapATAC/SnapATAC.250321/\"\n",
    "celltype = \"HCT116\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2938315 ranges and 3 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100034595-100034763      * |         AGL\n",
       "        [2]     chr1 100034595-100034763      * |      CDC14A\n",
       "        [3]     chr1 100034595-100034763      * |         DBT\n",
       "        [4]     chr1 100034595-100034763      * |        DPH5\n",
       "        [5]     chr1 100034595-100034763      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2938311]     chrX     9911462-9911841      * |        WWC3\n",
       "  [2938312]     chrX     9995795-9996095      * |       CLCN4\n",
       "  [2938313]     chrX     9995795-9996095      * |       FAM9B\n",
       "  [2938314]     chrX     9995795-9996095      * |       TBL1X\n",
       "  [2938315]     chrX     9995795-9996095      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-100034595-10003.. chr1-100034595-10003..\n",
       "        [2] chr1-100034595-10003.. chr1-100034595-10003..\n",
       "        [3] chr1-100034595-10003.. chr1-100034595-10003..\n",
       "        [4] chr1-100034595-10003.. chr1-100034595-10003..\n",
       "        [5] chr1-100034595-10003.. chr1-100034595-10003..\n",
       "        ...                    ...                    ...\n",
       "  [2938311]   chrX-9911462-9911841 chrX-9911462-9911841..\n",
       "  [2938312]   chrX-9995795-9996095 chrX-9995795-9996095..\n",
       "  [2938313]   chrX-9995795-9996095 chrX-9995795-9996095..\n",
       "  [2938314]   chrX-9995795-9996095 chrX-9995795-9996095..\n",
       "  [2938315]   chrX-9995795-9996095 chrX-9995795-9996095..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = BinarizeCounts(matrix.atac_count)\n",
    "rm(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna_count = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna_count)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4986774c-cbbf-4b72-a299-d681bddca80e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "85b165d0-f1db-4548-beac-235f8950107b",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "655ebba3-731e-49d8-a04e-1ebb5becd567",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Run SnapATAC prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "Sys.setenv(OPENBLAS_NUM_THREADS=1)\n",
    "PredictionSnapATAC = function(pairs.E2G,\n",
    "                              data.RNA,\n",
    "                              data.ATAC,\n",
    "                              n.core = 16){\n",
    "  my.cluster <- parallel::makeCluster(\n",
    "    n.core,\n",
    "    type = \"PSOCK\"\n",
    "  )\n",
    "  doParallel::registerDoParallel(cl = my.cluster)\n",
    "  \n",
    "  pairs.E2G =  \n",
    "    foreach (index.tmp = 1:length(pairs.E2G),\n",
    "             .combine = 'c',\n",
    "             .packages = c(\"Matrix\",\n",
    "                           \"GenomicRanges\")) %dopar% {\n",
    "\n",
    "                             pairs.E2G.tmp = pairs.E2G[index.tmp]\n",
    "                             \n",
    "                             model.tmp = summary(stats::glm(data.ATAC[pairs.E2G.tmp$PeakName,] ~ data.RNA[pairs.E2G.tmp$TargetGene,], \n",
    "                                                            family = binomial(link = \"logit\")))[[\"coefficients\"]]\n",
    "                             mcols(pairs.E2G.tmp)[,c(\"beta\",\"sd\",\"z\",\"p\")] = \n",
    "                               t(data.frame(model.tmp[2,]))\n",
    "                             \n",
    "                             pairs.E2G.tmp\n",
    "                           }\n",
    "  \n",
    "  parallel::stopCluster(cl = my.cluster)\n",
    "  \n",
    "  return(pairs.E2G)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "[1] \"chr10\"\n",
      "[1] \"chr11\"\n",
      "[1] \"chr12\"\n",
      "[1] \"chr13\"\n",
      "[1] \"chr14\"\n",
      "[1] \"chr15\"\n",
      "[1] \"chr16\"\n",
      "[1] \"chr17\"\n",
      "[1] \"chr18\"\n",
      "[1] \"chr19\"\n",
      "[1] \"chr2\"\n",
      "[1] \"chr20\"\n",
      "[1] \"chr21\"\n",
      "[1] \"chr22\"\n",
      "[1] \"chr3\"\n",
      "[1] \"chr4\"\n",
      "[1] \"chr5\"\n",
      "[1] \"chr6\"\n",
      "[1] \"chr7\"\n",
      "[1] \"chr8\"\n",
      "[1] \"chr9\"\n",
      "[1] \"chrX\"\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  mcols(pairs.E2G.chr.res)[,c(\"beta\",\"sd\",\"z\",\"p\")] = NA\n",
    "  pairs.E2G.chr.res = PredictionSnapATAC(pairs.E2G.chr.res,\n",
    "                                         matrix.rna.rename[rownames(matrix.rna.rename) %in% pairs.E2G.chr.res$TargetGene,],\n",
    "                                         matrix.atac[rownames(matrix.atac) %in% pairs.E2G.chr.res$PeakName,],\n",
    "                                         n.cores)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2843332 ranges and 7 metadata columns:\n",
       "       seqnames              ranges strand |  TargetGene               PeakName\n",
       "          <Rle>           <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 100034595-100034763      * |         AGL chr1-100034595-10003..\n",
       "  chr1     chr1 100034595-100034763      * |      CDC14A chr1-100034595-10003..\n",
       "  chr1     chr1 100034595-100034763      * |         DBT chr1-100034595-10003..\n",
       "  chr1     chr1 100034595-100034763      * |        DPH5 chr1-100034595-10003..\n",
       "  chr1     chr1 100034595-100034763      * |       EXTL2 chr1-100034595-10003..\n",
       "   ...      ...                 ...    ... .         ...                    ...\n",
       "  chrX     chrX     9911462-9911841      * |        WWC3   chrX-9911462-9911841\n",
       "  chrX     chrX     9995795-9996095      * |       CLCN4   chrX-9995795-9996095\n",
       "  chrX     chrX     9995795-9996095      * |       FAM9B   chrX-9995795-9996095\n",
       "  chrX     chrX     9995795-9996095      * |       TBL1X   chrX-9995795-9996095\n",
       "  chrX     chrX     9995795-9996095      * |        WWC3   chrX-9995795-9996095\n",
       "                     PairName        beta          sd          z         p\n",
       "                  <character>   <numeric>   <numeric>  <numeric> <numeric>\n",
       "  chr1 chr1-100034595-10003..   0.0966871    0.302473   0.319655  0.749230\n",
       "  chr1 chr1-100034595-10003..   0.1835532    0.366508   0.500816  0.616501\n",
       "  chr1 chr1-100034595-10003..   0.0844564    0.380384   0.222029  0.824291\n",
       "  chr1 chr1-100034595-10003..  -0.4662489    0.483889  -0.963546  0.335274\n",
       "  chr1 chr1-100034595-10003..  -0.2125478    0.379381  -0.560249  0.575309\n",
       "   ...                    ...         ...         ...        ...       ...\n",
       "  chrX chrX-9911462-9911841.. -45.3172777 3771.511893 -0.0120157  0.990413\n",
       "  chrX chrX-9995795-9996095.. -42.8950758 2629.351508 -0.0163139  0.986984\n",
       "  chrX chrX-9995795-9996095.. -15.0864058  829.081672 -0.0181965  0.985482\n",
       "  chrX chrX-9995795-9996095..  -0.0466583    0.169166 -0.2758141  0.782691\n",
       "  chrX chrX-9995795-9996095..  -0.6165867    0.697468 -0.8840353  0.376677\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2843332 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>beta</th><th scope=col>sd</th><th scope=col>z</th><th scope=col>p</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>AGL     </td><td>HCT116</td><td>  0.09668708</td><td>3.024730e-01</td><td> 0.31965524</td><td>0.74922970</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>CDC14A  </td><td>HCT116</td><td>  0.18355315</td><td>3.665083e-01</td><td> 0.50081583</td><td>0.61650075</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>DBT     </td><td>HCT116</td><td>  0.08445645</td><td>3.803844e-01</td><td> 0.22202923</td><td>0.82429113</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>DPH5    </td><td>HCT116</td><td> -0.46624892</td><td>4.838887e-01</td><td>-0.96354587</td><td>0.33527366</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>EXTL2   </td><td>HCT116</td><td> -0.21254782</td><td>3.793808e-01</td><td>-0.56024934</td><td>0.57530937</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>FRRS1   </td><td>HCT116</td><td> -0.55990653</td><td>5.504667e-01</td><td>-1.01714885</td><td>0.30908263</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>MFSD14A </td><td>HCT116</td><td>  0.29025694</td><td>2.799339e-01</td><td> 1.03687682</td><td>0.29979327</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>PLPPR4  </td><td>HCT116</td><td>-29.82663007</td><td>2.014787e+03</td><td>-0.01480386</td><td>0.98818866</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>RTCA    </td><td>HCT116</td><td>  0.43047533</td><td>3.061686e-01</td><td> 1.40600753</td><td>0.15972189</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>RTCA-AS1</td><td>HCT116</td><td>-30.06039198</td><td>1.934728e+03</td><td>-0.01553727</td><td>0.98760355</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>SASS6   </td><td>HCT116</td><td> -0.13588317</td><td>3.712150e-01</td><td>-0.36604980</td><td>0.71432791</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>SLC30A7 </td><td>HCT116</td><td> -0.82078821</td><td>3.374615e-01</td><td>-2.43224234</td><td>0.01500566</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>SLC35A3 </td><td>HCT116</td><td>  0.15579472</td><td>2.742344e-01</td><td> 0.56810782</td><td>0.56996176</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>TRMT13  </td><td>HCT116</td><td>  0.32210966</td><td>3.403561e-01</td><td> 0.94639006</td><td>0.34394968</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>AGL     </td><td>HCT116</td><td>  0.09376110</td><td>8.227838e-02</td><td> 1.13955941</td><td>0.25446990</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>CDC14A  </td><td>HCT116</td><td>  0.05771182</td><td>1.061016e-01</td><td> 0.54392955</td><td>0.58648995</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>DBT     </td><td>HCT116</td><td>  0.09744878</td><td>1.029016e-01</td><td> 0.94700951</td><td>0.34363393</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>DPH5    </td><td>HCT116</td><td> -0.02599482</td><td>1.021325e-01</td><td>-0.25452044</td><td>0.79909351</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>EXTL2   </td><td>HCT116</td><td>  0.05226019</td><td>9.006044e-02</td><td> 0.58027910</td><td>0.56172642</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>FRRS1   </td><td>HCT116</td><td> -0.06148027</td><td>1.105925e-01</td><td>-0.55591698</td><td>0.57826761</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>MFSD14A </td><td>HCT116</td><td> -0.01203135</td><td>8.579972e-02</td><td>-0.14022599</td><td>0.88848143</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>PLPPR4  </td><td>HCT116</td><td>  0.05074201</td><td>3.795120e-01</td><td> 0.13370332</td><td>0.89363718</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>RTCA    </td><td>HCT116</td><td> -0.03646012</td><td>1.034141e-01</td><td>-0.35256431</td><td>0.72441510</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>RTCA-AS1</td><td>HCT116</td><td> -0.64489886</td><td>3.189826e-01</td><td>-2.02173692</td><td>0.04320354</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>SASS6   </td><td>HCT116</td><td>  0.15471803</td><td>8.817360e-02</td><td> 1.75469786</td><td>0.07931100</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>SLC30A7 </td><td>HCT116</td><td> -0.04060774</td><td>6.536440e-02</td><td>-0.62125155</td><td>0.53443413</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>SLC35A3 </td><td>HCT116</td><td>  0.04266143</td><td>7.740949e-02</td><td> 0.55111371</td><td>0.58155572</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>TRMT13  </td><td>HCT116</td><td>  0.02646403</td><td>1.061844e-01</td><td> 0.24922711</td><td>0.80318511</td></tr>\n",
       "\t<tr><td>chr1</td><td>100046025</td><td>100046641</td><td>AGL     </td><td>HCT116</td><td> -0.02388234</td><td>2.915017e-01</td><td>-0.08192866</td><td>0.93470345</td></tr>\n",
       "\t<tr><td>chr1</td><td>100046025</td><td>100046641</td><td>CDC14A  </td><td>HCT116</td><td> -0.57411922</td><td>5.072311e-01</td><td>-1.13186916</td><td>0.25768945</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>ASMTL    </td><td>HCT116</td><td> 1.369647e-04</td><td>   0.2068302</td><td> 0.0006622086</td><td>0.9994716340</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>CSF2RA   </td><td>HCT116</td><td>-1.353126e-01</td><td>   0.3313158</td><td>-0.4084098425</td><td>0.6829728086</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>GTPBP6   </td><td>HCT116</td><td>-3.269883e-01</td><td>   0.3686049</td><td>-0.8870969807</td><td>0.3750266895</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>IL3RA    </td><td>HCT116</td><td> 6.482838e-01</td><td>   0.3881852</td><td> 1.6700373457</td><td>0.0949119749</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>LINC00106</td><td>HCT116</td><td> 2.984671e-01</td><td>   0.2983762</td><td> 1.0003046380</td><td>0.3171631034</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>P2RY8    </td><td>HCT116</td><td>-3.376590e+01</td><td>1312.5040843</td><td>-0.0257263215</td><td>0.9794756293</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>PLCXD1   </td><td>HCT116</td><td>-5.126421e-03</td><td>   0.3320241</td><td>-0.0154399055</td><td>0.9876812272</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>PPP2R3B  </td><td>HCT116</td><td> 2.292070e-01</td><td>   0.3457275</td><td> 0.6629700795</td><td>0.5073497170</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>SHOX     </td><td>HCT116</td><td>-1.189372e+01</td><td> 494.7870814</td><td>-0.0240380633</td><td>0.9808222473</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>SLC25A6  </td><td>HCT116</td><td> 2.514408e-02</td><td>   0.1136494</td><td> 0.2212424297</td><td>0.8249036736</td></tr>\n",
       "\t<tr><td>chrX</td><td>9902739</td><td>9902992</td><td>CLCN4    </td><td>HCT116</td><td>-4.507918e-01</td><td>   1.9075254</td><td>-0.2363228176</td><td>0.8131821824</td></tr>\n",
       "\t<tr><td>chrX</td><td>9902739</td><td>9902992</td><td>FAM9B    </td><td>HCT116</td><td>-1.515349e+01</td><td>1367.9308804</td><td>-0.0110776688</td><td>0.9911614799</td></tr>\n",
       "\t<tr><td>chrX</td><td>9902739</td><td>9902992</td><td>TBL1X    </td><td>HCT116</td><td>-2.656844e-01</td><td>   0.2852540</td><td>-0.9313957692</td><td>0.3516488786</td></tr>\n",
       "\t<tr><td>chrX</td><td>9902739</td><td>9902992</td><td>WWC3     </td><td>HCT116</td><td> 2.532484e-01</td><td>   0.6805315</td><td> 0.3721333003</td><td>0.7097936039</td></tr>\n",
       "\t<tr><td>chrX</td><td>9903251</td><td>9903473</td><td>CLCN4    </td><td>HCT116</td><td>-4.197738e+01</td><td>4336.9719362</td><td>-0.0096789614</td><td>0.9922774267</td></tr>\n",
       "\t<tr><td>chrX</td><td>9903251</td><td>9903473</td><td>FAM9B    </td><td>HCT116</td><td>-1.473992e+01</td><td>1367.0718523</td><td>-0.0107821102</td><td>0.9913972875</td></tr>\n",
       "\t<tr><td>chrX</td><td>9903251</td><td>9903473</td><td>TBL1X    </td><td>HCT116</td><td>-4.272078e-01</td><td>   0.3404950</td><td>-1.2546667892</td><td>0.2095997467</td></tr>\n",
       "\t<tr><td>chrX</td><td>9903251</td><td>9903473</td><td>WWC3     </td><td>HCT116</td><td> 1.121024e+00</td><td>   0.4800354</td><td> 2.3352943569</td><td>0.0195280488</td></tr>\n",
       "\t<tr><td>chrX</td><td>9906862</td><td>9908314</td><td>CLCN4    </td><td>HCT116</td><td>-3.048051e-01</td><td>   0.6802224</td><td>-0.4480963178</td><td>0.6540836854</td></tr>\n",
       "\t<tr><td>chrX</td><td>9906862</td><td>9908314</td><td>FAM9B    </td><td>HCT116</td><td>-1.501650e+01</td><td> 502.0195436</td><td>-0.0299121859</td><td>0.9761370872</td></tr>\n",
       "\t<tr><td>chrX</td><td>9906862</td><td>9908314</td><td>TBL1X    </td><td>HCT116</td><td>-1.122924e-01</td><td>   0.1079954</td><td>-1.0397887226</td><td>0.2984380697</td></tr>\n",
       "\t<tr><td>chrX</td><td>9906862</td><td>9908314</td><td>WWC3     </td><td>HCT116</td><td> 3.019850e-01</td><td>   0.2631925</td><td> 1.1473920719</td><td>0.2512196163</td></tr>\n",
       "\t<tr><td>chrX</td><td>9911462</td><td>9911841</td><td>CLCN4    </td><td>HCT116</td><td>-4.043799e+01</td><td>2629.6920671</td><td>-0.0153774614</td><td>0.9877310445</td></tr>\n",
       "\t<tr><td>chrX</td><td>9911462</td><td>9911841</td><td>FAM9B    </td><td>HCT116</td><td> 2.594126e+00</td><td>   0.7262598</td><td> 3.5718979047</td><td>0.0003544036</td></tr>\n",
       "\t<tr><td>chrX</td><td>9911462</td><td>9911841</td><td>TBL1X    </td><td>HCT116</td><td>-1.332225e-01</td><td>   0.2291654</td><td>-0.5813378407</td><td>0.5610127819</td></tr>\n",
       "\t<tr><td>chrX</td><td>9911462</td><td>9911841</td><td>WWC3     </td><td>HCT116</td><td>-4.531728e+01</td><td>3771.5118932</td><td>-0.0120156794</td><td>0.9904131056</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995795</td><td>9996095</td><td>CLCN4    </td><td>HCT116</td><td>-4.289508e+01</td><td>2629.3515084</td><td>-0.0163139374</td><td>0.9869839386</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995795</td><td>9996095</td><td>FAM9B    </td><td>HCT116</td><td>-1.508641e+01</td><td> 829.0816720</td><td>-0.0181965255</td><td>0.9854820744</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995795</td><td>9996095</td><td>TBL1X    </td><td>HCT116</td><td>-4.665829e-02</td><td>   0.1691657</td><td>-0.2758141327</td><td>0.7826908283</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995795</td><td>9996095</td><td>WWC3     </td><td>HCT116</td><td>-6.165867e-01</td><td>   0.6974684</td><td>-0.8840353350</td><td>0.3766771386</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2843332 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & beta & sd & z & p\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100034595 & 100034763 & AGL      & HCT116 &   0.09668708 & 3.024730e-01 &  0.31965524 & 0.74922970\\\\\n",
       "\t chr1 & 100034595 & 100034763 & CDC14A   & HCT116 &   0.18355315 & 3.665083e-01 &  0.50081583 & 0.61650075\\\\\n",
       "\t chr1 & 100034595 & 100034763 & DBT      & HCT116 &   0.08445645 & 3.803844e-01 &  0.22202923 & 0.82429113\\\\\n",
       "\t chr1 & 100034595 & 100034763 & DPH5     & HCT116 &  -0.46624892 & 4.838887e-01 & -0.96354587 & 0.33527366\\\\\n",
       "\t chr1 & 100034595 & 100034763 & EXTL2    & HCT116 &  -0.21254782 & 3.793808e-01 & -0.56024934 & 0.57530937\\\\\n",
       "\t chr1 & 100034595 & 100034763 & FRRS1    & HCT116 &  -0.55990653 & 5.504667e-01 & -1.01714885 & 0.30908263\\\\\n",
       "\t chr1 & 100034595 & 100034763 & MFSD14A  & HCT116 &   0.29025694 & 2.799339e-01 &  1.03687682 & 0.29979327\\\\\n",
       "\t chr1 & 100034595 & 100034763 & PLPPR4   & HCT116 & -29.82663007 & 2.014787e+03 & -0.01480386 & 0.98818866\\\\\n",
       "\t chr1 & 100034595 & 100034763 & RTCA     & HCT116 &   0.43047533 & 3.061686e-01 &  1.40600753 & 0.15972189\\\\\n",
       "\t chr1 & 100034595 & 100034763 & RTCA-AS1 & HCT116 & -30.06039198 & 1.934728e+03 & -0.01553727 & 0.98760355\\\\\n",
       "\t chr1 & 100034595 & 100034763 & SASS6    & HCT116 &  -0.13588317 & 3.712150e-01 & -0.36604980 & 0.71432791\\\\\n",
       "\t chr1 & 100034595 & 100034763 & SLC30A7  & HCT116 &  -0.82078821 & 3.374615e-01 & -2.43224234 & 0.01500566\\\\\n",
       "\t chr1 & 100034595 & 100034763 & SLC35A3  & HCT116 &   0.15579472 & 2.742344e-01 &  0.56810782 & 0.56996176\\\\\n",
       "\t chr1 & 100034595 & 100034763 & TRMT13   & HCT116 &   0.32210966 & 3.403561e-01 &  0.94639006 & 0.34394968\\\\\n",
       "\t chr1 & 100037727 & 100038876 & AGL      & HCT116 &   0.09376110 & 8.227838e-02 &  1.13955941 & 0.25446990\\\\\n",
       "\t chr1 & 100037727 & 100038876 & CDC14A   & HCT116 &   0.05771182 & 1.061016e-01 &  0.54392955 & 0.58648995\\\\\n",
       "\t chr1 & 100037727 & 100038876 & DBT      & HCT116 &   0.09744878 & 1.029016e-01 &  0.94700951 & 0.34363393\\\\\n",
       "\t chr1 & 100037727 & 100038876 & DPH5     & HCT116 &  -0.02599482 & 1.021325e-01 & -0.25452044 & 0.79909351\\\\\n",
       "\t chr1 & 100037727 & 100038876 & EXTL2    & HCT116 &   0.05226019 & 9.006044e-02 &  0.58027910 & 0.56172642\\\\\n",
       "\t chr1 & 100037727 & 100038876 & FRRS1    & HCT116 &  -0.06148027 & 1.105925e-01 & -0.55591698 & 0.57826761\\\\\n",
       "\t chr1 & 100037727 & 100038876 & MFSD14A  & HCT116 &  -0.01203135 & 8.579972e-02 & -0.14022599 & 0.88848143\\\\\n",
       "\t chr1 & 100037727 & 100038876 & PLPPR4   & HCT116 &   0.05074201 & 3.795120e-01 &  0.13370332 & 0.89363718\\\\\n",
       "\t chr1 & 100037727 & 100038876 & RTCA     & HCT116 &  -0.03646012 & 1.034141e-01 & -0.35256431 & 0.72441510\\\\\n",
       "\t chr1 & 100037727 & 100038876 & RTCA-AS1 & HCT116 &  -0.64489886 & 3.189826e-01 & -2.02173692 & 0.04320354\\\\\n",
       "\t chr1 & 100037727 & 100038876 & SASS6    & HCT116 &   0.15471803 & 8.817360e-02 &  1.75469786 & 0.07931100\\\\\n",
       "\t chr1 & 100037727 & 100038876 & SLC30A7  & HCT116 &  -0.04060774 & 6.536440e-02 & -0.62125155 & 0.53443413\\\\\n",
       "\t chr1 & 100037727 & 100038876 & SLC35A3  & HCT116 &   0.04266143 & 7.740949e-02 &  0.55111371 & 0.58155572\\\\\n",
       "\t chr1 & 100037727 & 100038876 & TRMT13   & HCT116 &   0.02646403 & 1.061844e-01 &  0.24922711 & 0.80318511\\\\\n",
       "\t chr1 & 100046025 & 100046641 & AGL      & HCT116 &  -0.02388234 & 2.915017e-01 & -0.08192866 & 0.93470345\\\\\n",
       "\t chr1 & 100046025 & 100046641 & CDC14A   & HCT116 &  -0.57411922 & 5.072311e-01 & -1.13186916 & 0.25768945\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  989884 &  990608 & ASMTL     & HCT116 &  1.369647e-04 &    0.2068302 &  0.0006622086 & 0.9994716340\\\\\n",
       "\t chrX &  989884 &  990608 & CSF2RA    & HCT116 & -1.353126e-01 &    0.3313158 & -0.4084098425 & 0.6829728086\\\\\n",
       "\t chrX &  989884 &  990608 & GTPBP6    & HCT116 & -3.269883e-01 &    0.3686049 & -0.8870969807 & 0.3750266895\\\\\n",
       "\t chrX &  989884 &  990608 & IL3RA     & HCT116 &  6.482838e-01 &    0.3881852 &  1.6700373457 & 0.0949119749\\\\\n",
       "\t chrX &  989884 &  990608 & LINC00106 & HCT116 &  2.984671e-01 &    0.2983762 &  1.0003046380 & 0.3171631034\\\\\n",
       "\t chrX &  989884 &  990608 & P2RY8     & HCT116 & -3.376590e+01 & 1312.5040843 & -0.0257263215 & 0.9794756293\\\\\n",
       "\t chrX &  989884 &  990608 & PLCXD1    & HCT116 & -5.126421e-03 &    0.3320241 & -0.0154399055 & 0.9876812272\\\\\n",
       "\t chrX &  989884 &  990608 & PPP2R3B   & HCT116 &  2.292070e-01 &    0.3457275 &  0.6629700795 & 0.5073497170\\\\\n",
       "\t chrX &  989884 &  990608 & SHOX      & HCT116 & -1.189372e+01 &  494.7870814 & -0.0240380633 & 0.9808222473\\\\\n",
       "\t chrX &  989884 &  990608 & SLC25A6   & HCT116 &  2.514408e-02 &    0.1136494 &  0.2212424297 & 0.8249036736\\\\\n",
       "\t chrX & 9902739 & 9902992 & CLCN4     & HCT116 & -4.507918e-01 &    1.9075254 & -0.2363228176 & 0.8131821824\\\\\n",
       "\t chrX & 9902739 & 9902992 & FAM9B     & HCT116 & -1.515349e+01 & 1367.9308804 & -0.0110776688 & 0.9911614799\\\\\n",
       "\t chrX & 9902739 & 9902992 & TBL1X     & HCT116 & -2.656844e-01 &    0.2852540 & -0.9313957692 & 0.3516488786\\\\\n",
       "\t chrX & 9902739 & 9902992 & WWC3      & HCT116 &  2.532484e-01 &    0.6805315 &  0.3721333003 & 0.7097936039\\\\\n",
       "\t chrX & 9903251 & 9903473 & CLCN4     & HCT116 & -4.197738e+01 & 4336.9719362 & -0.0096789614 & 0.9922774267\\\\\n",
       "\t chrX & 9903251 & 9903473 & FAM9B     & HCT116 & -1.473992e+01 & 1367.0718523 & -0.0107821102 & 0.9913972875\\\\\n",
       "\t chrX & 9903251 & 9903473 & TBL1X     & HCT116 & -4.272078e-01 &    0.3404950 & -1.2546667892 & 0.2095997467\\\\\n",
       "\t chrX & 9903251 & 9903473 & WWC3      & HCT116 &  1.121024e+00 &    0.4800354 &  2.3352943569 & 0.0195280488\\\\\n",
       "\t chrX & 9906862 & 9908314 & CLCN4     & HCT116 & -3.048051e-01 &    0.6802224 & -0.4480963178 & 0.6540836854\\\\\n",
       "\t chrX & 9906862 & 9908314 & FAM9B     & HCT116 & -1.501650e+01 &  502.0195436 & -0.0299121859 & 0.9761370872\\\\\n",
       "\t chrX & 9906862 & 9908314 & TBL1X     & HCT116 & -1.122924e-01 &    0.1079954 & -1.0397887226 & 0.2984380697\\\\\n",
       "\t chrX & 9906862 & 9908314 & WWC3      & HCT116 &  3.019850e-01 &    0.2631925 &  1.1473920719 & 0.2512196163\\\\\n",
       "\t chrX & 9911462 & 9911841 & CLCN4     & HCT116 & -4.043799e+01 & 2629.6920671 & -0.0153774614 & 0.9877310445\\\\\n",
       "\t chrX & 9911462 & 9911841 & FAM9B     & HCT116 &  2.594126e+00 &    0.7262598 &  3.5718979047 & 0.0003544036\\\\\n",
       "\t chrX & 9911462 & 9911841 & TBL1X     & HCT116 & -1.332225e-01 &    0.2291654 & -0.5813378407 & 0.5610127819\\\\\n",
       "\t chrX & 9911462 & 9911841 & WWC3      & HCT116 & -4.531728e+01 & 3771.5118932 & -0.0120156794 & 0.9904131056\\\\\n",
       "\t chrX & 9995795 & 9996095 & CLCN4     & HCT116 & -4.289508e+01 & 2629.3515084 & -0.0163139374 & 0.9869839386\\\\\n",
       "\t chrX & 9995795 & 9996095 & FAM9B     & HCT116 & -1.508641e+01 &  829.0816720 & -0.0181965255 & 0.9854820744\\\\\n",
       "\t chrX & 9995795 & 9996095 & TBL1X     & HCT116 & -4.665829e-02 &    0.1691657 & -0.2758141327 & 0.7826908283\\\\\n",
       "\t chrX & 9995795 & 9996095 & WWC3      & HCT116 & -6.165867e-01 &    0.6974684 & -0.8840353350 & 0.3766771386\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2843332 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | beta &lt;dbl&gt; | sd &lt;dbl&gt; | z &lt;dbl&gt; | p &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100034595 | 100034763 | AGL      | HCT116 |   0.09668708 | 3.024730e-01 |  0.31965524 | 0.74922970 |\n",
       "| chr1 | 100034595 | 100034763 | CDC14A   | HCT116 |   0.18355315 | 3.665083e-01 |  0.50081583 | 0.61650075 |\n",
       "| chr1 | 100034595 | 100034763 | DBT      | HCT116 |   0.08445645 | 3.803844e-01 |  0.22202923 | 0.82429113 |\n",
       "| chr1 | 100034595 | 100034763 | DPH5     | HCT116 |  -0.46624892 | 4.838887e-01 | -0.96354587 | 0.33527366 |\n",
       "| chr1 | 100034595 | 100034763 | EXTL2    | HCT116 |  -0.21254782 | 3.793808e-01 | -0.56024934 | 0.57530937 |\n",
       "| chr1 | 100034595 | 100034763 | FRRS1    | HCT116 |  -0.55990653 | 5.504667e-01 | -1.01714885 | 0.30908263 |\n",
       "| chr1 | 100034595 | 100034763 | MFSD14A  | HCT116 |   0.29025694 | 2.799339e-01 |  1.03687682 | 0.29979327 |\n",
       "| chr1 | 100034595 | 100034763 | PLPPR4   | HCT116 | -29.82663007 | 2.014787e+03 | -0.01480386 | 0.98818866 |\n",
       "| chr1 | 100034595 | 100034763 | RTCA     | HCT116 |   0.43047533 | 3.061686e-01 |  1.40600753 | 0.15972189 |\n",
       "| chr1 | 100034595 | 100034763 | RTCA-AS1 | HCT116 | -30.06039198 | 1.934728e+03 | -0.01553727 | 0.98760355 |\n",
       "| chr1 | 100034595 | 100034763 | SASS6    | HCT116 |  -0.13588317 | 3.712150e-01 | -0.36604980 | 0.71432791 |\n",
       "| chr1 | 100034595 | 100034763 | SLC30A7  | HCT116 |  -0.82078821 | 3.374615e-01 | -2.43224234 | 0.01500566 |\n",
       "| chr1 | 100034595 | 100034763 | SLC35A3  | HCT116 |   0.15579472 | 2.742344e-01 |  0.56810782 | 0.56996176 |\n",
       "| chr1 | 100034595 | 100034763 | TRMT13   | HCT116 |   0.32210966 | 3.403561e-01 |  0.94639006 | 0.34394968 |\n",
       "| chr1 | 100037727 | 100038876 | AGL      | HCT116 |   0.09376110 | 8.227838e-02 |  1.13955941 | 0.25446990 |\n",
       "| chr1 | 100037727 | 100038876 | CDC14A   | HCT116 |   0.05771182 | 1.061016e-01 |  0.54392955 | 0.58648995 |\n",
       "| chr1 | 100037727 | 100038876 | DBT      | HCT116 |   0.09744878 | 1.029016e-01 |  0.94700951 | 0.34363393 |\n",
       "| chr1 | 100037727 | 100038876 | DPH5     | HCT116 |  -0.02599482 | 1.021325e-01 | -0.25452044 | 0.79909351 |\n",
       "| chr1 | 100037727 | 100038876 | EXTL2    | HCT116 |   0.05226019 | 9.006044e-02 |  0.58027910 | 0.56172642 |\n",
       "| chr1 | 100037727 | 100038876 | FRRS1    | HCT116 |  -0.06148027 | 1.105925e-01 | -0.55591698 | 0.57826761 |\n",
       "| chr1 | 100037727 | 100038876 | MFSD14A  | HCT116 |  -0.01203135 | 8.579972e-02 | -0.14022599 | 0.88848143 |\n",
       "| chr1 | 100037727 | 100038876 | PLPPR4   | HCT116 |   0.05074201 | 3.795120e-01 |  0.13370332 | 0.89363718 |\n",
       "| chr1 | 100037727 | 100038876 | RTCA     | HCT116 |  -0.03646012 | 1.034141e-01 | -0.35256431 | 0.72441510 |\n",
       "| chr1 | 100037727 | 100038876 | RTCA-AS1 | HCT116 |  -0.64489886 | 3.189826e-01 | -2.02173692 | 0.04320354 |\n",
       "| chr1 | 100037727 | 100038876 | SASS6    | HCT116 |   0.15471803 | 8.817360e-02 |  1.75469786 | 0.07931100 |\n",
       "| chr1 | 100037727 | 100038876 | SLC30A7  | HCT116 |  -0.04060774 | 6.536440e-02 | -0.62125155 | 0.53443413 |\n",
       "| chr1 | 100037727 | 100038876 | SLC35A3  | HCT116 |   0.04266143 | 7.740949e-02 |  0.55111371 | 0.58155572 |\n",
       "| chr1 | 100037727 | 100038876 | TRMT13   | HCT116 |   0.02646403 | 1.061844e-01 |  0.24922711 | 0.80318511 |\n",
       "| chr1 | 100046025 | 100046641 | AGL      | HCT116 |  -0.02388234 | 2.915017e-01 | -0.08192866 | 0.93470345 |\n",
       "| chr1 | 100046025 | 100046641 | CDC14A   | HCT116 |  -0.57411922 | 5.072311e-01 | -1.13186916 | 0.25768945 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  989884 |  990608 | ASMTL     | HCT116 |  1.369647e-04 |    0.2068302 |  0.0006622086 | 0.9994716340 |\n",
       "| chrX |  989884 |  990608 | CSF2RA    | HCT116 | -1.353126e-01 |    0.3313158 | -0.4084098425 | 0.6829728086 |\n",
       "| chrX |  989884 |  990608 | GTPBP6    | HCT116 | -3.269883e-01 |    0.3686049 | -0.8870969807 | 0.3750266895 |\n",
       "| chrX |  989884 |  990608 | IL3RA     | HCT116 |  6.482838e-01 |    0.3881852 |  1.6700373457 | 0.0949119749 |\n",
       "| chrX |  989884 |  990608 | LINC00106 | HCT116 |  2.984671e-01 |    0.2983762 |  1.0003046380 | 0.3171631034 |\n",
       "| chrX |  989884 |  990608 | P2RY8     | HCT116 | -3.376590e+01 | 1312.5040843 | -0.0257263215 | 0.9794756293 |\n",
       "| chrX |  989884 |  990608 | PLCXD1    | HCT116 | -5.126421e-03 |    0.3320241 | -0.0154399055 | 0.9876812272 |\n",
       "| chrX |  989884 |  990608 | PPP2R3B   | HCT116 |  2.292070e-01 |    0.3457275 |  0.6629700795 | 0.5073497170 |\n",
       "| chrX |  989884 |  990608 | SHOX      | HCT116 | -1.189372e+01 |  494.7870814 | -0.0240380633 | 0.9808222473 |\n",
       "| chrX |  989884 |  990608 | SLC25A6   | HCT116 |  2.514408e-02 |    0.1136494 |  0.2212424297 | 0.8249036736 |\n",
       "| chrX | 9902739 | 9902992 | CLCN4     | HCT116 | -4.507918e-01 |    1.9075254 | -0.2363228176 | 0.8131821824 |\n",
       "| chrX | 9902739 | 9902992 | FAM9B     | HCT116 | -1.515349e+01 | 1367.9308804 | -0.0110776688 | 0.9911614799 |\n",
       "| chrX | 9902739 | 9902992 | TBL1X     | HCT116 | -2.656844e-01 |    0.2852540 | -0.9313957692 | 0.3516488786 |\n",
       "| chrX | 9902739 | 9902992 | WWC3      | HCT116 |  2.532484e-01 |    0.6805315 |  0.3721333003 | 0.7097936039 |\n",
       "| chrX | 9903251 | 9903473 | CLCN4     | HCT116 | -4.197738e+01 | 4336.9719362 | -0.0096789614 | 0.9922774267 |\n",
       "| chrX | 9903251 | 9903473 | FAM9B     | HCT116 | -1.473992e+01 | 1367.0718523 | -0.0107821102 | 0.9913972875 |\n",
       "| chrX | 9903251 | 9903473 | TBL1X     | HCT116 | -4.272078e-01 |    0.3404950 | -1.2546667892 | 0.2095997467 |\n",
       "| chrX | 9903251 | 9903473 | WWC3      | HCT116 |  1.121024e+00 |    0.4800354 |  2.3352943569 | 0.0195280488 |\n",
       "| chrX | 9906862 | 9908314 | CLCN4     | HCT116 | -3.048051e-01 |    0.6802224 | -0.4480963178 | 0.6540836854 |\n",
       "| chrX | 9906862 | 9908314 | FAM9B     | HCT116 | -1.501650e+01 |  502.0195436 | -0.0299121859 | 0.9761370872 |\n",
       "| chrX | 9906862 | 9908314 | TBL1X     | HCT116 | -1.122924e-01 |    0.1079954 | -1.0397887226 | 0.2984380697 |\n",
       "| chrX | 9906862 | 9908314 | WWC3      | HCT116 |  3.019850e-01 |    0.2631925 |  1.1473920719 | 0.2512196163 |\n",
       "| chrX | 9911462 | 9911841 | CLCN4     | HCT116 | -4.043799e+01 | 2629.6920671 | -0.0153774614 | 0.9877310445 |\n",
       "| chrX | 9911462 | 9911841 | FAM9B     | HCT116 |  2.594126e+00 |    0.7262598 |  3.5718979047 | 0.0003544036 |\n",
       "| chrX | 9911462 | 9911841 | TBL1X     | HCT116 | -1.332225e-01 |    0.2291654 | -0.5813378407 | 0.5610127819 |\n",
       "| chrX | 9911462 | 9911841 | WWC3      | HCT116 | -4.531728e+01 | 3771.5118932 | -0.0120156794 | 0.9904131056 |\n",
       "| chrX | 9995795 | 9996095 | CLCN4     | HCT116 | -4.289508e+01 | 2629.3515084 | -0.0163139374 | 0.9869839386 |\n",
       "| chrX | 9995795 | 9996095 | FAM9B     | HCT116 | -1.508641e+01 |  829.0816720 | -0.0181965255 | 0.9854820744 |\n",
       "| chrX | 9995795 | 9996095 | TBL1X     | HCT116 | -4.665829e-02 |    0.1691657 | -0.2758141327 | 0.7826908283 |\n",
       "| chrX | 9995795 | 9996095 | WWC3      | HCT116 | -6.165867e-01 |    0.6974684 | -0.8840353350 | 0.3766771386 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType beta          sd          \n",
       "1       chr1 100034595 100034763 AGL        HCT116     0.09668708  3.024730e-01\n",
       "2       chr1 100034595 100034763 CDC14A     HCT116     0.18355315  3.665083e-01\n",
       "3       chr1 100034595 100034763 DBT        HCT116     0.08445645  3.803844e-01\n",
       "4       chr1 100034595 100034763 DPH5       HCT116    -0.46624892  4.838887e-01\n",
       "5       chr1 100034595 100034763 EXTL2      HCT116    -0.21254782  3.793808e-01\n",
       "6       chr1 100034595 100034763 FRRS1      HCT116    -0.55990653  5.504667e-01\n",
       "7       chr1 100034595 100034763 MFSD14A    HCT116     0.29025694  2.799339e-01\n",
       "8       chr1 100034595 100034763 PLPPR4     HCT116   -29.82663007  2.014787e+03\n",
       "9       chr1 100034595 100034763 RTCA       HCT116     0.43047533  3.061686e-01\n",
       "10      chr1 100034595 100034763 RTCA-AS1   HCT116   -30.06039198  1.934728e+03\n",
       "11      chr1 100034595 100034763 SASS6      HCT116    -0.13588317  3.712150e-01\n",
       "12      chr1 100034595 100034763 SLC30A7    HCT116    -0.82078821  3.374615e-01\n",
       "13      chr1 100034595 100034763 SLC35A3    HCT116     0.15579472  2.742344e-01\n",
       "14      chr1 100034595 100034763 TRMT13     HCT116     0.32210966  3.403561e-01\n",
       "15      chr1 100037727 100038876 AGL        HCT116     0.09376110  8.227838e-02\n",
       "16      chr1 100037727 100038876 CDC14A     HCT116     0.05771182  1.061016e-01\n",
       "17      chr1 100037727 100038876 DBT        HCT116     0.09744878  1.029016e-01\n",
       "18      chr1 100037727 100038876 DPH5       HCT116    -0.02599482  1.021325e-01\n",
       "19      chr1 100037727 100038876 EXTL2      HCT116     0.05226019  9.006044e-02\n",
       "20      chr1 100037727 100038876 FRRS1      HCT116    -0.06148027  1.105925e-01\n",
       "21      chr1 100037727 100038876 MFSD14A    HCT116    -0.01203135  8.579972e-02\n",
       "22      chr1 100037727 100038876 PLPPR4     HCT116     0.05074201  3.795120e-01\n",
       "23      chr1 100037727 100038876 RTCA       HCT116    -0.03646012  1.034141e-01\n",
       "24      chr1 100037727 100038876 RTCA-AS1   HCT116    -0.64489886  3.189826e-01\n",
       "25      chr1 100037727 100038876 SASS6      HCT116     0.15471803  8.817360e-02\n",
       "26      chr1 100037727 100038876 SLC30A7    HCT116    -0.04060774  6.536440e-02\n",
       "27      chr1 100037727 100038876 SLC35A3    HCT116     0.04266143  7.740949e-02\n",
       "28      chr1 100037727 100038876 TRMT13     HCT116     0.02646403  1.061844e-01\n",
       "29      chr1 100046025 100046641 AGL        HCT116    -0.02388234  2.915017e-01\n",
       "30      chr1 100046025 100046641 CDC14A     HCT116    -0.57411922  5.072311e-01\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮           \n",
       "2843303 chrX  989884    990608   ASMTL      HCT116    1.369647e-04    0.2068302\n",
       "2843304 chrX  989884    990608   CSF2RA     HCT116   -1.353126e-01    0.3313158\n",
       "2843305 chrX  989884    990608   GTPBP6     HCT116   -3.269883e-01    0.3686049\n",
       "2843306 chrX  989884    990608   IL3RA      HCT116    6.482838e-01    0.3881852\n",
       "2843307 chrX  989884    990608   LINC00106  HCT116    2.984671e-01    0.2983762\n",
       "2843308 chrX  989884    990608   P2RY8      HCT116   -3.376590e+01 1312.5040843\n",
       "2843309 chrX  989884    990608   PLCXD1     HCT116   -5.126421e-03    0.3320241\n",
       "2843310 chrX  989884    990608   PPP2R3B    HCT116    2.292070e-01    0.3457275\n",
       "2843311 chrX  989884    990608   SHOX       HCT116   -1.189372e+01  494.7870814\n",
       "2843312 chrX  989884    990608   SLC25A6    HCT116    2.514408e-02    0.1136494\n",
       "2843313 chrX 9902739   9902992   CLCN4      HCT116   -4.507918e-01    1.9075254\n",
       "2843314 chrX 9902739   9902992   FAM9B      HCT116   -1.515349e+01 1367.9308804\n",
       "2843315 chrX 9902739   9902992   TBL1X      HCT116   -2.656844e-01    0.2852540\n",
       "2843316 chrX 9902739   9902992   WWC3       HCT116    2.532484e-01    0.6805315\n",
       "2843317 chrX 9903251   9903473   CLCN4      HCT116   -4.197738e+01 4336.9719362\n",
       "2843318 chrX 9903251   9903473   FAM9B      HCT116   -1.473992e+01 1367.0718523\n",
       "2843319 chrX 9903251   9903473   TBL1X      HCT116   -4.272078e-01    0.3404950\n",
       "2843320 chrX 9903251   9903473   WWC3       HCT116    1.121024e+00    0.4800354\n",
       "2843321 chrX 9906862   9908314   CLCN4      HCT116   -3.048051e-01    0.6802224\n",
       "2843322 chrX 9906862   9908314   FAM9B      HCT116   -1.501650e+01  502.0195436\n",
       "2843323 chrX 9906862   9908314   TBL1X      HCT116   -1.122924e-01    0.1079954\n",
       "2843324 chrX 9906862   9908314   WWC3       HCT116    3.019850e-01    0.2631925\n",
       "2843325 chrX 9911462   9911841   CLCN4      HCT116   -4.043799e+01 2629.6920671\n",
       "2843326 chrX 9911462   9911841   FAM9B      HCT116    2.594126e+00    0.7262598\n",
       "2843327 chrX 9911462   9911841   TBL1X      HCT116   -1.332225e-01    0.2291654\n",
       "2843328 chrX 9911462   9911841   WWC3       HCT116   -4.531728e+01 3771.5118932\n",
       "2843329 chrX 9995795   9996095   CLCN4      HCT116   -4.289508e+01 2629.3515084\n",
       "2843330 chrX 9995795   9996095   FAM9B      HCT116   -1.508641e+01  829.0816720\n",
       "2843331 chrX 9995795   9996095   TBL1X      HCT116   -4.665829e-02    0.1691657\n",
       "2843332 chrX 9995795   9996095   WWC3       HCT116   -6.165867e-01    0.6974684\n",
       "        z             p           \n",
       "1        0.31965524   0.74922970  \n",
       "2        0.50081583   0.61650075  \n",
       "3        0.22202923   0.82429113  \n",
       "4       -0.96354587   0.33527366  \n",
       "5       -0.56024934   0.57530937  \n",
       "6       -1.01714885   0.30908263  \n",
       "7        1.03687682   0.29979327  \n",
       "8       -0.01480386   0.98818866  \n",
       "9        1.40600753   0.15972189  \n",
       "10      -0.01553727   0.98760355  \n",
       "11      -0.36604980   0.71432791  \n",
       "12      -2.43224234   0.01500566  \n",
       "13       0.56810782   0.56996176  \n",
       "14       0.94639006   0.34394968  \n",
       "15       1.13955941   0.25446990  \n",
       "16       0.54392955   0.58648995  \n",
       "17       0.94700951   0.34363393  \n",
       "18      -0.25452044   0.79909351  \n",
       "19       0.58027910   0.56172642  \n",
       "20      -0.55591698   0.57826761  \n",
       "21      -0.14022599   0.88848143  \n",
       "22       0.13370332   0.89363718  \n",
       "23      -0.35256431   0.72441510  \n",
       "24      -2.02173692   0.04320354  \n",
       "25       1.75469786   0.07931100  \n",
       "26      -0.62125155   0.53443413  \n",
       "27       0.55111371   0.58155572  \n",
       "28       0.24922711   0.80318511  \n",
       "29      -0.08192866   0.93470345  \n",
       "30      -1.13186916   0.25768945  \n",
       "⋮       ⋮             ⋮           \n",
       "2843303  0.0006622086 0.9994716340\n",
       "2843304 -0.4084098425 0.6829728086\n",
       "2843305 -0.8870969807 0.3750266895\n",
       "2843306  1.6700373457 0.0949119749\n",
       "2843307  1.0003046380 0.3171631034\n",
       "2843308 -0.0257263215 0.9794756293\n",
       "2843309 -0.0154399055 0.9876812272\n",
       "2843310  0.6629700795 0.5073497170\n",
       "2843311 -0.0240380633 0.9808222473\n",
       "2843312  0.2212424297 0.8249036736\n",
       "2843313 -0.2363228176 0.8131821824\n",
       "2843314 -0.0110776688 0.9911614799\n",
       "2843315 -0.9313957692 0.3516488786\n",
       "2843316  0.3721333003 0.7097936039\n",
       "2843317 -0.0096789614 0.9922774267\n",
       "2843318 -0.0107821102 0.9913972875\n",
       "2843319 -1.2546667892 0.2095997467\n",
       "2843320  2.3352943569 0.0195280488\n",
       "2843321 -0.4480963178 0.6540836854\n",
       "2843322 -0.0299121859 0.9761370872\n",
       "2843323 -1.0397887226 0.2984380697\n",
       "2843324  1.1473920719 0.2512196163\n",
       "2843325 -0.0153774614 0.9877310445\n",
       "2843326  3.5718979047 0.0003544036\n",
       "2843327 -0.5813378407 0.5610127819\n",
       "2843328 -0.0120156794 0.9904131056\n",
       "2843329 -0.0163139374 0.9869839386\n",
       "2843330 -0.0181965255 0.9854820744\n",
       "2843331 -0.2758141327 0.7826908283\n",
       "2843332 -0.8840353350 0.3766771386"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = celltype\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"beta\",\n",
    "                         \"sd\",\n",
    "                         \"z\",\n",
    "                         \"p\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] data.table_1.15.2    dplyr_1.1.4          foreach_1.5.2       \n",
       " [4] GenomicRanges_1.54.1 GenomeInfoDb_1.38.1  IRanges_2.36.0      \n",
       " [7] S4Vectors_0.40.2     BiocGenerics_0.48.1  genomation_1.34.0   \n",
       "[10] Signac_1.13.0        Seurat_5.0.3         SeuratObject_5.0.1  \n",
       "[13] sp_2.1-4             SnapATAC_1.0.0       rhdf5_2.44.0        \n",
       "[16] Matrix_1.6-5        \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 BiocIO_1.10.0              \n",
       "  [5] pbdZMQ_0.3-11               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                polyclip_1.10-6            \n",
       "  [9] XML_3.99-0.16.1             fastDummies_1.7.3          \n",
       " [11] lifecycle_1.0.4             tcltk_4.3.3                \n",
       " [13] edgeR_3.42.4                doParallel_1.0.17          \n",
       " [15] vroom_1.6.5                 globals_0.16.3             \n",
       " [17] lattice_0.22-6              MASS_7.3-60                \n",
       " [19] magrittr_2.0.3              limma_3.56.2               \n",
       " [21] plotly_4.10.4               plotrix_3.8-4              \n",
       " [23] yaml_2.3.8                  plot3D_1.4.1               \n",
       " [25] httpuv_1.6.15               sctransform_0.4.1          \n",
       " [27] spam_2.10-0                 spatstat.sparse_3.0-3      \n",
       " [29] reticulate_1.36.1           cowplot_1.1.3              \n",
       " [31] pbapply_1.7-2               RColorBrewer_1.1-3         \n",
       " [33] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [35] Rtsne_0.17                  purrr_1.0.2                \n",
       " [37] RCurl_1.98-1.14             misc3d_0.9-1               \n",
       " [39] GenomeInfoDbData_1.2.11     ggrepel_0.9.5              \n",
       " [41] irlba_2.3.5.1               listenv_0.9.1              \n",
       " [43] spatstat.utils_3.0-4        goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             bigmemory_4.6.4            \n",
       " [47] spatstat.random_3.2-3       fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           DelayedArray_0.26.7        \n",
       " [51] leiden_0.4.3.1              codetools_0.2-20           \n",
       " [53] RcppRoll_0.3.0              tidyselect_1.2.1           \n",
       " [55] viridis_0.6.5               matrixStats_1.3.0          \n",
       " [57] base64enc_0.1-3             spatstat.explore_3.2-6     \n",
       " [59] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [61] progressr_0.14.0            ggridges_0.5.6             \n",
       " [63] survival_3.6-4              iterators_1.0.14           \n",
       " [65] tools_4.3.3                 snow_0.4-4                 \n",
       " [67] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [69] glue_1.7.0                  gridExtra_2.3              \n",
       " [71] MatrixGenerics_1.12.3       IRdisplay_1.1              \n",
       " [73] withr_3.0.0                 fastmap_1.1.1              \n",
       " [75] rhdf5filters_1.12.1         fansi_1.0.6                \n",
       " [77] digest_0.6.35               R6_2.5.1                   \n",
       " [79] mime_0.12                   seqPattern_1.34.0          \n",
       " [81] colorspace_2.1-0            scattermore_1.2            \n",
       " [83] tensor_1.5                  spatstat.data_3.0-4        \n",
       " [85] utf8_1.2.4                  tidyr_1.3.1                \n",
       " [87] generics_0.1.3              rtracklayer_1.60.1         \n",
       " [89] S4Arrays_1.2.1              httr_1.4.7                 \n",
       " [91] htmlwidgets_1.6.4           uwot_0.1.16                \n",
       " [93] pkgconfig_2.0.3             gtable_0.3.5               \n",
       " [95] impute_1.76.0               lmtest_0.9-40              \n",
       " [97] XVector_0.42.0              htmltools_0.5.8.1          \n",
       " [99] dotCall64_1.1-1             Biobase_2.60.0             \n",
       "[101] scales_1.3.0                png_0.1-8                  \n",
       "[103] doSNOW_1.0.20               bigmemory.sri_0.1.8        \n",
       "[105] tzdb_0.4.0                  rjson_0.2.21               \n",
       "[107] reshape2_1.4.4              uuid_1.2-0                 \n",
       "[109] nlme_3.1-164                repr_1.1.7                 \n",
       "[111] zoo_1.8-12                  stringr_1.5.1              \n",
       "[113] KernSmooth_2.23-22          miniUI_0.1.1.1             \n",
       "[115] restfulr_0.0.15             pillar_1.9.0               \n",
       "[117] vctrs_0.6.5                 RANN_2.6.1                 \n",
       "[119] promises_1.3.0              xtable_1.8-4               \n",
       "[121] cluster_2.1.6               evaluate_0.23              \n",
       "[123] readr_2.1.5                 cli_3.6.2                  \n",
       "[125] locfit_1.5-9.9              compiler_4.3.3             \n",
       "[127] Rsamtools_2.18.0            rlang_1.1.3                \n",
       "[129] crayon_1.5.2                future.apply_1.11.2        \n",
       "[131] plyr_1.8.9                  stringi_1.8.3              \n",
       "[133] gridBase_0.4-7              viridisLite_0.4.2          \n",
       "[135] deldir_2.0-4                BiocParallel_1.36.0        \n",
       "[137] munsell_0.5.1               Biostrings_2.70.1          \n",
       "[139] lazyeval_0.2.2              spatstat.geom_3.2-9        \n",
       "[141] IRkernel_1.3.2              BSgenome_1.68.0            \n",
       "[143] RcppHNSW_0.6.0              hms_1.1.3                  \n",
       "[145] patchwork_1.2.0             bit64_4.0.5                \n",
       "[147] future_1.33.2               ggplot2_3.5.1              \n",
       "[149] Rhdf5lib_1.22.1             shiny_1.8.1.1              \n",
       "[151] SummarizedExperiment_1.30.2 ROCR_1.0-11                \n",
       "[153] igraph_2.0.3                fastmatch_1.1-4            \n",
       "[155] bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
