{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Wang/K562/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Wang/K562/Predictions/EnhancerPredictionsAllPutative.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Wang/K562/Kendall/atac_matrix.csv.gz\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Wang/1.prepare_data/1.seurat_pipeline.240507/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Wang/3.Genome_wide_prediction/FigR/FigR.250211/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2813245 ranges and 3 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100028738-100029317      * |         AGL\n",
       "        [2]     chr1 100028738-100029317      * |      CDC14A\n",
       "        [3]     chr1 100028738-100029317      * |         DBT\n",
       "        [4]     chr1 100028738-100029317      * |        DPH5\n",
       "        [5]     chr1 100028738-100029317      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2813241]     chrX     9940445-9940756      * |       TBL1X\n",
       "  [2813242]     chrX     9940445-9940756      * |        WWC3\n",
       "  [2813243]     chrX     9995833-9996097      * |       FAM9B\n",
       "  [2813244]     chrX     9995833-9996097      * |       TBL1X\n",
       "  [2813245]     chrX     9995833-9996097      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-100028738-10002.. chr1-100028738-10002..\n",
       "        [2] chr1-100028738-10002.. chr1-100028738-10002..\n",
       "        [3] chr1-100028738-10002.. chr1-100028738-10002..\n",
       "        [4] chr1-100028738-10002.. chr1-100028738-10002..\n",
       "        [5] chr1-100028738-10002.. chr1-100028738-10002..\n",
       "        ...                    ...                    ...\n",
       "  [2813241]   chrX-9940445-9940756 chrX-9940445-9940756..\n",
       "  [2813242]   chrX-9940445-9940756 chrX-9940445-9940756..\n",
       "  [2813243]   chrX-9995833-9996097 chrX-9995833-9996097..\n",
       "  [2813244]   chrX-9995833-9996097 chrX-9995833-9996097..\n",
       "  [2813245]   chrX-9995833-9996097 chrX-9995833-9996097..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  4000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  4001  to  4574 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = read.csv(path.matrix.atac_count,\n",
    "                             row.names = 1,\n",
    "                             check.names = F)\n",
    "matrix.atac_count = Matrix(as.matrix(matrix.atac_count), sparse = TRUE)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dd80eda-2392-4d59-8250-1ae62d7e1d1e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b9e26c70-11fd-47c4-8e05-9c96b17d3f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a76b1603-6b00-4ac0-9399-fd341c5a7b88",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 153715 4574 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(153715): chr1-100028738-100029317 chr1-100029410-100029652 ...\n",
       "  chrX-9940445-9940756 chrX-9995833-9996097\n",
       "rowData names(2): PeakName bias\n",
       "colnames(4574): K562_Wang_AAACAGCCAATGCCCG-1\n",
       "  K562_Wang_AAACAGCCACCCTCAC-1 ... K562_Wang_TCCATATTCAAACCTA-1\n",
       "  K562_Wang_TGACTTCGTGGATTAT-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna.rename),\n",
    "                       index = 1:nrow(matrix.rna.rename))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  35.7120168209076 secs \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Time Elapsed:  59.856890753905 mins \n",
      "\n",
      "[1] \"chr10\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.6048517227173 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  18.0268725236257 mins \n",
      "\n",
      "[1] \"chr11\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  18.6449263095856 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  33.5491221547127 mins \n",
      "\n",
      "[1] \"chr12\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  17.0292477607727 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  26.9740383585294 mins \n",
      "\n",
      "[1] \"chr13\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.13601803779602 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  7.58141947587331 mins \n",
      "\n",
      "[1] \"chr14\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.10056185722351 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  14.3497748732567 mins \n",
      "\n",
      "[1] \"chr15\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  11.8666491508484 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  17.2385411659877 mins \n",
      "\n",
      "[1] \"chr16\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  16.3450484275818 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  33.3765254060427 mins \n",
      "\n",
      "[1] \"chr17\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  17.4515886306763 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  29.8326275030772 mins \n",
      "\n",
      "[1] \"chr18\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.35673308372498 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  6.38108371496201 mins \n",
      "\n",
      "[1] \"chr19\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  28.4481761455536 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  47.8527673721313 mins \n",
      "\n",
      "[1] \"chr2\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  18.6223936080933 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  27.2411881446838 mins \n",
      "\n",
      "[1] \"chr20\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  11.3508789539337 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  18.6554848631223 mins \n",
      "\n",
      "[1] \"chr21\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.18722629547119 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.35894639492035 mins \n",
      "\n",
      "[1] \"chr22\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  11.6145324707031 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  19.8030316034953 mins \n",
      "\n",
      "[1] \"chr3\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  15.1599161624908 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  25.9534785469373 mins \n",
      "\n",
      "[1] \"chr4\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.72961831092834 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  15.6418761372566 mins \n",
      "\n",
      "[1] \"chr5\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  13.5911931991577 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  21.1439281105995 mins \n",
      "\n",
      "[1] \"chr6\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  22.4325635433197 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  34.357839179039 mins \n",
      "\n",
      "[1] \"chr7\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  16.9589097499847 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  29.417960357666 mins \n",
      "\n",
      "[1] \"chr8\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  9.70329904556274 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  17.8925644437472 mins \n",
      "\n",
      "[1] \"chr9\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  13.2766599655151 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  23.2331319014231 mins \n",
      "\n",
      "[1] \"chrX\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.297749042511 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  18.3292199850082 mins \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna.rename,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2735933 ranges and 7 metadata columns:\n",
       "       seqnames              ranges strand |  TargetGene               PeakName\n",
       "          <Rle>           <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 100028738-100029317      * |         AGL chr1-100028738-10002..\n",
       "  chr1     chr1 100028738-100029317      * |      CDC14A chr1-100028738-10002..\n",
       "  chr1     chr1 100028738-100029317      * |         DBT chr1-100028738-10002..\n",
       "  chr1     chr1 100028738-100029317      * |        DPH5 chr1-100028738-10002..\n",
       "  chr1     chr1 100028738-100029317      * |       EXTL2 chr1-100028738-10002..\n",
       "   ...      ...                 ...    ... .         ...                    ...\n",
       "  chrX     chrX     9940445-9940756      * |       TBL1X   chrX-9940445-9940756\n",
       "  chrX     chrX     9940445-9940756      * |        WWC3   chrX-9940445-9940756\n",
       "  chrX     chrX     9995833-9996097      * |       FAM9B   chrX-9995833-9996097\n",
       "  chrX     chrX     9995833-9996097      * |       TBL1X   chrX-9995833-9996097\n",
       "  chrX     chrX     9995833-9996097      * |        WWC3   chrX-9995833-9996097\n",
       "                     PairName        rObs     rBgSD      rBgMean     pvalZ\n",
       "                  <character>   <numeric> <numeric>    <numeric> <numeric>\n",
       "  chr1 chr1-100028738-10002..  0.00472059 0.0138516  0.004489001  0.493330\n",
       "  chr1 chr1-100028738-10002.. -0.00629845 0.0146085 -0.003737446  0.569582\n",
       "  chr1 chr1-100028738-10002.. -0.00871077 0.0151337  0.000345098  0.725211\n",
       "  chr1 chr1-100028738-10002.. -0.00330845 0.0142716  0.002383941  0.655003\n",
       "  chr1 chr1-100028738-10002.. -0.01065292 0.0138044  0.000315880  0.786573\n",
       "   ...                    ...         ...       ...          ...       ...\n",
       "  chrX chrX-9940445-9940756..  0.01560896 0.0148306  -0.00140242  0.125682\n",
       "  chrX chrX-9940445-9940756.. -0.01068949 0.0134039   0.00123612  0.813189\n",
       "  chrX chrX-9995833-9996097..  0.03426305 0.0204543   0.01139896  0.131824\n",
       "  chrX chrX-9995833-9996097..  0.00382596 0.0125007  -0.00533177  0.231907\n",
       "  chrX chrX-9995833-9996097.. -0.01667955 0.0166122   0.00384050  0.891629\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2735933 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>AGL     </td><td>K562</td><td> 0.0047205854</td><td>0.013851552</td><td> 4.489001e-03</td><td>0.493330384</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>CDC14A  </td><td>K562</td><td>-0.0062984471</td><td>0.014608498</td><td>-3.737446e-03</td><td>0.569581566</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>DBT     </td><td>K562</td><td>-0.0087107728</td><td>0.015133699</td><td> 3.450980e-04</td><td>0.725210502</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>DPH5    </td><td>K562</td><td>-0.0033084520</td><td>0.014271566</td><td> 2.383941e-03</td><td>0.655002762</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>EXTL2   </td><td>K562</td><td>-0.0106529164</td><td>0.013804385</td><td> 3.158801e-04</td><td>0.786573354</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>FRRS1   </td><td>K562</td><td>-0.0140837852</td><td>0.015243026</td><td>-1.132415e-03</td><td>0.802242576</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>LRRC39  </td><td>K562</td><td> 0.0148327044</td><td>0.014888913</td><td> 4.068826e-03</td><td>0.234856551</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>MFSD14A </td><td>K562</td><td> 0.0269955676</td><td>0.016018736</td><td> 7.509079e-03</td><td>0.111900860</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>PALMD   </td><td>K562</td><td> 0.0089904398</td><td>0.015091760</td><td> 8.810720e-04</td><td>0.295517267</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>PLPPR4  </td><td>K562</td><td> 0.0423076624</td><td>0.016506769</td><td> 6.633906e-04</td><td>0.005820232</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>RTCA    </td><td>K562</td><td> 0.0130209429</td><td>0.014316152</td><td> 1.939669e-03</td><td>0.219453562</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>RTCA-AS1</td><td>K562</td><td>-0.0002573075</td><td>0.015781668</td><td> 4.007851e-03</td><td>0.606520009</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>SASS6   </td><td>K562</td><td> 0.0105523314</td><td>0.016149575</td><td> 6.626603e-03</td><td>0.403969559</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>SLC30A7 </td><td>K562</td><td> 0.0005850693</td><td>0.014468586</td><td> 1.665634e-03</td><td>0.529766736</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>SLC35A3 </td><td>K562</td><td>-0.0089382474</td><td>0.014272966</td><td> 3.441832e-03</td><td>0.807132973</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>TRMT13  </td><td>K562</td><td>-0.0060118839</td><td>0.014258681</td><td> 3.737736e-03</td><td>0.752938944</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>AGL     </td><td>K562</td><td> 0.0068971276</td><td>0.013502943</td><td> 2.734003e-03</td><td>0.378922323</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>CDC14A  </td><td>K562</td><td> 0.0131722051</td><td>0.014264149</td><td>-6.118940e-04</td><td>0.166935613</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>DBT     </td><td>K562</td><td>-0.0199082864</td><td>0.013746739</td><td> 1.320443e-03</td><td>0.938738997</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>DPH5    </td><td>K562</td><td>-0.0002186223</td><td>0.014933580</td><td> 2.883092e-03</td><td>0.582268644</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>EXTL2   </td><td>K562</td><td> 0.0197419863</td><td>0.013685458</td><td>-1.274360e-04</td><td>0.073269728</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>FRRS1   </td><td>K562</td><td>-0.0119640277</td><td>0.015702522</td><td>-1.925265e-03</td><td>0.738689026</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>LRRC39  </td><td>K562</td><td> 0.0214076248</td><td>0.014745812</td><td>-9.756608e-04</td><td>0.064514528</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>MFSD14A </td><td>K562</td><td> 0.0111976403</td><td>0.015078275</td><td> 2.580706e-03</td><td>0.283837122</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>PALMD   </td><td>K562</td><td>-0.0084516115</td><td>0.015703538</td><td>-1.296950e-05</td><td>0.704494670</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>PLPPR4  </td><td>K562</td><td>-0.0025195385</td><td>0.007557136</td><td>-1.943826e-03</td><td>0.530362570</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>RTCA    </td><td>K562</td><td>-0.0050826516</td><td>0.014106587</td><td>-2.373059e-03</td><td>0.576160226</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>RTCA-AS1</td><td>K562</td><td> 0.0065758344</td><td>0.014222561</td><td>-3.244719e-07</td><td>0.321906160</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>SASS6   </td><td>K562</td><td>-0.0022461967</td><td>0.015147027</td><td> 2.417460e-03</td><td>0.620917941</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>SLC30A7 </td><td>K562</td><td> 0.0082130824</td><td>0.015050004</td><td>-9.715164e-04</td><td>0.270840767</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9864063</td><td>9864529</td><td>TBL1X    </td><td>K562</td><td>-0.0040138347</td><td>0.01373375</td><td>-0.005583988</td><td>0.454488866</td></tr>\n",
       "\t<tr><td>chrX</td><td>9864063</td><td>9864529</td><td>WWC3     </td><td>K562</td><td> 0.0246353382</td><td>0.01610863</td><td> 0.006245860</td><td>0.126811819</td></tr>\n",
       "\t<tr><td>chrX</td><td>9874033</td><td>9874283</td><td>FAM9B    </td><td>K562</td><td> 0.0224928812</td><td>0.01611845</td><td> 0.007247343</td><td>0.172114044</td></tr>\n",
       "\t<tr><td>chrX</td><td>9874033</td><td>9874283</td><td>TBL1X    </td><td>K562</td><td>-0.0007651788</td><td>0.01413708</td><td>-0.002532682</td><td>0.450251463</td></tr>\n",
       "\t<tr><td>chrX</td><td>9874033</td><td>9874283</td><td>WWC3     </td><td>K562</td><td> 0.0023039813</td><td>0.01557011</td><td> 0.003161885</td><td>0.521970360</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882684</td><td>9883146</td><td>FAM9B    </td><td>K562</td><td>-0.0056623219</td><td>0.01864811</td><td> 0.013727961</td><td>0.850783289</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882684</td><td>9883146</td><td>TBL1X    </td><td>K562</td><td> 0.0005373820</td><td>0.01633870</td><td>-0.005059875</td><td>0.365958456</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882684</td><td>9883146</td><td>WWC3     </td><td>K562</td><td> 0.0269535323</td><td>0.01542460</td><td> 0.004478402</td><td>0.072544870</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885199</td><td>9885463</td><td>FAM9B    </td><td>K562</td><td> 0.0026406256</td><td>0.01729513</td><td> 0.010052671</td><td>0.665879192</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885199</td><td>9885463</td><td>TBL1X    </td><td>K562</td><td>-0.0049799836</td><td>0.01769587</td><td>-0.005119067</td><td>0.496864475</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885199</td><td>9885463</td><td>WWC3     </td><td>K562</td><td> 0.0256010058</td><td>0.01554113</td><td> 0.004344306</td><td>0.085691902</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894498</td><td>9895136</td><td>FAM9B    </td><td>K562</td><td>-0.0083305906</td><td>0.02289624</td><td> 0.010907225</td><td>0.799606736</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894498</td><td>9895136</td><td>TBL1X    </td><td>K562</td><td>-0.0061796709</td><td>0.01524818</td><td>-0.004621935</td><td>0.540684688</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894498</td><td>9895136</td><td>WWC3     </td><td>K562</td><td>-0.0015254686</td><td>0.01389721</td><td> 0.002020488</td><td>0.600698680</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>AKAP17A  </td><td>K562</td><td> 0.0438793145</td><td>0.01350371</td><td> 0.012114167</td><td>0.009328157</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>ASMT     </td><td>K562</td><td> 0.0409807849</td><td>0.01575378</td><td> 0.006177742</td><td>0.013580804</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>ASMTL    </td><td>K562</td><td> 0.0231346780</td><td>0.01430504</td><td> 0.005537010</td><td>0.109316197</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>GTPBP6   </td><td>K562</td><td> 0.0192363155</td><td>0.01517947</td><td> 0.013268890</td><td>0.347113629</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>IL3RA    </td><td>K562</td><td>-0.0068971428</td><td>0.01332706</td><td>-0.004542858</td><td>0.570110120</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>LINC00106</td><td>K562</td><td> 0.0009951114</td><td>0.01511970</td><td> 0.009148941</td><td>0.705154900</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>P2RY8    </td><td>K562</td><td> 0.0223772932</td><td>0.01751157</td><td> 0.002964413</td><td>0.133806834</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>PLCXD1   </td><td>K562</td><td> 0.0264926047</td><td>0.01474863</td><td> 0.010665462</td><td>0.141607203</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>PPP2R3B  </td><td>K562</td><td> 0.0116986221</td><td>0.01697001</td><td> 0.024904988</td><td>0.781779792</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>SLC25A6  </td><td>K562</td><td> 0.0250474347</td><td>0.01647551</td><td> 0.019002397</td><td>0.356842873</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940445</td><td>9940756</td><td>FAM9B    </td><td>K562</td><td>-0.0132319729</td><td>0.02016957</td><td> 0.006904864</td><td>0.840951768</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940445</td><td>9940756</td><td>TBL1X    </td><td>K562</td><td> 0.0156089585</td><td>0.01483061</td><td>-0.001402415</td><td>0.125681629</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940445</td><td>9940756</td><td>WWC3     </td><td>K562</td><td>-0.0106894901</td><td>0.01340392</td><td> 0.001236118</td><td>0.813189329</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995833</td><td>9996097</td><td>FAM9B    </td><td>K562</td><td> 0.0342630502</td><td>0.02045432</td><td> 0.011398960</td><td>0.131823541</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995833</td><td>9996097</td><td>TBL1X    </td><td>K562</td><td> 0.0038259623</td><td>0.01250067</td><td>-0.005331775</td><td>0.231907431</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995833</td><td>9996097</td><td>WWC3     </td><td>K562</td><td>-0.0166795494</td><td>0.01661220</td><td> 0.003840498</td><td>0.891629325</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2735933 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100028738 & 100029317 & AGL      & K562 &  0.0047205854 & 0.013851552 &  4.489001e-03 & 0.493330384\\\\\n",
       "\t chr1 & 100028738 & 100029317 & CDC14A   & K562 & -0.0062984471 & 0.014608498 & -3.737446e-03 & 0.569581566\\\\\n",
       "\t chr1 & 100028738 & 100029317 & DBT      & K562 & -0.0087107728 & 0.015133699 &  3.450980e-04 & 0.725210502\\\\\n",
       "\t chr1 & 100028738 & 100029317 & DPH5     & K562 & -0.0033084520 & 0.014271566 &  2.383941e-03 & 0.655002762\\\\\n",
       "\t chr1 & 100028738 & 100029317 & EXTL2    & K562 & -0.0106529164 & 0.013804385 &  3.158801e-04 & 0.786573354\\\\\n",
       "\t chr1 & 100028738 & 100029317 & FRRS1    & K562 & -0.0140837852 & 0.015243026 & -1.132415e-03 & 0.802242576\\\\\n",
       "\t chr1 & 100028738 & 100029317 & LRRC39   & K562 &  0.0148327044 & 0.014888913 &  4.068826e-03 & 0.234856551\\\\\n",
       "\t chr1 & 100028738 & 100029317 & MFSD14A  & K562 &  0.0269955676 & 0.016018736 &  7.509079e-03 & 0.111900860\\\\\n",
       "\t chr1 & 100028738 & 100029317 & PALMD    & K562 &  0.0089904398 & 0.015091760 &  8.810720e-04 & 0.295517267\\\\\n",
       "\t chr1 & 100028738 & 100029317 & PLPPR4   & K562 &  0.0423076624 & 0.016506769 &  6.633906e-04 & 0.005820232\\\\\n",
       "\t chr1 & 100028738 & 100029317 & RTCA     & K562 &  0.0130209429 & 0.014316152 &  1.939669e-03 & 0.219453562\\\\\n",
       "\t chr1 & 100028738 & 100029317 & RTCA-AS1 & K562 & -0.0002573075 & 0.015781668 &  4.007851e-03 & 0.606520009\\\\\n",
       "\t chr1 & 100028738 & 100029317 & SASS6    & K562 &  0.0105523314 & 0.016149575 &  6.626603e-03 & 0.403969559\\\\\n",
       "\t chr1 & 100028738 & 100029317 & SLC30A7  & K562 &  0.0005850693 & 0.014468586 &  1.665634e-03 & 0.529766736\\\\\n",
       "\t chr1 & 100028738 & 100029317 & SLC35A3  & K562 & -0.0089382474 & 0.014272966 &  3.441832e-03 & 0.807132973\\\\\n",
       "\t chr1 & 100028738 & 100029317 & TRMT13   & K562 & -0.0060118839 & 0.014258681 &  3.737736e-03 & 0.752938944\\\\\n",
       "\t chr1 & 100029410 & 100029652 & AGL      & K562 &  0.0068971276 & 0.013502943 &  2.734003e-03 & 0.378922323\\\\\n",
       "\t chr1 & 100029410 & 100029652 & CDC14A   & K562 &  0.0131722051 & 0.014264149 & -6.118940e-04 & 0.166935613\\\\\n",
       "\t chr1 & 100029410 & 100029652 & DBT      & K562 & -0.0199082864 & 0.013746739 &  1.320443e-03 & 0.938738997\\\\\n",
       "\t chr1 & 100029410 & 100029652 & DPH5     & K562 & -0.0002186223 & 0.014933580 &  2.883092e-03 & 0.582268644\\\\\n",
       "\t chr1 & 100029410 & 100029652 & EXTL2    & K562 &  0.0197419863 & 0.013685458 & -1.274360e-04 & 0.073269728\\\\\n",
       "\t chr1 & 100029410 & 100029652 & FRRS1    & K562 & -0.0119640277 & 0.015702522 & -1.925265e-03 & 0.738689026\\\\\n",
       "\t chr1 & 100029410 & 100029652 & LRRC39   & K562 &  0.0214076248 & 0.014745812 & -9.756608e-04 & 0.064514528\\\\\n",
       "\t chr1 & 100029410 & 100029652 & MFSD14A  & K562 &  0.0111976403 & 0.015078275 &  2.580706e-03 & 0.283837122\\\\\n",
       "\t chr1 & 100029410 & 100029652 & PALMD    & K562 & -0.0084516115 & 0.015703538 & -1.296950e-05 & 0.704494670\\\\\n",
       "\t chr1 & 100029410 & 100029652 & PLPPR4   & K562 & -0.0025195385 & 0.007557136 & -1.943826e-03 & 0.530362570\\\\\n",
       "\t chr1 & 100029410 & 100029652 & RTCA     & K562 & -0.0050826516 & 0.014106587 & -2.373059e-03 & 0.576160226\\\\\n",
       "\t chr1 & 100029410 & 100029652 & RTCA-AS1 & K562 &  0.0065758344 & 0.014222561 & -3.244719e-07 & 0.321906160\\\\\n",
       "\t chr1 & 100029410 & 100029652 & SASS6    & K562 & -0.0022461967 & 0.015147027 &  2.417460e-03 & 0.620917941\\\\\n",
       "\t chr1 & 100029410 & 100029652 & SLC30A7  & K562 &  0.0082130824 & 0.015050004 & -9.715164e-04 & 0.270840767\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9864063 & 9864529 & TBL1X     & K562 & -0.0040138347 & 0.01373375 & -0.005583988 & 0.454488866\\\\\n",
       "\t chrX & 9864063 & 9864529 & WWC3      & K562 &  0.0246353382 & 0.01610863 &  0.006245860 & 0.126811819\\\\\n",
       "\t chrX & 9874033 & 9874283 & FAM9B     & K562 &  0.0224928812 & 0.01611845 &  0.007247343 & 0.172114044\\\\\n",
       "\t chrX & 9874033 & 9874283 & TBL1X     & K562 & -0.0007651788 & 0.01413708 & -0.002532682 & 0.450251463\\\\\n",
       "\t chrX & 9874033 & 9874283 & WWC3      & K562 &  0.0023039813 & 0.01557011 &  0.003161885 & 0.521970360\\\\\n",
       "\t chrX & 9882684 & 9883146 & FAM9B     & K562 & -0.0056623219 & 0.01864811 &  0.013727961 & 0.850783289\\\\\n",
       "\t chrX & 9882684 & 9883146 & TBL1X     & K562 &  0.0005373820 & 0.01633870 & -0.005059875 & 0.365958456\\\\\n",
       "\t chrX & 9882684 & 9883146 & WWC3      & K562 &  0.0269535323 & 0.01542460 &  0.004478402 & 0.072544870\\\\\n",
       "\t chrX & 9885199 & 9885463 & FAM9B     & K562 &  0.0026406256 & 0.01729513 &  0.010052671 & 0.665879192\\\\\n",
       "\t chrX & 9885199 & 9885463 & TBL1X     & K562 & -0.0049799836 & 0.01769587 & -0.005119067 & 0.496864475\\\\\n",
       "\t chrX & 9885199 & 9885463 & WWC3      & K562 &  0.0256010058 & 0.01554113 &  0.004344306 & 0.085691902\\\\\n",
       "\t chrX & 9894498 & 9895136 & FAM9B     & K562 & -0.0083305906 & 0.02289624 &  0.010907225 & 0.799606736\\\\\n",
       "\t chrX & 9894498 & 9895136 & TBL1X     & K562 & -0.0061796709 & 0.01524818 & -0.004621935 & 0.540684688\\\\\n",
       "\t chrX & 9894498 & 9895136 & WWC3      & K562 & -0.0015254686 & 0.01389721 &  0.002020488 & 0.600698680\\\\\n",
       "\t chrX &  989856 &  990611 & AKAP17A   & K562 &  0.0438793145 & 0.01350371 &  0.012114167 & 0.009328157\\\\\n",
       "\t chrX &  989856 &  990611 & ASMT      & K562 &  0.0409807849 & 0.01575378 &  0.006177742 & 0.013580804\\\\\n",
       "\t chrX &  989856 &  990611 & ASMTL     & K562 &  0.0231346780 & 0.01430504 &  0.005537010 & 0.109316197\\\\\n",
       "\t chrX &  989856 &  990611 & GTPBP6    & K562 &  0.0192363155 & 0.01517947 &  0.013268890 & 0.347113629\\\\\n",
       "\t chrX &  989856 &  990611 & IL3RA     & K562 & -0.0068971428 & 0.01332706 & -0.004542858 & 0.570110120\\\\\n",
       "\t chrX &  989856 &  990611 & LINC00106 & K562 &  0.0009951114 & 0.01511970 &  0.009148941 & 0.705154900\\\\\n",
       "\t chrX &  989856 &  990611 & P2RY8     & K562 &  0.0223772932 & 0.01751157 &  0.002964413 & 0.133806834\\\\\n",
       "\t chrX &  989856 &  990611 & PLCXD1    & K562 &  0.0264926047 & 0.01474863 &  0.010665462 & 0.141607203\\\\\n",
       "\t chrX &  989856 &  990611 & PPP2R3B   & K562 &  0.0116986221 & 0.01697001 &  0.024904988 & 0.781779792\\\\\n",
       "\t chrX &  989856 &  990611 & SLC25A6   & K562 &  0.0250474347 & 0.01647551 &  0.019002397 & 0.356842873\\\\\n",
       "\t chrX & 9940445 & 9940756 & FAM9B     & K562 & -0.0132319729 & 0.02016957 &  0.006904864 & 0.840951768\\\\\n",
       "\t chrX & 9940445 & 9940756 & TBL1X     & K562 &  0.0156089585 & 0.01483061 & -0.001402415 & 0.125681629\\\\\n",
       "\t chrX & 9940445 & 9940756 & WWC3      & K562 & -0.0106894901 & 0.01340392 &  0.001236118 & 0.813189329\\\\\n",
       "\t chrX & 9995833 & 9996097 & FAM9B     & K562 &  0.0342630502 & 0.02045432 &  0.011398960 & 0.131823541\\\\\n",
       "\t chrX & 9995833 & 9996097 & TBL1X     & K562 &  0.0038259623 & 0.01250067 & -0.005331775 & 0.231907431\\\\\n",
       "\t chrX & 9995833 & 9996097 & WWC3      & K562 & -0.0166795494 & 0.01661220 &  0.003840498 & 0.891629325\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2735933 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100028738 | 100029317 | AGL      | K562 |  0.0047205854 | 0.013851552 |  4.489001e-03 | 0.493330384 |\n",
       "| chr1 | 100028738 | 100029317 | CDC14A   | K562 | -0.0062984471 | 0.014608498 | -3.737446e-03 | 0.569581566 |\n",
       "| chr1 | 100028738 | 100029317 | DBT      | K562 | -0.0087107728 | 0.015133699 |  3.450980e-04 | 0.725210502 |\n",
       "| chr1 | 100028738 | 100029317 | DPH5     | K562 | -0.0033084520 | 0.014271566 |  2.383941e-03 | 0.655002762 |\n",
       "| chr1 | 100028738 | 100029317 | EXTL2    | K562 | -0.0106529164 | 0.013804385 |  3.158801e-04 | 0.786573354 |\n",
       "| chr1 | 100028738 | 100029317 | FRRS1    | K562 | -0.0140837852 | 0.015243026 | -1.132415e-03 | 0.802242576 |\n",
       "| chr1 | 100028738 | 100029317 | LRRC39   | K562 |  0.0148327044 | 0.014888913 |  4.068826e-03 | 0.234856551 |\n",
       "| chr1 | 100028738 | 100029317 | MFSD14A  | K562 |  0.0269955676 | 0.016018736 |  7.509079e-03 | 0.111900860 |\n",
       "| chr1 | 100028738 | 100029317 | PALMD    | K562 |  0.0089904398 | 0.015091760 |  8.810720e-04 | 0.295517267 |\n",
       "| chr1 | 100028738 | 100029317 | PLPPR4   | K562 |  0.0423076624 | 0.016506769 |  6.633906e-04 | 0.005820232 |\n",
       "| chr1 | 100028738 | 100029317 | RTCA     | K562 |  0.0130209429 | 0.014316152 |  1.939669e-03 | 0.219453562 |\n",
       "| chr1 | 100028738 | 100029317 | RTCA-AS1 | K562 | -0.0002573075 | 0.015781668 |  4.007851e-03 | 0.606520009 |\n",
       "| chr1 | 100028738 | 100029317 | SASS6    | K562 |  0.0105523314 | 0.016149575 |  6.626603e-03 | 0.403969559 |\n",
       "| chr1 | 100028738 | 100029317 | SLC30A7  | K562 |  0.0005850693 | 0.014468586 |  1.665634e-03 | 0.529766736 |\n",
       "| chr1 | 100028738 | 100029317 | SLC35A3  | K562 | -0.0089382474 | 0.014272966 |  3.441832e-03 | 0.807132973 |\n",
       "| chr1 | 100028738 | 100029317 | TRMT13   | K562 | -0.0060118839 | 0.014258681 |  3.737736e-03 | 0.752938944 |\n",
       "| chr1 | 100029410 | 100029652 | AGL      | K562 |  0.0068971276 | 0.013502943 |  2.734003e-03 | 0.378922323 |\n",
       "| chr1 | 100029410 | 100029652 | CDC14A   | K562 |  0.0131722051 | 0.014264149 | -6.118940e-04 | 0.166935613 |\n",
       "| chr1 | 100029410 | 100029652 | DBT      | K562 | -0.0199082864 | 0.013746739 |  1.320443e-03 | 0.938738997 |\n",
       "| chr1 | 100029410 | 100029652 | DPH5     | K562 | -0.0002186223 | 0.014933580 |  2.883092e-03 | 0.582268644 |\n",
       "| chr1 | 100029410 | 100029652 | EXTL2    | K562 |  0.0197419863 | 0.013685458 | -1.274360e-04 | 0.073269728 |\n",
       "| chr1 | 100029410 | 100029652 | FRRS1    | K562 | -0.0119640277 | 0.015702522 | -1.925265e-03 | 0.738689026 |\n",
       "| chr1 | 100029410 | 100029652 | LRRC39   | K562 |  0.0214076248 | 0.014745812 | -9.756608e-04 | 0.064514528 |\n",
       "| chr1 | 100029410 | 100029652 | MFSD14A  | K562 |  0.0111976403 | 0.015078275 |  2.580706e-03 | 0.283837122 |\n",
       "| chr1 | 100029410 | 100029652 | PALMD    | K562 | -0.0084516115 | 0.015703538 | -1.296950e-05 | 0.704494670 |\n",
       "| chr1 | 100029410 | 100029652 | PLPPR4   | K562 | -0.0025195385 | 0.007557136 | -1.943826e-03 | 0.530362570 |\n",
       "| chr1 | 100029410 | 100029652 | RTCA     | K562 | -0.0050826516 | 0.014106587 | -2.373059e-03 | 0.576160226 |\n",
       "| chr1 | 100029410 | 100029652 | RTCA-AS1 | K562 |  0.0065758344 | 0.014222561 | -3.244719e-07 | 0.321906160 |\n",
       "| chr1 | 100029410 | 100029652 | SASS6    | K562 | -0.0022461967 | 0.015147027 |  2.417460e-03 | 0.620917941 |\n",
       "| chr1 | 100029410 | 100029652 | SLC30A7  | K562 |  0.0082130824 | 0.015050004 | -9.715164e-04 | 0.270840767 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9864063 | 9864529 | TBL1X     | K562 | -0.0040138347 | 0.01373375 | -0.005583988 | 0.454488866 |\n",
       "| chrX | 9864063 | 9864529 | WWC3      | K562 |  0.0246353382 | 0.01610863 |  0.006245860 | 0.126811819 |\n",
       "| chrX | 9874033 | 9874283 | FAM9B     | K562 |  0.0224928812 | 0.01611845 |  0.007247343 | 0.172114044 |\n",
       "| chrX | 9874033 | 9874283 | TBL1X     | K562 | -0.0007651788 | 0.01413708 | -0.002532682 | 0.450251463 |\n",
       "| chrX | 9874033 | 9874283 | WWC3      | K562 |  0.0023039813 | 0.01557011 |  0.003161885 | 0.521970360 |\n",
       "| chrX | 9882684 | 9883146 | FAM9B     | K562 | -0.0056623219 | 0.01864811 |  0.013727961 | 0.850783289 |\n",
       "| chrX | 9882684 | 9883146 | TBL1X     | K562 |  0.0005373820 | 0.01633870 | -0.005059875 | 0.365958456 |\n",
       "| chrX | 9882684 | 9883146 | WWC3      | K562 |  0.0269535323 | 0.01542460 |  0.004478402 | 0.072544870 |\n",
       "| chrX | 9885199 | 9885463 | FAM9B     | K562 |  0.0026406256 | 0.01729513 |  0.010052671 | 0.665879192 |\n",
       "| chrX | 9885199 | 9885463 | TBL1X     | K562 | -0.0049799836 | 0.01769587 | -0.005119067 | 0.496864475 |\n",
       "| chrX | 9885199 | 9885463 | WWC3      | K562 |  0.0256010058 | 0.01554113 |  0.004344306 | 0.085691902 |\n",
       "| chrX | 9894498 | 9895136 | FAM9B     | K562 | -0.0083305906 | 0.02289624 |  0.010907225 | 0.799606736 |\n",
       "| chrX | 9894498 | 9895136 | TBL1X     | K562 | -0.0061796709 | 0.01524818 | -0.004621935 | 0.540684688 |\n",
       "| chrX | 9894498 | 9895136 | WWC3      | K562 | -0.0015254686 | 0.01389721 |  0.002020488 | 0.600698680 |\n",
       "| chrX |  989856 |  990611 | AKAP17A   | K562 |  0.0438793145 | 0.01350371 |  0.012114167 | 0.009328157 |\n",
       "| chrX |  989856 |  990611 | ASMT      | K562 |  0.0409807849 | 0.01575378 |  0.006177742 | 0.013580804 |\n",
       "| chrX |  989856 |  990611 | ASMTL     | K562 |  0.0231346780 | 0.01430504 |  0.005537010 | 0.109316197 |\n",
       "| chrX |  989856 |  990611 | GTPBP6    | K562 |  0.0192363155 | 0.01517947 |  0.013268890 | 0.347113629 |\n",
       "| chrX |  989856 |  990611 | IL3RA     | K562 | -0.0068971428 | 0.01332706 | -0.004542858 | 0.570110120 |\n",
       "| chrX |  989856 |  990611 | LINC00106 | K562 |  0.0009951114 | 0.01511970 |  0.009148941 | 0.705154900 |\n",
       "| chrX |  989856 |  990611 | P2RY8     | K562 |  0.0223772932 | 0.01751157 |  0.002964413 | 0.133806834 |\n",
       "| chrX |  989856 |  990611 | PLCXD1    | K562 |  0.0264926047 | 0.01474863 |  0.010665462 | 0.141607203 |\n",
       "| chrX |  989856 |  990611 | PPP2R3B   | K562 |  0.0116986221 | 0.01697001 |  0.024904988 | 0.781779792 |\n",
       "| chrX |  989856 |  990611 | SLC25A6   | K562 |  0.0250474347 | 0.01647551 |  0.019002397 | 0.356842873 |\n",
       "| chrX | 9940445 | 9940756 | FAM9B     | K562 | -0.0132319729 | 0.02016957 |  0.006904864 | 0.840951768 |\n",
       "| chrX | 9940445 | 9940756 | TBL1X     | K562 |  0.0156089585 | 0.01483061 | -0.001402415 | 0.125681629 |\n",
       "| chrX | 9940445 | 9940756 | WWC3      | K562 | -0.0106894901 | 0.01340392 |  0.001236118 | 0.813189329 |\n",
       "| chrX | 9995833 | 9996097 | FAM9B     | K562 |  0.0342630502 | 0.02045432 |  0.011398960 | 0.131823541 |\n",
       "| chrX | 9995833 | 9996097 | TBL1X     | K562 |  0.0038259623 | 0.01250067 | -0.005331775 | 0.231907431 |\n",
       "| chrX | 9995833 | 9996097 | WWC3      | K562 | -0.0166795494 | 0.01661220 |  0.003840498 | 0.891629325 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType rObs          rBgSD      \n",
       "1       chr1 100028738 100029317 AGL        K562      0.0047205854 0.013851552\n",
       "2       chr1 100028738 100029317 CDC14A     K562     -0.0062984471 0.014608498\n",
       "3       chr1 100028738 100029317 DBT        K562     -0.0087107728 0.015133699\n",
       "4       chr1 100028738 100029317 DPH5       K562     -0.0033084520 0.014271566\n",
       "5       chr1 100028738 100029317 EXTL2      K562     -0.0106529164 0.013804385\n",
       "6       chr1 100028738 100029317 FRRS1      K562     -0.0140837852 0.015243026\n",
       "7       chr1 100028738 100029317 LRRC39     K562      0.0148327044 0.014888913\n",
       "8       chr1 100028738 100029317 MFSD14A    K562      0.0269955676 0.016018736\n",
       "9       chr1 100028738 100029317 PALMD      K562      0.0089904398 0.015091760\n",
       "10      chr1 100028738 100029317 PLPPR4     K562      0.0423076624 0.016506769\n",
       "11      chr1 100028738 100029317 RTCA       K562      0.0130209429 0.014316152\n",
       "12      chr1 100028738 100029317 RTCA-AS1   K562     -0.0002573075 0.015781668\n",
       "13      chr1 100028738 100029317 SASS6      K562      0.0105523314 0.016149575\n",
       "14      chr1 100028738 100029317 SLC30A7    K562      0.0005850693 0.014468586\n",
       "15      chr1 100028738 100029317 SLC35A3    K562     -0.0089382474 0.014272966\n",
       "16      chr1 100028738 100029317 TRMT13     K562     -0.0060118839 0.014258681\n",
       "17      chr1 100029410 100029652 AGL        K562      0.0068971276 0.013502943\n",
       "18      chr1 100029410 100029652 CDC14A     K562      0.0131722051 0.014264149\n",
       "19      chr1 100029410 100029652 DBT        K562     -0.0199082864 0.013746739\n",
       "20      chr1 100029410 100029652 DPH5       K562     -0.0002186223 0.014933580\n",
       "21      chr1 100029410 100029652 EXTL2      K562      0.0197419863 0.013685458\n",
       "22      chr1 100029410 100029652 FRRS1      K562     -0.0119640277 0.015702522\n",
       "23      chr1 100029410 100029652 LRRC39     K562      0.0214076248 0.014745812\n",
       "24      chr1 100029410 100029652 MFSD14A    K562      0.0111976403 0.015078275\n",
       "25      chr1 100029410 100029652 PALMD      K562     -0.0084516115 0.015703538\n",
       "26      chr1 100029410 100029652 PLPPR4     K562     -0.0025195385 0.007557136\n",
       "27      chr1 100029410 100029652 RTCA       K562     -0.0050826516 0.014106587\n",
       "28      chr1 100029410 100029652 RTCA-AS1   K562      0.0065758344 0.014222561\n",
       "29      chr1 100029410 100029652 SASS6      K562     -0.0022461967 0.015147027\n",
       "30      chr1 100029410 100029652 SLC30A7    K562      0.0082130824 0.015050004\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮          \n",
       "2735904 chrX 9864063   9864529   TBL1X      K562     -0.0040138347 0.01373375 \n",
       "2735905 chrX 9864063   9864529   WWC3       K562      0.0246353382 0.01610863 \n",
       "2735906 chrX 9874033   9874283   FAM9B      K562      0.0224928812 0.01611845 \n",
       "2735907 chrX 9874033   9874283   TBL1X      K562     -0.0007651788 0.01413708 \n",
       "2735908 chrX 9874033   9874283   WWC3       K562      0.0023039813 0.01557011 \n",
       "2735909 chrX 9882684   9883146   FAM9B      K562     -0.0056623219 0.01864811 \n",
       "2735910 chrX 9882684   9883146   TBL1X      K562      0.0005373820 0.01633870 \n",
       "2735911 chrX 9882684   9883146   WWC3       K562      0.0269535323 0.01542460 \n",
       "2735912 chrX 9885199   9885463   FAM9B      K562      0.0026406256 0.01729513 \n",
       "2735913 chrX 9885199   9885463   TBL1X      K562     -0.0049799836 0.01769587 \n",
       "2735914 chrX 9885199   9885463   WWC3       K562      0.0256010058 0.01554113 \n",
       "2735915 chrX 9894498   9895136   FAM9B      K562     -0.0083305906 0.02289624 \n",
       "2735916 chrX 9894498   9895136   TBL1X      K562     -0.0061796709 0.01524818 \n",
       "2735917 chrX 9894498   9895136   WWC3       K562     -0.0015254686 0.01389721 \n",
       "2735918 chrX  989856    990611   AKAP17A    K562      0.0438793145 0.01350371 \n",
       "2735919 chrX  989856    990611   ASMT       K562      0.0409807849 0.01575378 \n",
       "2735920 chrX  989856    990611   ASMTL      K562      0.0231346780 0.01430504 \n",
       "2735921 chrX  989856    990611   GTPBP6     K562      0.0192363155 0.01517947 \n",
       "2735922 chrX  989856    990611   IL3RA      K562     -0.0068971428 0.01332706 \n",
       "2735923 chrX  989856    990611   LINC00106  K562      0.0009951114 0.01511970 \n",
       "2735924 chrX  989856    990611   P2RY8      K562      0.0223772932 0.01751157 \n",
       "2735925 chrX  989856    990611   PLCXD1     K562      0.0264926047 0.01474863 \n",
       "2735926 chrX  989856    990611   PPP2R3B    K562      0.0116986221 0.01697001 \n",
       "2735927 chrX  989856    990611   SLC25A6    K562      0.0250474347 0.01647551 \n",
       "2735928 chrX 9940445   9940756   FAM9B      K562     -0.0132319729 0.02016957 \n",
       "2735929 chrX 9940445   9940756   TBL1X      K562      0.0156089585 0.01483061 \n",
       "2735930 chrX 9940445   9940756   WWC3       K562     -0.0106894901 0.01340392 \n",
       "2735931 chrX 9995833   9996097   FAM9B      K562      0.0342630502 0.02045432 \n",
       "2735932 chrX 9995833   9996097   TBL1X      K562      0.0038259623 0.01250067 \n",
       "2735933 chrX 9995833   9996097   WWC3       K562     -0.0166795494 0.01661220 \n",
       "        rBgMean       pvalZ      \n",
       "1        4.489001e-03 0.493330384\n",
       "2       -3.737446e-03 0.569581566\n",
       "3        3.450980e-04 0.725210502\n",
       "4        2.383941e-03 0.655002762\n",
       "5        3.158801e-04 0.786573354\n",
       "6       -1.132415e-03 0.802242576\n",
       "7        4.068826e-03 0.234856551\n",
       "8        7.509079e-03 0.111900860\n",
       "9        8.810720e-04 0.295517267\n",
       "10       6.633906e-04 0.005820232\n",
       "11       1.939669e-03 0.219453562\n",
       "12       4.007851e-03 0.606520009\n",
       "13       6.626603e-03 0.403969559\n",
       "14       1.665634e-03 0.529766736\n",
       "15       3.441832e-03 0.807132973\n",
       "16       3.737736e-03 0.752938944\n",
       "17       2.734003e-03 0.378922323\n",
       "18      -6.118940e-04 0.166935613\n",
       "19       1.320443e-03 0.938738997\n",
       "20       2.883092e-03 0.582268644\n",
       "21      -1.274360e-04 0.073269728\n",
       "22      -1.925265e-03 0.738689026\n",
       "23      -9.756608e-04 0.064514528\n",
       "24       2.580706e-03 0.283837122\n",
       "25      -1.296950e-05 0.704494670\n",
       "26      -1.943826e-03 0.530362570\n",
       "27      -2.373059e-03 0.576160226\n",
       "28      -3.244719e-07 0.321906160\n",
       "29       2.417460e-03 0.620917941\n",
       "30      -9.715164e-04 0.270840767\n",
       "⋮       ⋮             ⋮          \n",
       "2735904 -0.005583988  0.454488866\n",
       "2735905  0.006245860  0.126811819\n",
       "2735906  0.007247343  0.172114044\n",
       "2735907 -0.002532682  0.450251463\n",
       "2735908  0.003161885  0.521970360\n",
       "2735909  0.013727961  0.850783289\n",
       "2735910 -0.005059875  0.365958456\n",
       "2735911  0.004478402  0.072544870\n",
       "2735912  0.010052671  0.665879192\n",
       "2735913 -0.005119067  0.496864475\n",
       "2735914  0.004344306  0.085691902\n",
       "2735915  0.010907225  0.799606736\n",
       "2735916 -0.004621935  0.540684688\n",
       "2735917  0.002020488  0.600698680\n",
       "2735918  0.012114167  0.009328157\n",
       "2735919  0.006177742  0.013580804\n",
       "2735920  0.005537010  0.109316197\n",
       "2735921  0.013268890  0.347113629\n",
       "2735922 -0.004542858  0.570110120\n",
       "2735923  0.009148941  0.705154900\n",
       "2735924  0.002964413  0.133806834\n",
       "2735925  0.010665462  0.141607203\n",
       "2735926  0.024904988  0.781779792\n",
       "2735927  0.019002397  0.356842873\n",
       "2735928  0.006904864  0.840951768\n",
       "2735929 -0.001402415  0.125681629\n",
       "2735930  0.001236118  0.813189329\n",
       "2735931  0.011398960  0.131823541\n",
       "2735932 -0.005331775  0.231907431\n",
       "2735933  0.003840498  0.891629325"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   data.table_1.15.2                \n",
       " [3] foreach_1.5.2                     BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] genomation_1.34.0                 Seurat_5.0.3                     \n",
       "[11] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[13] FigR_0.1.0                        motifmatchr_1.24.0               \n",
       "[15] chromVAR_1.24.0                   cowplot_1.1.3                    \n",
       "[17] ggplot2_3.5.1                     dplyr_1.1.4                      \n",
       "[19] SummarizedExperiment_1.30.2       Biobase_2.60.0                   \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               MatrixGenerics_1.12.3            \n",
       "[27] matrixStats_1.3.0                 Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             vroom_1.6.5                \n",
       " [15] globals_0.16.3              lattice_0.22-6             \n",
       " [17] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [19] plotly_4.10.4               plotrix_3.8-4              \n",
       " [21] yaml_2.3.8                  httpuv_1.6.15              \n",
       " [23] sctransform_0.4.1           spam_2.10-0                \n",
       " [25] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [27] pbapply_1.7-2               DBI_1.2.2                  \n",
       " [29] CNEr_1.38.0                 RColorBrewer_1.1-3         \n",
       " [31] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [33] Rtsne_0.17                  purrr_1.0.2                \n",
       " [35] R.utils_2.12.3              RCurl_1.98-1.14            \n",
       " [37] pracma_2.4.4                GenomeInfoDbData_1.2.11    \n",
       " [39] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [41] spatstat.utils_3.0-4        listenv_0.9.1              \n",
       " [43] seqLogo_1.68.0              goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             spatstat.random_3.2-3      \n",
       " [47] annotate_1.80.0             fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           leiden_0.4.3.1             \n",
       " [51] codetools_0.2-20            DelayedArray_0.26.7        \n",
       " [53] DT_0.33                     tidyselect_1.2.1           \n",
       " [55] spatstat.explore_3.2-6      base64enc_0.1-3            \n",
       " [57] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [59] progressr_0.14.0            iterators_1.0.14           \n",
       " [61] ggridges_0.5.6              survival_3.6-4             \n",
       " [63] tools_4.3.3                 TFMPvalue_0.0.9            \n",
       " [65] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [67] glue_1.7.0                  gridExtra_2.3              \n",
       " [69] IRdisplay_1.1               withr_3.0.0                \n",
       " [71] fastmap_1.1.1               fansi_1.0.6                \n",
       " [73] caTools_1.18.2              digest_0.6.35              \n",
       " [75] R6_2.5.1                    mime_0.12                  \n",
       " [77] nabor_0.5.0                 seqPattern_1.34.0          \n",
       " [79] colorspace_2.1-0            scattermore_1.2            \n",
       " [81] GO.db_3.18.0                tensor_1.5                 \n",
       " [83] gtools_3.9.5                poweRlaw_0.80.0            \n",
       " [85] spatstat.data_3.0-4         RSQLite_2.3.6              \n",
       " [87] R.methodsS3_1.8.2           utf8_1.2.4                 \n",
       " [89] tidyr_1.3.1                 generics_0.1.3             \n",
       " [91] httr_1.4.7                  htmlwidgets_1.6.4          \n",
       " [93] S4Arrays_1.2.1              TFBSTools_1.40.0           \n",
       " [95] uwot_0.1.16                 pkgconfig_2.0.3            \n",
       " [97] gtable_0.3.5                blob_1.2.4                 \n",
       " [99] impute_1.76.0               lmtest_0.9-40              \n",
       "[101] htmltools_0.5.8.1           dotCall64_1.1-1            \n",
       "[103] scales_1.3.0                png_0.1-8                  \n",
       "[105] tzdb_0.4.0                  reshape2_1.4.4             \n",
       "[107] rjson_0.2.21                uuid_1.2-0                 \n",
       "[109] nlme_3.1-164                repr_1.1.7                 \n",
       "[111] cachem_1.0.8                zoo_1.8-12                 \n",
       "[113] stringr_1.5.1               KernSmooth_2.23-22         \n",
       "[115] miniUI_0.1.1.1              AnnotationDbi_1.64.1       \n",
       "[117] restfulr_0.0.15             pillar_1.9.0               \n",
       "[119] vctrs_0.6.5                 RANN_2.6.1                 \n",
       "[121] promises_1.3.0              xtable_1.8-4               \n",
       "[123] cluster_2.1.6               evaluate_0.23              \n",
       "[125] readr_2.1.5                 cli_3.6.2                  \n",
       "[127] compiler_4.3.3              Rsamtools_2.18.0           \n",
       "[129] rlang_1.1.3                 crayon_1.5.2               \n",
       "[131] future.apply_1.11.2         plyr_1.8.9                 \n",
       "[133] stringi_1.8.3               gridBase_0.4-7             \n",
       "[135] deldir_2.0-4                viridisLite_0.4.2          \n",
       "[137] BiocParallel_1.36.0         munsell_0.5.1              \n",
       "[139] lazyeval_0.2.2              spatstat.geom_3.2-9        \n",
       "[141] IRkernel_1.3.2              RcppHNSW_0.6.0             \n",
       "[143] hms_1.1.3                   patchwork_1.2.0            \n",
       "[145] bit64_4.0.5                 future_1.33.2              \n",
       "[147] KEGGREST_1.40.1             shiny_1.8.1.1              \n",
       "[149] ROCR_1.0-11                 igraph_2.0.3               \n",
       "[151] memoise_2.0.1               bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
