{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/PBMC/1.prepare_data/2.merge_peaks.240624/pairs.rds\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/PBMC/1.prepare_data/2.merge_peaks.240624/matrix.atac.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/PBMC/1.prepare_data/2.merge_peaks.240624/matrix.rna.rds\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/PBMC/2.Genome_wide_prediction/FigR/PBMC.240624/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 32"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readRDS(path.pairs.E2G)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  4000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  4001  to  5000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  5001  to  6000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  6001  to  7000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  7001  to  8000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  8001  to  9000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  9001  to  10000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  10001  to  11000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  11001  to  12000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  12001  to  13000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  13001  to  14000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  14001  to  15000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  15001  to  16000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  16001  to  17000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  17001  to  18000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  18001  to  18183 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = readRDS(path.matrix.rna_count)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter = pairs.E2G[pairs.E2G$TargetGene %in% rownames(matrix.rna) &\n",
    "                              pairs.E2G$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 303411 18183 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(303411): chr1-10057-10601 chr1-180665-181882 ...\n",
       "  chrX-156029605-156030380 chrX-156030498-156030791\n",
       "rowData names(2): PeakName bias\n",
       "colnames(18183): Multiome10x_05_AAACAGCCAACAACAA-1\n",
       "  Multiome10x_05_AAACAGCCACCGGCTA-1 ...\n",
       "  Multiome10x_06_TTTGTTGGTACCCACC-1 Multiome10x_06_TTTGTTGGTACCGAAC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna),\n",
    "                       index = 1:nrow(matrix.rna))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  1.48875176509221 mins \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Time Elapsed:  2.4894782128599 hours \n",
      "\n",
      "[1] \"chr10\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  22.358166217804 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  47.297016398112 mins \n",
      "\n",
      "[1] \"chr11\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  52.5654158592224 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.47915652566486 hours \n",
      "\n",
      "[1] \"chr12\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  47.0683612823486 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.21033205105199 hours \n",
      "\n",
      "[1] \"chr13\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  18.4556124210358 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  26.9496510148048 mins \n",
      "\n",
      "[1] \"chr14\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  36.107225894928 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  55.3340686321259 mins \n",
      "\n",
      "[1] \"chr15\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  23.9545197486877 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  44.817246166865 mins \n",
      "\n",
      "[1] \"chr16\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  52.0667195320129 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.56945303791099 hours \n",
      "\n",
      "[1] \"chr17\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  1.01077580054601 mins \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.61288167178631 hours \n",
      "\n",
      "[1] \"chr18\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  12.2151446342468 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  20.617389969031 mins \n",
      "\n",
      "[1] \"chr19\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  1.02607193787893 mins \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.81788400696384 hours \n",
      "\n",
      "[1] \"chr2\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  34.9657821655273 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.01109292785327 hours \n",
      "\n",
      "[1] \"chr20\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  26.051956653595 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  44.4013279358546 mins \n",
      "\n",
      "[1] \"chr21\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  14.1352727413177 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  22.5699072519938 mins \n",
      "\n",
      "[1] \"chr22\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  38.1759307384491 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  42.5497344930967 mins \n",
      "\n",
      "[1] \"chr3\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  35.2581076622009 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.04224292867714 hours \n",
      "\n",
      "[1] \"chr4\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  23.1086747646332 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  36.3262584527334 mins \n",
      "\n",
      "[1] \"chr5\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  23.9719579219818 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  43.1502835591634 mins \n",
      "\n",
      "[1] \"chr6\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  32.2050845623016 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  57.3823140104612 mins \n",
      "\n",
      "[1] \"chr7\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  28.6496775150299 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  48.3189757307371 mins \n",
      "\n",
      "[1] \"chr8\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  20.33265376091 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  35.5336506485939 mins \n",
      "\n",
      "[1] \"chr9\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  29.1112413406372 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  49.7705272873243 mins \n",
      "\n",
      "[1] \"chrX\"\n",
      "Running in parallel using  32 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  20.3316633701324 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  32.7874371131261 mins \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter[seqnames(pairs.E2G.filter) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 5424230 ranges and 7 metadata columns:\n",
       "       seqnames              ranges strand |  TargetGene               PeakName\n",
       "          <Rle>           <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 100028331-100029533      * |         AGL chr1-100028331-10002..\n",
       "  chr1     chr1 100028331-100029533      * |      CDC14A chr1-100028331-10002..\n",
       "  chr1     chr1 100028331-100029533      * |         DBT chr1-100028331-10002..\n",
       "  chr1     chr1 100028331-100029533      * |        DPH5 chr1-100028331-10002..\n",
       "  chr1     chr1 100028331-100029533      * |       EXTL2 chr1-100028331-10002..\n",
       "   ...      ...                 ...    ... .         ...                    ...\n",
       "  chrX     chrX     9995651-9996434      * |        WWC3   chrX-9995651-9996434\n",
       "  chrX     chrX     9997538-9997892      * |       CLCN4   chrX-9997538-9997892\n",
       "  chrX     chrX     9997538-9997892      * |      GPR143   chrX-9997538-9997892\n",
       "  chrX     chrX     9997538-9997892      * |       TBL1X   chrX-9997538-9997892\n",
       "  chrX     chrX     9997538-9997892      * |        WWC3   chrX-9997538-9997892\n",
       "                     PairName         rObs      rBgSD      rBgMean     pvalZ\n",
       "                  <character>    <numeric>  <numeric>    <numeric> <numeric>\n",
       "  chr1 chr1-100028331-10002..   0.00783415 0.01357713  -0.00158158 0.2439980\n",
       "  chr1 chr1-100028331-10002..  -0.02281754 0.06365834  -0.03952376 0.3964927\n",
       "  chr1 chr1-100028331-10002..   0.00185825 0.00786541   0.00547256 0.6770693\n",
       "  chr1 chr1-100028331-10002..   0.01989010 0.01544746  -0.00232977 0.0751581\n",
       "  chr1 chr1-100028331-10002..   0.01462640 0.00888672   0.00752278 0.2120431\n",
       "   ...                    ...          ...        ...          ...       ...\n",
       "  chrX chrX-9995651-9996434..  0.014235543 0.00744980  0.006943347 0.1638283\n",
       "  chrX chrX-9997538-9997892..  0.016935133 0.00978972  0.005613377 0.1237394\n",
       "  chrX chrX-9997538-9997892..  0.014742341 0.00674703 -0.000482596 0.0120184\n",
       "  chrX chrX-9997538-9997892.. -0.000989114 0.01482842  0.012504955 0.8185924\n",
       "  chrX chrX-9997538-9997892..  0.008704360 0.00700415  0.004177575 0.2590425\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5424230 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>AGL     </td><td>K562</td><td> 0.0078341525</td><td>0.013577127</td><td>-0.0015815836</td><td>0.24399795</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>CDC14A  </td><td>K562</td><td>-0.0228175386</td><td>0.063658339</td><td>-0.0395237640</td><td>0.39649275</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>DBT     </td><td>K562</td><td> 0.0018582526</td><td>0.007865414</td><td> 0.0054725617</td><td>0.67706934</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>DPH5    </td><td>K562</td><td> 0.0198901028</td><td>0.015447465</td><td>-0.0023297737</td><td>0.07515807</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>EXTL2   </td><td>K562</td><td> 0.0146264016</td><td>0.008886718</td><td> 0.0075227843</td><td>0.21204311</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>FRRS1   </td><td>K562</td><td>-0.0036368475</td><td>0.020406558</td><td> 0.0154815205</td><td>0.82558824</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>GPR88   </td><td>K562</td><td>-0.0026525468</td><td>0.009457548</td><td> 0.0013105063</td><td>0.66240508</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>MFSD14A </td><td>K562</td><td>-0.0088375880</td><td>0.006931090</td><td> 0.0031270674</td><td>0.95784697</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>RTCA    </td><td>K562</td><td> 0.0147787854</td><td>0.006930339</td><td> 0.0072043021</td><td>0.13720885</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>RTCA-AS1</td><td>K562</td><td>-0.0024411066</td><td>0.008035791</td><td> 0.0057421774</td><td>0.84574526</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>SASS6   </td><td>K562</td><td> 0.0183661392</td><td>0.008308769</td><td> 0.0006598667</td><td>0.01654316</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>SLC30A7 </td><td>K562</td><td>-0.0009802877</td><td>0.010378247</td><td> 0.0157497491</td><td>0.94652221</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>SLC35A3 </td><td>K562</td><td> 0.0025537225</td><td>0.008114936</td><td> 0.0094359258</td><td>0.80180633</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>TRMT13  </td><td>K562</td><td> 0.0085382999</td><td>0.012048989</td><td> 0.0009090877</td><td>0.26330715</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>AGL     </td><td>K562</td><td> 0.0252912654</td><td>0.009121610</td><td> 0.0134717601</td><td>0.09752742</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>CDC14A  </td><td>K562</td><td> 0.0976187201</td><td>0.029041227</td><td> 0.0510578984</td><td>0.05443793</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>DBT     </td><td>K562</td><td> 0.0082957779</td><td>0.005837887</td><td> 0.0049608171</td><td>0.28391115</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>DPH5    </td><td>K562</td><td> 0.0130808371</td><td>0.008476571</td><td> 0.0118986221</td><td>0.44453996</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>EXTL2   </td><td>K562</td><td> 0.0101170915</td><td>0.006783694</td><td> 0.0020523559</td><td>0.11725104</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>FRRS1   </td><td>K562</td><td>-0.0150448622</td><td>0.010971285</td><td>-0.0135920129</td><td>0.55267510</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>GPR88   </td><td>K562</td><td>-0.0055974188</td><td>0.007510672</td><td>-0.0038437107</td><td>0.59231166</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>MFSD14A </td><td>K562</td><td>-0.0054194069</td><td>0.007177446</td><td> 0.0031600301</td><td>0.88402149</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>RTCA    </td><td>K562</td><td>-0.0078052972</td><td>0.006787551</td><td> 0.0016697050</td><td>0.91863345</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>RTCA-AS1</td><td>K562</td><td>-0.0212093737</td><td>0.007132062</td><td> 0.0004945195</td><td>0.99882940</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>SASS6   </td><td>K562</td><td> 0.0026136083</td><td>0.006386352</td><td> 0.0072432756</td><td>0.76575294</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>SLC30A7 </td><td>K562</td><td> 0.0184780003</td><td>0.007582846</td><td> 0.0027047660</td><td>0.01875724</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>SLC35A3 </td><td>K562</td><td> 0.0127661338</td><td>0.007086684</td><td> 0.0052968957</td><td>0.14594557</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>TRMT13  </td><td>K562</td><td> 0.0218326533</td><td>0.007411628</td><td> 0.0101568919</td><td>0.05759012</td></tr>\n",
       "\t<tr><td>chr1</td><td>100044514</td><td>100044850</td><td>AGL     </td><td>K562</td><td> 0.0125142301</td><td>0.008873161</td><td> 0.0016335324</td><td>0.11005263</td></tr>\n",
       "\t<tr><td>chr1</td><td>100044514</td><td>100044850</td><td>CDC14A  </td><td>K562</td><td> 0.0004396872</td><td>0.015254556</td><td>-0.0081728473</td><td>0.28617712</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>AKAP17A  </td><td>K562</td><td>-0.0100166443</td><td>0.007818366</td><td> 0.0030581394</td><td>0.952769061</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>ASMT     </td><td>K562</td><td>-0.0013149372</td><td>0.004708502</td><td>-0.0007724019</td><td>0.545866449</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>ASMTL    </td><td>K562</td><td> 0.0001601504</td><td>0.008112325</td><td> 0.0022825383</td><td>0.603194761</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>CSF2RA   </td><td>K562</td><td> 0.0064858931</td><td>0.011159056</td><td> 0.0064432526</td><td>0.498475582</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>GTPBP6   </td><td>K562</td><td> 0.0013480666</td><td>0.006634397</td><td> 0.0022619760</td><td>0.554782269</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>IL3RA    </td><td>K562</td><td> 0.0036831612</td><td>0.008303235</td><td> 0.0024571668</td><td>0.441308467</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>LINC00106</td><td>K562</td><td> 0.0171906634</td><td>0.011070342</td><td> 0.0063330011</td><td>0.163348562</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>P2RY8    </td><td>K562</td><td> 0.0065786972</td><td>0.006430348</td><td> 0.0008736880</td><td>0.187485445</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>PLCXD1   </td><td>K562</td><td> 0.0014973300</td><td>0.007136451</td><td> 0.0007033834</td><td>0.455708143</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>PPP2R3B  </td><td>K562</td><td> 0.0170103080</td><td>0.007203074</td><td>-0.0001724716</td><td>0.008528458</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>SLC25A6  </td><td>K562</td><td> 0.0002857254</td><td>0.008257469</td><td> 0.0014631888</td><td>0.556694482</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>AKAP17A  </td><td>K562</td><td>-0.0053356686</td><td>0.007751039</td><td> 0.0040323523</td><td>0.886594563</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>ASMT     </td><td>K562</td><td>-0.0018603695</td><td>0.010417029</td><td> 0.0017406188</td><td>0.635209471</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>ASMTL    </td><td>K562</td><td>-0.0031390689</td><td>0.007901701</td><td> 0.0027979899</td><td>0.773783392</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>CSF2RA   </td><td>K562</td><td>-0.0058282870</td><td>0.013121997</td><td> 0.0095082030</td><td>0.878750203</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>GTPBP6   </td><td>K562</td><td>-0.0018860789</td><td>0.008991638</td><td> 0.0042383316</td><td>0.752103139</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>IL3RA    </td><td>K562</td><td> 0.0102547348</td><td>0.008846145</td><td> 0.0038417992</td><td>0.234244029</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>LINC00106</td><td>K562</td><td> 0.0241222491</td><td>0.009084283</td><td> 0.0065312650</td><td>0.026408159</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>P2RY8    </td><td>K562</td><td> 0.0104815495</td><td>0.007434975</td><td>-0.0001671759</td><td>0.076036468</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>PLCXD1   </td><td>K562</td><td>-0.0036885610</td><td>0.007951101</td><td> 0.0032490256</td><td>0.808540769</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>PPP2R3B  </td><td>K562</td><td> 0.0161612413</td><td>0.009159334</td><td> 0.0013372423</td><td>0.052781945</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>SLC25A6  </td><td>K562</td><td> 0.0131691609</td><td>0.008061788</td><td> 0.0029846775</td><td>0.103240086</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995651</td><td>9996434</td><td>CLCN4    </td><td>K562</td><td> 0.0090697954</td><td>0.018218238</td><td> 0.0113132275</td><td>0.549002713</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995651</td><td>9996434</td><td>GPR143   </td><td>K562</td><td> 0.0021010659</td><td>0.007557823</td><td> 0.0003748135</td><td>0.409665295</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995651</td><td>9996434</td><td>TBL1X    </td><td>K562</td><td> 0.0043513127</td><td>0.052325655</td><td> 0.0382900712</td><td>0.741703615</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995651</td><td>9996434</td><td>WWC3     </td><td>K562</td><td> 0.0142355426</td><td>0.007449798</td><td> 0.0069433467</td><td>0.163828339</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997538</td><td>9997892</td><td>CLCN4    </td><td>K562</td><td> 0.0169351328</td><td>0.009789716</td><td> 0.0056133765</td><td>0.123739390</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997538</td><td>9997892</td><td>GPR143   </td><td>K562</td><td> 0.0147423414</td><td>0.006747029</td><td>-0.0004825961</td><td>0.012018433</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997538</td><td>9997892</td><td>TBL1X    </td><td>K562</td><td>-0.0009891140</td><td>0.014828419</td><td> 0.0125049547</td><td>0.818592426</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997538</td><td>9997892</td><td>WWC3     </td><td>K562</td><td> 0.0087043601</td><td>0.007004155</td><td> 0.0041775750</td><td>0.259042548</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5424230 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100028331 & 100029533 & AGL      & K562 &  0.0078341525 & 0.013577127 & -0.0015815836 & 0.24399795\\\\\n",
       "\t chr1 & 100028331 & 100029533 & CDC14A   & K562 & -0.0228175386 & 0.063658339 & -0.0395237640 & 0.39649275\\\\\n",
       "\t chr1 & 100028331 & 100029533 & DBT      & K562 &  0.0018582526 & 0.007865414 &  0.0054725617 & 0.67706934\\\\\n",
       "\t chr1 & 100028331 & 100029533 & DPH5     & K562 &  0.0198901028 & 0.015447465 & -0.0023297737 & 0.07515807\\\\\n",
       "\t chr1 & 100028331 & 100029533 & EXTL2    & K562 &  0.0146264016 & 0.008886718 &  0.0075227843 & 0.21204311\\\\\n",
       "\t chr1 & 100028331 & 100029533 & FRRS1    & K562 & -0.0036368475 & 0.020406558 &  0.0154815205 & 0.82558824\\\\\n",
       "\t chr1 & 100028331 & 100029533 & GPR88    & K562 & -0.0026525468 & 0.009457548 &  0.0013105063 & 0.66240508\\\\\n",
       "\t chr1 & 100028331 & 100029533 & MFSD14A  & K562 & -0.0088375880 & 0.006931090 &  0.0031270674 & 0.95784697\\\\\n",
       "\t chr1 & 100028331 & 100029533 & RTCA     & K562 &  0.0147787854 & 0.006930339 &  0.0072043021 & 0.13720885\\\\\n",
       "\t chr1 & 100028331 & 100029533 & RTCA-AS1 & K562 & -0.0024411066 & 0.008035791 &  0.0057421774 & 0.84574526\\\\\n",
       "\t chr1 & 100028331 & 100029533 & SASS6    & K562 &  0.0183661392 & 0.008308769 &  0.0006598667 & 0.01654316\\\\\n",
       "\t chr1 & 100028331 & 100029533 & SLC30A7  & K562 & -0.0009802877 & 0.010378247 &  0.0157497491 & 0.94652221\\\\\n",
       "\t chr1 & 100028331 & 100029533 & SLC35A3  & K562 &  0.0025537225 & 0.008114936 &  0.0094359258 & 0.80180633\\\\\n",
       "\t chr1 & 100028331 & 100029533 & TRMT13   & K562 &  0.0085382999 & 0.012048989 &  0.0009090877 & 0.26330715\\\\\n",
       "\t chr1 & 100037542 & 100039303 & AGL      & K562 &  0.0252912654 & 0.009121610 &  0.0134717601 & 0.09752742\\\\\n",
       "\t chr1 & 100037542 & 100039303 & CDC14A   & K562 &  0.0976187201 & 0.029041227 &  0.0510578984 & 0.05443793\\\\\n",
       "\t chr1 & 100037542 & 100039303 & DBT      & K562 &  0.0082957779 & 0.005837887 &  0.0049608171 & 0.28391115\\\\\n",
       "\t chr1 & 100037542 & 100039303 & DPH5     & K562 &  0.0130808371 & 0.008476571 &  0.0118986221 & 0.44453996\\\\\n",
       "\t chr1 & 100037542 & 100039303 & EXTL2    & K562 &  0.0101170915 & 0.006783694 &  0.0020523559 & 0.11725104\\\\\n",
       "\t chr1 & 100037542 & 100039303 & FRRS1    & K562 & -0.0150448622 & 0.010971285 & -0.0135920129 & 0.55267510\\\\\n",
       "\t chr1 & 100037542 & 100039303 & GPR88    & K562 & -0.0055974188 & 0.007510672 & -0.0038437107 & 0.59231166\\\\\n",
       "\t chr1 & 100037542 & 100039303 & MFSD14A  & K562 & -0.0054194069 & 0.007177446 &  0.0031600301 & 0.88402149\\\\\n",
       "\t chr1 & 100037542 & 100039303 & RTCA     & K562 & -0.0078052972 & 0.006787551 &  0.0016697050 & 0.91863345\\\\\n",
       "\t chr1 & 100037542 & 100039303 & RTCA-AS1 & K562 & -0.0212093737 & 0.007132062 &  0.0004945195 & 0.99882940\\\\\n",
       "\t chr1 & 100037542 & 100039303 & SASS6    & K562 &  0.0026136083 & 0.006386352 &  0.0072432756 & 0.76575294\\\\\n",
       "\t chr1 & 100037542 & 100039303 & SLC30A7  & K562 &  0.0184780003 & 0.007582846 &  0.0027047660 & 0.01875724\\\\\n",
       "\t chr1 & 100037542 & 100039303 & SLC35A3  & K562 &  0.0127661338 & 0.007086684 &  0.0052968957 & 0.14594557\\\\\n",
       "\t chr1 & 100037542 & 100039303 & TRMT13   & K562 &  0.0218326533 & 0.007411628 &  0.0101568919 & 0.05759012\\\\\n",
       "\t chr1 & 100044514 & 100044850 & AGL      & K562 &  0.0125142301 & 0.008873161 &  0.0016335324 & 0.11005263\\\\\n",
       "\t chr1 & 100044514 & 100044850 & CDC14A   & K562 &  0.0004396872 & 0.015254556 & -0.0081728473 & 0.28617712\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  999238 &  999388 & AKAP17A   & K562 & -0.0100166443 & 0.007818366 &  0.0030581394 & 0.952769061\\\\\n",
       "\t chrX &  999238 &  999388 & ASMT      & K562 & -0.0013149372 & 0.004708502 & -0.0007724019 & 0.545866449\\\\\n",
       "\t chrX &  999238 &  999388 & ASMTL     & K562 &  0.0001601504 & 0.008112325 &  0.0022825383 & 0.603194761\\\\\n",
       "\t chrX &  999238 &  999388 & CSF2RA    & K562 &  0.0064858931 & 0.011159056 &  0.0064432526 & 0.498475582\\\\\n",
       "\t chrX &  999238 &  999388 & GTPBP6    & K562 &  0.0013480666 & 0.006634397 &  0.0022619760 & 0.554782269\\\\\n",
       "\t chrX &  999238 &  999388 & IL3RA     & K562 &  0.0036831612 & 0.008303235 &  0.0024571668 & 0.441308467\\\\\n",
       "\t chrX &  999238 &  999388 & LINC00106 & K562 &  0.0171906634 & 0.011070342 &  0.0063330011 & 0.163348562\\\\\n",
       "\t chrX &  999238 &  999388 & P2RY8     & K562 &  0.0065786972 & 0.006430348 &  0.0008736880 & 0.187485445\\\\\n",
       "\t chrX &  999238 &  999388 & PLCXD1    & K562 &  0.0014973300 & 0.007136451 &  0.0007033834 & 0.455708143\\\\\n",
       "\t chrX &  999238 &  999388 & PPP2R3B   & K562 &  0.0170103080 & 0.007203074 & -0.0001724716 & 0.008528458\\\\\n",
       "\t chrX &  999238 &  999388 & SLC25A6   & K562 &  0.0002857254 & 0.008257469 &  0.0014631888 & 0.556694482\\\\\n",
       "\t chrX &  999443 &  999815 & AKAP17A   & K562 & -0.0053356686 & 0.007751039 &  0.0040323523 & 0.886594563\\\\\n",
       "\t chrX &  999443 &  999815 & ASMT      & K562 & -0.0018603695 & 0.010417029 &  0.0017406188 & 0.635209471\\\\\n",
       "\t chrX &  999443 &  999815 & ASMTL     & K562 & -0.0031390689 & 0.007901701 &  0.0027979899 & 0.773783392\\\\\n",
       "\t chrX &  999443 &  999815 & CSF2RA    & K562 & -0.0058282870 & 0.013121997 &  0.0095082030 & 0.878750203\\\\\n",
       "\t chrX &  999443 &  999815 & GTPBP6    & K562 & -0.0018860789 & 0.008991638 &  0.0042383316 & 0.752103139\\\\\n",
       "\t chrX &  999443 &  999815 & IL3RA     & K562 &  0.0102547348 & 0.008846145 &  0.0038417992 & 0.234244029\\\\\n",
       "\t chrX &  999443 &  999815 & LINC00106 & K562 &  0.0241222491 & 0.009084283 &  0.0065312650 & 0.026408159\\\\\n",
       "\t chrX &  999443 &  999815 & P2RY8     & K562 &  0.0104815495 & 0.007434975 & -0.0001671759 & 0.076036468\\\\\n",
       "\t chrX &  999443 &  999815 & PLCXD1    & K562 & -0.0036885610 & 0.007951101 &  0.0032490256 & 0.808540769\\\\\n",
       "\t chrX &  999443 &  999815 & PPP2R3B   & K562 &  0.0161612413 & 0.009159334 &  0.0013372423 & 0.052781945\\\\\n",
       "\t chrX &  999443 &  999815 & SLC25A6   & K562 &  0.0131691609 & 0.008061788 &  0.0029846775 & 0.103240086\\\\\n",
       "\t chrX & 9995651 & 9996434 & CLCN4     & K562 &  0.0090697954 & 0.018218238 &  0.0113132275 & 0.549002713\\\\\n",
       "\t chrX & 9995651 & 9996434 & GPR143    & K562 &  0.0021010659 & 0.007557823 &  0.0003748135 & 0.409665295\\\\\n",
       "\t chrX & 9995651 & 9996434 & TBL1X     & K562 &  0.0043513127 & 0.052325655 &  0.0382900712 & 0.741703615\\\\\n",
       "\t chrX & 9995651 & 9996434 & WWC3      & K562 &  0.0142355426 & 0.007449798 &  0.0069433467 & 0.163828339\\\\\n",
       "\t chrX & 9997538 & 9997892 & CLCN4     & K562 &  0.0169351328 & 0.009789716 &  0.0056133765 & 0.123739390\\\\\n",
       "\t chrX & 9997538 & 9997892 & GPR143    & K562 &  0.0147423414 & 0.006747029 & -0.0004825961 & 0.012018433\\\\\n",
       "\t chrX & 9997538 & 9997892 & TBL1X     & K562 & -0.0009891140 & 0.014828419 &  0.0125049547 & 0.818592426\\\\\n",
       "\t chrX & 9997538 & 9997892 & WWC3      & K562 &  0.0087043601 & 0.007004155 &  0.0041775750 & 0.259042548\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5424230 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100028331 | 100029533 | AGL      | K562 |  0.0078341525 | 0.013577127 | -0.0015815836 | 0.24399795 |\n",
       "| chr1 | 100028331 | 100029533 | CDC14A   | K562 | -0.0228175386 | 0.063658339 | -0.0395237640 | 0.39649275 |\n",
       "| chr1 | 100028331 | 100029533 | DBT      | K562 |  0.0018582526 | 0.007865414 |  0.0054725617 | 0.67706934 |\n",
       "| chr1 | 100028331 | 100029533 | DPH5     | K562 |  0.0198901028 | 0.015447465 | -0.0023297737 | 0.07515807 |\n",
       "| chr1 | 100028331 | 100029533 | EXTL2    | K562 |  0.0146264016 | 0.008886718 |  0.0075227843 | 0.21204311 |\n",
       "| chr1 | 100028331 | 100029533 | FRRS1    | K562 | -0.0036368475 | 0.020406558 |  0.0154815205 | 0.82558824 |\n",
       "| chr1 | 100028331 | 100029533 | GPR88    | K562 | -0.0026525468 | 0.009457548 |  0.0013105063 | 0.66240508 |\n",
       "| chr1 | 100028331 | 100029533 | MFSD14A  | K562 | -0.0088375880 | 0.006931090 |  0.0031270674 | 0.95784697 |\n",
       "| chr1 | 100028331 | 100029533 | RTCA     | K562 |  0.0147787854 | 0.006930339 |  0.0072043021 | 0.13720885 |\n",
       "| chr1 | 100028331 | 100029533 | RTCA-AS1 | K562 | -0.0024411066 | 0.008035791 |  0.0057421774 | 0.84574526 |\n",
       "| chr1 | 100028331 | 100029533 | SASS6    | K562 |  0.0183661392 | 0.008308769 |  0.0006598667 | 0.01654316 |\n",
       "| chr1 | 100028331 | 100029533 | SLC30A7  | K562 | -0.0009802877 | 0.010378247 |  0.0157497491 | 0.94652221 |\n",
       "| chr1 | 100028331 | 100029533 | SLC35A3  | K562 |  0.0025537225 | 0.008114936 |  0.0094359258 | 0.80180633 |\n",
       "| chr1 | 100028331 | 100029533 | TRMT13   | K562 |  0.0085382999 | 0.012048989 |  0.0009090877 | 0.26330715 |\n",
       "| chr1 | 100037542 | 100039303 | AGL      | K562 |  0.0252912654 | 0.009121610 |  0.0134717601 | 0.09752742 |\n",
       "| chr1 | 100037542 | 100039303 | CDC14A   | K562 |  0.0976187201 | 0.029041227 |  0.0510578984 | 0.05443793 |\n",
       "| chr1 | 100037542 | 100039303 | DBT      | K562 |  0.0082957779 | 0.005837887 |  0.0049608171 | 0.28391115 |\n",
       "| chr1 | 100037542 | 100039303 | DPH5     | K562 |  0.0130808371 | 0.008476571 |  0.0118986221 | 0.44453996 |\n",
       "| chr1 | 100037542 | 100039303 | EXTL2    | K562 |  0.0101170915 | 0.006783694 |  0.0020523559 | 0.11725104 |\n",
       "| chr1 | 100037542 | 100039303 | FRRS1    | K562 | -0.0150448622 | 0.010971285 | -0.0135920129 | 0.55267510 |\n",
       "| chr1 | 100037542 | 100039303 | GPR88    | K562 | -0.0055974188 | 0.007510672 | -0.0038437107 | 0.59231166 |\n",
       "| chr1 | 100037542 | 100039303 | MFSD14A  | K562 | -0.0054194069 | 0.007177446 |  0.0031600301 | 0.88402149 |\n",
       "| chr1 | 100037542 | 100039303 | RTCA     | K562 | -0.0078052972 | 0.006787551 |  0.0016697050 | 0.91863345 |\n",
       "| chr1 | 100037542 | 100039303 | RTCA-AS1 | K562 | -0.0212093737 | 0.007132062 |  0.0004945195 | 0.99882940 |\n",
       "| chr1 | 100037542 | 100039303 | SASS6    | K562 |  0.0026136083 | 0.006386352 |  0.0072432756 | 0.76575294 |\n",
       "| chr1 | 100037542 | 100039303 | SLC30A7  | K562 |  0.0184780003 | 0.007582846 |  0.0027047660 | 0.01875724 |\n",
       "| chr1 | 100037542 | 100039303 | SLC35A3  | K562 |  0.0127661338 | 0.007086684 |  0.0052968957 | 0.14594557 |\n",
       "| chr1 | 100037542 | 100039303 | TRMT13   | K562 |  0.0218326533 | 0.007411628 |  0.0101568919 | 0.05759012 |\n",
       "| chr1 | 100044514 | 100044850 | AGL      | K562 |  0.0125142301 | 0.008873161 |  0.0016335324 | 0.11005263 |\n",
       "| chr1 | 100044514 | 100044850 | CDC14A   | K562 |  0.0004396872 | 0.015254556 | -0.0081728473 | 0.28617712 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  999238 |  999388 | AKAP17A   | K562 | -0.0100166443 | 0.007818366 |  0.0030581394 | 0.952769061 |\n",
       "| chrX |  999238 |  999388 | ASMT      | K562 | -0.0013149372 | 0.004708502 | -0.0007724019 | 0.545866449 |\n",
       "| chrX |  999238 |  999388 | ASMTL     | K562 |  0.0001601504 | 0.008112325 |  0.0022825383 | 0.603194761 |\n",
       "| chrX |  999238 |  999388 | CSF2RA    | K562 |  0.0064858931 | 0.011159056 |  0.0064432526 | 0.498475582 |\n",
       "| chrX |  999238 |  999388 | GTPBP6    | K562 |  0.0013480666 | 0.006634397 |  0.0022619760 | 0.554782269 |\n",
       "| chrX |  999238 |  999388 | IL3RA     | K562 |  0.0036831612 | 0.008303235 |  0.0024571668 | 0.441308467 |\n",
       "| chrX |  999238 |  999388 | LINC00106 | K562 |  0.0171906634 | 0.011070342 |  0.0063330011 | 0.163348562 |\n",
       "| chrX |  999238 |  999388 | P2RY8     | K562 |  0.0065786972 | 0.006430348 |  0.0008736880 | 0.187485445 |\n",
       "| chrX |  999238 |  999388 | PLCXD1    | K562 |  0.0014973300 | 0.007136451 |  0.0007033834 | 0.455708143 |\n",
       "| chrX |  999238 |  999388 | PPP2R3B   | K562 |  0.0170103080 | 0.007203074 | -0.0001724716 | 0.008528458 |\n",
       "| chrX |  999238 |  999388 | SLC25A6   | K562 |  0.0002857254 | 0.008257469 |  0.0014631888 | 0.556694482 |\n",
       "| chrX |  999443 |  999815 | AKAP17A   | K562 | -0.0053356686 | 0.007751039 |  0.0040323523 | 0.886594563 |\n",
       "| chrX |  999443 |  999815 | ASMT      | K562 | -0.0018603695 | 0.010417029 |  0.0017406188 | 0.635209471 |\n",
       "| chrX |  999443 |  999815 | ASMTL     | K562 | -0.0031390689 | 0.007901701 |  0.0027979899 | 0.773783392 |\n",
       "| chrX |  999443 |  999815 | CSF2RA    | K562 | -0.0058282870 | 0.013121997 |  0.0095082030 | 0.878750203 |\n",
       "| chrX |  999443 |  999815 | GTPBP6    | K562 | -0.0018860789 | 0.008991638 |  0.0042383316 | 0.752103139 |\n",
       "| chrX |  999443 |  999815 | IL3RA     | K562 |  0.0102547348 | 0.008846145 |  0.0038417992 | 0.234244029 |\n",
       "| chrX |  999443 |  999815 | LINC00106 | K562 |  0.0241222491 | 0.009084283 |  0.0065312650 | 0.026408159 |\n",
       "| chrX |  999443 |  999815 | P2RY8     | K562 |  0.0104815495 | 0.007434975 | -0.0001671759 | 0.076036468 |\n",
       "| chrX |  999443 |  999815 | PLCXD1    | K562 | -0.0036885610 | 0.007951101 |  0.0032490256 | 0.808540769 |\n",
       "| chrX |  999443 |  999815 | PPP2R3B   | K562 |  0.0161612413 | 0.009159334 |  0.0013372423 | 0.052781945 |\n",
       "| chrX |  999443 |  999815 | SLC25A6   | K562 |  0.0131691609 | 0.008061788 |  0.0029846775 | 0.103240086 |\n",
       "| chrX | 9995651 | 9996434 | CLCN4     | K562 |  0.0090697954 | 0.018218238 |  0.0113132275 | 0.549002713 |\n",
       "| chrX | 9995651 | 9996434 | GPR143    | K562 |  0.0021010659 | 0.007557823 |  0.0003748135 | 0.409665295 |\n",
       "| chrX | 9995651 | 9996434 | TBL1X     | K562 |  0.0043513127 | 0.052325655 |  0.0382900712 | 0.741703615 |\n",
       "| chrX | 9995651 | 9996434 | WWC3      | K562 |  0.0142355426 | 0.007449798 |  0.0069433467 | 0.163828339 |\n",
       "| chrX | 9997538 | 9997892 | CLCN4     | K562 |  0.0169351328 | 0.009789716 |  0.0056133765 | 0.123739390 |\n",
       "| chrX | 9997538 | 9997892 | GPR143    | K562 |  0.0147423414 | 0.006747029 | -0.0004825961 | 0.012018433 |\n",
       "| chrX | 9997538 | 9997892 | TBL1X     | K562 | -0.0009891140 | 0.014828419 |  0.0125049547 | 0.818592426 |\n",
       "| chrX | 9997538 | 9997892 | WWC3      | K562 |  0.0087043601 | 0.007004155 |  0.0041775750 | 0.259042548 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType rObs          rBgSD      \n",
       "1       chr1 100028331 100029533 AGL        K562      0.0078341525 0.013577127\n",
       "2       chr1 100028331 100029533 CDC14A     K562     -0.0228175386 0.063658339\n",
       "3       chr1 100028331 100029533 DBT        K562      0.0018582526 0.007865414\n",
       "4       chr1 100028331 100029533 DPH5       K562      0.0198901028 0.015447465\n",
       "5       chr1 100028331 100029533 EXTL2      K562      0.0146264016 0.008886718\n",
       "6       chr1 100028331 100029533 FRRS1      K562     -0.0036368475 0.020406558\n",
       "7       chr1 100028331 100029533 GPR88      K562     -0.0026525468 0.009457548\n",
       "8       chr1 100028331 100029533 MFSD14A    K562     -0.0088375880 0.006931090\n",
       "9       chr1 100028331 100029533 RTCA       K562      0.0147787854 0.006930339\n",
       "10      chr1 100028331 100029533 RTCA-AS1   K562     -0.0024411066 0.008035791\n",
       "11      chr1 100028331 100029533 SASS6      K562      0.0183661392 0.008308769\n",
       "12      chr1 100028331 100029533 SLC30A7    K562     -0.0009802877 0.010378247\n",
       "13      chr1 100028331 100029533 SLC35A3    K562      0.0025537225 0.008114936\n",
       "14      chr1 100028331 100029533 TRMT13     K562      0.0085382999 0.012048989\n",
       "15      chr1 100037542 100039303 AGL        K562      0.0252912654 0.009121610\n",
       "16      chr1 100037542 100039303 CDC14A     K562      0.0976187201 0.029041227\n",
       "17      chr1 100037542 100039303 DBT        K562      0.0082957779 0.005837887\n",
       "18      chr1 100037542 100039303 DPH5       K562      0.0130808371 0.008476571\n",
       "19      chr1 100037542 100039303 EXTL2      K562      0.0101170915 0.006783694\n",
       "20      chr1 100037542 100039303 FRRS1      K562     -0.0150448622 0.010971285\n",
       "21      chr1 100037542 100039303 GPR88      K562     -0.0055974188 0.007510672\n",
       "22      chr1 100037542 100039303 MFSD14A    K562     -0.0054194069 0.007177446\n",
       "23      chr1 100037542 100039303 RTCA       K562     -0.0078052972 0.006787551\n",
       "24      chr1 100037542 100039303 RTCA-AS1   K562     -0.0212093737 0.007132062\n",
       "25      chr1 100037542 100039303 SASS6      K562      0.0026136083 0.006386352\n",
       "26      chr1 100037542 100039303 SLC30A7    K562      0.0184780003 0.007582846\n",
       "27      chr1 100037542 100039303 SLC35A3    K562      0.0127661338 0.007086684\n",
       "28      chr1 100037542 100039303 TRMT13     K562      0.0218326533 0.007411628\n",
       "29      chr1 100044514 100044850 AGL        K562      0.0125142301 0.008873161\n",
       "30      chr1 100044514 100044850 CDC14A     K562      0.0004396872 0.015254556\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮          \n",
       "5424201 chrX  999238    999388   AKAP17A    K562     -0.0100166443 0.007818366\n",
       "5424202 chrX  999238    999388   ASMT       K562     -0.0013149372 0.004708502\n",
       "5424203 chrX  999238    999388   ASMTL      K562      0.0001601504 0.008112325\n",
       "5424204 chrX  999238    999388   CSF2RA     K562      0.0064858931 0.011159056\n",
       "5424205 chrX  999238    999388   GTPBP6     K562      0.0013480666 0.006634397\n",
       "5424206 chrX  999238    999388   IL3RA      K562      0.0036831612 0.008303235\n",
       "5424207 chrX  999238    999388   LINC00106  K562      0.0171906634 0.011070342\n",
       "5424208 chrX  999238    999388   P2RY8      K562      0.0065786972 0.006430348\n",
       "5424209 chrX  999238    999388   PLCXD1     K562      0.0014973300 0.007136451\n",
       "5424210 chrX  999238    999388   PPP2R3B    K562      0.0170103080 0.007203074\n",
       "5424211 chrX  999238    999388   SLC25A6    K562      0.0002857254 0.008257469\n",
       "5424212 chrX  999443    999815   AKAP17A    K562     -0.0053356686 0.007751039\n",
       "5424213 chrX  999443    999815   ASMT       K562     -0.0018603695 0.010417029\n",
       "5424214 chrX  999443    999815   ASMTL      K562     -0.0031390689 0.007901701\n",
       "5424215 chrX  999443    999815   CSF2RA     K562     -0.0058282870 0.013121997\n",
       "5424216 chrX  999443    999815   GTPBP6     K562     -0.0018860789 0.008991638\n",
       "5424217 chrX  999443    999815   IL3RA      K562      0.0102547348 0.008846145\n",
       "5424218 chrX  999443    999815   LINC00106  K562      0.0241222491 0.009084283\n",
       "5424219 chrX  999443    999815   P2RY8      K562      0.0104815495 0.007434975\n",
       "5424220 chrX  999443    999815   PLCXD1     K562     -0.0036885610 0.007951101\n",
       "5424221 chrX  999443    999815   PPP2R3B    K562      0.0161612413 0.009159334\n",
       "5424222 chrX  999443    999815   SLC25A6    K562      0.0131691609 0.008061788\n",
       "5424223 chrX 9995651   9996434   CLCN4      K562      0.0090697954 0.018218238\n",
       "5424224 chrX 9995651   9996434   GPR143     K562      0.0021010659 0.007557823\n",
       "5424225 chrX 9995651   9996434   TBL1X      K562      0.0043513127 0.052325655\n",
       "5424226 chrX 9995651   9996434   WWC3       K562      0.0142355426 0.007449798\n",
       "5424227 chrX 9997538   9997892   CLCN4      K562      0.0169351328 0.009789716\n",
       "5424228 chrX 9997538   9997892   GPR143     K562      0.0147423414 0.006747029\n",
       "5424229 chrX 9997538   9997892   TBL1X      K562     -0.0009891140 0.014828419\n",
       "5424230 chrX 9997538   9997892   WWC3       K562      0.0087043601 0.007004155\n",
       "        rBgMean       pvalZ      \n",
       "1       -0.0015815836 0.24399795 \n",
       "2       -0.0395237640 0.39649275 \n",
       "3        0.0054725617 0.67706934 \n",
       "4       -0.0023297737 0.07515807 \n",
       "5        0.0075227843 0.21204311 \n",
       "6        0.0154815205 0.82558824 \n",
       "7        0.0013105063 0.66240508 \n",
       "8        0.0031270674 0.95784697 \n",
       "9        0.0072043021 0.13720885 \n",
       "10       0.0057421774 0.84574526 \n",
       "11       0.0006598667 0.01654316 \n",
       "12       0.0157497491 0.94652221 \n",
       "13       0.0094359258 0.80180633 \n",
       "14       0.0009090877 0.26330715 \n",
       "15       0.0134717601 0.09752742 \n",
       "16       0.0510578984 0.05443793 \n",
       "17       0.0049608171 0.28391115 \n",
       "18       0.0118986221 0.44453996 \n",
       "19       0.0020523559 0.11725104 \n",
       "20      -0.0135920129 0.55267510 \n",
       "21      -0.0038437107 0.59231166 \n",
       "22       0.0031600301 0.88402149 \n",
       "23       0.0016697050 0.91863345 \n",
       "24       0.0004945195 0.99882940 \n",
       "25       0.0072432756 0.76575294 \n",
       "26       0.0027047660 0.01875724 \n",
       "27       0.0052968957 0.14594557 \n",
       "28       0.0101568919 0.05759012 \n",
       "29       0.0016335324 0.11005263 \n",
       "30      -0.0081728473 0.28617712 \n",
       "⋮       ⋮             ⋮          \n",
       "5424201  0.0030581394 0.952769061\n",
       "5424202 -0.0007724019 0.545866449\n",
       "5424203  0.0022825383 0.603194761\n",
       "5424204  0.0064432526 0.498475582\n",
       "5424205  0.0022619760 0.554782269\n",
       "5424206  0.0024571668 0.441308467\n",
       "5424207  0.0063330011 0.163348562\n",
       "5424208  0.0008736880 0.187485445\n",
       "5424209  0.0007033834 0.455708143\n",
       "5424210 -0.0001724716 0.008528458\n",
       "5424211  0.0014631888 0.556694482\n",
       "5424212  0.0040323523 0.886594563\n",
       "5424213  0.0017406188 0.635209471\n",
       "5424214  0.0027979899 0.773783392\n",
       "5424215  0.0095082030 0.878750203\n",
       "5424216  0.0042383316 0.752103139\n",
       "5424217  0.0038417992 0.234244029\n",
       "5424218  0.0065312650 0.026408159\n",
       "5424219 -0.0001671759 0.076036468\n",
       "5424220  0.0032490256 0.808540769\n",
       "5424221  0.0013372423 0.052781945\n",
       "5424222  0.0029846775 0.103240086\n",
       "5424223  0.0113132275 0.549002713\n",
       "5424224  0.0003748135 0.409665295\n",
       "5424225  0.0382900712 0.741703615\n",
       "5424226  0.0069433467 0.163828339\n",
       "5424227  0.0056133765 0.123739390\n",
       "5424228 -0.0004825961 0.012018433\n",
       "5424229  0.0125049547 0.818592426\n",
       "5424230  0.0041775750 0.259042548"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.9 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   foreach_1.5.2                    \n",
       " [3] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.68.0                  \n",
       " [5] rtracklayer_1.60.1                Biostrings_2.70.1                \n",
       " [7] XVector_0.42.0                    genomation_1.34.0                \n",
       " [9] Seurat_5.0.3                      SeuratObject_5.0.1               \n",
       "[11] sp_2.1-4                          FigR_0.1.0                       \n",
       "[13] motifmatchr_1.24.0                chromVAR_1.24.0                  \n",
       "[15] cowplot_1.1.3                     ggplot2_3.5.1                    \n",
       "[17] dplyr_1.1.4                       SummarizedExperiment_1.30.2      \n",
       "[19] Biobase_2.60.0                    GenomicRanges_1.54.1             \n",
       "[21] GenomeInfoDb_1.38.1               IRanges_2.36.0                   \n",
       "[23] S4Vectors_0.40.2                  BiocGenerics_0.48.1              \n",
       "[25] MatrixGenerics_1.12.3             matrixStats_1.3.0                \n",
       "[27] Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             globals_0.16.3             \n",
       " [15] lattice_0.22-6              MASS_7.3-60                \n",
       " [17] magrittr_2.0.3              plotly_4.10.4              \n",
       " [19] plotrix_3.8-4               yaml_2.3.8                 \n",
       " [21] httpuv_1.6.15               sctransform_0.4.1          \n",
       " [23] spam_2.10-0                 spatstat.sparse_3.0-3      \n",
       " [25] reticulate_1.36.1           pbapply_1.7-2              \n",
       " [27] DBI_1.2.2                   CNEr_1.38.0                \n",
       " [29] RColorBrewer_1.1-3          abind_1.4-5                \n",
       " [31] zlibbioc_1.48.0             Rtsne_0.17                 \n",
       " [33] purrr_1.0.2                 R.utils_2.12.3             \n",
       " [35] RCurl_1.98-1.14             pracma_2.4.4               \n",
       " [37] GenomeInfoDbData_1.2.11     ggrepel_0.9.5              \n",
       " [39] irlba_2.3.5.1               spatstat.utils_3.0-4       \n",
       " [41] listenv_0.9.1               seqLogo_1.68.0             \n",
       " [43] goftest_1.2-3               RSpectra_0.16-1            \n",
       " [45] spatstat.random_3.2-3       annotate_1.80.0            \n",
       " [47] fitdistrplus_1.1-11         parallelly_1.37.1          \n",
       " [49] leiden_0.4.3.1              codetools_0.2-20           \n",
       " [51] DelayedArray_0.26.7         DT_0.33                    \n",
       " [53] tidyselect_1.2.1            spatstat.explore_3.2-6     \n",
       " [55] base64enc_0.1-3             GenomicAlignments_1.36.0   \n",
       " [57] jsonlite_1.8.8              progressr_0.14.0           \n",
       " [59] iterators_1.0.14            ggridges_0.5.6             \n",
       " [61] survival_3.6-4              tools_4.3.3                \n",
       " [63] TFMPvalue_0.0.9             ica_1.0-3                  \n",
       " [65] Rcpp_1.0.12                 glue_1.7.0                 \n",
       " [67] gridExtra_2.3               IRdisplay_1.1              \n",
       " [69] withr_3.0.0                 fastmap_1.1.1              \n",
       " [71] fansi_1.0.6                 caTools_1.18.2             \n",
       " [73] digest_0.6.35               R6_2.5.1                   \n",
       " [75] mime_0.12                   nabor_0.5.0                \n",
       " [77] seqPattern_1.34.0           colorspace_2.1-0           \n",
       " [79] scattermore_1.2             GO.db_3.18.0               \n",
       " [81] tensor_1.5                  gtools_3.9.5               \n",
       " [83] poweRlaw_0.80.0             spatstat.data_3.0-4        \n",
       " [85] RSQLite_2.3.6               R.methodsS3_1.8.2          \n",
       " [87] utf8_1.2.4                  tidyr_1.3.1                \n",
       " [89] generics_0.1.3              data.table_1.15.2          \n",
       " [91] httr_1.4.7                  htmlwidgets_1.6.4          \n",
       " [93] S4Arrays_1.2.1              TFBSTools_1.40.0           \n",
       " [95] uwot_0.1.16                 pkgconfig_2.0.3            \n",
       " [97] gtable_0.3.5                blob_1.2.4                 \n",
       " [99] impute_1.76.0               lmtest_0.9-40              \n",
       "[101] htmltools_0.5.8.1           dotCall64_1.1-1            \n",
       "[103] scales_1.3.0                png_0.1-8                  \n",
       "[105] tzdb_0.4.0                  reshape2_1.4.4             \n",
       "[107] rjson_0.2.21                uuid_1.2-0                 \n",
       "[109] nlme_3.1-164                repr_1.1.7                 \n",
       "[111] cachem_1.0.8                zoo_1.8-12                 \n",
       "[113] stringr_1.5.1               KernSmooth_2.23-22         \n",
       "[115] miniUI_0.1.1.1              AnnotationDbi_1.64.1       \n",
       "[117] restfulr_0.0.15             pillar_1.9.0               \n",
       "[119] vctrs_0.6.5                 RANN_2.6.1                 \n",
       "[121] promises_1.3.0              xtable_1.8-4               \n",
       "[123] cluster_2.1.6               evaluate_0.23              \n",
       "[125] readr_2.1.5                 cli_3.6.2                  \n",
       "[127] compiler_4.3.3              Rsamtools_2.18.0           \n",
       "[129] rlang_1.1.3                 crayon_1.5.2               \n",
       "[131] future.apply_1.11.2         plyr_1.8.9                 \n",
       "[133] stringi_1.8.3               gridBase_0.4-7             \n",
       "[135] deldir_2.0-4                viridisLite_0.4.2          \n",
       "[137] BiocParallel_1.36.0         munsell_0.5.1              \n",
       "[139] lazyeval_0.2.2              spatstat.geom_3.2-9        \n",
       "[141] IRkernel_1.3.2              RcppHNSW_0.6.0             \n",
       "[143] hms_1.1.3                   patchwork_1.2.0            \n",
       "[145] bit64_4.0.5                 future_1.33.2              \n",
       "[147] KEGGREST_1.40.1             shiny_1.8.1.1              \n",
       "[149] ROCR_1.0-11                 igraph_2.0.3               \n",
       "[151] memoise_2.0.1               bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
