{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/PBMC/1.prepare_data/2.merge_peaks.240624/pairs.rds\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/PBMC/1.prepare_data/2.merge_peaks.240624/matrix.atac.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/PBMC/1.prepare_data/2.merge_peaks.240624/matrix.rna.rds\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/PBMC/2.Genome_wide_prediction/FigR/PBMC.250211/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readRDS(path.pairs.E2G)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  4000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  4001  to  5000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  5001  to  6000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  6001  to  7000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  7001  to  8000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  8001  to  9000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  9001  to  10000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  10001  to  11000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  11001  to  12000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  12001  to  13000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  13001  to  14000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  14001  to  15000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  15001  to  16000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  16001  to  17000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  17001  to  18000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  18001  to  18183 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = readRDS(path.matrix.rna_count)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dd80eda-2392-4d59-8250-1ae62d7e1d1e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b9e26c70-11fd-47c4-8e05-9c96b17d3f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a76b1603-6b00-4ac0-9399-fd341c5a7b88",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter = pairs.E2G[pairs.E2G$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                              pairs.E2G$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 303411 18183 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(303411): chr1-10057-10601 chr1-180665-181882 ...\n",
       "  chrX-156029605-156030380 chrX-156030498-156030791\n",
       "rowData names(2): PeakName bias\n",
       "colnames(18183): Multiome10x_05_AAACAGCCAACAACAA-1\n",
       "  Multiome10x_05_AAACAGCCACCGGCTA-1 ...\n",
       "  Multiome10x_06_TTTGTTGGTACCCACC-1 Multiome10x_06_TTTGTTGGTACCGAAC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna.rename),\n",
    "                       index = 1:nrow(matrix.rna.rename))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  1.62410396734873 mins \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Time Elapsed:  2.72668618553215 hours \n",
      "\n",
      "[1] \"chr10\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  20.6637232303619 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  34.871554283301 mins \n",
      "\n",
      "[1] \"chr11\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  51.213508605957 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.38856114142471 hours \n",
      "\n",
      "[1] \"chr12\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  40.7999241352081 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.12655633933014 hours \n",
      "\n",
      "[1] \"chr13\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.45288801193237 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  15.9909170508385 mins \n",
      "\n",
      "[1] \"chr14\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  27.9292759895325 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  41.4936949650447 mins \n",
      "\n",
      "[1] \"chr15\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  21.3922600746155 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  36.9412338336309 mins \n",
      "\n",
      "[1] \"chr16\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  53.7053487300873 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.58649879051579 hours \n",
      "\n",
      "[1] \"chr17\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  1.41813081502914 mins \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  2.41940255827374 hours \n",
      "\n",
      "[1] \"chr18\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  9.4580545425415 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  15.3379411617915 mins \n",
      "\n",
      "[1] \"chr19\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  1.70074468453725 mins \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  2.91259224149916 hours \n",
      "\n",
      "[1] \"chr2\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  53.0857503414154 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.46560393604967 hours \n",
      "\n",
      "[1] \"chr20\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  31.7677569389343 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  48.7622462550799 mins \n",
      "\n",
      "[1] \"chr21\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.6729509830475 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  17.2570799946785 mins \n",
      "\n",
      "[1] \"chr22\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  26.776967048645 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  46.8155753175418 mins \n",
      "\n",
      "[1] \"chr3\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  43.6323647499084 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.18416850209236 hours \n",
      "\n",
      "[1] \"chr4\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  20.2219109535217 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  31.721953757604 mins \n",
      "\n",
      "[1] \"chr5\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  26.191844701767 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  43.9351766109467 mins \n",
      "\n",
      "[1] \"chr6\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  48.7394027709961 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.33622198647923 hours \n",
      "\n",
      "[1] \"chr7\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  36.1888520717621 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.01919773942894 hours \n",
      "\n",
      "[1] \"chr8\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  25.0318477153778 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  38.5877778967222 mins \n",
      "\n",
      "[1] \"chr9\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  31.7370851039886 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  57.1004153966904 mins \n",
      "\n",
      "[1] \"chrX\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  18.2109022140503 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  30.9652296702067 mins \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter[seqnames(pairs.E2G.filter) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 5657294 ranges and 7 metadata columns:\n",
       "       seqnames              ranges strand |  TargetGene               PeakName\n",
       "          <Rle>           <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 100028331-100029533      * |         AGL chr1-100028331-10002..\n",
       "  chr1     chr1 100028331-100029533      * |      CDC14A chr1-100028331-10002..\n",
       "  chr1     chr1 100028331-100029533      * |         DBT chr1-100028331-10002..\n",
       "  chr1     chr1 100028331-100029533      * |        DPH5 chr1-100028331-10002..\n",
       "  chr1     chr1 100028331-100029533      * |       EXTL2 chr1-100028331-10002..\n",
       "   ...      ...                 ...    ... .         ...                    ...\n",
       "  chrX     chrX     9995651-9996434      * |        WWC3   chrX-9995651-9996434\n",
       "  chrX     chrX     9997538-9997892      * |       CLCN4   chrX-9997538-9997892\n",
       "  chrX     chrX     9997538-9997892      * |      GPR143   chrX-9997538-9997892\n",
       "  chrX     chrX     9997538-9997892      * |       TBL1X   chrX-9997538-9997892\n",
       "  chrX     chrX     9997538-9997892      * |        WWC3   chrX-9997538-9997892\n",
       "                     PairName        rObs      rBgSD     rBgMean     pvalZ\n",
       "                  <character>   <numeric>  <numeric>   <numeric> <numeric>\n",
       "  chr1 chr1-100028331-10002.. -0.00501817 0.00798765 0.002179933  0.816247\n",
       "  chr1 chr1-100028331-10002..  0.00369559 0.03090626 0.032197281  0.821787\n",
       "  chr1 chr1-100028331-10002..  0.00911630 0.00743671 0.003032279  0.206648\n",
       "  chr1 chr1-100028331-10002.. -0.00265255 0.00756198 0.000536637  0.663392\n",
       "  chr1 chr1-100028331-10002.. -0.00453344 0.02126859 0.024508055  0.913946\n",
       "   ...                    ...         ...        ...         ...       ...\n",
       "  chrX chrX-9995651-9996434..  0.02315685 0.00815983 0.005565846 0.0155494\n",
       "  chrX chrX-9997538-9997892.. -0.00559375 0.00769795 0.000576932 0.7886082\n",
       "  chrX chrX-9997538-9997892..  0.01433572 0.00760240 0.002221890 0.0555329\n",
       "  chrX chrX-9997538-9997892..  0.01355366 0.00727943 0.003156343 0.0766007\n",
       "  chrX chrX-9997538-9997892..  0.00278404 0.00857366 0.002714148 0.4967480\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5657294 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>AGL     </td><td>K562</td><td>-0.0050181653</td><td>0.007987651</td><td> 2.179933e-03</td><td>0.8162465867</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>CDC14A  </td><td>K562</td><td> 0.0036955917</td><td>0.030906259</td><td> 3.219728e-02</td><td>0.8217873344</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>DBT     </td><td>K562</td><td> 0.0091162973</td><td>0.007436708</td><td> 3.032279e-03</td><td>0.2066482430</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>DPH5    </td><td>K562</td><td>-0.0026525468</td><td>0.007561983</td><td> 5.366374e-04</td><td>0.6633922762</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>EXTL2   </td><td>K562</td><td>-0.0045334379</td><td>0.021268590</td><td> 2.450805e-02</td><td>0.9139463611</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>FRRS1   </td><td>K562</td><td>-0.0015313644</td><td>0.008081235</td><td> 2.561305e-04</td><td>0.5875280610</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>GPR88   </td><td>K562</td><td>-0.0015313644</td><td>0.009889356</td><td> 1.510757e-03</td><td>0.6208126240</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>MFSD14A </td><td>K562</td><td> 0.0451657425</td><td>0.011669937</td><td> 2.753077e-03</td><td>0.0001393399</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>RTCA    </td><td>K562</td><td>-0.0046274763</td><td>0.049646276</td><td> 4.758904e-02</td><td>0.8535477197</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>RTCA-AS1</td><td>K562</td><td>-0.0040522778</td><td>0.006600259</td><td>-6.558806e-04</td><td>0.6965786414</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>SASS6   </td><td>K562</td><td>-0.0017179241</td><td>0.010781323</td><td> 1.135754e-02</td><td>0.8873946946</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>SLC30A7 </td><td>K562</td><td> 0.0058553821</td><td>0.007615087</td><td> 1.405448e-03</td><td>0.2794898730</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>SLC35A3 </td><td>K562</td><td>-0.0010976998</td><td>0.007743991</td><td> 3.530910e-03</td><td>0.7249810905</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028331</td><td>100029533</td><td>TRMT13  </td><td>K562</td><td> 0.0147316463</td><td>0.011725586</td><td> 1.588729e-02</td><td>0.5392551258</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>AGL     </td><td>K562</td><td> 0.0108488630</td><td>0.007530731</td><td> 9.014075e-05</td><td>0.0765535181</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>CDC14A  </td><td>K562</td><td>-0.0247834430</td><td>0.017895107</td><td>-1.598203e-02</td><td>0.6885815690</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>DBT     </td><td>K562</td><td>-0.0004113671</td><td>0.006252183</td><td>-8.240457e-04</td><td>0.4736867161</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>DPH5    </td><td>K562</td><td> 0.0008235060</td><td>0.006577198</td><td>-5.204511e-04</td><td>0.4190455338</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>EXTL2   </td><td>K562</td><td>-0.0287062466</td><td>0.012808807</td><td>-8.190748e-03</td><td>0.9453858393</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>FRRS1   </td><td>K562</td><td> 0.0032652636</td><td>0.007182956</td><td> 1.516581e-03</td><td>0.4038288347</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>GPR88   </td><td>K562</td><td>-0.0024677366</td><td>0.005486828</td><td> 1.825377e-03</td><td>0.7830219925</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>MFSD14A </td><td>K562</td><td> 0.0086466358</td><td>0.007234104</td><td> 2.621171e-03</td><td>0.2024435648</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>RTCA    </td><td>K562</td><td>-0.0674599341</td><td>0.029702721</td><td>-2.969564e-02</td><td>0.8982083415</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>RTCA-AS1</td><td>K562</td><td> 0.0019094735</td><td>0.007093550</td><td> 3.784006e-03</td><td>0.6042097264</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>SASS6   </td><td>K562</td><td>-0.0003141682</td><td>0.007207255</td><td> 2.031387e-03</td><td>0.6275773539</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>SLC30A7 </td><td>K562</td><td> 0.0011427521</td><td>0.007279983</td><td> 3.174989e-03</td><td>0.6099367031</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>SLC35A3 </td><td>K562</td><td>-0.0041140843</td><td>0.007153400</td><td> 3.346893e-03</td><td>0.8515252488</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037542</td><td>100039303</td><td>TRMT13  </td><td>K562</td><td>-0.0208109820</td><td>0.009896219</td><td>-1.263894e-05</td><td>0.9822078226</td></tr>\n",
       "\t<tr><td>chr1</td><td>100044514</td><td>100044850</td><td>AGL     </td><td>K562</td><td>-0.0034716827</td><td>0.006949563</td><td> 4.058359e-04</td><td>0.7115612360</td></tr>\n",
       "\t<tr><td>chr1</td><td>100044514</td><td>100044850</td><td>CDC14A  </td><td>K562</td><td> 0.0024280150</td><td>0.010695842</td><td> 5.557945e-03</td><td>0.6150977142</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>AKAP17A  </td><td>K562</td><td>-0.0033856289</td><td>1.117241e-02</td><td> 4.083875e-03</td><td>0.74811401</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>ASMT     </td><td>K562</td><td>-0.0033440352</td><td>9.933152e-03</td><td> 2.797580e-03</td><td>0.73180943</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>ASMTL    </td><td>K562</td><td>-0.0019457644</td><td>7.424325e-03</td><td> 1.906809e-04</td><td>0.61323588</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>CSF2RA   </td><td>K562</td><td>-0.0017474199</td><td>8.992056e-03</td><td> 9.836667e-04</td><td>0.61933022</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>GTPBP6   </td><td>K562</td><td>-0.0016942688</td><td>4.832382e-03</td><td>-9.103320e-04</td><td>0.56443597</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>IL3RA    </td><td>K562</td><td>-0.0004765451</td><td>3.951421e-05</td><td>-4.491850e-04</td><td>0.75566009</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>LINC00106</td><td>K562</td><td>-0.0020823636</td><td>1.137080e-02</td><td> 2.202803e-03</td><td>0.64686015</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>P2RY8    </td><td>K562</td><td> 0.0010005886</td><td>8.659052e-03</td><td> 1.578235e-03</td><td>0.52659374</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>PLCXD1   </td><td>K562</td><td>-0.0016942689</td><td>8.112355e-03</td><td> 4.436169e-04</td><td>0.60393070</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>PPP2R3B  </td><td>K562</td><td>-0.0002130940</td><td>1.766935e-05</td><td>-2.008595e-04</td><td>0.75566009</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999238</td><td> 999388</td><td>SLC25A6  </td><td>K562</td><td>-0.0006028368</td><td>4.998609e-05</td><td>-5.682260e-04</td><td>0.75566009</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>AKAP17A  </td><td>K562</td><td> 0.0068970667</td><td>8.691860e-03</td><td> 2.612956e-03</td><td>0.31104598</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>ASMT     </td><td>K562</td><td>-0.0047311317</td><td>9.099562e-03</td><td> 3.386014e-03</td><td>0.81381344</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>ASMTL    </td><td>K562</td><td>-0.0027528622</td><td>9.206964e-03</td><td> 1.581830e-03</td><td>0.68111027</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>CSF2RA   </td><td>K562</td><td>-0.0024722449</td><td>7.853285e-03</td><td> 9.390831e-04</td><td>0.66799457</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>GTPBP6   </td><td>K562</td><td>-0.0023970469</td><td>7.360123e-03</td><td> 3.040663e-04</td><td>0.64318786</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>IL3RA    </td><td>K562</td><td>-0.0006742146</td><td>7.466152e-03</td><td> 3.854896e-05</td><td>0.53802765</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>LINC00106</td><td>K562</td><td>-0.0029461223</td><td>9.752139e-03</td><td> 1.756233e-03</td><td>0.68516346</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>P2RY8    </td><td>K562</td><td>-0.0042022663</td><td>8.282583e-03</td><td> 1.120951e-03</td><td>0.73979066</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>PLCXD1   </td><td>K562</td><td>-0.0023970469</td><td>6.773332e-03</td><td>-4.517879e-04</td><td>0.61301803</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>PPP2R3B  </td><td>K562</td><td>-0.0003014848</td><td>2.841732e-05</td><td>-3.167470e-04</td><td>0.29560808</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999443</td><td> 999815</td><td>SLC25A6  </td><td>K562</td><td> 0.0636364375</td><td>8.039179e-05</td><td>-8.960683e-04</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995651</td><td>9996434</td><td>CLCN4    </td><td>K562</td><td>-0.0008446387</td><td>7.235586e-03</td><td> 3.267752e-03</td><td>0.71510346</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995651</td><td>9996434</td><td>GPR143   </td><td>K562</td><td> 0.0007496020</td><td>7.362814e-03</td><td> 4.433983e-03</td><td>0.69160467</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995651</td><td>9996434</td><td>TBL1X    </td><td>K562</td><td> 0.0106753380</td><td>8.085883e-03</td><td> 5.943217e-03</td><td>0.27919575</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995651</td><td>9996434</td><td>WWC3     </td><td>K562</td><td> 0.0231568496</td><td>8.159827e-03</td><td> 5.565846e-03</td><td>0.01554941</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997538</td><td>9997892</td><td>CLCN4    </td><td>K562</td><td>-0.0055937534</td><td>7.697949e-03</td><td> 5.769322e-04</td><td>0.78860818</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997538</td><td>9997892</td><td>GPR143   </td><td>K562</td><td> 0.0143357176</td><td>7.602404e-03</td><td> 2.221890e-03</td><td>0.05553292</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997538</td><td>9997892</td><td>TBL1X    </td><td>K562</td><td> 0.0135536632</td><td>7.279433e-03</td><td> 3.156343e-03</td><td>0.07660068</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997538</td><td>9997892</td><td>WWC3     </td><td>K562</td><td> 0.0027840381</td><td>8.573662e-03</td><td> 2.714148e-03</td><td>0.49674796</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5657294 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100028331 & 100029533 & AGL      & K562 & -0.0050181653 & 0.007987651 &  2.179933e-03 & 0.8162465867\\\\\n",
       "\t chr1 & 100028331 & 100029533 & CDC14A   & K562 &  0.0036955917 & 0.030906259 &  3.219728e-02 & 0.8217873344\\\\\n",
       "\t chr1 & 100028331 & 100029533 & DBT      & K562 &  0.0091162973 & 0.007436708 &  3.032279e-03 & 0.2066482430\\\\\n",
       "\t chr1 & 100028331 & 100029533 & DPH5     & K562 & -0.0026525468 & 0.007561983 &  5.366374e-04 & 0.6633922762\\\\\n",
       "\t chr1 & 100028331 & 100029533 & EXTL2    & K562 & -0.0045334379 & 0.021268590 &  2.450805e-02 & 0.9139463611\\\\\n",
       "\t chr1 & 100028331 & 100029533 & FRRS1    & K562 & -0.0015313644 & 0.008081235 &  2.561305e-04 & 0.5875280610\\\\\n",
       "\t chr1 & 100028331 & 100029533 & GPR88    & K562 & -0.0015313644 & 0.009889356 &  1.510757e-03 & 0.6208126240\\\\\n",
       "\t chr1 & 100028331 & 100029533 & MFSD14A  & K562 &  0.0451657425 & 0.011669937 &  2.753077e-03 & 0.0001393399\\\\\n",
       "\t chr1 & 100028331 & 100029533 & RTCA     & K562 & -0.0046274763 & 0.049646276 &  4.758904e-02 & 0.8535477197\\\\\n",
       "\t chr1 & 100028331 & 100029533 & RTCA-AS1 & K562 & -0.0040522778 & 0.006600259 & -6.558806e-04 & 0.6965786414\\\\\n",
       "\t chr1 & 100028331 & 100029533 & SASS6    & K562 & -0.0017179241 & 0.010781323 &  1.135754e-02 & 0.8873946946\\\\\n",
       "\t chr1 & 100028331 & 100029533 & SLC30A7  & K562 &  0.0058553821 & 0.007615087 &  1.405448e-03 & 0.2794898730\\\\\n",
       "\t chr1 & 100028331 & 100029533 & SLC35A3  & K562 & -0.0010976998 & 0.007743991 &  3.530910e-03 & 0.7249810905\\\\\n",
       "\t chr1 & 100028331 & 100029533 & TRMT13   & K562 &  0.0147316463 & 0.011725586 &  1.588729e-02 & 0.5392551258\\\\\n",
       "\t chr1 & 100037542 & 100039303 & AGL      & K562 &  0.0108488630 & 0.007530731 &  9.014075e-05 & 0.0765535181\\\\\n",
       "\t chr1 & 100037542 & 100039303 & CDC14A   & K562 & -0.0247834430 & 0.017895107 & -1.598203e-02 & 0.6885815690\\\\\n",
       "\t chr1 & 100037542 & 100039303 & DBT      & K562 & -0.0004113671 & 0.006252183 & -8.240457e-04 & 0.4736867161\\\\\n",
       "\t chr1 & 100037542 & 100039303 & DPH5     & K562 &  0.0008235060 & 0.006577198 & -5.204511e-04 & 0.4190455338\\\\\n",
       "\t chr1 & 100037542 & 100039303 & EXTL2    & K562 & -0.0287062466 & 0.012808807 & -8.190748e-03 & 0.9453858393\\\\\n",
       "\t chr1 & 100037542 & 100039303 & FRRS1    & K562 &  0.0032652636 & 0.007182956 &  1.516581e-03 & 0.4038288347\\\\\n",
       "\t chr1 & 100037542 & 100039303 & GPR88    & K562 & -0.0024677366 & 0.005486828 &  1.825377e-03 & 0.7830219925\\\\\n",
       "\t chr1 & 100037542 & 100039303 & MFSD14A  & K562 &  0.0086466358 & 0.007234104 &  2.621171e-03 & 0.2024435648\\\\\n",
       "\t chr1 & 100037542 & 100039303 & RTCA     & K562 & -0.0674599341 & 0.029702721 & -2.969564e-02 & 0.8982083415\\\\\n",
       "\t chr1 & 100037542 & 100039303 & RTCA-AS1 & K562 &  0.0019094735 & 0.007093550 &  3.784006e-03 & 0.6042097264\\\\\n",
       "\t chr1 & 100037542 & 100039303 & SASS6    & K562 & -0.0003141682 & 0.007207255 &  2.031387e-03 & 0.6275773539\\\\\n",
       "\t chr1 & 100037542 & 100039303 & SLC30A7  & K562 &  0.0011427521 & 0.007279983 &  3.174989e-03 & 0.6099367031\\\\\n",
       "\t chr1 & 100037542 & 100039303 & SLC35A3  & K562 & -0.0041140843 & 0.007153400 &  3.346893e-03 & 0.8515252488\\\\\n",
       "\t chr1 & 100037542 & 100039303 & TRMT13   & K562 & -0.0208109820 & 0.009896219 & -1.263894e-05 & 0.9822078226\\\\\n",
       "\t chr1 & 100044514 & 100044850 & AGL      & K562 & -0.0034716827 & 0.006949563 &  4.058359e-04 & 0.7115612360\\\\\n",
       "\t chr1 & 100044514 & 100044850 & CDC14A   & K562 &  0.0024280150 & 0.010695842 &  5.557945e-03 & 0.6150977142\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  999238 &  999388 & AKAP17A   & K562 & -0.0033856289 & 1.117241e-02 &  4.083875e-03 & 0.74811401\\\\\n",
       "\t chrX &  999238 &  999388 & ASMT      & K562 & -0.0033440352 & 9.933152e-03 &  2.797580e-03 & 0.73180943\\\\\n",
       "\t chrX &  999238 &  999388 & ASMTL     & K562 & -0.0019457644 & 7.424325e-03 &  1.906809e-04 & 0.61323588\\\\\n",
       "\t chrX &  999238 &  999388 & CSF2RA    & K562 & -0.0017474199 & 8.992056e-03 &  9.836667e-04 & 0.61933022\\\\\n",
       "\t chrX &  999238 &  999388 & GTPBP6    & K562 & -0.0016942688 & 4.832382e-03 & -9.103320e-04 & 0.56443597\\\\\n",
       "\t chrX &  999238 &  999388 & IL3RA     & K562 & -0.0004765451 & 3.951421e-05 & -4.491850e-04 & 0.75566009\\\\\n",
       "\t chrX &  999238 &  999388 & LINC00106 & K562 & -0.0020823636 & 1.137080e-02 &  2.202803e-03 & 0.64686015\\\\\n",
       "\t chrX &  999238 &  999388 & P2RY8     & K562 &  0.0010005886 & 8.659052e-03 &  1.578235e-03 & 0.52659374\\\\\n",
       "\t chrX &  999238 &  999388 & PLCXD1    & K562 & -0.0016942689 & 8.112355e-03 &  4.436169e-04 & 0.60393070\\\\\n",
       "\t chrX &  999238 &  999388 & PPP2R3B   & K562 & -0.0002130940 & 1.766935e-05 & -2.008595e-04 & 0.75566009\\\\\n",
       "\t chrX &  999238 &  999388 & SLC25A6   & K562 & -0.0006028368 & 4.998609e-05 & -5.682260e-04 & 0.75566009\\\\\n",
       "\t chrX &  999443 &  999815 & AKAP17A   & K562 &  0.0068970667 & 8.691860e-03 &  2.612956e-03 & 0.31104598\\\\\n",
       "\t chrX &  999443 &  999815 & ASMT      & K562 & -0.0047311317 & 9.099562e-03 &  3.386014e-03 & 0.81381344\\\\\n",
       "\t chrX &  999443 &  999815 & ASMTL     & K562 & -0.0027528622 & 9.206964e-03 &  1.581830e-03 & 0.68111027\\\\\n",
       "\t chrX &  999443 &  999815 & CSF2RA    & K562 & -0.0024722449 & 7.853285e-03 &  9.390831e-04 & 0.66799457\\\\\n",
       "\t chrX &  999443 &  999815 & GTPBP6    & K562 & -0.0023970469 & 7.360123e-03 &  3.040663e-04 & 0.64318786\\\\\n",
       "\t chrX &  999443 &  999815 & IL3RA     & K562 & -0.0006742146 & 7.466152e-03 &  3.854896e-05 & 0.53802765\\\\\n",
       "\t chrX &  999443 &  999815 & LINC00106 & K562 & -0.0029461223 & 9.752139e-03 &  1.756233e-03 & 0.68516346\\\\\n",
       "\t chrX &  999443 &  999815 & P2RY8     & K562 & -0.0042022663 & 8.282583e-03 &  1.120951e-03 & 0.73979066\\\\\n",
       "\t chrX &  999443 &  999815 & PLCXD1    & K562 & -0.0023970469 & 6.773332e-03 & -4.517879e-04 & 0.61301803\\\\\n",
       "\t chrX &  999443 &  999815 & PPP2R3B   & K562 & -0.0003014848 & 2.841732e-05 & -3.167470e-04 & 0.29560808\\\\\n",
       "\t chrX &  999443 &  999815 & SLC25A6   & K562 &  0.0636364375 & 8.039179e-05 & -8.960683e-04 & 0.00000000\\\\\n",
       "\t chrX & 9995651 & 9996434 & CLCN4     & K562 & -0.0008446387 & 7.235586e-03 &  3.267752e-03 & 0.71510346\\\\\n",
       "\t chrX & 9995651 & 9996434 & GPR143    & K562 &  0.0007496020 & 7.362814e-03 &  4.433983e-03 & 0.69160467\\\\\n",
       "\t chrX & 9995651 & 9996434 & TBL1X     & K562 &  0.0106753380 & 8.085883e-03 &  5.943217e-03 & 0.27919575\\\\\n",
       "\t chrX & 9995651 & 9996434 & WWC3      & K562 &  0.0231568496 & 8.159827e-03 &  5.565846e-03 & 0.01554941\\\\\n",
       "\t chrX & 9997538 & 9997892 & CLCN4     & K562 & -0.0055937534 & 7.697949e-03 &  5.769322e-04 & 0.78860818\\\\\n",
       "\t chrX & 9997538 & 9997892 & GPR143    & K562 &  0.0143357176 & 7.602404e-03 &  2.221890e-03 & 0.05553292\\\\\n",
       "\t chrX & 9997538 & 9997892 & TBL1X     & K562 &  0.0135536632 & 7.279433e-03 &  3.156343e-03 & 0.07660068\\\\\n",
       "\t chrX & 9997538 & 9997892 & WWC3      & K562 &  0.0027840381 & 8.573662e-03 &  2.714148e-03 & 0.49674796\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5657294 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100028331 | 100029533 | AGL      | K562 | -0.0050181653 | 0.007987651 |  2.179933e-03 | 0.8162465867 |\n",
       "| chr1 | 100028331 | 100029533 | CDC14A   | K562 |  0.0036955917 | 0.030906259 |  3.219728e-02 | 0.8217873344 |\n",
       "| chr1 | 100028331 | 100029533 | DBT      | K562 |  0.0091162973 | 0.007436708 |  3.032279e-03 | 0.2066482430 |\n",
       "| chr1 | 100028331 | 100029533 | DPH5     | K562 | -0.0026525468 | 0.007561983 |  5.366374e-04 | 0.6633922762 |\n",
       "| chr1 | 100028331 | 100029533 | EXTL2    | K562 | -0.0045334379 | 0.021268590 |  2.450805e-02 | 0.9139463611 |\n",
       "| chr1 | 100028331 | 100029533 | FRRS1    | K562 | -0.0015313644 | 0.008081235 |  2.561305e-04 | 0.5875280610 |\n",
       "| chr1 | 100028331 | 100029533 | GPR88    | K562 | -0.0015313644 | 0.009889356 |  1.510757e-03 | 0.6208126240 |\n",
       "| chr1 | 100028331 | 100029533 | MFSD14A  | K562 |  0.0451657425 | 0.011669937 |  2.753077e-03 | 0.0001393399 |\n",
       "| chr1 | 100028331 | 100029533 | RTCA     | K562 | -0.0046274763 | 0.049646276 |  4.758904e-02 | 0.8535477197 |\n",
       "| chr1 | 100028331 | 100029533 | RTCA-AS1 | K562 | -0.0040522778 | 0.006600259 | -6.558806e-04 | 0.6965786414 |\n",
       "| chr1 | 100028331 | 100029533 | SASS6    | K562 | -0.0017179241 | 0.010781323 |  1.135754e-02 | 0.8873946946 |\n",
       "| chr1 | 100028331 | 100029533 | SLC30A7  | K562 |  0.0058553821 | 0.007615087 |  1.405448e-03 | 0.2794898730 |\n",
       "| chr1 | 100028331 | 100029533 | SLC35A3  | K562 | -0.0010976998 | 0.007743991 |  3.530910e-03 | 0.7249810905 |\n",
       "| chr1 | 100028331 | 100029533 | TRMT13   | K562 |  0.0147316463 | 0.011725586 |  1.588729e-02 | 0.5392551258 |\n",
       "| chr1 | 100037542 | 100039303 | AGL      | K562 |  0.0108488630 | 0.007530731 |  9.014075e-05 | 0.0765535181 |\n",
       "| chr1 | 100037542 | 100039303 | CDC14A   | K562 | -0.0247834430 | 0.017895107 | -1.598203e-02 | 0.6885815690 |\n",
       "| chr1 | 100037542 | 100039303 | DBT      | K562 | -0.0004113671 | 0.006252183 | -8.240457e-04 | 0.4736867161 |\n",
       "| chr1 | 100037542 | 100039303 | DPH5     | K562 |  0.0008235060 | 0.006577198 | -5.204511e-04 | 0.4190455338 |\n",
       "| chr1 | 100037542 | 100039303 | EXTL2    | K562 | -0.0287062466 | 0.012808807 | -8.190748e-03 | 0.9453858393 |\n",
       "| chr1 | 100037542 | 100039303 | FRRS1    | K562 |  0.0032652636 | 0.007182956 |  1.516581e-03 | 0.4038288347 |\n",
       "| chr1 | 100037542 | 100039303 | GPR88    | K562 | -0.0024677366 | 0.005486828 |  1.825377e-03 | 0.7830219925 |\n",
       "| chr1 | 100037542 | 100039303 | MFSD14A  | K562 |  0.0086466358 | 0.007234104 |  2.621171e-03 | 0.2024435648 |\n",
       "| chr1 | 100037542 | 100039303 | RTCA     | K562 | -0.0674599341 | 0.029702721 | -2.969564e-02 | 0.8982083415 |\n",
       "| chr1 | 100037542 | 100039303 | RTCA-AS1 | K562 |  0.0019094735 | 0.007093550 |  3.784006e-03 | 0.6042097264 |\n",
       "| chr1 | 100037542 | 100039303 | SASS6    | K562 | -0.0003141682 | 0.007207255 |  2.031387e-03 | 0.6275773539 |\n",
       "| chr1 | 100037542 | 100039303 | SLC30A7  | K562 |  0.0011427521 | 0.007279983 |  3.174989e-03 | 0.6099367031 |\n",
       "| chr1 | 100037542 | 100039303 | SLC35A3  | K562 | -0.0041140843 | 0.007153400 |  3.346893e-03 | 0.8515252488 |\n",
       "| chr1 | 100037542 | 100039303 | TRMT13   | K562 | -0.0208109820 | 0.009896219 | -1.263894e-05 | 0.9822078226 |\n",
       "| chr1 | 100044514 | 100044850 | AGL      | K562 | -0.0034716827 | 0.006949563 |  4.058359e-04 | 0.7115612360 |\n",
       "| chr1 | 100044514 | 100044850 | CDC14A   | K562 |  0.0024280150 | 0.010695842 |  5.557945e-03 | 0.6150977142 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  999238 |  999388 | AKAP17A   | K562 | -0.0033856289 | 1.117241e-02 |  4.083875e-03 | 0.74811401 |\n",
       "| chrX |  999238 |  999388 | ASMT      | K562 | -0.0033440352 | 9.933152e-03 |  2.797580e-03 | 0.73180943 |\n",
       "| chrX |  999238 |  999388 | ASMTL     | K562 | -0.0019457644 | 7.424325e-03 |  1.906809e-04 | 0.61323588 |\n",
       "| chrX |  999238 |  999388 | CSF2RA    | K562 | -0.0017474199 | 8.992056e-03 |  9.836667e-04 | 0.61933022 |\n",
       "| chrX |  999238 |  999388 | GTPBP6    | K562 | -0.0016942688 | 4.832382e-03 | -9.103320e-04 | 0.56443597 |\n",
       "| chrX |  999238 |  999388 | IL3RA     | K562 | -0.0004765451 | 3.951421e-05 | -4.491850e-04 | 0.75566009 |\n",
       "| chrX |  999238 |  999388 | LINC00106 | K562 | -0.0020823636 | 1.137080e-02 |  2.202803e-03 | 0.64686015 |\n",
       "| chrX |  999238 |  999388 | P2RY8     | K562 |  0.0010005886 | 8.659052e-03 |  1.578235e-03 | 0.52659374 |\n",
       "| chrX |  999238 |  999388 | PLCXD1    | K562 | -0.0016942689 | 8.112355e-03 |  4.436169e-04 | 0.60393070 |\n",
       "| chrX |  999238 |  999388 | PPP2R3B   | K562 | -0.0002130940 | 1.766935e-05 | -2.008595e-04 | 0.75566009 |\n",
       "| chrX |  999238 |  999388 | SLC25A6   | K562 | -0.0006028368 | 4.998609e-05 | -5.682260e-04 | 0.75566009 |\n",
       "| chrX |  999443 |  999815 | AKAP17A   | K562 |  0.0068970667 | 8.691860e-03 |  2.612956e-03 | 0.31104598 |\n",
       "| chrX |  999443 |  999815 | ASMT      | K562 | -0.0047311317 | 9.099562e-03 |  3.386014e-03 | 0.81381344 |\n",
       "| chrX |  999443 |  999815 | ASMTL     | K562 | -0.0027528622 | 9.206964e-03 |  1.581830e-03 | 0.68111027 |\n",
       "| chrX |  999443 |  999815 | CSF2RA    | K562 | -0.0024722449 | 7.853285e-03 |  9.390831e-04 | 0.66799457 |\n",
       "| chrX |  999443 |  999815 | GTPBP6    | K562 | -0.0023970469 | 7.360123e-03 |  3.040663e-04 | 0.64318786 |\n",
       "| chrX |  999443 |  999815 | IL3RA     | K562 | -0.0006742146 | 7.466152e-03 |  3.854896e-05 | 0.53802765 |\n",
       "| chrX |  999443 |  999815 | LINC00106 | K562 | -0.0029461223 | 9.752139e-03 |  1.756233e-03 | 0.68516346 |\n",
       "| chrX |  999443 |  999815 | P2RY8     | K562 | -0.0042022663 | 8.282583e-03 |  1.120951e-03 | 0.73979066 |\n",
       "| chrX |  999443 |  999815 | PLCXD1    | K562 | -0.0023970469 | 6.773332e-03 | -4.517879e-04 | 0.61301803 |\n",
       "| chrX |  999443 |  999815 | PPP2R3B   | K562 | -0.0003014848 | 2.841732e-05 | -3.167470e-04 | 0.29560808 |\n",
       "| chrX |  999443 |  999815 | SLC25A6   | K562 |  0.0636364375 | 8.039179e-05 | -8.960683e-04 | 0.00000000 |\n",
       "| chrX | 9995651 | 9996434 | CLCN4     | K562 | -0.0008446387 | 7.235586e-03 |  3.267752e-03 | 0.71510346 |\n",
       "| chrX | 9995651 | 9996434 | GPR143    | K562 |  0.0007496020 | 7.362814e-03 |  4.433983e-03 | 0.69160467 |\n",
       "| chrX | 9995651 | 9996434 | TBL1X     | K562 |  0.0106753380 | 8.085883e-03 |  5.943217e-03 | 0.27919575 |\n",
       "| chrX | 9995651 | 9996434 | WWC3      | K562 |  0.0231568496 | 8.159827e-03 |  5.565846e-03 | 0.01554941 |\n",
       "| chrX | 9997538 | 9997892 | CLCN4     | K562 | -0.0055937534 | 7.697949e-03 |  5.769322e-04 | 0.78860818 |\n",
       "| chrX | 9997538 | 9997892 | GPR143    | K562 |  0.0143357176 | 7.602404e-03 |  2.221890e-03 | 0.05553292 |\n",
       "| chrX | 9997538 | 9997892 | TBL1X     | K562 |  0.0135536632 | 7.279433e-03 |  3.156343e-03 | 0.07660068 |\n",
       "| chrX | 9997538 | 9997892 | WWC3      | K562 |  0.0027840381 | 8.573662e-03 |  2.714148e-03 | 0.49674796 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType rObs          rBgSD       \n",
       "1       chr1 100028331 100029533 AGL        K562     -0.0050181653 0.007987651 \n",
       "2       chr1 100028331 100029533 CDC14A     K562      0.0036955917 0.030906259 \n",
       "3       chr1 100028331 100029533 DBT        K562      0.0091162973 0.007436708 \n",
       "4       chr1 100028331 100029533 DPH5       K562     -0.0026525468 0.007561983 \n",
       "5       chr1 100028331 100029533 EXTL2      K562     -0.0045334379 0.021268590 \n",
       "6       chr1 100028331 100029533 FRRS1      K562     -0.0015313644 0.008081235 \n",
       "7       chr1 100028331 100029533 GPR88      K562     -0.0015313644 0.009889356 \n",
       "8       chr1 100028331 100029533 MFSD14A    K562      0.0451657425 0.011669937 \n",
       "9       chr1 100028331 100029533 RTCA       K562     -0.0046274763 0.049646276 \n",
       "10      chr1 100028331 100029533 RTCA-AS1   K562     -0.0040522778 0.006600259 \n",
       "11      chr1 100028331 100029533 SASS6      K562     -0.0017179241 0.010781323 \n",
       "12      chr1 100028331 100029533 SLC30A7    K562      0.0058553821 0.007615087 \n",
       "13      chr1 100028331 100029533 SLC35A3    K562     -0.0010976998 0.007743991 \n",
       "14      chr1 100028331 100029533 TRMT13     K562      0.0147316463 0.011725586 \n",
       "15      chr1 100037542 100039303 AGL        K562      0.0108488630 0.007530731 \n",
       "16      chr1 100037542 100039303 CDC14A     K562     -0.0247834430 0.017895107 \n",
       "17      chr1 100037542 100039303 DBT        K562     -0.0004113671 0.006252183 \n",
       "18      chr1 100037542 100039303 DPH5       K562      0.0008235060 0.006577198 \n",
       "19      chr1 100037542 100039303 EXTL2      K562     -0.0287062466 0.012808807 \n",
       "20      chr1 100037542 100039303 FRRS1      K562      0.0032652636 0.007182956 \n",
       "21      chr1 100037542 100039303 GPR88      K562     -0.0024677366 0.005486828 \n",
       "22      chr1 100037542 100039303 MFSD14A    K562      0.0086466358 0.007234104 \n",
       "23      chr1 100037542 100039303 RTCA       K562     -0.0674599341 0.029702721 \n",
       "24      chr1 100037542 100039303 RTCA-AS1   K562      0.0019094735 0.007093550 \n",
       "25      chr1 100037542 100039303 SASS6      K562     -0.0003141682 0.007207255 \n",
       "26      chr1 100037542 100039303 SLC30A7    K562      0.0011427521 0.007279983 \n",
       "27      chr1 100037542 100039303 SLC35A3    K562     -0.0041140843 0.007153400 \n",
       "28      chr1 100037542 100039303 TRMT13     K562     -0.0208109820 0.009896219 \n",
       "29      chr1 100044514 100044850 AGL        K562     -0.0034716827 0.006949563 \n",
       "30      chr1 100044514 100044850 CDC14A     K562      0.0024280150 0.010695842 \n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮           \n",
       "5657265 chrX  999238    999388   AKAP17A    K562     -0.0033856289 1.117241e-02\n",
       "5657266 chrX  999238    999388   ASMT       K562     -0.0033440352 9.933152e-03\n",
       "5657267 chrX  999238    999388   ASMTL      K562     -0.0019457644 7.424325e-03\n",
       "5657268 chrX  999238    999388   CSF2RA     K562     -0.0017474199 8.992056e-03\n",
       "5657269 chrX  999238    999388   GTPBP6     K562     -0.0016942688 4.832382e-03\n",
       "5657270 chrX  999238    999388   IL3RA      K562     -0.0004765451 3.951421e-05\n",
       "5657271 chrX  999238    999388   LINC00106  K562     -0.0020823636 1.137080e-02\n",
       "5657272 chrX  999238    999388   P2RY8      K562      0.0010005886 8.659052e-03\n",
       "5657273 chrX  999238    999388   PLCXD1     K562     -0.0016942689 8.112355e-03\n",
       "5657274 chrX  999238    999388   PPP2R3B    K562     -0.0002130940 1.766935e-05\n",
       "5657275 chrX  999238    999388   SLC25A6    K562     -0.0006028368 4.998609e-05\n",
       "5657276 chrX  999443    999815   AKAP17A    K562      0.0068970667 8.691860e-03\n",
       "5657277 chrX  999443    999815   ASMT       K562     -0.0047311317 9.099562e-03\n",
       "5657278 chrX  999443    999815   ASMTL      K562     -0.0027528622 9.206964e-03\n",
       "5657279 chrX  999443    999815   CSF2RA     K562     -0.0024722449 7.853285e-03\n",
       "5657280 chrX  999443    999815   GTPBP6     K562     -0.0023970469 7.360123e-03\n",
       "5657281 chrX  999443    999815   IL3RA      K562     -0.0006742146 7.466152e-03\n",
       "5657282 chrX  999443    999815   LINC00106  K562     -0.0029461223 9.752139e-03\n",
       "5657283 chrX  999443    999815   P2RY8      K562     -0.0042022663 8.282583e-03\n",
       "5657284 chrX  999443    999815   PLCXD1     K562     -0.0023970469 6.773332e-03\n",
       "5657285 chrX  999443    999815   PPP2R3B    K562     -0.0003014848 2.841732e-05\n",
       "5657286 chrX  999443    999815   SLC25A6    K562      0.0636364375 8.039179e-05\n",
       "5657287 chrX 9995651   9996434   CLCN4      K562     -0.0008446387 7.235586e-03\n",
       "5657288 chrX 9995651   9996434   GPR143     K562      0.0007496020 7.362814e-03\n",
       "5657289 chrX 9995651   9996434   TBL1X      K562      0.0106753380 8.085883e-03\n",
       "5657290 chrX 9995651   9996434   WWC3       K562      0.0231568496 8.159827e-03\n",
       "5657291 chrX 9997538   9997892   CLCN4      K562     -0.0055937534 7.697949e-03\n",
       "5657292 chrX 9997538   9997892   GPR143     K562      0.0143357176 7.602404e-03\n",
       "5657293 chrX 9997538   9997892   TBL1X      K562      0.0135536632 7.279433e-03\n",
       "5657294 chrX 9997538   9997892   WWC3       K562      0.0027840381 8.573662e-03\n",
       "        rBgMean       pvalZ       \n",
       "1        2.179933e-03 0.8162465867\n",
       "2        3.219728e-02 0.8217873344\n",
       "3        3.032279e-03 0.2066482430\n",
       "4        5.366374e-04 0.6633922762\n",
       "5        2.450805e-02 0.9139463611\n",
       "6        2.561305e-04 0.5875280610\n",
       "7        1.510757e-03 0.6208126240\n",
       "8        2.753077e-03 0.0001393399\n",
       "9        4.758904e-02 0.8535477197\n",
       "10      -6.558806e-04 0.6965786414\n",
       "11       1.135754e-02 0.8873946946\n",
       "12       1.405448e-03 0.2794898730\n",
       "13       3.530910e-03 0.7249810905\n",
       "14       1.588729e-02 0.5392551258\n",
       "15       9.014075e-05 0.0765535181\n",
       "16      -1.598203e-02 0.6885815690\n",
       "17      -8.240457e-04 0.4736867161\n",
       "18      -5.204511e-04 0.4190455338\n",
       "19      -8.190748e-03 0.9453858393\n",
       "20       1.516581e-03 0.4038288347\n",
       "21       1.825377e-03 0.7830219925\n",
       "22       2.621171e-03 0.2024435648\n",
       "23      -2.969564e-02 0.8982083415\n",
       "24       3.784006e-03 0.6042097264\n",
       "25       2.031387e-03 0.6275773539\n",
       "26       3.174989e-03 0.6099367031\n",
       "27       3.346893e-03 0.8515252488\n",
       "28      -1.263894e-05 0.9822078226\n",
       "29       4.058359e-04 0.7115612360\n",
       "30       5.557945e-03 0.6150977142\n",
       "⋮       ⋮             ⋮           \n",
       "5657265  4.083875e-03 0.74811401  \n",
       "5657266  2.797580e-03 0.73180943  \n",
       "5657267  1.906809e-04 0.61323588  \n",
       "5657268  9.836667e-04 0.61933022  \n",
       "5657269 -9.103320e-04 0.56443597  \n",
       "5657270 -4.491850e-04 0.75566009  \n",
       "5657271  2.202803e-03 0.64686015  \n",
       "5657272  1.578235e-03 0.52659374  \n",
       "5657273  4.436169e-04 0.60393070  \n",
       "5657274 -2.008595e-04 0.75566009  \n",
       "5657275 -5.682260e-04 0.75566009  \n",
       "5657276  2.612956e-03 0.31104598  \n",
       "5657277  3.386014e-03 0.81381344  \n",
       "5657278  1.581830e-03 0.68111027  \n",
       "5657279  9.390831e-04 0.66799457  \n",
       "5657280  3.040663e-04 0.64318786  \n",
       "5657281  3.854896e-05 0.53802765  \n",
       "5657282  1.756233e-03 0.68516346  \n",
       "5657283  1.120951e-03 0.73979066  \n",
       "5657284 -4.517879e-04 0.61301803  \n",
       "5657285 -3.167470e-04 0.29560808  \n",
       "5657286 -8.960683e-04 0.00000000  \n",
       "5657287  3.267752e-03 0.71510346  \n",
       "5657288  4.433983e-03 0.69160467  \n",
       "5657289  5.943217e-03 0.27919575  \n",
       "5657290  5.565846e-03 0.01554941  \n",
       "5657291  5.769322e-04 0.78860818  \n",
       "5657292  2.221890e-03 0.05553292  \n",
       "5657293  3.156343e-03 0.07660068  \n",
       "5657294  2.714148e-03 0.49674796  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   data.table_1.15.2                \n",
       " [3] foreach_1.5.2                     BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] genomation_1.34.0                 Seurat_5.0.3                     \n",
       "[11] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[13] FigR_0.1.0                        motifmatchr_1.24.0               \n",
       "[15] chromVAR_1.24.0                   cowplot_1.1.3                    \n",
       "[17] ggplot2_3.5.1                     dplyr_1.1.4                      \n",
       "[19] SummarizedExperiment_1.30.2       Biobase_2.60.0                   \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               MatrixGenerics_1.12.3            \n",
       "[27] matrixStats_1.3.0                 Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             globals_0.16.3             \n",
       " [15] lattice_0.22-6              MASS_7.3-60                \n",
       " [17] magrittr_2.0.3              plotly_4.10.4              \n",
       " [19] plotrix_3.8-4               yaml_2.3.8                 \n",
       " [21] httpuv_1.6.15               sctransform_0.4.1          \n",
       " [23] spam_2.10-0                 spatstat.sparse_3.0-3      \n",
       " [25] reticulate_1.36.1           pbapply_1.7-2              \n",
       " [27] DBI_1.2.2                   CNEr_1.38.0                \n",
       " [29] RColorBrewer_1.1-3          abind_1.4-5                \n",
       " [31] zlibbioc_1.48.0             Rtsne_0.17                 \n",
       " [33] purrr_1.0.2                 R.utils_2.12.3             \n",
       " [35] RCurl_1.98-1.14             pracma_2.4.4               \n",
       " [37] GenomeInfoDbData_1.2.11     ggrepel_0.9.5              \n",
       " [39] irlba_2.3.5.1               spatstat.utils_3.0-4       \n",
       " [41] listenv_0.9.1               seqLogo_1.68.0             \n",
       " [43] goftest_1.2-3               RSpectra_0.16-1            \n",
       " [45] spatstat.random_3.2-3       annotate_1.80.0            \n",
       " [47] fitdistrplus_1.1-11         parallelly_1.37.1          \n",
       " [49] leiden_0.4.3.1              codetools_0.2-20           \n",
       " [51] DelayedArray_0.26.7         DT_0.33                    \n",
       " [53] tidyselect_1.2.1            spatstat.explore_3.2-6     \n",
       " [55] base64enc_0.1-3             GenomicAlignments_1.36.0   \n",
       " [57] jsonlite_1.8.8              progressr_0.14.0           \n",
       " [59] iterators_1.0.14            ggridges_0.5.6             \n",
       " [61] survival_3.6-4              tools_4.3.3                \n",
       " [63] TFMPvalue_0.0.9             ica_1.0-3                  \n",
       " [65] Rcpp_1.0.12                 glue_1.7.0                 \n",
       " [67] gridExtra_2.3               IRdisplay_1.1              \n",
       " [69] withr_3.0.0                 fastmap_1.1.1              \n",
       " [71] fansi_1.0.6                 caTools_1.18.2             \n",
       " [73] digest_0.6.35               R6_2.5.1                   \n",
       " [75] mime_0.12                   nabor_0.5.0                \n",
       " [77] seqPattern_1.34.0           colorspace_2.1-0           \n",
       " [79] scattermore_1.2             GO.db_3.18.0               \n",
       " [81] tensor_1.5                  gtools_3.9.5               \n",
       " [83] poweRlaw_0.80.0             spatstat.data_3.0-4        \n",
       " [85] RSQLite_2.3.6               R.methodsS3_1.8.2          \n",
       " [87] utf8_1.2.4                  tidyr_1.3.1                \n",
       " [89] generics_0.1.3              httr_1.4.7                 \n",
       " [91] htmlwidgets_1.6.4           S4Arrays_1.2.1             \n",
       " [93] TFBSTools_1.40.0            uwot_0.1.16                \n",
       " [95] pkgconfig_2.0.3             gtable_0.3.5               \n",
       " [97] blob_1.2.4                  impute_1.76.0              \n",
       " [99] lmtest_0.9-40               htmltools_0.5.8.1          \n",
       "[101] dotCall64_1.1-1             scales_1.3.0               \n",
       "[103] png_0.1-8                   tzdb_0.4.0                 \n",
       "[105] reshape2_1.4.4              rjson_0.2.21               \n",
       "[107] uuid_1.2-0                  nlme_3.1-164               \n",
       "[109] repr_1.1.7                  cachem_1.0.8               \n",
       "[111] zoo_1.8-12                  stringr_1.5.1              \n",
       "[113] KernSmooth_2.23-22          miniUI_0.1.1.1             \n",
       "[115] AnnotationDbi_1.64.1        restfulr_0.0.15            \n",
       "[117] pillar_1.9.0                vctrs_0.6.5                \n",
       "[119] RANN_2.6.1                  promises_1.3.0             \n",
       "[121] xtable_1.8-4                cluster_2.1.6              \n",
       "[123] evaluate_0.23               readr_2.1.5                \n",
       "[125] cli_3.6.2                   compiler_4.3.3             \n",
       "[127] Rsamtools_2.18.0            rlang_1.1.3                \n",
       "[129] crayon_1.5.2                future.apply_1.11.2        \n",
       "[131] plyr_1.8.9                  stringi_1.8.3              \n",
       "[133] gridBase_0.4-7              deldir_2.0-4               \n",
       "[135] viridisLite_0.4.2           BiocParallel_1.36.0        \n",
       "[137] munsell_0.5.1               lazyeval_0.2.2             \n",
       "[139] spatstat.geom_3.2-9         IRkernel_1.3.2             \n",
       "[141] RcppHNSW_0.6.0              hms_1.1.3                  \n",
       "[143] patchwork_1.2.0             bit64_4.0.5                \n",
       "[145] future_1.33.2               KEGGREST_1.40.1            \n",
       "[147] shiny_1.8.1.1               ROCR_1.0-11                \n",
       "[149] igraph_2.0.3                memoise_2.0.1              \n",
       "[151] bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
