{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“replacing previous import ‘Hmisc::capitalize’ by ‘R.utils::capitalize’ when loading ‘SCENT’”\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SCENT)\n",
    "library(Signac)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(parallel)\n",
    "library(data.table)\n",
    "library(dplyr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/SCENT/BMMC5.Myeloid.2500.250324/pairs.E2G.res.rds\"\n",
    "path.pairs.E2G.reuse = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/SCENT/BMMC5.Myeloid.25000.240612/pairs.E2G.res.rds\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_BMMC5/Myeloid/Kendall/atac_matrix.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/3.merge_cell_type.240522/RNA_count_matrix/Myeloid.rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/SCENT/BMMC5.Myeloid.25000.250329/\"\n",
    "celltype = \"Myeloid\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 48"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readRDS(path.pairs.E2G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f80cb711-6490-4d95-8710-c71b7287efa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.reuse = readRDS(path.pairs.E2G.reuse)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.atac = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = BinarizeCounts(matrix.atac)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna = read.csv(path.matrix.rna_count,\n",
    "                      row.names = 1,\n",
    "                      check.names = F)\n",
    "matrix.rna = Matrix(as.matrix(matrix.rna), sparse = TRUE)\n",
    "matrix.rna = matrix.rna[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna[rowSums(matrix.rna) > 0,]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62718ef8-a3bd-4755-a95f-be216f16517b",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0f255a65-03a4-4255-967e-0c644bfe66cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ca760678-9db9-418e-b7ac-29984eddad7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare SCENT meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "827059dd-6885-441e-973e-d7c49e5fd534",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 14048 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>cell</th><th scope=col>nUMI</th><th scope=col>nMito</th><th scope=col>celltype</th><th scope=col>log.nUMI</th><th scope=col>percent.mito</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>s1d1_CTATGGCCATAACGGG-1</th><td>s1d1_CTATGGCCATAACGGG-1</td><td> 2157</td><td> 69</td><td>Myeloid</td><td> 7.676474</td><td>0.03198887</td></tr>\n",
       "\t<tr><th scope=row>s1d1_ACCACATAGGTGTCCA-1</th><td>s1d1_ACCACATAGGTGTCCA-1</td><td> 2128</td><td> 64</td><td>Myeloid</td><td> 7.662938</td><td>0.03007519</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TGCCGGTAGTCGCAAT-1</th><td>s1d1_TGCCGGTAGTCGCAAT-1</td><td> 1920</td><td> 67</td><td>Myeloid</td><td> 7.560080</td><td>0.03489583</td></tr>\n",
       "\t<tr><th scope=row>s1d1_CATCCCAGTCTTTATG-1</th><td>s1d1_CATCCCAGTCTTTATG-1</td><td> 3211</td><td>127</td><td>Myeloid</td><td> 8.074338</td><td>0.03955154</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TAGTGAGAGCTGCACA-1</th><td>s1d1_TAGTGAGAGCTGCACA-1</td><td> 3568</td><td> 85</td><td>Myeloid</td><td> 8.179760</td><td>0.02382287</td></tr>\n",
       "\t<tr><th scope=row>s1d1_CCTGTTGGTTATAGCG-1</th><td>s1d1_CCTGTTGGTTATAGCG-1</td><td> 4044</td><td> 42</td><td>Myeloid</td><td> 8.304990</td><td>0.01038576</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TTTAGCTTCTGGCTTC-1</th><td>s1d1_TTTAGCTTCTGGCTTC-1</td><td> 2728</td><td>105</td><td>Myeloid</td><td> 7.911324</td><td>0.03848974</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TGTGAAACAGGATTAA-1</th><td>s1d1_TGTGAAACAGGATTAA-1</td><td> 2800</td><td> 86</td><td>Myeloid</td><td> 7.937375</td><td>0.03071429</td></tr>\n",
       "\t<tr><th scope=row>s1d1_ACCAGGCTCTTGGATA-1</th><td>s1d1_ACCAGGCTCTTGGATA-1</td><td> 1015</td><td> 94</td><td>Myeloid</td><td> 6.922644</td><td>0.09261084</td></tr>\n",
       "\t<tr><th scope=row>s1d1_CACAGGGAGACAGGTA-1</th><td>s1d1_CACAGGGAGACAGGTA-1</td><td> 3092</td><td> 69</td><td>Myeloid</td><td> 8.036573</td><td>0.02231565</td></tr>\n",
       "\t<tr><th scope=row>s1d1_GCATCCTTCGCTATGG-1</th><td>s1d1_GCATCCTTCGCTATGG-1</td><td> 3112</td><td> 77</td><td>Myeloid</td><td> 8.043021</td><td>0.02474293</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TGCTATGCAGAGGGAG-1</th><td>s1d1_TGCTATGCAGAGGGAG-1</td><td> 2814</td><td> 78</td><td>Myeloid</td><td> 7.942362</td><td>0.02771855</td></tr>\n",
       "\t<tr><th scope=row>s1d1_CCCGCAACAAAGCGCA-1</th><td>s1d1_CCCGCAACAAAGCGCA-1</td><td> 3234</td><td> 58</td><td>Myeloid</td><td> 8.081475</td><td>0.01793445</td></tr>\n",
       "\t<tr><th scope=row>s1d1_AATCTTGAGAAGCTAG-1</th><td>s1d1_AATCTTGAGAAGCTAG-1</td><td> 1647</td><td> 67</td><td>Myeloid</td><td> 7.406711</td><td>0.04068002</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TTGTCCATCATTAGGC-1</th><td>s1d1_TTGTCCATCATTAGGC-1</td><td> 4110</td><td> 79</td><td>Myeloid</td><td> 8.321178</td><td>0.01922141</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TCCAGGATCCTAACGG-1</th><td>s1d1_TCCAGGATCCTAACGG-1</td><td> 3484</td><td>100</td><td>Myeloid</td><td> 8.155936</td><td>0.02870264</td></tr>\n",
       "\t<tr><th scope=row>s1d1_GGTATGTTCTAACCAA-1</th><td>s1d1_GGTATGTTCTAACCAA-1</td><td> 2759</td><td> 61</td><td>Myeloid</td><td> 7.922624</td><td>0.02210946</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TGGATTCAGTTTCCGC-1</th><td>s1d1_TGGATTCAGTTTCCGC-1</td><td> 3639</td><td> 72</td><td>Myeloid</td><td> 8.199464</td><td>0.01978566</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TGCTTGCTCAATCTCT-1</th><td>s1d1_TGCTTGCTCAATCTCT-1</td><td> 1934</td><td> 71</td><td>Myeloid</td><td> 7.567346</td><td>0.03671148</td></tr>\n",
       "\t<tr><th scope=row>s1d1_CGATGTCCATACCCGG-1</th><td>s1d1_CGATGTCCATACCCGG-1</td><td> 3296</td><td> 54</td><td>Myeloid</td><td> 8.100465</td><td>0.01638350</td></tr>\n",
       "\t<tr><th scope=row>s1d1_ACGCCTAAGCGCCTTT-1</th><td>s1d1_ACGCCTAAGCGCCTTT-1</td><td> 3236</td><td> 92</td><td>Myeloid</td><td> 8.082093</td><td>0.02843016</td></tr>\n",
       "\t<tr><th scope=row>s1d1_GAGTAATAGCTTAACA-1</th><td>s1d1_GAGTAATAGCTTAACA-1</td><td> 4308</td><td>100</td><td>Myeloid</td><td> 8.368229</td><td>0.02321263</td></tr>\n",
       "\t<tr><th scope=row>s1d1_ATTGGCTAGGGACCTC-1</th><td>s1d1_ATTGGCTAGGGACCTC-1</td><td>22348</td><td>863</td><td>Myeloid</td><td>10.014492</td><td>0.03861643</td></tr>\n",
       "\t<tr><th scope=row>s1d1_CCTGTAACATTGTGCA-1</th><td>s1d1_CCTGTAACATTGTGCA-1</td><td> 3940</td><td>127</td><td>Myeloid</td><td> 8.278936</td><td>0.03223350</td></tr>\n",
       "\t<tr><th scope=row>s1d1_AGATGAAGTGAGGTGA-1</th><td>s1d1_AGATGAAGTGAGGTGA-1</td><td> 2271</td><td> 69</td><td>Myeloid</td><td> 7.727976</td><td>0.03038309</td></tr>\n",
       "\t<tr><th scope=row>s1d1_GTTAACGGTAGCTGGT-1</th><td>s1d1_GTTAACGGTAGCTGGT-1</td><td> 1222</td><td> 72</td><td>Myeloid</td><td> 7.108244</td><td>0.05891980</td></tr>\n",
       "\t<tr><th scope=row>s1d1_GAACACAAGTAACCAC-1</th><td>s1d1_GAACACAAGTAACCAC-1</td><td> 3170</td><td> 64</td><td>Myeloid</td><td> 8.061487</td><td>0.02018927</td></tr>\n",
       "\t<tr><th scope=row>s1d1_GGTGCTGGTACTTCAC-1</th><td>s1d1_GGTGCTGGTACTTCAC-1</td><td> 4133</td><td>103</td><td>Myeloid</td><td> 8.326759</td><td>0.02492136</td></tr>\n",
       "\t<tr><th scope=row>s1d1_AGCTGCTCATTGTTGG-1</th><td>s1d1_AGCTGCTCATTGTTGG-1</td><td> 1308</td><td>102</td><td>Myeloid</td><td> 7.176255</td><td>0.07798165</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TCCAGGATCAACAAGG-1</th><td>s1d1_TCCAGGATCAACAAGG-1</td><td> 3218</td><td> 66</td><td>Myeloid</td><td> 8.076515</td><td>0.02050963</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GTTGCTGAGTAATCCA-1</th><td>s4d9_GTTGCTGAGTAATCCA-1</td><td>2839</td><td>168</td><td>Myeloid</td><td>7.951207</td><td>0.05917577</td></tr>\n",
       "\t<tr><th scope=row>s4d9_AACTAGTGTTACTTCA-1</th><td>s4d9_AACTAGTGTTACTTCA-1</td><td>1455</td><td>198</td><td>Myeloid</td><td>7.282761</td><td>0.13608247</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TGCTTCCAGTCCTTTG-1</th><td>s4d9_TGCTTCCAGTCCTTTG-1</td><td>1450</td><td>138</td><td>Myeloid</td><td>7.279319</td><td>0.09517241</td></tr>\n",
       "\t<tr><th scope=row>s4d9_AAGGATCCACAGGATG-1</th><td>s4d9_AAGGATCCACAGGATG-1</td><td>1789</td><td>229</td><td>Myeloid</td><td>7.489412</td><td>0.12800447</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TGTGCGCAGCCGCAGT-1</th><td>s4d9_TGTGCGCAGCCGCAGT-1</td><td>1686</td><td>138</td><td>Myeloid</td><td>7.430114</td><td>0.08185053</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CCAGACTCAATCATGT-1</th><td>s4d9_CCAGACTCAATCATGT-1</td><td>1834</td><td> 52</td><td>Myeloid</td><td>7.514255</td><td>0.02835333</td></tr>\n",
       "\t<tr><th scope=row>s4d9_AGTAAGTAGTTAGAGG-1</th><td>s4d9_AGTAAGTAGTTAGAGG-1</td><td>1826</td><td>194</td><td>Myeloid</td><td>7.509883</td><td>0.10624315</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TGCGGTTCAACAGCCT-1</th><td>s4d9_TGCGGTTCAACAGCCT-1</td><td>1048</td><td>124</td><td>Myeloid</td><td>6.954639</td><td>0.11832061</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GATGACTTCGCTAGTG-1</th><td>s4d9_GATGACTTCGCTAGTG-1</td><td> 848</td><td> 94</td><td>Myeloid</td><td>6.742881</td><td>0.11084906</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CTAGTCGAGCTATTGA-1</th><td>s4d9_CTAGTCGAGCTATTGA-1</td><td> 829</td><td> 52</td><td>Myeloid</td><td>6.720220</td><td>0.06272618</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TTATGACCATTTAAGC-1</th><td>s4d9_TTATGACCATTTAAGC-1</td><td>3534</td><td>336</td><td>Myeloid</td><td>8.170186</td><td>0.09507640</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GTGCTCAAGCAAGATG-1</th><td>s4d9_GTGCTCAAGCAAGATG-1</td><td>3857</td><td>573</td><td>Myeloid</td><td>8.257645</td><td>0.14856106</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GGATATTGTGTCACGG-1</th><td>s4d9_GGATATTGTGTCACGG-1</td><td>2140</td><td>285</td><td>Myeloid</td><td>7.668561</td><td>0.13317757</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TGGCCTGCAATTATGC-1</th><td>s4d9_TGGCCTGCAATTATGC-1</td><td>3184</td><td>332</td><td>Myeloid</td><td>8.065894</td><td>0.10427136</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GAGGGAGCATAACGGG-1</th><td>s4d9_GAGGGAGCATAACGGG-1</td><td>2488</td><td>358</td><td>Myeloid</td><td>7.819234</td><td>0.14389068</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GGTACAAAGTTGCCTC-1</th><td>s4d9_GGTACAAAGTTGCCTC-1</td><td>2427</td><td>276</td><td>Myeloid</td><td>7.794411</td><td>0.11372064</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GGAAACTCATAGCTGC-1</th><td>s4d9_GGAAACTCATAGCTGC-1</td><td>1561</td><td>182</td><td>Myeloid</td><td>7.353082</td><td>0.11659193</td></tr>\n",
       "\t<tr><th scope=row>s4d9_AGAAAGGCAATGCCCG-1</th><td>s4d9_AGAAAGGCAATGCCCG-1</td><td>1773</td><td> 38</td><td>Myeloid</td><td>7.480428</td><td>0.02143260</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CTACAACAGCGAGTAA-1</th><td>s4d9_CTACAACAGCGAGTAA-1</td><td>1161</td><td> 96</td><td>Myeloid</td><td>7.057037</td><td>0.08268734</td></tr>\n",
       "\t<tr><th scope=row>s4d9_ACATTGCAGGGTCTAT-1</th><td>s4d9_ACATTGCAGGGTCTAT-1</td><td>2255</td><td>255</td><td>Myeloid</td><td>7.720905</td><td>0.11308204</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CTCCAACTCTAACCAA-1</th><td>s4d9_CTCCAACTCTAACCAA-1</td><td>3392</td><td>399</td><td>Myeloid</td><td>8.129175</td><td>0.11762972</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CAGGGCTTCCGCCAAA-1</th><td>s4d9_CAGGGCTTCCGCCAAA-1</td><td>1837</td><td>253</td><td>Myeloid</td><td>7.515889</td><td>0.13772455</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TAAACAGCACCTAAGC-1</th><td>s4d9_TAAACAGCACCTAAGC-1</td><td>1474</td><td>171</td><td>Myeloid</td><td>7.295735</td><td>0.11601085</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TGGTCCTTCAGTGAAG-1</th><td>s4d9_TGGTCCTTCAGTGAAG-1</td><td>2021</td><td>227</td><td>Myeloid</td><td>7.611348</td><td>0.11232063</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GGATACTTCTTACTCG-1</th><td>s4d9_GGATACTTCTTACTCG-1</td><td>2211</td><td>192</td><td>Myeloid</td><td>7.701200</td><td>0.08683853</td></tr>\n",
       "\t<tr><th scope=row>s4d9_ACAGGAATCCGTCCAT-1</th><td>s4d9_ACAGGAATCCGTCCAT-1</td><td> 747</td><td> 61</td><td>Myeloid</td><td>6.616065</td><td>0.08165997</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TGTGGCCAGCTGGCTA-1</th><td>s4d9_TGTGGCCAGCTGGCTA-1</td><td>1062</td><td>140</td><td>Myeloid</td><td>6.967909</td><td>0.13182674</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CCGGTTAAGCCTGTTC-1</th><td>s4d9_CCGGTTAAGCCTGTTC-1</td><td>1496</td><td> 58</td><td>Myeloid</td><td>7.310550</td><td>0.03877005</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CAAGCTAGTCAACAAT-1</th><td>s4d9_CAAGCTAGTCAACAAT-1</td><td>1664</td><td>179</td><td>Myeloid</td><td>7.416980</td><td>0.10757212</td></tr>\n",
       "\t<tr><th scope=row>s4d9_ACAGGTAAGGTTACGT-1</th><td>s4d9_ACAGGTAAGGTTACGT-1</td><td>1221</td><td> 75</td><td>Myeloid</td><td>7.107425</td><td>0.06142506</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 14048 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & cell & nUMI & nMito & celltype & log.nUMI & percent.mito\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\ts1d1\\_CTATGGCCATAACGGG-1 & s1d1\\_CTATGGCCATAACGGG-1 &  2157 &  69 & Myeloid &  7.676474 & 0.03198887\\\\\n",
       "\ts1d1\\_ACCACATAGGTGTCCA-1 & s1d1\\_ACCACATAGGTGTCCA-1 &  2128 &  64 & Myeloid &  7.662938 & 0.03007519\\\\\n",
       "\ts1d1\\_TGCCGGTAGTCGCAAT-1 & s1d1\\_TGCCGGTAGTCGCAAT-1 &  1920 &  67 & Myeloid &  7.560080 & 0.03489583\\\\\n",
       "\ts1d1\\_CATCCCAGTCTTTATG-1 & s1d1\\_CATCCCAGTCTTTATG-1 &  3211 & 127 & Myeloid &  8.074338 & 0.03955154\\\\\n",
       "\ts1d1\\_TAGTGAGAGCTGCACA-1 & s1d1\\_TAGTGAGAGCTGCACA-1 &  3568 &  85 & Myeloid &  8.179760 & 0.02382287\\\\\n",
       "\ts1d1\\_CCTGTTGGTTATAGCG-1 & s1d1\\_CCTGTTGGTTATAGCG-1 &  4044 &  42 & Myeloid &  8.304990 & 0.01038576\\\\\n",
       "\ts1d1\\_TTTAGCTTCTGGCTTC-1 & s1d1\\_TTTAGCTTCTGGCTTC-1 &  2728 & 105 & Myeloid &  7.911324 & 0.03848974\\\\\n",
       "\ts1d1\\_TGTGAAACAGGATTAA-1 & s1d1\\_TGTGAAACAGGATTAA-1 &  2800 &  86 & Myeloid &  7.937375 & 0.03071429\\\\\n",
       "\ts1d1\\_ACCAGGCTCTTGGATA-1 & s1d1\\_ACCAGGCTCTTGGATA-1 &  1015 &  94 & Myeloid &  6.922644 & 0.09261084\\\\\n",
       "\ts1d1\\_CACAGGGAGACAGGTA-1 & s1d1\\_CACAGGGAGACAGGTA-1 &  3092 &  69 & Myeloid &  8.036573 & 0.02231565\\\\\n",
       "\ts1d1\\_GCATCCTTCGCTATGG-1 & s1d1\\_GCATCCTTCGCTATGG-1 &  3112 &  77 & Myeloid &  8.043021 & 0.02474293\\\\\n",
       "\ts1d1\\_TGCTATGCAGAGGGAG-1 & s1d1\\_TGCTATGCAGAGGGAG-1 &  2814 &  78 & Myeloid &  7.942362 & 0.02771855\\\\\n",
       "\ts1d1\\_CCCGCAACAAAGCGCA-1 & s1d1\\_CCCGCAACAAAGCGCA-1 &  3234 &  58 & Myeloid &  8.081475 & 0.01793445\\\\\n",
       "\ts1d1\\_AATCTTGAGAAGCTAG-1 & s1d1\\_AATCTTGAGAAGCTAG-1 &  1647 &  67 & Myeloid &  7.406711 & 0.04068002\\\\\n",
       "\ts1d1\\_TTGTCCATCATTAGGC-1 & s1d1\\_TTGTCCATCATTAGGC-1 &  4110 &  79 & Myeloid &  8.321178 & 0.01922141\\\\\n",
       "\ts1d1\\_TCCAGGATCCTAACGG-1 & s1d1\\_TCCAGGATCCTAACGG-1 &  3484 & 100 & Myeloid &  8.155936 & 0.02870264\\\\\n",
       "\ts1d1\\_GGTATGTTCTAACCAA-1 & s1d1\\_GGTATGTTCTAACCAA-1 &  2759 &  61 & Myeloid &  7.922624 & 0.02210946\\\\\n",
       "\ts1d1\\_TGGATTCAGTTTCCGC-1 & s1d1\\_TGGATTCAGTTTCCGC-1 &  3639 &  72 & Myeloid &  8.199464 & 0.01978566\\\\\n",
       "\ts1d1\\_TGCTTGCTCAATCTCT-1 & s1d1\\_TGCTTGCTCAATCTCT-1 &  1934 &  71 & Myeloid &  7.567346 & 0.03671148\\\\\n",
       "\ts1d1\\_CGATGTCCATACCCGG-1 & s1d1\\_CGATGTCCATACCCGG-1 &  3296 &  54 & Myeloid &  8.100465 & 0.01638350\\\\\n",
       "\ts1d1\\_ACGCCTAAGCGCCTTT-1 & s1d1\\_ACGCCTAAGCGCCTTT-1 &  3236 &  92 & Myeloid &  8.082093 & 0.02843016\\\\\n",
       "\ts1d1\\_GAGTAATAGCTTAACA-1 & s1d1\\_GAGTAATAGCTTAACA-1 &  4308 & 100 & Myeloid &  8.368229 & 0.02321263\\\\\n",
       "\ts1d1\\_ATTGGCTAGGGACCTC-1 & s1d1\\_ATTGGCTAGGGACCTC-1 & 22348 & 863 & Myeloid & 10.014492 & 0.03861643\\\\\n",
       "\ts1d1\\_CCTGTAACATTGTGCA-1 & s1d1\\_CCTGTAACATTGTGCA-1 &  3940 & 127 & Myeloid &  8.278936 & 0.03223350\\\\\n",
       "\ts1d1\\_AGATGAAGTGAGGTGA-1 & s1d1\\_AGATGAAGTGAGGTGA-1 &  2271 &  69 & Myeloid &  7.727976 & 0.03038309\\\\\n",
       "\ts1d1\\_GTTAACGGTAGCTGGT-1 & s1d1\\_GTTAACGGTAGCTGGT-1 &  1222 &  72 & Myeloid &  7.108244 & 0.05891980\\\\\n",
       "\ts1d1\\_GAACACAAGTAACCAC-1 & s1d1\\_GAACACAAGTAACCAC-1 &  3170 &  64 & Myeloid &  8.061487 & 0.02018927\\\\\n",
       "\ts1d1\\_GGTGCTGGTACTTCAC-1 & s1d1\\_GGTGCTGGTACTTCAC-1 &  4133 & 103 & Myeloid &  8.326759 & 0.02492136\\\\\n",
       "\ts1d1\\_AGCTGCTCATTGTTGG-1 & s1d1\\_AGCTGCTCATTGTTGG-1 &  1308 & 102 & Myeloid &  7.176255 & 0.07798165\\\\\n",
       "\ts1d1\\_TCCAGGATCAACAAGG-1 & s1d1\\_TCCAGGATCAACAAGG-1 &  3218 &  66 & Myeloid &  8.076515 & 0.02050963\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\ts4d9\\_GTTGCTGAGTAATCCA-1 & s4d9\\_GTTGCTGAGTAATCCA-1 & 2839 & 168 & Myeloid & 7.951207 & 0.05917577\\\\\n",
       "\ts4d9\\_AACTAGTGTTACTTCA-1 & s4d9\\_AACTAGTGTTACTTCA-1 & 1455 & 198 & Myeloid & 7.282761 & 0.13608247\\\\\n",
       "\ts4d9\\_TGCTTCCAGTCCTTTG-1 & s4d9\\_TGCTTCCAGTCCTTTG-1 & 1450 & 138 & Myeloid & 7.279319 & 0.09517241\\\\\n",
       "\ts4d9\\_AAGGATCCACAGGATG-1 & s4d9\\_AAGGATCCACAGGATG-1 & 1789 & 229 & Myeloid & 7.489412 & 0.12800447\\\\\n",
       "\ts4d9\\_TGTGCGCAGCCGCAGT-1 & s4d9\\_TGTGCGCAGCCGCAGT-1 & 1686 & 138 & Myeloid & 7.430114 & 0.08185053\\\\\n",
       "\ts4d9\\_CCAGACTCAATCATGT-1 & s4d9\\_CCAGACTCAATCATGT-1 & 1834 &  52 & Myeloid & 7.514255 & 0.02835333\\\\\n",
       "\ts4d9\\_AGTAAGTAGTTAGAGG-1 & s4d9\\_AGTAAGTAGTTAGAGG-1 & 1826 & 194 & Myeloid & 7.509883 & 0.10624315\\\\\n",
       "\ts4d9\\_TGCGGTTCAACAGCCT-1 & s4d9\\_TGCGGTTCAACAGCCT-1 & 1048 & 124 & Myeloid & 6.954639 & 0.11832061\\\\\n",
       "\ts4d9\\_GATGACTTCGCTAGTG-1 & s4d9\\_GATGACTTCGCTAGTG-1 &  848 &  94 & Myeloid & 6.742881 & 0.11084906\\\\\n",
       "\ts4d9\\_CTAGTCGAGCTATTGA-1 & s4d9\\_CTAGTCGAGCTATTGA-1 &  829 &  52 & Myeloid & 6.720220 & 0.06272618\\\\\n",
       "\ts4d9\\_TTATGACCATTTAAGC-1 & s4d9\\_TTATGACCATTTAAGC-1 & 3534 & 336 & Myeloid & 8.170186 & 0.09507640\\\\\n",
       "\ts4d9\\_GTGCTCAAGCAAGATG-1 & s4d9\\_GTGCTCAAGCAAGATG-1 & 3857 & 573 & Myeloid & 8.257645 & 0.14856106\\\\\n",
       "\ts4d9\\_GGATATTGTGTCACGG-1 & s4d9\\_GGATATTGTGTCACGG-1 & 2140 & 285 & Myeloid & 7.668561 & 0.13317757\\\\\n",
       "\ts4d9\\_TGGCCTGCAATTATGC-1 & s4d9\\_TGGCCTGCAATTATGC-1 & 3184 & 332 & Myeloid & 8.065894 & 0.10427136\\\\\n",
       "\ts4d9\\_GAGGGAGCATAACGGG-1 & s4d9\\_GAGGGAGCATAACGGG-1 & 2488 & 358 & Myeloid & 7.819234 & 0.14389068\\\\\n",
       "\ts4d9\\_GGTACAAAGTTGCCTC-1 & s4d9\\_GGTACAAAGTTGCCTC-1 & 2427 & 276 & Myeloid & 7.794411 & 0.11372064\\\\\n",
       "\ts4d9\\_GGAAACTCATAGCTGC-1 & s4d9\\_GGAAACTCATAGCTGC-1 & 1561 & 182 & Myeloid & 7.353082 & 0.11659193\\\\\n",
       "\ts4d9\\_AGAAAGGCAATGCCCG-1 & s4d9\\_AGAAAGGCAATGCCCG-1 & 1773 &  38 & Myeloid & 7.480428 & 0.02143260\\\\\n",
       "\ts4d9\\_CTACAACAGCGAGTAA-1 & s4d9\\_CTACAACAGCGAGTAA-1 & 1161 &  96 & Myeloid & 7.057037 & 0.08268734\\\\\n",
       "\ts4d9\\_ACATTGCAGGGTCTAT-1 & s4d9\\_ACATTGCAGGGTCTAT-1 & 2255 & 255 & Myeloid & 7.720905 & 0.11308204\\\\\n",
       "\ts4d9\\_CTCCAACTCTAACCAA-1 & s4d9\\_CTCCAACTCTAACCAA-1 & 3392 & 399 & Myeloid & 8.129175 & 0.11762972\\\\\n",
       "\ts4d9\\_CAGGGCTTCCGCCAAA-1 & s4d9\\_CAGGGCTTCCGCCAAA-1 & 1837 & 253 & Myeloid & 7.515889 & 0.13772455\\\\\n",
       "\ts4d9\\_TAAACAGCACCTAAGC-1 & s4d9\\_TAAACAGCACCTAAGC-1 & 1474 & 171 & Myeloid & 7.295735 & 0.11601085\\\\\n",
       "\ts4d9\\_TGGTCCTTCAGTGAAG-1 & s4d9\\_TGGTCCTTCAGTGAAG-1 & 2021 & 227 & Myeloid & 7.611348 & 0.11232063\\\\\n",
       "\ts4d9\\_GGATACTTCTTACTCG-1 & s4d9\\_GGATACTTCTTACTCG-1 & 2211 & 192 & Myeloid & 7.701200 & 0.08683853\\\\\n",
       "\ts4d9\\_ACAGGAATCCGTCCAT-1 & s4d9\\_ACAGGAATCCGTCCAT-1 &  747 &  61 & Myeloid & 6.616065 & 0.08165997\\\\\n",
       "\ts4d9\\_TGTGGCCAGCTGGCTA-1 & s4d9\\_TGTGGCCAGCTGGCTA-1 & 1062 & 140 & Myeloid & 6.967909 & 0.13182674\\\\\n",
       "\ts4d9\\_CCGGTTAAGCCTGTTC-1 & s4d9\\_CCGGTTAAGCCTGTTC-1 & 1496 &  58 & Myeloid & 7.310550 & 0.03877005\\\\\n",
       "\ts4d9\\_CAAGCTAGTCAACAAT-1 & s4d9\\_CAAGCTAGTCAACAAT-1 & 1664 & 179 & Myeloid & 7.416980 & 0.10757212\\\\\n",
       "\ts4d9\\_ACAGGTAAGGTTACGT-1 & s4d9\\_ACAGGTAAGGTTACGT-1 & 1221 &  75 & Myeloid & 7.107425 & 0.06142506\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 14048 × 6\n",
       "\n",
       "| <!--/--> | cell &lt;chr&gt; | nUMI &lt;dbl&gt; | nMito &lt;dbl&gt; | celltype &lt;chr&gt; | log.nUMI &lt;dbl&gt; | percent.mito &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| s1d1_CTATGGCCATAACGGG-1 | s1d1_CTATGGCCATAACGGG-1 |  2157 |  69 | Myeloid |  7.676474 | 0.03198887 |\n",
       "| s1d1_ACCACATAGGTGTCCA-1 | s1d1_ACCACATAGGTGTCCA-1 |  2128 |  64 | Myeloid |  7.662938 | 0.03007519 |\n",
       "| s1d1_TGCCGGTAGTCGCAAT-1 | s1d1_TGCCGGTAGTCGCAAT-1 |  1920 |  67 | Myeloid |  7.560080 | 0.03489583 |\n",
       "| s1d1_CATCCCAGTCTTTATG-1 | s1d1_CATCCCAGTCTTTATG-1 |  3211 | 127 | Myeloid |  8.074338 | 0.03955154 |\n",
       "| s1d1_TAGTGAGAGCTGCACA-1 | s1d1_TAGTGAGAGCTGCACA-1 |  3568 |  85 | Myeloid |  8.179760 | 0.02382287 |\n",
       "| s1d1_CCTGTTGGTTATAGCG-1 | s1d1_CCTGTTGGTTATAGCG-1 |  4044 |  42 | Myeloid |  8.304990 | 0.01038576 |\n",
       "| s1d1_TTTAGCTTCTGGCTTC-1 | s1d1_TTTAGCTTCTGGCTTC-1 |  2728 | 105 | Myeloid |  7.911324 | 0.03848974 |\n",
       "| s1d1_TGTGAAACAGGATTAA-1 | s1d1_TGTGAAACAGGATTAA-1 |  2800 |  86 | Myeloid |  7.937375 | 0.03071429 |\n",
       "| s1d1_ACCAGGCTCTTGGATA-1 | s1d1_ACCAGGCTCTTGGATA-1 |  1015 |  94 | Myeloid |  6.922644 | 0.09261084 |\n",
       "| s1d1_CACAGGGAGACAGGTA-1 | s1d1_CACAGGGAGACAGGTA-1 |  3092 |  69 | Myeloid |  8.036573 | 0.02231565 |\n",
       "| s1d1_GCATCCTTCGCTATGG-1 | s1d1_GCATCCTTCGCTATGG-1 |  3112 |  77 | Myeloid |  8.043021 | 0.02474293 |\n",
       "| s1d1_TGCTATGCAGAGGGAG-1 | s1d1_TGCTATGCAGAGGGAG-1 |  2814 |  78 | Myeloid |  7.942362 | 0.02771855 |\n",
       "| s1d1_CCCGCAACAAAGCGCA-1 | s1d1_CCCGCAACAAAGCGCA-1 |  3234 |  58 | Myeloid |  8.081475 | 0.01793445 |\n",
       "| s1d1_AATCTTGAGAAGCTAG-1 | s1d1_AATCTTGAGAAGCTAG-1 |  1647 |  67 | Myeloid |  7.406711 | 0.04068002 |\n",
       "| s1d1_TTGTCCATCATTAGGC-1 | s1d1_TTGTCCATCATTAGGC-1 |  4110 |  79 | Myeloid |  8.321178 | 0.01922141 |\n",
       "| s1d1_TCCAGGATCCTAACGG-1 | s1d1_TCCAGGATCCTAACGG-1 |  3484 | 100 | Myeloid |  8.155936 | 0.02870264 |\n",
       "| s1d1_GGTATGTTCTAACCAA-1 | s1d1_GGTATGTTCTAACCAA-1 |  2759 |  61 | Myeloid |  7.922624 | 0.02210946 |\n",
       "| s1d1_TGGATTCAGTTTCCGC-1 | s1d1_TGGATTCAGTTTCCGC-1 |  3639 |  72 | Myeloid |  8.199464 | 0.01978566 |\n",
       "| s1d1_TGCTTGCTCAATCTCT-1 | s1d1_TGCTTGCTCAATCTCT-1 |  1934 |  71 | Myeloid |  7.567346 | 0.03671148 |\n",
       "| s1d1_CGATGTCCATACCCGG-1 | s1d1_CGATGTCCATACCCGG-1 |  3296 |  54 | Myeloid |  8.100465 | 0.01638350 |\n",
       "| s1d1_ACGCCTAAGCGCCTTT-1 | s1d1_ACGCCTAAGCGCCTTT-1 |  3236 |  92 | Myeloid |  8.082093 | 0.02843016 |\n",
       "| s1d1_GAGTAATAGCTTAACA-1 | s1d1_GAGTAATAGCTTAACA-1 |  4308 | 100 | Myeloid |  8.368229 | 0.02321263 |\n",
       "| s1d1_ATTGGCTAGGGACCTC-1 | s1d1_ATTGGCTAGGGACCTC-1 | 22348 | 863 | Myeloid | 10.014492 | 0.03861643 |\n",
       "| s1d1_CCTGTAACATTGTGCA-1 | s1d1_CCTGTAACATTGTGCA-1 |  3940 | 127 | Myeloid |  8.278936 | 0.03223350 |\n",
       "| s1d1_AGATGAAGTGAGGTGA-1 | s1d1_AGATGAAGTGAGGTGA-1 |  2271 |  69 | Myeloid |  7.727976 | 0.03038309 |\n",
       "| s1d1_GTTAACGGTAGCTGGT-1 | s1d1_GTTAACGGTAGCTGGT-1 |  1222 |  72 | Myeloid |  7.108244 | 0.05891980 |\n",
       "| s1d1_GAACACAAGTAACCAC-1 | s1d1_GAACACAAGTAACCAC-1 |  3170 |  64 | Myeloid |  8.061487 | 0.02018927 |\n",
       "| s1d1_GGTGCTGGTACTTCAC-1 | s1d1_GGTGCTGGTACTTCAC-1 |  4133 | 103 | Myeloid |  8.326759 | 0.02492136 |\n",
       "| s1d1_AGCTGCTCATTGTTGG-1 | s1d1_AGCTGCTCATTGTTGG-1 |  1308 | 102 | Myeloid |  7.176255 | 0.07798165 |\n",
       "| s1d1_TCCAGGATCAACAAGG-1 | s1d1_TCCAGGATCAACAAGG-1 |  3218 |  66 | Myeloid |  8.076515 | 0.02050963 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| s4d9_GTTGCTGAGTAATCCA-1 | s4d9_GTTGCTGAGTAATCCA-1 | 2839 | 168 | Myeloid | 7.951207 | 0.05917577 |\n",
       "| s4d9_AACTAGTGTTACTTCA-1 | s4d9_AACTAGTGTTACTTCA-1 | 1455 | 198 | Myeloid | 7.282761 | 0.13608247 |\n",
       "| s4d9_TGCTTCCAGTCCTTTG-1 | s4d9_TGCTTCCAGTCCTTTG-1 | 1450 | 138 | Myeloid | 7.279319 | 0.09517241 |\n",
       "| s4d9_AAGGATCCACAGGATG-1 | s4d9_AAGGATCCACAGGATG-1 | 1789 | 229 | Myeloid | 7.489412 | 0.12800447 |\n",
       "| s4d9_TGTGCGCAGCCGCAGT-1 | s4d9_TGTGCGCAGCCGCAGT-1 | 1686 | 138 | Myeloid | 7.430114 | 0.08185053 |\n",
       "| s4d9_CCAGACTCAATCATGT-1 | s4d9_CCAGACTCAATCATGT-1 | 1834 |  52 | Myeloid | 7.514255 | 0.02835333 |\n",
       "| s4d9_AGTAAGTAGTTAGAGG-1 | s4d9_AGTAAGTAGTTAGAGG-1 | 1826 | 194 | Myeloid | 7.509883 | 0.10624315 |\n",
       "| s4d9_TGCGGTTCAACAGCCT-1 | s4d9_TGCGGTTCAACAGCCT-1 | 1048 | 124 | Myeloid | 6.954639 | 0.11832061 |\n",
       "| s4d9_GATGACTTCGCTAGTG-1 | s4d9_GATGACTTCGCTAGTG-1 |  848 |  94 | Myeloid | 6.742881 | 0.11084906 |\n",
       "| s4d9_CTAGTCGAGCTATTGA-1 | s4d9_CTAGTCGAGCTATTGA-1 |  829 |  52 | Myeloid | 6.720220 | 0.06272618 |\n",
       "| s4d9_TTATGACCATTTAAGC-1 | s4d9_TTATGACCATTTAAGC-1 | 3534 | 336 | Myeloid | 8.170186 | 0.09507640 |\n",
       "| s4d9_GTGCTCAAGCAAGATG-1 | s4d9_GTGCTCAAGCAAGATG-1 | 3857 | 573 | Myeloid | 8.257645 | 0.14856106 |\n",
       "| s4d9_GGATATTGTGTCACGG-1 | s4d9_GGATATTGTGTCACGG-1 | 2140 | 285 | Myeloid | 7.668561 | 0.13317757 |\n",
       "| s4d9_TGGCCTGCAATTATGC-1 | s4d9_TGGCCTGCAATTATGC-1 | 3184 | 332 | Myeloid | 8.065894 | 0.10427136 |\n",
       "| s4d9_GAGGGAGCATAACGGG-1 | s4d9_GAGGGAGCATAACGGG-1 | 2488 | 358 | Myeloid | 7.819234 | 0.14389068 |\n",
       "| s4d9_GGTACAAAGTTGCCTC-1 | s4d9_GGTACAAAGTTGCCTC-1 | 2427 | 276 | Myeloid | 7.794411 | 0.11372064 |\n",
       "| s4d9_GGAAACTCATAGCTGC-1 | s4d9_GGAAACTCATAGCTGC-1 | 1561 | 182 | Myeloid | 7.353082 | 0.11659193 |\n",
       "| s4d9_AGAAAGGCAATGCCCG-1 | s4d9_AGAAAGGCAATGCCCG-1 | 1773 |  38 | Myeloid | 7.480428 | 0.02143260 |\n",
       "| s4d9_CTACAACAGCGAGTAA-1 | s4d9_CTACAACAGCGAGTAA-1 | 1161 |  96 | Myeloid | 7.057037 | 0.08268734 |\n",
       "| s4d9_ACATTGCAGGGTCTAT-1 | s4d9_ACATTGCAGGGTCTAT-1 | 2255 | 255 | Myeloid | 7.720905 | 0.11308204 |\n",
       "| s4d9_CTCCAACTCTAACCAA-1 | s4d9_CTCCAACTCTAACCAA-1 | 3392 | 399 | Myeloid | 8.129175 | 0.11762972 |\n",
       "| s4d9_CAGGGCTTCCGCCAAA-1 | s4d9_CAGGGCTTCCGCCAAA-1 | 1837 | 253 | Myeloid | 7.515889 | 0.13772455 |\n",
       "| s4d9_TAAACAGCACCTAAGC-1 | s4d9_TAAACAGCACCTAAGC-1 | 1474 | 171 | Myeloid | 7.295735 | 0.11601085 |\n",
       "| s4d9_TGGTCCTTCAGTGAAG-1 | s4d9_TGGTCCTTCAGTGAAG-1 | 2021 | 227 | Myeloid | 7.611348 | 0.11232063 |\n",
       "| s4d9_GGATACTTCTTACTCG-1 | s4d9_GGATACTTCTTACTCG-1 | 2211 | 192 | Myeloid | 7.701200 | 0.08683853 |\n",
       "| s4d9_ACAGGAATCCGTCCAT-1 | s4d9_ACAGGAATCCGTCCAT-1 |  747 |  61 | Myeloid | 6.616065 | 0.08165997 |\n",
       "| s4d9_TGTGGCCAGCTGGCTA-1 | s4d9_TGTGGCCAGCTGGCTA-1 | 1062 | 140 | Myeloid | 6.967909 | 0.13182674 |\n",
       "| s4d9_CCGGTTAAGCCTGTTC-1 | s4d9_CCGGTTAAGCCTGTTC-1 | 1496 |  58 | Myeloid | 7.310550 | 0.03877005 |\n",
       "| s4d9_CAAGCTAGTCAACAAT-1 | s4d9_CAAGCTAGTCAACAAT-1 | 1664 | 179 | Myeloid | 7.416980 | 0.10757212 |\n",
       "| s4d9_ACAGGTAAGGTTACGT-1 | s4d9_ACAGGTAAGGTTACGT-1 | 1221 |  75 | Myeloid | 7.107425 | 0.06142506 |\n",
       "\n"
      ],
      "text/plain": [
       "                        cell                    nUMI  nMito celltype log.nUMI \n",
       "s1d1_CTATGGCCATAACGGG-1 s1d1_CTATGGCCATAACGGG-1  2157  69   Myeloid   7.676474\n",
       "s1d1_ACCACATAGGTGTCCA-1 s1d1_ACCACATAGGTGTCCA-1  2128  64   Myeloid   7.662938\n",
       "s1d1_TGCCGGTAGTCGCAAT-1 s1d1_TGCCGGTAGTCGCAAT-1  1920  67   Myeloid   7.560080\n",
       "s1d1_CATCCCAGTCTTTATG-1 s1d1_CATCCCAGTCTTTATG-1  3211 127   Myeloid   8.074338\n",
       "s1d1_TAGTGAGAGCTGCACA-1 s1d1_TAGTGAGAGCTGCACA-1  3568  85   Myeloid   8.179760\n",
       "s1d1_CCTGTTGGTTATAGCG-1 s1d1_CCTGTTGGTTATAGCG-1  4044  42   Myeloid   8.304990\n",
       "s1d1_TTTAGCTTCTGGCTTC-1 s1d1_TTTAGCTTCTGGCTTC-1  2728 105   Myeloid   7.911324\n",
       "s1d1_TGTGAAACAGGATTAA-1 s1d1_TGTGAAACAGGATTAA-1  2800  86   Myeloid   7.937375\n",
       "s1d1_ACCAGGCTCTTGGATA-1 s1d1_ACCAGGCTCTTGGATA-1  1015  94   Myeloid   6.922644\n",
       "s1d1_CACAGGGAGACAGGTA-1 s1d1_CACAGGGAGACAGGTA-1  3092  69   Myeloid   8.036573\n",
       "s1d1_GCATCCTTCGCTATGG-1 s1d1_GCATCCTTCGCTATGG-1  3112  77   Myeloid   8.043021\n",
       "s1d1_TGCTATGCAGAGGGAG-1 s1d1_TGCTATGCAGAGGGAG-1  2814  78   Myeloid   7.942362\n",
       "s1d1_CCCGCAACAAAGCGCA-1 s1d1_CCCGCAACAAAGCGCA-1  3234  58   Myeloid   8.081475\n",
       "s1d1_AATCTTGAGAAGCTAG-1 s1d1_AATCTTGAGAAGCTAG-1  1647  67   Myeloid   7.406711\n",
       "s1d1_TTGTCCATCATTAGGC-1 s1d1_TTGTCCATCATTAGGC-1  4110  79   Myeloid   8.321178\n",
       "s1d1_TCCAGGATCCTAACGG-1 s1d1_TCCAGGATCCTAACGG-1  3484 100   Myeloid   8.155936\n",
       "s1d1_GGTATGTTCTAACCAA-1 s1d1_GGTATGTTCTAACCAA-1  2759  61   Myeloid   7.922624\n",
       "s1d1_TGGATTCAGTTTCCGC-1 s1d1_TGGATTCAGTTTCCGC-1  3639  72   Myeloid   8.199464\n",
       "s1d1_TGCTTGCTCAATCTCT-1 s1d1_TGCTTGCTCAATCTCT-1  1934  71   Myeloid   7.567346\n",
       "s1d1_CGATGTCCATACCCGG-1 s1d1_CGATGTCCATACCCGG-1  3296  54   Myeloid   8.100465\n",
       "s1d1_ACGCCTAAGCGCCTTT-1 s1d1_ACGCCTAAGCGCCTTT-1  3236  92   Myeloid   8.082093\n",
       "s1d1_GAGTAATAGCTTAACA-1 s1d1_GAGTAATAGCTTAACA-1  4308 100   Myeloid   8.368229\n",
       "s1d1_ATTGGCTAGGGACCTC-1 s1d1_ATTGGCTAGGGACCTC-1 22348 863   Myeloid  10.014492\n",
       "s1d1_CCTGTAACATTGTGCA-1 s1d1_CCTGTAACATTGTGCA-1  3940 127   Myeloid   8.278936\n",
       "s1d1_AGATGAAGTGAGGTGA-1 s1d1_AGATGAAGTGAGGTGA-1  2271  69   Myeloid   7.727976\n",
       "s1d1_GTTAACGGTAGCTGGT-1 s1d1_GTTAACGGTAGCTGGT-1  1222  72   Myeloid   7.108244\n",
       "s1d1_GAACACAAGTAACCAC-1 s1d1_GAACACAAGTAACCAC-1  3170  64   Myeloid   8.061487\n",
       "s1d1_GGTGCTGGTACTTCAC-1 s1d1_GGTGCTGGTACTTCAC-1  4133 103   Myeloid   8.326759\n",
       "s1d1_AGCTGCTCATTGTTGG-1 s1d1_AGCTGCTCATTGTTGG-1  1308 102   Myeloid   7.176255\n",
       "s1d1_TCCAGGATCAACAAGG-1 s1d1_TCCAGGATCAACAAGG-1  3218  66   Myeloid   8.076515\n",
       "⋮                       ⋮                       ⋮     ⋮     ⋮        ⋮        \n",
       "s4d9_GTTGCTGAGTAATCCA-1 s4d9_GTTGCTGAGTAATCCA-1 2839  168   Myeloid  7.951207 \n",
       "s4d9_AACTAGTGTTACTTCA-1 s4d9_AACTAGTGTTACTTCA-1 1455  198   Myeloid  7.282761 \n",
       "s4d9_TGCTTCCAGTCCTTTG-1 s4d9_TGCTTCCAGTCCTTTG-1 1450  138   Myeloid  7.279319 \n",
       "s4d9_AAGGATCCACAGGATG-1 s4d9_AAGGATCCACAGGATG-1 1789  229   Myeloid  7.489412 \n",
       "s4d9_TGTGCGCAGCCGCAGT-1 s4d9_TGTGCGCAGCCGCAGT-1 1686  138   Myeloid  7.430114 \n",
       "s4d9_CCAGACTCAATCATGT-1 s4d9_CCAGACTCAATCATGT-1 1834   52   Myeloid  7.514255 \n",
       "s4d9_AGTAAGTAGTTAGAGG-1 s4d9_AGTAAGTAGTTAGAGG-1 1826  194   Myeloid  7.509883 \n",
       "s4d9_TGCGGTTCAACAGCCT-1 s4d9_TGCGGTTCAACAGCCT-1 1048  124   Myeloid  6.954639 \n",
       "s4d9_GATGACTTCGCTAGTG-1 s4d9_GATGACTTCGCTAGTG-1  848   94   Myeloid  6.742881 \n",
       "s4d9_CTAGTCGAGCTATTGA-1 s4d9_CTAGTCGAGCTATTGA-1  829   52   Myeloid  6.720220 \n",
       "s4d9_TTATGACCATTTAAGC-1 s4d9_TTATGACCATTTAAGC-1 3534  336   Myeloid  8.170186 \n",
       "s4d9_GTGCTCAAGCAAGATG-1 s4d9_GTGCTCAAGCAAGATG-1 3857  573   Myeloid  8.257645 \n",
       "s4d9_GGATATTGTGTCACGG-1 s4d9_GGATATTGTGTCACGG-1 2140  285   Myeloid  7.668561 \n",
       "s4d9_TGGCCTGCAATTATGC-1 s4d9_TGGCCTGCAATTATGC-1 3184  332   Myeloid  8.065894 \n",
       "s4d9_GAGGGAGCATAACGGG-1 s4d9_GAGGGAGCATAACGGG-1 2488  358   Myeloid  7.819234 \n",
       "s4d9_GGTACAAAGTTGCCTC-1 s4d9_GGTACAAAGTTGCCTC-1 2427  276   Myeloid  7.794411 \n",
       "s4d9_GGAAACTCATAGCTGC-1 s4d9_GGAAACTCATAGCTGC-1 1561  182   Myeloid  7.353082 \n",
       "s4d9_AGAAAGGCAATGCCCG-1 s4d9_AGAAAGGCAATGCCCG-1 1773   38   Myeloid  7.480428 \n",
       "s4d9_CTACAACAGCGAGTAA-1 s4d9_CTACAACAGCGAGTAA-1 1161   96   Myeloid  7.057037 \n",
       "s4d9_ACATTGCAGGGTCTAT-1 s4d9_ACATTGCAGGGTCTAT-1 2255  255   Myeloid  7.720905 \n",
       "s4d9_CTCCAACTCTAACCAA-1 s4d9_CTCCAACTCTAACCAA-1 3392  399   Myeloid  8.129175 \n",
       "s4d9_CAGGGCTTCCGCCAAA-1 s4d9_CAGGGCTTCCGCCAAA-1 1837  253   Myeloid  7.515889 \n",
       "s4d9_TAAACAGCACCTAAGC-1 s4d9_TAAACAGCACCTAAGC-1 1474  171   Myeloid  7.295735 \n",
       "s4d9_TGGTCCTTCAGTGAAG-1 s4d9_TGGTCCTTCAGTGAAG-1 2021  227   Myeloid  7.611348 \n",
       "s4d9_GGATACTTCTTACTCG-1 s4d9_GGATACTTCTTACTCG-1 2211  192   Myeloid  7.701200 \n",
       "s4d9_ACAGGAATCCGTCCAT-1 s4d9_ACAGGAATCCGTCCAT-1  747   61   Myeloid  6.616065 \n",
       "s4d9_TGTGGCCAGCTGGCTA-1 s4d9_TGTGGCCAGCTGGCTA-1 1062  140   Myeloid  6.967909 \n",
       "s4d9_CCGGTTAAGCCTGTTC-1 s4d9_CCGGTTAAGCCTGTTC-1 1496   58   Myeloid  7.310550 \n",
       "s4d9_CAAGCTAGTCAACAAT-1 s4d9_CAAGCTAGTCAACAAT-1 1664  179   Myeloid  7.416980 \n",
       "s4d9_ACAGGTAAGGTTACGT-1 s4d9_ACAGGTAAGGTTACGT-1 1221   75   Myeloid  7.107425 \n",
       "                        percent.mito\n",
       "s1d1_CTATGGCCATAACGGG-1 0.03198887  \n",
       "s1d1_ACCACATAGGTGTCCA-1 0.03007519  \n",
       "s1d1_TGCCGGTAGTCGCAAT-1 0.03489583  \n",
       "s1d1_CATCCCAGTCTTTATG-1 0.03955154  \n",
       "s1d1_TAGTGAGAGCTGCACA-1 0.02382287  \n",
       "s1d1_CCTGTTGGTTATAGCG-1 0.01038576  \n",
       "s1d1_TTTAGCTTCTGGCTTC-1 0.03848974  \n",
       "s1d1_TGTGAAACAGGATTAA-1 0.03071429  \n",
       "s1d1_ACCAGGCTCTTGGATA-1 0.09261084  \n",
       "s1d1_CACAGGGAGACAGGTA-1 0.02231565  \n",
       "s1d1_GCATCCTTCGCTATGG-1 0.02474293  \n",
       "s1d1_TGCTATGCAGAGGGAG-1 0.02771855  \n",
       "s1d1_CCCGCAACAAAGCGCA-1 0.01793445  \n",
       "s1d1_AATCTTGAGAAGCTAG-1 0.04068002  \n",
       "s1d1_TTGTCCATCATTAGGC-1 0.01922141  \n",
       "s1d1_TCCAGGATCCTAACGG-1 0.02870264  \n",
       "s1d1_GGTATGTTCTAACCAA-1 0.02210946  \n",
       "s1d1_TGGATTCAGTTTCCGC-1 0.01978566  \n",
       "s1d1_TGCTTGCTCAATCTCT-1 0.03671148  \n",
       "s1d1_CGATGTCCATACCCGG-1 0.01638350  \n",
       "s1d1_ACGCCTAAGCGCCTTT-1 0.02843016  \n",
       "s1d1_GAGTAATAGCTTAACA-1 0.02321263  \n",
       "s1d1_ATTGGCTAGGGACCTC-1 0.03861643  \n",
       "s1d1_CCTGTAACATTGTGCA-1 0.03223350  \n",
       "s1d1_AGATGAAGTGAGGTGA-1 0.03038309  \n",
       "s1d1_GTTAACGGTAGCTGGT-1 0.05891980  \n",
       "s1d1_GAACACAAGTAACCAC-1 0.02018927  \n",
       "s1d1_GGTGCTGGTACTTCAC-1 0.02492136  \n",
       "s1d1_AGCTGCTCATTGTTGG-1 0.07798165  \n",
       "s1d1_TCCAGGATCAACAAGG-1 0.02050963  \n",
       "⋮                       ⋮           \n",
       "s4d9_GTTGCTGAGTAATCCA-1 0.05917577  \n",
       "s4d9_AACTAGTGTTACTTCA-1 0.13608247  \n",
       "s4d9_TGCTTCCAGTCCTTTG-1 0.09517241  \n",
       "s4d9_AAGGATCCACAGGATG-1 0.12800447  \n",
       "s4d9_TGTGCGCAGCCGCAGT-1 0.08185053  \n",
       "s4d9_CCAGACTCAATCATGT-1 0.02835333  \n",
       "s4d9_AGTAAGTAGTTAGAGG-1 0.10624315  \n",
       "s4d9_TGCGGTTCAACAGCCT-1 0.11832061  \n",
       "s4d9_GATGACTTCGCTAGTG-1 0.11084906  \n",
       "s4d9_CTAGTCGAGCTATTGA-1 0.06272618  \n",
       "s4d9_TTATGACCATTTAAGC-1 0.09507640  \n",
       "s4d9_GTGCTCAAGCAAGATG-1 0.14856106  \n",
       "s4d9_GGATATTGTGTCACGG-1 0.13317757  \n",
       "s4d9_TGGCCTGCAATTATGC-1 0.10427136  \n",
       "s4d9_GAGGGAGCATAACGGG-1 0.14389068  \n",
       "s4d9_GGTACAAAGTTGCCTC-1 0.11372064  \n",
       "s4d9_GGAAACTCATAGCTGC-1 0.11659193  \n",
       "s4d9_AGAAAGGCAATGCCCG-1 0.02143260  \n",
       "s4d9_CTACAACAGCGAGTAA-1 0.08268734  \n",
       "s4d9_ACATTGCAGGGTCTAT-1 0.11308204  \n",
       "s4d9_CTCCAACTCTAACCAA-1 0.11762972  \n",
       "s4d9_CAGGGCTTCCGCCAAA-1 0.13772455  \n",
       "s4d9_TAAACAGCACCTAAGC-1 0.11601085  \n",
       "s4d9_TGGTCCTTCAGTGAAG-1 0.11232063  \n",
       "s4d9_GGATACTTCTTACTCG-1 0.08683853  \n",
       "s4d9_ACAGGAATCCGTCCAT-1 0.08165997  \n",
       "s4d9_TGTGGCCAGCTGGCTA-1 0.13182674  \n",
       "s4d9_CCGGTTAAGCCTGTTC-1 0.03877005  \n",
       "s4d9_CAAGCTAGTCAACAAT-1 0.10757212  \n",
       "s4d9_ACAGGTAAGGTTACGT-1 0.06142506  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meta.data = data.frame(cell = colnames(matrix.rna),\n",
    "                       nUMI = colSums(matrix.rna),\n",
    "                       nMito = colSums(matrix.rna[grep(\"MT-\",rownames(matrix.rna)),]),\n",
    "                       celltype = celltype)\n",
    "meta.data[,\"log.nUMI\"] = log(meta.data[,\"nUMI\"])\n",
    "meta.data[,\"percent.mito\"] = meta.data[,\"nMito\"] / meta.data[,\"nUMI\"]\n",
    "meta.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b823cb4a-3ce6-40b1-9a0d-d5d1e4aa3912",
   "metadata": {},
   "source": [
    "RUN SCENT prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f46765dc-9de5-4411-9288-40b59435f9fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "RhpcBLASctl::blas_set_num_threads(1)\n",
    "Sys.setenv(OPENBLAS_NUM_THREADS=1)\n",
    "SCENT_algorithm.modified = function (object, \n",
    "                                     # celltype, \n",
    "                                     # ncores, \n",
    "                                     # regr = \"poisson\", \n",
    "                                     # bin = TRUE) \n",
    "                                     regr = \"poisson\") {\n",
    "    res <- data.frame()\n",
    "    for (n in 1:nrow(object@peak.info)) {        \n",
    "        gene <- object@peak.info[n, 1]\n",
    "        this_peak <- object@peak.info[n, 2]\n",
    "        atac_target <- data.frame(cell = colnames(object@atac), \n",
    "            atac = object@atac[this_peak, ])\n",
    "        # if (bin) {\n",
    "        #     atac_target[atac_target$atac > 0, ]$atac <- 1\n",
    "        # }\n",
    "        mrna_target <- object@rna[gene, ]\n",
    "        df <- data.frame(cell = names(mrna_target), exprs = as.numeric(mrna_target))\n",
    "        df <- merge(df, atac_target, by = \"cell\")\n",
    "        df <- merge(df, object@meta.data, by = \"cell\")\n",
    "        # df2 <- df[df[[object@celltypes]] == celltype, ]\n",
    "        df2 <- df\n",
    "        # nonzero_m <- length(df2$exprs[df2$exprs > 0])/length(df2$exprs)\n",
    "        # nonzero_a <- length(df2$atac[df2$atac > 0])/length(df2$atac)\n",
    "        # if (nonzero_m > 0.05 & nonzero_a > 0.05) {\n",
    "            res_var <- \"exprs\"\n",
    "            pred_var <- c(\"atac\", object@covariates)\n",
    "            formula <- as.formula(paste(res_var, paste(pred_var, \n",
    "                collapse = \"+\"), sep = \"~\"))\n",
    "            if (regr == \"poisson\") {\n",
    "                base = glm(formula, family = \"poisson\", data = df2)\n",
    "                coefs <- summary(base)$coefficients[\"atac\", ]\n",
    "                # assoc <- assoc_poisson\n",
    "                assoc <- SCENT::assoc_poisson\n",
    "            }\n",
    "            else if (regr == \"negbin\") {\n",
    "                base = glm.nb(formula, data = df2)\n",
    "                coefs <- summary(base)$coefficients[\"atac\", ]\n",
    "                # assoc <- assoc_negbin\n",
    "                assoc <- SCENT::assoc_negbin\n",
    "            }\n",
    "            # bs = boot::boot(df2, assoc, R = 100, formula = formula, \n",
    "            #     stype = \"i\", parallel = \"no\", ncpus = 1)\n",
    "            # p0 = basic_p(bs$t0[1], bs$t[, 1])\n",
    "            # p0 = SCENT::basic_p(bs$t0[1], bs$t[, 1])\n",
    "            # if (p0 < 0.1) {\n",
    "            #     bs = boot::boot(df2, assoc, R = 500, formula = formula, \n",
    "            #       stype = \"i\", \n",
    "            #       # parallel = \"multicore\", ncpus = ncores)\n",
    "            #       parallel = \"no\", ncpus = 1)\n",
    "            #     # p0 = basic_p(bs$t0[1], bs$t[, 1])\n",
    "            #     p0 = SCENT::basic_p(bs$t0[1], bs$t[, 1])\n",
    "            # }\n",
    "            # if (p0 < 0.05) {\n",
    "            #     bs = boot::boot(df2, assoc, R = 2500, formula = formula, \n",
    "            #       stype = \"i\", \n",
    "            #       # parallel = \"multicore\", ncpus = ncores)\n",
    "            #       parallel = \"no\", ncpus = 1)\n",
    "            #     # p0 = basic_p(bs$t0[1], bs$t[, 1])\n",
    "            #     p0 = SCENT::basic_p(bs$t0[1], bs$t[, 1])\n",
    "            # }\n",
    "            # if (p0 < 0.01) {\n",
    "                bs = boot::boot(df2, assoc, R = 25000, formula = formula, \n",
    "                  stype = \"i\", \n",
    "                  # parallel = \"multicore\", ncpus = ncores)\n",
    "                  parallel = \"no\", ncpus = 1)\n",
    "                # p0 = basic_p(bs$t0[1], bs$t[, 1])\n",
    "                p0 = SCENT::basic_p(bs$t0[1], bs$t[, 1])\n",
    "            # }\n",
    "            # if (p0 < 0.001) {\n",
    "            #     bs = boot::boot(df2, assoc, R = 50000, formula = formula, \n",
    "            #       stype = \"i\", \n",
    "            #       # parallel = \"multicore\", ncpus = ncores)\n",
    "            #       parallel = \"no\", ncpus = 1)\n",
    "            #     # p0 = basic_p(bs$t0[1], bs$t[, 1])\n",
    "            #     p0 = SCENT::basic_p(bs$t0[1], bs$t[, 1])\n",
    "            # }\n",
    "            out <- data.frame(gene = gene, peak = this_peak, \n",
    "                beta = coefs[1], se = coefs[2], z = coefs[3], \n",
    "                p = coefs[4], boot_basic_p = p0)           \n",
    "            res <- rbind(res, out)\n",
    "        # }\n",
    "    }\n",
    "    object@SCENT.result <- res\n",
    "    return(object)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "563771b9-b96a-44d5-a5fe-fa30a7b9851b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(dir.output, \"chr\", sep = \"/\"), recursive = T):\n",
      "“'/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/SCENT/BMMC5.Myeloid.25000.250329//chr' already exists”\n"
     ]
    }
   ],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0d2266ce-b92d-40d2-9b78-f39bd33a5079",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71283133-9bcd-4177-b7ae-a14b3776e27b",
   "metadata": {},
   "source": [
    "Filter genes which can be reused"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e869e16b-9f63-4666-b464-2b2c2c8c1784",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_gene_names_reuse <- function(gene_gtf_path, abc_genes_path){\n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "    abc_genes.reuse = abc_genes[abc_genes$abc_name == abc_genes$gene_ref_name,]\n",
    "\treturn(abc_genes.reuse)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b374ec27-ffbd-4069-adf2-1200c35bfcfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "abc_genes.reuse = map_gene_names_reuse(gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "21af9f9c-a03d-4cde-ac06-26cbc0eef1fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>19669</li><li>3</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 19669\n",
       "\\item 3\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 19669\n",
       "2. 3\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 19669     3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 19669 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>abc_name</th><th scope=col>Ensembl_ID</th><th scope=col>gene_ref_name</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>FAM138A  </td><td>ENSG00000237613</td><td>FAM138A  </td></tr>\n",
       "\t<tr><td>FAM138F  </td><td>ENSG00000282591</td><td>FAM138F  </td></tr>\n",
       "\t<tr><td>OR4F5    </td><td>ENSG00000186092</td><td>OR4F5    </td></tr>\n",
       "\t<tr><td>FAM87B   </td><td>ENSG00000177757</td><td>FAM87B   </td></tr>\n",
       "\t<tr><td>LINC00115</td><td>ENSG00000225880</td><td>LINC00115</td></tr>\n",
       "\t<tr><td>LINC01128</td><td>ENSG00000228794</td><td>LINC01128</td></tr>\n",
       "\t<tr><td>FAM41C   </td><td>ENSG00000230368</td><td>FAM41C   </td></tr>\n",
       "\t<tr><td>SAMD11   </td><td>ENSG00000187634</td><td>SAMD11   </td></tr>\n",
       "\t<tr><td>NOC2L    </td><td>ENSG00000188976</td><td>NOC2L    </td></tr>\n",
       "\t<tr><td>KLHL17   </td><td>ENSG00000187961</td><td>KLHL17   </td></tr>\n",
       "\t<tr><td>PLEKHN1  </td><td>ENSG00000187583</td><td>PLEKHN1  </td></tr>\n",
       "\t<tr><td>PERM1    </td><td>ENSG00000187642</td><td>PERM1    </td></tr>\n",
       "\t<tr><td>HES4     </td><td>ENSG00000188290</td><td>HES4     </td></tr>\n",
       "\t<tr><td>ISG15    </td><td>ENSG00000187608</td><td>ISG15    </td></tr>\n",
       "\t<tr><td>AGRN     </td><td>ENSG00000188157</td><td>AGRN     </td></tr>\n",
       "\t<tr><td>RNF223   </td><td>ENSG00000237330</td><td>RNF223   </td></tr>\n",
       "\t<tr><td>C1orf159 </td><td>ENSG00000131591</td><td>C1orf159 </td></tr>\n",
       "\t<tr><td>LINC01342</td><td>ENSG00000223823</td><td>LINC01342</td></tr>\n",
       "\t<tr><td>TTLL10   </td><td>ENSG00000162571</td><td>TTLL10   </td></tr>\n",
       "\t<tr><td>TNFRSF18 </td><td>ENSG00000186891</td><td>TNFRSF18 </td></tr>\n",
       "\t<tr><td>TNFRSF4  </td><td>ENSG00000186827</td><td>TNFRSF4  </td></tr>\n",
       "\t<tr><td>SDF4     </td><td>ENSG00000078808</td><td>SDF4     </td></tr>\n",
       "\t<tr><td>B3GALT6  </td><td>ENSG00000176022</td><td>B3GALT6  </td></tr>\n",
       "\t<tr><td>C1QTNF12 </td><td>ENSG00000184163</td><td>C1QTNF12 </td></tr>\n",
       "\t<tr><td>UBE2J2   </td><td>ENSG00000160087</td><td>UBE2J2   </td></tr>\n",
       "\t<tr><td>SCNN1D   </td><td>ENSG00000162572</td><td>SCNN1D   </td></tr>\n",
       "\t<tr><td>ACAP3    </td><td>ENSG00000131584</td><td>ACAP3    </td></tr>\n",
       "\t<tr><td>PUSL1    </td><td>ENSG00000169972</td><td>PUSL1    </td></tr>\n",
       "\t<tr><td>INTS11   </td><td>ENSG00000127054</td><td>INTS11   </td></tr>\n",
       "\t<tr><td>CPTP     </td><td>ENSG00000224051</td><td>CPTP     </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>NA      </td><td>NA             </td><td>NA      </td></tr>\n",
       "\t<tr><td>TTTY12  </td><td>ENSG00000237048</td><td>TTTY12  </td></tr>\n",
       "\t<tr><td>NA      </td><td>NA             </td><td>NA      </td></tr>\n",
       "\t<tr><td>TTTY18  </td><td>ENSG00000233699</td><td>TTTY18  </td></tr>\n",
       "\t<tr><td>TTTY19  </td><td>ENSG00000232419</td><td>TTTY19  </td></tr>\n",
       "\t<tr><td>TTTY11  </td><td>ENSG00000180910</td><td>TTTY11  </td></tr>\n",
       "\t<tr><td>TSPY10  </td><td>ENSG00000236424</td><td>TSPY10  </td></tr>\n",
       "\t<tr><td>TSPY4   </td><td>ENSG00000233803</td><td>TSPY4   </td></tr>\n",
       "\t<tr><td>TSPY8   </td><td>ENSG00000229549</td><td>TSPY8   </td></tr>\n",
       "\t<tr><td>FAM197Y5</td><td>ENSG00000225516</td><td>FAM197Y5</td></tr>\n",
       "\t<tr><td>TSPY3   </td><td>ENSG00000228927</td><td>TSPY3   </td></tr>\n",
       "\t<tr><td>TSPY1   </td><td>ENSG00000258992</td><td>TSPY1   </td></tr>\n",
       "\t<tr><td>TTTY22  </td><td>ENSG00000224075</td><td>TTTY22  </td></tr>\n",
       "\t<tr><td>NA      </td><td>NA             </td><td>NA      </td></tr>\n",
       "\t<tr><td>USP9Y   </td><td>ENSG00000114374</td><td>USP9Y   </td></tr>\n",
       "\t<tr><td>DDX3Y   </td><td>ENSG00000067048</td><td>DDX3Y   </td></tr>\n",
       "\t<tr><td>UTY     </td><td>ENSG00000183878</td><td>UTY     </td></tr>\n",
       "\t<tr><td>TMSB4Y  </td><td>ENSG00000154620</td><td>TMSB4Y  </td></tr>\n",
       "\t<tr><td>NLGN4Y  </td><td>ENSG00000165246</td><td>NLGN4Y  </td></tr>\n",
       "\t<tr><td>TTTY14  </td><td>ENSG00000176728</td><td>TTTY14  </td></tr>\n",
       "\t<tr><td>CD24    </td><td>ENSG00000272398</td><td>CD24    </td></tr>\n",
       "\t<tr><td>KDM5D   </td><td>ENSG00000012817</td><td>KDM5D   </td></tr>\n",
       "\t<tr><td>TTTY10  </td><td>ENSG00000229236</td><td>TTTY10  </td></tr>\n",
       "\t<tr><td>EIF1AY  </td><td>ENSG00000198692</td><td>EIF1AY  </td></tr>\n",
       "\t<tr><td>RPS4Y2  </td><td>ENSG00000280969</td><td>RPS4Y2  </td></tr>\n",
       "\t<tr><td>PRORY   </td><td>ENSG00000183146</td><td>PRORY   </td></tr>\n",
       "\t<tr><td>TTTY13  </td><td>ENSG00000184991</td><td>TTTY13  </td></tr>\n",
       "\t<tr><td>TTTY5   </td><td>ENSG00000215560</td><td>TTTY5   </td></tr>\n",
       "\t<tr><td>DAZ1    </td><td>ENSG00000188120</td><td>DAZ1    </td></tr>\n",
       "\t<tr><td>DAZ2    </td><td>ENSG00000205944</td><td>DAZ2    </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 19669 × 3\n",
       "\\begin{tabular}{lll}\n",
       " abc\\_name & Ensembl\\_ID & gene\\_ref\\_name\\\\\n",
       " <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t FAM138A   & ENSG00000237613 & FAM138A  \\\\\n",
       "\t FAM138F   & ENSG00000282591 & FAM138F  \\\\\n",
       "\t OR4F5     & ENSG00000186092 & OR4F5    \\\\\n",
       "\t FAM87B    & ENSG00000177757 & FAM87B   \\\\\n",
       "\t LINC00115 & ENSG00000225880 & LINC00115\\\\\n",
       "\t LINC01128 & ENSG00000228794 & LINC01128\\\\\n",
       "\t FAM41C    & ENSG00000230368 & FAM41C   \\\\\n",
       "\t SAMD11    & ENSG00000187634 & SAMD11   \\\\\n",
       "\t NOC2L     & ENSG00000188976 & NOC2L    \\\\\n",
       "\t KLHL17    & ENSG00000187961 & KLHL17   \\\\\n",
       "\t PLEKHN1   & ENSG00000187583 & PLEKHN1  \\\\\n",
       "\t PERM1     & ENSG00000187642 & PERM1    \\\\\n",
       "\t HES4      & ENSG00000188290 & HES4     \\\\\n",
       "\t ISG15     & ENSG00000187608 & ISG15    \\\\\n",
       "\t AGRN      & ENSG00000188157 & AGRN     \\\\\n",
       "\t RNF223    & ENSG00000237330 & RNF223   \\\\\n",
       "\t C1orf159  & ENSG00000131591 & C1orf159 \\\\\n",
       "\t LINC01342 & ENSG00000223823 & LINC01342\\\\\n",
       "\t TTLL10    & ENSG00000162571 & TTLL10   \\\\\n",
       "\t TNFRSF18  & ENSG00000186891 & TNFRSF18 \\\\\n",
       "\t TNFRSF4   & ENSG00000186827 & TNFRSF4  \\\\\n",
       "\t SDF4      & ENSG00000078808 & SDF4     \\\\\n",
       "\t B3GALT6   & ENSG00000176022 & B3GALT6  \\\\\n",
       "\t C1QTNF12  & ENSG00000184163 & C1QTNF12 \\\\\n",
       "\t UBE2J2    & ENSG00000160087 & UBE2J2   \\\\\n",
       "\t SCNN1D    & ENSG00000162572 & SCNN1D   \\\\\n",
       "\t ACAP3     & ENSG00000131584 & ACAP3    \\\\\n",
       "\t PUSL1     & ENSG00000169972 & PUSL1    \\\\\n",
       "\t INTS11    & ENSG00000127054 & INTS11   \\\\\n",
       "\t CPTP      & ENSG00000224051 & CPTP     \\\\\n",
       "\t ⋮ & ⋮ & ⋮\\\\\n",
       "\t NA       & NA              & NA      \\\\\n",
       "\t TTTY12   & ENSG00000237048 & TTTY12  \\\\\n",
       "\t NA       & NA              & NA      \\\\\n",
       "\t TTTY18   & ENSG00000233699 & TTTY18  \\\\\n",
       "\t TTTY19   & ENSG00000232419 & TTTY19  \\\\\n",
       "\t TTTY11   & ENSG00000180910 & TTTY11  \\\\\n",
       "\t TSPY10   & ENSG00000236424 & TSPY10  \\\\\n",
       "\t TSPY4    & ENSG00000233803 & TSPY4   \\\\\n",
       "\t TSPY8    & ENSG00000229549 & TSPY8   \\\\\n",
       "\t FAM197Y5 & ENSG00000225516 & FAM197Y5\\\\\n",
       "\t TSPY3    & ENSG00000228927 & TSPY3   \\\\\n",
       "\t TSPY1    & ENSG00000258992 & TSPY1   \\\\\n",
       "\t TTTY22   & ENSG00000224075 & TTTY22  \\\\\n",
       "\t NA       & NA              & NA      \\\\\n",
       "\t USP9Y    & ENSG00000114374 & USP9Y   \\\\\n",
       "\t DDX3Y    & ENSG00000067048 & DDX3Y   \\\\\n",
       "\t UTY      & ENSG00000183878 & UTY     \\\\\n",
       "\t TMSB4Y   & ENSG00000154620 & TMSB4Y  \\\\\n",
       "\t NLGN4Y   & ENSG00000165246 & NLGN4Y  \\\\\n",
       "\t TTTY14   & ENSG00000176728 & TTTY14  \\\\\n",
       "\t CD24     & ENSG00000272398 & CD24    \\\\\n",
       "\t KDM5D    & ENSG00000012817 & KDM5D   \\\\\n",
       "\t TTTY10   & ENSG00000229236 & TTTY10  \\\\\n",
       "\t EIF1AY   & ENSG00000198692 & EIF1AY  \\\\\n",
       "\t RPS4Y2   & ENSG00000280969 & RPS4Y2  \\\\\n",
       "\t PRORY    & ENSG00000183146 & PRORY   \\\\\n",
       "\t TTTY13   & ENSG00000184991 & TTTY13  \\\\\n",
       "\t TTTY5    & ENSG00000215560 & TTTY5   \\\\\n",
       "\t DAZ1     & ENSG00000188120 & DAZ1    \\\\\n",
       "\t DAZ2     & ENSG00000205944 & DAZ2    \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 19669 × 3\n",
       "\n",
       "| abc_name &lt;chr&gt; | Ensembl_ID &lt;chr&gt; | gene_ref_name &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| FAM138A   | ENSG00000237613 | FAM138A   |\n",
       "| FAM138F   | ENSG00000282591 | FAM138F   |\n",
       "| OR4F5     | ENSG00000186092 | OR4F5     |\n",
       "| FAM87B    | ENSG00000177757 | FAM87B    |\n",
       "| LINC00115 | ENSG00000225880 | LINC00115 |\n",
       "| LINC01128 | ENSG00000228794 | LINC01128 |\n",
       "| FAM41C    | ENSG00000230368 | FAM41C    |\n",
       "| SAMD11    | ENSG00000187634 | SAMD11    |\n",
       "| NOC2L     | ENSG00000188976 | NOC2L     |\n",
       "| KLHL17    | ENSG00000187961 | KLHL17    |\n",
       "| PLEKHN1   | ENSG00000187583 | PLEKHN1   |\n",
       "| PERM1     | ENSG00000187642 | PERM1     |\n",
       "| HES4      | ENSG00000188290 | HES4      |\n",
       "| ISG15     | ENSG00000187608 | ISG15     |\n",
       "| AGRN      | ENSG00000188157 | AGRN      |\n",
       "| RNF223    | ENSG00000237330 | RNF223    |\n",
       "| C1orf159  | ENSG00000131591 | C1orf159  |\n",
       "| LINC01342 | ENSG00000223823 | LINC01342 |\n",
       "| TTLL10    | ENSG00000162571 | TTLL10    |\n",
       "| TNFRSF18  | ENSG00000186891 | TNFRSF18  |\n",
       "| TNFRSF4   | ENSG00000186827 | TNFRSF4   |\n",
       "| SDF4      | ENSG00000078808 | SDF4      |\n",
       "| B3GALT6   | ENSG00000176022 | B3GALT6   |\n",
       "| C1QTNF12  | ENSG00000184163 | C1QTNF12  |\n",
       "| UBE2J2    | ENSG00000160087 | UBE2J2    |\n",
       "| SCNN1D    | ENSG00000162572 | SCNN1D    |\n",
       "| ACAP3     | ENSG00000131584 | ACAP3     |\n",
       "| PUSL1     | ENSG00000169972 | PUSL1     |\n",
       "| INTS11    | ENSG00000127054 | INTS11    |\n",
       "| CPTP      | ENSG00000224051 | CPTP      |\n",
       "| ⋮ | ⋮ | ⋮ |\n",
       "| NA       | NA              | NA       |\n",
       "| TTTY12   | ENSG00000237048 | TTTY12   |\n",
       "| NA       | NA              | NA       |\n",
       "| TTTY18   | ENSG00000233699 | TTTY18   |\n",
       "| TTTY19   | ENSG00000232419 | TTTY19   |\n",
       "| TTTY11   | ENSG00000180910 | TTTY11   |\n",
       "| TSPY10   | ENSG00000236424 | TSPY10   |\n",
       "| TSPY4    | ENSG00000233803 | TSPY4    |\n",
       "| TSPY8    | ENSG00000229549 | TSPY8    |\n",
       "| FAM197Y5 | ENSG00000225516 | FAM197Y5 |\n",
       "| TSPY3    | ENSG00000228927 | TSPY3    |\n",
       "| TSPY1    | ENSG00000258992 | TSPY1    |\n",
       "| TTTY22   | ENSG00000224075 | TTTY22   |\n",
       "| NA       | NA              | NA       |\n",
       "| USP9Y    | ENSG00000114374 | USP9Y    |\n",
       "| DDX3Y    | ENSG00000067048 | DDX3Y    |\n",
       "| UTY      | ENSG00000183878 | UTY      |\n",
       "| TMSB4Y   | ENSG00000154620 | TMSB4Y   |\n",
       "| NLGN4Y   | ENSG00000165246 | NLGN4Y   |\n",
       "| TTTY14   | ENSG00000176728 | TTTY14   |\n",
       "| CD24     | ENSG00000272398 | CD24     |\n",
       "| KDM5D    | ENSG00000012817 | KDM5D    |\n",
       "| TTTY10   | ENSG00000229236 | TTTY10   |\n",
       "| EIF1AY   | ENSG00000198692 | EIF1AY   |\n",
       "| RPS4Y2   | ENSG00000280969 | RPS4Y2   |\n",
       "| PRORY    | ENSG00000183146 | PRORY    |\n",
       "| TTTY13   | ENSG00000184991 | TTTY13   |\n",
       "| TTTY5    | ENSG00000215560 | TTTY5    |\n",
       "| DAZ1     | ENSG00000188120 | DAZ1     |\n",
       "| DAZ2     | ENSG00000205944 | DAZ2     |\n",
       "\n"
      ],
      "text/plain": [
       "      abc_name  Ensembl_ID      gene_ref_name\n",
       "1     FAM138A   ENSG00000237613 FAM138A      \n",
       "2     FAM138F   ENSG00000282591 FAM138F      \n",
       "3     OR4F5     ENSG00000186092 OR4F5        \n",
       "4     FAM87B    ENSG00000177757 FAM87B       \n",
       "5     LINC00115 ENSG00000225880 LINC00115    \n",
       "6     LINC01128 ENSG00000228794 LINC01128    \n",
       "7     FAM41C    ENSG00000230368 FAM41C       \n",
       "8     SAMD11    ENSG00000187634 SAMD11       \n",
       "9     NOC2L     ENSG00000188976 NOC2L        \n",
       "10    KLHL17    ENSG00000187961 KLHL17       \n",
       "11    PLEKHN1   ENSG00000187583 PLEKHN1      \n",
       "12    PERM1     ENSG00000187642 PERM1        \n",
       "13    HES4      ENSG00000188290 HES4         \n",
       "14    ISG15     ENSG00000187608 ISG15        \n",
       "15    AGRN      ENSG00000188157 AGRN         \n",
       "16    RNF223    ENSG00000237330 RNF223       \n",
       "17    C1orf159  ENSG00000131591 C1orf159     \n",
       "18    LINC01342 ENSG00000223823 LINC01342    \n",
       "19    TTLL10    ENSG00000162571 TTLL10       \n",
       "20    TNFRSF18  ENSG00000186891 TNFRSF18     \n",
       "21    TNFRSF4   ENSG00000186827 TNFRSF4      \n",
       "22    SDF4      ENSG00000078808 SDF4         \n",
       "23    B3GALT6   ENSG00000176022 B3GALT6      \n",
       "24    C1QTNF12  ENSG00000184163 C1QTNF12     \n",
       "25    UBE2J2    ENSG00000160087 UBE2J2       \n",
       "26    SCNN1D    ENSG00000162572 SCNN1D       \n",
       "27    ACAP3     ENSG00000131584 ACAP3        \n",
       "28    PUSL1     ENSG00000169972 PUSL1        \n",
       "29    INTS11    ENSG00000127054 INTS11       \n",
       "30    CPTP      ENSG00000224051 CPTP         \n",
       "⋮     ⋮         ⋮               ⋮            \n",
       "19640 NA        NA              NA           \n",
       "19641 TTTY12    ENSG00000237048 TTTY12       \n",
       "19642 NA        NA              NA           \n",
       "19643 TTTY18    ENSG00000233699 TTTY18       \n",
       "19644 TTTY19    ENSG00000232419 TTTY19       \n",
       "19645 TTTY11    ENSG00000180910 TTTY11       \n",
       "19646 TSPY10    ENSG00000236424 TSPY10       \n",
       "19647 TSPY4     ENSG00000233803 TSPY4        \n",
       "19648 TSPY8     ENSG00000229549 TSPY8        \n",
       "19649 FAM197Y5  ENSG00000225516 FAM197Y5     \n",
       "19650 TSPY3     ENSG00000228927 TSPY3        \n",
       "19651 TSPY1     ENSG00000258992 TSPY1        \n",
       "19652 TTTY22    ENSG00000224075 TTTY22       \n",
       "19653 NA        NA              NA           \n",
       "19654 USP9Y     ENSG00000114374 USP9Y        \n",
       "19655 DDX3Y     ENSG00000067048 DDX3Y        \n",
       "19656 UTY       ENSG00000183878 UTY          \n",
       "19657 TMSB4Y    ENSG00000154620 TMSB4Y       \n",
       "19658 NLGN4Y    ENSG00000165246 NLGN4Y       \n",
       "19659 TTTY14    ENSG00000176728 TTTY14       \n",
       "19660 CD24      ENSG00000272398 CD24         \n",
       "19661 KDM5D     ENSG00000012817 KDM5D        \n",
       "19662 TTTY10    ENSG00000229236 TTTY10       \n",
       "19663 EIF1AY    ENSG00000198692 EIF1AY       \n",
       "19664 RPS4Y2    ENSG00000280969 RPS4Y2       \n",
       "19665 PRORY     ENSG00000183146 PRORY        \n",
       "19666 TTTY13    ENSG00000184991 TTTY13       \n",
       "19667 TTTY5     ENSG00000215560 TTTY5        \n",
       "19668 DAZ1      ENSG00000188120 DAZ1         \n",
       "19669 DAZ2      ENSG00000205944 DAZ2         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(abc_genes.reuse)\n",
    "abc_genes.reuse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7f3f250f-fb93-49e3-a813-9886243ddadf",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.reuse.25000 = pairs.E2G.reuse[pairs.E2G.reuse$boot_basic_p < 0.01 & \n",
    "                                        pairs.E2G.reuse$TargetGene %in% abc_genes.reuse$gene_ref_name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1127d717-5fc4-4f46-ae77-aad414bea55f",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(pairs.E2G.reuse.25000) = pairs.E2G.reuse.25000$PairName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "bedf4221-dd29-42e9-a38c-e08be387cd41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 36081 ranges and 11 metadata columns:\n",
       "                                  seqnames              ranges strand |\n",
       "                                     <Rle>           <IRanges>  <Rle> |\n",
       "      chr1-1001860-1002126_CDK11B     chr1     1001860-1002126      * |\n",
       "     chr1-10022225-10023589_KIF1B     chr1   10022225-10023589      * |\n",
       "  chr1-10022225-10023589_SLC25A33     chr1   10022225-10023589      * |\n",
       "  chr1-100265575-100266826_CDC14A     chr1 100265575-100266826      * |\n",
       "  chr1-10032414-10033799_SLC25A33     chr1   10032414-10033799      * |\n",
       "                              ...      ...                 ...    ... .\n",
       "   chrX-77899134-77899778_CYSLTR1     chrX   77899134-77899778      * |\n",
       "   chrX-81120681-81122301_SH3BGRL     chrX   81120681-81122301      * |\n",
       "       chrX-9378667-9379857_TBL1X     chrX     9378667-9379857      * |\n",
       "    chrX-96722435-96723511_DIAPH2     chrX   96722435-96723511      * |\n",
       "         chrX-989778-990654_IL3RA     chrX       989778-990654      * |\n",
       "                                   TargetGene               PeakName\n",
       "                                  <character>            <character>\n",
       "      chr1-1001860-1002126_CDK11B      CDK11B   chr1-1001860-1002126\n",
       "     chr1-10022225-10023589_KIF1B       KIF1B chr1-10022225-10023589\n",
       "  chr1-10022225-10023589_SLC25A33    SLC25A33 chr1-10022225-10023589\n",
       "  chr1-100265575-100266826_CDC14A      CDC14A chr1-100265575-10026..\n",
       "  chr1-10032414-10033799_SLC25A33    SLC25A33 chr1-10032414-10033799\n",
       "                              ...         ...                    ...\n",
       "   chrX-77899134-77899778_CYSLTR1     CYSLTR1 chrX-77899134-77899778\n",
       "   chrX-81120681-81122301_SH3BGRL     SH3BGRL chrX-81120681-81122301\n",
       "       chrX-9378667-9379857_TBL1X       TBL1X   chrX-9378667-9379857\n",
       "    chrX-96722435-96723511_DIAPH2      DIAPH2 chrX-96722435-96723511\n",
       "         chrX-989778-990654_IL3RA       IL3RA     chrX-989778-990654\n",
       "                                                PairName       beta        se\n",
       "                                             <character>  <numeric> <numeric>\n",
       "      chr1-1001860-1002126_CDK11B chr1-1001860-1002126..   0.168775 0.0655577\n",
       "     chr1-10022225-10023589_KIF1B chr1-10022225-100235..   0.127238 0.0430439\n",
       "  chr1-10022225-10023589_SLC25A33 chr1-10022225-100235..   0.286707 0.0643334\n",
       "  chr1-100265575-100266826_CDC14A chr1-100265575-10026..   0.110917 0.0253869\n",
       "  chr1-10032414-10033799_SLC25A33 chr1-10032414-100337..   0.132958 0.0336388\n",
       "                              ...                    ...        ...       ...\n",
       "   chrX-77899134-77899778_CYSLTR1 chrX-77899134-778997.. -0.4792389 0.0950917\n",
       "   chrX-81120681-81122301_SH3BGRL chrX-81120681-811223.. -0.1499563 0.0329712\n",
       "       chrX-9378667-9379857_TBL1X chrX-9378667-9379857..  0.1630958 0.0264363\n",
       "    chrX-96722435-96723511_DIAPH2 chrX-96722435-967235..  0.0833827 0.0201098\n",
       "         chrX-989778-990654_IL3RA chrX-989778-990654_I..  0.3102046 0.0796438\n",
       "                                          z           p boot_basic_p\n",
       "                                  <numeric>   <numeric>    <numeric>\n",
       "      chr1-1001860-1002126_CDK11B   2.57444 1.00401e-02      0.00976\n",
       "     chr1-10022225-10023589_KIF1B   2.95600 3.11658e-03      0.00616\n",
       "  chr1-10022225-10023589_SLC25A33   4.45657 8.32801e-06      0.00008\n",
       "  chr1-100265575-100266826_CDC14A   4.36905 1.24786e-05      0.00080\n",
       "  chr1-10032414-10033799_SLC25A33   3.95251 7.73355e-05      0.00208\n",
       "                              ...       ...         ...          ...\n",
       "   chrX-77899134-77899778_CYSLTR1  -5.03975 4.66132e-07      0.00032\n",
       "   chrX-81120681-81122301_SH3BGRL  -4.54809 5.41340e-06      0.00048\n",
       "       chrX-9378667-9379857_TBL1X   6.16938 6.85587e-10      0.00008\n",
       "    chrX-96722435-96723511_DIAPH2   4.14637 3.37783e-05      0.00208\n",
       "         chrX-989778-990654_IL3RA   3.89490 9.82400e-05      0.00016\n",
       "                                  boot_basic_p.log10      sign\n",
       "                                           <numeric> <numeric>\n",
       "      chr1-1001860-1002126_CDK11B            2.01055         1\n",
       "     chr1-10022225-10023589_KIF1B            2.21042         1\n",
       "  chr1-10022225-10023589_SLC25A33            4.09691         1\n",
       "  chr1-100265575-100266826_CDC14A            3.09691         1\n",
       "  chr1-10032414-10033799_SLC25A33            2.68194         1\n",
       "                              ...                ...       ...\n",
       "   chrX-77899134-77899778_CYSLTR1            3.49485        -1\n",
       "   chrX-81120681-81122301_SH3BGRL            3.31876        -1\n",
       "       chrX-9378667-9379857_TBL1X            4.09691         1\n",
       "    chrX-96722435-96723511_DIAPH2            2.68194         1\n",
       "         chrX-989778-990654_IL3RA            3.79588         1\n",
       "                                  boot_basic_p.log10.signed\n",
       "                                                  <numeric>\n",
       "      chr1-1001860-1002126_CDK11B                   2.01055\n",
       "     chr1-10022225-10023589_KIF1B                   2.21042\n",
       "  chr1-10022225-10023589_SLC25A33                   4.09691\n",
       "  chr1-100265575-100266826_CDC14A                   3.09691\n",
       "  chr1-10032414-10033799_SLC25A33                   2.68194\n",
       "                              ...                       ...\n",
       "   chrX-77899134-77899778_CYSLTR1                  -3.49485\n",
       "   chrX-81120681-81122301_SH3BGRL                  -3.31876\n",
       "       chrX-9378667-9379857_TBL1X                   4.09691\n",
       "    chrX-96722435-96723511_DIAPH2                   2.68194\n",
       "         chrX-989778-990654_IL3RA                   3.79588\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.reuse.25000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "49dab076-95ff-4c9b-ae00-15bbb4666c5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "FALSE  TRUE \n",
       " 3486 35150 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "table(pairs.E2G[pairs.E2G$boot_basic_p < 0.01]$PairName %in% pairs.E2G.reuse.25000$PairName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2d41238d-e833-4e0e-add5-bb56bb615c1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr11\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr12\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr13\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr14\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr15\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr16\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr17\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr18\"\n",
      "[1] \"chr19\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr2\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr20\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr21\"\n",
      "[1] \"chr22\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr3\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr4\"\n",
      "[1] \"chr5\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr6\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr7\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr8\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr9\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chrX\"\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr = pairs.E2G[seqnames(pairs.E2G) == chr.tmp]\n",
    "  pairs.E2G.chr.res.0 = pairs.E2G.chr[!is.na(pairs.E2G.chr$boot_basic_p) & pairs.E2G.chr$boot_basic_p < 0.01]\n",
    "  pairs.E2G.chr.keep.0 = pairs.E2G.chr[!pairs.E2G.chr$PairName %in% pairs.E2G.chr.res.0$PairName]\n",
    "  pairs.E2G.chr.res = pairs.E2G.chr.res.0[!pairs.E2G.chr.res.0$PairName %in% pairs.E2G.reuse.25000$PairName]\n",
    "  pairs.E2G.chr.keep.1 = pairs.E2G.reuse.25000[pairs.E2G.reuse.25000$PairName %in% pairs.E2G.chr.res.0$PairName]\n",
    "\n",
    "  gene_peak.chr = as.data.frame(mcols(pairs.E2G.chr.res)[,c(\"TargetGene\",\"PeakName\")])\n",
    "  list.gene_peak.chr <- split(gene_peak.chr, seq_len(min(n.cores,floor(length(pairs.E2G.chr.res)/2))))\n",
    "  \n",
    "  list.obj.SCENT.chr <- lapply(list.gene_peak.chr, function(gene_peak.tmp) {\n",
    "    obj.SCENT.tmp <- CreateSCENTObj(\n",
    "      rna = matrix.rna.rename[rownames(matrix.rna.rename) %in% gene_peak.tmp$TargetGene,],\n",
    "      atac = matrix.atac[rownames(matrix.atac) %in% gene_peak.tmp$PeakName,], \n",
    "      meta.data = meta.data,\n",
    "      peak.info = gene_peak.tmp,\n",
    "      covariates = c(\"log.nUMI\",\"percent.mito\"), \n",
    "      celltypes = \"celltype\"\n",
    "    )\n",
    "  })\n",
    "  \n",
    "  cl <- makeCluster(min(n.cores,floor(length(pairs.E2G.chr.res)/2)))\n",
    "  clusterExport(cl, varlist=c(\"SCENT_algorithm.modified\"))\n",
    "  list.SCENT.result.chr <- parLapply(cl, list.obj.SCENT.chr, function(obj.SCENT.tmp) {\n",
    "    obj.SCENT.tmp <- SCENT_algorithm.modified(object = obj.SCENT.tmp)\n",
    "    return(obj.SCENT.tmp@SCENT.result)\n",
    "  })\n",
    "  stopCluster(cl)\n",
    "  \n",
    "  \n",
    "  df.SCENT.result.chr = do.call(rbind,list.SCENT.result.chr)\n",
    "  rownames(df.SCENT.result.chr) = paste(df.SCENT.result.chr$peak,\n",
    "                                        df.SCENT.result.chr$gene,\n",
    "                                        sep = \"_\")\n",
    "\n",
    "  pairs.E2G.chr.res = pairs.E2G.chr.res[pairs.E2G.chr.res$PairName %in% rownames(df.SCENT.result.chr)]\n",
    "  mcols(pairs.E2G.chr.res)[,c(\"beta\",\"se\",\"z\",\"p\",\"boot_basic_p\")] = \n",
    "    df.SCENT.result.chr[pairs.E2G.chr.res$PairName,c(\"beta\",\"se\",\"z\",\"p\",\"boot_basic_p\")]\n",
    "  pairs.E2G.chr.res$boot_basic_p.log10 = -log10(pairs.E2G.chr.res$boot_basic_p)\n",
    "  pairs.E2G.chr.res$sign = 1\n",
    "  pairs.E2G.chr.res$sign[pairs.E2G.chr.res$beta < 0] = -1\n",
    "  pairs.E2G.chr.res$boot_basic_p.log10.signed = pairs.E2G.chr.res$boot_basic_p.log10 * pairs.E2G.chr.res$sign\n",
    "  pairs.E2G.chr.merge = c(pairs.E2G.chr.res,\n",
    "                          pairs.E2G.chr.keep.0,\n",
    "                          pairs.E2G.chr.keep.1)\n",
    "  saveRDS(pairs.E2G.chr.merge,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.merge)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00278bea-8baf-4ed0-93d5-1289f05fa36a",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 318489 ranges and 11 metadata columns:\n",
       "                                      seqnames            ranges strand |\n",
       "                                         <Rle>         <IRanges>  <Rle> |\n",
       "                       chr1.chr1.chr1     chr1   1012920-1014120      * |\n",
       "                       chr1.chr1.chr1     chr1 10147738-10148559      * |\n",
       "                       chr1.chr1.chr1     chr1   1019144-1020371      * |\n",
       "                       chr1.chr1.chr1     chr1   1038601-1039080      * |\n",
       "                       chr1.chr1.chr1     chr1   1058756-1060110      * |\n",
       "                                  ...      ...               ...    ... .\n",
       "  chrX.chrX-77899134-77899778_CYSLTR1     chrX 77899134-77899778      * |\n",
       "  chrX.chrX-81120681-81122301_SH3BGRL     chrX 81120681-81122301      * |\n",
       "      chrX.chrX-9378667-9379857_TBL1X     chrX   9378667-9379857      * |\n",
       "   chrX.chrX-96722435-96723511_DIAPH2     chrX 96722435-96723511      * |\n",
       "        chrX.chrX-989778-990654_IL3RA     chrX     989778-990654      * |\n",
       "                                       TargetGene               PeakName\n",
       "                                      <character>            <character>\n",
       "                       chr1.chr1.chr1      UBE2J2   chr1-1012920-1014120\n",
       "                       chr1.chr1.chr1       UBE4B chr1-10147738-10148559\n",
       "                       chr1.chr1.chr1        NADK   chr1-1019144-1020371\n",
       "                       chr1.chr1.chr1       ISG15   chr1-1038601-1039080\n",
       "                       chr1.chr1.chr1        NADK   chr1-1058756-1060110\n",
       "                                  ...         ...                    ...\n",
       "  chrX.chrX-77899134-77899778_CYSLTR1     CYSLTR1 chrX-77899134-77899778\n",
       "  chrX.chrX-81120681-81122301_SH3BGRL     SH3BGRL chrX-81120681-81122301\n",
       "      chrX.chrX-9378667-9379857_TBL1X       TBL1X   chrX-9378667-9379857\n",
       "   chrX.chrX-96722435-96723511_DIAPH2      DIAPH2 chrX-96722435-96723511\n",
       "        chrX.chrX-989778-990654_IL3RA       IL3RA     chrX-989778-990654\n",
       "                                                    PairName       beta\n",
       "                                                 <character>  <numeric>\n",
       "                       chr1.chr1.chr1 chr1-1012920-1014120..   0.152425\n",
       "                       chr1.chr1.chr1 chr1-10147738-101485..   0.132616\n",
       "                       chr1.chr1.chr1 chr1-1019144-1020371..  -0.104505\n",
       "                       chr1.chr1.chr1 chr1-1038601-1039080..  -0.374559\n",
       "                       chr1.chr1.chr1 chr1-1058756-1060110..  -0.128388\n",
       "                                  ...                    ...        ...\n",
       "  chrX.chrX-77899134-77899778_CYSLTR1 chrX-77899134-778997.. -0.4792389\n",
       "  chrX.chrX-81120681-81122301_SH3BGRL chrX-81120681-811223.. -0.1499563\n",
       "      chrX.chrX-9378667-9379857_TBL1X chrX-9378667-9379857..  0.1630958\n",
       "   chrX.chrX-96722435-96723511_DIAPH2 chrX-96722435-967235..  0.0833827\n",
       "        chrX.chrX-989778-990654_IL3RA chrX-989778-990654_I..  0.3102046\n",
       "                                             se         z           p\n",
       "                                      <numeric> <numeric>   <numeric>\n",
       "                       chr1.chr1.chr1 0.0620578   2.45618 0.014042137\n",
       "                       chr1.chr1.chr1 0.0471038   2.81539 0.004871749\n",
       "                       chr1.chr1.chr1 0.0384375  -2.71883 0.006551309\n",
       "                       chr1.chr1.chr1 0.1028667  -3.64121 0.000271359\n",
       "                       chr1.chr1.chr1 0.0467201  -2.74802 0.005995713\n",
       "                                  ...       ...       ...         ...\n",
       "  chrX.chrX-77899134-77899778_CYSLTR1 0.0950917  -5.03975 4.66132e-07\n",
       "  chrX.chrX-81120681-81122301_SH3BGRL 0.0329712  -4.54809 5.41340e-06\n",
       "      chrX.chrX-9378667-9379857_TBL1X 0.0264363   6.16938 6.85587e-10\n",
       "   chrX.chrX-96722435-96723511_DIAPH2 0.0201098   4.14637 3.37783e-05\n",
       "        chrX.chrX-989778-990654_IL3RA 0.0796438   3.89490 9.82400e-05\n",
       "                                      boot_basic_p boot_basic_p.log10      sign\n",
       "                                         <numeric>          <numeric> <numeric>\n",
       "                       chr1.chr1.chr1      0.01104            1.95703         1\n",
       "                       chr1.chr1.chr1      0.01552            1.80911         1\n",
       "                       chr1.chr1.chr1      0.01072            1.96981        -1\n",
       "                       chr1.chr1.chr1      0.01368            1.86391        -1\n",
       "                       chr1.chr1.chr1      0.01160            1.93554        -1\n",
       "                                  ...          ...                ...       ...\n",
       "  chrX.chrX-77899134-77899778_CYSLTR1      0.00032            3.49485        -1\n",
       "  chrX.chrX-81120681-81122301_SH3BGRL      0.00048            3.31876        -1\n",
       "      chrX.chrX-9378667-9379857_TBL1X      0.00008            4.09691         1\n",
       "   chrX.chrX-96722435-96723511_DIAPH2      0.00208            2.68194         1\n",
       "        chrX.chrX-989778-990654_IL3RA      0.00016            3.79588         1\n",
       "                                      boot_basic_p.log10.signed\n",
       "                                                      <numeric>\n",
       "                       chr1.chr1.chr1                   1.95703\n",
       "                       chr1.chr1.chr1                   1.80911\n",
       "                       chr1.chr1.chr1                  -1.96981\n",
       "                       chr1.chr1.chr1                  -1.86391\n",
       "                       chr1.chr1.chr1                  -1.93554\n",
       "                                  ...                       ...\n",
       "  chrX.chrX-77899134-77899778_CYSLTR1                  -3.49485\n",
       "  chrX.chrX-81120681-81122301_SH3BGRL                  -3.31876\n",
       "      chrX.chrX-9378667-9379857_TBL1X                   4.09691\n",
       "   chrX.chrX-96722435-96723511_DIAPH2                   2.68194\n",
       "        chrX.chrX-989778-990654_IL3RA                   3.79588\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 318489 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>beta</th><th scope=col>se</th><th scope=col>z</th><th scope=col>p</th><th scope=col>boot_basic_p</th><th scope=col>boot_basic_p.log10</th><th scope=col>sign</th><th scope=col>boot_basic_p.log10.signed</th><th scope=col>FDR</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>  1012920</td><td>  1014120</td><td>UBE2J2     </td><td>Myeloid</td><td> 0.15242525</td><td>0.06205776</td><td> 2.456184</td><td>0.0140421368</td><td>0.01104</td><td>1.957031</td><td> 1</td><td> 1.957031</td><td>0.09000269</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10147738</td><td> 10148559</td><td>UBE4B      </td><td>Myeloid</td><td> 0.13261564</td><td>0.04710376</td><td> 2.815394</td><td>0.0048717487</td><td>0.01552</td><td>1.809108</td><td> 1</td><td> 1.809108</td><td>0.11035832</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1019144</td><td>  1020371</td><td>NADK       </td><td>Myeloid</td><td>-0.10450517</td><td>0.03843754</td><td>-2.718831</td><td>0.0065513095</td><td>0.01072</td><td>1.969805</td><td>-1</td><td>-1.969805</td><td>0.08870131</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1038601</td><td>  1039080</td><td>ISG15      </td><td>Myeloid</td><td>-0.37455939</td><td>0.10286672</td><td>-3.641210</td><td>0.0002713592</td><td>0.01368</td><td>1.863914</td><td>-1</td><td>-1.863914</td><td>0.10151280</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1058756</td><td>  1060110</td><td>NADK       </td><td>Myeloid</td><td>-0.12838767</td><td>0.04672014</td><td>-2.748016</td><td>0.0059957130</td><td>0.01160</td><td>1.935542</td><td>-1</td><td>-1.935542</td><td>0.09284460</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1079374</td><td>  1080601</td><td>CDK11B     </td><td>Myeloid</td><td> 0.16449572</td><td>0.06110268</td><td> 2.692120</td><td>0.0070999496</td><td>0.00712</td><td>2.147520</td><td> 1</td><td> 2.147520</td><td>0.06637324</td></tr>\n",
       "\t<tr><td>chr1</td><td>109040911</td><td>109042568</td><td>CLCC1      </td><td>Myeloid</td><td> 0.14251012</td><td>0.05488633</td><td> 2.596459</td><td>0.0094190065</td><td>0.01040</td><td>1.982967</td><td> 1</td><td> 1.982967</td><td>0.08635188</td></tr>\n",
       "\t<tr><td>chr1</td><td>109091688</td><td>109092177</td><td>GPSM2      </td><td>Myeloid</td><td> 0.28891353</td><td>0.10194700</td><td> 2.833958</td><td>0.0045975345</td><td>0.01520</td><td>1.818156</td><td> 1</td><td> 1.818156</td><td>0.10814084</td></tr>\n",
       "\t<tr><td>chr1</td><td>109133440</td><td>109134866</td><td>GPSM2      </td><td>Myeloid</td><td> 0.29310711</td><td>0.09617894</td><td> 3.047519</td><td>0.0023073926</td><td>0.01088</td><td>1.963371</td><td> 1</td><td> 1.963371</td><td>0.08987810</td></tr>\n",
       "\t<tr><td>chr1</td><td>109213476</td><td>109214665</td><td>SARS       </td><td>Myeloid</td><td> 0.21115222</td><td>0.07072164</td><td> 2.985681</td><td>0.0028294796</td><td>0.00272</td><td>2.565431</td><td> 1</td><td> 2.565431</td><td>0.03363841</td></tr>\n",
       "\t<tr><td>chr1</td><td>109573504</td><td>109574558</td><td>AHCYL1     </td><td>Myeloid</td><td>-0.27273708</td><td>0.09505956</td><td>-2.869118</td><td>0.0041161823</td><td>0.00992</td><td>2.003488</td><td>-1</td><td>-2.003488</td><td>0.08506532</td></tr>\n",
       "\t<tr><td>chr1</td><td>109643175</td><td>109643836</td><td>AMPD2      </td><td>Myeloid</td><td> 0.16632329</td><td>0.06444675</td><td> 2.580786</td><td>0.0098575594</td><td>0.01136</td><td>1.944622</td><td> 1</td><td> 1.944622</td><td>0.09112061</td></tr>\n",
       "\t<tr><td>chr1</td><td>110034075</td><td>110035478</td><td>LAMTOR5-AS1</td><td>Myeloid</td><td> 0.15030050</td><td>0.05475852</td><td> 2.744787</td><td>0.0060550092</td><td>0.01104</td><td>1.957031</td><td> 1</td><td> 1.957031</td><td>0.09000269</td></tr>\n",
       "\t<tr><td>chr1</td><td>110068666</td><td>110070653</td><td>AMPD2      </td><td>Myeloid</td><td> 0.15144738</td><td>0.05508265</td><td> 2.749457</td><td>0.0059694128</td><td>0.01000</td><td>2.000000</td><td> 1</td><td> 2.000000</td><td>0.08567751</td></tr>\n",
       "\t<tr><td>chr1</td><td>110424462</td><td>110425171</td><td>STRIP1     </td><td>Myeloid</td><td> 0.26196256</td><td>0.11672272</td><td> 2.244315</td><td>0.0248121302</td><td>0.01616</td><td>1.791559</td><td> 1</td><td> 1.791559</td><td>0.11279630</td></tr>\n",
       "\t<tr><td>chr1</td><td>110619222</td><td>110620626</td><td>RBM15      </td><td>Myeloid</td><td> 0.20249370</td><td>0.07707231</td><td> 2.627321</td><td>0.0086060090</td><td>0.00832</td><td>2.079877</td><td> 1</td><td> 2.079877</td><td>0.07428107</td></tr>\n",
       "\t<tr><td>chr1</td><td>110627679</td><td>110628608</td><td>CEPT1      </td><td>Myeloid</td><td>-0.14393662</td><td>0.05169942</td><td>-2.784105</td><td>0.0053675675</td><td>0.01152</td><td>1.938548</td><td>-1</td><td>-1.938548</td><td>0.09227386</td></tr>\n",
       "\t<tr><td>chr1</td><td>110963175</td><td>110964482</td><td>FAM212B    </td><td>Myeloid</td><td>-0.19962694</td><td>0.05904812</td><td>-3.380750</td><td>0.0007228823</td><td>0.00192</td><td>2.716699</td><td>-1</td><td>-2.716699</td><td>0.02617718</td></tr>\n",
       "\t<tr><td>chr1</td><td>111139345</td><td>111140838</td><td>CEPT1      </td><td>Myeloid</td><td>-0.10139644</td><td>0.03657225</td><td>-2.772497</td><td>0.0055628110</td><td>0.01392</td><td>1.856361</td><td>-1</td><td>-1.856361</td><td>0.10319515</td></tr>\n",
       "\t<tr><td>chr1</td><td>111139345</td><td>111140838</td><td>FAM212B    </td><td>Myeloid</td><td>-0.24226979</td><td>0.06737893</td><td>-3.595631</td><td>0.0003236058</td><td>0.00128</td><td>2.892790</td><td>-1</td><td>-2.892790</td><td>0.01952049</td></tr>\n",
       "\t<tr><td>chr1</td><td>111616317</td><td>111616816</td><td>CEPT1      </td><td>Myeloid</td><td>-0.15529944</td><td>0.05521685</td><td>-2.812537</td><td>0.0049152311</td><td>0.01208</td><td>1.917933</td><td>-1</td><td>-1.917933</td><td>0.09412931</td></tr>\n",
       "\t<tr><td>chr1</td><td>111738407</td><td>111740718</td><td>ATP5F1     </td><td>Myeloid</td><td>-0.16005882</td><td>0.05621385</td><td>-2.847320</td><td>0.0044089029</td><td>0.00968</td><td>2.014125</td><td>-1</td><td>-2.014125</td><td>0.08344539</td></tr>\n",
       "\t<tr><td>chr1</td><td>111738407</td><td>111740718</td><td>FAM212B    </td><td>Myeloid</td><td>-0.20705799</td><td>0.06124127</td><td>-3.381021</td><td>0.0007221707</td><td>0.00232</td><td>2.634512</td><td>-1</td><td>-2.634512</td><td>0.03002050</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1121806</td><td>  1123448</td><td>GNB1       </td><td>Myeloid</td><td> 0.07788593</td><td>0.02661965</td><td> 2.925881</td><td>0.0034348180</td><td>0.00736</td><td>2.133122</td><td> 1</td><td> 2.133122</td><td>0.06794038</td></tr>\n",
       "\t<tr><td>chr1</td><td>112706534</td><td>112707637</td><td>FAM212B    </td><td>Myeloid</td><td>-0.34345561</td><td>0.09626244</td><td>-3.567909</td><td>0.0003598416</td><td>0.00232</td><td>2.634512</td><td>-1</td><td>-2.634512</td><td>0.03002050</td></tr>\n",
       "\t<tr><td>chr1</td><td>113140382</td><td>113141992</td><td>PTPN22     </td><td>Myeloid</td><td> 0.15969368</td><td>0.05463950</td><td> 2.922678</td><td>0.0034703502</td><td>0.01192</td><td>1.923724</td><td> 1</td><td> 1.923724</td><td>0.09356087</td></tr>\n",
       "\t<tr><td>chr1</td><td>113389901</td><td>113391532</td><td>PHTF1      </td><td>Myeloid</td><td> 0.14690124</td><td>0.04562050</td><td> 3.220071</td><td>0.0012815902</td><td>0.00984</td><td>2.007005</td><td> 1</td><td> 2.007005</td><td>0.08453407</td></tr>\n",
       "\t<tr><td>chr1</td><td>113389901</td><td>113391532</td><td>PTPN22     </td><td>Myeloid</td><td> 0.11077072</td><td>0.03523556</td><td> 3.143719</td><td>0.0016681536</td><td>0.00808</td><td>2.092589</td><td> 1</td><td> 2.092589</td><td>0.07274398</td></tr>\n",
       "\t<tr><td>chr1</td><td>113928993</td><td>113931084</td><td>PHTF1      </td><td>Myeloid</td><td> 0.14114251</td><td>0.04262542</td><td> 3.311229</td><td>0.0009288722</td><td>0.01000</td><td>2.000000</td><td> 1</td><td> 2.000000</td><td>0.08567751</td></tr>\n",
       "\t<tr><td>chr1</td><td>114040080</td><td>114041198</td><td>PTPN22     </td><td>Myeloid</td><td> 0.14492475</td><td>0.04467912</td><td> 3.243680</td><td>0.0011799633</td><td>0.00920</td><td>2.036212</td><td> 1</td><td> 2.036212</td><td>0.08017564</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>53071574</td><td>53073601</td><td>SMC1A  </td><td>Myeloid</td><td> 0.19046290</td><td>0.05344579</td><td>  3.563665</td><td> 3.657126e-04</td><td>0.00112</td><td>2.950782</td><td> 1</td><td> 2.950782</td><td>0.017663168</td></tr>\n",
       "\t<tr><td>chrX</td><td>53081843</td><td>53082907</td><td>SMC1A  </td><td>Myeloid</td><td> 0.13253351</td><td>0.04073004</td><td>  3.253950</td><td> 1.138123e-03</td><td>0.00240</td><td>2.619789</td><td> 1</td><td> 2.619789</td><td>0.030759501</td></tr>\n",
       "\t<tr><td>chrX</td><td>53224384</td><td>53225810</td><td>SMC1A  </td><td>Myeloid</td><td> 0.13246754</td><td>0.03962539</td><td>  3.342997</td><td> 8.287887e-04</td><td>0.00216</td><td>2.665546</td><td> 1</td><td> 2.665546</td><td>0.028515492</td></tr>\n",
       "\t<tr><td>chrX</td><td>53280164</td><td>53281893</td><td>KDM5C  </td><td>Myeloid</td><td> 0.27843651</td><td>0.09001215</td><td>  3.093321</td><td> 1.979296e-03</td><td>0.00208</td><td>2.681937</td><td> 1</td><td> 2.681937</td><td>0.027737601</td></tr>\n",
       "\t<tr><td>chrX</td><td>53287345</td><td>53288709</td><td>SMC1A  </td><td>Myeloid</td><td> 0.23446749</td><td>0.05263760</td><td>  4.454373</td><td> 8.413885e-06</td><td>0.00008</td><td>4.096910</td><td> 1</td><td> 4.096910</td><td>0.002128937</td></tr>\n",
       "\t<tr><td>chrX</td><td>53374180</td><td>53375241</td><td>SMC1A  </td><td>Myeloid</td><td> 0.21616396</td><td>0.05898753</td><td>  3.664571</td><td> 2.477537e-04</td><td>0.00072</td><td>3.142668</td><td> 1</td><td> 3.142668</td><td>0.012651003</td></tr>\n",
       "\t<tr><td>chrX</td><td>53421910</td><td>53423211</td><td>KDM5C  </td><td>Myeloid</td><td> 0.19170571</td><td>0.04925167</td><td>  3.892370</td><td> 9.926981e-05</td><td>0.00008</td><td>4.096910</td><td> 1</td><td> 4.096910</td><td>0.002128937</td></tr>\n",
       "\t<tr><td>chrX</td><td>53421910</td><td>53423211</td><td>SMC1A  </td><td>Myeloid</td><td> 0.11762703</td><td>0.03621192</td><td>  3.248295</td><td> 1.160986e-03</td><td>0.00376</td><td>2.424812</td><td> 1</td><td> 2.424812</td><td>0.042366045</td></tr>\n",
       "\t<tr><td>chrX</td><td>54182281</td><td>54183769</td><td>SMC1A  </td><td>Myeloid</td><td> 0.15326493</td><td>0.04967353</td><td>  3.085444</td><td> 2.032482e-03</td><td>0.00488</td><td>2.311580</td><td> 1</td><td> 2.311580</td><td>0.050918173</td></tr>\n",
       "\t<tr><td>chrX</td><td>65997710</td><td>65998828</td><td>MSN    </td><td>Myeloid</td><td>-0.11387141</td><td>0.03133574</td><td> -3.633915</td><td> 2.791529e-04</td><td>0.00360</td><td>2.443697</td><td>-1</td><td>-2.443697</td><td>0.041124835</td></tr>\n",
       "\t<tr><td>chrX</td><td>71067720</td><td>71068957</td><td>NHSL2  </td><td>Myeloid</td><td>-0.19718401</td><td>0.04503518</td><td> -4.378444</td><td> 1.195299e-05</td><td>0.00008</td><td>4.096910</td><td>-1</td><td>-4.096910</td><td>0.002128937</td></tr>\n",
       "\t<tr><td>chrX</td><td>71111138</td><td>71112478</td><td>NHSL2  </td><td>Myeloid</td><td>-0.22457700</td><td>0.06742250</td><td> -3.330891</td><td> 8.656843e-04</td><td>0.00560</td><td>2.251812</td><td>-1</td><td>-2.251812</td><td>0.055987519</td></tr>\n",
       "\t<tr><td>chrX</td><td>71253872</td><td>71255695</td><td>NHSL2  </td><td>Myeloid</td><td>-0.30942868</td><td>0.05915155</td><td> -5.231117</td><td> 1.684888e-07</td><td>0.00008</td><td>4.096910</td><td>-1</td><td>-4.096910</td><td>0.002128937</td></tr>\n",
       "\t<tr><td>chrX</td><td>71253872</td><td>71255695</td><td>NONO   </td><td>Myeloid</td><td> 0.18692709</td><td>0.05182659</td><td>  3.606779</td><td> 3.100209e-04</td><td>0.00088</td><td>3.055517</td><td> 1</td><td> 3.055517</td><td>0.014729363</td></tr>\n",
       "\t<tr><td>chrX</td><td>71253872</td><td>71255695</td><td>OGT    </td><td>Myeloid</td><td> 0.10905648</td><td>0.02955893</td><td>  3.689459</td><td> 2.247310e-04</td><td>0.00040</td><td>3.397940</td><td> 1</td><td> 3.397940</td><td>0.008014822</td></tr>\n",
       "\t<tr><td>chrX</td><td>71282778</td><td>71284216</td><td>NONO   </td><td>Myeloid</td><td> 0.14348201</td><td>0.04543674</td><td>  3.157841</td><td> 1.589421e-03</td><td>0.00296</td><td>2.528708</td><td> 1</td><td> 2.528708</td><td>0.035693149</td></tr>\n",
       "\t<tr><td>chrX</td><td>71365597</td><td>71367182</td><td>OGT    </td><td>Myeloid</td><td> 0.07541851</td><td>0.02416346</td><td>  3.121180</td><td> 1.801279e-03</td><td>0.00368</td><td>2.434152</td><td> 1</td><td> 2.434152</td><td>0.041748220</td></tr>\n",
       "\t<tr><td>chrX</td><td>71540871</td><td>71542054</td><td>RPS4X  </td><td>Myeloid</td><td>-0.18319556</td><td>0.03024068</td><td> -6.057919</td><td> 1.378942e-09</td><td>0.00480</td><td>2.318759</td><td>-1</td><td>-2.318759</td><td>0.050317530</td></tr>\n",
       "\t<tr><td>chrX</td><td>72107322</td><td>72108644</td><td>RPS4X  </td><td>Myeloid</td><td>-0.30474042</td><td>0.03040675</td><td>-10.022132</td><td> 1.218459e-23</td><td>0.00008</td><td>4.096910</td><td>-1</td><td>-4.096910</td><td>0.002128937</td></tr>\n",
       "\t<tr><td>chrX</td><td>72181025</td><td>72182115</td><td>NHSL2  </td><td>Myeloid</td><td>-0.17016975</td><td>0.05128859</td><td> -3.317887</td><td> 9.070110e-04</td><td>0.00368</td><td>2.434152</td><td>-1</td><td>-2.434152</td><td>0.041748220</td></tr>\n",
       "\t<tr><td>chrX</td><td>73850802</td><td>73851931</td><td>JPX    </td><td>Myeloid</td><td> 0.17253377</td><td>0.03542301</td><td>  4.870669</td><td> 1.112208e-06</td><td>0.00016</td><td>3.795880</td><td> 1</td><td> 3.795880</td><td>0.003831735</td></tr>\n",
       "\t<tr><td>chrX</td><td>73850802</td><td>73851931</td><td>XIST   </td><td>Myeloid</td><td> 0.54370458</td><td>0.01779508</td><td> 30.553648</td><td>5.057667e-205</td><td>0.00008</td><td>4.096910</td><td> 1</td><td> 4.096910</td><td>0.002128937</td></tr>\n",
       "\t<tr><td>chrX</td><td>73943694</td><td>73944847</td><td>XIST   </td><td>Myeloid</td><td> 0.18747851</td><td>0.01836553</td><td> 10.208175</td><td> 1.822602e-24</td><td>0.00008</td><td>4.096910</td><td> 1</td><td> 4.096910</td><td>0.002128937</td></tr>\n",
       "\t<tr><td>chrX</td><td>74291819</td><td>74293013</td><td>XIST   </td><td>Myeloid</td><td> 0.11040114</td><td>0.01830518</td><td>  6.031141</td><td> 1.628058e-09</td><td>0.00024</td><td>3.619789</td><td> 1</td><td> 3.619789</td><td>0.005347514</td></tr>\n",
       "\t<tr><td>chrX</td><td>74613836</td><td>74615059</td><td>XIST   </td><td>Myeloid</td><td> 0.07637641</td><td>0.01694361</td><td>  4.507683</td><td> 6.553951e-06</td><td>0.00792</td><td>2.101275</td><td> 1</td><td> 2.101275</td><td>0.071688537</td></tr>\n",
       "\t<tr><td>chrX</td><td>77899134</td><td>77899778</td><td>CYSLTR1</td><td>Myeloid</td><td>-0.47923887</td><td>0.09509173</td><td> -5.039753</td><td> 4.661319e-07</td><td>0.00032</td><td>3.494850</td><td>-1</td><td>-3.494850</td><td>0.006722723</td></tr>\n",
       "\t<tr><td>chrX</td><td>81120681</td><td>81122301</td><td>SH3BGRL</td><td>Myeloid</td><td>-0.14995632</td><td>0.03297125</td><td> -4.548094</td><td> 5.413401e-06</td><td>0.00048</td><td>3.318759</td><td>-1</td><td>-3.318759</td><td>0.009255598</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9378667</td><td> 9379857</td><td>TBL1X  </td><td>Myeloid</td><td> 0.16309584</td><td>0.02643635</td><td>  6.169379</td><td> 6.855870e-10</td><td>0.00008</td><td>4.096910</td><td> 1</td><td> 4.096910</td><td>0.002128937</td></tr>\n",
       "\t<tr><td>chrX</td><td>96722435</td><td>96723511</td><td>DIAPH2 </td><td>Myeloid</td><td> 0.08338271</td><td>0.02010979</td><td>  4.146373</td><td> 3.377832e-05</td><td>0.00208</td><td>2.681937</td><td> 1</td><td> 2.681937</td><td>0.027737601</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989778</td><td>  990654</td><td>IL3RA  </td><td>Myeloid</td><td> 0.31020460</td><td>0.07964383</td><td>  3.894898</td><td> 9.824001e-05</td><td>0.00016</td><td>3.795880</td><td> 1</td><td> 3.795880</td><td>0.003831735</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 318489 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " chr & start & end & TargetGene & CellType & beta & se & z & p & boot\\_basic\\_p & boot\\_basic\\_p.log10 & sign & boot\\_basic\\_p.log10.signed & FDR\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &   1012920 &   1014120 & UBE2J2      & Myeloid &  0.15242525 & 0.06205776 &  2.456184 & 0.0140421368 & 0.01104 & 1.957031 &  1 &  1.957031 & 0.09000269\\\\\n",
       "\t chr1 &  10147738 &  10148559 & UBE4B       & Myeloid &  0.13261564 & 0.04710376 &  2.815394 & 0.0048717487 & 0.01552 & 1.809108 &  1 &  1.809108 & 0.11035832\\\\\n",
       "\t chr1 &   1019144 &   1020371 & NADK        & Myeloid & -0.10450517 & 0.03843754 & -2.718831 & 0.0065513095 & 0.01072 & 1.969805 & -1 & -1.969805 & 0.08870131\\\\\n",
       "\t chr1 &   1038601 &   1039080 & ISG15       & Myeloid & -0.37455939 & 0.10286672 & -3.641210 & 0.0002713592 & 0.01368 & 1.863914 & -1 & -1.863914 & 0.10151280\\\\\n",
       "\t chr1 &   1058756 &   1060110 & NADK        & Myeloid & -0.12838767 & 0.04672014 & -2.748016 & 0.0059957130 & 0.01160 & 1.935542 & -1 & -1.935542 & 0.09284460\\\\\n",
       "\t chr1 &   1079374 &   1080601 & CDK11B      & Myeloid &  0.16449572 & 0.06110268 &  2.692120 & 0.0070999496 & 0.00712 & 2.147520 &  1 &  2.147520 & 0.06637324\\\\\n",
       "\t chr1 & 109040911 & 109042568 & CLCC1       & Myeloid &  0.14251012 & 0.05488633 &  2.596459 & 0.0094190065 & 0.01040 & 1.982967 &  1 &  1.982967 & 0.08635188\\\\\n",
       "\t chr1 & 109091688 & 109092177 & GPSM2       & Myeloid &  0.28891353 & 0.10194700 &  2.833958 & 0.0045975345 & 0.01520 & 1.818156 &  1 &  1.818156 & 0.10814084\\\\\n",
       "\t chr1 & 109133440 & 109134866 & GPSM2       & Myeloid &  0.29310711 & 0.09617894 &  3.047519 & 0.0023073926 & 0.01088 & 1.963371 &  1 &  1.963371 & 0.08987810\\\\\n",
       "\t chr1 & 109213476 & 109214665 & SARS        & Myeloid &  0.21115222 & 0.07072164 &  2.985681 & 0.0028294796 & 0.00272 & 2.565431 &  1 &  2.565431 & 0.03363841\\\\\n",
       "\t chr1 & 109573504 & 109574558 & AHCYL1      & Myeloid & -0.27273708 & 0.09505956 & -2.869118 & 0.0041161823 & 0.00992 & 2.003488 & -1 & -2.003488 & 0.08506532\\\\\n",
       "\t chr1 & 109643175 & 109643836 & AMPD2       & Myeloid &  0.16632329 & 0.06444675 &  2.580786 & 0.0098575594 & 0.01136 & 1.944622 &  1 &  1.944622 & 0.09112061\\\\\n",
       "\t chr1 & 110034075 & 110035478 & LAMTOR5-AS1 & Myeloid &  0.15030050 & 0.05475852 &  2.744787 & 0.0060550092 & 0.01104 & 1.957031 &  1 &  1.957031 & 0.09000269\\\\\n",
       "\t chr1 & 110068666 & 110070653 & AMPD2       & Myeloid &  0.15144738 & 0.05508265 &  2.749457 & 0.0059694128 & 0.01000 & 2.000000 &  1 &  2.000000 & 0.08567751\\\\\n",
       "\t chr1 & 110424462 & 110425171 & STRIP1      & Myeloid &  0.26196256 & 0.11672272 &  2.244315 & 0.0248121302 & 0.01616 & 1.791559 &  1 &  1.791559 & 0.11279630\\\\\n",
       "\t chr1 & 110619222 & 110620626 & RBM15       & Myeloid &  0.20249370 & 0.07707231 &  2.627321 & 0.0086060090 & 0.00832 & 2.079877 &  1 &  2.079877 & 0.07428107\\\\\n",
       "\t chr1 & 110627679 & 110628608 & CEPT1       & Myeloid & -0.14393662 & 0.05169942 & -2.784105 & 0.0053675675 & 0.01152 & 1.938548 & -1 & -1.938548 & 0.09227386\\\\\n",
       "\t chr1 & 110963175 & 110964482 & FAM212B     & Myeloid & -0.19962694 & 0.05904812 & -3.380750 & 0.0007228823 & 0.00192 & 2.716699 & -1 & -2.716699 & 0.02617718\\\\\n",
       "\t chr1 & 111139345 & 111140838 & CEPT1       & Myeloid & -0.10139644 & 0.03657225 & -2.772497 & 0.0055628110 & 0.01392 & 1.856361 & -1 & -1.856361 & 0.10319515\\\\\n",
       "\t chr1 & 111139345 & 111140838 & FAM212B     & Myeloid & -0.24226979 & 0.06737893 & -3.595631 & 0.0003236058 & 0.00128 & 2.892790 & -1 & -2.892790 & 0.01952049\\\\\n",
       "\t chr1 & 111616317 & 111616816 & CEPT1       & Myeloid & -0.15529944 & 0.05521685 & -2.812537 & 0.0049152311 & 0.01208 & 1.917933 & -1 & -1.917933 & 0.09412931\\\\\n",
       "\t chr1 & 111738407 & 111740718 & ATP5F1      & Myeloid & -0.16005882 & 0.05621385 & -2.847320 & 0.0044089029 & 0.00968 & 2.014125 & -1 & -2.014125 & 0.08344539\\\\\n",
       "\t chr1 & 111738407 & 111740718 & FAM212B     & Myeloid & -0.20705799 & 0.06124127 & -3.381021 & 0.0007221707 & 0.00232 & 2.634512 & -1 & -2.634512 & 0.03002050\\\\\n",
       "\t chr1 &   1121806 &   1123448 & GNB1        & Myeloid &  0.07788593 & 0.02661965 &  2.925881 & 0.0034348180 & 0.00736 & 2.133122 &  1 &  2.133122 & 0.06794038\\\\\n",
       "\t chr1 & 112706534 & 112707637 & FAM212B     & Myeloid & -0.34345561 & 0.09626244 & -3.567909 & 0.0003598416 & 0.00232 & 2.634512 & -1 & -2.634512 & 0.03002050\\\\\n",
       "\t chr1 & 113140382 & 113141992 & PTPN22      & Myeloid &  0.15969368 & 0.05463950 &  2.922678 & 0.0034703502 & 0.01192 & 1.923724 &  1 &  1.923724 & 0.09356087\\\\\n",
       "\t chr1 & 113389901 & 113391532 & PHTF1       & Myeloid &  0.14690124 & 0.04562050 &  3.220071 & 0.0012815902 & 0.00984 & 2.007005 &  1 &  2.007005 & 0.08453407\\\\\n",
       "\t chr1 & 113389901 & 113391532 & PTPN22      & Myeloid &  0.11077072 & 0.03523556 &  3.143719 & 0.0016681536 & 0.00808 & 2.092589 &  1 &  2.092589 & 0.07274398\\\\\n",
       "\t chr1 & 113928993 & 113931084 & PHTF1       & Myeloid &  0.14114251 & 0.04262542 &  3.311229 & 0.0009288722 & 0.01000 & 2.000000 &  1 &  2.000000 & 0.08567751\\\\\n",
       "\t chr1 & 114040080 & 114041198 & PTPN22      & Myeloid &  0.14492475 & 0.04467912 &  3.243680 & 0.0011799633 & 0.00920 & 2.036212 &  1 &  2.036212 & 0.08017564\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 53071574 & 53073601 & SMC1A   & Myeloid &  0.19046290 & 0.05344579 &   3.563665 &  3.657126e-04 & 0.00112 & 2.950782 &  1 &  2.950782 & 0.017663168\\\\\n",
       "\t chrX & 53081843 & 53082907 & SMC1A   & Myeloid &  0.13253351 & 0.04073004 &   3.253950 &  1.138123e-03 & 0.00240 & 2.619789 &  1 &  2.619789 & 0.030759501\\\\\n",
       "\t chrX & 53224384 & 53225810 & SMC1A   & Myeloid &  0.13246754 & 0.03962539 &   3.342997 &  8.287887e-04 & 0.00216 & 2.665546 &  1 &  2.665546 & 0.028515492\\\\\n",
       "\t chrX & 53280164 & 53281893 & KDM5C   & Myeloid &  0.27843651 & 0.09001215 &   3.093321 &  1.979296e-03 & 0.00208 & 2.681937 &  1 &  2.681937 & 0.027737601\\\\\n",
       "\t chrX & 53287345 & 53288709 & SMC1A   & Myeloid &  0.23446749 & 0.05263760 &   4.454373 &  8.413885e-06 & 0.00008 & 4.096910 &  1 &  4.096910 & 0.002128937\\\\\n",
       "\t chrX & 53374180 & 53375241 & SMC1A   & Myeloid &  0.21616396 & 0.05898753 &   3.664571 &  2.477537e-04 & 0.00072 & 3.142668 &  1 &  3.142668 & 0.012651003\\\\\n",
       "\t chrX & 53421910 & 53423211 & KDM5C   & Myeloid &  0.19170571 & 0.04925167 &   3.892370 &  9.926981e-05 & 0.00008 & 4.096910 &  1 &  4.096910 & 0.002128937\\\\\n",
       "\t chrX & 53421910 & 53423211 & SMC1A   & Myeloid &  0.11762703 & 0.03621192 &   3.248295 &  1.160986e-03 & 0.00376 & 2.424812 &  1 &  2.424812 & 0.042366045\\\\\n",
       "\t chrX & 54182281 & 54183769 & SMC1A   & Myeloid &  0.15326493 & 0.04967353 &   3.085444 &  2.032482e-03 & 0.00488 & 2.311580 &  1 &  2.311580 & 0.050918173\\\\\n",
       "\t chrX & 65997710 & 65998828 & MSN     & Myeloid & -0.11387141 & 0.03133574 &  -3.633915 &  2.791529e-04 & 0.00360 & 2.443697 & -1 & -2.443697 & 0.041124835\\\\\n",
       "\t chrX & 71067720 & 71068957 & NHSL2   & Myeloid & -0.19718401 & 0.04503518 &  -4.378444 &  1.195299e-05 & 0.00008 & 4.096910 & -1 & -4.096910 & 0.002128937\\\\\n",
       "\t chrX & 71111138 & 71112478 & NHSL2   & Myeloid & -0.22457700 & 0.06742250 &  -3.330891 &  8.656843e-04 & 0.00560 & 2.251812 & -1 & -2.251812 & 0.055987519\\\\\n",
       "\t chrX & 71253872 & 71255695 & NHSL2   & Myeloid & -0.30942868 & 0.05915155 &  -5.231117 &  1.684888e-07 & 0.00008 & 4.096910 & -1 & -4.096910 & 0.002128937\\\\\n",
       "\t chrX & 71253872 & 71255695 & NONO    & Myeloid &  0.18692709 & 0.05182659 &   3.606779 &  3.100209e-04 & 0.00088 & 3.055517 &  1 &  3.055517 & 0.014729363\\\\\n",
       "\t chrX & 71253872 & 71255695 & OGT     & Myeloid &  0.10905648 & 0.02955893 &   3.689459 &  2.247310e-04 & 0.00040 & 3.397940 &  1 &  3.397940 & 0.008014822\\\\\n",
       "\t chrX & 71282778 & 71284216 & NONO    & Myeloid &  0.14348201 & 0.04543674 &   3.157841 &  1.589421e-03 & 0.00296 & 2.528708 &  1 &  2.528708 & 0.035693149\\\\\n",
       "\t chrX & 71365597 & 71367182 & OGT     & Myeloid &  0.07541851 & 0.02416346 &   3.121180 &  1.801279e-03 & 0.00368 & 2.434152 &  1 &  2.434152 & 0.041748220\\\\\n",
       "\t chrX & 71540871 & 71542054 & RPS4X   & Myeloid & -0.18319556 & 0.03024068 &  -6.057919 &  1.378942e-09 & 0.00480 & 2.318759 & -1 & -2.318759 & 0.050317530\\\\\n",
       "\t chrX & 72107322 & 72108644 & RPS4X   & Myeloid & -0.30474042 & 0.03040675 & -10.022132 &  1.218459e-23 & 0.00008 & 4.096910 & -1 & -4.096910 & 0.002128937\\\\\n",
       "\t chrX & 72181025 & 72182115 & NHSL2   & Myeloid & -0.17016975 & 0.05128859 &  -3.317887 &  9.070110e-04 & 0.00368 & 2.434152 & -1 & -2.434152 & 0.041748220\\\\\n",
       "\t chrX & 73850802 & 73851931 & JPX     & Myeloid &  0.17253377 & 0.03542301 &   4.870669 &  1.112208e-06 & 0.00016 & 3.795880 &  1 &  3.795880 & 0.003831735\\\\\n",
       "\t chrX & 73850802 & 73851931 & XIST    & Myeloid &  0.54370458 & 0.01779508 &  30.553648 & 5.057667e-205 & 0.00008 & 4.096910 &  1 &  4.096910 & 0.002128937\\\\\n",
       "\t chrX & 73943694 & 73944847 & XIST    & Myeloid &  0.18747851 & 0.01836553 &  10.208175 &  1.822602e-24 & 0.00008 & 4.096910 &  1 &  4.096910 & 0.002128937\\\\\n",
       "\t chrX & 74291819 & 74293013 & XIST    & Myeloid &  0.11040114 & 0.01830518 &   6.031141 &  1.628058e-09 & 0.00024 & 3.619789 &  1 &  3.619789 & 0.005347514\\\\\n",
       "\t chrX & 74613836 & 74615059 & XIST    & Myeloid &  0.07637641 & 0.01694361 &   4.507683 &  6.553951e-06 & 0.00792 & 2.101275 &  1 &  2.101275 & 0.071688537\\\\\n",
       "\t chrX & 77899134 & 77899778 & CYSLTR1 & Myeloid & -0.47923887 & 0.09509173 &  -5.039753 &  4.661319e-07 & 0.00032 & 3.494850 & -1 & -3.494850 & 0.006722723\\\\\n",
       "\t chrX & 81120681 & 81122301 & SH3BGRL & Myeloid & -0.14995632 & 0.03297125 &  -4.548094 &  5.413401e-06 & 0.00048 & 3.318759 & -1 & -3.318759 & 0.009255598\\\\\n",
       "\t chrX &  9378667 &  9379857 & TBL1X   & Myeloid &  0.16309584 & 0.02643635 &   6.169379 &  6.855870e-10 & 0.00008 & 4.096910 &  1 &  4.096910 & 0.002128937\\\\\n",
       "\t chrX & 96722435 & 96723511 & DIAPH2  & Myeloid &  0.08338271 & 0.02010979 &   4.146373 &  3.377832e-05 & 0.00208 & 2.681937 &  1 &  2.681937 & 0.027737601\\\\\n",
       "\t chrX &   989778 &   990654 & IL3RA   & Myeloid &  0.31020460 & 0.07964383 &   3.894898 &  9.824001e-05 & 0.00016 & 3.795880 &  1 &  3.795880 & 0.003831735\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 318489 × 14\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | beta &lt;dbl&gt; | se &lt;dbl&gt; | z &lt;dbl&gt; | p &lt;dbl&gt; | boot_basic_p &lt;dbl&gt; | boot_basic_p.log10 &lt;dbl&gt; | sign &lt;dbl&gt; | boot_basic_p.log10.signed &lt;dbl&gt; | FDR &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |   1012920 |   1014120 | UBE2J2      | Myeloid |  0.15242525 | 0.06205776 |  2.456184 | 0.0140421368 | 0.01104 | 1.957031 |  1 |  1.957031 | 0.09000269 |\n",
       "| chr1 |  10147738 |  10148559 | UBE4B       | Myeloid |  0.13261564 | 0.04710376 |  2.815394 | 0.0048717487 | 0.01552 | 1.809108 |  1 |  1.809108 | 0.11035832 |\n",
       "| chr1 |   1019144 |   1020371 | NADK        | Myeloid | -0.10450517 | 0.03843754 | -2.718831 | 0.0065513095 | 0.01072 | 1.969805 | -1 | -1.969805 | 0.08870131 |\n",
       "| chr1 |   1038601 |   1039080 | ISG15       | Myeloid | -0.37455939 | 0.10286672 | -3.641210 | 0.0002713592 | 0.01368 | 1.863914 | -1 | -1.863914 | 0.10151280 |\n",
       "| chr1 |   1058756 |   1060110 | NADK        | Myeloid | -0.12838767 | 0.04672014 | -2.748016 | 0.0059957130 | 0.01160 | 1.935542 | -1 | -1.935542 | 0.09284460 |\n",
       "| chr1 |   1079374 |   1080601 | CDK11B      | Myeloid |  0.16449572 | 0.06110268 |  2.692120 | 0.0070999496 | 0.00712 | 2.147520 |  1 |  2.147520 | 0.06637324 |\n",
       "| chr1 | 109040911 | 109042568 | CLCC1       | Myeloid |  0.14251012 | 0.05488633 |  2.596459 | 0.0094190065 | 0.01040 | 1.982967 |  1 |  1.982967 | 0.08635188 |\n",
       "| chr1 | 109091688 | 109092177 | GPSM2       | Myeloid |  0.28891353 | 0.10194700 |  2.833958 | 0.0045975345 | 0.01520 | 1.818156 |  1 |  1.818156 | 0.10814084 |\n",
       "| chr1 | 109133440 | 109134866 | GPSM2       | Myeloid |  0.29310711 | 0.09617894 |  3.047519 | 0.0023073926 | 0.01088 | 1.963371 |  1 |  1.963371 | 0.08987810 |\n",
       "| chr1 | 109213476 | 109214665 | SARS        | Myeloid |  0.21115222 | 0.07072164 |  2.985681 | 0.0028294796 | 0.00272 | 2.565431 |  1 |  2.565431 | 0.03363841 |\n",
       "| chr1 | 109573504 | 109574558 | AHCYL1      | Myeloid | -0.27273708 | 0.09505956 | -2.869118 | 0.0041161823 | 0.00992 | 2.003488 | -1 | -2.003488 | 0.08506532 |\n",
       "| chr1 | 109643175 | 109643836 | AMPD2       | Myeloid |  0.16632329 | 0.06444675 |  2.580786 | 0.0098575594 | 0.01136 | 1.944622 |  1 |  1.944622 | 0.09112061 |\n",
       "| chr1 | 110034075 | 110035478 | LAMTOR5-AS1 | Myeloid |  0.15030050 | 0.05475852 |  2.744787 | 0.0060550092 | 0.01104 | 1.957031 |  1 |  1.957031 | 0.09000269 |\n",
       "| chr1 | 110068666 | 110070653 | AMPD2       | Myeloid |  0.15144738 | 0.05508265 |  2.749457 | 0.0059694128 | 0.01000 | 2.000000 |  1 |  2.000000 | 0.08567751 |\n",
       "| chr1 | 110424462 | 110425171 | STRIP1      | Myeloid |  0.26196256 | 0.11672272 |  2.244315 | 0.0248121302 | 0.01616 | 1.791559 |  1 |  1.791559 | 0.11279630 |\n",
       "| chr1 | 110619222 | 110620626 | RBM15       | Myeloid |  0.20249370 | 0.07707231 |  2.627321 | 0.0086060090 | 0.00832 | 2.079877 |  1 |  2.079877 | 0.07428107 |\n",
       "| chr1 | 110627679 | 110628608 | CEPT1       | Myeloid | -0.14393662 | 0.05169942 | -2.784105 | 0.0053675675 | 0.01152 | 1.938548 | -1 | -1.938548 | 0.09227386 |\n",
       "| chr1 | 110963175 | 110964482 | FAM212B     | Myeloid | -0.19962694 | 0.05904812 | -3.380750 | 0.0007228823 | 0.00192 | 2.716699 | -1 | -2.716699 | 0.02617718 |\n",
       "| chr1 | 111139345 | 111140838 | CEPT1       | Myeloid | -0.10139644 | 0.03657225 | -2.772497 | 0.0055628110 | 0.01392 | 1.856361 | -1 | -1.856361 | 0.10319515 |\n",
       "| chr1 | 111139345 | 111140838 | FAM212B     | Myeloid | -0.24226979 | 0.06737893 | -3.595631 | 0.0003236058 | 0.00128 | 2.892790 | -1 | -2.892790 | 0.01952049 |\n",
       "| chr1 | 111616317 | 111616816 | CEPT1       | Myeloid | -0.15529944 | 0.05521685 | -2.812537 | 0.0049152311 | 0.01208 | 1.917933 | -1 | -1.917933 | 0.09412931 |\n",
       "| chr1 | 111738407 | 111740718 | ATP5F1      | Myeloid | -0.16005882 | 0.05621385 | -2.847320 | 0.0044089029 | 0.00968 | 2.014125 | -1 | -2.014125 | 0.08344539 |\n",
       "| chr1 | 111738407 | 111740718 | FAM212B     | Myeloid | -0.20705799 | 0.06124127 | -3.381021 | 0.0007221707 | 0.00232 | 2.634512 | -1 | -2.634512 | 0.03002050 |\n",
       "| chr1 |   1121806 |   1123448 | GNB1        | Myeloid |  0.07788593 | 0.02661965 |  2.925881 | 0.0034348180 | 0.00736 | 2.133122 |  1 |  2.133122 | 0.06794038 |\n",
       "| chr1 | 112706534 | 112707637 | FAM212B     | Myeloid | -0.34345561 | 0.09626244 | -3.567909 | 0.0003598416 | 0.00232 | 2.634512 | -1 | -2.634512 | 0.03002050 |\n",
       "| chr1 | 113140382 | 113141992 | PTPN22      | Myeloid |  0.15969368 | 0.05463950 |  2.922678 | 0.0034703502 | 0.01192 | 1.923724 |  1 |  1.923724 | 0.09356087 |\n",
       "| chr1 | 113389901 | 113391532 | PHTF1       | Myeloid |  0.14690124 | 0.04562050 |  3.220071 | 0.0012815902 | 0.00984 | 2.007005 |  1 |  2.007005 | 0.08453407 |\n",
       "| chr1 | 113389901 | 113391532 | PTPN22      | Myeloid |  0.11077072 | 0.03523556 |  3.143719 | 0.0016681536 | 0.00808 | 2.092589 |  1 |  2.092589 | 0.07274398 |\n",
       "| chr1 | 113928993 | 113931084 | PHTF1       | Myeloid |  0.14114251 | 0.04262542 |  3.311229 | 0.0009288722 | 0.01000 | 2.000000 |  1 |  2.000000 | 0.08567751 |\n",
       "| chr1 | 114040080 | 114041198 | PTPN22      | Myeloid |  0.14492475 | 0.04467912 |  3.243680 | 0.0011799633 | 0.00920 | 2.036212 |  1 |  2.036212 | 0.08017564 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 53071574 | 53073601 | SMC1A   | Myeloid |  0.19046290 | 0.05344579 |   3.563665 |  3.657126e-04 | 0.00112 | 2.950782 |  1 |  2.950782 | 0.017663168 |\n",
       "| chrX | 53081843 | 53082907 | SMC1A   | Myeloid |  0.13253351 | 0.04073004 |   3.253950 |  1.138123e-03 | 0.00240 | 2.619789 |  1 |  2.619789 | 0.030759501 |\n",
       "| chrX | 53224384 | 53225810 | SMC1A   | Myeloid |  0.13246754 | 0.03962539 |   3.342997 |  8.287887e-04 | 0.00216 | 2.665546 |  1 |  2.665546 | 0.028515492 |\n",
       "| chrX | 53280164 | 53281893 | KDM5C   | Myeloid |  0.27843651 | 0.09001215 |   3.093321 |  1.979296e-03 | 0.00208 | 2.681937 |  1 |  2.681937 | 0.027737601 |\n",
       "| chrX | 53287345 | 53288709 | SMC1A   | Myeloid |  0.23446749 | 0.05263760 |   4.454373 |  8.413885e-06 | 0.00008 | 4.096910 |  1 |  4.096910 | 0.002128937 |\n",
       "| chrX | 53374180 | 53375241 | SMC1A   | Myeloid |  0.21616396 | 0.05898753 |   3.664571 |  2.477537e-04 | 0.00072 | 3.142668 |  1 |  3.142668 | 0.012651003 |\n",
       "| chrX | 53421910 | 53423211 | KDM5C   | Myeloid |  0.19170571 | 0.04925167 |   3.892370 |  9.926981e-05 | 0.00008 | 4.096910 |  1 |  4.096910 | 0.002128937 |\n",
       "| chrX | 53421910 | 53423211 | SMC1A   | Myeloid |  0.11762703 | 0.03621192 |   3.248295 |  1.160986e-03 | 0.00376 | 2.424812 |  1 |  2.424812 | 0.042366045 |\n",
       "| chrX | 54182281 | 54183769 | SMC1A   | Myeloid |  0.15326493 | 0.04967353 |   3.085444 |  2.032482e-03 | 0.00488 | 2.311580 |  1 |  2.311580 | 0.050918173 |\n",
       "| chrX | 65997710 | 65998828 | MSN     | Myeloid | -0.11387141 | 0.03133574 |  -3.633915 |  2.791529e-04 | 0.00360 | 2.443697 | -1 | -2.443697 | 0.041124835 |\n",
       "| chrX | 71067720 | 71068957 | NHSL2   | Myeloid | -0.19718401 | 0.04503518 |  -4.378444 |  1.195299e-05 | 0.00008 | 4.096910 | -1 | -4.096910 | 0.002128937 |\n",
       "| chrX | 71111138 | 71112478 | NHSL2   | Myeloid | -0.22457700 | 0.06742250 |  -3.330891 |  8.656843e-04 | 0.00560 | 2.251812 | -1 | -2.251812 | 0.055987519 |\n",
       "| chrX | 71253872 | 71255695 | NHSL2   | Myeloid | -0.30942868 | 0.05915155 |  -5.231117 |  1.684888e-07 | 0.00008 | 4.096910 | -1 | -4.096910 | 0.002128937 |\n",
       "| chrX | 71253872 | 71255695 | NONO    | Myeloid |  0.18692709 | 0.05182659 |   3.606779 |  3.100209e-04 | 0.00088 | 3.055517 |  1 |  3.055517 | 0.014729363 |\n",
       "| chrX | 71253872 | 71255695 | OGT     | Myeloid |  0.10905648 | 0.02955893 |   3.689459 |  2.247310e-04 | 0.00040 | 3.397940 |  1 |  3.397940 | 0.008014822 |\n",
       "| chrX | 71282778 | 71284216 | NONO    | Myeloid |  0.14348201 | 0.04543674 |   3.157841 |  1.589421e-03 | 0.00296 | 2.528708 |  1 |  2.528708 | 0.035693149 |\n",
       "| chrX | 71365597 | 71367182 | OGT     | Myeloid |  0.07541851 | 0.02416346 |   3.121180 |  1.801279e-03 | 0.00368 | 2.434152 |  1 |  2.434152 | 0.041748220 |\n",
       "| chrX | 71540871 | 71542054 | RPS4X   | Myeloid | -0.18319556 | 0.03024068 |  -6.057919 |  1.378942e-09 | 0.00480 | 2.318759 | -1 | -2.318759 | 0.050317530 |\n",
       "| chrX | 72107322 | 72108644 | RPS4X   | Myeloid | -0.30474042 | 0.03040675 | -10.022132 |  1.218459e-23 | 0.00008 | 4.096910 | -1 | -4.096910 | 0.002128937 |\n",
       "| chrX | 72181025 | 72182115 | NHSL2   | Myeloid | -0.17016975 | 0.05128859 |  -3.317887 |  9.070110e-04 | 0.00368 | 2.434152 | -1 | -2.434152 | 0.041748220 |\n",
       "| chrX | 73850802 | 73851931 | JPX     | Myeloid |  0.17253377 | 0.03542301 |   4.870669 |  1.112208e-06 | 0.00016 | 3.795880 |  1 |  3.795880 | 0.003831735 |\n",
       "| chrX | 73850802 | 73851931 | XIST    | Myeloid |  0.54370458 | 0.01779508 |  30.553648 | 5.057667e-205 | 0.00008 | 4.096910 |  1 |  4.096910 | 0.002128937 |\n",
       "| chrX | 73943694 | 73944847 | XIST    | Myeloid |  0.18747851 | 0.01836553 |  10.208175 |  1.822602e-24 | 0.00008 | 4.096910 |  1 |  4.096910 | 0.002128937 |\n",
       "| chrX | 74291819 | 74293013 | XIST    | Myeloid |  0.11040114 | 0.01830518 |   6.031141 |  1.628058e-09 | 0.00024 | 3.619789 |  1 |  3.619789 | 0.005347514 |\n",
       "| chrX | 74613836 | 74615059 | XIST    | Myeloid |  0.07637641 | 0.01694361 |   4.507683 |  6.553951e-06 | 0.00792 | 2.101275 |  1 |  2.101275 | 0.071688537 |\n",
       "| chrX | 77899134 | 77899778 | CYSLTR1 | Myeloid | -0.47923887 | 0.09509173 |  -5.039753 |  4.661319e-07 | 0.00032 | 3.494850 | -1 | -3.494850 | 0.006722723 |\n",
       "| chrX | 81120681 | 81122301 | SH3BGRL | Myeloid | -0.14995632 | 0.03297125 |  -4.548094 |  5.413401e-06 | 0.00048 | 3.318759 | -1 | -3.318759 | 0.009255598 |\n",
       "| chrX |  9378667 |  9379857 | TBL1X   | Myeloid |  0.16309584 | 0.02643635 |   6.169379 |  6.855870e-10 | 0.00008 | 4.096910 |  1 |  4.096910 | 0.002128937 |\n",
       "| chrX | 96722435 | 96723511 | DIAPH2  | Myeloid |  0.08338271 | 0.02010979 |   4.146373 |  3.377832e-05 | 0.00208 | 2.681937 |  1 |  2.681937 | 0.027737601 |\n",
       "| chrX |   989778 |   990654 | IL3RA   | Myeloid |  0.31020460 | 0.07964383 |   3.894898 |  9.824001e-05 | 0.00016 | 3.795880 |  1 |  3.795880 | 0.003831735 |\n",
       "\n"
      ],
      "text/plain": [
       "       chr  start     end       TargetGene  CellType beta        se        \n",
       "1      chr1   1012920   1014120 UBE2J2      Myeloid   0.15242525 0.06205776\n",
       "2      chr1  10147738  10148559 UBE4B       Myeloid   0.13261564 0.04710376\n",
       "3      chr1   1019144   1020371 NADK        Myeloid  -0.10450517 0.03843754\n",
       "4      chr1   1038601   1039080 ISG15       Myeloid  -0.37455939 0.10286672\n",
       "5      chr1   1058756   1060110 NADK        Myeloid  -0.12838767 0.04672014\n",
       "6      chr1   1079374   1080601 CDK11B      Myeloid   0.16449572 0.06110268\n",
       "7      chr1 109040911 109042568 CLCC1       Myeloid   0.14251012 0.05488633\n",
       "8      chr1 109091688 109092177 GPSM2       Myeloid   0.28891353 0.10194700\n",
       "9      chr1 109133440 109134866 GPSM2       Myeloid   0.29310711 0.09617894\n",
       "10     chr1 109213476 109214665 SARS        Myeloid   0.21115222 0.07072164\n",
       "11     chr1 109573504 109574558 AHCYL1      Myeloid  -0.27273708 0.09505956\n",
       "12     chr1 109643175 109643836 AMPD2       Myeloid   0.16632329 0.06444675\n",
       "13     chr1 110034075 110035478 LAMTOR5-AS1 Myeloid   0.15030050 0.05475852\n",
       "14     chr1 110068666 110070653 AMPD2       Myeloid   0.15144738 0.05508265\n",
       "15     chr1 110424462 110425171 STRIP1      Myeloid   0.26196256 0.11672272\n",
       "16     chr1 110619222 110620626 RBM15       Myeloid   0.20249370 0.07707231\n",
       "17     chr1 110627679 110628608 CEPT1       Myeloid  -0.14393662 0.05169942\n",
       "18     chr1 110963175 110964482 FAM212B     Myeloid  -0.19962694 0.05904812\n",
       "19     chr1 111139345 111140838 CEPT1       Myeloid  -0.10139644 0.03657225\n",
       "20     chr1 111139345 111140838 FAM212B     Myeloid  -0.24226979 0.06737893\n",
       "21     chr1 111616317 111616816 CEPT1       Myeloid  -0.15529944 0.05521685\n",
       "22     chr1 111738407 111740718 ATP5F1      Myeloid  -0.16005882 0.05621385\n",
       "23     chr1 111738407 111740718 FAM212B     Myeloid  -0.20705799 0.06124127\n",
       "24     chr1   1121806   1123448 GNB1        Myeloid   0.07788593 0.02661965\n",
       "25     chr1 112706534 112707637 FAM212B     Myeloid  -0.34345561 0.09626244\n",
       "26     chr1 113140382 113141992 PTPN22      Myeloid   0.15969368 0.05463950\n",
       "27     chr1 113389901 113391532 PHTF1       Myeloid   0.14690124 0.04562050\n",
       "28     chr1 113389901 113391532 PTPN22      Myeloid   0.11077072 0.03523556\n",
       "29     chr1 113928993 113931084 PHTF1       Myeloid   0.14114251 0.04262542\n",
       "30     chr1 114040080 114041198 PTPN22      Myeloid   0.14492475 0.04467912\n",
       "⋮      ⋮    ⋮         ⋮         ⋮           ⋮        ⋮           ⋮         \n",
       "318460 chrX 53071574  53073601  SMC1A       Myeloid   0.19046290 0.05344579\n",
       "318461 chrX 53081843  53082907  SMC1A       Myeloid   0.13253351 0.04073004\n",
       "318462 chrX 53224384  53225810  SMC1A       Myeloid   0.13246754 0.03962539\n",
       "318463 chrX 53280164  53281893  KDM5C       Myeloid   0.27843651 0.09001215\n",
       "318464 chrX 53287345  53288709  SMC1A       Myeloid   0.23446749 0.05263760\n",
       "318465 chrX 53374180  53375241  SMC1A       Myeloid   0.21616396 0.05898753\n",
       "318466 chrX 53421910  53423211  KDM5C       Myeloid   0.19170571 0.04925167\n",
       "318467 chrX 53421910  53423211  SMC1A       Myeloid   0.11762703 0.03621192\n",
       "318468 chrX 54182281  54183769  SMC1A       Myeloid   0.15326493 0.04967353\n",
       "318469 chrX 65997710  65998828  MSN         Myeloid  -0.11387141 0.03133574\n",
       "318470 chrX 71067720  71068957  NHSL2       Myeloid  -0.19718401 0.04503518\n",
       "318471 chrX 71111138  71112478  NHSL2       Myeloid  -0.22457700 0.06742250\n",
       "318472 chrX 71253872  71255695  NHSL2       Myeloid  -0.30942868 0.05915155\n",
       "318473 chrX 71253872  71255695  NONO        Myeloid   0.18692709 0.05182659\n",
       "318474 chrX 71253872  71255695  OGT         Myeloid   0.10905648 0.02955893\n",
       "318475 chrX 71282778  71284216  NONO        Myeloid   0.14348201 0.04543674\n",
       "318476 chrX 71365597  71367182  OGT         Myeloid   0.07541851 0.02416346\n",
       "318477 chrX 71540871  71542054  RPS4X       Myeloid  -0.18319556 0.03024068\n",
       "318478 chrX 72107322  72108644  RPS4X       Myeloid  -0.30474042 0.03040675\n",
       "318479 chrX 72181025  72182115  NHSL2       Myeloid  -0.17016975 0.05128859\n",
       "318480 chrX 73850802  73851931  JPX         Myeloid   0.17253377 0.03542301\n",
       "318481 chrX 73850802  73851931  XIST        Myeloid   0.54370458 0.01779508\n",
       "318482 chrX 73943694  73944847  XIST        Myeloid   0.18747851 0.01836553\n",
       "318483 chrX 74291819  74293013  XIST        Myeloid   0.11040114 0.01830518\n",
       "318484 chrX 74613836  74615059  XIST        Myeloid   0.07637641 0.01694361\n",
       "318485 chrX 77899134  77899778  CYSLTR1     Myeloid  -0.47923887 0.09509173\n",
       "318486 chrX 81120681  81122301  SH3BGRL     Myeloid  -0.14995632 0.03297125\n",
       "318487 chrX  9378667   9379857  TBL1X       Myeloid   0.16309584 0.02643635\n",
       "318488 chrX 96722435  96723511  DIAPH2      Myeloid   0.08338271 0.02010979\n",
       "318489 chrX   989778    990654  IL3RA       Myeloid   0.31020460 0.07964383\n",
       "       z          p             boot_basic_p boot_basic_p.log10 sign\n",
       "1       2.456184  0.0140421368  0.01104      1.957031            1  \n",
       "2       2.815394  0.0048717487  0.01552      1.809108            1  \n",
       "3      -2.718831  0.0065513095  0.01072      1.969805           -1  \n",
       "4      -3.641210  0.0002713592  0.01368      1.863914           -1  \n",
       "5      -2.748016  0.0059957130  0.01160      1.935542           -1  \n",
       "6       2.692120  0.0070999496  0.00712      2.147520            1  \n",
       "7       2.596459  0.0094190065  0.01040      1.982967            1  \n",
       "8       2.833958  0.0045975345  0.01520      1.818156            1  \n",
       "9       3.047519  0.0023073926  0.01088      1.963371            1  \n",
       "10      2.985681  0.0028294796  0.00272      2.565431            1  \n",
       "11     -2.869118  0.0041161823  0.00992      2.003488           -1  \n",
       "12      2.580786  0.0098575594  0.01136      1.944622            1  \n",
       "13      2.744787  0.0060550092  0.01104      1.957031            1  \n",
       "14      2.749457  0.0059694128  0.01000      2.000000            1  \n",
       "15      2.244315  0.0248121302  0.01616      1.791559            1  \n",
       "16      2.627321  0.0086060090  0.00832      2.079877            1  \n",
       "17     -2.784105  0.0053675675  0.01152      1.938548           -1  \n",
       "18     -3.380750  0.0007228823  0.00192      2.716699           -1  \n",
       "19     -2.772497  0.0055628110  0.01392      1.856361           -1  \n",
       "20     -3.595631  0.0003236058  0.00128      2.892790           -1  \n",
       "21     -2.812537  0.0049152311  0.01208      1.917933           -1  \n",
       "22     -2.847320  0.0044089029  0.00968      2.014125           -1  \n",
       "23     -3.381021  0.0007221707  0.00232      2.634512           -1  \n",
       "24      2.925881  0.0034348180  0.00736      2.133122            1  \n",
       "25     -3.567909  0.0003598416  0.00232      2.634512           -1  \n",
       "26      2.922678  0.0034703502  0.01192      1.923724            1  \n",
       "27      3.220071  0.0012815902  0.00984      2.007005            1  \n",
       "28      3.143719  0.0016681536  0.00808      2.092589            1  \n",
       "29      3.311229  0.0009288722  0.01000      2.000000            1  \n",
       "30      3.243680  0.0011799633  0.00920      2.036212            1  \n",
       "⋮      ⋮          ⋮             ⋮            ⋮                  ⋮   \n",
       "318460   3.563665  3.657126e-04 0.00112      2.950782            1  \n",
       "318461   3.253950  1.138123e-03 0.00240      2.619789            1  \n",
       "318462   3.342997  8.287887e-04 0.00216      2.665546            1  \n",
       "318463   3.093321  1.979296e-03 0.00208      2.681937            1  \n",
       "318464   4.454373  8.413885e-06 0.00008      4.096910            1  \n",
       "318465   3.664571  2.477537e-04 0.00072      3.142668            1  \n",
       "318466   3.892370  9.926981e-05 0.00008      4.096910            1  \n",
       "318467   3.248295  1.160986e-03 0.00376      2.424812            1  \n",
       "318468   3.085444  2.032482e-03 0.00488      2.311580            1  \n",
       "318469  -3.633915  2.791529e-04 0.00360      2.443697           -1  \n",
       "318470  -4.378444  1.195299e-05 0.00008      4.096910           -1  \n",
       "318471  -3.330891  8.656843e-04 0.00560      2.251812           -1  \n",
       "318472  -5.231117  1.684888e-07 0.00008      4.096910           -1  \n",
       "318473   3.606779  3.100209e-04 0.00088      3.055517            1  \n",
       "318474   3.689459  2.247310e-04 0.00040      3.397940            1  \n",
       "318475   3.157841  1.589421e-03 0.00296      2.528708            1  \n",
       "318476   3.121180  1.801279e-03 0.00368      2.434152            1  \n",
       "318477  -6.057919  1.378942e-09 0.00480      2.318759           -1  \n",
       "318478 -10.022132  1.218459e-23 0.00008      4.096910           -1  \n",
       "318479  -3.317887  9.070110e-04 0.00368      2.434152           -1  \n",
       "318480   4.870669  1.112208e-06 0.00016      3.795880            1  \n",
       "318481  30.553648 5.057667e-205 0.00008      4.096910            1  \n",
       "318482  10.208175  1.822602e-24 0.00008      4.096910            1  \n",
       "318483   6.031141  1.628058e-09 0.00024      3.619789            1  \n",
       "318484   4.507683  6.553951e-06 0.00792      2.101275            1  \n",
       "318485  -5.039753  4.661319e-07 0.00032      3.494850           -1  \n",
       "318486  -4.548094  5.413401e-06 0.00048      3.318759           -1  \n",
       "318487   6.169379  6.855870e-10 0.00008      4.096910            1  \n",
       "318488   4.146373  3.377832e-05 0.00208      2.681937            1  \n",
       "318489   3.894898  9.824001e-05 0.00016      3.795880            1  \n",
       "       boot_basic_p.log10.signed FDR        \n",
       "1       1.957031                 0.09000269 \n",
       "2       1.809108                 0.11035832 \n",
       "3      -1.969805                 0.08870131 \n",
       "4      -1.863914                 0.10151280 \n",
       "5      -1.935542                 0.09284460 \n",
       "6       2.147520                 0.06637324 \n",
       "7       1.982967                 0.08635188 \n",
       "8       1.818156                 0.10814084 \n",
       "9       1.963371                 0.08987810 \n",
       "10      2.565431                 0.03363841 \n",
       "11     -2.003488                 0.08506532 \n",
       "12      1.944622                 0.09112061 \n",
       "13      1.957031                 0.09000269 \n",
       "14      2.000000                 0.08567751 \n",
       "15      1.791559                 0.11279630 \n",
       "16      2.079877                 0.07428107 \n",
       "17     -1.938548                 0.09227386 \n",
       "18     -2.716699                 0.02617718 \n",
       "19     -1.856361                 0.10319515 \n",
       "20     -2.892790                 0.01952049 \n",
       "21     -1.917933                 0.09412931 \n",
       "22     -2.014125                 0.08344539 \n",
       "23     -2.634512                 0.03002050 \n",
       "24      2.133122                 0.06794038 \n",
       "25     -2.634512                 0.03002050 \n",
       "26      1.923724                 0.09356087 \n",
       "27      2.007005                 0.08453407 \n",
       "28      2.092589                 0.07274398 \n",
       "29      2.000000                 0.08567751 \n",
       "30      2.036212                 0.08017564 \n",
       "⋮      ⋮                         ⋮          \n",
       "318460  2.950782                 0.017663168\n",
       "318461  2.619789                 0.030759501\n",
       "318462  2.665546                 0.028515492\n",
       "318463  2.681937                 0.027737601\n",
       "318464  4.096910                 0.002128937\n",
       "318465  3.142668                 0.012651003\n",
       "318466  4.096910                 0.002128937\n",
       "318467  2.424812                 0.042366045\n",
       "318468  2.311580                 0.050918173\n",
       "318469 -2.443697                 0.041124835\n",
       "318470 -4.096910                 0.002128937\n",
       "318471 -2.251812                 0.055987519\n",
       "318472 -4.096910                 0.002128937\n",
       "318473  3.055517                 0.014729363\n",
       "318474  3.397940                 0.008014822\n",
       "318475  2.528708                 0.035693149\n",
       "318476  2.434152                 0.041748220\n",
       "318477 -2.318759                 0.050317530\n",
       "318478 -4.096910                 0.002128937\n",
       "318479 -2.434152                 0.041748220\n",
       "318480  3.795880                 0.003831735\n",
       "318481  4.096910                 0.002128937\n",
       "318482  4.096910                 0.002128937\n",
       "318483  3.619789                 0.005347514\n",
       "318484  2.101275                 0.071688537\n",
       "318485 -3.494850                 0.006722723\n",
       "318486 -3.318759                 0.009255598\n",
       "318487  4.096910                 0.002128937\n",
       "318488  2.681937                 0.027737601\n",
       "318489  3.795880                 0.003831735"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = celltype\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"beta\",\"se\",\"z\",\"p\",\"boot_basic_p\",\n",
    "                         \"boot_basic_p.log10\",\n",
    "                         \"sign\",\n",
    "                         \"boot_basic_p.log10.signed\")]\n",
    "df.output[,\"FDR\"] = p.adjust(df.output$boot_basic_p, method = \"BH\")\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] dplyr_1.1.4          data.table_1.15.2    Matrix_1.6-5        \n",
       " [4] GenomicRanges_1.54.1 GenomeInfoDb_1.38.1  IRanges_2.36.0      \n",
       " [7] S4Vectors_0.40.2     BiocGenerics_0.48.1  genomation_1.34.0   \n",
       "[10] Signac_1.13.0        SCENT_1.0.0         \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] rstudioapi_0.16.0           jsonlite_1.8.8             \n",
       "  [3] magrittr_2.0.3              nloptr_2.0.3               \n",
       "  [5] rmarkdown_2.26              BiocIO_1.10.0              \n",
       "  [7] zlibbioc_1.48.0             vctrs_0.6.5                \n",
       "  [9] minqa_1.2.6                 Rsamtools_2.18.0           \n",
       " [11] RCurl_1.98-1.14             base64enc_0.1-3            \n",
       " [13] RcppRoll_0.3.0              htmltools_0.5.8.1          \n",
       " [15] S4Arrays_1.2.1              plotrix_3.8-4              \n",
       " [17] Formula_1.2-5               parallelly_1.37.1          \n",
       " [19] KernSmooth_2.23-22          htmlwidgets_1.6.4          \n",
       " [21] plyr_1.8.9                  impute_1.76.0              \n",
       " [23] uuid_1.2-0                  GenomicAlignments_1.36.0   \n",
       " [25] lifecycle_1.0.4             pkgconfig_2.0.3            \n",
       " [27] R6_2.5.1                    fastmap_1.1.1              \n",
       " [29] GenomeInfoDbData_1.2.11     MatrixGenerics_1.12.3      \n",
       " [31] future_1.33.2               digest_0.6.35              \n",
       " [33] colorspace_2.1-0            patchwork_1.2.0            \n",
       " [35] irlba_2.3.5.1               Hmisc_5.1-2                \n",
       " [37] progressr_0.14.0            fansi_1.0.6                \n",
       " [39] abind_1.4-5                 compiler_4.3.3             \n",
       " [41] withr_3.0.0                 htmlTable_2.4.2            \n",
       " [43] backports_1.4.1             BiocParallel_1.36.0        \n",
       " [45] R.utils_2.12.3              MASS_7.3-60                \n",
       " [47] DelayedArray_0.26.7         rjson_0.2.21               \n",
       " [49] tools_4.3.3                 foreign_0.8-86             \n",
       " [51] future.apply_1.11.2         nnet_7.3-19                \n",
       " [53] R.oo_1.26.0                 glue_1.7.0                 \n",
       " [55] restfulr_0.0.15             nlme_3.1-164               \n",
       " [57] pbdZMQ_0.3-11               checkmate_2.3.1            \n",
       " [59] gridBase_0.4-7              cluster_2.1.6              \n",
       " [61] reshape2_1.4.4              generics_0.1.3             \n",
       " [63] gtable_0.3.5                BSgenome_1.68.0            \n",
       " [65] tzdb_0.4.0                  R.methodsS3_1.8.2          \n",
       " [67] tidyr_1.3.1                 seqPattern_1.34.0          \n",
       " [69] hms_1.1.3                   sp_2.1-4                   \n",
       " [71] utf8_1.2.4                  XVector_0.42.0             \n",
       " [73] pillar_1.9.0                stringr_1.5.1              \n",
       " [75] spam_2.10-0                 IRdisplay_1.1              \n",
       " [77] splines_4.3.3               lattice_0.22-6             \n",
       " [79] rtracklayer_1.60.1          tidyselect_1.2.1           \n",
       " [81] Biostrings_2.70.1           pbapply_1.7-2              \n",
       " [83] knitr_1.46                  gridExtra_2.3              \n",
       " [85] SummarizedExperiment_1.30.2 RhpcBLASctl_0.23-42        \n",
       " [87] xfun_0.43                   Biobase_2.60.0             \n",
       " [89] matrixStats_1.3.0           stringi_1.8.3              \n",
       " [91] yaml_2.3.8                  boot_1.3-30                \n",
       " [93] evaluate_0.23               codetools_0.2-20           \n",
       " [95] tibble_3.2.1                cli_3.6.2                  \n",
       " [97] IRkernel_1.3.2              rpart_4.1.23               \n",
       " [99] repr_1.1.7                  munsell_0.5.1              \n",
       "[101] Rcpp_1.0.12                 globals_0.16.3             \n",
       "[103] XML_3.99-0.16.1             ggplot2_3.5.1              \n",
       "[105] readr_2.1.5                 dotCall64_1.1-1            \n",
       "[107] bitops_1.0-7                lme4_1.1-35.3              \n",
       "[109] listenv_0.9.1               scales_1.3.0               \n",
       "[111] SeuratObject_5.0.1          purrr_1.0.2                \n",
       "[113] crayon_1.5.2                rlang_1.1.3                \n",
       "[115] fastmatch_1.1-4            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
