{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“replacing previous import ‘Hmisc::capitalize’ by ‘R.utils::capitalize’ when loading ‘SCENT’”\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SCENT)\n",
    "library(Signac)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(parallel)\n",
    "library(data.table)\n",
    "library(dplyr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/SCENT/BMMC5.Myeloid.500.250204/pairs.E2G.res.rds\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_BMMC5/Myeloid/Kendall/atac_matrix.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/3.merge_cell_type.240522/RNA_count_matrix/Myeloid.rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/SCENT/BMMC5.Myeloid.2500.250324/\"\n",
    "celltype = \"Myeloid\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 48"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readRDS(path.pairs.E2G)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.atac = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = BinarizeCounts(matrix.atac)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna = read.csv(path.matrix.rna_count,\n",
    "                      row.names = 1,\n",
    "                      check.names = F)\n",
    "matrix.rna = Matrix(as.matrix(matrix.rna), sparse = TRUE)\n",
    "matrix.rna = matrix.rna[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna[rowSums(matrix.rna) > 0,]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62718ef8-a3bd-4755-a95f-be216f16517b",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0f255a65-03a4-4255-967e-0c644bfe66cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ca760678-9db9-418e-b7ac-29984eddad7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare SCENT meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "827059dd-6885-441e-973e-d7c49e5fd534",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 14048 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>cell</th><th scope=col>nUMI</th><th scope=col>nMito</th><th scope=col>celltype</th><th scope=col>log.nUMI</th><th scope=col>percent.mito</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>s1d1_CTATGGCCATAACGGG-1</th><td>s1d1_CTATGGCCATAACGGG-1</td><td> 2157</td><td> 69</td><td>Myeloid</td><td> 7.676474</td><td>0.03198887</td></tr>\n",
       "\t<tr><th scope=row>s1d1_ACCACATAGGTGTCCA-1</th><td>s1d1_ACCACATAGGTGTCCA-1</td><td> 2128</td><td> 64</td><td>Myeloid</td><td> 7.662938</td><td>0.03007519</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TGCCGGTAGTCGCAAT-1</th><td>s1d1_TGCCGGTAGTCGCAAT-1</td><td> 1920</td><td> 67</td><td>Myeloid</td><td> 7.560080</td><td>0.03489583</td></tr>\n",
       "\t<tr><th scope=row>s1d1_CATCCCAGTCTTTATG-1</th><td>s1d1_CATCCCAGTCTTTATG-1</td><td> 3211</td><td>127</td><td>Myeloid</td><td> 8.074338</td><td>0.03955154</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TAGTGAGAGCTGCACA-1</th><td>s1d1_TAGTGAGAGCTGCACA-1</td><td> 3568</td><td> 85</td><td>Myeloid</td><td> 8.179760</td><td>0.02382287</td></tr>\n",
       "\t<tr><th scope=row>s1d1_CCTGTTGGTTATAGCG-1</th><td>s1d1_CCTGTTGGTTATAGCG-1</td><td> 4044</td><td> 42</td><td>Myeloid</td><td> 8.304990</td><td>0.01038576</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TTTAGCTTCTGGCTTC-1</th><td>s1d1_TTTAGCTTCTGGCTTC-1</td><td> 2728</td><td>105</td><td>Myeloid</td><td> 7.911324</td><td>0.03848974</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TGTGAAACAGGATTAA-1</th><td>s1d1_TGTGAAACAGGATTAA-1</td><td> 2800</td><td> 86</td><td>Myeloid</td><td> 7.937375</td><td>0.03071429</td></tr>\n",
       "\t<tr><th scope=row>s1d1_ACCAGGCTCTTGGATA-1</th><td>s1d1_ACCAGGCTCTTGGATA-1</td><td> 1015</td><td> 94</td><td>Myeloid</td><td> 6.922644</td><td>0.09261084</td></tr>\n",
       "\t<tr><th scope=row>s1d1_CACAGGGAGACAGGTA-1</th><td>s1d1_CACAGGGAGACAGGTA-1</td><td> 3092</td><td> 69</td><td>Myeloid</td><td> 8.036573</td><td>0.02231565</td></tr>\n",
       "\t<tr><th scope=row>s1d1_GCATCCTTCGCTATGG-1</th><td>s1d1_GCATCCTTCGCTATGG-1</td><td> 3112</td><td> 77</td><td>Myeloid</td><td> 8.043021</td><td>0.02474293</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TGCTATGCAGAGGGAG-1</th><td>s1d1_TGCTATGCAGAGGGAG-1</td><td> 2814</td><td> 78</td><td>Myeloid</td><td> 7.942362</td><td>0.02771855</td></tr>\n",
       "\t<tr><th scope=row>s1d1_CCCGCAACAAAGCGCA-1</th><td>s1d1_CCCGCAACAAAGCGCA-1</td><td> 3234</td><td> 58</td><td>Myeloid</td><td> 8.081475</td><td>0.01793445</td></tr>\n",
       "\t<tr><th scope=row>s1d1_AATCTTGAGAAGCTAG-1</th><td>s1d1_AATCTTGAGAAGCTAG-1</td><td> 1647</td><td> 67</td><td>Myeloid</td><td> 7.406711</td><td>0.04068002</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TTGTCCATCATTAGGC-1</th><td>s1d1_TTGTCCATCATTAGGC-1</td><td> 4110</td><td> 79</td><td>Myeloid</td><td> 8.321178</td><td>0.01922141</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TCCAGGATCCTAACGG-1</th><td>s1d1_TCCAGGATCCTAACGG-1</td><td> 3484</td><td>100</td><td>Myeloid</td><td> 8.155936</td><td>0.02870264</td></tr>\n",
       "\t<tr><th scope=row>s1d1_GGTATGTTCTAACCAA-1</th><td>s1d1_GGTATGTTCTAACCAA-1</td><td> 2759</td><td> 61</td><td>Myeloid</td><td> 7.922624</td><td>0.02210946</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TGGATTCAGTTTCCGC-1</th><td>s1d1_TGGATTCAGTTTCCGC-1</td><td> 3639</td><td> 72</td><td>Myeloid</td><td> 8.199464</td><td>0.01978566</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TGCTTGCTCAATCTCT-1</th><td>s1d1_TGCTTGCTCAATCTCT-1</td><td> 1934</td><td> 71</td><td>Myeloid</td><td> 7.567346</td><td>0.03671148</td></tr>\n",
       "\t<tr><th scope=row>s1d1_CGATGTCCATACCCGG-1</th><td>s1d1_CGATGTCCATACCCGG-1</td><td> 3296</td><td> 54</td><td>Myeloid</td><td> 8.100465</td><td>0.01638350</td></tr>\n",
       "\t<tr><th scope=row>s1d1_ACGCCTAAGCGCCTTT-1</th><td>s1d1_ACGCCTAAGCGCCTTT-1</td><td> 3236</td><td> 92</td><td>Myeloid</td><td> 8.082093</td><td>0.02843016</td></tr>\n",
       "\t<tr><th scope=row>s1d1_GAGTAATAGCTTAACA-1</th><td>s1d1_GAGTAATAGCTTAACA-1</td><td> 4308</td><td>100</td><td>Myeloid</td><td> 8.368229</td><td>0.02321263</td></tr>\n",
       "\t<tr><th scope=row>s1d1_ATTGGCTAGGGACCTC-1</th><td>s1d1_ATTGGCTAGGGACCTC-1</td><td>22348</td><td>863</td><td>Myeloid</td><td>10.014492</td><td>0.03861643</td></tr>\n",
       "\t<tr><th scope=row>s1d1_CCTGTAACATTGTGCA-1</th><td>s1d1_CCTGTAACATTGTGCA-1</td><td> 3940</td><td>127</td><td>Myeloid</td><td> 8.278936</td><td>0.03223350</td></tr>\n",
       "\t<tr><th scope=row>s1d1_AGATGAAGTGAGGTGA-1</th><td>s1d1_AGATGAAGTGAGGTGA-1</td><td> 2271</td><td> 69</td><td>Myeloid</td><td> 7.727976</td><td>0.03038309</td></tr>\n",
       "\t<tr><th scope=row>s1d1_GTTAACGGTAGCTGGT-1</th><td>s1d1_GTTAACGGTAGCTGGT-1</td><td> 1222</td><td> 72</td><td>Myeloid</td><td> 7.108244</td><td>0.05891980</td></tr>\n",
       "\t<tr><th scope=row>s1d1_GAACACAAGTAACCAC-1</th><td>s1d1_GAACACAAGTAACCAC-1</td><td> 3170</td><td> 64</td><td>Myeloid</td><td> 8.061487</td><td>0.02018927</td></tr>\n",
       "\t<tr><th scope=row>s1d1_GGTGCTGGTACTTCAC-1</th><td>s1d1_GGTGCTGGTACTTCAC-1</td><td> 4133</td><td>103</td><td>Myeloid</td><td> 8.326759</td><td>0.02492136</td></tr>\n",
       "\t<tr><th scope=row>s1d1_AGCTGCTCATTGTTGG-1</th><td>s1d1_AGCTGCTCATTGTTGG-1</td><td> 1308</td><td>102</td><td>Myeloid</td><td> 7.176255</td><td>0.07798165</td></tr>\n",
       "\t<tr><th scope=row>s1d1_TCCAGGATCAACAAGG-1</th><td>s1d1_TCCAGGATCAACAAGG-1</td><td> 3218</td><td> 66</td><td>Myeloid</td><td> 8.076515</td><td>0.02050963</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GTTGCTGAGTAATCCA-1</th><td>s4d9_GTTGCTGAGTAATCCA-1</td><td>2839</td><td>168</td><td>Myeloid</td><td>7.951207</td><td>0.05917577</td></tr>\n",
       "\t<tr><th scope=row>s4d9_AACTAGTGTTACTTCA-1</th><td>s4d9_AACTAGTGTTACTTCA-1</td><td>1455</td><td>198</td><td>Myeloid</td><td>7.282761</td><td>0.13608247</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TGCTTCCAGTCCTTTG-1</th><td>s4d9_TGCTTCCAGTCCTTTG-1</td><td>1450</td><td>138</td><td>Myeloid</td><td>7.279319</td><td>0.09517241</td></tr>\n",
       "\t<tr><th scope=row>s4d9_AAGGATCCACAGGATG-1</th><td>s4d9_AAGGATCCACAGGATG-1</td><td>1789</td><td>229</td><td>Myeloid</td><td>7.489412</td><td>0.12800447</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TGTGCGCAGCCGCAGT-1</th><td>s4d9_TGTGCGCAGCCGCAGT-1</td><td>1686</td><td>138</td><td>Myeloid</td><td>7.430114</td><td>0.08185053</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CCAGACTCAATCATGT-1</th><td>s4d9_CCAGACTCAATCATGT-1</td><td>1834</td><td> 52</td><td>Myeloid</td><td>7.514255</td><td>0.02835333</td></tr>\n",
       "\t<tr><th scope=row>s4d9_AGTAAGTAGTTAGAGG-1</th><td>s4d9_AGTAAGTAGTTAGAGG-1</td><td>1826</td><td>194</td><td>Myeloid</td><td>7.509883</td><td>0.10624315</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TGCGGTTCAACAGCCT-1</th><td>s4d9_TGCGGTTCAACAGCCT-1</td><td>1048</td><td>124</td><td>Myeloid</td><td>6.954639</td><td>0.11832061</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GATGACTTCGCTAGTG-1</th><td>s4d9_GATGACTTCGCTAGTG-1</td><td> 848</td><td> 94</td><td>Myeloid</td><td>6.742881</td><td>0.11084906</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CTAGTCGAGCTATTGA-1</th><td>s4d9_CTAGTCGAGCTATTGA-1</td><td> 829</td><td> 52</td><td>Myeloid</td><td>6.720220</td><td>0.06272618</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TTATGACCATTTAAGC-1</th><td>s4d9_TTATGACCATTTAAGC-1</td><td>3534</td><td>336</td><td>Myeloid</td><td>8.170186</td><td>0.09507640</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GTGCTCAAGCAAGATG-1</th><td>s4d9_GTGCTCAAGCAAGATG-1</td><td>3857</td><td>573</td><td>Myeloid</td><td>8.257645</td><td>0.14856106</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GGATATTGTGTCACGG-1</th><td>s4d9_GGATATTGTGTCACGG-1</td><td>2140</td><td>285</td><td>Myeloid</td><td>7.668561</td><td>0.13317757</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TGGCCTGCAATTATGC-1</th><td>s4d9_TGGCCTGCAATTATGC-1</td><td>3184</td><td>332</td><td>Myeloid</td><td>8.065894</td><td>0.10427136</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GAGGGAGCATAACGGG-1</th><td>s4d9_GAGGGAGCATAACGGG-1</td><td>2488</td><td>358</td><td>Myeloid</td><td>7.819234</td><td>0.14389068</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GGTACAAAGTTGCCTC-1</th><td>s4d9_GGTACAAAGTTGCCTC-1</td><td>2427</td><td>276</td><td>Myeloid</td><td>7.794411</td><td>0.11372064</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GGAAACTCATAGCTGC-1</th><td>s4d9_GGAAACTCATAGCTGC-1</td><td>1561</td><td>182</td><td>Myeloid</td><td>7.353082</td><td>0.11659193</td></tr>\n",
       "\t<tr><th scope=row>s4d9_AGAAAGGCAATGCCCG-1</th><td>s4d9_AGAAAGGCAATGCCCG-1</td><td>1773</td><td> 38</td><td>Myeloid</td><td>7.480428</td><td>0.02143260</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CTACAACAGCGAGTAA-1</th><td>s4d9_CTACAACAGCGAGTAA-1</td><td>1161</td><td> 96</td><td>Myeloid</td><td>7.057037</td><td>0.08268734</td></tr>\n",
       "\t<tr><th scope=row>s4d9_ACATTGCAGGGTCTAT-1</th><td>s4d9_ACATTGCAGGGTCTAT-1</td><td>2255</td><td>255</td><td>Myeloid</td><td>7.720905</td><td>0.11308204</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CTCCAACTCTAACCAA-1</th><td>s4d9_CTCCAACTCTAACCAA-1</td><td>3392</td><td>399</td><td>Myeloid</td><td>8.129175</td><td>0.11762972</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CAGGGCTTCCGCCAAA-1</th><td>s4d9_CAGGGCTTCCGCCAAA-1</td><td>1837</td><td>253</td><td>Myeloid</td><td>7.515889</td><td>0.13772455</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TAAACAGCACCTAAGC-1</th><td>s4d9_TAAACAGCACCTAAGC-1</td><td>1474</td><td>171</td><td>Myeloid</td><td>7.295735</td><td>0.11601085</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TGGTCCTTCAGTGAAG-1</th><td>s4d9_TGGTCCTTCAGTGAAG-1</td><td>2021</td><td>227</td><td>Myeloid</td><td>7.611348</td><td>0.11232063</td></tr>\n",
       "\t<tr><th scope=row>s4d9_GGATACTTCTTACTCG-1</th><td>s4d9_GGATACTTCTTACTCG-1</td><td>2211</td><td>192</td><td>Myeloid</td><td>7.701200</td><td>0.08683853</td></tr>\n",
       "\t<tr><th scope=row>s4d9_ACAGGAATCCGTCCAT-1</th><td>s4d9_ACAGGAATCCGTCCAT-1</td><td> 747</td><td> 61</td><td>Myeloid</td><td>6.616065</td><td>0.08165997</td></tr>\n",
       "\t<tr><th scope=row>s4d9_TGTGGCCAGCTGGCTA-1</th><td>s4d9_TGTGGCCAGCTGGCTA-1</td><td>1062</td><td>140</td><td>Myeloid</td><td>6.967909</td><td>0.13182674</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CCGGTTAAGCCTGTTC-1</th><td>s4d9_CCGGTTAAGCCTGTTC-1</td><td>1496</td><td> 58</td><td>Myeloid</td><td>7.310550</td><td>0.03877005</td></tr>\n",
       "\t<tr><th scope=row>s4d9_CAAGCTAGTCAACAAT-1</th><td>s4d9_CAAGCTAGTCAACAAT-1</td><td>1664</td><td>179</td><td>Myeloid</td><td>7.416980</td><td>0.10757212</td></tr>\n",
       "\t<tr><th scope=row>s4d9_ACAGGTAAGGTTACGT-1</th><td>s4d9_ACAGGTAAGGTTACGT-1</td><td>1221</td><td> 75</td><td>Myeloid</td><td>7.107425</td><td>0.06142506</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 14048 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & cell & nUMI & nMito & celltype & log.nUMI & percent.mito\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\ts1d1\\_CTATGGCCATAACGGG-1 & s1d1\\_CTATGGCCATAACGGG-1 &  2157 &  69 & Myeloid &  7.676474 & 0.03198887\\\\\n",
       "\ts1d1\\_ACCACATAGGTGTCCA-1 & s1d1\\_ACCACATAGGTGTCCA-1 &  2128 &  64 & Myeloid &  7.662938 & 0.03007519\\\\\n",
       "\ts1d1\\_TGCCGGTAGTCGCAAT-1 & s1d1\\_TGCCGGTAGTCGCAAT-1 &  1920 &  67 & Myeloid &  7.560080 & 0.03489583\\\\\n",
       "\ts1d1\\_CATCCCAGTCTTTATG-1 & s1d1\\_CATCCCAGTCTTTATG-1 &  3211 & 127 & Myeloid &  8.074338 & 0.03955154\\\\\n",
       "\ts1d1\\_TAGTGAGAGCTGCACA-1 & s1d1\\_TAGTGAGAGCTGCACA-1 &  3568 &  85 & Myeloid &  8.179760 & 0.02382287\\\\\n",
       "\ts1d1\\_CCTGTTGGTTATAGCG-1 & s1d1\\_CCTGTTGGTTATAGCG-1 &  4044 &  42 & Myeloid &  8.304990 & 0.01038576\\\\\n",
       "\ts1d1\\_TTTAGCTTCTGGCTTC-1 & s1d1\\_TTTAGCTTCTGGCTTC-1 &  2728 & 105 & Myeloid &  7.911324 & 0.03848974\\\\\n",
       "\ts1d1\\_TGTGAAACAGGATTAA-1 & s1d1\\_TGTGAAACAGGATTAA-1 &  2800 &  86 & Myeloid &  7.937375 & 0.03071429\\\\\n",
       "\ts1d1\\_ACCAGGCTCTTGGATA-1 & s1d1\\_ACCAGGCTCTTGGATA-1 &  1015 &  94 & Myeloid &  6.922644 & 0.09261084\\\\\n",
       "\ts1d1\\_CACAGGGAGACAGGTA-1 & s1d1\\_CACAGGGAGACAGGTA-1 &  3092 &  69 & Myeloid &  8.036573 & 0.02231565\\\\\n",
       "\ts1d1\\_GCATCCTTCGCTATGG-1 & s1d1\\_GCATCCTTCGCTATGG-1 &  3112 &  77 & Myeloid &  8.043021 & 0.02474293\\\\\n",
       "\ts1d1\\_TGCTATGCAGAGGGAG-1 & s1d1\\_TGCTATGCAGAGGGAG-1 &  2814 &  78 & Myeloid &  7.942362 & 0.02771855\\\\\n",
       "\ts1d1\\_CCCGCAACAAAGCGCA-1 & s1d1\\_CCCGCAACAAAGCGCA-1 &  3234 &  58 & Myeloid &  8.081475 & 0.01793445\\\\\n",
       "\ts1d1\\_AATCTTGAGAAGCTAG-1 & s1d1\\_AATCTTGAGAAGCTAG-1 &  1647 &  67 & Myeloid &  7.406711 & 0.04068002\\\\\n",
       "\ts1d1\\_TTGTCCATCATTAGGC-1 & s1d1\\_TTGTCCATCATTAGGC-1 &  4110 &  79 & Myeloid &  8.321178 & 0.01922141\\\\\n",
       "\ts1d1\\_TCCAGGATCCTAACGG-1 & s1d1\\_TCCAGGATCCTAACGG-1 &  3484 & 100 & Myeloid &  8.155936 & 0.02870264\\\\\n",
       "\ts1d1\\_GGTATGTTCTAACCAA-1 & s1d1\\_GGTATGTTCTAACCAA-1 &  2759 &  61 & Myeloid &  7.922624 & 0.02210946\\\\\n",
       "\ts1d1\\_TGGATTCAGTTTCCGC-1 & s1d1\\_TGGATTCAGTTTCCGC-1 &  3639 &  72 & Myeloid &  8.199464 & 0.01978566\\\\\n",
       "\ts1d1\\_TGCTTGCTCAATCTCT-1 & s1d1\\_TGCTTGCTCAATCTCT-1 &  1934 &  71 & Myeloid &  7.567346 & 0.03671148\\\\\n",
       "\ts1d1\\_CGATGTCCATACCCGG-1 & s1d1\\_CGATGTCCATACCCGG-1 &  3296 &  54 & Myeloid &  8.100465 & 0.01638350\\\\\n",
       "\ts1d1\\_ACGCCTAAGCGCCTTT-1 & s1d1\\_ACGCCTAAGCGCCTTT-1 &  3236 &  92 & Myeloid &  8.082093 & 0.02843016\\\\\n",
       "\ts1d1\\_GAGTAATAGCTTAACA-1 & s1d1\\_GAGTAATAGCTTAACA-1 &  4308 & 100 & Myeloid &  8.368229 & 0.02321263\\\\\n",
       "\ts1d1\\_ATTGGCTAGGGACCTC-1 & s1d1\\_ATTGGCTAGGGACCTC-1 & 22348 & 863 & Myeloid & 10.014492 & 0.03861643\\\\\n",
       "\ts1d1\\_CCTGTAACATTGTGCA-1 & s1d1\\_CCTGTAACATTGTGCA-1 &  3940 & 127 & Myeloid &  8.278936 & 0.03223350\\\\\n",
       "\ts1d1\\_AGATGAAGTGAGGTGA-1 & s1d1\\_AGATGAAGTGAGGTGA-1 &  2271 &  69 & Myeloid &  7.727976 & 0.03038309\\\\\n",
       "\ts1d1\\_GTTAACGGTAGCTGGT-1 & s1d1\\_GTTAACGGTAGCTGGT-1 &  1222 &  72 & Myeloid &  7.108244 & 0.05891980\\\\\n",
       "\ts1d1\\_GAACACAAGTAACCAC-1 & s1d1\\_GAACACAAGTAACCAC-1 &  3170 &  64 & Myeloid &  8.061487 & 0.02018927\\\\\n",
       "\ts1d1\\_GGTGCTGGTACTTCAC-1 & s1d1\\_GGTGCTGGTACTTCAC-1 &  4133 & 103 & Myeloid &  8.326759 & 0.02492136\\\\\n",
       "\ts1d1\\_AGCTGCTCATTGTTGG-1 & s1d1\\_AGCTGCTCATTGTTGG-1 &  1308 & 102 & Myeloid &  7.176255 & 0.07798165\\\\\n",
       "\ts1d1\\_TCCAGGATCAACAAGG-1 & s1d1\\_TCCAGGATCAACAAGG-1 &  3218 &  66 & Myeloid &  8.076515 & 0.02050963\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\ts4d9\\_GTTGCTGAGTAATCCA-1 & s4d9\\_GTTGCTGAGTAATCCA-1 & 2839 & 168 & Myeloid & 7.951207 & 0.05917577\\\\\n",
       "\ts4d9\\_AACTAGTGTTACTTCA-1 & s4d9\\_AACTAGTGTTACTTCA-1 & 1455 & 198 & Myeloid & 7.282761 & 0.13608247\\\\\n",
       "\ts4d9\\_TGCTTCCAGTCCTTTG-1 & s4d9\\_TGCTTCCAGTCCTTTG-1 & 1450 & 138 & Myeloid & 7.279319 & 0.09517241\\\\\n",
       "\ts4d9\\_AAGGATCCACAGGATG-1 & s4d9\\_AAGGATCCACAGGATG-1 & 1789 & 229 & Myeloid & 7.489412 & 0.12800447\\\\\n",
       "\ts4d9\\_TGTGCGCAGCCGCAGT-1 & s4d9\\_TGTGCGCAGCCGCAGT-1 & 1686 & 138 & Myeloid & 7.430114 & 0.08185053\\\\\n",
       "\ts4d9\\_CCAGACTCAATCATGT-1 & s4d9\\_CCAGACTCAATCATGT-1 & 1834 &  52 & Myeloid & 7.514255 & 0.02835333\\\\\n",
       "\ts4d9\\_AGTAAGTAGTTAGAGG-1 & s4d9\\_AGTAAGTAGTTAGAGG-1 & 1826 & 194 & Myeloid & 7.509883 & 0.10624315\\\\\n",
       "\ts4d9\\_TGCGGTTCAACAGCCT-1 & s4d9\\_TGCGGTTCAACAGCCT-1 & 1048 & 124 & Myeloid & 6.954639 & 0.11832061\\\\\n",
       "\ts4d9\\_GATGACTTCGCTAGTG-1 & s4d9\\_GATGACTTCGCTAGTG-1 &  848 &  94 & Myeloid & 6.742881 & 0.11084906\\\\\n",
       "\ts4d9\\_CTAGTCGAGCTATTGA-1 & s4d9\\_CTAGTCGAGCTATTGA-1 &  829 &  52 & Myeloid & 6.720220 & 0.06272618\\\\\n",
       "\ts4d9\\_TTATGACCATTTAAGC-1 & s4d9\\_TTATGACCATTTAAGC-1 & 3534 & 336 & Myeloid & 8.170186 & 0.09507640\\\\\n",
       "\ts4d9\\_GTGCTCAAGCAAGATG-1 & s4d9\\_GTGCTCAAGCAAGATG-1 & 3857 & 573 & Myeloid & 8.257645 & 0.14856106\\\\\n",
       "\ts4d9\\_GGATATTGTGTCACGG-1 & s4d9\\_GGATATTGTGTCACGG-1 & 2140 & 285 & Myeloid & 7.668561 & 0.13317757\\\\\n",
       "\ts4d9\\_TGGCCTGCAATTATGC-1 & s4d9\\_TGGCCTGCAATTATGC-1 & 3184 & 332 & Myeloid & 8.065894 & 0.10427136\\\\\n",
       "\ts4d9\\_GAGGGAGCATAACGGG-1 & s4d9\\_GAGGGAGCATAACGGG-1 & 2488 & 358 & Myeloid & 7.819234 & 0.14389068\\\\\n",
       "\ts4d9\\_GGTACAAAGTTGCCTC-1 & s4d9\\_GGTACAAAGTTGCCTC-1 & 2427 & 276 & Myeloid & 7.794411 & 0.11372064\\\\\n",
       "\ts4d9\\_GGAAACTCATAGCTGC-1 & s4d9\\_GGAAACTCATAGCTGC-1 & 1561 & 182 & Myeloid & 7.353082 & 0.11659193\\\\\n",
       "\ts4d9\\_AGAAAGGCAATGCCCG-1 & s4d9\\_AGAAAGGCAATGCCCG-1 & 1773 &  38 & Myeloid & 7.480428 & 0.02143260\\\\\n",
       "\ts4d9\\_CTACAACAGCGAGTAA-1 & s4d9\\_CTACAACAGCGAGTAA-1 & 1161 &  96 & Myeloid & 7.057037 & 0.08268734\\\\\n",
       "\ts4d9\\_ACATTGCAGGGTCTAT-1 & s4d9\\_ACATTGCAGGGTCTAT-1 & 2255 & 255 & Myeloid & 7.720905 & 0.11308204\\\\\n",
       "\ts4d9\\_CTCCAACTCTAACCAA-1 & s4d9\\_CTCCAACTCTAACCAA-1 & 3392 & 399 & Myeloid & 8.129175 & 0.11762972\\\\\n",
       "\ts4d9\\_CAGGGCTTCCGCCAAA-1 & s4d9\\_CAGGGCTTCCGCCAAA-1 & 1837 & 253 & Myeloid & 7.515889 & 0.13772455\\\\\n",
       "\ts4d9\\_TAAACAGCACCTAAGC-1 & s4d9\\_TAAACAGCACCTAAGC-1 & 1474 & 171 & Myeloid & 7.295735 & 0.11601085\\\\\n",
       "\ts4d9\\_TGGTCCTTCAGTGAAG-1 & s4d9\\_TGGTCCTTCAGTGAAG-1 & 2021 & 227 & Myeloid & 7.611348 & 0.11232063\\\\\n",
       "\ts4d9\\_GGATACTTCTTACTCG-1 & s4d9\\_GGATACTTCTTACTCG-1 & 2211 & 192 & Myeloid & 7.701200 & 0.08683853\\\\\n",
       "\ts4d9\\_ACAGGAATCCGTCCAT-1 & s4d9\\_ACAGGAATCCGTCCAT-1 &  747 &  61 & Myeloid & 6.616065 & 0.08165997\\\\\n",
       "\ts4d9\\_TGTGGCCAGCTGGCTA-1 & s4d9\\_TGTGGCCAGCTGGCTA-1 & 1062 & 140 & Myeloid & 6.967909 & 0.13182674\\\\\n",
       "\ts4d9\\_CCGGTTAAGCCTGTTC-1 & s4d9\\_CCGGTTAAGCCTGTTC-1 & 1496 &  58 & Myeloid & 7.310550 & 0.03877005\\\\\n",
       "\ts4d9\\_CAAGCTAGTCAACAAT-1 & s4d9\\_CAAGCTAGTCAACAAT-1 & 1664 & 179 & Myeloid & 7.416980 & 0.10757212\\\\\n",
       "\ts4d9\\_ACAGGTAAGGTTACGT-1 & s4d9\\_ACAGGTAAGGTTACGT-1 & 1221 &  75 & Myeloid & 7.107425 & 0.06142506\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 14048 × 6\n",
       "\n",
       "| <!--/--> | cell &lt;chr&gt; | nUMI &lt;dbl&gt; | nMito &lt;dbl&gt; | celltype &lt;chr&gt; | log.nUMI &lt;dbl&gt; | percent.mito &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| s1d1_CTATGGCCATAACGGG-1 | s1d1_CTATGGCCATAACGGG-1 |  2157 |  69 | Myeloid |  7.676474 | 0.03198887 |\n",
       "| s1d1_ACCACATAGGTGTCCA-1 | s1d1_ACCACATAGGTGTCCA-1 |  2128 |  64 | Myeloid |  7.662938 | 0.03007519 |\n",
       "| s1d1_TGCCGGTAGTCGCAAT-1 | s1d1_TGCCGGTAGTCGCAAT-1 |  1920 |  67 | Myeloid |  7.560080 | 0.03489583 |\n",
       "| s1d1_CATCCCAGTCTTTATG-1 | s1d1_CATCCCAGTCTTTATG-1 |  3211 | 127 | Myeloid |  8.074338 | 0.03955154 |\n",
       "| s1d1_TAGTGAGAGCTGCACA-1 | s1d1_TAGTGAGAGCTGCACA-1 |  3568 |  85 | Myeloid |  8.179760 | 0.02382287 |\n",
       "| s1d1_CCTGTTGGTTATAGCG-1 | s1d1_CCTGTTGGTTATAGCG-1 |  4044 |  42 | Myeloid |  8.304990 | 0.01038576 |\n",
       "| s1d1_TTTAGCTTCTGGCTTC-1 | s1d1_TTTAGCTTCTGGCTTC-1 |  2728 | 105 | Myeloid |  7.911324 | 0.03848974 |\n",
       "| s1d1_TGTGAAACAGGATTAA-1 | s1d1_TGTGAAACAGGATTAA-1 |  2800 |  86 | Myeloid |  7.937375 | 0.03071429 |\n",
       "| s1d1_ACCAGGCTCTTGGATA-1 | s1d1_ACCAGGCTCTTGGATA-1 |  1015 |  94 | Myeloid |  6.922644 | 0.09261084 |\n",
       "| s1d1_CACAGGGAGACAGGTA-1 | s1d1_CACAGGGAGACAGGTA-1 |  3092 |  69 | Myeloid |  8.036573 | 0.02231565 |\n",
       "| s1d1_GCATCCTTCGCTATGG-1 | s1d1_GCATCCTTCGCTATGG-1 |  3112 |  77 | Myeloid |  8.043021 | 0.02474293 |\n",
       "| s1d1_TGCTATGCAGAGGGAG-1 | s1d1_TGCTATGCAGAGGGAG-1 |  2814 |  78 | Myeloid |  7.942362 | 0.02771855 |\n",
       "| s1d1_CCCGCAACAAAGCGCA-1 | s1d1_CCCGCAACAAAGCGCA-1 |  3234 |  58 | Myeloid |  8.081475 | 0.01793445 |\n",
       "| s1d1_AATCTTGAGAAGCTAG-1 | s1d1_AATCTTGAGAAGCTAG-1 |  1647 |  67 | Myeloid |  7.406711 | 0.04068002 |\n",
       "| s1d1_TTGTCCATCATTAGGC-1 | s1d1_TTGTCCATCATTAGGC-1 |  4110 |  79 | Myeloid |  8.321178 | 0.01922141 |\n",
       "| s1d1_TCCAGGATCCTAACGG-1 | s1d1_TCCAGGATCCTAACGG-1 |  3484 | 100 | Myeloid |  8.155936 | 0.02870264 |\n",
       "| s1d1_GGTATGTTCTAACCAA-1 | s1d1_GGTATGTTCTAACCAA-1 |  2759 |  61 | Myeloid |  7.922624 | 0.02210946 |\n",
       "| s1d1_TGGATTCAGTTTCCGC-1 | s1d1_TGGATTCAGTTTCCGC-1 |  3639 |  72 | Myeloid |  8.199464 | 0.01978566 |\n",
       "| s1d1_TGCTTGCTCAATCTCT-1 | s1d1_TGCTTGCTCAATCTCT-1 |  1934 |  71 | Myeloid |  7.567346 | 0.03671148 |\n",
       "| s1d1_CGATGTCCATACCCGG-1 | s1d1_CGATGTCCATACCCGG-1 |  3296 |  54 | Myeloid |  8.100465 | 0.01638350 |\n",
       "| s1d1_ACGCCTAAGCGCCTTT-1 | s1d1_ACGCCTAAGCGCCTTT-1 |  3236 |  92 | Myeloid |  8.082093 | 0.02843016 |\n",
       "| s1d1_GAGTAATAGCTTAACA-1 | s1d1_GAGTAATAGCTTAACA-1 |  4308 | 100 | Myeloid |  8.368229 | 0.02321263 |\n",
       "| s1d1_ATTGGCTAGGGACCTC-1 | s1d1_ATTGGCTAGGGACCTC-1 | 22348 | 863 | Myeloid | 10.014492 | 0.03861643 |\n",
       "| s1d1_CCTGTAACATTGTGCA-1 | s1d1_CCTGTAACATTGTGCA-1 |  3940 | 127 | Myeloid |  8.278936 | 0.03223350 |\n",
       "| s1d1_AGATGAAGTGAGGTGA-1 | s1d1_AGATGAAGTGAGGTGA-1 |  2271 |  69 | Myeloid |  7.727976 | 0.03038309 |\n",
       "| s1d1_GTTAACGGTAGCTGGT-1 | s1d1_GTTAACGGTAGCTGGT-1 |  1222 |  72 | Myeloid |  7.108244 | 0.05891980 |\n",
       "| s1d1_GAACACAAGTAACCAC-1 | s1d1_GAACACAAGTAACCAC-1 |  3170 |  64 | Myeloid |  8.061487 | 0.02018927 |\n",
       "| s1d1_GGTGCTGGTACTTCAC-1 | s1d1_GGTGCTGGTACTTCAC-1 |  4133 | 103 | Myeloid |  8.326759 | 0.02492136 |\n",
       "| s1d1_AGCTGCTCATTGTTGG-1 | s1d1_AGCTGCTCATTGTTGG-1 |  1308 | 102 | Myeloid |  7.176255 | 0.07798165 |\n",
       "| s1d1_TCCAGGATCAACAAGG-1 | s1d1_TCCAGGATCAACAAGG-1 |  3218 |  66 | Myeloid |  8.076515 | 0.02050963 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| s4d9_GTTGCTGAGTAATCCA-1 | s4d9_GTTGCTGAGTAATCCA-1 | 2839 | 168 | Myeloid | 7.951207 | 0.05917577 |\n",
       "| s4d9_AACTAGTGTTACTTCA-1 | s4d9_AACTAGTGTTACTTCA-1 | 1455 | 198 | Myeloid | 7.282761 | 0.13608247 |\n",
       "| s4d9_TGCTTCCAGTCCTTTG-1 | s4d9_TGCTTCCAGTCCTTTG-1 | 1450 | 138 | Myeloid | 7.279319 | 0.09517241 |\n",
       "| s4d9_AAGGATCCACAGGATG-1 | s4d9_AAGGATCCACAGGATG-1 | 1789 | 229 | Myeloid | 7.489412 | 0.12800447 |\n",
       "| s4d9_TGTGCGCAGCCGCAGT-1 | s4d9_TGTGCGCAGCCGCAGT-1 | 1686 | 138 | Myeloid | 7.430114 | 0.08185053 |\n",
       "| s4d9_CCAGACTCAATCATGT-1 | s4d9_CCAGACTCAATCATGT-1 | 1834 |  52 | Myeloid | 7.514255 | 0.02835333 |\n",
       "| s4d9_AGTAAGTAGTTAGAGG-1 | s4d9_AGTAAGTAGTTAGAGG-1 | 1826 | 194 | Myeloid | 7.509883 | 0.10624315 |\n",
       "| s4d9_TGCGGTTCAACAGCCT-1 | s4d9_TGCGGTTCAACAGCCT-1 | 1048 | 124 | Myeloid | 6.954639 | 0.11832061 |\n",
       "| s4d9_GATGACTTCGCTAGTG-1 | s4d9_GATGACTTCGCTAGTG-1 |  848 |  94 | Myeloid | 6.742881 | 0.11084906 |\n",
       "| s4d9_CTAGTCGAGCTATTGA-1 | s4d9_CTAGTCGAGCTATTGA-1 |  829 |  52 | Myeloid | 6.720220 | 0.06272618 |\n",
       "| s4d9_TTATGACCATTTAAGC-1 | s4d9_TTATGACCATTTAAGC-1 | 3534 | 336 | Myeloid | 8.170186 | 0.09507640 |\n",
       "| s4d9_GTGCTCAAGCAAGATG-1 | s4d9_GTGCTCAAGCAAGATG-1 | 3857 | 573 | Myeloid | 8.257645 | 0.14856106 |\n",
       "| s4d9_GGATATTGTGTCACGG-1 | s4d9_GGATATTGTGTCACGG-1 | 2140 | 285 | Myeloid | 7.668561 | 0.13317757 |\n",
       "| s4d9_TGGCCTGCAATTATGC-1 | s4d9_TGGCCTGCAATTATGC-1 | 3184 | 332 | Myeloid | 8.065894 | 0.10427136 |\n",
       "| s4d9_GAGGGAGCATAACGGG-1 | s4d9_GAGGGAGCATAACGGG-1 | 2488 | 358 | Myeloid | 7.819234 | 0.14389068 |\n",
       "| s4d9_GGTACAAAGTTGCCTC-1 | s4d9_GGTACAAAGTTGCCTC-1 | 2427 | 276 | Myeloid | 7.794411 | 0.11372064 |\n",
       "| s4d9_GGAAACTCATAGCTGC-1 | s4d9_GGAAACTCATAGCTGC-1 | 1561 | 182 | Myeloid | 7.353082 | 0.11659193 |\n",
       "| s4d9_AGAAAGGCAATGCCCG-1 | s4d9_AGAAAGGCAATGCCCG-1 | 1773 |  38 | Myeloid | 7.480428 | 0.02143260 |\n",
       "| s4d9_CTACAACAGCGAGTAA-1 | s4d9_CTACAACAGCGAGTAA-1 | 1161 |  96 | Myeloid | 7.057037 | 0.08268734 |\n",
       "| s4d9_ACATTGCAGGGTCTAT-1 | s4d9_ACATTGCAGGGTCTAT-1 | 2255 | 255 | Myeloid | 7.720905 | 0.11308204 |\n",
       "| s4d9_CTCCAACTCTAACCAA-1 | s4d9_CTCCAACTCTAACCAA-1 | 3392 | 399 | Myeloid | 8.129175 | 0.11762972 |\n",
       "| s4d9_CAGGGCTTCCGCCAAA-1 | s4d9_CAGGGCTTCCGCCAAA-1 | 1837 | 253 | Myeloid | 7.515889 | 0.13772455 |\n",
       "| s4d9_TAAACAGCACCTAAGC-1 | s4d9_TAAACAGCACCTAAGC-1 | 1474 | 171 | Myeloid | 7.295735 | 0.11601085 |\n",
       "| s4d9_TGGTCCTTCAGTGAAG-1 | s4d9_TGGTCCTTCAGTGAAG-1 | 2021 | 227 | Myeloid | 7.611348 | 0.11232063 |\n",
       "| s4d9_GGATACTTCTTACTCG-1 | s4d9_GGATACTTCTTACTCG-1 | 2211 | 192 | Myeloid | 7.701200 | 0.08683853 |\n",
       "| s4d9_ACAGGAATCCGTCCAT-1 | s4d9_ACAGGAATCCGTCCAT-1 |  747 |  61 | Myeloid | 6.616065 | 0.08165997 |\n",
       "| s4d9_TGTGGCCAGCTGGCTA-1 | s4d9_TGTGGCCAGCTGGCTA-1 | 1062 | 140 | Myeloid | 6.967909 | 0.13182674 |\n",
       "| s4d9_CCGGTTAAGCCTGTTC-1 | s4d9_CCGGTTAAGCCTGTTC-1 | 1496 |  58 | Myeloid | 7.310550 | 0.03877005 |\n",
       "| s4d9_CAAGCTAGTCAACAAT-1 | s4d9_CAAGCTAGTCAACAAT-1 | 1664 | 179 | Myeloid | 7.416980 | 0.10757212 |\n",
       "| s4d9_ACAGGTAAGGTTACGT-1 | s4d9_ACAGGTAAGGTTACGT-1 | 1221 |  75 | Myeloid | 7.107425 | 0.06142506 |\n",
       "\n"
      ],
      "text/plain": [
       "                        cell                    nUMI  nMito celltype log.nUMI \n",
       "s1d1_CTATGGCCATAACGGG-1 s1d1_CTATGGCCATAACGGG-1  2157  69   Myeloid   7.676474\n",
       "s1d1_ACCACATAGGTGTCCA-1 s1d1_ACCACATAGGTGTCCA-1  2128  64   Myeloid   7.662938\n",
       "s1d1_TGCCGGTAGTCGCAAT-1 s1d1_TGCCGGTAGTCGCAAT-1  1920  67   Myeloid   7.560080\n",
       "s1d1_CATCCCAGTCTTTATG-1 s1d1_CATCCCAGTCTTTATG-1  3211 127   Myeloid   8.074338\n",
       "s1d1_TAGTGAGAGCTGCACA-1 s1d1_TAGTGAGAGCTGCACA-1  3568  85   Myeloid   8.179760\n",
       "s1d1_CCTGTTGGTTATAGCG-1 s1d1_CCTGTTGGTTATAGCG-1  4044  42   Myeloid   8.304990\n",
       "s1d1_TTTAGCTTCTGGCTTC-1 s1d1_TTTAGCTTCTGGCTTC-1  2728 105   Myeloid   7.911324\n",
       "s1d1_TGTGAAACAGGATTAA-1 s1d1_TGTGAAACAGGATTAA-1  2800  86   Myeloid   7.937375\n",
       "s1d1_ACCAGGCTCTTGGATA-1 s1d1_ACCAGGCTCTTGGATA-1  1015  94   Myeloid   6.922644\n",
       "s1d1_CACAGGGAGACAGGTA-1 s1d1_CACAGGGAGACAGGTA-1  3092  69   Myeloid   8.036573\n",
       "s1d1_GCATCCTTCGCTATGG-1 s1d1_GCATCCTTCGCTATGG-1  3112  77   Myeloid   8.043021\n",
       "s1d1_TGCTATGCAGAGGGAG-1 s1d1_TGCTATGCAGAGGGAG-1  2814  78   Myeloid   7.942362\n",
       "s1d1_CCCGCAACAAAGCGCA-1 s1d1_CCCGCAACAAAGCGCA-1  3234  58   Myeloid   8.081475\n",
       "s1d1_AATCTTGAGAAGCTAG-1 s1d1_AATCTTGAGAAGCTAG-1  1647  67   Myeloid   7.406711\n",
       "s1d1_TTGTCCATCATTAGGC-1 s1d1_TTGTCCATCATTAGGC-1  4110  79   Myeloid   8.321178\n",
       "s1d1_TCCAGGATCCTAACGG-1 s1d1_TCCAGGATCCTAACGG-1  3484 100   Myeloid   8.155936\n",
       "s1d1_GGTATGTTCTAACCAA-1 s1d1_GGTATGTTCTAACCAA-1  2759  61   Myeloid   7.922624\n",
       "s1d1_TGGATTCAGTTTCCGC-1 s1d1_TGGATTCAGTTTCCGC-1  3639  72   Myeloid   8.199464\n",
       "s1d1_TGCTTGCTCAATCTCT-1 s1d1_TGCTTGCTCAATCTCT-1  1934  71   Myeloid   7.567346\n",
       "s1d1_CGATGTCCATACCCGG-1 s1d1_CGATGTCCATACCCGG-1  3296  54   Myeloid   8.100465\n",
       "s1d1_ACGCCTAAGCGCCTTT-1 s1d1_ACGCCTAAGCGCCTTT-1  3236  92   Myeloid   8.082093\n",
       "s1d1_GAGTAATAGCTTAACA-1 s1d1_GAGTAATAGCTTAACA-1  4308 100   Myeloid   8.368229\n",
       "s1d1_ATTGGCTAGGGACCTC-1 s1d1_ATTGGCTAGGGACCTC-1 22348 863   Myeloid  10.014492\n",
       "s1d1_CCTGTAACATTGTGCA-1 s1d1_CCTGTAACATTGTGCA-1  3940 127   Myeloid   8.278936\n",
       "s1d1_AGATGAAGTGAGGTGA-1 s1d1_AGATGAAGTGAGGTGA-1  2271  69   Myeloid   7.727976\n",
       "s1d1_GTTAACGGTAGCTGGT-1 s1d1_GTTAACGGTAGCTGGT-1  1222  72   Myeloid   7.108244\n",
       "s1d1_GAACACAAGTAACCAC-1 s1d1_GAACACAAGTAACCAC-1  3170  64   Myeloid   8.061487\n",
       "s1d1_GGTGCTGGTACTTCAC-1 s1d1_GGTGCTGGTACTTCAC-1  4133 103   Myeloid   8.326759\n",
       "s1d1_AGCTGCTCATTGTTGG-1 s1d1_AGCTGCTCATTGTTGG-1  1308 102   Myeloid   7.176255\n",
       "s1d1_TCCAGGATCAACAAGG-1 s1d1_TCCAGGATCAACAAGG-1  3218  66   Myeloid   8.076515\n",
       "⋮                       ⋮                       ⋮     ⋮     ⋮        ⋮        \n",
       "s4d9_GTTGCTGAGTAATCCA-1 s4d9_GTTGCTGAGTAATCCA-1 2839  168   Myeloid  7.951207 \n",
       "s4d9_AACTAGTGTTACTTCA-1 s4d9_AACTAGTGTTACTTCA-1 1455  198   Myeloid  7.282761 \n",
       "s4d9_TGCTTCCAGTCCTTTG-1 s4d9_TGCTTCCAGTCCTTTG-1 1450  138   Myeloid  7.279319 \n",
       "s4d9_AAGGATCCACAGGATG-1 s4d9_AAGGATCCACAGGATG-1 1789  229   Myeloid  7.489412 \n",
       "s4d9_TGTGCGCAGCCGCAGT-1 s4d9_TGTGCGCAGCCGCAGT-1 1686  138   Myeloid  7.430114 \n",
       "s4d9_CCAGACTCAATCATGT-1 s4d9_CCAGACTCAATCATGT-1 1834   52   Myeloid  7.514255 \n",
       "s4d9_AGTAAGTAGTTAGAGG-1 s4d9_AGTAAGTAGTTAGAGG-1 1826  194   Myeloid  7.509883 \n",
       "s4d9_TGCGGTTCAACAGCCT-1 s4d9_TGCGGTTCAACAGCCT-1 1048  124   Myeloid  6.954639 \n",
       "s4d9_GATGACTTCGCTAGTG-1 s4d9_GATGACTTCGCTAGTG-1  848   94   Myeloid  6.742881 \n",
       "s4d9_CTAGTCGAGCTATTGA-1 s4d9_CTAGTCGAGCTATTGA-1  829   52   Myeloid  6.720220 \n",
       "s4d9_TTATGACCATTTAAGC-1 s4d9_TTATGACCATTTAAGC-1 3534  336   Myeloid  8.170186 \n",
       "s4d9_GTGCTCAAGCAAGATG-1 s4d9_GTGCTCAAGCAAGATG-1 3857  573   Myeloid  8.257645 \n",
       "s4d9_GGATATTGTGTCACGG-1 s4d9_GGATATTGTGTCACGG-1 2140  285   Myeloid  7.668561 \n",
       "s4d9_TGGCCTGCAATTATGC-1 s4d9_TGGCCTGCAATTATGC-1 3184  332   Myeloid  8.065894 \n",
       "s4d9_GAGGGAGCATAACGGG-1 s4d9_GAGGGAGCATAACGGG-1 2488  358   Myeloid  7.819234 \n",
       "s4d9_GGTACAAAGTTGCCTC-1 s4d9_GGTACAAAGTTGCCTC-1 2427  276   Myeloid  7.794411 \n",
       "s4d9_GGAAACTCATAGCTGC-1 s4d9_GGAAACTCATAGCTGC-1 1561  182   Myeloid  7.353082 \n",
       "s4d9_AGAAAGGCAATGCCCG-1 s4d9_AGAAAGGCAATGCCCG-1 1773   38   Myeloid  7.480428 \n",
       "s4d9_CTACAACAGCGAGTAA-1 s4d9_CTACAACAGCGAGTAA-1 1161   96   Myeloid  7.057037 \n",
       "s4d9_ACATTGCAGGGTCTAT-1 s4d9_ACATTGCAGGGTCTAT-1 2255  255   Myeloid  7.720905 \n",
       "s4d9_CTCCAACTCTAACCAA-1 s4d9_CTCCAACTCTAACCAA-1 3392  399   Myeloid  8.129175 \n",
       "s4d9_CAGGGCTTCCGCCAAA-1 s4d9_CAGGGCTTCCGCCAAA-1 1837  253   Myeloid  7.515889 \n",
       "s4d9_TAAACAGCACCTAAGC-1 s4d9_TAAACAGCACCTAAGC-1 1474  171   Myeloid  7.295735 \n",
       "s4d9_TGGTCCTTCAGTGAAG-1 s4d9_TGGTCCTTCAGTGAAG-1 2021  227   Myeloid  7.611348 \n",
       "s4d9_GGATACTTCTTACTCG-1 s4d9_GGATACTTCTTACTCG-1 2211  192   Myeloid  7.701200 \n",
       "s4d9_ACAGGAATCCGTCCAT-1 s4d9_ACAGGAATCCGTCCAT-1  747   61   Myeloid  6.616065 \n",
       "s4d9_TGTGGCCAGCTGGCTA-1 s4d9_TGTGGCCAGCTGGCTA-1 1062  140   Myeloid  6.967909 \n",
       "s4d9_CCGGTTAAGCCTGTTC-1 s4d9_CCGGTTAAGCCTGTTC-1 1496   58   Myeloid  7.310550 \n",
       "s4d9_CAAGCTAGTCAACAAT-1 s4d9_CAAGCTAGTCAACAAT-1 1664  179   Myeloid  7.416980 \n",
       "s4d9_ACAGGTAAGGTTACGT-1 s4d9_ACAGGTAAGGTTACGT-1 1221   75   Myeloid  7.107425 \n",
       "                        percent.mito\n",
       "s1d1_CTATGGCCATAACGGG-1 0.03198887  \n",
       "s1d1_ACCACATAGGTGTCCA-1 0.03007519  \n",
       "s1d1_TGCCGGTAGTCGCAAT-1 0.03489583  \n",
       "s1d1_CATCCCAGTCTTTATG-1 0.03955154  \n",
       "s1d1_TAGTGAGAGCTGCACA-1 0.02382287  \n",
       "s1d1_CCTGTTGGTTATAGCG-1 0.01038576  \n",
       "s1d1_TTTAGCTTCTGGCTTC-1 0.03848974  \n",
       "s1d1_TGTGAAACAGGATTAA-1 0.03071429  \n",
       "s1d1_ACCAGGCTCTTGGATA-1 0.09261084  \n",
       "s1d1_CACAGGGAGACAGGTA-1 0.02231565  \n",
       "s1d1_GCATCCTTCGCTATGG-1 0.02474293  \n",
       "s1d1_TGCTATGCAGAGGGAG-1 0.02771855  \n",
       "s1d1_CCCGCAACAAAGCGCA-1 0.01793445  \n",
       "s1d1_AATCTTGAGAAGCTAG-1 0.04068002  \n",
       "s1d1_TTGTCCATCATTAGGC-1 0.01922141  \n",
       "s1d1_TCCAGGATCCTAACGG-1 0.02870264  \n",
       "s1d1_GGTATGTTCTAACCAA-1 0.02210946  \n",
       "s1d1_TGGATTCAGTTTCCGC-1 0.01978566  \n",
       "s1d1_TGCTTGCTCAATCTCT-1 0.03671148  \n",
       "s1d1_CGATGTCCATACCCGG-1 0.01638350  \n",
       "s1d1_ACGCCTAAGCGCCTTT-1 0.02843016  \n",
       "s1d1_GAGTAATAGCTTAACA-1 0.02321263  \n",
       "s1d1_ATTGGCTAGGGACCTC-1 0.03861643  \n",
       "s1d1_CCTGTAACATTGTGCA-1 0.03223350  \n",
       "s1d1_AGATGAAGTGAGGTGA-1 0.03038309  \n",
       "s1d1_GTTAACGGTAGCTGGT-1 0.05891980  \n",
       "s1d1_GAACACAAGTAACCAC-1 0.02018927  \n",
       "s1d1_GGTGCTGGTACTTCAC-1 0.02492136  \n",
       "s1d1_AGCTGCTCATTGTTGG-1 0.07798165  \n",
       "s1d1_TCCAGGATCAACAAGG-1 0.02050963  \n",
       "⋮                       ⋮           \n",
       "s4d9_GTTGCTGAGTAATCCA-1 0.05917577  \n",
       "s4d9_AACTAGTGTTACTTCA-1 0.13608247  \n",
       "s4d9_TGCTTCCAGTCCTTTG-1 0.09517241  \n",
       "s4d9_AAGGATCCACAGGATG-1 0.12800447  \n",
       "s4d9_TGTGCGCAGCCGCAGT-1 0.08185053  \n",
       "s4d9_CCAGACTCAATCATGT-1 0.02835333  \n",
       "s4d9_AGTAAGTAGTTAGAGG-1 0.10624315  \n",
       "s4d9_TGCGGTTCAACAGCCT-1 0.11832061  \n",
       "s4d9_GATGACTTCGCTAGTG-1 0.11084906  \n",
       "s4d9_CTAGTCGAGCTATTGA-1 0.06272618  \n",
       "s4d9_TTATGACCATTTAAGC-1 0.09507640  \n",
       "s4d9_GTGCTCAAGCAAGATG-1 0.14856106  \n",
       "s4d9_GGATATTGTGTCACGG-1 0.13317757  \n",
       "s4d9_TGGCCTGCAATTATGC-1 0.10427136  \n",
       "s4d9_GAGGGAGCATAACGGG-1 0.14389068  \n",
       "s4d9_GGTACAAAGTTGCCTC-1 0.11372064  \n",
       "s4d9_GGAAACTCATAGCTGC-1 0.11659193  \n",
       "s4d9_AGAAAGGCAATGCCCG-1 0.02143260  \n",
       "s4d9_CTACAACAGCGAGTAA-1 0.08268734  \n",
       "s4d9_ACATTGCAGGGTCTAT-1 0.11308204  \n",
       "s4d9_CTCCAACTCTAACCAA-1 0.11762972  \n",
       "s4d9_CAGGGCTTCCGCCAAA-1 0.13772455  \n",
       "s4d9_TAAACAGCACCTAAGC-1 0.11601085  \n",
       "s4d9_TGGTCCTTCAGTGAAG-1 0.11232063  \n",
       "s4d9_GGATACTTCTTACTCG-1 0.08683853  \n",
       "s4d9_ACAGGAATCCGTCCAT-1 0.08165997  \n",
       "s4d9_TGTGGCCAGCTGGCTA-1 0.13182674  \n",
       "s4d9_CCGGTTAAGCCTGTTC-1 0.03877005  \n",
       "s4d9_CAAGCTAGTCAACAAT-1 0.10757212  \n",
       "s4d9_ACAGGTAAGGTTACGT-1 0.06142506  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meta.data = data.frame(cell = colnames(matrix.rna),\n",
    "                       nUMI = colSums(matrix.rna),\n",
    "                       nMito = colSums(matrix.rna[grep(\"MT-\",rownames(matrix.rna)),]),\n",
    "                       celltype = celltype)\n",
    "meta.data[,\"log.nUMI\"] = log(meta.data[,\"nUMI\"])\n",
    "meta.data[,\"percent.mito\"] = meta.data[,\"nMito\"] / meta.data[,\"nUMI\"]\n",
    "meta.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b823cb4a-3ce6-40b1-9a0d-d5d1e4aa3912",
   "metadata": {},
   "source": [
    "RUN SCENT prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f46765dc-9de5-4411-9288-40b59435f9fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "RhpcBLASctl::blas_set_num_threads(1)\n",
    "Sys.setenv(OPENBLAS_NUM_THREADS=1)\n",
    "SCENT_algorithm.modified = function (object, \n",
    "                                     # celltype, \n",
    "                                     # ncores, \n",
    "                                     # regr = \"poisson\", \n",
    "                                     # bin = TRUE) \n",
    "                                     regr = \"poisson\") {\n",
    "    res <- data.frame()\n",
    "    for (n in 1:nrow(object@peak.info)) {        \n",
    "        gene <- object@peak.info[n, 1]\n",
    "        this_peak <- object@peak.info[n, 2]\n",
    "        atac_target <- data.frame(cell = colnames(object@atac), \n",
    "            atac = object@atac[this_peak, ])\n",
    "        # if (bin) {\n",
    "        #     atac_target[atac_target$atac > 0, ]$atac <- 1\n",
    "        # }\n",
    "        mrna_target <- object@rna[gene, ]\n",
    "        df <- data.frame(cell = names(mrna_target), exprs = as.numeric(mrna_target))\n",
    "        df <- merge(df, atac_target, by = \"cell\")\n",
    "        df <- merge(df, object@meta.data, by = \"cell\")\n",
    "        # df2 <- df[df[[object@celltypes]] == celltype, ]\n",
    "        df2 <- df\n",
    "        # nonzero_m <- length(df2$exprs[df2$exprs > 0])/length(df2$exprs)\n",
    "        # nonzero_a <- length(df2$atac[df2$atac > 0])/length(df2$atac)\n",
    "        # if (nonzero_m > 0.05 & nonzero_a > 0.05) {\n",
    "            res_var <- \"exprs\"\n",
    "            pred_var <- c(\"atac\", object@covariates)\n",
    "            formula <- as.formula(paste(res_var, paste(pred_var, \n",
    "                collapse = \"+\"), sep = \"~\"))\n",
    "            if (regr == \"poisson\") {\n",
    "                base = glm(formula, family = \"poisson\", data = df2)\n",
    "                coefs <- summary(base)$coefficients[\"atac\", ]\n",
    "                # assoc <- assoc_poisson\n",
    "                assoc <- SCENT::assoc_poisson\n",
    "            }\n",
    "            else if (regr == \"negbin\") {\n",
    "                base = glm.nb(formula, data = df2)\n",
    "                coefs <- summary(base)$coefficients[\"atac\", ]\n",
    "                # assoc <- assoc_negbin\n",
    "                assoc <- SCENT::assoc_negbin\n",
    "            }\n",
    "            # bs = boot::boot(df2, assoc, R = 100, formula = formula, \n",
    "            #     stype = \"i\", parallel = \"no\", ncpus = 1)\n",
    "            # p0 = basic_p(bs$t0[1], bs$t[, 1])\n",
    "            # p0 = SCENT::basic_p(bs$t0[1], bs$t[, 1])\n",
    "            # if (p0 < 0.1) {\n",
    "            #     bs = boot::boot(df2, assoc, R = 500, formula = formula, \n",
    "            #       stype = \"i\", \n",
    "            #       # parallel = \"multicore\", ncpus = ncores)\n",
    "            #       parallel = \"no\", ncpus = 1)\n",
    "            #     # p0 = basic_p(bs$t0[1], bs$t[, 1])\n",
    "            #     p0 = SCENT::basic_p(bs$t0[1], bs$t[, 1])\n",
    "            # }\n",
    "            # if (p0 < 0.05) {\n",
    "                bs = boot::boot(df2, assoc, R = 2500, formula = formula, \n",
    "                  stype = \"i\", \n",
    "                  # parallel = \"multicore\", ncpus = ncores)\n",
    "                  parallel = \"no\", ncpus = 1)\n",
    "                # p0 = basic_p(bs$t0[1], bs$t[, 1])\n",
    "                p0 = SCENT::basic_p(bs$t0[1], bs$t[, 1])\n",
    "            # }\n",
    "            # if (p0 < 0.01) {\n",
    "            #     bs = boot::boot(df2, assoc, R = 25000, formula = formula, \n",
    "            #       stype = \"i\", \n",
    "            #       # parallel = \"multicore\", ncpus = ncores)\n",
    "            #       parallel = \"no\", ncpus = 1)\n",
    "            #     # p0 = basic_p(bs$t0[1], bs$t[, 1])\n",
    "            #     p0 = SCENT::basic_p(bs$t0[1], bs$t[, 1])\n",
    "            # }\n",
    "            # if (p0 < 0.001) {\n",
    "            #     bs = boot::boot(df2, assoc, R = 50000, formula = formula, \n",
    "            #       stype = \"i\", \n",
    "            #       # parallel = \"multicore\", ncpus = ncores)\n",
    "            #       parallel = \"no\", ncpus = 1)\n",
    "            #     # p0 = basic_p(bs$t0[1], bs$t[, 1])\n",
    "            #     p0 = SCENT::basic_p(bs$t0[1], bs$t[, 1])\n",
    "            # }\n",
    "            out <- data.frame(gene = gene, peak = this_peak, \n",
    "                beta = coefs[1], se = coefs[2], z = coefs[3], \n",
    "                p = coefs[4], boot_basic_p = p0)           \n",
    "            res <- rbind(res, out)\n",
    "        # }\n",
    "    }\n",
    "    object@SCENT.result <- res\n",
    "    return(object)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "563771b9-b96a-44d5-a5fe-fa30a7b9851b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(dir.output, \"chr\", sep = \"/\"), recursive = T):\n",
      "“'/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/SCENT/BMMC5.Myeloid.2500.250324//chr' already exists”\n"
     ]
    }
   ],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0d2266ce-b92d-40d2-9b78-f39bd33a5079",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2d41238d-e833-4e0e-add5-bb56bb615c1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr10\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr11\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr12\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr13\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr14\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr15\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr16\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr17\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr18\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr19\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr2\"\n",
      "[1] \"chr20\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr21\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr22\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr3\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr4\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr5\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr6\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr7\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr8\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr9\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chrX\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in split.default(x = seq_len(nrow(x)), f = f, drop = drop, ...):\n",
      "“data length is not a multiple of split variable”\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr = pairs.E2G[seqnames(pairs.E2G) == chr.tmp]\n",
    "  pairs.E2G.chr.res = pairs.E2G.chr[!is.na(pairs.E2G.chr$boot_basic_p) & pairs.E2G.chr$boot_basic_p < 0.05]\n",
    "  pairs.E2G.chr.keep = pairs.E2G.chr[!pairs.E2G.chr$PairName %in% pairs.E2G.chr.res$PairName]\n",
    "  gene_peak.chr = as.data.frame(mcols(pairs.E2G.chr.res)[,c(\"TargetGene\",\"PeakName\")])\n",
    "  list.gene_peak.chr <- split(gene_peak.chr, seq_len(n.cores))\n",
    "  \n",
    "  list.obj.SCENT.chr <- lapply(list.gene_peak.chr, function(gene_peak.tmp) {\n",
    "    obj.SCENT.tmp <- CreateSCENTObj(\n",
    "      rna = matrix.rna.rename[rownames(matrix.rna.rename) %in% gene_peak.tmp$TargetGene,],\n",
    "      atac = matrix.atac[rownames(matrix.atac) %in% gene_peak.tmp$PeakName,], \n",
    "      meta.data = meta.data,\n",
    "      peak.info = gene_peak.tmp,\n",
    "      covariates = c(\"log.nUMI\",\"percent.mito\"), \n",
    "      celltypes = \"celltype\"\n",
    "    )\n",
    "  })\n",
    "  \n",
    "  cl <- makeCluster(n.cores)\n",
    "  clusterExport(cl, varlist=c(\"SCENT_algorithm.modified\"))\n",
    "  list.SCENT.result.chr <- parLapply(cl, list.obj.SCENT.chr, function(obj.SCENT.tmp) {\n",
    "    obj.SCENT.tmp <- SCENT_algorithm.modified(object = obj.SCENT.tmp)\n",
    "    return(obj.SCENT.tmp@SCENT.result)\n",
    "  })\n",
    "  stopCluster(cl)\n",
    "  \n",
    "  \n",
    "  df.SCENT.result.chr = do.call(rbind,list.SCENT.result.chr)\n",
    "  rownames(df.SCENT.result.chr) = paste(df.SCENT.result.chr$peak,\n",
    "                                        df.SCENT.result.chr$gene,\n",
    "                                        sep = \"_\")\n",
    "\n",
    "  pairs.E2G.chr.res = pairs.E2G.chr.res[pairs.E2G.chr.res$PairName %in% rownames(df.SCENT.result.chr)]\n",
    "  mcols(pairs.E2G.chr.res)[,c(\"beta\",\"se\",\"z\",\"p\",\"boot_basic_p\")] = \n",
    "    df.SCENT.result.chr[pairs.E2G.chr.res$PairName,c(\"beta\",\"se\",\"z\",\"p\",\"boot_basic_p\")]\n",
    "  pairs.E2G.chr.res$boot_basic_p.log10 = -log10(pairs.E2G.chr.res$boot_basic_p)\n",
    "  pairs.E2G.chr.res$sign = 1\n",
    "  pairs.E2G.chr.res$sign[pairs.E2G.chr.res$beta < 0] = -1\n",
    "  pairs.E2G.chr.res$boot_basic_p.log10.signed = pairs.E2G.chr.res$boot_basic_p.log10 * pairs.E2G.chr.res$sign\n",
    "  pairs.E2G.chr.merge = c(pairs.E2G.chr.res,\n",
    "                          pairs.E2G.chr.keep)\n",
    "  saveRDS(pairs.E2G.chr.merge,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.merge)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00278bea-8baf-4ed0-93d5-1289f05fa36a",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 318489 ranges and 11 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "  chr1.chr1     chr1 100132371-100133496      * |     SLC30A7\n",
       "  chr1.chr1     chr1     1001860-1002126      * |    AURKAIP1\n",
       "  chr1.chr1     chr1     1001860-1002126      * |      CDK11B\n",
       "  chr1.chr1     chr1     1001860-1002126      * |        GNB1\n",
       "  chr1.chr1     chr1     1001860-1002126      * |      UBE2J2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  chrX.chrX     chrX       989778-990654      * |      CSF2RA\n",
       "  chrX.chrX     chrX       989778-990654      * |      GTPBP6\n",
       "  chrX.chrX     chrX       989778-990654      * |   LINC00106\n",
       "  chrX.chrX     chrX       989778-990654      * |       P2RY8\n",
       "  chrX.chrX     chrX     9980873-9982164      * |       TBL1X\n",
       "                          PeakName               PairName       beta        se\n",
       "                       <character>            <character>  <numeric> <numeric>\n",
       "  chr1.chr1 chr1-100132371-10013.. chr1-100132371-10013.. -0.0771667 0.0333876\n",
       "  chr1.chr1   chr1-1001860-1002126 chr1-1001860-1002126.. -0.2862840 0.1195895\n",
       "  chr1.chr1   chr1-1001860-1002126 chr1-1001860-1002126..  0.1687747 0.0655577\n",
       "  chr1.chr1   chr1-1001860-1002126 chr1-1001860-1002126..  0.0838553 0.0318446\n",
       "  chr1.chr1   chr1-1001860-1002126 chr1-1001860-1002126..  0.1885768 0.0915052\n",
       "        ...                    ...                    ...        ...       ...\n",
       "  chrX.chrX     chrX-989778-990654 chrX-989778-990654_C..  0.0442413 0.0414866\n",
       "  chrX.chrX     chrX-989778-990654 chrX-989778-990654_G.. -0.0995624 0.1204362\n",
       "  chrX.chrX     chrX-989778-990654 chrX-989778-990654_L..  0.1363353 0.1143497\n",
       "  chrX.chrX     chrX-989778-990654 chrX-989778-990654_P..  0.0860696 0.0446429\n",
       "  chrX.chrX   chrX-9980873-9982164 chrX-9980873-9982164..  0.0596522 0.0308013\n",
       "                    z          p boot_basic_p boot_basic_p.log10      sign\n",
       "            <numeric>  <numeric>    <numeric>          <numeric> <numeric>\n",
       "  chr1.chr1  -2.31124 0.02081975       0.0448            1.34872        -1\n",
       "  chr1.chr1  -2.39389 0.01667080       0.0304            1.51713        -1\n",
       "  chr1.chr1   2.57444 0.01004011       0.0184            1.73518         1\n",
       "  chr1.chr1   2.63327 0.00845671       0.0144            1.84164         1\n",
       "  chr1.chr1   2.06083 0.03931910       0.0312            1.50585         1\n",
       "        ...       ...        ...          ...                ...       ...\n",
       "  chrX.chrX  1.066402  0.2862421         0.36           0.443697         1\n",
       "  chrX.chrX -0.826681  0.4084177         0.44           0.356547        -1\n",
       "  chrX.chrX  1.192266  0.2331571         0.18           0.744727         1\n",
       "  chrX.chrX  1.927958  0.0538604         0.12           0.920819         1\n",
       "  chrX.chrX  1.936678  0.0527848         0.16           0.795880         1\n",
       "            boot_basic_p.log10.signed\n",
       "                            <numeric>\n",
       "  chr1.chr1                  -1.34872\n",
       "  chr1.chr1                  -1.51713\n",
       "  chr1.chr1                   1.73518\n",
       "  chr1.chr1                   1.84164\n",
       "  chr1.chr1                   1.50585\n",
       "        ...                       ...\n",
       "  chrX.chrX                  0.443697\n",
       "  chrX.chrX                 -0.356547\n",
       "  chrX.chrX                  0.744727\n",
       "  chrX.chrX                  0.920819\n",
       "  chrX.chrX                  0.795880\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 318489 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>beta</th><th scope=col>se</th><th scope=col>z</th><th scope=col>p</th><th scope=col>boot_basic_p</th><th scope=col>boot_basic_p.log10</th><th scope=col>sign</th><th scope=col>boot_basic_p.log10.signed</th><th scope=col>FDR</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100132371</td><td>100133496</td><td>SLC30A7 </td><td>Myeloid</td><td>-0.07716672</td><td>0.03338762</td><td>-2.311237</td><td>2.081975e-02</td><td>0.0448</td><td>1.348722</td><td>-1</td><td>-1.348722</td><td>0.22165738</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1001860</td><td>  1002126</td><td>AURKAIP1</td><td>Myeloid</td><td>-0.28628402</td><td>0.11958952</td><td>-2.393889</td><td>1.667080e-02</td><td>0.0304</td><td>1.517126</td><td>-1</td><td>-1.517126</td><td>0.16928168</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1001860</td><td>  1002126</td><td>CDK11B  </td><td>Myeloid</td><td> 0.16877471</td><td>0.06555771</td><td> 2.574445</td><td>1.004011e-02</td><td>0.0184</td><td>1.735182</td><td> 1</td><td> 1.735182</td><td>0.12222750</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1001860</td><td>  1002126</td><td>GNB1    </td><td>Myeloid</td><td> 0.08385533</td><td>0.03184456</td><td> 2.633270</td><td>8.456712e-03</td><td>0.0144</td><td>1.841638</td><td> 1</td><td> 1.841638</td><td>0.10410264</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1001860</td><td>  1002126</td><td>UBE2J2  </td><td>Myeloid</td><td> 0.18857677</td><td>0.09150518</td><td> 2.060832</td><td>3.931910e-02</td><td>0.0312</td><td>1.505845</td><td> 1</td><td> 1.505845</td><td>0.17211149</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10022225</td><td> 10023589</td><td>KIF1B   </td><td>Myeloid</td><td> 0.12723788</td><td>0.04304395</td><td> 2.955999</td><td>3.116576e-03</td><td>0.0096</td><td>2.017729</td><td> 1</td><td> 2.017729</td><td>0.07913589</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10022225</td><td> 10023589</td><td>SLC25A33</td><td>Myeloid</td><td> 0.28670652</td><td>0.06433340</td><td> 4.456573</td><td>8.328009e-06</td><td>0.0008</td><td>3.096910</td><td> 1</td><td> 3.096910</td><td>0.01213523</td></tr>\n",
       "\t<tr><td>chr1</td><td>100265575</td><td>100266826</td><td>CDC14A  </td><td>Myeloid</td><td> 0.11091695</td><td>0.02538695</td><td> 4.369054</td><td>1.247859e-05</td><td>0.0008</td><td>3.096910</td><td> 1</td><td> 3.096910</td><td>0.01213523</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10032414</td><td> 10033799</td><td>CTNNBIP1</td><td>Myeloid</td><td> 0.11434006</td><td>0.05150244</td><td> 2.220090</td><td>2.641265e-02</td><td>0.0448</td><td>1.348722</td><td> 1</td><td> 1.348722</td><td>0.22165738</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10032414</td><td> 10033799</td><td>KIF1B   </td><td>Myeloid</td><td> 0.06142459</td><td>0.02098237</td><td> 2.927439</td><td>3.417666e-03</td><td>0.0136</td><td>1.866461</td><td> 1</td><td> 1.866461</td><td>0.10022330</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10032414</td><td> 10033799</td><td>SLC25A33</td><td>Myeloid</td><td> 0.13295755</td><td>0.03363876</td><td> 3.952510</td><td>7.733554e-05</td><td>0.0008</td><td>3.096910</td><td> 1</td><td> 3.096910</td><td>0.01213523</td></tr>\n",
       "\t<tr><td>chr1</td><td>100359946</td><td>100360614</td><td>DBT     </td><td>Myeloid</td><td> 0.30450473</td><td>0.11614588</td><td> 2.621744</td><td>8.748119e-03</td><td>0.0072</td><td>2.142668</td><td> 1</td><td> 2.142668</td><td>0.06492966</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10055378</td><td> 10056588</td><td>KIF1B   </td><td>Myeloid</td><td> 0.11391135</td><td>0.03561935</td><td> 3.198019</td><td>1.383750e-03</td><td>0.0104</td><td>1.982967</td><td> 1</td><td> 1.982967</td><td>0.08354231</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10055378</td><td> 10056588</td><td>UBE4B   </td><td>Myeloid</td><td> 0.09166147</td><td>0.04553599</td><td> 2.012945</td><td>4.412037e-02</td><td>0.0552</td><td>1.258061</td><td> 1</td><td> 1.258061</td><td>0.25172377</td></tr>\n",
       "\t<tr><td>chr1</td><td>101002570</td><td>101003099</td><td>CDC14A  </td><td>Myeloid</td><td> 0.10621286</td><td>0.04361466</td><td> 2.435256</td><td>1.488126e-02</td><td>0.0544</td><td>1.264401</td><td> 1</td><td> 1.264401</td><td>0.25154333</td></tr>\n",
       "\t<tr><td>chr1</td><td>101111709</td><td>101112781</td><td>RTCA    </td><td>Myeloid</td><td> 0.15033471</td><td>0.07170388</td><td> 2.096605</td><td>3.602857e-02</td><td>0.0240</td><td>1.619789</td><td> 1</td><td> 1.619789</td><td>0.14540664</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1012920</td><td>  1014120</td><td>CDK11A  </td><td>Myeloid</td><td> 0.16007107</td><td>0.07943768</td><td> 2.015052</td><td>4.389919e-02</td><td>0.0440</td><td>1.356547</td><td> 1</td><td> 1.356547</td><td>0.21869807</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1012920</td><td>  1014120</td><td>CDK11B  </td><td>Myeloid</td><td> 0.09748166</td><td>0.04454643</td><td> 2.188316</td><td>2.864661e-02</td><td>0.0448</td><td>1.348722</td><td> 1</td><td> 1.348722</td><td>0.22165738</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1012920</td><td>  1014120</td><td>GNB1    </td><td>Myeloid</td><td> 0.07781772</td><td>0.02099667</td><td> 3.706193</td><td>2.103980e-04</td><td>0.0008</td><td>3.096910</td><td> 1</td><td> 3.096910</td><td>0.01213523</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1012920</td><td>  1014120</td><td>UBE2J2  </td><td>Myeloid</td><td> 0.15242525</td><td>0.06205776</td><td> 2.456184</td><td>1.404214e-02</td><td>0.0080</td><td>2.096910</td><td> 1</td><td> 2.096910</td><td>0.06973894</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10147738</td><td> 10148559</td><td>KIF1B   </td><td>Myeloid</td><td> 0.15370624</td><td>0.03673388</td><td> 4.184318</td><td>2.860229e-05</td><td>0.0008</td><td>3.096910</td><td> 1</td><td> 3.096910</td><td>0.01213523</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10147738</td><td> 10148559</td><td>SLC25A33</td><td>Myeloid</td><td> 0.14003480</td><td>0.05933841</td><td> 2.359935</td><td>1.827813e-02</td><td>0.0440</td><td>1.356547</td><td> 1</td><td> 1.356547</td><td>0.21869807</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10147738</td><td> 10148559</td><td>UBE4B   </td><td>Myeloid</td><td> 0.13261564</td><td>0.04710376</td><td> 2.815394</td><td>4.871749e-03</td><td>0.0056</td><td>2.251812</td><td> 1</td><td> 2.251812</td><td>0.05432317</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1019144</td><td>  1020371</td><td>NADK    </td><td>Myeloid</td><td>-0.10450517</td><td>0.03843754</td><td>-2.718831</td><td>6.551309e-03</td><td>0.0088</td><td>2.055517</td><td>-1</td><td>-2.055517</td><td>0.07454196</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1019144</td><td>  1020371</td><td>NOC2L   </td><td>Myeloid</td><td> 0.18708163</td><td>0.06767898</td><td> 2.764250</td><td>5.705374e-03</td><td>0.0040</td><td>2.397940</td><td> 1</td><td> 2.397940</td><td>0.04261150</td></tr>\n",
       "\t<tr><td>chr1</td><td>  1019144</td><td>  1020371</td><td>SDF4    </td><td>Myeloid</td><td>-0.19247865</td><td>0.07495379</td><td>-2.567964</td><td>1.022977e-02</td><td>0.0144</td><td>1.841638</td><td>-1</td><td>-1.841638</td><td>0.10410264</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10208605</td><td> 10208948</td><td>KIF1B   </td><td>Myeloid</td><td> 0.12156606</td><td>0.02931510</td><td> 4.146875</td><td>3.370442e-05</td><td>0.0008</td><td>3.096910</td><td> 1</td><td> 3.096910</td><td>0.01213523</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10208605</td><td> 10208948</td><td>SLC25A33</td><td>Myeloid</td><td> 0.12055483</td><td>0.04681708</td><td> 2.575018</td><td>1.002348e-02</td><td>0.0392</td><td>1.406714</td><td> 1</td><td> 1.406714</td><td>0.20118876</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10209811</td><td> 10211231</td><td>KIF1B   </td><td>Myeloid</td><td> 0.09177111</td><td>0.02090705</td><td> 4.389482</td><td>1.136212e-05</td><td>0.0008</td><td>3.096910</td><td> 1</td><td> 3.096910</td><td>0.01213523</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10209811</td><td> 10211231</td><td>SLC25A33</td><td>Myeloid</td><td> 0.18355943</td><td>0.03353770</td><td> 5.473227</td><td>4.419140e-08</td><td>0.0008</td><td>3.096910</td><td> 1</td><td> 3.096910</td><td>0.01213523</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>81201555</td><td>81202882</td><td>SH3BGRL  </td><td>Myeloid</td><td>-0.054485013</td><td>0.02888583</td><td>-1.88621917</td><td>0.05926543</td><td>0.240</td><td>0.619788758</td><td>-1</td><td>-0.619788758</td><td>0.5262288</td></tr>\n",
       "\t<tr><td>chrX</td><td>84501665</td><td>84502766</td><td>APOOL    </td><td>Myeloid</td><td> 0.024098304</td><td>0.08779844</td><td> 0.27447303</td><td>0.78372113</td><td>0.860</td><td>0.065501549</td><td> 1</td><td> 0.065501549</td><td>0.9412034</td></tr>\n",
       "\t<tr><td>chrX</td><td>84501665</td><td>84502766</td><td>HDX      </td><td>Myeloid</td><td>-0.065263847</td><td>0.07563744</td><td>-0.86285109</td><td>0.38821934</td><td>0.480</td><td>0.318758763</td><td>-1</td><td>-0.318758763</td><td>0.7389215</td></tr>\n",
       "\t<tr><td>chrX</td><td>85003051</td><td>85004561</td><td>APOOL    </td><td>Myeloid</td><td>-0.042399589</td><td>0.09415923</td><td>-0.45029669</td><td>0.65249653</td><td>0.760</td><td>0.119186408</td><td>-1</td><td>-0.119186408</td><td>0.8977078</td></tr>\n",
       "\t<tr><td>chrX</td><td>85003051</td><td>85004561</td><td>HDX      </td><td>Myeloid</td><td> 0.106646183</td><td>0.07443353</td><td> 1.43277067</td><td>0.15192338</td><td>0.300</td><td>0.522878745</td><td> 1</td><td> 0.522878745</td><td>0.5898527</td></tr>\n",
       "\t<tr><td>chrX</td><td>86047049</td><td>86048147</td><td>CHM      </td><td>Myeloid</td><td>-0.111582609</td><td>0.07106164</td><td>-1.57022272</td><td>0.11636331</td><td>0.200</td><td>0.698970004</td><td>-1</td><td>-0.698970004</td><td>0.4776559</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9315905</td><td> 9317571</td><td>WWC3     </td><td>Myeloid</td><td>-0.055094369</td><td>0.10400033</td><td>-0.52975187</td><td>0.59628398</td><td>0.420</td><td>0.376750710</td><td>-1</td><td>-0.376750710</td><td>0.6945892</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9340651</td><td> 9342271</td><td>TBL1X    </td><td>Myeloid</td><td> 0.007413409</td><td>0.02164750</td><td> 0.34246023</td><td>0.73200457</td><td>0.860</td><td>0.065501549</td><td> 1</td><td> 0.065501549</td><td>0.9412034</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9340651</td><td> 9342271</td><td>WWC3     </td><td>Myeloid</td><td>-0.004781536</td><td>0.08118011</td><td>-0.05890034</td><td>0.95303149</td><td>0.960</td><td>0.017728767</td><td>-1</td><td>-0.017728767</td><td>0.9797652</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9342463</td><td> 9343112</td><td>TBL1X    </td><td>Myeloid</td><td> 0.040075778</td><td>0.02844898</td><td> 1.40868936</td><td>0.15892704</td><td>0.300</td><td>0.522878745</td><td> 1</td><td> 0.522878745</td><td>0.5898527</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9342463</td><td> 9343112</td><td>WWC3     </td><td>Myeloid</td><td> 0.143612719</td><td>0.10218414</td><td> 1.40543061</td><td>0.15989327</td><td>0.100</td><td>1.000000000</td><td> 1</td><td> 1.000000000</td><td>0.3427414</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9352376</td><td> 9353538</td><td>TBL1X    </td><td>Myeloid</td><td> 0.048158526</td><td>0.02272511</td><td> 2.11917706</td><td>0.03407551</td><td>0.144</td><td>0.841637508</td><td> 1</td><td> 0.841637508</td><td>0.4082393</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9352376</td><td> 9353538</td><td>WWC3     </td><td>Myeloid</td><td>-0.001300558</td><td>0.08642824</td><td>-0.01504783</td><td>0.98799402</td><td>1.000</td><td>0.000000000</td><td>-1</td><td> 0.000000000</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9378667</td><td> 9379857</td><td>WWC3     </td><td>Myeloid</td><td> 0.014837751</td><td>0.10494230</td><td> 0.14138961</td><td>0.88756216</td><td>0.840</td><td>0.075720714</td><td> 1</td><td> 0.075720714</td><td>0.9330830</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9456233</td><td> 9457102</td><td>WWC3     </td><td>Myeloid</td><td>-0.018933833</td><td>0.10602909</td><td>-0.17857206</td><td>0.85827373</td><td>0.860</td><td>0.065501549</td><td>-1</td><td>-0.065501549</td><td>0.9412034</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9462835</td><td> 9463831</td><td>TBL1X    </td><td>Myeloid</td><td> 0.042352592</td><td>0.02004615</td><td> 2.11275443</td><td>0.03462179</td><td>0.160</td><td>0.795880017</td><td> 1</td><td> 0.795880017</td><td>0.4247865</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9462835</td><td> 9463831</td><td>WWC3     </td><td>Myeloid</td><td> 0.126254124</td><td>0.07291450</td><td> 1.73153663</td><td>0.08335610</td><td>0.140</td><td>0.853871964</td><td> 1</td><td> 0.853871964</td><td>0.3977348</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9464712</td><td> 9466516</td><td>TBL1X    </td><td>Myeloid</td><td> 0.046263749</td><td>0.02257962</td><td> 2.04891643</td><td>0.04047029</td><td>0.140</td><td>0.853871964</td><td> 1</td><td> 0.853871964</td><td>0.3977348</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9464712</td><td> 9466516</td><td>WWC3     </td><td>Myeloid</td><td> 0.014892608</td><td>0.08515457</td><td> 0.17488911</td><td>0.86116677</td><td>0.920</td><td>0.036212173</td><td> 1</td><td> 0.036212173</td><td>0.9648703</td></tr>\n",
       "\t<tr><td>chrX</td><td>96684334</td><td>96686300</td><td>DIAPH2   </td><td>Myeloid</td><td>-0.011097063</td><td>0.01185132</td><td>-0.93635654</td><td>0.34908965</td><td>0.460</td><td>0.337242168</td><td>-1</td><td>-0.337242168</td><td>0.7245761</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9676951</td><td> 9677902</td><td>TBL1X    </td><td>Myeloid</td><td> 0.038449505</td><td>0.03149465</td><td> 1.22082661</td><td>0.22215168</td><td>0.320</td><td>0.494850022</td><td> 1</td><td> 0.494850022</td><td>0.6091512</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9676951</td><td> 9677902</td><td>WWC3     </td><td>Myeloid</td><td> 0.084453461</td><td>0.11485791</td><td> 0.73528642</td><td>0.46216508</td><td>0.380</td><td>0.420216403</td><td> 1</td><td> 0.420216403</td><td>0.6624292</td></tr>\n",
       "\t<tr><td>chrX</td><td>97563508</td><td>97564917</td><td>DIAPH2   </td><td>Myeloid</td><td>-0.001867729</td><td>0.01688986</td><td>-0.11058285</td><td>0.91194715</td><td>0.980</td><td>0.008773924</td><td>-1</td><td>-0.008773924</td><td>0.9866762</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989778</td><td>  990654</td><td>AKAP17A  </td><td>Myeloid</td><td> 0.033821414</td><td>0.08631921</td><td> 0.39181793</td><td>0.69519275</td><td>0.480</td><td>0.318758763</td><td> 1</td><td> 0.318758763</td><td>0.7389215</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989778</td><td>  990654</td><td>ASMTL    </td><td>Myeloid</td><td> 0.121904220</td><td>0.08973006</td><td> 1.35856617</td><td>0.17428410</td><td>0.200</td><td>0.698970004</td><td> 1</td><td> 0.698970004</td><td>0.4776559</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989778</td><td>  990654</td><td>CSF2RA   </td><td>Myeloid</td><td> 0.044241339</td><td>0.04148656</td><td> 1.06640171</td><td>0.28624209</td><td>0.360</td><td>0.443697499</td><td> 1</td><td> 0.443697499</td><td>0.6453678</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989778</td><td>  990654</td><td>GTPBP6   </td><td>Myeloid</td><td>-0.099562375</td><td>0.12043621</td><td>-0.82668140</td><td>0.40841766</td><td>0.440</td><td>0.356547324</td><td>-1</td><td>-0.356547324</td><td>0.7101096</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989778</td><td>  990654</td><td>LINC00106</td><td>Myeloid</td><td> 0.136335255</td><td>0.11434973</td><td> 1.19226561</td><td>0.23315712</td><td>0.180</td><td>0.744727495</td><td> 1</td><td> 0.744727495</td><td>0.4516009</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989778</td><td>  990654</td><td>P2RY8    </td><td>Myeloid</td><td> 0.086069636</td><td>0.04464291</td><td> 1.92795770</td><td>0.05386039</td><td>0.120</td><td>0.920818754</td><td> 1</td><td> 0.920818754</td><td>0.3702284</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9980873</td><td> 9982164</td><td>TBL1X    </td><td>Myeloid</td><td> 0.059652190</td><td>0.03080130</td><td> 1.93667771</td><td>0.05278475</td><td>0.160</td><td>0.795880017</td><td> 1</td><td> 0.795880017</td><td>0.4247865</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 318489 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " chr & start & end & TargetGene & CellType & beta & se & z & p & boot\\_basic\\_p & boot\\_basic\\_p.log10 & sign & boot\\_basic\\_p.log10.signed & FDR\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100132371 & 100133496 & SLC30A7  & Myeloid & -0.07716672 & 0.03338762 & -2.311237 & 2.081975e-02 & 0.0448 & 1.348722 & -1 & -1.348722 & 0.22165738\\\\\n",
       "\t chr1 &   1001860 &   1002126 & AURKAIP1 & Myeloid & -0.28628402 & 0.11958952 & -2.393889 & 1.667080e-02 & 0.0304 & 1.517126 & -1 & -1.517126 & 0.16928168\\\\\n",
       "\t chr1 &   1001860 &   1002126 & CDK11B   & Myeloid &  0.16877471 & 0.06555771 &  2.574445 & 1.004011e-02 & 0.0184 & 1.735182 &  1 &  1.735182 & 0.12222750\\\\\n",
       "\t chr1 &   1001860 &   1002126 & GNB1     & Myeloid &  0.08385533 & 0.03184456 &  2.633270 & 8.456712e-03 & 0.0144 & 1.841638 &  1 &  1.841638 & 0.10410264\\\\\n",
       "\t chr1 &   1001860 &   1002126 & UBE2J2   & Myeloid &  0.18857677 & 0.09150518 &  2.060832 & 3.931910e-02 & 0.0312 & 1.505845 &  1 &  1.505845 & 0.17211149\\\\\n",
       "\t chr1 &  10022225 &  10023589 & KIF1B    & Myeloid &  0.12723788 & 0.04304395 &  2.955999 & 3.116576e-03 & 0.0096 & 2.017729 &  1 &  2.017729 & 0.07913589\\\\\n",
       "\t chr1 &  10022225 &  10023589 & SLC25A33 & Myeloid &  0.28670652 & 0.06433340 &  4.456573 & 8.328009e-06 & 0.0008 & 3.096910 &  1 &  3.096910 & 0.01213523\\\\\n",
       "\t chr1 & 100265575 & 100266826 & CDC14A   & Myeloid &  0.11091695 & 0.02538695 &  4.369054 & 1.247859e-05 & 0.0008 & 3.096910 &  1 &  3.096910 & 0.01213523\\\\\n",
       "\t chr1 &  10032414 &  10033799 & CTNNBIP1 & Myeloid &  0.11434006 & 0.05150244 &  2.220090 & 2.641265e-02 & 0.0448 & 1.348722 &  1 &  1.348722 & 0.22165738\\\\\n",
       "\t chr1 &  10032414 &  10033799 & KIF1B    & Myeloid &  0.06142459 & 0.02098237 &  2.927439 & 3.417666e-03 & 0.0136 & 1.866461 &  1 &  1.866461 & 0.10022330\\\\\n",
       "\t chr1 &  10032414 &  10033799 & SLC25A33 & Myeloid &  0.13295755 & 0.03363876 &  3.952510 & 7.733554e-05 & 0.0008 & 3.096910 &  1 &  3.096910 & 0.01213523\\\\\n",
       "\t chr1 & 100359946 & 100360614 & DBT      & Myeloid &  0.30450473 & 0.11614588 &  2.621744 & 8.748119e-03 & 0.0072 & 2.142668 &  1 &  2.142668 & 0.06492966\\\\\n",
       "\t chr1 &  10055378 &  10056588 & KIF1B    & Myeloid &  0.11391135 & 0.03561935 &  3.198019 & 1.383750e-03 & 0.0104 & 1.982967 &  1 &  1.982967 & 0.08354231\\\\\n",
       "\t chr1 &  10055378 &  10056588 & UBE4B    & Myeloid &  0.09166147 & 0.04553599 &  2.012945 & 4.412037e-02 & 0.0552 & 1.258061 &  1 &  1.258061 & 0.25172377\\\\\n",
       "\t chr1 & 101002570 & 101003099 & CDC14A   & Myeloid &  0.10621286 & 0.04361466 &  2.435256 & 1.488126e-02 & 0.0544 & 1.264401 &  1 &  1.264401 & 0.25154333\\\\\n",
       "\t chr1 & 101111709 & 101112781 & RTCA     & Myeloid &  0.15033471 & 0.07170388 &  2.096605 & 3.602857e-02 & 0.0240 & 1.619789 &  1 &  1.619789 & 0.14540664\\\\\n",
       "\t chr1 &   1012920 &   1014120 & CDK11A   & Myeloid &  0.16007107 & 0.07943768 &  2.015052 & 4.389919e-02 & 0.0440 & 1.356547 &  1 &  1.356547 & 0.21869807\\\\\n",
       "\t chr1 &   1012920 &   1014120 & CDK11B   & Myeloid &  0.09748166 & 0.04454643 &  2.188316 & 2.864661e-02 & 0.0448 & 1.348722 &  1 &  1.348722 & 0.22165738\\\\\n",
       "\t chr1 &   1012920 &   1014120 & GNB1     & Myeloid &  0.07781772 & 0.02099667 &  3.706193 & 2.103980e-04 & 0.0008 & 3.096910 &  1 &  3.096910 & 0.01213523\\\\\n",
       "\t chr1 &   1012920 &   1014120 & UBE2J2   & Myeloid &  0.15242525 & 0.06205776 &  2.456184 & 1.404214e-02 & 0.0080 & 2.096910 &  1 &  2.096910 & 0.06973894\\\\\n",
       "\t chr1 &  10147738 &  10148559 & KIF1B    & Myeloid &  0.15370624 & 0.03673388 &  4.184318 & 2.860229e-05 & 0.0008 & 3.096910 &  1 &  3.096910 & 0.01213523\\\\\n",
       "\t chr1 &  10147738 &  10148559 & SLC25A33 & Myeloid &  0.14003480 & 0.05933841 &  2.359935 & 1.827813e-02 & 0.0440 & 1.356547 &  1 &  1.356547 & 0.21869807\\\\\n",
       "\t chr1 &  10147738 &  10148559 & UBE4B    & Myeloid &  0.13261564 & 0.04710376 &  2.815394 & 4.871749e-03 & 0.0056 & 2.251812 &  1 &  2.251812 & 0.05432317\\\\\n",
       "\t chr1 &   1019144 &   1020371 & NADK     & Myeloid & -0.10450517 & 0.03843754 & -2.718831 & 6.551309e-03 & 0.0088 & 2.055517 & -1 & -2.055517 & 0.07454196\\\\\n",
       "\t chr1 &   1019144 &   1020371 & NOC2L    & Myeloid &  0.18708163 & 0.06767898 &  2.764250 & 5.705374e-03 & 0.0040 & 2.397940 &  1 &  2.397940 & 0.04261150\\\\\n",
       "\t chr1 &   1019144 &   1020371 & SDF4     & Myeloid & -0.19247865 & 0.07495379 & -2.567964 & 1.022977e-02 & 0.0144 & 1.841638 & -1 & -1.841638 & 0.10410264\\\\\n",
       "\t chr1 &  10208605 &  10208948 & KIF1B    & Myeloid &  0.12156606 & 0.02931510 &  4.146875 & 3.370442e-05 & 0.0008 & 3.096910 &  1 &  3.096910 & 0.01213523\\\\\n",
       "\t chr1 &  10208605 &  10208948 & SLC25A33 & Myeloid &  0.12055483 & 0.04681708 &  2.575018 & 1.002348e-02 & 0.0392 & 1.406714 &  1 &  1.406714 & 0.20118876\\\\\n",
       "\t chr1 &  10209811 &  10211231 & KIF1B    & Myeloid &  0.09177111 & 0.02090705 &  4.389482 & 1.136212e-05 & 0.0008 & 3.096910 &  1 &  3.096910 & 0.01213523\\\\\n",
       "\t chr1 &  10209811 &  10211231 & SLC25A33 & Myeloid &  0.18355943 & 0.03353770 &  5.473227 & 4.419140e-08 & 0.0008 & 3.096910 &  1 &  3.096910 & 0.01213523\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 81201555 & 81202882 & SH3BGRL   & Myeloid & -0.054485013 & 0.02888583 & -1.88621917 & 0.05926543 & 0.240 & 0.619788758 & -1 & -0.619788758 & 0.5262288\\\\\n",
       "\t chrX & 84501665 & 84502766 & APOOL     & Myeloid &  0.024098304 & 0.08779844 &  0.27447303 & 0.78372113 & 0.860 & 0.065501549 &  1 &  0.065501549 & 0.9412034\\\\\n",
       "\t chrX & 84501665 & 84502766 & HDX       & Myeloid & -0.065263847 & 0.07563744 & -0.86285109 & 0.38821934 & 0.480 & 0.318758763 & -1 & -0.318758763 & 0.7389215\\\\\n",
       "\t chrX & 85003051 & 85004561 & APOOL     & Myeloid & -0.042399589 & 0.09415923 & -0.45029669 & 0.65249653 & 0.760 & 0.119186408 & -1 & -0.119186408 & 0.8977078\\\\\n",
       "\t chrX & 85003051 & 85004561 & HDX       & Myeloid &  0.106646183 & 0.07443353 &  1.43277067 & 0.15192338 & 0.300 & 0.522878745 &  1 &  0.522878745 & 0.5898527\\\\\n",
       "\t chrX & 86047049 & 86048147 & CHM       & Myeloid & -0.111582609 & 0.07106164 & -1.57022272 & 0.11636331 & 0.200 & 0.698970004 & -1 & -0.698970004 & 0.4776559\\\\\n",
       "\t chrX &  9315905 &  9317571 & WWC3      & Myeloid & -0.055094369 & 0.10400033 & -0.52975187 & 0.59628398 & 0.420 & 0.376750710 & -1 & -0.376750710 & 0.6945892\\\\\n",
       "\t chrX &  9340651 &  9342271 & TBL1X     & Myeloid &  0.007413409 & 0.02164750 &  0.34246023 & 0.73200457 & 0.860 & 0.065501549 &  1 &  0.065501549 & 0.9412034\\\\\n",
       "\t chrX &  9340651 &  9342271 & WWC3      & Myeloid & -0.004781536 & 0.08118011 & -0.05890034 & 0.95303149 & 0.960 & 0.017728767 & -1 & -0.017728767 & 0.9797652\\\\\n",
       "\t chrX &  9342463 &  9343112 & TBL1X     & Myeloid &  0.040075778 & 0.02844898 &  1.40868936 & 0.15892704 & 0.300 & 0.522878745 &  1 &  0.522878745 & 0.5898527\\\\\n",
       "\t chrX &  9342463 &  9343112 & WWC3      & Myeloid &  0.143612719 & 0.10218414 &  1.40543061 & 0.15989327 & 0.100 & 1.000000000 &  1 &  1.000000000 & 0.3427414\\\\\n",
       "\t chrX &  9352376 &  9353538 & TBL1X     & Myeloid &  0.048158526 & 0.02272511 &  2.11917706 & 0.03407551 & 0.144 & 0.841637508 &  1 &  0.841637508 & 0.4082393\\\\\n",
       "\t chrX &  9352376 &  9353538 & WWC3      & Myeloid & -0.001300558 & 0.08642824 & -0.01504783 & 0.98799402 & 1.000 & 0.000000000 & -1 &  0.000000000 & 1.0000000\\\\\n",
       "\t chrX &  9378667 &  9379857 & WWC3      & Myeloid &  0.014837751 & 0.10494230 &  0.14138961 & 0.88756216 & 0.840 & 0.075720714 &  1 &  0.075720714 & 0.9330830\\\\\n",
       "\t chrX &  9456233 &  9457102 & WWC3      & Myeloid & -0.018933833 & 0.10602909 & -0.17857206 & 0.85827373 & 0.860 & 0.065501549 & -1 & -0.065501549 & 0.9412034\\\\\n",
       "\t chrX &  9462835 &  9463831 & TBL1X     & Myeloid &  0.042352592 & 0.02004615 &  2.11275443 & 0.03462179 & 0.160 & 0.795880017 &  1 &  0.795880017 & 0.4247865\\\\\n",
       "\t chrX &  9462835 &  9463831 & WWC3      & Myeloid &  0.126254124 & 0.07291450 &  1.73153663 & 0.08335610 & 0.140 & 0.853871964 &  1 &  0.853871964 & 0.3977348\\\\\n",
       "\t chrX &  9464712 &  9466516 & TBL1X     & Myeloid &  0.046263749 & 0.02257962 &  2.04891643 & 0.04047029 & 0.140 & 0.853871964 &  1 &  0.853871964 & 0.3977348\\\\\n",
       "\t chrX &  9464712 &  9466516 & WWC3      & Myeloid &  0.014892608 & 0.08515457 &  0.17488911 & 0.86116677 & 0.920 & 0.036212173 &  1 &  0.036212173 & 0.9648703\\\\\n",
       "\t chrX & 96684334 & 96686300 & DIAPH2    & Myeloid & -0.011097063 & 0.01185132 & -0.93635654 & 0.34908965 & 0.460 & 0.337242168 & -1 & -0.337242168 & 0.7245761\\\\\n",
       "\t chrX &  9676951 &  9677902 & TBL1X     & Myeloid &  0.038449505 & 0.03149465 &  1.22082661 & 0.22215168 & 0.320 & 0.494850022 &  1 &  0.494850022 & 0.6091512\\\\\n",
       "\t chrX &  9676951 &  9677902 & WWC3      & Myeloid &  0.084453461 & 0.11485791 &  0.73528642 & 0.46216508 & 0.380 & 0.420216403 &  1 &  0.420216403 & 0.6624292\\\\\n",
       "\t chrX & 97563508 & 97564917 & DIAPH2    & Myeloid & -0.001867729 & 0.01688986 & -0.11058285 & 0.91194715 & 0.980 & 0.008773924 & -1 & -0.008773924 & 0.9866762\\\\\n",
       "\t chrX &   989778 &   990654 & AKAP17A   & Myeloid &  0.033821414 & 0.08631921 &  0.39181793 & 0.69519275 & 0.480 & 0.318758763 &  1 &  0.318758763 & 0.7389215\\\\\n",
       "\t chrX &   989778 &   990654 & ASMTL     & Myeloid &  0.121904220 & 0.08973006 &  1.35856617 & 0.17428410 & 0.200 & 0.698970004 &  1 &  0.698970004 & 0.4776559\\\\\n",
       "\t chrX &   989778 &   990654 & CSF2RA    & Myeloid &  0.044241339 & 0.04148656 &  1.06640171 & 0.28624209 & 0.360 & 0.443697499 &  1 &  0.443697499 & 0.6453678\\\\\n",
       "\t chrX &   989778 &   990654 & GTPBP6    & Myeloid & -0.099562375 & 0.12043621 & -0.82668140 & 0.40841766 & 0.440 & 0.356547324 & -1 & -0.356547324 & 0.7101096\\\\\n",
       "\t chrX &   989778 &   990654 & LINC00106 & Myeloid &  0.136335255 & 0.11434973 &  1.19226561 & 0.23315712 & 0.180 & 0.744727495 &  1 &  0.744727495 & 0.4516009\\\\\n",
       "\t chrX &   989778 &   990654 & P2RY8     & Myeloid &  0.086069636 & 0.04464291 &  1.92795770 & 0.05386039 & 0.120 & 0.920818754 &  1 &  0.920818754 & 0.3702284\\\\\n",
       "\t chrX &  9980873 &  9982164 & TBL1X     & Myeloid &  0.059652190 & 0.03080130 &  1.93667771 & 0.05278475 & 0.160 & 0.795880017 &  1 &  0.795880017 & 0.4247865\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 318489 × 14\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | beta &lt;dbl&gt; | se &lt;dbl&gt; | z &lt;dbl&gt; | p &lt;dbl&gt; | boot_basic_p &lt;dbl&gt; | boot_basic_p.log10 &lt;dbl&gt; | sign &lt;dbl&gt; | boot_basic_p.log10.signed &lt;dbl&gt; | FDR &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100132371 | 100133496 | SLC30A7  | Myeloid | -0.07716672 | 0.03338762 | -2.311237 | 2.081975e-02 | 0.0448 | 1.348722 | -1 | -1.348722 | 0.22165738 |\n",
       "| chr1 |   1001860 |   1002126 | AURKAIP1 | Myeloid | -0.28628402 | 0.11958952 | -2.393889 | 1.667080e-02 | 0.0304 | 1.517126 | -1 | -1.517126 | 0.16928168 |\n",
       "| chr1 |   1001860 |   1002126 | CDK11B   | Myeloid |  0.16877471 | 0.06555771 |  2.574445 | 1.004011e-02 | 0.0184 | 1.735182 |  1 |  1.735182 | 0.12222750 |\n",
       "| chr1 |   1001860 |   1002126 | GNB1     | Myeloid |  0.08385533 | 0.03184456 |  2.633270 | 8.456712e-03 | 0.0144 | 1.841638 |  1 |  1.841638 | 0.10410264 |\n",
       "| chr1 |   1001860 |   1002126 | UBE2J2   | Myeloid |  0.18857677 | 0.09150518 |  2.060832 | 3.931910e-02 | 0.0312 | 1.505845 |  1 |  1.505845 | 0.17211149 |\n",
       "| chr1 |  10022225 |  10023589 | KIF1B    | Myeloid |  0.12723788 | 0.04304395 |  2.955999 | 3.116576e-03 | 0.0096 | 2.017729 |  1 |  2.017729 | 0.07913589 |\n",
       "| chr1 |  10022225 |  10023589 | SLC25A33 | Myeloid |  0.28670652 | 0.06433340 |  4.456573 | 8.328009e-06 | 0.0008 | 3.096910 |  1 |  3.096910 | 0.01213523 |\n",
       "| chr1 | 100265575 | 100266826 | CDC14A   | Myeloid |  0.11091695 | 0.02538695 |  4.369054 | 1.247859e-05 | 0.0008 | 3.096910 |  1 |  3.096910 | 0.01213523 |\n",
       "| chr1 |  10032414 |  10033799 | CTNNBIP1 | Myeloid |  0.11434006 | 0.05150244 |  2.220090 | 2.641265e-02 | 0.0448 | 1.348722 |  1 |  1.348722 | 0.22165738 |\n",
       "| chr1 |  10032414 |  10033799 | KIF1B    | Myeloid |  0.06142459 | 0.02098237 |  2.927439 | 3.417666e-03 | 0.0136 | 1.866461 |  1 |  1.866461 | 0.10022330 |\n",
       "| chr1 |  10032414 |  10033799 | SLC25A33 | Myeloid |  0.13295755 | 0.03363876 |  3.952510 | 7.733554e-05 | 0.0008 | 3.096910 |  1 |  3.096910 | 0.01213523 |\n",
       "| chr1 | 100359946 | 100360614 | DBT      | Myeloid |  0.30450473 | 0.11614588 |  2.621744 | 8.748119e-03 | 0.0072 | 2.142668 |  1 |  2.142668 | 0.06492966 |\n",
       "| chr1 |  10055378 |  10056588 | KIF1B    | Myeloid |  0.11391135 | 0.03561935 |  3.198019 | 1.383750e-03 | 0.0104 | 1.982967 |  1 |  1.982967 | 0.08354231 |\n",
       "| chr1 |  10055378 |  10056588 | UBE4B    | Myeloid |  0.09166147 | 0.04553599 |  2.012945 | 4.412037e-02 | 0.0552 | 1.258061 |  1 |  1.258061 | 0.25172377 |\n",
       "| chr1 | 101002570 | 101003099 | CDC14A   | Myeloid |  0.10621286 | 0.04361466 |  2.435256 | 1.488126e-02 | 0.0544 | 1.264401 |  1 |  1.264401 | 0.25154333 |\n",
       "| chr1 | 101111709 | 101112781 | RTCA     | Myeloid |  0.15033471 | 0.07170388 |  2.096605 | 3.602857e-02 | 0.0240 | 1.619789 |  1 |  1.619789 | 0.14540664 |\n",
       "| chr1 |   1012920 |   1014120 | CDK11A   | Myeloid |  0.16007107 | 0.07943768 |  2.015052 | 4.389919e-02 | 0.0440 | 1.356547 |  1 |  1.356547 | 0.21869807 |\n",
       "| chr1 |   1012920 |   1014120 | CDK11B   | Myeloid |  0.09748166 | 0.04454643 |  2.188316 | 2.864661e-02 | 0.0448 | 1.348722 |  1 |  1.348722 | 0.22165738 |\n",
       "| chr1 |   1012920 |   1014120 | GNB1     | Myeloid |  0.07781772 | 0.02099667 |  3.706193 | 2.103980e-04 | 0.0008 | 3.096910 |  1 |  3.096910 | 0.01213523 |\n",
       "| chr1 |   1012920 |   1014120 | UBE2J2   | Myeloid |  0.15242525 | 0.06205776 |  2.456184 | 1.404214e-02 | 0.0080 | 2.096910 |  1 |  2.096910 | 0.06973894 |\n",
       "| chr1 |  10147738 |  10148559 | KIF1B    | Myeloid |  0.15370624 | 0.03673388 |  4.184318 | 2.860229e-05 | 0.0008 | 3.096910 |  1 |  3.096910 | 0.01213523 |\n",
       "| chr1 |  10147738 |  10148559 | SLC25A33 | Myeloid |  0.14003480 | 0.05933841 |  2.359935 | 1.827813e-02 | 0.0440 | 1.356547 |  1 |  1.356547 | 0.21869807 |\n",
       "| chr1 |  10147738 |  10148559 | UBE4B    | Myeloid |  0.13261564 | 0.04710376 |  2.815394 | 4.871749e-03 | 0.0056 | 2.251812 |  1 |  2.251812 | 0.05432317 |\n",
       "| chr1 |   1019144 |   1020371 | NADK     | Myeloid | -0.10450517 | 0.03843754 | -2.718831 | 6.551309e-03 | 0.0088 | 2.055517 | -1 | -2.055517 | 0.07454196 |\n",
       "| chr1 |   1019144 |   1020371 | NOC2L    | Myeloid |  0.18708163 | 0.06767898 |  2.764250 | 5.705374e-03 | 0.0040 | 2.397940 |  1 |  2.397940 | 0.04261150 |\n",
       "| chr1 |   1019144 |   1020371 | SDF4     | Myeloid | -0.19247865 | 0.07495379 | -2.567964 | 1.022977e-02 | 0.0144 | 1.841638 | -1 | -1.841638 | 0.10410264 |\n",
       "| chr1 |  10208605 |  10208948 | KIF1B    | Myeloid |  0.12156606 | 0.02931510 |  4.146875 | 3.370442e-05 | 0.0008 | 3.096910 |  1 |  3.096910 | 0.01213523 |\n",
       "| chr1 |  10208605 |  10208948 | SLC25A33 | Myeloid |  0.12055483 | 0.04681708 |  2.575018 | 1.002348e-02 | 0.0392 | 1.406714 |  1 |  1.406714 | 0.20118876 |\n",
       "| chr1 |  10209811 |  10211231 | KIF1B    | Myeloid |  0.09177111 | 0.02090705 |  4.389482 | 1.136212e-05 | 0.0008 | 3.096910 |  1 |  3.096910 | 0.01213523 |\n",
       "| chr1 |  10209811 |  10211231 | SLC25A33 | Myeloid |  0.18355943 | 0.03353770 |  5.473227 | 4.419140e-08 | 0.0008 | 3.096910 |  1 |  3.096910 | 0.01213523 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 81201555 | 81202882 | SH3BGRL   | Myeloid | -0.054485013 | 0.02888583 | -1.88621917 | 0.05926543 | 0.240 | 0.619788758 | -1 | -0.619788758 | 0.5262288 |\n",
       "| chrX | 84501665 | 84502766 | APOOL     | Myeloid |  0.024098304 | 0.08779844 |  0.27447303 | 0.78372113 | 0.860 | 0.065501549 |  1 |  0.065501549 | 0.9412034 |\n",
       "| chrX | 84501665 | 84502766 | HDX       | Myeloid | -0.065263847 | 0.07563744 | -0.86285109 | 0.38821934 | 0.480 | 0.318758763 | -1 | -0.318758763 | 0.7389215 |\n",
       "| chrX | 85003051 | 85004561 | APOOL     | Myeloid | -0.042399589 | 0.09415923 | -0.45029669 | 0.65249653 | 0.760 | 0.119186408 | -1 | -0.119186408 | 0.8977078 |\n",
       "| chrX | 85003051 | 85004561 | HDX       | Myeloid |  0.106646183 | 0.07443353 |  1.43277067 | 0.15192338 | 0.300 | 0.522878745 |  1 |  0.522878745 | 0.5898527 |\n",
       "| chrX | 86047049 | 86048147 | CHM       | Myeloid | -0.111582609 | 0.07106164 | -1.57022272 | 0.11636331 | 0.200 | 0.698970004 | -1 | -0.698970004 | 0.4776559 |\n",
       "| chrX |  9315905 |  9317571 | WWC3      | Myeloid | -0.055094369 | 0.10400033 | -0.52975187 | 0.59628398 | 0.420 | 0.376750710 | -1 | -0.376750710 | 0.6945892 |\n",
       "| chrX |  9340651 |  9342271 | TBL1X     | Myeloid |  0.007413409 | 0.02164750 |  0.34246023 | 0.73200457 | 0.860 | 0.065501549 |  1 |  0.065501549 | 0.9412034 |\n",
       "| chrX |  9340651 |  9342271 | WWC3      | Myeloid | -0.004781536 | 0.08118011 | -0.05890034 | 0.95303149 | 0.960 | 0.017728767 | -1 | -0.017728767 | 0.9797652 |\n",
       "| chrX |  9342463 |  9343112 | TBL1X     | Myeloid |  0.040075778 | 0.02844898 |  1.40868936 | 0.15892704 | 0.300 | 0.522878745 |  1 |  0.522878745 | 0.5898527 |\n",
       "| chrX |  9342463 |  9343112 | WWC3      | Myeloid |  0.143612719 | 0.10218414 |  1.40543061 | 0.15989327 | 0.100 | 1.000000000 |  1 |  1.000000000 | 0.3427414 |\n",
       "| chrX |  9352376 |  9353538 | TBL1X     | Myeloid |  0.048158526 | 0.02272511 |  2.11917706 | 0.03407551 | 0.144 | 0.841637508 |  1 |  0.841637508 | 0.4082393 |\n",
       "| chrX |  9352376 |  9353538 | WWC3      | Myeloid | -0.001300558 | 0.08642824 | -0.01504783 | 0.98799402 | 1.000 | 0.000000000 | -1 |  0.000000000 | 1.0000000 |\n",
       "| chrX |  9378667 |  9379857 | WWC3      | Myeloid |  0.014837751 | 0.10494230 |  0.14138961 | 0.88756216 | 0.840 | 0.075720714 |  1 |  0.075720714 | 0.9330830 |\n",
       "| chrX |  9456233 |  9457102 | WWC3      | Myeloid | -0.018933833 | 0.10602909 | -0.17857206 | 0.85827373 | 0.860 | 0.065501549 | -1 | -0.065501549 | 0.9412034 |\n",
       "| chrX |  9462835 |  9463831 | TBL1X     | Myeloid |  0.042352592 | 0.02004615 |  2.11275443 | 0.03462179 | 0.160 | 0.795880017 |  1 |  0.795880017 | 0.4247865 |\n",
       "| chrX |  9462835 |  9463831 | WWC3      | Myeloid |  0.126254124 | 0.07291450 |  1.73153663 | 0.08335610 | 0.140 | 0.853871964 |  1 |  0.853871964 | 0.3977348 |\n",
       "| chrX |  9464712 |  9466516 | TBL1X     | Myeloid |  0.046263749 | 0.02257962 |  2.04891643 | 0.04047029 | 0.140 | 0.853871964 |  1 |  0.853871964 | 0.3977348 |\n",
       "| chrX |  9464712 |  9466516 | WWC3      | Myeloid |  0.014892608 | 0.08515457 |  0.17488911 | 0.86116677 | 0.920 | 0.036212173 |  1 |  0.036212173 | 0.9648703 |\n",
       "| chrX | 96684334 | 96686300 | DIAPH2    | Myeloid | -0.011097063 | 0.01185132 | -0.93635654 | 0.34908965 | 0.460 | 0.337242168 | -1 | -0.337242168 | 0.7245761 |\n",
       "| chrX |  9676951 |  9677902 | TBL1X     | Myeloid |  0.038449505 | 0.03149465 |  1.22082661 | 0.22215168 | 0.320 | 0.494850022 |  1 |  0.494850022 | 0.6091512 |\n",
       "| chrX |  9676951 |  9677902 | WWC3      | Myeloid |  0.084453461 | 0.11485791 |  0.73528642 | 0.46216508 | 0.380 | 0.420216403 |  1 |  0.420216403 | 0.6624292 |\n",
       "| chrX | 97563508 | 97564917 | DIAPH2    | Myeloid | -0.001867729 | 0.01688986 | -0.11058285 | 0.91194715 | 0.980 | 0.008773924 | -1 | -0.008773924 | 0.9866762 |\n",
       "| chrX |   989778 |   990654 | AKAP17A   | Myeloid |  0.033821414 | 0.08631921 |  0.39181793 | 0.69519275 | 0.480 | 0.318758763 |  1 |  0.318758763 | 0.7389215 |\n",
       "| chrX |   989778 |   990654 | ASMTL     | Myeloid |  0.121904220 | 0.08973006 |  1.35856617 | 0.17428410 | 0.200 | 0.698970004 |  1 |  0.698970004 | 0.4776559 |\n",
       "| chrX |   989778 |   990654 | CSF2RA    | Myeloid |  0.044241339 | 0.04148656 |  1.06640171 | 0.28624209 | 0.360 | 0.443697499 |  1 |  0.443697499 | 0.6453678 |\n",
       "| chrX |   989778 |   990654 | GTPBP6    | Myeloid | -0.099562375 | 0.12043621 | -0.82668140 | 0.40841766 | 0.440 | 0.356547324 | -1 | -0.356547324 | 0.7101096 |\n",
       "| chrX |   989778 |   990654 | LINC00106 | Myeloid |  0.136335255 | 0.11434973 |  1.19226561 | 0.23315712 | 0.180 | 0.744727495 |  1 |  0.744727495 | 0.4516009 |\n",
       "| chrX |   989778 |   990654 | P2RY8     | Myeloid |  0.086069636 | 0.04464291 |  1.92795770 | 0.05386039 | 0.120 | 0.920818754 |  1 |  0.920818754 | 0.3702284 |\n",
       "| chrX |  9980873 |  9982164 | TBL1X     | Myeloid |  0.059652190 | 0.03080130 |  1.93667771 | 0.05278475 | 0.160 | 0.795880017 |  1 |  0.795880017 | 0.4247865 |\n",
       "\n"
      ],
      "text/plain": [
       "       chr  start     end       TargetGene CellType beta         se        \n",
       "1      chr1 100132371 100133496 SLC30A7    Myeloid  -0.07716672  0.03338762\n",
       "2      chr1   1001860   1002126 AURKAIP1   Myeloid  -0.28628402  0.11958952\n",
       "3      chr1   1001860   1002126 CDK11B     Myeloid   0.16877471  0.06555771\n",
       "4      chr1   1001860   1002126 GNB1       Myeloid   0.08385533  0.03184456\n",
       "5      chr1   1001860   1002126 UBE2J2     Myeloid   0.18857677  0.09150518\n",
       "6      chr1  10022225  10023589 KIF1B      Myeloid   0.12723788  0.04304395\n",
       "7      chr1  10022225  10023589 SLC25A33   Myeloid   0.28670652  0.06433340\n",
       "8      chr1 100265575 100266826 CDC14A     Myeloid   0.11091695  0.02538695\n",
       "9      chr1  10032414  10033799 CTNNBIP1   Myeloid   0.11434006  0.05150244\n",
       "10     chr1  10032414  10033799 KIF1B      Myeloid   0.06142459  0.02098237\n",
       "11     chr1  10032414  10033799 SLC25A33   Myeloid   0.13295755  0.03363876\n",
       "12     chr1 100359946 100360614 DBT        Myeloid   0.30450473  0.11614588\n",
       "13     chr1  10055378  10056588 KIF1B      Myeloid   0.11391135  0.03561935\n",
       "14     chr1  10055378  10056588 UBE4B      Myeloid   0.09166147  0.04553599\n",
       "15     chr1 101002570 101003099 CDC14A     Myeloid   0.10621286  0.04361466\n",
       "16     chr1 101111709 101112781 RTCA       Myeloid   0.15033471  0.07170388\n",
       "17     chr1   1012920   1014120 CDK11A     Myeloid   0.16007107  0.07943768\n",
       "18     chr1   1012920   1014120 CDK11B     Myeloid   0.09748166  0.04454643\n",
       "19     chr1   1012920   1014120 GNB1       Myeloid   0.07781772  0.02099667\n",
       "20     chr1   1012920   1014120 UBE2J2     Myeloid   0.15242525  0.06205776\n",
       "21     chr1  10147738  10148559 KIF1B      Myeloid   0.15370624  0.03673388\n",
       "22     chr1  10147738  10148559 SLC25A33   Myeloid   0.14003480  0.05933841\n",
       "23     chr1  10147738  10148559 UBE4B      Myeloid   0.13261564  0.04710376\n",
       "24     chr1   1019144   1020371 NADK       Myeloid  -0.10450517  0.03843754\n",
       "25     chr1   1019144   1020371 NOC2L      Myeloid   0.18708163  0.06767898\n",
       "26     chr1   1019144   1020371 SDF4       Myeloid  -0.19247865  0.07495379\n",
       "27     chr1  10208605  10208948 KIF1B      Myeloid   0.12156606  0.02931510\n",
       "28     chr1  10208605  10208948 SLC25A33   Myeloid   0.12055483  0.04681708\n",
       "29     chr1  10209811  10211231 KIF1B      Myeloid   0.09177111  0.02090705\n",
       "30     chr1  10209811  10211231 SLC25A33   Myeloid   0.18355943  0.03353770\n",
       "⋮      ⋮    ⋮         ⋮         ⋮          ⋮        ⋮            ⋮         \n",
       "318460 chrX 81201555  81202882  SH3BGRL    Myeloid  -0.054485013 0.02888583\n",
       "318461 chrX 84501665  84502766  APOOL      Myeloid   0.024098304 0.08779844\n",
       "318462 chrX 84501665  84502766  HDX        Myeloid  -0.065263847 0.07563744\n",
       "318463 chrX 85003051  85004561  APOOL      Myeloid  -0.042399589 0.09415923\n",
       "318464 chrX 85003051  85004561  HDX        Myeloid   0.106646183 0.07443353\n",
       "318465 chrX 86047049  86048147  CHM        Myeloid  -0.111582609 0.07106164\n",
       "318466 chrX  9315905   9317571  WWC3       Myeloid  -0.055094369 0.10400033\n",
       "318467 chrX  9340651   9342271  TBL1X      Myeloid   0.007413409 0.02164750\n",
       "318468 chrX  9340651   9342271  WWC3       Myeloid  -0.004781536 0.08118011\n",
       "318469 chrX  9342463   9343112  TBL1X      Myeloid   0.040075778 0.02844898\n",
       "318470 chrX  9342463   9343112  WWC3       Myeloid   0.143612719 0.10218414\n",
       "318471 chrX  9352376   9353538  TBL1X      Myeloid   0.048158526 0.02272511\n",
       "318472 chrX  9352376   9353538  WWC3       Myeloid  -0.001300558 0.08642824\n",
       "318473 chrX  9378667   9379857  WWC3       Myeloid   0.014837751 0.10494230\n",
       "318474 chrX  9456233   9457102  WWC3       Myeloid  -0.018933833 0.10602909\n",
       "318475 chrX  9462835   9463831  TBL1X      Myeloid   0.042352592 0.02004615\n",
       "318476 chrX  9462835   9463831  WWC3       Myeloid   0.126254124 0.07291450\n",
       "318477 chrX  9464712   9466516  TBL1X      Myeloid   0.046263749 0.02257962\n",
       "318478 chrX  9464712   9466516  WWC3       Myeloid   0.014892608 0.08515457\n",
       "318479 chrX 96684334  96686300  DIAPH2     Myeloid  -0.011097063 0.01185132\n",
       "318480 chrX  9676951   9677902  TBL1X      Myeloid   0.038449505 0.03149465\n",
       "318481 chrX  9676951   9677902  WWC3       Myeloid   0.084453461 0.11485791\n",
       "318482 chrX 97563508  97564917  DIAPH2     Myeloid  -0.001867729 0.01688986\n",
       "318483 chrX   989778    990654  AKAP17A    Myeloid   0.033821414 0.08631921\n",
       "318484 chrX   989778    990654  ASMTL      Myeloid   0.121904220 0.08973006\n",
       "318485 chrX   989778    990654  CSF2RA     Myeloid   0.044241339 0.04148656\n",
       "318486 chrX   989778    990654  GTPBP6     Myeloid  -0.099562375 0.12043621\n",
       "318487 chrX   989778    990654  LINC00106  Myeloid   0.136335255 0.11434973\n",
       "318488 chrX   989778    990654  P2RY8      Myeloid   0.086069636 0.04464291\n",
       "318489 chrX  9980873   9982164  TBL1X      Myeloid   0.059652190 0.03080130\n",
       "       z           p            boot_basic_p boot_basic_p.log10 sign\n",
       "1      -2.311237   2.081975e-02 0.0448       1.348722           -1  \n",
       "2      -2.393889   1.667080e-02 0.0304       1.517126           -1  \n",
       "3       2.574445   1.004011e-02 0.0184       1.735182            1  \n",
       "4       2.633270   8.456712e-03 0.0144       1.841638            1  \n",
       "5       2.060832   3.931910e-02 0.0312       1.505845            1  \n",
       "6       2.955999   3.116576e-03 0.0096       2.017729            1  \n",
       "7       4.456573   8.328009e-06 0.0008       3.096910            1  \n",
       "8       4.369054   1.247859e-05 0.0008       3.096910            1  \n",
       "9       2.220090   2.641265e-02 0.0448       1.348722            1  \n",
       "10      2.927439   3.417666e-03 0.0136       1.866461            1  \n",
       "11      3.952510   7.733554e-05 0.0008       3.096910            1  \n",
       "12      2.621744   8.748119e-03 0.0072       2.142668            1  \n",
       "13      3.198019   1.383750e-03 0.0104       1.982967            1  \n",
       "14      2.012945   4.412037e-02 0.0552       1.258061            1  \n",
       "15      2.435256   1.488126e-02 0.0544       1.264401            1  \n",
       "16      2.096605   3.602857e-02 0.0240       1.619789            1  \n",
       "17      2.015052   4.389919e-02 0.0440       1.356547            1  \n",
       "18      2.188316   2.864661e-02 0.0448       1.348722            1  \n",
       "19      3.706193   2.103980e-04 0.0008       3.096910            1  \n",
       "20      2.456184   1.404214e-02 0.0080       2.096910            1  \n",
       "21      4.184318   2.860229e-05 0.0008       3.096910            1  \n",
       "22      2.359935   1.827813e-02 0.0440       1.356547            1  \n",
       "23      2.815394   4.871749e-03 0.0056       2.251812            1  \n",
       "24     -2.718831   6.551309e-03 0.0088       2.055517           -1  \n",
       "25      2.764250   5.705374e-03 0.0040       2.397940            1  \n",
       "26     -2.567964   1.022977e-02 0.0144       1.841638           -1  \n",
       "27      4.146875   3.370442e-05 0.0008       3.096910            1  \n",
       "28      2.575018   1.002348e-02 0.0392       1.406714            1  \n",
       "29      4.389482   1.136212e-05 0.0008       3.096910            1  \n",
       "30      5.473227   4.419140e-08 0.0008       3.096910            1  \n",
       "⋮      ⋮           ⋮            ⋮            ⋮                  ⋮   \n",
       "318460 -1.88621917 0.05926543   0.240        0.619788758        -1  \n",
       "318461  0.27447303 0.78372113   0.860        0.065501549         1  \n",
       "318462 -0.86285109 0.38821934   0.480        0.318758763        -1  \n",
       "318463 -0.45029669 0.65249653   0.760        0.119186408        -1  \n",
       "318464  1.43277067 0.15192338   0.300        0.522878745         1  \n",
       "318465 -1.57022272 0.11636331   0.200        0.698970004        -1  \n",
       "318466 -0.52975187 0.59628398   0.420        0.376750710        -1  \n",
       "318467  0.34246023 0.73200457   0.860        0.065501549         1  \n",
       "318468 -0.05890034 0.95303149   0.960        0.017728767        -1  \n",
       "318469  1.40868936 0.15892704   0.300        0.522878745         1  \n",
       "318470  1.40543061 0.15989327   0.100        1.000000000         1  \n",
       "318471  2.11917706 0.03407551   0.144        0.841637508         1  \n",
       "318472 -0.01504783 0.98799402   1.000        0.000000000        -1  \n",
       "318473  0.14138961 0.88756216   0.840        0.075720714         1  \n",
       "318474 -0.17857206 0.85827373   0.860        0.065501549        -1  \n",
       "318475  2.11275443 0.03462179   0.160        0.795880017         1  \n",
       "318476  1.73153663 0.08335610   0.140        0.853871964         1  \n",
       "318477  2.04891643 0.04047029   0.140        0.853871964         1  \n",
       "318478  0.17488911 0.86116677   0.920        0.036212173         1  \n",
       "318479 -0.93635654 0.34908965   0.460        0.337242168        -1  \n",
       "318480  1.22082661 0.22215168   0.320        0.494850022         1  \n",
       "318481  0.73528642 0.46216508   0.380        0.420216403         1  \n",
       "318482 -0.11058285 0.91194715   0.980        0.008773924        -1  \n",
       "318483  0.39181793 0.69519275   0.480        0.318758763         1  \n",
       "318484  1.35856617 0.17428410   0.200        0.698970004         1  \n",
       "318485  1.06640171 0.28624209   0.360        0.443697499         1  \n",
       "318486 -0.82668140 0.40841766   0.440        0.356547324        -1  \n",
       "318487  1.19226561 0.23315712   0.180        0.744727495         1  \n",
       "318488  1.92795770 0.05386039   0.120        0.920818754         1  \n",
       "318489  1.93667771 0.05278475   0.160        0.795880017         1  \n",
       "       boot_basic_p.log10.signed FDR       \n",
       "1      -1.348722                 0.22165738\n",
       "2      -1.517126                 0.16928168\n",
       "3       1.735182                 0.12222750\n",
       "4       1.841638                 0.10410264\n",
       "5       1.505845                 0.17211149\n",
       "6       2.017729                 0.07913589\n",
       "7       3.096910                 0.01213523\n",
       "8       3.096910                 0.01213523\n",
       "9       1.348722                 0.22165738\n",
       "10      1.866461                 0.10022330\n",
       "11      3.096910                 0.01213523\n",
       "12      2.142668                 0.06492966\n",
       "13      1.982967                 0.08354231\n",
       "14      1.258061                 0.25172377\n",
       "15      1.264401                 0.25154333\n",
       "16      1.619789                 0.14540664\n",
       "17      1.356547                 0.21869807\n",
       "18      1.348722                 0.22165738\n",
       "19      3.096910                 0.01213523\n",
       "20      2.096910                 0.06973894\n",
       "21      3.096910                 0.01213523\n",
       "22      1.356547                 0.21869807\n",
       "23      2.251812                 0.05432317\n",
       "24     -2.055517                 0.07454196\n",
       "25      2.397940                 0.04261150\n",
       "26     -1.841638                 0.10410264\n",
       "27      3.096910                 0.01213523\n",
       "28      1.406714                 0.20118876\n",
       "29      3.096910                 0.01213523\n",
       "30      3.096910                 0.01213523\n",
       "⋮      ⋮                         ⋮         \n",
       "318460 -0.619788758              0.5262288 \n",
       "318461  0.065501549              0.9412034 \n",
       "318462 -0.318758763              0.7389215 \n",
       "318463 -0.119186408              0.8977078 \n",
       "318464  0.522878745              0.5898527 \n",
       "318465 -0.698970004              0.4776559 \n",
       "318466 -0.376750710              0.6945892 \n",
       "318467  0.065501549              0.9412034 \n",
       "318468 -0.017728767              0.9797652 \n",
       "318469  0.522878745              0.5898527 \n",
       "318470  1.000000000              0.3427414 \n",
       "318471  0.841637508              0.4082393 \n",
       "318472  0.000000000              1.0000000 \n",
       "318473  0.075720714              0.9330830 \n",
       "318474 -0.065501549              0.9412034 \n",
       "318475  0.795880017              0.4247865 \n",
       "318476  0.853871964              0.3977348 \n",
       "318477  0.853871964              0.3977348 \n",
       "318478  0.036212173              0.9648703 \n",
       "318479 -0.337242168              0.7245761 \n",
       "318480  0.494850022              0.6091512 \n",
       "318481  0.420216403              0.6624292 \n",
       "318482 -0.008773924              0.9866762 \n",
       "318483  0.318758763              0.7389215 \n",
       "318484  0.698970004              0.4776559 \n",
       "318485  0.443697499              0.6453678 \n",
       "318486 -0.356547324              0.7101096 \n",
       "318487  0.744727495              0.4516009 \n",
       "318488  0.920818754              0.3702284 \n",
       "318489  0.795880017              0.4247865 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = celltype\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"beta\",\"se\",\"z\",\"p\",\"boot_basic_p\",\n",
    "                         \"boot_basic_p.log10\",\n",
    "                         \"sign\",\n",
    "                         \"boot_basic_p.log10.signed\")]\n",
    "df.output[,\"FDR\"] = p.adjust(df.output$boot_basic_p, method = \"BH\")\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] dplyr_1.1.4          data.table_1.15.2    Matrix_1.6-5        \n",
       " [4] GenomicRanges_1.54.1 GenomeInfoDb_1.38.1  IRanges_2.36.0      \n",
       " [7] S4Vectors_0.40.2     BiocGenerics_0.48.1  genomation_1.34.0   \n",
       "[10] Signac_1.13.0        SCENT_1.0.0         \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] rstudioapi_0.16.0           jsonlite_1.8.8             \n",
       "  [3] magrittr_2.0.3              nloptr_2.0.3               \n",
       "  [5] rmarkdown_2.26              BiocIO_1.10.0              \n",
       "  [7] zlibbioc_1.48.0             vctrs_0.6.5                \n",
       "  [9] minqa_1.2.6                 Rsamtools_2.18.0           \n",
       " [11] RCurl_1.98-1.14             base64enc_0.1-3            \n",
       " [13] RcppRoll_0.3.0              htmltools_0.5.8.1          \n",
       " [15] S4Arrays_1.2.1              plotrix_3.8-4              \n",
       " [17] Formula_1.2-5               parallelly_1.37.1          \n",
       " [19] KernSmooth_2.23-22          htmlwidgets_1.6.4          \n",
       " [21] plyr_1.8.9                  impute_1.76.0              \n",
       " [23] uuid_1.2-0                  GenomicAlignments_1.36.0   \n",
       " [25] lifecycle_1.0.4             pkgconfig_2.0.3            \n",
       " [27] R6_2.5.1                    fastmap_1.1.1              \n",
       " [29] GenomeInfoDbData_1.2.11     MatrixGenerics_1.12.3      \n",
       " [31] future_1.33.2               digest_0.6.35              \n",
       " [33] colorspace_2.1-0            patchwork_1.2.0            \n",
       " [35] irlba_2.3.5.1               Hmisc_5.1-2                \n",
       " [37] progressr_0.14.0            fansi_1.0.6                \n",
       " [39] abind_1.4-5                 compiler_4.3.3             \n",
       " [41] withr_3.0.0                 htmlTable_2.4.2            \n",
       " [43] backports_1.4.1             BiocParallel_1.36.0        \n",
       " [45] R.utils_2.12.3              MASS_7.3-60                \n",
       " [47] DelayedArray_0.26.7         rjson_0.2.21               \n",
       " [49] tools_4.3.3                 foreign_0.8-86             \n",
       " [51] future.apply_1.11.2         nnet_7.3-19                \n",
       " [53] R.oo_1.26.0                 glue_1.7.0                 \n",
       " [55] restfulr_0.0.15             nlme_3.1-164               \n",
       " [57] pbdZMQ_0.3-11               checkmate_2.3.1            \n",
       " [59] gridBase_0.4-7              cluster_2.1.6              \n",
       " [61] reshape2_1.4.4              generics_0.1.3             \n",
       " [63] gtable_0.3.5                BSgenome_1.68.0            \n",
       " [65] tzdb_0.4.0                  R.methodsS3_1.8.2          \n",
       " [67] tidyr_1.3.1                 seqPattern_1.34.0          \n",
       " [69] hms_1.1.3                   sp_2.1-4                   \n",
       " [71] utf8_1.2.4                  XVector_0.42.0             \n",
       " [73] pillar_1.9.0                stringr_1.5.1              \n",
       " [75] spam_2.10-0                 IRdisplay_1.1              \n",
       " [77] splines_4.3.3               lattice_0.22-6             \n",
       " [79] rtracklayer_1.60.1          tidyselect_1.2.1           \n",
       " [81] Biostrings_2.70.1           pbapply_1.7-2              \n",
       " [83] knitr_1.46                  gridExtra_2.3              \n",
       " [85] SummarizedExperiment_1.30.2 RhpcBLASctl_0.23-42        \n",
       " [87] xfun_0.43                   Biobase_2.60.0             \n",
       " [89] matrixStats_1.3.0           stringi_1.8.3              \n",
       " [91] yaml_2.3.8                  boot_1.3-30                \n",
       " [93] evaluate_0.23               codetools_0.2-20           \n",
       " [95] tibble_3.2.1                cli_3.6.2                  \n",
       " [97] IRkernel_1.3.2              rpart_4.1.23               \n",
       " [99] repr_1.1.7                  munsell_0.5.1              \n",
       "[101] Rcpp_1.0.12                 globals_0.16.3             \n",
       "[103] XML_3.99-0.16.1             ggplot2_3.5.1              \n",
       "[105] readr_2.1.5                 dotCall64_1.1-1            \n",
       "[107] bitops_1.0-7                lme4_1.1-35.3              \n",
       "[109] listenv_0.9.1               scales_1.3.0               \n",
       "[111] SeuratObject_5.0.1          purrr_1.0.2                \n",
       "[113] crayon_1.5.2                rlang_1.1.3                \n",
       "[115] fastmatch_1.1-4            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
