{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8466b35-cae5-4f62-8138-37b2de484a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.2\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : grid v4.3.3\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "\tLoading Package : gtools v3.9.5\n",
      "\n",
      "\tLoading Package : gtable v0.3.5\n",
      "\n",
      "\tLoading Package : ggplot2 v3.5.1\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "\tLoading Package : plyr v1.8.9\n",
      "\n",
      "\tLoading Package : stringr v1.5.1\n",
      "\n",
      "\tLoading Package : data.table v1.15.2\n",
      "\n",
      "\tLoading Package : matrixStats v1.3.0\n",
      "\n",
      "\tLoading Package : S4Vectors v0.40.2\n",
      "\n",
      "\tLoading Package : GenomicRanges v1.54.1\n",
      "\n",
      "\tLoading Package : BiocGenerics v0.48.1\n",
      "\n",
      "\tLoading Package : Matrix v1.6.5\n",
      "\n",
      "\tLoading Package : Rcpp v1.0.12\n",
      "\n",
      "\tLoading Package : SummarizedExperiment v1.30.2\n",
      "\n",
      "\tLoading Package : rhdf5 v2.44.0\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(magrittr)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e85020d7-d358-492d-9358-c81270f4b624",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1bec6690-52ec-450d-bba7-7d4528a185cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.seurat = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Wang/1.prepare_data/1.seurat_pipeline.240507/obj.seurat.qc.rds\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Wang/K562/Kendall/Pairs.tsv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Wang/3.Genome_wide_prediction/ArchR_250226/ArchR.250226/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "340dac32-44ac-4231-b964-b5f3b9e2dbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16\n",
    "maxDist = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da973385-4c81-4b1c-9669-cf5933b23a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 16.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRThreads(threads = n.cores) \n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeb26496-d3b9-4778-92ed-9c63a5018aef",
   "metadata": {},
   "source": [
    "Import seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c9b19a09-be95-4d9a-bdc3-0530980d4757",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.seurat = readRDS(path.seurat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86be3e66-c954-4620-9e64-c6a1c0fcb84c",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6f363969-bf0b-45ce-858c-a8cb64cc210a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3538ffdc-4fed-48eb-9c9d-86805a91cf06",
   "metadata": {},
   "source": [
    "Create an ArchRProject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0300f485-fafa-4cef-b9d0-534108ec6f95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 1\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_230720/geo/cellranger_res/GSE1\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"ed3e8fdfffa0e7cea75003307869d5ce\" \"124adcbdbfec73546f348dde2eb879e6\"\n",
      "  .. ..@ cells: Named chr [1:4574] \"AAACAGCCAATGCCCG-1\" \"AAACAGCCACCCTCAC-1\" \"AAACAGCCACCTACTT-1\" \"AAACATGCAACTAACT-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:4574] \"K562_Wang_AAACAGCCAATGCCCG-1\" \"K562_Wang_AAACAGCCACCCTCAC-1\" \"K562_Wang_AAACAGCCACCTACTT-1\" \"K562_Wang_AAACATGCAACTAACT-1\" ...\n"
     ]
    }
   ],
   "source": [
    "str(Fragments(obj.seurat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1753ffaa-89e2-43bf-bcf6-19937782d63a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-createArrows-2a52567c171baa-Date-2025-02-26_Time-14-48-49.211254.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Cleaning Temporary Files\n",
      "\n",
      "2025-02-26 14:48:49.697407 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "(K562_Wang : 1 of 1) Determining Arrow Method to use!\n",
      "\n",
      "2025-02-26 14:48:49.764073 : (K562_Wang : 1 of 1) Reading In Fragments from inputFiles (readMethod = tabix), 0.001 mins elapsed.\n",
      "\n",
      "2025-02-26 14:48:49.774087 : (K562_Wang : 1 of 1) Tabix Bed To Temporary File, 0.001 mins elapsed.\n",
      "\n",
      "2025-02-26 14:50:03.289682 : (K562_Wang : 1 of 1) Successful creation of Temporary File, 1.227 mins elapsed.\n",
      "\n",
      "2025-02-26 14:50:03.298966 : (K562_Wang : 1 of 1) Creating ArrowFile From Temporary File, 1.227 mins elapsed.\n",
      "\n",
      "2025-02-26 14:50:33.374936 : (K562_Wang : 1 of 1) Successful creation of Arrow File, 1.728 mins elapsed.\n",
      "\n",
      "2025-02-26 14:50:54.766449 : Continuing through after error ggplot for Fragment Size Distribution, 2.084 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-02-26 14:51:26.674739 : (K562_Wang : 1 of 1) CellStats : Number of Cells Pass Filter = 4551 , 2.616 mins elapsed.\n",
      "\n",
      "2025-02-26 14:51:26.681231 : (K562_Wang : 1 of 1) CellStats : Median Frags = 15951 , 2.616 mins elapsed.\n",
      "\n",
      "2025-02-26 14:51:26.689396 : (K562_Wang : 1 of 1) CellStats : Median TSS Enrichment = 13.899 , 2.617 mins elapsed.\n",
      "\n",
      "2025-02-26 14:51:28.127735 : Continuing through after error ggplot for TSS by Frags, 2.641 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-02-26 14:51:31.364013 : (K562_Wang : 1 of 1) Adding Additional Feature Counts!, 2.694 mins elapsed.\n",
      "\n",
      "2025-02-26 14:51:56.585064 : (K562_Wang : 1 of 1) Removing Fragments from Filtered Cells, 3.115 mins elapsed.\n",
      "\n",
      "2025-02-26 14:51:56.599586 : (K562_Wang : 1 of 1) Creating Filtered Arrow File, 3.115 mins elapsed.\n",
      "\n",
      "2025-02-26 14:53:10.621516 : (K562_Wang : 1 of 1) Finished Constructing Filtered Arrow File!, 4.349 mins elapsed.\n",
      "\n",
      "2025-02-26 14:53:10.841158 : (K562_Wang : 1 of 1) Adding TileMatrix!, 4.352 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:31.64037 : (K562_Wang : 1 of 1) Finished Creating Arrow File, 8.699 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-createArrows-2a52567c171baa-Date-2025-02-26_Time-14-48-49.211254.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ArrowFiles <- createArrowFiles(\n",
    "  inputFiles = c(Fragments(obj.seurat)[[1]]@path),\n",
    "  sampleNames = c(\"K562_Wang\"),\n",
    "  validBarcodes = list(K562_Wang = unname(Fragments(obj.seurat)[[1]]@cells)),\n",
    "  addTileMat = T,\n",
    "  addGeneScoreMat = F,\n",
    "  force = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ada3db45-d050-4659-a7cf-2891e04b1249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'K562_Wang.arrow'"
      ],
      "text/latex": [
       "'K562\\_Wang.arrow'"
      ],
      "text/markdown": [
       "'K562_Wang.arrow'"
      ],
      "text/plain": [
       "[1] \"K562_Wang.arrow\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ArrowFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5ba28283-d488-4982-8d4c-745fc3f59365",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"ArchR\",sep=\"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aa217507-8323-43bf-b23c-0ec25ee33c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Validating Arrows...\n",
      "\n",
      "Getting SampleNames...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Copying ArrowFiles to Ouptut Directory! If you want to save disk space set copyArrows = FALSE\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Getting Cell Metadata...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Merging Cell Metadata...\n",
      "\n",
      "Initializing ArchRProject...\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR <- ArchRProject(\n",
    "  ArrowFiles = ArrowFiles, \n",
    "  outputDirectory = paste(dir.output,\"ArchR\",sep=\"/\"),\n",
    "  copyArrows = TRUE \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f9795403-545f-4385-aa38-2c70b1b2fb21",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: ArchRProject \n",
       "outputDirectory: /maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Wang/3.Genome_wide_prediction/ArchR_250226/ArchR.250226/ArchR \n",
       "samples(1): K562_Wang\n",
       "sampleColData names(1): ArrowFiles\n",
       "cellColData names(13): Sample TSSEnrichment ... nDiFrags BlacklistRatio\n",
       "numberOfCells(1): 4551\n",
       "medianTSS(1): 13.899\n",
       "medianFrags(1): 15951"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obj.ArchR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ee876dd5-0b22-46fd-9015-7e22144c5024",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking Inputs...\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addIterativeLSI-2a52565921cd7e-Date-2025-02-26_Time-14-57-40.372031.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 14:57:41.014359 : Computing Total Across All Features, 0 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:42.121436 : Computing Top Features, 0.019 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-26 14:57:45.072613 : Running LSI (1 of 2) on Top Features, 0.068 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-02-26 14:57:45.086149 : Creating Partial Matrix, 0.068 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:59.726903 : Computing LSI, 0.312 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:38.496402 : Identifying Clusters, 0.958 mins elapsed.\n",
      "\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "2025-02-26 14:58:43.473487 : Identified 6 Clusters, 1.041 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:43.489659 : Saving LSI Iteration, 1.042 mins elapsed.\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************************************************************\n",
      "2025-02-26 14:58:58.568446 : ERROR Found in .saveIteration for  \n",
      "LogFile = ArchRLogs/ArchR-addIterativeLSI-2a52565921cd7e-Date-2025-02-26_Time-14-57-40.372031.log\n",
      "\n",
      "<simpleError in g$grobs[[legend]]: no such index at level 2\n",
      ">\n",
      "\n",
      "************************************************************\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-26 14:58:58.570436 : Creating Cluster Matrix on the total Group Features, 1.293 mins elapsed.\n",
      "\n",
      "2025-02-26 14:59:16.951706 : Computing Variable Features, 1.599 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-26 14:59:17.136908 : Running LSI (2 of 2) on Variable Features, 1.602 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-02-26 14:59:17.219321 : Creating Partial Matrix, 1.604 mins elapsed.\n",
      "\n",
      "2025-02-26 14:59:32.772078 : Computing LSI, 1.863 mins elapsed.\n",
      "\n",
      "2025-02-26 15:00:09.969709 : Finished Running IterativeLSI, 2.483 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addIterativeLSI(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    useMatrix = \"TileMatrix\", \n",
    "    name = \"IterativeLSI\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2c1d334b-7fd7-425a-bb8d-7be4cbedd3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.ArchR = addPeakSet(obj.ArchR,\n",
    "                       peakSet = reduce(pairs.E2G))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e25c14db-eb4e-4b94-8862-eb34246a365c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeakMatrix-2a5256127da5d6-Date-2025-02-26_Time-15-00-17.811932.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 15:00:18.044651 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-02-26 15:00:18.788761 : Adding K562_Wang to PeakMatrix for Chr (1 of 23)!, 0.012 mins elapsed.\n",
      "\n",
      "2025-02-26 15:00:30.942185 : Adding K562_Wang to PeakMatrix for Chr (2 of 23)!, 0.214 mins elapsed.\n",
      "\n",
      "2025-02-26 15:00:38.731667 : Adding K562_Wang to PeakMatrix for Chr (3 of 23)!, 0.344 mins elapsed.\n",
      "\n",
      "2025-02-26 15:00:46.256022 : Adding K562_Wang to PeakMatrix for Chr (4 of 23)!, 0.469 mins elapsed.\n",
      "\n",
      "2025-02-26 15:00:53.816314 : Adding K562_Wang to PeakMatrix for Chr (5 of 23)!, 0.595 mins elapsed.\n",
      "\n",
      "2025-02-26 15:01:00.992153 : Adding K562_Wang to PeakMatrix for Chr (6 of 23)!, 0.715 mins elapsed.\n",
      "\n",
      "2025-02-26 15:01:08.770417 : Adding K562_Wang to PeakMatrix for Chr (7 of 23)!, 0.845 mins elapsed.\n",
      "\n",
      "2025-02-26 15:01:15.93715 : Adding K562_Wang to PeakMatrix for Chr (8 of 23)!, 0.964 mins elapsed.\n",
      "\n",
      "2025-02-26 15:01:21.919728 : Adding K562_Wang to PeakMatrix for Chr (9 of 23)!, 1.064 mins elapsed.\n",
      "\n",
      "2025-02-26 15:01:28.498098 : Adding K562_Wang to PeakMatrix for Chr (10 of 23)!, 1.173 mins elapsed.\n",
      "\n",
      "2025-02-26 15:01:35.095934 : Adding K562_Wang to PeakMatrix for Chr (11 of 23)!, 1.283 mins elapsed.\n",
      "\n",
      "2025-02-26 15:01:44.644716 : Adding K562_Wang to PeakMatrix for Chr (12 of 23)!, 1.443 mins elapsed.\n",
      "\n",
      "2025-02-26 15:01:52.000415 : Adding K562_Wang to PeakMatrix for Chr (13 of 23)!, 1.565 mins elapsed.\n",
      "\n",
      "2025-02-26 15:01:58.637954 : Adding K562_Wang to PeakMatrix for Chr (14 of 23)!, 1.676 mins elapsed.\n",
      "\n",
      "2025-02-26 15:02:04.880145 : Adding K562_Wang to PeakMatrix for Chr (15 of 23)!, 1.78 mins elapsed.\n",
      "\n",
      "2025-02-26 15:02:11.570413 : Adding K562_Wang to PeakMatrix for Chr (16 of 23)!, 1.891 mins elapsed.\n",
      "\n",
      "2025-02-26 15:02:18.28038 : Adding K562_Wang to PeakMatrix for Chr (17 of 23)!, 2.003 mins elapsed.\n",
      "\n",
      "2025-02-26 15:02:25.518059 : Adding K562_Wang to PeakMatrix for Chr (18 of 23)!, 2.124 mins elapsed.\n",
      "\n",
      "2025-02-26 15:02:30.773291 : Adding K562_Wang to PeakMatrix for Chr (19 of 23)!, 2.211 mins elapsed.\n",
      "\n",
      "2025-02-26 15:02:38.109254 : Adding K562_Wang to PeakMatrix for Chr (20 of 23)!, 2.334 mins elapsed.\n",
      "\n",
      "2025-02-26 15:02:44.343987 : Adding K562_Wang to PeakMatrix for Chr (21 of 23)!, 2.438 mins elapsed.\n",
      "\n",
      "2025-02-26 15:02:53.994424 : Adding K562_Wang to PeakMatrix for Chr (22 of 23)!, 2.598 mins elapsed.\n",
      "\n",
      "2025-02-26 15:03:01.377013 : Adding K562_Wang to PeakMatrix for Chr (23 of 23)!, 2.721 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeakMatrix-2a5256127da5d6-Date-2025-02-26_Time-15-00-17.811932.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addPeakMatrix(obj.ArchR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3b913333-8823-435e-a677-42fa6d8777ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seRNA = import10xFeatureMatrix(c(\"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_230720/geo/cellranger_res/GSE162798_01/outs/raw_feature_bc_matrix.h5\"),\n",
    "                               c(\"K562_Wang\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "895234f3-e668-4e83-adf0-562024a6711b",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA = seRNA[,rownames(obj.ArchR@cellColData)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4656d98c-a9cc-486a-afaa-fe59ca6ef055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 4551 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(4551): K562_Wang#CCTATATTCAATGAGG-1\n",
       "  K562_Wang#TGTGGCTCATTTAAGC-1 ... K562_Wang#AGGATATAGTGAACGA-1\n",
       "  K562_Wang#GACTAATGTGAACAAA-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fe1e71d-e9b5-45c8-bdb0-e7665205e1ce",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "70dfbb6c-643d-47e2-9cb4-1d917697c3e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1882f38e-98df-4393-96cb-6444954dfe60",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biostrings’:\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seRNA.rename = map_gene_names(seRNA,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "65cb2097-9075-4ee4-9e54-dd288dc50736",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addGeneExpressionMatrix-2a52565acb9b06-Date-2025-02-26_Time-15-03-49.799199.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Overlap w/ scATAC = 1\n",
      "\n",
      "2025-02-26 15:03:50.32429 : \n",
      "\n",
      "Overlap Per Sample w/ scATAC : K562_Wang=4551\n",
      "\n",
      "2025-02-26 15:03:50.336121 : \n",
      "\n",
      "2025-02-26 15:03:52.88409 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-02-26 15:03:53.694549 : Adding K562_Wang to GeneExpressionMatrix for Chr (1 of 23)!, 0.014 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:00.017695 : Adding K562_Wang to GeneExpressionMatrix for Chr (2 of 23)!, 0.119 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:05.048598 : Adding K562_Wang to GeneExpressionMatrix for Chr (3 of 23)!, 0.203 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:12.569874 : Adding K562_Wang to GeneExpressionMatrix for Chr (4 of 23)!, 0.328 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:19.773825 : Adding K562_Wang to GeneExpressionMatrix for Chr (5 of 23)!, 0.448 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:25.220606 : Adding K562_Wang to GeneExpressionMatrix for Chr (6 of 23)!, 0.539 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:32.296193 : Adding K562_Wang to GeneExpressionMatrix for Chr (7 of 23)!, 0.657 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:37.361938 : Adding K562_Wang to GeneExpressionMatrix for Chr (8 of 23)!, 0.741 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:42.34223 : Adding K562_Wang to GeneExpressionMatrix for Chr (9 of 23)!, 0.824 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:49.965047 : Adding K562_Wang to GeneExpressionMatrix for Chr (10 of 23)!, 0.951 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:54.856775 : Adding K562_Wang to GeneExpressionMatrix for Chr (11 of 23)!, 1.033 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:59.811157 : Adding K562_Wang to GeneExpressionMatrix for Chr (12 of 23)!, 1.115 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:07.039864 : Adding K562_Wang to GeneExpressionMatrix for Chr (13 of 23)!, 1.236 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:13.308235 : Adding K562_Wang to GeneExpressionMatrix for Chr (14 of 23)!, 1.34 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:21.219947 : Adding K562_Wang to GeneExpressionMatrix for Chr (15 of 23)!, 1.472 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:30.765077 : Adding K562_Wang to GeneExpressionMatrix for Chr (16 of 23)!, 1.631 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:37.556681 : Adding K562_Wang to GeneExpressionMatrix for Chr (17 of 23)!, 1.745 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:43.284647 : Adding K562_Wang to GeneExpressionMatrix for Chr (18 of 23)!, 1.84 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:49.984863 : Adding K562_Wang to GeneExpressionMatrix for Chr (19 of 23)!, 1.952 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:54.871717 : Adding K562_Wang to GeneExpressionMatrix for Chr (20 of 23)!, 2.033 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:59.572073 : Adding K562_Wang to GeneExpressionMatrix for Chr (21 of 23)!, 2.111 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:06.502748 : Adding K562_Wang to GeneExpressionMatrix for Chr (22 of 23)!, 2.227 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:11.090966 : Adding K562_Wang to GeneExpressionMatrix for Chr (23 of 23)!, 2.303 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addGeneExpressionMatrix-2a52565acb9b06-Date-2025-02-26_Time-15-03-49.799199.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addGeneExpressionMatrix(obj.ArchR,\n",
    "                                    seRNA = seRNA.rename,\n",
    "                                    strictMatch = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4318703-a143-40c0-b159-274d05033bbc",
   "metadata": {},
   "source": [
    "Run ArchR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "47c8fd5f-19ec-46e6-af87-57502c1e1ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeak2GeneLinks-2a525674881716-Date-2025-02-26_Time-15-06-17.529598.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 15:06:17.842516 : Getting Available Matrices, 0.005 mins elapsed.\n",
      "\n",
      "No predictionScore found. Continuing without predictionScore!\n",
      "\n",
      "2025-02-26 15:06:18.175507 : Filtered Low Prediction Score Cells (0 of 4551, 0), 0 mins elapsed.\n",
      "\n",
      "Filtering 1 dims correlated > 0.75 to log10(depth + 1)\n",
      "\n",
      "2025-02-26 15:06:18.276014 : Computing KNN, 0.002 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:18.735977 : Identifying Non-Overlapping KNN pairs, 0.01 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:20.576406 : Identified 500 Groupings!, 0.04 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:20.617809 : Getting Group RNA Matrix, 0.041 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:37.158858 : Getting Group ATAC Matrix, 0.317 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:59.080997 : Normalizing Group Matrices, 0.682 mins elapsed.\n",
      "\n",
      "2025-02-26 15:07:04.50558 : Finding Peak Gene Pairings, 0.773 mins elapsed.\n",
      "\n",
      "2025-02-26 15:07:05.30347 : Computing Correlations, 0.786 mins elapsed.\n",
      "\n",
      "2025-02-26 15:07:20.528713 : Completed Peak2Gene Correlations!, 1.04 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeak2GeneLinks-2a525674881716-Date-2025-02-26_Time-15-06-17.529598.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "obj.ArchR = addPeak2GeneLinks(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    reducedDims = \"IterativeLSI\",\n",
    "    useMatrix = \"GeneExpressionMatrix\",\n",
    "    maxDist = maxDist,\n",
    "    threads = n.cores\n",
    ")\n",
    "end_time <- Sys.time()\n",
    "execution_time <- end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "128d0065-5f22-495b-b00d-743fc9843337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 1.050559 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "execution_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "36bd2df2-6b32-4b81-bd4f-39ef8913b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g = metadata(obj.ArchR@peakSet)$Peak2GeneLinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e7730413-8fd4-4952-bb13-0df7a3b15b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata(p2g)[[1]]$PeakName = paste(seqnames(metadata(p2g)[[1]]),\n",
    "                                    ranges(metadata(p2g)[[1]]),sep=\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "aeea3810-6735-4aff-9bd7-1697bef979d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g$PairName = paste(metadata(p2g)[[1]][p2g$idxATAC]$PeakName,\n",
    "                     metadata(p2g)[[2]][p2g$idxRNA]$name,\n",
    "                     sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "87f975ec-6ba4-4696-ad6c-2af65da7aa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 3608501 rows and 7 columns\n",
       "                                idxATAC    idxRNA Correlation         FDR\n",
       "                              <integer> <integer>   <numeric>   <numeric>\n",
       "chr1-10076-10384_FAM138A              1         1         NaN         NaN\n",
       "chr1-115531-115905_FAM138A            2         1         NaN         NaN\n",
       "chr1-138931-139302_FAM138A            3         1         NaN         NaN\n",
       "chr1-180745-181204_FAM138A            4         1         NaN         NaN\n",
       "chr1-181343-181653_FAM138A            5         1         NaN         NaN\n",
       "...                                 ...       ...         ...         ...\n",
       "chrX-156001062-156001946_IL9R    153711     20494   0.2027357 3.94427e-05\n",
       "chrX-156002345-156003202_IL9R    153712     20494   0.1766650 4.21111e-04\n",
       "chrX-156025240-156025503_IL9R    153713     20494   0.0708646 2.07143e-01\n",
       "chrX-156029994-156030343_IL9R    153714     20494  -0.0323022 6.03283e-01\n",
       "chrX-156030510-156030708_IL9R    153715     20494  -0.0169088 7.98782e-01\n",
       "                               VarQATAC   VarQRNA               PairName\n",
       "                              <numeric> <numeric>            <character>\n",
       "chr1-10076-10384_FAM138A      0.1690987 0.0507466 chr1-10076-10384_FAM..\n",
       "chr1-115531-115905_FAM138A    0.7556712 0.0507466 chr1-115531-115905_F..\n",
       "chr1-138931-139302_FAM138A    0.1692613 0.0507466 chr1-138931-139302_F..\n",
       "chr1-180745-181204_FAM138A    0.3299808 0.0507466 chr1-180745-181204_F..\n",
       "chr1-181343-181653_FAM138A    0.0189702 0.0507466 chr1-181343-181653_F..\n",
       "...                                 ...       ...                    ...\n",
       "chrX-156001062-156001946_IL9R 0.9059949  0.382307 chrX-156001062-15600..\n",
       "chrX-156002345-156003202_IL9R 0.8314478  0.382307 chrX-156002345-15600..\n",
       "chrX-156025240-156025503_IL9R 0.0233679  0.382307 chrX-156025240-15602..\n",
       "chrX-156029994-156030343_IL9R 0.1616563  0.382307 chrX-156029994-15603..\n",
       "chrX-156030510-156030708_IL9R 0.0420128  0.382307 chrX-156030510-15603.."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(p2g) = p2g$PairName\n",
    "p2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c7209a32-3904-4734-93af-9c22049f652a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = pairs.E2G[pairs.E2G$PairName %in% p2g$PairName]\n",
    "mcols(pairs.E2G.res)[,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")] = p2g[pairs.E2G.res$PairName,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1ab4ed70-ce9e-4857-bd1b-63a81d27a401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2803268 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100028738-100029317      * |         AGL\n",
       "        [2]     chr1 100028738-100029317      * |      CDC14A\n",
       "        [3]     chr1 100028738-100029317      * |         DBT\n",
       "        [4]     chr1 100028738-100029317      * |        DPH5\n",
       "        [5]     chr1 100028738-100029317      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2803264]     chrX     9940445-9940756      * |       TBL1X\n",
       "  [2803265]     chrX     9940445-9940756      * |        WWC3\n",
       "  [2803266]     chrX     9995833-9996097      * |       FAM9B\n",
       "  [2803267]     chrX     9995833-9996097      * |       TBL1X\n",
       "  [2803268]     chrX     9995833-9996097      * |        WWC3\n",
       "                          PeakName               PairName Correlation       FDR\n",
       "                       <character>            <character>   <numeric> <numeric>\n",
       "        [1] chr1-100028738-10002.. chr1-100028738-10002..   0.0785478 0.1562845\n",
       "        [2] chr1-100028738-10002.. chr1-100028738-10002..  -0.0101480 0.8825721\n",
       "        [3] chr1-100028738-10002.. chr1-100028738-10002..  -0.0432920 0.4702945\n",
       "        [4] chr1-100028738-10002.. chr1-100028738-10002..   0.1266586 0.0150749\n",
       "        [5] chr1-100028738-10002.. chr1-100028738-10002..   0.0730293 0.1917841\n",
       "        ...                    ...                    ...         ...       ...\n",
       "  [2803264]   chrX-9940445-9940756 chrX-9940445-9940756..   0.0745559 0.1814526\n",
       "  [2803265]   chrX-9940445-9940756 chrX-9940445-9940756..  -0.0420563 0.4846622\n",
       "  [2803266]   chrX-9995833-9996097 chrX-9995833-9996097..   0.0920342 0.0898625\n",
       "  [2803267]   chrX-9995833-9996097 chrX-9995833-9996097..   0.0564851 0.3293163\n",
       "  [2803268]   chrX-9995833-9996097 chrX-9995833-9996097..  -0.1545007 0.0023940\n",
       "             VarQATAC   VarQRNA\n",
       "            <numeric> <numeric>\n",
       "        [1]  0.292489  0.798966\n",
       "        [2]  0.292489  0.804626\n",
       "        [3]  0.292489  0.716942\n",
       "        [4]  0.292489  0.900117\n",
       "        [5]  0.292489  0.744364\n",
       "        ...       ...       ...\n",
       "  [2803264] 0.0117815  0.892652\n",
       "  [2803265] 0.0117815  0.406314\n",
       "  [2803266] 0.1573366  0.420172\n",
       "  [2803267] 0.1573366  0.892652\n",
       "  [2803268] 0.1573366  0.406314\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46485351-50d7-461a-bf09-d795d5bdab84",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6d424e21-6a41-4422-b76c-6f1c0bda2222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2803268 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100028738-100029317      * |         AGL\n",
       "        [2]     chr1 100028738-100029317      * |      CDC14A\n",
       "        [3]     chr1 100028738-100029317      * |         DBT\n",
       "        [4]     chr1 100028738-100029317      * |        DPH5\n",
       "        [5]     chr1 100028738-100029317      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2803264]     chrX     9940445-9940756      * |       TBL1X\n",
       "  [2803265]     chrX     9940445-9940756      * |        WWC3\n",
       "  [2803266]     chrX     9995833-9996097      * |       FAM9B\n",
       "  [2803267]     chrX     9995833-9996097      * |       TBL1X\n",
       "  [2803268]     chrX     9995833-9996097      * |        WWC3\n",
       "                          PeakName               PairName Correlation       FDR\n",
       "                       <character>            <character>   <numeric> <numeric>\n",
       "        [1] chr1-100028738-10002.. chr1-100028738-10002..   0.0785478 0.1562845\n",
       "        [2] chr1-100028738-10002.. chr1-100028738-10002..  -0.0101480 0.8825721\n",
       "        [3] chr1-100028738-10002.. chr1-100028738-10002..  -0.0432920 0.4702945\n",
       "        [4] chr1-100028738-10002.. chr1-100028738-10002..   0.1266586 0.0150749\n",
       "        [5] chr1-100028738-10002.. chr1-100028738-10002..   0.0730293 0.1917841\n",
       "        ...                    ...                    ...         ...       ...\n",
       "  [2803264]   chrX-9940445-9940756 chrX-9940445-9940756..   0.0745559 0.1814526\n",
       "  [2803265]   chrX-9940445-9940756 chrX-9940445-9940756..  -0.0420563 0.4846622\n",
       "  [2803266]   chrX-9995833-9996097 chrX-9995833-9996097..   0.0920342 0.0898625\n",
       "  [2803267]   chrX-9995833-9996097 chrX-9995833-9996097..   0.0564851 0.3293163\n",
       "  [2803268]   chrX-9995833-9996097 chrX-9995833-9996097..  -0.1545007 0.0023940\n",
       "             VarQATAC   VarQRNA\n",
       "            <numeric> <numeric>\n",
       "        [1]  0.292489  0.798966\n",
       "        [2]  0.292489  0.804626\n",
       "        [3]  0.292489  0.716942\n",
       "        [4]  0.292489  0.900117\n",
       "        [5]  0.292489  0.744364\n",
       "        ...       ...       ...\n",
       "  [2803264] 0.0117815  0.892652\n",
       "  [2803265] 0.0117815  0.406314\n",
       "  [2803266] 0.1573366  0.420172\n",
       "  [2803267] 0.1573366  0.892652\n",
       "  [2803268] 0.1573366  0.406314\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2803268 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>Correlation</th><th scope=col>FDR</th><th scope=col>VarQATAC</th><th scope=col>VarQRNA</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>AGL     </td><td>K562</td><td> 0.078547814</td><td>1.562845e-01</td><td>0.2924893</td><td>0.7989656</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>CDC14A  </td><td>K562</td><td>-0.010147960</td><td>8.825721e-01</td><td>0.2924893</td><td>0.8046257</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>DBT     </td><td>K562</td><td>-0.043291982</td><td>4.702945e-01</td><td>0.2924893</td><td>0.7169415</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>DPH5    </td><td>K562</td><td> 0.126658591</td><td>1.507493e-02</td><td>0.2924893</td><td>0.9001171</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>EXTL2   </td><td>K562</td><td> 0.073029257</td><td>1.917841e-01</td><td>0.2924893</td><td>0.7443642</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>FRRS1   </td><td>K562</td><td>-0.070390730</td><td>2.106054e-01</td><td>0.2924893</td><td>0.8625451</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>LRRC39  </td><td>K562</td><td> 0.054417442</td><td>3.497189e-01</td><td>0.2924893</td><td>0.5480628</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>MFSD14A </td><td>K562</td><td>-0.006558291</td><td>9.254416e-01</td><td>0.2924893</td><td>0.3859666</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>PALMD   </td><td>K562</td><td> 0.032048465</td><td>6.065120e-01</td><td>0.2924893</td><td>0.7882795</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>PLPPR4  </td><td>K562</td><td>-0.094822031</td><td>7.940056e-02</td><td>0.2924893</td><td>0.1753684</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>RTCA    </td><td>K562</td><td>-0.015177726</td><td>8.205285e-01</td><td>0.2924893</td><td>0.7514882</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>RTCA-AS1</td><td>K562</td><td>-0.076268292</td><td>1.703323e-01</td><td>0.2924893</td><td>0.6602908</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>SASS6   </td><td>K562</td><td>-0.092018877</td><td>8.992333e-02</td><td>0.2924893</td><td>0.8987996</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>SLC30A7 </td><td>K562</td><td> 0.242729147</td><td>5.096125e-07</td><td>0.2924893</td><td>0.8388797</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>SLC35A3 </td><td>K562</td><td> 0.037842545</td><td>5.348336e-01</td><td>0.2924893</td><td>0.6553138</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028738</td><td>100029317</td><td>TRMT13  </td><td>K562</td><td> 0.081494068</td><td>1.394223e-01</td><td>0.2924893</td><td>0.6979604</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>AGL     </td><td>K562</td><td>-0.019793479</td><td>7.622929e-01</td><td>0.0143057</td><td>0.7989656</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>CDC14A  </td><td>K562</td><td>-0.070635686</td><td>2.088023e-01</td><td>0.0143057</td><td>0.8046257</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>DBT     </td><td>K562</td><td> 0.012598462</td><td>8.525403e-01</td><td>0.0143057</td><td>0.7169415</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>DPH5    </td><td>K562</td><td>-0.233695830</td><td>1.473402e-06</td><td>0.0143057</td><td>0.9001171</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>EXTL2   </td><td>K562</td><td>-0.065704927</td><td>2.469354e-01</td><td>0.0143057</td><td>0.7443642</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>FRRS1   </td><td>K562</td><td> 0.045809225</td><td>4.415628e-01</td><td>0.0143057</td><td>0.8625451</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>LRRC39  </td><td>K562</td><td>-0.102253754</td><td>5.610481e-02</td><td>0.0143057</td><td>0.5480628</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>MFSD14A </td><td>K562</td><td> 0.089044000</td><td>1.022856e-01</td><td>0.0143057</td><td>0.3859666</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>PALMD   </td><td>K562</td><td>-0.301849317</td><td>1.488568e-10</td><td>0.0143057</td><td>0.7882795</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>PLPPR4  </td><td>K562</td><td>-0.145532768</td><td>4.513512e-03</td><td>0.0143057</td><td>0.1753684</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>RTCA    </td><td>K562</td><td>-0.069798139</td><td>2.149792e-01</td><td>0.0143057</td><td>0.7514882</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>RTCA-AS1</td><td>K562</td><td> 0.105090081</td><td>4.883099e-02</td><td>0.0143057</td><td>0.6602908</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>SASS6   </td><td>K562</td><td> 0.038630780</td><td>5.253219e-01</td><td>0.0143057</td><td>0.8987996</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029410</td><td>100029652</td><td>SLC30A7 </td><td>K562</td><td>-0.082578060</td><td>1.335438e-01</td><td>0.0143057</td><td>0.8388797</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9864063</td><td>9864529</td><td>TBL1X    </td><td>K562</td><td>-0.132060587</td><td>1.086291e-02</td><td>0.51134892</td><td>0.8926515</td></tr>\n",
       "\t<tr><td>chrX</td><td>9864063</td><td>9864529</td><td>WWC3     </td><td>K562</td><td> 0.059899400</td><td>2.971978e-01</td><td>0.51134892</td><td>0.4063140</td></tr>\n",
       "\t<tr><td>chrX</td><td>9874033</td><td>9874283</td><td>FAM9B    </td><td>K562</td><td> 0.010235225</td><td>8.815172e-01</td><td>0.03608626</td><td>0.4201718</td></tr>\n",
       "\t<tr><td>chrX</td><td>9874033</td><td>9874283</td><td>TBL1X    </td><td>K562</td><td>-0.026983064</td><td>6.706844e-01</td><td>0.03608626</td><td>0.8926515</td></tr>\n",
       "\t<tr><td>chrX</td><td>9874033</td><td>9874283</td><td>WWC3     </td><td>K562</td><td> 0.129753423</td><td>1.251778e-02</td><td>0.03608626</td><td>0.4063140</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882684</td><td>9883146</td><td>FAM9B    </td><td>K562</td><td> 0.070174841</td><td>2.121953e-01</td><td>0.27314185</td><td>0.4201718</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882684</td><td>9883146</td><td>TBL1X    </td><td>K562</td><td> 0.204683710</td><td>3.256565e-05</td><td>0.27314185</td><td>0.8926515</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882684</td><td>9883146</td><td>WWC3     </td><td>K562</td><td> 0.159862033</td><td>1.608991e-03</td><td>0.27314185</td><td>0.4063140</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885199</td><td>9885463</td><td>FAM9B    </td><td>K562</td><td>-0.039101809</td><td>5.197063e-01</td><td>0.10931269</td><td>0.4201718</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885199</td><td>9885463</td><td>TBL1X    </td><td>K562</td><td>-0.085110828</td><td>1.205360e-01</td><td>0.10931269</td><td>0.8926515</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885199</td><td>9885463</td><td>WWC3     </td><td>K562</td><td>-0.022742719</td><td>7.248081e-01</td><td>0.10931269</td><td>0.4063140</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894498</td><td>9895136</td><td>FAM9B    </td><td>K562</td><td> 0.003652618</td><td>9.591882e-01</td><td>0.35031715</td><td>0.4201718</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894498</td><td>9895136</td><td>TBL1X    </td><td>K562</td><td>-0.051243269</td><td>3.823236e-01</td><td>0.35031715</td><td>0.8926515</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894498</td><td>9895136</td><td>WWC3     </td><td>K562</td><td>-0.110918108</td><td>3.629741e-02</td><td>0.35031715</td><td>0.4063140</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>AKAP17A  </td><td>K562</td><td> 0.202442629</td><td>4.058665e-05</td><td>0.94445565</td><td>0.6909827</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>ASMT     </td><td>K562</td><td> 0.149247950</td><td>3.486628e-03</td><td>0.94445565</td><td>0.1842978</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>ASMTL    </td><td>K562</td><td> 0.061122430</td><td>2.861070e-01</td><td>0.94445565</td><td>0.5703621</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>GTPBP6   </td><td>K562</td><td> 0.231942160</td><td>1.800711e-06</td><td>0.94445565</td><td>0.5604079</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>IL3RA    </td><td>K562</td><td> 0.048867239</td><td>4.077065e-01</td><td>0.94445565</td><td>0.3749878</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>LINC00106</td><td>K562</td><td>-0.267271597</td><td>2.249642e-08</td><td>0.94445565</td><td>0.5051722</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>P2RY8    </td><td>K562</td><td> 0.088440034</td><td>1.049511e-01</td><td>0.94445565</td><td>0.4737972</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>PLCXD1   </td><td>K562</td><td> 0.037820031</td><td>5.351016e-01</td><td>0.94445565</td><td>0.7363131</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>PPP2R3B  </td><td>K562</td><td>-0.342588803</td><td>1.495773e-13</td><td>0.94445565</td><td>0.5828535</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989856</td><td> 990611</td><td>SLC25A6  </td><td>K562</td><td> 0.102993546</td><td>5.412816e-02</td><td>0.94445565</td><td>0.8420025</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940445</td><td>9940756</td><td>FAM9B    </td><td>K562</td><td>-0.049341842</td><td>4.025928e-01</td><td>0.01178154</td><td>0.4201718</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940445</td><td>9940756</td><td>TBL1X    </td><td>K562</td><td> 0.074555909</td><td>1.814526e-01</td><td>0.01178154</td><td>0.8926515</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940445</td><td>9940756</td><td>WWC3     </td><td>K562</td><td>-0.042056278</td><td>4.846622e-01</td><td>0.01178154</td><td>0.4063140</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995833</td><td>9996097</td><td>FAM9B    </td><td>K562</td><td> 0.092034219</td><td>8.986252e-02</td><td>0.15733663</td><td>0.4201718</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995833</td><td>9996097</td><td>TBL1X    </td><td>K562</td><td> 0.056485087</td><td>3.293163e-01</td><td>0.15733663</td><td>0.8926515</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995833</td><td>9996097</td><td>WWC3     </td><td>K562</td><td>-0.154500741</td><td>2.393998e-03</td><td>0.15733663</td><td>0.4063140</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2803268 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & Correlation & FDR & VarQATAC & VarQRNA\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100028738 & 100029317 & AGL      & K562 &  0.078547814 & 1.562845e-01 & 0.2924893 & 0.7989656\\\\\n",
       "\t chr1 & 100028738 & 100029317 & CDC14A   & K562 & -0.010147960 & 8.825721e-01 & 0.2924893 & 0.8046257\\\\\n",
       "\t chr1 & 100028738 & 100029317 & DBT      & K562 & -0.043291982 & 4.702945e-01 & 0.2924893 & 0.7169415\\\\\n",
       "\t chr1 & 100028738 & 100029317 & DPH5     & K562 &  0.126658591 & 1.507493e-02 & 0.2924893 & 0.9001171\\\\\n",
       "\t chr1 & 100028738 & 100029317 & EXTL2    & K562 &  0.073029257 & 1.917841e-01 & 0.2924893 & 0.7443642\\\\\n",
       "\t chr1 & 100028738 & 100029317 & FRRS1    & K562 & -0.070390730 & 2.106054e-01 & 0.2924893 & 0.8625451\\\\\n",
       "\t chr1 & 100028738 & 100029317 & LRRC39   & K562 &  0.054417442 & 3.497189e-01 & 0.2924893 & 0.5480628\\\\\n",
       "\t chr1 & 100028738 & 100029317 & MFSD14A  & K562 & -0.006558291 & 9.254416e-01 & 0.2924893 & 0.3859666\\\\\n",
       "\t chr1 & 100028738 & 100029317 & PALMD    & K562 &  0.032048465 & 6.065120e-01 & 0.2924893 & 0.7882795\\\\\n",
       "\t chr1 & 100028738 & 100029317 & PLPPR4   & K562 & -0.094822031 & 7.940056e-02 & 0.2924893 & 0.1753684\\\\\n",
       "\t chr1 & 100028738 & 100029317 & RTCA     & K562 & -0.015177726 & 8.205285e-01 & 0.2924893 & 0.7514882\\\\\n",
       "\t chr1 & 100028738 & 100029317 & RTCA-AS1 & K562 & -0.076268292 & 1.703323e-01 & 0.2924893 & 0.6602908\\\\\n",
       "\t chr1 & 100028738 & 100029317 & SASS6    & K562 & -0.092018877 & 8.992333e-02 & 0.2924893 & 0.8987996\\\\\n",
       "\t chr1 & 100028738 & 100029317 & SLC30A7  & K562 &  0.242729147 & 5.096125e-07 & 0.2924893 & 0.8388797\\\\\n",
       "\t chr1 & 100028738 & 100029317 & SLC35A3  & K562 &  0.037842545 & 5.348336e-01 & 0.2924893 & 0.6553138\\\\\n",
       "\t chr1 & 100028738 & 100029317 & TRMT13   & K562 &  0.081494068 & 1.394223e-01 & 0.2924893 & 0.6979604\\\\\n",
       "\t chr1 & 100029410 & 100029652 & AGL      & K562 & -0.019793479 & 7.622929e-01 & 0.0143057 & 0.7989656\\\\\n",
       "\t chr1 & 100029410 & 100029652 & CDC14A   & K562 & -0.070635686 & 2.088023e-01 & 0.0143057 & 0.8046257\\\\\n",
       "\t chr1 & 100029410 & 100029652 & DBT      & K562 &  0.012598462 & 8.525403e-01 & 0.0143057 & 0.7169415\\\\\n",
       "\t chr1 & 100029410 & 100029652 & DPH5     & K562 & -0.233695830 & 1.473402e-06 & 0.0143057 & 0.9001171\\\\\n",
       "\t chr1 & 100029410 & 100029652 & EXTL2    & K562 & -0.065704927 & 2.469354e-01 & 0.0143057 & 0.7443642\\\\\n",
       "\t chr1 & 100029410 & 100029652 & FRRS1    & K562 &  0.045809225 & 4.415628e-01 & 0.0143057 & 0.8625451\\\\\n",
       "\t chr1 & 100029410 & 100029652 & LRRC39   & K562 & -0.102253754 & 5.610481e-02 & 0.0143057 & 0.5480628\\\\\n",
       "\t chr1 & 100029410 & 100029652 & MFSD14A  & K562 &  0.089044000 & 1.022856e-01 & 0.0143057 & 0.3859666\\\\\n",
       "\t chr1 & 100029410 & 100029652 & PALMD    & K562 & -0.301849317 & 1.488568e-10 & 0.0143057 & 0.7882795\\\\\n",
       "\t chr1 & 100029410 & 100029652 & PLPPR4   & K562 & -0.145532768 & 4.513512e-03 & 0.0143057 & 0.1753684\\\\\n",
       "\t chr1 & 100029410 & 100029652 & RTCA     & K562 & -0.069798139 & 2.149792e-01 & 0.0143057 & 0.7514882\\\\\n",
       "\t chr1 & 100029410 & 100029652 & RTCA-AS1 & K562 &  0.105090081 & 4.883099e-02 & 0.0143057 & 0.6602908\\\\\n",
       "\t chr1 & 100029410 & 100029652 & SASS6    & K562 &  0.038630780 & 5.253219e-01 & 0.0143057 & 0.8987996\\\\\n",
       "\t chr1 & 100029410 & 100029652 & SLC30A7  & K562 & -0.082578060 & 1.335438e-01 & 0.0143057 & 0.8388797\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9864063 & 9864529 & TBL1X     & K562 & -0.132060587 & 1.086291e-02 & 0.51134892 & 0.8926515\\\\\n",
       "\t chrX & 9864063 & 9864529 & WWC3      & K562 &  0.059899400 & 2.971978e-01 & 0.51134892 & 0.4063140\\\\\n",
       "\t chrX & 9874033 & 9874283 & FAM9B     & K562 &  0.010235225 & 8.815172e-01 & 0.03608626 & 0.4201718\\\\\n",
       "\t chrX & 9874033 & 9874283 & TBL1X     & K562 & -0.026983064 & 6.706844e-01 & 0.03608626 & 0.8926515\\\\\n",
       "\t chrX & 9874033 & 9874283 & WWC3      & K562 &  0.129753423 & 1.251778e-02 & 0.03608626 & 0.4063140\\\\\n",
       "\t chrX & 9882684 & 9883146 & FAM9B     & K562 &  0.070174841 & 2.121953e-01 & 0.27314185 & 0.4201718\\\\\n",
       "\t chrX & 9882684 & 9883146 & TBL1X     & K562 &  0.204683710 & 3.256565e-05 & 0.27314185 & 0.8926515\\\\\n",
       "\t chrX & 9882684 & 9883146 & WWC3      & K562 &  0.159862033 & 1.608991e-03 & 0.27314185 & 0.4063140\\\\\n",
       "\t chrX & 9885199 & 9885463 & FAM9B     & K562 & -0.039101809 & 5.197063e-01 & 0.10931269 & 0.4201718\\\\\n",
       "\t chrX & 9885199 & 9885463 & TBL1X     & K562 & -0.085110828 & 1.205360e-01 & 0.10931269 & 0.8926515\\\\\n",
       "\t chrX & 9885199 & 9885463 & WWC3      & K562 & -0.022742719 & 7.248081e-01 & 0.10931269 & 0.4063140\\\\\n",
       "\t chrX & 9894498 & 9895136 & FAM9B     & K562 &  0.003652618 & 9.591882e-01 & 0.35031715 & 0.4201718\\\\\n",
       "\t chrX & 9894498 & 9895136 & TBL1X     & K562 & -0.051243269 & 3.823236e-01 & 0.35031715 & 0.8926515\\\\\n",
       "\t chrX & 9894498 & 9895136 & WWC3      & K562 & -0.110918108 & 3.629741e-02 & 0.35031715 & 0.4063140\\\\\n",
       "\t chrX &  989856 &  990611 & AKAP17A   & K562 &  0.202442629 & 4.058665e-05 & 0.94445565 & 0.6909827\\\\\n",
       "\t chrX &  989856 &  990611 & ASMT      & K562 &  0.149247950 & 3.486628e-03 & 0.94445565 & 0.1842978\\\\\n",
       "\t chrX &  989856 &  990611 & ASMTL     & K562 &  0.061122430 & 2.861070e-01 & 0.94445565 & 0.5703621\\\\\n",
       "\t chrX &  989856 &  990611 & GTPBP6    & K562 &  0.231942160 & 1.800711e-06 & 0.94445565 & 0.5604079\\\\\n",
       "\t chrX &  989856 &  990611 & IL3RA     & K562 &  0.048867239 & 4.077065e-01 & 0.94445565 & 0.3749878\\\\\n",
       "\t chrX &  989856 &  990611 & LINC00106 & K562 & -0.267271597 & 2.249642e-08 & 0.94445565 & 0.5051722\\\\\n",
       "\t chrX &  989856 &  990611 & P2RY8     & K562 &  0.088440034 & 1.049511e-01 & 0.94445565 & 0.4737972\\\\\n",
       "\t chrX &  989856 &  990611 & PLCXD1    & K562 &  0.037820031 & 5.351016e-01 & 0.94445565 & 0.7363131\\\\\n",
       "\t chrX &  989856 &  990611 & PPP2R3B   & K562 & -0.342588803 & 1.495773e-13 & 0.94445565 & 0.5828535\\\\\n",
       "\t chrX &  989856 &  990611 & SLC25A6   & K562 &  0.102993546 & 5.412816e-02 & 0.94445565 & 0.8420025\\\\\n",
       "\t chrX & 9940445 & 9940756 & FAM9B     & K562 & -0.049341842 & 4.025928e-01 & 0.01178154 & 0.4201718\\\\\n",
       "\t chrX & 9940445 & 9940756 & TBL1X     & K562 &  0.074555909 & 1.814526e-01 & 0.01178154 & 0.8926515\\\\\n",
       "\t chrX & 9940445 & 9940756 & WWC3      & K562 & -0.042056278 & 4.846622e-01 & 0.01178154 & 0.4063140\\\\\n",
       "\t chrX & 9995833 & 9996097 & FAM9B     & K562 &  0.092034219 & 8.986252e-02 & 0.15733663 & 0.4201718\\\\\n",
       "\t chrX & 9995833 & 9996097 & TBL1X     & K562 &  0.056485087 & 3.293163e-01 & 0.15733663 & 0.8926515\\\\\n",
       "\t chrX & 9995833 & 9996097 & WWC3      & K562 & -0.154500741 & 2.393998e-03 & 0.15733663 & 0.4063140\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2803268 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | Correlation &lt;dbl&gt; | FDR &lt;dbl&gt; | VarQATAC &lt;dbl&gt; | VarQRNA &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100028738 | 100029317 | AGL      | K562 |  0.078547814 | 1.562845e-01 | 0.2924893 | 0.7989656 |\n",
       "| chr1 | 100028738 | 100029317 | CDC14A   | K562 | -0.010147960 | 8.825721e-01 | 0.2924893 | 0.8046257 |\n",
       "| chr1 | 100028738 | 100029317 | DBT      | K562 | -0.043291982 | 4.702945e-01 | 0.2924893 | 0.7169415 |\n",
       "| chr1 | 100028738 | 100029317 | DPH5     | K562 |  0.126658591 | 1.507493e-02 | 0.2924893 | 0.9001171 |\n",
       "| chr1 | 100028738 | 100029317 | EXTL2    | K562 |  0.073029257 | 1.917841e-01 | 0.2924893 | 0.7443642 |\n",
       "| chr1 | 100028738 | 100029317 | FRRS1    | K562 | -0.070390730 | 2.106054e-01 | 0.2924893 | 0.8625451 |\n",
       "| chr1 | 100028738 | 100029317 | LRRC39   | K562 |  0.054417442 | 3.497189e-01 | 0.2924893 | 0.5480628 |\n",
       "| chr1 | 100028738 | 100029317 | MFSD14A  | K562 | -0.006558291 | 9.254416e-01 | 0.2924893 | 0.3859666 |\n",
       "| chr1 | 100028738 | 100029317 | PALMD    | K562 |  0.032048465 | 6.065120e-01 | 0.2924893 | 0.7882795 |\n",
       "| chr1 | 100028738 | 100029317 | PLPPR4   | K562 | -0.094822031 | 7.940056e-02 | 0.2924893 | 0.1753684 |\n",
       "| chr1 | 100028738 | 100029317 | RTCA     | K562 | -0.015177726 | 8.205285e-01 | 0.2924893 | 0.7514882 |\n",
       "| chr1 | 100028738 | 100029317 | RTCA-AS1 | K562 | -0.076268292 | 1.703323e-01 | 0.2924893 | 0.6602908 |\n",
       "| chr1 | 100028738 | 100029317 | SASS6    | K562 | -0.092018877 | 8.992333e-02 | 0.2924893 | 0.8987996 |\n",
       "| chr1 | 100028738 | 100029317 | SLC30A7  | K562 |  0.242729147 | 5.096125e-07 | 0.2924893 | 0.8388797 |\n",
       "| chr1 | 100028738 | 100029317 | SLC35A3  | K562 |  0.037842545 | 5.348336e-01 | 0.2924893 | 0.6553138 |\n",
       "| chr1 | 100028738 | 100029317 | TRMT13   | K562 |  0.081494068 | 1.394223e-01 | 0.2924893 | 0.6979604 |\n",
       "| chr1 | 100029410 | 100029652 | AGL      | K562 | -0.019793479 | 7.622929e-01 | 0.0143057 | 0.7989656 |\n",
       "| chr1 | 100029410 | 100029652 | CDC14A   | K562 | -0.070635686 | 2.088023e-01 | 0.0143057 | 0.8046257 |\n",
       "| chr1 | 100029410 | 100029652 | DBT      | K562 |  0.012598462 | 8.525403e-01 | 0.0143057 | 0.7169415 |\n",
       "| chr1 | 100029410 | 100029652 | DPH5     | K562 | -0.233695830 | 1.473402e-06 | 0.0143057 | 0.9001171 |\n",
       "| chr1 | 100029410 | 100029652 | EXTL2    | K562 | -0.065704927 | 2.469354e-01 | 0.0143057 | 0.7443642 |\n",
       "| chr1 | 100029410 | 100029652 | FRRS1    | K562 |  0.045809225 | 4.415628e-01 | 0.0143057 | 0.8625451 |\n",
       "| chr1 | 100029410 | 100029652 | LRRC39   | K562 | -0.102253754 | 5.610481e-02 | 0.0143057 | 0.5480628 |\n",
       "| chr1 | 100029410 | 100029652 | MFSD14A  | K562 |  0.089044000 | 1.022856e-01 | 0.0143057 | 0.3859666 |\n",
       "| chr1 | 100029410 | 100029652 | PALMD    | K562 | -0.301849317 | 1.488568e-10 | 0.0143057 | 0.7882795 |\n",
       "| chr1 | 100029410 | 100029652 | PLPPR4   | K562 | -0.145532768 | 4.513512e-03 | 0.0143057 | 0.1753684 |\n",
       "| chr1 | 100029410 | 100029652 | RTCA     | K562 | -0.069798139 | 2.149792e-01 | 0.0143057 | 0.7514882 |\n",
       "| chr1 | 100029410 | 100029652 | RTCA-AS1 | K562 |  0.105090081 | 4.883099e-02 | 0.0143057 | 0.6602908 |\n",
       "| chr1 | 100029410 | 100029652 | SASS6    | K562 |  0.038630780 | 5.253219e-01 | 0.0143057 | 0.8987996 |\n",
       "| chr1 | 100029410 | 100029652 | SLC30A7  | K562 | -0.082578060 | 1.335438e-01 | 0.0143057 | 0.8388797 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9864063 | 9864529 | TBL1X     | K562 | -0.132060587 | 1.086291e-02 | 0.51134892 | 0.8926515 |\n",
       "| chrX | 9864063 | 9864529 | WWC3      | K562 |  0.059899400 | 2.971978e-01 | 0.51134892 | 0.4063140 |\n",
       "| chrX | 9874033 | 9874283 | FAM9B     | K562 |  0.010235225 | 8.815172e-01 | 0.03608626 | 0.4201718 |\n",
       "| chrX | 9874033 | 9874283 | TBL1X     | K562 | -0.026983064 | 6.706844e-01 | 0.03608626 | 0.8926515 |\n",
       "| chrX | 9874033 | 9874283 | WWC3      | K562 |  0.129753423 | 1.251778e-02 | 0.03608626 | 0.4063140 |\n",
       "| chrX | 9882684 | 9883146 | FAM9B     | K562 |  0.070174841 | 2.121953e-01 | 0.27314185 | 0.4201718 |\n",
       "| chrX | 9882684 | 9883146 | TBL1X     | K562 |  0.204683710 | 3.256565e-05 | 0.27314185 | 0.8926515 |\n",
       "| chrX | 9882684 | 9883146 | WWC3      | K562 |  0.159862033 | 1.608991e-03 | 0.27314185 | 0.4063140 |\n",
       "| chrX | 9885199 | 9885463 | FAM9B     | K562 | -0.039101809 | 5.197063e-01 | 0.10931269 | 0.4201718 |\n",
       "| chrX | 9885199 | 9885463 | TBL1X     | K562 | -0.085110828 | 1.205360e-01 | 0.10931269 | 0.8926515 |\n",
       "| chrX | 9885199 | 9885463 | WWC3      | K562 | -0.022742719 | 7.248081e-01 | 0.10931269 | 0.4063140 |\n",
       "| chrX | 9894498 | 9895136 | FAM9B     | K562 |  0.003652618 | 9.591882e-01 | 0.35031715 | 0.4201718 |\n",
       "| chrX | 9894498 | 9895136 | TBL1X     | K562 | -0.051243269 | 3.823236e-01 | 0.35031715 | 0.8926515 |\n",
       "| chrX | 9894498 | 9895136 | WWC3      | K562 | -0.110918108 | 3.629741e-02 | 0.35031715 | 0.4063140 |\n",
       "| chrX |  989856 |  990611 | AKAP17A   | K562 |  0.202442629 | 4.058665e-05 | 0.94445565 | 0.6909827 |\n",
       "| chrX |  989856 |  990611 | ASMT      | K562 |  0.149247950 | 3.486628e-03 | 0.94445565 | 0.1842978 |\n",
       "| chrX |  989856 |  990611 | ASMTL     | K562 |  0.061122430 | 2.861070e-01 | 0.94445565 | 0.5703621 |\n",
       "| chrX |  989856 |  990611 | GTPBP6    | K562 |  0.231942160 | 1.800711e-06 | 0.94445565 | 0.5604079 |\n",
       "| chrX |  989856 |  990611 | IL3RA     | K562 |  0.048867239 | 4.077065e-01 | 0.94445565 | 0.3749878 |\n",
       "| chrX |  989856 |  990611 | LINC00106 | K562 | -0.267271597 | 2.249642e-08 | 0.94445565 | 0.5051722 |\n",
       "| chrX |  989856 |  990611 | P2RY8     | K562 |  0.088440034 | 1.049511e-01 | 0.94445565 | 0.4737972 |\n",
       "| chrX |  989856 |  990611 | PLCXD1    | K562 |  0.037820031 | 5.351016e-01 | 0.94445565 | 0.7363131 |\n",
       "| chrX |  989856 |  990611 | PPP2R3B   | K562 | -0.342588803 | 1.495773e-13 | 0.94445565 | 0.5828535 |\n",
       "| chrX |  989856 |  990611 | SLC25A6   | K562 |  0.102993546 | 5.412816e-02 | 0.94445565 | 0.8420025 |\n",
       "| chrX | 9940445 | 9940756 | FAM9B     | K562 | -0.049341842 | 4.025928e-01 | 0.01178154 | 0.4201718 |\n",
       "| chrX | 9940445 | 9940756 | TBL1X     | K562 |  0.074555909 | 1.814526e-01 | 0.01178154 | 0.8926515 |\n",
       "| chrX | 9940445 | 9940756 | WWC3      | K562 | -0.042056278 | 4.846622e-01 | 0.01178154 | 0.4063140 |\n",
       "| chrX | 9995833 | 9996097 | FAM9B     | K562 |  0.092034219 | 8.986252e-02 | 0.15733663 | 0.4201718 |\n",
       "| chrX | 9995833 | 9996097 | TBL1X     | K562 |  0.056485087 | 3.293163e-01 | 0.15733663 | 0.8926515 |\n",
       "| chrX | 9995833 | 9996097 | WWC3      | K562 | -0.154500741 | 2.393998e-03 | 0.15733663 | 0.4063140 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType Correlation  FDR         \n",
       "1       chr1 100028738 100029317 AGL        K562      0.078547814 1.562845e-01\n",
       "2       chr1 100028738 100029317 CDC14A     K562     -0.010147960 8.825721e-01\n",
       "3       chr1 100028738 100029317 DBT        K562     -0.043291982 4.702945e-01\n",
       "4       chr1 100028738 100029317 DPH5       K562      0.126658591 1.507493e-02\n",
       "5       chr1 100028738 100029317 EXTL2      K562      0.073029257 1.917841e-01\n",
       "6       chr1 100028738 100029317 FRRS1      K562     -0.070390730 2.106054e-01\n",
       "7       chr1 100028738 100029317 LRRC39     K562      0.054417442 3.497189e-01\n",
       "8       chr1 100028738 100029317 MFSD14A    K562     -0.006558291 9.254416e-01\n",
       "9       chr1 100028738 100029317 PALMD      K562      0.032048465 6.065120e-01\n",
       "10      chr1 100028738 100029317 PLPPR4     K562     -0.094822031 7.940056e-02\n",
       "11      chr1 100028738 100029317 RTCA       K562     -0.015177726 8.205285e-01\n",
       "12      chr1 100028738 100029317 RTCA-AS1   K562     -0.076268292 1.703323e-01\n",
       "13      chr1 100028738 100029317 SASS6      K562     -0.092018877 8.992333e-02\n",
       "14      chr1 100028738 100029317 SLC30A7    K562      0.242729147 5.096125e-07\n",
       "15      chr1 100028738 100029317 SLC35A3    K562      0.037842545 5.348336e-01\n",
       "16      chr1 100028738 100029317 TRMT13     K562      0.081494068 1.394223e-01\n",
       "17      chr1 100029410 100029652 AGL        K562     -0.019793479 7.622929e-01\n",
       "18      chr1 100029410 100029652 CDC14A     K562     -0.070635686 2.088023e-01\n",
       "19      chr1 100029410 100029652 DBT        K562      0.012598462 8.525403e-01\n",
       "20      chr1 100029410 100029652 DPH5       K562     -0.233695830 1.473402e-06\n",
       "21      chr1 100029410 100029652 EXTL2      K562     -0.065704927 2.469354e-01\n",
       "22      chr1 100029410 100029652 FRRS1      K562      0.045809225 4.415628e-01\n",
       "23      chr1 100029410 100029652 LRRC39     K562     -0.102253754 5.610481e-02\n",
       "24      chr1 100029410 100029652 MFSD14A    K562      0.089044000 1.022856e-01\n",
       "25      chr1 100029410 100029652 PALMD      K562     -0.301849317 1.488568e-10\n",
       "26      chr1 100029410 100029652 PLPPR4     K562     -0.145532768 4.513512e-03\n",
       "27      chr1 100029410 100029652 RTCA       K562     -0.069798139 2.149792e-01\n",
       "28      chr1 100029410 100029652 RTCA-AS1   K562      0.105090081 4.883099e-02\n",
       "29      chr1 100029410 100029652 SASS6      K562      0.038630780 5.253219e-01\n",
       "30      chr1 100029410 100029652 SLC30A7    K562     -0.082578060 1.335438e-01\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮            ⋮           \n",
       "2803239 chrX 9864063   9864529   TBL1X      K562     -0.132060587 1.086291e-02\n",
       "2803240 chrX 9864063   9864529   WWC3       K562      0.059899400 2.971978e-01\n",
       "2803241 chrX 9874033   9874283   FAM9B      K562      0.010235225 8.815172e-01\n",
       "2803242 chrX 9874033   9874283   TBL1X      K562     -0.026983064 6.706844e-01\n",
       "2803243 chrX 9874033   9874283   WWC3       K562      0.129753423 1.251778e-02\n",
       "2803244 chrX 9882684   9883146   FAM9B      K562      0.070174841 2.121953e-01\n",
       "2803245 chrX 9882684   9883146   TBL1X      K562      0.204683710 3.256565e-05\n",
       "2803246 chrX 9882684   9883146   WWC3       K562      0.159862033 1.608991e-03\n",
       "2803247 chrX 9885199   9885463   FAM9B      K562     -0.039101809 5.197063e-01\n",
       "2803248 chrX 9885199   9885463   TBL1X      K562     -0.085110828 1.205360e-01\n",
       "2803249 chrX 9885199   9885463   WWC3       K562     -0.022742719 7.248081e-01\n",
       "2803250 chrX 9894498   9895136   FAM9B      K562      0.003652618 9.591882e-01\n",
       "2803251 chrX 9894498   9895136   TBL1X      K562     -0.051243269 3.823236e-01\n",
       "2803252 chrX 9894498   9895136   WWC3       K562     -0.110918108 3.629741e-02\n",
       "2803253 chrX  989856    990611   AKAP17A    K562      0.202442629 4.058665e-05\n",
       "2803254 chrX  989856    990611   ASMT       K562      0.149247950 3.486628e-03\n",
       "2803255 chrX  989856    990611   ASMTL      K562      0.061122430 2.861070e-01\n",
       "2803256 chrX  989856    990611   GTPBP6     K562      0.231942160 1.800711e-06\n",
       "2803257 chrX  989856    990611   IL3RA      K562      0.048867239 4.077065e-01\n",
       "2803258 chrX  989856    990611   LINC00106  K562     -0.267271597 2.249642e-08\n",
       "2803259 chrX  989856    990611   P2RY8      K562      0.088440034 1.049511e-01\n",
       "2803260 chrX  989856    990611   PLCXD1     K562      0.037820031 5.351016e-01\n",
       "2803261 chrX  989856    990611   PPP2R3B    K562     -0.342588803 1.495773e-13\n",
       "2803262 chrX  989856    990611   SLC25A6    K562      0.102993546 5.412816e-02\n",
       "2803263 chrX 9940445   9940756   FAM9B      K562     -0.049341842 4.025928e-01\n",
       "2803264 chrX 9940445   9940756   TBL1X      K562      0.074555909 1.814526e-01\n",
       "2803265 chrX 9940445   9940756   WWC3       K562     -0.042056278 4.846622e-01\n",
       "2803266 chrX 9995833   9996097   FAM9B      K562      0.092034219 8.986252e-02\n",
       "2803267 chrX 9995833   9996097   TBL1X      K562      0.056485087 3.293163e-01\n",
       "2803268 chrX 9995833   9996097   WWC3       K562     -0.154500741 2.393998e-03\n",
       "        VarQATAC   VarQRNA  \n",
       "1       0.2924893  0.7989656\n",
       "2       0.2924893  0.8046257\n",
       "3       0.2924893  0.7169415\n",
       "4       0.2924893  0.9001171\n",
       "5       0.2924893  0.7443642\n",
       "6       0.2924893  0.8625451\n",
       "7       0.2924893  0.5480628\n",
       "8       0.2924893  0.3859666\n",
       "9       0.2924893  0.7882795\n",
       "10      0.2924893  0.1753684\n",
       "11      0.2924893  0.7514882\n",
       "12      0.2924893  0.6602908\n",
       "13      0.2924893  0.8987996\n",
       "14      0.2924893  0.8388797\n",
       "15      0.2924893  0.6553138\n",
       "16      0.2924893  0.6979604\n",
       "17      0.0143057  0.7989656\n",
       "18      0.0143057  0.8046257\n",
       "19      0.0143057  0.7169415\n",
       "20      0.0143057  0.9001171\n",
       "21      0.0143057  0.7443642\n",
       "22      0.0143057  0.8625451\n",
       "23      0.0143057  0.5480628\n",
       "24      0.0143057  0.3859666\n",
       "25      0.0143057  0.7882795\n",
       "26      0.0143057  0.1753684\n",
       "27      0.0143057  0.7514882\n",
       "28      0.0143057  0.6602908\n",
       "29      0.0143057  0.8987996\n",
       "30      0.0143057  0.8388797\n",
       "⋮       ⋮          ⋮        \n",
       "2803239 0.51134892 0.8926515\n",
       "2803240 0.51134892 0.4063140\n",
       "2803241 0.03608626 0.4201718\n",
       "2803242 0.03608626 0.8926515\n",
       "2803243 0.03608626 0.4063140\n",
       "2803244 0.27314185 0.4201718\n",
       "2803245 0.27314185 0.8926515\n",
       "2803246 0.27314185 0.4063140\n",
       "2803247 0.10931269 0.4201718\n",
       "2803248 0.10931269 0.8926515\n",
       "2803249 0.10931269 0.4063140\n",
       "2803250 0.35031715 0.4201718\n",
       "2803251 0.35031715 0.8926515\n",
       "2803252 0.35031715 0.4063140\n",
       "2803253 0.94445565 0.6909827\n",
       "2803254 0.94445565 0.1842978\n",
       "2803255 0.94445565 0.5703621\n",
       "2803256 0.94445565 0.5604079\n",
       "2803257 0.94445565 0.3749878\n",
       "2803258 0.94445565 0.5051722\n",
       "2803259 0.94445565 0.4737972\n",
       "2803260 0.94445565 0.7363131\n",
       "2803261 0.94445565 0.5828535\n",
       "2803262 0.94445565 0.8420025\n",
       "2803263 0.01178154 0.4201718\n",
       "2803264 0.01178154 0.8926515\n",
       "2803265 0.01178154 0.4063140\n",
       "2803266 0.15733663 0.4201718\n",
       "2803267 0.15733663 0.8926515\n",
       "2803268 0.15733663 0.4063140"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(obj.ArchR,\n",
    "        paste(dir.output,\"obj.ArchR.rds\",sep = \"/\"))\n",
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"Correlation\",\n",
    "                         \"FDR\",\n",
    "                         \"VarQATAC\",\n",
    "                         \"VarQRNA\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "pairs.E2G.res\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4c909ed5-7321-4e69-8208-1544b371da6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] nabor_0.5.0                       dplyr_1.1.4                      \n",
       " [3] uwot_0.1.16                       Rsamtools_2.18.0                 \n",
       " [5] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.68.0                  \n",
       " [7] rtracklayer_1.60.1                Biostrings_2.70.1                \n",
       " [9] XVector_0.42.0                    genomation_1.34.0                \n",
       "[11] Signac_1.13.0                     Seurat_5.0.3                     \n",
       "[13] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[15] rhdf5_2.44.0                      SummarizedExperiment_1.30.2      \n",
       "[17] Biobase_2.60.0                    MatrixGenerics_1.12.3            \n",
       "[19] Rcpp_1.0.12                       Matrix_1.6-5                     \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               matrixStats_1.3.0                \n",
       "[27] data.table_1.15.2                 stringr_1.5.1                    \n",
       "[29] plyr_1.8.9                        magrittr_2.0.3                   \n",
       "[31] ggplot2_3.5.1                     gtable_0.3.5                     \n",
       "[33] gtools_3.9.5                      gridExtra_2.3                    \n",
       "[35] ArchR_1.0.2                      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22         splines_4.3.3            later_1.3.2             \n",
       "  [4] pbdZMQ_0.3-11            BiocIO_1.10.0            bitops_1.0-7            \n",
       "  [7] tibble_3.2.1             polyclip_1.10-6          XML_3.99-0.16.1         \n",
       " [10] fastDummies_1.7.3        lifecycle_1.0.4          vroom_1.6.5             \n",
       " [13] globals_0.16.3           lattice_0.22-6           MASS_7.3-60             \n",
       " [16] plotly_4.10.4            plotrix_3.8-4            yaml_2.3.8              \n",
       " [19] httpuv_1.6.15            sctransform_0.4.1        spam_2.10-0             \n",
       " [22] spatstat.sparse_3.0-3    reticulate_1.36.1        cowplot_1.1.3           \n",
       " [25] pbapply_1.7-2            RColorBrewer_1.1-3       abind_1.4-5             \n",
       " [28] zlibbioc_1.48.0          Rtsne_0.17               purrr_1.0.2             \n",
       " [31] RCurl_1.98-1.14          GenomeInfoDbData_1.2.11  ggrepel_0.9.5           \n",
       " [34] irlba_2.3.5.1            listenv_0.9.1            spatstat.utils_3.0-4    \n",
       " [37] goftest_1.2-3            RSpectra_0.16-1          spatstat.random_3.2-3   \n",
       " [40] fitdistrplus_1.1-11      parallelly_1.37.1        leiden_0.4.3.1          \n",
       " [43] codetools_0.2-20         DelayedArray_0.26.7      RcppRoll_0.3.0          \n",
       " [46] tidyselect_1.2.1         farver_2.1.1             base64enc_0.1-3         \n",
       " [49] spatstat.explore_3.2-6   GenomicAlignments_1.36.0 jsonlite_1.8.8          \n",
       " [52] progressr_0.14.0         ggridges_0.5.6           survival_3.6-4          \n",
       " [55] tools_4.3.3              ica_1.0-3                glue_1.7.0              \n",
       " [58] IRdisplay_1.1            withr_3.0.0              fastmap_1.1.1           \n",
       " [61] rhdf5filters_1.12.1      fansi_1.0.6              digest_0.6.35           \n",
       " [64] R6_2.5.1                 mime_0.12                seqPattern_1.34.0       \n",
       " [67] colorspace_2.1-0         Cairo_1.6-2              scattermore_1.2         \n",
       " [70] tensor_1.5               spatstat.data_3.0-4      utf8_1.2.4              \n",
       " [73] tidyr_1.3.1              generics_0.1.3           httr_1.4.7              \n",
       " [76] htmlwidgets_1.6.4        S4Arrays_1.2.1           pkgconfig_2.0.3         \n",
       " [79] impute_1.76.0            lmtest_0.9-40            htmltools_0.5.8.1       \n",
       " [82] dotCall64_1.1-1          scales_1.3.0             png_0.1-8               \n",
       " [85] tzdb_0.4.0               reshape2_1.4.4           rjson_0.2.21            \n",
       " [88] uuid_1.2-0               nlme_3.1-164             repr_1.1.7              \n",
       " [91] zoo_1.8-12               KernSmooth_2.23-22       miniUI_0.1.1.1          \n",
       " [94] restfulr_0.0.15          pillar_1.9.0             vctrs_0.6.5             \n",
       " [97] RANN_2.6.1               promises_1.3.0           xtable_1.8-4            \n",
       "[100] cluster_2.1.6            evaluate_0.23            readr_2.1.5             \n",
       "[103] cli_3.6.2                compiler_4.3.3           rlang_1.1.3             \n",
       "[106] crayon_1.5.2             future.apply_1.11.2      labeling_0.4.3          \n",
       "[109] stringi_1.8.3            gridBase_0.4-7           viridisLite_0.4.2       \n",
       "[112] deldir_2.0-4             BiocParallel_1.36.0      munsell_0.5.1           \n",
       "[115] lazyeval_0.2.2           spatstat.geom_3.2-9      IRkernel_1.3.2          \n",
       "[118] RcppHNSW_0.6.0           hms_1.1.3                patchwork_1.2.0         \n",
       "[121] bit64_4.0.5              future_1.33.2            Rhdf5lib_1.22.1         \n",
       "[124] shiny_1.8.1.1            ROCR_1.0-11              igraph_2.0.3            \n",
       "[127] fastmatch_1.1-4          bit_4.0.5               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b54012ba-20d1-4787-8d0d-d851ef1744c4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
