{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: rhdf5\n",
      "\n",
      "Warning message:\n",
      "“no DISPLAY variable so Tk is not available”\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SnapATAC)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(parallel)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Xu/K562/Kendall/atac_matrix.csv.gz\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/1.prepare_data/1.seurat_pipeline.240507/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/3.Genome_wide_prediction/SnapATAC.default/SnapATAC.250212/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.atac_count = read.csv(path.matrix.atac_count,\n",
    "                             row.names = 1,\n",
    "                             check.names = F)\n",
    "matrix.atac_count = Matrix(as.matrix(matrix.atac_count), sparse = TRUE)\n",
    "matrix.atac = BinarizeCounts(matrix.atac_count)\n",
    "rm(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna_count = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna_count)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db56e579-aab1-4d6c-8a6a-ea231c11514f",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "607b378e-8627-429d-b2ed-4b77de3b9217",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2b7ec113-3336-4f1b-b71d-72d9029093d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60dce6a2-0dda-4835-aa84-8a58212c4322",
   "metadata": {},
   "source": [
    "Create SnapATAC object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6bd966aa-47a5-479d-8b34-117932f78294",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.peaks = as.data.frame(do.call(rbind,strsplit(rownames(matrix.atac),\"-\")))\n",
    "colnames(df.peaks) = c(\"chr\",\"start\",\"end\")\n",
    "bed.peaks = makeGRangesFromDataFrame(df.peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a0d07fbd-af2c-455f-b9e3-4f85b5ff3dee",
   "metadata": {},
   "outputs": [],
   "source": [
    "x.sp = createSnapFromPmat(mat = t(matrix.atac), \n",
    "                          barcodes = colnames(matrix.atac), \n",
    "                          peaks = bed.peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3eee1c80-91bc-49a1-b2be-dae1311f91ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "x.sp@gmat = t(matrix.rna.rename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ed130d1d-9653-4401-9b4b-b80c1bf55987",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "number of barcodes: 7821\n",
       "number of bins: 0\n",
       "number of genes: 18235\n",
       "number of peaks: 157600\n",
       "number of motifs: 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "x.sp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "105b7791-4cff-44c4-870c-e0078dcf6172",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 20666 ranges and 4 metadata columns:\n",
       "          seqnames            ranges strand |        name     score\n",
       "             <Rle>         <IRanges>  <Rle> | <character> <integer>\n",
       "      [1]     chr1       35831-36331      * |     FAM138A         0\n",
       "      [2]     chr1       35831-36331      * |     FAM138F         0\n",
       "      [3]     chr1       68840-69340      * |       OR4F5         0\n",
       "      [4]     chr1     817120-817620      * |      FAM87B         0\n",
       "      [5]     chr1     827272-827772      * |   LINC00115         0\n",
       "      ...      ...               ...    ... .         ...       ...\n",
       "  [20662]     chrY 21386110-21386610      * |       PRORY         0\n",
       "  [20663]     chrY 21594416-21594916      * |      TTTY13         0\n",
       "  [20664]     chrY 22298626-22299126      * |       TTTY5         0\n",
       "  [20665]     chrY 23198842-23199342      * |        DAZ1         0\n",
       "  [20666]     chrY 23219206-23219706      * |        DAZ2         0\n",
       "               Ensembl_ID      gene_type\n",
       "              <character>    <character>\n",
       "      [1] ENSG00000237613        lincRNA\n",
       "      [2] ENSG00000282591        lincRNA\n",
       "      [3] ENSG00000186092 protein_coding\n",
       "      [4] ENSG00000177757        lincRNA\n",
       "      [5] ENSG00000225880        lincRNA\n",
       "      ...             ...            ...\n",
       "  [20662] ENSG00000183146 protein_coding\n",
       "  [20663] ENSG00000184991        lincRNA\n",
       "  [20664] ENSG00000215560        lincRNA\n",
       "  [20665] ENSG00000188120 protein_coding\n",
       "  [20666] ENSG00000205944 protein_coding\n",
       "  -------\n",
       "  seqinfo: 24 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bed.tss.500 = readGeneric(\"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\",\n",
    "                          header = T,\n",
    "                          keep.all.metadata = T)\n",
    "bed.tss.500"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "045e3d33-0e97-4694-a2c9-af3171fbb0bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 20666 ranges and 4 metadata columns:\n",
       "          seqnames            ranges strand |        name     score\n",
       "             <Rle>         <IRanges>  <Rle> | <character> <integer>\n",
       "      [1]     chr1    -213919-286080      * |     FAM138A         0\n",
       "      [2]     chr1    -213919-286080      * |     FAM138F         0\n",
       "      [3]     chr1    -180910-319089      * |       OR4F5         0\n",
       "      [4]     chr1    567370-1067369      * |      FAM87B         0\n",
       "      [5]     chr1    577522-1077521      * |   LINC00115         0\n",
       "      ...      ...               ...    ... .         ...       ...\n",
       "  [20662]     chrY 21136360-21636359      * |       PRORY         0\n",
       "  [20663]     chrY 21344666-21844665      * |      TTTY13         0\n",
       "  [20664]     chrY 22048876-22548875      * |       TTTY5         0\n",
       "  [20665]     chrY 22949092-23449091      * |        DAZ1         0\n",
       "  [20666]     chrY 22969456-23469455      * |        DAZ2         0\n",
       "               Ensembl_ID      gene_type\n",
       "              <character>    <character>\n",
       "      [1] ENSG00000237613        lincRNA\n",
       "      [2] ENSG00000282591        lincRNA\n",
       "      [3] ENSG00000186092 protein_coding\n",
       "      [4] ENSG00000177757        lincRNA\n",
       "      [5] ENSG00000225880        lincRNA\n",
       "      ...             ...            ...\n",
       "  [20662] ENSG00000183146 protein_coding\n",
       "  [20663] ENSG00000184991        lincRNA\n",
       "  [20664] ENSG00000215560        lincRNA\n",
       "  [20665] ENSG00000188120 protein_coding\n",
       "  [20666] ENSG00000205944 protein_coding\n",
       "  -------\n",
       "  seqinfo: 24 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bed.tss.1Mb = resize(bed.tss.500, width=500000, fix=\"center\")\n",
    "bed.tss.1Mb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1edf75d9-df5a-41fa-82ee-e1792cea86a7",
   "metadata": {},
   "source": [
    "The original predictGenePeakPair function encounter bug when there is only one peak within TSS±500kb.\n",
    "Modify the predictGenePeakPair function to fix the bug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "afa20136-004c-486d-8dce-2a93370d679c",
   "metadata": {},
   "outputs": [],
   "source": [
    "predictGenePeakPair.modified = function (obj, input.mat = c(\"pmat\", \"bmat\"), gene.name = NULL, \n",
    "    gene.loci = NULL, do.par = FALSE, num.cores = 1) \n",
    "{\n",
    "    cat(\"Epoch: checking input parameters ... \\n\", file = stderr())\n",
    "    if (missing(obj)) {\n",
    "        stop(\"obj is missing\")\n",
    "    }\n",
    "    else {\n",
    "        if (!is(obj, \"snap\")) {\n",
    "            stop(\"obj is not a snap obj\")\n",
    "        }\n",
    "    }\n",
    "    gene.mat = obj@gmat\n",
    "    if ((x = nrow(gene.mat)) == 1L) {\n",
    "        stop(\"cell by gene matrix is empty!\")\n",
    "    }\n",
    "    if (!(gene.name %in% colnames(gene.mat))) {\n",
    "        stop(\"gene.name does not exist in the cell by gene matrix\")\n",
    "    }\n",
    "    else {\n",
    "        gene.val = gene.mat[, gene.name]\n",
    "    }\n",
    "    input.mat = match.arg(input.mat)\n",
    "    if (input.mat == \"bmat\") {\n",
    "        data.use = obj@bmat\n",
    "        peak.use = obj@feature\n",
    "    }\n",
    "    else if (input.mat == \"pmat\") {\n",
    "        data.use = obj@pmat\n",
    "        peak.use = obj@peak\n",
    "    }\n",
    "    else {\n",
    "        stop(\"input.mat does not exist in obj\")\n",
    "    }\n",
    "    if ((x = max(data.use)) > 1L) {\n",
    "        stop(\"input matrix is not binarized, run 'makeBinary' first!\")\n",
    "    }\n",
    "    if ((x = length(peak.use)) == 1L) {\n",
    "        stop(\"peak is empty!\")\n",
    "    }\n",
    "    if ((x = nrow(data.use)) == 1L) {\n",
    "        stop(\"input matrix is empty!\")\n",
    "    }\n",
    "    ncell = nrow(data.use)\n",
    "    if (!is(gene.loci, \"GRanges\")) {\n",
    "        stop(\"gene.loci is not genomic range object\")\n",
    "    }\n",
    "    else {\n",
    "        if ((x = width(gene.loci)) == 0L) {\n",
    "            stop(\"length of gene flanking window is 0\")\n",
    "        }\n",
    "    }\n",
    "    idy = queryHits(findOverlaps(peak.use, gene.loci))\n",
    "    if ((x = length(idy)) == 0L) {\n",
    "        stop(\"no peaks are found within the gene flanking window\")\n",
    "    }\n",
    "    else {\n",
    "        # data.use = data.use[, idy]\n",
    "        # add \", drop = F\" to make sure that data.use is a matrix\n",
    "        data.use = data.use[, idy, drop = F]\n",
    "        peak.use = peak.use[idy]\n",
    "        idy = which(Matrix::colSums(data.use) > 0)\n",
    "        if ((x = length(idy)) == 0L) {\n",
    "            stop(\"no peaks are found within the gene flanking window\")\n",
    "        }\n",
    "        # data.use = data.use[, idy]\n",
    "        # add \", drop = F\" to make sure that data.use is a matrix\n",
    "        data.use = data.use[, idy, drop = F]\n",
    "        peak.use = peak.use[idy]\n",
    "    }\n",
    "    cat(\"Epoch: performing statitical test ... \\n\", file = stderr())\n",
    "    if (do.par) {\n",
    "        if (num.cores > 1) {\n",
    "            if (num.cores == 1) {\n",
    "                num.cores = 1\n",
    "            }\n",
    "            else if (num.cores > detectCores()) {\n",
    "                num.cores <- detectCores() - 1\n",
    "                warning(paste0(\"num.cores set greater than number of available cores(\", \n",
    "                  parallel::detectCores(), \"). Setting num.cores to \", \n",
    "                  num.cores, \".\"))\n",
    "            }\n",
    "        }\n",
    "        else if (num.cores != 1) {\n",
    "            num.cores <- 1\n",
    "        }\n",
    "        cl <- makeCluster(num.cores, type = \"SOCK\")\n",
    "        registerDoSNOW(cl)\n",
    "        clusterEvalQ(cl, library(stats))\n",
    "        peaks.id = seq(ncol(data.use))\n",
    "        clusterExport(cl, c(\"data.use\", \"gene.val\"), envir = environment())\n",
    "        opts <- list(preschedule = TRUE)\n",
    "        clusterSetRNGStream(cl, 10)\n",
    "        models <- suppressWarnings(llply(.data = peaks.id, .fun = function(t) summary(stats::glm(y ~ \n",
    "            x, data = data.frame(y = data.use[, t], x = gene.val), \n",
    "            family = binomial(link = \"logit\")))[[\"coefficients\"]][\"x\", \n",
    "            ], .parallel = TRUE, .paropts = list(.options.snow = opts), \n",
    "            .inform = FALSE))\n",
    "        models <- do.call(rbind, models)\n",
    "        stopCluster(cl)\n",
    "    }\n",
    "    else {\n",
    "        peaks.id = seq(ncol(data.use))\n",
    "        models = lapply(peaks.id, function(t) {\n",
    "            summary(stats::glm(y ~ x, data = data.frame(y = data.use[, \n",
    "                t], x = gene.val), family = binomial(link = \"logit\")))[[\"coefficients\"]][\"x\", \n",
    "                ]\n",
    "        })\n",
    "        models <- do.call(rbind, models)\n",
    "    }\n",
    "    models[models[, \"z value\"] < 0, \"Pr(>|z|)\"] = 1\n",
    "    peak.use$beta = models[, \"Estimate\"]\n",
    "    peak.use$zvalue = models[, \"z value\"]\n",
    "    peak.use$stde = models[, \"Std. Error\"]\n",
    "    peak.use$Pval = models[, \"Pr(>|z|)\"]\n",
    "    peak.use$logPval = -log10(peak.use$Pval)\n",
    "    cat(\"Epoch: Done ... \\n\", file = stderr())\n",
    "    return(peak.use)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Run SnapATAC prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3092d715-6ee1-461d-8d70-4d622d3d9017",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "18235"
      ],
      "text/latex": [
       "18235"
      ],
      "text/markdown": [
       "18235"
      ],
      "text/plain": [
       "[1] 18235"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene.use = rownames(matrix.rna.rename[rowSums(matrix.rna.rename) > 0,])\n",
    "gene.use = gene.use[gene.use %in% bed.tss.1Mb$name]\n",
    "length(gene.use)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c2813e99-53dc-46ad-b86e-1e8b5639cb23",
   "metadata": {},
   "outputs": [],
   "source": [
    "my.cluster <- parallel::makeCluster(\n",
    "  n.cores,\n",
    "  type = \"PSOCK\"\n",
    ")\n",
    "doParallel::registerDoParallel(cl = my.cluster)\n",
    "\n",
    "pairs.E2G =  \n",
    "  foreach (gene.tmp = gene.use,\n",
    "           .combine = 'c',\n",
    "           .packages = c(\"SnapATAC\",\n",
    "                         \"genomation\",\n",
    "                         \"GenomicRanges\")) %dopar% {\n",
    "                           bed.tss.1Mb.tmp = bed.tss.1Mb[bed.tss.1Mb$name == gene.tmp]\n",
    "                           if(length(findOverlaps(bed.peaks, bed.tss.1Mb.tmp)) == 0) {\n",
    "                             pairs = NULL\n",
    "                           } else {\n",
    "                             \n",
    "                             pairs = predictGenePeakPair.modified(\n",
    "                               x.sp, \n",
    "                               input.mat=\"pmat\",\n",
    "                               gene.name=gene.tmp, \n",
    "                               gene.loci=bed.tss.1Mb.tmp,\n",
    "                               do.par=FALSE\n",
    "                             )\n",
    "                             pairs$TargetGene = gene.tmp\n",
    "                           }\n",
    "                           pairs\n",
    "                           \n",
    "                         }\n",
    "\n",
    "parallel::stopCluster(cl = my.cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5c588146-a132-4ee7-852c-f802bcbaed77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 1041014 ranges and 6 metadata columns:\n",
       "            seqnames              ranges strand |       beta     zvalue\n",
       "               <Rle>           <IRanges>  <Rle> |  <numeric>  <numeric>\n",
       "        [1]     chr1       115493-115961      * |  1.0665967  0.7975380\n",
       "        [2]     chr1       135076-135269      * |  1.9248347  1.0107396\n",
       "        [3]     chr1       136330-137277      * |  0.7032392  0.6309758\n",
       "        [4]     chr1       137692-138124      * |  0.0661497  0.0278072\n",
       "        [5]     chr1       138272-139644      * | -0.5042218 -0.3458275\n",
       "        ...      ...                 ...    ... .        ...        ...\n",
       "  [1041010]     chrX 156009722-156010137      * | -74.075252  -0.021592\n",
       "  [1041011]     chrX 156016191-156016823      * |   1.005953   0.693722\n",
       "  [1041012]     chrX 156019811-156020191      * |   2.068836   1.204477\n",
       "  [1041013]     chrX 156025028-156025591      * |  -1.559864  -0.397004\n",
       "  [1041014]     chrX 156030021-156030865      * |   0.701188   0.743172\n",
       "                   stde      Pval    logPval  TargetGene\n",
       "              <numeric> <numeric>  <numeric> <character>\n",
       "        [1]     1.33736  0.425139 0.37146942       OR4F5\n",
       "        [2]     1.90438  0.312141 0.50564908       OR4F5\n",
       "        [3]     1.11453  0.528056 0.27731971       OR4F5\n",
       "        [4]     2.37887  0.977816 0.00974291       OR4F5\n",
       "        [5]     1.45802  1.000000 0.00000000       OR4F5\n",
       "        ...         ...       ...        ...         ...\n",
       "  [1041010] 3430.675187  1.000000   0.000000        IL9R\n",
       "  [1041011]    1.450080  0.487857   0.311708        IL9R\n",
       "  [1041012]    1.717621  0.228405   0.641294        IL9R\n",
       "  [1041013]    3.929089  1.000000   0.000000        IL9R\n",
       "  [1041014]    0.943507  0.457378   0.339725        IL9R\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "58859bda-726c-4b37-89c4-85cd7985e5b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = pairs.E2G"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b9d6acb4-82b3-4941-a1b1-3b1c427bde7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(dir.output,recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 1041014 ranges and 6 metadata columns:\n",
       "            seqnames              ranges strand |       beta     zvalue\n",
       "               <Rle>           <IRanges>  <Rle> |  <numeric>  <numeric>\n",
       "        [1]     chr1       115493-115961      * |  1.0665967  0.7975380\n",
       "        [2]     chr1       135076-135269      * |  1.9248347  1.0107396\n",
       "        [3]     chr1       136330-137277      * |  0.7032392  0.6309758\n",
       "        [4]     chr1       137692-138124      * |  0.0661497  0.0278072\n",
       "        [5]     chr1       138272-139644      * | -0.5042218 -0.3458275\n",
       "        ...      ...                 ...    ... .        ...        ...\n",
       "  [1041010]     chrX 156009722-156010137      * | -74.075252  -0.021592\n",
       "  [1041011]     chrX 156016191-156016823      * |   1.005953   0.693722\n",
       "  [1041012]     chrX 156019811-156020191      * |   2.068836   1.204477\n",
       "  [1041013]     chrX 156025028-156025591      * |  -1.559864  -0.397004\n",
       "  [1041014]     chrX 156030021-156030865      * |   0.701188   0.743172\n",
       "                   stde      Pval    logPval  TargetGene\n",
       "              <numeric> <numeric>  <numeric> <character>\n",
       "        [1]     1.33736  0.425139 0.37146942       OR4F5\n",
       "        [2]     1.90438  0.312141 0.50564908       OR4F5\n",
       "        [3]     1.11453  0.528056 0.27731971       OR4F5\n",
       "        [4]     2.37887  0.977816 0.00974291       OR4F5\n",
       "        [5]     1.45802  1.000000 0.00000000       OR4F5\n",
       "        ...         ...       ...        ...         ...\n",
       "  [1041010] 3430.675187  1.000000   0.000000        IL9R\n",
       "  [1041011]    1.450080  0.487857   0.311708        IL9R\n",
       "  [1041012]    1.717621  0.228405   0.641294        IL9R\n",
       "  [1041013]    3.929089  1.000000   0.000000        IL9R\n",
       "  [1041014]    0.943507  0.457378   0.339725        IL9R\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1041014 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>beta</th><th scope=col>zvalue</th><th scope=col>stde</th><th scope=col>Pval</th><th scope=col>logPval</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 115493</td><td> 115961</td><td>OR4F5 </td><td>K562</td><td>  1.06659666</td><td> 0.79753800</td><td>   1.337362</td><td>0.42513865</td><td>0.371469416</td></tr>\n",
       "\t<tr><td>chr1</td><td> 135076</td><td> 135269</td><td>OR4F5 </td><td>K562</td><td>  1.92483470</td><td> 1.01073960</td><td>   1.904382</td><td>0.31214108</td><td>0.505649076</td></tr>\n",
       "\t<tr><td>chr1</td><td> 136330</td><td> 137277</td><td>OR4F5 </td><td>K562</td><td>  0.70323918</td><td> 0.63097575</td><td>   1.114526</td><td>0.52805638</td><td>0.277319707</td></tr>\n",
       "\t<tr><td>chr1</td><td> 137692</td><td> 138124</td><td>OR4F5 </td><td>K562</td><td>  0.06614971</td><td> 0.02780723</td><td>   2.378867</td><td>0.97781590</td><td>0.009742906</td></tr>\n",
       "\t<tr><td>chr1</td><td> 138272</td><td> 139644</td><td>OR4F5 </td><td>K562</td><td> -0.50422184</td><td>-0.34582746</td><td>   1.458016</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td>  15894</td><td>  16509</td><td>OR4F5 </td><td>K562</td><td>-70.65753028</td><td>-0.02599705</td><td>2717.905418</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td>  17271</td><td>  17713</td><td>OR4F5 </td><td>K562</td><td>-72.44242573</td><td>-0.02666247</td><td>2717.018805</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td> 180712</td><td> 181916</td><td>OR4F5 </td><td>K562</td><td>  0.60861363</td><td> 0.38847602</td><td>   1.566670</td><td>0.69766380</td><td>0.156353811</td></tr>\n",
       "\t<tr><td>chr1</td><td> 183179</td><td> 184664</td><td>OR4F5 </td><td>K562</td><td> -1.95632595</td><td>-0.66503496</td><td>   2.941689</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td> 186287</td><td> 187021</td><td>OR4F5 </td><td>K562</td><td>-71.59753613</td><td>-0.02634713</td><td>2717.470118</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td> 190808</td><td> 191845</td><td>OR4F5 </td><td>K562</td><td>  1.61145308</td><td> 0.87764585</td><td>   1.836109</td><td>0.38013594</td><td>0.420061070</td></tr>\n",
       "\t<tr><td>chr1</td><td> 203150</td><td> 203455</td><td>OR4F5 </td><td>K562</td><td>-68.04768953</td><td>-0.02502873</td><td>2718.783084</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td> 267856</td><td> 268810</td><td>OR4F5 </td><td>K562</td><td> -0.35181525</td><td>-0.16384893</td><td>   2.147193</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td> 268888</td><td> 269533</td><td>OR4F5 </td><td>K562</td><td>-71.37227734</td><td>-0.02626317</td><td>2717.580682</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td> 292088</td><td> 292440</td><td>OR4F5 </td><td>K562</td><td>-75.19576797</td><td>-0.02769546</td><td>2715.093623</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td>  29230</td><td>  29495</td><td>OR4F5 </td><td>K562</td><td>-70.24965604</td><td>-0.02584538</td><td>2718.073760</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td>   9926</td><td>  10618</td><td>OR4F5 </td><td>K562</td><td>  2.11792107</td><td> 1.92553065</td><td>   1.099916</td><td>0.05416301</td><td>1.266297247</td></tr>\n",
       "\t<tr><td>chr1</td><td>1001483</td><td>1002126</td><td>FAM87B</td><td>K562</td><td> -1.22462980</td><td>-0.64115652</td><td>   1.910033</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td>1005000</td><td>1005582</td><td>FAM87B</td><td>K562</td><td> -3.02902807</td><td>-0.72004521</td><td>   4.206719</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td>1012915</td><td>1014502</td><td>FAM87B</td><td>K562</td><td>  2.35844676</td><td> 1.52495889</td><td>   1.546564</td><td>0.12726935</td><td>0.895276171</td></tr>\n",
       "\t<tr><td>chr1</td><td>1019151</td><td>1020996</td><td>FAM87B</td><td>K562</td><td> -1.56272631</td><td>-1.06472513</td><td>   1.467727</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td>1021107</td><td>1021435</td><td>FAM87B</td><td>K562</td><td> -2.98045118</td><td>-0.70943985</td><td>   4.201133</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td>1032480</td><td>1033769</td><td>FAM87B</td><td>K562</td><td>  0.64674218</td><td> 0.44463186</td><td>   1.454557</td><td>0.65658582</td><td>0.182708499</td></tr>\n",
       "\t<tr><td>chr1</td><td>1039998</td><td>1041550</td><td>FAM87B</td><td>K562</td><td>  1.11917654</td><td> 0.76810140</td><td>   1.457069</td><td>0.44242695</td><td>0.354158429</td></tr>\n",
       "\t<tr><td>chr1</td><td>1046875</td><td>1047212</td><td>FAM87B</td><td>K562</td><td>-71.54925518</td><td>-0.02723066</td><td>2627.525179</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td>1047293</td><td>1047528</td><td>FAM87B</td><td>K562</td><td>-70.87620820</td><td>-0.02697263</td><td>2627.708142</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td>1058323</td><td>1059984</td><td>FAM87B</td><td>K562</td><td>  4.22802905</td><td> 1.65064208</td><td>   2.561445</td><td>0.09881168</td><td>1.005191707</td></tr>\n",
       "\t<tr><td>chr1</td><td>1062604</td><td>1062770</td><td>FAM87B</td><td>K562</td><td> -1.56848484</td><td>-0.38623755</td><td>   4.060933</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>chr1</td><td>1063870</td><td>1064538</td><td>FAM87B</td><td>K562</td><td>  1.98094311</td><td> 1.34312587</td><td>   1.474875</td><td>0.17923122</td><td>0.746586340</td></tr>\n",
       "\t<tr><td>chr1</td><td> 586023</td><td> 586316</td><td>FAM87B</td><td>K562</td><td>-77.20456136</td><td>-0.02941288</td><td>2624.855252</td><td>1.00000000</td><td>0.000000000</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>156008485</td><td>156009305</td><td>VAMP7</td><td>K562</td><td>  0.28081236</td><td> 1.24282650</td><td>2.259466e-01</td><td>0.2139318</td><td>0.66972471</td></tr>\n",
       "\t<tr><td>chrX</td><td>156009722</td><td>156010137</td><td>VAMP7</td><td>K562</td><td> -0.25233019</td><td>-0.80224971</td><td>3.145282e-01</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>156016191</td><td>156016823</td><td>VAMP7</td><td>K562</td><td>  0.13604546</td><td> 0.82671542</td><td>1.645614e-01</td><td>0.4083984</td><td>0.38891599</td></tr>\n",
       "\t<tr><td>chrX</td><td>156019811</td><td>156020191</td><td>VAMP7</td><td>K562</td><td>  0.38123775</td><td> 1.53188106</td><td>2.488690e-01</td><td>0.1255518</td><td>0.90117710</td></tr>\n",
       "\t<tr><td>chrX</td><td>156025028</td><td>156025591</td><td>VAMP7</td><td>K562</td><td>  0.31388638</td><td> 1.33598316</td><td>2.349479e-01</td><td>0.1815548</td><td>0.74099232</td></tr>\n",
       "\t<tr><td>chrX</td><td>156030021</td><td>156030865</td><td>VAMP7</td><td>K562</td><td>  0.01997592</td><td> 0.20519718</td><td>9.734987e-02</td><td>0.8374181</td><td>0.07705767</td></tr>\n",
       "\t<tr><td>chrX</td><td>155767101</td><td>155768235</td><td>IL9R </td><td>K562</td><td> -1.80179637</td><td>-1.03847444</td><td>1.735042e+00</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>155819988</td><td>155820456</td><td>IL9R </td><td>K562</td><td>-72.89707127</td><td>-0.03506749</td><td>2.078765e+03</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>155857665</td><td>155857903</td><td>IL9R </td><td>K562</td><td>-72.93954583</td><td>-0.02126330</td><td>3.430301e+03</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>155874510</td><td>155875164</td><td>IL9R </td><td>K562</td><td> -0.45835562</td><td>-0.14337907</td><td>3.196810e+00</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>155880679</td><td>155881916</td><td>IL9R </td><td>K562</td><td>  0.31085763</td><td> 0.39675719</td><td>7.834959e-01</td><td>0.6915465</td><td>0.16017860</td></tr>\n",
       "\t<tr><td>chrX</td><td>155887539</td><td>155888544</td><td>IL9R </td><td>K562</td><td>-71.98858777</td><td>-0.03463079</td><td>2.078745e+03</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>155889413</td><td>155889681</td><td>IL9R </td><td>K562</td><td>-74.07525222</td><td>-0.02159203</td><td>3.430675e+03</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>155893193</td><td>155894187</td><td>IL9R </td><td>K562</td><td> -0.28337634</td><td>-0.14167662</td><td>2.000163e+00</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>155944742</td><td>155945240</td><td>IL9R </td><td>K562</td><td>-74.77771243</td><td>-0.02179552</td><td>3.430875e+03</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>155957800</td><td>155958249</td><td>IL9R </td><td>K562</td><td>-72.45650043</td><td>-0.02112358</td><td>3.430124e+03</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>155958731</td><td>155959597</td><td>IL9R </td><td>K562</td><td>  0.34877397</td><td> 0.16060426</td><td>2.171636e+00</td><td>0.8724051</td><td>0.05928180</td></tr>\n",
       "\t<tr><td>chrX</td><td>155966876</td><td>155967347</td><td>IL9R </td><td>K562</td><td>-73.78141481</td><td>-0.03549331</td><td>2.078742e+03</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>155968122</td><td>155968569</td><td>IL9R </td><td>K562</td><td>-77.70123091</td><td>-0.02264400</td><td>3.431428e+03</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>155986821</td><td>155987372</td><td>IL9R </td><td>K562</td><td>  1.81535689</td><td> 0.81036635</td><td>2.240168e+00</td><td>0.4177297</td><td>0.37910470</td></tr>\n",
       "\t<tr><td>chrX</td><td>155997283</td><td>155998084</td><td>IL9R </td><td>K562</td><td> -1.11755911</td><td>-0.60778061</td><td>1.838754e+00</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>156000614</td><td>156000806</td><td>IL9R </td><td>K562</td><td>-78.77911769</td><td>-0.02295761</td><td>3.431503e+03</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>156000905</td><td>156002182</td><td>IL9R </td><td>K562</td><td>  1.18912142</td><td> 1.33629262</td><td>8.898660e-01</td><td>0.1814536</td><td>0.74123430</td></tr>\n",
       "\t<tr><td>chrX</td><td>156002257</td><td>156003190</td><td>IL9R </td><td>K562</td><td>  1.29950262</td><td> 1.25101536</td><td>1.038758e+00</td><td>0.2109289</td><td>0.67586397</td></tr>\n",
       "\t<tr><td>chrX</td><td>156008485</td><td>156009305</td><td>IL9R </td><td>K562</td><td>  0.68049080</td><td> 0.31445778</td><td>2.164013e+00</td><td>0.7531734</td><td>0.12310504</td></tr>\n",
       "\t<tr><td>chrX</td><td>156009722</td><td>156010137</td><td>IL9R </td><td>K562</td><td>-74.07525222</td><td>-0.02159203</td><td>3.430675e+03</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>156016191</td><td>156016823</td><td>IL9R </td><td>K562</td><td>  1.00595252</td><td> 0.69372195</td><td>1.450080e+00</td><td>0.4878566</td><td>0.31170782</td></tr>\n",
       "\t<tr><td>chrX</td><td>156019811</td><td>156020191</td><td>IL9R </td><td>K562</td><td>  2.06883592</td><td> 1.20447729</td><td>1.717621e+00</td><td>0.2284052</td><td>0.64129410</td></tr>\n",
       "\t<tr><td>chrX</td><td>156025028</td><td>156025591</td><td>IL9R </td><td>K562</td><td> -1.55986428</td><td>-0.39700402</td><td>3.929089e+00</td><td>1.0000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>chrX</td><td>156030021</td><td>156030865</td><td>IL9R </td><td>K562</td><td>  0.70118791</td><td> 0.74317182</td><td>9.435071e-01</td><td>0.4573777</td><td>0.33972505</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1041014 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " chr & start & end & TargetGene & CellType & beta & zvalue & stde & Pval & logPval\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  115493 &  115961 & OR4F5  & K562 &   1.06659666 &  0.79753800 &    1.337362 & 0.42513865 & 0.371469416\\\\\n",
       "\t chr1 &  135076 &  135269 & OR4F5  & K562 &   1.92483470 &  1.01073960 &    1.904382 & 0.31214108 & 0.505649076\\\\\n",
       "\t chr1 &  136330 &  137277 & OR4F5  & K562 &   0.70323918 &  0.63097575 &    1.114526 & 0.52805638 & 0.277319707\\\\\n",
       "\t chr1 &  137692 &  138124 & OR4F5  & K562 &   0.06614971 &  0.02780723 &    2.378867 & 0.97781590 & 0.009742906\\\\\n",
       "\t chr1 &  138272 &  139644 & OR4F5  & K562 &  -0.50422184 & -0.34582746 &    1.458016 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 &   15894 &   16509 & OR4F5  & K562 & -70.65753028 & -0.02599705 & 2717.905418 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 &   17271 &   17713 & OR4F5  & K562 & -72.44242573 & -0.02666247 & 2717.018805 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 &  180712 &  181916 & OR4F5  & K562 &   0.60861363 &  0.38847602 &    1.566670 & 0.69766380 & 0.156353811\\\\\n",
       "\t chr1 &  183179 &  184664 & OR4F5  & K562 &  -1.95632595 & -0.66503496 &    2.941689 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 &  186287 &  187021 & OR4F5  & K562 & -71.59753613 & -0.02634713 & 2717.470118 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 &  190808 &  191845 & OR4F5  & K562 &   1.61145308 &  0.87764585 &    1.836109 & 0.38013594 & 0.420061070\\\\\n",
       "\t chr1 &  203150 &  203455 & OR4F5  & K562 & -68.04768953 & -0.02502873 & 2718.783084 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 &  267856 &  268810 & OR4F5  & K562 &  -0.35181525 & -0.16384893 &    2.147193 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 &  268888 &  269533 & OR4F5  & K562 & -71.37227734 & -0.02626317 & 2717.580682 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 &  292088 &  292440 & OR4F5  & K562 & -75.19576797 & -0.02769546 & 2715.093623 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 &   29230 &   29495 & OR4F5  & K562 & -70.24965604 & -0.02584538 & 2718.073760 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 &    9926 &   10618 & OR4F5  & K562 &   2.11792107 &  1.92553065 &    1.099916 & 0.05416301 & 1.266297247\\\\\n",
       "\t chr1 & 1001483 & 1002126 & FAM87B & K562 &  -1.22462980 & -0.64115652 &    1.910033 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 & 1005000 & 1005582 & FAM87B & K562 &  -3.02902807 & -0.72004521 &    4.206719 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 & 1012915 & 1014502 & FAM87B & K562 &   2.35844676 &  1.52495889 &    1.546564 & 0.12726935 & 0.895276171\\\\\n",
       "\t chr1 & 1019151 & 1020996 & FAM87B & K562 &  -1.56272631 & -1.06472513 &    1.467727 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 & 1021107 & 1021435 & FAM87B & K562 &  -2.98045118 & -0.70943985 &    4.201133 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 & 1032480 & 1033769 & FAM87B & K562 &   0.64674218 &  0.44463186 &    1.454557 & 0.65658582 & 0.182708499\\\\\n",
       "\t chr1 & 1039998 & 1041550 & FAM87B & K562 &   1.11917654 &  0.76810140 &    1.457069 & 0.44242695 & 0.354158429\\\\\n",
       "\t chr1 & 1046875 & 1047212 & FAM87B & K562 & -71.54925518 & -0.02723066 & 2627.525179 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 & 1047293 & 1047528 & FAM87B & K562 & -70.87620820 & -0.02697263 & 2627.708142 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 & 1058323 & 1059984 & FAM87B & K562 &   4.22802905 &  1.65064208 &    2.561445 & 0.09881168 & 1.005191707\\\\\n",
       "\t chr1 & 1062604 & 1062770 & FAM87B & K562 &  -1.56848484 & -0.38623755 &    4.060933 & 1.00000000 & 0.000000000\\\\\n",
       "\t chr1 & 1063870 & 1064538 & FAM87B & K562 &   1.98094311 &  1.34312587 &    1.474875 & 0.17923122 & 0.746586340\\\\\n",
       "\t chr1 &  586023 &  586316 & FAM87B & K562 & -77.20456136 & -0.02941288 & 2624.855252 & 1.00000000 & 0.000000000\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 156008485 & 156009305 & VAMP7 & K562 &   0.28081236 &  1.24282650 & 2.259466e-01 & 0.2139318 & 0.66972471\\\\\n",
       "\t chrX & 156009722 & 156010137 & VAMP7 & K562 &  -0.25233019 & -0.80224971 & 3.145282e-01 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 156016191 & 156016823 & VAMP7 & K562 &   0.13604546 &  0.82671542 & 1.645614e-01 & 0.4083984 & 0.38891599\\\\\n",
       "\t chrX & 156019811 & 156020191 & VAMP7 & K562 &   0.38123775 &  1.53188106 & 2.488690e-01 & 0.1255518 & 0.90117710\\\\\n",
       "\t chrX & 156025028 & 156025591 & VAMP7 & K562 &   0.31388638 &  1.33598316 & 2.349479e-01 & 0.1815548 & 0.74099232\\\\\n",
       "\t chrX & 156030021 & 156030865 & VAMP7 & K562 &   0.01997592 &  0.20519718 & 9.734987e-02 & 0.8374181 & 0.07705767\\\\\n",
       "\t chrX & 155767101 & 155768235 & IL9R  & K562 &  -1.80179637 & -1.03847444 & 1.735042e+00 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 155819988 & 155820456 & IL9R  & K562 & -72.89707127 & -0.03506749 & 2.078765e+03 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 155857665 & 155857903 & IL9R  & K562 & -72.93954583 & -0.02126330 & 3.430301e+03 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 155874510 & 155875164 & IL9R  & K562 &  -0.45835562 & -0.14337907 & 3.196810e+00 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 155880679 & 155881916 & IL9R  & K562 &   0.31085763 &  0.39675719 & 7.834959e-01 & 0.6915465 & 0.16017860\\\\\n",
       "\t chrX & 155887539 & 155888544 & IL9R  & K562 & -71.98858777 & -0.03463079 & 2.078745e+03 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 155889413 & 155889681 & IL9R  & K562 & -74.07525222 & -0.02159203 & 3.430675e+03 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 155893193 & 155894187 & IL9R  & K562 &  -0.28337634 & -0.14167662 & 2.000163e+00 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 155944742 & 155945240 & IL9R  & K562 & -74.77771243 & -0.02179552 & 3.430875e+03 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 155957800 & 155958249 & IL9R  & K562 & -72.45650043 & -0.02112358 & 3.430124e+03 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 155958731 & 155959597 & IL9R  & K562 &   0.34877397 &  0.16060426 & 2.171636e+00 & 0.8724051 & 0.05928180\\\\\n",
       "\t chrX & 155966876 & 155967347 & IL9R  & K562 & -73.78141481 & -0.03549331 & 2.078742e+03 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 155968122 & 155968569 & IL9R  & K562 & -77.70123091 & -0.02264400 & 3.431428e+03 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 155986821 & 155987372 & IL9R  & K562 &   1.81535689 &  0.81036635 & 2.240168e+00 & 0.4177297 & 0.37910470\\\\\n",
       "\t chrX & 155997283 & 155998084 & IL9R  & K562 &  -1.11755911 & -0.60778061 & 1.838754e+00 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 156000614 & 156000806 & IL9R  & K562 & -78.77911769 & -0.02295761 & 3.431503e+03 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 156000905 & 156002182 & IL9R  & K562 &   1.18912142 &  1.33629262 & 8.898660e-01 & 0.1814536 & 0.74123430\\\\\n",
       "\t chrX & 156002257 & 156003190 & IL9R  & K562 &   1.29950262 &  1.25101536 & 1.038758e+00 & 0.2109289 & 0.67586397\\\\\n",
       "\t chrX & 156008485 & 156009305 & IL9R  & K562 &   0.68049080 &  0.31445778 & 2.164013e+00 & 0.7531734 & 0.12310504\\\\\n",
       "\t chrX & 156009722 & 156010137 & IL9R  & K562 & -74.07525222 & -0.02159203 & 3.430675e+03 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 156016191 & 156016823 & IL9R  & K562 &   1.00595252 &  0.69372195 & 1.450080e+00 & 0.4878566 & 0.31170782\\\\\n",
       "\t chrX & 156019811 & 156020191 & IL9R  & K562 &   2.06883592 &  1.20447729 & 1.717621e+00 & 0.2284052 & 0.64129410\\\\\n",
       "\t chrX & 156025028 & 156025591 & IL9R  & K562 &  -1.55986428 & -0.39700402 & 3.929089e+00 & 1.0000000 & 0.00000000\\\\\n",
       "\t chrX & 156030021 & 156030865 & IL9R  & K562 &   0.70118791 &  0.74317182 & 9.435071e-01 & 0.4573777 & 0.33972505\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1041014 × 10\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | beta &lt;dbl&gt; | zvalue &lt;dbl&gt; | stde &lt;dbl&gt; | Pval &lt;dbl&gt; | logPval &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  115493 |  115961 | OR4F5  | K562 |   1.06659666 |  0.79753800 |    1.337362 | 0.42513865 | 0.371469416 |\n",
       "| chr1 |  135076 |  135269 | OR4F5  | K562 |   1.92483470 |  1.01073960 |    1.904382 | 0.31214108 | 0.505649076 |\n",
       "| chr1 |  136330 |  137277 | OR4F5  | K562 |   0.70323918 |  0.63097575 |    1.114526 | 0.52805638 | 0.277319707 |\n",
       "| chr1 |  137692 |  138124 | OR4F5  | K562 |   0.06614971 |  0.02780723 |    2.378867 | 0.97781590 | 0.009742906 |\n",
       "| chr1 |  138272 |  139644 | OR4F5  | K562 |  -0.50422184 | -0.34582746 |    1.458016 | 1.00000000 | 0.000000000 |\n",
       "| chr1 |   15894 |   16509 | OR4F5  | K562 | -70.65753028 | -0.02599705 | 2717.905418 | 1.00000000 | 0.000000000 |\n",
       "| chr1 |   17271 |   17713 | OR4F5  | K562 | -72.44242573 | -0.02666247 | 2717.018805 | 1.00000000 | 0.000000000 |\n",
       "| chr1 |  180712 |  181916 | OR4F5  | K562 |   0.60861363 |  0.38847602 |    1.566670 | 0.69766380 | 0.156353811 |\n",
       "| chr1 |  183179 |  184664 | OR4F5  | K562 |  -1.95632595 | -0.66503496 |    2.941689 | 1.00000000 | 0.000000000 |\n",
       "| chr1 |  186287 |  187021 | OR4F5  | K562 | -71.59753613 | -0.02634713 | 2717.470118 | 1.00000000 | 0.000000000 |\n",
       "| chr1 |  190808 |  191845 | OR4F5  | K562 |   1.61145308 |  0.87764585 |    1.836109 | 0.38013594 | 0.420061070 |\n",
       "| chr1 |  203150 |  203455 | OR4F5  | K562 | -68.04768953 | -0.02502873 | 2718.783084 | 1.00000000 | 0.000000000 |\n",
       "| chr1 |  267856 |  268810 | OR4F5  | K562 |  -0.35181525 | -0.16384893 |    2.147193 | 1.00000000 | 0.000000000 |\n",
       "| chr1 |  268888 |  269533 | OR4F5  | K562 | -71.37227734 | -0.02626317 | 2717.580682 | 1.00000000 | 0.000000000 |\n",
       "| chr1 |  292088 |  292440 | OR4F5  | K562 | -75.19576797 | -0.02769546 | 2715.093623 | 1.00000000 | 0.000000000 |\n",
       "| chr1 |   29230 |   29495 | OR4F5  | K562 | -70.24965604 | -0.02584538 | 2718.073760 | 1.00000000 | 0.000000000 |\n",
       "| chr1 |    9926 |   10618 | OR4F5  | K562 |   2.11792107 |  1.92553065 |    1.099916 | 0.05416301 | 1.266297247 |\n",
       "| chr1 | 1001483 | 1002126 | FAM87B | K562 |  -1.22462980 | -0.64115652 |    1.910033 | 1.00000000 | 0.000000000 |\n",
       "| chr1 | 1005000 | 1005582 | FAM87B | K562 |  -3.02902807 | -0.72004521 |    4.206719 | 1.00000000 | 0.000000000 |\n",
       "| chr1 | 1012915 | 1014502 | FAM87B | K562 |   2.35844676 |  1.52495889 |    1.546564 | 0.12726935 | 0.895276171 |\n",
       "| chr1 | 1019151 | 1020996 | FAM87B | K562 |  -1.56272631 | -1.06472513 |    1.467727 | 1.00000000 | 0.000000000 |\n",
       "| chr1 | 1021107 | 1021435 | FAM87B | K562 |  -2.98045118 | -0.70943985 |    4.201133 | 1.00000000 | 0.000000000 |\n",
       "| chr1 | 1032480 | 1033769 | FAM87B | K562 |   0.64674218 |  0.44463186 |    1.454557 | 0.65658582 | 0.182708499 |\n",
       "| chr1 | 1039998 | 1041550 | FAM87B | K562 |   1.11917654 |  0.76810140 |    1.457069 | 0.44242695 | 0.354158429 |\n",
       "| chr1 | 1046875 | 1047212 | FAM87B | K562 | -71.54925518 | -0.02723066 | 2627.525179 | 1.00000000 | 0.000000000 |\n",
       "| chr1 | 1047293 | 1047528 | FAM87B | K562 | -70.87620820 | -0.02697263 | 2627.708142 | 1.00000000 | 0.000000000 |\n",
       "| chr1 | 1058323 | 1059984 | FAM87B | K562 |   4.22802905 |  1.65064208 |    2.561445 | 0.09881168 | 1.005191707 |\n",
       "| chr1 | 1062604 | 1062770 | FAM87B | K562 |  -1.56848484 | -0.38623755 |    4.060933 | 1.00000000 | 0.000000000 |\n",
       "| chr1 | 1063870 | 1064538 | FAM87B | K562 |   1.98094311 |  1.34312587 |    1.474875 | 0.17923122 | 0.746586340 |\n",
       "| chr1 |  586023 |  586316 | FAM87B | K562 | -77.20456136 | -0.02941288 | 2624.855252 | 1.00000000 | 0.000000000 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 156008485 | 156009305 | VAMP7 | K562 |   0.28081236 |  1.24282650 | 2.259466e-01 | 0.2139318 | 0.66972471 |\n",
       "| chrX | 156009722 | 156010137 | VAMP7 | K562 |  -0.25233019 | -0.80224971 | 3.145282e-01 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 156016191 | 156016823 | VAMP7 | K562 |   0.13604546 |  0.82671542 | 1.645614e-01 | 0.4083984 | 0.38891599 |\n",
       "| chrX | 156019811 | 156020191 | VAMP7 | K562 |   0.38123775 |  1.53188106 | 2.488690e-01 | 0.1255518 | 0.90117710 |\n",
       "| chrX | 156025028 | 156025591 | VAMP7 | K562 |   0.31388638 |  1.33598316 | 2.349479e-01 | 0.1815548 | 0.74099232 |\n",
       "| chrX | 156030021 | 156030865 | VAMP7 | K562 |   0.01997592 |  0.20519718 | 9.734987e-02 | 0.8374181 | 0.07705767 |\n",
       "| chrX | 155767101 | 155768235 | IL9R  | K562 |  -1.80179637 | -1.03847444 | 1.735042e+00 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 155819988 | 155820456 | IL9R  | K562 | -72.89707127 | -0.03506749 | 2.078765e+03 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 155857665 | 155857903 | IL9R  | K562 | -72.93954583 | -0.02126330 | 3.430301e+03 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 155874510 | 155875164 | IL9R  | K562 |  -0.45835562 | -0.14337907 | 3.196810e+00 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 155880679 | 155881916 | IL9R  | K562 |   0.31085763 |  0.39675719 | 7.834959e-01 | 0.6915465 | 0.16017860 |\n",
       "| chrX | 155887539 | 155888544 | IL9R  | K562 | -71.98858777 | -0.03463079 | 2.078745e+03 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 155889413 | 155889681 | IL9R  | K562 | -74.07525222 | -0.02159203 | 3.430675e+03 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 155893193 | 155894187 | IL9R  | K562 |  -0.28337634 | -0.14167662 | 2.000163e+00 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 155944742 | 155945240 | IL9R  | K562 | -74.77771243 | -0.02179552 | 3.430875e+03 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 155957800 | 155958249 | IL9R  | K562 | -72.45650043 | -0.02112358 | 3.430124e+03 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 155958731 | 155959597 | IL9R  | K562 |   0.34877397 |  0.16060426 | 2.171636e+00 | 0.8724051 | 0.05928180 |\n",
       "| chrX | 155966876 | 155967347 | IL9R  | K562 | -73.78141481 | -0.03549331 | 2.078742e+03 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 155968122 | 155968569 | IL9R  | K562 | -77.70123091 | -0.02264400 | 3.431428e+03 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 155986821 | 155987372 | IL9R  | K562 |   1.81535689 |  0.81036635 | 2.240168e+00 | 0.4177297 | 0.37910470 |\n",
       "| chrX | 155997283 | 155998084 | IL9R  | K562 |  -1.11755911 | -0.60778061 | 1.838754e+00 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 156000614 | 156000806 | IL9R  | K562 | -78.77911769 | -0.02295761 | 3.431503e+03 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 156000905 | 156002182 | IL9R  | K562 |   1.18912142 |  1.33629262 | 8.898660e-01 | 0.1814536 | 0.74123430 |\n",
       "| chrX | 156002257 | 156003190 | IL9R  | K562 |   1.29950262 |  1.25101536 | 1.038758e+00 | 0.2109289 | 0.67586397 |\n",
       "| chrX | 156008485 | 156009305 | IL9R  | K562 |   0.68049080 |  0.31445778 | 2.164013e+00 | 0.7531734 | 0.12310504 |\n",
       "| chrX | 156009722 | 156010137 | IL9R  | K562 | -74.07525222 | -0.02159203 | 3.430675e+03 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 156016191 | 156016823 | IL9R  | K562 |   1.00595252 |  0.69372195 | 1.450080e+00 | 0.4878566 | 0.31170782 |\n",
       "| chrX | 156019811 | 156020191 | IL9R  | K562 |   2.06883592 |  1.20447729 | 1.717621e+00 | 0.2284052 | 0.64129410 |\n",
       "| chrX | 156025028 | 156025591 | IL9R  | K562 |  -1.55986428 | -0.39700402 | 3.929089e+00 | 1.0000000 | 0.00000000 |\n",
       "| chrX | 156030021 | 156030865 | IL9R  | K562 |   0.70118791 |  0.74317182 | 9.435071e-01 | 0.4573777 | 0.33972505 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType beta         zvalue     \n",
       "1       chr1  115493    115961   OR4F5      K562       1.06659666  0.79753800\n",
       "2       chr1  135076    135269   OR4F5      K562       1.92483470  1.01073960\n",
       "3       chr1  136330    137277   OR4F5      K562       0.70323918  0.63097575\n",
       "4       chr1  137692    138124   OR4F5      K562       0.06614971  0.02780723\n",
       "5       chr1  138272    139644   OR4F5      K562      -0.50422184 -0.34582746\n",
       "6       chr1   15894     16509   OR4F5      K562     -70.65753028 -0.02599705\n",
       "7       chr1   17271     17713   OR4F5      K562     -72.44242573 -0.02666247\n",
       "8       chr1  180712    181916   OR4F5      K562       0.60861363  0.38847602\n",
       "9       chr1  183179    184664   OR4F5      K562      -1.95632595 -0.66503496\n",
       "10      chr1  186287    187021   OR4F5      K562     -71.59753613 -0.02634713\n",
       "11      chr1  190808    191845   OR4F5      K562       1.61145308  0.87764585\n",
       "12      chr1  203150    203455   OR4F5      K562     -68.04768953 -0.02502873\n",
       "13      chr1  267856    268810   OR4F5      K562      -0.35181525 -0.16384893\n",
       "14      chr1  268888    269533   OR4F5      K562     -71.37227734 -0.02626317\n",
       "15      chr1  292088    292440   OR4F5      K562     -75.19576797 -0.02769546\n",
       "16      chr1   29230     29495   OR4F5      K562     -70.24965604 -0.02584538\n",
       "17      chr1    9926     10618   OR4F5      K562       2.11792107  1.92553065\n",
       "18      chr1 1001483   1002126   FAM87B     K562      -1.22462980 -0.64115652\n",
       "19      chr1 1005000   1005582   FAM87B     K562      -3.02902807 -0.72004521\n",
       "20      chr1 1012915   1014502   FAM87B     K562       2.35844676  1.52495889\n",
       "21      chr1 1019151   1020996   FAM87B     K562      -1.56272631 -1.06472513\n",
       "22      chr1 1021107   1021435   FAM87B     K562      -2.98045118 -0.70943985\n",
       "23      chr1 1032480   1033769   FAM87B     K562       0.64674218  0.44463186\n",
       "24      chr1 1039998   1041550   FAM87B     K562       1.11917654  0.76810140\n",
       "25      chr1 1046875   1047212   FAM87B     K562     -71.54925518 -0.02723066\n",
       "26      chr1 1047293   1047528   FAM87B     K562     -70.87620820 -0.02697263\n",
       "27      chr1 1058323   1059984   FAM87B     K562       4.22802905  1.65064208\n",
       "28      chr1 1062604   1062770   FAM87B     K562      -1.56848484 -0.38623755\n",
       "29      chr1 1063870   1064538   FAM87B     K562       1.98094311  1.34312587\n",
       "30      chr1  586023    586316   FAM87B     K562     -77.20456136 -0.02941288\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮            ⋮          \n",
       "1040985 chrX 156008485 156009305 VAMP7      K562       0.28081236  1.24282650\n",
       "1040986 chrX 156009722 156010137 VAMP7      K562      -0.25233019 -0.80224971\n",
       "1040987 chrX 156016191 156016823 VAMP7      K562       0.13604546  0.82671542\n",
       "1040988 chrX 156019811 156020191 VAMP7      K562       0.38123775  1.53188106\n",
       "1040989 chrX 156025028 156025591 VAMP7      K562       0.31388638  1.33598316\n",
       "1040990 chrX 156030021 156030865 VAMP7      K562       0.01997592  0.20519718\n",
       "1040991 chrX 155767101 155768235 IL9R       K562      -1.80179637 -1.03847444\n",
       "1040992 chrX 155819988 155820456 IL9R       K562     -72.89707127 -0.03506749\n",
       "1040993 chrX 155857665 155857903 IL9R       K562     -72.93954583 -0.02126330\n",
       "1040994 chrX 155874510 155875164 IL9R       K562      -0.45835562 -0.14337907\n",
       "1040995 chrX 155880679 155881916 IL9R       K562       0.31085763  0.39675719\n",
       "1040996 chrX 155887539 155888544 IL9R       K562     -71.98858777 -0.03463079\n",
       "1040997 chrX 155889413 155889681 IL9R       K562     -74.07525222 -0.02159203\n",
       "1040998 chrX 155893193 155894187 IL9R       K562      -0.28337634 -0.14167662\n",
       "1040999 chrX 155944742 155945240 IL9R       K562     -74.77771243 -0.02179552\n",
       "1041000 chrX 155957800 155958249 IL9R       K562     -72.45650043 -0.02112358\n",
       "1041001 chrX 155958731 155959597 IL9R       K562       0.34877397  0.16060426\n",
       "1041002 chrX 155966876 155967347 IL9R       K562     -73.78141481 -0.03549331\n",
       "1041003 chrX 155968122 155968569 IL9R       K562     -77.70123091 -0.02264400\n",
       "1041004 chrX 155986821 155987372 IL9R       K562       1.81535689  0.81036635\n",
       "1041005 chrX 155997283 155998084 IL9R       K562      -1.11755911 -0.60778061\n",
       "1041006 chrX 156000614 156000806 IL9R       K562     -78.77911769 -0.02295761\n",
       "1041007 chrX 156000905 156002182 IL9R       K562       1.18912142  1.33629262\n",
       "1041008 chrX 156002257 156003190 IL9R       K562       1.29950262  1.25101536\n",
       "1041009 chrX 156008485 156009305 IL9R       K562       0.68049080  0.31445778\n",
       "1041010 chrX 156009722 156010137 IL9R       K562     -74.07525222 -0.02159203\n",
       "1041011 chrX 156016191 156016823 IL9R       K562       1.00595252  0.69372195\n",
       "1041012 chrX 156019811 156020191 IL9R       K562       2.06883592  1.20447729\n",
       "1041013 chrX 156025028 156025591 IL9R       K562      -1.55986428 -0.39700402\n",
       "1041014 chrX 156030021 156030865 IL9R       K562       0.70118791  0.74317182\n",
       "        stde         Pval       logPval    \n",
       "1          1.337362  0.42513865 0.371469416\n",
       "2          1.904382  0.31214108 0.505649076\n",
       "3          1.114526  0.52805638 0.277319707\n",
       "4          2.378867  0.97781590 0.009742906\n",
       "5          1.458016  1.00000000 0.000000000\n",
       "6       2717.905418  1.00000000 0.000000000\n",
       "7       2717.018805  1.00000000 0.000000000\n",
       "8          1.566670  0.69766380 0.156353811\n",
       "9          2.941689  1.00000000 0.000000000\n",
       "10      2717.470118  1.00000000 0.000000000\n",
       "11         1.836109  0.38013594 0.420061070\n",
       "12      2718.783084  1.00000000 0.000000000\n",
       "13         2.147193  1.00000000 0.000000000\n",
       "14      2717.580682  1.00000000 0.000000000\n",
       "15      2715.093623  1.00000000 0.000000000\n",
       "16      2718.073760  1.00000000 0.000000000\n",
       "17         1.099916  0.05416301 1.266297247\n",
       "18         1.910033  1.00000000 0.000000000\n",
       "19         4.206719  1.00000000 0.000000000\n",
       "20         1.546564  0.12726935 0.895276171\n",
       "21         1.467727  1.00000000 0.000000000\n",
       "22         4.201133  1.00000000 0.000000000\n",
       "23         1.454557  0.65658582 0.182708499\n",
       "24         1.457069  0.44242695 0.354158429\n",
       "25      2627.525179  1.00000000 0.000000000\n",
       "26      2627.708142  1.00000000 0.000000000\n",
       "27         2.561445  0.09881168 1.005191707\n",
       "28         4.060933  1.00000000 0.000000000\n",
       "29         1.474875  0.17923122 0.746586340\n",
       "30      2624.855252  1.00000000 0.000000000\n",
       "⋮       ⋮            ⋮          ⋮          \n",
       "1040985 2.259466e-01 0.2139318  0.66972471 \n",
       "1040986 3.145282e-01 1.0000000  0.00000000 \n",
       "1040987 1.645614e-01 0.4083984  0.38891599 \n",
       "1040988 2.488690e-01 0.1255518  0.90117710 \n",
       "1040989 2.349479e-01 0.1815548  0.74099232 \n",
       "1040990 9.734987e-02 0.8374181  0.07705767 \n",
       "1040991 1.735042e+00 1.0000000  0.00000000 \n",
       "1040992 2.078765e+03 1.0000000  0.00000000 \n",
       "1040993 3.430301e+03 1.0000000  0.00000000 \n",
       "1040994 3.196810e+00 1.0000000  0.00000000 \n",
       "1040995 7.834959e-01 0.6915465  0.16017860 \n",
       "1040996 2.078745e+03 1.0000000  0.00000000 \n",
       "1040997 3.430675e+03 1.0000000  0.00000000 \n",
       "1040998 2.000163e+00 1.0000000  0.00000000 \n",
       "1040999 3.430875e+03 1.0000000  0.00000000 \n",
       "1041000 3.430124e+03 1.0000000  0.00000000 \n",
       "1041001 2.171636e+00 0.8724051  0.05928180 \n",
       "1041002 2.078742e+03 1.0000000  0.00000000 \n",
       "1041003 3.431428e+03 1.0000000  0.00000000 \n",
       "1041004 2.240168e+00 0.4177297  0.37910470 \n",
       "1041005 1.838754e+00 1.0000000  0.00000000 \n",
       "1041006 3.431503e+03 1.0000000  0.00000000 \n",
       "1041007 8.898660e-01 0.1814536  0.74123430 \n",
       "1041008 1.038758e+00 0.2109289  0.67586397 \n",
       "1041009 2.164013e+00 0.7531734  0.12310504 \n",
       "1041010 3.430675e+03 1.0000000  0.00000000 \n",
       "1041011 1.450080e+00 0.4878566  0.31170782 \n",
       "1041012 1.717621e+00 0.2284052  0.64129410 \n",
       "1041013 3.929089e+00 1.0000000  0.00000000 \n",
       "1041014 9.435071e-01 0.4573777  0.33972505 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"beta\",\n",
    "                         \"zvalue\",\n",
    "                         \"stde\",\n",
    "                         \"Pval\",\n",
    "                         \"logPval\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] data.table_1.15.2    dplyr_1.1.4          foreach_1.5.2       \n",
       " [4] GenomicRanges_1.54.1 GenomeInfoDb_1.38.1  IRanges_2.36.0      \n",
       " [7] S4Vectors_0.40.2     BiocGenerics_0.48.1  genomation_1.34.0   \n",
       "[10] Signac_1.13.0        Seurat_5.0.3         SeuratObject_5.0.1  \n",
       "[13] sp_2.1-4             SnapATAC_1.0.0       rhdf5_2.44.0        \n",
       "[16] Matrix_1.6-5        \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 BiocIO_1.10.0              \n",
       "  [5] pbdZMQ_0.3-11               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                polyclip_1.10-6            \n",
       "  [9] XML_3.99-0.16.1             fastDummies_1.7.3          \n",
       " [11] lifecycle_1.0.4             tcltk_4.3.3                \n",
       " [13] edgeR_3.42.4                doParallel_1.0.17          \n",
       " [15] vroom_1.6.5                 globals_0.16.3             \n",
       " [17] lattice_0.22-6              MASS_7.3-60                \n",
       " [19] magrittr_2.0.3              limma_3.56.2               \n",
       " [21] plotly_4.10.4               plotrix_3.8-4              \n",
       " [23] yaml_2.3.8                  plot3D_1.4.1               \n",
       " [25] httpuv_1.6.15               sctransform_0.4.1          \n",
       " [27] spam_2.10-0                 spatstat.sparse_3.0-3      \n",
       " [29] reticulate_1.36.1           cowplot_1.1.3              \n",
       " [31] pbapply_1.7-2               RColorBrewer_1.1-3         \n",
       " [33] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [35] Rtsne_0.17                  purrr_1.0.2                \n",
       " [37] RCurl_1.98-1.14             misc3d_0.9-1               \n",
       " [39] GenomeInfoDbData_1.2.11     ggrepel_0.9.5              \n",
       " [41] irlba_2.3.5.1               listenv_0.9.1              \n",
       " [43] spatstat.utils_3.0-4        goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             bigmemory_4.6.4            \n",
       " [47] spatstat.random_3.2-3       fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           DelayedArray_0.26.7        \n",
       " [51] leiden_0.4.3.1              codetools_0.2-20           \n",
       " [53] RcppRoll_0.3.0              tidyselect_1.2.1           \n",
       " [55] viridis_0.6.5               matrixStats_1.3.0          \n",
       " [57] base64enc_0.1-3             spatstat.explore_3.2-6     \n",
       " [59] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [61] progressr_0.14.0            ggridges_0.5.6             \n",
       " [63] survival_3.6-4              iterators_1.0.14           \n",
       " [65] tools_4.3.3                 snow_0.4-4                 \n",
       " [67] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [69] glue_1.7.0                  gridExtra_2.3              \n",
       " [71] MatrixGenerics_1.12.3       IRdisplay_1.1              \n",
       " [73] withr_3.0.0                 fastmap_1.1.1              \n",
       " [75] rhdf5filters_1.12.1         fansi_1.0.6                \n",
       " [77] digest_0.6.35               R6_2.5.1                   \n",
       " [79] mime_0.12                   seqPattern_1.34.0          \n",
       " [81] colorspace_2.1-0            scattermore_1.2            \n",
       " [83] tensor_1.5                  spatstat.data_3.0-4        \n",
       " [85] utf8_1.2.4                  tidyr_1.3.1                \n",
       " [87] generics_0.1.3              rtracklayer_1.60.1         \n",
       " [89] S4Arrays_1.2.1              httr_1.4.7                 \n",
       " [91] htmlwidgets_1.6.4           uwot_0.1.16                \n",
       " [93] pkgconfig_2.0.3             gtable_0.3.5               \n",
       " [95] impute_1.76.0               lmtest_0.9-40              \n",
       " [97] XVector_0.42.0              htmltools_0.5.8.1          \n",
       " [99] dotCall64_1.1-1             Biobase_2.60.0             \n",
       "[101] scales_1.3.0                png_0.1-8                  \n",
       "[103] doSNOW_1.0.20               bigmemory.sri_0.1.8        \n",
       "[105] tzdb_0.4.0                  rjson_0.2.21               \n",
       "[107] reshape2_1.4.4              uuid_1.2-0                 \n",
       "[109] nlme_3.1-164                repr_1.1.7                 \n",
       "[111] zoo_1.8-12                  stringr_1.5.1              \n",
       "[113] KernSmooth_2.23-22          miniUI_0.1.1.1             \n",
       "[115] restfulr_0.0.15             pillar_1.9.0               \n",
       "[117] vctrs_0.6.5                 RANN_2.6.1                 \n",
       "[119] promises_1.3.0              xtable_1.8-4               \n",
       "[121] cluster_2.1.6               evaluate_0.23              \n",
       "[123] readr_2.1.5                 cli_3.6.2                  \n",
       "[125] locfit_1.5-9.9              compiler_4.3.3             \n",
       "[127] Rsamtools_2.18.0            rlang_1.1.3                \n",
       "[129] crayon_1.5.2                future.apply_1.11.2        \n",
       "[131] plyr_1.8.9                  stringi_1.8.3              \n",
       "[133] gridBase_0.4-7              viridisLite_0.4.2          \n",
       "[135] deldir_2.0-4                BiocParallel_1.36.0        \n",
       "[137] munsell_0.5.1               Biostrings_2.70.1          \n",
       "[139] lazyeval_0.2.2              spatstat.geom_3.2-9        \n",
       "[141] IRkernel_1.3.2              BSgenome_1.68.0            \n",
       "[143] RcppHNSW_0.6.0              hms_1.1.3                  \n",
       "[145] patchwork_1.2.0             bit64_4.0.5                \n",
       "[147] future_1.33.2               ggplot2_3.5.1              \n",
       "[149] Rhdf5lib_1.22.1             shiny_1.8.1.1              \n",
       "[151] SummarizedExperiment_1.30.2 ROCR_1.0-11                \n",
       "[153] igraph_2.0.3                fastmatch_1.1-4            \n",
       "[155] bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
