{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: rhdf5\n",
      "\n",
      "Warning message:\n",
      "“no DISPLAY variable so Tk is not available”\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SnapATAC)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(parallel)\n",
    "library(foreach)\n",
    "Sys.setenv(OPENBLAS_NUM_THREADS=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/5.merge_peaks_from_5_super_groups.240524/pairs.rds\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/5.merge_peaks_from_5_super_groups.240524/matrix.atac.5_super_groups.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/5.merge_peaks_from_5_super_groups.240524/matrix.rna.5_super_groups.rds\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/SnapATAC/BMMC5.merge.250212/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readRDS(path.pairs.E2G)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = BinarizeCounts(matrix.atac_count)\n",
    "rm(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = readRDS(path.matrix.rna_count)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna_count = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna_count)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4986774c-cbbf-4b72-a299-d681bddca80e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "85b165d0-f1db-4548-beac-235f8950107b",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "655ebba3-731e-49d8-a04e-1ebb5becd567",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter = pairs.E2G[pairs.E2G$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                             pairs.E2G$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Run SnapATAC prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PredictionSnapATAC = function(pairs.E2G,\n",
    "                              data.RNA,\n",
    "                              data.ATAC,\n",
    "                              n.core = 16){\n",
    "  my.cluster <- parallel::makeCluster(\n",
    "    n.core,\n",
    "    type = \"PSOCK\"\n",
    "  )\n",
    "  doParallel::registerDoParallel(cl = my.cluster)\n",
    "  \n",
    "  pairs.E2G =  \n",
    "    foreach (index.tmp = 1:length(pairs.E2G),\n",
    "             .combine = 'c',\n",
    "             .packages = c(\"Matrix\",\n",
    "                           \"GenomicRanges\")) %dopar% {\n",
    "\n",
    "                             pairs.E2G.tmp = pairs.E2G[index.tmp]\n",
    "                             \n",
    "                             model.tmp = summary(stats::glm(data.ATAC[pairs.E2G.tmp$PeakName,] ~ data.RNA[pairs.E2G.tmp$TargetGene,], \n",
    "                                                            family = binomial(link = \"logit\")))[[\"coefficients\"]]\n",
    "                             mcols(pairs.E2G.tmp)[,c(\"beta\",\"sd\",\"z\",\"p\")] = \n",
    "                               t(data.frame(model.tmp[2,]))\n",
    "                             \n",
    "                             pairs.E2G.tmp\n",
    "                           }\n",
    "  \n",
    "  parallel::stopCluster(cl = my.cluster)\n",
    "  \n",
    "  return(pairs.E2G)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(dir.output, \"chr\", sep = \"/\"), recursive = T):\n",
      "“'/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/SnapATAC/BMMC5.merge.250212//chr' already exists”\n"
     ]
    }
   ],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "[1] \"chr10\"\n",
      "[1] \"chr11\"\n",
      "[1] \"chr12\"\n",
      "[1] \"chr13\"\n",
      "[1] \"chr14\"\n",
      "[1] \"chr15\"\n",
      "[1] \"chr16\"\n",
      "[1] \"chr17\"\n",
      "[1] \"chr18\"\n",
      "[1] \"chr19\"\n",
      "[1] \"chr2\"\n",
      "[1] \"chr20\"\n",
      "[1] \"chr21\"\n",
      "[1] \"chr22\"\n",
      "[1] \"chr3\"\n",
      "[1] \"chr4\"\n",
      "[1] \"chr5\"\n",
      "[1] \"chr6\"\n",
      "[1] \"chr7\"\n",
      "[1] \"chr8\"\n",
      "[1] \"chr9\"\n",
      "[1] \"chrX\"\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter[seqnames(pairs.E2G.filter) == chr.tmp]\n",
    "  mcols(pairs.E2G.chr.res)[,c(\"beta\",\"sd\",\"z\",\"p\")] = NA\n",
    "  pairs.E2G.chr.res = PredictionSnapATAC(pairs.E2G.chr.res,\n",
    "                                         matrix.rna.rename[rownames(matrix.rna.rename) %in% pairs.E2G.chr.res$TargetGene,],\n",
    "                                         matrix.atac[rownames(matrix.atac) %in% pairs.E2G.chr.res$PeakName,],\n",
    "                                         n.cores)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 5660858 ranges and 7 metadata columns:\n",
       "       seqnames              ranges strand |  TargetGene               PeakName\n",
       "          <Rle>           <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 100015450-100016026      * |         AGL chr1-100015450-10001..\n",
       "  chr1     chr1 100015450-100016026      * |      CDC14A chr1-100015450-10001..\n",
       "  chr1     chr1 100015450-100016026      * |         DBT chr1-100015450-10001..\n",
       "  chr1     chr1 100015450-100016026      * |       EXTL2 chr1-100015450-10001..\n",
       "  chr1     chr1 100015450-100016026      * |     MFSD14A chr1-100015450-10001..\n",
       "   ...      ...                 ...    ... .         ...                    ...\n",
       "  chrX     chrX     9995615-9996434      * |        WWC3   chrX-9995615-9996434\n",
       "  chrX     chrX     9997010-9998101      * |       CLCN4   chrX-9997010-9998101\n",
       "  chrX     chrX     9997010-9998101      * |     SHROOM2   chrX-9997010-9998101\n",
       "  chrX     chrX     9997010-9998101      * |       TBL1X   chrX-9997010-9998101\n",
       "  chrX     chrX     9997010-9998101      * |        WWC3   chrX-9997010-9998101\n",
       "                     PairName       beta        sd         z           p\n",
       "                  <character>  <numeric> <numeric> <numeric>   <numeric>\n",
       "  chr1 chr1-100015450-10001..   0.186203 0.1439239  1.293759   0.1957487\n",
       "  chr1 chr1-100015450-10001..   0.133354 0.0686637  1.942138   0.0521204\n",
       "  chr1 chr1-100015450-10001..  -0.411852 0.2881150 -1.429470   0.1528692\n",
       "  chr1 chr1-100015450-10001..   0.265307 0.3061121  0.866698   0.3861077\n",
       "  chr1 chr1-100015450-10001..   0.163397 0.4894818  0.333816   0.7385182\n",
       "   ...                    ...        ...       ...       ...         ...\n",
       "  chrX chrX-9995615-9996434..  0.0563888 0.0438936  1.284672 1.98907e-01\n",
       "  chrX chrX-9997010-9998101.. -0.2671559 0.1567812 -1.704004 8.83804e-02\n",
       "  chrX chrX-9997010-9998101..  0.0870743 0.4031991  0.215959 8.29020e-01\n",
       "  chrX chrX-9997010-9998101..  0.2191688 0.0371879  5.893558 3.77968e-09\n",
       "  chrX chrX-9997010-9998101..  0.0278290 0.0806543  0.345041 7.30064e-01\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5660858 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>beta</th><th scope=col>sd</th><th scope=col>z</th><th scope=col>p</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>AGL     </td><td>K562</td><td> 0.186202797</td><td>0.143923884</td><td> 1.29375884</td><td>1.957487e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>CDC14A  </td><td>K562</td><td> 0.133354374</td><td>0.068663708</td><td> 1.94213768</td><td>5.212043e-02</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>DBT     </td><td>K562</td><td>-0.411851772</td><td>0.288115031</td><td>-1.42946992</td><td>1.528692e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>EXTL2   </td><td>K562</td><td> 0.265306585</td><td>0.306112058</td><td> 0.86669760</td><td>3.861077e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>MFSD14A </td><td>K562</td><td> 0.163396983</td><td>0.489481785</td><td> 0.33381627</td><td>7.385182e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>RTCA    </td><td>K562</td><td> 0.090986595</td><td>0.136889000</td><td> 0.66467426</td><td>5.062589e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>RTCA-AS1</td><td>K562</td><td> 0.014138722</td><td>0.413738974</td><td> 0.03417305</td><td>9.727392e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>SASS6   </td><td>K562</td><td> 0.015875834</td><td>0.183221132</td><td> 0.08664849</td><td>9.309509e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>SLC30A7 </td><td>K562</td><td> 0.101410986</td><td>0.100720610</td><td> 1.00685437</td><td>3.140048e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>SLC35A3 </td><td>K562</td><td>-0.087917219</td><td>0.224578483</td><td>-0.39147659</td><td>6.954450e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>TRMT13  </td><td>K562</td><td>-0.113919226</td><td>0.186336451</td><td>-0.61136308</td><td>5.409592e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>AGL     </td><td>K562</td><td> 0.187385214</td><td>0.042191001</td><td> 4.44135504</td><td>8.939416e-06</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>CDC14A  </td><td>K562</td><td> 0.097553272</td><td>0.020248843</td><td> 4.81772071</td><td>1.452074e-06</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>DBT     </td><td>K562</td><td> 0.032055857</td><td>0.055813799</td><td> 0.57433570</td><td>5.657407e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>DPH5    </td><td>K562</td><td> 0.149823444</td><td>0.057667647</td><td> 2.59805022</td><td>9.375478e-03</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>EXTL2   </td><td>K562</td><td> 0.114501301</td><td>0.102801945</td><td> 1.11380481</td><td>2.653629e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>FRRS1   </td><td>K562</td><td>-0.770525949</td><td>0.147181099</td><td>-5.23522352</td><td>1.647853e-07</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>MFSD14A </td><td>K562</td><td> 0.196956118</td><td>0.139249995</td><td> 1.41440665</td><td>1.572425e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>RTCA    </td><td>K562</td><td> 0.090545549</td><td>0.039855502</td><td> 2.27184567</td><td>2.309583e-02</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>RTCA-AS1</td><td>K562</td><td> 0.034972974</td><td>0.117734570</td><td> 0.29704933</td><td>7.664289e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>SASS6   </td><td>K562</td><td>-0.096079077</td><td>0.058327134</td><td>-1.64724495</td><td>9.950771e-02</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>SLC30A7 </td><td>K562</td><td> 0.007023095</td><td>0.031200077</td><td> 0.22509864</td><td>8.219025e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>SLC35A3 </td><td>K562</td><td> 0.121783939</td><td>0.054457127</td><td> 2.23632690</td><td>2.533037e-02</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>TRMT13  </td><td>K562</td><td> 0.025684388</td><td>0.047758628</td><td> 0.53779575</td><td>5.907181e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>VCAM1   </td><td>K562</td><td> 0.544125314</td><td>0.143223236</td><td> 3.79914134</td><td>1.451983e-04</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>AGL     </td><td>K562</td><td> 0.072809383</td><td>0.018596372</td><td> 3.91524669</td><td>9.031182e-05</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>CDC14A  </td><td>K562</td><td> 0.093677870</td><td>0.007995935</td><td>11.71568697</td><td>1.059291e-31</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>DBT     </td><td>K562</td><td> 0.025361401</td><td>0.021810102</td><td> 1.16282821</td><td>2.448992e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>DPH5    </td><td>K562</td><td> 0.112824614</td><td>0.024391368</td><td> 4.62559590</td><td>3.735229e-06</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>EXTL2   </td><td>K562</td><td> 0.080957578</td><td>0.042460253</td><td> 1.90666738</td><td>5.656367e-02</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9986538</td><td>9987289</td><td>CLCN4    </td><td>K562</td><td>-1.608619e-02</td><td>  0.09986950</td><td>-0.16107214</td><td>8.720366e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>9986538</td><td>9987289</td><td>SHROOM2  </td><td>K562</td><td> 5.488062e-01</td><td>  0.20953585</td><td> 2.61915174</td><td>8.814873e-03</td></tr>\n",
       "\t<tr><td>chrX</td><td>9986538</td><td>9987289</td><td>TBL1X    </td><td>K562</td><td> 3.825249e-01</td><td>  0.02841623</td><td>13.46149333</td><td>2.635357e-41</td></tr>\n",
       "\t<tr><td>chrX</td><td>9986538</td><td>9987289</td><td>WWC3     </td><td>K562</td><td>-3.905878e-02</td><td>  0.06888749</td><td>-0.56699380</td><td>5.707184e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>9987345</td><td>9987562</td><td>CLCN4    </td><td>K562</td><td> 1.773150e-02</td><td>  0.25764974</td><td> 0.06882019</td><td>9.451327e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>9987345</td><td>9987562</td><td>TBL1X    </td><td>K562</td><td> 1.673476e-01</td><td>  0.08150312</td><td> 2.05326592</td><td>4.004679e-02</td></tr>\n",
       "\t<tr><td>chrX</td><td>9987345</td><td>9987562</td><td>WWC3     </td><td>K562</td><td>-3.615037e-01</td><td>  0.24597549</td><td>-1.46967364</td><td>1.416502e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988628</td><td>9989100</td><td>CLCN4    </td><td>K562</td><td> 1.629022e-01</td><td>  0.15046136</td><td> 1.08268456</td><td>2.789485e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988628</td><td>9989100</td><td>SHROOM2  </td><td>K562</td><td> 5.959272e-01</td><td>  0.34787510</td><td> 1.71304925</td><td>8.670348e-02</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988628</td><td>9989100</td><td>TBL1X    </td><td>K562</td><td> 1.469698e-01</td><td>  0.05523129</td><td> 2.66098841</td><td>7.791164e-03</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988628</td><td>9989100</td><td>WWC3     </td><td>K562</td><td> 1.603043e-02</td><td>  0.11726105</td><td> 0.13670718</td><td>8.912623e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>AKAP17A  </td><td>K562</td><td> 1.270236e-01</td><td>  0.12520607</td><td> 1.01451612</td><td>3.103365e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>ASMT     </td><td>K562</td><td>-1.191942e+01</td><td>299.03078684</td><td>-0.03986019</td><td>9.682046e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>ASMTL    </td><td>K562</td><td> 2.831859e-01</td><td>  0.12091872</td><td> 2.34195278</td><td>1.918314e-02</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>CSF2RA   </td><td>K562</td><td>-2.039934e-01</td><td>  0.13851564</td><td>-1.47271042</td><td>1.408291e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>GTPBP6   </td><td>K562</td><td>-6.236828e-02</td><td>  0.18538232</td><td>-0.33643058</td><td>7.365462e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>IL3RA    </td><td>K562</td><td> 2.351789e-01</td><td>  0.12519168</td><td> 1.87855035</td><td>6.030592e-02</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>LINC00106</td><td>K562</td><td> 1.519153e-01</td><td>  0.21182872</td><td> 0.71716113</td><td>4.732747e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>P2RY8    </td><td>K562</td><td> 8.787939e-03</td><td>  0.06983985</td><td> 0.12582986</td><td>8.998666e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>PLCXD1   </td><td>K562</td><td>-1.794655e-01</td><td>  0.26096875</td><td>-0.68768965</td><td>4.916482e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>PPP2R3B  </td><td>K562</td><td> 5.871419e-01</td><td>  0.14631624</td><td> 4.01282780</td><td>5.999568e-05</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>SLC25A6  </td><td>K562</td><td>-1.827568e-01</td><td>  0.08911021</td><td>-2.05090697</td><td>4.027601e-02</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995615</td><td>9996434</td><td>CLCN4    </td><td>K562</td><td>-9.818911e-04</td><td>  0.06801909</td><td>-0.01443552</td><td>9.884825e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995615</td><td>9996434</td><td>SHROOM2  </td><td>K562</td><td> 9.887375e-02</td><td>  0.22303112</td><td> 0.44331818</td><td>6.575356e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995615</td><td>9996434</td><td>TBL1X    </td><td>K562</td><td> 3.913810e-02</td><td>  0.02265325</td><td> 1.72770334</td><td>8.404142e-02</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995615</td><td>9996434</td><td>WWC3     </td><td>K562</td><td> 5.638883e-02</td><td>  0.04389357</td><td> 1.28467185</td><td>1.989070e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997010</td><td>9998101</td><td>CLCN4    </td><td>K562</td><td>-2.671559e-01</td><td>  0.15678124</td><td>-1.70400394</td><td>8.838035e-02</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997010</td><td>9998101</td><td>SHROOM2  </td><td>K562</td><td> 8.707427e-02</td><td>  0.40319907</td><td> 0.21595852</td><td>8.290201e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997010</td><td>9998101</td><td>TBL1X    </td><td>K562</td><td> 2.191688e-01</td><td>  0.03718786</td><td> 5.89355786</td><td>3.779677e-09</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997010</td><td>9998101</td><td>WWC3     </td><td>K562</td><td> 2.782902e-02</td><td>  0.08065427</td><td> 0.34504089</td><td>7.300636e-01</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5660858 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & beta & sd & z & p\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100015450 & 100016026 & AGL      & K562 &  0.186202797 & 0.143923884 &  1.29375884 & 1.957487e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & CDC14A   & K562 &  0.133354374 & 0.068663708 &  1.94213768 & 5.212043e-02\\\\\n",
       "\t chr1 & 100015450 & 100016026 & DBT      & K562 & -0.411851772 & 0.288115031 & -1.42946992 & 1.528692e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & EXTL2    & K562 &  0.265306585 & 0.306112058 &  0.86669760 & 3.861077e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & MFSD14A  & K562 &  0.163396983 & 0.489481785 &  0.33381627 & 7.385182e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & RTCA     & K562 &  0.090986595 & 0.136889000 &  0.66467426 & 5.062589e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & RTCA-AS1 & K562 &  0.014138722 & 0.413738974 &  0.03417305 & 9.727392e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & SASS6    & K562 &  0.015875834 & 0.183221132 &  0.08664849 & 9.309509e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & SLC30A7  & K562 &  0.101410986 & 0.100720610 &  1.00685437 & 3.140048e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & SLC35A3  & K562 & -0.087917219 & 0.224578483 & -0.39147659 & 6.954450e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & TRMT13   & K562 & -0.113919226 & 0.186336451 & -0.61136308 & 5.409592e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & AGL      & K562 &  0.187385214 & 0.042191001 &  4.44135504 & 8.939416e-06\\\\\n",
       "\t chr1 & 100028395 & 100029373 & CDC14A   & K562 &  0.097553272 & 0.020248843 &  4.81772071 & 1.452074e-06\\\\\n",
       "\t chr1 & 100028395 & 100029373 & DBT      & K562 &  0.032055857 & 0.055813799 &  0.57433570 & 5.657407e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & DPH5     & K562 &  0.149823444 & 0.057667647 &  2.59805022 & 9.375478e-03\\\\\n",
       "\t chr1 & 100028395 & 100029373 & EXTL2    & K562 &  0.114501301 & 0.102801945 &  1.11380481 & 2.653629e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & FRRS1    & K562 & -0.770525949 & 0.147181099 & -5.23522352 & 1.647853e-07\\\\\n",
       "\t chr1 & 100028395 & 100029373 & MFSD14A  & K562 &  0.196956118 & 0.139249995 &  1.41440665 & 1.572425e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & RTCA     & K562 &  0.090545549 & 0.039855502 &  2.27184567 & 2.309583e-02\\\\\n",
       "\t chr1 & 100028395 & 100029373 & RTCA-AS1 & K562 &  0.034972974 & 0.117734570 &  0.29704933 & 7.664289e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & SASS6    & K562 & -0.096079077 & 0.058327134 & -1.64724495 & 9.950771e-02\\\\\n",
       "\t chr1 & 100028395 & 100029373 & SLC30A7  & K562 &  0.007023095 & 0.031200077 &  0.22509864 & 8.219025e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & SLC35A3  & K562 &  0.121783939 & 0.054457127 &  2.23632690 & 2.533037e-02\\\\\n",
       "\t chr1 & 100028395 & 100029373 & TRMT13   & K562 &  0.025684388 & 0.047758628 &  0.53779575 & 5.907181e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & VCAM1    & K562 &  0.544125314 & 0.143223236 &  3.79914134 & 1.451983e-04\\\\\n",
       "\t chr1 & 100037560 & 100039288 & AGL      & K562 &  0.072809383 & 0.018596372 &  3.91524669 & 9.031182e-05\\\\\n",
       "\t chr1 & 100037560 & 100039288 & CDC14A   & K562 &  0.093677870 & 0.007995935 & 11.71568697 & 1.059291e-31\\\\\n",
       "\t chr1 & 100037560 & 100039288 & DBT      & K562 &  0.025361401 & 0.021810102 &  1.16282821 & 2.448992e-01\\\\\n",
       "\t chr1 & 100037560 & 100039288 & DPH5     & K562 &  0.112824614 & 0.024391368 &  4.62559590 & 3.735229e-06\\\\\n",
       "\t chr1 & 100037560 & 100039288 & EXTL2    & K562 &  0.080957578 & 0.042460253 &  1.90666738 & 5.656367e-02\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9986538 & 9987289 & CLCN4     & K562 & -1.608619e-02 &   0.09986950 & -0.16107214 & 8.720366e-01\\\\\n",
       "\t chrX & 9986538 & 9987289 & SHROOM2   & K562 &  5.488062e-01 &   0.20953585 &  2.61915174 & 8.814873e-03\\\\\n",
       "\t chrX & 9986538 & 9987289 & TBL1X     & K562 &  3.825249e-01 &   0.02841623 & 13.46149333 & 2.635357e-41\\\\\n",
       "\t chrX & 9986538 & 9987289 & WWC3      & K562 & -3.905878e-02 &   0.06888749 & -0.56699380 & 5.707184e-01\\\\\n",
       "\t chrX & 9987345 & 9987562 & CLCN4     & K562 &  1.773150e-02 &   0.25764974 &  0.06882019 & 9.451327e-01\\\\\n",
       "\t chrX & 9987345 & 9987562 & TBL1X     & K562 &  1.673476e-01 &   0.08150312 &  2.05326592 & 4.004679e-02\\\\\n",
       "\t chrX & 9987345 & 9987562 & WWC3      & K562 & -3.615037e-01 &   0.24597549 & -1.46967364 & 1.416502e-01\\\\\n",
       "\t chrX & 9988628 & 9989100 & CLCN4     & K562 &  1.629022e-01 &   0.15046136 &  1.08268456 & 2.789485e-01\\\\\n",
       "\t chrX & 9988628 & 9989100 & SHROOM2   & K562 &  5.959272e-01 &   0.34787510 &  1.71304925 & 8.670348e-02\\\\\n",
       "\t chrX & 9988628 & 9989100 & TBL1X     & K562 &  1.469698e-01 &   0.05523129 &  2.66098841 & 7.791164e-03\\\\\n",
       "\t chrX & 9988628 & 9989100 & WWC3      & K562 &  1.603043e-02 &   0.11726105 &  0.13670718 & 8.912623e-01\\\\\n",
       "\t chrX &  999358 &  999880 & AKAP17A   & K562 &  1.270236e-01 &   0.12520607 &  1.01451612 & 3.103365e-01\\\\\n",
       "\t chrX &  999358 &  999880 & ASMT      & K562 & -1.191942e+01 & 299.03078684 & -0.03986019 & 9.682046e-01\\\\\n",
       "\t chrX &  999358 &  999880 & ASMTL     & K562 &  2.831859e-01 &   0.12091872 &  2.34195278 & 1.918314e-02\\\\\n",
       "\t chrX &  999358 &  999880 & CSF2RA    & K562 & -2.039934e-01 &   0.13851564 & -1.47271042 & 1.408291e-01\\\\\n",
       "\t chrX &  999358 &  999880 & GTPBP6    & K562 & -6.236828e-02 &   0.18538232 & -0.33643058 & 7.365462e-01\\\\\n",
       "\t chrX &  999358 &  999880 & IL3RA     & K562 &  2.351789e-01 &   0.12519168 &  1.87855035 & 6.030592e-02\\\\\n",
       "\t chrX &  999358 &  999880 & LINC00106 & K562 &  1.519153e-01 &   0.21182872 &  0.71716113 & 4.732747e-01\\\\\n",
       "\t chrX &  999358 &  999880 & P2RY8     & K562 &  8.787939e-03 &   0.06983985 &  0.12582986 & 8.998666e-01\\\\\n",
       "\t chrX &  999358 &  999880 & PLCXD1    & K562 & -1.794655e-01 &   0.26096875 & -0.68768965 & 4.916482e-01\\\\\n",
       "\t chrX &  999358 &  999880 & PPP2R3B   & K562 &  5.871419e-01 &   0.14631624 &  4.01282780 & 5.999568e-05\\\\\n",
       "\t chrX &  999358 &  999880 & SLC25A6   & K562 & -1.827568e-01 &   0.08911021 & -2.05090697 & 4.027601e-02\\\\\n",
       "\t chrX & 9995615 & 9996434 & CLCN4     & K562 & -9.818911e-04 &   0.06801909 & -0.01443552 & 9.884825e-01\\\\\n",
       "\t chrX & 9995615 & 9996434 & SHROOM2   & K562 &  9.887375e-02 &   0.22303112 &  0.44331818 & 6.575356e-01\\\\\n",
       "\t chrX & 9995615 & 9996434 & TBL1X     & K562 &  3.913810e-02 &   0.02265325 &  1.72770334 & 8.404142e-02\\\\\n",
       "\t chrX & 9995615 & 9996434 & WWC3      & K562 &  5.638883e-02 &   0.04389357 &  1.28467185 & 1.989070e-01\\\\\n",
       "\t chrX & 9997010 & 9998101 & CLCN4     & K562 & -2.671559e-01 &   0.15678124 & -1.70400394 & 8.838035e-02\\\\\n",
       "\t chrX & 9997010 & 9998101 & SHROOM2   & K562 &  8.707427e-02 &   0.40319907 &  0.21595852 & 8.290201e-01\\\\\n",
       "\t chrX & 9997010 & 9998101 & TBL1X     & K562 &  2.191688e-01 &   0.03718786 &  5.89355786 & 3.779677e-09\\\\\n",
       "\t chrX & 9997010 & 9998101 & WWC3      & K562 &  2.782902e-02 &   0.08065427 &  0.34504089 & 7.300636e-01\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5660858 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | beta &lt;dbl&gt; | sd &lt;dbl&gt; | z &lt;dbl&gt; | p &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100015450 | 100016026 | AGL      | K562 |  0.186202797 | 0.143923884 |  1.29375884 | 1.957487e-01 |\n",
       "| chr1 | 100015450 | 100016026 | CDC14A   | K562 |  0.133354374 | 0.068663708 |  1.94213768 | 5.212043e-02 |\n",
       "| chr1 | 100015450 | 100016026 | DBT      | K562 | -0.411851772 | 0.288115031 | -1.42946992 | 1.528692e-01 |\n",
       "| chr1 | 100015450 | 100016026 | EXTL2    | K562 |  0.265306585 | 0.306112058 |  0.86669760 | 3.861077e-01 |\n",
       "| chr1 | 100015450 | 100016026 | MFSD14A  | K562 |  0.163396983 | 0.489481785 |  0.33381627 | 7.385182e-01 |\n",
       "| chr1 | 100015450 | 100016026 | RTCA     | K562 |  0.090986595 | 0.136889000 |  0.66467426 | 5.062589e-01 |\n",
       "| chr1 | 100015450 | 100016026 | RTCA-AS1 | K562 |  0.014138722 | 0.413738974 |  0.03417305 | 9.727392e-01 |\n",
       "| chr1 | 100015450 | 100016026 | SASS6    | K562 |  0.015875834 | 0.183221132 |  0.08664849 | 9.309509e-01 |\n",
       "| chr1 | 100015450 | 100016026 | SLC30A7  | K562 |  0.101410986 | 0.100720610 |  1.00685437 | 3.140048e-01 |\n",
       "| chr1 | 100015450 | 100016026 | SLC35A3  | K562 | -0.087917219 | 0.224578483 | -0.39147659 | 6.954450e-01 |\n",
       "| chr1 | 100015450 | 100016026 | TRMT13   | K562 | -0.113919226 | 0.186336451 | -0.61136308 | 5.409592e-01 |\n",
       "| chr1 | 100028395 | 100029373 | AGL      | K562 |  0.187385214 | 0.042191001 |  4.44135504 | 8.939416e-06 |\n",
       "| chr1 | 100028395 | 100029373 | CDC14A   | K562 |  0.097553272 | 0.020248843 |  4.81772071 | 1.452074e-06 |\n",
       "| chr1 | 100028395 | 100029373 | DBT      | K562 |  0.032055857 | 0.055813799 |  0.57433570 | 5.657407e-01 |\n",
       "| chr1 | 100028395 | 100029373 | DPH5     | K562 |  0.149823444 | 0.057667647 |  2.59805022 | 9.375478e-03 |\n",
       "| chr1 | 100028395 | 100029373 | EXTL2    | K562 |  0.114501301 | 0.102801945 |  1.11380481 | 2.653629e-01 |\n",
       "| chr1 | 100028395 | 100029373 | FRRS1    | K562 | -0.770525949 | 0.147181099 | -5.23522352 | 1.647853e-07 |\n",
       "| chr1 | 100028395 | 100029373 | MFSD14A  | K562 |  0.196956118 | 0.139249995 |  1.41440665 | 1.572425e-01 |\n",
       "| chr1 | 100028395 | 100029373 | RTCA     | K562 |  0.090545549 | 0.039855502 |  2.27184567 | 2.309583e-02 |\n",
       "| chr1 | 100028395 | 100029373 | RTCA-AS1 | K562 |  0.034972974 | 0.117734570 |  0.29704933 | 7.664289e-01 |\n",
       "| chr1 | 100028395 | 100029373 | SASS6    | K562 | -0.096079077 | 0.058327134 | -1.64724495 | 9.950771e-02 |\n",
       "| chr1 | 100028395 | 100029373 | SLC30A7  | K562 |  0.007023095 | 0.031200077 |  0.22509864 | 8.219025e-01 |\n",
       "| chr1 | 100028395 | 100029373 | SLC35A3  | K562 |  0.121783939 | 0.054457127 |  2.23632690 | 2.533037e-02 |\n",
       "| chr1 | 100028395 | 100029373 | TRMT13   | K562 |  0.025684388 | 0.047758628 |  0.53779575 | 5.907181e-01 |\n",
       "| chr1 | 100028395 | 100029373 | VCAM1    | K562 |  0.544125314 | 0.143223236 |  3.79914134 | 1.451983e-04 |\n",
       "| chr1 | 100037560 | 100039288 | AGL      | K562 |  0.072809383 | 0.018596372 |  3.91524669 | 9.031182e-05 |\n",
       "| chr1 | 100037560 | 100039288 | CDC14A   | K562 |  0.093677870 | 0.007995935 | 11.71568697 | 1.059291e-31 |\n",
       "| chr1 | 100037560 | 100039288 | DBT      | K562 |  0.025361401 | 0.021810102 |  1.16282821 | 2.448992e-01 |\n",
       "| chr1 | 100037560 | 100039288 | DPH5     | K562 |  0.112824614 | 0.024391368 |  4.62559590 | 3.735229e-06 |\n",
       "| chr1 | 100037560 | 100039288 | EXTL2    | K562 |  0.080957578 | 0.042460253 |  1.90666738 | 5.656367e-02 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9986538 | 9987289 | CLCN4     | K562 | -1.608619e-02 |   0.09986950 | -0.16107214 | 8.720366e-01 |\n",
       "| chrX | 9986538 | 9987289 | SHROOM2   | K562 |  5.488062e-01 |   0.20953585 |  2.61915174 | 8.814873e-03 |\n",
       "| chrX | 9986538 | 9987289 | TBL1X     | K562 |  3.825249e-01 |   0.02841623 | 13.46149333 | 2.635357e-41 |\n",
       "| chrX | 9986538 | 9987289 | WWC3      | K562 | -3.905878e-02 |   0.06888749 | -0.56699380 | 5.707184e-01 |\n",
       "| chrX | 9987345 | 9987562 | CLCN4     | K562 |  1.773150e-02 |   0.25764974 |  0.06882019 | 9.451327e-01 |\n",
       "| chrX | 9987345 | 9987562 | TBL1X     | K562 |  1.673476e-01 |   0.08150312 |  2.05326592 | 4.004679e-02 |\n",
       "| chrX | 9987345 | 9987562 | WWC3      | K562 | -3.615037e-01 |   0.24597549 | -1.46967364 | 1.416502e-01 |\n",
       "| chrX | 9988628 | 9989100 | CLCN4     | K562 |  1.629022e-01 |   0.15046136 |  1.08268456 | 2.789485e-01 |\n",
       "| chrX | 9988628 | 9989100 | SHROOM2   | K562 |  5.959272e-01 |   0.34787510 |  1.71304925 | 8.670348e-02 |\n",
       "| chrX | 9988628 | 9989100 | TBL1X     | K562 |  1.469698e-01 |   0.05523129 |  2.66098841 | 7.791164e-03 |\n",
       "| chrX | 9988628 | 9989100 | WWC3      | K562 |  1.603043e-02 |   0.11726105 |  0.13670718 | 8.912623e-01 |\n",
       "| chrX |  999358 |  999880 | AKAP17A   | K562 |  1.270236e-01 |   0.12520607 |  1.01451612 | 3.103365e-01 |\n",
       "| chrX |  999358 |  999880 | ASMT      | K562 | -1.191942e+01 | 299.03078684 | -0.03986019 | 9.682046e-01 |\n",
       "| chrX |  999358 |  999880 | ASMTL     | K562 |  2.831859e-01 |   0.12091872 |  2.34195278 | 1.918314e-02 |\n",
       "| chrX |  999358 |  999880 | CSF2RA    | K562 | -2.039934e-01 |   0.13851564 | -1.47271042 | 1.408291e-01 |\n",
       "| chrX |  999358 |  999880 | GTPBP6    | K562 | -6.236828e-02 |   0.18538232 | -0.33643058 | 7.365462e-01 |\n",
       "| chrX |  999358 |  999880 | IL3RA     | K562 |  2.351789e-01 |   0.12519168 |  1.87855035 | 6.030592e-02 |\n",
       "| chrX |  999358 |  999880 | LINC00106 | K562 |  1.519153e-01 |   0.21182872 |  0.71716113 | 4.732747e-01 |\n",
       "| chrX |  999358 |  999880 | P2RY8     | K562 |  8.787939e-03 |   0.06983985 |  0.12582986 | 8.998666e-01 |\n",
       "| chrX |  999358 |  999880 | PLCXD1    | K562 | -1.794655e-01 |   0.26096875 | -0.68768965 | 4.916482e-01 |\n",
       "| chrX |  999358 |  999880 | PPP2R3B   | K562 |  5.871419e-01 |   0.14631624 |  4.01282780 | 5.999568e-05 |\n",
       "| chrX |  999358 |  999880 | SLC25A6   | K562 | -1.827568e-01 |   0.08911021 | -2.05090697 | 4.027601e-02 |\n",
       "| chrX | 9995615 | 9996434 | CLCN4     | K562 | -9.818911e-04 |   0.06801909 | -0.01443552 | 9.884825e-01 |\n",
       "| chrX | 9995615 | 9996434 | SHROOM2   | K562 |  9.887375e-02 |   0.22303112 |  0.44331818 | 6.575356e-01 |\n",
       "| chrX | 9995615 | 9996434 | TBL1X     | K562 |  3.913810e-02 |   0.02265325 |  1.72770334 | 8.404142e-02 |\n",
       "| chrX | 9995615 | 9996434 | WWC3      | K562 |  5.638883e-02 |   0.04389357 |  1.28467185 | 1.989070e-01 |\n",
       "| chrX | 9997010 | 9998101 | CLCN4     | K562 | -2.671559e-01 |   0.15678124 | -1.70400394 | 8.838035e-02 |\n",
       "| chrX | 9997010 | 9998101 | SHROOM2   | K562 |  8.707427e-02 |   0.40319907 |  0.21595852 | 8.290201e-01 |\n",
       "| chrX | 9997010 | 9998101 | TBL1X     | K562 |  2.191688e-01 |   0.03718786 |  5.89355786 | 3.779677e-09 |\n",
       "| chrX | 9997010 | 9998101 | WWC3      | K562 |  2.782902e-02 |   0.08065427 |  0.34504089 | 7.300636e-01 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType beta          sd          \n",
       "1       chr1 100015450 100016026 AGL        K562      0.186202797  0.143923884 \n",
       "2       chr1 100015450 100016026 CDC14A     K562      0.133354374  0.068663708 \n",
       "3       chr1 100015450 100016026 DBT        K562     -0.411851772  0.288115031 \n",
       "4       chr1 100015450 100016026 EXTL2      K562      0.265306585  0.306112058 \n",
       "5       chr1 100015450 100016026 MFSD14A    K562      0.163396983  0.489481785 \n",
       "6       chr1 100015450 100016026 RTCA       K562      0.090986595  0.136889000 \n",
       "7       chr1 100015450 100016026 RTCA-AS1   K562      0.014138722  0.413738974 \n",
       "8       chr1 100015450 100016026 SASS6      K562      0.015875834  0.183221132 \n",
       "9       chr1 100015450 100016026 SLC30A7    K562      0.101410986  0.100720610 \n",
       "10      chr1 100015450 100016026 SLC35A3    K562     -0.087917219  0.224578483 \n",
       "11      chr1 100015450 100016026 TRMT13     K562     -0.113919226  0.186336451 \n",
       "12      chr1 100028395 100029373 AGL        K562      0.187385214  0.042191001 \n",
       "13      chr1 100028395 100029373 CDC14A     K562      0.097553272  0.020248843 \n",
       "14      chr1 100028395 100029373 DBT        K562      0.032055857  0.055813799 \n",
       "15      chr1 100028395 100029373 DPH5       K562      0.149823444  0.057667647 \n",
       "16      chr1 100028395 100029373 EXTL2      K562      0.114501301  0.102801945 \n",
       "17      chr1 100028395 100029373 FRRS1      K562     -0.770525949  0.147181099 \n",
       "18      chr1 100028395 100029373 MFSD14A    K562      0.196956118  0.139249995 \n",
       "19      chr1 100028395 100029373 RTCA       K562      0.090545549  0.039855502 \n",
       "20      chr1 100028395 100029373 RTCA-AS1   K562      0.034972974  0.117734570 \n",
       "21      chr1 100028395 100029373 SASS6      K562     -0.096079077  0.058327134 \n",
       "22      chr1 100028395 100029373 SLC30A7    K562      0.007023095  0.031200077 \n",
       "23      chr1 100028395 100029373 SLC35A3    K562      0.121783939  0.054457127 \n",
       "24      chr1 100028395 100029373 TRMT13     K562      0.025684388  0.047758628 \n",
       "25      chr1 100028395 100029373 VCAM1      K562      0.544125314  0.143223236 \n",
       "26      chr1 100037560 100039288 AGL        K562      0.072809383  0.018596372 \n",
       "27      chr1 100037560 100039288 CDC14A     K562      0.093677870  0.007995935 \n",
       "28      chr1 100037560 100039288 DBT        K562      0.025361401  0.021810102 \n",
       "29      chr1 100037560 100039288 DPH5       K562      0.112824614  0.024391368 \n",
       "30      chr1 100037560 100039288 EXTL2      K562      0.080957578  0.042460253 \n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮           \n",
       "5660829 chrX 9986538   9987289   CLCN4      K562     -1.608619e-02   0.09986950\n",
       "5660830 chrX 9986538   9987289   SHROOM2    K562      5.488062e-01   0.20953585\n",
       "5660831 chrX 9986538   9987289   TBL1X      K562      3.825249e-01   0.02841623\n",
       "5660832 chrX 9986538   9987289   WWC3       K562     -3.905878e-02   0.06888749\n",
       "5660833 chrX 9987345   9987562   CLCN4      K562      1.773150e-02   0.25764974\n",
       "5660834 chrX 9987345   9987562   TBL1X      K562      1.673476e-01   0.08150312\n",
       "5660835 chrX 9987345   9987562   WWC3       K562     -3.615037e-01   0.24597549\n",
       "5660836 chrX 9988628   9989100   CLCN4      K562      1.629022e-01   0.15046136\n",
       "5660837 chrX 9988628   9989100   SHROOM2    K562      5.959272e-01   0.34787510\n",
       "5660838 chrX 9988628   9989100   TBL1X      K562      1.469698e-01   0.05523129\n",
       "5660839 chrX 9988628   9989100   WWC3       K562      1.603043e-02   0.11726105\n",
       "5660840 chrX  999358    999880   AKAP17A    K562      1.270236e-01   0.12520607\n",
       "5660841 chrX  999358    999880   ASMT       K562     -1.191942e+01 299.03078684\n",
       "5660842 chrX  999358    999880   ASMTL      K562      2.831859e-01   0.12091872\n",
       "5660843 chrX  999358    999880   CSF2RA     K562     -2.039934e-01   0.13851564\n",
       "5660844 chrX  999358    999880   GTPBP6     K562     -6.236828e-02   0.18538232\n",
       "5660845 chrX  999358    999880   IL3RA      K562      2.351789e-01   0.12519168\n",
       "5660846 chrX  999358    999880   LINC00106  K562      1.519153e-01   0.21182872\n",
       "5660847 chrX  999358    999880   P2RY8      K562      8.787939e-03   0.06983985\n",
       "5660848 chrX  999358    999880   PLCXD1     K562     -1.794655e-01   0.26096875\n",
       "5660849 chrX  999358    999880   PPP2R3B    K562      5.871419e-01   0.14631624\n",
       "5660850 chrX  999358    999880   SLC25A6    K562     -1.827568e-01   0.08911021\n",
       "5660851 chrX 9995615   9996434   CLCN4      K562     -9.818911e-04   0.06801909\n",
       "5660852 chrX 9995615   9996434   SHROOM2    K562      9.887375e-02   0.22303112\n",
       "5660853 chrX 9995615   9996434   TBL1X      K562      3.913810e-02   0.02265325\n",
       "5660854 chrX 9995615   9996434   WWC3       K562      5.638883e-02   0.04389357\n",
       "5660855 chrX 9997010   9998101   CLCN4      K562     -2.671559e-01   0.15678124\n",
       "5660856 chrX 9997010   9998101   SHROOM2    K562      8.707427e-02   0.40319907\n",
       "5660857 chrX 9997010   9998101   TBL1X      K562      2.191688e-01   0.03718786\n",
       "5660858 chrX 9997010   9998101   WWC3       K562      2.782902e-02   0.08065427\n",
       "        z           p           \n",
       "1        1.29375884 1.957487e-01\n",
       "2        1.94213768 5.212043e-02\n",
       "3       -1.42946992 1.528692e-01\n",
       "4        0.86669760 3.861077e-01\n",
       "5        0.33381627 7.385182e-01\n",
       "6        0.66467426 5.062589e-01\n",
       "7        0.03417305 9.727392e-01\n",
       "8        0.08664849 9.309509e-01\n",
       "9        1.00685437 3.140048e-01\n",
       "10      -0.39147659 6.954450e-01\n",
       "11      -0.61136308 5.409592e-01\n",
       "12       4.44135504 8.939416e-06\n",
       "13       4.81772071 1.452074e-06\n",
       "14       0.57433570 5.657407e-01\n",
       "15       2.59805022 9.375478e-03\n",
       "16       1.11380481 2.653629e-01\n",
       "17      -5.23522352 1.647853e-07\n",
       "18       1.41440665 1.572425e-01\n",
       "19       2.27184567 2.309583e-02\n",
       "20       0.29704933 7.664289e-01\n",
       "21      -1.64724495 9.950771e-02\n",
       "22       0.22509864 8.219025e-01\n",
       "23       2.23632690 2.533037e-02\n",
       "24       0.53779575 5.907181e-01\n",
       "25       3.79914134 1.451983e-04\n",
       "26       3.91524669 9.031182e-05\n",
       "27      11.71568697 1.059291e-31\n",
       "28       1.16282821 2.448992e-01\n",
       "29       4.62559590 3.735229e-06\n",
       "30       1.90666738 5.656367e-02\n",
       "⋮       ⋮           ⋮           \n",
       "5660829 -0.16107214 8.720366e-01\n",
       "5660830  2.61915174 8.814873e-03\n",
       "5660831 13.46149333 2.635357e-41\n",
       "5660832 -0.56699380 5.707184e-01\n",
       "5660833  0.06882019 9.451327e-01\n",
       "5660834  2.05326592 4.004679e-02\n",
       "5660835 -1.46967364 1.416502e-01\n",
       "5660836  1.08268456 2.789485e-01\n",
       "5660837  1.71304925 8.670348e-02\n",
       "5660838  2.66098841 7.791164e-03\n",
       "5660839  0.13670718 8.912623e-01\n",
       "5660840  1.01451612 3.103365e-01\n",
       "5660841 -0.03986019 9.682046e-01\n",
       "5660842  2.34195278 1.918314e-02\n",
       "5660843 -1.47271042 1.408291e-01\n",
       "5660844 -0.33643058 7.365462e-01\n",
       "5660845  1.87855035 6.030592e-02\n",
       "5660846  0.71716113 4.732747e-01\n",
       "5660847  0.12582986 8.998666e-01\n",
       "5660848 -0.68768965 4.916482e-01\n",
       "5660849  4.01282780 5.999568e-05\n",
       "5660850 -2.05090697 4.027601e-02\n",
       "5660851 -0.01443552 9.884825e-01\n",
       "5660852  0.44331818 6.575356e-01\n",
       "5660853  1.72770334 8.404142e-02\n",
       "5660854  1.28467185 1.989070e-01\n",
       "5660855 -1.70400394 8.838035e-02\n",
       "5660856  0.21595852 8.290201e-01\n",
       "5660857  5.89355786 3.779677e-09\n",
       "5660858  0.34504089 7.300636e-01"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"beta\",\n",
    "                         \"sd\",\n",
    "                         \"z\",\n",
    "                         \"p\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] data.table_1.15.2    dplyr_1.1.4          foreach_1.5.2       \n",
       " [4] GenomicRanges_1.54.1 GenomeInfoDb_1.38.1  IRanges_2.36.0      \n",
       " [7] S4Vectors_0.40.2     BiocGenerics_0.48.1  genomation_1.34.0   \n",
       "[10] Signac_1.13.0        Seurat_5.0.3         SeuratObject_5.0.1  \n",
       "[13] sp_2.1-4             SnapATAC_1.0.0       rhdf5_2.44.0        \n",
       "[16] Matrix_1.6-5        \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 BiocIO_1.10.0              \n",
       "  [5] pbdZMQ_0.3-11               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                polyclip_1.10-6            \n",
       "  [9] XML_3.99-0.16.1             fastDummies_1.7.3          \n",
       " [11] lifecycle_1.0.4             tcltk_4.3.3                \n",
       " [13] edgeR_3.42.4                doParallel_1.0.17          \n",
       " [15] globals_0.16.3              lattice_0.22-6             \n",
       " [17] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [19] limma_3.56.2                plotly_4.10.4              \n",
       " [21] plotrix_3.8-4               yaml_2.3.8                 \n",
       " [23] plot3D_1.4.1                httpuv_1.6.15              \n",
       " [25] sctransform_0.4.1           spam_2.10-0                \n",
       " [27] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [29] cowplot_1.1.3               pbapply_1.7-2              \n",
       " [31] RColorBrewer_1.1-3          abind_1.4-5                \n",
       " [33] zlibbioc_1.48.0             Rtsne_0.17                 \n",
       " [35] purrr_1.0.2                 RCurl_1.98-1.14            \n",
       " [37] misc3d_0.9-1                GenomeInfoDbData_1.2.11    \n",
       " [39] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [41] listenv_0.9.1               spatstat.utils_3.0-4       \n",
       " [43] goftest_1.2-3               RSpectra_0.16-1            \n",
       " [45] bigmemory_4.6.4             spatstat.random_3.2-3      \n",
       " [47] fitdistrplus_1.1-11         parallelly_1.37.1          \n",
       " [49] DelayedArray_0.26.7         leiden_0.4.3.1             \n",
       " [51] codetools_0.2-20            RcppRoll_0.3.0             \n",
       " [53] tidyselect_1.2.1            viridis_0.6.5              \n",
       " [55] matrixStats_1.3.0           base64enc_0.1-3            \n",
       " [57] spatstat.explore_3.2-6      GenomicAlignments_1.36.0   \n",
       " [59] jsonlite_1.8.8              progressr_0.14.0           \n",
       " [61] ggridges_0.5.6              survival_3.6-4             \n",
       " [63] iterators_1.0.14            tools_4.3.3                \n",
       " [65] snow_0.4-4                  ica_1.0-3                  \n",
       " [67] Rcpp_1.0.12                 glue_1.7.0                 \n",
       " [69] gridExtra_2.3               MatrixGenerics_1.12.3      \n",
       " [71] IRdisplay_1.1               withr_3.0.0                \n",
       " [73] fastmap_1.1.1               rhdf5filters_1.12.1        \n",
       " [75] fansi_1.0.6                 digest_0.6.35              \n",
       " [77] R6_2.5.1                    mime_0.12                  \n",
       " [79] seqPattern_1.34.0           colorspace_2.1-0           \n",
       " [81] scattermore_1.2             tensor_1.5                 \n",
       " [83] spatstat.data_3.0-4         utf8_1.2.4                 \n",
       " [85] tidyr_1.3.1                 generics_0.1.3             \n",
       " [87] rtracklayer_1.60.1          S4Arrays_1.2.1             \n",
       " [89] httr_1.4.7                  htmlwidgets_1.6.4          \n",
       " [91] uwot_0.1.16                 pkgconfig_2.0.3            \n",
       " [93] gtable_0.3.5                impute_1.76.0              \n",
       " [95] lmtest_0.9-40               XVector_0.42.0             \n",
       " [97] htmltools_0.5.8.1           dotCall64_1.1-1            \n",
       " [99] Biobase_2.60.0              scales_1.3.0               \n",
       "[101] png_0.1-8                   doSNOW_1.0.20              \n",
       "[103] bigmemory.sri_0.1.8         tzdb_0.4.0                 \n",
       "[105] rjson_0.2.21                reshape2_1.4.4             \n",
       "[107] uuid_1.2-0                  nlme_3.1-164               \n",
       "[109] repr_1.1.7                  zoo_1.8-12                 \n",
       "[111] stringr_1.5.1               KernSmooth_2.23-22         \n",
       "[113] miniUI_0.1.1.1              restfulr_0.0.15            \n",
       "[115] pillar_1.9.0                vctrs_0.6.5                \n",
       "[117] RANN_2.6.1                  promises_1.3.0             \n",
       "[119] xtable_1.8-4                cluster_2.1.6              \n",
       "[121] evaluate_0.23               readr_2.1.5                \n",
       "[123] cli_3.6.2                   locfit_1.5-9.9             \n",
       "[125] compiler_4.3.3              Rsamtools_2.18.0           \n",
       "[127] rlang_1.1.3                 crayon_1.5.2               \n",
       "[129] future.apply_1.11.2         plyr_1.8.9                 \n",
       "[131] stringi_1.8.3               gridBase_0.4-7             \n",
       "[133] viridisLite_0.4.2           deldir_2.0-4               \n",
       "[135] BiocParallel_1.36.0         munsell_0.5.1              \n",
       "[137] Biostrings_2.70.1           lazyeval_0.2.2             \n",
       "[139] spatstat.geom_3.2-9         IRkernel_1.3.2             \n",
       "[141] BSgenome_1.68.0             RcppHNSW_0.6.0             \n",
       "[143] hms_1.1.3                   patchwork_1.2.0            \n",
       "[145] future_1.33.2               ggplot2_3.5.1              \n",
       "[147] Rhdf5lib_1.22.1             shiny_1.8.1.1              \n",
       "[149] SummarizedExperiment_1.30.2 ROCR_1.0-11                \n",
       "[151] igraph_2.0.3                fastmatch_1.1-4            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
