{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: rhdf5\n",
      "\n",
      "Warning message:\n",
      "“no DISPLAY variable so Tk is not available”\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SnapATAC)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(parallel)\n",
    "library(foreach)\n",
    "Sys.setenv(OPENBLAS_NUM_THREADS=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/4.merge_peaks_from_22_cell_types.240523/pairs.rds\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/4.merge_peaks_from_22_cell_types.240523/matrix.atac.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/4.merge_peaks_from_22_cell_types.240523/matrix.rna.rds\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/SnapATAC/BMMC22.merge.250212/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readRDS(path.pairs.E2G)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = BinarizeCounts(matrix.atac_count)\n",
    "rm(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = readRDS(path.matrix.rna_count)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna_count = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna_count)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4986774c-cbbf-4b72-a299-d681bddca80e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "85b165d0-f1db-4548-beac-235f8950107b",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "655ebba3-731e-49d8-a04e-1ebb5becd567",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter = pairs.E2G[pairs.E2G$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                             pairs.E2G$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Run SnapATAC prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PredictionSnapATAC = function(pairs.E2G,\n",
    "                              data.RNA,\n",
    "                              data.ATAC,\n",
    "                              n.core = 16){\n",
    "  my.cluster <- parallel::makeCluster(\n",
    "    n.core,\n",
    "    type = \"PSOCK\"\n",
    "  )\n",
    "  doParallel::registerDoParallel(cl = my.cluster)\n",
    "  \n",
    "  pairs.E2G =  \n",
    "    foreach (index.tmp = 1:length(pairs.E2G),\n",
    "             .combine = 'c',\n",
    "             .packages = c(\"Matrix\",\n",
    "                           \"GenomicRanges\")) %dopar% {\n",
    "\n",
    "                             pairs.E2G.tmp = pairs.E2G[index.tmp]\n",
    "                             \n",
    "                             model.tmp = summary(stats::glm(data.ATAC[pairs.E2G.tmp$PeakName,] ~ data.RNA[pairs.E2G.tmp$TargetGene,], \n",
    "                                                            family = binomial(link = \"logit\")))[[\"coefficients\"]]\n",
    "                             mcols(pairs.E2G.tmp)[,c(\"beta\",\"sd\",\"z\",\"p\")] = \n",
    "                               t(data.frame(model.tmp[2,]))\n",
    "                             \n",
    "                             pairs.E2G.tmp\n",
    "                           }\n",
    "  \n",
    "  parallel::stopCluster(cl = my.cluster)\n",
    "  \n",
    "  return(pairs.E2G)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(dir.output, \"chr\", sep = \"/\"), recursive = T):\n",
      "“'/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/SnapATAC/BMMC22.merge.250212//chr' already exists”\n"
     ]
    }
   ],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "[1] \"chr10\"\n",
      "[1] \"chr11\"\n",
      "[1] \"chr12\"\n",
      "[1] \"chr13\"\n",
      "[1] \"chr14\"\n",
      "[1] \"chr15\"\n",
      "[1] \"chr16\"\n",
      "[1] \"chr17\"\n",
      "[1] \"chr18\"\n",
      "[1] \"chr19\"\n",
      "[1] \"chr2\"\n",
      "[1] \"chr20\"\n",
      "[1] \"chr21\"\n",
      "[1] \"chr22\"\n",
      "[1] \"chr3\"\n",
      "[1] \"chr4\"\n",
      "[1] \"chr5\"\n",
      "[1] \"chr6\"\n",
      "[1] \"chr7\"\n",
      "[1] \"chr8\"\n",
      "[1] \"chr9\"\n",
      "[1] \"chrX\"\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter[seqnames(pairs.E2G.filter) == chr.tmp]\n",
    "  mcols(pairs.E2G.chr.res)[,c(\"beta\",\"sd\",\"z\",\"p\")] = NA\n",
    "  pairs.E2G.chr.res = PredictionSnapATAC(pairs.E2G.chr.res,\n",
    "                                         matrix.rna.rename[rownames(matrix.rna.rename) %in% pairs.E2G.chr.res$TargetGene,],\n",
    "                                         matrix.atac[rownames(matrix.atac) %in% pairs.E2G.chr.res$PeakName,],\n",
    "                                         n.cores)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 9727174 ranges and 7 metadata columns:\n",
       "       seqnames              ranges strand |  TargetGene               PeakName\n",
       "          <Rle>           <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 100004103-100004301      * |         AGL chr1-100004103-10000..\n",
       "  chr1     chr1 100004103-100004301      * |      CDC14A chr1-100004103-10000..\n",
       "  chr1     chr1 100004103-100004301      * |         DBT chr1-100004103-10000..\n",
       "  chr1     chr1 100004103-100004301      * |       EXTL2 chr1-100004103-10000..\n",
       "  chr1     chr1 100004103-100004301      * |       FRRS1 chr1-100004103-10000..\n",
       "   ...      ...                 ...    ... .         ...                    ...\n",
       "  chrX     chrX   99980528-99980848      * |      PCDH19 chrX-99980528-99980848\n",
       "  chrX     chrX   99980528-99980848      * |      TSPAN6 chrX-99980528-99980848\n",
       "  chrX     chrX   99980528-99980848      * |        XKRX chrX-99980528-99980848\n",
       "  chrX     chrX     9999312-9999592      * |       TBL1X   chrX-9999312-9999592\n",
       "  chrX     chrX     9999312-9999592      * |        WWC3   chrX-9999312-9999592\n",
       "                     PairName       beta          sd          z         p\n",
       "                  <character>  <numeric>   <numeric>  <numeric> <numeric>\n",
       "  chr1 chr1-100004103-10000..  0.4571315    0.199535   2.290986 0.0219642\n",
       "  chr1 chr1-100004103-10000..  0.0748561    0.127891   0.585310 0.5583391\n",
       "  chr1 chr1-100004103-10000.. -0.6556284    0.639632  -1.025008 0.3053592\n",
       "  chr1 chr1-100004103-10000.. -0.3459728    0.880083  -0.393114 0.6942354\n",
       "  chr1 chr1-100004103-10000.. -0.2621136    0.633400  -0.413820 0.6790058\n",
       "   ...                    ...        ...         ...        ...       ...\n",
       "  chrX chrX-99980528-999808.. -15.345517 1241.830012 -0.0123572  0.990141\n",
       "  chrX chrX-99980528-999808.. -19.442721 1865.082197 -0.0104246  0.991683\n",
       "  chrX chrX-99980528-999808.. -15.904810 1210.071570 -0.0131437  0.989513\n",
       "  chrX chrX-9999312-9999592..   0.160232    0.102074  1.5697579  0.116471\n",
       "  chrX chrX-9999312-9999592..   0.201931    0.188512  1.0711884  0.284085\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 9727174 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>beta</th><th scope=col>sd</th><th scope=col>z</th><th scope=col>p</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>AGL        </td><td>K562</td><td> 0.457131511</td><td>  0.19953486</td><td> 2.29098569</td><td>0.02196424</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>CDC14A     </td><td>K562</td><td> 0.074856093</td><td>  0.12789128</td><td> 0.58531036</td><td>0.55833905</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>DBT        </td><td>K562</td><td>-0.655628429</td><td>  0.63963226</td><td>-1.02500838</td><td>0.30535923</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>EXTL2      </td><td>K562</td><td>-0.345972818</td><td>  0.88008305</td><td>-0.39311383</td><td>0.69423541</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>FRRS1      </td><td>K562</td><td>-0.262113620</td><td>  0.63339983</td><td>-0.41382016</td><td>0.67900582</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>MFSD14A    </td><td>K562</td><td>-0.046556068</td><td>  1.02902606</td><td>-0.04524285</td><td>0.96391374</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>PLPPR5     </td><td>K562</td><td> 1.064230198</td><td>  0.52153143</td><td> 2.04058688</td><td>0.04129191</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>RTCA       </td><td>K562</td><td>-0.003990392</td><td>  0.26336186</td><td>-0.01515175</td><td>0.98791112</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>RTCA-AS1   </td><td>K562</td><td> 0.024454937</td><td>  0.72139816</td><td> 0.03389936</td><td>0.97295740</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>SASS6      </td><td>K562</td><td> 0.478958268</td><td>  0.21958646</td><td> 2.18118302</td><td>0.02916988</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>SLC30A7    </td><td>K562</td><td> 0.170716318</td><td>  0.17185899</td><td> 0.99335113</td><td>0.32053887</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>SLC35A3    </td><td>K562</td><td>-0.385101235</td><td>  0.51159813</td><td>-0.75274168</td><td>0.45160516</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>TRMT13     </td><td>K562</td><td>-0.473817506</td><td>  0.45064019</td><td>-1.05143198</td><td>0.29306024</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>AGL        </td><td>K562</td><td> 0.100658564</td><td>  0.13176197</td><td> 0.76394247</td><td>0.44490152</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>CDC14A     </td><td>K562</td><td> 0.061930828</td><td>  0.06180056</td><td> 1.00210785</td><td>0.31629151</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>DBT        </td><td>K562</td><td> 0.128080932</td><td>  0.15167720</td><td> 0.84443101</td><td>0.39842860</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>EXTL2      </td><td>K562</td><td> 0.271037154</td><td>  0.24230387</td><td> 1.11858367</td><td>0.26331779</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>FRRS1      </td><td>K562</td><td> 0.194968190</td><td>  0.20396612</td><td> 0.95588518</td><td>0.33913025</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>GPR88      </td><td>K562</td><td>-9.326765521</td><td>314.00867345</td><td>-0.02970225</td><td>0.97630451</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>MFSD14A    </td><td>K562</td><td>-0.491938413</td><td>  0.74430404</td><td>-0.66093744</td><td>0.50865243</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>RTCA       </td><td>K562</td><td>-0.041605736</td><td>  0.13041590</td><td>-0.31902349</td><td>0.74970870</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>RTCA-AS1   </td><td>K562</td><td> 0.456613536</td><td>  0.22870540</td><td> 1.99651400</td><td>0.04587800</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>SASS6      </td><td>K562</td><td>-0.013200767</td><td>  0.16147591</td><td>-0.08175069</td><td>0.93484497</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>SLC30A7    </td><td>K562</td><td> 0.104259684</td><td>  0.08633671</td><td> 1.20759388</td><td>0.22720351</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>SLC35A3    </td><td>K562</td><td> 0.132256033</td><td>  0.15583125</td><td> 0.84871317</td><td>0.39604092</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>TRMT13     </td><td>K562</td><td>-0.026169687</td><td>  0.14566140</td><td>-0.17966110</td><td>0.85741863</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001440</td><td> 10001830</td><td>APITD1-CORT</td><td>K562</td><td> 1.001306699</td><td>  0.55288905</td><td> 1.81104454</td><td>0.07013396</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001440</td><td> 10001830</td><td>CENPS      </td><td>K562</td><td>-0.195202385</td><td>  1.02687668</td><td>-0.19009331</td><td>0.84923601</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001440</td><td> 10001830</td><td>CLSTN1     </td><td>K562</td><td> 0.004962965</td><td>  0.14622498</td><td> 0.03394061</td><td>0.97292451</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001440</td><td> 10001830</td><td>CTNNBIP1   </td><td>K562</td><td> 0.265333568</td><td>  0.18006297</td><td> 1.47355990</td><td>0.14060012</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9992130</td><td> 9992304</td><td>WWC3     </td><td>K562</td><td>  0.440577637</td><td>1.430916e-01</td><td> 3.07898956</td><td>2.077040e-03</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>AKAP17A  </td><td>K562</td><td>  0.048439172</td><td>1.049150e-01</td><td> 0.46169934</td><td>6.442969e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>ASMT     </td><td>K562</td><td>  0.210805014</td><td>6.133866e-01</td><td> 0.34367400</td><td>7.310915e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>ASMTL    </td><td>K562</td><td>  0.156090099</td><td>1.063782e-01</td><td> 1.46731252</td><td>1.422911e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>CSF2RA   </td><td>K562</td><td> -0.181719570</td><td>1.189538e-01</td><td>-1.52764803</td><td>1.265999e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>GTPBP6   </td><td>K562</td><td> -0.026414961</td><td>1.419104e-01</td><td>-0.18613835</td><td>8.523363e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>IL3RA    </td><td>K562</td><td>  0.197814962</td><td>1.071269e-01</td><td> 1.84654787</td><td>6.481270e-02</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>LINC00106</td><td>K562</td><td>  0.185572928</td><td>1.698959e-01</td><td> 1.09227425</td><td>2.747126e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>P2RY8    </td><td>K562</td><td> -0.011915616</td><td>5.739606e-02</td><td>-0.20760336</td><td>8.355387e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>PLCXD1   </td><td>K562</td><td> -0.004166042</td><td>1.782306e-01</td><td>-0.02337445</td><td>9.813516e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>PPP2R3B  </td><td>K562</td><td>  0.631945853</td><td>1.126565e-01</td><td> 5.60949430</td><td>2.029187e-08</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>SHOX     </td><td>K562</td><td>-10.203892238</td><td>2.353793e+02</td><td>-0.04335084</td><td>9.654219e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>SLC25A6  </td><td>K562</td><td> -0.009238020</td><td>6.403557e-02</td><td>-0.14426389</td><td>8.852921e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>CLCN4    </td><td>K562</td><td>  0.006258711</td><td>6.490849e-02</td><td> 0.09642361</td><td>9.231841e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>FAM9B    </td><td>K562</td><td>  0.045752222</td><td>1.535455e-01</td><td> 0.29797182</td><td>7.657247e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>GPR143   </td><td>K562</td><td> -0.233540209</td><td>3.622750e-01</td><td>-0.64464892</td><td>5.191547e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>SHROOM2  </td><td>K562</td><td>  0.118672260</td><td>1.887302e-01</td><td> 0.62879301</td><td>5.294846e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>TBL1X    </td><td>K562</td><td>  0.026033112</td><td>2.059323e-02</td><td> 1.26415902</td><td>2.061730e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>WWC3     </td><td>K562</td><td>  0.057659020</td><td>4.025994e-02</td><td> 1.43216867</td><td>1.520956e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>CLCN4    </td><td>K562</td><td> -0.228042253</td><td>1.467925e-01</td><td>-1.55350059</td><td>1.203036e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>GPR143   </td><td>K562</td><td> -0.544363049</td><td>8.909577e-01</td><td>-0.61098640</td><td>5.412086e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>SHROOM2  </td><td>K562</td><td>  0.079025419</td><td>3.546140e-01</td><td> 0.22284913</td><td>8.236529e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>TBL1X    </td><td>K562</td><td>  0.199258783</td><td>3.421840e-02</td><td> 5.82314752</td><td>5.774954e-09</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>WWC3     </td><td>K562</td><td>  0.052105985</td><td>7.336479e-02</td><td> 0.71023154</td><td>4.775606e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>99980528</td><td>99980848</td><td>CSTF2    </td><td>K562</td><td>  0.421534303</td><td>5.526727e-01</td><td> 0.76271959</td><td>4.456306e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>99980528</td><td>99980848</td><td>PCDH19   </td><td>K562</td><td>-15.345517138</td><td>1.241830e+03</td><td>-0.01235718</td><td>9.901406e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>99980528</td><td>99980848</td><td>TSPAN6   </td><td>K562</td><td>-19.442721494</td><td>1.865082e+03</td><td>-0.01042459</td><td>9.916825e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td>99980528</td><td>99980848</td><td>XKRX     </td><td>K562</td><td>-15.904810429</td><td>1.210072e+03</td><td>-0.01314369</td><td>9.895132e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9999312</td><td> 9999592</td><td>TBL1X    </td><td>K562</td><td>  0.160231973</td><td>1.020743e-01</td><td> 1.56975785</td><td>1.164715e-01</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9999312</td><td> 9999592</td><td>WWC3     </td><td>K562</td><td>  0.201931459</td><td>1.885116e-01</td><td> 1.07118837</td><td>2.840847e-01</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 9727174 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & beta & sd & z & p\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100004103 & 100004301 & AGL         & K562 &  0.457131511 &   0.19953486 &  2.29098569 & 0.02196424\\\\\n",
       "\t chr1 & 100004103 & 100004301 & CDC14A      & K562 &  0.074856093 &   0.12789128 &  0.58531036 & 0.55833905\\\\\n",
       "\t chr1 & 100004103 & 100004301 & DBT         & K562 & -0.655628429 &   0.63963226 & -1.02500838 & 0.30535923\\\\\n",
       "\t chr1 & 100004103 & 100004301 & EXTL2       & K562 & -0.345972818 &   0.88008305 & -0.39311383 & 0.69423541\\\\\n",
       "\t chr1 & 100004103 & 100004301 & FRRS1       & K562 & -0.262113620 &   0.63339983 & -0.41382016 & 0.67900582\\\\\n",
       "\t chr1 & 100004103 & 100004301 & MFSD14A     & K562 & -0.046556068 &   1.02902606 & -0.04524285 & 0.96391374\\\\\n",
       "\t chr1 & 100004103 & 100004301 & PLPPR5      & K562 &  1.064230198 &   0.52153143 &  2.04058688 & 0.04129191\\\\\n",
       "\t chr1 & 100004103 & 100004301 & RTCA        & K562 & -0.003990392 &   0.26336186 & -0.01515175 & 0.98791112\\\\\n",
       "\t chr1 & 100004103 & 100004301 & RTCA-AS1    & K562 &  0.024454937 &   0.72139816 &  0.03389936 & 0.97295740\\\\\n",
       "\t chr1 & 100004103 & 100004301 & SASS6       & K562 &  0.478958268 &   0.21958646 &  2.18118302 & 0.02916988\\\\\n",
       "\t chr1 & 100004103 & 100004301 & SLC30A7     & K562 &  0.170716318 &   0.17185899 &  0.99335113 & 0.32053887\\\\\n",
       "\t chr1 & 100004103 & 100004301 & SLC35A3     & K562 & -0.385101235 &   0.51159813 & -0.75274168 & 0.45160516\\\\\n",
       "\t chr1 & 100004103 & 100004301 & TRMT13      & K562 & -0.473817506 &   0.45064019 & -1.05143198 & 0.29306024\\\\\n",
       "\t chr1 & 100010532 & 100010875 & AGL         & K562 &  0.100658564 &   0.13176197 &  0.76394247 & 0.44490152\\\\\n",
       "\t chr1 & 100010532 & 100010875 & CDC14A      & K562 &  0.061930828 &   0.06180056 &  1.00210785 & 0.31629151\\\\\n",
       "\t chr1 & 100010532 & 100010875 & DBT         & K562 &  0.128080932 &   0.15167720 &  0.84443101 & 0.39842860\\\\\n",
       "\t chr1 & 100010532 & 100010875 & EXTL2       & K562 &  0.271037154 &   0.24230387 &  1.11858367 & 0.26331779\\\\\n",
       "\t chr1 & 100010532 & 100010875 & FRRS1       & K562 &  0.194968190 &   0.20396612 &  0.95588518 & 0.33913025\\\\\n",
       "\t chr1 & 100010532 & 100010875 & GPR88       & K562 & -9.326765521 & 314.00867345 & -0.02970225 & 0.97630451\\\\\n",
       "\t chr1 & 100010532 & 100010875 & MFSD14A     & K562 & -0.491938413 &   0.74430404 & -0.66093744 & 0.50865243\\\\\n",
       "\t chr1 & 100010532 & 100010875 & RTCA        & K562 & -0.041605736 &   0.13041590 & -0.31902349 & 0.74970870\\\\\n",
       "\t chr1 & 100010532 & 100010875 & RTCA-AS1    & K562 &  0.456613536 &   0.22870540 &  1.99651400 & 0.04587800\\\\\n",
       "\t chr1 & 100010532 & 100010875 & SASS6       & K562 & -0.013200767 &   0.16147591 & -0.08175069 & 0.93484497\\\\\n",
       "\t chr1 & 100010532 & 100010875 & SLC30A7     & K562 &  0.104259684 &   0.08633671 &  1.20759388 & 0.22720351\\\\\n",
       "\t chr1 & 100010532 & 100010875 & SLC35A3     & K562 &  0.132256033 &   0.15583125 &  0.84871317 & 0.39604092\\\\\n",
       "\t chr1 & 100010532 & 100010875 & TRMT13      & K562 & -0.026169687 &   0.14566140 & -0.17966110 & 0.85741863\\\\\n",
       "\t chr1 &  10001440 &  10001830 & APITD1-CORT & K562 &  1.001306699 &   0.55288905 &  1.81104454 & 0.07013396\\\\\n",
       "\t chr1 &  10001440 &  10001830 & CENPS       & K562 & -0.195202385 &   1.02687668 & -0.19009331 & 0.84923601\\\\\n",
       "\t chr1 &  10001440 &  10001830 & CLSTN1      & K562 &  0.004962965 &   0.14622498 &  0.03394061 & 0.97292451\\\\\n",
       "\t chr1 &  10001440 &  10001830 & CTNNBIP1    & K562 &  0.265333568 &   0.18006297 &  1.47355990 & 0.14060012\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  9992130 &  9992304 & WWC3      & K562 &   0.440577637 & 1.430916e-01 &  3.07898956 & 2.077040e-03\\\\\n",
       "\t chrX &   999313 &  1000203 & AKAP17A   & K562 &   0.048439172 & 1.049150e-01 &  0.46169934 & 6.442969e-01\\\\\n",
       "\t chrX &   999313 &  1000203 & ASMT      & K562 &   0.210805014 & 6.133866e-01 &  0.34367400 & 7.310915e-01\\\\\n",
       "\t chrX &   999313 &  1000203 & ASMTL     & K562 &   0.156090099 & 1.063782e-01 &  1.46731252 & 1.422911e-01\\\\\n",
       "\t chrX &   999313 &  1000203 & CSF2RA    & K562 &  -0.181719570 & 1.189538e-01 & -1.52764803 & 1.265999e-01\\\\\n",
       "\t chrX &   999313 &  1000203 & GTPBP6    & K562 &  -0.026414961 & 1.419104e-01 & -0.18613835 & 8.523363e-01\\\\\n",
       "\t chrX &   999313 &  1000203 & IL3RA     & K562 &   0.197814962 & 1.071269e-01 &  1.84654787 & 6.481270e-02\\\\\n",
       "\t chrX &   999313 &  1000203 & LINC00106 & K562 &   0.185572928 & 1.698959e-01 &  1.09227425 & 2.747126e-01\\\\\n",
       "\t chrX &   999313 &  1000203 & P2RY8     & K562 &  -0.011915616 & 5.739606e-02 & -0.20760336 & 8.355387e-01\\\\\n",
       "\t chrX &   999313 &  1000203 & PLCXD1    & K562 &  -0.004166042 & 1.782306e-01 & -0.02337445 & 9.813516e-01\\\\\n",
       "\t chrX &   999313 &  1000203 & PPP2R3B   & K562 &   0.631945853 & 1.126565e-01 &  5.60949430 & 2.029187e-08\\\\\n",
       "\t chrX &   999313 &  1000203 & SHOX      & K562 & -10.203892238 & 2.353793e+02 & -0.04335084 & 9.654219e-01\\\\\n",
       "\t chrX &   999313 &  1000203 & SLC25A6   & K562 &  -0.009238020 & 6.403557e-02 & -0.14426389 & 8.852921e-01\\\\\n",
       "\t chrX &  9995566 &  9996680 & CLCN4     & K562 &   0.006258711 & 6.490849e-02 &  0.09642361 & 9.231841e-01\\\\\n",
       "\t chrX &  9995566 &  9996680 & FAM9B     & K562 &   0.045752222 & 1.535455e-01 &  0.29797182 & 7.657247e-01\\\\\n",
       "\t chrX &  9995566 &  9996680 & GPR143    & K562 &  -0.233540209 & 3.622750e-01 & -0.64464892 & 5.191547e-01\\\\\n",
       "\t chrX &  9995566 &  9996680 & SHROOM2   & K562 &   0.118672260 & 1.887302e-01 &  0.62879301 & 5.294846e-01\\\\\n",
       "\t chrX &  9995566 &  9996680 & TBL1X     & K562 &   0.026033112 & 2.059323e-02 &  1.26415902 & 2.061730e-01\\\\\n",
       "\t chrX &  9995566 &  9996680 & WWC3      & K562 &   0.057659020 & 4.025994e-02 &  1.43216867 & 1.520956e-01\\\\\n",
       "\t chrX &  9996996 &  9998157 & CLCN4     & K562 &  -0.228042253 & 1.467925e-01 & -1.55350059 & 1.203036e-01\\\\\n",
       "\t chrX &  9996996 &  9998157 & GPR143    & K562 &  -0.544363049 & 8.909577e-01 & -0.61098640 & 5.412086e-01\\\\\n",
       "\t chrX &  9996996 &  9998157 & SHROOM2   & K562 &   0.079025419 & 3.546140e-01 &  0.22284913 & 8.236529e-01\\\\\n",
       "\t chrX &  9996996 &  9998157 & TBL1X     & K562 &   0.199258783 & 3.421840e-02 &  5.82314752 & 5.774954e-09\\\\\n",
       "\t chrX &  9996996 &  9998157 & WWC3      & K562 &   0.052105985 & 7.336479e-02 &  0.71023154 & 4.775606e-01\\\\\n",
       "\t chrX & 99980528 & 99980848 & CSTF2     & K562 &   0.421534303 & 5.526727e-01 &  0.76271959 & 4.456306e-01\\\\\n",
       "\t chrX & 99980528 & 99980848 & PCDH19    & K562 & -15.345517138 & 1.241830e+03 & -0.01235718 & 9.901406e-01\\\\\n",
       "\t chrX & 99980528 & 99980848 & TSPAN6    & K562 & -19.442721494 & 1.865082e+03 & -0.01042459 & 9.916825e-01\\\\\n",
       "\t chrX & 99980528 & 99980848 & XKRX      & K562 & -15.904810429 & 1.210072e+03 & -0.01314369 & 9.895132e-01\\\\\n",
       "\t chrX &  9999312 &  9999592 & TBL1X     & K562 &   0.160231973 & 1.020743e-01 &  1.56975785 & 1.164715e-01\\\\\n",
       "\t chrX &  9999312 &  9999592 & WWC3      & K562 &   0.201931459 & 1.885116e-01 &  1.07118837 & 2.840847e-01\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 9727174 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | beta &lt;dbl&gt; | sd &lt;dbl&gt; | z &lt;dbl&gt; | p &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100004103 | 100004301 | AGL         | K562 |  0.457131511 |   0.19953486 |  2.29098569 | 0.02196424 |\n",
       "| chr1 | 100004103 | 100004301 | CDC14A      | K562 |  0.074856093 |   0.12789128 |  0.58531036 | 0.55833905 |\n",
       "| chr1 | 100004103 | 100004301 | DBT         | K562 | -0.655628429 |   0.63963226 | -1.02500838 | 0.30535923 |\n",
       "| chr1 | 100004103 | 100004301 | EXTL2       | K562 | -0.345972818 |   0.88008305 | -0.39311383 | 0.69423541 |\n",
       "| chr1 | 100004103 | 100004301 | FRRS1       | K562 | -0.262113620 |   0.63339983 | -0.41382016 | 0.67900582 |\n",
       "| chr1 | 100004103 | 100004301 | MFSD14A     | K562 | -0.046556068 |   1.02902606 | -0.04524285 | 0.96391374 |\n",
       "| chr1 | 100004103 | 100004301 | PLPPR5      | K562 |  1.064230198 |   0.52153143 |  2.04058688 | 0.04129191 |\n",
       "| chr1 | 100004103 | 100004301 | RTCA        | K562 | -0.003990392 |   0.26336186 | -0.01515175 | 0.98791112 |\n",
       "| chr1 | 100004103 | 100004301 | RTCA-AS1    | K562 |  0.024454937 |   0.72139816 |  0.03389936 | 0.97295740 |\n",
       "| chr1 | 100004103 | 100004301 | SASS6       | K562 |  0.478958268 |   0.21958646 |  2.18118302 | 0.02916988 |\n",
       "| chr1 | 100004103 | 100004301 | SLC30A7     | K562 |  0.170716318 |   0.17185899 |  0.99335113 | 0.32053887 |\n",
       "| chr1 | 100004103 | 100004301 | SLC35A3     | K562 | -0.385101235 |   0.51159813 | -0.75274168 | 0.45160516 |\n",
       "| chr1 | 100004103 | 100004301 | TRMT13      | K562 | -0.473817506 |   0.45064019 | -1.05143198 | 0.29306024 |\n",
       "| chr1 | 100010532 | 100010875 | AGL         | K562 |  0.100658564 |   0.13176197 |  0.76394247 | 0.44490152 |\n",
       "| chr1 | 100010532 | 100010875 | CDC14A      | K562 |  0.061930828 |   0.06180056 |  1.00210785 | 0.31629151 |\n",
       "| chr1 | 100010532 | 100010875 | DBT         | K562 |  0.128080932 |   0.15167720 |  0.84443101 | 0.39842860 |\n",
       "| chr1 | 100010532 | 100010875 | EXTL2       | K562 |  0.271037154 |   0.24230387 |  1.11858367 | 0.26331779 |\n",
       "| chr1 | 100010532 | 100010875 | FRRS1       | K562 |  0.194968190 |   0.20396612 |  0.95588518 | 0.33913025 |\n",
       "| chr1 | 100010532 | 100010875 | GPR88       | K562 | -9.326765521 | 314.00867345 | -0.02970225 | 0.97630451 |\n",
       "| chr1 | 100010532 | 100010875 | MFSD14A     | K562 | -0.491938413 |   0.74430404 | -0.66093744 | 0.50865243 |\n",
       "| chr1 | 100010532 | 100010875 | RTCA        | K562 | -0.041605736 |   0.13041590 | -0.31902349 | 0.74970870 |\n",
       "| chr1 | 100010532 | 100010875 | RTCA-AS1    | K562 |  0.456613536 |   0.22870540 |  1.99651400 | 0.04587800 |\n",
       "| chr1 | 100010532 | 100010875 | SASS6       | K562 | -0.013200767 |   0.16147591 | -0.08175069 | 0.93484497 |\n",
       "| chr1 | 100010532 | 100010875 | SLC30A7     | K562 |  0.104259684 |   0.08633671 |  1.20759388 | 0.22720351 |\n",
       "| chr1 | 100010532 | 100010875 | SLC35A3     | K562 |  0.132256033 |   0.15583125 |  0.84871317 | 0.39604092 |\n",
       "| chr1 | 100010532 | 100010875 | TRMT13      | K562 | -0.026169687 |   0.14566140 | -0.17966110 | 0.85741863 |\n",
       "| chr1 |  10001440 |  10001830 | APITD1-CORT | K562 |  1.001306699 |   0.55288905 |  1.81104454 | 0.07013396 |\n",
       "| chr1 |  10001440 |  10001830 | CENPS       | K562 | -0.195202385 |   1.02687668 | -0.19009331 | 0.84923601 |\n",
       "| chr1 |  10001440 |  10001830 | CLSTN1      | K562 |  0.004962965 |   0.14622498 |  0.03394061 | 0.97292451 |\n",
       "| chr1 |  10001440 |  10001830 | CTNNBIP1    | K562 |  0.265333568 |   0.18006297 |  1.47355990 | 0.14060012 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  9992130 |  9992304 | WWC3      | K562 |   0.440577637 | 1.430916e-01 |  3.07898956 | 2.077040e-03 |\n",
       "| chrX |   999313 |  1000203 | AKAP17A   | K562 |   0.048439172 | 1.049150e-01 |  0.46169934 | 6.442969e-01 |\n",
       "| chrX |   999313 |  1000203 | ASMT      | K562 |   0.210805014 | 6.133866e-01 |  0.34367400 | 7.310915e-01 |\n",
       "| chrX |   999313 |  1000203 | ASMTL     | K562 |   0.156090099 | 1.063782e-01 |  1.46731252 | 1.422911e-01 |\n",
       "| chrX |   999313 |  1000203 | CSF2RA    | K562 |  -0.181719570 | 1.189538e-01 | -1.52764803 | 1.265999e-01 |\n",
       "| chrX |   999313 |  1000203 | GTPBP6    | K562 |  -0.026414961 | 1.419104e-01 | -0.18613835 | 8.523363e-01 |\n",
       "| chrX |   999313 |  1000203 | IL3RA     | K562 |   0.197814962 | 1.071269e-01 |  1.84654787 | 6.481270e-02 |\n",
       "| chrX |   999313 |  1000203 | LINC00106 | K562 |   0.185572928 | 1.698959e-01 |  1.09227425 | 2.747126e-01 |\n",
       "| chrX |   999313 |  1000203 | P2RY8     | K562 |  -0.011915616 | 5.739606e-02 | -0.20760336 | 8.355387e-01 |\n",
       "| chrX |   999313 |  1000203 | PLCXD1    | K562 |  -0.004166042 | 1.782306e-01 | -0.02337445 | 9.813516e-01 |\n",
       "| chrX |   999313 |  1000203 | PPP2R3B   | K562 |   0.631945853 | 1.126565e-01 |  5.60949430 | 2.029187e-08 |\n",
       "| chrX |   999313 |  1000203 | SHOX      | K562 | -10.203892238 | 2.353793e+02 | -0.04335084 | 9.654219e-01 |\n",
       "| chrX |   999313 |  1000203 | SLC25A6   | K562 |  -0.009238020 | 6.403557e-02 | -0.14426389 | 8.852921e-01 |\n",
       "| chrX |  9995566 |  9996680 | CLCN4     | K562 |   0.006258711 | 6.490849e-02 |  0.09642361 | 9.231841e-01 |\n",
       "| chrX |  9995566 |  9996680 | FAM9B     | K562 |   0.045752222 | 1.535455e-01 |  0.29797182 | 7.657247e-01 |\n",
       "| chrX |  9995566 |  9996680 | GPR143    | K562 |  -0.233540209 | 3.622750e-01 | -0.64464892 | 5.191547e-01 |\n",
       "| chrX |  9995566 |  9996680 | SHROOM2   | K562 |   0.118672260 | 1.887302e-01 |  0.62879301 | 5.294846e-01 |\n",
       "| chrX |  9995566 |  9996680 | TBL1X     | K562 |   0.026033112 | 2.059323e-02 |  1.26415902 | 2.061730e-01 |\n",
       "| chrX |  9995566 |  9996680 | WWC3      | K562 |   0.057659020 | 4.025994e-02 |  1.43216867 | 1.520956e-01 |\n",
       "| chrX |  9996996 |  9998157 | CLCN4     | K562 |  -0.228042253 | 1.467925e-01 | -1.55350059 | 1.203036e-01 |\n",
       "| chrX |  9996996 |  9998157 | GPR143    | K562 |  -0.544363049 | 8.909577e-01 | -0.61098640 | 5.412086e-01 |\n",
       "| chrX |  9996996 |  9998157 | SHROOM2   | K562 |   0.079025419 | 3.546140e-01 |  0.22284913 | 8.236529e-01 |\n",
       "| chrX |  9996996 |  9998157 | TBL1X     | K562 |   0.199258783 | 3.421840e-02 |  5.82314752 | 5.774954e-09 |\n",
       "| chrX |  9996996 |  9998157 | WWC3      | K562 |   0.052105985 | 7.336479e-02 |  0.71023154 | 4.775606e-01 |\n",
       "| chrX | 99980528 | 99980848 | CSTF2     | K562 |   0.421534303 | 5.526727e-01 |  0.76271959 | 4.456306e-01 |\n",
       "| chrX | 99980528 | 99980848 | PCDH19    | K562 | -15.345517138 | 1.241830e+03 | -0.01235718 | 9.901406e-01 |\n",
       "| chrX | 99980528 | 99980848 | TSPAN6    | K562 | -19.442721494 | 1.865082e+03 | -0.01042459 | 9.916825e-01 |\n",
       "| chrX | 99980528 | 99980848 | XKRX      | K562 | -15.904810429 | 1.210072e+03 | -0.01314369 | 9.895132e-01 |\n",
       "| chrX |  9999312 |  9999592 | TBL1X     | K562 |   0.160231973 | 1.020743e-01 |  1.56975785 | 1.164715e-01 |\n",
       "| chrX |  9999312 |  9999592 | WWC3      | K562 |   0.201931459 | 1.885116e-01 |  1.07118837 | 2.840847e-01 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene  CellType beta         \n",
       "1       chr1 100004103 100004301 AGL         K562      0.457131511 \n",
       "2       chr1 100004103 100004301 CDC14A      K562      0.074856093 \n",
       "3       chr1 100004103 100004301 DBT         K562     -0.655628429 \n",
       "4       chr1 100004103 100004301 EXTL2       K562     -0.345972818 \n",
       "5       chr1 100004103 100004301 FRRS1       K562     -0.262113620 \n",
       "6       chr1 100004103 100004301 MFSD14A     K562     -0.046556068 \n",
       "7       chr1 100004103 100004301 PLPPR5      K562      1.064230198 \n",
       "8       chr1 100004103 100004301 RTCA        K562     -0.003990392 \n",
       "9       chr1 100004103 100004301 RTCA-AS1    K562      0.024454937 \n",
       "10      chr1 100004103 100004301 SASS6       K562      0.478958268 \n",
       "11      chr1 100004103 100004301 SLC30A7     K562      0.170716318 \n",
       "12      chr1 100004103 100004301 SLC35A3     K562     -0.385101235 \n",
       "13      chr1 100004103 100004301 TRMT13      K562     -0.473817506 \n",
       "14      chr1 100010532 100010875 AGL         K562      0.100658564 \n",
       "15      chr1 100010532 100010875 CDC14A      K562      0.061930828 \n",
       "16      chr1 100010532 100010875 DBT         K562      0.128080932 \n",
       "17      chr1 100010532 100010875 EXTL2       K562      0.271037154 \n",
       "18      chr1 100010532 100010875 FRRS1       K562      0.194968190 \n",
       "19      chr1 100010532 100010875 GPR88       K562     -9.326765521 \n",
       "20      chr1 100010532 100010875 MFSD14A     K562     -0.491938413 \n",
       "21      chr1 100010532 100010875 RTCA        K562     -0.041605736 \n",
       "22      chr1 100010532 100010875 RTCA-AS1    K562      0.456613536 \n",
       "23      chr1 100010532 100010875 SASS6       K562     -0.013200767 \n",
       "24      chr1 100010532 100010875 SLC30A7     K562      0.104259684 \n",
       "25      chr1 100010532 100010875 SLC35A3     K562      0.132256033 \n",
       "26      chr1 100010532 100010875 TRMT13      K562     -0.026169687 \n",
       "27      chr1  10001440  10001830 APITD1-CORT K562      1.001306699 \n",
       "28      chr1  10001440  10001830 CENPS       K562     -0.195202385 \n",
       "29      chr1  10001440  10001830 CLSTN1      K562      0.004962965 \n",
       "30      chr1  10001440  10001830 CTNNBIP1    K562      0.265333568 \n",
       "⋮       ⋮    ⋮         ⋮         ⋮           ⋮        ⋮            \n",
       "9727145 chrX  9992130   9992304  WWC3        K562       0.440577637\n",
       "9727146 chrX   999313   1000203  AKAP17A     K562       0.048439172\n",
       "9727147 chrX   999313   1000203  ASMT        K562       0.210805014\n",
       "9727148 chrX   999313   1000203  ASMTL       K562       0.156090099\n",
       "9727149 chrX   999313   1000203  CSF2RA      K562      -0.181719570\n",
       "9727150 chrX   999313   1000203  GTPBP6      K562      -0.026414961\n",
       "9727151 chrX   999313   1000203  IL3RA       K562       0.197814962\n",
       "9727152 chrX   999313   1000203  LINC00106   K562       0.185572928\n",
       "9727153 chrX   999313   1000203  P2RY8       K562      -0.011915616\n",
       "9727154 chrX   999313   1000203  PLCXD1      K562      -0.004166042\n",
       "9727155 chrX   999313   1000203  PPP2R3B     K562       0.631945853\n",
       "9727156 chrX   999313   1000203  SHOX        K562     -10.203892238\n",
       "9727157 chrX   999313   1000203  SLC25A6     K562      -0.009238020\n",
       "9727158 chrX  9995566   9996680  CLCN4       K562       0.006258711\n",
       "9727159 chrX  9995566   9996680  FAM9B       K562       0.045752222\n",
       "9727160 chrX  9995566   9996680  GPR143      K562      -0.233540209\n",
       "9727161 chrX  9995566   9996680  SHROOM2     K562       0.118672260\n",
       "9727162 chrX  9995566   9996680  TBL1X       K562       0.026033112\n",
       "9727163 chrX  9995566   9996680  WWC3        K562       0.057659020\n",
       "9727164 chrX  9996996   9998157  CLCN4       K562      -0.228042253\n",
       "9727165 chrX  9996996   9998157  GPR143      K562      -0.544363049\n",
       "9727166 chrX  9996996   9998157  SHROOM2     K562       0.079025419\n",
       "9727167 chrX  9996996   9998157  TBL1X       K562       0.199258783\n",
       "9727168 chrX  9996996   9998157  WWC3        K562       0.052105985\n",
       "9727169 chrX 99980528  99980848  CSTF2       K562       0.421534303\n",
       "9727170 chrX 99980528  99980848  PCDH19      K562     -15.345517138\n",
       "9727171 chrX 99980528  99980848  TSPAN6      K562     -19.442721494\n",
       "9727172 chrX 99980528  99980848  XKRX        K562     -15.904810429\n",
       "9727173 chrX  9999312   9999592  TBL1X       K562       0.160231973\n",
       "9727174 chrX  9999312   9999592  WWC3        K562       0.201931459\n",
       "        sd           z           p           \n",
       "1         0.19953486  2.29098569 0.02196424  \n",
       "2         0.12789128  0.58531036 0.55833905  \n",
       "3         0.63963226 -1.02500838 0.30535923  \n",
       "4         0.88008305 -0.39311383 0.69423541  \n",
       "5         0.63339983 -0.41382016 0.67900582  \n",
       "6         1.02902606 -0.04524285 0.96391374  \n",
       "7         0.52153143  2.04058688 0.04129191  \n",
       "8         0.26336186 -0.01515175 0.98791112  \n",
       "9         0.72139816  0.03389936 0.97295740  \n",
       "10        0.21958646  2.18118302 0.02916988  \n",
       "11        0.17185899  0.99335113 0.32053887  \n",
       "12        0.51159813 -0.75274168 0.45160516  \n",
       "13        0.45064019 -1.05143198 0.29306024  \n",
       "14        0.13176197  0.76394247 0.44490152  \n",
       "15        0.06180056  1.00210785 0.31629151  \n",
       "16        0.15167720  0.84443101 0.39842860  \n",
       "17        0.24230387  1.11858367 0.26331779  \n",
       "18        0.20396612  0.95588518 0.33913025  \n",
       "19      314.00867345 -0.02970225 0.97630451  \n",
       "20        0.74430404 -0.66093744 0.50865243  \n",
       "21        0.13041590 -0.31902349 0.74970870  \n",
       "22        0.22870540  1.99651400 0.04587800  \n",
       "23        0.16147591 -0.08175069 0.93484497  \n",
       "24        0.08633671  1.20759388 0.22720351  \n",
       "25        0.15583125  0.84871317 0.39604092  \n",
       "26        0.14566140 -0.17966110 0.85741863  \n",
       "27        0.55288905  1.81104454 0.07013396  \n",
       "28        1.02687668 -0.19009331 0.84923601  \n",
       "29        0.14622498  0.03394061 0.97292451  \n",
       "30        0.18006297  1.47355990 0.14060012  \n",
       "⋮       ⋮            ⋮           ⋮           \n",
       "9727145 1.430916e-01  3.07898956 2.077040e-03\n",
       "9727146 1.049150e-01  0.46169934 6.442969e-01\n",
       "9727147 6.133866e-01  0.34367400 7.310915e-01\n",
       "9727148 1.063782e-01  1.46731252 1.422911e-01\n",
       "9727149 1.189538e-01 -1.52764803 1.265999e-01\n",
       "9727150 1.419104e-01 -0.18613835 8.523363e-01\n",
       "9727151 1.071269e-01  1.84654787 6.481270e-02\n",
       "9727152 1.698959e-01  1.09227425 2.747126e-01\n",
       "9727153 5.739606e-02 -0.20760336 8.355387e-01\n",
       "9727154 1.782306e-01 -0.02337445 9.813516e-01\n",
       "9727155 1.126565e-01  5.60949430 2.029187e-08\n",
       "9727156 2.353793e+02 -0.04335084 9.654219e-01\n",
       "9727157 6.403557e-02 -0.14426389 8.852921e-01\n",
       "9727158 6.490849e-02  0.09642361 9.231841e-01\n",
       "9727159 1.535455e-01  0.29797182 7.657247e-01\n",
       "9727160 3.622750e-01 -0.64464892 5.191547e-01\n",
       "9727161 1.887302e-01  0.62879301 5.294846e-01\n",
       "9727162 2.059323e-02  1.26415902 2.061730e-01\n",
       "9727163 4.025994e-02  1.43216867 1.520956e-01\n",
       "9727164 1.467925e-01 -1.55350059 1.203036e-01\n",
       "9727165 8.909577e-01 -0.61098640 5.412086e-01\n",
       "9727166 3.546140e-01  0.22284913 8.236529e-01\n",
       "9727167 3.421840e-02  5.82314752 5.774954e-09\n",
       "9727168 7.336479e-02  0.71023154 4.775606e-01\n",
       "9727169 5.526727e-01  0.76271959 4.456306e-01\n",
       "9727170 1.241830e+03 -0.01235718 9.901406e-01\n",
       "9727171 1.865082e+03 -0.01042459 9.916825e-01\n",
       "9727172 1.210072e+03 -0.01314369 9.895132e-01\n",
       "9727173 1.020743e-01  1.56975785 1.164715e-01\n",
       "9727174 1.885116e-01  1.07118837 2.840847e-01"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"beta\",\n",
    "                         \"sd\",\n",
    "                         \"z\",\n",
    "                         \"p\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] data.table_1.15.2    dplyr_1.1.4          foreach_1.5.2       \n",
       " [4] GenomicRanges_1.54.1 GenomeInfoDb_1.38.1  IRanges_2.36.0      \n",
       " [7] S4Vectors_0.40.2     BiocGenerics_0.48.1  genomation_1.34.0   \n",
       "[10] Signac_1.13.0        Seurat_5.0.3         SeuratObject_5.0.1  \n",
       "[13] sp_2.1-4             SnapATAC_1.0.0       rhdf5_2.44.0        \n",
       "[16] Matrix_1.6-5        \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 BiocIO_1.10.0              \n",
       "  [5] pbdZMQ_0.3-11               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                polyclip_1.10-6            \n",
       "  [9] XML_3.99-0.16.1             fastDummies_1.7.3          \n",
       " [11] lifecycle_1.0.4             tcltk_4.3.3                \n",
       " [13] edgeR_3.42.4                doParallel_1.0.17          \n",
       " [15] globals_0.16.3              lattice_0.22-6             \n",
       " [17] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [19] limma_3.56.2                plotly_4.10.4              \n",
       " [21] plotrix_3.8-4               yaml_2.3.8                 \n",
       " [23] plot3D_1.4.1                httpuv_1.6.15              \n",
       " [25] sctransform_0.4.1           spam_2.10-0                \n",
       " [27] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [29] cowplot_1.1.3               pbapply_1.7-2              \n",
       " [31] RColorBrewer_1.1-3          abind_1.4-5                \n",
       " [33] zlibbioc_1.48.0             Rtsne_0.17                 \n",
       " [35] purrr_1.0.2                 RCurl_1.98-1.14            \n",
       " [37] misc3d_0.9-1                GenomeInfoDbData_1.2.11    \n",
       " [39] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [41] listenv_0.9.1               spatstat.utils_3.0-4       \n",
       " [43] goftest_1.2-3               RSpectra_0.16-1            \n",
       " [45] bigmemory_4.6.4             spatstat.random_3.2-3      \n",
       " [47] fitdistrplus_1.1-11         parallelly_1.37.1          \n",
       " [49] DelayedArray_0.26.7         leiden_0.4.3.1             \n",
       " [51] codetools_0.2-20            RcppRoll_0.3.0             \n",
       " [53] tidyselect_1.2.1            viridis_0.6.5              \n",
       " [55] matrixStats_1.3.0           base64enc_0.1-3            \n",
       " [57] spatstat.explore_3.2-6      GenomicAlignments_1.36.0   \n",
       " [59] jsonlite_1.8.8              progressr_0.14.0           \n",
       " [61] ggridges_0.5.6              survival_3.6-4             \n",
       " [63] iterators_1.0.14            tools_4.3.3                \n",
       " [65] snow_0.4-4                  ica_1.0-3                  \n",
       " [67] Rcpp_1.0.12                 glue_1.7.0                 \n",
       " [69] gridExtra_2.3               MatrixGenerics_1.12.3      \n",
       " [71] IRdisplay_1.1               withr_3.0.0                \n",
       " [73] fastmap_1.1.1               rhdf5filters_1.12.1        \n",
       " [75] fansi_1.0.6                 digest_0.6.35              \n",
       " [77] R6_2.5.1                    mime_0.12                  \n",
       " [79] seqPattern_1.34.0           colorspace_2.1-0           \n",
       " [81] scattermore_1.2             tensor_1.5                 \n",
       " [83] spatstat.data_3.0-4         utf8_1.2.4                 \n",
       " [85] tidyr_1.3.1                 generics_0.1.3             \n",
       " [87] rtracklayer_1.60.1          S4Arrays_1.2.1             \n",
       " [89] httr_1.4.7                  htmlwidgets_1.6.4          \n",
       " [91] uwot_0.1.16                 pkgconfig_2.0.3            \n",
       " [93] gtable_0.3.5                impute_1.76.0              \n",
       " [95] lmtest_0.9-40               XVector_0.42.0             \n",
       " [97] htmltools_0.5.8.1           dotCall64_1.1-1            \n",
       " [99] Biobase_2.60.0              scales_1.3.0               \n",
       "[101] png_0.1-8                   doSNOW_1.0.20              \n",
       "[103] bigmemory.sri_0.1.8         tzdb_0.4.0                 \n",
       "[105] rjson_0.2.21                reshape2_1.4.4             \n",
       "[107] uuid_1.2-0                  nlme_3.1-164               \n",
       "[109] repr_1.1.7                  zoo_1.8-12                 \n",
       "[111] stringr_1.5.1               KernSmooth_2.23-22         \n",
       "[113] miniUI_0.1.1.1              restfulr_0.0.15            \n",
       "[115] pillar_1.9.0                vctrs_0.6.5                \n",
       "[117] RANN_2.6.1                  promises_1.3.0             \n",
       "[119] xtable_1.8-4                cluster_2.1.6              \n",
       "[121] evaluate_0.23               readr_2.1.5                \n",
       "[123] cli_3.6.2                   locfit_1.5-9.9             \n",
       "[125] compiler_4.3.3              Rsamtools_2.18.0           \n",
       "[127] rlang_1.1.3                 crayon_1.5.2               \n",
       "[129] future.apply_1.11.2         plyr_1.8.9                 \n",
       "[131] stringi_1.8.3               gridBase_0.4-7             \n",
       "[133] viridisLite_0.4.2           deldir_2.0-4               \n",
       "[135] BiocParallel_1.36.0         munsell_0.5.1              \n",
       "[137] Biostrings_2.70.1           lazyeval_0.2.2             \n",
       "[139] spatstat.geom_3.2-9         IRkernel_1.3.2             \n",
       "[141] BSgenome_1.68.0             RcppHNSW_0.6.0             \n",
       "[143] hms_1.1.3                   patchwork_1.2.0            \n",
       "[145] future_1.33.2               ggplot2_3.5.1              \n",
       "[147] Rhdf5lib_1.22.1             shiny_1.8.1.1              \n",
       "[149] SummarizedExperiment_1.30.2 ROCR_1.0-11                \n",
       "[151] igraph_2.0.3                fastmatch_1.1-4            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
