{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: rhdf5\n",
      "\n",
      "Warning message:\n",
      "“no DISPLAY variable so Tk is not available”\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SnapATAC)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(parallel)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/Jurkat/Jurkat/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/Jurkat/Jurkat/multiome_powerlaw_v3/encode_e2g_predictions.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/Jurkat/Jurkat/Kendall/atac_matrix.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/scE2G_input_Maya/Jurkat/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/Jurkat/1.Genome_wide_prediction/SnapATAC/SnapATAC.250321/\"\n",
    "celltype = \"Jurkat\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2793593 ranges and 3 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10001055-10001570      * | APITD1-CORT\n",
       "        [2]     chr1 10001055-10001570      * |       CASZ1\n",
       "        [3]     chr1 10001055-10001570      * |       CENPS\n",
       "        [4]     chr1 10001055-10001570      * |      CLSTN1\n",
       "        [5]     chr1 10001055-10001570      * |    CTNNBIP1\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [2793589]     chrX   9941883-9942743      * |        WWC3\n",
       "  [2793590]     chrX 99603024-99603776      * |      PCDH19\n",
       "  [2793591]     chrX 99632013-99632219      * |      PCDH19\n",
       "  [2793592]     chrX   9995829-9996207      * |       TBL1X\n",
       "  [2793593]     chrX   9995829-9996207      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-10001055-10001570 chr1-10001055-100015..\n",
       "        [2] chr1-10001055-10001570 chr1-10001055-100015..\n",
       "        [3] chr1-10001055-10001570 chr1-10001055-100015..\n",
       "        [4] chr1-10001055-10001570 chr1-10001055-100015..\n",
       "        [5] chr1-10001055-10001570 chr1-10001055-100015..\n",
       "        ...                    ...                    ...\n",
       "  [2793589]   chrX-9941883-9942743 chrX-9941883-9942743..\n",
       "  [2793590] chrX-99603024-99603776 chrX-99603024-996037..\n",
       "  [2793591] chrX-99632013-99632219 chrX-99632013-996322..\n",
       "  [2793592]   chrX-9995829-9996207 chrX-9995829-9996207..\n",
       "  [2793593]   chrX-9995829-9996207 chrX-9995829-9996207..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = BinarizeCounts(matrix.atac_count)\n",
    "rm(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna_count = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna_count)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4986774c-cbbf-4b72-a299-d681bddca80e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "85b165d0-f1db-4548-beac-235f8950107b",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "655ebba3-731e-49d8-a04e-1ebb5becd567",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Run SnapATAC prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "Sys.setenv(OPENBLAS_NUM_THREADS=1)\n",
    "PredictionSnapATAC = function(pairs.E2G,\n",
    "                              data.RNA,\n",
    "                              data.ATAC,\n",
    "                              n.core = 16){\n",
    "  my.cluster <- parallel::makeCluster(\n",
    "    n.core,\n",
    "    type = \"PSOCK\"\n",
    "  )\n",
    "  doParallel::registerDoParallel(cl = my.cluster)\n",
    "  \n",
    "  pairs.E2G =  \n",
    "    foreach (index.tmp = 1:length(pairs.E2G),\n",
    "             .combine = 'c',\n",
    "             .packages = c(\"Matrix\",\n",
    "                           \"GenomicRanges\")) %dopar% {\n",
    "\n",
    "                             pairs.E2G.tmp = pairs.E2G[index.tmp]\n",
    "                             \n",
    "                             model.tmp = summary(stats::glm(data.ATAC[pairs.E2G.tmp$PeakName,] ~ data.RNA[pairs.E2G.tmp$TargetGene,], \n",
    "                                                            family = binomial(link = \"logit\")))[[\"coefficients\"]]\n",
    "                             mcols(pairs.E2G.tmp)[,c(\"beta\",\"sd\",\"z\",\"p\")] = \n",
    "                               t(data.frame(model.tmp[2,]))\n",
    "                             \n",
    "                             pairs.E2G.tmp\n",
    "                           }\n",
    "  \n",
    "  parallel::stopCluster(cl = my.cluster)\n",
    "  \n",
    "  return(pairs.E2G)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "[1] \"chr10\"\n",
      "[1] \"chr11\"\n",
      "[1] \"chr12\"\n",
      "[1] \"chr13\"\n",
      "[1] \"chr14\"\n",
      "[1] \"chr15\"\n",
      "[1] \"chr16\"\n",
      "[1] \"chr17\"\n",
      "[1] \"chr18\"\n",
      "[1] \"chr19\"\n",
      "[1] \"chr2\"\n",
      "[1] \"chr20\"\n",
      "[1] \"chr21\"\n",
      "[1] \"chr22\"\n",
      "[1] \"chr3\"\n",
      "[1] \"chr4\"\n",
      "[1] \"chr5\"\n",
      "[1] \"chr6\"\n",
      "[1] \"chr7\"\n",
      "[1] \"chr8\"\n",
      "[1] \"chr9\"\n",
      "[1] \"chrX\"\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  mcols(pairs.E2G.chr.res)[,c(\"beta\",\"sd\",\"z\",\"p\")] = NA\n",
    "  pairs.E2G.chr.res = PredictionSnapATAC(pairs.E2G.chr.res,\n",
    "                                         matrix.rna.rename[rownames(matrix.rna.rename) %in% pairs.E2G.chr.res$TargetGene,],\n",
    "                                         matrix.atac[rownames(matrix.atac) %in% pairs.E2G.chr.res$PeakName,],\n",
    "                                         n.cores)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2590409 ranges and 7 metadata columns:\n",
       "       seqnames            ranges strand |  TargetGene               PeakName\n",
       "          <Rle>         <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 10001055-10001570      * | APITD1-CORT chr1-10001055-10001570\n",
       "  chr1     chr1 10001055-10001570      * |       CASZ1 chr1-10001055-10001570\n",
       "  chr1     chr1 10001055-10001570      * |      CLSTN1 chr1-10001055-10001570\n",
       "  chr1     chr1 10001055-10001570      * |    CTNNBIP1 chr1-10001055-10001570\n",
       "  chr1     chr1 10001055-10001570      * |        DFFA chr1-10001055-10001570\n",
       "   ...      ...               ...    ... .         ...                    ...\n",
       "  chrX     chrX   9941883-9942743      * |        WWC3   chrX-9941883-9942743\n",
       "  chrX     chrX 99603024-99603776      * |      PCDH19 chrX-99603024-99603776\n",
       "  chrX     chrX 99632013-99632219      * |      PCDH19 chrX-99632013-99632219\n",
       "  chrX     chrX   9995829-9996207      * |       TBL1X   chrX-9995829-9996207\n",
       "  chrX     chrX   9995829-9996207      * |        WWC3   chrX-9995829-9996207\n",
       "                     PairName       beta          sd          z         p\n",
       "                  <character>  <numeric>   <numeric>  <numeric> <numeric>\n",
       "  chr1 chr1-10001055-100015.. -23.335571 1404.802952 -0.0166113 0.9867467\n",
       "  chr1 chr1-10001055-100015..  -0.371286    1.241065 -0.2991671 0.7648125\n",
       "  chr1 chr1-10001055-100015..   0.451226    0.300843  1.4998742 0.1336470\n",
       "  chr1 chr1-10001055-100015..  -1.069015    0.568060 -1.8818689 0.0598538\n",
       "  chr1 chr1-10001055-100015..  -0.698861    0.493907 -1.4149634 0.1570792\n",
       "   ...                    ...        ...         ...        ...       ...\n",
       "  chrX chrX-9941883-9942743..   0.274618    0.320617  0.8565292  0.391705\n",
       "  chrX chrX-99603024-996037.. -36.197595 2881.754293 -0.0125610  0.989978\n",
       "  chrX chrX-99632013-996322.. -32.303206 2867.434427 -0.0112655  0.991012\n",
       "  chrX chrX-9995829-9996207..  -0.137887    0.255164 -0.5403861  0.588931\n",
       "  chrX chrX-9995829-9996207..   0.439169    0.471568  0.9312952  0.351701\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2590409 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>beta</th><th scope=col>sd</th><th scope=col>z</th><th scope=col>p</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>APITD1-CORT</td><td>Jurkat</td><td>-23.33557068</td><td>1.404803e+03</td><td>-0.01661128</td><td>0.98674673</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>CASZ1      </td><td>Jurkat</td><td> -0.37128583</td><td>1.241065e+00</td><td>-0.29916714</td><td>0.76481252</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>CLSTN1     </td><td>Jurkat</td><td>  0.45122592</td><td>3.008425e-01</td><td> 1.49987419</td><td>0.13364699</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>CTNNBIP1   </td><td>Jurkat</td><td> -1.06901479</td><td>5.680602e-01</td><td>-1.88186895</td><td>0.05985381</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>DFFA       </td><td>Jurkat</td><td> -0.69886070</td><td>4.939073e-01</td><td>-1.41496340</td><td>0.15707923</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>GPR157     </td><td>Jurkat</td><td>  0.72019151</td><td>6.347292e-01</td><td> 1.13464375</td><td>0.25652462</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>H6PD       </td><td>Jurkat</td><td> -0.32546245</td><td>8.083685e-01</td><td>-0.40261643</td><td>0.68723042</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>KIF1B      </td><td>Jurkat</td><td>  0.17339554</td><td>2.951582e-01</td><td> 0.58746648</td><td>0.55689045</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>LINC01759  </td><td>Jurkat</td><td>-16.73549270</td><td>1.180793e+03</td><td>-0.01417310</td><td>0.98869188</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>LZIC       </td><td>Jurkat</td><td> -0.83144376</td><td>4.999534e-01</td><td>-1.66304241</td><td>0.09630395</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>MIR34AHG   </td><td>Jurkat</td><td>-15.97593779</td><td>1.025017e+03</td><td>-0.01558602</td><td>0.98756466</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>NMNAT1     </td><td>Jurkat</td><td> -0.48361336</td><td>1.107481e+00</td><td>-0.43667865</td><td>0.66234441</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>PEX14      </td><td>Jurkat</td><td>  0.25627843</td><td>2.970535e-01</td><td> 0.86273481</td><td>0.38828329</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>PGD        </td><td>Jurkat</td><td> -0.05181250</td><td>2.936617e-01</td><td>-0.17643600</td><td>0.85995143</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>PIK3CD     </td><td>Jurkat</td><td>  0.49548326</td><td>4.512738e-01</td><td> 1.09796588</td><td>0.27221939</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>SLC25A33   </td><td>Jurkat</td><td> -0.02200935</td><td>3.805832e-01</td><td>-0.05783059</td><td>0.95388357</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>SPSB1      </td><td>Jurkat</td><td> -0.02086543</td><td>1.166714e+00</td><td>-0.01788392</td><td>0.98573145</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>TMEM201    </td><td>Jurkat</td><td>  0.55538350</td><td>5.921901e-01</td><td> 0.93784658</td><td>0.34832326</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>UBE4B      </td><td>Jurkat</td><td>  0.30821438</td><td>2.863096e-01</td><td> 1.07650747</td><td>0.28170036</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>AGL        </td><td>Jurkat</td><td>  0.00928901</td><td>5.785458e-02</td><td> 0.16055789</td><td>0.87244162</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>CDC14A     </td><td>Jurkat</td><td>  0.03767467</td><td>5.340894e-02</td><td> 0.70540004</td><td>0.48056132</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>DBT        </td><td>Jurkat</td><td>  0.03004789</td><td>7.126554e-02</td><td> 0.42163275</td><td>0.67329310</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>DPH5       </td><td>Jurkat</td><td>  0.10085758</td><td>6.752651e-02</td><td> 1.49359975</td><td>0.13528026</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>EXTL2      </td><td>Jurkat</td><td>  0.12930622</td><td>6.765497e-02</td><td> 1.91125955</td><td>0.05597123</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>FRRS1      </td><td>Jurkat</td><td>  0.10765048</td><td>1.030189e-01</td><td> 1.04495879</td><td>0.29604202</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>MFSD14A    </td><td>Jurkat</td><td>-11.09340350</td><td>2.071577e+02</td><td>-0.05355052</td><td>0.95729328</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>RTCA       </td><td>Jurkat</td><td>  0.04336320</td><td>6.673008e-02</td><td> 0.64982990</td><td>0.51580210</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>RTCA-AS1   </td><td>Jurkat</td><td>  0.09365090</td><td>1.404237e-01</td><td> 0.66691653</td><td>0.50482545</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>SASS6      </td><td>Jurkat</td><td>  0.02674776</td><td>5.497944e-02</td><td> 0.48650478</td><td>0.62660931</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>SLC30A7    </td><td>Jurkat</td><td>  0.02562122</td><td>4.747476e-02</td><td> 0.53968085</td><td>0.58941715</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>IL3RA    </td><td>Jurkat</td><td>-28.60413936</td><td>2.064815e+03</td><td>-0.01385312</td><td>0.98894716</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>LINC00106</td><td>Jurkat</td><td>  0.19482473</td><td>1.314032e+00</td><td> 0.14826488</td><td>0.88213373</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>P2RY8    </td><td>Jurkat</td><td>  0.33777979</td><td>3.447763e-01</td><td> 0.97970711</td><td>0.32723072</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>PLCXD1   </td><td>Jurkat</td><td>  0.08742649</td><td>6.724470e-01</td><td> 0.13001247</td><td>0.89655656</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>PPP2R3B  </td><td>Jurkat</td><td> -0.88269523</td><td>9.916087e-01</td><td>-0.89016486</td><td>0.37337737</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>SHOX     </td><td>Jurkat</td><td>-19.23873281</td><td>1.636486e+03</td><td>-0.01175613</td><td>0.99062018</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>SLC25A6  </td><td>Jurkat</td><td> -0.28160142</td><td>3.379915e-01</td><td>-0.83316135</td><td>0.40475374</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9842372</td><td> 9843073</td><td>TBL1X    </td><td>Jurkat</td><td> -0.12525948</td><td>2.575620e-01</td><td>-0.48632741</td><td>0.62673505</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9842372</td><td> 9843073</td><td>WWC3     </td><td>Jurkat</td><td>-47.29171095</td><td>3.515520e+03</td><td>-0.01345227</td><td>0.98926697</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9873846</td><td> 9874297</td><td>TBL1X    </td><td>Jurkat</td><td> -0.51419511</td><td>3.559851e-01</td><td>-1.44442889</td><td>0.14861837</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9873846</td><td> 9874297</td><td>WWC3     </td><td>Jurkat</td><td> -0.06005241</td><td>7.478092e-01</td><td>-0.08030445</td><td>0.93599512</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9885186</td><td> 9885487</td><td>TBL1X    </td><td>Jurkat</td><td>  0.28033199</td><td>2.449863e-01</td><td> 1.14427622</td><td>0.25250910</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9885186</td><td> 9885487</td><td>WWC3     </td><td>Jurkat</td><td> -0.22017978</td><td>6.495738e-01</td><td>-0.33896039</td><td>0.73463957</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>AKAP17A  </td><td>Jurkat</td><td>  0.03569388</td><td>1.310651e-01</td><td> 0.27233710</td><td>0.78536283</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>ASMT     </td><td>Jurkat</td><td>  1.61163806</td><td>7.926542e-01</td><td> 2.03321721</td><td>0.04203058</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>ASMTL    </td><td>Jurkat</td><td> -0.02572605</td><td>1.645188e-01</td><td>-0.15637153</td><td>0.87574017</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>GTPBP6   </td><td>Jurkat</td><td>  0.17701549</td><td>1.416908e-01</td><td> 1.24930801</td><td>0.21155244</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>IL3RA    </td><td>Jurkat</td><td>  0.31105639</td><td>2.533586e-01</td><td> 1.22773186</td><td>0.21954764</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>LINC00106</td><td>Jurkat</td><td>  0.18187483</td><td>2.871707e-01</td><td> 0.63333351</td><td>0.52651587</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>P2RY8    </td><td>Jurkat</td><td> -0.06355849</td><td>8.023537e-02</td><td>-0.79215057</td><td>0.42827288</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>PLCXD1   </td><td>Jurkat</td><td>  0.24683237</td><td>1.378217e-01</td><td> 1.79095469</td><td>0.07330057</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>PPP2R3B  </td><td>Jurkat</td><td>  0.05430420</td><td>1.440059e-01</td><td> 0.37709704</td><td>0.70610149</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>SHOX     </td><td>Jurkat</td><td> -0.14242059</td><td>1.098825e+00</td><td>-0.12961176</td><td>0.89687360</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>SLC25A6  </td><td>Jurkat</td><td> -0.05655356</td><td>6.920936e-02</td><td>-0.81713744</td><td>0.41384989</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9941883</td><td> 9942743</td><td>TBL1X    </td><td>Jurkat</td><td>  0.09851419</td><td>1.542987e-01</td><td> 0.63846423</td><td>0.52317153</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9941883</td><td> 9942743</td><td>WWC3     </td><td>Jurkat</td><td>  0.27461778</td><td>3.206169e-01</td><td> 0.85652925</td><td>0.39170510</td></tr>\n",
       "\t<tr><td>chrX</td><td>99603024</td><td>99603776</td><td>PCDH19   </td><td>Jurkat</td><td>-36.19759507</td><td>2.881754e+03</td><td>-0.01256096</td><td>0.98997807</td></tr>\n",
       "\t<tr><td>chrX</td><td>99632013</td><td>99632219</td><td>PCDH19   </td><td>Jurkat</td><td>-32.30320638</td><td>2.867434e+03</td><td>-0.01126554</td><td>0.99101159</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995829</td><td> 9996207</td><td>TBL1X    </td><td>Jurkat</td><td> -0.13788683</td><td>2.551635e-01</td><td>-0.54038610</td><td>0.58893079</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995829</td><td> 9996207</td><td>WWC3     </td><td>Jurkat</td><td>  0.43916916</td><td>4.715682e-01</td><td> 0.93129520</td><td>0.35170089</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2590409 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & beta & sd & z & p\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  10001055 &  10001570 & APITD1-CORT & Jurkat & -23.33557068 & 1.404803e+03 & -0.01661128 & 0.98674673\\\\\n",
       "\t chr1 &  10001055 &  10001570 & CASZ1       & Jurkat &  -0.37128583 & 1.241065e+00 & -0.29916714 & 0.76481252\\\\\n",
       "\t chr1 &  10001055 &  10001570 & CLSTN1      & Jurkat &   0.45122592 & 3.008425e-01 &  1.49987419 & 0.13364699\\\\\n",
       "\t chr1 &  10001055 &  10001570 & CTNNBIP1    & Jurkat &  -1.06901479 & 5.680602e-01 & -1.88186895 & 0.05985381\\\\\n",
       "\t chr1 &  10001055 &  10001570 & DFFA        & Jurkat &  -0.69886070 & 4.939073e-01 & -1.41496340 & 0.15707923\\\\\n",
       "\t chr1 &  10001055 &  10001570 & GPR157      & Jurkat &   0.72019151 & 6.347292e-01 &  1.13464375 & 0.25652462\\\\\n",
       "\t chr1 &  10001055 &  10001570 & H6PD        & Jurkat &  -0.32546245 & 8.083685e-01 & -0.40261643 & 0.68723042\\\\\n",
       "\t chr1 &  10001055 &  10001570 & KIF1B       & Jurkat &   0.17339554 & 2.951582e-01 &  0.58746648 & 0.55689045\\\\\n",
       "\t chr1 &  10001055 &  10001570 & LINC01759   & Jurkat & -16.73549270 & 1.180793e+03 & -0.01417310 & 0.98869188\\\\\n",
       "\t chr1 &  10001055 &  10001570 & LZIC        & Jurkat &  -0.83144376 & 4.999534e-01 & -1.66304241 & 0.09630395\\\\\n",
       "\t chr1 &  10001055 &  10001570 & MIR34AHG    & Jurkat & -15.97593779 & 1.025017e+03 & -0.01558602 & 0.98756466\\\\\n",
       "\t chr1 &  10001055 &  10001570 & NMNAT1      & Jurkat &  -0.48361336 & 1.107481e+00 & -0.43667865 & 0.66234441\\\\\n",
       "\t chr1 &  10001055 &  10001570 & PEX14       & Jurkat &   0.25627843 & 2.970535e-01 &  0.86273481 & 0.38828329\\\\\n",
       "\t chr1 &  10001055 &  10001570 & PGD         & Jurkat &  -0.05181250 & 2.936617e-01 & -0.17643600 & 0.85995143\\\\\n",
       "\t chr1 &  10001055 &  10001570 & PIK3CD      & Jurkat &   0.49548326 & 4.512738e-01 &  1.09796588 & 0.27221939\\\\\n",
       "\t chr1 &  10001055 &  10001570 & SLC25A33    & Jurkat &  -0.02200935 & 3.805832e-01 & -0.05783059 & 0.95388357\\\\\n",
       "\t chr1 &  10001055 &  10001570 & SPSB1       & Jurkat &  -0.02086543 & 1.166714e+00 & -0.01788392 & 0.98573145\\\\\n",
       "\t chr1 &  10001055 &  10001570 & TMEM201     & Jurkat &   0.55538350 & 5.921901e-01 &  0.93784658 & 0.34832326\\\\\n",
       "\t chr1 &  10001055 &  10001570 & UBE4B       & Jurkat &   0.30821438 & 2.863096e-01 &  1.07650747 & 0.28170036\\\\\n",
       "\t chr1 & 100037799 & 100039086 & AGL         & Jurkat &   0.00928901 & 5.785458e-02 &  0.16055789 & 0.87244162\\\\\n",
       "\t chr1 & 100037799 & 100039086 & CDC14A      & Jurkat &   0.03767467 & 5.340894e-02 &  0.70540004 & 0.48056132\\\\\n",
       "\t chr1 & 100037799 & 100039086 & DBT         & Jurkat &   0.03004789 & 7.126554e-02 &  0.42163275 & 0.67329310\\\\\n",
       "\t chr1 & 100037799 & 100039086 & DPH5        & Jurkat &   0.10085758 & 6.752651e-02 &  1.49359975 & 0.13528026\\\\\n",
       "\t chr1 & 100037799 & 100039086 & EXTL2       & Jurkat &   0.12930622 & 6.765497e-02 &  1.91125955 & 0.05597123\\\\\n",
       "\t chr1 & 100037799 & 100039086 & FRRS1       & Jurkat &   0.10765048 & 1.030189e-01 &  1.04495879 & 0.29604202\\\\\n",
       "\t chr1 & 100037799 & 100039086 & MFSD14A     & Jurkat & -11.09340350 & 2.071577e+02 & -0.05355052 & 0.95729328\\\\\n",
       "\t chr1 & 100037799 & 100039086 & RTCA        & Jurkat &   0.04336320 & 6.673008e-02 &  0.64982990 & 0.51580210\\\\\n",
       "\t chr1 & 100037799 & 100039086 & RTCA-AS1    & Jurkat &   0.09365090 & 1.404237e-01 &  0.66691653 & 0.50482545\\\\\n",
       "\t chr1 & 100037799 & 100039086 & SASS6       & Jurkat &   0.02674776 & 5.497944e-02 &  0.48650478 & 0.62660931\\\\\n",
       "\t chr1 & 100037799 & 100039086 & SLC30A7     & Jurkat &   0.02562122 & 4.747476e-02 &  0.53968085 & 0.58941715\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &   978694 &   979118 & IL3RA     & Jurkat & -28.60413936 & 2.064815e+03 & -0.01385312 & 0.98894716\\\\\n",
       "\t chrX &   978694 &   979118 & LINC00106 & Jurkat &   0.19482473 & 1.314032e+00 &  0.14826488 & 0.88213373\\\\\n",
       "\t chrX &   978694 &   979118 & P2RY8     & Jurkat &   0.33777979 & 3.447763e-01 &  0.97970711 & 0.32723072\\\\\n",
       "\t chrX &   978694 &   979118 & PLCXD1    & Jurkat &   0.08742649 & 6.724470e-01 &  0.13001247 & 0.89655656\\\\\n",
       "\t chrX &   978694 &   979118 & PPP2R3B   & Jurkat &  -0.88269523 & 9.916087e-01 & -0.89016486 & 0.37337737\\\\\n",
       "\t chrX &   978694 &   979118 & SHOX      & Jurkat & -19.23873281 & 1.636486e+03 & -0.01175613 & 0.99062018\\\\\n",
       "\t chrX &   978694 &   979118 & SLC25A6   & Jurkat &  -0.28160142 & 3.379915e-01 & -0.83316135 & 0.40475374\\\\\n",
       "\t chrX &  9842372 &  9843073 & TBL1X     & Jurkat &  -0.12525948 & 2.575620e-01 & -0.48632741 & 0.62673505\\\\\n",
       "\t chrX &  9842372 &  9843073 & WWC3      & Jurkat & -47.29171095 & 3.515520e+03 & -0.01345227 & 0.98926697\\\\\n",
       "\t chrX &  9873846 &  9874297 & TBL1X     & Jurkat &  -0.51419511 & 3.559851e-01 & -1.44442889 & 0.14861837\\\\\n",
       "\t chrX &  9873846 &  9874297 & WWC3      & Jurkat &  -0.06005241 & 7.478092e-01 & -0.08030445 & 0.93599512\\\\\n",
       "\t chrX &  9885186 &  9885487 & TBL1X     & Jurkat &   0.28033199 & 2.449863e-01 &  1.14427622 & 0.25250910\\\\\n",
       "\t chrX &  9885186 &  9885487 & WWC3      & Jurkat &  -0.22017978 & 6.495738e-01 & -0.33896039 & 0.73463957\\\\\n",
       "\t chrX &   989869 &   990628 & AKAP17A   & Jurkat &   0.03569388 & 1.310651e-01 &  0.27233710 & 0.78536283\\\\\n",
       "\t chrX &   989869 &   990628 & ASMT      & Jurkat &   1.61163806 & 7.926542e-01 &  2.03321721 & 0.04203058\\\\\n",
       "\t chrX &   989869 &   990628 & ASMTL     & Jurkat &  -0.02572605 & 1.645188e-01 & -0.15637153 & 0.87574017\\\\\n",
       "\t chrX &   989869 &   990628 & GTPBP6    & Jurkat &   0.17701549 & 1.416908e-01 &  1.24930801 & 0.21155244\\\\\n",
       "\t chrX &   989869 &   990628 & IL3RA     & Jurkat &   0.31105639 & 2.533586e-01 &  1.22773186 & 0.21954764\\\\\n",
       "\t chrX &   989869 &   990628 & LINC00106 & Jurkat &   0.18187483 & 2.871707e-01 &  0.63333351 & 0.52651587\\\\\n",
       "\t chrX &   989869 &   990628 & P2RY8     & Jurkat &  -0.06355849 & 8.023537e-02 & -0.79215057 & 0.42827288\\\\\n",
       "\t chrX &   989869 &   990628 & PLCXD1    & Jurkat &   0.24683237 & 1.378217e-01 &  1.79095469 & 0.07330057\\\\\n",
       "\t chrX &   989869 &   990628 & PPP2R3B   & Jurkat &   0.05430420 & 1.440059e-01 &  0.37709704 & 0.70610149\\\\\n",
       "\t chrX &   989869 &   990628 & SHOX      & Jurkat &  -0.14242059 & 1.098825e+00 & -0.12961176 & 0.89687360\\\\\n",
       "\t chrX &   989869 &   990628 & SLC25A6   & Jurkat &  -0.05655356 & 6.920936e-02 & -0.81713744 & 0.41384989\\\\\n",
       "\t chrX &  9941883 &  9942743 & TBL1X     & Jurkat &   0.09851419 & 1.542987e-01 &  0.63846423 & 0.52317153\\\\\n",
       "\t chrX &  9941883 &  9942743 & WWC3      & Jurkat &   0.27461778 & 3.206169e-01 &  0.85652925 & 0.39170510\\\\\n",
       "\t chrX & 99603024 & 99603776 & PCDH19    & Jurkat & -36.19759507 & 2.881754e+03 & -0.01256096 & 0.98997807\\\\\n",
       "\t chrX & 99632013 & 99632219 & PCDH19    & Jurkat & -32.30320638 & 2.867434e+03 & -0.01126554 & 0.99101159\\\\\n",
       "\t chrX &  9995829 &  9996207 & TBL1X     & Jurkat &  -0.13788683 & 2.551635e-01 & -0.54038610 & 0.58893079\\\\\n",
       "\t chrX &  9995829 &  9996207 & WWC3      & Jurkat &   0.43916916 & 4.715682e-01 &  0.93129520 & 0.35170089\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2590409 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | beta &lt;dbl&gt; | sd &lt;dbl&gt; | z &lt;dbl&gt; | p &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  10001055 |  10001570 | APITD1-CORT | Jurkat | -23.33557068 | 1.404803e+03 | -0.01661128 | 0.98674673 |\n",
       "| chr1 |  10001055 |  10001570 | CASZ1       | Jurkat |  -0.37128583 | 1.241065e+00 | -0.29916714 | 0.76481252 |\n",
       "| chr1 |  10001055 |  10001570 | CLSTN1      | Jurkat |   0.45122592 | 3.008425e-01 |  1.49987419 | 0.13364699 |\n",
       "| chr1 |  10001055 |  10001570 | CTNNBIP1    | Jurkat |  -1.06901479 | 5.680602e-01 | -1.88186895 | 0.05985381 |\n",
       "| chr1 |  10001055 |  10001570 | DFFA        | Jurkat |  -0.69886070 | 4.939073e-01 | -1.41496340 | 0.15707923 |\n",
       "| chr1 |  10001055 |  10001570 | GPR157      | Jurkat |   0.72019151 | 6.347292e-01 |  1.13464375 | 0.25652462 |\n",
       "| chr1 |  10001055 |  10001570 | H6PD        | Jurkat |  -0.32546245 | 8.083685e-01 | -0.40261643 | 0.68723042 |\n",
       "| chr1 |  10001055 |  10001570 | KIF1B       | Jurkat |   0.17339554 | 2.951582e-01 |  0.58746648 | 0.55689045 |\n",
       "| chr1 |  10001055 |  10001570 | LINC01759   | Jurkat | -16.73549270 | 1.180793e+03 | -0.01417310 | 0.98869188 |\n",
       "| chr1 |  10001055 |  10001570 | LZIC        | Jurkat |  -0.83144376 | 4.999534e-01 | -1.66304241 | 0.09630395 |\n",
       "| chr1 |  10001055 |  10001570 | MIR34AHG    | Jurkat | -15.97593779 | 1.025017e+03 | -0.01558602 | 0.98756466 |\n",
       "| chr1 |  10001055 |  10001570 | NMNAT1      | Jurkat |  -0.48361336 | 1.107481e+00 | -0.43667865 | 0.66234441 |\n",
       "| chr1 |  10001055 |  10001570 | PEX14       | Jurkat |   0.25627843 | 2.970535e-01 |  0.86273481 | 0.38828329 |\n",
       "| chr1 |  10001055 |  10001570 | PGD         | Jurkat |  -0.05181250 | 2.936617e-01 | -0.17643600 | 0.85995143 |\n",
       "| chr1 |  10001055 |  10001570 | PIK3CD      | Jurkat |   0.49548326 | 4.512738e-01 |  1.09796588 | 0.27221939 |\n",
       "| chr1 |  10001055 |  10001570 | SLC25A33    | Jurkat |  -0.02200935 | 3.805832e-01 | -0.05783059 | 0.95388357 |\n",
       "| chr1 |  10001055 |  10001570 | SPSB1       | Jurkat |  -0.02086543 | 1.166714e+00 | -0.01788392 | 0.98573145 |\n",
       "| chr1 |  10001055 |  10001570 | TMEM201     | Jurkat |   0.55538350 | 5.921901e-01 |  0.93784658 | 0.34832326 |\n",
       "| chr1 |  10001055 |  10001570 | UBE4B       | Jurkat |   0.30821438 | 2.863096e-01 |  1.07650747 | 0.28170036 |\n",
       "| chr1 | 100037799 | 100039086 | AGL         | Jurkat |   0.00928901 | 5.785458e-02 |  0.16055789 | 0.87244162 |\n",
       "| chr1 | 100037799 | 100039086 | CDC14A      | Jurkat |   0.03767467 | 5.340894e-02 |  0.70540004 | 0.48056132 |\n",
       "| chr1 | 100037799 | 100039086 | DBT         | Jurkat |   0.03004789 | 7.126554e-02 |  0.42163275 | 0.67329310 |\n",
       "| chr1 | 100037799 | 100039086 | DPH5        | Jurkat |   0.10085758 | 6.752651e-02 |  1.49359975 | 0.13528026 |\n",
       "| chr1 | 100037799 | 100039086 | EXTL2       | Jurkat |   0.12930622 | 6.765497e-02 |  1.91125955 | 0.05597123 |\n",
       "| chr1 | 100037799 | 100039086 | FRRS1       | Jurkat |   0.10765048 | 1.030189e-01 |  1.04495879 | 0.29604202 |\n",
       "| chr1 | 100037799 | 100039086 | MFSD14A     | Jurkat | -11.09340350 | 2.071577e+02 | -0.05355052 | 0.95729328 |\n",
       "| chr1 | 100037799 | 100039086 | RTCA        | Jurkat |   0.04336320 | 6.673008e-02 |  0.64982990 | 0.51580210 |\n",
       "| chr1 | 100037799 | 100039086 | RTCA-AS1    | Jurkat |   0.09365090 | 1.404237e-01 |  0.66691653 | 0.50482545 |\n",
       "| chr1 | 100037799 | 100039086 | SASS6       | Jurkat |   0.02674776 | 5.497944e-02 |  0.48650478 | 0.62660931 |\n",
       "| chr1 | 100037799 | 100039086 | SLC30A7     | Jurkat |   0.02562122 | 4.747476e-02 |  0.53968085 | 0.58941715 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |   978694 |   979118 | IL3RA     | Jurkat | -28.60413936 | 2.064815e+03 | -0.01385312 | 0.98894716 |\n",
       "| chrX |   978694 |   979118 | LINC00106 | Jurkat |   0.19482473 | 1.314032e+00 |  0.14826488 | 0.88213373 |\n",
       "| chrX |   978694 |   979118 | P2RY8     | Jurkat |   0.33777979 | 3.447763e-01 |  0.97970711 | 0.32723072 |\n",
       "| chrX |   978694 |   979118 | PLCXD1    | Jurkat |   0.08742649 | 6.724470e-01 |  0.13001247 | 0.89655656 |\n",
       "| chrX |   978694 |   979118 | PPP2R3B   | Jurkat |  -0.88269523 | 9.916087e-01 | -0.89016486 | 0.37337737 |\n",
       "| chrX |   978694 |   979118 | SHOX      | Jurkat | -19.23873281 | 1.636486e+03 | -0.01175613 | 0.99062018 |\n",
       "| chrX |   978694 |   979118 | SLC25A6   | Jurkat |  -0.28160142 | 3.379915e-01 | -0.83316135 | 0.40475374 |\n",
       "| chrX |  9842372 |  9843073 | TBL1X     | Jurkat |  -0.12525948 | 2.575620e-01 | -0.48632741 | 0.62673505 |\n",
       "| chrX |  9842372 |  9843073 | WWC3      | Jurkat | -47.29171095 | 3.515520e+03 | -0.01345227 | 0.98926697 |\n",
       "| chrX |  9873846 |  9874297 | TBL1X     | Jurkat |  -0.51419511 | 3.559851e-01 | -1.44442889 | 0.14861837 |\n",
       "| chrX |  9873846 |  9874297 | WWC3      | Jurkat |  -0.06005241 | 7.478092e-01 | -0.08030445 | 0.93599512 |\n",
       "| chrX |  9885186 |  9885487 | TBL1X     | Jurkat |   0.28033199 | 2.449863e-01 |  1.14427622 | 0.25250910 |\n",
       "| chrX |  9885186 |  9885487 | WWC3      | Jurkat |  -0.22017978 | 6.495738e-01 | -0.33896039 | 0.73463957 |\n",
       "| chrX |   989869 |   990628 | AKAP17A   | Jurkat |   0.03569388 | 1.310651e-01 |  0.27233710 | 0.78536283 |\n",
       "| chrX |   989869 |   990628 | ASMT      | Jurkat |   1.61163806 | 7.926542e-01 |  2.03321721 | 0.04203058 |\n",
       "| chrX |   989869 |   990628 | ASMTL     | Jurkat |  -0.02572605 | 1.645188e-01 | -0.15637153 | 0.87574017 |\n",
       "| chrX |   989869 |   990628 | GTPBP6    | Jurkat |   0.17701549 | 1.416908e-01 |  1.24930801 | 0.21155244 |\n",
       "| chrX |   989869 |   990628 | IL3RA     | Jurkat |   0.31105639 | 2.533586e-01 |  1.22773186 | 0.21954764 |\n",
       "| chrX |   989869 |   990628 | LINC00106 | Jurkat |   0.18187483 | 2.871707e-01 |  0.63333351 | 0.52651587 |\n",
       "| chrX |   989869 |   990628 | P2RY8     | Jurkat |  -0.06355849 | 8.023537e-02 | -0.79215057 | 0.42827288 |\n",
       "| chrX |   989869 |   990628 | PLCXD1    | Jurkat |   0.24683237 | 1.378217e-01 |  1.79095469 | 0.07330057 |\n",
       "| chrX |   989869 |   990628 | PPP2R3B   | Jurkat |   0.05430420 | 1.440059e-01 |  0.37709704 | 0.70610149 |\n",
       "| chrX |   989869 |   990628 | SHOX      | Jurkat |  -0.14242059 | 1.098825e+00 | -0.12961176 | 0.89687360 |\n",
       "| chrX |   989869 |   990628 | SLC25A6   | Jurkat |  -0.05655356 | 6.920936e-02 | -0.81713744 | 0.41384989 |\n",
       "| chrX |  9941883 |  9942743 | TBL1X     | Jurkat |   0.09851419 | 1.542987e-01 |  0.63846423 | 0.52317153 |\n",
       "| chrX |  9941883 |  9942743 | WWC3      | Jurkat |   0.27461778 | 3.206169e-01 |  0.85652925 | 0.39170510 |\n",
       "| chrX | 99603024 | 99603776 | PCDH19    | Jurkat | -36.19759507 | 2.881754e+03 | -0.01256096 | 0.98997807 |\n",
       "| chrX | 99632013 | 99632219 | PCDH19    | Jurkat | -32.30320638 | 2.867434e+03 | -0.01126554 | 0.99101159 |\n",
       "| chrX |  9995829 |  9996207 | TBL1X     | Jurkat |  -0.13788683 | 2.551635e-01 | -0.54038610 | 0.58893079 |\n",
       "| chrX |  9995829 |  9996207 | WWC3      | Jurkat |   0.43916916 | 4.715682e-01 |  0.93129520 | 0.35170089 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene  CellType beta         sd          \n",
       "1       chr1  10001055  10001570 APITD1-CORT Jurkat   -23.33557068 1.404803e+03\n",
       "2       chr1  10001055  10001570 CASZ1       Jurkat    -0.37128583 1.241065e+00\n",
       "3       chr1  10001055  10001570 CLSTN1      Jurkat     0.45122592 3.008425e-01\n",
       "4       chr1  10001055  10001570 CTNNBIP1    Jurkat    -1.06901479 5.680602e-01\n",
       "5       chr1  10001055  10001570 DFFA        Jurkat    -0.69886070 4.939073e-01\n",
       "6       chr1  10001055  10001570 GPR157      Jurkat     0.72019151 6.347292e-01\n",
       "7       chr1  10001055  10001570 H6PD        Jurkat    -0.32546245 8.083685e-01\n",
       "8       chr1  10001055  10001570 KIF1B       Jurkat     0.17339554 2.951582e-01\n",
       "9       chr1  10001055  10001570 LINC01759   Jurkat   -16.73549270 1.180793e+03\n",
       "10      chr1  10001055  10001570 LZIC        Jurkat    -0.83144376 4.999534e-01\n",
       "11      chr1  10001055  10001570 MIR34AHG    Jurkat   -15.97593779 1.025017e+03\n",
       "12      chr1  10001055  10001570 NMNAT1      Jurkat    -0.48361336 1.107481e+00\n",
       "13      chr1  10001055  10001570 PEX14       Jurkat     0.25627843 2.970535e-01\n",
       "14      chr1  10001055  10001570 PGD         Jurkat    -0.05181250 2.936617e-01\n",
       "15      chr1  10001055  10001570 PIK3CD      Jurkat     0.49548326 4.512738e-01\n",
       "16      chr1  10001055  10001570 SLC25A33    Jurkat    -0.02200935 3.805832e-01\n",
       "17      chr1  10001055  10001570 SPSB1       Jurkat    -0.02086543 1.166714e+00\n",
       "18      chr1  10001055  10001570 TMEM201     Jurkat     0.55538350 5.921901e-01\n",
       "19      chr1  10001055  10001570 UBE4B       Jurkat     0.30821438 2.863096e-01\n",
       "20      chr1 100037799 100039086 AGL         Jurkat     0.00928901 5.785458e-02\n",
       "21      chr1 100037799 100039086 CDC14A      Jurkat     0.03767467 5.340894e-02\n",
       "22      chr1 100037799 100039086 DBT         Jurkat     0.03004789 7.126554e-02\n",
       "23      chr1 100037799 100039086 DPH5        Jurkat     0.10085758 6.752651e-02\n",
       "24      chr1 100037799 100039086 EXTL2       Jurkat     0.12930622 6.765497e-02\n",
       "25      chr1 100037799 100039086 FRRS1       Jurkat     0.10765048 1.030189e-01\n",
       "26      chr1 100037799 100039086 MFSD14A     Jurkat   -11.09340350 2.071577e+02\n",
       "27      chr1 100037799 100039086 RTCA        Jurkat     0.04336320 6.673008e-02\n",
       "28      chr1 100037799 100039086 RTCA-AS1    Jurkat     0.09365090 1.404237e-01\n",
       "29      chr1 100037799 100039086 SASS6       Jurkat     0.02674776 5.497944e-02\n",
       "30      chr1 100037799 100039086 SLC30A7     Jurkat     0.02562122 4.747476e-02\n",
       "⋮       ⋮    ⋮         ⋮         ⋮           ⋮        ⋮            ⋮           \n",
       "2590380 chrX   978694    979118  IL3RA       Jurkat   -28.60413936 2.064815e+03\n",
       "2590381 chrX   978694    979118  LINC00106   Jurkat     0.19482473 1.314032e+00\n",
       "2590382 chrX   978694    979118  P2RY8       Jurkat     0.33777979 3.447763e-01\n",
       "2590383 chrX   978694    979118  PLCXD1      Jurkat     0.08742649 6.724470e-01\n",
       "2590384 chrX   978694    979118  PPP2R3B     Jurkat    -0.88269523 9.916087e-01\n",
       "2590385 chrX   978694    979118  SHOX        Jurkat   -19.23873281 1.636486e+03\n",
       "2590386 chrX   978694    979118  SLC25A6     Jurkat    -0.28160142 3.379915e-01\n",
       "2590387 chrX  9842372   9843073  TBL1X       Jurkat    -0.12525948 2.575620e-01\n",
       "2590388 chrX  9842372   9843073  WWC3        Jurkat   -47.29171095 3.515520e+03\n",
       "2590389 chrX  9873846   9874297  TBL1X       Jurkat    -0.51419511 3.559851e-01\n",
       "2590390 chrX  9873846   9874297  WWC3        Jurkat    -0.06005241 7.478092e-01\n",
       "2590391 chrX  9885186   9885487  TBL1X       Jurkat     0.28033199 2.449863e-01\n",
       "2590392 chrX  9885186   9885487  WWC3        Jurkat    -0.22017978 6.495738e-01\n",
       "2590393 chrX   989869    990628  AKAP17A     Jurkat     0.03569388 1.310651e-01\n",
       "2590394 chrX   989869    990628  ASMT        Jurkat     1.61163806 7.926542e-01\n",
       "2590395 chrX   989869    990628  ASMTL       Jurkat    -0.02572605 1.645188e-01\n",
       "2590396 chrX   989869    990628  GTPBP6      Jurkat     0.17701549 1.416908e-01\n",
       "2590397 chrX   989869    990628  IL3RA       Jurkat     0.31105639 2.533586e-01\n",
       "2590398 chrX   989869    990628  LINC00106   Jurkat     0.18187483 2.871707e-01\n",
       "2590399 chrX   989869    990628  P2RY8       Jurkat    -0.06355849 8.023537e-02\n",
       "2590400 chrX   989869    990628  PLCXD1      Jurkat     0.24683237 1.378217e-01\n",
       "2590401 chrX   989869    990628  PPP2R3B     Jurkat     0.05430420 1.440059e-01\n",
       "2590402 chrX   989869    990628  SHOX        Jurkat    -0.14242059 1.098825e+00\n",
       "2590403 chrX   989869    990628  SLC25A6     Jurkat    -0.05655356 6.920936e-02\n",
       "2590404 chrX  9941883   9942743  TBL1X       Jurkat     0.09851419 1.542987e-01\n",
       "2590405 chrX  9941883   9942743  WWC3        Jurkat     0.27461778 3.206169e-01\n",
       "2590406 chrX 99603024  99603776  PCDH19      Jurkat   -36.19759507 2.881754e+03\n",
       "2590407 chrX 99632013  99632219  PCDH19      Jurkat   -32.30320638 2.867434e+03\n",
       "2590408 chrX  9995829   9996207  TBL1X       Jurkat    -0.13788683 2.551635e-01\n",
       "2590409 chrX  9995829   9996207  WWC3        Jurkat     0.43916916 4.715682e-01\n",
       "        z           p         \n",
       "1       -0.01661128 0.98674673\n",
       "2       -0.29916714 0.76481252\n",
       "3        1.49987419 0.13364699\n",
       "4       -1.88186895 0.05985381\n",
       "5       -1.41496340 0.15707923\n",
       "6        1.13464375 0.25652462\n",
       "7       -0.40261643 0.68723042\n",
       "8        0.58746648 0.55689045\n",
       "9       -0.01417310 0.98869188\n",
       "10      -1.66304241 0.09630395\n",
       "11      -0.01558602 0.98756466\n",
       "12      -0.43667865 0.66234441\n",
       "13       0.86273481 0.38828329\n",
       "14      -0.17643600 0.85995143\n",
       "15       1.09796588 0.27221939\n",
       "16      -0.05783059 0.95388357\n",
       "17      -0.01788392 0.98573145\n",
       "18       0.93784658 0.34832326\n",
       "19       1.07650747 0.28170036\n",
       "20       0.16055789 0.87244162\n",
       "21       0.70540004 0.48056132\n",
       "22       0.42163275 0.67329310\n",
       "23       1.49359975 0.13528026\n",
       "24       1.91125955 0.05597123\n",
       "25       1.04495879 0.29604202\n",
       "26      -0.05355052 0.95729328\n",
       "27       0.64982990 0.51580210\n",
       "28       0.66691653 0.50482545\n",
       "29       0.48650478 0.62660931\n",
       "30       0.53968085 0.58941715\n",
       "⋮       ⋮           ⋮         \n",
       "2590380 -0.01385312 0.98894716\n",
       "2590381  0.14826488 0.88213373\n",
       "2590382  0.97970711 0.32723072\n",
       "2590383  0.13001247 0.89655656\n",
       "2590384 -0.89016486 0.37337737\n",
       "2590385 -0.01175613 0.99062018\n",
       "2590386 -0.83316135 0.40475374\n",
       "2590387 -0.48632741 0.62673505\n",
       "2590388 -0.01345227 0.98926697\n",
       "2590389 -1.44442889 0.14861837\n",
       "2590390 -0.08030445 0.93599512\n",
       "2590391  1.14427622 0.25250910\n",
       "2590392 -0.33896039 0.73463957\n",
       "2590393  0.27233710 0.78536283\n",
       "2590394  2.03321721 0.04203058\n",
       "2590395 -0.15637153 0.87574017\n",
       "2590396  1.24930801 0.21155244\n",
       "2590397  1.22773186 0.21954764\n",
       "2590398  0.63333351 0.52651587\n",
       "2590399 -0.79215057 0.42827288\n",
       "2590400  1.79095469 0.07330057\n",
       "2590401  0.37709704 0.70610149\n",
       "2590402 -0.12961176 0.89687360\n",
       "2590403 -0.81713744 0.41384989\n",
       "2590404  0.63846423 0.52317153\n",
       "2590405  0.85652925 0.39170510\n",
       "2590406 -0.01256096 0.98997807\n",
       "2590407 -0.01126554 0.99101159\n",
       "2590408 -0.54038610 0.58893079\n",
       "2590409  0.93129520 0.35170089"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = celltype\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"beta\",\n",
    "                         \"sd\",\n",
    "                         \"z\",\n",
    "                         \"p\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] data.table_1.15.2    dplyr_1.1.4          foreach_1.5.2       \n",
       " [4] GenomicRanges_1.54.1 GenomeInfoDb_1.38.1  IRanges_2.36.0      \n",
       " [7] S4Vectors_0.40.2     BiocGenerics_0.48.1  genomation_1.34.0   \n",
       "[10] Signac_1.13.0        Seurat_5.0.3         SeuratObject_5.0.1  \n",
       "[13] sp_2.1-4             SnapATAC_1.0.0       rhdf5_2.44.0        \n",
       "[16] Matrix_1.6-5        \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 BiocIO_1.10.0              \n",
       "  [5] pbdZMQ_0.3-11               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                polyclip_1.10-6            \n",
       "  [9] XML_3.99-0.16.1             fastDummies_1.7.3          \n",
       " [11] lifecycle_1.0.4             tcltk_4.3.3                \n",
       " [13] edgeR_3.42.4                doParallel_1.0.17          \n",
       " [15] vroom_1.6.5                 globals_0.16.3             \n",
       " [17] lattice_0.22-6              MASS_7.3-60                \n",
       " [19] magrittr_2.0.3              limma_3.56.2               \n",
       " [21] plotly_4.10.4               plotrix_3.8-4              \n",
       " [23] yaml_2.3.8                  plot3D_1.4.1               \n",
       " [25] httpuv_1.6.15               sctransform_0.4.1          \n",
       " [27] spam_2.10-0                 spatstat.sparse_3.0-3      \n",
       " [29] reticulate_1.36.1           cowplot_1.1.3              \n",
       " [31] pbapply_1.7-2               RColorBrewer_1.1-3         \n",
       " [33] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [35] Rtsne_0.17                  purrr_1.0.2                \n",
       " [37] RCurl_1.98-1.14             misc3d_0.9-1               \n",
       " [39] GenomeInfoDbData_1.2.11     ggrepel_0.9.5              \n",
       " [41] irlba_2.3.5.1               listenv_0.9.1              \n",
       " [43] spatstat.utils_3.0-4        goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             bigmemory_4.6.4            \n",
       " [47] spatstat.random_3.2-3       fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           DelayedArray_0.26.7        \n",
       " [51] leiden_0.4.3.1              codetools_0.2-20           \n",
       " [53] RcppRoll_0.3.0              tidyselect_1.2.1           \n",
       " [55] viridis_0.6.5               matrixStats_1.3.0          \n",
       " [57] base64enc_0.1-3             spatstat.explore_3.2-6     \n",
       " [59] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [61] progressr_0.14.0            ggridges_0.5.6             \n",
       " [63] survival_3.6-4              iterators_1.0.14           \n",
       " [65] tools_4.3.3                 snow_0.4-4                 \n",
       " [67] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [69] glue_1.7.0                  gridExtra_2.3              \n",
       " [71] MatrixGenerics_1.12.3       IRdisplay_1.1              \n",
       " [73] withr_3.0.0                 fastmap_1.1.1              \n",
       " [75] rhdf5filters_1.12.1         fansi_1.0.6                \n",
       " [77] digest_0.6.35               R6_2.5.1                   \n",
       " [79] mime_0.12                   seqPattern_1.34.0          \n",
       " [81] colorspace_2.1-0            scattermore_1.2            \n",
       " [83] tensor_1.5                  spatstat.data_3.0-4        \n",
       " [85] utf8_1.2.4                  tidyr_1.3.1                \n",
       " [87] generics_0.1.3              rtracklayer_1.60.1         \n",
       " [89] S4Arrays_1.2.1              httr_1.4.7                 \n",
       " [91] htmlwidgets_1.6.4           uwot_0.1.16                \n",
       " [93] pkgconfig_2.0.3             gtable_0.3.5               \n",
       " [95] impute_1.76.0               lmtest_0.9-40              \n",
       " [97] XVector_0.42.0              htmltools_0.5.8.1          \n",
       " [99] dotCall64_1.1-1             Biobase_2.60.0             \n",
       "[101] scales_1.3.0                png_0.1-8                  \n",
       "[103] doSNOW_1.0.20               bigmemory.sri_0.1.8        \n",
       "[105] tzdb_0.4.0                  rjson_0.2.21               \n",
       "[107] reshape2_1.4.4              uuid_1.2-0                 \n",
       "[109] nlme_3.1-164                repr_1.1.7                 \n",
       "[111] zoo_1.8-12                  stringr_1.5.1              \n",
       "[113] KernSmooth_2.23-22          miniUI_0.1.1.1             \n",
       "[115] restfulr_0.0.15             pillar_1.9.0               \n",
       "[117] vctrs_0.6.5                 RANN_2.6.1                 \n",
       "[119] promises_1.3.0              xtable_1.8-4               \n",
       "[121] cluster_2.1.6               evaluate_0.23              \n",
       "[123] readr_2.1.5                 cli_3.6.2                  \n",
       "[125] locfit_1.5-9.9              compiler_4.3.3             \n",
       "[127] Rsamtools_2.18.0            rlang_1.1.3                \n",
       "[129] crayon_1.5.2                future.apply_1.11.2        \n",
       "[131] plyr_1.8.9                  stringi_1.8.3              \n",
       "[133] gridBase_0.4-7              viridisLite_0.4.2          \n",
       "[135] deldir_2.0-4                BiocParallel_1.36.0        \n",
       "[137] munsell_0.5.1               Biostrings_2.70.1          \n",
       "[139] lazyeval_0.2.2              spatstat.geom_3.2-9        \n",
       "[141] IRkernel_1.3.2              BSgenome_1.68.0            \n",
       "[143] RcppHNSW_0.6.0              hms_1.1.3                  \n",
       "[145] patchwork_1.2.0             bit64_4.0.5                \n",
       "[147] future_1.33.2               ggplot2_3.5.1              \n",
       "[149] Rhdf5lib_1.22.1             shiny_1.8.1.1              \n",
       "[151] SummarizedExperiment_1.30.2 ROCR_1.0-11                \n",
       "[153] igraph_2.0.3                fastmatch_1.1-4            \n",
       "[155] bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
