{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8466b35-cae5-4f62-8138-37b2de484a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.2\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : grid v4.3.3\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "\tLoading Package : gtools v3.9.5\n",
      "\n",
      "\tLoading Package : gtable v0.3.5\n",
      "\n",
      "\tLoading Package : ggplot2 v3.5.1\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "\tLoading Package : plyr v1.8.9\n",
      "\n",
      "\tLoading Package : stringr v1.5.1\n",
      "\n",
      "\tLoading Package : data.table v1.15.2\n",
      "\n",
      "\tLoading Package : matrixStats v1.3.0\n",
      "\n",
      "\tLoading Package : S4Vectors v0.40.2\n",
      "\n",
      "\tLoading Package : GenomicRanges v1.54.1\n",
      "\n",
      "\tLoading Package : BiocGenerics v0.48.1\n",
      "\n",
      "\tLoading Package : Matrix v1.6.5\n",
      "\n",
      "\tLoading Package : Rcpp v1.0.12\n",
      "\n",
      "\tLoading Package : SummarizedExperiment v1.30.2\n",
      "\n",
      "\tLoading Package : rhdf5 v2.44.0\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(magrittr)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e85020d7-d358-492d-9358-c81270f4b624",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1bec6690-52ec-450d-bba7-7d4528a185cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.seurat = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/1.prepare_data/1.seurat_pipeline.240507/obj.seurat.qc.rds\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Xu/K562/Kendall/Pairs.tsv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/3.Genome_wide_prediction/ArchR_250226/ArchR_250226/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "60b75d89-aa5d-485e-b1a3-ec1988eb7a3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "inputFiles = c(\"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_230720/ArrayExpress/cellranger_res/EMTAB11264_01/outs/atac_fragments.tsv.gz\")\n",
    "names(inputFiles) = c(\"K562_Xu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "340dac32-44ac-4231-b964-b5f3b9e2dbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16\n",
    "maxDist = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "da973385-4c81-4b1c-9669-cf5933b23a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 16.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRThreads(threads = n.cores) \n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeb26496-d3b9-4778-92ed-9c63a5018aef",
   "metadata": {},
   "source": [
    "Import seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c9b19a09-be95-4d9a-bdc3-0530980d4757",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.seurat = readRDS(path.seurat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86be3e66-c954-4620-9e64-c6a1c0fcb84c",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6f363969-bf0b-45ce-858c-a8cb64cc210a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3538ffdc-4fed-48eb-9c9d-86805a91cf06",
   "metadata": {},
   "source": [
    "Create an ArchRProject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0300f485-fafa-4cef-b9d0-534108ec6f95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 1\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_230720/ArrayExpress/cellranger\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"c8366071805c37835af9f47f43ae7435\" \"af5268737bf82cc3e7978e6407c893ea\"\n",
      "  .. ..@ cells: Named chr [1:7821] \"AAACAGCCAAGCGATG-1\" \"AAACAGCCACATGCTA-1\" \"AAACAGCCACGGTACT-1\" \"AAACAGCCATAAAGCA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:7821] \"K562_Xu_AAACAGCCAAGCGATG-1\" \"K562_Xu_AAACAGCCACATGCTA-1\" \"K562_Xu_AAACAGCCACGGTACT-1\" \"K562_Xu_AAACAGCCATAAAGCA-1\" ...\n"
     ]
    }
   ],
   "source": [
    "str(Fragments(obj.seurat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1753ffaa-89e2-43bf-bcf6-19937782d63a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-createArrows-29e71b59ad4ca4-Date-2025-02-26_Time-13-32-01.140829.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Cleaning Temporary Files\n",
      "\n",
      "2025-02-26 13:32:01.569983 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "(K562_Xu : 1 of 1) Determining Arrow Method to use!\n",
      "\n",
      "2025-02-26 13:32:01.633727 : (K562_Xu : 1 of 1) Reading In Fragments from inputFiles (readMethod = tabix), 0.001 mins elapsed.\n",
      "\n",
      "2025-02-26 13:32:01.644588 : (K562_Xu : 1 of 1) Tabix Bed To Temporary File, 0.001 mins elapsed.\n",
      "\n",
      "2025-02-26 13:33:57.453146 : (K562_Xu : 1 of 1) Successful creation of Temporary File, 1.931 mins elapsed.\n",
      "\n",
      "2025-02-26 13:33:57.46188 : (K562_Xu : 1 of 1) Creating ArrowFile From Temporary File, 1.932 mins elapsed.\n",
      "\n",
      "2025-02-26 13:34:35.317216 : (K562_Xu : 1 of 1) Successful creation of Arrow File, 2.562 mins elapsed.\n",
      "\n",
      "2025-02-26 13:35:21.15107 : Continuing through after error ggplot for Fragment Size Distribution, 3.326 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-02-26 13:35:53.825384 : (K562_Xu : 1 of 1) CellStats : Number of Cells Pass Filter = 7699 , 3.871 mins elapsed.\n",
      "\n",
      "2025-02-26 13:35:53.829247 : (K562_Xu : 1 of 1) CellStats : Median Frags = 31472 , 3.871 mins elapsed.\n",
      "\n",
      "2025-02-26 13:35:53.834211 : (K562_Xu : 1 of 1) CellStats : Median TSS Enrichment = 9.17 , 3.871 mins elapsed.\n",
      "\n",
      "2025-02-26 13:35:54.755789 : Continuing through after error ggplot for TSS by Frags, 3.886 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-02-26 13:35:57.336609 : (K562_Xu : 1 of 1) Adding Additional Feature Counts!, 3.929 mins elapsed.\n",
      "\n",
      "2025-02-26 13:36:24.112168 : (K562_Xu : 1 of 1) Removing Fragments from Filtered Cells, 4.376 mins elapsed.\n",
      "\n",
      "2025-02-26 13:36:24.121519 : (K562_Xu : 1 of 1) Creating Filtered Arrow File, 4.376 mins elapsed.\n",
      "\n",
      "2025-02-26 13:38:22.012406 : (K562_Xu : 1 of 1) Finished Constructing Filtered Arrow File!, 6.341 mins elapsed.\n",
      "\n",
      "2025-02-26 13:38:22.206577 : (K562_Xu : 1 of 1) Adding TileMatrix!, 6.344 mins elapsed.\n",
      "\n",
      "2025-02-26 13:47:58.55422 : (K562_Xu : 1 of 1) Finished Creating Arrow File, 15.95 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-createArrows-29e71b59ad4ca4-Date-2025-02-26_Time-13-32-01.140829.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ArrowFiles <- createArrowFiles(\n",
    "  inputFiles = c(Fragments(obj.seurat)[[1]]@path),\n",
    "  sampleNames = c(\"K562_Xu\"),\n",
    "  validBarcodes = list(K562_Xu = unname(Fragments(obj.seurat)[[1]]@cells)),\n",
    "  addTileMat = T,\n",
    "  addGeneScoreMat = F,\n",
    "  force = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ada3db45-d050-4659-a7cf-2891e04b1249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'K562_Xu.arrow'"
      ],
      "text/latex": [
       "'K562\\_Xu.arrow'"
      ],
      "text/markdown": [
       "'K562_Xu.arrow'"
      ],
      "text/plain": [
       "[1] \"K562_Xu.arrow\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ArrowFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5ba28283-d488-4982-8d4c-745fc3f59365",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"ArchR\",sep=\"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "aa217507-8323-43bf-b23c-0ec25ee33c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Validating Arrows...\n",
      "\n",
      "Getting SampleNames...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Copying ArrowFiles to Ouptut Directory! If you want to save disk space set copyArrows = FALSE\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Getting Cell Metadata...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Merging Cell Metadata...\n",
      "\n",
      "Initializing ArchRProject...\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR <- ArchRProject(\n",
    "  ArrowFiles = ArrowFiles, \n",
    "  outputDirectory = paste(dir.output,\"ArchR\",sep=\"/\"),\n",
    "  copyArrows = TRUE \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f9795403-545f-4385-aa38-2c70b1b2fb21",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: ArchRProject \n",
       "outputDirectory: /maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/3.Genome_wide_prediction/ArchR_250226/ArchR_250226/ArchR \n",
       "samples(1): K562_Xu\n",
       "sampleColData names(1): ArrowFiles\n",
       "cellColData names(13): Sample TSSEnrichment ... nDiFrags BlacklistRatio\n",
       "numberOfCells(1): 7699\n",
       "medianTSS(1): 9.17\n",
       "medianFrags(1): 31472"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obj.ArchR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ee876dd5-0b22-46fd-9015-7e22144c5024",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking Inputs...\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addIterativeLSI-29e71b60652919-Date-2025-02-26_Time-13-48-12.167445.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 13:48:12.6913 : Computing Total Across All Features, 0 mins elapsed.\n",
      "\n",
      "2025-02-26 13:48:13.629304 : Computing Top Features, 0.016 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-26 13:48:16.674102 : Running LSI (1 of 2) on Top Features, 0.067 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-02-26 13:48:16.688989 : Creating Partial Matrix, 0.067 mins elapsed.\n",
      "\n",
      "2025-02-26 13:48:57.385737 : Computing LSI, 0.745 mins elapsed.\n",
      "\n",
      "2025-02-26 13:49:58.707984 : Identifying Clusters, 1.767 mins elapsed.\n",
      "\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "2025-02-26 13:50:06.67247 : Identified 6 Clusters, 1.9 mins elapsed.\n",
      "\n",
      "2025-02-26 13:50:06.686766 : Saving LSI Iteration, 1.9 mins elapsed.\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************************************************************\n",
      "2025-02-26 13:50:21.717164 : ERROR Found in .saveIteration for  \n",
      "LogFile = ArchRLogs/ArchR-addIterativeLSI-29e71b60652919-Date-2025-02-26_Time-13-48-12.167445.log\n",
      "\n",
      "<simpleError in g$grobs[[legend]]: no such index at level 2\n",
      ">\n",
      "\n",
      "************************************************************\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-26 13:50:21.720224 : Creating Cluster Matrix on the total Group Features, 2.151 mins elapsed.\n",
      "\n",
      "2025-02-26 13:50:44.123919 : Computing Variable Features, 2.524 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-26 13:50:44.332517 : Running LSI (2 of 2) on Variable Features, 2.528 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-02-26 13:50:44.349766 : Creating Partial Matrix, 2.528 mins elapsed.\n",
      "\n",
      "2025-02-26 13:51:15.491405 : Computing LSI, 3.047 mins elapsed.\n",
      "\n",
      "2025-02-26 13:51:59.072084 : Finished Running IterativeLSI, 3.773 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addIterativeLSI(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    useMatrix = \"TileMatrix\", \n",
    "    name = \"IterativeLSI\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2c1d334b-7fd7-425a-bb8d-7be4cbedd3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.ArchR = addPeakSet(obj.ArchR,\n",
    "                       peakSet = reduce(pairs.E2G))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e25c14db-eb4e-4b94-8862-eb34246a365c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeakMatrix-29e71b21129653-Date-2025-02-26_Time-13-52-02.21826.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 13:52:02.512289 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-02-26 13:52:03.726298 : Adding K562_Xu to PeakMatrix for Chr (1 of 23)!, 0.019 mins elapsed.\n",
      "\n",
      "2025-02-26 13:52:24.774975 : Adding K562_Xu to PeakMatrix for Chr (2 of 23)!, 0.37 mins elapsed.\n",
      "\n",
      "2025-02-26 13:52:37.187572 : Adding K562_Xu to PeakMatrix for Chr (3 of 23)!, 0.577 mins elapsed.\n",
      "\n",
      "2025-02-26 13:52:49.231514 : Adding K562_Xu to PeakMatrix for Chr (4 of 23)!, 0.778 mins elapsed.\n",
      "\n",
      "2025-02-26 13:52:57.890532 : Adding K562_Xu to PeakMatrix for Chr (5 of 23)!, 0.922 mins elapsed.\n",
      "\n",
      "2025-02-26 13:53:07.718132 : Adding K562_Xu to PeakMatrix for Chr (6 of 23)!, 1.086 mins elapsed.\n",
      "\n",
      "2025-02-26 13:53:20.299965 : Adding K562_Xu to PeakMatrix for Chr (7 of 23)!, 1.296 mins elapsed.\n",
      "\n",
      "2025-02-26 13:53:32.594468 : Adding K562_Xu to PeakMatrix for Chr (8 of 23)!, 1.5 mins elapsed.\n",
      "\n",
      "2025-02-26 13:53:41.316232 : Adding K562_Xu to PeakMatrix for Chr (9 of 23)!, 1.646 mins elapsed.\n",
      "\n",
      "2025-02-26 13:53:50.618288 : Adding K562_Xu to PeakMatrix for Chr (10 of 23)!, 1.801 mins elapsed.\n",
      "\n",
      "2025-02-26 13:54:00.455773 : Adding K562_Xu to PeakMatrix for Chr (11 of 23)!, 1.965 mins elapsed.\n",
      "\n",
      "2025-02-26 13:54:11.800082 : Adding K562_Xu to PeakMatrix for Chr (12 of 23)!, 2.154 mins elapsed.\n",
      "\n",
      "2025-02-26 13:54:21.53926 : Adding K562_Xu to PeakMatrix for Chr (13 of 23)!, 2.316 mins elapsed.\n",
      "\n",
      "2025-02-26 13:54:27.930806 : Adding K562_Xu to PeakMatrix for Chr (14 of 23)!, 2.423 mins elapsed.\n",
      "\n",
      "2025-02-26 13:54:35.42848 : Adding K562_Xu to PeakMatrix for Chr (15 of 23)!, 2.548 mins elapsed.\n",
      "\n",
      "2025-02-26 13:54:44.424733 : Adding K562_Xu to PeakMatrix for Chr (16 of 23)!, 2.698 mins elapsed.\n",
      "\n",
      "2025-02-26 13:54:53.912621 : Adding K562_Xu to PeakMatrix for Chr (17 of 23)!, 2.856 mins elapsed.\n",
      "\n",
      "2025-02-26 13:55:04.387863 : Adding K562_Xu to PeakMatrix for Chr (18 of 23)!, 3.03 mins elapsed.\n",
      "\n",
      "2025-02-26 13:55:11.051249 : Adding K562_Xu to PeakMatrix for Chr (19 of 23)!, 3.141 mins elapsed.\n",
      "\n",
      "2025-02-26 13:55:23.279206 : Adding K562_Xu to PeakMatrix for Chr (20 of 23)!, 3.345 mins elapsed.\n",
      "\n",
      "2025-02-26 13:55:30.968552 : Adding K562_Xu to PeakMatrix for Chr (21 of 23)!, 3.473 mins elapsed.\n",
      "\n",
      "2025-02-26 13:55:37.316509 : Adding K562_Xu to PeakMatrix for Chr (22 of 23)!, 3.579 mins elapsed.\n",
      "\n",
      "2025-02-26 13:55:45.453896 : Adding K562_Xu to PeakMatrix for Chr (23 of 23)!, 3.715 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeakMatrix-29e71b21129653-Date-2025-02-26_Time-13-52-02.21826.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addPeakMatrix(obj.ArchR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3b913333-8823-435e-a677-42fa6d8777ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seRNA = import10xFeatureMatrix(c(\"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_230720/ArrayExpress/cellranger_res/EMTAB11264_01/outs/filtered_feature_bc_matrix.h5\"),\n",
    "                               c(\"K562_Xu\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "895234f3-e668-4e83-adf0-562024a6711b",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA = seRNA[,rownames(obj.ArchR@cellColData)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "491fc2c1-0ac8-4968-97fc-b82389488943",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "31d772b0-1dd0-4d88-89b1-48fb00919cf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "aa2f8f62-d3dd-4731-87ac-b83f12d7a42f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biostrings’:\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seRNA.rename = map_gene_names(seRNA,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "76a543f9-13a3-46a0-89d1-e99ffad20a06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 20532 7699 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(20532): FAM138A OR4F5 ... FAM231B PRODH\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(7699): K562_Xu#GTCATGCCAGGATGGC-1 K562_Xu#TATCGCGAGGTTAGAG-1\n",
       "  ... K562_Xu#CGTGTTACAGTCTATG-1 K562_Xu#AAGTGAAGTTGCCTCA-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA.rename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "65cb2097-9075-4ee4-9e54-dd288dc50736",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addGeneExpressionMatrix-29e71b30d20752-Date-2025-02-26_Time-14-00-11.767381.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Overlap w/ scATAC = 1\n",
      "\n",
      "2025-02-26 14:00:12.244878 : \n",
      "\n",
      "Overlap Per Sample w/ scATAC : K562_Xu=7699\n",
      "\n",
      "2025-02-26 14:00:12.25844 : \n",
      "\n",
      "2025-02-26 14:00:15.518987 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-02-26 14:00:16.437684 : Adding K562_Xu to GeneExpressionMatrix for Chr (1 of 23)!, 0.015 mins elapsed.\n",
      "\n",
      "2025-02-26 14:00:22.339061 : Adding K562_Xu to GeneExpressionMatrix for Chr (2 of 23)!, 0.114 mins elapsed.\n",
      "\n",
      "2025-02-26 14:00:28.147598 : Adding K562_Xu to GeneExpressionMatrix for Chr (3 of 23)!, 0.21 mins elapsed.\n",
      "\n",
      "2025-02-26 14:00:35.967326 : Adding K562_Xu to GeneExpressionMatrix for Chr (4 of 23)!, 0.341 mins elapsed.\n",
      "\n",
      "2025-02-26 14:00:41.263313 : Adding K562_Xu to GeneExpressionMatrix for Chr (5 of 23)!, 0.429 mins elapsed.\n",
      "\n",
      "2025-02-26 14:00:46.557264 : Adding K562_Xu to GeneExpressionMatrix for Chr (6 of 23)!, 0.517 mins elapsed.\n",
      "\n",
      "2025-02-26 14:00:54.39689 : Adding K562_Xu to GeneExpressionMatrix for Chr (7 of 23)!, 0.648 mins elapsed.\n",
      "\n",
      "2025-02-26 14:00:59.750889 : Adding K562_Xu to GeneExpressionMatrix for Chr (8 of 23)!, 0.737 mins elapsed.\n",
      "\n",
      "2025-02-26 14:01:04.91832 : Adding K562_Xu to GeneExpressionMatrix for Chr (9 of 23)!, 0.823 mins elapsed.\n",
      "\n",
      "2025-02-26 14:01:12.500128 : Adding K562_Xu to GeneExpressionMatrix for Chr (10 of 23)!, 0.95 mins elapsed.\n",
      "\n",
      "2025-02-26 14:01:18.025806 : Adding K562_Xu to GeneExpressionMatrix for Chr (11 of 23)!, 1.042 mins elapsed.\n",
      "\n",
      "2025-02-26 14:01:23.48922 : Adding K562_Xu to GeneExpressionMatrix for Chr (12 of 23)!, 1.133 mins elapsed.\n",
      "\n",
      "2025-02-26 14:01:31.237604 : Adding K562_Xu to GeneExpressionMatrix for Chr (13 of 23)!, 1.262 mins elapsed.\n",
      "\n",
      "2025-02-26 14:01:36.225421 : Adding K562_Xu to GeneExpressionMatrix for Chr (14 of 23)!, 1.345 mins elapsed.\n",
      "\n",
      "2025-02-26 14:01:40.996923 : Adding K562_Xu to GeneExpressionMatrix for Chr (15 of 23)!, 1.425 mins elapsed.\n",
      "\n",
      "2025-02-26 14:01:48.457208 : Adding K562_Xu to GeneExpressionMatrix for Chr (16 of 23)!, 1.549 mins elapsed.\n",
      "\n",
      "2025-02-26 14:01:53.711766 : Adding K562_Xu to GeneExpressionMatrix for Chr (17 of 23)!, 1.637 mins elapsed.\n",
      "\n",
      "2025-02-26 14:01:59.172243 : Adding K562_Xu to GeneExpressionMatrix for Chr (18 of 23)!, 1.728 mins elapsed.\n",
      "\n",
      "2025-02-26 14:02:06.551659 : Adding K562_Xu to GeneExpressionMatrix for Chr (19 of 23)!, 1.851 mins elapsed.\n",
      "\n",
      "2025-02-26 14:02:12.006849 : Adding K562_Xu to GeneExpressionMatrix for Chr (20 of 23)!, 1.941 mins elapsed.\n",
      "\n",
      "2025-02-26 14:02:17.436091 : Adding K562_Xu to GeneExpressionMatrix for Chr (21 of 23)!, 2.032 mins elapsed.\n",
      "\n",
      "2025-02-26 14:02:24.845948 : Adding K562_Xu to GeneExpressionMatrix for Chr (22 of 23)!, 2.155 mins elapsed.\n",
      "\n",
      "2025-02-26 14:02:30.094731 : Adding K562_Xu to GeneExpressionMatrix for Chr (23 of 23)!, 2.243 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addGeneExpressionMatrix-29e71b30d20752-Date-2025-02-26_Time-14-00-11.767381.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addGeneExpressionMatrix(obj.ArchR,\n",
    "                                    seRNA = seRNA.rename,\n",
    "                                    strictMatch = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4318703-a143-40c0-b159-274d05033bbc",
   "metadata": {},
   "source": [
    "Run ArchR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "47c8fd5f-19ec-46e6-af87-57502c1e1ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeak2GeneLinks-29e71b2fb6054e-Date-2025-02-26_Time-14-02-37.506703.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 14:02:37.906859 : Getting Available Matrices, 0.007 mins elapsed.\n",
      "\n",
      "No predictionScore found. Continuing without predictionScore!\n",
      "\n",
      "2025-02-26 14:02:38.248078 : Filtered Low Prediction Score Cells (0 of 7699, 0), 0 mins elapsed.\n",
      "\n",
      "Filtering 1 dims correlated > 0.75 to log10(depth + 1)\n",
      "\n",
      "2025-02-26 14:02:38.335571 : Computing KNN, 0.002 mins elapsed.\n",
      "\n",
      "2025-02-26 14:02:38.670088 : Identifying Non-Overlapping KNN pairs, 0.007 mins elapsed.\n",
      "\n",
      "2025-02-26 14:02:40.301665 : Identified 494 Groupings!, 0.035 mins elapsed.\n",
      "\n",
      "2025-02-26 14:02:40.349501 : Getting Group RNA Matrix, 0.035 mins elapsed.\n",
      "\n",
      "2025-02-26 14:02:58.447593 : Getting Group ATAC Matrix, 0.337 mins elapsed.\n",
      "\n",
      "2025-02-26 14:03:29.133093 : Normalizing Group Matrices, 0.848 mins elapsed.\n",
      "\n",
      "2025-02-26 14:03:36.012158 : Finding Peak Gene Pairings, 0.963 mins elapsed.\n",
      "\n",
      "2025-02-26 14:03:37.165642 : Computing Correlations, 0.982 mins elapsed.\n",
      "\n",
      "2025-02-26 14:03:52.930975 : Completed Peak2Gene Correlations!, 1.245 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeak2GeneLinks-29e71b2fb6054e-Date-2025-02-26_Time-14-02-37.506703.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "obj.ArchR = addPeak2GeneLinks(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    reducedDims = \"IterativeLSI\",\n",
    "    useMatrix = \"GeneExpressionMatrix\",\n",
    "    maxDist = maxDist,\n",
    "    threads = n.cores\n",
    ")\n",
    "end_time <- Sys.time()\n",
    "execution_time <- end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "128d0065-5f22-495b-b00d-743fc9843337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 1.25748 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "execution_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "36bd2df2-6b32-4b81-bd4f-39ef8913b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g = metadata(obj.ArchR@peakSet)$Peak2GeneLinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e7730413-8fd4-4952-bb13-0df7a3b15b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata(p2g)[[1]]$PeakName = paste(seqnames(metadata(p2g)[[1]]),\n",
    "                                    ranges(metadata(p2g)[[1]]),sep=\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "aeea3810-6735-4aff-9bd7-1697bef979d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g$PairName = paste(metadata(p2g)[[1]][p2g$idxATAC]$PeakName,\n",
    "                     metadata(p2g)[[2]][p2g$idxRNA]$name,\n",
    "                     sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "87f975ec-6ba4-4696-ad6c-2af65da7aa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 3921433 rows and 7 columns\n",
       "                                idxATAC    idxRNA Correlation         FDR\n",
       "                              <integer> <integer>   <numeric>   <numeric>\n",
       "chr1-9926-10618_FAM138A               1         1         NaN         NaN\n",
       "chr1-15894-16509_FAM138A              2         1         NaN         NaN\n",
       "chr1-17271-17713_FAM138A              3         1         NaN         NaN\n",
       "chr1-29230-29495_FAM138A              4         1         NaN         NaN\n",
       "chr1-115493-115961_FAM138A            5         1         NaN         NaN\n",
       "...                                 ...       ...         ...         ...\n",
       "chrX-156009722-156010137_IL9R    157596     20493   0.0787222 0.141580666\n",
       "chrX-156016191-156016823_IL9R    157597     20493   0.0747140 0.165590419\n",
       "chrX-156019811-156020191_IL9R    157598     20493   0.0323247 0.585302269\n",
       "chrX-156025028-156025591_IL9R    157599     20493  -0.1003445 0.054041307\n",
       "chrX-156030021-156030865_IL9R    157600     20493   0.1683037 0.000643542\n",
       "                               VarQATAC   VarQRNA               PairName\n",
       "                              <numeric> <numeric>            <character>\n",
       "chr1-9926-10618_FAM138A        0.780032 0.0597765 chr1-9926-10618_FAM1..\n",
       "chr1-15894-16509_FAM138A       0.274651 0.0597765 chr1-15894-16509_FAM..\n",
       "chr1-17271-17713_FAM138A       0.338503 0.0597765 chr1-17271-17713_FAM..\n",
       "chr1-29230-29495_FAM138A       0.222716 0.0597765 chr1-29230-29495_FAM..\n",
       "chr1-115493-115961_FAM138A     0.843997 0.0597765 chr1-115493-115961_F..\n",
       "...                                 ...       ...                    ...\n",
       "chrX-156009722-156010137_IL9R 0.0590863  0.349485 chrX-156009722-15601..\n",
       "chrX-156016191-156016823_IL9R 0.5508439  0.349485 chrX-156016191-15601..\n",
       "chrX-156019811-156020191_IL9R 0.1883376  0.349485 chrX-156019811-15602..\n",
       "chrX-156025028-156025591_IL9R 0.4855330  0.349485 chrX-156025028-15602..\n",
       "chrX-156030021-156030865_IL9R 0.8127475  0.349485 chrX-156030021-15603.."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(p2g) = p2g$PairName\n",
    "p2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c7209a32-3904-4734-93af-9c22049f652a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = pairs.E2G[pairs.E2G$PairName %in% p2g$PairName]\n",
    "mcols(pairs.E2G.res)[,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")] = p2g[pairs.E2G.res$PairName,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1ab4ed70-ce9e-4857-bd1b-63a81d27a401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3138687 ranges and 7 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10001196-10001745      * | APITD1-CORT\n",
       "        [2]     chr1 10001196-10001745      * |       CASZ1\n",
       "        [3]     chr1 10001196-10001745      * |       CENPS\n",
       "        [4]     chr1 10001196-10001745      * |      CLSTN1\n",
       "        [5]     chr1 10001196-10001745      * |        CORT\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [3138683]     chrX     989754-991253      * |     SLC25A6\n",
       "  [3138684]     chrX   9995827-9996208      * |       CLCN4\n",
       "  [3138685]     chrX   9995827-9996208      * |       FAM9B\n",
       "  [3138686]     chrX   9995827-9996208      * |       TBL1X\n",
       "  [3138687]     chrX   9995827-9996208      * |        WWC3\n",
       "                          PeakName               PairName Correlation       FDR\n",
       "                       <character>            <character>   <numeric> <numeric>\n",
       "        [1] chr1-10001196-10001745 chr1-10001196-100017..  0.00775952  0.905464\n",
       "        [2] chr1-10001196-10001745 chr1-10001196-100017.. -0.06787976  0.212933\n",
       "        [3] chr1-10001196-10001745 chr1-10001196-100017.. -0.05437093  0.330969\n",
       "        [4] chr1-10001196-10001745 chr1-10001196-100017.. -0.00856967  0.895217\n",
       "        [5] chr1-10001196-10001745 chr1-10001196-100017..  0.02389720  0.694848\n",
       "        ...                    ...                    ...         ...       ...\n",
       "  [3138683]     chrX-989754-991253 chrX-989754-991253_S.. -0.13882537 0.0057679\n",
       "  [3138684]   chrX-9995827-9996208 chrX-9995827-9996208.. -0.02470933 0.6841444\n",
       "  [3138685]   chrX-9995827-9996208 chrX-9995827-9996208.. -0.05067681 0.3687553\n",
       "  [3138686]   chrX-9995827-9996208 chrX-9995827-9996208.. -0.00962123 0.8818629\n",
       "  [3138687]   chrX-9995827-9996208 chrX-9995827-9996208.. -0.03551510 0.5448640\n",
       "             VarQATAC   VarQRNA\n",
       "            <numeric> <numeric>\n",
       "        [1]  0.121796  0.403894\n",
       "        [2]  0.121796  0.403064\n",
       "        [3]  0.121796  0.533792\n",
       "        [4]  0.121796  0.664959\n",
       "        [5]  0.121796  0.188796\n",
       "        ...       ...       ...\n",
       "  [3138683]  0.913065  0.828234\n",
       "  [3138684]  0.339911  0.425365\n",
       "  [3138685]  0.339911  0.426146\n",
       "  [3138686]  0.339911  0.917142\n",
       "  [3138687]  0.339911  0.461231\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46485351-50d7-461a-bf09-d795d5bdab84",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6d424e21-6a41-4422-b76c-6f1c0bda2222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3138687 ranges and 7 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10001196-10001745      * | APITD1-CORT\n",
       "        [2]     chr1 10001196-10001745      * |       CASZ1\n",
       "        [3]     chr1 10001196-10001745      * |       CENPS\n",
       "        [4]     chr1 10001196-10001745      * |      CLSTN1\n",
       "        [5]     chr1 10001196-10001745      * |        CORT\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [3138683]     chrX     989754-991253      * |     SLC25A6\n",
       "  [3138684]     chrX   9995827-9996208      * |       CLCN4\n",
       "  [3138685]     chrX   9995827-9996208      * |       FAM9B\n",
       "  [3138686]     chrX   9995827-9996208      * |       TBL1X\n",
       "  [3138687]     chrX   9995827-9996208      * |        WWC3\n",
       "                          PeakName               PairName Correlation       FDR\n",
       "                       <character>            <character>   <numeric> <numeric>\n",
       "        [1] chr1-10001196-10001745 chr1-10001196-100017..  0.00775952  0.905464\n",
       "        [2] chr1-10001196-10001745 chr1-10001196-100017.. -0.06787976  0.212933\n",
       "        [3] chr1-10001196-10001745 chr1-10001196-100017.. -0.05437093  0.330969\n",
       "        [4] chr1-10001196-10001745 chr1-10001196-100017.. -0.00856967  0.895217\n",
       "        [5] chr1-10001196-10001745 chr1-10001196-100017..  0.02389720  0.694848\n",
       "        ...                    ...                    ...         ...       ...\n",
       "  [3138683]     chrX-989754-991253 chrX-989754-991253_S.. -0.13882537 0.0057679\n",
       "  [3138684]   chrX-9995827-9996208 chrX-9995827-9996208.. -0.02470933 0.6841444\n",
       "  [3138685]   chrX-9995827-9996208 chrX-9995827-9996208.. -0.05067681 0.3687553\n",
       "  [3138686]   chrX-9995827-9996208 chrX-9995827-9996208.. -0.00962123 0.8818629\n",
       "  [3138687]   chrX-9995827-9996208 chrX-9995827-9996208.. -0.03551510 0.5448640\n",
       "             VarQATAC   VarQRNA\n",
       "            <numeric> <numeric>\n",
       "        [1]  0.121796  0.403894\n",
       "        [2]  0.121796  0.403064\n",
       "        [3]  0.121796  0.533792\n",
       "        [4]  0.121796  0.664959\n",
       "        [5]  0.121796  0.188796\n",
       "        ...       ...       ...\n",
       "  [3138683]  0.913065  0.828234\n",
       "  [3138684]  0.339911  0.425365\n",
       "  [3138685]  0.339911  0.426146\n",
       "  [3138686]  0.339911  0.917142\n",
       "  [3138687]  0.339911  0.461231\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3138687 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>Correlation</th><th scope=col>FDR</th><th scope=col>VarQATAC</th><th scope=col>VarQRNA</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>APITD1-CORT</td><td>K562</td><td> 0.0077595237</td><td>9.054638e-01</td><td>0.1217957</td><td>0.40389401</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CASZ1      </td><td>K562</td><td>-0.0678797647</td><td>2.129327e-01</td><td>0.1217957</td><td>0.40306446</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CENPS      </td><td>K562</td><td>-0.0543709313</td><td>3.309685e-01</td><td>0.1217957</td><td>0.53379203</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CLSTN1     </td><td>K562</td><td>-0.0085696713</td><td>8.952172e-01</td><td>0.1217957</td><td>0.66495877</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CORT       </td><td>K562</td><td> 0.0238971961</td><td>6.948478e-01</td><td>0.1217957</td><td>0.18879617</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CTNNBIP1   </td><td>K562</td><td> 0.1092765086</td><td>3.417134e-02</td><td>0.1217957</td><td>0.69531059</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>DFFA       </td><td>K562</td><td> 0.0474884439</td><td>4.031689e-01</td><td>0.1217957</td><td>0.89010882</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>GPR157     </td><td>K562</td><td>-0.0864822012</td><td>1.025556e-01</td><td>0.1217957</td><td>0.39598887</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>H6PD       </td><td>K562</td><td>-0.0001260889</td><td>9.985013e-01</td><td>0.1217957</td><td>0.64397599</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>KIF1B      </td><td>K562</td><td>-0.0857113472</td><td>1.060060e-01</td><td>0.1217957</td><td>0.85917142</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>LINC01759  </td><td>K562</td><td>          NaN</td><td>         NaN</td><td>0.1217957</td><td>0.05977651</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>LZIC       </td><td>K562</td><td>-0.1538909633</td><td>1.984068e-03</td><td>0.1217957</td><td>0.70687552</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>MIR34AHG   </td><td>K562</td><td>-0.0074112973</td><td>9.098202e-01</td><td>0.1217957</td><td>0.20777827</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>NMNAT1     </td><td>K562</td><td> 0.0088056946</td><td>8.922335e-01</td><td>0.1217957</td><td>0.71351193</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>PEX14      </td><td>K562</td><td> 0.0626662605</td><td>2.546270e-01</td><td>0.1217957</td><td>0.86083053</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>PGD        </td><td>K562</td><td> 0.1524246606</td><td>2.211931e-03</td><td>0.1217957</td><td>0.90396721</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>PIK3CD     </td><td>K562</td><td> 0.0275055699</td><td>6.475386e-01</td><td>0.1217957</td><td>0.62504270</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>RBP7       </td><td>K562</td><td> 0.0554304252</td><td>3.205579e-01</td><td>0.1217957</td><td>0.24457132</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>SLC25A33   </td><td>K562</td><td>-0.0180378423</td><td>7.722424e-01</td><td>0.1217957</td><td>0.83164983</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>SPSB1      </td><td>K562</td><td>-0.0391685899</td><td>4.998308e-01</td><td>0.1217957</td><td>0.36470990</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>TMEM201    </td><td>K562</td><td> 0.1214595118</td><td>1.724086e-02</td><td>0.1217957</td><td>0.63836432</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>UBE4B      </td><td>K562</td><td> 0.1395134164</td><td>5.507033e-03</td><td>0.1217957</td><td>0.83482165</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>AGL        </td><td>K562</td><td>-0.0608430722</td><td>2.703814e-01</td><td>0.6677602</td><td>0.71507344</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>CDC14A     </td><td>K562</td><td> 0.1762017358</td><td>3.323644e-04</td><td>0.6677602</td><td>0.88791295</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>DBT        </td><td>K562</td><td> 0.1133995174</td><td>2.732068e-02</td><td>0.6677602</td><td>0.68603914</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>DPH5       </td><td>K562</td><td> 0.2329096982</td><td>1.109738e-06</td><td>0.6677602</td><td>0.81349729</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>EXTL2      </td><td>K562</td><td> 0.0499924636</td><td>3.760218e-01</td><td>0.6677602</td><td>0.72405212</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>FRRS1      </td><td>K562</td><td> 0.1424196744</td><td>4.516584e-03</td><td>0.6677602</td><td>0.81915776</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>MFSD14A    </td><td>K562</td><td>-0.0337651515</td><td>5.669177e-01</td><td>0.6677602</td><td>0.49499829</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>PALMD      </td><td>K562</td><td>-0.0839180500</td><td>1.144077e-01</td><td>0.6677602</td><td>0.51939687</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>CLCN4    </td><td>K562</td><td> 0.009178360</td><td>8.875162e-01</td><td>0.2749175</td><td>0.4253648</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>FAM9B    </td><td>K562</td><td> 0.134837610</td><td>7.512612e-03</td><td>0.2749175</td><td>0.4261455</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>TBL1X    </td><td>K562</td><td>-0.068333669</td><td>2.095207e-01</td><td>0.2749175</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>WWC3     </td><td>K562</td><td> 0.075693801</td><td>1.594750e-01</td><td>0.2749175</td><td>0.4612307</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>CLCN4    </td><td>K562</td><td> 0.141609290</td><td>4.775212e-03</td><td>0.4712373</td><td>0.4253648</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>FAM9B    </td><td>K562</td><td> 0.036509916</td><td>5.324533e-01</td><td>0.4712373</td><td>0.4261455</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>TBL1X    </td><td>K562</td><td>-0.236986617</td><td>6.888422e-07</td><td>0.4712373</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>WWC3     </td><td>K562</td><td> 0.018624767</td><td>7.644783e-01</td><td>0.4712373</td><td>0.4612307</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>CLCN4    </td><td>K562</td><td> 0.099906619</td><td>5.521721e-02</td><td>0.1015990</td><td>0.4253648</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>FAM9B    </td><td>K562</td><td> 0.202028342</td><td>3.061945e-05</td><td>0.1015990</td><td>0.4261455</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>TBL1X    </td><td>K562</td><td>-0.001133045</td><td>9.866563e-01</td><td>0.1015990</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>WWC3     </td><td>K562</td><td> 0.002865084</td><td>9.659557e-01</td><td>0.1015990</td><td>0.4612307</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>CLCN4    </td><td>K562</td><td> 0.046537863</td><td>4.137481e-01</td><td>0.3171320</td><td>0.4253648</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>FAM9B    </td><td>K562</td><td> 0.079525083</td><td>1.370903e-01</td><td>0.3171320</td><td>0.4261455</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>TBL1X    </td><td>K562</td><td> 0.050948419</td><td>3.658880e-01</td><td>0.3171320</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>WWC3     </td><td>K562</td><td> 0.001953204</td><td>9.769148e-01</td><td>0.3171320</td><td>0.4612307</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>AKAP17A  </td><td>K562</td><td> 0.197527146</td><td>4.758354e-05</td><td>0.9130647</td><td>0.7429366</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>ASMT     </td><td>K562</td><td>-0.056662832</td><td>3.087044e-01</td><td>0.9130647</td><td>0.2353487</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>ASMTL    </td><td>K562</td><td> 0.080577450</td><td>1.313708e-01</td><td>0.9130647</td><td>0.5432587</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>GTPBP6   </td><td>K562</td><td> 0.177694227</td><td>2.923763e-04</td><td>0.9130647</td><td>0.5998634</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>IL3RA    </td><td>K562</td><td> 0.098622088</td><td>5.877578e-02</td><td>0.9130647</td><td>0.3884741</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>LINC00106</td><td>K562</td><td>-0.049059749</td><td>3.859967e-01</td><td>0.9130647</td><td>0.5148099</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>P2RY8    </td><td>K562</td><td> 0.093105868</td><td>7.630134e-02</td><td>0.9130647</td><td>0.4362465</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>PLCXD1   </td><td>K562</td><td>-0.209146063</td><td>1.492249e-05</td><td>0.9130647</td><td>0.7567950</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>PPP2R3B  </td><td>K562</td><td> 0.039915262</td><td>4.908143e-01</td><td>0.9130647</td><td>0.5489679</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>SLC25A6  </td><td>K562</td><td>-0.138825365</td><td>5.767899e-03</td><td>0.9130647</td><td>0.8282340</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>CLCN4    </td><td>K562</td><td>-0.024709327</td><td>6.841444e-01</td><td>0.3399112</td><td>0.4253648</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>FAM9B    </td><td>K562</td><td>-0.050676807</td><td>3.687553e-01</td><td>0.3399112</td><td>0.4261455</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>TBL1X    </td><td>K562</td><td>-0.009621227</td><td>8.818629e-01</td><td>0.3399112</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>WWC3     </td><td>K562</td><td>-0.035515096</td><td>5.448640e-01</td><td>0.3399112</td><td>0.4612307</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3138687 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & Correlation & FDR & VarQATAC & VarQRNA\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  10001196 &  10001745 & APITD1-CORT & K562 &  0.0077595237 & 9.054638e-01 & 0.1217957 & 0.40389401\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CASZ1       & K562 & -0.0678797647 & 2.129327e-01 & 0.1217957 & 0.40306446\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CENPS       & K562 & -0.0543709313 & 3.309685e-01 & 0.1217957 & 0.53379203\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CLSTN1      & K562 & -0.0085696713 & 8.952172e-01 & 0.1217957 & 0.66495877\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CORT        & K562 &  0.0238971961 & 6.948478e-01 & 0.1217957 & 0.18879617\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CTNNBIP1    & K562 &  0.1092765086 & 3.417134e-02 & 0.1217957 & 0.69531059\\\\\n",
       "\t chr1 &  10001196 &  10001745 & DFFA        & K562 &  0.0474884439 & 4.031689e-01 & 0.1217957 & 0.89010882\\\\\n",
       "\t chr1 &  10001196 &  10001745 & GPR157      & K562 & -0.0864822012 & 1.025556e-01 & 0.1217957 & 0.39598887\\\\\n",
       "\t chr1 &  10001196 &  10001745 & H6PD        & K562 & -0.0001260889 & 9.985013e-01 & 0.1217957 & 0.64397599\\\\\n",
       "\t chr1 &  10001196 &  10001745 & KIF1B       & K562 & -0.0857113472 & 1.060060e-01 & 0.1217957 & 0.85917142\\\\\n",
       "\t chr1 &  10001196 &  10001745 & LINC01759   & K562 &           NaN &          NaN & 0.1217957 & 0.05977651\\\\\n",
       "\t chr1 &  10001196 &  10001745 & LZIC        & K562 & -0.1538909633 & 1.984068e-03 & 0.1217957 & 0.70687552\\\\\n",
       "\t chr1 &  10001196 &  10001745 & MIR34AHG    & K562 & -0.0074112973 & 9.098202e-01 & 0.1217957 & 0.20777827\\\\\n",
       "\t chr1 &  10001196 &  10001745 & NMNAT1      & K562 &  0.0088056946 & 8.922335e-01 & 0.1217957 & 0.71351193\\\\\n",
       "\t chr1 &  10001196 &  10001745 & PEX14       & K562 &  0.0626662605 & 2.546270e-01 & 0.1217957 & 0.86083053\\\\\n",
       "\t chr1 &  10001196 &  10001745 & PGD         & K562 &  0.1524246606 & 2.211931e-03 & 0.1217957 & 0.90396721\\\\\n",
       "\t chr1 &  10001196 &  10001745 & PIK3CD      & K562 &  0.0275055699 & 6.475386e-01 & 0.1217957 & 0.62504270\\\\\n",
       "\t chr1 &  10001196 &  10001745 & RBP7        & K562 &  0.0554304252 & 3.205579e-01 & 0.1217957 & 0.24457132\\\\\n",
       "\t chr1 &  10001196 &  10001745 & SLC25A33    & K562 & -0.0180378423 & 7.722424e-01 & 0.1217957 & 0.83164983\\\\\n",
       "\t chr1 &  10001196 &  10001745 & SPSB1       & K562 & -0.0391685899 & 4.998308e-01 & 0.1217957 & 0.36470990\\\\\n",
       "\t chr1 &  10001196 &  10001745 & TMEM201     & K562 &  0.1214595118 & 1.724086e-02 & 0.1217957 & 0.63836432\\\\\n",
       "\t chr1 &  10001196 &  10001745 & UBE4B       & K562 &  0.1395134164 & 5.507033e-03 & 0.1217957 & 0.83482165\\\\\n",
       "\t chr1 & 100027917 & 100029695 & AGL         & K562 & -0.0608430722 & 2.703814e-01 & 0.6677602 & 0.71507344\\\\\n",
       "\t chr1 & 100027917 & 100029695 & CDC14A      & K562 &  0.1762017358 & 3.323644e-04 & 0.6677602 & 0.88791295\\\\\n",
       "\t chr1 & 100027917 & 100029695 & DBT         & K562 &  0.1133995174 & 2.732068e-02 & 0.6677602 & 0.68603914\\\\\n",
       "\t chr1 & 100027917 & 100029695 & DPH5        & K562 &  0.2329096982 & 1.109738e-06 & 0.6677602 & 0.81349729\\\\\n",
       "\t chr1 & 100027917 & 100029695 & EXTL2       & K562 &  0.0499924636 & 3.760218e-01 & 0.6677602 & 0.72405212\\\\\n",
       "\t chr1 & 100027917 & 100029695 & FRRS1       & K562 &  0.1424196744 & 4.516584e-03 & 0.6677602 & 0.81915776\\\\\n",
       "\t chr1 & 100027917 & 100029695 & MFSD14A     & K562 & -0.0337651515 & 5.669177e-01 & 0.6677602 & 0.49499829\\\\\n",
       "\t chr1 & 100027917 & 100029695 & PALMD       & K562 & -0.0839180500 & 1.144077e-01 & 0.6677602 & 0.51939687\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9873821 & 9874426 & CLCN4     & K562 &  0.009178360 & 8.875162e-01 & 0.2749175 & 0.4253648\\\\\n",
       "\t chrX & 9873821 & 9874426 & FAM9B     & K562 &  0.134837610 & 7.512612e-03 & 0.2749175 & 0.4261455\\\\\n",
       "\t chrX & 9873821 & 9874426 & TBL1X     & K562 & -0.068333669 & 2.095207e-01 & 0.2749175 & 0.9171424\\\\\n",
       "\t chrX & 9873821 & 9874426 & WWC3      & K562 &  0.075693801 & 1.594750e-01 & 0.2749175 & 0.4612307\\\\\n",
       "\t chrX & 9882635 & 9883318 & CLCN4     & K562 &  0.141609290 & 4.775212e-03 & 0.4712373 & 0.4253648\\\\\n",
       "\t chrX & 9882635 & 9883318 & FAM9B     & K562 &  0.036509916 & 5.324533e-01 & 0.4712373 & 0.4261455\\\\\n",
       "\t chrX & 9882635 & 9883318 & TBL1X     & K562 & -0.236986617 & 6.888422e-07 & 0.4712373 & 0.9171424\\\\\n",
       "\t chrX & 9882635 & 9883318 & WWC3      & K562 &  0.018624767 & 7.644783e-01 & 0.4712373 & 0.4612307\\\\\n",
       "\t chrX & 9885160 & 9885475 & CLCN4     & K562 &  0.099906619 & 5.521721e-02 & 0.1015990 & 0.4253648\\\\\n",
       "\t chrX & 9885160 & 9885475 & FAM9B     & K562 &  0.202028342 & 3.061945e-05 & 0.1015990 & 0.4261455\\\\\n",
       "\t chrX & 9885160 & 9885475 & TBL1X     & K562 & -0.001133045 & 9.866563e-01 & 0.1015990 & 0.9171424\\\\\n",
       "\t chrX & 9885160 & 9885475 & WWC3      & K562 &  0.002865084 & 9.659557e-01 & 0.1015990 & 0.4612307\\\\\n",
       "\t chrX & 9894484 & 9895274 & CLCN4     & K562 &  0.046537863 & 4.137481e-01 & 0.3171320 & 0.4253648\\\\\n",
       "\t chrX & 9894484 & 9895274 & FAM9B     & K562 &  0.079525083 & 1.370903e-01 & 0.3171320 & 0.4261455\\\\\n",
       "\t chrX & 9894484 & 9895274 & TBL1X     & K562 &  0.050948419 & 3.658880e-01 & 0.3171320 & 0.9171424\\\\\n",
       "\t chrX & 9894484 & 9895274 & WWC3      & K562 &  0.001953204 & 9.769148e-01 & 0.3171320 & 0.4612307\\\\\n",
       "\t chrX &  989754 &  991253 & AKAP17A   & K562 &  0.197527146 & 4.758354e-05 & 0.9130647 & 0.7429366\\\\\n",
       "\t chrX &  989754 &  991253 & ASMT      & K562 & -0.056662832 & 3.087044e-01 & 0.9130647 & 0.2353487\\\\\n",
       "\t chrX &  989754 &  991253 & ASMTL     & K562 &  0.080577450 & 1.313708e-01 & 0.9130647 & 0.5432587\\\\\n",
       "\t chrX &  989754 &  991253 & GTPBP6    & K562 &  0.177694227 & 2.923763e-04 & 0.9130647 & 0.5998634\\\\\n",
       "\t chrX &  989754 &  991253 & IL3RA     & K562 &  0.098622088 & 5.877578e-02 & 0.9130647 & 0.3884741\\\\\n",
       "\t chrX &  989754 &  991253 & LINC00106 & K562 & -0.049059749 & 3.859967e-01 & 0.9130647 & 0.5148099\\\\\n",
       "\t chrX &  989754 &  991253 & P2RY8     & K562 &  0.093105868 & 7.630134e-02 & 0.9130647 & 0.4362465\\\\\n",
       "\t chrX &  989754 &  991253 & PLCXD1    & K562 & -0.209146063 & 1.492249e-05 & 0.9130647 & 0.7567950\\\\\n",
       "\t chrX &  989754 &  991253 & PPP2R3B   & K562 &  0.039915262 & 4.908143e-01 & 0.9130647 & 0.5489679\\\\\n",
       "\t chrX &  989754 &  991253 & SLC25A6   & K562 & -0.138825365 & 5.767899e-03 & 0.9130647 & 0.8282340\\\\\n",
       "\t chrX & 9995827 & 9996208 & CLCN4     & K562 & -0.024709327 & 6.841444e-01 & 0.3399112 & 0.4253648\\\\\n",
       "\t chrX & 9995827 & 9996208 & FAM9B     & K562 & -0.050676807 & 3.687553e-01 & 0.3399112 & 0.4261455\\\\\n",
       "\t chrX & 9995827 & 9996208 & TBL1X     & K562 & -0.009621227 & 8.818629e-01 & 0.3399112 & 0.9171424\\\\\n",
       "\t chrX & 9995827 & 9996208 & WWC3      & K562 & -0.035515096 & 5.448640e-01 & 0.3399112 & 0.4612307\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3138687 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | Correlation &lt;dbl&gt; | FDR &lt;dbl&gt; | VarQATAC &lt;dbl&gt; | VarQRNA &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  10001196 |  10001745 | APITD1-CORT | K562 |  0.0077595237 | 9.054638e-01 | 0.1217957 | 0.40389401 |\n",
       "| chr1 |  10001196 |  10001745 | CASZ1       | K562 | -0.0678797647 | 2.129327e-01 | 0.1217957 | 0.40306446 |\n",
       "| chr1 |  10001196 |  10001745 | CENPS       | K562 | -0.0543709313 | 3.309685e-01 | 0.1217957 | 0.53379203 |\n",
       "| chr1 |  10001196 |  10001745 | CLSTN1      | K562 | -0.0085696713 | 8.952172e-01 | 0.1217957 | 0.66495877 |\n",
       "| chr1 |  10001196 |  10001745 | CORT        | K562 |  0.0238971961 | 6.948478e-01 | 0.1217957 | 0.18879617 |\n",
       "| chr1 |  10001196 |  10001745 | CTNNBIP1    | K562 |  0.1092765086 | 3.417134e-02 | 0.1217957 | 0.69531059 |\n",
       "| chr1 |  10001196 |  10001745 | DFFA        | K562 |  0.0474884439 | 4.031689e-01 | 0.1217957 | 0.89010882 |\n",
       "| chr1 |  10001196 |  10001745 | GPR157      | K562 | -0.0864822012 | 1.025556e-01 | 0.1217957 | 0.39598887 |\n",
       "| chr1 |  10001196 |  10001745 | H6PD        | K562 | -0.0001260889 | 9.985013e-01 | 0.1217957 | 0.64397599 |\n",
       "| chr1 |  10001196 |  10001745 | KIF1B       | K562 | -0.0857113472 | 1.060060e-01 | 0.1217957 | 0.85917142 |\n",
       "| chr1 |  10001196 |  10001745 | LINC01759   | K562 |           NaN |          NaN | 0.1217957 | 0.05977651 |\n",
       "| chr1 |  10001196 |  10001745 | LZIC        | K562 | -0.1538909633 | 1.984068e-03 | 0.1217957 | 0.70687552 |\n",
       "| chr1 |  10001196 |  10001745 | MIR34AHG    | K562 | -0.0074112973 | 9.098202e-01 | 0.1217957 | 0.20777827 |\n",
       "| chr1 |  10001196 |  10001745 | NMNAT1      | K562 |  0.0088056946 | 8.922335e-01 | 0.1217957 | 0.71351193 |\n",
       "| chr1 |  10001196 |  10001745 | PEX14       | K562 |  0.0626662605 | 2.546270e-01 | 0.1217957 | 0.86083053 |\n",
       "| chr1 |  10001196 |  10001745 | PGD         | K562 |  0.1524246606 | 2.211931e-03 | 0.1217957 | 0.90396721 |\n",
       "| chr1 |  10001196 |  10001745 | PIK3CD      | K562 |  0.0275055699 | 6.475386e-01 | 0.1217957 | 0.62504270 |\n",
       "| chr1 |  10001196 |  10001745 | RBP7        | K562 |  0.0554304252 | 3.205579e-01 | 0.1217957 | 0.24457132 |\n",
       "| chr1 |  10001196 |  10001745 | SLC25A33    | K562 | -0.0180378423 | 7.722424e-01 | 0.1217957 | 0.83164983 |\n",
       "| chr1 |  10001196 |  10001745 | SPSB1       | K562 | -0.0391685899 | 4.998308e-01 | 0.1217957 | 0.36470990 |\n",
       "| chr1 |  10001196 |  10001745 | TMEM201     | K562 |  0.1214595118 | 1.724086e-02 | 0.1217957 | 0.63836432 |\n",
       "| chr1 |  10001196 |  10001745 | UBE4B       | K562 |  0.1395134164 | 5.507033e-03 | 0.1217957 | 0.83482165 |\n",
       "| chr1 | 100027917 | 100029695 | AGL         | K562 | -0.0608430722 | 2.703814e-01 | 0.6677602 | 0.71507344 |\n",
       "| chr1 | 100027917 | 100029695 | CDC14A      | K562 |  0.1762017358 | 3.323644e-04 | 0.6677602 | 0.88791295 |\n",
       "| chr1 | 100027917 | 100029695 | DBT         | K562 |  0.1133995174 | 2.732068e-02 | 0.6677602 | 0.68603914 |\n",
       "| chr1 | 100027917 | 100029695 | DPH5        | K562 |  0.2329096982 | 1.109738e-06 | 0.6677602 | 0.81349729 |\n",
       "| chr1 | 100027917 | 100029695 | EXTL2       | K562 |  0.0499924636 | 3.760218e-01 | 0.6677602 | 0.72405212 |\n",
       "| chr1 | 100027917 | 100029695 | FRRS1       | K562 |  0.1424196744 | 4.516584e-03 | 0.6677602 | 0.81915776 |\n",
       "| chr1 | 100027917 | 100029695 | MFSD14A     | K562 | -0.0337651515 | 5.669177e-01 | 0.6677602 | 0.49499829 |\n",
       "| chr1 | 100027917 | 100029695 | PALMD       | K562 | -0.0839180500 | 1.144077e-01 | 0.6677602 | 0.51939687 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9873821 | 9874426 | CLCN4     | K562 |  0.009178360 | 8.875162e-01 | 0.2749175 | 0.4253648 |\n",
       "| chrX | 9873821 | 9874426 | FAM9B     | K562 |  0.134837610 | 7.512612e-03 | 0.2749175 | 0.4261455 |\n",
       "| chrX | 9873821 | 9874426 | TBL1X     | K562 | -0.068333669 | 2.095207e-01 | 0.2749175 | 0.9171424 |\n",
       "| chrX | 9873821 | 9874426 | WWC3      | K562 |  0.075693801 | 1.594750e-01 | 0.2749175 | 0.4612307 |\n",
       "| chrX | 9882635 | 9883318 | CLCN4     | K562 |  0.141609290 | 4.775212e-03 | 0.4712373 | 0.4253648 |\n",
       "| chrX | 9882635 | 9883318 | FAM9B     | K562 |  0.036509916 | 5.324533e-01 | 0.4712373 | 0.4261455 |\n",
       "| chrX | 9882635 | 9883318 | TBL1X     | K562 | -0.236986617 | 6.888422e-07 | 0.4712373 | 0.9171424 |\n",
       "| chrX | 9882635 | 9883318 | WWC3      | K562 |  0.018624767 | 7.644783e-01 | 0.4712373 | 0.4612307 |\n",
       "| chrX | 9885160 | 9885475 | CLCN4     | K562 |  0.099906619 | 5.521721e-02 | 0.1015990 | 0.4253648 |\n",
       "| chrX | 9885160 | 9885475 | FAM9B     | K562 |  0.202028342 | 3.061945e-05 | 0.1015990 | 0.4261455 |\n",
       "| chrX | 9885160 | 9885475 | TBL1X     | K562 | -0.001133045 | 9.866563e-01 | 0.1015990 | 0.9171424 |\n",
       "| chrX | 9885160 | 9885475 | WWC3      | K562 |  0.002865084 | 9.659557e-01 | 0.1015990 | 0.4612307 |\n",
       "| chrX | 9894484 | 9895274 | CLCN4     | K562 |  0.046537863 | 4.137481e-01 | 0.3171320 | 0.4253648 |\n",
       "| chrX | 9894484 | 9895274 | FAM9B     | K562 |  0.079525083 | 1.370903e-01 | 0.3171320 | 0.4261455 |\n",
       "| chrX | 9894484 | 9895274 | TBL1X     | K562 |  0.050948419 | 3.658880e-01 | 0.3171320 | 0.9171424 |\n",
       "| chrX | 9894484 | 9895274 | WWC3      | K562 |  0.001953204 | 9.769148e-01 | 0.3171320 | 0.4612307 |\n",
       "| chrX |  989754 |  991253 | AKAP17A   | K562 |  0.197527146 | 4.758354e-05 | 0.9130647 | 0.7429366 |\n",
       "| chrX |  989754 |  991253 | ASMT      | K562 | -0.056662832 | 3.087044e-01 | 0.9130647 | 0.2353487 |\n",
       "| chrX |  989754 |  991253 | ASMTL     | K562 |  0.080577450 | 1.313708e-01 | 0.9130647 | 0.5432587 |\n",
       "| chrX |  989754 |  991253 | GTPBP6    | K562 |  0.177694227 | 2.923763e-04 | 0.9130647 | 0.5998634 |\n",
       "| chrX |  989754 |  991253 | IL3RA     | K562 |  0.098622088 | 5.877578e-02 | 0.9130647 | 0.3884741 |\n",
       "| chrX |  989754 |  991253 | LINC00106 | K562 | -0.049059749 | 3.859967e-01 | 0.9130647 | 0.5148099 |\n",
       "| chrX |  989754 |  991253 | P2RY8     | K562 |  0.093105868 | 7.630134e-02 | 0.9130647 | 0.4362465 |\n",
       "| chrX |  989754 |  991253 | PLCXD1    | K562 | -0.209146063 | 1.492249e-05 | 0.9130647 | 0.7567950 |\n",
       "| chrX |  989754 |  991253 | PPP2R3B   | K562 |  0.039915262 | 4.908143e-01 | 0.9130647 | 0.5489679 |\n",
       "| chrX |  989754 |  991253 | SLC25A6   | K562 | -0.138825365 | 5.767899e-03 | 0.9130647 | 0.8282340 |\n",
       "| chrX | 9995827 | 9996208 | CLCN4     | K562 | -0.024709327 | 6.841444e-01 | 0.3399112 | 0.4253648 |\n",
       "| chrX | 9995827 | 9996208 | FAM9B     | K562 | -0.050676807 | 3.687553e-01 | 0.3399112 | 0.4261455 |\n",
       "| chrX | 9995827 | 9996208 | TBL1X     | K562 | -0.009621227 | 8.818629e-01 | 0.3399112 | 0.9171424 |\n",
       "| chrX | 9995827 | 9996208 | WWC3      | K562 | -0.035515096 | 5.448640e-01 | 0.3399112 | 0.4612307 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene  CellType Correlation  \n",
       "1       chr1  10001196  10001745 APITD1-CORT K562      0.0077595237\n",
       "2       chr1  10001196  10001745 CASZ1       K562     -0.0678797647\n",
       "3       chr1  10001196  10001745 CENPS       K562     -0.0543709313\n",
       "4       chr1  10001196  10001745 CLSTN1      K562     -0.0085696713\n",
       "5       chr1  10001196  10001745 CORT        K562      0.0238971961\n",
       "6       chr1  10001196  10001745 CTNNBIP1    K562      0.1092765086\n",
       "7       chr1  10001196  10001745 DFFA        K562      0.0474884439\n",
       "8       chr1  10001196  10001745 GPR157      K562     -0.0864822012\n",
       "9       chr1  10001196  10001745 H6PD        K562     -0.0001260889\n",
       "10      chr1  10001196  10001745 KIF1B       K562     -0.0857113472\n",
       "11      chr1  10001196  10001745 LINC01759   K562               NaN\n",
       "12      chr1  10001196  10001745 LZIC        K562     -0.1538909633\n",
       "13      chr1  10001196  10001745 MIR34AHG    K562     -0.0074112973\n",
       "14      chr1  10001196  10001745 NMNAT1      K562      0.0088056946\n",
       "15      chr1  10001196  10001745 PEX14       K562      0.0626662605\n",
       "16      chr1  10001196  10001745 PGD         K562      0.1524246606\n",
       "17      chr1  10001196  10001745 PIK3CD      K562      0.0275055699\n",
       "18      chr1  10001196  10001745 RBP7        K562      0.0554304252\n",
       "19      chr1  10001196  10001745 SLC25A33    K562     -0.0180378423\n",
       "20      chr1  10001196  10001745 SPSB1       K562     -0.0391685899\n",
       "21      chr1  10001196  10001745 TMEM201     K562      0.1214595118\n",
       "22      chr1  10001196  10001745 UBE4B       K562      0.1395134164\n",
       "23      chr1 100027917 100029695 AGL         K562     -0.0608430722\n",
       "24      chr1 100027917 100029695 CDC14A      K562      0.1762017358\n",
       "25      chr1 100027917 100029695 DBT         K562      0.1133995174\n",
       "26      chr1 100027917 100029695 DPH5        K562      0.2329096982\n",
       "27      chr1 100027917 100029695 EXTL2       K562      0.0499924636\n",
       "28      chr1 100027917 100029695 FRRS1       K562      0.1424196744\n",
       "29      chr1 100027917 100029695 MFSD14A     K562     -0.0337651515\n",
       "30      chr1 100027917 100029695 PALMD       K562     -0.0839180500\n",
       "⋮       ⋮    ⋮         ⋮         ⋮           ⋮        ⋮            \n",
       "3138658 chrX 9873821   9874426   CLCN4       K562      0.009178360 \n",
       "3138659 chrX 9873821   9874426   FAM9B       K562      0.134837610 \n",
       "3138660 chrX 9873821   9874426   TBL1X       K562     -0.068333669 \n",
       "3138661 chrX 9873821   9874426   WWC3        K562      0.075693801 \n",
       "3138662 chrX 9882635   9883318   CLCN4       K562      0.141609290 \n",
       "3138663 chrX 9882635   9883318   FAM9B       K562      0.036509916 \n",
       "3138664 chrX 9882635   9883318   TBL1X       K562     -0.236986617 \n",
       "3138665 chrX 9882635   9883318   WWC3        K562      0.018624767 \n",
       "3138666 chrX 9885160   9885475   CLCN4       K562      0.099906619 \n",
       "3138667 chrX 9885160   9885475   FAM9B       K562      0.202028342 \n",
       "3138668 chrX 9885160   9885475   TBL1X       K562     -0.001133045 \n",
       "3138669 chrX 9885160   9885475   WWC3        K562      0.002865084 \n",
       "3138670 chrX 9894484   9895274   CLCN4       K562      0.046537863 \n",
       "3138671 chrX 9894484   9895274   FAM9B       K562      0.079525083 \n",
       "3138672 chrX 9894484   9895274   TBL1X       K562      0.050948419 \n",
       "3138673 chrX 9894484   9895274   WWC3        K562      0.001953204 \n",
       "3138674 chrX  989754    991253   AKAP17A     K562      0.197527146 \n",
       "3138675 chrX  989754    991253   ASMT        K562     -0.056662832 \n",
       "3138676 chrX  989754    991253   ASMTL       K562      0.080577450 \n",
       "3138677 chrX  989754    991253   GTPBP6      K562      0.177694227 \n",
       "3138678 chrX  989754    991253   IL3RA       K562      0.098622088 \n",
       "3138679 chrX  989754    991253   LINC00106   K562     -0.049059749 \n",
       "3138680 chrX  989754    991253   P2RY8       K562      0.093105868 \n",
       "3138681 chrX  989754    991253   PLCXD1      K562     -0.209146063 \n",
       "3138682 chrX  989754    991253   PPP2R3B     K562      0.039915262 \n",
       "3138683 chrX  989754    991253   SLC25A6     K562     -0.138825365 \n",
       "3138684 chrX 9995827   9996208   CLCN4       K562     -0.024709327 \n",
       "3138685 chrX 9995827   9996208   FAM9B       K562     -0.050676807 \n",
       "3138686 chrX 9995827   9996208   TBL1X       K562     -0.009621227 \n",
       "3138687 chrX 9995827   9996208   WWC3        K562     -0.035515096 \n",
       "        FDR          VarQATAC  VarQRNA   \n",
       "1       9.054638e-01 0.1217957 0.40389401\n",
       "2       2.129327e-01 0.1217957 0.40306446\n",
       "3       3.309685e-01 0.1217957 0.53379203\n",
       "4       8.952172e-01 0.1217957 0.66495877\n",
       "5       6.948478e-01 0.1217957 0.18879617\n",
       "6       3.417134e-02 0.1217957 0.69531059\n",
       "7       4.031689e-01 0.1217957 0.89010882\n",
       "8       1.025556e-01 0.1217957 0.39598887\n",
       "9       9.985013e-01 0.1217957 0.64397599\n",
       "10      1.060060e-01 0.1217957 0.85917142\n",
       "11               NaN 0.1217957 0.05977651\n",
       "12      1.984068e-03 0.1217957 0.70687552\n",
       "13      9.098202e-01 0.1217957 0.20777827\n",
       "14      8.922335e-01 0.1217957 0.71351193\n",
       "15      2.546270e-01 0.1217957 0.86083053\n",
       "16      2.211931e-03 0.1217957 0.90396721\n",
       "17      6.475386e-01 0.1217957 0.62504270\n",
       "18      3.205579e-01 0.1217957 0.24457132\n",
       "19      7.722424e-01 0.1217957 0.83164983\n",
       "20      4.998308e-01 0.1217957 0.36470990\n",
       "21      1.724086e-02 0.1217957 0.63836432\n",
       "22      5.507033e-03 0.1217957 0.83482165\n",
       "23      2.703814e-01 0.6677602 0.71507344\n",
       "24      3.323644e-04 0.6677602 0.88791295\n",
       "25      2.732068e-02 0.6677602 0.68603914\n",
       "26      1.109738e-06 0.6677602 0.81349729\n",
       "27      3.760218e-01 0.6677602 0.72405212\n",
       "28      4.516584e-03 0.6677602 0.81915776\n",
       "29      5.669177e-01 0.6677602 0.49499829\n",
       "30      1.144077e-01 0.6677602 0.51939687\n",
       "⋮       ⋮            ⋮         ⋮         \n",
       "3138658 8.875162e-01 0.2749175 0.4253648 \n",
       "3138659 7.512612e-03 0.2749175 0.4261455 \n",
       "3138660 2.095207e-01 0.2749175 0.9171424 \n",
       "3138661 1.594750e-01 0.2749175 0.4612307 \n",
       "3138662 4.775212e-03 0.4712373 0.4253648 \n",
       "3138663 5.324533e-01 0.4712373 0.4261455 \n",
       "3138664 6.888422e-07 0.4712373 0.9171424 \n",
       "3138665 7.644783e-01 0.4712373 0.4612307 \n",
       "3138666 5.521721e-02 0.1015990 0.4253648 \n",
       "3138667 3.061945e-05 0.1015990 0.4261455 \n",
       "3138668 9.866563e-01 0.1015990 0.9171424 \n",
       "3138669 9.659557e-01 0.1015990 0.4612307 \n",
       "3138670 4.137481e-01 0.3171320 0.4253648 \n",
       "3138671 1.370903e-01 0.3171320 0.4261455 \n",
       "3138672 3.658880e-01 0.3171320 0.9171424 \n",
       "3138673 9.769148e-01 0.3171320 0.4612307 \n",
       "3138674 4.758354e-05 0.9130647 0.7429366 \n",
       "3138675 3.087044e-01 0.9130647 0.2353487 \n",
       "3138676 1.313708e-01 0.9130647 0.5432587 \n",
       "3138677 2.923763e-04 0.9130647 0.5998634 \n",
       "3138678 5.877578e-02 0.9130647 0.3884741 \n",
       "3138679 3.859967e-01 0.9130647 0.5148099 \n",
       "3138680 7.630134e-02 0.9130647 0.4362465 \n",
       "3138681 1.492249e-05 0.9130647 0.7567950 \n",
       "3138682 4.908143e-01 0.9130647 0.5489679 \n",
       "3138683 5.767899e-03 0.9130647 0.8282340 \n",
       "3138684 6.841444e-01 0.3399112 0.4253648 \n",
       "3138685 3.687553e-01 0.3399112 0.4261455 \n",
       "3138686 8.818629e-01 0.3399112 0.9171424 \n",
       "3138687 5.448640e-01 0.3399112 0.4612307 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(obj.ArchR,\n",
    "        paste(dir.output,\"obj.ArchR.rds\",sep = \"/\"))\n",
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"Correlation\",\n",
    "                         \"FDR\",\n",
    "                         \"VarQATAC\",\n",
    "                         \"VarQRNA\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "pairs.E2G.res\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4c909ed5-7321-4e69-8208-1544b371da6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] nabor_0.5.0                       dplyr_1.1.4                      \n",
       " [3] uwot_0.1.16                       Rsamtools_2.18.0                 \n",
       " [5] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.68.0                  \n",
       " [7] rtracklayer_1.60.1                Biostrings_2.70.1                \n",
       " [9] XVector_0.42.0                    genomation_1.34.0                \n",
       "[11] Signac_1.13.0                     Seurat_5.0.3                     \n",
       "[13] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[15] rhdf5_2.44.0                      SummarizedExperiment_1.30.2      \n",
       "[17] Biobase_2.60.0                    MatrixGenerics_1.12.3            \n",
       "[19] Rcpp_1.0.12                       Matrix_1.6-5                     \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               matrixStats_1.3.0                \n",
       "[27] data.table_1.15.2                 stringr_1.5.1                    \n",
       "[29] plyr_1.8.9                        magrittr_2.0.3                   \n",
       "[31] ggplot2_3.5.1                     gtable_0.3.5                     \n",
       "[33] gtools_3.9.5                      gridExtra_2.3                    \n",
       "[35] ArchR_1.0.2                      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22         splines_4.3.3            later_1.3.2             \n",
       "  [4] pbdZMQ_0.3-11            BiocIO_1.10.0            bitops_1.0-7            \n",
       "  [7] tibble_3.2.1             polyclip_1.10-6          XML_3.99-0.16.1         \n",
       " [10] fastDummies_1.7.3        lifecycle_1.0.4          vroom_1.6.5             \n",
       " [13] globals_0.16.3           lattice_0.22-6           MASS_7.3-60             \n",
       " [16] plotly_4.10.4            plotrix_3.8-4            yaml_2.3.8              \n",
       " [19] httpuv_1.6.15            sctransform_0.4.1        spam_2.10-0             \n",
       " [22] spatstat.sparse_3.0-3    reticulate_1.36.1        cowplot_1.1.3           \n",
       " [25] pbapply_1.7-2            RColorBrewer_1.1-3       abind_1.4-5             \n",
       " [28] zlibbioc_1.48.0          Rtsne_0.17               purrr_1.0.2             \n",
       " [31] RCurl_1.98-1.14          GenomeInfoDbData_1.2.11  ggrepel_0.9.5           \n",
       " [34] irlba_2.3.5.1            listenv_0.9.1            spatstat.utils_3.0-4    \n",
       " [37] goftest_1.2-3            RSpectra_0.16-1          spatstat.random_3.2-3   \n",
       " [40] fitdistrplus_1.1-11      parallelly_1.37.1        leiden_0.4.3.1          \n",
       " [43] codetools_0.2-20         DelayedArray_0.26.7      RcppRoll_0.3.0          \n",
       " [46] tidyselect_1.2.1         farver_2.1.1             base64enc_0.1-3         \n",
       " [49] spatstat.explore_3.2-6   GenomicAlignments_1.36.0 jsonlite_1.8.8          \n",
       " [52] progressr_0.14.0         ggridges_0.5.6           survival_3.6-4          \n",
       " [55] tools_4.3.3              ica_1.0-3                glue_1.7.0              \n",
       " [58] IRdisplay_1.1            withr_3.0.0              fastmap_1.1.1           \n",
       " [61] rhdf5filters_1.12.1      fansi_1.0.6              digest_0.6.35           \n",
       " [64] R6_2.5.1                 mime_0.12                seqPattern_1.34.0       \n",
       " [67] colorspace_2.1-0         Cairo_1.6-2              scattermore_1.2         \n",
       " [70] tensor_1.5               spatstat.data_3.0-4      utf8_1.2.4              \n",
       " [73] tidyr_1.3.1              generics_0.1.3           httr_1.4.7              \n",
       " [76] htmlwidgets_1.6.4        S4Arrays_1.2.1           pkgconfig_2.0.3         \n",
       " [79] impute_1.76.0            lmtest_0.9-40            htmltools_0.5.8.1       \n",
       " [82] dotCall64_1.1-1          scales_1.3.0             png_0.1-8               \n",
       " [85] tzdb_0.4.0               reshape2_1.4.4           rjson_0.2.21            \n",
       " [88] uuid_1.2-0               nlme_3.1-164             repr_1.1.7              \n",
       " [91] zoo_1.8-12               KernSmooth_2.23-22       miniUI_0.1.1.1          \n",
       " [94] restfulr_0.0.15          pillar_1.9.0             vctrs_0.6.5             \n",
       " [97] RANN_2.6.1               promises_1.3.0           xtable_1.8-4            \n",
       "[100] cluster_2.1.6            evaluate_0.23            readr_2.1.5             \n",
       "[103] cli_3.6.2                compiler_4.3.3           rlang_1.1.3             \n",
       "[106] crayon_1.5.2             future.apply_1.11.2      labeling_0.4.3          \n",
       "[109] stringi_1.8.3            gridBase_0.4-7           viridisLite_0.4.2       \n",
       "[112] deldir_2.0-4             BiocParallel_1.36.0      munsell_0.5.1           \n",
       "[115] lazyeval_0.2.2           spatstat.geom_3.2-9      IRkernel_1.3.2          \n",
       "[118] RcppHNSW_0.6.0           hms_1.1.3                patchwork_1.2.0         \n",
       "[121] bit64_4.0.5              future_1.33.2            Rhdf5lib_1.22.1         \n",
       "[124] shiny_1.8.1.1            ROCR_1.0-11              igraph_2.0.3            \n",
       "[127] fastmatch_1.1-4          bit_4.0.5               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b54012ba-20d1-4787-8d0d-d851ef1744c4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
