{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Xu/K562/Kendall/atac_matrix.csv.gz\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/1.prepare_data/1.seurat_pipeline.240507/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/3.Genome_wide_prediction/FigR.default/FigR.240612/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.atac = read.csv(path.matrix.atac_count,\n",
    "                       row.names = 1,\n",
    "                       check.names = F)\n",
    "matrix.atac = Matrix(as.matrix(matrix.atac), \n",
    "                     sparse = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84a583f5-4b9c-4e4c-b3d0-14a0cc5ae73f",
   "metadata": {},
   "source": [
    "Make ATAC SummarizedExperiment object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a0ac6707-ebb4-46e4-95cb-92c656fb7b32",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.peaks = as.data.frame(do.call(rbind,strsplit(rownames(matrix.atac),\"-\")))\n",
    "colnames(df.peaks) = c(\"chr\",\"start\",\"end\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "44d0cd2c-cc3a-43de-995b-a3893067aa65",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peaks = makeGRangesFromDataFrame(df.peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4445cbf6-f4c2-4f58-9ee7-327546c09607",
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC.se <- SummarizedExperiment(assays=SimpleList(counts=matrix.atac),\n",
    "                                rowRanges=bed.peaks)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Assuming paired scATAC/scRNA-seq data ..\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  4000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  4001  to  5000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  5001  to  6000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  6001  to  7000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  7001  to  7821 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n",
      "Number of peaks in ATAC data: 157600 \n",
      "Number of genes in RNA data: 37141 \n",
      "\n",
      "Num genes overlapping TSS annotation and RNA matrix being considered:  18526 \n",
      "\n",
      "Taking peak summits from peak windows ..\n",
      "Finding overlapping peak-gene pairs ..\n",
      "Found  234556 total gene-peak pairs for given TSS window ..\n",
      "Number of peak summits that overlap any gene TSS window:  111062 \n",
      "Number of gene TSS windows that overlap any peak summit:  17593 \n",
      "\n",
      "Determining background peaks ..\n",
      "Using  100  iterations ..\n",
      "\n",
      "Computing gene-peak correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: iterators\n",
      "\n",
      "Loading required package: parallel\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cisCor <- runGenePeakcorr(ATAC.se = ATAC.se,\n",
    "                          RNAmat = matrix.rna,\n",
    "                          genome = \"hg38\", \n",
    "                          nCores = n.cores, \n",
    "                          p.cut = NULL,\n",
    "                          keepPosCorOnly = F)\n",
    "cisCor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1306279a-873c-4d2f-8f67-29da980aa482",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 111348 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Peak</th><th scope=col>PeakRanges</th><th scope=col>Gene</th><th scope=col>rObs</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td> 1338</td><td>chr1:115493-115961  </td><td>OR4F5    </td><td> 0.0157355382</td><td>0.111991475</td></tr>\n",
       "\t<tr><td>17878</td><td>chr1:804348-805153  </td><td>FAM87B   </td><td> 0.0254473489</td><td>0.050114786</td></tr>\n",
       "\t<tr><td>17901</td><td>chr1:816701-817594  </td><td>FAM87B   </td><td> 0.0266505827</td><td>0.028215881</td></tr>\n",
       "\t<tr><td>17808</td><td>chr1:778120-779401  </td><td>LINC01128</td><td> 0.0249139102</td><td>0.020607569</td></tr>\n",
       "\t<tr><td>17909</td><td>chr1:818917-819221  </td><td>LINC01128</td><td> 0.0096345124</td><td>0.337101803</td></tr>\n",
       "\t<tr><td>17973</td><td>chr1:842493-843467  </td><td>LINC01128</td><td> 0.0418846845</td><td>0.003623764</td></tr>\n",
       "\t<tr><td>18036</td><td>chr1:849559-850153  </td><td>LINC01128</td><td>-0.0005044386</td><td>0.670700760</td></tr>\n",
       "\t<tr><td>18079</td><td>chr1:856218-856698  </td><td>LINC01128</td><td> 0.0005674452</td><td>0.602986234</td></tr>\n",
       "\t<tr><td>18092</td><td>chr1:860170-861174  </td><td>LINC01128</td><td> 0.0091232008</td><td>0.385078608</td></tr>\n",
       "\t<tr><td>18146</td><td>chr1:869578-870262  </td><td>LINC01128</td><td> 0.0151104067</td><td>0.204736225</td></tr>\n",
       "\t<tr><td>18234</td><td>chr1:885999-887141  </td><td>SAMD11   </td><td> 0.0234631743</td><td>0.040165907</td></tr>\n",
       "\t<tr><td>18340</td><td>chr1:897236-897537  </td><td>SAMD11   </td><td> 0.0015554781</td><td>0.580280522</td></tr>\n",
       "\t<tr><td>18345</td><td>chr1:897706-898129  </td><td>SAMD11   </td><td>-0.0029978186</td><td>0.722901208</td></tr>\n",
       "\t<tr><td>18411</td><td>chr1:904067-905237  </td><td>SAMD11   </td><td>-0.0047263047</td><td>0.794375315</td></tr>\n",
       "\t<tr><td>18420</td><td>chr1:906793-907288  </td><td>SAMD11   </td><td> 0.0226430078</td><td>0.081565027</td></tr>\n",
       "\t<tr><td>18464</td><td>chr1:912712-913841  </td><td>SAMD11   </td><td> 0.0133513044</td><td>0.326932441</td></tr>\n",
       "\t<tr><td>18775</td><td>chr1:939995-941085  </td><td>SAMD11   </td><td> 0.0205339126</td><td>0.095705570</td></tr>\n",
       "\t<tr><td>18457</td><td>chr1:910941-911743  </td><td>KLHL17   </td><td> 0.0231678176</td><td>0.286159532</td></tr>\n",
       "\t<tr><td>18504</td><td>chr1:916605-916851  </td><td>KLHL17   </td><td> 0.0202676060</td><td>0.285377222</td></tr>\n",
       "\t<tr><td>18561</td><td>chr1:920895-921548  </td><td>KLHL17   </td><td> 0.0264510507</td><td>0.366994956</td></tr>\n",
       "\t<tr><td>18570</td><td>chr1:923564-924643  </td><td>KLHL17   </td><td> 0.0401515643</td><td>0.090042933</td></tr>\n",
       "\t<tr><td>18587</td><td>chr1:925567-925763  </td><td>KLHL17   </td><td> 0.0210008988</td><td>0.203132826</td></tr>\n",
       "\t<tr><td>18654</td><td>chr1:930551-931222  </td><td>KLHL17   </td><td> 0.0191831667</td><td>0.409348722</td></tr>\n",
       "\t<tr><td>18677</td><td>chr1:931921-932263  </td><td>KLHL17   </td><td> 0.0266950216</td><td>0.162548323</td></tr>\n",
       "\t<tr><td>18748</td><td>chr1:938208-938358  </td><td>KLHL17   </td><td> 0.0297271819</td><td>0.104676600</td></tr>\n",
       "\t<tr><td>18769</td><td>chr1:939194-939374  </td><td>KLHL17   </td><td> 0.0437762830</td><td>0.017832559</td></tr>\n",
       "\t<tr><td>18923</td><td>chr1:982389-982818  </td><td>KLHL17   </td><td> 0.0317919311</td><td>0.050629301</td></tr>\n",
       "\t<tr><td>18955</td><td>chr1:993306-993654  </td><td>KLHL17   </td><td> 0.0152605845</td><td>0.325262283</td></tr>\n",
       "\t<tr><td>  156</td><td>chr1:1012915-1014502</td><td>PLEKHN1  </td><td> 0.0227086367</td><td>0.078223118</td></tr>\n",
       "\t<tr><td>18649</td><td>chr1:930308-930467  </td><td>PLEKHN1  </td><td> 0.0212739466</td><td>0.054493579</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>155093</td><td>chrX:154812460-154812819</td><td>MPP1    </td><td> 0.011999489</td><td>0.201638759</td></tr>\n",
       "\t<tr><td>155094</td><td>chrX:154815322-154815631</td><td>MPP1    </td><td> 0.022366071</td><td>0.064356056</td></tr>\n",
       "\t<tr><td>155095</td><td>chrX:154815707-154815857</td><td>MPP1    </td><td> 0.013773417</td><td>0.260166156</td></tr>\n",
       "\t<tr><td>155098</td><td>chrX:154818257-154818559</td><td>MPP1    </td><td> 0.011741403</td><td>0.245858769</td></tr>\n",
       "\t<tr><td>155100</td><td>chrX:154827231-154828278</td><td>MPP1    </td><td>-0.004119386</td><td>0.726869174</td></tr>\n",
       "\t<tr><td>155101</td><td>chrX:154829509-154830513</td><td>MPP1    </td><td> 0.011158519</td><td>0.280236603</td></tr>\n",
       "\t<tr><td>155103</td><td>chrX:154834579-154834772</td><td>MPP1    </td><td> 0.002888720</td><td>0.452996027</td></tr>\n",
       "\t<tr><td>155104</td><td>chrX:154836057-154836437</td><td>MPP1    </td><td> 0.002853249</td><td>0.479367029</td></tr>\n",
       "\t<tr><td>155087</td><td>chrX:154791986-154792298</td><td>SMIM9   </td><td> 0.027897543</td><td>0.020213167</td></tr>\n",
       "\t<tr><td>155096</td><td>chrX:154816475-154816702</td><td>SMIM9   </td><td>-0.006730153</td><td>0.708034616</td></tr>\n",
       "\t<tr><td>155097</td><td>chrX:154817284-154817570</td><td>SMIM9   </td><td> 0.012246082</td><td>0.147693116</td></tr>\n",
       "\t<tr><td>155099</td><td>chrX:154818880-154820481</td><td>SMIM9   </td><td> 0.029118679</td><td>0.005153993</td></tr>\n",
       "\t<tr><td>155102</td><td>chrX:154833474-154834469</td><td>SMIM9   </td><td> 0.012812935</td><td>0.073982441</td></tr>\n",
       "\t<tr><td>155107</td><td>chrX:154840613-154841413</td><td>SMIM9   </td><td> 0.009056907</td><td>0.177356435</td></tr>\n",
       "\t<tr><td>155105</td><td>chrX:154837474-154837679</td><td>F8      </td><td> 0.007825327</td><td>0.345724664</td></tr>\n",
       "\t<tr><td>155108</td><td>chrX:154910458-154911012</td><td>F8      </td><td> 0.014368714</td><td>0.276274353</td></tr>\n",
       "\t<tr><td>155109</td><td>chrX:154917842-154918632</td><td>F8      </td><td> 0.037911760</td><td>0.018113805</td></tr>\n",
       "\t<tr><td>155125</td><td>chrX:155070725-155072173</td><td>MTCP1   </td><td> 0.019776559</td><td>0.374901485</td></tr>\n",
       "\t<tr><td>155137</td><td>chrX:155215583-155217032</td><td>RAB39B  </td><td> 0.013552040</td><td>0.100575640</td></tr>\n",
       "\t<tr><td>155143</td><td>chrX:155246578-155247364</td><td>RAB39B  </td><td> 0.024990334</td><td>0.006569273</td></tr>\n",
       "\t<tr><td>155145</td><td>chrX:155313817-155314432</td><td>CLIC2   </td><td> 0.018319026</td><td>0.093897379</td></tr>\n",
       "\t<tr><td>155146</td><td>chrX:155333791-155335669</td><td>CLIC2   </td><td> 0.034747011</td><td>0.010289553</td></tr>\n",
       "\t<tr><td>155172</td><td>chrX:155966876-155967347</td><td>WASIR1  </td><td>-0.007168906</td><td>0.830064525</td></tr>\n",
       "\t<tr><td>155173</td><td>chrX:155968122-155968569</td><td>WASIR1  </td><td> 0.030026099</td><td>0.014325011</td></tr>\n",
       "\t<tr><td>155176</td><td>chrX:156000614-156000806</td><td>WASIR1  </td><td> 0.012648940</td><td>0.217951905</td></tr>\n",
       "\t<tr><td>155177</td><td>chrX:156000905-156002182</td><td>WASIR1  </td><td> 0.020380207</td><td>0.117262101</td></tr>\n",
       "\t<tr><td>155178</td><td>chrX:156002257-156003190</td><td>WASIR1  </td><td> 0.018800919</td><td>0.119473828</td></tr>\n",
       "\t<tr><td>155181</td><td>chrX:156016191-156016823</td><td>WASIR1  </td><td> 0.016568172</td><td>0.211021802</td></tr>\n",
       "\t<tr><td>155175</td><td>chrX:155997283-155998084</td><td>DDX11L16</td><td> 0.041291155</td><td>0.001763403</td></tr>\n",
       "\t<tr><td>155180</td><td>chrX:156009722-156010137</td><td>DDX11L16</td><td>-0.001273811</td><td>0.493750653</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 111348 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " Peak & PeakRanges & Gene & rObs & pvalZ\\\\\n",
       " <dbl> & <chr> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t  1338 & chr1:115493-115961   & OR4F5     &  0.0157355382 & 0.111991475\\\\\n",
       "\t 17878 & chr1:804348-805153   & FAM87B    &  0.0254473489 & 0.050114786\\\\\n",
       "\t 17901 & chr1:816701-817594   & FAM87B    &  0.0266505827 & 0.028215881\\\\\n",
       "\t 17808 & chr1:778120-779401   & LINC01128 &  0.0249139102 & 0.020607569\\\\\n",
       "\t 17909 & chr1:818917-819221   & LINC01128 &  0.0096345124 & 0.337101803\\\\\n",
       "\t 17973 & chr1:842493-843467   & LINC01128 &  0.0418846845 & 0.003623764\\\\\n",
       "\t 18036 & chr1:849559-850153   & LINC01128 & -0.0005044386 & 0.670700760\\\\\n",
       "\t 18079 & chr1:856218-856698   & LINC01128 &  0.0005674452 & 0.602986234\\\\\n",
       "\t 18092 & chr1:860170-861174   & LINC01128 &  0.0091232008 & 0.385078608\\\\\n",
       "\t 18146 & chr1:869578-870262   & LINC01128 &  0.0151104067 & 0.204736225\\\\\n",
       "\t 18234 & chr1:885999-887141   & SAMD11    &  0.0234631743 & 0.040165907\\\\\n",
       "\t 18340 & chr1:897236-897537   & SAMD11    &  0.0015554781 & 0.580280522\\\\\n",
       "\t 18345 & chr1:897706-898129   & SAMD11    & -0.0029978186 & 0.722901208\\\\\n",
       "\t 18411 & chr1:904067-905237   & SAMD11    & -0.0047263047 & 0.794375315\\\\\n",
       "\t 18420 & chr1:906793-907288   & SAMD11    &  0.0226430078 & 0.081565027\\\\\n",
       "\t 18464 & chr1:912712-913841   & SAMD11    &  0.0133513044 & 0.326932441\\\\\n",
       "\t 18775 & chr1:939995-941085   & SAMD11    &  0.0205339126 & 0.095705570\\\\\n",
       "\t 18457 & chr1:910941-911743   & KLHL17    &  0.0231678176 & 0.286159532\\\\\n",
       "\t 18504 & chr1:916605-916851   & KLHL17    &  0.0202676060 & 0.285377222\\\\\n",
       "\t 18561 & chr1:920895-921548   & KLHL17    &  0.0264510507 & 0.366994956\\\\\n",
       "\t 18570 & chr1:923564-924643   & KLHL17    &  0.0401515643 & 0.090042933\\\\\n",
       "\t 18587 & chr1:925567-925763   & KLHL17    &  0.0210008988 & 0.203132826\\\\\n",
       "\t 18654 & chr1:930551-931222   & KLHL17    &  0.0191831667 & 0.409348722\\\\\n",
       "\t 18677 & chr1:931921-932263   & KLHL17    &  0.0266950216 & 0.162548323\\\\\n",
       "\t 18748 & chr1:938208-938358   & KLHL17    &  0.0297271819 & 0.104676600\\\\\n",
       "\t 18769 & chr1:939194-939374   & KLHL17    &  0.0437762830 & 0.017832559\\\\\n",
       "\t 18923 & chr1:982389-982818   & KLHL17    &  0.0317919311 & 0.050629301\\\\\n",
       "\t 18955 & chr1:993306-993654   & KLHL17    &  0.0152605845 & 0.325262283\\\\\n",
       "\t   156 & chr1:1012915-1014502 & PLEKHN1   &  0.0227086367 & 0.078223118\\\\\n",
       "\t 18649 & chr1:930308-930467   & PLEKHN1   &  0.0212739466 & 0.054493579\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 155093 & chrX:154812460-154812819 & MPP1     &  0.011999489 & 0.201638759\\\\\n",
       "\t 155094 & chrX:154815322-154815631 & MPP1     &  0.022366071 & 0.064356056\\\\\n",
       "\t 155095 & chrX:154815707-154815857 & MPP1     &  0.013773417 & 0.260166156\\\\\n",
       "\t 155098 & chrX:154818257-154818559 & MPP1     &  0.011741403 & 0.245858769\\\\\n",
       "\t 155100 & chrX:154827231-154828278 & MPP1     & -0.004119386 & 0.726869174\\\\\n",
       "\t 155101 & chrX:154829509-154830513 & MPP1     &  0.011158519 & 0.280236603\\\\\n",
       "\t 155103 & chrX:154834579-154834772 & MPP1     &  0.002888720 & 0.452996027\\\\\n",
       "\t 155104 & chrX:154836057-154836437 & MPP1     &  0.002853249 & 0.479367029\\\\\n",
       "\t 155087 & chrX:154791986-154792298 & SMIM9    &  0.027897543 & 0.020213167\\\\\n",
       "\t 155096 & chrX:154816475-154816702 & SMIM9    & -0.006730153 & 0.708034616\\\\\n",
       "\t 155097 & chrX:154817284-154817570 & SMIM9    &  0.012246082 & 0.147693116\\\\\n",
       "\t 155099 & chrX:154818880-154820481 & SMIM9    &  0.029118679 & 0.005153993\\\\\n",
       "\t 155102 & chrX:154833474-154834469 & SMIM9    &  0.012812935 & 0.073982441\\\\\n",
       "\t 155107 & chrX:154840613-154841413 & SMIM9    &  0.009056907 & 0.177356435\\\\\n",
       "\t 155105 & chrX:154837474-154837679 & F8       &  0.007825327 & 0.345724664\\\\\n",
       "\t 155108 & chrX:154910458-154911012 & F8       &  0.014368714 & 0.276274353\\\\\n",
       "\t 155109 & chrX:154917842-154918632 & F8       &  0.037911760 & 0.018113805\\\\\n",
       "\t 155125 & chrX:155070725-155072173 & MTCP1    &  0.019776559 & 0.374901485\\\\\n",
       "\t 155137 & chrX:155215583-155217032 & RAB39B   &  0.013552040 & 0.100575640\\\\\n",
       "\t 155143 & chrX:155246578-155247364 & RAB39B   &  0.024990334 & 0.006569273\\\\\n",
       "\t 155145 & chrX:155313817-155314432 & CLIC2    &  0.018319026 & 0.093897379\\\\\n",
       "\t 155146 & chrX:155333791-155335669 & CLIC2    &  0.034747011 & 0.010289553\\\\\n",
       "\t 155172 & chrX:155966876-155967347 & WASIR1   & -0.007168906 & 0.830064525\\\\\n",
       "\t 155173 & chrX:155968122-155968569 & WASIR1   &  0.030026099 & 0.014325011\\\\\n",
       "\t 155176 & chrX:156000614-156000806 & WASIR1   &  0.012648940 & 0.217951905\\\\\n",
       "\t 155177 & chrX:156000905-156002182 & WASIR1   &  0.020380207 & 0.117262101\\\\\n",
       "\t 155178 & chrX:156002257-156003190 & WASIR1   &  0.018800919 & 0.119473828\\\\\n",
       "\t 155181 & chrX:156016191-156016823 & WASIR1   &  0.016568172 & 0.211021802\\\\\n",
       "\t 155175 & chrX:155997283-155998084 & DDX11L16 &  0.041291155 & 0.001763403\\\\\n",
       "\t 155180 & chrX:156009722-156010137 & DDX11L16 & -0.001273811 & 0.493750653\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 111348 × 5\n",
       "\n",
       "| Peak &lt;dbl&gt; | PeakRanges &lt;chr&gt; | Gene &lt;chr&gt; | rObs &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "|  1338 | chr1:115493-115961   | OR4F5     |  0.0157355382 | 0.111991475 |\n",
       "| 17878 | chr1:804348-805153   | FAM87B    |  0.0254473489 | 0.050114786 |\n",
       "| 17901 | chr1:816701-817594   | FAM87B    |  0.0266505827 | 0.028215881 |\n",
       "| 17808 | chr1:778120-779401   | LINC01128 |  0.0249139102 | 0.020607569 |\n",
       "| 17909 | chr1:818917-819221   | LINC01128 |  0.0096345124 | 0.337101803 |\n",
       "| 17973 | chr1:842493-843467   | LINC01128 |  0.0418846845 | 0.003623764 |\n",
       "| 18036 | chr1:849559-850153   | LINC01128 | -0.0005044386 | 0.670700760 |\n",
       "| 18079 | chr1:856218-856698   | LINC01128 |  0.0005674452 | 0.602986234 |\n",
       "| 18092 | chr1:860170-861174   | LINC01128 |  0.0091232008 | 0.385078608 |\n",
       "| 18146 | chr1:869578-870262   | LINC01128 |  0.0151104067 | 0.204736225 |\n",
       "| 18234 | chr1:885999-887141   | SAMD11    |  0.0234631743 | 0.040165907 |\n",
       "| 18340 | chr1:897236-897537   | SAMD11    |  0.0015554781 | 0.580280522 |\n",
       "| 18345 | chr1:897706-898129   | SAMD11    | -0.0029978186 | 0.722901208 |\n",
       "| 18411 | chr1:904067-905237   | SAMD11    | -0.0047263047 | 0.794375315 |\n",
       "| 18420 | chr1:906793-907288   | SAMD11    |  0.0226430078 | 0.081565027 |\n",
       "| 18464 | chr1:912712-913841   | SAMD11    |  0.0133513044 | 0.326932441 |\n",
       "| 18775 | chr1:939995-941085   | SAMD11    |  0.0205339126 | 0.095705570 |\n",
       "| 18457 | chr1:910941-911743   | KLHL17    |  0.0231678176 | 0.286159532 |\n",
       "| 18504 | chr1:916605-916851   | KLHL17    |  0.0202676060 | 0.285377222 |\n",
       "| 18561 | chr1:920895-921548   | KLHL17    |  0.0264510507 | 0.366994956 |\n",
       "| 18570 | chr1:923564-924643   | KLHL17    |  0.0401515643 | 0.090042933 |\n",
       "| 18587 | chr1:925567-925763   | KLHL17    |  0.0210008988 | 0.203132826 |\n",
       "| 18654 | chr1:930551-931222   | KLHL17    |  0.0191831667 | 0.409348722 |\n",
       "| 18677 | chr1:931921-932263   | KLHL17    |  0.0266950216 | 0.162548323 |\n",
       "| 18748 | chr1:938208-938358   | KLHL17    |  0.0297271819 | 0.104676600 |\n",
       "| 18769 | chr1:939194-939374   | KLHL17    |  0.0437762830 | 0.017832559 |\n",
       "| 18923 | chr1:982389-982818   | KLHL17    |  0.0317919311 | 0.050629301 |\n",
       "| 18955 | chr1:993306-993654   | KLHL17    |  0.0152605845 | 0.325262283 |\n",
       "|   156 | chr1:1012915-1014502 | PLEKHN1   |  0.0227086367 | 0.078223118 |\n",
       "| 18649 | chr1:930308-930467   | PLEKHN1   |  0.0212739466 | 0.054493579 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 155093 | chrX:154812460-154812819 | MPP1     |  0.011999489 | 0.201638759 |\n",
       "| 155094 | chrX:154815322-154815631 | MPP1     |  0.022366071 | 0.064356056 |\n",
       "| 155095 | chrX:154815707-154815857 | MPP1     |  0.013773417 | 0.260166156 |\n",
       "| 155098 | chrX:154818257-154818559 | MPP1     |  0.011741403 | 0.245858769 |\n",
       "| 155100 | chrX:154827231-154828278 | MPP1     | -0.004119386 | 0.726869174 |\n",
       "| 155101 | chrX:154829509-154830513 | MPP1     |  0.011158519 | 0.280236603 |\n",
       "| 155103 | chrX:154834579-154834772 | MPP1     |  0.002888720 | 0.452996027 |\n",
       "| 155104 | chrX:154836057-154836437 | MPP1     |  0.002853249 | 0.479367029 |\n",
       "| 155087 | chrX:154791986-154792298 | SMIM9    |  0.027897543 | 0.020213167 |\n",
       "| 155096 | chrX:154816475-154816702 | SMIM9    | -0.006730153 | 0.708034616 |\n",
       "| 155097 | chrX:154817284-154817570 | SMIM9    |  0.012246082 | 0.147693116 |\n",
       "| 155099 | chrX:154818880-154820481 | SMIM9    |  0.029118679 | 0.005153993 |\n",
       "| 155102 | chrX:154833474-154834469 | SMIM9    |  0.012812935 | 0.073982441 |\n",
       "| 155107 | chrX:154840613-154841413 | SMIM9    |  0.009056907 | 0.177356435 |\n",
       "| 155105 | chrX:154837474-154837679 | F8       |  0.007825327 | 0.345724664 |\n",
       "| 155108 | chrX:154910458-154911012 | F8       |  0.014368714 | 0.276274353 |\n",
       "| 155109 | chrX:154917842-154918632 | F8       |  0.037911760 | 0.018113805 |\n",
       "| 155125 | chrX:155070725-155072173 | MTCP1    |  0.019776559 | 0.374901485 |\n",
       "| 155137 | chrX:155215583-155217032 | RAB39B   |  0.013552040 | 0.100575640 |\n",
       "| 155143 | chrX:155246578-155247364 | RAB39B   |  0.024990334 | 0.006569273 |\n",
       "| 155145 | chrX:155313817-155314432 | CLIC2    |  0.018319026 | 0.093897379 |\n",
       "| 155146 | chrX:155333791-155335669 | CLIC2    |  0.034747011 | 0.010289553 |\n",
       "| 155172 | chrX:155966876-155967347 | WASIR1   | -0.007168906 | 0.830064525 |\n",
       "| 155173 | chrX:155968122-155968569 | WASIR1   |  0.030026099 | 0.014325011 |\n",
       "| 155176 | chrX:156000614-156000806 | WASIR1   |  0.012648940 | 0.217951905 |\n",
       "| 155177 | chrX:156000905-156002182 | WASIR1   |  0.020380207 | 0.117262101 |\n",
       "| 155178 | chrX:156002257-156003190 | WASIR1   |  0.018800919 | 0.119473828 |\n",
       "| 155181 | chrX:156016191-156016823 | WASIR1   |  0.016568172 | 0.211021802 |\n",
       "| 155175 | chrX:155997283-155998084 | DDX11L16 |  0.041291155 | 0.001763403 |\n",
       "| 155180 | chrX:156009722-156010137 | DDX11L16 | -0.001273811 | 0.493750653 |\n",
       "\n"
      ],
      "text/plain": [
       "       Peak   PeakRanges               Gene      rObs          pvalZ      \n",
       "1       1338  chr1:115493-115961       OR4F5      0.0157355382 0.111991475\n",
       "2      17878  chr1:804348-805153       FAM87B     0.0254473489 0.050114786\n",
       "3      17901  chr1:816701-817594       FAM87B     0.0266505827 0.028215881\n",
       "4      17808  chr1:778120-779401       LINC01128  0.0249139102 0.020607569\n",
       "5      17909  chr1:818917-819221       LINC01128  0.0096345124 0.337101803\n",
       "6      17973  chr1:842493-843467       LINC01128  0.0418846845 0.003623764\n",
       "7      18036  chr1:849559-850153       LINC01128 -0.0005044386 0.670700760\n",
       "8      18079  chr1:856218-856698       LINC01128  0.0005674452 0.602986234\n",
       "9      18092  chr1:860170-861174       LINC01128  0.0091232008 0.385078608\n",
       "10     18146  chr1:869578-870262       LINC01128  0.0151104067 0.204736225\n",
       "11     18234  chr1:885999-887141       SAMD11     0.0234631743 0.040165907\n",
       "12     18340  chr1:897236-897537       SAMD11     0.0015554781 0.580280522\n",
       "13     18345  chr1:897706-898129       SAMD11    -0.0029978186 0.722901208\n",
       "14     18411  chr1:904067-905237       SAMD11    -0.0047263047 0.794375315\n",
       "15     18420  chr1:906793-907288       SAMD11     0.0226430078 0.081565027\n",
       "16     18464  chr1:912712-913841       SAMD11     0.0133513044 0.326932441\n",
       "17     18775  chr1:939995-941085       SAMD11     0.0205339126 0.095705570\n",
       "18     18457  chr1:910941-911743       KLHL17     0.0231678176 0.286159532\n",
       "19     18504  chr1:916605-916851       KLHL17     0.0202676060 0.285377222\n",
       "20     18561  chr1:920895-921548       KLHL17     0.0264510507 0.366994956\n",
       "21     18570  chr1:923564-924643       KLHL17     0.0401515643 0.090042933\n",
       "22     18587  chr1:925567-925763       KLHL17     0.0210008988 0.203132826\n",
       "23     18654  chr1:930551-931222       KLHL17     0.0191831667 0.409348722\n",
       "24     18677  chr1:931921-932263       KLHL17     0.0266950216 0.162548323\n",
       "25     18748  chr1:938208-938358       KLHL17     0.0297271819 0.104676600\n",
       "26     18769  chr1:939194-939374       KLHL17     0.0437762830 0.017832559\n",
       "27     18923  chr1:982389-982818       KLHL17     0.0317919311 0.050629301\n",
       "28     18955  chr1:993306-993654       KLHL17     0.0152605845 0.325262283\n",
       "29       156  chr1:1012915-1014502     PLEKHN1    0.0227086367 0.078223118\n",
       "30     18649  chr1:930308-930467       PLEKHN1    0.0212739466 0.054493579\n",
       "⋮      ⋮      ⋮                        ⋮         ⋮             ⋮          \n",
       "111319 155093 chrX:154812460-154812819 MPP1       0.011999489  0.201638759\n",
       "111320 155094 chrX:154815322-154815631 MPP1       0.022366071  0.064356056\n",
       "111321 155095 chrX:154815707-154815857 MPP1       0.013773417  0.260166156\n",
       "111322 155098 chrX:154818257-154818559 MPP1       0.011741403  0.245858769\n",
       "111323 155100 chrX:154827231-154828278 MPP1      -0.004119386  0.726869174\n",
       "111324 155101 chrX:154829509-154830513 MPP1       0.011158519  0.280236603\n",
       "111325 155103 chrX:154834579-154834772 MPP1       0.002888720  0.452996027\n",
       "111326 155104 chrX:154836057-154836437 MPP1       0.002853249  0.479367029\n",
       "111327 155087 chrX:154791986-154792298 SMIM9      0.027897543  0.020213167\n",
       "111328 155096 chrX:154816475-154816702 SMIM9     -0.006730153  0.708034616\n",
       "111329 155097 chrX:154817284-154817570 SMIM9      0.012246082  0.147693116\n",
       "111330 155099 chrX:154818880-154820481 SMIM9      0.029118679  0.005153993\n",
       "111331 155102 chrX:154833474-154834469 SMIM9      0.012812935  0.073982441\n",
       "111332 155107 chrX:154840613-154841413 SMIM9      0.009056907  0.177356435\n",
       "111333 155105 chrX:154837474-154837679 F8         0.007825327  0.345724664\n",
       "111334 155108 chrX:154910458-154911012 F8         0.014368714  0.276274353\n",
       "111335 155109 chrX:154917842-154918632 F8         0.037911760  0.018113805\n",
       "111336 155125 chrX:155070725-155072173 MTCP1      0.019776559  0.374901485\n",
       "111337 155137 chrX:155215583-155217032 RAB39B     0.013552040  0.100575640\n",
       "111338 155143 chrX:155246578-155247364 RAB39B     0.024990334  0.006569273\n",
       "111339 155145 chrX:155313817-155314432 CLIC2      0.018319026  0.093897379\n",
       "111340 155146 chrX:155333791-155335669 CLIC2      0.034747011  0.010289553\n",
       "111341 155172 chrX:155966876-155967347 WASIR1    -0.007168906  0.830064525\n",
       "111342 155173 chrX:155968122-155968569 WASIR1     0.030026099  0.014325011\n",
       "111343 155176 chrX:156000614-156000806 WASIR1     0.012648940  0.217951905\n",
       "111344 155177 chrX:156000905-156002182 WASIR1     0.020380207  0.117262101\n",
       "111345 155178 chrX:156002257-156003190 WASIR1     0.018800919  0.119473828\n",
       "111346 155181 chrX:156016191-156016823 WASIR1     0.016568172  0.211021802\n",
       "111347 155175 chrX:155997283-155998084 DDX11L16   0.041291155  0.001763403\n",
       "111348 155180 chrX:156009722-156010137 DDX11L16  -0.001273811  0.493750653"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cisCor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bf338729-a01f-44ce-b8b1-ca732f4cdf84",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.peak.res = do.call(rbind,strsplit(gsub(\":\",\"-\",cisCor$PeakRanges),\"-\"))\n",
    "df.peak.res = as.data.frame(df.peak.res)\n",
    "colnames(df.peak.res) = c(\"chr\",\"start\",\"end\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "796d844e-0dd7-48ce-a85b-51c2b43ed8b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = makeGRangesFromDataFrame(df.peak.res)\n",
    "mcols(pairs.E2G.res) = cisCor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "60e71061-ebd4-42e9-8998-8618cf6dcd56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 111348 ranges and 8 metadata columns:\n",
       "           seqnames              ranges strand |      Peak\n",
       "              <Rle>           <IRanges>  <Rle> | <numeric>\n",
       "       [1]     chr1       115493-115961      * |      1338\n",
       "       [2]     chr1       804348-805153      * |     17878\n",
       "       [3]     chr1       816701-817594      * |     17901\n",
       "       [4]     chr1       778120-779401      * |     17808\n",
       "       [5]     chr1       818917-819221      * |     17909\n",
       "       ...      ...                 ...    ... .       ...\n",
       "  [111344]     chrX 156000905-156002182      * |    155177\n",
       "  [111345]     chrX 156002257-156003190      * |    155178\n",
       "  [111346]     chrX 156016191-156016823      * |    155181\n",
       "  [111347]     chrX 155997283-155998084      * |    155175\n",
       "  [111348]     chrX 156009722-156010137      * |    155180\n",
       "                       PeakRanges        Gene        rObs     pvalZ  TargetGene\n",
       "                      <character> <character>   <numeric> <numeric> <character>\n",
       "       [1]     chr1:115493-115961       OR4F5  0.01573554 0.1119915       OR4F5\n",
       "       [2]     chr1:804348-805153      FAM87B  0.02544735 0.0501148      FAM87B\n",
       "       [3]     chr1:816701-817594      FAM87B  0.02665058 0.0282159      FAM87B\n",
       "       [4]     chr1:778120-779401   LINC01128  0.02491391 0.0206076   LINC01128\n",
       "       [5]     chr1:818917-819221   LINC01128  0.00963451 0.3371018   LINC01128\n",
       "       ...                    ...         ...         ...       ...         ...\n",
       "  [111344] chrX:156000905-15600..      WASIR1  0.02038021 0.1172621      WASIR1\n",
       "  [111345] chrX:156002257-15600..      WASIR1  0.01880092 0.1194738      WASIR1\n",
       "  [111346] chrX:156016191-15601..      WASIR1  0.01656817 0.2110218      WASIR1\n",
       "  [111347] chrX:155997283-15599..    DDX11L16  0.04129116 0.0017634    DDX11L16\n",
       "  [111348] chrX:156009722-15601..    DDX11L16 -0.00127381 0.4937507    DDX11L16\n",
       "                         PeakName               PairName\n",
       "                      <character>            <character>\n",
       "       [1]     chr1-115493-115961 chr1-115493-115961_O..\n",
       "       [2]     chr1-804348-805153 chr1-804348-805153_F..\n",
       "       [3]     chr1-816701-817594 chr1-816701-817594_F..\n",
       "       [4]     chr1-778120-779401 chr1-778120-779401_L..\n",
       "       [5]     chr1-818917-819221 chr1-818917-819221_L..\n",
       "       ...                    ...                    ...\n",
       "  [111344] chrX-156000905-15600.. chrX-156000905-15600..\n",
       "  [111345] chrX-156002257-15600.. chrX-156002257-15600..\n",
       "  [111346] chrX-156016191-15601.. chrX-156016191-15601..\n",
       "  [111347] chrX-155997283-15599.. chrX-155997283-15599..\n",
       "  [111348] chrX-156009722-15601.. chrX-156009722-15601..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.res$TargetGene = pairs.E2G.res$Gene\n",
    "pairs.E2G.res$PeakName = gsub(\":\",\"-\",pairs.E2G.res$PeakRanges)\n",
    "pairs.E2G.res$PairName = paste(pairs.E2G.res$PeakName,\n",
    "                               pairs.E2G.res$TargetGene,\n",
    "                               sep = \"_\")\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8fa35fbc-5fea-48f2-8ea7-c85e5a5deb56",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(dir.output,recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 111348 ranges and 8 metadata columns:\n",
       "           seqnames              ranges strand |      Peak\n",
       "              <Rle>           <IRanges>  <Rle> | <numeric>\n",
       "       [1]     chr1       115493-115961      * |      1338\n",
       "       [2]     chr1       804348-805153      * |     17878\n",
       "       [3]     chr1       816701-817594      * |     17901\n",
       "       [4]     chr1       778120-779401      * |     17808\n",
       "       [5]     chr1       818917-819221      * |     17909\n",
       "       ...      ...                 ...    ... .       ...\n",
       "  [111344]     chrX 156000905-156002182      * |    155177\n",
       "  [111345]     chrX 156002257-156003190      * |    155178\n",
       "  [111346]     chrX 156016191-156016823      * |    155181\n",
       "  [111347]     chrX 155997283-155998084      * |    155175\n",
       "  [111348]     chrX 156009722-156010137      * |    155180\n",
       "                       PeakRanges        Gene        rObs     pvalZ  TargetGene\n",
       "                      <character> <character>   <numeric> <numeric> <character>\n",
       "       [1]     chr1:115493-115961       OR4F5  0.01573554 0.1119915       OR4F5\n",
       "       [2]     chr1:804348-805153      FAM87B  0.02544735 0.0501148      FAM87B\n",
       "       [3]     chr1:816701-817594      FAM87B  0.02665058 0.0282159      FAM87B\n",
       "       [4]     chr1:778120-779401   LINC01128  0.02491391 0.0206076   LINC01128\n",
       "       [5]     chr1:818917-819221   LINC01128  0.00963451 0.3371018   LINC01128\n",
       "       ...                    ...         ...         ...       ...         ...\n",
       "  [111344] chrX:156000905-15600..      WASIR1  0.02038021 0.1172621      WASIR1\n",
       "  [111345] chrX:156002257-15600..      WASIR1  0.01880092 0.1194738      WASIR1\n",
       "  [111346] chrX:156016191-15601..      WASIR1  0.01656817 0.2110218      WASIR1\n",
       "  [111347] chrX:155997283-15599..    DDX11L16  0.04129116 0.0017634    DDX11L16\n",
       "  [111348] chrX:156009722-15601..    DDX11L16 -0.00127381 0.4937507    DDX11L16\n",
       "                         PeakName               PairName\n",
       "                      <character>            <character>\n",
       "       [1]     chr1-115493-115961 chr1-115493-115961_O..\n",
       "       [2]     chr1-804348-805153 chr1-804348-805153_F..\n",
       "       [3]     chr1-816701-817594 chr1-816701-817594_F..\n",
       "       [4]     chr1-778120-779401 chr1-778120-779401_L..\n",
       "       [5]     chr1-818917-819221 chr1-818917-819221_L..\n",
       "       ...                    ...                    ...\n",
       "  [111344] chrX-156000905-15600.. chrX-156000905-15600..\n",
       "  [111345] chrX-156002257-15600.. chrX-156002257-15600..\n",
       "  [111346] chrX-156016191-15601.. chrX-156016191-15601..\n",
       "  [111347] chrX-155997283-15599.. chrX-155997283-15599..\n",
       "  [111348] chrX-156009722-15601.. chrX-156009722-15601..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 111348 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 115493</td><td> 115961</td><td>OR4F5    </td><td>K562</td><td> 0.0157355382</td><td>0.111991475</td></tr>\n",
       "\t<tr><td>chr1</td><td> 804348</td><td> 805153</td><td>FAM87B   </td><td>K562</td><td> 0.0254473489</td><td>0.050114786</td></tr>\n",
       "\t<tr><td>chr1</td><td> 816701</td><td> 817594</td><td>FAM87B   </td><td>K562</td><td> 0.0266505827</td><td>0.028215881</td></tr>\n",
       "\t<tr><td>chr1</td><td> 778120</td><td> 779401</td><td>LINC01128</td><td>K562</td><td> 0.0249139102</td><td>0.020607569</td></tr>\n",
       "\t<tr><td>chr1</td><td> 818917</td><td> 819221</td><td>LINC01128</td><td>K562</td><td> 0.0096345124</td><td>0.337101803</td></tr>\n",
       "\t<tr><td>chr1</td><td> 842493</td><td> 843467</td><td>LINC01128</td><td>K562</td><td> 0.0418846845</td><td>0.003623764</td></tr>\n",
       "\t<tr><td>chr1</td><td> 849559</td><td> 850153</td><td>LINC01128</td><td>K562</td><td>-0.0005044386</td><td>0.670700760</td></tr>\n",
       "\t<tr><td>chr1</td><td> 856218</td><td> 856698</td><td>LINC01128</td><td>K562</td><td> 0.0005674452</td><td>0.602986234</td></tr>\n",
       "\t<tr><td>chr1</td><td> 860170</td><td> 861174</td><td>LINC01128</td><td>K562</td><td> 0.0091232008</td><td>0.385078608</td></tr>\n",
       "\t<tr><td>chr1</td><td> 869578</td><td> 870262</td><td>LINC01128</td><td>K562</td><td> 0.0151104067</td><td>0.204736225</td></tr>\n",
       "\t<tr><td>chr1</td><td> 885999</td><td> 887141</td><td>SAMD11   </td><td>K562</td><td> 0.0234631743</td><td>0.040165907</td></tr>\n",
       "\t<tr><td>chr1</td><td> 897236</td><td> 897537</td><td>SAMD11   </td><td>K562</td><td> 0.0015554781</td><td>0.580280522</td></tr>\n",
       "\t<tr><td>chr1</td><td> 897706</td><td> 898129</td><td>SAMD11   </td><td>K562</td><td>-0.0029978186</td><td>0.722901208</td></tr>\n",
       "\t<tr><td>chr1</td><td> 904067</td><td> 905237</td><td>SAMD11   </td><td>K562</td><td>-0.0047263047</td><td>0.794375315</td></tr>\n",
       "\t<tr><td>chr1</td><td> 906793</td><td> 907288</td><td>SAMD11   </td><td>K562</td><td> 0.0226430078</td><td>0.081565027</td></tr>\n",
       "\t<tr><td>chr1</td><td> 912712</td><td> 913841</td><td>SAMD11   </td><td>K562</td><td> 0.0133513044</td><td>0.326932441</td></tr>\n",
       "\t<tr><td>chr1</td><td> 939995</td><td> 941085</td><td>SAMD11   </td><td>K562</td><td> 0.0205339126</td><td>0.095705570</td></tr>\n",
       "\t<tr><td>chr1</td><td> 910941</td><td> 911743</td><td>KLHL17   </td><td>K562</td><td> 0.0231678176</td><td>0.286159532</td></tr>\n",
       "\t<tr><td>chr1</td><td> 916605</td><td> 916851</td><td>KLHL17   </td><td>K562</td><td> 0.0202676060</td><td>0.285377222</td></tr>\n",
       "\t<tr><td>chr1</td><td> 920895</td><td> 921548</td><td>KLHL17   </td><td>K562</td><td> 0.0264510507</td><td>0.366994956</td></tr>\n",
       "\t<tr><td>chr1</td><td> 923564</td><td> 924643</td><td>KLHL17   </td><td>K562</td><td> 0.0401515643</td><td>0.090042933</td></tr>\n",
       "\t<tr><td>chr1</td><td> 925567</td><td> 925763</td><td>KLHL17   </td><td>K562</td><td> 0.0210008988</td><td>0.203132826</td></tr>\n",
       "\t<tr><td>chr1</td><td> 930551</td><td> 931222</td><td>KLHL17   </td><td>K562</td><td> 0.0191831667</td><td>0.409348722</td></tr>\n",
       "\t<tr><td>chr1</td><td> 931921</td><td> 932263</td><td>KLHL17   </td><td>K562</td><td> 0.0266950216</td><td>0.162548323</td></tr>\n",
       "\t<tr><td>chr1</td><td> 938208</td><td> 938358</td><td>KLHL17   </td><td>K562</td><td> 0.0297271819</td><td>0.104676600</td></tr>\n",
       "\t<tr><td>chr1</td><td> 939194</td><td> 939374</td><td>KLHL17   </td><td>K562</td><td> 0.0437762830</td><td>0.017832559</td></tr>\n",
       "\t<tr><td>chr1</td><td> 982389</td><td> 982818</td><td>KLHL17   </td><td>K562</td><td> 0.0317919311</td><td>0.050629301</td></tr>\n",
       "\t<tr><td>chr1</td><td> 993306</td><td> 993654</td><td>KLHL17   </td><td>K562</td><td> 0.0152605845</td><td>0.325262283</td></tr>\n",
       "\t<tr><td>chr1</td><td>1012915</td><td>1014502</td><td>PLEKHN1  </td><td>K562</td><td> 0.0227086367</td><td>0.078223118</td></tr>\n",
       "\t<tr><td>chr1</td><td> 930308</td><td> 930467</td><td>PLEKHN1  </td><td>K562</td><td> 0.0212739466</td><td>0.054493579</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>154812460</td><td>154812819</td><td>MPP1    </td><td>K562</td><td> 0.011999489</td><td>0.201638759</td></tr>\n",
       "\t<tr><td>chrX</td><td>154815322</td><td>154815631</td><td>MPP1    </td><td>K562</td><td> 0.022366071</td><td>0.064356056</td></tr>\n",
       "\t<tr><td>chrX</td><td>154815707</td><td>154815857</td><td>MPP1    </td><td>K562</td><td> 0.013773417</td><td>0.260166156</td></tr>\n",
       "\t<tr><td>chrX</td><td>154818257</td><td>154818559</td><td>MPP1    </td><td>K562</td><td> 0.011741403</td><td>0.245858769</td></tr>\n",
       "\t<tr><td>chrX</td><td>154827231</td><td>154828278</td><td>MPP1    </td><td>K562</td><td>-0.004119386</td><td>0.726869174</td></tr>\n",
       "\t<tr><td>chrX</td><td>154829509</td><td>154830513</td><td>MPP1    </td><td>K562</td><td> 0.011158519</td><td>0.280236603</td></tr>\n",
       "\t<tr><td>chrX</td><td>154834579</td><td>154834772</td><td>MPP1    </td><td>K562</td><td> 0.002888720</td><td>0.452996027</td></tr>\n",
       "\t<tr><td>chrX</td><td>154836057</td><td>154836437</td><td>MPP1    </td><td>K562</td><td> 0.002853249</td><td>0.479367029</td></tr>\n",
       "\t<tr><td>chrX</td><td>154791986</td><td>154792298</td><td>SMIM9   </td><td>K562</td><td> 0.027897543</td><td>0.020213167</td></tr>\n",
       "\t<tr><td>chrX</td><td>154816475</td><td>154816702</td><td>SMIM9   </td><td>K562</td><td>-0.006730153</td><td>0.708034616</td></tr>\n",
       "\t<tr><td>chrX</td><td>154817284</td><td>154817570</td><td>SMIM9   </td><td>K562</td><td> 0.012246082</td><td>0.147693116</td></tr>\n",
       "\t<tr><td>chrX</td><td>154818880</td><td>154820481</td><td>SMIM9   </td><td>K562</td><td> 0.029118679</td><td>0.005153993</td></tr>\n",
       "\t<tr><td>chrX</td><td>154833474</td><td>154834469</td><td>SMIM9   </td><td>K562</td><td> 0.012812935</td><td>0.073982441</td></tr>\n",
       "\t<tr><td>chrX</td><td>154840613</td><td>154841413</td><td>SMIM9   </td><td>K562</td><td> 0.009056907</td><td>0.177356435</td></tr>\n",
       "\t<tr><td>chrX</td><td>154837474</td><td>154837679</td><td>F8      </td><td>K562</td><td> 0.007825327</td><td>0.345724664</td></tr>\n",
       "\t<tr><td>chrX</td><td>154910458</td><td>154911012</td><td>F8      </td><td>K562</td><td> 0.014368714</td><td>0.276274353</td></tr>\n",
       "\t<tr><td>chrX</td><td>154917842</td><td>154918632</td><td>F8      </td><td>K562</td><td> 0.037911760</td><td>0.018113805</td></tr>\n",
       "\t<tr><td>chrX</td><td>155070725</td><td>155072173</td><td>MTCP1   </td><td>K562</td><td> 0.019776559</td><td>0.374901485</td></tr>\n",
       "\t<tr><td>chrX</td><td>155215583</td><td>155217032</td><td>RAB39B  </td><td>K562</td><td> 0.013552040</td><td>0.100575640</td></tr>\n",
       "\t<tr><td>chrX</td><td>155246578</td><td>155247364</td><td>RAB39B  </td><td>K562</td><td> 0.024990334</td><td>0.006569273</td></tr>\n",
       "\t<tr><td>chrX</td><td>155313817</td><td>155314432</td><td>CLIC2   </td><td>K562</td><td> 0.018319026</td><td>0.093897379</td></tr>\n",
       "\t<tr><td>chrX</td><td>155333791</td><td>155335669</td><td>CLIC2   </td><td>K562</td><td> 0.034747011</td><td>0.010289553</td></tr>\n",
       "\t<tr><td>chrX</td><td>155966876</td><td>155967347</td><td>WASIR1  </td><td>K562</td><td>-0.007168906</td><td>0.830064525</td></tr>\n",
       "\t<tr><td>chrX</td><td>155968122</td><td>155968569</td><td>WASIR1  </td><td>K562</td><td> 0.030026099</td><td>0.014325011</td></tr>\n",
       "\t<tr><td>chrX</td><td>156000614</td><td>156000806</td><td>WASIR1  </td><td>K562</td><td> 0.012648940</td><td>0.217951905</td></tr>\n",
       "\t<tr><td>chrX</td><td>156000905</td><td>156002182</td><td>WASIR1  </td><td>K562</td><td> 0.020380207</td><td>0.117262101</td></tr>\n",
       "\t<tr><td>chrX</td><td>156002257</td><td>156003190</td><td>WASIR1  </td><td>K562</td><td> 0.018800919</td><td>0.119473828</td></tr>\n",
       "\t<tr><td>chrX</td><td>156016191</td><td>156016823</td><td>WASIR1  </td><td>K562</td><td> 0.016568172</td><td>0.211021802</td></tr>\n",
       "\t<tr><td>chrX</td><td>155997283</td><td>155998084</td><td>DDX11L16</td><td>K562</td><td> 0.041291155</td><td>0.001763403</td></tr>\n",
       "\t<tr><td>chrX</td><td>156009722</td><td>156010137</td><td>DDX11L16</td><td>K562</td><td>-0.001273811</td><td>0.493750653</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 111348 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  115493 &  115961 & OR4F5     & K562 &  0.0157355382 & 0.111991475\\\\\n",
       "\t chr1 &  804348 &  805153 & FAM87B    & K562 &  0.0254473489 & 0.050114786\\\\\n",
       "\t chr1 &  816701 &  817594 & FAM87B    & K562 &  0.0266505827 & 0.028215881\\\\\n",
       "\t chr1 &  778120 &  779401 & LINC01128 & K562 &  0.0249139102 & 0.020607569\\\\\n",
       "\t chr1 &  818917 &  819221 & LINC01128 & K562 &  0.0096345124 & 0.337101803\\\\\n",
       "\t chr1 &  842493 &  843467 & LINC01128 & K562 &  0.0418846845 & 0.003623764\\\\\n",
       "\t chr1 &  849559 &  850153 & LINC01128 & K562 & -0.0005044386 & 0.670700760\\\\\n",
       "\t chr1 &  856218 &  856698 & LINC01128 & K562 &  0.0005674452 & 0.602986234\\\\\n",
       "\t chr1 &  860170 &  861174 & LINC01128 & K562 &  0.0091232008 & 0.385078608\\\\\n",
       "\t chr1 &  869578 &  870262 & LINC01128 & K562 &  0.0151104067 & 0.204736225\\\\\n",
       "\t chr1 &  885999 &  887141 & SAMD11    & K562 &  0.0234631743 & 0.040165907\\\\\n",
       "\t chr1 &  897236 &  897537 & SAMD11    & K562 &  0.0015554781 & 0.580280522\\\\\n",
       "\t chr1 &  897706 &  898129 & SAMD11    & K562 & -0.0029978186 & 0.722901208\\\\\n",
       "\t chr1 &  904067 &  905237 & SAMD11    & K562 & -0.0047263047 & 0.794375315\\\\\n",
       "\t chr1 &  906793 &  907288 & SAMD11    & K562 &  0.0226430078 & 0.081565027\\\\\n",
       "\t chr1 &  912712 &  913841 & SAMD11    & K562 &  0.0133513044 & 0.326932441\\\\\n",
       "\t chr1 &  939995 &  941085 & SAMD11    & K562 &  0.0205339126 & 0.095705570\\\\\n",
       "\t chr1 &  910941 &  911743 & KLHL17    & K562 &  0.0231678176 & 0.286159532\\\\\n",
       "\t chr1 &  916605 &  916851 & KLHL17    & K562 &  0.0202676060 & 0.285377222\\\\\n",
       "\t chr1 &  920895 &  921548 & KLHL17    & K562 &  0.0264510507 & 0.366994956\\\\\n",
       "\t chr1 &  923564 &  924643 & KLHL17    & K562 &  0.0401515643 & 0.090042933\\\\\n",
       "\t chr1 &  925567 &  925763 & KLHL17    & K562 &  0.0210008988 & 0.203132826\\\\\n",
       "\t chr1 &  930551 &  931222 & KLHL17    & K562 &  0.0191831667 & 0.409348722\\\\\n",
       "\t chr1 &  931921 &  932263 & KLHL17    & K562 &  0.0266950216 & 0.162548323\\\\\n",
       "\t chr1 &  938208 &  938358 & KLHL17    & K562 &  0.0297271819 & 0.104676600\\\\\n",
       "\t chr1 &  939194 &  939374 & KLHL17    & K562 &  0.0437762830 & 0.017832559\\\\\n",
       "\t chr1 &  982389 &  982818 & KLHL17    & K562 &  0.0317919311 & 0.050629301\\\\\n",
       "\t chr1 &  993306 &  993654 & KLHL17    & K562 &  0.0152605845 & 0.325262283\\\\\n",
       "\t chr1 & 1012915 & 1014502 & PLEKHN1   & K562 &  0.0227086367 & 0.078223118\\\\\n",
       "\t chr1 &  930308 &  930467 & PLEKHN1   & K562 &  0.0212739466 & 0.054493579\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 154812460 & 154812819 & MPP1     & K562 &  0.011999489 & 0.201638759\\\\\n",
       "\t chrX & 154815322 & 154815631 & MPP1     & K562 &  0.022366071 & 0.064356056\\\\\n",
       "\t chrX & 154815707 & 154815857 & MPP1     & K562 &  0.013773417 & 0.260166156\\\\\n",
       "\t chrX & 154818257 & 154818559 & MPP1     & K562 &  0.011741403 & 0.245858769\\\\\n",
       "\t chrX & 154827231 & 154828278 & MPP1     & K562 & -0.004119386 & 0.726869174\\\\\n",
       "\t chrX & 154829509 & 154830513 & MPP1     & K562 &  0.011158519 & 0.280236603\\\\\n",
       "\t chrX & 154834579 & 154834772 & MPP1     & K562 &  0.002888720 & 0.452996027\\\\\n",
       "\t chrX & 154836057 & 154836437 & MPP1     & K562 &  0.002853249 & 0.479367029\\\\\n",
       "\t chrX & 154791986 & 154792298 & SMIM9    & K562 &  0.027897543 & 0.020213167\\\\\n",
       "\t chrX & 154816475 & 154816702 & SMIM9    & K562 & -0.006730153 & 0.708034616\\\\\n",
       "\t chrX & 154817284 & 154817570 & SMIM9    & K562 &  0.012246082 & 0.147693116\\\\\n",
       "\t chrX & 154818880 & 154820481 & SMIM9    & K562 &  0.029118679 & 0.005153993\\\\\n",
       "\t chrX & 154833474 & 154834469 & SMIM9    & K562 &  0.012812935 & 0.073982441\\\\\n",
       "\t chrX & 154840613 & 154841413 & SMIM9    & K562 &  0.009056907 & 0.177356435\\\\\n",
       "\t chrX & 154837474 & 154837679 & F8       & K562 &  0.007825327 & 0.345724664\\\\\n",
       "\t chrX & 154910458 & 154911012 & F8       & K562 &  0.014368714 & 0.276274353\\\\\n",
       "\t chrX & 154917842 & 154918632 & F8       & K562 &  0.037911760 & 0.018113805\\\\\n",
       "\t chrX & 155070725 & 155072173 & MTCP1    & K562 &  0.019776559 & 0.374901485\\\\\n",
       "\t chrX & 155215583 & 155217032 & RAB39B   & K562 &  0.013552040 & 0.100575640\\\\\n",
       "\t chrX & 155246578 & 155247364 & RAB39B   & K562 &  0.024990334 & 0.006569273\\\\\n",
       "\t chrX & 155313817 & 155314432 & CLIC2    & K562 &  0.018319026 & 0.093897379\\\\\n",
       "\t chrX & 155333791 & 155335669 & CLIC2    & K562 &  0.034747011 & 0.010289553\\\\\n",
       "\t chrX & 155966876 & 155967347 & WASIR1   & K562 & -0.007168906 & 0.830064525\\\\\n",
       "\t chrX & 155968122 & 155968569 & WASIR1   & K562 &  0.030026099 & 0.014325011\\\\\n",
       "\t chrX & 156000614 & 156000806 & WASIR1   & K562 &  0.012648940 & 0.217951905\\\\\n",
       "\t chrX & 156000905 & 156002182 & WASIR1   & K562 &  0.020380207 & 0.117262101\\\\\n",
       "\t chrX & 156002257 & 156003190 & WASIR1   & K562 &  0.018800919 & 0.119473828\\\\\n",
       "\t chrX & 156016191 & 156016823 & WASIR1   & K562 &  0.016568172 & 0.211021802\\\\\n",
       "\t chrX & 155997283 & 155998084 & DDX11L16 & K562 &  0.041291155 & 0.001763403\\\\\n",
       "\t chrX & 156009722 & 156010137 & DDX11L16 & K562 & -0.001273811 & 0.493750653\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 111348 × 7\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| chr1 |  115493 |  115961 | OR4F5     | K562 |  0.0157355382 | 0.111991475 |\n",
       "| chr1 |  804348 |  805153 | FAM87B    | K562 |  0.0254473489 | 0.050114786 |\n",
       "| chr1 |  816701 |  817594 | FAM87B    | K562 |  0.0266505827 | 0.028215881 |\n",
       "| chr1 |  778120 |  779401 | LINC01128 | K562 |  0.0249139102 | 0.020607569 |\n",
       "| chr1 |  818917 |  819221 | LINC01128 | K562 |  0.0096345124 | 0.337101803 |\n",
       "| chr1 |  842493 |  843467 | LINC01128 | K562 |  0.0418846845 | 0.003623764 |\n",
       "| chr1 |  849559 |  850153 | LINC01128 | K562 | -0.0005044386 | 0.670700760 |\n",
       "| chr1 |  856218 |  856698 | LINC01128 | K562 |  0.0005674452 | 0.602986234 |\n",
       "| chr1 |  860170 |  861174 | LINC01128 | K562 |  0.0091232008 | 0.385078608 |\n",
       "| chr1 |  869578 |  870262 | LINC01128 | K562 |  0.0151104067 | 0.204736225 |\n",
       "| chr1 |  885999 |  887141 | SAMD11    | K562 |  0.0234631743 | 0.040165907 |\n",
       "| chr1 |  897236 |  897537 | SAMD11    | K562 |  0.0015554781 | 0.580280522 |\n",
       "| chr1 |  897706 |  898129 | SAMD11    | K562 | -0.0029978186 | 0.722901208 |\n",
       "| chr1 |  904067 |  905237 | SAMD11    | K562 | -0.0047263047 | 0.794375315 |\n",
       "| chr1 |  906793 |  907288 | SAMD11    | K562 |  0.0226430078 | 0.081565027 |\n",
       "| chr1 |  912712 |  913841 | SAMD11    | K562 |  0.0133513044 | 0.326932441 |\n",
       "| chr1 |  939995 |  941085 | SAMD11    | K562 |  0.0205339126 | 0.095705570 |\n",
       "| chr1 |  910941 |  911743 | KLHL17    | K562 |  0.0231678176 | 0.286159532 |\n",
       "| chr1 |  916605 |  916851 | KLHL17    | K562 |  0.0202676060 | 0.285377222 |\n",
       "| chr1 |  920895 |  921548 | KLHL17    | K562 |  0.0264510507 | 0.366994956 |\n",
       "| chr1 |  923564 |  924643 | KLHL17    | K562 |  0.0401515643 | 0.090042933 |\n",
       "| chr1 |  925567 |  925763 | KLHL17    | K562 |  0.0210008988 | 0.203132826 |\n",
       "| chr1 |  930551 |  931222 | KLHL17    | K562 |  0.0191831667 | 0.409348722 |\n",
       "| chr1 |  931921 |  932263 | KLHL17    | K562 |  0.0266950216 | 0.162548323 |\n",
       "| chr1 |  938208 |  938358 | KLHL17    | K562 |  0.0297271819 | 0.104676600 |\n",
       "| chr1 |  939194 |  939374 | KLHL17    | K562 |  0.0437762830 | 0.017832559 |\n",
       "| chr1 |  982389 |  982818 | KLHL17    | K562 |  0.0317919311 | 0.050629301 |\n",
       "| chr1 |  993306 |  993654 | KLHL17    | K562 |  0.0152605845 | 0.325262283 |\n",
       "| chr1 | 1012915 | 1014502 | PLEKHN1   | K562 |  0.0227086367 | 0.078223118 |\n",
       "| chr1 |  930308 |  930467 | PLEKHN1   | K562 |  0.0212739466 | 0.054493579 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 154812460 | 154812819 | MPP1     | K562 |  0.011999489 | 0.201638759 |\n",
       "| chrX | 154815322 | 154815631 | MPP1     | K562 |  0.022366071 | 0.064356056 |\n",
       "| chrX | 154815707 | 154815857 | MPP1     | K562 |  0.013773417 | 0.260166156 |\n",
       "| chrX | 154818257 | 154818559 | MPP1     | K562 |  0.011741403 | 0.245858769 |\n",
       "| chrX | 154827231 | 154828278 | MPP1     | K562 | -0.004119386 | 0.726869174 |\n",
       "| chrX | 154829509 | 154830513 | MPP1     | K562 |  0.011158519 | 0.280236603 |\n",
       "| chrX | 154834579 | 154834772 | MPP1     | K562 |  0.002888720 | 0.452996027 |\n",
       "| chrX | 154836057 | 154836437 | MPP1     | K562 |  0.002853249 | 0.479367029 |\n",
       "| chrX | 154791986 | 154792298 | SMIM9    | K562 |  0.027897543 | 0.020213167 |\n",
       "| chrX | 154816475 | 154816702 | SMIM9    | K562 | -0.006730153 | 0.708034616 |\n",
       "| chrX | 154817284 | 154817570 | SMIM9    | K562 |  0.012246082 | 0.147693116 |\n",
       "| chrX | 154818880 | 154820481 | SMIM9    | K562 |  0.029118679 | 0.005153993 |\n",
       "| chrX | 154833474 | 154834469 | SMIM9    | K562 |  0.012812935 | 0.073982441 |\n",
       "| chrX | 154840613 | 154841413 | SMIM9    | K562 |  0.009056907 | 0.177356435 |\n",
       "| chrX | 154837474 | 154837679 | F8       | K562 |  0.007825327 | 0.345724664 |\n",
       "| chrX | 154910458 | 154911012 | F8       | K562 |  0.014368714 | 0.276274353 |\n",
       "| chrX | 154917842 | 154918632 | F8       | K562 |  0.037911760 | 0.018113805 |\n",
       "| chrX | 155070725 | 155072173 | MTCP1    | K562 |  0.019776559 | 0.374901485 |\n",
       "| chrX | 155215583 | 155217032 | RAB39B   | K562 |  0.013552040 | 0.100575640 |\n",
       "| chrX | 155246578 | 155247364 | RAB39B   | K562 |  0.024990334 | 0.006569273 |\n",
       "| chrX | 155313817 | 155314432 | CLIC2    | K562 |  0.018319026 | 0.093897379 |\n",
       "| chrX | 155333791 | 155335669 | CLIC2    | K562 |  0.034747011 | 0.010289553 |\n",
       "| chrX | 155966876 | 155967347 | WASIR1   | K562 | -0.007168906 | 0.830064525 |\n",
       "| chrX | 155968122 | 155968569 | WASIR1   | K562 |  0.030026099 | 0.014325011 |\n",
       "| chrX | 156000614 | 156000806 | WASIR1   | K562 |  0.012648940 | 0.217951905 |\n",
       "| chrX | 156000905 | 156002182 | WASIR1   | K562 |  0.020380207 | 0.117262101 |\n",
       "| chrX | 156002257 | 156003190 | WASIR1   | K562 |  0.018800919 | 0.119473828 |\n",
       "| chrX | 156016191 | 156016823 | WASIR1   | K562 |  0.016568172 | 0.211021802 |\n",
       "| chrX | 155997283 | 155998084 | DDX11L16 | K562 |  0.041291155 | 0.001763403 |\n",
       "| chrX | 156009722 | 156010137 | DDX11L16 | K562 | -0.001273811 | 0.493750653 |\n",
       "\n"
      ],
      "text/plain": [
       "       chr  start     end       TargetGene CellType rObs          pvalZ      \n",
       "1      chr1  115493    115961   OR4F5      K562      0.0157355382 0.111991475\n",
       "2      chr1  804348    805153   FAM87B     K562      0.0254473489 0.050114786\n",
       "3      chr1  816701    817594   FAM87B     K562      0.0266505827 0.028215881\n",
       "4      chr1  778120    779401   LINC01128  K562      0.0249139102 0.020607569\n",
       "5      chr1  818917    819221   LINC01128  K562      0.0096345124 0.337101803\n",
       "6      chr1  842493    843467   LINC01128  K562      0.0418846845 0.003623764\n",
       "7      chr1  849559    850153   LINC01128  K562     -0.0005044386 0.670700760\n",
       "8      chr1  856218    856698   LINC01128  K562      0.0005674452 0.602986234\n",
       "9      chr1  860170    861174   LINC01128  K562      0.0091232008 0.385078608\n",
       "10     chr1  869578    870262   LINC01128  K562      0.0151104067 0.204736225\n",
       "11     chr1  885999    887141   SAMD11     K562      0.0234631743 0.040165907\n",
       "12     chr1  897236    897537   SAMD11     K562      0.0015554781 0.580280522\n",
       "13     chr1  897706    898129   SAMD11     K562     -0.0029978186 0.722901208\n",
       "14     chr1  904067    905237   SAMD11     K562     -0.0047263047 0.794375315\n",
       "15     chr1  906793    907288   SAMD11     K562      0.0226430078 0.081565027\n",
       "16     chr1  912712    913841   SAMD11     K562      0.0133513044 0.326932441\n",
       "17     chr1  939995    941085   SAMD11     K562      0.0205339126 0.095705570\n",
       "18     chr1  910941    911743   KLHL17     K562      0.0231678176 0.286159532\n",
       "19     chr1  916605    916851   KLHL17     K562      0.0202676060 0.285377222\n",
       "20     chr1  920895    921548   KLHL17     K562      0.0264510507 0.366994956\n",
       "21     chr1  923564    924643   KLHL17     K562      0.0401515643 0.090042933\n",
       "22     chr1  925567    925763   KLHL17     K562      0.0210008988 0.203132826\n",
       "23     chr1  930551    931222   KLHL17     K562      0.0191831667 0.409348722\n",
       "24     chr1  931921    932263   KLHL17     K562      0.0266950216 0.162548323\n",
       "25     chr1  938208    938358   KLHL17     K562      0.0297271819 0.104676600\n",
       "26     chr1  939194    939374   KLHL17     K562      0.0437762830 0.017832559\n",
       "27     chr1  982389    982818   KLHL17     K562      0.0317919311 0.050629301\n",
       "28     chr1  993306    993654   KLHL17     K562      0.0152605845 0.325262283\n",
       "29     chr1 1012915   1014502   PLEKHN1    K562      0.0227086367 0.078223118\n",
       "30     chr1  930308    930467   PLEKHN1    K562      0.0212739466 0.054493579\n",
       "⋮      ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮          \n",
       "111319 chrX 154812460 154812819 MPP1       K562      0.011999489  0.201638759\n",
       "111320 chrX 154815322 154815631 MPP1       K562      0.022366071  0.064356056\n",
       "111321 chrX 154815707 154815857 MPP1       K562      0.013773417  0.260166156\n",
       "111322 chrX 154818257 154818559 MPP1       K562      0.011741403  0.245858769\n",
       "111323 chrX 154827231 154828278 MPP1       K562     -0.004119386  0.726869174\n",
       "111324 chrX 154829509 154830513 MPP1       K562      0.011158519  0.280236603\n",
       "111325 chrX 154834579 154834772 MPP1       K562      0.002888720  0.452996027\n",
       "111326 chrX 154836057 154836437 MPP1       K562      0.002853249  0.479367029\n",
       "111327 chrX 154791986 154792298 SMIM9      K562      0.027897543  0.020213167\n",
       "111328 chrX 154816475 154816702 SMIM9      K562     -0.006730153  0.708034616\n",
       "111329 chrX 154817284 154817570 SMIM9      K562      0.012246082  0.147693116\n",
       "111330 chrX 154818880 154820481 SMIM9      K562      0.029118679  0.005153993\n",
       "111331 chrX 154833474 154834469 SMIM9      K562      0.012812935  0.073982441\n",
       "111332 chrX 154840613 154841413 SMIM9      K562      0.009056907  0.177356435\n",
       "111333 chrX 154837474 154837679 F8         K562      0.007825327  0.345724664\n",
       "111334 chrX 154910458 154911012 F8         K562      0.014368714  0.276274353\n",
       "111335 chrX 154917842 154918632 F8         K562      0.037911760  0.018113805\n",
       "111336 chrX 155070725 155072173 MTCP1      K562      0.019776559  0.374901485\n",
       "111337 chrX 155215583 155217032 RAB39B     K562      0.013552040  0.100575640\n",
       "111338 chrX 155246578 155247364 RAB39B     K562      0.024990334  0.006569273\n",
       "111339 chrX 155313817 155314432 CLIC2      K562      0.018319026  0.093897379\n",
       "111340 chrX 155333791 155335669 CLIC2      K562      0.034747011  0.010289553\n",
       "111341 chrX 155966876 155967347 WASIR1     K562     -0.007168906  0.830064525\n",
       "111342 chrX 155968122 155968569 WASIR1     K562      0.030026099  0.014325011\n",
       "111343 chrX 156000614 156000806 WASIR1     K562      0.012648940  0.217951905\n",
       "111344 chrX 156000905 156002182 WASIR1     K562      0.020380207  0.117262101\n",
       "111345 chrX 156002257 156003190 WASIR1     K562      0.018800919  0.119473828\n",
       "111346 chrX 156016191 156016823 WASIR1     K562      0.016568172  0.211021802\n",
       "111347 chrX 155997283 155998084 DDX11L16   K562      0.041291155  0.001763403\n",
       "111348 chrX 156009722 156010137 DDX11L16   K562     -0.001273811  0.493750653"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.9 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   doParallel_1.0.17                \n",
       " [3] iterators_1.0.14                  foreach_1.5.2                    \n",
       " [5] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.68.0                  \n",
       " [7] rtracklayer_1.60.1                Biostrings_2.70.1                \n",
       " [9] XVector_0.42.0                    genomation_1.34.0                \n",
       "[11] Seurat_5.0.3                      SeuratObject_5.0.1               \n",
       "[13] sp_2.1-4                          FigR_0.1.0                       \n",
       "[15] motifmatchr_1.24.0                chromVAR_1.24.0                  \n",
       "[17] cowplot_1.1.3                     ggplot2_3.5.1                    \n",
       "[19] dplyr_1.1.4                       SummarizedExperiment_1.30.2      \n",
       "[21] Biobase_2.60.0                    GenomicRanges_1.54.1             \n",
       "[23] GenomeInfoDb_1.38.1               IRanges_2.36.0                   \n",
       "[25] S4Vectors_0.40.2                  BiocGenerics_0.48.1              \n",
       "[27] MatrixGenerics_1.12.3             matrixStats_1.3.0                \n",
       "[29] Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             globals_0.16.3             \n",
       " [15] lattice_0.22-6              MASS_7.3-60                \n",
       " [17] magrittr_2.0.3              plotly_4.10.4              \n",
       " [19] plotrix_3.8-4               yaml_2.3.8                 \n",
       " [21] httpuv_1.6.15               sctransform_0.4.1          \n",
       " [23] spam_2.10-0                 spatstat.sparse_3.0-3      \n",
       " [25] reticulate_1.36.1           pbapply_1.7-2              \n",
       " [27] DBI_1.2.2                   CNEr_1.38.0                \n",
       " [29] RColorBrewer_1.1-3          abind_1.4-5                \n",
       " [31] zlibbioc_1.48.0             Rtsne_0.17                 \n",
       " [33] purrr_1.0.2                 R.utils_2.12.3             \n",
       " [35] RCurl_1.98-1.14             pracma_2.4.4               \n",
       " [37] GenomeInfoDbData_1.2.11     ggrepel_0.9.5              \n",
       " [39] irlba_2.3.5.1               spatstat.utils_3.0-4       \n",
       " [41] listenv_0.9.1               seqLogo_1.68.0             \n",
       " [43] goftest_1.2-3               RSpectra_0.16-1            \n",
       " [45] spatstat.random_3.2-3       annotate_1.80.0            \n",
       " [47] fitdistrplus_1.1-11         parallelly_1.37.1          \n",
       " [49] leiden_0.4.3.1              codetools_0.2-20           \n",
       " [51] DelayedArray_0.26.7         DT_0.33                    \n",
       " [53] tidyselect_1.2.1            spatstat.explore_3.2-6     \n",
       " [55] base64enc_0.1-3             GenomicAlignments_1.36.0   \n",
       " [57] jsonlite_1.8.8              progressr_0.14.0           \n",
       " [59] ggridges_0.5.6              survival_3.6-4             \n",
       " [61] tools_4.3.3                 TFMPvalue_0.0.9            \n",
       " [63] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [65] glue_1.7.0                  gridExtra_2.3              \n",
       " [67] IRdisplay_1.1               withr_3.0.0                \n",
       " [69] fastmap_1.1.1               fansi_1.0.6                \n",
       " [71] caTools_1.18.2              digest_0.6.35              \n",
       " [73] R6_2.5.1                    mime_0.12                  \n",
       " [75] nabor_0.5.0                 seqPattern_1.34.0          \n",
       " [77] colorspace_2.1-0            scattermore_1.2            \n",
       " [79] GO.db_3.18.0                tensor_1.5                 \n",
       " [81] gtools_3.9.5                poweRlaw_0.80.0            \n",
       " [83] spatstat.data_3.0-4         RSQLite_2.3.6              \n",
       " [85] R.methodsS3_1.8.2           utf8_1.2.4                 \n",
       " [87] tidyr_1.3.1                 generics_0.1.3             \n",
       " [89] data.table_1.15.2           httr_1.4.7                 \n",
       " [91] htmlwidgets_1.6.4           S4Arrays_1.2.1             \n",
       " [93] TFBSTools_1.40.0            uwot_0.1.16                \n",
       " [95] pkgconfig_2.0.3             gtable_0.3.5               \n",
       " [97] blob_1.2.4                  impute_1.76.0              \n",
       " [99] lmtest_0.9-40               htmltools_0.5.8.1          \n",
       "[101] dotCall64_1.1-1             scales_1.3.0               \n",
       "[103] png_0.1-8                   tzdb_0.4.0                 \n",
       "[105] reshape2_1.4.4              rjson_0.2.21               \n",
       "[107] uuid_1.2-0                  nlme_3.1-164               \n",
       "[109] repr_1.1.7                  cachem_1.0.8               \n",
       "[111] zoo_1.8-12                  stringr_1.5.1              \n",
       "[113] KernSmooth_2.23-22          miniUI_0.1.1.1             \n",
       "[115] AnnotationDbi_1.64.1        restfulr_0.0.15            \n",
       "[117] pillar_1.9.0                vctrs_0.6.5                \n",
       "[119] RANN_2.6.1                  promises_1.3.0             \n",
       "[121] xtable_1.8-4                cluster_2.1.6              \n",
       "[123] evaluate_0.23               readr_2.1.5                \n",
       "[125] cli_3.6.2                   compiler_4.3.3             \n",
       "[127] Rsamtools_2.18.0            rlang_1.1.3                \n",
       "[129] crayon_1.5.2                future.apply_1.11.2        \n",
       "[131] plyr_1.8.9                  stringi_1.8.3              \n",
       "[133] gridBase_0.4-7              deldir_2.0-4               \n",
       "[135] viridisLite_0.4.2           BiocParallel_1.36.0        \n",
       "[137] munsell_0.5.1               lazyeval_0.2.2             \n",
       "[139] spatstat.geom_3.2-9         IRkernel_1.3.2             \n",
       "[141] RcppHNSW_0.6.0              hms_1.1.3                  \n",
       "[143] patchwork_1.2.0             bit64_4.0.5                \n",
       "[145] future_1.33.2               KEGGREST_1.40.1            \n",
       "[147] shiny_1.8.1.1               ROCR_1.0-11                \n",
       "[149] igraph_2.0.3                memoise_2.0.1              \n",
       "[151] bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
