{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dc89309f-ebd0-4f9c-a1f1-5d354337fd99",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Matrix'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(Matrix)\n",
    "library(parallel)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9817830-01bd-4fe7-bee7-45b48c6d680a",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "83e2e91c-dce2-468e-826e-26a7fcd2e2e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Xu/K562/Kendall/atac_matrix.csv.gz\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/1.prepare_data/1.seurat_pipeline.240507/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/3.Genome_wide_prediction/Signac.default/Signac.250210/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b37042e-23d6-4cd8-8110-9b6a0a1b51e8",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0b752808-a40f-40fe-b92a-5c404763fe8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.atac = read.csv(path.matrix.atac_count,\n",
    "                       row.names = 1,\n",
    "                       check.names = F)\n",
    "matrix.atac = Matrix(as.matrix(matrix.atac), sparse = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deda5367-2072-4a08-8c2f-b7b1077c8964",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8c3df281-40e5-4514-802e-0cb9c4178dab",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3b3c75a-aea2-47b1-baa7-de3f1885b17b",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0d3df0d1-aa41-479d-9b85-fc2ab82e8da6",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4cb63b91-8816-4341-997b-5e6c2d77bff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ensembldb':\n",
      "\n",
      "    filter, select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Biostrings':\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:XVector':\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'data.table'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna_count.rename = map_gene_names(matrix.rna_count,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8fed8ea-0592-4f54-b8ab-968c851c7364",
   "metadata": {},
   "source": [
    "Create Seurat object, follow https://stuartlab.org/signac/articles/pbmc_multiomic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "79f541f2-8b14-4591-83c9-12f5c2d0ce44",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n"
     ]
    }
   ],
   "source": [
    "# get gene annotations for hg38\n",
    "annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)\n",
    "seqlevels(annotation) <- paste0('chr', seqlevels(annotation))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "772d22ad-c4fe-4626-8b21-fd7ba795ace5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a Seurat object containing the RNA adata\n",
    "obj.seurat <- CreateSeuratObject(\n",
    "  counts = matrix.rna_count.rename,\n",
    "  assay = \"RNA\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dc72ee8e-03a9-47b6-bce4-4f235fc108a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create ATAC assay and add it to the object\n",
    "obj.seurat[[\"ATAC\"]] <- CreateChromatinAssay(\n",
    "  counts = matrix.atac,\n",
    "  sep = c(\"-\", \"-\"),\n",
    "  annotation = annotation\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7d98b321-9004-4eb9-ac0e-add0f0c38993",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running SCTransform on assay: RNA\n",
      "\n",
      "Running SCTransform on layer: counts\n",
      "\n",
      "vst.flavor='v2' set. Using model with fixed slope and excluding poisson genes.\n",
      "\n",
      "`vst.flavor` is set to 'v2' but could not find glmGamPoi installed.\n",
      "Please install the glmGamPoi package for much faster estimation.\n",
      "--------------------------------------------\n",
      "install.packages('BiocManager')\n",
      "BiocManager::install('glmGamPoi')\n",
      "--------------------------------------------\n",
      "Falling back to native (slower) implementation.\n",
      "\n",
      "\n",
      "Variance stabilizing transformation of count matrix of size 16128 by 7821\n",
      "\n",
      "Model formula is y ~ log_umi\n",
      "\n",
      "Get Negative Binomial regression parameters per gene\n",
      "\n",
      "Using 2000 genes, 5000 cells\n",
      "\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in glm.nb(formula = as.formula(new_formula), data = data):\n",
      "\"alternation limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in glm.nb(formula = as.formula(new_formula), data = data):\n",
      "\"alternation limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in glm.nb(formula = as.formula(new_formula), data = data):\n",
      "\"alternation limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in glm.nb(formula = as.formula(new_formula), data = data):\n",
      "\"alternation limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "\"iteration limit reached\"\n",
      "Found 69 outliers - those will be ignored in fitting/regularization step\n",
      "\n",
      "\n",
      "Second step: Get residuals using fitted parameters for 16128 genes\n",
      "\n",
      "Computing corrected count matrix for 16128 genes\n",
      "\n",
      "Calculating gene attributes\n",
      "\n",
      "Wall clock passed: Time difference of 6.295701 mins\n",
      "\n",
      "Determine variable features\n",
      "\n",
      "Centering data matrix\n",
      "\n",
      "Getting residuals for block 1(of 2) for counts dataset\n",
      "\n",
      "Getting residuals for block 2(of 2) for counts dataset\n",
      "\n",
      "Centering data matrix\n",
      "\n",
      "Finished calculating residuals for counts\n",
      "\n",
      "Set default assay to SCT\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  MALAT1, PPP1R15A, KLF6, TXNIP, VIM, SQSTM1, CPEB4, KLF10, BACH2, SLC3A2 \n",
      "\t   ADGRB3, UBE2H, MIR3681HG, AAK1, BTG1, JUN, NEU1, TALAM1, BTG2, RB1CC1 \n",
      "\t   ARRDC4, DDIT3, RAPGEF2, DUSP8, BMP2K, TRPS1, TPM4, UPP1, GDF15, CALB1 \n",
      "Negative:  HBG2, RPLP1, RPS12, RPS18, HBZ, RPLP2, RPS24, RPL13A, CCDC26, RPS19 \n",
      "\t   RPS11, RPS2, RPL23A, RPL32, PTMA, RPS23, EEF1A1, RPL14, RPL19, RPS27A \n",
      "\t   HBA1, HSP90AA1, HSP90AB1, RPL28, RPS20, RPL10, RPL37, RPS21, HBA2, RPS8 \n",
      "PC_ 2 \n",
      "Positive:  CCDC26, IMMP2L, LINC00534, LVCAT5, KCNQ5, ARL15, STXBP5, XACT, ZFPM2, PCDH9 \n",
      "\t   ADGRB3, CPED1, SMYD3, RELN, ADGRV1, PRKCB, HTR1F, BMP2K, ROBO2, LINC02109 \n",
      "\t   MBNL1, SPATA5, RABGAP1L, PTH2R, MALAT1, LINC01194, VWDE, ISPD, SCN9A, ARHGAP6 \n",
      "Negative:  FTL, FTH1, PPP1R15A, KLF6, TXNIP, SQSTM1, RPL19, HBG2, SLC3A2, KLF10 \n",
      "\t   RPS15, RPL23A, RPS27A, RPL13A, DDIT3, MYL6, RPL32, RPL35, RPS18, RPS11 \n",
      "\t   JUN, RPL14, CLIC1, RPS24, H3F3B, UBC, NACA, NEU1, ACTG1, VIM \n",
      "PC_ 3 \n",
      "Positive:  ADGRB3, VIM, CD24, SCN9A, LINC00534, ANK3, STXBP5, BMP2K, MALAT1, ISPD \n",
      "\t   CD53, PTPRC, ANTXR2, PLD3, LYZ, DLK1, APOE, MIR3681HG, GPR85, HNMT \n",
      "\t   FNBP1, UTRN, LGALS1, CD84, DOCK2, ROBO2, TMSB4X, LAPTM5, CALB1, FTL \n",
      "Negative:  CCDC26, PPP1R15A, RELN, HBZ, KLF10, TXNIP, CPEB4, CR1L, CENPE, KLF6 \n",
      "\t   ASPM, LINC01029, SLC25A21, RB1CC1, RSRC2, NFIA, CSPP1, CENPF, WRN, SLC25A37 \n",
      "\t   USP15, SMC4, SLC3A2, KIF20B, SPTA1, TFRC, DDX10, DBF4, KIF14, HSP90AB1 \n",
      "PC_ 4 \n",
      "Positive:  MALAT1, SLCO2B1, TALAM1, SQSTM1, FTL, NEAT1, TNNI3, IMMP2L, LINC00534, SAMD4A \n",
      "\t   FTH1, SLC25A37, RELN, VIM, ANXA1, FHIT, HBE1, PTPRF, ACTG1, LVCAT5 \n",
      "\t   CPED1, S100A10, HBZ, ZBTB38, BACH2, CREB5, ZFPM2-AS1, TPM4, CCDC26, PTH2R \n",
      "Negative:  ADGRB3, SCN9A, CD24, ROBO2, HSPA1A, KLF10, UTRN, ANK3, DLK1, FNBP1 \n",
      "\t   JRKL-AS1, CENPF, PTPRC, CENPE, HSPA1B, ARID4A, APOE, CDKN2C, CDC20, CDC42BPA \n",
      "\t   HSPB1, TXNIP, NMU, LYZ, GPR85, TOP2A, SSBP2, RSRC2, ASPM, SOSTDC1 \n",
      "PC_ 5 \n",
      "Positive:  HBZ, DTL, HBE1, HBA2, HBG2, LINC01036, LINC00861, LINC01021, HBA1, POLA1 \n",
      "\t   LINC01194, BRIP1, MBD5, MMS22L, BRINP3, FBXL20, CLSPN, HELLS, WDR76, ABHD5 \n",
      "\t   LINC01876, CCNE2, SLC25A37, LINC00355, MSH6, RBBP8, HSP90AB1, GYPB, ATAD2, SLC44A5 \n",
      "Negative:  CCDC26, LINC00534, MALAT1, KIF14, CENPE, TOP2A, ASPM, PLK1, IMMP2L, XACT \n",
      "\t   CDCA8, STXBP5, PRKCB, KCNQ5, KIF2C, UBE2S, CDCA2, KPNA2, CDK1, CDC20 \n",
      "\t   NDC80, AURKA, CCNB1, KIF23, PIF1, KIF18A, ADGRV1, TPX2, TALAM1, MYO16 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "DefaultAssay(obj.seurat) <- \"RNA\"\n",
    "obj.seurat <- SCTransform(obj.seurat)\n",
    "obj.seurat <- RunPCA(obj.seurat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "db07a5e6-3d5c-45eb-9d19-55018e0a1a40",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n"
     ]
    }
   ],
   "source": [
    "DefaultAssay(obj.seurat) <- \"ATAC\"\n",
    "obj.seurat <- FindTopFeatures(obj.seurat, min.cutoff = 5)\n",
    "obj.seurat <- RunTFIDF(obj.seurat)\n",
    "obj.seurat <- RunSVD(obj.seurat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8a87b9d-2f0d-4093-b5d4-3cded4740acc",
   "metadata": {},
   "source": [
    "Linking peaks to genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1e83b1e2-1fb5-47de-8b3f-f26c33aaefcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(obj.seurat) <- \"ATAC\"\n",
    "\n",
    "# first compute the GC content for each peak\n",
    "obj.seurat <- RegionStats(obj.seurat, genome = BSgenome.Hsapiens.UCSC.hg38)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "19e4d078-5a87-467b-ae94-c446e8b8ce3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Testing 14980 genes and 157600 peaks\n",
      "\n",
      "Found gene coordinates for 14753 genes\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# link peaks to genes\n",
    "obj.seurat <- LinkPeaks(\n",
    "  object = obj.seurat,\n",
    "  peak.assay = \"ATAC\",\n",
    "  expression.assay = \"SCT\",\n",
    "  genes.use = rownames(matrix.rna_count.rename[rowSums(matrix.rna_count.rename) > 0,])\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8df0f6c8-01aa-4bd0-b9c0-6b9b42a66b5f",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "db9cb345-cd3b-4ac7-bdd1-c14818757396",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 21911 ranges and 5 metadata columns:\n",
       "          seqnames              ranges strand |     score        gene\n",
       "             <Rle>           <IRanges>  <Rle> | <numeric> <character>\n",
       "      [1]     chr1      778761-1000172      * | 0.0508541        HES4\n",
       "      [2]     chr1      820756-1020123      * | 0.0511646        AGRN\n",
       "      [3]     chr1      827285-1116361      * | 0.0546499    C1orf159\n",
       "      [4]     chr1      940540-1000172      * | 0.0505713        HES4\n",
       "      [5]     chr1       951869-959309      * | 0.0530122       NOC2L\n",
       "      ...      ...                 ...    ... .       ...         ...\n",
       "  [21907]     chrX 154762742-154799249      * | 0.0512082        DKC1\n",
       "  [21908]     chrX 154763237-155071420      * | 0.0604306       BRCC3\n",
       "  [21909]     chrX 154776030-154821007      * | 0.0527089        MPP1\n",
       "  [21910]     chrX 155264589-155334730      * | 0.0516050      RAB39B\n",
       "  [21911]     chrX 155334657-155334730      * | 0.0606970       CLIC2\n",
       "                            peak    zscore      pvalue\n",
       "                     <character> <numeric>   <numeric>\n",
       "      [1]     chr1-778120-779401   2.34342 0.009553860\n",
       "      [2]     chr1-819819-821692   3.09390 0.000987716\n",
       "      [3]     chr1-826642-827928   1.92627 0.027035550\n",
       "      [4]     chr1-939995-941085   1.79947 0.035972385\n",
       "      [5]     chr1-950610-953128   3.45354 0.000276642\n",
       "      ...                    ...       ...         ...\n",
       "  [21907] chrX-154798594-15479..   3.18269 7.29571e-04\n",
       "  [21908] chrX-154762337-15476..   2.67497 3.73678e-03\n",
       "  [21909] chrX-154775311-15477..   3.87754 5.27585e-05\n",
       "  [21910] chrX-155333791-15533..   4.33042 7.44139e-06\n",
       "  [21911] chrX-155333791-15533..   3.94684 3.95948e-05\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dir.create(dir.output,recursive = T)\n",
    "saveRDS(obj.seurat@assays$ATAC@links,\n",
    "        paste(dir.output,\"links.rds\",sep = \"/\"))\n",
    "obj.seurat@assays$ATAC@links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6260d206-3297-42c3-b8d7-af50cccbcc79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 21911 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>width</th><th scope=col>strand</th><th scope=col>score</th><th scope=col>TargetGene</th><th scope=col>PeakName</th><th scope=col>zscore</th><th scope=col>pvalue</th><th scope=col>CellType</th><th scope=col>PairName</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 778761</td><td>1000172</td><td>221412</td><td>*</td><td>0.05085411</td><td>HES4    </td><td>chr1-778120-779401  </td><td>2.343423</td><td>9.553860e-03</td><td>K562</td><td>chr1-778120-779401_HES4      </td></tr>\n",
       "\t<tr><td>chr1</td><td> 820756</td><td>1020123</td><td>199368</td><td>*</td><td>0.05116465</td><td>AGRN    </td><td>chr1-819819-821692  </td><td>3.093901</td><td>9.877161e-04</td><td>K562</td><td>chr1-819819-821692_AGRN      </td></tr>\n",
       "\t<tr><td>chr1</td><td> 827285</td><td>1116361</td><td>289077</td><td>*</td><td>0.05464988</td><td>C1orf159</td><td>chr1-826642-827928  </td><td>1.926267</td><td>2.703555e-02</td><td>K562</td><td>chr1-826642-827928_C1orf159  </td></tr>\n",
       "\t<tr><td>chr1</td><td> 940540</td><td>1000172</td><td> 59633</td><td>*</td><td>0.05057131</td><td>HES4    </td><td>chr1-939995-941085  </td><td>1.799467</td><td>3.597238e-02</td><td>K562</td><td>chr1-939995-941085_HES4      </td></tr>\n",
       "\t<tr><td>chr1</td><td> 951869</td><td> 959309</td><td>  7441</td><td>*</td><td>0.05301224</td><td>NOC2L   </td><td>chr1-950610-953128  </td><td>3.453538</td><td>2.766421e-04</td><td>K562</td><td>chr1-950610-953128_NOC2L     </td></tr>\n",
       "\t<tr><td>chr1</td><td> 959309</td><td>1080049</td><td>120741</td><td>*</td><td>0.05341145</td><td>NOC2L   </td><td>chr1-1079353-1080745</td><td>3.384546</td><td>3.564810e-04</td><td>K562</td><td>chr1-1079353-1080745_NOC2L   </td></tr>\n",
       "\t<tr><td>chr1</td><td> 959309</td><td>1331427</td><td>372119</td><td>*</td><td>0.06348661</td><td>NOC2L   </td><td>chr1-1330543-1332311</td><td>4.451339</td><td>4.266819e-06</td><td>K562</td><td>chr1-1330543-1332311_NOC2L   </td></tr>\n",
       "\t<tr><td>chr1</td><td> 960587</td><td>1000388</td><td> 39802</td><td>*</td><td>0.05638424</td><td>KLHL17  </td><td>chr1-999557-1001218 </td><td>2.216227</td><td>1.333799e-02</td><td>K562</td><td>chr1-999557-1001218_KLHL17   </td></tr>\n",
       "\t<tr><td>chr1</td><td> 960587</td><td>1308322</td><td>347736</td><td>*</td><td>0.05029331</td><td>KLHL17  </td><td>chr1-1307532-1309112</td><td>2.078736</td><td>1.882081e-02</td><td>K562</td><td>chr1-1307532-1309112_KLHL17  </td></tr>\n",
       "\t<tr><td>chr1</td><td> 960857</td><td>1000172</td><td> 39316</td><td>*</td><td>0.05322021</td><td>HES4    </td><td>chr1-960264-961450  </td><td>2.362046</td><td>9.087205e-03</td><td>K562</td><td>chr1-960264-961450_HES4      </td></tr>\n",
       "\t<tr><td>chr1</td><td> 966713</td><td>1000172</td><td> 33460</td><td>*</td><td>0.05778342</td><td>HES4    </td><td>chr1-965905-967520  </td><td>2.267999</td><td>1.166463e-02</td><td>K562</td><td>chr1-965905-967520_HES4      </td></tr>\n",
       "\t<tr><td>chr1</td><td> 966713</td><td>1399328</td><td>432616</td><td>*</td><td>0.07396319</td><td>CCNL2   </td><td>chr1-965905-967520  </td><td>2.610996</td><td>4.513943e-03</td><td>K562</td><td>chr1-965905-967520_CCNL2     </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000172</td><td>1000388</td><td>   217</td><td>*</td><td>0.06471269</td><td>HES4    </td><td>chr1-999557-1001218 </td><td>2.563924</td><td>5.174803e-03</td><td>K562</td><td>chr1-999557-1001218_HES4     </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000172</td><td>1001805</td><td>  1634</td><td>*</td><td>0.05643616</td><td>HES4    </td><td>chr1-1001483-1002126</td><td>2.218792</td><td>1.325043e-02</td><td>K562</td><td>chr1-1001483-1002126_HES4    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000172</td><td>1020074</td><td> 19903</td><td>*</td><td>0.06280860</td><td>HES4    </td><td>chr1-1019151-1020996</td><td>2.586115</td><td>4.853221e-03</td><td>K562</td><td>chr1-1019151-1020996_HES4    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000172</td><td>1059154</td><td> 58983</td><td>*</td><td>0.06285742</td><td>HES4    </td><td>chr1-1058323-1059984</td><td>2.356170</td><td>9.232227e-03</td><td>K562</td><td>chr1-1058323-1059984_HES4    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000172</td><td>1069597</td><td> 69426</td><td>*</td><td>0.05902627</td><td>HES4    </td><td>chr1-1068523-1070670</td><td>2.242326</td><td>1.247016e-02</td><td>K562</td><td>chr1-1068523-1070670_HES4    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000172</td><td>1157468</td><td>157297</td><td>*</td><td>0.05741698</td><td>HES4    </td><td>chr1-1156382-1158554</td><td>2.100162</td><td>1.785731e-02</td><td>K562</td><td>chr1-1156382-1158554_HES4    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000172</td><td>1273797</td><td>273626</td><td>*</td><td>0.05905527</td><td>HES4    </td><td>chr1-1272810-1274783</td><td>2.141628</td><td>1.611173e-02</td><td>K562</td><td>chr1-1272810-1274783_HES4    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000172</td><td>1349468</td><td>349297</td><td>*</td><td>0.05402978</td><td>HES4    </td><td>chr1-1348905-1350031</td><td>2.400730</td><td>8.181207e-03</td><td>K562</td><td>chr1-1348905-1350031_HES4    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000172</td><td>1399614</td><td>399443</td><td>*</td><td>0.05790177</td><td>HES4    </td><td>chr1-1398814-1400413</td><td>2.060970</td><td>1.965296e-02</td><td>K562</td><td>chr1-1398814-1400413_HES4    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000172</td><td>1433558</td><td>433387</td><td>*</td><td>0.05050301</td><td>HES4    </td><td>chr1-1433049-1434067</td><td>2.027773</td><td>2.129171e-02</td><td>K562</td><td>chr1-1433049-1434067_HES4    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000388</td><td>1116361</td><td>115974</td><td>*</td><td>0.07027118</td><td>C1orf159</td><td>chr1-999557-1001218 </td><td>2.074273</td><td>1.902699e-02</td><td>K562</td><td>chr1-999557-1001218_C1orf159 </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000388</td><td>1399328</td><td>398941</td><td>*</td><td>0.07830343</td><td>CCNL2   </td><td>chr1-999557-1001218 </td><td>2.746377</td><td>3.012878e-03</td><td>K562</td><td>chr1-999557-1001218_CCNL2    </td></tr>\n",
       "\t<tr><td>chr1</td><td>1000388</td><td>1471769</td><td>471382</td><td>*</td><td>0.07170709</td><td>ATAD3B  </td><td>chr1-999557-1001218 </td><td>3.215316</td><td>6.515039e-04</td><td>K562</td><td>chr1-999557-1001218_ATAD3B   </td></tr>\n",
       "\t<tr><td>chr1</td><td>1001138</td><td>1157468</td><td>156331</td><td>*</td><td>0.05259096</td><td>ISG15   </td><td>chr1-1156382-1158554</td><td>2.883984</td><td>1.963391e-03</td><td>K562</td><td>chr1-1156382-1158554_ISG15   </td></tr>\n",
       "\t<tr><td>chr1</td><td>1001138</td><td>1208319</td><td>207182</td><td>*</td><td>0.05553247</td><td>ISG15   </td><td>chr1-1208012-1208626</td><td>3.650349</td><td>1.309423e-04</td><td>K562</td><td>chr1-1208012-1208626_ISG15   </td></tr>\n",
       "\t<tr><td>chr1</td><td>1001138</td><td>1318289</td><td>317152</td><td>*</td><td>0.05696939</td><td>ISG15   </td><td>chr1-1317926-1318651</td><td>3.921849</td><td>4.393597e-05</td><td>K562</td><td>chr1-1317926-1318651_ISG15   </td></tr>\n",
       "\t<tr><td>chr1</td><td>1001805</td><td>1399328</td><td>397524</td><td>*</td><td>0.05311551</td><td>CCNL2   </td><td>chr1-1001483-1002126</td><td>1.831572</td><td>3.350764e-02</td><td>K562</td><td>chr1-1001483-1002126_CCNL2   </td></tr>\n",
       "\t<tr><td>chr1</td><td>1020074</td><td>1116361</td><td> 96288</td><td>*</td><td>0.05795131</td><td>C1orf159</td><td>chr1-1019151-1020996</td><td>1.730659</td><td>4.175630e-02</td><td>K562</td><td>chr1-1019151-1020996_C1orf159</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>154436913</td><td>154490635</td><td> 53723</td><td>*</td><td>0.06877406</td><td>GDI1  </td><td>chrX-154490012-154491257</td><td>2.032247</td><td>2.106431e-02</td><td>K562</td><td>chrX-154490012-154491257_GDI1  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154436913</td><td>154547314</td><td>110402</td><td>*</td><td>0.06671435</td><td>GDI1  </td><td>chrX-154546671-154547957</td><td>2.149107</td><td>1.581295e-02</td><td>K562</td><td>chrX-154546671-154547957_GDI1  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154436913</td><td>154776030</td><td>339118</td><td>*</td><td>0.05306943</td><td>GDI1  </td><td>chrX-154775311-154776748</td><td>2.274201</td><td>1.147694e-02</td><td>K562</td><td>chrX-154775311-154776748_GDI1  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154436913</td><td>154815477</td><td>378565</td><td>*</td><td>0.05584039</td><td>GDI1  </td><td>chrX-154815322-154815631</td><td>2.950147</td><td>1.588115e-03</td><td>K562</td><td>chrX-154815322-154815631_GDI1  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154436913</td><td>154815782</td><td>378870</td><td>*</td><td>0.05443369</td><td>GDI1  </td><td>chrX-154815707-154815857</td><td>3.609951</td><td>1.531276e-04</td><td>K562</td><td>chrX-154815707-154815857_GDI1  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154444126</td><td>154458894</td><td> 14769</td><td>*</td><td>0.05041422</td><td>FAM50A</td><td>chrX-154457558-154460229</td><td>1.936615</td><td>2.639619e-02</td><td>K562</td><td>chrX-154457558-154460229_FAM50A</td></tr>\n",
       "\t<tr><td>chrX</td><td>154478948</td><td>154486670</td><td>  7723</td><td>*</td><td>0.05441445</td><td>UBL4A </td><td>chrX-154478219-154479676</td><td>1.777295</td><td>3.775985e-02</td><td>K562</td><td>chrX-154478219-154479676_UBL4A </td></tr>\n",
       "\t<tr><td>chrX</td><td>154486552</td><td>154486670</td><td>   119</td><td>*</td><td>0.05458454</td><td>UBL4A </td><td>chrX-154486006-154487098</td><td>1.966755</td><td>2.460571e-02</td><td>K562</td><td>chrX-154486006-154487098_UBL4A </td></tr>\n",
       "\t<tr><td>chrX</td><td>154486552</td><td>154751583</td><td>265032</td><td>*</td><td>0.06137794</td><td>GAB3  </td><td>chrX-154486006-154487098</td><td>2.026130</td><td>2.137572e-02</td><td>K562</td><td>chrX-154486006-154487098_GAB3  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154486670</td><td>154516145</td><td> 29476</td><td>*</td><td>0.05149503</td><td>UBL4A </td><td>chrX-154515321-154516969</td><td>1.768604</td><td>3.848000e-02</td><td>K562</td><td>chrX-154515321-154516969_UBL4A </td></tr>\n",
       "\t<tr><td>chrX</td><td>154486670</td><td>154763237</td><td>276568</td><td>*</td><td>0.05075481</td><td>UBL4A </td><td>chrX-154762337-154764137</td><td>2.139178</td><td>1.621063e-02</td><td>K562</td><td>chrX-154762337-154764137_UBL4A </td></tr>\n",
       "\t<tr><td>chrX</td><td>154486670</td><td>154799249</td><td>312580</td><td>*</td><td>0.05313110</td><td>UBL4A </td><td>chrX-154798594-154799904</td><td>2.681901</td><td>3.660254e-03</td><td>K562</td><td>chrX-154798594-154799904_UBL4A </td></tr>\n",
       "\t<tr><td>chrX</td><td>154495942</td><td>154751583</td><td>255642</td><td>*</td><td>0.05310293</td><td>GAB3  </td><td>chrX-154495283-154496601</td><td>1.925464</td><td>2.708567e-02</td><td>K562</td><td>chrX-154495283-154496601_GAB3  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154516242</td><td>154547314</td><td> 31073</td><td>*</td><td>0.05218297</td><td>FAM3A </td><td>chrX-154546671-154547957</td><td>2.289939</td><td>1.101242e-02</td><td>K562</td><td>chrX-154546671-154547957_FAM3A </td></tr>\n",
       "\t<tr><td>chrX</td><td>154547314</td><td>154821007</td><td>273694</td><td>*</td><td>0.05301616</td><td>MPP1  </td><td>chrX-154546671-154547957</td><td>3.308351</td><td>4.692361e-04</td><td>K562</td><td>chrX-154546671-154547957_MPP1  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154751337</td><td>154751583</td><td>   247</td><td>*</td><td>0.08521963</td><td>GAB3  </td><td>chrX-154750373-154752301</td><td>3.520915</td><td>2.150301e-04</td><td>K562</td><td>chrX-154750373-154752301_GAB3  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154751337</td><td>154762742</td><td> 11406</td><td>*</td><td>0.06416246</td><td>DKC1  </td><td>chrX-154750373-154752301</td><td>4.162127</td><td>1.576482e-05</td><td>K562</td><td>chrX-154750373-154752301_DKC1  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154751583</td><td>154763237</td><td> 11655</td><td>*</td><td>0.06442513</td><td>GAB3  </td><td>chrX-154762337-154764137</td><td>2.625673</td><td>4.323896e-03</td><td>K562</td><td>chrX-154762337-154764137_GAB3  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154751583</td><td>154778206</td><td> 26624</td><td>*</td><td>0.05562112</td><td>GAB3  </td><td>chrX-154777776-154778636</td><td>2.127016</td><td>1.670937e-02</td><td>K562</td><td>chrX-154777776-154778636_GAB3  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154751583</td><td>155071449</td><td>319867</td><td>*</td><td>0.06132955</td><td>GAB3  </td><td>chrX-155070725-155072173</td><td>1.738662</td><td>4.104714e-02</td><td>K562</td><td>chrX-155070725-155072173_GAB3  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154751583</td><td>155086941</td><td>335359</td><td>*</td><td>0.05059728</td><td>GAB3  </td><td>chrX-155085964-155087918</td><td>1.680416</td><td>4.643823e-02</td><td>K562</td><td>chrX-155085964-155087918_GAB3  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154751583</td><td>155216308</td><td>464726</td><td>*</td><td>0.05497587</td><td>GAB3  </td><td>chrX-155215583-155217032</td><td>1.949908</td><td>2.559355e-02</td><td>K562</td><td>chrX-155215583-155217032_GAB3  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154751583</td><td>155218972</td><td>467390</td><td>*</td><td>0.05779409</td><td>GAB3  </td><td>chrX-155218826-155219117</td><td>2.944411</td><td>1.617849e-03</td><td>K562</td><td>chrX-155218826-155219117_GAB3  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154762742</td><td>154763237</td><td>   496</td><td>*</td><td>0.05202060</td><td>DKC1  </td><td>chrX-154762337-154764137</td><td>3.311179</td><td>4.645181e-04</td><td>K562</td><td>chrX-154762337-154764137_DKC1  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154762742</td><td>154785456</td><td> 22715</td><td>*</td><td>0.05343360</td><td>DKC1  </td><td>chrX-154784088-154786824</td><td>3.408487</td><td>3.266206e-04</td><td>K562</td><td>chrX-154784088-154786824_DKC1  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154762742</td><td>154799249</td><td> 36508</td><td>*</td><td>0.05120821</td><td>DKC1  </td><td>chrX-154798594-154799904</td><td>3.182689</td><td>7.295712e-04</td><td>K562</td><td>chrX-154798594-154799904_DKC1  </td></tr>\n",
       "\t<tr><td>chrX</td><td>154763237</td><td>155071420</td><td>308184</td><td>*</td><td>0.06043059</td><td>BRCC3 </td><td>chrX-154762337-154764137</td><td>2.674971</td><td>3.736781e-03</td><td>K562</td><td>chrX-154762337-154764137_BRCC3 </td></tr>\n",
       "\t<tr><td>chrX</td><td>154776030</td><td>154821007</td><td> 44978</td><td>*</td><td>0.05270889</td><td>MPP1  </td><td>chrX-154775311-154776748</td><td>3.877542</td><td>5.275849e-05</td><td>K562</td><td>chrX-154775311-154776748_MPP1  </td></tr>\n",
       "\t<tr><td>chrX</td><td>155264589</td><td>155334730</td><td> 70142</td><td>*</td><td>0.05160496</td><td>RAB39B</td><td>chrX-155333791-155335669</td><td>4.330416</td><td>7.441395e-06</td><td>K562</td><td>chrX-155333791-155335669_RAB39B</td></tr>\n",
       "\t<tr><td>chrX</td><td>155334657</td><td>155334730</td><td>    74</td><td>*</td><td>0.06069699</td><td>CLIC2 </td><td>chrX-155333791-155335669</td><td>3.946840</td><td>3.959475e-05</td><td>K562</td><td>chrX-155333791-155335669_CLIC2 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 21911 × 12\n",
       "\\begin{tabular}{llllllllllll}\n",
       " chr & start & end & width & strand & score & TargetGene & PeakName & zscore & pvalue & CellType & PairName\\\\\n",
       " <fct> & <int> & <int> & <int> & <fct> & <dbl> & <chr> & <chr> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t chr1 &  778761 & 1000172 & 221412 & * & 0.05085411 & HES4     & chr1-778120-779401   & 2.343423 & 9.553860e-03 & K562 & chr1-778120-779401\\_HES4      \\\\\n",
       "\t chr1 &  820756 & 1020123 & 199368 & * & 0.05116465 & AGRN     & chr1-819819-821692   & 3.093901 & 9.877161e-04 & K562 & chr1-819819-821692\\_AGRN      \\\\\n",
       "\t chr1 &  827285 & 1116361 & 289077 & * & 0.05464988 & C1orf159 & chr1-826642-827928   & 1.926267 & 2.703555e-02 & K562 & chr1-826642-827928\\_C1orf159  \\\\\n",
       "\t chr1 &  940540 & 1000172 &  59633 & * & 0.05057131 & HES4     & chr1-939995-941085   & 1.799467 & 3.597238e-02 & K562 & chr1-939995-941085\\_HES4      \\\\\n",
       "\t chr1 &  951869 &  959309 &   7441 & * & 0.05301224 & NOC2L    & chr1-950610-953128   & 3.453538 & 2.766421e-04 & K562 & chr1-950610-953128\\_NOC2L     \\\\\n",
       "\t chr1 &  959309 & 1080049 & 120741 & * & 0.05341145 & NOC2L    & chr1-1079353-1080745 & 3.384546 & 3.564810e-04 & K562 & chr1-1079353-1080745\\_NOC2L   \\\\\n",
       "\t chr1 &  959309 & 1331427 & 372119 & * & 0.06348661 & NOC2L    & chr1-1330543-1332311 & 4.451339 & 4.266819e-06 & K562 & chr1-1330543-1332311\\_NOC2L   \\\\\n",
       "\t chr1 &  960587 & 1000388 &  39802 & * & 0.05638424 & KLHL17   & chr1-999557-1001218  & 2.216227 & 1.333799e-02 & K562 & chr1-999557-1001218\\_KLHL17   \\\\\n",
       "\t chr1 &  960587 & 1308322 & 347736 & * & 0.05029331 & KLHL17   & chr1-1307532-1309112 & 2.078736 & 1.882081e-02 & K562 & chr1-1307532-1309112\\_KLHL17  \\\\\n",
       "\t chr1 &  960857 & 1000172 &  39316 & * & 0.05322021 & HES4     & chr1-960264-961450   & 2.362046 & 9.087205e-03 & K562 & chr1-960264-961450\\_HES4      \\\\\n",
       "\t chr1 &  966713 & 1000172 &  33460 & * & 0.05778342 & HES4     & chr1-965905-967520   & 2.267999 & 1.166463e-02 & K562 & chr1-965905-967520\\_HES4      \\\\\n",
       "\t chr1 &  966713 & 1399328 & 432616 & * & 0.07396319 & CCNL2    & chr1-965905-967520   & 2.610996 & 4.513943e-03 & K562 & chr1-965905-967520\\_CCNL2     \\\\\n",
       "\t chr1 & 1000172 & 1000388 &    217 & * & 0.06471269 & HES4     & chr1-999557-1001218  & 2.563924 & 5.174803e-03 & K562 & chr1-999557-1001218\\_HES4     \\\\\n",
       "\t chr1 & 1000172 & 1001805 &   1634 & * & 0.05643616 & HES4     & chr1-1001483-1002126 & 2.218792 & 1.325043e-02 & K562 & chr1-1001483-1002126\\_HES4    \\\\\n",
       "\t chr1 & 1000172 & 1020074 &  19903 & * & 0.06280860 & HES4     & chr1-1019151-1020996 & 2.586115 & 4.853221e-03 & K562 & chr1-1019151-1020996\\_HES4    \\\\\n",
       "\t chr1 & 1000172 & 1059154 &  58983 & * & 0.06285742 & HES4     & chr1-1058323-1059984 & 2.356170 & 9.232227e-03 & K562 & chr1-1058323-1059984\\_HES4    \\\\\n",
       "\t chr1 & 1000172 & 1069597 &  69426 & * & 0.05902627 & HES4     & chr1-1068523-1070670 & 2.242326 & 1.247016e-02 & K562 & chr1-1068523-1070670\\_HES4    \\\\\n",
       "\t chr1 & 1000172 & 1157468 & 157297 & * & 0.05741698 & HES4     & chr1-1156382-1158554 & 2.100162 & 1.785731e-02 & K562 & chr1-1156382-1158554\\_HES4    \\\\\n",
       "\t chr1 & 1000172 & 1273797 & 273626 & * & 0.05905527 & HES4     & chr1-1272810-1274783 & 2.141628 & 1.611173e-02 & K562 & chr1-1272810-1274783\\_HES4    \\\\\n",
       "\t chr1 & 1000172 & 1349468 & 349297 & * & 0.05402978 & HES4     & chr1-1348905-1350031 & 2.400730 & 8.181207e-03 & K562 & chr1-1348905-1350031\\_HES4    \\\\\n",
       "\t chr1 & 1000172 & 1399614 & 399443 & * & 0.05790177 & HES4     & chr1-1398814-1400413 & 2.060970 & 1.965296e-02 & K562 & chr1-1398814-1400413\\_HES4    \\\\\n",
       "\t chr1 & 1000172 & 1433558 & 433387 & * & 0.05050301 & HES4     & chr1-1433049-1434067 & 2.027773 & 2.129171e-02 & K562 & chr1-1433049-1434067\\_HES4    \\\\\n",
       "\t chr1 & 1000388 & 1116361 & 115974 & * & 0.07027118 & C1orf159 & chr1-999557-1001218  & 2.074273 & 1.902699e-02 & K562 & chr1-999557-1001218\\_C1orf159 \\\\\n",
       "\t chr1 & 1000388 & 1399328 & 398941 & * & 0.07830343 & CCNL2    & chr1-999557-1001218  & 2.746377 & 3.012878e-03 & K562 & chr1-999557-1001218\\_CCNL2    \\\\\n",
       "\t chr1 & 1000388 & 1471769 & 471382 & * & 0.07170709 & ATAD3B   & chr1-999557-1001218  & 3.215316 & 6.515039e-04 & K562 & chr1-999557-1001218\\_ATAD3B   \\\\\n",
       "\t chr1 & 1001138 & 1157468 & 156331 & * & 0.05259096 & ISG15    & chr1-1156382-1158554 & 2.883984 & 1.963391e-03 & K562 & chr1-1156382-1158554\\_ISG15   \\\\\n",
       "\t chr1 & 1001138 & 1208319 & 207182 & * & 0.05553247 & ISG15    & chr1-1208012-1208626 & 3.650349 & 1.309423e-04 & K562 & chr1-1208012-1208626\\_ISG15   \\\\\n",
       "\t chr1 & 1001138 & 1318289 & 317152 & * & 0.05696939 & ISG15    & chr1-1317926-1318651 & 3.921849 & 4.393597e-05 & K562 & chr1-1317926-1318651\\_ISG15   \\\\\n",
       "\t chr1 & 1001805 & 1399328 & 397524 & * & 0.05311551 & CCNL2    & chr1-1001483-1002126 & 1.831572 & 3.350764e-02 & K562 & chr1-1001483-1002126\\_CCNL2   \\\\\n",
       "\t chr1 & 1020074 & 1116361 &  96288 & * & 0.05795131 & C1orf159 & chr1-1019151-1020996 & 1.730659 & 4.175630e-02 & K562 & chr1-1019151-1020996\\_C1orf159\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 154436913 & 154490635 &  53723 & * & 0.06877406 & GDI1   & chrX-154490012-154491257 & 2.032247 & 2.106431e-02 & K562 & chrX-154490012-154491257\\_GDI1  \\\\\n",
       "\t chrX & 154436913 & 154547314 & 110402 & * & 0.06671435 & GDI1   & chrX-154546671-154547957 & 2.149107 & 1.581295e-02 & K562 & chrX-154546671-154547957\\_GDI1  \\\\\n",
       "\t chrX & 154436913 & 154776030 & 339118 & * & 0.05306943 & GDI1   & chrX-154775311-154776748 & 2.274201 & 1.147694e-02 & K562 & chrX-154775311-154776748\\_GDI1  \\\\\n",
       "\t chrX & 154436913 & 154815477 & 378565 & * & 0.05584039 & GDI1   & chrX-154815322-154815631 & 2.950147 & 1.588115e-03 & K562 & chrX-154815322-154815631\\_GDI1  \\\\\n",
       "\t chrX & 154436913 & 154815782 & 378870 & * & 0.05443369 & GDI1   & chrX-154815707-154815857 & 3.609951 & 1.531276e-04 & K562 & chrX-154815707-154815857\\_GDI1  \\\\\n",
       "\t chrX & 154444126 & 154458894 &  14769 & * & 0.05041422 & FAM50A & chrX-154457558-154460229 & 1.936615 & 2.639619e-02 & K562 & chrX-154457558-154460229\\_FAM50A\\\\\n",
       "\t chrX & 154478948 & 154486670 &   7723 & * & 0.05441445 & UBL4A  & chrX-154478219-154479676 & 1.777295 & 3.775985e-02 & K562 & chrX-154478219-154479676\\_UBL4A \\\\\n",
       "\t chrX & 154486552 & 154486670 &    119 & * & 0.05458454 & UBL4A  & chrX-154486006-154487098 & 1.966755 & 2.460571e-02 & K562 & chrX-154486006-154487098\\_UBL4A \\\\\n",
       "\t chrX & 154486552 & 154751583 & 265032 & * & 0.06137794 & GAB3   & chrX-154486006-154487098 & 2.026130 & 2.137572e-02 & K562 & chrX-154486006-154487098\\_GAB3  \\\\\n",
       "\t chrX & 154486670 & 154516145 &  29476 & * & 0.05149503 & UBL4A  & chrX-154515321-154516969 & 1.768604 & 3.848000e-02 & K562 & chrX-154515321-154516969\\_UBL4A \\\\\n",
       "\t chrX & 154486670 & 154763237 & 276568 & * & 0.05075481 & UBL4A  & chrX-154762337-154764137 & 2.139178 & 1.621063e-02 & K562 & chrX-154762337-154764137\\_UBL4A \\\\\n",
       "\t chrX & 154486670 & 154799249 & 312580 & * & 0.05313110 & UBL4A  & chrX-154798594-154799904 & 2.681901 & 3.660254e-03 & K562 & chrX-154798594-154799904\\_UBL4A \\\\\n",
       "\t chrX & 154495942 & 154751583 & 255642 & * & 0.05310293 & GAB3   & chrX-154495283-154496601 & 1.925464 & 2.708567e-02 & K562 & chrX-154495283-154496601\\_GAB3  \\\\\n",
       "\t chrX & 154516242 & 154547314 &  31073 & * & 0.05218297 & FAM3A  & chrX-154546671-154547957 & 2.289939 & 1.101242e-02 & K562 & chrX-154546671-154547957\\_FAM3A \\\\\n",
       "\t chrX & 154547314 & 154821007 & 273694 & * & 0.05301616 & MPP1   & chrX-154546671-154547957 & 3.308351 & 4.692361e-04 & K562 & chrX-154546671-154547957\\_MPP1  \\\\\n",
       "\t chrX & 154751337 & 154751583 &    247 & * & 0.08521963 & GAB3   & chrX-154750373-154752301 & 3.520915 & 2.150301e-04 & K562 & chrX-154750373-154752301\\_GAB3  \\\\\n",
       "\t chrX & 154751337 & 154762742 &  11406 & * & 0.06416246 & DKC1   & chrX-154750373-154752301 & 4.162127 & 1.576482e-05 & K562 & chrX-154750373-154752301\\_DKC1  \\\\\n",
       "\t chrX & 154751583 & 154763237 &  11655 & * & 0.06442513 & GAB3   & chrX-154762337-154764137 & 2.625673 & 4.323896e-03 & K562 & chrX-154762337-154764137\\_GAB3  \\\\\n",
       "\t chrX & 154751583 & 154778206 &  26624 & * & 0.05562112 & GAB3   & chrX-154777776-154778636 & 2.127016 & 1.670937e-02 & K562 & chrX-154777776-154778636\\_GAB3  \\\\\n",
       "\t chrX & 154751583 & 155071449 & 319867 & * & 0.06132955 & GAB3   & chrX-155070725-155072173 & 1.738662 & 4.104714e-02 & K562 & chrX-155070725-155072173\\_GAB3  \\\\\n",
       "\t chrX & 154751583 & 155086941 & 335359 & * & 0.05059728 & GAB3   & chrX-155085964-155087918 & 1.680416 & 4.643823e-02 & K562 & chrX-155085964-155087918\\_GAB3  \\\\\n",
       "\t chrX & 154751583 & 155216308 & 464726 & * & 0.05497587 & GAB3   & chrX-155215583-155217032 & 1.949908 & 2.559355e-02 & K562 & chrX-155215583-155217032\\_GAB3  \\\\\n",
       "\t chrX & 154751583 & 155218972 & 467390 & * & 0.05779409 & GAB3   & chrX-155218826-155219117 & 2.944411 & 1.617849e-03 & K562 & chrX-155218826-155219117\\_GAB3  \\\\\n",
       "\t chrX & 154762742 & 154763237 &    496 & * & 0.05202060 & DKC1   & chrX-154762337-154764137 & 3.311179 & 4.645181e-04 & K562 & chrX-154762337-154764137\\_DKC1  \\\\\n",
       "\t chrX & 154762742 & 154785456 &  22715 & * & 0.05343360 & DKC1   & chrX-154784088-154786824 & 3.408487 & 3.266206e-04 & K562 & chrX-154784088-154786824\\_DKC1  \\\\\n",
       "\t chrX & 154762742 & 154799249 &  36508 & * & 0.05120821 & DKC1   & chrX-154798594-154799904 & 3.182689 & 7.295712e-04 & K562 & chrX-154798594-154799904\\_DKC1  \\\\\n",
       "\t chrX & 154763237 & 155071420 & 308184 & * & 0.06043059 & BRCC3  & chrX-154762337-154764137 & 2.674971 & 3.736781e-03 & K562 & chrX-154762337-154764137\\_BRCC3 \\\\\n",
       "\t chrX & 154776030 & 154821007 &  44978 & * & 0.05270889 & MPP1   & chrX-154775311-154776748 & 3.877542 & 5.275849e-05 & K562 & chrX-154775311-154776748\\_MPP1  \\\\\n",
       "\t chrX & 155264589 & 155334730 &  70142 & * & 0.05160496 & RAB39B & chrX-155333791-155335669 & 4.330416 & 7.441395e-06 & K562 & chrX-155333791-155335669\\_RAB39B\\\\\n",
       "\t chrX & 155334657 & 155334730 &     74 & * & 0.06069699 & CLIC2  & chrX-155333791-155335669 & 3.946840 & 3.959475e-05 & K562 & chrX-155333791-155335669\\_CLIC2 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 21911 × 12\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | width &lt;int&gt; | strand &lt;fct&gt; | score &lt;dbl&gt; | TargetGene &lt;chr&gt; | PeakName &lt;chr&gt; | zscore &lt;dbl&gt; | pvalue &lt;dbl&gt; | CellType &lt;chr&gt; | PairName &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  778761 | 1000172 | 221412 | * | 0.05085411 | HES4     | chr1-778120-779401   | 2.343423 | 9.553860e-03 | K562 | chr1-778120-779401_HES4       |\n",
       "| chr1 |  820756 | 1020123 | 199368 | * | 0.05116465 | AGRN     | chr1-819819-821692   | 3.093901 | 9.877161e-04 | K562 | chr1-819819-821692_AGRN       |\n",
       "| chr1 |  827285 | 1116361 | 289077 | * | 0.05464988 | C1orf159 | chr1-826642-827928   | 1.926267 | 2.703555e-02 | K562 | chr1-826642-827928_C1orf159   |\n",
       "| chr1 |  940540 | 1000172 |  59633 | * | 0.05057131 | HES4     | chr1-939995-941085   | 1.799467 | 3.597238e-02 | K562 | chr1-939995-941085_HES4       |\n",
       "| chr1 |  951869 |  959309 |   7441 | * | 0.05301224 | NOC2L    | chr1-950610-953128   | 3.453538 | 2.766421e-04 | K562 | chr1-950610-953128_NOC2L      |\n",
       "| chr1 |  959309 | 1080049 | 120741 | * | 0.05341145 | NOC2L    | chr1-1079353-1080745 | 3.384546 | 3.564810e-04 | K562 | chr1-1079353-1080745_NOC2L    |\n",
       "| chr1 |  959309 | 1331427 | 372119 | * | 0.06348661 | NOC2L    | chr1-1330543-1332311 | 4.451339 | 4.266819e-06 | K562 | chr1-1330543-1332311_NOC2L    |\n",
       "| chr1 |  960587 | 1000388 |  39802 | * | 0.05638424 | KLHL17   | chr1-999557-1001218  | 2.216227 | 1.333799e-02 | K562 | chr1-999557-1001218_KLHL17    |\n",
       "| chr1 |  960587 | 1308322 | 347736 | * | 0.05029331 | KLHL17   | chr1-1307532-1309112 | 2.078736 | 1.882081e-02 | K562 | chr1-1307532-1309112_KLHL17   |\n",
       "| chr1 |  960857 | 1000172 |  39316 | * | 0.05322021 | HES4     | chr1-960264-961450   | 2.362046 | 9.087205e-03 | K562 | chr1-960264-961450_HES4       |\n",
       "| chr1 |  966713 | 1000172 |  33460 | * | 0.05778342 | HES4     | chr1-965905-967520   | 2.267999 | 1.166463e-02 | K562 | chr1-965905-967520_HES4       |\n",
       "| chr1 |  966713 | 1399328 | 432616 | * | 0.07396319 | CCNL2    | chr1-965905-967520   | 2.610996 | 4.513943e-03 | K562 | chr1-965905-967520_CCNL2      |\n",
       "| chr1 | 1000172 | 1000388 |    217 | * | 0.06471269 | HES4     | chr1-999557-1001218  | 2.563924 | 5.174803e-03 | K562 | chr1-999557-1001218_HES4      |\n",
       "| chr1 | 1000172 | 1001805 |   1634 | * | 0.05643616 | HES4     | chr1-1001483-1002126 | 2.218792 | 1.325043e-02 | K562 | chr1-1001483-1002126_HES4     |\n",
       "| chr1 | 1000172 | 1020074 |  19903 | * | 0.06280860 | HES4     | chr1-1019151-1020996 | 2.586115 | 4.853221e-03 | K562 | chr1-1019151-1020996_HES4     |\n",
       "| chr1 | 1000172 | 1059154 |  58983 | * | 0.06285742 | HES4     | chr1-1058323-1059984 | 2.356170 | 9.232227e-03 | K562 | chr1-1058323-1059984_HES4     |\n",
       "| chr1 | 1000172 | 1069597 |  69426 | * | 0.05902627 | HES4     | chr1-1068523-1070670 | 2.242326 | 1.247016e-02 | K562 | chr1-1068523-1070670_HES4     |\n",
       "| chr1 | 1000172 | 1157468 | 157297 | * | 0.05741698 | HES4     | chr1-1156382-1158554 | 2.100162 | 1.785731e-02 | K562 | chr1-1156382-1158554_HES4     |\n",
       "| chr1 | 1000172 | 1273797 | 273626 | * | 0.05905527 | HES4     | chr1-1272810-1274783 | 2.141628 | 1.611173e-02 | K562 | chr1-1272810-1274783_HES4     |\n",
       "| chr1 | 1000172 | 1349468 | 349297 | * | 0.05402978 | HES4     | chr1-1348905-1350031 | 2.400730 | 8.181207e-03 | K562 | chr1-1348905-1350031_HES4     |\n",
       "| chr1 | 1000172 | 1399614 | 399443 | * | 0.05790177 | HES4     | chr1-1398814-1400413 | 2.060970 | 1.965296e-02 | K562 | chr1-1398814-1400413_HES4     |\n",
       "| chr1 | 1000172 | 1433558 | 433387 | * | 0.05050301 | HES4     | chr1-1433049-1434067 | 2.027773 | 2.129171e-02 | K562 | chr1-1433049-1434067_HES4     |\n",
       "| chr1 | 1000388 | 1116361 | 115974 | * | 0.07027118 | C1orf159 | chr1-999557-1001218  | 2.074273 | 1.902699e-02 | K562 | chr1-999557-1001218_C1orf159  |\n",
       "| chr1 | 1000388 | 1399328 | 398941 | * | 0.07830343 | CCNL2    | chr1-999557-1001218  | 2.746377 | 3.012878e-03 | K562 | chr1-999557-1001218_CCNL2     |\n",
       "| chr1 | 1000388 | 1471769 | 471382 | * | 0.07170709 | ATAD3B   | chr1-999557-1001218  | 3.215316 | 6.515039e-04 | K562 | chr1-999557-1001218_ATAD3B    |\n",
       "| chr1 | 1001138 | 1157468 | 156331 | * | 0.05259096 | ISG15    | chr1-1156382-1158554 | 2.883984 | 1.963391e-03 | K562 | chr1-1156382-1158554_ISG15    |\n",
       "| chr1 | 1001138 | 1208319 | 207182 | * | 0.05553247 | ISG15    | chr1-1208012-1208626 | 3.650349 | 1.309423e-04 | K562 | chr1-1208012-1208626_ISG15    |\n",
       "| chr1 | 1001138 | 1318289 | 317152 | * | 0.05696939 | ISG15    | chr1-1317926-1318651 | 3.921849 | 4.393597e-05 | K562 | chr1-1317926-1318651_ISG15    |\n",
       "| chr1 | 1001805 | 1399328 | 397524 | * | 0.05311551 | CCNL2    | chr1-1001483-1002126 | 1.831572 | 3.350764e-02 | K562 | chr1-1001483-1002126_CCNL2    |\n",
       "| chr1 | 1020074 | 1116361 |  96288 | * | 0.05795131 | C1orf159 | chr1-1019151-1020996 | 1.730659 | 4.175630e-02 | K562 | chr1-1019151-1020996_C1orf159 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 154436913 | 154490635 |  53723 | * | 0.06877406 | GDI1   | chrX-154490012-154491257 | 2.032247 | 2.106431e-02 | K562 | chrX-154490012-154491257_GDI1   |\n",
       "| chrX | 154436913 | 154547314 | 110402 | * | 0.06671435 | GDI1   | chrX-154546671-154547957 | 2.149107 | 1.581295e-02 | K562 | chrX-154546671-154547957_GDI1   |\n",
       "| chrX | 154436913 | 154776030 | 339118 | * | 0.05306943 | GDI1   | chrX-154775311-154776748 | 2.274201 | 1.147694e-02 | K562 | chrX-154775311-154776748_GDI1   |\n",
       "| chrX | 154436913 | 154815477 | 378565 | * | 0.05584039 | GDI1   | chrX-154815322-154815631 | 2.950147 | 1.588115e-03 | K562 | chrX-154815322-154815631_GDI1   |\n",
       "| chrX | 154436913 | 154815782 | 378870 | * | 0.05443369 | GDI1   | chrX-154815707-154815857 | 3.609951 | 1.531276e-04 | K562 | chrX-154815707-154815857_GDI1   |\n",
       "| chrX | 154444126 | 154458894 |  14769 | * | 0.05041422 | FAM50A | chrX-154457558-154460229 | 1.936615 | 2.639619e-02 | K562 | chrX-154457558-154460229_FAM50A |\n",
       "| chrX | 154478948 | 154486670 |   7723 | * | 0.05441445 | UBL4A  | chrX-154478219-154479676 | 1.777295 | 3.775985e-02 | K562 | chrX-154478219-154479676_UBL4A  |\n",
       "| chrX | 154486552 | 154486670 |    119 | * | 0.05458454 | UBL4A  | chrX-154486006-154487098 | 1.966755 | 2.460571e-02 | K562 | chrX-154486006-154487098_UBL4A  |\n",
       "| chrX | 154486552 | 154751583 | 265032 | * | 0.06137794 | GAB3   | chrX-154486006-154487098 | 2.026130 | 2.137572e-02 | K562 | chrX-154486006-154487098_GAB3   |\n",
       "| chrX | 154486670 | 154516145 |  29476 | * | 0.05149503 | UBL4A  | chrX-154515321-154516969 | 1.768604 | 3.848000e-02 | K562 | chrX-154515321-154516969_UBL4A  |\n",
       "| chrX | 154486670 | 154763237 | 276568 | * | 0.05075481 | UBL4A  | chrX-154762337-154764137 | 2.139178 | 1.621063e-02 | K562 | chrX-154762337-154764137_UBL4A  |\n",
       "| chrX | 154486670 | 154799249 | 312580 | * | 0.05313110 | UBL4A  | chrX-154798594-154799904 | 2.681901 | 3.660254e-03 | K562 | chrX-154798594-154799904_UBL4A  |\n",
       "| chrX | 154495942 | 154751583 | 255642 | * | 0.05310293 | GAB3   | chrX-154495283-154496601 | 1.925464 | 2.708567e-02 | K562 | chrX-154495283-154496601_GAB3   |\n",
       "| chrX | 154516242 | 154547314 |  31073 | * | 0.05218297 | FAM3A  | chrX-154546671-154547957 | 2.289939 | 1.101242e-02 | K562 | chrX-154546671-154547957_FAM3A  |\n",
       "| chrX | 154547314 | 154821007 | 273694 | * | 0.05301616 | MPP1   | chrX-154546671-154547957 | 3.308351 | 4.692361e-04 | K562 | chrX-154546671-154547957_MPP1   |\n",
       "| chrX | 154751337 | 154751583 |    247 | * | 0.08521963 | GAB3   | chrX-154750373-154752301 | 3.520915 | 2.150301e-04 | K562 | chrX-154750373-154752301_GAB3   |\n",
       "| chrX | 154751337 | 154762742 |  11406 | * | 0.06416246 | DKC1   | chrX-154750373-154752301 | 4.162127 | 1.576482e-05 | K562 | chrX-154750373-154752301_DKC1   |\n",
       "| chrX | 154751583 | 154763237 |  11655 | * | 0.06442513 | GAB3   | chrX-154762337-154764137 | 2.625673 | 4.323896e-03 | K562 | chrX-154762337-154764137_GAB3   |\n",
       "| chrX | 154751583 | 154778206 |  26624 | * | 0.05562112 | GAB3   | chrX-154777776-154778636 | 2.127016 | 1.670937e-02 | K562 | chrX-154777776-154778636_GAB3   |\n",
       "| chrX | 154751583 | 155071449 | 319867 | * | 0.06132955 | GAB3   | chrX-155070725-155072173 | 1.738662 | 4.104714e-02 | K562 | chrX-155070725-155072173_GAB3   |\n",
       "| chrX | 154751583 | 155086941 | 335359 | * | 0.05059728 | GAB3   | chrX-155085964-155087918 | 1.680416 | 4.643823e-02 | K562 | chrX-155085964-155087918_GAB3   |\n",
       "| chrX | 154751583 | 155216308 | 464726 | * | 0.05497587 | GAB3   | chrX-155215583-155217032 | 1.949908 | 2.559355e-02 | K562 | chrX-155215583-155217032_GAB3   |\n",
       "| chrX | 154751583 | 155218972 | 467390 | * | 0.05779409 | GAB3   | chrX-155218826-155219117 | 2.944411 | 1.617849e-03 | K562 | chrX-155218826-155219117_GAB3   |\n",
       "| chrX | 154762742 | 154763237 |    496 | * | 0.05202060 | DKC1   | chrX-154762337-154764137 | 3.311179 | 4.645181e-04 | K562 | chrX-154762337-154764137_DKC1   |\n",
       "| chrX | 154762742 | 154785456 |  22715 | * | 0.05343360 | DKC1   | chrX-154784088-154786824 | 3.408487 | 3.266206e-04 | K562 | chrX-154784088-154786824_DKC1   |\n",
       "| chrX | 154762742 | 154799249 |  36508 | * | 0.05120821 | DKC1   | chrX-154798594-154799904 | 3.182689 | 7.295712e-04 | K562 | chrX-154798594-154799904_DKC1   |\n",
       "| chrX | 154763237 | 155071420 | 308184 | * | 0.06043059 | BRCC3  | chrX-154762337-154764137 | 2.674971 | 3.736781e-03 | K562 | chrX-154762337-154764137_BRCC3  |\n",
       "| chrX | 154776030 | 154821007 |  44978 | * | 0.05270889 | MPP1   | chrX-154775311-154776748 | 3.877542 | 5.275849e-05 | K562 | chrX-154775311-154776748_MPP1   |\n",
       "| chrX | 155264589 | 155334730 |  70142 | * | 0.05160496 | RAB39B | chrX-155333791-155335669 | 4.330416 | 7.441395e-06 | K562 | chrX-155333791-155335669_RAB39B |\n",
       "| chrX | 155334657 | 155334730 |     74 | * | 0.06069699 | CLIC2  | chrX-155333791-155335669 | 3.946840 | 3.959475e-05 | K562 | chrX-155333791-155335669_CLIC2  |\n",
       "\n"
      ],
      "text/plain": [
       "      chr  start     end       width  strand score      TargetGene\n",
       "1     chr1  778761   1000172   221412 *      0.05085411 HES4      \n",
       "2     chr1  820756   1020123   199368 *      0.05116465 AGRN      \n",
       "3     chr1  827285   1116361   289077 *      0.05464988 C1orf159  \n",
       "4     chr1  940540   1000172    59633 *      0.05057131 HES4      \n",
       "5     chr1  951869    959309     7441 *      0.05301224 NOC2L     \n",
       "6     chr1  959309   1080049   120741 *      0.05341145 NOC2L     \n",
       "7     chr1  959309   1331427   372119 *      0.06348661 NOC2L     \n",
       "8     chr1  960587   1000388    39802 *      0.05638424 KLHL17    \n",
       "9     chr1  960587   1308322   347736 *      0.05029331 KLHL17    \n",
       "10    chr1  960857   1000172    39316 *      0.05322021 HES4      \n",
       "11    chr1  966713   1000172    33460 *      0.05778342 HES4      \n",
       "12    chr1  966713   1399328   432616 *      0.07396319 CCNL2     \n",
       "13    chr1 1000172   1000388      217 *      0.06471269 HES4      \n",
       "14    chr1 1000172   1001805     1634 *      0.05643616 HES4      \n",
       "15    chr1 1000172   1020074    19903 *      0.06280860 HES4      \n",
       "16    chr1 1000172   1059154    58983 *      0.06285742 HES4      \n",
       "17    chr1 1000172   1069597    69426 *      0.05902627 HES4      \n",
       "18    chr1 1000172   1157468   157297 *      0.05741698 HES4      \n",
       "19    chr1 1000172   1273797   273626 *      0.05905527 HES4      \n",
       "20    chr1 1000172   1349468   349297 *      0.05402978 HES4      \n",
       "21    chr1 1000172   1399614   399443 *      0.05790177 HES4      \n",
       "22    chr1 1000172   1433558   433387 *      0.05050301 HES4      \n",
       "23    chr1 1000388   1116361   115974 *      0.07027118 C1orf159  \n",
       "24    chr1 1000388   1399328   398941 *      0.07830343 CCNL2     \n",
       "25    chr1 1000388   1471769   471382 *      0.07170709 ATAD3B    \n",
       "26    chr1 1001138   1157468   156331 *      0.05259096 ISG15     \n",
       "27    chr1 1001138   1208319   207182 *      0.05553247 ISG15     \n",
       "28    chr1 1001138   1318289   317152 *      0.05696939 ISG15     \n",
       "29    chr1 1001805   1399328   397524 *      0.05311551 CCNL2     \n",
       "30    chr1 1020074   1116361    96288 *      0.05795131 C1orf159  \n",
       "⋮     ⋮    ⋮         ⋮         ⋮      ⋮      ⋮          ⋮         \n",
       "21882 chrX 154436913 154490635  53723 *      0.06877406 GDI1      \n",
       "21883 chrX 154436913 154547314 110402 *      0.06671435 GDI1      \n",
       "21884 chrX 154436913 154776030 339118 *      0.05306943 GDI1      \n",
       "21885 chrX 154436913 154815477 378565 *      0.05584039 GDI1      \n",
       "21886 chrX 154436913 154815782 378870 *      0.05443369 GDI1      \n",
       "21887 chrX 154444126 154458894  14769 *      0.05041422 FAM50A    \n",
       "21888 chrX 154478948 154486670   7723 *      0.05441445 UBL4A     \n",
       "21889 chrX 154486552 154486670    119 *      0.05458454 UBL4A     \n",
       "21890 chrX 154486552 154751583 265032 *      0.06137794 GAB3      \n",
       "21891 chrX 154486670 154516145  29476 *      0.05149503 UBL4A     \n",
       "21892 chrX 154486670 154763237 276568 *      0.05075481 UBL4A     \n",
       "21893 chrX 154486670 154799249 312580 *      0.05313110 UBL4A     \n",
       "21894 chrX 154495942 154751583 255642 *      0.05310293 GAB3      \n",
       "21895 chrX 154516242 154547314  31073 *      0.05218297 FAM3A     \n",
       "21896 chrX 154547314 154821007 273694 *      0.05301616 MPP1      \n",
       "21897 chrX 154751337 154751583    247 *      0.08521963 GAB3      \n",
       "21898 chrX 154751337 154762742  11406 *      0.06416246 DKC1      \n",
       "21899 chrX 154751583 154763237  11655 *      0.06442513 GAB3      \n",
       "21900 chrX 154751583 154778206  26624 *      0.05562112 GAB3      \n",
       "21901 chrX 154751583 155071449 319867 *      0.06132955 GAB3      \n",
       "21902 chrX 154751583 155086941 335359 *      0.05059728 GAB3      \n",
       "21903 chrX 154751583 155216308 464726 *      0.05497587 GAB3      \n",
       "21904 chrX 154751583 155218972 467390 *      0.05779409 GAB3      \n",
       "21905 chrX 154762742 154763237    496 *      0.05202060 DKC1      \n",
       "21906 chrX 154762742 154785456  22715 *      0.05343360 DKC1      \n",
       "21907 chrX 154762742 154799249  36508 *      0.05120821 DKC1      \n",
       "21908 chrX 154763237 155071420 308184 *      0.06043059 BRCC3     \n",
       "21909 chrX 154776030 154821007  44978 *      0.05270889 MPP1      \n",
       "21910 chrX 155264589 155334730  70142 *      0.05160496 RAB39B    \n",
       "21911 chrX 155334657 155334730     74 *      0.06069699 CLIC2     \n",
       "      PeakName                 zscore   pvalue       CellType\n",
       "1     chr1-778120-779401       2.343423 9.553860e-03 K562    \n",
       "2     chr1-819819-821692       3.093901 9.877161e-04 K562    \n",
       "3     chr1-826642-827928       1.926267 2.703555e-02 K562    \n",
       "4     chr1-939995-941085       1.799467 3.597238e-02 K562    \n",
       "5     chr1-950610-953128       3.453538 2.766421e-04 K562    \n",
       "6     chr1-1079353-1080745     3.384546 3.564810e-04 K562    \n",
       "7     chr1-1330543-1332311     4.451339 4.266819e-06 K562    \n",
       "8     chr1-999557-1001218      2.216227 1.333799e-02 K562    \n",
       "9     chr1-1307532-1309112     2.078736 1.882081e-02 K562    \n",
       "10    chr1-960264-961450       2.362046 9.087205e-03 K562    \n",
       "11    chr1-965905-967520       2.267999 1.166463e-02 K562    \n",
       "12    chr1-965905-967520       2.610996 4.513943e-03 K562    \n",
       "13    chr1-999557-1001218      2.563924 5.174803e-03 K562    \n",
       "14    chr1-1001483-1002126     2.218792 1.325043e-02 K562    \n",
       "15    chr1-1019151-1020996     2.586115 4.853221e-03 K562    \n",
       "16    chr1-1058323-1059984     2.356170 9.232227e-03 K562    \n",
       "17    chr1-1068523-1070670     2.242326 1.247016e-02 K562    \n",
       "18    chr1-1156382-1158554     2.100162 1.785731e-02 K562    \n",
       "19    chr1-1272810-1274783     2.141628 1.611173e-02 K562    \n",
       "20    chr1-1348905-1350031     2.400730 8.181207e-03 K562    \n",
       "21    chr1-1398814-1400413     2.060970 1.965296e-02 K562    \n",
       "22    chr1-1433049-1434067     2.027773 2.129171e-02 K562    \n",
       "23    chr1-999557-1001218      2.074273 1.902699e-02 K562    \n",
       "24    chr1-999557-1001218      2.746377 3.012878e-03 K562    \n",
       "25    chr1-999557-1001218      3.215316 6.515039e-04 K562    \n",
       "26    chr1-1156382-1158554     2.883984 1.963391e-03 K562    \n",
       "27    chr1-1208012-1208626     3.650349 1.309423e-04 K562    \n",
       "28    chr1-1317926-1318651     3.921849 4.393597e-05 K562    \n",
       "29    chr1-1001483-1002126     1.831572 3.350764e-02 K562    \n",
       "30    chr1-1019151-1020996     1.730659 4.175630e-02 K562    \n",
       "⋮     ⋮                        ⋮        ⋮            ⋮       \n",
       "21882 chrX-154490012-154491257 2.032247 2.106431e-02 K562    \n",
       "21883 chrX-154546671-154547957 2.149107 1.581295e-02 K562    \n",
       "21884 chrX-154775311-154776748 2.274201 1.147694e-02 K562    \n",
       "21885 chrX-154815322-154815631 2.950147 1.588115e-03 K562    \n",
       "21886 chrX-154815707-154815857 3.609951 1.531276e-04 K562    \n",
       "21887 chrX-154457558-154460229 1.936615 2.639619e-02 K562    \n",
       "21888 chrX-154478219-154479676 1.777295 3.775985e-02 K562    \n",
       "21889 chrX-154486006-154487098 1.966755 2.460571e-02 K562    \n",
       "21890 chrX-154486006-154487098 2.026130 2.137572e-02 K562    \n",
       "21891 chrX-154515321-154516969 1.768604 3.848000e-02 K562    \n",
       "21892 chrX-154762337-154764137 2.139178 1.621063e-02 K562    \n",
       "21893 chrX-154798594-154799904 2.681901 3.660254e-03 K562    \n",
       "21894 chrX-154495283-154496601 1.925464 2.708567e-02 K562    \n",
       "21895 chrX-154546671-154547957 2.289939 1.101242e-02 K562    \n",
       "21896 chrX-154546671-154547957 3.308351 4.692361e-04 K562    \n",
       "21897 chrX-154750373-154752301 3.520915 2.150301e-04 K562    \n",
       "21898 chrX-154750373-154752301 4.162127 1.576482e-05 K562    \n",
       "21899 chrX-154762337-154764137 2.625673 4.323896e-03 K562    \n",
       "21900 chrX-154777776-154778636 2.127016 1.670937e-02 K562    \n",
       "21901 chrX-155070725-155072173 1.738662 4.104714e-02 K562    \n",
       "21902 chrX-155085964-155087918 1.680416 4.643823e-02 K562    \n",
       "21903 chrX-155215583-155217032 1.949908 2.559355e-02 K562    \n",
       "21904 chrX-155218826-155219117 2.944411 1.617849e-03 K562    \n",
       "21905 chrX-154762337-154764137 3.311179 4.645181e-04 K562    \n",
       "21906 chrX-154784088-154786824 3.408487 3.266206e-04 K562    \n",
       "21907 chrX-154798594-154799904 3.182689 7.295712e-04 K562    \n",
       "21908 chrX-154762337-154764137 2.674971 3.736781e-03 K562    \n",
       "21909 chrX-154775311-154776748 3.877542 5.275849e-05 K562    \n",
       "21910 chrX-155333791-155335669 4.330416 7.441395e-06 K562    \n",
       "21911 chrX-155333791-155335669 3.946840 3.959475e-05 K562    \n",
       "      PairName                       \n",
       "1     chr1-778120-779401_HES4        \n",
       "2     chr1-819819-821692_AGRN        \n",
       "3     chr1-826642-827928_C1orf159    \n",
       "4     chr1-939995-941085_HES4        \n",
       "5     chr1-950610-953128_NOC2L       \n",
       "6     chr1-1079353-1080745_NOC2L     \n",
       "7     chr1-1330543-1332311_NOC2L     \n",
       "8     chr1-999557-1001218_KLHL17     \n",
       "9     chr1-1307532-1309112_KLHL17    \n",
       "10    chr1-960264-961450_HES4        \n",
       "11    chr1-965905-967520_HES4        \n",
       "12    chr1-965905-967520_CCNL2       \n",
       "13    chr1-999557-1001218_HES4       \n",
       "14    chr1-1001483-1002126_HES4      \n",
       "15    chr1-1019151-1020996_HES4      \n",
       "16    chr1-1058323-1059984_HES4      \n",
       "17    chr1-1068523-1070670_HES4      \n",
       "18    chr1-1156382-1158554_HES4      \n",
       "19    chr1-1272810-1274783_HES4      \n",
       "20    chr1-1348905-1350031_HES4      \n",
       "21    chr1-1398814-1400413_HES4      \n",
       "22    chr1-1433049-1434067_HES4      \n",
       "23    chr1-999557-1001218_C1orf159   \n",
       "24    chr1-999557-1001218_CCNL2      \n",
       "25    chr1-999557-1001218_ATAD3B     \n",
       "26    chr1-1156382-1158554_ISG15     \n",
       "27    chr1-1208012-1208626_ISG15     \n",
       "28    chr1-1317926-1318651_ISG15     \n",
       "29    chr1-1001483-1002126_CCNL2     \n",
       "30    chr1-1019151-1020996_C1orf159  \n",
       "⋮     ⋮                              \n",
       "21882 chrX-154490012-154491257_GDI1  \n",
       "21883 chrX-154546671-154547957_GDI1  \n",
       "21884 chrX-154775311-154776748_GDI1  \n",
       "21885 chrX-154815322-154815631_GDI1  \n",
       "21886 chrX-154815707-154815857_GDI1  \n",
       "21887 chrX-154457558-154460229_FAM50A\n",
       "21888 chrX-154478219-154479676_UBL4A \n",
       "21889 chrX-154486006-154487098_UBL4A \n",
       "21890 chrX-154486006-154487098_GAB3  \n",
       "21891 chrX-154515321-154516969_UBL4A \n",
       "21892 chrX-154762337-154764137_UBL4A \n",
       "21893 chrX-154798594-154799904_UBL4A \n",
       "21894 chrX-154495283-154496601_GAB3  \n",
       "21895 chrX-154546671-154547957_FAM3A \n",
       "21896 chrX-154546671-154547957_MPP1  \n",
       "21897 chrX-154750373-154752301_GAB3  \n",
       "21898 chrX-154750373-154752301_DKC1  \n",
       "21899 chrX-154762337-154764137_GAB3  \n",
       "21900 chrX-154777776-154778636_GAB3  \n",
       "21901 chrX-155070725-155072173_GAB3  \n",
       "21902 chrX-155085964-155087918_GAB3  \n",
       "21903 chrX-155215583-155217032_GAB3  \n",
       "21904 chrX-155218826-155219117_GAB3  \n",
       "21905 chrX-154762337-154764137_DKC1  \n",
       "21906 chrX-154784088-154786824_DKC1  \n",
       "21907 chrX-154798594-154799904_DKC1  \n",
       "21908 chrX-154762337-154764137_BRCC3 \n",
       "21909 chrX-154775311-154776748_MPP1  \n",
       "21910 chrX-155333791-155335669_RAB39B\n",
       "21911 chrX-155333791-155335669_CLIC2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(obj.seurat@assays$ATAC@links)\n",
    "colnames(df.output) = c('chr',\n",
    "                        'start',\n",
    "                        'end',\n",
    "                        'width',\n",
    "                        'strand',\n",
    "                        'score',\n",
    "                        'TargetGene',\n",
    "                        'PeakName',\n",
    "                        'zscore',\n",
    "                        'pvalue')\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output[,\"PairName\"] = paste(df.output[,\"PeakName\"],\n",
    "                               df.output[,\"TargetGene\"],\n",
    "                               sep = \"_\")\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1932e7b2-e909-4df3-8097-650f387c9ea2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] data.table_1.15.2                 dplyr_1.1.4                      \n",
       " [3] foreach_1.5.2                     Matrix_1.6-5                     \n",
       " [5] EnsDb.Hsapiens.v86_2.99.0         ensembldb_2.24.1                 \n",
       " [7] AnnotationFilter_1.24.0           GenomicFeatures_1.52.2           \n",
       " [9] AnnotationDbi_1.64.1              Biobase_2.60.0                   \n",
       "[11] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.68.0                  \n",
       "[13] rtracklayer_1.60.1                Biostrings_2.70.1                \n",
       "[15] XVector_0.42.0                    GenomicRanges_1.54.1             \n",
       "[17] GenomeInfoDb_1.38.1               IRanges_2.36.0                   \n",
       "[19] S4Vectors_0.40.2                  BiocGenerics_0.48.1              \n",
       "[21] genomation_1.34.0                 Signac_1.13.0                    \n",
       "[23] Seurat_5.0.3                      SeuratObject_5.0.1               \n",
       "[25] sp_2.1-4                         \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               filelock_1.0.3             \n",
       "  [7] bitops_1.0-7                tibble_3.2.1               \n",
       "  [9] polyclip_1.10-6             rpart_4.1.23               \n",
       " [11] XML_3.99-0.16.1             fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             globals_0.16.3             \n",
       " [15] lattice_0.22-6              MASS_7.3-60                \n",
       " [17] backports_1.4.1             magrittr_2.0.3             \n",
       " [19] rmarkdown_2.26              Hmisc_5.1-2                \n",
       " [21] plotly_4.10.4               yaml_2.3.8                 \n",
       " [23] plotrix_3.8-4               httpuv_1.6.15              \n",
       " [25] sctransform_0.4.1           spam_2.10-0                \n",
       " [27] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [29] cowplot_1.1.3               pbapply_1.7-2              \n",
       " [31] DBI_1.2.2                   RColorBrewer_1.1-3         \n",
       " [33] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [35] Rtsne_0.17                  purrr_1.0.2                \n",
       " [37] biovizBase_1.48.0           RCurl_1.98-1.14            \n",
       " [39] nnet_7.3-19                 VariantAnnotation_1.46.0   \n",
       " [41] rappdirs_0.3.3              GenomeInfoDbData_1.2.11    \n",
       " [43] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [45] listenv_0.9.1               spatstat.utils_3.0-4       \n",
       " [47] goftest_1.2-3               RSpectra_0.16-1            \n",
       " [49] spatstat.random_3.2-3       fitdistrplus_1.1-11        \n",
       " [51] parallelly_1.37.1           leiden_0.4.3.1             \n",
       " [53] codetools_0.2-20            DelayedArray_0.26.7        \n",
       " [55] RcppRoll_0.3.0              xml2_1.3.6                 \n",
       " [57] tidyselect_1.2.1            BiocFileCache_2.8.0        \n",
       " [59] matrixStats_1.3.0           base64enc_0.1-3            \n",
       " [61] spatstat.explore_3.2-6      GenomicAlignments_1.36.0   \n",
       " [63] jsonlite_1.8.8              Formula_1.2-5              \n",
       " [65] progressr_0.14.0            iterators_1.0.14           \n",
       " [67] ggridges_0.5.6              survival_3.6-4             \n",
       " [69] progress_1.2.3              tools_4.3.3                \n",
       " [71] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [73] glue_1.7.0                  gridExtra_2.3              \n",
       " [75] xfun_0.43                   MatrixGenerics_1.12.3      \n",
       " [77] IRdisplay_1.1               withr_3.0.0                \n",
       " [79] fastmap_1.1.1               fansi_1.0.6                \n",
       " [81] digest_0.6.35               R6_2.5.1                   \n",
       " [83] mime_0.12                   seqPattern_1.34.0          \n",
       " [85] colorspace_2.1-0            scattermore_1.2            \n",
       " [87] tensor_1.5                  dichromat_2.0-0.1          \n",
       " [89] biomaRt_2.56.1              RSQLite_2.3.6              \n",
       " [91] spatstat.data_3.0-4         utf8_1.2.4                 \n",
       " [93] tidyr_1.3.1                 generics_0.1.3             \n",
       " [95] prettyunits_1.2.0           httr_1.4.7                 \n",
       " [97] htmlwidgets_1.6.4           S4Arrays_1.2.1             \n",
       " [99] uwot_0.1.16                 pkgconfig_2.0.3            \n",
       "[101] gtable_0.3.5                blob_1.2.4                 \n",
       "[103] lmtest_0.9-40               impute_1.76.0              \n",
       "[105] htmltools_0.5.8.1           dotCall64_1.1-1            \n",
       "[107] ProtGenerics_1.32.0         scales_1.3.0               \n",
       "[109] png_0.1-8                   rstudioapi_0.16.0          \n",
       "[111] knitr_1.46                  tzdb_0.4.0                 \n",
       "[113] reshape2_1.4.4              rjson_0.2.21               \n",
       "[115] uuid_1.2-0                  checkmate_2.3.1            \n",
       "[117] curl_5.1.0                  nlme_3.1-164               \n",
       "[119] cachem_1.0.8                repr_1.1.7                 \n",
       "[121] zoo_1.8-12                  stringr_1.5.1              \n",
       "[123] KernSmooth_2.23-22          miniUI_0.1.1.1             \n",
       "[125] foreign_0.8-86              restfulr_0.0.15            \n",
       "[127] pillar_1.9.0                vctrs_0.6.5                \n",
       "[129] RANN_2.6.1                  promises_1.3.0             \n",
       "[131] dbplyr_2.5.0                xtable_1.8-4               \n",
       "[133] cluster_2.1.6               htmlTable_2.4.2            \n",
       "[135] evaluate_0.23               readr_2.1.5                \n",
       "[137] cli_3.6.2                   compiler_4.3.3             \n",
       "[139] Rsamtools_2.18.0            rlang_1.1.3                \n",
       "[141] crayon_1.5.2                future.apply_1.11.2        \n",
       "[143] plyr_1.8.9                  stringi_1.8.3              \n",
       "[145] viridisLite_0.4.2           deldir_2.0-4               \n",
       "[147] gridBase_0.4-7              BiocParallel_1.36.0        \n",
       "[149] munsell_0.5.1               lazyeval_0.2.2             \n",
       "[151] spatstat.geom_3.2-9         IRkernel_1.3.2             \n",
       "[153] RcppHNSW_0.6.0              hms_1.1.3                  \n",
       "[155] patchwork_1.2.0             bit64_4.0.5                \n",
       "[157] future_1.33.2               ggplot2_3.5.1              \n",
       "[159] KEGGREST_1.40.1             shiny_1.8.1.1              \n",
       "[161] SummarizedExperiment_1.30.2 ROCR_1.0-11                \n",
       "[163] memoise_2.0.1               igraph_2.0.3               \n",
       "[165] fastmatch_1.1-4             bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1c849d6-1307-4b9e-a1b2-2d1ae9445932",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
