{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/HCT116/HCT116/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/HCT116/HCT116/multiome_powerlaw_v3/encode_e2g_predictions.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/HCT116/HCT116/Kendall/atac_matrix.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/scE2G_input_Maya/HCT116/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/HCT116/1.Genome_wide_prediction/FigR/FigR.250321/\"\n",
    "celltype = \"HCT116\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2938315 ranges and 3 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100034595-100034763      * |         AGL\n",
       "        [2]     chr1 100034595-100034763      * |      CDC14A\n",
       "        [3]     chr1 100034595-100034763      * |         DBT\n",
       "        [4]     chr1 100034595-100034763      * |        DPH5\n",
       "        [5]     chr1 100034595-100034763      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2938311]     chrX     9911462-9911841      * |        WWC3\n",
       "  [2938312]     chrX     9995795-9996095      * |       CLCN4\n",
       "  [2938313]     chrX     9995795-9996095      * |       FAM9B\n",
       "  [2938314]     chrX     9995795-9996095      * |       TBL1X\n",
       "  [2938315]     chrX     9995795-9996095      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-100034595-10003.. chr1-100034595-10003..\n",
       "        [2] chr1-100034595-10003.. chr1-100034595-10003..\n",
       "        [3] chr1-100034595-10003.. chr1-100034595-10003..\n",
       "        [4] chr1-100034595-10003.. chr1-100034595-10003..\n",
       "        [5] chr1-100034595-10003.. chr1-100034595-10003..\n",
       "        ...                    ...                    ...\n",
       "  [2938311]   chrX-9911462-9911841 chrX-9911462-9911841..\n",
       "  [2938312]   chrX-9995795-9996095 chrX-9995795-9996095..\n",
       "  [2938313]   chrX-9995795-9996095 chrX-9995795-9996095..\n",
       "  [2938314]   chrX-9995795-9996095 chrX-9995795-9996095..\n",
       "  [2938315]   chrX-9995795-9996095 chrX-9995795-9996095..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  4000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  4001  to  4825 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dd80eda-2392-4d59-8250-1ae62d7e1d1e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b9e26c70-11fd-47c4-8e05-9c96b17d3f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a76b1603-6b00-4ac0-9399-fd341c5a7b88",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 153354 4825 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(153354): chr1-100034595-100034763 chr1-100037727-100038876 ...\n",
       "  chrX-9911462-9911841 chrX-9995795-9996095\n",
       "rowData names(2): PeakName bias\n",
       "colnames(4825): e18l1_AAACCGGTCGCCTGTG-1 e18l1_AAACCGGTCGTTGTCA-1 ...\n",
       "  e18l4_TTTCAGTAGTGAGGCC-1 e18l4_TTTGCTCTCCGCCATG-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna.rename),\n",
    "                       index = 1:nrow(matrix.rna.rename))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "Sys.setenv(OPENBLAS_NUM_THREADS=1)\n",
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  14.4301578998566 secs \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Time Elapsed:  26.8420498728752 mins \n",
      "\n",
      "[1] \"chr10\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.62695264816284 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.43612165053685 mins \n",
      "\n",
      "[1] \"chr11\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.5477042198181 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  18.503330330054 mins \n",
      "\n",
      "[1] \"chr12\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.64825892448425 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  13.7046528379122 mins \n",
      "\n",
      "[1] \"chr13\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.36532211303711 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  4.3112881064415 mins \n",
      "\n",
      "[1] \"chr14\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.32240796089172 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.73857932885488 mins \n",
      "\n",
      "[1] \"chr15\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.03470659255981 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.31312845945358 mins \n",
      "\n",
      "[1] \"chr16\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  12.8429148197174 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  20.8745757420858 mins \n",
      "\n",
      "[1] \"chr17\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  17.1247696876526 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  29.2427265286446 mins \n",
      "\n",
      "[1] \"chr18\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.5071165561676 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  4.41781097253164 mins \n",
      "\n",
      "[1] \"chr19\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  18.6043479442596 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  31.8640242775281 mins \n",
      "\n",
      "[1] \"chr2\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.65208697319031 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  14.8154784917831 mins \n",
      "\n",
      "[1] \"chr20\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.16059017181396 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  13.4507542451223 mins \n",
      "\n",
      "[1] \"chr21\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.58098316192627 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  4.44366375605265 mins \n",
      "\n",
      "[1] \"chr22\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.0627818107605 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.8522477666537 mins \n",
      "\n",
      "[1] \"chr3\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.78722453117371 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  13.0447757879893 mins \n",
      "\n",
      "[1] \"chr4\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.75123333930969 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.41826679309209 mins \n",
      "\n",
      "[1] \"chr5\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.98326110839844 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  10.2798255205154 mins \n",
      "\n",
      "[1] \"chr6\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.50509786605835 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  12.4158242901166 mins \n",
      "\n",
      "[1] \"chr7\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.72415614128113 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  12.514133254687 mins \n",
      "\n",
      "[1] \"chr8\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.52401399612427 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.2052518407504 mins \n",
      "\n",
      "[1] \"chr9\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.90440344810486 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  13.8933863838514 mins \n",
      "\n",
      "[1] \"chrX\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.3862521648407 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.0125435312589 mins \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna.rename,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2843332 ranges and 7 metadata columns:\n",
       "       seqnames              ranges strand |  TargetGene               PeakName\n",
       "          <Rle>           <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 100034595-100034763      * |         AGL chr1-100034595-10003..\n",
       "  chr1     chr1 100034595-100034763      * |      CDC14A chr1-100034595-10003..\n",
       "  chr1     chr1 100034595-100034763      * |         DBT chr1-100034595-10003..\n",
       "  chr1     chr1 100034595-100034763      * |        DPH5 chr1-100034595-10003..\n",
       "  chr1     chr1 100034595-100034763      * |       EXTL2 chr1-100034595-10003..\n",
       "   ...      ...                 ...    ... .         ...                    ...\n",
       "  chrX     chrX     9911462-9911841      * |        WWC3   chrX-9911462-9911841\n",
       "  chrX     chrX     9995795-9996095      * |       CLCN4   chrX-9995795-9996095\n",
       "  chrX     chrX     9995795-9996095      * |       FAM9B   chrX-9995795-9996095\n",
       "  chrX     chrX     9995795-9996095      * |       TBL1X   chrX-9995795-9996095\n",
       "  chrX     chrX     9995795-9996095      * |        WWC3   chrX-9995795-9996095\n",
       "                     PairName         rObs     rBgSD     rBgMean     pvalZ\n",
       "                  <character>    <numeric> <numeric>   <numeric> <numeric>\n",
       "  chr1 chr1-100034595-10003..  0.005040022 0.0161666 0.007302964  0.555661\n",
       "  chr1 chr1-100034595-10003..  0.014188951 0.0139871 0.000293221  0.160242\n",
       "  chr1 chr1-100034595-10003.. -0.000123853 0.0136493 0.004784390  0.640426\n",
       "  chr1 chr1-100034595-10003.. -0.016393486 0.0158015 0.003836168  0.899768\n",
       "  chr1 chr1-100034595-10003.. -0.006810514 0.0162936 0.006676814  0.796099\n",
       "   ...                    ...          ...       ...         ...       ...\n",
       "  chrX chrX-9911462-9911841..  -0.02098700 0.0171618  0.00673206  0.946862\n",
       "  chrX chrX-9995795-9996095..  -0.01419599 0.0141376  0.00311088  0.889557\n",
       "  chrX chrX-9995795-9996095..  -0.00445570 0.0133737 -0.00118103  0.596717\n",
       "  chrX chrX-9995795-9996095..  -0.00501025 0.0165579  0.00257623  0.676588\n",
       "  chrX chrX-9995795-9996095..  -0.01232410 0.0161614  0.00811787  0.897040\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2843332 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>AGL     </td><td>HCT116</td><td> 0.0050400223</td><td>0.01616658</td><td>0.0073029637</td><td>0.55566072</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>CDC14A  </td><td>HCT116</td><td> 0.0141889509</td><td>0.01398714</td><td>0.0002932206</td><td>0.16024180</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>DBT     </td><td>HCT116</td><td>-0.0001238528</td><td>0.01364927</td><td>0.0047843902</td><td>0.64042590</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>DPH5    </td><td>HCT116</td><td>-0.0163934858</td><td>0.01580155</td><td>0.0038361685</td><td>0.89976829</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>EXTL2   </td><td>HCT116</td><td>-0.0068105141</td><td>0.01629358</td><td>0.0066768142</td><td>0.79609945</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>FRRS1   </td><td>HCT116</td><td>-0.0180925943</td><td>0.01434001</td><td>0.0034641909</td><td>0.93361414</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>MFSD14A </td><td>HCT116</td><td> 0.0178185181</td><td>0.01472433</td><td>0.0030524887</td><td>0.15797094</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>PLPPR4  </td><td>HCT116</td><td>-0.0089699035</td><td>0.01575691</td><td>0.0010479894</td><td>0.73753940</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>RTCA    </td><td>HCT116</td><td> 0.0295133880</td><td>0.01580357</td><td>0.0053641107</td><td>0.06324513</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>RTCA-AS1</td><td>HCT116</td><td>-0.0147241072</td><td>0.01558145</td><td>0.0040766378</td><td>0.88620889</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>SASS6   </td><td>HCT116</td><td>-0.0039188789</td><td>0.01551349</td><td>0.0028756640</td><td>0.66929829</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>SLC30A7 </td><td>HCT116</td><td>-0.0357777822</td><td>0.01419918</td><td>0.0047461359</td><td>0.99784112</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>SLC35A3 </td><td>HCT116</td><td> 0.0093719800</td><td>0.01606637</td><td>0.0027847697</td><td>0.34090299</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>TRMT13  </td><td>HCT116</td><td> 0.0129998407</td><td>0.01562570</td><td>0.0036223934</td><td>0.27420987</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>AGL     </td><td>HCT116</td><td> 0.0243556687</td><td>0.01569431</td><td>0.0101133694</td><td>0.18207599</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>CDC14A  </td><td>HCT116</td><td> 0.0152480782</td><td>0.01329053</td><td>0.0054948536</td><td>0.23152081</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>DBT     </td><td>HCT116</td><td> 0.0152141760</td><td>0.01405810</td><td>0.0115274866</td><td>0.39656563</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>DPH5    </td><td>HCT116</td><td> 0.0024098791</td><td>0.01333417</td><td>0.0082324553</td><td>0.66882315</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>EXTL2   </td><td>HCT116</td><td> 0.0164913521</td><td>0.01440219</td><td>0.0158709666</td><td>0.48282056</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>FRRS1   </td><td>HCT116</td><td>-0.0022498411</td><td>0.01266587</td><td>0.0031121802</td><td>0.66397784</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>MFSD14A </td><td>HCT116</td><td> 0.0067903419</td><td>0.01395098</td><td>0.0126126170</td><td>0.66178442</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>PLPPR4  </td><td>HCT116</td><td>-0.0005688027</td><td>0.01556156</td><td>0.0050089879</td><td>0.63999066</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>RTCA    </td><td>HCT116</td><td> 0.0040791528</td><td>0.01307826</td><td>0.0107840530</td><td>0.69591073</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>RTCA-AS1</td><td>HCT116</td><td>-0.0249087958</td><td>0.01460240</td><td>0.0010782260</td><td>0.96243264</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>SASS6   </td><td>HCT116</td><td> 0.0294145880</td><td>0.01736045</td><td>0.0222455789</td><td>0.33982139</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>SLC30A7 </td><td>HCT116</td><td>-0.0097983219</td><td>0.01463933</td><td>0.0142596526</td><td>0.94984781</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>SLC35A3 </td><td>HCT116</td><td> 0.0099716939</td><td>0.01420939</td><td>0.0125128735</td><td>0.57096754</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>TRMT13  </td><td>HCT116</td><td> 0.0070494852</td><td>0.01270121</td><td>0.0118967716</td><td>0.64863574</td></tr>\n",
       "\t<tr><td>chr1</td><td>100046025</td><td>100046641</td><td>AGL     </td><td>HCT116</td><td> 0.0031032834</td><td>0.01524022</td><td>0.0061113677</td><td>0.57823413</td></tr>\n",
       "\t<tr><td>chr1</td><td>100046025</td><td>100046641</td><td>CDC14A  </td><td>HCT116</td><td>-0.0121750800</td><td>0.01397360</td><td>0.0006327480</td><td>0.82031689</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>ASMTL    </td><td>HCT116</td><td> 0.003952418</td><td>0.0148517286</td><td> 0.0093814987</td><td>0.6426503303</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>CSF2RA   </td><td>HCT116</td><td> 0.003445228</td><td>0.0157337318</td><td> 0.0041007269</td><td>0.5166159325</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>GTPBP6   </td><td>HCT116</td><td>-0.011086705</td><td>0.0140485327</td><td> 0.0032285525</td><td>0.8458951799</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>IL3RA    </td><td>HCT116</td><td> 0.014231312</td><td>0.0150745427</td><td> 0.0023006160</td><td>0.2143417103</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>LINC00106</td><td>HCT116</td><td> 0.015691484</td><td>0.0152868696</td><td> 0.0062515992</td><td>0.2684472635</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>P2RY8    </td><td>HCT116</td><td>-0.011085038</td><td>0.0139393423</td><td> 0.0005890353</td><td>0.7988416696</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>PLCXD1   </td><td>HCT116</td><td> 0.013712258</td><td>0.0142326486</td><td> 0.0022705909</td><td>0.2107265377</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>PPP2R3B  </td><td>HCT116</td><td> 0.014587699</td><td>0.0146551250</td><td> 0.0060971991</td><td>0.2811752823</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>SHOX     </td><td>HCT116</td><td>-0.004951215</td><td>0.0128875109</td><td>-0.0007166742</td><td>0.6287623125</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>SLC25A6  </td><td>HCT116</td><td> 0.006951821</td><td>0.0142101164</td><td> 0.0059842131</td><td>0.4728558518</td></tr>\n",
       "\t<tr><td>chrX</td><td>9902739</td><td>9902992</td><td>CLCN4    </td><td>HCT116</td><td> 0.014815425</td><td>0.0160579607</td><td> 0.0022118474</td><td>0.2162618608</td></tr>\n",
       "\t<tr><td>chrX</td><td>9902739</td><td>9902992</td><td>FAM9B    </td><td>HCT116</td><td>-0.002758100</td><td>0.0206972171</td><td> 0.0032270809</td><td>0.6137774363</td></tr>\n",
       "\t<tr><td>chrX</td><td>9902739</td><td>9902992</td><td>TBL1X    </td><td>HCT116</td><td>-0.012858781</td><td>0.0153783700</td><td> 0.0007509701</td><td>0.8119198275</td></tr>\n",
       "\t<tr><td>chrX</td><td>9902739</td><td>9902992</td><td>WWC3     </td><td>HCT116</td><td> 0.007937758</td><td>0.0148231419</td><td> 0.0046010171</td><td>0.4109493765</td></tr>\n",
       "\t<tr><td>chrX</td><td>9903251</td><td>9903473</td><td>CLCN4    </td><td>HCT116</td><td>-0.007720546</td><td>0.0156064150</td><td> 0.0008304726</td><td>0.7081255092</td></tr>\n",
       "\t<tr><td>chrX</td><td>9903251</td><td>9903473</td><td>FAM9B    </td><td>HCT116</td><td>-0.002423248</td><td>0.0001896353</td><td>-0.0024936829</td><td>0.3551606360</td></tr>\n",
       "\t<tr><td>chrX</td><td>9903251</td><td>9903473</td><td>TBL1X    </td><td>HCT116</td><td>-0.017486091</td><td>0.0131449755</td><td> 0.0019882760</td><td>0.9307641791</td></tr>\n",
       "\t<tr><td>chrX</td><td>9903251</td><td>9903473</td><td>WWC3     </td><td>HCT116</td><td> 0.042423638</td><td>0.0182114220</td><td> 0.0057069685</td><td>0.0218929746</td></tr>\n",
       "\t<tr><td>chrX</td><td>9906862</td><td>9908314</td><td>CLCN4    </td><td>HCT116</td><td> 0.004822346</td><td>0.0150428244</td><td> 0.0035527851</td><td>0.4663706216</td></tr>\n",
       "\t<tr><td>chrX</td><td>9906862</td><td>9908314</td><td>FAM9B    </td><td>HCT116</td><td>-0.007197606</td><td>0.0143622836</td><td> 0.0003927634</td><td>0.7014214644</td></tr>\n",
       "\t<tr><td>chrX</td><td>9906862</td><td>9908314</td><td>TBL1X    </td><td>HCT116</td><td>-0.014559052</td><td>0.0134908173</td><td> 0.0081316420</td><td>0.9537093919</td></tr>\n",
       "\t<tr><td>chrX</td><td>9906862</td><td>9908314</td><td>WWC3     </td><td>HCT116</td><td> 0.027832196</td><td>0.0137228171</td><td> 0.0113309466</td><td>0.1145911112</td></tr>\n",
       "\t<tr><td>chrX</td><td>9911462</td><td>9911841</td><td>CLCN4    </td><td>HCT116</td><td>-0.010609007</td><td>0.0174102374</td><td> 0.0032912636</td><td>0.7876797457</td></tr>\n",
       "\t<tr><td>chrX</td><td>9911462</td><td>9911841</td><td>FAM9B    </td><td>HCT116</td><td> 0.059898809</td><td>0.0178920487</td><td> 0.0023088041</td><td>0.0006437568</td></tr>\n",
       "\t<tr><td>chrX</td><td>9911462</td><td>9911841</td><td>TBL1X    </td><td>HCT116</td><td>-0.009438068</td><td>0.0140744591</td><td> 0.0025539527</td><td>0.8029044300</td></tr>\n",
       "\t<tr><td>chrX</td><td>9911462</td><td>9911841</td><td>WWC3     </td><td>HCT116</td><td>-0.020986996</td><td>0.0171618490</td><td> 0.0067320633</td><td>0.9468615151</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995795</td><td>9996095</td><td>CLCN4    </td><td>HCT116</td><td>-0.014195992</td><td>0.0141375659</td><td> 0.0031108815</td><td>0.8895571398</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995795</td><td>9996095</td><td>FAM9B    </td><td>HCT116</td><td>-0.004455696</td><td>0.0133737329</td><td>-0.0011810341</td><td>0.5967167061</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995795</td><td>9996095</td><td>TBL1X    </td><td>HCT116</td><td>-0.005010254</td><td>0.0165578613</td><td> 0.0025762293</td><td>0.6765884943</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995795</td><td>9996095</td><td>WWC3     </td><td>HCT116</td><td>-0.012324096</td><td>0.0161613571</td><td> 0.0081178653</td><td>0.8970404297</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2843332 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100034595 & 100034763 & AGL      & HCT116 &  0.0050400223 & 0.01616658 & 0.0073029637 & 0.55566072\\\\\n",
       "\t chr1 & 100034595 & 100034763 & CDC14A   & HCT116 &  0.0141889509 & 0.01398714 & 0.0002932206 & 0.16024180\\\\\n",
       "\t chr1 & 100034595 & 100034763 & DBT      & HCT116 & -0.0001238528 & 0.01364927 & 0.0047843902 & 0.64042590\\\\\n",
       "\t chr1 & 100034595 & 100034763 & DPH5     & HCT116 & -0.0163934858 & 0.01580155 & 0.0038361685 & 0.89976829\\\\\n",
       "\t chr1 & 100034595 & 100034763 & EXTL2    & HCT116 & -0.0068105141 & 0.01629358 & 0.0066768142 & 0.79609945\\\\\n",
       "\t chr1 & 100034595 & 100034763 & FRRS1    & HCT116 & -0.0180925943 & 0.01434001 & 0.0034641909 & 0.93361414\\\\\n",
       "\t chr1 & 100034595 & 100034763 & MFSD14A  & HCT116 &  0.0178185181 & 0.01472433 & 0.0030524887 & 0.15797094\\\\\n",
       "\t chr1 & 100034595 & 100034763 & PLPPR4   & HCT116 & -0.0089699035 & 0.01575691 & 0.0010479894 & 0.73753940\\\\\n",
       "\t chr1 & 100034595 & 100034763 & RTCA     & HCT116 &  0.0295133880 & 0.01580357 & 0.0053641107 & 0.06324513\\\\\n",
       "\t chr1 & 100034595 & 100034763 & RTCA-AS1 & HCT116 & -0.0147241072 & 0.01558145 & 0.0040766378 & 0.88620889\\\\\n",
       "\t chr1 & 100034595 & 100034763 & SASS6    & HCT116 & -0.0039188789 & 0.01551349 & 0.0028756640 & 0.66929829\\\\\n",
       "\t chr1 & 100034595 & 100034763 & SLC30A7  & HCT116 & -0.0357777822 & 0.01419918 & 0.0047461359 & 0.99784112\\\\\n",
       "\t chr1 & 100034595 & 100034763 & SLC35A3  & HCT116 &  0.0093719800 & 0.01606637 & 0.0027847697 & 0.34090299\\\\\n",
       "\t chr1 & 100034595 & 100034763 & TRMT13   & HCT116 &  0.0129998407 & 0.01562570 & 0.0036223934 & 0.27420987\\\\\n",
       "\t chr1 & 100037727 & 100038876 & AGL      & HCT116 &  0.0243556687 & 0.01569431 & 0.0101133694 & 0.18207599\\\\\n",
       "\t chr1 & 100037727 & 100038876 & CDC14A   & HCT116 &  0.0152480782 & 0.01329053 & 0.0054948536 & 0.23152081\\\\\n",
       "\t chr1 & 100037727 & 100038876 & DBT      & HCT116 &  0.0152141760 & 0.01405810 & 0.0115274866 & 0.39656563\\\\\n",
       "\t chr1 & 100037727 & 100038876 & DPH5     & HCT116 &  0.0024098791 & 0.01333417 & 0.0082324553 & 0.66882315\\\\\n",
       "\t chr1 & 100037727 & 100038876 & EXTL2    & HCT116 &  0.0164913521 & 0.01440219 & 0.0158709666 & 0.48282056\\\\\n",
       "\t chr1 & 100037727 & 100038876 & FRRS1    & HCT116 & -0.0022498411 & 0.01266587 & 0.0031121802 & 0.66397784\\\\\n",
       "\t chr1 & 100037727 & 100038876 & MFSD14A  & HCT116 &  0.0067903419 & 0.01395098 & 0.0126126170 & 0.66178442\\\\\n",
       "\t chr1 & 100037727 & 100038876 & PLPPR4   & HCT116 & -0.0005688027 & 0.01556156 & 0.0050089879 & 0.63999066\\\\\n",
       "\t chr1 & 100037727 & 100038876 & RTCA     & HCT116 &  0.0040791528 & 0.01307826 & 0.0107840530 & 0.69591073\\\\\n",
       "\t chr1 & 100037727 & 100038876 & RTCA-AS1 & HCT116 & -0.0249087958 & 0.01460240 & 0.0010782260 & 0.96243264\\\\\n",
       "\t chr1 & 100037727 & 100038876 & SASS6    & HCT116 &  0.0294145880 & 0.01736045 & 0.0222455789 & 0.33982139\\\\\n",
       "\t chr1 & 100037727 & 100038876 & SLC30A7  & HCT116 & -0.0097983219 & 0.01463933 & 0.0142596526 & 0.94984781\\\\\n",
       "\t chr1 & 100037727 & 100038876 & SLC35A3  & HCT116 &  0.0099716939 & 0.01420939 & 0.0125128735 & 0.57096754\\\\\n",
       "\t chr1 & 100037727 & 100038876 & TRMT13   & HCT116 &  0.0070494852 & 0.01270121 & 0.0118967716 & 0.64863574\\\\\n",
       "\t chr1 & 100046025 & 100046641 & AGL      & HCT116 &  0.0031032834 & 0.01524022 & 0.0061113677 & 0.57823413\\\\\n",
       "\t chr1 & 100046025 & 100046641 & CDC14A   & HCT116 & -0.0121750800 & 0.01397360 & 0.0006327480 & 0.82031689\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  989884 &  990608 & ASMTL     & HCT116 &  0.003952418 & 0.0148517286 &  0.0093814987 & 0.6426503303\\\\\n",
       "\t chrX &  989884 &  990608 & CSF2RA    & HCT116 &  0.003445228 & 0.0157337318 &  0.0041007269 & 0.5166159325\\\\\n",
       "\t chrX &  989884 &  990608 & GTPBP6    & HCT116 & -0.011086705 & 0.0140485327 &  0.0032285525 & 0.8458951799\\\\\n",
       "\t chrX &  989884 &  990608 & IL3RA     & HCT116 &  0.014231312 & 0.0150745427 &  0.0023006160 & 0.2143417103\\\\\n",
       "\t chrX &  989884 &  990608 & LINC00106 & HCT116 &  0.015691484 & 0.0152868696 &  0.0062515992 & 0.2684472635\\\\\n",
       "\t chrX &  989884 &  990608 & P2RY8     & HCT116 & -0.011085038 & 0.0139393423 &  0.0005890353 & 0.7988416696\\\\\n",
       "\t chrX &  989884 &  990608 & PLCXD1    & HCT116 &  0.013712258 & 0.0142326486 &  0.0022705909 & 0.2107265377\\\\\n",
       "\t chrX &  989884 &  990608 & PPP2R3B   & HCT116 &  0.014587699 & 0.0146551250 &  0.0060971991 & 0.2811752823\\\\\n",
       "\t chrX &  989884 &  990608 & SHOX      & HCT116 & -0.004951215 & 0.0128875109 & -0.0007166742 & 0.6287623125\\\\\n",
       "\t chrX &  989884 &  990608 & SLC25A6   & HCT116 &  0.006951821 & 0.0142101164 &  0.0059842131 & 0.4728558518\\\\\n",
       "\t chrX & 9902739 & 9902992 & CLCN4     & HCT116 &  0.014815425 & 0.0160579607 &  0.0022118474 & 0.2162618608\\\\\n",
       "\t chrX & 9902739 & 9902992 & FAM9B     & HCT116 & -0.002758100 & 0.0206972171 &  0.0032270809 & 0.6137774363\\\\\n",
       "\t chrX & 9902739 & 9902992 & TBL1X     & HCT116 & -0.012858781 & 0.0153783700 &  0.0007509701 & 0.8119198275\\\\\n",
       "\t chrX & 9902739 & 9902992 & WWC3      & HCT116 &  0.007937758 & 0.0148231419 &  0.0046010171 & 0.4109493765\\\\\n",
       "\t chrX & 9903251 & 9903473 & CLCN4     & HCT116 & -0.007720546 & 0.0156064150 &  0.0008304726 & 0.7081255092\\\\\n",
       "\t chrX & 9903251 & 9903473 & FAM9B     & HCT116 & -0.002423248 & 0.0001896353 & -0.0024936829 & 0.3551606360\\\\\n",
       "\t chrX & 9903251 & 9903473 & TBL1X     & HCT116 & -0.017486091 & 0.0131449755 &  0.0019882760 & 0.9307641791\\\\\n",
       "\t chrX & 9903251 & 9903473 & WWC3      & HCT116 &  0.042423638 & 0.0182114220 &  0.0057069685 & 0.0218929746\\\\\n",
       "\t chrX & 9906862 & 9908314 & CLCN4     & HCT116 &  0.004822346 & 0.0150428244 &  0.0035527851 & 0.4663706216\\\\\n",
       "\t chrX & 9906862 & 9908314 & FAM9B     & HCT116 & -0.007197606 & 0.0143622836 &  0.0003927634 & 0.7014214644\\\\\n",
       "\t chrX & 9906862 & 9908314 & TBL1X     & HCT116 & -0.014559052 & 0.0134908173 &  0.0081316420 & 0.9537093919\\\\\n",
       "\t chrX & 9906862 & 9908314 & WWC3      & HCT116 &  0.027832196 & 0.0137228171 &  0.0113309466 & 0.1145911112\\\\\n",
       "\t chrX & 9911462 & 9911841 & CLCN4     & HCT116 & -0.010609007 & 0.0174102374 &  0.0032912636 & 0.7876797457\\\\\n",
       "\t chrX & 9911462 & 9911841 & FAM9B     & HCT116 &  0.059898809 & 0.0178920487 &  0.0023088041 & 0.0006437568\\\\\n",
       "\t chrX & 9911462 & 9911841 & TBL1X     & HCT116 & -0.009438068 & 0.0140744591 &  0.0025539527 & 0.8029044300\\\\\n",
       "\t chrX & 9911462 & 9911841 & WWC3      & HCT116 & -0.020986996 & 0.0171618490 &  0.0067320633 & 0.9468615151\\\\\n",
       "\t chrX & 9995795 & 9996095 & CLCN4     & HCT116 & -0.014195992 & 0.0141375659 &  0.0031108815 & 0.8895571398\\\\\n",
       "\t chrX & 9995795 & 9996095 & FAM9B     & HCT116 & -0.004455696 & 0.0133737329 & -0.0011810341 & 0.5967167061\\\\\n",
       "\t chrX & 9995795 & 9996095 & TBL1X     & HCT116 & -0.005010254 & 0.0165578613 &  0.0025762293 & 0.6765884943\\\\\n",
       "\t chrX & 9995795 & 9996095 & WWC3      & HCT116 & -0.012324096 & 0.0161613571 &  0.0081178653 & 0.8970404297\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2843332 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100034595 | 100034763 | AGL      | HCT116 |  0.0050400223 | 0.01616658 | 0.0073029637 | 0.55566072 |\n",
       "| chr1 | 100034595 | 100034763 | CDC14A   | HCT116 |  0.0141889509 | 0.01398714 | 0.0002932206 | 0.16024180 |\n",
       "| chr1 | 100034595 | 100034763 | DBT      | HCT116 | -0.0001238528 | 0.01364927 | 0.0047843902 | 0.64042590 |\n",
       "| chr1 | 100034595 | 100034763 | DPH5     | HCT116 | -0.0163934858 | 0.01580155 | 0.0038361685 | 0.89976829 |\n",
       "| chr1 | 100034595 | 100034763 | EXTL2    | HCT116 | -0.0068105141 | 0.01629358 | 0.0066768142 | 0.79609945 |\n",
       "| chr1 | 100034595 | 100034763 | FRRS1    | HCT116 | -0.0180925943 | 0.01434001 | 0.0034641909 | 0.93361414 |\n",
       "| chr1 | 100034595 | 100034763 | MFSD14A  | HCT116 |  0.0178185181 | 0.01472433 | 0.0030524887 | 0.15797094 |\n",
       "| chr1 | 100034595 | 100034763 | PLPPR4   | HCT116 | -0.0089699035 | 0.01575691 | 0.0010479894 | 0.73753940 |\n",
       "| chr1 | 100034595 | 100034763 | RTCA     | HCT116 |  0.0295133880 | 0.01580357 | 0.0053641107 | 0.06324513 |\n",
       "| chr1 | 100034595 | 100034763 | RTCA-AS1 | HCT116 | -0.0147241072 | 0.01558145 | 0.0040766378 | 0.88620889 |\n",
       "| chr1 | 100034595 | 100034763 | SASS6    | HCT116 | -0.0039188789 | 0.01551349 | 0.0028756640 | 0.66929829 |\n",
       "| chr1 | 100034595 | 100034763 | SLC30A7  | HCT116 | -0.0357777822 | 0.01419918 | 0.0047461359 | 0.99784112 |\n",
       "| chr1 | 100034595 | 100034763 | SLC35A3  | HCT116 |  0.0093719800 | 0.01606637 | 0.0027847697 | 0.34090299 |\n",
       "| chr1 | 100034595 | 100034763 | TRMT13   | HCT116 |  0.0129998407 | 0.01562570 | 0.0036223934 | 0.27420987 |\n",
       "| chr1 | 100037727 | 100038876 | AGL      | HCT116 |  0.0243556687 | 0.01569431 | 0.0101133694 | 0.18207599 |\n",
       "| chr1 | 100037727 | 100038876 | CDC14A   | HCT116 |  0.0152480782 | 0.01329053 | 0.0054948536 | 0.23152081 |\n",
       "| chr1 | 100037727 | 100038876 | DBT      | HCT116 |  0.0152141760 | 0.01405810 | 0.0115274866 | 0.39656563 |\n",
       "| chr1 | 100037727 | 100038876 | DPH5     | HCT116 |  0.0024098791 | 0.01333417 | 0.0082324553 | 0.66882315 |\n",
       "| chr1 | 100037727 | 100038876 | EXTL2    | HCT116 |  0.0164913521 | 0.01440219 | 0.0158709666 | 0.48282056 |\n",
       "| chr1 | 100037727 | 100038876 | FRRS1    | HCT116 | -0.0022498411 | 0.01266587 | 0.0031121802 | 0.66397784 |\n",
       "| chr1 | 100037727 | 100038876 | MFSD14A  | HCT116 |  0.0067903419 | 0.01395098 | 0.0126126170 | 0.66178442 |\n",
       "| chr1 | 100037727 | 100038876 | PLPPR4   | HCT116 | -0.0005688027 | 0.01556156 | 0.0050089879 | 0.63999066 |\n",
       "| chr1 | 100037727 | 100038876 | RTCA     | HCT116 |  0.0040791528 | 0.01307826 | 0.0107840530 | 0.69591073 |\n",
       "| chr1 | 100037727 | 100038876 | RTCA-AS1 | HCT116 | -0.0249087958 | 0.01460240 | 0.0010782260 | 0.96243264 |\n",
       "| chr1 | 100037727 | 100038876 | SASS6    | HCT116 |  0.0294145880 | 0.01736045 | 0.0222455789 | 0.33982139 |\n",
       "| chr1 | 100037727 | 100038876 | SLC30A7  | HCT116 | -0.0097983219 | 0.01463933 | 0.0142596526 | 0.94984781 |\n",
       "| chr1 | 100037727 | 100038876 | SLC35A3  | HCT116 |  0.0099716939 | 0.01420939 | 0.0125128735 | 0.57096754 |\n",
       "| chr1 | 100037727 | 100038876 | TRMT13   | HCT116 |  0.0070494852 | 0.01270121 | 0.0118967716 | 0.64863574 |\n",
       "| chr1 | 100046025 | 100046641 | AGL      | HCT116 |  0.0031032834 | 0.01524022 | 0.0061113677 | 0.57823413 |\n",
       "| chr1 | 100046025 | 100046641 | CDC14A   | HCT116 | -0.0121750800 | 0.01397360 | 0.0006327480 | 0.82031689 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  989884 |  990608 | ASMTL     | HCT116 |  0.003952418 | 0.0148517286 |  0.0093814987 | 0.6426503303 |\n",
       "| chrX |  989884 |  990608 | CSF2RA    | HCT116 |  0.003445228 | 0.0157337318 |  0.0041007269 | 0.5166159325 |\n",
       "| chrX |  989884 |  990608 | GTPBP6    | HCT116 | -0.011086705 | 0.0140485327 |  0.0032285525 | 0.8458951799 |\n",
       "| chrX |  989884 |  990608 | IL3RA     | HCT116 |  0.014231312 | 0.0150745427 |  0.0023006160 | 0.2143417103 |\n",
       "| chrX |  989884 |  990608 | LINC00106 | HCT116 |  0.015691484 | 0.0152868696 |  0.0062515992 | 0.2684472635 |\n",
       "| chrX |  989884 |  990608 | P2RY8     | HCT116 | -0.011085038 | 0.0139393423 |  0.0005890353 | 0.7988416696 |\n",
       "| chrX |  989884 |  990608 | PLCXD1    | HCT116 |  0.013712258 | 0.0142326486 |  0.0022705909 | 0.2107265377 |\n",
       "| chrX |  989884 |  990608 | PPP2R3B   | HCT116 |  0.014587699 | 0.0146551250 |  0.0060971991 | 0.2811752823 |\n",
       "| chrX |  989884 |  990608 | SHOX      | HCT116 | -0.004951215 | 0.0128875109 | -0.0007166742 | 0.6287623125 |\n",
       "| chrX |  989884 |  990608 | SLC25A6   | HCT116 |  0.006951821 | 0.0142101164 |  0.0059842131 | 0.4728558518 |\n",
       "| chrX | 9902739 | 9902992 | CLCN4     | HCT116 |  0.014815425 | 0.0160579607 |  0.0022118474 | 0.2162618608 |\n",
       "| chrX | 9902739 | 9902992 | FAM9B     | HCT116 | -0.002758100 | 0.0206972171 |  0.0032270809 | 0.6137774363 |\n",
       "| chrX | 9902739 | 9902992 | TBL1X     | HCT116 | -0.012858781 | 0.0153783700 |  0.0007509701 | 0.8119198275 |\n",
       "| chrX | 9902739 | 9902992 | WWC3      | HCT116 |  0.007937758 | 0.0148231419 |  0.0046010171 | 0.4109493765 |\n",
       "| chrX | 9903251 | 9903473 | CLCN4     | HCT116 | -0.007720546 | 0.0156064150 |  0.0008304726 | 0.7081255092 |\n",
       "| chrX | 9903251 | 9903473 | FAM9B     | HCT116 | -0.002423248 | 0.0001896353 | -0.0024936829 | 0.3551606360 |\n",
       "| chrX | 9903251 | 9903473 | TBL1X     | HCT116 | -0.017486091 | 0.0131449755 |  0.0019882760 | 0.9307641791 |\n",
       "| chrX | 9903251 | 9903473 | WWC3      | HCT116 |  0.042423638 | 0.0182114220 |  0.0057069685 | 0.0218929746 |\n",
       "| chrX | 9906862 | 9908314 | CLCN4     | HCT116 |  0.004822346 | 0.0150428244 |  0.0035527851 | 0.4663706216 |\n",
       "| chrX | 9906862 | 9908314 | FAM9B     | HCT116 | -0.007197606 | 0.0143622836 |  0.0003927634 | 0.7014214644 |\n",
       "| chrX | 9906862 | 9908314 | TBL1X     | HCT116 | -0.014559052 | 0.0134908173 |  0.0081316420 | 0.9537093919 |\n",
       "| chrX | 9906862 | 9908314 | WWC3      | HCT116 |  0.027832196 | 0.0137228171 |  0.0113309466 | 0.1145911112 |\n",
       "| chrX | 9911462 | 9911841 | CLCN4     | HCT116 | -0.010609007 | 0.0174102374 |  0.0032912636 | 0.7876797457 |\n",
       "| chrX | 9911462 | 9911841 | FAM9B     | HCT116 |  0.059898809 | 0.0178920487 |  0.0023088041 | 0.0006437568 |\n",
       "| chrX | 9911462 | 9911841 | TBL1X     | HCT116 | -0.009438068 | 0.0140744591 |  0.0025539527 | 0.8029044300 |\n",
       "| chrX | 9911462 | 9911841 | WWC3      | HCT116 | -0.020986996 | 0.0171618490 |  0.0067320633 | 0.9468615151 |\n",
       "| chrX | 9995795 | 9996095 | CLCN4     | HCT116 | -0.014195992 | 0.0141375659 |  0.0031108815 | 0.8895571398 |\n",
       "| chrX | 9995795 | 9996095 | FAM9B     | HCT116 | -0.004455696 | 0.0133737329 | -0.0011810341 | 0.5967167061 |\n",
       "| chrX | 9995795 | 9996095 | TBL1X     | HCT116 | -0.005010254 | 0.0165578613 |  0.0025762293 | 0.6765884943 |\n",
       "| chrX | 9995795 | 9996095 | WWC3      | HCT116 | -0.012324096 | 0.0161613571 |  0.0081178653 | 0.8970404297 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType rObs          rBgSD       \n",
       "1       chr1 100034595 100034763 AGL        HCT116    0.0050400223 0.01616658  \n",
       "2       chr1 100034595 100034763 CDC14A     HCT116    0.0141889509 0.01398714  \n",
       "3       chr1 100034595 100034763 DBT        HCT116   -0.0001238528 0.01364927  \n",
       "4       chr1 100034595 100034763 DPH5       HCT116   -0.0163934858 0.01580155  \n",
       "5       chr1 100034595 100034763 EXTL2      HCT116   -0.0068105141 0.01629358  \n",
       "6       chr1 100034595 100034763 FRRS1      HCT116   -0.0180925943 0.01434001  \n",
       "7       chr1 100034595 100034763 MFSD14A    HCT116    0.0178185181 0.01472433  \n",
       "8       chr1 100034595 100034763 PLPPR4     HCT116   -0.0089699035 0.01575691  \n",
       "9       chr1 100034595 100034763 RTCA       HCT116    0.0295133880 0.01580357  \n",
       "10      chr1 100034595 100034763 RTCA-AS1   HCT116   -0.0147241072 0.01558145  \n",
       "11      chr1 100034595 100034763 SASS6      HCT116   -0.0039188789 0.01551349  \n",
       "12      chr1 100034595 100034763 SLC30A7    HCT116   -0.0357777822 0.01419918  \n",
       "13      chr1 100034595 100034763 SLC35A3    HCT116    0.0093719800 0.01606637  \n",
       "14      chr1 100034595 100034763 TRMT13     HCT116    0.0129998407 0.01562570  \n",
       "15      chr1 100037727 100038876 AGL        HCT116    0.0243556687 0.01569431  \n",
       "16      chr1 100037727 100038876 CDC14A     HCT116    0.0152480782 0.01329053  \n",
       "17      chr1 100037727 100038876 DBT        HCT116    0.0152141760 0.01405810  \n",
       "18      chr1 100037727 100038876 DPH5       HCT116    0.0024098791 0.01333417  \n",
       "19      chr1 100037727 100038876 EXTL2      HCT116    0.0164913521 0.01440219  \n",
       "20      chr1 100037727 100038876 FRRS1      HCT116   -0.0022498411 0.01266587  \n",
       "21      chr1 100037727 100038876 MFSD14A    HCT116    0.0067903419 0.01395098  \n",
       "22      chr1 100037727 100038876 PLPPR4     HCT116   -0.0005688027 0.01556156  \n",
       "23      chr1 100037727 100038876 RTCA       HCT116    0.0040791528 0.01307826  \n",
       "24      chr1 100037727 100038876 RTCA-AS1   HCT116   -0.0249087958 0.01460240  \n",
       "25      chr1 100037727 100038876 SASS6      HCT116    0.0294145880 0.01736045  \n",
       "26      chr1 100037727 100038876 SLC30A7    HCT116   -0.0097983219 0.01463933  \n",
       "27      chr1 100037727 100038876 SLC35A3    HCT116    0.0099716939 0.01420939  \n",
       "28      chr1 100037727 100038876 TRMT13     HCT116    0.0070494852 0.01270121  \n",
       "29      chr1 100046025 100046641 AGL        HCT116    0.0031032834 0.01524022  \n",
       "30      chr1 100046025 100046641 CDC14A     HCT116   -0.0121750800 0.01397360  \n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮           \n",
       "2843303 chrX  989884    990608   ASMTL      HCT116    0.003952418  0.0148517286\n",
       "2843304 chrX  989884    990608   CSF2RA     HCT116    0.003445228  0.0157337318\n",
       "2843305 chrX  989884    990608   GTPBP6     HCT116   -0.011086705  0.0140485327\n",
       "2843306 chrX  989884    990608   IL3RA      HCT116    0.014231312  0.0150745427\n",
       "2843307 chrX  989884    990608   LINC00106  HCT116    0.015691484  0.0152868696\n",
       "2843308 chrX  989884    990608   P2RY8      HCT116   -0.011085038  0.0139393423\n",
       "2843309 chrX  989884    990608   PLCXD1     HCT116    0.013712258  0.0142326486\n",
       "2843310 chrX  989884    990608   PPP2R3B    HCT116    0.014587699  0.0146551250\n",
       "2843311 chrX  989884    990608   SHOX       HCT116   -0.004951215  0.0128875109\n",
       "2843312 chrX  989884    990608   SLC25A6    HCT116    0.006951821  0.0142101164\n",
       "2843313 chrX 9902739   9902992   CLCN4      HCT116    0.014815425  0.0160579607\n",
       "2843314 chrX 9902739   9902992   FAM9B      HCT116   -0.002758100  0.0206972171\n",
       "2843315 chrX 9902739   9902992   TBL1X      HCT116   -0.012858781  0.0153783700\n",
       "2843316 chrX 9902739   9902992   WWC3       HCT116    0.007937758  0.0148231419\n",
       "2843317 chrX 9903251   9903473   CLCN4      HCT116   -0.007720546  0.0156064150\n",
       "2843318 chrX 9903251   9903473   FAM9B      HCT116   -0.002423248  0.0001896353\n",
       "2843319 chrX 9903251   9903473   TBL1X      HCT116   -0.017486091  0.0131449755\n",
       "2843320 chrX 9903251   9903473   WWC3       HCT116    0.042423638  0.0182114220\n",
       "2843321 chrX 9906862   9908314   CLCN4      HCT116    0.004822346  0.0150428244\n",
       "2843322 chrX 9906862   9908314   FAM9B      HCT116   -0.007197606  0.0143622836\n",
       "2843323 chrX 9906862   9908314   TBL1X      HCT116   -0.014559052  0.0134908173\n",
       "2843324 chrX 9906862   9908314   WWC3       HCT116    0.027832196  0.0137228171\n",
       "2843325 chrX 9911462   9911841   CLCN4      HCT116   -0.010609007  0.0174102374\n",
       "2843326 chrX 9911462   9911841   FAM9B      HCT116    0.059898809  0.0178920487\n",
       "2843327 chrX 9911462   9911841   TBL1X      HCT116   -0.009438068  0.0140744591\n",
       "2843328 chrX 9911462   9911841   WWC3       HCT116   -0.020986996  0.0171618490\n",
       "2843329 chrX 9995795   9996095   CLCN4      HCT116   -0.014195992  0.0141375659\n",
       "2843330 chrX 9995795   9996095   FAM9B      HCT116   -0.004455696  0.0133737329\n",
       "2843331 chrX 9995795   9996095   TBL1X      HCT116   -0.005010254  0.0165578613\n",
       "2843332 chrX 9995795   9996095   WWC3       HCT116   -0.012324096  0.0161613571\n",
       "        rBgMean       pvalZ       \n",
       "1       0.0073029637  0.55566072  \n",
       "2       0.0002932206  0.16024180  \n",
       "3       0.0047843902  0.64042590  \n",
       "4       0.0038361685  0.89976829  \n",
       "5       0.0066768142  0.79609945  \n",
       "6       0.0034641909  0.93361414  \n",
       "7       0.0030524887  0.15797094  \n",
       "8       0.0010479894  0.73753940  \n",
       "9       0.0053641107  0.06324513  \n",
       "10      0.0040766378  0.88620889  \n",
       "11      0.0028756640  0.66929829  \n",
       "12      0.0047461359  0.99784112  \n",
       "13      0.0027847697  0.34090299  \n",
       "14      0.0036223934  0.27420987  \n",
       "15      0.0101133694  0.18207599  \n",
       "16      0.0054948536  0.23152081  \n",
       "17      0.0115274866  0.39656563  \n",
       "18      0.0082324553  0.66882315  \n",
       "19      0.0158709666  0.48282056  \n",
       "20      0.0031121802  0.66397784  \n",
       "21      0.0126126170  0.66178442  \n",
       "22      0.0050089879  0.63999066  \n",
       "23      0.0107840530  0.69591073  \n",
       "24      0.0010782260  0.96243264  \n",
       "25      0.0222455789  0.33982139  \n",
       "26      0.0142596526  0.94984781  \n",
       "27      0.0125128735  0.57096754  \n",
       "28      0.0118967716  0.64863574  \n",
       "29      0.0061113677  0.57823413  \n",
       "30      0.0006327480  0.82031689  \n",
       "⋮       ⋮             ⋮           \n",
       "2843303  0.0093814987 0.6426503303\n",
       "2843304  0.0041007269 0.5166159325\n",
       "2843305  0.0032285525 0.8458951799\n",
       "2843306  0.0023006160 0.2143417103\n",
       "2843307  0.0062515992 0.2684472635\n",
       "2843308  0.0005890353 0.7988416696\n",
       "2843309  0.0022705909 0.2107265377\n",
       "2843310  0.0060971991 0.2811752823\n",
       "2843311 -0.0007166742 0.6287623125\n",
       "2843312  0.0059842131 0.4728558518\n",
       "2843313  0.0022118474 0.2162618608\n",
       "2843314  0.0032270809 0.6137774363\n",
       "2843315  0.0007509701 0.8119198275\n",
       "2843316  0.0046010171 0.4109493765\n",
       "2843317  0.0008304726 0.7081255092\n",
       "2843318 -0.0024936829 0.3551606360\n",
       "2843319  0.0019882760 0.9307641791\n",
       "2843320  0.0057069685 0.0218929746\n",
       "2843321  0.0035527851 0.4663706216\n",
       "2843322  0.0003927634 0.7014214644\n",
       "2843323  0.0081316420 0.9537093919\n",
       "2843324  0.0113309466 0.1145911112\n",
       "2843325  0.0032912636 0.7876797457\n",
       "2843326  0.0023088041 0.0006437568\n",
       "2843327  0.0025539527 0.8029044300\n",
       "2843328  0.0067320633 0.9468615151\n",
       "2843329  0.0031108815 0.8895571398\n",
       "2843330 -0.0011810341 0.5967167061\n",
       "2843331  0.0025762293 0.6765884943\n",
       "2843332  0.0081178653 0.8970404297"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = celltype\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   data.table_1.15.2                \n",
       " [3] foreach_1.5.2                     BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] genomation_1.34.0                 Seurat_5.0.3                     \n",
       "[11] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[13] FigR_0.1.0                        motifmatchr_1.24.0               \n",
       "[15] chromVAR_1.24.0                   cowplot_1.1.3                    \n",
       "[17] ggplot2_3.5.1                     dplyr_1.1.4                      \n",
       "[19] SummarizedExperiment_1.30.2       Biobase_2.60.0                   \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               MatrixGenerics_1.12.3            \n",
       "[27] matrixStats_1.3.0                 Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             vroom_1.6.5                \n",
       " [15] globals_0.16.3              lattice_0.22-6             \n",
       " [17] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [19] plotly_4.10.4               plotrix_3.8-4              \n",
       " [21] yaml_2.3.8                  httpuv_1.6.15              \n",
       " [23] sctransform_0.4.1           spam_2.10-0                \n",
       " [25] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [27] pbapply_1.7-2               DBI_1.2.2                  \n",
       " [29] CNEr_1.38.0                 RColorBrewer_1.1-3         \n",
       " [31] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [33] Rtsne_0.17                  purrr_1.0.2                \n",
       " [35] R.utils_2.12.3              RCurl_1.98-1.14            \n",
       " [37] pracma_2.4.4                GenomeInfoDbData_1.2.11    \n",
       " [39] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [41] spatstat.utils_3.0-4        listenv_0.9.1              \n",
       " [43] seqLogo_1.68.0              goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             spatstat.random_3.2-3      \n",
       " [47] annotate_1.80.0             fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           leiden_0.4.3.1             \n",
       " [51] codetools_0.2-20            DelayedArray_0.26.7        \n",
       " [53] DT_0.33                     tidyselect_1.2.1           \n",
       " [55] spatstat.explore_3.2-6      base64enc_0.1-3            \n",
       " [57] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [59] progressr_0.14.0            iterators_1.0.14           \n",
       " [61] ggridges_0.5.6              survival_3.6-4             \n",
       " [63] tools_4.3.3                 TFMPvalue_0.0.9            \n",
       " [65] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [67] glue_1.7.0                  gridExtra_2.3              \n",
       " [69] IRdisplay_1.1               withr_3.0.0                \n",
       " [71] fastmap_1.1.1               fansi_1.0.6                \n",
       " [73] caTools_1.18.2              digest_0.6.35              \n",
       " [75] R6_2.5.1                    mime_0.12                  \n",
       " [77] nabor_0.5.0                 seqPattern_1.34.0          \n",
       " [79] colorspace_2.1-0            scattermore_1.2            \n",
       " [81] GO.db_3.18.0                tensor_1.5                 \n",
       " [83] gtools_3.9.5                poweRlaw_0.80.0            \n",
       " [85] spatstat.data_3.0-4         RSQLite_2.3.6              \n",
       " [87] R.methodsS3_1.8.2           utf8_1.2.4                 \n",
       " [89] tidyr_1.3.1                 generics_0.1.3             \n",
       " [91] httr_1.4.7                  htmlwidgets_1.6.4          \n",
       " [93] S4Arrays_1.2.1              TFBSTools_1.40.0           \n",
       " [95] uwot_0.1.16                 pkgconfig_2.0.3            \n",
       " [97] gtable_0.3.5                blob_1.2.4                 \n",
       " [99] impute_1.76.0               lmtest_0.9-40              \n",
       "[101] htmltools_0.5.8.1           dotCall64_1.1-1            \n",
       "[103] scales_1.3.0                png_0.1-8                  \n",
       "[105] tzdb_0.4.0                  reshape2_1.4.4             \n",
       "[107] rjson_0.2.21                uuid_1.2-0                 \n",
       "[109] nlme_3.1-164                repr_1.1.7                 \n",
       "[111] cachem_1.0.8                zoo_1.8-12                 \n",
       "[113] stringr_1.5.1               KernSmooth_2.23-22         \n",
       "[115] miniUI_0.1.1.1              AnnotationDbi_1.64.1       \n",
       "[117] restfulr_0.0.15             pillar_1.9.0               \n",
       "[119] vctrs_0.6.5                 RANN_2.6.1                 \n",
       "[121] promises_1.3.0              xtable_1.8-4               \n",
       "[123] cluster_2.1.6               evaluate_0.23              \n",
       "[125] readr_2.1.5                 cli_3.6.2                  \n",
       "[127] compiler_4.3.3              Rsamtools_2.18.0           \n",
       "[129] rlang_1.1.3                 crayon_1.5.2               \n",
       "[131] future.apply_1.11.2         plyr_1.8.9                 \n",
       "[133] stringi_1.8.3               gridBase_0.4-7             \n",
       "[135] deldir_2.0-4                viridisLite_0.4.2          \n",
       "[137] BiocParallel_1.36.0         munsell_0.5.1              \n",
       "[139] lazyeval_0.2.2              spatstat.geom_3.2-9        \n",
       "[141] IRkernel_1.3.2              RcppHNSW_0.6.0             \n",
       "[143] hms_1.1.3                   patchwork_1.2.0            \n",
       "[145] bit64_4.0.5                 future_1.33.2              \n",
       "[147] KEGGREST_1.40.1             shiny_1.8.1.1              \n",
       "[149] ROCR_1.0-11                 igraph_2.0.3               \n",
       "[151] memoise_2.0.1               bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
