{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8466b35-cae5-4f62-8138-37b2de484a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.2\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : grid v4.3.3\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "\tLoading Package : gtools v3.9.5\n",
      "\n",
      "\tLoading Package : gtable v0.3.5\n",
      "\n",
      "\tLoading Package : ggplot2 v3.5.1\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "\tLoading Package : plyr v1.8.9\n",
      "\n",
      "\tLoading Package : stringr v1.5.1\n",
      "\n",
      "\tLoading Package : data.table v1.15.2\n",
      "\n",
      "\tLoading Package : matrixStats v1.3.0\n",
      "\n",
      "\tLoading Package : S4Vectors v0.40.2\n",
      "\n",
      "\tLoading Package : GenomicRanges v1.54.1\n",
      "\n",
      "\tLoading Package : BiocGenerics v0.48.1\n",
      "\n",
      "\tLoading Package : Matrix v1.6.5\n",
      "\n",
      "\tLoading Package : Rcpp v1.0.12\n",
      "\n",
      "\tLoading Package : SummarizedExperiment v1.30.2\n",
      "\n",
      "\tLoading Package : rhdf5 v2.44.0\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(magrittr)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(ggplot2)\n",
    "library(dplyr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e85020d7-d358-492d-9358-c81270f4b624",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1bec6690-52ec-450d-bba7-7d4528a185cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/cellranger_arc/refdata-cellranger-arc-GRCh38-2020-A-2.0.0/genes/genes.gtf.gz\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "TSS_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.bed\"\n",
    "path.atac_frag = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/scE2G_input_Maya/WTC11_HPC_lenient/cluster_5/atac_fragments.tsv.gz\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/scE2G_input_Maya/WTC11_HPC_lenient/cluster_5/rna_count_matrix.csv.gz\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/WTC11_cluster5/WTC11_cluster5/Kendall/Pairs.tsv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/WTC11/1.Genome_wide_prediction/ArchR_250325/ArchR_250325/\"\n",
    "celltype = \"WTC11_cluster5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "340dac32-44ac-4231-b964-b5f3b9e2dbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16\n",
    "maxDist = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da973385-4c81-4b1c-9669-cf5933b23a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 16.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRThreads(threads = n.cores) \n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86be3e66-c954-4620-9e64-c6a1c0fcb84c",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6f363969-bf0b-45ce-858c-a8cb64cc210a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1ff4f07-c695-45a1-bf15-dc523d20497f",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2a104a7a-b6b9-4315-8b15-6a22c48e142a",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna = read.csv(path.matrix.rna_count,\n",
    "                      row.names = 1,\n",
    "                      check.names = F)\n",
    "matrix.rna = Matrix(as.matrix(matrix.rna), sparse = TRUE)\n",
    "# matrix.rna = matrix.rna[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna[rowSums(matrix.rna) > 0,]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "287f8d61-9ad5-4f60-97b6-9e7b71795042",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ad6e79b8-2429-425a-ac78-6773abe4dbc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6ddc5a2b-371a-435d-9250-751ecffa65c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65a5d8fc-a99f-481c-9eec-8e88c8875dd3",
   "metadata": {},
   "source": [
    "make TSS GRanges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "815b37df-d305-4ba7-bd48-2270290ca3b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tss <- fread(TSS_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\"))\n",
    "df.tss = as.data.frame(df.tss)\n",
    "rownames(df.tss) = df.tss$name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fce150b6-ee55-46d7-9035-cd9698b5ee9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tss[df.tss$strand == \"+\",\"end\"] = df.tss[df.tss$strand == \"+\",\"start\"]\n",
    "df.tss[df.tss$strand == \"-\",\"start\"] = df.tss[df.tss$strand == \"-\",\"end\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7d73721a-8ec9-4b65-91b5-29cca6872927",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tss = df.tss[unname(rownames(matrix.rna.rename)),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a5a5da89-e7fd-4ffb-b477-48beff459909",
   "metadata": {},
   "outputs": [],
   "source": [
    "gr.tss <- makeGRangesFromDataFrame(\n",
    "  df.tss,\n",
    "  seqnames.field = \"chr\",\n",
    "  start.field = \"start\",\n",
    "  end.field = \"end\",\n",
    "  strand.field = \"strand\",    \n",
    "  keep.extra.columns = TRUE   \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a2259f88-32ce-4299-9592-b39c2273f03e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 17524 ranges and 4 metadata columns:\n",
       "            seqnames    ranges strand |        name     score      Ensembl_ID\n",
       "               <Rle> <IRanges>  <Rle> | <character> <integer>     <character>\n",
       "  LINC01128     chr1    827590      + |   LINC01128         0 ENSG00000228794\n",
       "  LINC00115     chr1    827522      - |   LINC00115         0 ENSG00000225880\n",
       "     FAM41C     chr1    876802      - |      FAM41C         0 ENSG00000230368\n",
       "     SAMD11     chr1    925740      + |      SAMD11         0 ENSG00000187634\n",
       "      NOC2L     chr1    959299      - |       NOC2L         0 ENSG00000188976\n",
       "        ...      ...       ...    ... .         ...       ...             ...\n",
       "     NLGN4Y     chrY  14523745      + |      NLGN4Y         0 ENSG00000165246\n",
       "     TTTY14     chrY  19077547      - |      TTTY14         0 ENSG00000176728\n",
       "      KDM5D     chrY  19744939      - |       KDM5D         0 ENSG00000012817\n",
       "     TTTY10     chrY  20519228      - |      TTTY10         0 ENSG00000229236\n",
       "     EIF1AY     chrY  20575710      + |      EIF1AY         0 ENSG00000198692\n",
       "                       gene_type\n",
       "                     <character>\n",
       "  LINC01128 processed_transcript\n",
       "  LINC00115              lincRNA\n",
       "     FAM41C              lincRNA\n",
       "     SAMD11       protein_coding\n",
       "      NOC2L       protein_coding\n",
       "        ...                  ...\n",
       "     NLGN4Y       protein_coding\n",
       "     TTTY14              lincRNA\n",
       "      KDM5D       protein_coding\n",
       "     TTTY10              lincRNA\n",
       "     EIF1AY       protein_coding\n",
       "  -------\n",
       "  seqinfo: 24 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gr.tss"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3538ffdc-4fed-48eb-9c9d-86805a91cf06",
   "metadata": {},
   "source": [
    "Create an ArchRProject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1753ffaa-89e2-43bf-bcf6-19937782d63a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-createArrows-2b40a56cf3220f-Date-2025-03-25_Time-16-39-45.884567.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Cleaning Temporary Files\n",
      "\n",
      "2025-03-25 16:39:46.30563 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "(WTC11_cluster5 : 1 of 1) Determining Arrow Method to use!\n",
      "\n",
      "2025-03-25 16:39:46.381244 : (WTC11_cluster5 : 1 of 1) Reading In Fragments from inputFiles (readMethod = tabix), 0.001 mins elapsed.\n",
      "\n",
      "2025-03-25 16:39:46.392265 : (WTC11_cluster5 : 1 of 1) Tabix Bed To Temporary File, 0.001 mins elapsed.\n",
      "\n",
      "2025-03-25 16:40:51.968513 : (WTC11_cluster5 : 1 of 1) Successful creation of Temporary File, 1.094 mins elapsed.\n",
      "\n",
      "2025-03-25 16:40:51.976239 : (WTC11_cluster5 : 1 of 1) Creating ArrowFile From Temporary File, 1.095 mins elapsed.\n",
      "\n",
      "2025-03-25 16:41:17.049677 : (WTC11_cluster5 : 1 of 1) Successful creation of Arrow File, 1.512 mins elapsed.\n",
      "\n",
      "2025-03-25 16:41:32.40054 : Continuing through after error ggplot for Fragment Size Distribution, 1.768 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-03-25 16:42:07.322507 : (WTC11_cluster5 : 1 of 1) CellStats : Number of Cells Pass Filter = 791 , 2.35 mins elapsed.\n",
      "\n",
      "2025-03-25 16:42:07.328703 : (WTC11_cluster5 : 1 of 1) CellStats : Median Frags = 53648 , 2.35 mins elapsed.\n",
      "\n",
      "2025-03-25 16:42:07.336488 : (WTC11_cluster5 : 1 of 1) CellStats : Median TSS Enrichment = 11.204 , 2.351 mins elapsed.\n",
      "\n",
      "2025-03-25 16:42:08.284086 : Continuing through after error ggplot for TSS by Frags, 2.366 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-03-25 16:42:11.309921 : (WTC11_cluster5 : 1 of 1) Adding Additional Feature Counts!, 2.417 mins elapsed.\n",
      "\n",
      "2025-03-25 16:42:39.864155 : (WTC11_cluster5 : 1 of 1) Removing Fragments from Filtered Cells, 2.893 mins elapsed.\n",
      "\n",
      "2025-03-25 16:42:39.874972 : (WTC11_cluster5 : 1 of 1) Creating Filtered Arrow File, 2.893 mins elapsed.\n",
      "\n",
      "2025-03-25 16:43:55.473704 : (WTC11_cluster5 : 1 of 1) Finished Constructing Filtered Arrow File!, 4.153 mins elapsed.\n",
      "\n",
      "2025-03-25 16:43:55.658919 : (WTC11_cluster5 : 1 of 1) Adding TileMatrix!, 4.156 mins elapsed.\n",
      "\n",
      "2025-03-25 16:48:14.463362 : (WTC11_cluster5 : 1 of 1) Finished Creating Arrow File, 8.469 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-createArrows-2b40a56cf3220f-Date-2025-03-25_Time-16-39-45.884567.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ArrowFiles <- createArrowFiles(\n",
    "  inputFiles = c(path.atac_frag),\n",
    "  sampleNames = c(celltype),\n",
    "  validBarcodes = list(WTC11_cluster5 = colnames(matrix.rna.rename)),\n",
    "  addTileMat = T,\n",
    "  addGeneScoreMat = F,\n",
    "  force = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ada3db45-d050-4659-a7cf-2891e04b1249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'WTC11_cluster5.arrow'"
      ],
      "text/latex": [
       "'WTC11\\_cluster5.arrow'"
      ],
      "text/markdown": [
       "'WTC11_cluster5.arrow'"
      ],
      "text/plain": [
       "[1] \"WTC11_cluster5.arrow\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ArrowFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5ba28283-d488-4982-8d4c-745fc3f59365",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"ArchR\",sep=\"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "aa217507-8323-43bf-b23c-0ec25ee33c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Validating Arrows...\n",
      "\n",
      "Getting SampleNames...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Copying ArrowFiles to Ouptut Directory! If you want to save disk space set copyArrows = FALSE\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Getting Cell Metadata...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Merging Cell Metadata...\n",
      "\n",
      "Initializing ArchRProject...\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR <- ArchRProject(\n",
    "  ArrowFiles = ArrowFiles, \n",
    "  outputDirectory = paste(dir.output,\"ArchR\",sep=\"/\"),\n",
    "  copyArrows = TRUE \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f9795403-545f-4385-aa38-2c70b1b2fb21",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: ArchRProject \n",
       "outputDirectory: /maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/WTC11/1.Genome_wide_prediction/ArchR_250325/ArchR_250325/ArchR \n",
       "samples(1): WTC11_cluster5\n",
       "sampleColData names(1): ArrowFiles\n",
       "cellColData names(13): Sample TSSEnrichment ... nDiFrags BlacklistRatio\n",
       "numberOfCells(1): 791\n",
       "medianTSS(1): 11.204\n",
       "medianFrags(1): 53648"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obj.ArchR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ee876dd5-0b22-46fd-9015-7e22144c5024",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking Inputs...\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addIterativeLSI-2b40a55e5c876b-Date-2025-03-25_Time-16-48-19.663651.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-03-25 16:48:20.293555 : Computing Total Across All Features, 0.001 mins elapsed.\n",
      "\n",
      "2025-03-25 16:48:22.5281 : Computing Top Features, 0.038 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-03-25 16:48:25.532104 : Running LSI (1 of 2) on Top Features, 0.088 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-03-25 16:48:25.552024 : Creating Partial Matrix, 0.088 mins elapsed.\n",
      "\n",
      "2025-03-25 16:48:41.135012 : Computing LSI, 0.348 mins elapsed.\n",
      "\n",
      "2025-03-25 16:49:27.774069 : Identifying Clusters, 1.125 mins elapsed.\n",
      "\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "2025-03-25 16:49:32.817006 : Identified 6 Clusters, 1.209 mins elapsed.\n",
      "\n",
      "2025-03-25 16:49:32.837105 : Saving LSI Iteration, 1.21 mins elapsed.\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************************************************************\n",
      "2025-03-25 16:49:45.519584 : ERROR Found in .saveIteration for  \n",
      "LogFile = ArchRLogs/ArchR-addIterativeLSI-2b40a55e5c876b-Date-2025-03-25_Time-16-48-19.663651.log\n",
      "\n",
      "<simpleError in g$grobs[[legend]]: no such index at level 2\n",
      ">\n",
      "\n",
      "************************************************************\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-03-25 16:49:45.523005 : Creating Cluster Matrix on the total Group Features, 1.421 mins elapsed.\n",
      "\n",
      "2025-03-25 16:50:02.598005 : Computing Variable Features, 1.706 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-03-25 16:50:02.778947 : Running LSI (2 of 2) on Variable Features, 1.709 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-03-25 16:50:02.79814 : Creating Partial Matrix, 1.709 mins elapsed.\n",
      "\n",
      "2025-03-25 16:50:12.254971 : Computing LSI, 1.867 mins elapsed.\n",
      "\n",
      "2025-03-25 16:50:59.329499 : Finished Running IterativeLSI, 2.651 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addIterativeLSI(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    useMatrix = \"TileMatrix\", \n",
    "    name = \"IterativeLSI\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2c1d334b-7fd7-425a-bb8d-7be4cbedd3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.ArchR = addPeakSet(obj.ArchR,\n",
    "                       peakSet = reduce(pairs.E2G))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e25c14db-eb4e-4b94-8862-eb34246a365c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeakMatrix-2b40a5a79c8e5-Date-2025-03-25_Time-16-51-03.178492.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-03-25 16:51:03.458099 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-03-25 16:51:04.209887 : Adding WTC11_cluster5 to PeakMatrix for Chr (1 of 23)!, 0.011 mins elapsed.\n",
      "\n",
      "2025-03-25 16:51:13.771046 : Adding WTC11_cluster5 to PeakMatrix for Chr (2 of 23)!, 0.171 mins elapsed.\n",
      "\n",
      "2025-03-25 16:51:21.871066 : Adding WTC11_cluster5 to PeakMatrix for Chr (3 of 23)!, 0.306 mins elapsed.\n",
      "\n",
      "2025-03-25 16:51:29.052075 : Adding WTC11_cluster5 to PeakMatrix for Chr (4 of 23)!, 0.426 mins elapsed.\n",
      "\n",
      "2025-03-25 16:51:35.899617 : Adding WTC11_cluster5 to PeakMatrix for Chr (5 of 23)!, 0.54 mins elapsed.\n",
      "\n",
      "2025-03-25 16:51:42.286648 : Adding WTC11_cluster5 to PeakMatrix for Chr (6 of 23)!, 0.646 mins elapsed.\n",
      "\n",
      "2025-03-25 16:51:48.759189 : Adding WTC11_cluster5 to PeakMatrix for Chr (7 of 23)!, 0.754 mins elapsed.\n",
      "\n",
      "2025-03-25 16:51:55.17953 : Adding WTC11_cluster5 to PeakMatrix for Chr (8 of 23)!, 0.861 mins elapsed.\n",
      "\n",
      "2025-03-25 16:52:01.573796 : Adding WTC11_cluster5 to PeakMatrix for Chr (9 of 23)!, 0.968 mins elapsed.\n",
      "\n",
      "2025-03-25 16:52:07.848156 : Adding WTC11_cluster5 to PeakMatrix for Chr (10 of 23)!, 1.072 mins elapsed.\n",
      "\n",
      "2025-03-25 16:52:14.518016 : Adding WTC11_cluster5 to PeakMatrix for Chr (11 of 23)!, 1.183 mins elapsed.\n",
      "\n",
      "2025-03-25 16:52:22.139038 : Adding WTC11_cluster5 to PeakMatrix for Chr (12 of 23)!, 1.31 mins elapsed.\n",
      "\n",
      "2025-03-25 16:52:30.090418 : Adding WTC11_cluster5 to PeakMatrix for Chr (13 of 23)!, 1.443 mins elapsed.\n",
      "\n",
      "2025-03-25 16:52:37.588739 : Adding WTC11_cluster5 to PeakMatrix for Chr (14 of 23)!, 1.568 mins elapsed.\n",
      "\n",
      "2025-03-25 16:52:43.893023 : Adding WTC11_cluster5 to PeakMatrix for Chr (15 of 23)!, 1.673 mins elapsed.\n",
      "\n",
      "2025-03-25 16:52:50.903627 : Adding WTC11_cluster5 to PeakMatrix for Chr (16 of 23)!, 1.79 mins elapsed.\n",
      "\n",
      "2025-03-25 16:52:58.187229 : Adding WTC11_cluster5 to PeakMatrix for Chr (17 of 23)!, 1.911 mins elapsed.\n",
      "\n",
      "2025-03-25 16:53:05.834661 : Adding WTC11_cluster5 to PeakMatrix for Chr (18 of 23)!, 2.039 mins elapsed.\n",
      "\n",
      "2025-03-25 16:53:13.171304 : Adding WTC11_cluster5 to PeakMatrix for Chr (19 of 23)!, 2.161 mins elapsed.\n",
      "\n",
      "2025-03-25 16:53:20.61841 : Adding WTC11_cluster5 to PeakMatrix for Chr (20 of 23)!, 2.285 mins elapsed.\n",
      "\n",
      "2025-03-25 16:53:27.053885 : Adding WTC11_cluster5 to PeakMatrix for Chr (21 of 23)!, 2.392 mins elapsed.\n",
      "\n",
      "2025-03-25 16:53:34.300892 : Adding WTC11_cluster5 to PeakMatrix for Chr (22 of 23)!, 2.513 mins elapsed.\n",
      "\n",
      "2025-03-25 16:53:42.03479 : Adding WTC11_cluster5 to PeakMatrix for Chr (23 of 23)!, 2.642 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeakMatrix-2b40a5a79c8e5-Date-2025-03-25_Time-16-51-03.178492.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addPeakMatrix(obj.ArchR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "550747c0-ded3-4a82-bc47-9d259a41ca2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna.rename2 = matrix.rna.rename\n",
    "colnames(matrix.rna.rename2) = paste(celltype,colnames(matrix.rna.rename),sep=\"#\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "92b20885-7b28-419c-a66d-74640e850733",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA <- SummarizedExperiment(\n",
    "  assays = SimpleList(counts = matrix.rna.rename2),\n",
    "  rowRanges = gr.tss\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "834ae1ff-160a-4e09-b201-ffe7ab9919eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA = sort.GenomicRanges(sortSeqlevels(seRNA), \n",
    "                           ignore.strand = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4b4e7086-362b-4c79-8e58-327684d0b787",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 17524 791 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(17524): LINC00115 LINC01128 ... TTTY10 EIF1AY\n",
       "rowData names(4): name score Ensembl_ID gene_type\n",
       "colnames(791): WTC11_cluster5#AAACAGCCAAACCCTA-1\n",
       "  WTC11_cluster5#AAACCAACAACATAAG-1 ...\n",
       "  WTC11_cluster5#TTTGTCTAGATAACCC-1 WTC11_cluster5#TTTGTGGCAGGCAAGC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "65cb2097-9075-4ee4-9e54-dd288dc50736",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addGeneExpressionMatrix-2b40a52ba866f3-Date-2025-03-25_Time-16-53-48.961231.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Overlap w/ scATAC = 1\n",
      "\n",
      "2025-03-25 16:53:49.45887 : \n",
      "\n",
      "Overlap Per Sample w/ scATAC : WTC11_cluster5=791\n",
      "\n",
      "2025-03-25 16:53:49.46997 : \n",
      "\n",
      "2025-03-25 16:53:50.984147 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-03-25 16:53:51.612918 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (1 of 23)!, 0.01 mins elapsed.\n",
      "\n",
      "2025-03-25 16:53:58.516363 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (2 of 23)!, 0.126 mins elapsed.\n",
      "\n",
      "2025-03-25 16:54:05.131032 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (3 of 23)!, 0.236 mins elapsed.\n",
      "\n",
      "2025-03-25 16:54:14.47154 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (4 of 23)!, 0.391 mins elapsed.\n",
      "\n",
      "2025-03-25 16:54:20.630713 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (5 of 23)!, 0.494 mins elapsed.\n",
      "\n",
      "2025-03-25 16:54:26.26633 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (6 of 23)!, 0.588 mins elapsed.\n",
      "\n",
      "2025-03-25 16:54:35.143389 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (7 of 23)!, 0.736 mins elapsed.\n",
      "\n",
      "2025-03-25 16:54:40.616988 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (8 of 23)!, 0.827 mins elapsed.\n",
      "\n",
      "2025-03-25 16:54:47.435985 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (9 of 23)!, 0.941 mins elapsed.\n",
      "\n",
      "2025-03-25 16:54:56.780934 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (10 of 23)!, 1.097 mins elapsed.\n",
      "\n",
      "2025-03-25 16:55:03.695162 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (11 of 23)!, 1.212 mins elapsed.\n",
      "\n",
      "2025-03-25 16:55:10.714801 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (12 of 23)!, 1.329 mins elapsed.\n",
      "\n",
      "2025-03-25 16:55:19.886366 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (13 of 23)!, 1.482 mins elapsed.\n",
      "\n",
      "2025-03-25 16:55:26.919824 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (14 of 23)!, 1.599 mins elapsed.\n",
      "\n",
      "2025-03-25 16:55:33.897621 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (15 of 23)!, 1.715 mins elapsed.\n",
      "\n",
      "2025-03-25 16:55:42.180455 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (16 of 23)!, 1.853 mins elapsed.\n",
      "\n",
      "2025-03-25 16:55:48.498166 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (17 of 23)!, 1.959 mins elapsed.\n",
      "\n",
      "2025-03-25 16:55:53.87465 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (18 of 23)!, 2.048 mins elapsed.\n",
      "\n",
      "2025-03-25 16:56:02.64508 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (19 of 23)!, 2.194 mins elapsed.\n",
      "\n",
      "2025-03-25 16:56:08.573662 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (20 of 23)!, 2.293 mins elapsed.\n",
      "\n",
      "2025-03-25 16:56:13.999344 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (21 of 23)!, 2.384 mins elapsed.\n",
      "\n",
      "2025-03-25 16:56:22.042999 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (22 of 23)!, 2.518 mins elapsed.\n",
      "\n",
      "2025-03-25 16:56:27.776997 : Adding WTC11_cluster5 to GeneExpressionMatrix for Chr (23 of 23)!, 2.613 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addGeneExpressionMatrix-2b40a52ba866f3-Date-2025-03-25_Time-16-53-48.961231.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addGeneExpressionMatrix(obj.ArchR,\n",
    "                                    seRNA = seRNA,\n",
    "                                    strictMatch = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4318703-a143-40c0-b159-274d05033bbc",
   "metadata": {},
   "source": [
    "Run ArchR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "47c8fd5f-19ec-46e6-af87-57502c1e1ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeak2GeneLinks-2b40a56b06f58f-Date-2025-03-25_Time-16-56-36.17321.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-03-25 16:56:36.605271 : Getting Available Matrices, 0.007 mins elapsed.\n",
      "\n",
      "No predictionScore found. Continuing without predictionScore!\n",
      "\n",
      "2025-03-25 16:56:36.920839 : Filtered Low Prediction Score Cells (0 of 791, 0), 0 mins elapsed.\n",
      "\n",
      "2025-03-25 16:56:37.031521 : Computing KNN, 0.002 mins elapsed.\n",
      "\n",
      "2025-03-25 16:56:37.073392 : Identifying Non-Overlapping KNN pairs, 0.003 mins elapsed.\n",
      "\n",
      "2025-03-25 16:56:38.597563 : Identified 489 Groupings!, 0.028 mins elapsed.\n",
      "\n",
      "2025-03-25 16:56:38.635398 : Getting Group RNA Matrix, 0.029 mins elapsed.\n",
      "\n",
      "2025-03-25 16:56:58.645332 : Getting Group ATAC Matrix, 0.362 mins elapsed.\n",
      "\n",
      "2025-03-25 16:57:54.148563 : Normalizing Group Matrices, 1.287 mins elapsed.\n",
      "\n",
      "2025-03-25 16:58:03.018608 : Finding Peak Gene Pairings, 1.435 mins elapsed.\n",
      "\n",
      "2025-03-25 16:58:04.2269 : Computing Correlations, 1.455 mins elapsed.\n",
      "\n",
      "2025-03-25 16:58:23.026047 : Completed Peak2Gene Correlations!, 1.769 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeak2GeneLinks-2b40a56b06f58f-Date-2025-03-25_Time-16-56-36.17321.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "obj.ArchR = addPeak2GeneLinks(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    reducedDims = \"IterativeLSI\",\n",
    "    useMatrix = \"GeneExpressionMatrix\",\n",
    "    maxDist = maxDist,\n",
    "    threads = n.cores\n",
    ")\n",
    "end_time <- Sys.time()\n",
    "execution_time <- end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "128d0065-5f22-495b-b00d-743fc9843337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 1.781427 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "execution_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "36bd2df2-6b32-4b81-bd4f-39ef8913b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g = metadata(obj.ArchR@peakSet)$Peak2GeneLinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e7730413-8fd4-4952-bb13-0df7a3b15b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata(p2g)[[1]]$PeakName = paste(seqnames(metadata(p2g)[[1]]),\n",
    "                                    ranges(metadata(p2g)[[1]]),sep=\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "aeea3810-6735-4aff-9bd7-1697bef979d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g$PairName = paste(metadata(p2g)[[1]][p2g$idxATAC]$PeakName,\n",
    "                     metadata(p2g)[[2]][p2g$idxRNA]$name,\n",
    "                     sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "87f975ec-6ba4-4696-ad6c-2af65da7aa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 2657764 rows and 7 columns\n",
       "                                idxATAC    idxRNA Correlation         FDR\n",
       "                              <integer> <integer>   <numeric>   <numeric>\n",
       "chr1-10021-10505_LINC00115            1         1  -0.1141958  0.03608648\n",
       "chr1-180699-181904_LINC00115          2         1   0.1644095  0.00169764\n",
       "chr1-191227-191617_LINC00115          3         1   0.0509998  0.39804127\n",
       "chr1-267886-268117_LINC00115          4         1   0.0578537  0.32961268\n",
       "chr1-778320-779311_LINC00115          5         1  -0.1181020  0.02954749\n",
       "...                                 ...       ...         ...         ...\n",
       "chrX-155966905-155967206_IL9R    152400     17509  -0.0042053 9.54278e-01\n",
       "chrX-155971502-155972157_IL9R    152401     17509   0.0704830 2.22976e-01\n",
       "chrX-156000454-156000883_IL9R    152402     17509  -0.0390446 5.31823e-01\n",
       "chrX-156029831-156030296_IL9R    152403     17509  -0.0449059 4.64193e-01\n",
       "chrX-156030503-156030719_IL9R    152404     17509   0.3442538 1.19022e-12\n",
       "                               VarQATAC   VarQRNA               PairName\n",
       "                              <numeric> <numeric>            <character>\n",
       "chr1-10021-10505_LINC00115     0.220053  0.138043 chr1-10021-10505_LIN..\n",
       "chr1-180699-181904_LINC00115   0.825313  0.138043 chr1-180699-181904_L..\n",
       "chr1-191227-191617_LINC00115   0.176852  0.138043 chr1-191227-191617_L..\n",
       "chr1-267886-268117_LINC00115   0.305458  0.138043 chr1-267886-268117_L..\n",
       "chr1-778320-779311_LINC00115   0.941196  0.138043 chr1-778320-779311_L..\n",
       "...                                 ...       ...                    ...\n",
       "chrX-155966905-155967206_IL9R 0.6070116 0.0210178 chrX-155966905-15596..\n",
       "chrX-155971502-155972157_IL9R 0.1229758 0.0210178 chrX-155971502-15597..\n",
       "chrX-156000454-156000883_IL9R 0.0697882 0.0210178 chrX-156000454-15600..\n",
       "chrX-156029831-156030296_IL9R 0.0851618 0.0210178 chrX-156029831-15603..\n",
       "chrX-156030503-156030719_IL9R 0.0287197 0.0210178 chrX-156030503-15603.."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(p2g) = p2g$PairName\n",
    "p2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c7209a32-3904-4734-93af-9c22049f652a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = pairs.E2G[pairs.E2G$PairName %in% p2g$PairName]\n",
    "mcols(pairs.E2G.res)[,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")] = p2g[pairs.E2G.res$PairName,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "1ab4ed70-ce9e-4857-bd1b-63a81d27a401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2154019 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100034573-100034728      * |         AGL\n",
       "        [2]     chr1 100034573-100034728      * |      CDC14A\n",
       "        [3]     chr1 100034573-100034728      * |         DBT\n",
       "        [4]     chr1 100034573-100034728      * |        DPH5\n",
       "        [5]     chr1 100034573-100034728      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2154015]     chrX     9995697-9996195      * |       CLCN4\n",
       "  [2154016]     chrX     9995697-9996195      * |      GPR143\n",
       "  [2154017]     chrX     9995697-9996195      * |     SHROOM2\n",
       "  [2154018]     chrX     9995697-9996195      * |       TBL1X\n",
       "  [2154019]     chrX     9995697-9996195      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-100034573-10003.. chr1-100034573-10003..  -0.0436631\n",
       "        [2] chr1-100034573-10003.. chr1-100034573-10003..   0.1292986\n",
       "        [3] chr1-100034573-10003.. chr1-100034573-10003..  -0.0286173\n",
       "        [4] chr1-100034573-10003.. chr1-100034573-10003..   0.0842369\n",
       "        [5] chr1-100034573-10003.. chr1-100034573-10003..   0.1165768\n",
       "        ...                    ...                    ...         ...\n",
       "  [2154015]   chrX-9995697-9996195 chrX-9995697-9996195..   0.1802298\n",
       "  [2154016]   chrX-9995697-9996195 chrX-9995697-9996195..   0.0389414\n",
       "  [2154017]   chrX-9995697-9996195 chrX-9995697-9996195..   0.0941964\n",
       "  [2154018]   chrX-9995697-9996195 chrX-9995697-9996195..   0.0352221\n",
       "  [2154019]   chrX-9995697-9996195 chrX-9995697-9996195..  -0.0768434\n",
       "                    FDR  VarQATAC   VarQRNA\n",
       "              <numeric> <numeric> <numeric>\n",
       "        [1]   0.4782771 0.0424005  0.824376\n",
       "        [2]   0.0161018 0.0424005  0.857787\n",
       "        [3]   0.6586870 0.0424005  0.637558\n",
       "        [4]   0.1361776 0.0424005  0.731567\n",
       "        [5]   0.0319716 0.0424005  0.862071\n",
       "        ...         ...       ...       ...\n",
       "  [2154015] 0.000515646  0.718419  0.451825\n",
       "  [2154016] 0.533041457  0.718419  0.546290\n",
       "  [2154017] 0.091070120  0.718419  0.560854\n",
       "  [2154018] 0.577580154  0.718419  0.898167\n",
       "  [2154019] 0.179091195  0.718419  0.709407\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46485351-50d7-461a-bf09-d795d5bdab84",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6d424e21-6a41-4422-b76c-6f1c0bda2222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2154019 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100034573-100034728      * |         AGL\n",
       "        [2]     chr1 100034573-100034728      * |      CDC14A\n",
       "        [3]     chr1 100034573-100034728      * |         DBT\n",
       "        [4]     chr1 100034573-100034728      * |        DPH5\n",
       "        [5]     chr1 100034573-100034728      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2154015]     chrX     9995697-9996195      * |       CLCN4\n",
       "  [2154016]     chrX     9995697-9996195      * |      GPR143\n",
       "  [2154017]     chrX     9995697-9996195      * |     SHROOM2\n",
       "  [2154018]     chrX     9995697-9996195      * |       TBL1X\n",
       "  [2154019]     chrX     9995697-9996195      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-100034573-10003.. chr1-100034573-10003..  -0.0436631\n",
       "        [2] chr1-100034573-10003.. chr1-100034573-10003..   0.1292986\n",
       "        [3] chr1-100034573-10003.. chr1-100034573-10003..  -0.0286173\n",
       "        [4] chr1-100034573-10003.. chr1-100034573-10003..   0.0842369\n",
       "        [5] chr1-100034573-10003.. chr1-100034573-10003..   0.1165768\n",
       "        ...                    ...                    ...         ...\n",
       "  [2154015]   chrX-9995697-9996195 chrX-9995697-9996195..   0.1802298\n",
       "  [2154016]   chrX-9995697-9996195 chrX-9995697-9996195..   0.0389414\n",
       "  [2154017]   chrX-9995697-9996195 chrX-9995697-9996195..   0.0941964\n",
       "  [2154018]   chrX-9995697-9996195 chrX-9995697-9996195..   0.0352221\n",
       "  [2154019]   chrX-9995697-9996195 chrX-9995697-9996195..  -0.0768434\n",
       "                    FDR  VarQATAC   VarQRNA\n",
       "              <numeric> <numeric> <numeric>\n",
       "        [1]   0.4782771 0.0424005  0.824376\n",
       "        [2]   0.0161018 0.0424005  0.857787\n",
       "        [3]   0.6586870 0.0424005  0.637558\n",
       "        [4]   0.1361776 0.0424005  0.731567\n",
       "        [5]   0.0319716 0.0424005  0.862071\n",
       "        ...         ...       ...       ...\n",
       "  [2154015] 0.000515646  0.718419  0.451825\n",
       "  [2154016] 0.533041457  0.718419  0.546290\n",
       "  [2154017] 0.091070120  0.718419  0.560854\n",
       "  [2154018] 0.577580154  0.718419  0.898167\n",
       "  [2154019] 0.179091195  0.718419  0.709407\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2154019 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>Correlation</th><th scope=col>FDR</th><th scope=col>VarQATAC</th><th scope=col>VarQRNA</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>AGL     </td><td>WTC11_cluster5</td><td>-0.0436631367</td><td>4.782771e-01</td><td>0.04240046</td><td>0.8243760</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>CDC14A  </td><td>WTC11_cluster5</td><td> 0.1292985683</td><td>1.610183e-02</td><td>0.04240046</td><td>0.8577874</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>DBT     </td><td>WTC11_cluster5</td><td>-0.0286173399</td><td>6.586870e-01</td><td>0.04240046</td><td>0.6375578</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>DPH5    </td><td>WTC11_cluster5</td><td> 0.0842369437</td><td>1.361776e-01</td><td>0.04240046</td><td>0.7315666</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>EXTL2   </td><td>WTC11_cluster5</td><td> 0.1165767724</td><td>3.197162e-02</td><td>0.04240046</td><td>0.8620709</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>MFSD14A </td><td>WTC11_cluster5</td><td>-0.0036682992</td><td>9.602563e-01</td><td>0.04240046</td><td>0.7301388</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>PLPPR4  </td><td>WTC11_cluster5</td><td> 0.0140590809</td><td>8.389499e-01</td><td>0.04240046</td><td>0.2813981</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>RTCA    </td><td>WTC11_cluster5</td><td>-0.1676390107</td><td>1.344091e-03</td><td>0.04240046</td><td>0.7156891</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>RTCA-AS1</td><td>WTC11_cluster5</td><td>-0.0000366005</td><td>9.996019e-01</td><td>0.04240046</td><td>0.2836256</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>SASS6   </td><td>WTC11_cluster5</td><td> 0.1291174213</td><td>1.626581e-02</td><td>0.04240046</td><td>0.8397967</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>SLC30A7 </td><td>WTC11_cluster5</td><td> 0.1257370369</td><td>1.964259e-02</td><td>0.04240046</td><td>0.8011308</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>SLC35A3 </td><td>WTC11_cluster5</td><td> 0.1039480771</td><td>5.914219e-02</td><td>0.04240046</td><td>0.8179793</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>TRMT13  </td><td>WTC11_cluster5</td><td> 0.1416209101</td><td>7.771283e-03</td><td>0.04240046</td><td>0.7520704</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>AGL     </td><td>WTC11_cluster5</td><td> 0.2217185446</td><td>1.296994e-05</td><td>0.92519225</td><td>0.8243760</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>CDC14A  </td><td>WTC11_cluster5</td><td>-0.0042253361</td><td>9.540620e-01</td><td>0.92519225</td><td>0.8577874</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>DBT     </td><td>WTC11_cluster5</td><td> 0.0822803831</td><td>1.467216e-01</td><td>0.92519225</td><td>0.6375578</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>DPH5    </td><td>WTC11_cluster5</td><td>-0.1210332343</td><td>2.533576e-02</td><td>0.92519225</td><td>0.7315666</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>EXTL2   </td><td>WTC11_cluster5</td><td>-0.0849823190</td><td>1.322990e-01</td><td>0.92519225</td><td>0.8620709</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>MFSD14A </td><td>WTC11_cluster5</td><td> 0.0959792255</td><td>8.442660e-02</td><td>0.92519225</td><td>0.7301388</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>PLPPR4  </td><td>WTC11_cluster5</td><td>-0.0449581666</td><td>4.636104e-01</td><td>0.92519225</td><td>0.2813981</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>RTCA    </td><td>WTC11_cluster5</td><td> 0.1602238501</td><td>2.284494e-03</td><td>0.92519225</td><td>0.7156891</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>RTCA-AS1</td><td>WTC11_cluster5</td><td>-0.0334886215</td><td>5.986461e-01</td><td>0.92519225</td><td>0.2836256</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>SASS6   </td><td>WTC11_cluster5</td><td>-0.0085833311</td><td>9.042741e-01</td><td>0.92519225</td><td>0.8397967</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>SLC30A7 </td><td>WTC11_cluster5</td><td>-0.0057909173</td><td>9.363773e-01</td><td>0.92519225</td><td>0.8011308</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>SLC35A3 </td><td>WTC11_cluster5</td><td> 0.0227404008</td><td>7.318988e-01</td><td>0.92519225</td><td>0.8179793</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>TRMT13  </td><td>WTC11_cluster5</td><td>-0.0145714531</td><td>8.327468e-01</td><td>0.92519225</td><td>0.7520704</td></tr>\n",
       "\t<tr><td>chr1</td><td>100049383</td><td>100049815</td><td>AGL     </td><td>WTC11_cluster5</td><td> 0.2032411099</td><td>7.390019e-05</td><td>0.30306291</td><td>0.8243760</td></tr>\n",
       "\t<tr><td>chr1</td><td>100049383</td><td>100049815</td><td>CDC14A  </td><td>WTC11_cluster5</td><td>-0.1574957830</td><td>2.759017e-03</td><td>0.30306291</td><td>0.8577874</td></tr>\n",
       "\t<tr><td>chr1</td><td>100049383</td><td>100049815</td><td>DBT     </td><td>WTC11_cluster5</td><td>-0.0459677483</td><td>4.523386e-01</td><td>0.30306291</td><td>0.6375578</td></tr>\n",
       "\t<tr><td>chr1</td><td>100049383</td><td>100049815</td><td>DPH5    </td><td>WTC11_cluster5</td><td>-0.0931370503</td><td>9.520986e-02</td><td>0.30306291</td><td>0.7315666</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9959363</td><td> 9959794</td><td>CLCN4  </td><td>WTC11_cluster5</td><td> 0.12428255</td><td>0.0212719838</td><td>0.01940894</td><td>0.45182478</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9959363</td><td> 9959794</td><td>GPR143 </td><td>WTC11_cluster5</td><td> 0.16598025</td><td>0.0015162579</td><td>0.01940894</td><td>0.54629048</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9959363</td><td> 9959794</td><td>SHROOM2</td><td>WTC11_cluster5</td><td>-0.04925139</td><td>0.4166074721</td><td>0.01940894</td><td>0.56085442</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9959363</td><td> 9959794</td><td>TBL1X  </td><td>WTC11_cluster5</td><td>-0.05459818</td><td>0.3612279064</td><td>0.01940894</td><td>0.89816666</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9959363</td><td> 9959794</td><td>WWC3   </td><td>WTC11_cluster5</td><td>-0.19886584</td><td>0.0001090320</td><td>0.01940894</td><td>0.70940659</td></tr>\n",
       "\t<tr><td>chrX</td><td>99603193</td><td>99603678</td><td>PCDH19 </td><td>WTC11_cluster5</td><td>-0.08599342</td><td>0.1271751248</td><td>0.09914438</td><td>0.38340282</td></tr>\n",
       "\t<tr><td>chrX</td><td>99722173</td><td>99722445</td><td>PCDH19 </td><td>WTC11_cluster5</td><td>-0.14456944</td><td>0.0064621144</td><td>0.03349650</td><td>0.38340282</td></tr>\n",
       "\t<tr><td>chrX</td><td>99722173</td><td>99722445</td><td>TSPAN6 </td><td>WTC11_cluster5</td><td>-0.05347348</td><td>0.3725406655</td><td>0.03349650</td><td>0.73139528</td></tr>\n",
       "\t<tr><td>chrX</td><td>  997278</td><td>  998378</td><td>AKAP17A</td><td>WTC11_cluster5</td><td> 0.06391359</td><td>0.2752157467</td><td>0.64267342</td><td>0.63710092</td></tr>\n",
       "\t<tr><td>chrX</td><td>  997278</td><td>  998378</td><td>ASMTL  </td><td>WTC11_cluster5</td><td>-0.14908233</td><td>0.0048393800</td><td>0.64267342</td><td>0.71660289</td></tr>\n",
       "\t<tr><td>chrX</td><td>  997278</td><td>  998378</td><td>PLCXD1 </td><td>WTC11_cluster5</td><td> 0.06524432</td><td>0.2640904912</td><td>0.64267342</td><td>0.79153578</td></tr>\n",
       "\t<tr><td>chrX</td><td>  997278</td><td>  998378</td><td>PPP2R3B</td><td>WTC11_cluster5</td><td> 0.02146459</td><td>0.7477904600</td><td>0.64267342</td><td>0.76263636</td></tr>\n",
       "\t<tr><td>chrX</td><td>  997278</td><td>  998378</td><td>SHOX   </td><td>WTC11_cluster5</td><td>-0.10556719</td><td>0.0548485720</td><td>0.64267342</td><td>0.02787138</td></tr>\n",
       "\t<tr><td>chrX</td><td>  997278</td><td>  998378</td><td>SLC25A6</td><td>WTC11_cluster5</td><td>-0.12298267</td><td>0.0228267126</td><td>0.64267342</td><td>0.81500942</td></tr>\n",
       "\t<tr><td>chrX</td><td>99736932</td><td>99737301</td><td>PCDH19 </td><td>WTC11_cluster5</td><td>-0.19437227</td><td>0.0001608724</td><td>0.01893651</td><td>0.38340282</td></tr>\n",
       "\t<tr><td>chrX</td><td>99736932</td><td>99737301</td><td>SYTL4  </td><td>WTC11_cluster5</td><td> 0.05171126</td><td>0.3906047715</td><td>0.01893651</td><td>0.18727512</td></tr>\n",
       "\t<tr><td>chrX</td><td>99736932</td><td>99737301</td><td>TSPAN6 </td><td>WTC11_cluster5</td><td>-0.04478881</td><td>0.4655027024</td><td>0.01893651</td><td>0.73139528</td></tr>\n",
       "\t<tr><td>chrX</td><td>99777683</td><td>99778218</td><td>PCDH19 </td><td>WTC11_cluster5</td><td>-0.15480879</td><td>0.0033128878</td><td>0.02204667</td><td>0.38340282</td></tr>\n",
       "\t<tr><td>chrX</td><td>99777683</td><td>99778218</td><td>SYTL4  </td><td>WTC11_cluster5</td><td> 0.15785553</td><td>0.0026918960</td><td>0.02204667</td><td>0.18727512</td></tr>\n",
       "\t<tr><td>chrX</td><td>99777683</td><td>99778218</td><td>TSPAN6 </td><td>WTC11_cluster5</td><td> 0.09397730</td><td>0.0919082071</td><td>0.02204667</td><td>0.73139528</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9981419</td><td> 9981826</td><td>CLCN4  </td><td>WTC11_cluster5</td><td>-0.01996782</td><td>0.7663041389</td><td>0.01017690</td><td>0.45182478</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9981419</td><td> 9981826</td><td>GPR143 </td><td>WTC11_cluster5</td><td> 0.06241711</td><td>0.2881017563</td><td>0.01017690</td><td>0.54629048</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9981419</td><td> 9981826</td><td>SHROOM2</td><td>WTC11_cluster5</td><td>-0.11495784</td><td>0.0347235127</td><td>0.01017690</td><td>0.56085442</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9981419</td><td> 9981826</td><td>TBL1X  </td><td>WTC11_cluster5</td><td> 0.11202722</td><td>0.0401985549</td><td>0.01017690</td><td>0.89816666</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9981419</td><td> 9981826</td><td>WWC3   </td><td>WTC11_cluster5</td><td> 0.11260583</td><td>0.0390688375</td><td>0.01017690</td><td>0.70940659</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995697</td><td> 9996195</td><td>CLCN4  </td><td>WTC11_cluster5</td><td> 0.18022978</td><td>0.0005156461</td><td>0.71841946</td><td>0.45182478</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995697</td><td> 9996195</td><td>GPR143 </td><td>WTC11_cluster5</td><td> 0.03894136</td><td>0.5330414571</td><td>0.71841946</td><td>0.54629048</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995697</td><td> 9996195</td><td>SHROOM2</td><td>WTC11_cluster5</td><td> 0.09419636</td><td>0.0910701205</td><td>0.71841946</td><td>0.56085442</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995697</td><td> 9996195</td><td>TBL1X  </td><td>WTC11_cluster5</td><td> 0.03522214</td><td>0.5775801538</td><td>0.71841946</td><td>0.89816666</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995697</td><td> 9996195</td><td>WWC3   </td><td>WTC11_cluster5</td><td>-0.07684341</td><td>0.1790911951</td><td>0.71841946</td><td>0.70940659</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2154019 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & Correlation & FDR & VarQATAC & VarQRNA\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100034573 & 100034728 & AGL      & WTC11\\_cluster5 & -0.0436631367 & 4.782771e-01 & 0.04240046 & 0.8243760\\\\\n",
       "\t chr1 & 100034573 & 100034728 & CDC14A   & WTC11\\_cluster5 &  0.1292985683 & 1.610183e-02 & 0.04240046 & 0.8577874\\\\\n",
       "\t chr1 & 100034573 & 100034728 & DBT      & WTC11\\_cluster5 & -0.0286173399 & 6.586870e-01 & 0.04240046 & 0.6375578\\\\\n",
       "\t chr1 & 100034573 & 100034728 & DPH5     & WTC11\\_cluster5 &  0.0842369437 & 1.361776e-01 & 0.04240046 & 0.7315666\\\\\n",
       "\t chr1 & 100034573 & 100034728 & EXTL2    & WTC11\\_cluster5 &  0.1165767724 & 3.197162e-02 & 0.04240046 & 0.8620709\\\\\n",
       "\t chr1 & 100034573 & 100034728 & MFSD14A  & WTC11\\_cluster5 & -0.0036682992 & 9.602563e-01 & 0.04240046 & 0.7301388\\\\\n",
       "\t chr1 & 100034573 & 100034728 & PLPPR4   & WTC11\\_cluster5 &  0.0140590809 & 8.389499e-01 & 0.04240046 & 0.2813981\\\\\n",
       "\t chr1 & 100034573 & 100034728 & RTCA     & WTC11\\_cluster5 & -0.1676390107 & 1.344091e-03 & 0.04240046 & 0.7156891\\\\\n",
       "\t chr1 & 100034573 & 100034728 & RTCA-AS1 & WTC11\\_cluster5 & -0.0000366005 & 9.996019e-01 & 0.04240046 & 0.2836256\\\\\n",
       "\t chr1 & 100034573 & 100034728 & SASS6    & WTC11\\_cluster5 &  0.1291174213 & 1.626581e-02 & 0.04240046 & 0.8397967\\\\\n",
       "\t chr1 & 100034573 & 100034728 & SLC30A7  & WTC11\\_cluster5 &  0.1257370369 & 1.964259e-02 & 0.04240046 & 0.8011308\\\\\n",
       "\t chr1 & 100034573 & 100034728 & SLC35A3  & WTC11\\_cluster5 &  0.1039480771 & 5.914219e-02 & 0.04240046 & 0.8179793\\\\\n",
       "\t chr1 & 100034573 & 100034728 & TRMT13   & WTC11\\_cluster5 &  0.1416209101 & 7.771283e-03 & 0.04240046 & 0.7520704\\\\\n",
       "\t chr1 & 100037552 & 100039045 & AGL      & WTC11\\_cluster5 &  0.2217185446 & 1.296994e-05 & 0.92519225 & 0.8243760\\\\\n",
       "\t chr1 & 100037552 & 100039045 & CDC14A   & WTC11\\_cluster5 & -0.0042253361 & 9.540620e-01 & 0.92519225 & 0.8577874\\\\\n",
       "\t chr1 & 100037552 & 100039045 & DBT      & WTC11\\_cluster5 &  0.0822803831 & 1.467216e-01 & 0.92519225 & 0.6375578\\\\\n",
       "\t chr1 & 100037552 & 100039045 & DPH5     & WTC11\\_cluster5 & -0.1210332343 & 2.533576e-02 & 0.92519225 & 0.7315666\\\\\n",
       "\t chr1 & 100037552 & 100039045 & EXTL2    & WTC11\\_cluster5 & -0.0849823190 & 1.322990e-01 & 0.92519225 & 0.8620709\\\\\n",
       "\t chr1 & 100037552 & 100039045 & MFSD14A  & WTC11\\_cluster5 &  0.0959792255 & 8.442660e-02 & 0.92519225 & 0.7301388\\\\\n",
       "\t chr1 & 100037552 & 100039045 & PLPPR4   & WTC11\\_cluster5 & -0.0449581666 & 4.636104e-01 & 0.92519225 & 0.2813981\\\\\n",
       "\t chr1 & 100037552 & 100039045 & RTCA     & WTC11\\_cluster5 &  0.1602238501 & 2.284494e-03 & 0.92519225 & 0.7156891\\\\\n",
       "\t chr1 & 100037552 & 100039045 & RTCA-AS1 & WTC11\\_cluster5 & -0.0334886215 & 5.986461e-01 & 0.92519225 & 0.2836256\\\\\n",
       "\t chr1 & 100037552 & 100039045 & SASS6    & WTC11\\_cluster5 & -0.0085833311 & 9.042741e-01 & 0.92519225 & 0.8397967\\\\\n",
       "\t chr1 & 100037552 & 100039045 & SLC30A7  & WTC11\\_cluster5 & -0.0057909173 & 9.363773e-01 & 0.92519225 & 0.8011308\\\\\n",
       "\t chr1 & 100037552 & 100039045 & SLC35A3  & WTC11\\_cluster5 &  0.0227404008 & 7.318988e-01 & 0.92519225 & 0.8179793\\\\\n",
       "\t chr1 & 100037552 & 100039045 & TRMT13   & WTC11\\_cluster5 & -0.0145714531 & 8.327468e-01 & 0.92519225 & 0.7520704\\\\\n",
       "\t chr1 & 100049383 & 100049815 & AGL      & WTC11\\_cluster5 &  0.2032411099 & 7.390019e-05 & 0.30306291 & 0.8243760\\\\\n",
       "\t chr1 & 100049383 & 100049815 & CDC14A   & WTC11\\_cluster5 & -0.1574957830 & 2.759017e-03 & 0.30306291 & 0.8577874\\\\\n",
       "\t chr1 & 100049383 & 100049815 & DBT      & WTC11\\_cluster5 & -0.0459677483 & 4.523386e-01 & 0.30306291 & 0.6375578\\\\\n",
       "\t chr1 & 100049383 & 100049815 & DPH5     & WTC11\\_cluster5 & -0.0931370503 & 9.520986e-02 & 0.30306291 & 0.7315666\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  9959363 &  9959794 & CLCN4   & WTC11\\_cluster5 &  0.12428255 & 0.0212719838 & 0.01940894 & 0.45182478\\\\\n",
       "\t chrX &  9959363 &  9959794 & GPR143  & WTC11\\_cluster5 &  0.16598025 & 0.0015162579 & 0.01940894 & 0.54629048\\\\\n",
       "\t chrX &  9959363 &  9959794 & SHROOM2 & WTC11\\_cluster5 & -0.04925139 & 0.4166074721 & 0.01940894 & 0.56085442\\\\\n",
       "\t chrX &  9959363 &  9959794 & TBL1X   & WTC11\\_cluster5 & -0.05459818 & 0.3612279064 & 0.01940894 & 0.89816666\\\\\n",
       "\t chrX &  9959363 &  9959794 & WWC3    & WTC11\\_cluster5 & -0.19886584 & 0.0001090320 & 0.01940894 & 0.70940659\\\\\n",
       "\t chrX & 99603193 & 99603678 & PCDH19  & WTC11\\_cluster5 & -0.08599342 & 0.1271751248 & 0.09914438 & 0.38340282\\\\\n",
       "\t chrX & 99722173 & 99722445 & PCDH19  & WTC11\\_cluster5 & -0.14456944 & 0.0064621144 & 0.03349650 & 0.38340282\\\\\n",
       "\t chrX & 99722173 & 99722445 & TSPAN6  & WTC11\\_cluster5 & -0.05347348 & 0.3725406655 & 0.03349650 & 0.73139528\\\\\n",
       "\t chrX &   997278 &   998378 & AKAP17A & WTC11\\_cluster5 &  0.06391359 & 0.2752157467 & 0.64267342 & 0.63710092\\\\\n",
       "\t chrX &   997278 &   998378 & ASMTL   & WTC11\\_cluster5 & -0.14908233 & 0.0048393800 & 0.64267342 & 0.71660289\\\\\n",
       "\t chrX &   997278 &   998378 & PLCXD1  & WTC11\\_cluster5 &  0.06524432 & 0.2640904912 & 0.64267342 & 0.79153578\\\\\n",
       "\t chrX &   997278 &   998378 & PPP2R3B & WTC11\\_cluster5 &  0.02146459 & 0.7477904600 & 0.64267342 & 0.76263636\\\\\n",
       "\t chrX &   997278 &   998378 & SHOX    & WTC11\\_cluster5 & -0.10556719 & 0.0548485720 & 0.64267342 & 0.02787138\\\\\n",
       "\t chrX &   997278 &   998378 & SLC25A6 & WTC11\\_cluster5 & -0.12298267 & 0.0228267126 & 0.64267342 & 0.81500942\\\\\n",
       "\t chrX & 99736932 & 99737301 & PCDH19  & WTC11\\_cluster5 & -0.19437227 & 0.0001608724 & 0.01893651 & 0.38340282\\\\\n",
       "\t chrX & 99736932 & 99737301 & SYTL4   & WTC11\\_cluster5 &  0.05171126 & 0.3906047715 & 0.01893651 & 0.18727512\\\\\n",
       "\t chrX & 99736932 & 99737301 & TSPAN6  & WTC11\\_cluster5 & -0.04478881 & 0.4655027024 & 0.01893651 & 0.73139528\\\\\n",
       "\t chrX & 99777683 & 99778218 & PCDH19  & WTC11\\_cluster5 & -0.15480879 & 0.0033128878 & 0.02204667 & 0.38340282\\\\\n",
       "\t chrX & 99777683 & 99778218 & SYTL4   & WTC11\\_cluster5 &  0.15785553 & 0.0026918960 & 0.02204667 & 0.18727512\\\\\n",
       "\t chrX & 99777683 & 99778218 & TSPAN6  & WTC11\\_cluster5 &  0.09397730 & 0.0919082071 & 0.02204667 & 0.73139528\\\\\n",
       "\t chrX &  9981419 &  9981826 & CLCN4   & WTC11\\_cluster5 & -0.01996782 & 0.7663041389 & 0.01017690 & 0.45182478\\\\\n",
       "\t chrX &  9981419 &  9981826 & GPR143  & WTC11\\_cluster5 &  0.06241711 & 0.2881017563 & 0.01017690 & 0.54629048\\\\\n",
       "\t chrX &  9981419 &  9981826 & SHROOM2 & WTC11\\_cluster5 & -0.11495784 & 0.0347235127 & 0.01017690 & 0.56085442\\\\\n",
       "\t chrX &  9981419 &  9981826 & TBL1X   & WTC11\\_cluster5 &  0.11202722 & 0.0401985549 & 0.01017690 & 0.89816666\\\\\n",
       "\t chrX &  9981419 &  9981826 & WWC3    & WTC11\\_cluster5 &  0.11260583 & 0.0390688375 & 0.01017690 & 0.70940659\\\\\n",
       "\t chrX &  9995697 &  9996195 & CLCN4   & WTC11\\_cluster5 &  0.18022978 & 0.0005156461 & 0.71841946 & 0.45182478\\\\\n",
       "\t chrX &  9995697 &  9996195 & GPR143  & WTC11\\_cluster5 &  0.03894136 & 0.5330414571 & 0.71841946 & 0.54629048\\\\\n",
       "\t chrX &  9995697 &  9996195 & SHROOM2 & WTC11\\_cluster5 &  0.09419636 & 0.0910701205 & 0.71841946 & 0.56085442\\\\\n",
       "\t chrX &  9995697 &  9996195 & TBL1X   & WTC11\\_cluster5 &  0.03522214 & 0.5775801538 & 0.71841946 & 0.89816666\\\\\n",
       "\t chrX &  9995697 &  9996195 & WWC3    & WTC11\\_cluster5 & -0.07684341 & 0.1790911951 & 0.71841946 & 0.70940659\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2154019 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | Correlation &lt;dbl&gt; | FDR &lt;dbl&gt; | VarQATAC &lt;dbl&gt; | VarQRNA &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100034573 | 100034728 | AGL      | WTC11_cluster5 | -0.0436631367 | 4.782771e-01 | 0.04240046 | 0.8243760 |\n",
       "| chr1 | 100034573 | 100034728 | CDC14A   | WTC11_cluster5 |  0.1292985683 | 1.610183e-02 | 0.04240046 | 0.8577874 |\n",
       "| chr1 | 100034573 | 100034728 | DBT      | WTC11_cluster5 | -0.0286173399 | 6.586870e-01 | 0.04240046 | 0.6375578 |\n",
       "| chr1 | 100034573 | 100034728 | DPH5     | WTC11_cluster5 |  0.0842369437 | 1.361776e-01 | 0.04240046 | 0.7315666 |\n",
       "| chr1 | 100034573 | 100034728 | EXTL2    | WTC11_cluster5 |  0.1165767724 | 3.197162e-02 | 0.04240046 | 0.8620709 |\n",
       "| chr1 | 100034573 | 100034728 | MFSD14A  | WTC11_cluster5 | -0.0036682992 | 9.602563e-01 | 0.04240046 | 0.7301388 |\n",
       "| chr1 | 100034573 | 100034728 | PLPPR4   | WTC11_cluster5 |  0.0140590809 | 8.389499e-01 | 0.04240046 | 0.2813981 |\n",
       "| chr1 | 100034573 | 100034728 | RTCA     | WTC11_cluster5 | -0.1676390107 | 1.344091e-03 | 0.04240046 | 0.7156891 |\n",
       "| chr1 | 100034573 | 100034728 | RTCA-AS1 | WTC11_cluster5 | -0.0000366005 | 9.996019e-01 | 0.04240046 | 0.2836256 |\n",
       "| chr1 | 100034573 | 100034728 | SASS6    | WTC11_cluster5 |  0.1291174213 | 1.626581e-02 | 0.04240046 | 0.8397967 |\n",
       "| chr1 | 100034573 | 100034728 | SLC30A7  | WTC11_cluster5 |  0.1257370369 | 1.964259e-02 | 0.04240046 | 0.8011308 |\n",
       "| chr1 | 100034573 | 100034728 | SLC35A3  | WTC11_cluster5 |  0.1039480771 | 5.914219e-02 | 0.04240046 | 0.8179793 |\n",
       "| chr1 | 100034573 | 100034728 | TRMT13   | WTC11_cluster5 |  0.1416209101 | 7.771283e-03 | 0.04240046 | 0.7520704 |\n",
       "| chr1 | 100037552 | 100039045 | AGL      | WTC11_cluster5 |  0.2217185446 | 1.296994e-05 | 0.92519225 | 0.8243760 |\n",
       "| chr1 | 100037552 | 100039045 | CDC14A   | WTC11_cluster5 | -0.0042253361 | 9.540620e-01 | 0.92519225 | 0.8577874 |\n",
       "| chr1 | 100037552 | 100039045 | DBT      | WTC11_cluster5 |  0.0822803831 | 1.467216e-01 | 0.92519225 | 0.6375578 |\n",
       "| chr1 | 100037552 | 100039045 | DPH5     | WTC11_cluster5 | -0.1210332343 | 2.533576e-02 | 0.92519225 | 0.7315666 |\n",
       "| chr1 | 100037552 | 100039045 | EXTL2    | WTC11_cluster5 | -0.0849823190 | 1.322990e-01 | 0.92519225 | 0.8620709 |\n",
       "| chr1 | 100037552 | 100039045 | MFSD14A  | WTC11_cluster5 |  0.0959792255 | 8.442660e-02 | 0.92519225 | 0.7301388 |\n",
       "| chr1 | 100037552 | 100039045 | PLPPR4   | WTC11_cluster5 | -0.0449581666 | 4.636104e-01 | 0.92519225 | 0.2813981 |\n",
       "| chr1 | 100037552 | 100039045 | RTCA     | WTC11_cluster5 |  0.1602238501 | 2.284494e-03 | 0.92519225 | 0.7156891 |\n",
       "| chr1 | 100037552 | 100039045 | RTCA-AS1 | WTC11_cluster5 | -0.0334886215 | 5.986461e-01 | 0.92519225 | 0.2836256 |\n",
       "| chr1 | 100037552 | 100039045 | SASS6    | WTC11_cluster5 | -0.0085833311 | 9.042741e-01 | 0.92519225 | 0.8397967 |\n",
       "| chr1 | 100037552 | 100039045 | SLC30A7  | WTC11_cluster5 | -0.0057909173 | 9.363773e-01 | 0.92519225 | 0.8011308 |\n",
       "| chr1 | 100037552 | 100039045 | SLC35A3  | WTC11_cluster5 |  0.0227404008 | 7.318988e-01 | 0.92519225 | 0.8179793 |\n",
       "| chr1 | 100037552 | 100039045 | TRMT13   | WTC11_cluster5 | -0.0145714531 | 8.327468e-01 | 0.92519225 | 0.7520704 |\n",
       "| chr1 | 100049383 | 100049815 | AGL      | WTC11_cluster5 |  0.2032411099 | 7.390019e-05 | 0.30306291 | 0.8243760 |\n",
       "| chr1 | 100049383 | 100049815 | CDC14A   | WTC11_cluster5 | -0.1574957830 | 2.759017e-03 | 0.30306291 | 0.8577874 |\n",
       "| chr1 | 100049383 | 100049815 | DBT      | WTC11_cluster5 | -0.0459677483 | 4.523386e-01 | 0.30306291 | 0.6375578 |\n",
       "| chr1 | 100049383 | 100049815 | DPH5     | WTC11_cluster5 | -0.0931370503 | 9.520986e-02 | 0.30306291 | 0.7315666 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  9959363 |  9959794 | CLCN4   | WTC11_cluster5 |  0.12428255 | 0.0212719838 | 0.01940894 | 0.45182478 |\n",
       "| chrX |  9959363 |  9959794 | GPR143  | WTC11_cluster5 |  0.16598025 | 0.0015162579 | 0.01940894 | 0.54629048 |\n",
       "| chrX |  9959363 |  9959794 | SHROOM2 | WTC11_cluster5 | -0.04925139 | 0.4166074721 | 0.01940894 | 0.56085442 |\n",
       "| chrX |  9959363 |  9959794 | TBL1X   | WTC11_cluster5 | -0.05459818 | 0.3612279064 | 0.01940894 | 0.89816666 |\n",
       "| chrX |  9959363 |  9959794 | WWC3    | WTC11_cluster5 | -0.19886584 | 0.0001090320 | 0.01940894 | 0.70940659 |\n",
       "| chrX | 99603193 | 99603678 | PCDH19  | WTC11_cluster5 | -0.08599342 | 0.1271751248 | 0.09914438 | 0.38340282 |\n",
       "| chrX | 99722173 | 99722445 | PCDH19  | WTC11_cluster5 | -0.14456944 | 0.0064621144 | 0.03349650 | 0.38340282 |\n",
       "| chrX | 99722173 | 99722445 | TSPAN6  | WTC11_cluster5 | -0.05347348 | 0.3725406655 | 0.03349650 | 0.73139528 |\n",
       "| chrX |   997278 |   998378 | AKAP17A | WTC11_cluster5 |  0.06391359 | 0.2752157467 | 0.64267342 | 0.63710092 |\n",
       "| chrX |   997278 |   998378 | ASMTL   | WTC11_cluster5 | -0.14908233 | 0.0048393800 | 0.64267342 | 0.71660289 |\n",
       "| chrX |   997278 |   998378 | PLCXD1  | WTC11_cluster5 |  0.06524432 | 0.2640904912 | 0.64267342 | 0.79153578 |\n",
       "| chrX |   997278 |   998378 | PPP2R3B | WTC11_cluster5 |  0.02146459 | 0.7477904600 | 0.64267342 | 0.76263636 |\n",
       "| chrX |   997278 |   998378 | SHOX    | WTC11_cluster5 | -0.10556719 | 0.0548485720 | 0.64267342 | 0.02787138 |\n",
       "| chrX |   997278 |   998378 | SLC25A6 | WTC11_cluster5 | -0.12298267 | 0.0228267126 | 0.64267342 | 0.81500942 |\n",
       "| chrX | 99736932 | 99737301 | PCDH19  | WTC11_cluster5 | -0.19437227 | 0.0001608724 | 0.01893651 | 0.38340282 |\n",
       "| chrX | 99736932 | 99737301 | SYTL4   | WTC11_cluster5 |  0.05171126 | 0.3906047715 | 0.01893651 | 0.18727512 |\n",
       "| chrX | 99736932 | 99737301 | TSPAN6  | WTC11_cluster5 | -0.04478881 | 0.4655027024 | 0.01893651 | 0.73139528 |\n",
       "| chrX | 99777683 | 99778218 | PCDH19  | WTC11_cluster5 | -0.15480879 | 0.0033128878 | 0.02204667 | 0.38340282 |\n",
       "| chrX | 99777683 | 99778218 | SYTL4   | WTC11_cluster5 |  0.15785553 | 0.0026918960 | 0.02204667 | 0.18727512 |\n",
       "| chrX | 99777683 | 99778218 | TSPAN6  | WTC11_cluster5 |  0.09397730 | 0.0919082071 | 0.02204667 | 0.73139528 |\n",
       "| chrX |  9981419 |  9981826 | CLCN4   | WTC11_cluster5 | -0.01996782 | 0.7663041389 | 0.01017690 | 0.45182478 |\n",
       "| chrX |  9981419 |  9981826 | GPR143  | WTC11_cluster5 |  0.06241711 | 0.2881017563 | 0.01017690 | 0.54629048 |\n",
       "| chrX |  9981419 |  9981826 | SHROOM2 | WTC11_cluster5 | -0.11495784 | 0.0347235127 | 0.01017690 | 0.56085442 |\n",
       "| chrX |  9981419 |  9981826 | TBL1X   | WTC11_cluster5 |  0.11202722 | 0.0401985549 | 0.01017690 | 0.89816666 |\n",
       "| chrX |  9981419 |  9981826 | WWC3    | WTC11_cluster5 |  0.11260583 | 0.0390688375 | 0.01017690 | 0.70940659 |\n",
       "| chrX |  9995697 |  9996195 | CLCN4   | WTC11_cluster5 |  0.18022978 | 0.0005156461 | 0.71841946 | 0.45182478 |\n",
       "| chrX |  9995697 |  9996195 | GPR143  | WTC11_cluster5 |  0.03894136 | 0.5330414571 | 0.71841946 | 0.54629048 |\n",
       "| chrX |  9995697 |  9996195 | SHROOM2 | WTC11_cluster5 |  0.09419636 | 0.0910701205 | 0.71841946 | 0.56085442 |\n",
       "| chrX |  9995697 |  9996195 | TBL1X   | WTC11_cluster5 |  0.03522214 | 0.5775801538 | 0.71841946 | 0.89816666 |\n",
       "| chrX |  9995697 |  9996195 | WWC3    | WTC11_cluster5 | -0.07684341 | 0.1790911951 | 0.71841946 | 0.70940659 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType       Correlation  \n",
       "1       chr1 100034573 100034728 AGL        WTC11_cluster5 -0.0436631367\n",
       "2       chr1 100034573 100034728 CDC14A     WTC11_cluster5  0.1292985683\n",
       "3       chr1 100034573 100034728 DBT        WTC11_cluster5 -0.0286173399\n",
       "4       chr1 100034573 100034728 DPH5       WTC11_cluster5  0.0842369437\n",
       "5       chr1 100034573 100034728 EXTL2      WTC11_cluster5  0.1165767724\n",
       "6       chr1 100034573 100034728 MFSD14A    WTC11_cluster5 -0.0036682992\n",
       "7       chr1 100034573 100034728 PLPPR4     WTC11_cluster5  0.0140590809\n",
       "8       chr1 100034573 100034728 RTCA       WTC11_cluster5 -0.1676390107\n",
       "9       chr1 100034573 100034728 RTCA-AS1   WTC11_cluster5 -0.0000366005\n",
       "10      chr1 100034573 100034728 SASS6      WTC11_cluster5  0.1291174213\n",
       "11      chr1 100034573 100034728 SLC30A7    WTC11_cluster5  0.1257370369\n",
       "12      chr1 100034573 100034728 SLC35A3    WTC11_cluster5  0.1039480771\n",
       "13      chr1 100034573 100034728 TRMT13     WTC11_cluster5  0.1416209101\n",
       "14      chr1 100037552 100039045 AGL        WTC11_cluster5  0.2217185446\n",
       "15      chr1 100037552 100039045 CDC14A     WTC11_cluster5 -0.0042253361\n",
       "16      chr1 100037552 100039045 DBT        WTC11_cluster5  0.0822803831\n",
       "17      chr1 100037552 100039045 DPH5       WTC11_cluster5 -0.1210332343\n",
       "18      chr1 100037552 100039045 EXTL2      WTC11_cluster5 -0.0849823190\n",
       "19      chr1 100037552 100039045 MFSD14A    WTC11_cluster5  0.0959792255\n",
       "20      chr1 100037552 100039045 PLPPR4     WTC11_cluster5 -0.0449581666\n",
       "21      chr1 100037552 100039045 RTCA       WTC11_cluster5  0.1602238501\n",
       "22      chr1 100037552 100039045 RTCA-AS1   WTC11_cluster5 -0.0334886215\n",
       "23      chr1 100037552 100039045 SASS6      WTC11_cluster5 -0.0085833311\n",
       "24      chr1 100037552 100039045 SLC30A7    WTC11_cluster5 -0.0057909173\n",
       "25      chr1 100037552 100039045 SLC35A3    WTC11_cluster5  0.0227404008\n",
       "26      chr1 100037552 100039045 TRMT13     WTC11_cluster5 -0.0145714531\n",
       "27      chr1 100049383 100049815 AGL        WTC11_cluster5  0.2032411099\n",
       "28      chr1 100049383 100049815 CDC14A     WTC11_cluster5 -0.1574957830\n",
       "29      chr1 100049383 100049815 DBT        WTC11_cluster5 -0.0459677483\n",
       "30      chr1 100049383 100049815 DPH5       WTC11_cluster5 -0.0931370503\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮              ⋮            \n",
       "2153990 chrX  9959363   9959794  CLCN4      WTC11_cluster5  0.12428255  \n",
       "2153991 chrX  9959363   9959794  GPR143     WTC11_cluster5  0.16598025  \n",
       "2153992 chrX  9959363   9959794  SHROOM2    WTC11_cluster5 -0.04925139  \n",
       "2153993 chrX  9959363   9959794  TBL1X      WTC11_cluster5 -0.05459818  \n",
       "2153994 chrX  9959363   9959794  WWC3       WTC11_cluster5 -0.19886584  \n",
       "2153995 chrX 99603193  99603678  PCDH19     WTC11_cluster5 -0.08599342  \n",
       "2153996 chrX 99722173  99722445  PCDH19     WTC11_cluster5 -0.14456944  \n",
       "2153997 chrX 99722173  99722445  TSPAN6     WTC11_cluster5 -0.05347348  \n",
       "2153998 chrX   997278    998378  AKAP17A    WTC11_cluster5  0.06391359  \n",
       "2153999 chrX   997278    998378  ASMTL      WTC11_cluster5 -0.14908233  \n",
       "2154000 chrX   997278    998378  PLCXD1     WTC11_cluster5  0.06524432  \n",
       "2154001 chrX   997278    998378  PPP2R3B    WTC11_cluster5  0.02146459  \n",
       "2154002 chrX   997278    998378  SHOX       WTC11_cluster5 -0.10556719  \n",
       "2154003 chrX   997278    998378  SLC25A6    WTC11_cluster5 -0.12298267  \n",
       "2154004 chrX 99736932  99737301  PCDH19     WTC11_cluster5 -0.19437227  \n",
       "2154005 chrX 99736932  99737301  SYTL4      WTC11_cluster5  0.05171126  \n",
       "2154006 chrX 99736932  99737301  TSPAN6     WTC11_cluster5 -0.04478881  \n",
       "2154007 chrX 99777683  99778218  PCDH19     WTC11_cluster5 -0.15480879  \n",
       "2154008 chrX 99777683  99778218  SYTL4      WTC11_cluster5  0.15785553  \n",
       "2154009 chrX 99777683  99778218  TSPAN6     WTC11_cluster5  0.09397730  \n",
       "2154010 chrX  9981419   9981826  CLCN4      WTC11_cluster5 -0.01996782  \n",
       "2154011 chrX  9981419   9981826  GPR143     WTC11_cluster5  0.06241711  \n",
       "2154012 chrX  9981419   9981826  SHROOM2    WTC11_cluster5 -0.11495784  \n",
       "2154013 chrX  9981419   9981826  TBL1X      WTC11_cluster5  0.11202722  \n",
       "2154014 chrX  9981419   9981826  WWC3       WTC11_cluster5  0.11260583  \n",
       "2154015 chrX  9995697   9996195  CLCN4      WTC11_cluster5  0.18022978  \n",
       "2154016 chrX  9995697   9996195  GPR143     WTC11_cluster5  0.03894136  \n",
       "2154017 chrX  9995697   9996195  SHROOM2    WTC11_cluster5  0.09419636  \n",
       "2154018 chrX  9995697   9996195  TBL1X      WTC11_cluster5  0.03522214  \n",
       "2154019 chrX  9995697   9996195  WWC3       WTC11_cluster5 -0.07684341  \n",
       "        FDR          VarQATAC   VarQRNA   \n",
       "1       4.782771e-01 0.04240046 0.8243760 \n",
       "2       1.610183e-02 0.04240046 0.8577874 \n",
       "3       6.586870e-01 0.04240046 0.6375578 \n",
       "4       1.361776e-01 0.04240046 0.7315666 \n",
       "5       3.197162e-02 0.04240046 0.8620709 \n",
       "6       9.602563e-01 0.04240046 0.7301388 \n",
       "7       8.389499e-01 0.04240046 0.2813981 \n",
       "8       1.344091e-03 0.04240046 0.7156891 \n",
       "9       9.996019e-01 0.04240046 0.2836256 \n",
       "10      1.626581e-02 0.04240046 0.8397967 \n",
       "11      1.964259e-02 0.04240046 0.8011308 \n",
       "12      5.914219e-02 0.04240046 0.8179793 \n",
       "13      7.771283e-03 0.04240046 0.7520704 \n",
       "14      1.296994e-05 0.92519225 0.8243760 \n",
       "15      9.540620e-01 0.92519225 0.8577874 \n",
       "16      1.467216e-01 0.92519225 0.6375578 \n",
       "17      2.533576e-02 0.92519225 0.7315666 \n",
       "18      1.322990e-01 0.92519225 0.8620709 \n",
       "19      8.442660e-02 0.92519225 0.7301388 \n",
       "20      4.636104e-01 0.92519225 0.2813981 \n",
       "21      2.284494e-03 0.92519225 0.7156891 \n",
       "22      5.986461e-01 0.92519225 0.2836256 \n",
       "23      9.042741e-01 0.92519225 0.8397967 \n",
       "24      9.363773e-01 0.92519225 0.8011308 \n",
       "25      7.318988e-01 0.92519225 0.8179793 \n",
       "26      8.327468e-01 0.92519225 0.7520704 \n",
       "27      7.390019e-05 0.30306291 0.8243760 \n",
       "28      2.759017e-03 0.30306291 0.8577874 \n",
       "29      4.523386e-01 0.30306291 0.6375578 \n",
       "30      9.520986e-02 0.30306291 0.7315666 \n",
       "⋮       ⋮            ⋮          ⋮         \n",
       "2153990 0.0212719838 0.01940894 0.45182478\n",
       "2153991 0.0015162579 0.01940894 0.54629048\n",
       "2153992 0.4166074721 0.01940894 0.56085442\n",
       "2153993 0.3612279064 0.01940894 0.89816666\n",
       "2153994 0.0001090320 0.01940894 0.70940659\n",
       "2153995 0.1271751248 0.09914438 0.38340282\n",
       "2153996 0.0064621144 0.03349650 0.38340282\n",
       "2153997 0.3725406655 0.03349650 0.73139528\n",
       "2153998 0.2752157467 0.64267342 0.63710092\n",
       "2153999 0.0048393800 0.64267342 0.71660289\n",
       "2154000 0.2640904912 0.64267342 0.79153578\n",
       "2154001 0.7477904600 0.64267342 0.76263636\n",
       "2154002 0.0548485720 0.64267342 0.02787138\n",
       "2154003 0.0228267126 0.64267342 0.81500942\n",
       "2154004 0.0001608724 0.01893651 0.38340282\n",
       "2154005 0.3906047715 0.01893651 0.18727512\n",
       "2154006 0.4655027024 0.01893651 0.73139528\n",
       "2154007 0.0033128878 0.02204667 0.38340282\n",
       "2154008 0.0026918960 0.02204667 0.18727512\n",
       "2154009 0.0919082071 0.02204667 0.73139528\n",
       "2154010 0.7663041389 0.01017690 0.45182478\n",
       "2154011 0.2881017563 0.01017690 0.54629048\n",
       "2154012 0.0347235127 0.01017690 0.56085442\n",
       "2154013 0.0401985549 0.01017690 0.89816666\n",
       "2154014 0.0390688375 0.01017690 0.70940659\n",
       "2154015 0.0005156461 0.71841946 0.45182478\n",
       "2154016 0.5330414571 0.71841946 0.54629048\n",
       "2154017 0.0910701205 0.71841946 0.56085442\n",
       "2154018 0.5775801538 0.71841946 0.89816666\n",
       "2154019 0.1790911951 0.71841946 0.70940659"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(obj.ArchR,\n",
    "        paste(dir.output,\"obj.ArchR.rds\",sep = \"/\"))\n",
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = celltype\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"Correlation\",\n",
    "                         \"FDR\",\n",
    "                         \"VarQATAC\",\n",
    "                         \"VarQRNA\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "pairs.E2G.res\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "4c909ed5-7321-4e69-8208-1544b371da6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] uwot_0.1.16                       nabor_0.5.0                      \n",
       " [3] Rsamtools_2.18.0                  BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] dplyr_1.1.4                       genomation_1.34.0                \n",
       "[11] Signac_1.13.0                     Seurat_5.0.3                     \n",
       "[13] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[15] rhdf5_2.44.0                      SummarizedExperiment_1.30.2      \n",
       "[17] Biobase_2.60.0                    MatrixGenerics_1.12.3            \n",
       "[19] Rcpp_1.0.12                       Matrix_1.6-5                     \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               matrixStats_1.3.0                \n",
       "[27] data.table_1.15.2                 stringr_1.5.1                    \n",
       "[29] plyr_1.8.9                        magrittr_2.0.3                   \n",
       "[31] ggplot2_3.5.1                     gtable_0.3.5                     \n",
       "[33] gtools_3.9.5                      gridExtra_2.3                    \n",
       "[35] ArchR_1.0.2                      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22         splines_4.3.3            later_1.3.2             \n",
       "  [4] pbdZMQ_0.3-11            BiocIO_1.10.0            bitops_1.0-7            \n",
       "  [7] R.oo_1.26.0              tibble_3.2.1             polyclip_1.10-6         \n",
       " [10] XML_3.99-0.16.1          fastDummies_1.7.3        lifecycle_1.0.4         \n",
       " [13] vroom_1.6.5              globals_0.16.3           lattice_0.22-6          \n",
       " [16] MASS_7.3-60              plotly_4.10.4            plotrix_3.8-4           \n",
       " [19] yaml_2.3.8               httpuv_1.6.15            sctransform_0.4.1       \n",
       " [22] spam_2.10-0              spatstat.sparse_3.0-3    reticulate_1.36.1       \n",
       " [25] cowplot_1.1.3            pbapply_1.7-2            RColorBrewer_1.1-3      \n",
       " [28] abind_1.4-5              zlibbioc_1.48.0          Rtsne_0.17              \n",
       " [31] R.utils_2.12.3           purrr_1.0.2              RCurl_1.98-1.14         \n",
       " [34] GenomeInfoDbData_1.2.11  ggrepel_0.9.5            irlba_2.3.5.1           \n",
       " [37] listenv_0.9.1            spatstat.utils_3.0-4     goftest_1.2-3           \n",
       " [40] RSpectra_0.16-1          spatstat.random_3.2-3    fitdistrplus_1.1-11     \n",
       " [43] parallelly_1.37.1        leiden_0.4.3.1           codetools_0.2-20        \n",
       " [46] DelayedArray_0.26.7      RcppRoll_0.3.0           tidyselect_1.2.1        \n",
       " [49] farver_2.1.1             base64enc_0.1-3          spatstat.explore_3.2-6  \n",
       " [52] GenomicAlignments_1.36.0 jsonlite_1.8.8           progressr_0.14.0        \n",
       " [55] ggridges_0.5.6           survival_3.6-4           tools_4.3.3             \n",
       " [58] ica_1.0-3                glue_1.7.0               IRdisplay_1.1           \n",
       " [61] withr_3.0.0              fastmap_1.1.1            rhdf5filters_1.12.1     \n",
       " [64] fansi_1.0.6              digest_0.6.35            R6_2.5.1                \n",
       " [67] mime_0.12                seqPattern_1.34.0        colorspace_2.1-0        \n",
       " [70] Cairo_1.6-2              scattermore_1.2          tensor_1.5              \n",
       " [73] spatstat.data_3.0-4      R.methodsS3_1.8.2        utf8_1.2.4              \n",
       " [76] tidyr_1.3.1              generics_0.1.3           httr_1.4.7              \n",
       " [79] htmlwidgets_1.6.4        S4Arrays_1.2.1           pkgconfig_2.0.3         \n",
       " [82] impute_1.76.0            lmtest_0.9-40            htmltools_0.5.8.1       \n",
       " [85] dotCall64_1.1-1          scales_1.3.0             png_0.1-8               \n",
       " [88] tzdb_0.4.0               reshape2_1.4.4           rjson_0.2.21            \n",
       " [91] uuid_1.2-0               nlme_3.1-164             repr_1.1.7              \n",
       " [94] zoo_1.8-12               KernSmooth_2.23-22       miniUI_0.1.1.1          \n",
       " [97] restfulr_0.0.15          pillar_1.9.0             vctrs_0.6.5             \n",
       "[100] RANN_2.6.1               promises_1.3.0           xtable_1.8-4            \n",
       "[103] cluster_2.1.6            evaluate_0.23            readr_2.1.5             \n",
       "[106] cli_3.6.2                compiler_4.3.3           rlang_1.1.3             \n",
       "[109] crayon_1.5.2             future.apply_1.11.2      labeling_0.4.3          \n",
       "[112] stringi_1.8.3            gridBase_0.4-7           viridisLite_0.4.2       \n",
       "[115] deldir_2.0-4             BiocParallel_1.36.0      munsell_0.5.1           \n",
       "[118] lazyeval_0.2.2           spatstat.geom_3.2-9      IRkernel_1.3.2          \n",
       "[121] RcppHNSW_0.6.0           hms_1.1.3                patchwork_1.2.0         \n",
       "[124] bit64_4.0.5              future_1.33.2            Rhdf5lib_1.22.1         \n",
       "[127] shiny_1.8.1.1            ROCR_1.0-11              igraph_2.0.3            \n",
       "[130] fastmatch_1.1-4          bit_4.0.5               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7690cee-49f1-4886-ae66-0b63a43ce524",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
