{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: rhdf5\n",
      "\n",
      "Warning message:\n",
      "“no DISPLAY variable so Tk is not available”\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SnapATAC)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(parallel)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Xu/K562/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Xu/K562/Predictions/EnhancerPredictionsAllPutative.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Xu/K562/Kendall/atac_matrix.csv.gz\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/1.prepare_data/1.seurat_pipeline.240507/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/3.Genome_wide_prediction/SnapATAC/SnapATAC.250212/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3150487 ranges and 3 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10001196-10001745      * | APITD1-CORT\n",
       "        [2]     chr1 10001196-10001745      * |       CASZ1\n",
       "        [3]     chr1 10001196-10001745      * |       CENPS\n",
       "        [4]     chr1 10001196-10001745      * |      CLSTN1\n",
       "        [5]     chr1 10001196-10001745      * |        CORT\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [3150483]     chrX     989754-991253      * |     SLC25A6\n",
       "  [3150484]     chrX   9995827-9996208      * |       CLCN4\n",
       "  [3150485]     chrX   9995827-9996208      * |       FAM9B\n",
       "  [3150486]     chrX   9995827-9996208      * |       TBL1X\n",
       "  [3150487]     chrX   9995827-9996208      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-10001196-10001745 chr1-10001196-100017..\n",
       "        [2] chr1-10001196-10001745 chr1-10001196-100017..\n",
       "        [3] chr1-10001196-10001745 chr1-10001196-100017..\n",
       "        [4] chr1-10001196-10001745 chr1-10001196-100017..\n",
       "        [5] chr1-10001196-10001745 chr1-10001196-100017..\n",
       "        ...                    ...                    ...\n",
       "  [3150483]     chrX-989754-991253 chrX-989754-991253_S..\n",
       "  [3150484]   chrX-9995827-9996208 chrX-9995827-9996208..\n",
       "  [3150485]   chrX-9995827-9996208 chrX-9995827-9996208..\n",
       "  [3150486]   chrX-9995827-9996208 chrX-9995827-9996208..\n",
       "  [3150487]   chrX-9995827-9996208 chrX-9995827-9996208..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.atac_count = read.csv(path.matrix.atac_count,\n",
    "                             row.names = 1,\n",
    "                             check.names = F)\n",
    "matrix.atac_count = Matrix(as.matrix(matrix.atac_count), sparse = TRUE)\n",
    "matrix.atac = BinarizeCounts(matrix.atac_count)\n",
    "rm(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna_count = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna_count)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4986774c-cbbf-4b72-a299-d681bddca80e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "85b165d0-f1db-4548-beac-235f8950107b",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "655ebba3-731e-49d8-a04e-1ebb5becd567",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Run SnapATAC prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PredictionSnapATAC = function(pairs.E2G,\n",
    "                              data.RNA,\n",
    "                              data.ATAC,\n",
    "                              n.core = 16){\n",
    "  my.cluster <- parallel::makeCluster(\n",
    "    n.core,\n",
    "    type = \"PSOCK\"\n",
    "  )\n",
    "  doParallel::registerDoParallel(cl = my.cluster)\n",
    "  \n",
    "  pairs.E2G =  \n",
    "    foreach (index.tmp = 1:length(pairs.E2G),\n",
    "             .combine = 'c',\n",
    "             .packages = c(\"Matrix\",\n",
    "                           \"GenomicRanges\")) %dopar% {\n",
    "\n",
    "                             pairs.E2G.tmp = pairs.E2G[index.tmp]\n",
    "                             \n",
    "                             model.tmp = summary(stats::glm(data.ATAC[pairs.E2G.tmp$PeakName,] ~ data.RNA[pairs.E2G.tmp$TargetGene,], \n",
    "                                                            family = binomial(link = \"logit\")))[[\"coefficients\"]]\n",
    "                             mcols(pairs.E2G.tmp)[,c(\"beta\",\"sd\",\"z\",\"p\")] = \n",
    "                               t(data.frame(model.tmp[2,]))\n",
    "                             \n",
    "                             pairs.E2G.tmp\n",
    "                           }\n",
    "  \n",
    "  parallel::stopCluster(cl = my.cluster)\n",
    "  \n",
    "  return(pairs.E2G)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "[1] \"chr10\"\n",
      "[1] \"chr11\"\n",
      "[1] \"chr12\"\n",
      "[1] \"chr13\"\n",
      "[1] \"chr14\"\n",
      "[1] \"chr15\"\n",
      "[1] \"chr16\"\n",
      "[1] \"chr17\"\n",
      "[1] \"chr18\"\n",
      "[1] \"chr19\"\n",
      "[1] \"chr2\"\n",
      "[1] \"chr20\"\n",
      "[1] \"chr21\"\n",
      "[1] \"chr22\"\n",
      "[1] \"chr3\"\n",
      "[1] \"chr4\"\n",
      "[1] \"chr5\"\n",
      "[1] \"chr6\"\n",
      "[1] \"chr7\"\n",
      "[1] \"chr8\"\n",
      "[1] \"chr9\"\n",
      "[1] \"chrX\"\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  mcols(pairs.E2G.chr.res)[,c(\"beta\",\"sd\",\"z\",\"p\")] = NA\n",
    "  pairs.E2G.chr.res = PredictionSnapATAC(pairs.E2G.chr.res,\n",
    "                                         matrix.rna.rename[rownames(matrix.rna.rename) %in% pairs.E2G.chr.res$TargetGene,],\n",
    "                                         matrix.atac[rownames(matrix.atac) %in% pairs.E2G.chr.res$PeakName,],\n",
    "                                         n.cores)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3046227 ranges and 7 metadata columns:\n",
       "       seqnames            ranges strand |  TargetGene               PeakName\n",
       "          <Rle>         <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 10001196-10001745      * | APITD1-CORT chr1-10001196-10001745\n",
       "  chr1     chr1 10001196-10001745      * |       CASZ1 chr1-10001196-10001745\n",
       "  chr1     chr1 10001196-10001745      * |       CENPS chr1-10001196-10001745\n",
       "  chr1     chr1 10001196-10001745      * |      CLSTN1 chr1-10001196-10001745\n",
       "  chr1     chr1 10001196-10001745      * |        CORT chr1-10001196-10001745\n",
       "   ...      ...               ...    ... .         ...                    ...\n",
       "  chrX     chrX     989754-991253      * |     SLC25A6     chrX-989754-991253\n",
       "  chrX     chrX   9995827-9996208      * |       CLCN4   chrX-9995827-9996208\n",
       "  chrX     chrX   9995827-9996208      * |       FAM9B   chrX-9995827-9996208\n",
       "  chrX     chrX   9995827-9996208      * |       TBL1X   chrX-9995827-9996208\n",
       "  chrX     chrX   9995827-9996208      * |        WWC3   chrX-9995827-9996208\n",
       "                     PairName       beta          sd          z         p\n",
       "                  <character>  <numeric>   <numeric>  <numeric> <numeric>\n",
       "  chr1 chr1-10001196-100017..   0.963677    1.020103  0.9446857  0.344819\n",
       "  chr1 chr1-10001196-100017..   1.456133    1.165428  1.2494406  0.211504\n",
       "  chr1 chr1-10001196-100017..   0.247096    0.389808  0.6338908  0.526152\n",
       "  chr1 chr1-10001196-100017..   0.196842    0.290920  0.6766184  0.498648\n",
       "  chr1 chr1-10001196-100017.. -92.390128 3762.167362 -0.0245577  0.980408\n",
       "   ...                    ...        ...         ...        ...       ...\n",
       "  chrX chrX-989754-991253_S.. -0.0259896   0.0572838 -0.4536981  0.650046\n",
       "  chrX chrX-9995827-9996208..  0.2209092   0.8080978  0.2733694  0.784569\n",
       "  chrX chrX-9995827-9996208..  0.0163483   1.0088909  0.0162042  0.987071\n",
       "  chrX chrX-9995827-9996208..  0.1679418   0.1438550  1.1674381  0.243033\n",
       "  chrX chrX-9995827-9996208..  0.2991297   0.6739398  0.4438523  0.657149\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3046227 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>beta</th><th scope=col>sd</th><th scope=col>z</th><th scope=col>p</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>APITD1-CORT</td><td>K562</td><td>  0.96367698</td><td>   1.0201033</td><td> 0.94468569</td><td>0.3448193640</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CASZ1      </td><td>K562</td><td>  1.45613277</td><td>   1.1654277</td><td> 1.24944063</td><td>0.2115039554</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CENPS      </td><td>K562</td><td>  0.24709586</td><td>   0.3898082</td><td> 0.63389079</td><td>0.5261520982</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CLSTN1     </td><td>K562</td><td>  0.19684217</td><td>   0.2909205</td><td> 0.67661845</td><td>0.4986480717</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CORT       </td><td>K562</td><td>-92.39012771</td><td>3762.1673621</td><td>-0.02455769</td><td>0.9804077702</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CTNNBIP1   </td><td>K562</td><td>  0.05705119</td><td>   0.2986616</td><td> 0.19102282</td><td>0.8485077169</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>DFFA       </td><td>K562</td><td>  0.16739230</td><td>   0.1958141</td><td> 0.85485302</td><td>0.3926325204</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>GPR157     </td><td>K562</td><td>-41.77387807</td><td>1439.2644515</td><td>-0.02902446</td><td>0.9768450800</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>H6PD       </td><td>K562</td><td>  0.01751790</td><td>   0.6373034</td><td> 0.02748754</td><td>0.9780708780</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>KIF1B      </td><td>K562</td><td>  0.40168515</td><td>   0.1904620</td><td> 2.10900390</td><td>0.0349442450</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>LINC01759  </td><td>K562</td><td>-21.30177128</td><td> 797.1302832</td><td>-0.02672307</td><td>0.9786806096</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>LZIC       </td><td>K562</td><td>  0.03222872</td><td>   0.2075816</td><td> 0.15525809</td><td>0.8766178553</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>MIR34AHG   </td><td>K562</td><td>-27.60609270</td><td> 935.9795156</td><td>-0.02949433</td><td>0.9764703376</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>NMNAT1     </td><td>K562</td><td>  0.16099811</td><td>   0.3037635</td><td> 0.53001134</td><td>0.5961040684</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>PEX14      </td><td>K562</td><td>  0.03211822</td><td>   0.1839919</td><td> 0.17456324</td><td>0.8614228349</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>PGD        </td><td>K562</td><td>  0.31550898</td><td>   0.1850671</td><td> 1.70483602</td><td>0.0882250107</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>PIK3CD     </td><td>K562</td><td>  0.04370438</td><td>   0.3586868</td><td> 0.12184551</td><td>0.9030213733</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>RBP7       </td><td>K562</td><td>  3.14833716</td><td>   2.2397150</td><td> 1.40568648</td><td>0.1598172438</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>SLC25A33   </td><td>K562</td><td>  0.23115551</td><td>   0.1905822</td><td> 1.21289164</td><td>0.2251712511</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>SPSB1      </td><td>K562</td><td>  2.34968569</td><td>   1.1835508</td><td> 1.98528496</td><td>0.0471127766</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>TMEM201    </td><td>K562</td><td>  0.37236389</td><td>   0.2974634</td><td> 1.25179741</td><td>0.2106436927</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>UBE4B      </td><td>K562</td><td>  0.46708599</td><td>   0.1783609</td><td> 2.61876966</td><td>0.0088247512</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>AGL        </td><td>K562</td><td>  0.32728787</td><td>   0.1201924</td><td> 2.72303355</td><td>0.0064685488</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>CDC14A     </td><td>K562</td><td>  0.31953495</td><td>   0.1020647</td><td> 3.13070977</td><td>0.0017438441</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>DBT        </td><td>K562</td><td>  0.41526266</td><td>   0.1306255</td><td> 3.17903164</td><td>0.0014776796</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>DPH5       </td><td>K562</td><td>  0.40476407</td><td>   0.1202017</td><td> 3.36737475</td><td>0.0007588749</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>EXTL2      </td><td>K562</td><td>  0.05575241</td><td>   0.1551514</td><td> 0.35934204</td><td>0.7193392273</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>FRRS1      </td><td>K562</td><td>  0.06186156</td><td>   0.1113732</td><td> 0.55544373</td><td>0.5785911851</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>MFSD14A    </td><td>K562</td><td>  0.04894811</td><td>   0.3028510</td><td> 0.16162438</td><td>0.8716016556</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>PALMD      </td><td>K562</td><td> -0.15325591</td><td>   0.4141157</td><td>-0.37007991</td><td>0.7113229507</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>CLCN4    </td><td>K562</td><td>-0.808520358</td><td>1.16917894</td><td>-0.691528331</td><td>0.4892335833</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>FAM9B    </td><td>K562</td><td> 0.685075219</td><td>0.91505422</td><td> 0.748671721</td><td>0.4540550925</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>TBL1X    </td><td>K562</td><td> 0.000177228</td><td>0.15834401</td><td> 0.001119259</td><td>0.9991069608</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>WWC3     </td><td>K562</td><td> 0.043300055</td><td>0.79571861</td><td> 0.054416291</td><td>0.9566035000</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>CLCN4    </td><td>K562</td><td> 0.784002869</td><td>0.80654518</td><td> 0.972050774</td><td>0.3310252897</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>FAM9B    </td><td>K562</td><td>-0.640227945</td><td>1.41567772</td><td>-0.452241309</td><td>0.6510951522</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>TBL1X    </td><td>K562</td><td> 0.028572555</td><td>0.16560668</td><td> 0.172532627</td><td>0.8630188136</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>WWC3     </td><td>K562</td><td>-1.384149996</td><td>1.19248795</td><td>-1.160724510</td><td>0.2457539512</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>CLCN4    </td><td>K562</td><td>-0.461618428</td><td>1.14003276</td><td>-0.404916809</td><td>0.6855386626</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>FAM9B    </td><td>K562</td><td> 0.792737191</td><td>0.94305132</td><td> 0.840608754</td><td>0.4005671532</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>TBL1X    </td><td>K562</td><td>-0.027164918</td><td>0.16868839</td><td>-0.161036082</td><td>0.8720649822</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>WWC3     </td><td>K562</td><td>-0.730664242</td><td>1.03280114</td><td>-0.707458786</td><td>0.4792814163</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>CLCN4    </td><td>K562</td><td> 0.502392094</td><td>0.81918955</td><td> 0.613279426</td><td>0.5396915950</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>FAM9B    </td><td>K562</td><td> 0.722703518</td><td>0.89187069</td><td> 0.810323206</td><td>0.4177544411</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>TBL1X    </td><td>K562</td><td> 0.278416996</td><td>0.15786675</td><td> 1.763620274</td><td>0.0777959320</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>WWC3     </td><td>K562</td><td> 0.233043009</td><td>0.74737975</td><td> 0.311813383</td><td>0.7551823524</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>AKAP17A  </td><td>K562</td><td>-0.025851055</td><td>0.08857008</td><td>-0.291871197</td><td>0.7703851068</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>ASMT     </td><td>K562</td><td>-0.258132088</td><td>1.88157623</td><td>-0.137189280</td><td>0.8908811843</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>ASMTL    </td><td>K562</td><td> 0.391001170</td><td>0.13388730</td><td> 2.920375354</td><td>0.0034961002</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>GTPBP6   </td><td>K562</td><td> 0.159632401</td><td>0.10977056</td><td> 1.454236894</td><td>0.1458806395</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>IL3RA    </td><td>K562</td><td> 0.225360336</td><td>0.40381568</td><td> 0.558077223</td><td>0.5767916534</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>LINC00106</td><td>K562</td><td> 0.218801129</td><td>0.18423488</td><td> 1.187620550</td><td>0.2349829381</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>P2RY8    </td><td>K562</td><td>-0.851947364</td><td>0.71116296</td><td>-1.197963635</td><td>0.2309311745</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>PLCXD1   </td><td>K562</td><td> 0.107541278</td><td>0.07781592</td><td> 1.381995776</td><td>0.1669729980</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>PPP2R3B  </td><td>K562</td><td> 0.503448363</td><td>0.14960606</td><td> 3.365160173</td><td>0.0007649924</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>SLC25A6  </td><td>K562</td><td>-0.025989563</td><td>0.05728382</td><td>-0.453698128</td><td>0.6500461148</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>CLCN4    </td><td>K562</td><td> 0.220909203</td><td>0.80809776</td><td> 0.273369404</td><td>0.7845692695</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>FAM9B    </td><td>K562</td><td> 0.016348253</td><td>1.00889092</td><td> 0.016204183</td><td>0.9870714984</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>TBL1X    </td><td>K562</td><td> 0.167941773</td><td>0.14385497</td><td> 1.167438096</td><td>0.2430334941</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>WWC3     </td><td>K562</td><td> 0.299129686</td><td>0.67393975</td><td> 0.443852264</td><td>0.6571493990</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3046227 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & beta & sd & z & p\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  10001196 &  10001745 & APITD1-CORT & K562 &   0.96367698 &    1.0201033 &  0.94468569 & 0.3448193640\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CASZ1       & K562 &   1.45613277 &    1.1654277 &  1.24944063 & 0.2115039554\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CENPS       & K562 &   0.24709586 &    0.3898082 &  0.63389079 & 0.5261520982\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CLSTN1      & K562 &   0.19684217 &    0.2909205 &  0.67661845 & 0.4986480717\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CORT        & K562 & -92.39012771 & 3762.1673621 & -0.02455769 & 0.9804077702\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CTNNBIP1    & K562 &   0.05705119 &    0.2986616 &  0.19102282 & 0.8485077169\\\\\n",
       "\t chr1 &  10001196 &  10001745 & DFFA        & K562 &   0.16739230 &    0.1958141 &  0.85485302 & 0.3926325204\\\\\n",
       "\t chr1 &  10001196 &  10001745 & GPR157      & K562 & -41.77387807 & 1439.2644515 & -0.02902446 & 0.9768450800\\\\\n",
       "\t chr1 &  10001196 &  10001745 & H6PD        & K562 &   0.01751790 &    0.6373034 &  0.02748754 & 0.9780708780\\\\\n",
       "\t chr1 &  10001196 &  10001745 & KIF1B       & K562 &   0.40168515 &    0.1904620 &  2.10900390 & 0.0349442450\\\\\n",
       "\t chr1 &  10001196 &  10001745 & LINC01759   & K562 & -21.30177128 &  797.1302832 & -0.02672307 & 0.9786806096\\\\\n",
       "\t chr1 &  10001196 &  10001745 & LZIC        & K562 &   0.03222872 &    0.2075816 &  0.15525809 & 0.8766178553\\\\\n",
       "\t chr1 &  10001196 &  10001745 & MIR34AHG    & K562 & -27.60609270 &  935.9795156 & -0.02949433 & 0.9764703376\\\\\n",
       "\t chr1 &  10001196 &  10001745 & NMNAT1      & K562 &   0.16099811 &    0.3037635 &  0.53001134 & 0.5961040684\\\\\n",
       "\t chr1 &  10001196 &  10001745 & PEX14       & K562 &   0.03211822 &    0.1839919 &  0.17456324 & 0.8614228349\\\\\n",
       "\t chr1 &  10001196 &  10001745 & PGD         & K562 &   0.31550898 &    0.1850671 &  1.70483602 & 0.0882250107\\\\\n",
       "\t chr1 &  10001196 &  10001745 & PIK3CD      & K562 &   0.04370438 &    0.3586868 &  0.12184551 & 0.9030213733\\\\\n",
       "\t chr1 &  10001196 &  10001745 & RBP7        & K562 &   3.14833716 &    2.2397150 &  1.40568648 & 0.1598172438\\\\\n",
       "\t chr1 &  10001196 &  10001745 & SLC25A33    & K562 &   0.23115551 &    0.1905822 &  1.21289164 & 0.2251712511\\\\\n",
       "\t chr1 &  10001196 &  10001745 & SPSB1       & K562 &   2.34968569 &    1.1835508 &  1.98528496 & 0.0471127766\\\\\n",
       "\t chr1 &  10001196 &  10001745 & TMEM201     & K562 &   0.37236389 &    0.2974634 &  1.25179741 & 0.2106436927\\\\\n",
       "\t chr1 &  10001196 &  10001745 & UBE4B       & K562 &   0.46708599 &    0.1783609 &  2.61876966 & 0.0088247512\\\\\n",
       "\t chr1 & 100027917 & 100029695 & AGL         & K562 &   0.32728787 &    0.1201924 &  2.72303355 & 0.0064685488\\\\\n",
       "\t chr1 & 100027917 & 100029695 & CDC14A      & K562 &   0.31953495 &    0.1020647 &  3.13070977 & 0.0017438441\\\\\n",
       "\t chr1 & 100027917 & 100029695 & DBT         & K562 &   0.41526266 &    0.1306255 &  3.17903164 & 0.0014776796\\\\\n",
       "\t chr1 & 100027917 & 100029695 & DPH5        & K562 &   0.40476407 &    0.1202017 &  3.36737475 & 0.0007588749\\\\\n",
       "\t chr1 & 100027917 & 100029695 & EXTL2       & K562 &   0.05575241 &    0.1551514 &  0.35934204 & 0.7193392273\\\\\n",
       "\t chr1 & 100027917 & 100029695 & FRRS1       & K562 &   0.06186156 &    0.1113732 &  0.55544373 & 0.5785911851\\\\\n",
       "\t chr1 & 100027917 & 100029695 & MFSD14A     & K562 &   0.04894811 &    0.3028510 &  0.16162438 & 0.8716016556\\\\\n",
       "\t chr1 & 100027917 & 100029695 & PALMD       & K562 &  -0.15325591 &    0.4141157 & -0.37007991 & 0.7113229507\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9873821 & 9874426 & CLCN4     & K562 & -0.808520358 & 1.16917894 & -0.691528331 & 0.4892335833\\\\\n",
       "\t chrX & 9873821 & 9874426 & FAM9B     & K562 &  0.685075219 & 0.91505422 &  0.748671721 & 0.4540550925\\\\\n",
       "\t chrX & 9873821 & 9874426 & TBL1X     & K562 &  0.000177228 & 0.15834401 &  0.001119259 & 0.9991069608\\\\\n",
       "\t chrX & 9873821 & 9874426 & WWC3      & K562 &  0.043300055 & 0.79571861 &  0.054416291 & 0.9566035000\\\\\n",
       "\t chrX & 9882635 & 9883318 & CLCN4     & K562 &  0.784002869 & 0.80654518 &  0.972050774 & 0.3310252897\\\\\n",
       "\t chrX & 9882635 & 9883318 & FAM9B     & K562 & -0.640227945 & 1.41567772 & -0.452241309 & 0.6510951522\\\\\n",
       "\t chrX & 9882635 & 9883318 & TBL1X     & K562 &  0.028572555 & 0.16560668 &  0.172532627 & 0.8630188136\\\\\n",
       "\t chrX & 9882635 & 9883318 & WWC3      & K562 & -1.384149996 & 1.19248795 & -1.160724510 & 0.2457539512\\\\\n",
       "\t chrX & 9885160 & 9885475 & CLCN4     & K562 & -0.461618428 & 1.14003276 & -0.404916809 & 0.6855386626\\\\\n",
       "\t chrX & 9885160 & 9885475 & FAM9B     & K562 &  0.792737191 & 0.94305132 &  0.840608754 & 0.4005671532\\\\\n",
       "\t chrX & 9885160 & 9885475 & TBL1X     & K562 & -0.027164918 & 0.16868839 & -0.161036082 & 0.8720649822\\\\\n",
       "\t chrX & 9885160 & 9885475 & WWC3      & K562 & -0.730664242 & 1.03280114 & -0.707458786 & 0.4792814163\\\\\n",
       "\t chrX & 9894484 & 9895274 & CLCN4     & K562 &  0.502392094 & 0.81918955 &  0.613279426 & 0.5396915950\\\\\n",
       "\t chrX & 9894484 & 9895274 & FAM9B     & K562 &  0.722703518 & 0.89187069 &  0.810323206 & 0.4177544411\\\\\n",
       "\t chrX & 9894484 & 9895274 & TBL1X     & K562 &  0.278416996 & 0.15786675 &  1.763620274 & 0.0777959320\\\\\n",
       "\t chrX & 9894484 & 9895274 & WWC3      & K562 &  0.233043009 & 0.74737975 &  0.311813383 & 0.7551823524\\\\\n",
       "\t chrX &  989754 &  991253 & AKAP17A   & K562 & -0.025851055 & 0.08857008 & -0.291871197 & 0.7703851068\\\\\n",
       "\t chrX &  989754 &  991253 & ASMT      & K562 & -0.258132088 & 1.88157623 & -0.137189280 & 0.8908811843\\\\\n",
       "\t chrX &  989754 &  991253 & ASMTL     & K562 &  0.391001170 & 0.13388730 &  2.920375354 & 0.0034961002\\\\\n",
       "\t chrX &  989754 &  991253 & GTPBP6    & K562 &  0.159632401 & 0.10977056 &  1.454236894 & 0.1458806395\\\\\n",
       "\t chrX &  989754 &  991253 & IL3RA     & K562 &  0.225360336 & 0.40381568 &  0.558077223 & 0.5767916534\\\\\n",
       "\t chrX &  989754 &  991253 & LINC00106 & K562 &  0.218801129 & 0.18423488 &  1.187620550 & 0.2349829381\\\\\n",
       "\t chrX &  989754 &  991253 & P2RY8     & K562 & -0.851947364 & 0.71116296 & -1.197963635 & 0.2309311745\\\\\n",
       "\t chrX &  989754 &  991253 & PLCXD1    & K562 &  0.107541278 & 0.07781592 &  1.381995776 & 0.1669729980\\\\\n",
       "\t chrX &  989754 &  991253 & PPP2R3B   & K562 &  0.503448363 & 0.14960606 &  3.365160173 & 0.0007649924\\\\\n",
       "\t chrX &  989754 &  991253 & SLC25A6   & K562 & -0.025989563 & 0.05728382 & -0.453698128 & 0.6500461148\\\\\n",
       "\t chrX & 9995827 & 9996208 & CLCN4     & K562 &  0.220909203 & 0.80809776 &  0.273369404 & 0.7845692695\\\\\n",
       "\t chrX & 9995827 & 9996208 & FAM9B     & K562 &  0.016348253 & 1.00889092 &  0.016204183 & 0.9870714984\\\\\n",
       "\t chrX & 9995827 & 9996208 & TBL1X     & K562 &  0.167941773 & 0.14385497 &  1.167438096 & 0.2430334941\\\\\n",
       "\t chrX & 9995827 & 9996208 & WWC3      & K562 &  0.299129686 & 0.67393975 &  0.443852264 & 0.6571493990\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3046227 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | beta &lt;dbl&gt; | sd &lt;dbl&gt; | z &lt;dbl&gt; | p &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  10001196 |  10001745 | APITD1-CORT | K562 |   0.96367698 |    1.0201033 |  0.94468569 | 0.3448193640 |\n",
       "| chr1 |  10001196 |  10001745 | CASZ1       | K562 |   1.45613277 |    1.1654277 |  1.24944063 | 0.2115039554 |\n",
       "| chr1 |  10001196 |  10001745 | CENPS       | K562 |   0.24709586 |    0.3898082 |  0.63389079 | 0.5261520982 |\n",
       "| chr1 |  10001196 |  10001745 | CLSTN1      | K562 |   0.19684217 |    0.2909205 |  0.67661845 | 0.4986480717 |\n",
       "| chr1 |  10001196 |  10001745 | CORT        | K562 | -92.39012771 | 3762.1673621 | -0.02455769 | 0.9804077702 |\n",
       "| chr1 |  10001196 |  10001745 | CTNNBIP1    | K562 |   0.05705119 |    0.2986616 |  0.19102282 | 0.8485077169 |\n",
       "| chr1 |  10001196 |  10001745 | DFFA        | K562 |   0.16739230 |    0.1958141 |  0.85485302 | 0.3926325204 |\n",
       "| chr1 |  10001196 |  10001745 | GPR157      | K562 | -41.77387807 | 1439.2644515 | -0.02902446 | 0.9768450800 |\n",
       "| chr1 |  10001196 |  10001745 | H6PD        | K562 |   0.01751790 |    0.6373034 |  0.02748754 | 0.9780708780 |\n",
       "| chr1 |  10001196 |  10001745 | KIF1B       | K562 |   0.40168515 |    0.1904620 |  2.10900390 | 0.0349442450 |\n",
       "| chr1 |  10001196 |  10001745 | LINC01759   | K562 | -21.30177128 |  797.1302832 | -0.02672307 | 0.9786806096 |\n",
       "| chr1 |  10001196 |  10001745 | LZIC        | K562 |   0.03222872 |    0.2075816 |  0.15525809 | 0.8766178553 |\n",
       "| chr1 |  10001196 |  10001745 | MIR34AHG    | K562 | -27.60609270 |  935.9795156 | -0.02949433 | 0.9764703376 |\n",
       "| chr1 |  10001196 |  10001745 | NMNAT1      | K562 |   0.16099811 |    0.3037635 |  0.53001134 | 0.5961040684 |\n",
       "| chr1 |  10001196 |  10001745 | PEX14       | K562 |   0.03211822 |    0.1839919 |  0.17456324 | 0.8614228349 |\n",
       "| chr1 |  10001196 |  10001745 | PGD         | K562 |   0.31550898 |    0.1850671 |  1.70483602 | 0.0882250107 |\n",
       "| chr1 |  10001196 |  10001745 | PIK3CD      | K562 |   0.04370438 |    0.3586868 |  0.12184551 | 0.9030213733 |\n",
       "| chr1 |  10001196 |  10001745 | RBP7        | K562 |   3.14833716 |    2.2397150 |  1.40568648 | 0.1598172438 |\n",
       "| chr1 |  10001196 |  10001745 | SLC25A33    | K562 |   0.23115551 |    0.1905822 |  1.21289164 | 0.2251712511 |\n",
       "| chr1 |  10001196 |  10001745 | SPSB1       | K562 |   2.34968569 |    1.1835508 |  1.98528496 | 0.0471127766 |\n",
       "| chr1 |  10001196 |  10001745 | TMEM201     | K562 |   0.37236389 |    0.2974634 |  1.25179741 | 0.2106436927 |\n",
       "| chr1 |  10001196 |  10001745 | UBE4B       | K562 |   0.46708599 |    0.1783609 |  2.61876966 | 0.0088247512 |\n",
       "| chr1 | 100027917 | 100029695 | AGL         | K562 |   0.32728787 |    0.1201924 |  2.72303355 | 0.0064685488 |\n",
       "| chr1 | 100027917 | 100029695 | CDC14A      | K562 |   0.31953495 |    0.1020647 |  3.13070977 | 0.0017438441 |\n",
       "| chr1 | 100027917 | 100029695 | DBT         | K562 |   0.41526266 |    0.1306255 |  3.17903164 | 0.0014776796 |\n",
       "| chr1 | 100027917 | 100029695 | DPH5        | K562 |   0.40476407 |    0.1202017 |  3.36737475 | 0.0007588749 |\n",
       "| chr1 | 100027917 | 100029695 | EXTL2       | K562 |   0.05575241 |    0.1551514 |  0.35934204 | 0.7193392273 |\n",
       "| chr1 | 100027917 | 100029695 | FRRS1       | K562 |   0.06186156 |    0.1113732 |  0.55544373 | 0.5785911851 |\n",
       "| chr1 | 100027917 | 100029695 | MFSD14A     | K562 |   0.04894811 |    0.3028510 |  0.16162438 | 0.8716016556 |\n",
       "| chr1 | 100027917 | 100029695 | PALMD       | K562 |  -0.15325591 |    0.4141157 | -0.37007991 | 0.7113229507 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9873821 | 9874426 | CLCN4     | K562 | -0.808520358 | 1.16917894 | -0.691528331 | 0.4892335833 |\n",
       "| chrX | 9873821 | 9874426 | FAM9B     | K562 |  0.685075219 | 0.91505422 |  0.748671721 | 0.4540550925 |\n",
       "| chrX | 9873821 | 9874426 | TBL1X     | K562 |  0.000177228 | 0.15834401 |  0.001119259 | 0.9991069608 |\n",
       "| chrX | 9873821 | 9874426 | WWC3      | K562 |  0.043300055 | 0.79571861 |  0.054416291 | 0.9566035000 |\n",
       "| chrX | 9882635 | 9883318 | CLCN4     | K562 |  0.784002869 | 0.80654518 |  0.972050774 | 0.3310252897 |\n",
       "| chrX | 9882635 | 9883318 | FAM9B     | K562 | -0.640227945 | 1.41567772 | -0.452241309 | 0.6510951522 |\n",
       "| chrX | 9882635 | 9883318 | TBL1X     | K562 |  0.028572555 | 0.16560668 |  0.172532627 | 0.8630188136 |\n",
       "| chrX | 9882635 | 9883318 | WWC3      | K562 | -1.384149996 | 1.19248795 | -1.160724510 | 0.2457539512 |\n",
       "| chrX | 9885160 | 9885475 | CLCN4     | K562 | -0.461618428 | 1.14003276 | -0.404916809 | 0.6855386626 |\n",
       "| chrX | 9885160 | 9885475 | FAM9B     | K562 |  0.792737191 | 0.94305132 |  0.840608754 | 0.4005671532 |\n",
       "| chrX | 9885160 | 9885475 | TBL1X     | K562 | -0.027164918 | 0.16868839 | -0.161036082 | 0.8720649822 |\n",
       "| chrX | 9885160 | 9885475 | WWC3      | K562 | -0.730664242 | 1.03280114 | -0.707458786 | 0.4792814163 |\n",
       "| chrX | 9894484 | 9895274 | CLCN4     | K562 |  0.502392094 | 0.81918955 |  0.613279426 | 0.5396915950 |\n",
       "| chrX | 9894484 | 9895274 | FAM9B     | K562 |  0.722703518 | 0.89187069 |  0.810323206 | 0.4177544411 |\n",
       "| chrX | 9894484 | 9895274 | TBL1X     | K562 |  0.278416996 | 0.15786675 |  1.763620274 | 0.0777959320 |\n",
       "| chrX | 9894484 | 9895274 | WWC3      | K562 |  0.233043009 | 0.74737975 |  0.311813383 | 0.7551823524 |\n",
       "| chrX |  989754 |  991253 | AKAP17A   | K562 | -0.025851055 | 0.08857008 | -0.291871197 | 0.7703851068 |\n",
       "| chrX |  989754 |  991253 | ASMT      | K562 | -0.258132088 | 1.88157623 | -0.137189280 | 0.8908811843 |\n",
       "| chrX |  989754 |  991253 | ASMTL     | K562 |  0.391001170 | 0.13388730 |  2.920375354 | 0.0034961002 |\n",
       "| chrX |  989754 |  991253 | GTPBP6    | K562 |  0.159632401 | 0.10977056 |  1.454236894 | 0.1458806395 |\n",
       "| chrX |  989754 |  991253 | IL3RA     | K562 |  0.225360336 | 0.40381568 |  0.558077223 | 0.5767916534 |\n",
       "| chrX |  989754 |  991253 | LINC00106 | K562 |  0.218801129 | 0.18423488 |  1.187620550 | 0.2349829381 |\n",
       "| chrX |  989754 |  991253 | P2RY8     | K562 | -0.851947364 | 0.71116296 | -1.197963635 | 0.2309311745 |\n",
       "| chrX |  989754 |  991253 | PLCXD1    | K562 |  0.107541278 | 0.07781592 |  1.381995776 | 0.1669729980 |\n",
       "| chrX |  989754 |  991253 | PPP2R3B   | K562 |  0.503448363 | 0.14960606 |  3.365160173 | 0.0007649924 |\n",
       "| chrX |  989754 |  991253 | SLC25A6   | K562 | -0.025989563 | 0.05728382 | -0.453698128 | 0.6500461148 |\n",
       "| chrX | 9995827 | 9996208 | CLCN4     | K562 |  0.220909203 | 0.80809776 |  0.273369404 | 0.7845692695 |\n",
       "| chrX | 9995827 | 9996208 | FAM9B     | K562 |  0.016348253 | 1.00889092 |  0.016204183 | 0.9870714984 |\n",
       "| chrX | 9995827 | 9996208 | TBL1X     | K562 |  0.167941773 | 0.14385497 |  1.167438096 | 0.2430334941 |\n",
       "| chrX | 9995827 | 9996208 | WWC3      | K562 |  0.299129686 | 0.67393975 |  0.443852264 | 0.6571493990 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene  CellType beta         sd          \n",
       "1       chr1  10001196  10001745 APITD1-CORT K562       0.96367698    1.0201033\n",
       "2       chr1  10001196  10001745 CASZ1       K562       1.45613277    1.1654277\n",
       "3       chr1  10001196  10001745 CENPS       K562       0.24709586    0.3898082\n",
       "4       chr1  10001196  10001745 CLSTN1      K562       0.19684217    0.2909205\n",
       "5       chr1  10001196  10001745 CORT        K562     -92.39012771 3762.1673621\n",
       "6       chr1  10001196  10001745 CTNNBIP1    K562       0.05705119    0.2986616\n",
       "7       chr1  10001196  10001745 DFFA        K562       0.16739230    0.1958141\n",
       "8       chr1  10001196  10001745 GPR157      K562     -41.77387807 1439.2644515\n",
       "9       chr1  10001196  10001745 H6PD        K562       0.01751790    0.6373034\n",
       "10      chr1  10001196  10001745 KIF1B       K562       0.40168515    0.1904620\n",
       "11      chr1  10001196  10001745 LINC01759   K562     -21.30177128  797.1302832\n",
       "12      chr1  10001196  10001745 LZIC        K562       0.03222872    0.2075816\n",
       "13      chr1  10001196  10001745 MIR34AHG    K562     -27.60609270  935.9795156\n",
       "14      chr1  10001196  10001745 NMNAT1      K562       0.16099811    0.3037635\n",
       "15      chr1  10001196  10001745 PEX14       K562       0.03211822    0.1839919\n",
       "16      chr1  10001196  10001745 PGD         K562       0.31550898    0.1850671\n",
       "17      chr1  10001196  10001745 PIK3CD      K562       0.04370438    0.3586868\n",
       "18      chr1  10001196  10001745 RBP7        K562       3.14833716    2.2397150\n",
       "19      chr1  10001196  10001745 SLC25A33    K562       0.23115551    0.1905822\n",
       "20      chr1  10001196  10001745 SPSB1       K562       2.34968569    1.1835508\n",
       "21      chr1  10001196  10001745 TMEM201     K562       0.37236389    0.2974634\n",
       "22      chr1  10001196  10001745 UBE4B       K562       0.46708599    0.1783609\n",
       "23      chr1 100027917 100029695 AGL         K562       0.32728787    0.1201924\n",
       "24      chr1 100027917 100029695 CDC14A      K562       0.31953495    0.1020647\n",
       "25      chr1 100027917 100029695 DBT         K562       0.41526266    0.1306255\n",
       "26      chr1 100027917 100029695 DPH5        K562       0.40476407    0.1202017\n",
       "27      chr1 100027917 100029695 EXTL2       K562       0.05575241    0.1551514\n",
       "28      chr1 100027917 100029695 FRRS1       K562       0.06186156    0.1113732\n",
       "29      chr1 100027917 100029695 MFSD14A     K562       0.04894811    0.3028510\n",
       "30      chr1 100027917 100029695 PALMD       K562      -0.15325591    0.4141157\n",
       "⋮       ⋮    ⋮         ⋮         ⋮           ⋮        ⋮            ⋮           \n",
       "3046198 chrX 9873821   9874426   CLCN4       K562     -0.808520358 1.16917894  \n",
       "3046199 chrX 9873821   9874426   FAM9B       K562      0.685075219 0.91505422  \n",
       "3046200 chrX 9873821   9874426   TBL1X       K562      0.000177228 0.15834401  \n",
       "3046201 chrX 9873821   9874426   WWC3        K562      0.043300055 0.79571861  \n",
       "3046202 chrX 9882635   9883318   CLCN4       K562      0.784002869 0.80654518  \n",
       "3046203 chrX 9882635   9883318   FAM9B       K562     -0.640227945 1.41567772  \n",
       "3046204 chrX 9882635   9883318   TBL1X       K562      0.028572555 0.16560668  \n",
       "3046205 chrX 9882635   9883318   WWC3        K562     -1.384149996 1.19248795  \n",
       "3046206 chrX 9885160   9885475   CLCN4       K562     -0.461618428 1.14003276  \n",
       "3046207 chrX 9885160   9885475   FAM9B       K562      0.792737191 0.94305132  \n",
       "3046208 chrX 9885160   9885475   TBL1X       K562     -0.027164918 0.16868839  \n",
       "3046209 chrX 9885160   9885475   WWC3        K562     -0.730664242 1.03280114  \n",
       "3046210 chrX 9894484   9895274   CLCN4       K562      0.502392094 0.81918955  \n",
       "3046211 chrX 9894484   9895274   FAM9B       K562      0.722703518 0.89187069  \n",
       "3046212 chrX 9894484   9895274   TBL1X       K562      0.278416996 0.15786675  \n",
       "3046213 chrX 9894484   9895274   WWC3        K562      0.233043009 0.74737975  \n",
       "3046214 chrX  989754    991253   AKAP17A     K562     -0.025851055 0.08857008  \n",
       "3046215 chrX  989754    991253   ASMT        K562     -0.258132088 1.88157623  \n",
       "3046216 chrX  989754    991253   ASMTL       K562      0.391001170 0.13388730  \n",
       "3046217 chrX  989754    991253   GTPBP6      K562      0.159632401 0.10977056  \n",
       "3046218 chrX  989754    991253   IL3RA       K562      0.225360336 0.40381568  \n",
       "3046219 chrX  989754    991253   LINC00106   K562      0.218801129 0.18423488  \n",
       "3046220 chrX  989754    991253   P2RY8       K562     -0.851947364 0.71116296  \n",
       "3046221 chrX  989754    991253   PLCXD1      K562      0.107541278 0.07781592  \n",
       "3046222 chrX  989754    991253   PPP2R3B     K562      0.503448363 0.14960606  \n",
       "3046223 chrX  989754    991253   SLC25A6     K562     -0.025989563 0.05728382  \n",
       "3046224 chrX 9995827   9996208   CLCN4       K562      0.220909203 0.80809776  \n",
       "3046225 chrX 9995827   9996208   FAM9B       K562      0.016348253 1.00889092  \n",
       "3046226 chrX 9995827   9996208   TBL1X       K562      0.167941773 0.14385497  \n",
       "3046227 chrX 9995827   9996208   WWC3        K562      0.299129686 0.67393975  \n",
       "        z            p           \n",
       "1        0.94468569  0.3448193640\n",
       "2        1.24944063  0.2115039554\n",
       "3        0.63389079  0.5261520982\n",
       "4        0.67661845  0.4986480717\n",
       "5       -0.02455769  0.9804077702\n",
       "6        0.19102282  0.8485077169\n",
       "7        0.85485302  0.3926325204\n",
       "8       -0.02902446  0.9768450800\n",
       "9        0.02748754  0.9780708780\n",
       "10       2.10900390  0.0349442450\n",
       "11      -0.02672307  0.9786806096\n",
       "12       0.15525809  0.8766178553\n",
       "13      -0.02949433  0.9764703376\n",
       "14       0.53001134  0.5961040684\n",
       "15       0.17456324  0.8614228349\n",
       "16       1.70483602  0.0882250107\n",
       "17       0.12184551  0.9030213733\n",
       "18       1.40568648  0.1598172438\n",
       "19       1.21289164  0.2251712511\n",
       "20       1.98528496  0.0471127766\n",
       "21       1.25179741  0.2106436927\n",
       "22       2.61876966  0.0088247512\n",
       "23       2.72303355  0.0064685488\n",
       "24       3.13070977  0.0017438441\n",
       "25       3.17903164  0.0014776796\n",
       "26       3.36737475  0.0007588749\n",
       "27       0.35934204  0.7193392273\n",
       "28       0.55544373  0.5785911851\n",
       "29       0.16162438  0.8716016556\n",
       "30      -0.37007991  0.7113229507\n",
       "⋮       ⋮            ⋮           \n",
       "3046198 -0.691528331 0.4892335833\n",
       "3046199  0.748671721 0.4540550925\n",
       "3046200  0.001119259 0.9991069608\n",
       "3046201  0.054416291 0.9566035000\n",
       "3046202  0.972050774 0.3310252897\n",
       "3046203 -0.452241309 0.6510951522\n",
       "3046204  0.172532627 0.8630188136\n",
       "3046205 -1.160724510 0.2457539512\n",
       "3046206 -0.404916809 0.6855386626\n",
       "3046207  0.840608754 0.4005671532\n",
       "3046208 -0.161036082 0.8720649822\n",
       "3046209 -0.707458786 0.4792814163\n",
       "3046210  0.613279426 0.5396915950\n",
       "3046211  0.810323206 0.4177544411\n",
       "3046212  1.763620274 0.0777959320\n",
       "3046213  0.311813383 0.7551823524\n",
       "3046214 -0.291871197 0.7703851068\n",
       "3046215 -0.137189280 0.8908811843\n",
       "3046216  2.920375354 0.0034961002\n",
       "3046217  1.454236894 0.1458806395\n",
       "3046218  0.558077223 0.5767916534\n",
       "3046219  1.187620550 0.2349829381\n",
       "3046220 -1.197963635 0.2309311745\n",
       "3046221  1.381995776 0.1669729980\n",
       "3046222  3.365160173 0.0007649924\n",
       "3046223 -0.453698128 0.6500461148\n",
       "3046224  0.273369404 0.7845692695\n",
       "3046225  0.016204183 0.9870714984\n",
       "3046226  1.167438096 0.2430334941\n",
       "3046227  0.443852264 0.6571493990"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"beta\",\n",
    "                         \"sd\",\n",
    "                         \"z\",\n",
    "                         \"p\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] data.table_1.15.2    dplyr_1.1.4          foreach_1.5.2       \n",
       " [4] GenomicRanges_1.54.1 GenomeInfoDb_1.38.1  IRanges_2.36.0      \n",
       " [7] S4Vectors_0.40.2     BiocGenerics_0.48.1  genomation_1.34.0   \n",
       "[10] Signac_1.13.0        Seurat_5.0.3         SeuratObject_5.0.1  \n",
       "[13] sp_2.1-4             SnapATAC_1.0.0       rhdf5_2.44.0        \n",
       "[16] Matrix_1.6-5        \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 BiocIO_1.10.0              \n",
       "  [5] pbdZMQ_0.3-11               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                polyclip_1.10-6            \n",
       "  [9] XML_3.99-0.16.1             fastDummies_1.7.3          \n",
       " [11] lifecycle_1.0.4             tcltk_4.3.3                \n",
       " [13] edgeR_3.42.4                doParallel_1.0.17          \n",
       " [15] vroom_1.6.5                 globals_0.16.3             \n",
       " [17] lattice_0.22-6              MASS_7.3-60                \n",
       " [19] magrittr_2.0.3              limma_3.56.2               \n",
       " [21] plotly_4.10.4               plotrix_3.8-4              \n",
       " [23] yaml_2.3.8                  plot3D_1.4.1               \n",
       " [25] httpuv_1.6.15               sctransform_0.4.1          \n",
       " [27] spam_2.10-0                 spatstat.sparse_3.0-3      \n",
       " [29] reticulate_1.36.1           cowplot_1.1.3              \n",
       " [31] pbapply_1.7-2               RColorBrewer_1.1-3         \n",
       " [33] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [35] Rtsne_0.17                  purrr_1.0.2                \n",
       " [37] RCurl_1.98-1.14             misc3d_0.9-1               \n",
       " [39] GenomeInfoDbData_1.2.11     ggrepel_0.9.5              \n",
       " [41] irlba_2.3.5.1               listenv_0.9.1              \n",
       " [43] spatstat.utils_3.0-4        goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             bigmemory_4.6.4            \n",
       " [47] spatstat.random_3.2-3       fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           DelayedArray_0.26.7        \n",
       " [51] leiden_0.4.3.1              codetools_0.2-20           \n",
       " [53] RcppRoll_0.3.0              tidyselect_1.2.1           \n",
       " [55] viridis_0.6.5               matrixStats_1.3.0          \n",
       " [57] base64enc_0.1-3             spatstat.explore_3.2-6     \n",
       " [59] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [61] progressr_0.14.0            ggridges_0.5.6             \n",
       " [63] survival_3.6-4              iterators_1.0.14           \n",
       " [65] tools_4.3.3                 snow_0.4-4                 \n",
       " [67] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [69] glue_1.7.0                  gridExtra_2.3              \n",
       " [71] MatrixGenerics_1.12.3       IRdisplay_1.1              \n",
       " [73] withr_3.0.0                 fastmap_1.1.1              \n",
       " [75] rhdf5filters_1.12.1         fansi_1.0.6                \n",
       " [77] digest_0.6.35               R6_2.5.1                   \n",
       " [79] mime_0.12                   seqPattern_1.34.0          \n",
       " [81] colorspace_2.1-0            scattermore_1.2            \n",
       " [83] tensor_1.5                  spatstat.data_3.0-4        \n",
       " [85] utf8_1.2.4                  tidyr_1.3.1                \n",
       " [87] generics_0.1.3              rtracklayer_1.60.1         \n",
       " [89] S4Arrays_1.2.1              httr_1.4.7                 \n",
       " [91] htmlwidgets_1.6.4           uwot_0.1.16                \n",
       " [93] pkgconfig_2.0.3             gtable_0.3.5               \n",
       " [95] impute_1.76.0               lmtest_0.9-40              \n",
       " [97] XVector_0.42.0              htmltools_0.5.8.1          \n",
       " [99] dotCall64_1.1-1             Biobase_2.60.0             \n",
       "[101] scales_1.3.0                png_0.1-8                  \n",
       "[103] doSNOW_1.0.20               bigmemory.sri_0.1.8        \n",
       "[105] tzdb_0.4.0                  rjson_0.2.21               \n",
       "[107] reshape2_1.4.4              uuid_1.2-0                 \n",
       "[109] nlme_3.1-164                repr_1.1.7                 \n",
       "[111] zoo_1.8-12                  stringr_1.5.1              \n",
       "[113] KernSmooth_2.23-22          miniUI_0.1.1.1             \n",
       "[115] restfulr_0.0.15             pillar_1.9.0               \n",
       "[117] vctrs_0.6.5                 RANN_2.6.1                 \n",
       "[119] promises_1.3.0              xtable_1.8-4               \n",
       "[121] cluster_2.1.6               evaluate_0.23              \n",
       "[123] readr_2.1.5                 cli_3.6.2                  \n",
       "[125] locfit_1.5-9.9              compiler_4.3.3             \n",
       "[127] Rsamtools_2.18.0            rlang_1.1.3                \n",
       "[129] crayon_1.5.2                future.apply_1.11.2        \n",
       "[131] plyr_1.8.9                  stringi_1.8.3              \n",
       "[133] gridBase_0.4-7              viridisLite_0.4.2          \n",
       "[135] deldir_2.0-4                BiocParallel_1.36.0        \n",
       "[137] munsell_0.5.1               Biostrings_2.70.1          \n",
       "[139] lazyeval_0.2.2              spatstat.geom_3.2-9        \n",
       "[141] IRkernel_1.3.2              BSgenome_1.68.0            \n",
       "[143] RcppHNSW_0.6.0              hms_1.1.3                  \n",
       "[145] patchwork_1.2.0             bit64_4.0.5                \n",
       "[147] future_1.33.2               ggplot2_3.5.1              \n",
       "[149] Rhdf5lib_1.22.1             shiny_1.8.1.1              \n",
       "[151] SummarizedExperiment_1.30.2 ROCR_1.0-11                \n",
       "[153] igraph_2.0.3                fastmatch_1.1-4            \n",
       "[155] bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
