{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8466b35-cae5-4f62-8138-37b2de484a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.2\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : grid v4.3.3\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "\tLoading Package : gtools v3.9.5\n",
      "\n",
      "\tLoading Package : gtable v0.3.5\n",
      "\n",
      "\tLoading Package : ggplot2 v3.5.1\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "\tLoading Package : plyr v1.8.9\n",
      "\n",
      "\tLoading Package : stringr v1.5.1\n",
      "\n",
      "\tLoading Package : data.table v1.15.2\n",
      "\n",
      "\tLoading Package : matrixStats v1.3.0\n",
      "\n",
      "\tLoading Package : S4Vectors v0.40.2\n",
      "\n",
      "\tLoading Package : GenomicRanges v1.54.1\n",
      "\n",
      "\tLoading Package : BiocGenerics v0.48.1\n",
      "\n",
      "\tLoading Package : Matrix v1.6.5\n",
      "\n",
      "\tLoading Package : Rcpp v1.0.12\n",
      "\n",
      "\tLoading Package : SummarizedExperiment v1.30.2\n",
      "\n",
      "\tLoading Package : rhdf5 v2.44.0\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(magrittr)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e85020d7-d358-492d-9358-c81270f4b624",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1bec6690-52ec-450d-bba7-7d4528a185cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "path.seurat = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/GM12878/1.prepare_data/1.seurat_pipeline.240615/obj.seurat.qc.rds\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240617/sc-E2G/test/results/GM12878/GM12878.multiome_7features/Kendall/Pairs.tsv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/GM12878/2.Genome_wide_prediction/ArchR/ArchR.240625/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "340dac32-44ac-4231-b964-b5f3b9e2dbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16\n",
    "maxDist = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da973385-4c81-4b1c-9669-cf5933b23a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 16.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRThreads(threads = n.cores) \n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeb26496-d3b9-4778-92ed-9c63a5018aef",
   "metadata": {},
   "source": [
    "Import seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c9b19a09-be95-4d9a-bdc3-0530980d4757",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.seurat = readRDS(path.seurat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86be3e66-c954-4620-9e64-c6a1c0fcb84c",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6f363969-bf0b-45ce-858c-a8cb64cc210a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3538ffdc-4fed-48eb-9c9d-86805a91cf06",
   "metadata": {},
   "source": [
    "Create an ArchRProject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0300f485-fafa-4cef-b9d0-534108ec6f95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 1\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240614/GM12878/cellranger_res/\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"a5984ec1fe8d8e28c00a2146ef41d26d\" \"6e3871eeb1cd823c9aaa4c7bb93361fd\"\n",
      "  .. ..@ cells: Named chr [1:3050] \"AAACAGCCACTAAGAA-1\" \"AAACAGCCATAATCAC-1\" \"AAACATGCACTTGTTC-1\" \"AAACATGCATAATCAC-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:3050] \"GM12878_AAACAGCCACTAAGAA-1\" \"GM12878_AAACAGCCATAATCAC-1\" \"GM12878_AAACATGCACTTGTTC-1\" \"GM12878_AAACATGCATAATCAC-1\" ...\n"
     ]
    }
   ],
   "source": [
    "str(Fragments(obj.seurat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1753ffaa-89e2-43bf-bcf6-19937782d63a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-createArrows-35cc641263fc3-Date-2024-06-25_Time-13-27-08.916718.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Cleaning Temporary Files\n",
      "\n",
      "2024-06-25 13:27:10.104111 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "(GM12878 : 1 of 1) Checking if completed file exists!\n",
      "\n",
      "2024-06-25 13:27:10.183488 : (GM12878 : 1 of 1) Arrow Exists! Overriding since force = TRUE!, 0.001 mins elapsed.\n",
      "\n",
      "(GM12878 : 1 of 1) Determining Arrow Method to use!\n",
      "\n",
      "2024-06-25 13:27:10.34584 : (GM12878 : 1 of 1) Reading In Fragments from inputFiles (readMethod = tabix), 0.004 mins elapsed.\n",
      "\n",
      "2024-06-25 13:27:10.355684 : (GM12878 : 1 of 1) Tabix Bed To Temporary File, 0.004 mins elapsed.\n",
      "\n",
      "2024-06-25 13:28:21.715038 : (GM12878 : 1 of 1) Successful creation of Temporary File, 1.194 mins elapsed.\n",
      "\n",
      "2024-06-25 13:28:21.862386 : (GM12878 : 1 of 1) Creating ArrowFile From Temporary File, 1.196 mins elapsed.\n",
      "\n",
      "2024-06-25 13:28:56.641499 : (GM12878 : 1 of 1) Successful creation of Arrow File, 1.776 mins elapsed.\n",
      "\n",
      "2024-06-25 13:29:21.732003 : Continuing through after error ggplot for Fragment Size Distribution, 2.194 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2024-06-25 13:29:58.861759 : (GM12878 : 1 of 1) CellStats : Number of Cells Pass Filter = 3032 , 2.813 mins elapsed.\n",
      "\n",
      "2024-06-25 13:29:58.914675 : (GM12878 : 1 of 1) CellStats : Median Frags = 14044.5 , 2.814 mins elapsed.\n",
      "\n",
      "2024-06-25 13:29:59.136229 : (GM12878 : 1 of 1) CellStats : Median TSS Enrichment = 15.2195 , 2.817 mins elapsed.\n",
      "\n",
      "2024-06-25 13:30:00.19685 : Continuing through after error ggplot for TSS by Frags, 2.835 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2024-06-25 13:30:04.192076 : (GM12878 : 1 of 1) Adding Additional Feature Counts!, 2.901 mins elapsed.\n",
      "\n",
      "2024-06-25 13:30:31.177917 : (GM12878 : 1 of 1) Removing Fragments from Filtered Cells, 3.351 mins elapsed.\n",
      "\n",
      "2024-06-25 13:30:31.187353 : (GM12878 : 1 of 1) Creating Filtered Arrow File, 3.351 mins elapsed.\n",
      "\n",
      "2024-06-25 13:31:58.723563 : (GM12878 : 1 of 1) Finished Constructing Filtered Arrow File!, 4.81 mins elapsed.\n",
      "\n",
      "2024-06-25 13:31:58.863007 : (GM12878 : 1 of 1) Adding TileMatrix!, 4.813 mins elapsed.\n",
      "\n",
      "2024-06-25 13:36:23.739444 : (GM12878 : 1 of 1) Finished Creating Arrow File, 9.227 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-createArrows-35cc641263fc3-Date-2024-06-25_Time-13-27-08.916718.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ArrowFiles <- createArrowFiles(\n",
    "  inputFiles = c(Fragments(obj.seurat)[[1]]@path),\n",
    "  sampleNames = c(\"GM12878\"),\n",
    "  validBarcodes = list(GM12878 = unname(Fragments(obj.seurat)[[1]]@cells)),\n",
    "  addTileMat = T,\n",
    "  addGeneScoreMat = F,\n",
    "  force = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ada3db45-d050-4659-a7cf-2891e04b1249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'GM12878.arrow'"
      ],
      "text/latex": [
       "'GM12878.arrow'"
      ],
      "text/markdown": [
       "'GM12878.arrow'"
      ],
      "text/plain": [
       "[1] \"GM12878.arrow\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ArrowFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5ba28283-d488-4982-8d4c-745fc3f59365",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(dir.output, \"ArchR\", sep = \"/\"), recursive = T):\n",
      "“'/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/GM12878/2.Genome_wide_prediction/ArchR/ArchR.240625//ArchR' already exists”\n"
     ]
    }
   ],
   "source": [
    "dir.create(paste(dir.output,\"ArchR\",sep=\"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aa217507-8323-43bf-b23c-0ec25ee33c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Validating Arrows...\n",
      "\n",
      "Getting SampleNames...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Copying ArrowFiles to Ouptut Directory! If you want to save disk space set copyArrows = FALSE\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Getting Cell Metadata...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Merging Cell Metadata...\n",
      "\n",
      "Initializing ArchRProject...\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR <- ArchRProject(\n",
    "  ArrowFiles = ArrowFiles, \n",
    "  outputDirectory = paste(dir.output,\"ArchR\",sep=\"/\"),\n",
    "  copyArrows = TRUE \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f9795403-545f-4385-aa38-2c70b1b2fb21",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: ArchRProject \n",
       "outputDirectory: /maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/GM12878/2.Genome_wide_prediction/ArchR/ArchR.240625/ArchR \n",
       "samples(1): GM12878\n",
       "sampleColData names(1): ArrowFiles\n",
       "cellColData names(13): Sample TSSEnrichment ... nDiFrags BlacklistRatio\n",
       "numberOfCells(1): 3032\n",
       "medianTSS(1): 15.2195\n",
       "medianFrags(1): 14044.5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obj.ArchR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ee876dd5-0b22-46fd-9015-7e22144c5024",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking Inputs...\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addIterativeLSI-35cc6426a396df-Date-2024-06-25_Time-13-36-30.690028.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2024-06-25 13:36:31.23978 : Computing Total Across All Features, 0 mins elapsed.\n",
      "\n",
      "2024-06-25 13:36:32.250376 : Computing Top Features, 0.017 mins elapsed.\n",
      "\n",
      "###########\n",
      "2024-06-25 13:36:34.923731 : Running LSI (1 of 2) on Top Features, 0.062 mins elapsed.\n",
      "###########\n",
      "\n",
      "2024-06-25 13:36:34.940605 : Creating Partial Matrix, 0.062 mins elapsed.\n",
      "\n",
      "2024-06-25 13:36:48.020571 : Computing LSI, 0.28 mins elapsed.\n",
      "\n",
      "2024-06-25 13:37:13.533605 : Identifying Clusters, 0.705 mins elapsed.\n",
      "\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "2024-06-25 13:37:18.710526 : Identified 6 Clusters, 0.792 mins elapsed.\n",
      "\n",
      "2024-06-25 13:37:18.72603 : Saving LSI Iteration, 0.792 mins elapsed.\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************************************************************\n",
      "2024-06-25 13:37:30.354725 : ERROR Found in .saveIteration for  \n",
      "LogFile = ArchRLogs/ArchR-addIterativeLSI-35cc6426a396df-Date-2024-06-25_Time-13-36-30.690028.log\n",
      "\n",
      "<simpleError in g$grobs[[legend]]: no such index at level 2\n",
      ">\n",
      "\n",
      "************************************************************\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-06-25 13:37:30.357136 : Creating Cluster Matrix on the total Group Features, 0.986 mins elapsed.\n",
      "\n",
      "2024-06-25 13:37:49.788725 : Computing Variable Features, 1.309 mins elapsed.\n",
      "\n",
      "###########\n",
      "2024-06-25 13:37:49.988034 : Running LSI (2 of 2) on Variable Features, 1.313 mins elapsed.\n",
      "###########\n",
      "\n",
      "2024-06-25 13:37:50.001458 : Creating Partial Matrix, 1.313 mins elapsed.\n",
      "\n",
      "2024-06-25 13:38:01.459062 : Computing LSI, 1.504 mins elapsed.\n",
      "\n",
      "2024-06-25 13:38:23.008416 : Finished Running IterativeLSI, 1.863 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addIterativeLSI(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    useMatrix = \"TileMatrix\", \n",
    "    name = \"IterativeLSI\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2c1d334b-7fd7-425a-bb8d-7be4cbedd3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.ArchR = addPeakSet(obj.ArchR,\n",
    "                       peakSet = reduce(pairs.E2G))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e25c14db-eb4e-4b94-8862-eb34246a365c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeakMatrix-35cc646a80423c-Date-2024-06-25_Time-13-38-26.331674.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2024-06-25 13:38:26.62199 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2024-06-25 13:38:27.472147 : Adding GM12878 to PeakMatrix for Chr (1 of 23)!, 0.014 mins elapsed.\n",
      "\n",
      "2024-06-25 13:38:36.959518 : Adding GM12878 to PeakMatrix for Chr (2 of 23)!, 0.172 mins elapsed.\n",
      "\n",
      "2024-06-25 13:38:45.396499 : Adding GM12878 to PeakMatrix for Chr (3 of 23)!, 0.312 mins elapsed.\n",
      "\n",
      "2024-06-25 13:38:51.924284 : Adding GM12878 to PeakMatrix for Chr (4 of 23)!, 0.421 mins elapsed.\n",
      "\n",
      "2024-06-25 13:38:59.811466 : Adding GM12878 to PeakMatrix for Chr (5 of 23)!, 0.553 mins elapsed.\n",
      "\n",
      "2024-06-25 13:39:06.878126 : Adding GM12878 to PeakMatrix for Chr (6 of 23)!, 0.67 mins elapsed.\n",
      "\n",
      "2024-06-25 13:39:13.145498 : Adding GM12878 to PeakMatrix for Chr (7 of 23)!, 0.775 mins elapsed.\n",
      "\n",
      "2024-06-25 13:39:19.699338 : Adding GM12878 to PeakMatrix for Chr (8 of 23)!, 0.884 mins elapsed.\n",
      "\n",
      "2024-06-25 13:39:26.120085 : Adding GM12878 to PeakMatrix for Chr (9 of 23)!, 0.991 mins elapsed.\n",
      "\n",
      "2024-06-25 13:39:32.104066 : Adding GM12878 to PeakMatrix for Chr (10 of 23)!, 1.091 mins elapsed.\n",
      "\n",
      "2024-06-25 13:39:38.330349 : Adding GM12878 to PeakMatrix for Chr (11 of 23)!, 1.194 mins elapsed.\n",
      "\n",
      "2024-06-25 13:39:46.223357 : Adding GM12878 to PeakMatrix for Chr (12 of 23)!, 1.326 mins elapsed.\n",
      "\n",
      "2024-06-25 13:39:54.145643 : Adding GM12878 to PeakMatrix for Chr (13 of 23)!, 1.458 mins elapsed.\n",
      "\n",
      "2024-06-25 13:39:59.3756 : Adding GM12878 to PeakMatrix for Chr (14 of 23)!, 1.545 mins elapsed.\n",
      "\n",
      "2024-06-25 13:40:06.865929 : Adding GM12878 to PeakMatrix for Chr (15 of 23)!, 1.67 mins elapsed.\n",
      "\n",
      "2024-06-25 13:40:14.523677 : Adding GM12878 to PeakMatrix for Chr (16 of 23)!, 1.798 mins elapsed.\n",
      "\n",
      "2024-06-25 13:40:23.77866 : Adding GM12878 to PeakMatrix for Chr (17 of 23)!, 1.952 mins elapsed.\n",
      "\n",
      "2024-06-25 13:40:36.793298 : Adding GM12878 to PeakMatrix for Chr (18 of 23)!, 2.169 mins elapsed.\n",
      "\n",
      "2024-06-25 13:40:46.07776 : Adding GM12878 to PeakMatrix for Chr (19 of 23)!, 2.324 mins elapsed.\n",
      "\n",
      "2024-06-25 13:40:53.120045 : Adding GM12878 to PeakMatrix for Chr (20 of 23)!, 2.441 mins elapsed.\n",
      "\n",
      "2024-06-25 13:40:58.519106 : Adding GM12878 to PeakMatrix for Chr (21 of 23)!, 2.531 mins elapsed.\n",
      "\n",
      "2024-06-25 13:41:04.476985 : Adding GM12878 to PeakMatrix for Chr (22 of 23)!, 2.63 mins elapsed.\n",
      "\n",
      "2024-06-25 13:41:09.632542 : Adding GM12878 to PeakMatrix for Chr (23 of 23)!, 2.716 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeakMatrix-35cc646a80423c-Date-2024-06-25_Time-13-38-26.331674.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addPeakMatrix(obj.ArchR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "aa6d686e-4205-4eda-8d76-ed4d9e3e30e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA.Xu = ArchR:::.importFM(featureMatrix = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_230720/ArrayExpress/cellranger_res/EMTAB11264_01/outs/filtered_feature_bc_matrix.h5\",\n",
    "                             name = \"K562_Xu\",\n",
    "                             featureType = \"Gene Expression\")\n",
    "rowdata.Xu = rowData(seRNA.Xu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d843eb16-9fed-4146-b76e-0fb053f1557f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene.intersect = intersect(rownames(obj.seurat@assays$RNA$counts),\n",
    "                           rowdata.Xu$name)\n",
    "rowdata.intersect = rowdata.Xu[gene.intersect,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fdf69211-b595-4599-ac8d-1c832875c062",
   "metadata": {},
   "outputs": [],
   "source": [
    "rr = GRanges(paste0(rowdata.intersect$interval))\n",
    "mcols(rr) <- rowdata.intersect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1e8f8d64-f6be-43c2-8e76-ea1cf4ad0453",
   "metadata": {},
   "outputs": [],
   "source": [
    "count.rna = obj.seurat@assays$RNA$counts[gene.intersect,]\n",
    "colnames(count.rna) = gsub(\"GM12878_\",\"GM12878#\",colnames(count.rna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "174a97a2-ab68-478a-9085-8e8fbc3bac29",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA <- SummarizedExperiment(assays = SimpleList(counts = count.rna), \n",
    "                              rowRanges = rr)\n",
    "seRNA = sort.GenomicRanges(sortSeqlevels(seRNA), \n",
    "                           ignore.strand = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4656d98c-a9cc-486a-afaa-fe59ca6ef055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 61214 3050 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61214): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(3050): GM12878#AAACAGCCACTAAGAA-1 GM12878#AAACAGCCATAATCAC-1\n",
       "  ... GM12878#TTTGTTGGTCACAAAT-1 GM12878#TTTGTTGGTTGGTGAC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "65cb2097-9075-4ee4-9e54-dd288dc50736",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addGeneExpressionMatrix-35cc6458b2b020-Date-2024-06-25_Time-13-41-54.528176.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Overlap w/ scATAC = 1\n",
      "\n",
      "2024-06-25 13:41:55.181055 : \n",
      "\n",
      "Overlap Per Sample w/ scATAC : GM12878=3032\n",
      "\n",
      "2024-06-25 13:41:55.19265 : \n",
      "\n",
      "2024-06-25 13:41:57.077192 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2024-06-25 13:41:58.795964 : Adding GM12878 to GeneExpressionMatrix for Chr (1 of 23)!, 0.029 mins elapsed.\n",
      "\n",
      "2024-06-25 13:42:07.145231 : Adding GM12878 to GeneExpressionMatrix for Chr (2 of 23)!, 0.168 mins elapsed.\n",
      "\n",
      "2024-06-25 13:42:13.005605 : Adding GM12878 to GeneExpressionMatrix for Chr (3 of 23)!, 0.265 mins elapsed.\n",
      "\n",
      "2024-06-25 13:42:22.003324 : Adding GM12878 to GeneExpressionMatrix for Chr (4 of 23)!, 0.415 mins elapsed.\n",
      "\n",
      "2024-06-25 13:42:27.181044 : Adding GM12878 to GeneExpressionMatrix for Chr (5 of 23)!, 0.502 mins elapsed.\n",
      "\n",
      "2024-06-25 13:42:32.803717 : Adding GM12878 to GeneExpressionMatrix for Chr (6 of 23)!, 0.595 mins elapsed.\n",
      "\n",
      "2024-06-25 13:42:41.550823 : Adding GM12878 to GeneExpressionMatrix for Chr (7 of 23)!, 0.741 mins elapsed.\n",
      "\n",
      "2024-06-25 13:42:48.573644 : Adding GM12878 to GeneExpressionMatrix for Chr (8 of 23)!, 0.858 mins elapsed.\n",
      "\n",
      "2024-06-25 13:42:57.297585 : Adding GM12878 to GeneExpressionMatrix for Chr (9 of 23)!, 1.004 mins elapsed.\n",
      "\n",
      "2024-06-25 13:43:04.821326 : Adding GM12878 to GeneExpressionMatrix for Chr (10 of 23)!, 1.129 mins elapsed.\n",
      "\n",
      "2024-06-25 13:43:10.440031 : Adding GM12878 to GeneExpressionMatrix for Chr (11 of 23)!, 1.223 mins elapsed.\n",
      "\n",
      "2024-06-25 13:43:15.564446 : Adding GM12878 to GeneExpressionMatrix for Chr (12 of 23)!, 1.308 mins elapsed.\n",
      "\n",
      "2024-06-25 13:43:23.699285 : Adding GM12878 to GeneExpressionMatrix for Chr (13 of 23)!, 1.444 mins elapsed.\n",
      "\n",
      "2024-06-25 13:43:29.589326 : Adding GM12878 to GeneExpressionMatrix for Chr (14 of 23)!, 1.542 mins elapsed.\n",
      "\n",
      "2024-06-25 13:43:36.312743 : Adding GM12878 to GeneExpressionMatrix for Chr (15 of 23)!, 1.654 mins elapsed.\n",
      "\n",
      "2024-06-25 13:43:44.335872 : Adding GM12878 to GeneExpressionMatrix for Chr (16 of 23)!, 1.788 mins elapsed.\n",
      "\n",
      "2024-06-25 13:43:49.774234 : Adding GM12878 to GeneExpressionMatrix for Chr (17 of 23)!, 1.878 mins elapsed.\n",
      "\n",
      "2024-06-25 13:43:55.216904 : Adding GM12878 to GeneExpressionMatrix for Chr (18 of 23)!, 1.969 mins elapsed.\n",
      "\n",
      "2024-06-25 13:44:01.978369 : Adding GM12878 to GeneExpressionMatrix for Chr (19 of 23)!, 2.082 mins elapsed.\n",
      "\n",
      "2024-06-25 13:44:06.602415 : Adding GM12878 to GeneExpressionMatrix for Chr (20 of 23)!, 2.159 mins elapsed.\n",
      "\n",
      "2024-06-25 13:44:11.476933 : Adding GM12878 to GeneExpressionMatrix for Chr (21 of 23)!, 2.24 mins elapsed.\n",
      "\n",
      "2024-06-25 13:44:18.201168 : Adding GM12878 to GeneExpressionMatrix for Chr (22 of 23)!, 2.352 mins elapsed.\n",
      "\n",
      "2024-06-25 13:44:22.679292 : Adding GM12878 to GeneExpressionMatrix for Chr (23 of 23)!, 2.427 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addGeneExpressionMatrix-35cc6458b2b020-Date-2024-06-25_Time-13-41-54.528176.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addGeneExpressionMatrix(obj.ArchR,\n",
    "                                    seRNA = seRNA,\n",
    "                                    strictMatch = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4318703-a143-40c0-b159-274d05033bbc",
   "metadata": {},
   "source": [
    "Run ArchR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "47c8fd5f-19ec-46e6-af87-57502c1e1ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeak2GeneLinks-35cc64745d0b8d-Date-2024-06-25_Time-13-44-30.920799.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2024-06-25 13:44:32.488818 : Getting Available Matrices, 0.026 mins elapsed.\n",
      "\n",
      "No predictionScore found. Continuing without predictionScore!\n",
      "\n",
      "2024-06-25 13:44:32.866099 : Filtered Low Prediction Score Cells (0 of 3032, 0), 0 mins elapsed.\n",
      "\n",
      "Filtering 1 dims correlated > 0.75 to log10(depth + 1)\n",
      "\n",
      "2024-06-25 13:44:33.06601 : Computing KNN, 0.004 mins elapsed.\n",
      "\n",
      "2024-06-25 13:44:33.50364 : Identifying Non-Overlapping KNN pairs, 0.011 mins elapsed.\n",
      "\n",
      "2024-06-25 13:44:35.228306 : Identified 499 Groupings!, 0.04 mins elapsed.\n",
      "\n",
      "2024-06-25 13:44:35.278798 : Getting Group RNA Matrix, 0.04 mins elapsed.\n",
      "\n",
      "2024-06-25 13:44:54.772631 : Getting Group ATAC Matrix, 0.365 mins elapsed.\n",
      "\n",
      "2024-06-25 13:45:22.584987 : Normalizing Group Matrices, 0.829 mins elapsed.\n",
      "\n",
      "2024-06-25 13:45:30.744961 : Finding Peak Gene Pairings, 0.965 mins elapsed.\n",
      "\n",
      "2024-06-25 13:45:32.601272 : Computing Correlations, 0.996 mins elapsed.\n",
      "\n",
      "2024-06-25 13:46:19.773261 : Completed Peak2Gene Correlations!, 1.782 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeak2GeneLinks-35cc64745d0b8d-Date-2024-06-25_Time-13-44-30.920799.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "obj.ArchR = addPeak2GeneLinks(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    reducedDims = \"IterativeLSI\",\n",
    "    useMatrix = \"GeneExpressionMatrix\",\n",
    "    maxDist = maxDist,\n",
    "    threads = n.cores\n",
    ")\n",
    "end_time <- Sys.time()\n",
    "execution_time <- end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "128d0065-5f22-495b-b00d-743fc9843337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 1.816291 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "execution_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "36bd2df2-6b32-4b81-bd4f-39ef8913b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g = metadata(obj.ArchR@peakSet)$Peak2GeneLinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e7730413-8fd4-4952-bb13-0df7a3b15b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata(p2g)[[1]]$PeakName = paste(seqnames(metadata(p2g)[[1]]),\n",
    "                                    ranges(metadata(p2g)[[1]]),sep=\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "aeea3810-6735-4aff-9bd7-1697bef979d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g$PairName = paste(metadata(p2g)[[1]][p2g$idxATAC]$PeakName,\n",
    "                     metadata(p2g)[[2]][p2g$idxRNA]$name,\n",
    "                     sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "87f975ec-6ba4-4696-ad6c-2af65da7aa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 8771125 rows and 7 columns\n",
       "                                  idxATAC    idxRNA Correlation         FDR\n",
       "                                <integer> <integer>   <numeric>   <numeric>\n",
       "chr1-180700-181061_DDX11L1              1         1         NaN         NaN\n",
       "chr1-181282-181914_DDX11L1              2         1         NaN         NaN\n",
       "chr1-267886-268118_DDX11L1              3         1         NaN         NaN\n",
       "chr1-586095-586295_DDX11L1              4         1         NaN         NaN\n",
       "chr1-778328-779254_DDX11L1              5         1         NaN         NaN\n",
       "...                                   ...       ...         ...         ...\n",
       "chrX-155981990-155982304_WASH6P    155266     60637    0.249031 1.17733e-07\n",
       "chrX-155997270-155998098_WASH6P    155267     60637    0.181218 1.83855e-04\n",
       "chrX-156000622-156000826_WASH6P    155268     60637   -0.328591 6.42281e-13\n",
       "chrX-156004204-156004495_WASH6P    155269     60637    0.187915 9.95324e-05\n",
       "chrX-156030527-156030695_WASH6P    155270     60637   -0.143765 3.70707e-03\n",
       "                                 VarQATAC   VarQRNA               PairName\n",
       "                                <numeric> <numeric>            <character>\n",
       "chr1-180700-181061_DDX11L1       0.239100  0.199218 chr1-180700-181061_D..\n",
       "chr1-181282-181914_DDX11L1       0.608553  0.199218 chr1-181282-181914_D..\n",
       "chr1-267886-268118_DDX11L1       0.426425  0.199218 chr1-267886-268118_D..\n",
       "chr1-586095-586295_DDX11L1       0.219231  0.199218 chr1-586095-586295_D..\n",
       "chr1-778328-779254_DDX11L1       0.986862  0.199218 chr1-778328-779254_D..\n",
       "...                                   ...       ...                    ...\n",
       "chrX-155981990-155982304_WASH6P 0.0676241  0.683757 chrX-155981990-15598..\n",
       "chrX-155997270-155998098_WASH6P 0.6945321  0.683757 chrX-155997270-15599..\n",
       "chrX-156000622-156000826_WASH6P 0.0529400  0.683757 chrX-156000622-15600..\n",
       "chrX-156004204-156004495_WASH6P 0.0526889  0.683757 chrX-156004204-15600..\n",
       "chrX-156030527-156030695_WASH6P 0.0675211  0.683757 chrX-156030527-15603.."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(p2g) = p2g$PairName\n",
    "p2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c7209a32-3904-4734-93af-9c22049f652a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = pairs.E2G[pairs.E2G$PairName %in% p2g$PairName]\n",
    "mcols(pairs.E2G.res)[,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")] = p2g[pairs.E2G.res$PairName,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1ab4ed70-ce9e-4857-bd1b-63a81d27a401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2513716 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100028533-100029101      * |         AGL\n",
       "        [2]     chr1 100028533-100029101      * |      CDC14A\n",
       "        [3]     chr1 100028533-100029101      * |         DBT\n",
       "        [4]     chr1 100028533-100029101      * |        DPH5\n",
       "        [5]     chr1 100028533-100029101      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2513712]     chrX     9995654-9996249      * |       TBL1X\n",
       "  [2513713]     chrX     9995654-9996249      * |        WWC3\n",
       "  [2513714]     chrX     9997430-9997980      * |       CLCN4\n",
       "  [2513715]     chrX     9997430-9997980      * |       TBL1X\n",
       "  [2513716]     chrX     9997430-9997980      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-100028533-10002.. chr1-100028533-10002..   0.2848537\n",
       "        [2] chr1-100028533-10002.. chr1-100028533-10002..   0.3289842\n",
       "        [3] chr1-100028533-10002.. chr1-100028533-10002..   0.0718081\n",
       "        [4] chr1-100028533-10002.. chr1-100028533-10002..   0.1567433\n",
       "        [5] chr1-100028533-10002.. chr1-100028533-10002..   0.3813759\n",
       "        ...                    ...                    ...         ...\n",
       "  [2513712]   chrX-9995654-9996249 chrX-9995654-9996249..  -0.0730499\n",
       "  [2513713]   chrX-9995654-9996249 chrX-9995654-9996249..  -0.0530663\n",
       "  [2513714]   chrX-9997430-9997980 chrX-9997430-9997980..  -0.0018111\n",
       "  [2513715]   chrX-9997430-9997980 chrX-9997430-9997980..  -0.3834368\n",
       "  [2513716]   chrX-9997430-9997980 chrX-9997430-9997980..   0.0284566\n",
       "                    FDR  VarQATAC   VarQRNA\n",
       "              <numeric> <numeric> <numeric>\n",
       "        [1] 8.24388e-10  0.306853  0.927618\n",
       "        [2] 5.98794e-13  0.306853  0.977275\n",
       "        [3] 1.77429e-01  0.306853  0.886060\n",
       "        [4] 1.41961e-03  0.306853  0.946996\n",
       "        [5] 1.95473e-17  0.306853  0.988110\n",
       "        ...         ...       ...       ...\n",
       "  [2513712] 1.69260e-01  0.704753  0.982981\n",
       "  [2513713] 3.34723e-01  0.704753  0.920197\n",
       "  [2513714] 9.77926e-01  0.835925  0.712073\n",
       "  [2513715] 1.24909e-17  0.835925  0.982981\n",
       "  [2513716] 6.26569e-01  0.835925  0.920197\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46485351-50d7-461a-bf09-d795d5bdab84",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6d424e21-6a41-4422-b76c-6f1c0bda2222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2513716 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100028533-100029101      * |         AGL\n",
       "        [2]     chr1 100028533-100029101      * |      CDC14A\n",
       "        [3]     chr1 100028533-100029101      * |         DBT\n",
       "        [4]     chr1 100028533-100029101      * |        DPH5\n",
       "        [5]     chr1 100028533-100029101      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2513712]     chrX     9995654-9996249      * |       TBL1X\n",
       "  [2513713]     chrX     9995654-9996249      * |        WWC3\n",
       "  [2513714]     chrX     9997430-9997980      * |       CLCN4\n",
       "  [2513715]     chrX     9997430-9997980      * |       TBL1X\n",
       "  [2513716]     chrX     9997430-9997980      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-100028533-10002.. chr1-100028533-10002..   0.2848537\n",
       "        [2] chr1-100028533-10002.. chr1-100028533-10002..   0.3289842\n",
       "        [3] chr1-100028533-10002.. chr1-100028533-10002..   0.0718081\n",
       "        [4] chr1-100028533-10002.. chr1-100028533-10002..   0.1567433\n",
       "        [5] chr1-100028533-10002.. chr1-100028533-10002..   0.3813759\n",
       "        ...                    ...                    ...         ...\n",
       "  [2513712]   chrX-9995654-9996249 chrX-9995654-9996249..  -0.0730499\n",
       "  [2513713]   chrX-9995654-9996249 chrX-9995654-9996249..  -0.0530663\n",
       "  [2513714]   chrX-9997430-9997980 chrX-9997430-9997980..  -0.0018111\n",
       "  [2513715]   chrX-9997430-9997980 chrX-9997430-9997980..  -0.3834368\n",
       "  [2513716]   chrX-9997430-9997980 chrX-9997430-9997980..   0.0284566\n",
       "                    FDR  VarQATAC   VarQRNA\n",
       "              <numeric> <numeric> <numeric>\n",
       "        [1] 8.24388e-10  0.306853  0.927618\n",
       "        [2] 5.98794e-13  0.306853  0.977275\n",
       "        [3] 1.77429e-01  0.306853  0.886060\n",
       "        [4] 1.41961e-03  0.306853  0.946996\n",
       "        [5] 1.95473e-17  0.306853  0.988110\n",
       "        ...         ...       ...       ...\n",
       "  [2513712] 1.69260e-01  0.704753  0.982981\n",
       "  [2513713] 3.34723e-01  0.704753  0.920197\n",
       "  [2513714] 9.77926e-01  0.835925  0.712073\n",
       "  [2513715] 1.24909e-17  0.835925  0.982981\n",
       "  [2513716] 6.26569e-01  0.835925  0.920197\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2513716 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>Correlation</th><th scope=col>FDR</th><th scope=col>VarQATAC</th><th scope=col>VarQRNA</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>AGL     </td><td>K562</td><td> 0.284853741</td><td>8.243878e-10</td><td>0.3068526</td><td>0.9276185</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>CDC14A  </td><td>K562</td><td> 0.328984179</td><td>5.987939e-13</td><td>0.3068526</td><td>0.9772746</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>DBT     </td><td>K562</td><td> 0.071808078</td><td>1.774288e-01</td><td>0.3068526</td><td>0.8860597</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>DPH5    </td><td>K562</td><td> 0.156743343</td><td>1.419607e-03</td><td>0.3068526</td><td>0.9469961</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>EXTL2   </td><td>K562</td><td> 0.381375856</td><td>1.954729e-17</td><td>0.3068526</td><td>0.9881096</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>FRRS1   </td><td>K562</td><td>-0.054004679</td><td>3.253500e-01</td><td>0.3068526</td><td>0.9274370</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>MFSD14A </td><td>K562</td><td> 0.274427925</td><td>3.787000e-09</td><td>0.3068526</td><td>0.7644672</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>RTCA    </td><td>K562</td><td>-0.105225608</td><td>3.954769e-02</td><td>0.3068526</td><td>0.8887148</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>RTCA-AS1</td><td>K562</td><td>-0.215846066</td><td>5.925808e-06</td><td>0.3068526</td><td>0.8494978</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>SASS6   </td><td>K562</td><td>-0.158213070</td><td>1.266497e-03</td><td>0.3068526</td><td>0.9607500</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>SLC30A7 </td><td>K562</td><td>-0.003540864</td><td>9.565045e-01</td><td>0.3068526</td><td>0.9231822</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>SLC35A3 </td><td>K562</td><td>-0.135256821</td><td>6.648412e-03</td><td>0.3068526</td><td>0.8725531</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>TRMT13  </td><td>K562</td><td>-0.110501216</td><td>2.981516e-02</td><td>0.3068526</td><td>0.8910731</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>VCAM1   </td><td>K562</td><td> 0.400591837</td><td>2.636342e-19</td><td>0.3068526</td><td>0.8634662</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>AGL     </td><td>K562</td><td> 0.136326908</td><td>6.189733e-03</td><td>0.0212533</td><td>0.9276185</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>CDC14A  </td><td>K562</td><td>-0.036105000</td><td>5.279213e-01</td><td>0.0212533</td><td>0.9772746</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>DBT     </td><td>K562</td><td> 0.065973355</td><td>2.195423e-01</td><td>0.0212533</td><td>0.8860597</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>DPH5    </td><td>K562</td><td> 0.057069912</td><td>2.957503e-01</td><td>0.0212533</td><td>0.9469961</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>EXTL2   </td><td>K562</td><td> 0.023694142</td><td>6.900479e-01</td><td>0.0212533</td><td>0.9881096</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>FRRS1   </td><td>K562</td><td>-0.004484065</td><td>9.446231e-01</td><td>0.0212533</td><td>0.9274370</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>MFSD14A </td><td>K562</td><td> 0.143498508</td><td>3.777886e-03</td><td>0.0212533</td><td>0.7644672</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>RTCA    </td><td>K562</td><td> 0.083568503</td><td>1.105951e-01</td><td>0.0212533</td><td>0.8887148</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>RTCA-AS1</td><td>K562</td><td> 0.021450388</td><td>7.201974e-01</td><td>0.0212533</td><td>0.8494978</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>SASS6   </td><td>K562</td><td> 0.047197232</td><td>3.967271e-01</td><td>0.0212533</td><td>0.9607500</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>SLC30A7 </td><td>K562</td><td>-0.072171361</td><td>1.750097e-01</td><td>0.0212533</td><td>0.9231822</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>SLC35A3 </td><td>K562</td><td>-0.009475548</td><td>8.802512e-01</td><td>0.0212533</td><td>0.8725531</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>TRMT13  </td><td>K562</td><td> 0.110611488</td><td>2.963584e-02</td><td>0.0212533</td><td>0.8910731</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>VCAM1   </td><td>K562</td><td>-0.023526482</td><td>6.922768e-01</td><td>0.0212533</td><td>0.8634662</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034578</td><td>100034758</td><td>AGL     </td><td>K562</td><td>-0.241548480</td><td>3.011215e-07</td><td>0.4688736</td><td>0.9276185</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034578</td><td>100034758</td><td>CDC14A  </td><td>K562</td><td>-0.254209922</td><td>6.025778e-08</td><td>0.4688736</td><td>0.9772746</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>ASMT     </td><td>K562</td><td> 0.115322838</td><td>2.276656e-02</td><td>0.8158949</td><td>0.6388674</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>ASMTL    </td><td>K562</td><td> 0.150615977</td><td>2.257349e-03</td><td>0.8158949</td><td>0.9546482</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>CSF2RA   </td><td>K562</td><td>-0.022939653</td><td>7.001363e-01</td><td>0.8158949</td><td>0.6812507</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>GTPBP6   </td><td>K562</td><td> 0.382944219</td><td>1.390690e-17</td><td>0.8158949</td><td>0.8713987</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>IL3RA    </td><td>K562</td><td> 0.264878616</td><td>1.443359e-08</td><td>0.8158949</td><td>0.9808533</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>LINC00106</td><td>K562</td><td>-0.272724697</td><td>4.827873e-09</td><td>0.8158949</td><td>0.8785395</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>P2RY8    </td><td>K562</td><td>-0.298637316</td><td>9.892163e-11</td><td>0.8158949</td><td>0.9946073</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>PLCXD1   </td><td>K562</td><td> 0.505884524</td><td>4.590465e-32</td><td>0.8158949</td><td>0.9108135</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>PPP2R3B  </td><td>K562</td><td> 0.288399802</td><td>4.831186e-10</td><td>0.8158949</td><td>0.8297574</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>SLC25A6  </td><td>K562</td><td> 0.355585790</td><td>4.053317e-15</td><td>0.8158949</td><td>0.9614922</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988858</td><td>9989079</td><td>CLCN4    </td><td>K562</td><td> 0.339381223</td><td>9.016555e-14</td><td>0.1029497</td><td>0.7120735</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988858</td><td>9989079</td><td>TBL1X    </td><td>K562</td><td>-0.006516291</td><td>9.186652e-01</td><td>0.1029497</td><td>0.9829807</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988858</td><td>9989079</td><td>WWC3     </td><td>K562</td><td> 0.190741281</td><td>7.627582e-05</td><td>0.1029497</td><td>0.9201972</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>AKAP17A  </td><td>K562</td><td> 0.319698001</td><td>3.050720e-12</td><td>0.8720745</td><td>0.9605192</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>ASMT     </td><td>K562</td><td> 0.004550612</td><td>9.437768e-01</td><td>0.8720745</td><td>0.6388674</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>ASMTL    </td><td>K562</td><td>-0.503109103</td><td>1.148410e-31</td><td>0.8720745</td><td>0.9546482</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>CSF2RA   </td><td>K562</td><td>-0.244502915</td><td>2.086309e-07</td><td>0.8720745</td><td>0.6812507</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>GTPBP6   </td><td>K562</td><td> 0.018742414</td><td>7.567248e-01</td><td>0.8720745</td><td>0.8713987</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>IL3RA    </td><td>K562</td><td>-0.542062010</td><td>1.323015e-37</td><td>0.8720745</td><td>0.9808533</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>LINC00106</td><td>K562</td><td>-0.492821758</td><td>3.183864e-30</td><td>0.8720745</td><td>0.8785395</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>P2RY8    </td><td>K562</td><td> 0.516580851</td><td>1.240740e-33</td><td>0.8720745</td><td>0.9946073</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>PLCXD1   </td><td>K562</td><td> 0.148830080</td><td>2.574179e-03</td><td>0.8720745</td><td>0.9108135</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>PPP2R3B  </td><td>K562</td><td>-0.208490550</td><td>1.298196e-05</td><td>0.8720745</td><td>0.8297574</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>SLC25A6  </td><td>K562</td><td> 0.372018981</td><td>1.434578e-16</td><td>0.8720745</td><td>0.9614922</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995654</td><td>9996249</td><td>CLCN4    </td><td>K562</td><td> 0.002714168</td><td>9.668057e-01</td><td>0.7047530</td><td>0.7120735</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995654</td><td>9996249</td><td>TBL1X    </td><td>K562</td><td>-0.073049930</td><td>1.692598e-01</td><td>0.7047530</td><td>0.9829807</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995654</td><td>9996249</td><td>WWC3     </td><td>K562</td><td>-0.053066275</td><td>3.347233e-01</td><td>0.7047530</td><td>0.9201972</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997430</td><td>9997980</td><td>CLCN4    </td><td>K562</td><td>-0.001811101</td><td>9.779257e-01</td><td>0.8359245</td><td>0.7120735</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997430</td><td>9997980</td><td>TBL1X    </td><td>K562</td><td>-0.383436845</td><td>1.249085e-17</td><td>0.8359245</td><td>0.9829807</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997430</td><td>9997980</td><td>WWC3     </td><td>K562</td><td> 0.028456646</td><td>6.265693e-01</td><td>0.8359245</td><td>0.9201972</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2513716 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & Correlation & FDR & VarQATAC & VarQRNA\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100028533 & 100029101 & AGL      & K562 &  0.284853741 & 8.243878e-10 & 0.3068526 & 0.9276185\\\\\n",
       "\t chr1 & 100028533 & 100029101 & CDC14A   & K562 &  0.328984179 & 5.987939e-13 & 0.3068526 & 0.9772746\\\\\n",
       "\t chr1 & 100028533 & 100029101 & DBT      & K562 &  0.071808078 & 1.774288e-01 & 0.3068526 & 0.8860597\\\\\n",
       "\t chr1 & 100028533 & 100029101 & DPH5     & K562 &  0.156743343 & 1.419607e-03 & 0.3068526 & 0.9469961\\\\\n",
       "\t chr1 & 100028533 & 100029101 & EXTL2    & K562 &  0.381375856 & 1.954729e-17 & 0.3068526 & 0.9881096\\\\\n",
       "\t chr1 & 100028533 & 100029101 & FRRS1    & K562 & -0.054004679 & 3.253500e-01 & 0.3068526 & 0.9274370\\\\\n",
       "\t chr1 & 100028533 & 100029101 & MFSD14A  & K562 &  0.274427925 & 3.787000e-09 & 0.3068526 & 0.7644672\\\\\n",
       "\t chr1 & 100028533 & 100029101 & RTCA     & K562 & -0.105225608 & 3.954769e-02 & 0.3068526 & 0.8887148\\\\\n",
       "\t chr1 & 100028533 & 100029101 & RTCA-AS1 & K562 & -0.215846066 & 5.925808e-06 & 0.3068526 & 0.8494978\\\\\n",
       "\t chr1 & 100028533 & 100029101 & SASS6    & K562 & -0.158213070 & 1.266497e-03 & 0.3068526 & 0.9607500\\\\\n",
       "\t chr1 & 100028533 & 100029101 & SLC30A7  & K562 & -0.003540864 & 9.565045e-01 & 0.3068526 & 0.9231822\\\\\n",
       "\t chr1 & 100028533 & 100029101 & SLC35A3  & K562 & -0.135256821 & 6.648412e-03 & 0.3068526 & 0.8725531\\\\\n",
       "\t chr1 & 100028533 & 100029101 & TRMT13   & K562 & -0.110501216 & 2.981516e-02 & 0.3068526 & 0.8910731\\\\\n",
       "\t chr1 & 100028533 & 100029101 & VCAM1    & K562 &  0.400591837 & 2.636342e-19 & 0.3068526 & 0.8634662\\\\\n",
       "\t chr1 & 100029136 & 100029435 & AGL      & K562 &  0.136326908 & 6.189733e-03 & 0.0212533 & 0.9276185\\\\\n",
       "\t chr1 & 100029136 & 100029435 & CDC14A   & K562 & -0.036105000 & 5.279213e-01 & 0.0212533 & 0.9772746\\\\\n",
       "\t chr1 & 100029136 & 100029435 & DBT      & K562 &  0.065973355 & 2.195423e-01 & 0.0212533 & 0.8860597\\\\\n",
       "\t chr1 & 100029136 & 100029435 & DPH5     & K562 &  0.057069912 & 2.957503e-01 & 0.0212533 & 0.9469961\\\\\n",
       "\t chr1 & 100029136 & 100029435 & EXTL2    & K562 &  0.023694142 & 6.900479e-01 & 0.0212533 & 0.9881096\\\\\n",
       "\t chr1 & 100029136 & 100029435 & FRRS1    & K562 & -0.004484065 & 9.446231e-01 & 0.0212533 & 0.9274370\\\\\n",
       "\t chr1 & 100029136 & 100029435 & MFSD14A  & K562 &  0.143498508 & 3.777886e-03 & 0.0212533 & 0.7644672\\\\\n",
       "\t chr1 & 100029136 & 100029435 & RTCA     & K562 &  0.083568503 & 1.105951e-01 & 0.0212533 & 0.8887148\\\\\n",
       "\t chr1 & 100029136 & 100029435 & RTCA-AS1 & K562 &  0.021450388 & 7.201974e-01 & 0.0212533 & 0.8494978\\\\\n",
       "\t chr1 & 100029136 & 100029435 & SASS6    & K562 &  0.047197232 & 3.967271e-01 & 0.0212533 & 0.9607500\\\\\n",
       "\t chr1 & 100029136 & 100029435 & SLC30A7  & K562 & -0.072171361 & 1.750097e-01 & 0.0212533 & 0.9231822\\\\\n",
       "\t chr1 & 100029136 & 100029435 & SLC35A3  & K562 & -0.009475548 & 8.802512e-01 & 0.0212533 & 0.8725531\\\\\n",
       "\t chr1 & 100029136 & 100029435 & TRMT13   & K562 &  0.110611488 & 2.963584e-02 & 0.0212533 & 0.8910731\\\\\n",
       "\t chr1 & 100029136 & 100029435 & VCAM1    & K562 & -0.023526482 & 6.922768e-01 & 0.0212533 & 0.8634662\\\\\n",
       "\t chr1 & 100034578 & 100034758 & AGL      & K562 & -0.241548480 & 3.011215e-07 & 0.4688736 & 0.9276185\\\\\n",
       "\t chr1 & 100034578 & 100034758 & CDC14A   & K562 & -0.254209922 & 6.025778e-08 & 0.4688736 & 0.9772746\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  997541 &  998270 & ASMT      & K562 &  0.115322838 & 2.276656e-02 & 0.8158949 & 0.6388674\\\\\n",
       "\t chrX &  997541 &  998270 & ASMTL     & K562 &  0.150615977 & 2.257349e-03 & 0.8158949 & 0.9546482\\\\\n",
       "\t chrX &  997541 &  998270 & CSF2RA    & K562 & -0.022939653 & 7.001363e-01 & 0.8158949 & 0.6812507\\\\\n",
       "\t chrX &  997541 &  998270 & GTPBP6    & K562 &  0.382944219 & 1.390690e-17 & 0.8158949 & 0.8713987\\\\\n",
       "\t chrX &  997541 &  998270 & IL3RA     & K562 &  0.264878616 & 1.443359e-08 & 0.8158949 & 0.9808533\\\\\n",
       "\t chrX &  997541 &  998270 & LINC00106 & K562 & -0.272724697 & 4.827873e-09 & 0.8158949 & 0.8785395\\\\\n",
       "\t chrX &  997541 &  998270 & P2RY8     & K562 & -0.298637316 & 9.892163e-11 & 0.8158949 & 0.9946073\\\\\n",
       "\t chrX &  997541 &  998270 & PLCXD1    & K562 &  0.505884524 & 4.590465e-32 & 0.8158949 & 0.9108135\\\\\n",
       "\t chrX &  997541 &  998270 & PPP2R3B   & K562 &  0.288399802 & 4.831186e-10 & 0.8158949 & 0.8297574\\\\\n",
       "\t chrX &  997541 &  998270 & SLC25A6   & K562 &  0.355585790 & 4.053317e-15 & 0.8158949 & 0.9614922\\\\\n",
       "\t chrX & 9988858 & 9989079 & CLCN4     & K562 &  0.339381223 & 9.016555e-14 & 0.1029497 & 0.7120735\\\\\n",
       "\t chrX & 9988858 & 9989079 & TBL1X     & K562 & -0.006516291 & 9.186652e-01 & 0.1029497 & 0.9829807\\\\\n",
       "\t chrX & 9988858 & 9989079 & WWC3      & K562 &  0.190741281 & 7.627582e-05 & 0.1029497 & 0.9201972\\\\\n",
       "\t chrX &  999293 &  999875 & AKAP17A   & K562 &  0.319698001 & 3.050720e-12 & 0.8720745 & 0.9605192\\\\\n",
       "\t chrX &  999293 &  999875 & ASMT      & K562 &  0.004550612 & 9.437768e-01 & 0.8720745 & 0.6388674\\\\\n",
       "\t chrX &  999293 &  999875 & ASMTL     & K562 & -0.503109103 & 1.148410e-31 & 0.8720745 & 0.9546482\\\\\n",
       "\t chrX &  999293 &  999875 & CSF2RA    & K562 & -0.244502915 & 2.086309e-07 & 0.8720745 & 0.6812507\\\\\n",
       "\t chrX &  999293 &  999875 & GTPBP6    & K562 &  0.018742414 & 7.567248e-01 & 0.8720745 & 0.8713987\\\\\n",
       "\t chrX &  999293 &  999875 & IL3RA     & K562 & -0.542062010 & 1.323015e-37 & 0.8720745 & 0.9808533\\\\\n",
       "\t chrX &  999293 &  999875 & LINC00106 & K562 & -0.492821758 & 3.183864e-30 & 0.8720745 & 0.8785395\\\\\n",
       "\t chrX &  999293 &  999875 & P2RY8     & K562 &  0.516580851 & 1.240740e-33 & 0.8720745 & 0.9946073\\\\\n",
       "\t chrX &  999293 &  999875 & PLCXD1    & K562 &  0.148830080 & 2.574179e-03 & 0.8720745 & 0.9108135\\\\\n",
       "\t chrX &  999293 &  999875 & PPP2R3B   & K562 & -0.208490550 & 1.298196e-05 & 0.8720745 & 0.8297574\\\\\n",
       "\t chrX &  999293 &  999875 & SLC25A6   & K562 &  0.372018981 & 1.434578e-16 & 0.8720745 & 0.9614922\\\\\n",
       "\t chrX & 9995654 & 9996249 & CLCN4     & K562 &  0.002714168 & 9.668057e-01 & 0.7047530 & 0.7120735\\\\\n",
       "\t chrX & 9995654 & 9996249 & TBL1X     & K562 & -0.073049930 & 1.692598e-01 & 0.7047530 & 0.9829807\\\\\n",
       "\t chrX & 9995654 & 9996249 & WWC3      & K562 & -0.053066275 & 3.347233e-01 & 0.7047530 & 0.9201972\\\\\n",
       "\t chrX & 9997430 & 9997980 & CLCN4     & K562 & -0.001811101 & 9.779257e-01 & 0.8359245 & 0.7120735\\\\\n",
       "\t chrX & 9997430 & 9997980 & TBL1X     & K562 & -0.383436845 & 1.249085e-17 & 0.8359245 & 0.9829807\\\\\n",
       "\t chrX & 9997430 & 9997980 & WWC3      & K562 &  0.028456646 & 6.265693e-01 & 0.8359245 & 0.9201972\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2513716 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | Correlation &lt;dbl&gt; | FDR &lt;dbl&gt; | VarQATAC &lt;dbl&gt; | VarQRNA &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100028533 | 100029101 | AGL      | K562 |  0.284853741 | 8.243878e-10 | 0.3068526 | 0.9276185 |\n",
       "| chr1 | 100028533 | 100029101 | CDC14A   | K562 |  0.328984179 | 5.987939e-13 | 0.3068526 | 0.9772746 |\n",
       "| chr1 | 100028533 | 100029101 | DBT      | K562 |  0.071808078 | 1.774288e-01 | 0.3068526 | 0.8860597 |\n",
       "| chr1 | 100028533 | 100029101 | DPH5     | K562 |  0.156743343 | 1.419607e-03 | 0.3068526 | 0.9469961 |\n",
       "| chr1 | 100028533 | 100029101 | EXTL2    | K562 |  0.381375856 | 1.954729e-17 | 0.3068526 | 0.9881096 |\n",
       "| chr1 | 100028533 | 100029101 | FRRS1    | K562 | -0.054004679 | 3.253500e-01 | 0.3068526 | 0.9274370 |\n",
       "| chr1 | 100028533 | 100029101 | MFSD14A  | K562 |  0.274427925 | 3.787000e-09 | 0.3068526 | 0.7644672 |\n",
       "| chr1 | 100028533 | 100029101 | RTCA     | K562 | -0.105225608 | 3.954769e-02 | 0.3068526 | 0.8887148 |\n",
       "| chr1 | 100028533 | 100029101 | RTCA-AS1 | K562 | -0.215846066 | 5.925808e-06 | 0.3068526 | 0.8494978 |\n",
       "| chr1 | 100028533 | 100029101 | SASS6    | K562 | -0.158213070 | 1.266497e-03 | 0.3068526 | 0.9607500 |\n",
       "| chr1 | 100028533 | 100029101 | SLC30A7  | K562 | -0.003540864 | 9.565045e-01 | 0.3068526 | 0.9231822 |\n",
       "| chr1 | 100028533 | 100029101 | SLC35A3  | K562 | -0.135256821 | 6.648412e-03 | 0.3068526 | 0.8725531 |\n",
       "| chr1 | 100028533 | 100029101 | TRMT13   | K562 | -0.110501216 | 2.981516e-02 | 0.3068526 | 0.8910731 |\n",
       "| chr1 | 100028533 | 100029101 | VCAM1    | K562 |  0.400591837 | 2.636342e-19 | 0.3068526 | 0.8634662 |\n",
       "| chr1 | 100029136 | 100029435 | AGL      | K562 |  0.136326908 | 6.189733e-03 | 0.0212533 | 0.9276185 |\n",
       "| chr1 | 100029136 | 100029435 | CDC14A   | K562 | -0.036105000 | 5.279213e-01 | 0.0212533 | 0.9772746 |\n",
       "| chr1 | 100029136 | 100029435 | DBT      | K562 |  0.065973355 | 2.195423e-01 | 0.0212533 | 0.8860597 |\n",
       "| chr1 | 100029136 | 100029435 | DPH5     | K562 |  0.057069912 | 2.957503e-01 | 0.0212533 | 0.9469961 |\n",
       "| chr1 | 100029136 | 100029435 | EXTL2    | K562 |  0.023694142 | 6.900479e-01 | 0.0212533 | 0.9881096 |\n",
       "| chr1 | 100029136 | 100029435 | FRRS1    | K562 | -0.004484065 | 9.446231e-01 | 0.0212533 | 0.9274370 |\n",
       "| chr1 | 100029136 | 100029435 | MFSD14A  | K562 |  0.143498508 | 3.777886e-03 | 0.0212533 | 0.7644672 |\n",
       "| chr1 | 100029136 | 100029435 | RTCA     | K562 |  0.083568503 | 1.105951e-01 | 0.0212533 | 0.8887148 |\n",
       "| chr1 | 100029136 | 100029435 | RTCA-AS1 | K562 |  0.021450388 | 7.201974e-01 | 0.0212533 | 0.8494978 |\n",
       "| chr1 | 100029136 | 100029435 | SASS6    | K562 |  0.047197232 | 3.967271e-01 | 0.0212533 | 0.9607500 |\n",
       "| chr1 | 100029136 | 100029435 | SLC30A7  | K562 | -0.072171361 | 1.750097e-01 | 0.0212533 | 0.9231822 |\n",
       "| chr1 | 100029136 | 100029435 | SLC35A3  | K562 | -0.009475548 | 8.802512e-01 | 0.0212533 | 0.8725531 |\n",
       "| chr1 | 100029136 | 100029435 | TRMT13   | K562 |  0.110611488 | 2.963584e-02 | 0.0212533 | 0.8910731 |\n",
       "| chr1 | 100029136 | 100029435 | VCAM1    | K562 | -0.023526482 | 6.922768e-01 | 0.0212533 | 0.8634662 |\n",
       "| chr1 | 100034578 | 100034758 | AGL      | K562 | -0.241548480 | 3.011215e-07 | 0.4688736 | 0.9276185 |\n",
       "| chr1 | 100034578 | 100034758 | CDC14A   | K562 | -0.254209922 | 6.025778e-08 | 0.4688736 | 0.9772746 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  997541 |  998270 | ASMT      | K562 |  0.115322838 | 2.276656e-02 | 0.8158949 | 0.6388674 |\n",
       "| chrX |  997541 |  998270 | ASMTL     | K562 |  0.150615977 | 2.257349e-03 | 0.8158949 | 0.9546482 |\n",
       "| chrX |  997541 |  998270 | CSF2RA    | K562 | -0.022939653 | 7.001363e-01 | 0.8158949 | 0.6812507 |\n",
       "| chrX |  997541 |  998270 | GTPBP6    | K562 |  0.382944219 | 1.390690e-17 | 0.8158949 | 0.8713987 |\n",
       "| chrX |  997541 |  998270 | IL3RA     | K562 |  0.264878616 | 1.443359e-08 | 0.8158949 | 0.9808533 |\n",
       "| chrX |  997541 |  998270 | LINC00106 | K562 | -0.272724697 | 4.827873e-09 | 0.8158949 | 0.8785395 |\n",
       "| chrX |  997541 |  998270 | P2RY8     | K562 | -0.298637316 | 9.892163e-11 | 0.8158949 | 0.9946073 |\n",
       "| chrX |  997541 |  998270 | PLCXD1    | K562 |  0.505884524 | 4.590465e-32 | 0.8158949 | 0.9108135 |\n",
       "| chrX |  997541 |  998270 | PPP2R3B   | K562 |  0.288399802 | 4.831186e-10 | 0.8158949 | 0.8297574 |\n",
       "| chrX |  997541 |  998270 | SLC25A6   | K562 |  0.355585790 | 4.053317e-15 | 0.8158949 | 0.9614922 |\n",
       "| chrX | 9988858 | 9989079 | CLCN4     | K562 |  0.339381223 | 9.016555e-14 | 0.1029497 | 0.7120735 |\n",
       "| chrX | 9988858 | 9989079 | TBL1X     | K562 | -0.006516291 | 9.186652e-01 | 0.1029497 | 0.9829807 |\n",
       "| chrX | 9988858 | 9989079 | WWC3      | K562 |  0.190741281 | 7.627582e-05 | 0.1029497 | 0.9201972 |\n",
       "| chrX |  999293 |  999875 | AKAP17A   | K562 |  0.319698001 | 3.050720e-12 | 0.8720745 | 0.9605192 |\n",
       "| chrX |  999293 |  999875 | ASMT      | K562 |  0.004550612 | 9.437768e-01 | 0.8720745 | 0.6388674 |\n",
       "| chrX |  999293 |  999875 | ASMTL     | K562 | -0.503109103 | 1.148410e-31 | 0.8720745 | 0.9546482 |\n",
       "| chrX |  999293 |  999875 | CSF2RA    | K562 | -0.244502915 | 2.086309e-07 | 0.8720745 | 0.6812507 |\n",
       "| chrX |  999293 |  999875 | GTPBP6    | K562 |  0.018742414 | 7.567248e-01 | 0.8720745 | 0.8713987 |\n",
       "| chrX |  999293 |  999875 | IL3RA     | K562 | -0.542062010 | 1.323015e-37 | 0.8720745 | 0.9808533 |\n",
       "| chrX |  999293 |  999875 | LINC00106 | K562 | -0.492821758 | 3.183864e-30 | 0.8720745 | 0.8785395 |\n",
       "| chrX |  999293 |  999875 | P2RY8     | K562 |  0.516580851 | 1.240740e-33 | 0.8720745 | 0.9946073 |\n",
       "| chrX |  999293 |  999875 | PLCXD1    | K562 |  0.148830080 | 2.574179e-03 | 0.8720745 | 0.9108135 |\n",
       "| chrX |  999293 |  999875 | PPP2R3B   | K562 | -0.208490550 | 1.298196e-05 | 0.8720745 | 0.8297574 |\n",
       "| chrX |  999293 |  999875 | SLC25A6   | K562 |  0.372018981 | 1.434578e-16 | 0.8720745 | 0.9614922 |\n",
       "| chrX | 9995654 | 9996249 | CLCN4     | K562 |  0.002714168 | 9.668057e-01 | 0.7047530 | 0.7120735 |\n",
       "| chrX | 9995654 | 9996249 | TBL1X     | K562 | -0.073049930 | 1.692598e-01 | 0.7047530 | 0.9829807 |\n",
       "| chrX | 9995654 | 9996249 | WWC3      | K562 | -0.053066275 | 3.347233e-01 | 0.7047530 | 0.9201972 |\n",
       "| chrX | 9997430 | 9997980 | CLCN4     | K562 | -0.001811101 | 9.779257e-01 | 0.8359245 | 0.7120735 |\n",
       "| chrX | 9997430 | 9997980 | TBL1X     | K562 | -0.383436845 | 1.249085e-17 | 0.8359245 | 0.9829807 |\n",
       "| chrX | 9997430 | 9997980 | WWC3      | K562 |  0.028456646 | 6.265693e-01 | 0.8359245 | 0.9201972 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType Correlation  FDR         \n",
       "1       chr1 100028533 100029101 AGL        K562      0.284853741 8.243878e-10\n",
       "2       chr1 100028533 100029101 CDC14A     K562      0.328984179 5.987939e-13\n",
       "3       chr1 100028533 100029101 DBT        K562      0.071808078 1.774288e-01\n",
       "4       chr1 100028533 100029101 DPH5       K562      0.156743343 1.419607e-03\n",
       "5       chr1 100028533 100029101 EXTL2      K562      0.381375856 1.954729e-17\n",
       "6       chr1 100028533 100029101 FRRS1      K562     -0.054004679 3.253500e-01\n",
       "7       chr1 100028533 100029101 MFSD14A    K562      0.274427925 3.787000e-09\n",
       "8       chr1 100028533 100029101 RTCA       K562     -0.105225608 3.954769e-02\n",
       "9       chr1 100028533 100029101 RTCA-AS1   K562     -0.215846066 5.925808e-06\n",
       "10      chr1 100028533 100029101 SASS6      K562     -0.158213070 1.266497e-03\n",
       "11      chr1 100028533 100029101 SLC30A7    K562     -0.003540864 9.565045e-01\n",
       "12      chr1 100028533 100029101 SLC35A3    K562     -0.135256821 6.648412e-03\n",
       "13      chr1 100028533 100029101 TRMT13     K562     -0.110501216 2.981516e-02\n",
       "14      chr1 100028533 100029101 VCAM1      K562      0.400591837 2.636342e-19\n",
       "15      chr1 100029136 100029435 AGL        K562      0.136326908 6.189733e-03\n",
       "16      chr1 100029136 100029435 CDC14A     K562     -0.036105000 5.279213e-01\n",
       "17      chr1 100029136 100029435 DBT        K562      0.065973355 2.195423e-01\n",
       "18      chr1 100029136 100029435 DPH5       K562      0.057069912 2.957503e-01\n",
       "19      chr1 100029136 100029435 EXTL2      K562      0.023694142 6.900479e-01\n",
       "20      chr1 100029136 100029435 FRRS1      K562     -0.004484065 9.446231e-01\n",
       "21      chr1 100029136 100029435 MFSD14A    K562      0.143498508 3.777886e-03\n",
       "22      chr1 100029136 100029435 RTCA       K562      0.083568503 1.105951e-01\n",
       "23      chr1 100029136 100029435 RTCA-AS1   K562      0.021450388 7.201974e-01\n",
       "24      chr1 100029136 100029435 SASS6      K562      0.047197232 3.967271e-01\n",
       "25      chr1 100029136 100029435 SLC30A7    K562     -0.072171361 1.750097e-01\n",
       "26      chr1 100029136 100029435 SLC35A3    K562     -0.009475548 8.802512e-01\n",
       "27      chr1 100029136 100029435 TRMT13     K562      0.110611488 2.963584e-02\n",
       "28      chr1 100029136 100029435 VCAM1      K562     -0.023526482 6.922768e-01\n",
       "29      chr1 100034578 100034758 AGL        K562     -0.241548480 3.011215e-07\n",
       "30      chr1 100034578 100034758 CDC14A     K562     -0.254209922 6.025778e-08\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮            ⋮           \n",
       "2513687 chrX  997541    998270   ASMT       K562      0.115322838 2.276656e-02\n",
       "2513688 chrX  997541    998270   ASMTL      K562      0.150615977 2.257349e-03\n",
       "2513689 chrX  997541    998270   CSF2RA     K562     -0.022939653 7.001363e-01\n",
       "2513690 chrX  997541    998270   GTPBP6     K562      0.382944219 1.390690e-17\n",
       "2513691 chrX  997541    998270   IL3RA      K562      0.264878616 1.443359e-08\n",
       "2513692 chrX  997541    998270   LINC00106  K562     -0.272724697 4.827873e-09\n",
       "2513693 chrX  997541    998270   P2RY8      K562     -0.298637316 9.892163e-11\n",
       "2513694 chrX  997541    998270   PLCXD1     K562      0.505884524 4.590465e-32\n",
       "2513695 chrX  997541    998270   PPP2R3B    K562      0.288399802 4.831186e-10\n",
       "2513696 chrX  997541    998270   SLC25A6    K562      0.355585790 4.053317e-15\n",
       "2513697 chrX 9988858   9989079   CLCN4      K562      0.339381223 9.016555e-14\n",
       "2513698 chrX 9988858   9989079   TBL1X      K562     -0.006516291 9.186652e-01\n",
       "2513699 chrX 9988858   9989079   WWC3       K562      0.190741281 7.627582e-05\n",
       "2513700 chrX  999293    999875   AKAP17A    K562      0.319698001 3.050720e-12\n",
       "2513701 chrX  999293    999875   ASMT       K562      0.004550612 9.437768e-01\n",
       "2513702 chrX  999293    999875   ASMTL      K562     -0.503109103 1.148410e-31\n",
       "2513703 chrX  999293    999875   CSF2RA     K562     -0.244502915 2.086309e-07\n",
       "2513704 chrX  999293    999875   GTPBP6     K562      0.018742414 7.567248e-01\n",
       "2513705 chrX  999293    999875   IL3RA      K562     -0.542062010 1.323015e-37\n",
       "2513706 chrX  999293    999875   LINC00106  K562     -0.492821758 3.183864e-30\n",
       "2513707 chrX  999293    999875   P2RY8      K562      0.516580851 1.240740e-33\n",
       "2513708 chrX  999293    999875   PLCXD1     K562      0.148830080 2.574179e-03\n",
       "2513709 chrX  999293    999875   PPP2R3B    K562     -0.208490550 1.298196e-05\n",
       "2513710 chrX  999293    999875   SLC25A6    K562      0.372018981 1.434578e-16\n",
       "2513711 chrX 9995654   9996249   CLCN4      K562      0.002714168 9.668057e-01\n",
       "2513712 chrX 9995654   9996249   TBL1X      K562     -0.073049930 1.692598e-01\n",
       "2513713 chrX 9995654   9996249   WWC3       K562     -0.053066275 3.347233e-01\n",
       "2513714 chrX 9997430   9997980   CLCN4      K562     -0.001811101 9.779257e-01\n",
       "2513715 chrX 9997430   9997980   TBL1X      K562     -0.383436845 1.249085e-17\n",
       "2513716 chrX 9997430   9997980   WWC3       K562      0.028456646 6.265693e-01\n",
       "        VarQATAC  VarQRNA  \n",
       "1       0.3068526 0.9276185\n",
       "2       0.3068526 0.9772746\n",
       "3       0.3068526 0.8860597\n",
       "4       0.3068526 0.9469961\n",
       "5       0.3068526 0.9881096\n",
       "6       0.3068526 0.9274370\n",
       "7       0.3068526 0.7644672\n",
       "8       0.3068526 0.8887148\n",
       "9       0.3068526 0.8494978\n",
       "10      0.3068526 0.9607500\n",
       "11      0.3068526 0.9231822\n",
       "12      0.3068526 0.8725531\n",
       "13      0.3068526 0.8910731\n",
       "14      0.3068526 0.8634662\n",
       "15      0.0212533 0.9276185\n",
       "16      0.0212533 0.9772746\n",
       "17      0.0212533 0.8860597\n",
       "18      0.0212533 0.9469961\n",
       "19      0.0212533 0.9881096\n",
       "20      0.0212533 0.9274370\n",
       "21      0.0212533 0.7644672\n",
       "22      0.0212533 0.8887148\n",
       "23      0.0212533 0.8494978\n",
       "24      0.0212533 0.9607500\n",
       "25      0.0212533 0.9231822\n",
       "26      0.0212533 0.8725531\n",
       "27      0.0212533 0.8910731\n",
       "28      0.0212533 0.8634662\n",
       "29      0.4688736 0.9276185\n",
       "30      0.4688736 0.9772746\n",
       "⋮       ⋮         ⋮        \n",
       "2513687 0.8158949 0.6388674\n",
       "2513688 0.8158949 0.9546482\n",
       "2513689 0.8158949 0.6812507\n",
       "2513690 0.8158949 0.8713987\n",
       "2513691 0.8158949 0.9808533\n",
       "2513692 0.8158949 0.8785395\n",
       "2513693 0.8158949 0.9946073\n",
       "2513694 0.8158949 0.9108135\n",
       "2513695 0.8158949 0.8297574\n",
       "2513696 0.8158949 0.9614922\n",
       "2513697 0.1029497 0.7120735\n",
       "2513698 0.1029497 0.9829807\n",
       "2513699 0.1029497 0.9201972\n",
       "2513700 0.8720745 0.9605192\n",
       "2513701 0.8720745 0.6388674\n",
       "2513702 0.8720745 0.9546482\n",
       "2513703 0.8720745 0.6812507\n",
       "2513704 0.8720745 0.8713987\n",
       "2513705 0.8720745 0.9808533\n",
       "2513706 0.8720745 0.8785395\n",
       "2513707 0.8720745 0.9946073\n",
       "2513708 0.8720745 0.9108135\n",
       "2513709 0.8720745 0.8297574\n",
       "2513710 0.8720745 0.9614922\n",
       "2513711 0.7047530 0.7120735\n",
       "2513712 0.7047530 0.9829807\n",
       "2513713 0.7047530 0.9201972\n",
       "2513714 0.8359245 0.7120735\n",
       "2513715 0.8359245 0.9829807\n",
       "2513716 0.8359245 0.9201972"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(obj.ArchR,\n",
    "        paste(dir.output,\"obj.ArchR.rds\",sep = \"/\"))\n",
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"Correlation\",\n",
    "                         \"FDR\",\n",
    "                         \"VarQATAC\",\n",
    "                         \"VarQRNA\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "pairs.E2G.res\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4c909ed5-7321-4e69-8208-1544b371da6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.9 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] nabor_0.5.0                       uwot_0.1.16                      \n",
       " [3] Rsamtools_2.18.0                  BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] genomation_1.34.0                 Signac_1.13.0                    \n",
       "[11] Seurat_5.0.3                      SeuratObject_5.0.1               \n",
       "[13] sp_2.1-4                          rhdf5_2.44.0                     \n",
       "[15] SummarizedExperiment_1.30.2       Biobase_2.60.0                   \n",
       "[17] MatrixGenerics_1.12.3             Rcpp_1.0.12                      \n",
       "[19] Matrix_1.6-5                      GenomicRanges_1.54.1             \n",
       "[21] GenomeInfoDb_1.38.1               IRanges_2.36.0                   \n",
       "[23] S4Vectors_0.40.2                  BiocGenerics_0.48.1              \n",
       "[25] matrixStats_1.3.0                 data.table_1.15.2                \n",
       "[27] stringr_1.5.1                     plyr_1.8.9                       \n",
       "[29] magrittr_2.0.3                    ggplot2_3.5.1                    \n",
       "[31] gtable_0.3.5                      gtools_3.9.5                     \n",
       "[33] gridExtra_2.3                     ArchR_1.0.2                      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22         splines_4.3.3            later_1.3.2             \n",
       "  [4] pbdZMQ_0.3-11            BiocIO_1.10.0            bitops_1.0-7            \n",
       "  [7] tibble_3.2.1             polyclip_1.10-6          XML_3.99-0.16.1         \n",
       " [10] fastDummies_1.7.3        lifecycle_1.0.4          vroom_1.6.5             \n",
       " [13] globals_0.16.3           lattice_0.22-6           MASS_7.3-60             \n",
       " [16] plotly_4.10.4            plotrix_3.8-4            yaml_2.3.8              \n",
       " [19] httpuv_1.6.15            sctransform_0.4.1        spam_2.10-0             \n",
       " [22] spatstat.sparse_3.0-3    reticulate_1.36.1        cowplot_1.1.3           \n",
       " [25] pbapply_1.7-2            RColorBrewer_1.1-3       abind_1.4-5             \n",
       " [28] zlibbioc_1.48.0          Rtsne_0.17               purrr_1.0.2             \n",
       " [31] RCurl_1.98-1.14          GenomeInfoDbData_1.2.11  ggrepel_0.9.5           \n",
       " [34] irlba_2.3.5.1            listenv_0.9.1            spatstat.utils_3.0-4    \n",
       " [37] goftest_1.2-3            RSpectra_0.16-1          spatstat.random_3.2-3   \n",
       " [40] fitdistrplus_1.1-11      parallelly_1.37.1        leiden_0.4.3.1          \n",
       " [43] codetools_0.2-20         DelayedArray_0.26.7      RcppRoll_0.3.0          \n",
       " [46] tidyselect_1.2.1         farver_2.1.1             base64enc_0.1-3         \n",
       " [49] spatstat.explore_3.2-6   GenomicAlignments_1.36.0 jsonlite_1.8.8          \n",
       " [52] progressr_0.14.0         ggridges_0.5.6           survival_3.6-4          \n",
       " [55] tools_4.3.3              ica_1.0-3                glue_1.7.0              \n",
       " [58] IRdisplay_1.1            dplyr_1.1.4              withr_3.0.0             \n",
       " [61] fastmap_1.1.1            rhdf5filters_1.12.1      fansi_1.0.6             \n",
       " [64] digest_0.6.35            R6_2.5.1                 mime_0.12               \n",
       " [67] seqPattern_1.34.0        colorspace_2.1-0         Cairo_1.6-2             \n",
       " [70] scattermore_1.2          tensor_1.5               spatstat.data_3.0-4     \n",
       " [73] utf8_1.2.4               tidyr_1.3.1              generics_0.1.3          \n",
       " [76] httr_1.4.7               htmlwidgets_1.6.4        S4Arrays_1.2.1          \n",
       " [79] pkgconfig_2.0.3          impute_1.76.0            lmtest_0.9-40           \n",
       " [82] htmltools_0.5.8.1        dotCall64_1.1-1          scales_1.3.0            \n",
       " [85] png_0.1-8                tzdb_0.4.0               reshape2_1.4.4          \n",
       " [88] rjson_0.2.21             uuid_1.2-0               nlme_3.1-164            \n",
       " [91] repr_1.1.7               zoo_1.8-12               KernSmooth_2.23-22      \n",
       " [94] miniUI_0.1.1.1           restfulr_0.0.15          pillar_1.9.0            \n",
       " [97] vctrs_0.6.5              RANN_2.6.1               promises_1.3.0          \n",
       "[100] xtable_1.8-4             cluster_2.1.6            evaluate_0.23           \n",
       "[103] readr_2.1.5              cli_3.6.2                compiler_4.3.3          \n",
       "[106] rlang_1.1.3              crayon_1.5.2             future.apply_1.11.2     \n",
       "[109] labeling_0.4.3           stringi_1.8.3            gridBase_0.4-7          \n",
       "[112] viridisLite_0.4.2        deldir_2.0-4             BiocParallel_1.36.0     \n",
       "[115] munsell_0.5.1            lazyeval_0.2.2           spatstat.geom_3.2-9     \n",
       "[118] IRkernel_1.3.2           RcppHNSW_0.6.0           hms_1.1.3               \n",
       "[121] patchwork_1.2.0          bit64_4.0.5              future_1.33.2           \n",
       "[124] Rhdf5lib_1.22.1          shiny_1.8.1.1            ROCR_1.0-11             \n",
       "[127] igraph_2.0.3             fastmatch_1.1-4          bit_4.0.5               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7690cee-49f1-4886-ae66-0b63a43ce524",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
