{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8466b35-cae5-4f62-8138-37b2de484a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.2\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : grid v4.3.3\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "\tLoading Package : gtools v3.9.5\n",
      "\n",
      "\tLoading Package : gtable v0.3.5\n",
      "\n",
      "\tLoading Package : ggplot2 v3.5.1\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "\tLoading Package : plyr v1.8.9\n",
      "\n",
      "\tLoading Package : stringr v1.5.1\n",
      "\n",
      "\tLoading Package : data.table v1.15.2\n",
      "\n",
      "\tLoading Package : matrixStats v1.3.0\n",
      "\n",
      "\tLoading Package : S4Vectors v0.40.2\n",
      "\n",
      "\tLoading Package : GenomicRanges v1.54.1\n",
      "\n",
      "\tLoading Package : BiocGenerics v0.48.1\n",
      "\n",
      "\tLoading Package : Matrix v1.6.5\n",
      "\n",
      "\tLoading Package : Rcpp v1.0.12\n",
      "\n",
      "\tLoading Package : SummarizedExperiment v1.30.2\n",
      "\n",
      "\tLoading Package : rhdf5 v2.44.0\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(magrittr)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e85020d7-d358-492d-9358-c81270f4b624",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1bec6690-52ec-450d-bba7-7d4528a185cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.seurat = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/GM12878/1.prepare_data/1.seurat_pipeline.240615/obj.seurat.qc.rds\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240617/sc-E2G/test/results/GM12878/GM12878.multiome_7features/Kendall/Pairs.tsv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/GM12878/2.Genome_wide_prediction/ArchR_250226/ArchR.250226/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "340dac32-44ac-4231-b964-b5f3b9e2dbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16\n",
    "maxDist = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da973385-4c81-4b1c-9669-cf5933b23a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 16.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRThreads(threads = n.cores) \n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeb26496-d3b9-4778-92ed-9c63a5018aef",
   "metadata": {},
   "source": [
    "Import seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c9b19a09-be95-4d9a-bdc3-0530980d4757",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.seurat = readRDS(path.seurat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86be3e66-c954-4620-9e64-c6a1c0fcb84c",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6f363969-bf0b-45ce-858c-a8cb64cc210a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3538ffdc-4fed-48eb-9c9d-86805a91cf06",
   "metadata": {},
   "source": [
    "Create an ArchRProject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0300f485-fafa-4cef-b9d0-534108ec6f95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 1\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240614/GM12878/cellranger_res/\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"a5984ec1fe8d8e28c00a2146ef41d26d\" \"6e3871eeb1cd823c9aaa4c7bb93361fd\"\n",
      "  .. ..@ cells: Named chr [1:3050] \"AAACAGCCACTAAGAA-1\" \"AAACAGCCATAATCAC-1\" \"AAACATGCACTTGTTC-1\" \"AAACATGCATAATCAC-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:3050] \"GM12878_AAACAGCCACTAAGAA-1\" \"GM12878_AAACAGCCATAATCAC-1\" \"GM12878_AAACATGCACTTGTTC-1\" \"GM12878_AAACATGCATAATCAC-1\" ...\n"
     ]
    }
   ],
   "source": [
    "str(Fragments(obj.seurat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1753ffaa-89e2-43bf-bcf6-19937782d63a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-createArrows-2a915716ecc94e-Date-2025-02-26_Time-14-54-23.95099.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Cleaning Temporary Files\n",
      "\n",
      "2025-02-26 14:54:24.464124 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "(GM12878 : 1 of 1) Determining Arrow Method to use!\n",
      "\n",
      "2025-02-26 14:54:24.528878 : (GM12878 : 1 of 1) Reading In Fragments from inputFiles (readMethod = tabix), 0.001 mins elapsed.\n",
      "\n",
      "2025-02-26 14:54:24.537632 : (GM12878 : 1 of 1) Tabix Bed To Temporary File, 0.001 mins elapsed.\n",
      "\n",
      "2025-02-26 14:55:20.567719 : (GM12878 : 1 of 1) Successful creation of Temporary File, 0.935 mins elapsed.\n",
      "\n",
      "2025-02-26 14:55:20.575225 : (GM12878 : 1 of 1) Creating ArrowFile From Temporary File, 0.935 mins elapsed.\n",
      "\n",
      "2025-02-26 14:55:47.069499 : (GM12878 : 1 of 1) Successful creation of Arrow File, 1.377 mins elapsed.\n",
      "\n",
      "2025-02-26 14:56:05.539683 : Continuing through after error ggplot for Fragment Size Distribution, 1.685 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-02-26 14:56:36.059578 : (GM12878 : 1 of 1) CellStats : Number of Cells Pass Filter = 3032 , 2.193 mins elapsed.\n",
      "\n",
      "2025-02-26 14:56:36.064623 : (GM12878 : 1 of 1) CellStats : Median Frags = 14044.5 , 2.193 mins elapsed.\n",
      "\n",
      "2025-02-26 14:56:36.069814 : (GM12878 : 1 of 1) CellStats : Median TSS Enrichment = 15.2195 , 2.193 mins elapsed.\n",
      "\n",
      "2025-02-26 14:56:36.802024 : Continuing through after error ggplot for TSS by Frags, 2.206 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-02-26 14:56:40.302992 : (GM12878 : 1 of 1) Adding Additional Feature Counts!, 2.264 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:03.240424 : (GM12878 : 1 of 1) Removing Fragments from Filtered Cells, 2.646 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:03.251744 : (GM12878 : 1 of 1) Creating Filtered Arrow File, 2.646 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:28.179487 : (GM12878 : 1 of 1) Finished Constructing Filtered Arrow File!, 4.062 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:28.400719 : (GM12878 : 1 of 1) Adding TileMatrix!, 4.066 mins elapsed.\n",
      "\n",
      "2025-02-26 15:02:43.376103 : (GM12878 : 1 of 1) Finished Creating Arrow File, 8.315 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-createArrows-2a915716ecc94e-Date-2025-02-26_Time-14-54-23.95099.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ArrowFiles <- createArrowFiles(\n",
    "  inputFiles = c(Fragments(obj.seurat)[[1]]@path),\n",
    "  sampleNames = c(\"GM12878\"),\n",
    "  validBarcodes = list(GM12878 = unname(Fragments(obj.seurat)[[1]]@cells)),\n",
    "  addTileMat = T,\n",
    "  addGeneScoreMat = F,\n",
    "  force = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ada3db45-d050-4659-a7cf-2891e04b1249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'GM12878.arrow'"
      ],
      "text/latex": [
       "'GM12878.arrow'"
      ],
      "text/markdown": [
       "'GM12878.arrow'"
      ],
      "text/plain": [
       "[1] \"GM12878.arrow\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ArrowFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5ba28283-d488-4982-8d4c-745fc3f59365",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"ArchR\",sep=\"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aa217507-8323-43bf-b23c-0ec25ee33c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Validating Arrows...\n",
      "\n",
      "Getting SampleNames...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Copying ArrowFiles to Ouptut Directory! If you want to save disk space set copyArrows = FALSE\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Getting Cell Metadata...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Merging Cell Metadata...\n",
      "\n",
      "Initializing ArchRProject...\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR <- ArchRProject(\n",
    "  ArrowFiles = ArrowFiles, \n",
    "  outputDirectory = paste(dir.output,\"ArchR\",sep=\"/\"),\n",
    "  copyArrows = TRUE \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f9795403-545f-4385-aa38-2c70b1b2fb21",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: ArchRProject \n",
       "outputDirectory: /maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/GM12878/2.Genome_wide_prediction/ArchR_250226/ArchR.250226/ArchR \n",
       "samples(1): GM12878\n",
       "sampleColData names(1): ArrowFiles\n",
       "cellColData names(13): Sample TSSEnrichment ... nDiFrags BlacklistRatio\n",
       "numberOfCells(1): 3032\n",
       "medianTSS(1): 15.2195\n",
       "medianFrags(1): 14044.5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obj.ArchR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ee876dd5-0b22-46fd-9015-7e22144c5024",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking Inputs...\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addIterativeLSI-2a915759d0365e-Date-2025-02-26_Time-15-02-49.477491.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 15:02:50.053065 : Computing Total Across All Features, 0 mins elapsed.\n",
      "\n",
      "2025-02-26 15:02:51.096116 : Computing Top Features, 0.018 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-26 15:02:53.487774 : Running LSI (1 of 2) on Top Features, 0.058 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-02-26 15:02:53.500573 : Creating Partial Matrix, 0.058 mins elapsed.\n",
      "\n",
      "2025-02-26 15:03:02.760445 : Computing LSI, 0.212 mins elapsed.\n",
      "\n",
      "2025-02-26 15:03:27.511592 : Identifying Clusters, 0.625 mins elapsed.\n",
      "\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "2025-02-26 15:03:31.249812 : Identified 6 Clusters, 0.687 mins elapsed.\n",
      "\n",
      "2025-02-26 15:03:31.267457 : Saving LSI Iteration, 0.687 mins elapsed.\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************************************************************\n",
      "2025-02-26 15:03:42.850188 : ERROR Found in .saveIteration for  \n",
      "LogFile = ArchRLogs/ArchR-addIterativeLSI-2a915759d0365e-Date-2025-02-26_Time-15-02-49.477491.log\n",
      "\n",
      "<simpleError in g$grobs[[legend]]: no such index at level 2\n",
      ">\n",
      "\n",
      "************************************************************\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-26 15:03:42.85221 : Creating Cluster Matrix on the total Group Features, 0.88 mins elapsed.\n",
      "\n",
      "2025-02-26 15:03:59.316927 : Computing Variable Features, 1.155 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-26 15:03:59.467777 : Running LSI (2 of 2) on Variable Features, 1.157 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-02-26 15:03:59.481074 : Creating Partial Matrix, 1.158 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:10.491966 : Computing LSI, 1.341 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:39.924722 : Finished Running IterativeLSI, 1.832 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addIterativeLSI(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    useMatrix = \"TileMatrix\", \n",
    "    name = \"IterativeLSI\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2c1d334b-7fd7-425a-bb8d-7be4cbedd3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.ArchR = addPeakSet(obj.ArchR,\n",
    "                       peakSet = reduce(pairs.E2G))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e25c14db-eb4e-4b94-8862-eb34246a365c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeakMatrix-2a91571392b5fe-Date-2025-02-26_Time-15-04-43.86654.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 15:04:44.112888 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:44.916435 : Adding GM12878 to PeakMatrix for Chr (1 of 23)!, 0.012 mins elapsed.\n",
      "\n",
      "2025-02-26 15:04:54.316475 : Adding GM12878 to PeakMatrix for Chr (2 of 23)!, 0.169 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:01.402203 : Adding GM12878 to PeakMatrix for Chr (3 of 23)!, 0.287 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:08.317939 : Adding GM12878 to PeakMatrix for Chr (4 of 23)!, 0.403 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:12.910614 : Adding GM12878 to PeakMatrix for Chr (5 of 23)!, 0.479 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:18.254376 : Adding GM12878 to PeakMatrix for Chr (6 of 23)!, 0.568 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:23.521483 : Adding GM12878 to PeakMatrix for Chr (7 of 23)!, 0.656 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:28.261906 : Adding GM12878 to PeakMatrix for Chr (8 of 23)!, 0.735 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:33.146007 : Adding GM12878 to PeakMatrix for Chr (9 of 23)!, 0.816 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:37.855247 : Adding GM12878 to PeakMatrix for Chr (10 of 23)!, 0.895 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:44.043865 : Adding GM12878 to PeakMatrix for Chr (11 of 23)!, 0.998 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:51.002447 : Adding GM12878 to PeakMatrix for Chr (12 of 23)!, 1.114 mins elapsed.\n",
      "\n",
      "2025-02-26 15:05:57.368225 : Adding GM12878 to PeakMatrix for Chr (13 of 23)!, 1.22 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:03.130856 : Adding GM12878 to PeakMatrix for Chr (14 of 23)!, 1.316 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:09.237327 : Adding GM12878 to PeakMatrix for Chr (15 of 23)!, 1.418 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:15.678652 : Adding GM12878 to PeakMatrix for Chr (16 of 23)!, 1.525 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:21.547006 : Adding GM12878 to PeakMatrix for Chr (17 of 23)!, 1.623 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:26.856956 : Adding GM12878 to PeakMatrix for Chr (18 of 23)!, 1.712 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:30.935669 : Adding GM12878 to PeakMatrix for Chr (19 of 23)!, 1.779 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:37.194748 : Adding GM12878 to PeakMatrix for Chr (20 of 23)!, 1.884 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:41.571336 : Adding GM12878 to PeakMatrix for Chr (21 of 23)!, 1.957 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:45.702896 : Adding GM12878 to PeakMatrix for Chr (22 of 23)!, 2.026 mins elapsed.\n",
      "\n",
      "2025-02-26 15:06:49.98816 : Adding GM12878 to PeakMatrix for Chr (23 of 23)!, 2.097 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeakMatrix-2a91571392b5fe-Date-2025-02-26_Time-15-04-43.86654.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addPeakMatrix(obj.ArchR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "aa6d686e-4205-4eda-8d76-ed4d9e3e30e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA.Xu = ArchR:::.importFM(featureMatrix = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_230720/ArrayExpress/cellranger_res/EMTAB11264_01/outs/filtered_feature_bc_matrix.h5\",\n",
    "                             name = \"K562_Xu\",\n",
    "                             featureType = \"Gene Expression\")\n",
    "rowdata.Xu = rowData(seRNA.Xu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d843eb16-9fed-4146-b76e-0fb053f1557f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene.intersect = intersect(rownames(obj.seurat@assays$RNA$counts),\n",
    "                           rowdata.Xu$name)\n",
    "rowdata.intersect = rowdata.Xu[gene.intersect,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fdf69211-b595-4599-ac8d-1c832875c062",
   "metadata": {},
   "outputs": [],
   "source": [
    "rr = GRanges(paste0(rowdata.intersect$interval))\n",
    "mcols(rr) <- rowdata.intersect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1e8f8d64-f6be-43c2-8e76-ea1cf4ad0453",
   "metadata": {},
   "outputs": [],
   "source": [
    "count.rna = obj.seurat@assays$RNA$counts[gene.intersect,]\n",
    "colnames(count.rna) = gsub(\"GM12878_\",\"GM12878#\",colnames(count.rna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "174a97a2-ab68-478a-9085-8e8fbc3bac29",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA <- SummarizedExperiment(assays = SimpleList(counts = count.rna), \n",
    "                              rowRanges = rr)\n",
    "seRNA = sort.GenomicRanges(sortSeqlevels(seRNA), \n",
    "                           ignore.strand = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4656d98c-a9cc-486a-afaa-fe59ca6ef055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 61214 3050 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61214): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(3050): GM12878#AAACAGCCACTAAGAA-1 GM12878#AAACAGCCATAATCAC-1\n",
       "  ... GM12878#TTTGTTGGTCACAAAT-1 GM12878#TTTGTTGGTTGGTGAC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "810220a5-fb60-41dc-8c29-f53edc2a206c",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "337ea70a-27ee-4ade-aa21-1c6367fd578a",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "740f849e-cecd-4293-a9e3-d12ae65d3c5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biostrings’:\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seRNA.rename = map_gene_names(seRNA,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "65cb2097-9075-4ee4-9e54-dd288dc50736",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addGeneExpressionMatrix-2a9157af66a61-Date-2025-02-26_Time-15-07-43.791545.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Overlap w/ scATAC = 1\n",
      "\n",
      "2025-02-26 15:07:44.153276 : \n",
      "\n",
      "Overlap Per Sample w/ scATAC : GM12878=3032\n",
      "\n",
      "2025-02-26 15:07:44.162591 : \n",
      "\n",
      "2025-02-26 15:07:45.122836 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-02-26 15:07:45.71278 : Adding GM12878 to GeneExpressionMatrix for Chr (1 of 23)!, 0.01 mins elapsed.\n",
      "\n",
      "2025-02-26 15:07:51.564012 : Adding GM12878 to GeneExpressionMatrix for Chr (2 of 23)!, 0.107 mins elapsed.\n",
      "\n",
      "2025-02-26 15:07:55.451348 : Adding GM12878 to GeneExpressionMatrix for Chr (3 of 23)!, 0.172 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:01.19988 : Adding GM12878 to GeneExpressionMatrix for Chr (4 of 23)!, 0.268 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:06.11187 : Adding GM12878 to GeneExpressionMatrix for Chr (5 of 23)!, 0.35 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:10.343095 : Adding GM12878 to GeneExpressionMatrix for Chr (6 of 23)!, 0.42 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:15.942293 : Adding GM12878 to GeneExpressionMatrix for Chr (7 of 23)!, 0.514 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:20.100099 : Adding GM12878 to GeneExpressionMatrix for Chr (8 of 23)!, 0.583 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:24.631925 : Adding GM12878 to GeneExpressionMatrix for Chr (9 of 23)!, 0.658 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:32.331074 : Adding GM12878 to GeneExpressionMatrix for Chr (10 of 23)!, 0.787 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:37.481727 : Adding GM12878 to GeneExpressionMatrix for Chr (11 of 23)!, 0.873 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:42.526609 : Adding GM12878 to GeneExpressionMatrix for Chr (12 of 23)!, 0.957 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:50.109278 : Adding GM12878 to GeneExpressionMatrix for Chr (13 of 23)!, 1.083 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:59.420269 : Adding GM12878 to GeneExpressionMatrix for Chr (14 of 23)!, 1.238 mins elapsed.\n",
      "\n",
      "2025-02-26 15:09:04.801277 : Adding GM12878 to GeneExpressionMatrix for Chr (15 of 23)!, 1.328 mins elapsed.\n",
      "\n",
      "2025-02-26 15:09:12.357201 : Adding GM12878 to GeneExpressionMatrix for Chr (16 of 23)!, 1.454 mins elapsed.\n",
      "\n",
      "2025-02-26 15:09:17.59419 : Adding GM12878 to GeneExpressionMatrix for Chr (17 of 23)!, 1.541 mins elapsed.\n",
      "\n",
      "2025-02-26 15:09:23.052462 : Adding GM12878 to GeneExpressionMatrix for Chr (18 of 23)!, 1.632 mins elapsed.\n",
      "\n",
      "2025-02-26 15:09:31.249243 : Adding GM12878 to GeneExpressionMatrix for Chr (19 of 23)!, 1.769 mins elapsed.\n",
      "\n",
      "2025-02-26 15:09:36.937898 : Adding GM12878 to GeneExpressionMatrix for Chr (20 of 23)!, 1.864 mins elapsed.\n",
      "\n",
      "2025-02-26 15:09:42.593445 : Adding GM12878 to GeneExpressionMatrix for Chr (21 of 23)!, 1.958 mins elapsed.\n",
      "\n",
      "2025-02-26 15:09:50.0072 : Adding GM12878 to GeneExpressionMatrix for Chr (22 of 23)!, 2.081 mins elapsed.\n",
      "\n",
      "2025-02-26 15:09:56.424115 : Adding GM12878 to GeneExpressionMatrix for Chr (23 of 23)!, 2.188 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addGeneExpressionMatrix-2a9157af66a61-Date-2025-02-26_Time-15-07-43.791545.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addGeneExpressionMatrix(obj.ArchR,\n",
    "                                    seRNA = seRNA.rename,\n",
    "                                    strictMatch = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4318703-a143-40c0-b159-274d05033bbc",
   "metadata": {},
   "source": [
    "Run ArchR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "47c8fd5f-19ec-46e6-af87-57502c1e1ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeak2GeneLinks-2a9157196e4fff-Date-2025-02-26_Time-15-10-04.278177.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 15:10:04.64529 : Getting Available Matrices, 0.006 mins elapsed.\n",
      "\n",
      "No predictionScore found. Continuing without predictionScore!\n",
      "\n",
      "2025-02-26 15:10:05.021225 : Filtered Low Prediction Score Cells (0 of 3032, 0), 0 mins elapsed.\n",
      "\n",
      "Filtering 1 dims correlated > 0.75 to log10(depth + 1)\n",
      "\n",
      "2025-02-26 15:10:05.119867 : Computing KNN, 0.002 mins elapsed.\n",
      "\n",
      "2025-02-26 15:10:05.389711 : Identifying Non-Overlapping KNN pairs, 0.006 mins elapsed.\n",
      "\n",
      "2025-02-26 15:10:07.170779 : Identified 499 Groupings!, 0.036 mins elapsed.\n",
      "\n",
      "2025-02-26 15:10:07.208972 : Getting Group RNA Matrix, 0.037 mins elapsed.\n",
      "\n",
      "2025-02-26 15:10:27.417978 : Getting Group ATAC Matrix, 0.374 mins elapsed.\n",
      "\n",
      "2025-02-26 15:10:52.171213 : Normalizing Group Matrices, 0.786 mins elapsed.\n",
      "\n",
      "2025-02-26 15:10:58.616725 : Finding Peak Gene Pairings, 0.894 mins elapsed.\n",
      "\n",
      "2025-02-26 15:10:59.415321 : Computing Correlations, 0.907 mins elapsed.\n",
      "\n",
      "2025-02-26 15:11:14.792764 : Completed Peak2Gene Correlations!, 1.163 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeak2GeneLinks-2a9157196e4fff-Date-2025-02-26_Time-15-10-04.278177.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "obj.ArchR = addPeak2GeneLinks(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    reducedDims = \"IterativeLSI\",\n",
    "    useMatrix = \"GeneExpressionMatrix\",\n",
    "    maxDist = maxDist,\n",
    "    threads = n.cores\n",
    ")\n",
    "end_time <- Sys.time()\n",
    "execution_time <- end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "128d0065-5f22-495b-b00d-743fc9843337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 1.175743 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "execution_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "36bd2df2-6b32-4b81-bd4f-39ef8913b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g = metadata(obj.ArchR@peakSet)$Peak2GeneLinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e7730413-8fd4-4952-bb13-0df7a3b15b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata(p2g)[[1]]$PeakName = paste(seqnames(metadata(p2g)[[1]]),\n",
    "                                    ranges(metadata(p2g)[[1]]),sep=\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "aeea3810-6735-4aff-9bd7-1697bef979d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g$PairName = paste(metadata(p2g)[[1]][p2g$idxATAC]$PeakName,\n",
    "                     metadata(p2g)[[2]][p2g$idxRNA]$name,\n",
    "                     sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "87f975ec-6ba4-4696-ad6c-2af65da7aa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 3417150 rows and 7 columns\n",
       "                                idxATAC    idxRNA Correlation       FDR\n",
       "                              <integer> <integer>   <numeric> <numeric>\n",
       "chr1-180700-181061_FAM138A            1         1         NaN       NaN\n",
       "chr1-181282-181914_FAM138A            2         1         NaN       NaN\n",
       "chr1-267886-268118_FAM138A            3         1         NaN       NaN\n",
       "chr1-586095-586295_FAM138A            4         1         NaN       NaN\n",
       "chr1-778328-779254_FAM138A            5         1         NaN       NaN\n",
       "...                                 ...       ...         ...       ...\n",
       "chrX-155981990-155982304_IL9R    155266     20493  -0.0474404  0.376047\n",
       "chrX-155997270-155998098_IL9R    155267     20493  -0.0829243  0.103852\n",
       "chrX-156000622-156000826_IL9R    155268     20493   0.0152770  0.793101\n",
       "chrX-156004204-156004495_IL9R    155269     20493  -0.0259362  0.645091\n",
       "chrX-156030527-156030695_IL9R    155270     20493   0.0245413  0.664273\n",
       "                               VarQATAC   VarQRNA               PairName\n",
       "                              <numeric> <numeric>            <character>\n",
       "chr1-180700-181061_FAM138A     0.222226 0.0639243 chr1-180700-181061_F..\n",
       "chr1-181282-181914_FAM138A     0.615637 0.0639243 chr1-181282-181914_F..\n",
       "chr1-267886-268118_FAM138A     0.447234 0.0639243 chr1-267886-268118_F..\n",
       "chr1-586095-586295_FAM138A     0.213338 0.0639243 chr1-586095-586295_F..\n",
       "chr1-778328-779254_FAM138A     0.988446 0.0639243 chr1-778328-779254_F..\n",
       "...                                 ...       ...                    ...\n",
       "chrX-155981990-155982304_IL9R 0.0518516  0.390377 chrX-155981990-15598..\n",
       "chrX-155997270-155998098_IL9R 0.7083918  0.390377 chrX-155997270-15599..\n",
       "chrX-156000622-156000826_IL9R 0.0232241  0.390377 chrX-156000622-15600..\n",
       "chrX-156004204-156004495_IL9R 0.0555033  0.390377 chrX-156004204-15600..\n",
       "chrX-156030527-156030695_IL9R 0.0943518  0.390377 chrX-156030527-15603.."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(p2g) = p2g$PairName\n",
    "p2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c7209a32-3904-4734-93af-9c22049f652a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = pairs.E2G[pairs.E2G$PairName %in% p2g$PairName]\n",
    "mcols(pairs.E2G.res)[,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")] = p2g[pairs.E2G.res$PairName,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1ab4ed70-ce9e-4857-bd1b-63a81d27a401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2628622 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100028533-100029101      * |         AGL\n",
       "        [2]     chr1 100028533-100029101      * |      CDC14A\n",
       "        [3]     chr1 100028533-100029101      * |         DBT\n",
       "        [4]     chr1 100028533-100029101      * |        DPH5\n",
       "        [5]     chr1 100028533-100029101      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2628618]     chrX     9995654-9996249      * |       TBL1X\n",
       "  [2628619]     chrX     9995654-9996249      * |        WWC3\n",
       "  [2628620]     chrX     9997430-9997980      * |       CLCN4\n",
       "  [2628621]     chrX     9997430-9997980      * |       TBL1X\n",
       "  [2628622]     chrX     9997430-9997980      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-100028533-10002.. chr1-100028533-10002..   0.2390602\n",
       "        [2] chr1-100028533-10002.. chr1-100028533-10002..   0.2506110\n",
       "        [3] chr1-100028533-10002.. chr1-100028533-10002..  -0.0063773\n",
       "        [4] chr1-100028533-10002.. chr1-100028533-10002..   0.1091928\n",
       "        [5] chr1-100028533-10002.. chr1-100028533-10002..   0.2977003\n",
       "        ...                    ...                    ...         ...\n",
       "  [2628618]   chrX-9995654-9996249 chrX-9995654-9996249..   -0.122068\n",
       "  [2628619]   chrX-9995654-9996249 chrX-9995654-9996249..   -0.114240\n",
       "  [2628620]   chrX-9997430-9997980 chrX-9997430-9997980..    0.104033\n",
       "  [2628621]   chrX-9997430-9997980 chrX-9997430-9997980..   -0.373740\n",
       "  [2628622]   chrX-9997430-9997980 chrX-9997430-9997980..   -0.109514\n",
       "                    FDR  VarQATAC   VarQRNA\n",
       "              <numeric> <numeric> <numeric>\n",
       "        [1] 3.12426e-07  0.250042  0.798029\n",
       "        [2] 7.28251e-08  0.250042  0.928951\n",
       "        [3] 9.15741e-01  0.250042  0.679256\n",
       "        [4] 2.82710e-02  0.250042  0.853413\n",
       "        [5] 8.38844e-11  0.250042  0.962768\n",
       "        ...         ...       ...       ...\n",
       "  [2628618] 1.33149e-02  0.705513  0.950471\n",
       "  [2628619] 2.12402e-02  0.705513  0.811106\n",
       "  [2628620] 3.74134e-02  0.831764  0.378178\n",
       "  [2628621] 7.04057e-17  0.831764  0.950471\n",
       "  [2628622] 2.77710e-02  0.831764  0.811106\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46485351-50d7-461a-bf09-d795d5bdab84",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6d424e21-6a41-4422-b76c-6f1c0bda2222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2628622 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100028533-100029101      * |         AGL\n",
       "        [2]     chr1 100028533-100029101      * |      CDC14A\n",
       "        [3]     chr1 100028533-100029101      * |         DBT\n",
       "        [4]     chr1 100028533-100029101      * |        DPH5\n",
       "        [5]     chr1 100028533-100029101      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2628618]     chrX     9995654-9996249      * |       TBL1X\n",
       "  [2628619]     chrX     9995654-9996249      * |        WWC3\n",
       "  [2628620]     chrX     9997430-9997980      * |       CLCN4\n",
       "  [2628621]     chrX     9997430-9997980      * |       TBL1X\n",
       "  [2628622]     chrX     9997430-9997980      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-100028533-10002.. chr1-100028533-10002..   0.2390602\n",
       "        [2] chr1-100028533-10002.. chr1-100028533-10002..   0.2506110\n",
       "        [3] chr1-100028533-10002.. chr1-100028533-10002..  -0.0063773\n",
       "        [4] chr1-100028533-10002.. chr1-100028533-10002..   0.1091928\n",
       "        [5] chr1-100028533-10002.. chr1-100028533-10002..   0.2977003\n",
       "        ...                    ...                    ...         ...\n",
       "  [2628618]   chrX-9995654-9996249 chrX-9995654-9996249..   -0.122068\n",
       "  [2628619]   chrX-9995654-9996249 chrX-9995654-9996249..   -0.114240\n",
       "  [2628620]   chrX-9997430-9997980 chrX-9997430-9997980..    0.104033\n",
       "  [2628621]   chrX-9997430-9997980 chrX-9997430-9997980..   -0.373740\n",
       "  [2628622]   chrX-9997430-9997980 chrX-9997430-9997980..   -0.109514\n",
       "                    FDR  VarQATAC   VarQRNA\n",
       "              <numeric> <numeric> <numeric>\n",
       "        [1] 3.12426e-07  0.250042  0.798029\n",
       "        [2] 7.28251e-08  0.250042  0.928951\n",
       "        [3] 9.15741e-01  0.250042  0.679256\n",
       "        [4] 2.82710e-02  0.250042  0.853413\n",
       "        [5] 8.38844e-11  0.250042  0.962768\n",
       "        ...         ...       ...       ...\n",
       "  [2628618] 1.33149e-02  0.705513  0.950471\n",
       "  [2628619] 2.12402e-02  0.705513  0.811106\n",
       "  [2628620] 3.74134e-02  0.831764  0.378178\n",
       "  [2628621] 7.04057e-17  0.831764  0.950471\n",
       "  [2628622] 2.77710e-02  0.831764  0.811106\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2628622 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>Correlation</th><th scope=col>FDR</th><th scope=col>VarQATAC</th><th scope=col>VarQRNA</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>AGL     </td><td>K562</td><td> 0.239060154</td><td>3.124260e-07</td><td>0.25004186</td><td>0.7980286</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>CDC14A  </td><td>K562</td><td> 0.250611041</td><td>7.282513e-08</td><td>0.25004186</td><td>0.9289513</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>DBT     </td><td>K562</td><td>-0.006377297</td><td>9.157406e-01</td><td>0.25004186</td><td>0.6792563</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>DPH5    </td><td>K562</td><td> 0.109192839</td><td>2.827099e-02</td><td>0.25004186</td><td>0.8534134</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>EXTL2   </td><td>K562</td><td> 0.297700272</td><td>8.388435e-11</td><td>0.25004186</td><td>0.9627678</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>FRRS1   </td><td>K562</td><td> 0.052411707</td><td>3.238124e-01</td><td>0.25004186</td><td>0.8201337</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>MFSD14A </td><td>K562</td><td> 0.259727933</td><td>2.182599e-08</td><td>0.25004186</td><td>0.4436149</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>RTCA    </td><td>K562</td><td>-0.093778868</td><td>6.298193e-02</td><td>0.25004186</td><td>0.6899917</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>RTCA-AS1</td><td>K562</td><td>-0.094458669</td><td>6.092956e-02</td><td>0.25004186</td><td>0.6208949</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>SASS6   </td><td>K562</td><td>-0.247961866</td><td>1.023916e-07</td><td>0.25004186</td><td>0.9006490</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>SLC30A7 </td><td>K562</td><td>-0.073527423</td><td>1.536087e-01</td><td>0.25004186</td><td>0.8160347</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>SLC35A3 </td><td>K562</td><td>-0.180999641</td><td>1.518199e-04</td><td>0.25004186</td><td>0.6425609</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>TRMT13  </td><td>K562</td><td>-0.121284659</td><td>1.397065e-02</td><td>0.25004186</td><td>0.7007271</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028533</td><td>100029101</td><td>VCAM1   </td><td>K562</td><td> 0.275252877</td><td>2.502730e-09</td><td>0.25004186</td><td>0.6057678</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>AGL     </td><td>K562</td><td> 0.096239802</td><td>5.582017e-02</td><td>0.02579378</td><td>0.7980286</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>CDC14A  </td><td>K562</td><td>-0.120592021</td><td>1.457354e-02</td><td>0.02579378</td><td>0.9289513</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>DBT     </td><td>K562</td><td>-0.017272916</td><td>7.652039e-01</td><td>0.02579378</td><td>0.6792563</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>DPH5    </td><td>K562</td><td>-0.011982116</td><td>8.388946e-01</td><td>0.02579378</td><td>0.8534134</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>EXTL2   </td><td>K562</td><td>-0.012492951</td><td>8.317856e-01</td><td>0.02579378</td><td>0.9627678</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>FRRS1   </td><td>K562</td><td> 0.048695609</td><td>3.624027e-01</td><td>0.02579378</td><td>0.8201337</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>MFSD14A </td><td>K562</td><td> 0.060060088</td><td>2.524295e-01</td><td>0.02579378</td><td>0.4436149</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>RTCA    </td><td>K562</td><td> 0.004180552</td><td>9.451665e-01</td><td>0.02579378</td><td>0.6899917</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>RTCA-AS1</td><td>K562</td><td> 0.143387096</td><td>3.219733e-03</td><td>0.02579378</td><td>0.6208949</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>SASS6   </td><td>K562</td><td> 0.031613771</td><td>5.686221e-01</td><td>0.02579378</td><td>0.9006490</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>SLC30A7 </td><td>K562</td><td>-0.072689798</td><td>1.587742e-01</td><td>0.02579378</td><td>0.8160347</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>SLC35A3 </td><td>K562</td><td>-0.119545915</td><td>1.552624e-02</td><td>0.02579378</td><td>0.6425609</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>TRMT13  </td><td>K562</td><td> 0.097198135</td><td>5.321544e-02</td><td>0.02579378</td><td>0.7007271</td></tr>\n",
       "\t<tr><td>chr1</td><td>100029136</td><td>100029435</td><td>VCAM1   </td><td>K562</td><td>-0.043401535</td><td>4.216773e-01</td><td>0.02579378</td><td>0.6057678</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034578</td><td>100034758</td><td>AGL     </td><td>K562</td><td>-0.177137591</td><td>2.147535e-04</td><td>0.47476654</td><td>0.7980286</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034578</td><td>100034758</td><td>CDC14A  </td><td>K562</td><td>-0.242069051</td><td>2.153961e-07</td><td>0.47476654</td><td>0.9289513</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>ASMT     </td><td>K562</td><td> 0.107038452</td><td>3.182215e-02</td><td>0.82948412</td><td>0.3051774</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>ASMTL    </td><td>K562</td><td> 0.121607580</td><td>1.369742e-02</td><td>0.82948412</td><td>0.8825453</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>CSF2RA   </td><td>K562</td><td> 0.001336011</td><td>9.827104e-01</td><td>0.82948412</td><td>0.3498268</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>GTPBP6   </td><td>K562</td><td> 0.350798315</td><td>7.352820e-15</td><td>0.82948412</td><td>0.6354853</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>IL3RA    </td><td>K562</td><td> 0.286757362</td><td>4.568795e-10</td><td>0.82948412</td><td>0.9442737</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>LINC00106</td><td>K562</td><td>-0.332056590</td><td>2.476046e-13</td><td>0.82948412</td><td>0.6546626</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>P2RY8    </td><td>K562</td><td>-0.351017308</td><td>7.045703e-15</td><td>0.82948412</td><td>0.9853609</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>PLCXD1   </td><td>K562</td><td> 0.511510173</td><td>4.779843e-33</td><td>0.82948412</td><td>0.7444005</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>PPP2R3B  </td><td>K562</td><td> 0.287686214</td><td>3.968925e-10</td><td>0.82948412</td><td>0.5513590</td></tr>\n",
       "\t<tr><td>chrX</td><td> 997541</td><td> 998270</td><td>SLC25A6  </td><td>K562</td><td> 0.303317090</td><td>3.412362e-11</td><td>0.82948412</td><td>0.8978676</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988858</td><td>9989079</td><td>CLCN4    </td><td>K562</td><td> 0.317437877</td><td>3.249789e-12</td><td>0.09842854</td><td>0.3781779</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988858</td><td>9989079</td><td>TBL1X    </td><td>K562</td><td>-0.052752862</td><td>3.203807e-01</td><td>0.09842854</td><td>0.9504709</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988858</td><td>9989079</td><td>WWC3     </td><td>K562</td><td> 0.154254665</td><td>1.433381e-03</td><td>0.09842854</td><td>0.8111062</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>AKAP17A  </td><td>K562</td><td> 0.374751807</td><td>5.685914e-17</td><td>0.87065756</td><td>0.9154833</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>ASMT     </td><td>K562</td><td>-0.022216379</td><td>6.963897e-01</td><td>0.87065756</td><td>0.3051774</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>ASMTL    </td><td>K562</td><td>-0.524880732</td><td>4.695757e-35</td><td>0.87065756</td><td>0.8825453</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>CSF2RA   </td><td>K562</td><td>-0.217192594</td><td>3.994496e-06</td><td>0.87065756</td><td>0.3498268</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>GTPBP6   </td><td>K562</td><td>-0.056599936</td><td>2.833746e-01</td><td>0.87065756</td><td>0.6354853</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>IL3RA    </td><td>K562</td><td>-0.568482053</td><td>2.836953e-42</td><td>0.87065756</td><td>0.9442737</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>LINC00106</td><td>K562</td><td>-0.350804221</td><td>7.344386e-15</td><td>0.87065756</td><td>0.6546626</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>P2RY8    </td><td>K562</td><td> 0.546970181</td><td>1.406170e-38</td><td>0.87065756</td><td>0.9853609</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>PLCXD1   </td><td>K562</td><td> 0.167609106</td><td>4.890576e-04</td><td>0.87065756</td><td>0.7444005</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>PPP2R3B  </td><td>K562</td><td>-0.214487248</td><td>5.372700e-06</td><td>0.87065756</td><td>0.5513590</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999293</td><td> 999875</td><td>SLC25A6  </td><td>K562</td><td> 0.402618679</td><td>1.149283e-19</td><td>0.87065756</td><td>0.8978676</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995654</td><td>9996249</td><td>CLCN4    </td><td>K562</td><td>-0.017752707</td><td>7.584762e-01</td><td>0.70551298</td><td>0.3781779</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995654</td><td>9996249</td><td>TBL1X    </td><td>K562</td><td>-0.122068385</td><td>1.331493e-02</td><td>0.70551298</td><td>0.9504709</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995654</td><td>9996249</td><td>WWC3     </td><td>K562</td><td>-0.114239506</td><td>2.124022e-02</td><td>0.70551298</td><td>0.8111062</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997430</td><td>9997980</td><td>CLCN4    </td><td>K562</td><td> 0.104033205</td><td>3.741343e-02</td><td>0.83176402</td><td>0.3781779</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997430</td><td>9997980</td><td>TBL1X    </td><td>K562</td><td>-0.373740032</td><td>7.040567e-17</td><td>0.83176402</td><td>0.9504709</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997430</td><td>9997980</td><td>WWC3     </td><td>K562</td><td>-0.109513741</td><td>2.777099e-02</td><td>0.83176402</td><td>0.8111062</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2628622 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & Correlation & FDR & VarQATAC & VarQRNA\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100028533 & 100029101 & AGL      & K562 &  0.239060154 & 3.124260e-07 & 0.25004186 & 0.7980286\\\\\n",
       "\t chr1 & 100028533 & 100029101 & CDC14A   & K562 &  0.250611041 & 7.282513e-08 & 0.25004186 & 0.9289513\\\\\n",
       "\t chr1 & 100028533 & 100029101 & DBT      & K562 & -0.006377297 & 9.157406e-01 & 0.25004186 & 0.6792563\\\\\n",
       "\t chr1 & 100028533 & 100029101 & DPH5     & K562 &  0.109192839 & 2.827099e-02 & 0.25004186 & 0.8534134\\\\\n",
       "\t chr1 & 100028533 & 100029101 & EXTL2    & K562 &  0.297700272 & 8.388435e-11 & 0.25004186 & 0.9627678\\\\\n",
       "\t chr1 & 100028533 & 100029101 & FRRS1    & K562 &  0.052411707 & 3.238124e-01 & 0.25004186 & 0.8201337\\\\\n",
       "\t chr1 & 100028533 & 100029101 & MFSD14A  & K562 &  0.259727933 & 2.182599e-08 & 0.25004186 & 0.4436149\\\\\n",
       "\t chr1 & 100028533 & 100029101 & RTCA     & K562 & -0.093778868 & 6.298193e-02 & 0.25004186 & 0.6899917\\\\\n",
       "\t chr1 & 100028533 & 100029101 & RTCA-AS1 & K562 & -0.094458669 & 6.092956e-02 & 0.25004186 & 0.6208949\\\\\n",
       "\t chr1 & 100028533 & 100029101 & SASS6    & K562 & -0.247961866 & 1.023916e-07 & 0.25004186 & 0.9006490\\\\\n",
       "\t chr1 & 100028533 & 100029101 & SLC30A7  & K562 & -0.073527423 & 1.536087e-01 & 0.25004186 & 0.8160347\\\\\n",
       "\t chr1 & 100028533 & 100029101 & SLC35A3  & K562 & -0.180999641 & 1.518199e-04 & 0.25004186 & 0.6425609\\\\\n",
       "\t chr1 & 100028533 & 100029101 & TRMT13   & K562 & -0.121284659 & 1.397065e-02 & 0.25004186 & 0.7007271\\\\\n",
       "\t chr1 & 100028533 & 100029101 & VCAM1    & K562 &  0.275252877 & 2.502730e-09 & 0.25004186 & 0.6057678\\\\\n",
       "\t chr1 & 100029136 & 100029435 & AGL      & K562 &  0.096239802 & 5.582017e-02 & 0.02579378 & 0.7980286\\\\\n",
       "\t chr1 & 100029136 & 100029435 & CDC14A   & K562 & -0.120592021 & 1.457354e-02 & 0.02579378 & 0.9289513\\\\\n",
       "\t chr1 & 100029136 & 100029435 & DBT      & K562 & -0.017272916 & 7.652039e-01 & 0.02579378 & 0.6792563\\\\\n",
       "\t chr1 & 100029136 & 100029435 & DPH5     & K562 & -0.011982116 & 8.388946e-01 & 0.02579378 & 0.8534134\\\\\n",
       "\t chr1 & 100029136 & 100029435 & EXTL2    & K562 & -0.012492951 & 8.317856e-01 & 0.02579378 & 0.9627678\\\\\n",
       "\t chr1 & 100029136 & 100029435 & FRRS1    & K562 &  0.048695609 & 3.624027e-01 & 0.02579378 & 0.8201337\\\\\n",
       "\t chr1 & 100029136 & 100029435 & MFSD14A  & K562 &  0.060060088 & 2.524295e-01 & 0.02579378 & 0.4436149\\\\\n",
       "\t chr1 & 100029136 & 100029435 & RTCA     & K562 &  0.004180552 & 9.451665e-01 & 0.02579378 & 0.6899917\\\\\n",
       "\t chr1 & 100029136 & 100029435 & RTCA-AS1 & K562 &  0.143387096 & 3.219733e-03 & 0.02579378 & 0.6208949\\\\\n",
       "\t chr1 & 100029136 & 100029435 & SASS6    & K562 &  0.031613771 & 5.686221e-01 & 0.02579378 & 0.9006490\\\\\n",
       "\t chr1 & 100029136 & 100029435 & SLC30A7  & K562 & -0.072689798 & 1.587742e-01 & 0.02579378 & 0.8160347\\\\\n",
       "\t chr1 & 100029136 & 100029435 & SLC35A3  & K562 & -0.119545915 & 1.552624e-02 & 0.02579378 & 0.6425609\\\\\n",
       "\t chr1 & 100029136 & 100029435 & TRMT13   & K562 &  0.097198135 & 5.321544e-02 & 0.02579378 & 0.7007271\\\\\n",
       "\t chr1 & 100029136 & 100029435 & VCAM1    & K562 & -0.043401535 & 4.216773e-01 & 0.02579378 & 0.6057678\\\\\n",
       "\t chr1 & 100034578 & 100034758 & AGL      & K562 & -0.177137591 & 2.147535e-04 & 0.47476654 & 0.7980286\\\\\n",
       "\t chr1 & 100034578 & 100034758 & CDC14A   & K562 & -0.242069051 & 2.153961e-07 & 0.47476654 & 0.9289513\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  997541 &  998270 & ASMT      & K562 &  0.107038452 & 3.182215e-02 & 0.82948412 & 0.3051774\\\\\n",
       "\t chrX &  997541 &  998270 & ASMTL     & K562 &  0.121607580 & 1.369742e-02 & 0.82948412 & 0.8825453\\\\\n",
       "\t chrX &  997541 &  998270 & CSF2RA    & K562 &  0.001336011 & 9.827104e-01 & 0.82948412 & 0.3498268\\\\\n",
       "\t chrX &  997541 &  998270 & GTPBP6    & K562 &  0.350798315 & 7.352820e-15 & 0.82948412 & 0.6354853\\\\\n",
       "\t chrX &  997541 &  998270 & IL3RA     & K562 &  0.286757362 & 4.568795e-10 & 0.82948412 & 0.9442737\\\\\n",
       "\t chrX &  997541 &  998270 & LINC00106 & K562 & -0.332056590 & 2.476046e-13 & 0.82948412 & 0.6546626\\\\\n",
       "\t chrX &  997541 &  998270 & P2RY8     & K562 & -0.351017308 & 7.045703e-15 & 0.82948412 & 0.9853609\\\\\n",
       "\t chrX &  997541 &  998270 & PLCXD1    & K562 &  0.511510173 & 4.779843e-33 & 0.82948412 & 0.7444005\\\\\n",
       "\t chrX &  997541 &  998270 & PPP2R3B   & K562 &  0.287686214 & 3.968925e-10 & 0.82948412 & 0.5513590\\\\\n",
       "\t chrX &  997541 &  998270 & SLC25A6   & K562 &  0.303317090 & 3.412362e-11 & 0.82948412 & 0.8978676\\\\\n",
       "\t chrX & 9988858 & 9989079 & CLCN4     & K562 &  0.317437877 & 3.249789e-12 & 0.09842854 & 0.3781779\\\\\n",
       "\t chrX & 9988858 & 9989079 & TBL1X     & K562 & -0.052752862 & 3.203807e-01 & 0.09842854 & 0.9504709\\\\\n",
       "\t chrX & 9988858 & 9989079 & WWC3      & K562 &  0.154254665 & 1.433381e-03 & 0.09842854 & 0.8111062\\\\\n",
       "\t chrX &  999293 &  999875 & AKAP17A   & K562 &  0.374751807 & 5.685914e-17 & 0.87065756 & 0.9154833\\\\\n",
       "\t chrX &  999293 &  999875 & ASMT      & K562 & -0.022216379 & 6.963897e-01 & 0.87065756 & 0.3051774\\\\\n",
       "\t chrX &  999293 &  999875 & ASMTL     & K562 & -0.524880732 & 4.695757e-35 & 0.87065756 & 0.8825453\\\\\n",
       "\t chrX &  999293 &  999875 & CSF2RA    & K562 & -0.217192594 & 3.994496e-06 & 0.87065756 & 0.3498268\\\\\n",
       "\t chrX &  999293 &  999875 & GTPBP6    & K562 & -0.056599936 & 2.833746e-01 & 0.87065756 & 0.6354853\\\\\n",
       "\t chrX &  999293 &  999875 & IL3RA     & K562 & -0.568482053 & 2.836953e-42 & 0.87065756 & 0.9442737\\\\\n",
       "\t chrX &  999293 &  999875 & LINC00106 & K562 & -0.350804221 & 7.344386e-15 & 0.87065756 & 0.6546626\\\\\n",
       "\t chrX &  999293 &  999875 & P2RY8     & K562 &  0.546970181 & 1.406170e-38 & 0.87065756 & 0.9853609\\\\\n",
       "\t chrX &  999293 &  999875 & PLCXD1    & K562 &  0.167609106 & 4.890576e-04 & 0.87065756 & 0.7444005\\\\\n",
       "\t chrX &  999293 &  999875 & PPP2R3B   & K562 & -0.214487248 & 5.372700e-06 & 0.87065756 & 0.5513590\\\\\n",
       "\t chrX &  999293 &  999875 & SLC25A6   & K562 &  0.402618679 & 1.149283e-19 & 0.87065756 & 0.8978676\\\\\n",
       "\t chrX & 9995654 & 9996249 & CLCN4     & K562 & -0.017752707 & 7.584762e-01 & 0.70551298 & 0.3781779\\\\\n",
       "\t chrX & 9995654 & 9996249 & TBL1X     & K562 & -0.122068385 & 1.331493e-02 & 0.70551298 & 0.9504709\\\\\n",
       "\t chrX & 9995654 & 9996249 & WWC3      & K562 & -0.114239506 & 2.124022e-02 & 0.70551298 & 0.8111062\\\\\n",
       "\t chrX & 9997430 & 9997980 & CLCN4     & K562 &  0.104033205 & 3.741343e-02 & 0.83176402 & 0.3781779\\\\\n",
       "\t chrX & 9997430 & 9997980 & TBL1X     & K562 & -0.373740032 & 7.040567e-17 & 0.83176402 & 0.9504709\\\\\n",
       "\t chrX & 9997430 & 9997980 & WWC3      & K562 & -0.109513741 & 2.777099e-02 & 0.83176402 & 0.8111062\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2628622 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | Correlation &lt;dbl&gt; | FDR &lt;dbl&gt; | VarQATAC &lt;dbl&gt; | VarQRNA &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100028533 | 100029101 | AGL      | K562 |  0.239060154 | 3.124260e-07 | 0.25004186 | 0.7980286 |\n",
       "| chr1 | 100028533 | 100029101 | CDC14A   | K562 |  0.250611041 | 7.282513e-08 | 0.25004186 | 0.9289513 |\n",
       "| chr1 | 100028533 | 100029101 | DBT      | K562 | -0.006377297 | 9.157406e-01 | 0.25004186 | 0.6792563 |\n",
       "| chr1 | 100028533 | 100029101 | DPH5     | K562 |  0.109192839 | 2.827099e-02 | 0.25004186 | 0.8534134 |\n",
       "| chr1 | 100028533 | 100029101 | EXTL2    | K562 |  0.297700272 | 8.388435e-11 | 0.25004186 | 0.9627678 |\n",
       "| chr1 | 100028533 | 100029101 | FRRS1    | K562 |  0.052411707 | 3.238124e-01 | 0.25004186 | 0.8201337 |\n",
       "| chr1 | 100028533 | 100029101 | MFSD14A  | K562 |  0.259727933 | 2.182599e-08 | 0.25004186 | 0.4436149 |\n",
       "| chr1 | 100028533 | 100029101 | RTCA     | K562 | -0.093778868 | 6.298193e-02 | 0.25004186 | 0.6899917 |\n",
       "| chr1 | 100028533 | 100029101 | RTCA-AS1 | K562 | -0.094458669 | 6.092956e-02 | 0.25004186 | 0.6208949 |\n",
       "| chr1 | 100028533 | 100029101 | SASS6    | K562 | -0.247961866 | 1.023916e-07 | 0.25004186 | 0.9006490 |\n",
       "| chr1 | 100028533 | 100029101 | SLC30A7  | K562 | -0.073527423 | 1.536087e-01 | 0.25004186 | 0.8160347 |\n",
       "| chr1 | 100028533 | 100029101 | SLC35A3  | K562 | -0.180999641 | 1.518199e-04 | 0.25004186 | 0.6425609 |\n",
       "| chr1 | 100028533 | 100029101 | TRMT13   | K562 | -0.121284659 | 1.397065e-02 | 0.25004186 | 0.7007271 |\n",
       "| chr1 | 100028533 | 100029101 | VCAM1    | K562 |  0.275252877 | 2.502730e-09 | 0.25004186 | 0.6057678 |\n",
       "| chr1 | 100029136 | 100029435 | AGL      | K562 |  0.096239802 | 5.582017e-02 | 0.02579378 | 0.7980286 |\n",
       "| chr1 | 100029136 | 100029435 | CDC14A   | K562 | -0.120592021 | 1.457354e-02 | 0.02579378 | 0.9289513 |\n",
       "| chr1 | 100029136 | 100029435 | DBT      | K562 | -0.017272916 | 7.652039e-01 | 0.02579378 | 0.6792563 |\n",
       "| chr1 | 100029136 | 100029435 | DPH5     | K562 | -0.011982116 | 8.388946e-01 | 0.02579378 | 0.8534134 |\n",
       "| chr1 | 100029136 | 100029435 | EXTL2    | K562 | -0.012492951 | 8.317856e-01 | 0.02579378 | 0.9627678 |\n",
       "| chr1 | 100029136 | 100029435 | FRRS1    | K562 |  0.048695609 | 3.624027e-01 | 0.02579378 | 0.8201337 |\n",
       "| chr1 | 100029136 | 100029435 | MFSD14A  | K562 |  0.060060088 | 2.524295e-01 | 0.02579378 | 0.4436149 |\n",
       "| chr1 | 100029136 | 100029435 | RTCA     | K562 |  0.004180552 | 9.451665e-01 | 0.02579378 | 0.6899917 |\n",
       "| chr1 | 100029136 | 100029435 | RTCA-AS1 | K562 |  0.143387096 | 3.219733e-03 | 0.02579378 | 0.6208949 |\n",
       "| chr1 | 100029136 | 100029435 | SASS6    | K562 |  0.031613771 | 5.686221e-01 | 0.02579378 | 0.9006490 |\n",
       "| chr1 | 100029136 | 100029435 | SLC30A7  | K562 | -0.072689798 | 1.587742e-01 | 0.02579378 | 0.8160347 |\n",
       "| chr1 | 100029136 | 100029435 | SLC35A3  | K562 | -0.119545915 | 1.552624e-02 | 0.02579378 | 0.6425609 |\n",
       "| chr1 | 100029136 | 100029435 | TRMT13   | K562 |  0.097198135 | 5.321544e-02 | 0.02579378 | 0.7007271 |\n",
       "| chr1 | 100029136 | 100029435 | VCAM1    | K562 | -0.043401535 | 4.216773e-01 | 0.02579378 | 0.6057678 |\n",
       "| chr1 | 100034578 | 100034758 | AGL      | K562 | -0.177137591 | 2.147535e-04 | 0.47476654 | 0.7980286 |\n",
       "| chr1 | 100034578 | 100034758 | CDC14A   | K562 | -0.242069051 | 2.153961e-07 | 0.47476654 | 0.9289513 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  997541 |  998270 | ASMT      | K562 |  0.107038452 | 3.182215e-02 | 0.82948412 | 0.3051774 |\n",
       "| chrX |  997541 |  998270 | ASMTL     | K562 |  0.121607580 | 1.369742e-02 | 0.82948412 | 0.8825453 |\n",
       "| chrX |  997541 |  998270 | CSF2RA    | K562 |  0.001336011 | 9.827104e-01 | 0.82948412 | 0.3498268 |\n",
       "| chrX |  997541 |  998270 | GTPBP6    | K562 |  0.350798315 | 7.352820e-15 | 0.82948412 | 0.6354853 |\n",
       "| chrX |  997541 |  998270 | IL3RA     | K562 |  0.286757362 | 4.568795e-10 | 0.82948412 | 0.9442737 |\n",
       "| chrX |  997541 |  998270 | LINC00106 | K562 | -0.332056590 | 2.476046e-13 | 0.82948412 | 0.6546626 |\n",
       "| chrX |  997541 |  998270 | P2RY8     | K562 | -0.351017308 | 7.045703e-15 | 0.82948412 | 0.9853609 |\n",
       "| chrX |  997541 |  998270 | PLCXD1    | K562 |  0.511510173 | 4.779843e-33 | 0.82948412 | 0.7444005 |\n",
       "| chrX |  997541 |  998270 | PPP2R3B   | K562 |  0.287686214 | 3.968925e-10 | 0.82948412 | 0.5513590 |\n",
       "| chrX |  997541 |  998270 | SLC25A6   | K562 |  0.303317090 | 3.412362e-11 | 0.82948412 | 0.8978676 |\n",
       "| chrX | 9988858 | 9989079 | CLCN4     | K562 |  0.317437877 | 3.249789e-12 | 0.09842854 | 0.3781779 |\n",
       "| chrX | 9988858 | 9989079 | TBL1X     | K562 | -0.052752862 | 3.203807e-01 | 0.09842854 | 0.9504709 |\n",
       "| chrX | 9988858 | 9989079 | WWC3      | K562 |  0.154254665 | 1.433381e-03 | 0.09842854 | 0.8111062 |\n",
       "| chrX |  999293 |  999875 | AKAP17A   | K562 |  0.374751807 | 5.685914e-17 | 0.87065756 | 0.9154833 |\n",
       "| chrX |  999293 |  999875 | ASMT      | K562 | -0.022216379 | 6.963897e-01 | 0.87065756 | 0.3051774 |\n",
       "| chrX |  999293 |  999875 | ASMTL     | K562 | -0.524880732 | 4.695757e-35 | 0.87065756 | 0.8825453 |\n",
       "| chrX |  999293 |  999875 | CSF2RA    | K562 | -0.217192594 | 3.994496e-06 | 0.87065756 | 0.3498268 |\n",
       "| chrX |  999293 |  999875 | GTPBP6    | K562 | -0.056599936 | 2.833746e-01 | 0.87065756 | 0.6354853 |\n",
       "| chrX |  999293 |  999875 | IL3RA     | K562 | -0.568482053 | 2.836953e-42 | 0.87065756 | 0.9442737 |\n",
       "| chrX |  999293 |  999875 | LINC00106 | K562 | -0.350804221 | 7.344386e-15 | 0.87065756 | 0.6546626 |\n",
       "| chrX |  999293 |  999875 | P2RY8     | K562 |  0.546970181 | 1.406170e-38 | 0.87065756 | 0.9853609 |\n",
       "| chrX |  999293 |  999875 | PLCXD1    | K562 |  0.167609106 | 4.890576e-04 | 0.87065756 | 0.7444005 |\n",
       "| chrX |  999293 |  999875 | PPP2R3B   | K562 | -0.214487248 | 5.372700e-06 | 0.87065756 | 0.5513590 |\n",
       "| chrX |  999293 |  999875 | SLC25A6   | K562 |  0.402618679 | 1.149283e-19 | 0.87065756 | 0.8978676 |\n",
       "| chrX | 9995654 | 9996249 | CLCN4     | K562 | -0.017752707 | 7.584762e-01 | 0.70551298 | 0.3781779 |\n",
       "| chrX | 9995654 | 9996249 | TBL1X     | K562 | -0.122068385 | 1.331493e-02 | 0.70551298 | 0.9504709 |\n",
       "| chrX | 9995654 | 9996249 | WWC3      | K562 | -0.114239506 | 2.124022e-02 | 0.70551298 | 0.8111062 |\n",
       "| chrX | 9997430 | 9997980 | CLCN4     | K562 |  0.104033205 | 3.741343e-02 | 0.83176402 | 0.3781779 |\n",
       "| chrX | 9997430 | 9997980 | TBL1X     | K562 | -0.373740032 | 7.040567e-17 | 0.83176402 | 0.9504709 |\n",
       "| chrX | 9997430 | 9997980 | WWC3      | K562 | -0.109513741 | 2.777099e-02 | 0.83176402 | 0.8111062 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType Correlation  FDR         \n",
       "1       chr1 100028533 100029101 AGL        K562      0.239060154 3.124260e-07\n",
       "2       chr1 100028533 100029101 CDC14A     K562      0.250611041 7.282513e-08\n",
       "3       chr1 100028533 100029101 DBT        K562     -0.006377297 9.157406e-01\n",
       "4       chr1 100028533 100029101 DPH5       K562      0.109192839 2.827099e-02\n",
       "5       chr1 100028533 100029101 EXTL2      K562      0.297700272 8.388435e-11\n",
       "6       chr1 100028533 100029101 FRRS1      K562      0.052411707 3.238124e-01\n",
       "7       chr1 100028533 100029101 MFSD14A    K562      0.259727933 2.182599e-08\n",
       "8       chr1 100028533 100029101 RTCA       K562     -0.093778868 6.298193e-02\n",
       "9       chr1 100028533 100029101 RTCA-AS1   K562     -0.094458669 6.092956e-02\n",
       "10      chr1 100028533 100029101 SASS6      K562     -0.247961866 1.023916e-07\n",
       "11      chr1 100028533 100029101 SLC30A7    K562     -0.073527423 1.536087e-01\n",
       "12      chr1 100028533 100029101 SLC35A3    K562     -0.180999641 1.518199e-04\n",
       "13      chr1 100028533 100029101 TRMT13     K562     -0.121284659 1.397065e-02\n",
       "14      chr1 100028533 100029101 VCAM1      K562      0.275252877 2.502730e-09\n",
       "15      chr1 100029136 100029435 AGL        K562      0.096239802 5.582017e-02\n",
       "16      chr1 100029136 100029435 CDC14A     K562     -0.120592021 1.457354e-02\n",
       "17      chr1 100029136 100029435 DBT        K562     -0.017272916 7.652039e-01\n",
       "18      chr1 100029136 100029435 DPH5       K562     -0.011982116 8.388946e-01\n",
       "19      chr1 100029136 100029435 EXTL2      K562     -0.012492951 8.317856e-01\n",
       "20      chr1 100029136 100029435 FRRS1      K562      0.048695609 3.624027e-01\n",
       "21      chr1 100029136 100029435 MFSD14A    K562      0.060060088 2.524295e-01\n",
       "22      chr1 100029136 100029435 RTCA       K562      0.004180552 9.451665e-01\n",
       "23      chr1 100029136 100029435 RTCA-AS1   K562      0.143387096 3.219733e-03\n",
       "24      chr1 100029136 100029435 SASS6      K562      0.031613771 5.686221e-01\n",
       "25      chr1 100029136 100029435 SLC30A7    K562     -0.072689798 1.587742e-01\n",
       "26      chr1 100029136 100029435 SLC35A3    K562     -0.119545915 1.552624e-02\n",
       "27      chr1 100029136 100029435 TRMT13     K562      0.097198135 5.321544e-02\n",
       "28      chr1 100029136 100029435 VCAM1      K562     -0.043401535 4.216773e-01\n",
       "29      chr1 100034578 100034758 AGL        K562     -0.177137591 2.147535e-04\n",
       "30      chr1 100034578 100034758 CDC14A     K562     -0.242069051 2.153961e-07\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮            ⋮           \n",
       "2628593 chrX  997541    998270   ASMT       K562      0.107038452 3.182215e-02\n",
       "2628594 chrX  997541    998270   ASMTL      K562      0.121607580 1.369742e-02\n",
       "2628595 chrX  997541    998270   CSF2RA     K562      0.001336011 9.827104e-01\n",
       "2628596 chrX  997541    998270   GTPBP6     K562      0.350798315 7.352820e-15\n",
       "2628597 chrX  997541    998270   IL3RA      K562      0.286757362 4.568795e-10\n",
       "2628598 chrX  997541    998270   LINC00106  K562     -0.332056590 2.476046e-13\n",
       "2628599 chrX  997541    998270   P2RY8      K562     -0.351017308 7.045703e-15\n",
       "2628600 chrX  997541    998270   PLCXD1     K562      0.511510173 4.779843e-33\n",
       "2628601 chrX  997541    998270   PPP2R3B    K562      0.287686214 3.968925e-10\n",
       "2628602 chrX  997541    998270   SLC25A6    K562      0.303317090 3.412362e-11\n",
       "2628603 chrX 9988858   9989079   CLCN4      K562      0.317437877 3.249789e-12\n",
       "2628604 chrX 9988858   9989079   TBL1X      K562     -0.052752862 3.203807e-01\n",
       "2628605 chrX 9988858   9989079   WWC3       K562      0.154254665 1.433381e-03\n",
       "2628606 chrX  999293    999875   AKAP17A    K562      0.374751807 5.685914e-17\n",
       "2628607 chrX  999293    999875   ASMT       K562     -0.022216379 6.963897e-01\n",
       "2628608 chrX  999293    999875   ASMTL      K562     -0.524880732 4.695757e-35\n",
       "2628609 chrX  999293    999875   CSF2RA     K562     -0.217192594 3.994496e-06\n",
       "2628610 chrX  999293    999875   GTPBP6     K562     -0.056599936 2.833746e-01\n",
       "2628611 chrX  999293    999875   IL3RA      K562     -0.568482053 2.836953e-42\n",
       "2628612 chrX  999293    999875   LINC00106  K562     -0.350804221 7.344386e-15\n",
       "2628613 chrX  999293    999875   P2RY8      K562      0.546970181 1.406170e-38\n",
       "2628614 chrX  999293    999875   PLCXD1     K562      0.167609106 4.890576e-04\n",
       "2628615 chrX  999293    999875   PPP2R3B    K562     -0.214487248 5.372700e-06\n",
       "2628616 chrX  999293    999875   SLC25A6    K562      0.402618679 1.149283e-19\n",
       "2628617 chrX 9995654   9996249   CLCN4      K562     -0.017752707 7.584762e-01\n",
       "2628618 chrX 9995654   9996249   TBL1X      K562     -0.122068385 1.331493e-02\n",
       "2628619 chrX 9995654   9996249   WWC3       K562     -0.114239506 2.124022e-02\n",
       "2628620 chrX 9997430   9997980   CLCN4      K562      0.104033205 3.741343e-02\n",
       "2628621 chrX 9997430   9997980   TBL1X      K562     -0.373740032 7.040567e-17\n",
       "2628622 chrX 9997430   9997980   WWC3       K562     -0.109513741 2.777099e-02\n",
       "        VarQATAC   VarQRNA  \n",
       "1       0.25004186 0.7980286\n",
       "2       0.25004186 0.9289513\n",
       "3       0.25004186 0.6792563\n",
       "4       0.25004186 0.8534134\n",
       "5       0.25004186 0.9627678\n",
       "6       0.25004186 0.8201337\n",
       "7       0.25004186 0.4436149\n",
       "8       0.25004186 0.6899917\n",
       "9       0.25004186 0.6208949\n",
       "10      0.25004186 0.9006490\n",
       "11      0.25004186 0.8160347\n",
       "12      0.25004186 0.6425609\n",
       "13      0.25004186 0.7007271\n",
       "14      0.25004186 0.6057678\n",
       "15      0.02579378 0.7980286\n",
       "16      0.02579378 0.9289513\n",
       "17      0.02579378 0.6792563\n",
       "18      0.02579378 0.8534134\n",
       "19      0.02579378 0.9627678\n",
       "20      0.02579378 0.8201337\n",
       "21      0.02579378 0.4436149\n",
       "22      0.02579378 0.6899917\n",
       "23      0.02579378 0.6208949\n",
       "24      0.02579378 0.9006490\n",
       "25      0.02579378 0.8160347\n",
       "26      0.02579378 0.6425609\n",
       "27      0.02579378 0.7007271\n",
       "28      0.02579378 0.6057678\n",
       "29      0.47476654 0.7980286\n",
       "30      0.47476654 0.9289513\n",
       "⋮       ⋮          ⋮        \n",
       "2628593 0.82948412 0.3051774\n",
       "2628594 0.82948412 0.8825453\n",
       "2628595 0.82948412 0.3498268\n",
       "2628596 0.82948412 0.6354853\n",
       "2628597 0.82948412 0.9442737\n",
       "2628598 0.82948412 0.6546626\n",
       "2628599 0.82948412 0.9853609\n",
       "2628600 0.82948412 0.7444005\n",
       "2628601 0.82948412 0.5513590\n",
       "2628602 0.82948412 0.8978676\n",
       "2628603 0.09842854 0.3781779\n",
       "2628604 0.09842854 0.9504709\n",
       "2628605 0.09842854 0.8111062\n",
       "2628606 0.87065756 0.9154833\n",
       "2628607 0.87065756 0.3051774\n",
       "2628608 0.87065756 0.8825453\n",
       "2628609 0.87065756 0.3498268\n",
       "2628610 0.87065756 0.6354853\n",
       "2628611 0.87065756 0.9442737\n",
       "2628612 0.87065756 0.6546626\n",
       "2628613 0.87065756 0.9853609\n",
       "2628614 0.87065756 0.7444005\n",
       "2628615 0.87065756 0.5513590\n",
       "2628616 0.87065756 0.8978676\n",
       "2628617 0.70551298 0.3781779\n",
       "2628618 0.70551298 0.9504709\n",
       "2628619 0.70551298 0.8111062\n",
       "2628620 0.83176402 0.3781779\n",
       "2628621 0.83176402 0.9504709\n",
       "2628622 0.83176402 0.8111062"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(obj.ArchR,\n",
    "        paste(dir.output,\"obj.ArchR.rds\",sep = \"/\"))\n",
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"Correlation\",\n",
    "                         \"FDR\",\n",
    "                         \"VarQATAC\",\n",
    "                         \"VarQRNA\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "pairs.E2G.res\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4c909ed5-7321-4e69-8208-1544b371da6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] nabor_0.5.0                       dplyr_1.1.4                      \n",
       " [3] uwot_0.1.16                       Rsamtools_2.18.0                 \n",
       " [5] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.68.0                  \n",
       " [7] rtracklayer_1.60.1                Biostrings_2.70.1                \n",
       " [9] XVector_0.42.0                    genomation_1.34.0                \n",
       "[11] Signac_1.13.0                     Seurat_5.0.3                     \n",
       "[13] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[15] rhdf5_2.44.0                      SummarizedExperiment_1.30.2      \n",
       "[17] Biobase_2.60.0                    MatrixGenerics_1.12.3            \n",
       "[19] Rcpp_1.0.12                       Matrix_1.6-5                     \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               matrixStats_1.3.0                \n",
       "[27] data.table_1.15.2                 stringr_1.5.1                    \n",
       "[29] plyr_1.8.9                        magrittr_2.0.3                   \n",
       "[31] ggplot2_3.5.1                     gtable_0.3.5                     \n",
       "[33] gtools_3.9.5                      gridExtra_2.3                    \n",
       "[35] ArchR_1.0.2                      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22         splines_4.3.3            later_1.3.2             \n",
       "  [4] pbdZMQ_0.3-11            BiocIO_1.10.0            bitops_1.0-7            \n",
       "  [7] tibble_3.2.1             polyclip_1.10-6          XML_3.99-0.16.1         \n",
       " [10] fastDummies_1.7.3        lifecycle_1.0.4          vroom_1.6.5             \n",
       " [13] globals_0.16.3           lattice_0.22-6           MASS_7.3-60             \n",
       " [16] plotly_4.10.4            plotrix_3.8-4            yaml_2.3.8              \n",
       " [19] httpuv_1.6.15            sctransform_0.4.1        spam_2.10-0             \n",
       " [22] spatstat.sparse_3.0-3    reticulate_1.36.1        cowplot_1.1.3           \n",
       " [25] pbapply_1.7-2            RColorBrewer_1.1-3       abind_1.4-5             \n",
       " [28] zlibbioc_1.48.0          Rtsne_0.17               purrr_1.0.2             \n",
       " [31] RCurl_1.98-1.14          GenomeInfoDbData_1.2.11  ggrepel_0.9.5           \n",
       " [34] irlba_2.3.5.1            listenv_0.9.1            spatstat.utils_3.0-4    \n",
       " [37] goftest_1.2-3            RSpectra_0.16-1          spatstat.random_3.2-3   \n",
       " [40] fitdistrplus_1.1-11      parallelly_1.37.1        leiden_0.4.3.1          \n",
       " [43] codetools_0.2-20         DelayedArray_0.26.7      RcppRoll_0.3.0          \n",
       " [46] tidyselect_1.2.1         farver_2.1.1             base64enc_0.1-3         \n",
       " [49] spatstat.explore_3.2-6   GenomicAlignments_1.36.0 jsonlite_1.8.8          \n",
       " [52] progressr_0.14.0         ggridges_0.5.6           survival_3.6-4          \n",
       " [55] tools_4.3.3              ica_1.0-3                glue_1.7.0              \n",
       " [58] IRdisplay_1.1            withr_3.0.0              fastmap_1.1.1           \n",
       " [61] rhdf5filters_1.12.1      fansi_1.0.6              digest_0.6.35           \n",
       " [64] R6_2.5.1                 mime_0.12                seqPattern_1.34.0       \n",
       " [67] colorspace_2.1-0         Cairo_1.6-2              scattermore_1.2         \n",
       " [70] tensor_1.5               spatstat.data_3.0-4      utf8_1.2.4              \n",
       " [73] tidyr_1.3.1              generics_0.1.3           httr_1.4.7              \n",
       " [76] htmlwidgets_1.6.4        S4Arrays_1.2.1           pkgconfig_2.0.3         \n",
       " [79] impute_1.76.0            lmtest_0.9-40            htmltools_0.5.8.1       \n",
       " [82] dotCall64_1.1-1          scales_1.3.0             png_0.1-8               \n",
       " [85] tzdb_0.4.0               reshape2_1.4.4           rjson_0.2.21            \n",
       " [88] uuid_1.2-0               nlme_3.1-164             repr_1.1.7              \n",
       " [91] zoo_1.8-12               KernSmooth_2.23-22       miniUI_0.1.1.1          \n",
       " [94] restfulr_0.0.15          pillar_1.9.0             vctrs_0.6.5             \n",
       " [97] RANN_2.6.1               promises_1.3.0           xtable_1.8-4            \n",
       "[100] cluster_2.1.6            evaluate_0.23            readr_2.1.5             \n",
       "[103] cli_3.6.2                compiler_4.3.3           rlang_1.1.3             \n",
       "[106] crayon_1.5.2             future.apply_1.11.2      labeling_0.4.3          \n",
       "[109] stringi_1.8.3            gridBase_0.4-7           viridisLite_0.4.2       \n",
       "[112] deldir_2.0-4             BiocParallel_1.36.0      munsell_0.5.1           \n",
       "[115] lazyeval_0.2.2           spatstat.geom_3.2-9      IRkernel_1.3.2          \n",
       "[118] RcppHNSW_0.6.0           hms_1.1.3                patchwork_1.2.0         \n",
       "[121] bit64_4.0.5              future_1.33.2            Rhdf5lib_1.22.1         \n",
       "[124] shiny_1.8.1.1            ROCR_1.0-11              igraph_2.0.3            \n",
       "[127] fastmatch_1.1-4          bit_4.0.5               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7690cee-49f1-4886-ae66-0b63a43ce524",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
