{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8466b35-cae5-4f62-8138-37b2de484a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.2\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : grid v4.3.3\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "\tLoading Package : gtools v3.9.5\n",
      "\n",
      "\tLoading Package : gtable v0.3.5\n",
      "\n",
      "\tLoading Package : ggplot2 v3.5.1\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "\tLoading Package : plyr v1.8.9\n",
      "\n",
      "\tLoading Package : stringr v1.5.1\n",
      "\n",
      "\tLoading Package : data.table v1.15.2\n",
      "\n",
      "\tLoading Package : matrixStats v1.3.0\n",
      "\n",
      "\tLoading Package : S4Vectors v0.40.2\n",
      "\n",
      "\tLoading Package : GenomicRanges v1.54.1\n",
      "\n",
      "\tLoading Package : BiocGenerics v0.48.1\n",
      "\n",
      "\tLoading Package : Matrix v1.6.5\n",
      "\n",
      "\tLoading Package : Rcpp v1.0.12\n",
      "\n",
      "\tLoading Package : SummarizedExperiment v1.30.2\n",
      "\n",
      "\tLoading Package : rhdf5 v2.44.0\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(magrittr)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(ggplot2)\n",
    "library(dplyr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e85020d7-d358-492d-9358-c81270f4b624",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1bec6690-52ec-450d-bba7-7d4528a185cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "TSS_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.bed\"\n",
    "path.atac_frag = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/scE2G_input_Maya/Jurkat/atac_fragments.tsv.gz\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/scE2G_input_Maya/Jurkat/rna_count_matrix.csv.gz\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/Jurkat/Jurkat/Kendall/Pairs.tsv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/Jurkat/1.Genome_wide_prediction/ArchR/ArchR/\"\n",
    "celltype = \"Jurkat\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "340dac32-44ac-4231-b964-b5f3b9e2dbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16\n",
    "maxDist = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da973385-4c81-4b1c-9669-cf5933b23a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 16.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRThreads(threads = n.cores) \n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86be3e66-c954-4620-9e64-c6a1c0fcb84c",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6f363969-bf0b-45ce-858c-a8cb64cc210a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1ff4f07-c695-45a1-bf15-dc523d20497f",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2a104a7a-b6b9-4315-8b15-6a22c48e142a",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna = read.csv(path.matrix.rna_count,\n",
    "                      row.names = 1,\n",
    "                      check.names = F)\n",
    "matrix.rna = Matrix(as.matrix(matrix.rna), sparse = TRUE)\n",
    "# matrix.rna = matrix.rna[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna[rowSums(matrix.rna) > 0,]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "287f8d61-9ad5-4f60-97b6-9e7b71795042",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ad6e79b8-2429-425a-ac78-6773abe4dbc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6ddc5a2b-371a-435d-9250-751ecffa65c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65a5d8fc-a99f-481c-9eec-8e88c8875dd3",
   "metadata": {},
   "source": [
    "make TSS GRanges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "815b37df-d305-4ba7-bd48-2270290ca3b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tss <- fread(TSS_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\"))\n",
    "df.tss = as.data.frame(df.tss)\n",
    "rownames(df.tss) = df.tss$name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fce150b6-ee55-46d7-9035-cd9698b5ee9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tss[df.tss$strand == \"+\",\"end\"] = df.tss[df.tss$strand == \"+\",\"start\"]\n",
    "df.tss[df.tss$strand == \"-\",\"start\"] = df.tss[df.tss$strand == \"-\",\"end\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7d73721a-8ec9-4b65-91b5-29cca6872927",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tss = df.tss[unname(rownames(matrix.rna.rename)),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a5a5da89-e7fd-4ffb-b477-48beff459909",
   "metadata": {},
   "outputs": [],
   "source": [
    "gr.tss <- makeGRangesFromDataFrame(\n",
    "  df.tss,\n",
    "  seqnames.field = \"chr\",\n",
    "  start.field = \"start\",\n",
    "  end.field = \"end\",\n",
    "  strand.field = \"strand\",    \n",
    "  keep.extra.columns = TRUE   \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a2259f88-32ce-4299-9592-b39c2273f03e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 17571 ranges and 4 metadata columns:\n",
       "            seqnames    ranges strand |        name     score      Ensembl_ID\n",
       "               <Rle> <IRanges>  <Rle> | <character> <integer>     <character>\n",
       "      OR4F5     chr1     69090      + |       OR4F5         0 ENSG00000186092\n",
       "     FAM87B     chr1    817370      + |      FAM87B         0 ENSG00000177757\n",
       "  LINC01128     chr1    827590      + |   LINC01128         0 ENSG00000228794\n",
       "     FAM41C     chr1    876802      - |      FAM41C         0 ENSG00000230368\n",
       "     SAMD11     chr1    925740      + |      SAMD11         0 ENSG00000187634\n",
       "        ...      ...       ...    ... .         ...       ...             ...\n",
       "     NLGN4Y     chrY  14523745      + |      NLGN4Y         0 ENSG00000165246\n",
       "     TTTY14     chrY  19077547      - |      TTTY14         0 ENSG00000176728\n",
       "      KDM5D     chrY  19744939      - |       KDM5D         0 ENSG00000012817\n",
       "     TTTY10     chrY  20519228      - |      TTTY10         0 ENSG00000229236\n",
       "     EIF1AY     chrY  20575710      + |      EIF1AY         0 ENSG00000198692\n",
       "                       gene_type\n",
       "                     <character>\n",
       "      OR4F5       protein_coding\n",
       "     FAM87B              lincRNA\n",
       "  LINC01128 processed_transcript\n",
       "     FAM41C              lincRNA\n",
       "     SAMD11       protein_coding\n",
       "        ...                  ...\n",
       "     NLGN4Y       protein_coding\n",
       "     TTTY14              lincRNA\n",
       "      KDM5D       protein_coding\n",
       "     TTTY10              lincRNA\n",
       "     EIF1AY       protein_coding\n",
       "  -------\n",
       "  seqinfo: 24 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gr.tss"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3538ffdc-4fed-48eb-9c9d-86805a91cf06",
   "metadata": {},
   "source": [
    "Create an ArchRProject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1753ffaa-89e2-43bf-bcf6-19937782d63a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-createArrows-6788d2a6a5d07-Date-2025-03-22_Time-16-52-34.894808.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Cleaning Temporary Files\n",
      "\n",
      "2025-03-22 16:52:35.292951 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "(Jurkat : 1 of 1) Determining Arrow Method to use!\n",
      "\n",
      "2025-03-22 16:52:35.34847 : (Jurkat : 1 of 1) Reading In Fragments from inputFiles (readMethod = tabix), 0.001 mins elapsed.\n",
      "\n",
      "2025-03-22 16:52:35.356103 : (Jurkat : 1 of 1) Tabix Bed To Temporary File, 0.001 mins elapsed.\n",
      "\n",
      "2025-03-22 16:53:41.194705 : (Jurkat : 1 of 1) Successful creation of Temporary File, 1.098 mins elapsed.\n",
      "\n",
      "2025-03-22 16:53:41.206138 : (Jurkat : 1 of 1) Creating ArrowFile From Temporary File, 1.099 mins elapsed.\n",
      "\n",
      "2025-03-22 16:54:07.606263 : (Jurkat : 1 of 1) Successful creation of Arrow File, 1.539 mins elapsed.\n",
      "\n",
      "2025-03-22 16:54:21.79745 : Continuing through after error ggplot for Fragment Size Distribution, 1.775 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-03-22 16:54:44.846012 : (Jurkat : 1 of 1) CellStats : Number of Cells Pass Filter = 5276 , 2.159 mins elapsed.\n",
      "\n",
      "2025-03-22 16:54:44.850795 : (Jurkat : 1 of 1) CellStats : Median Frags = 9551 , 2.159 mins elapsed.\n",
      "\n",
      "2025-03-22 16:54:44.857028 : (Jurkat : 1 of 1) CellStats : Median TSS Enrichment = 15.406 , 2.159 mins elapsed.\n",
      "\n",
      "2025-03-22 16:54:45.828795 : Continuing through after error ggplot for TSS by Frags, 2.176 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-03-22 16:54:48.487848 : (Jurkat : 1 of 1) Adding Additional Feature Counts!, 2.22 mins elapsed.\n",
      "\n",
      "2025-03-22 16:55:06.512588 : (Jurkat : 1 of 1) Removing Fragments from Filtered Cells, 2.52 mins elapsed.\n",
      "\n",
      "2025-03-22 16:55:06.521932 : (Jurkat : 1 of 1) Creating Filtered Arrow File, 2.52 mins elapsed.\n",
      "\n",
      "2025-03-22 16:56:05.424935 : (Jurkat : 1 of 1) Finished Constructing Filtered Arrow File!, 3.502 mins elapsed.\n",
      "\n",
      "2025-03-22 16:56:05.593074 : (Jurkat : 1 of 1) Adding TileMatrix!, 3.505 mins elapsed.\n",
      "\n",
      "2025-03-22 16:59:30.390549 : (Jurkat : 1 of 1) Finished Creating Arrow File, 6.918 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-createArrows-6788d2a6a5d07-Date-2025-03-22_Time-16-52-34.894808.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ArrowFiles <- createArrowFiles(\n",
    "  inputFiles = c(path.atac_frag),\n",
    "  sampleNames = c(celltype),\n",
    "  validBarcodes = list(Jurkat = colnames(matrix.rna.rename)),\n",
    "  addTileMat = T,\n",
    "  addGeneScoreMat = F,\n",
    "  force = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ada3db45-d050-4659-a7cf-2891e04b1249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'Jurkat.arrow'"
      ],
      "text/latex": [
       "'Jurkat.arrow'"
      ],
      "text/markdown": [
       "'Jurkat.arrow'"
      ],
      "text/plain": [
       "[1] \"Jurkat.arrow\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ArrowFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5ba28283-d488-4982-8d4c-745fc3f59365",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"ArchR\",sep=\"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "aa217507-8323-43bf-b23c-0ec25ee33c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Validating Arrows...\n",
      "\n",
      "Getting SampleNames...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Copying ArrowFiles to Ouptut Directory! If you want to save disk space set copyArrows = FALSE\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Getting Cell Metadata...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Merging Cell Metadata...\n",
      "\n",
      "Initializing ArchRProject...\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR <- ArchRProject(\n",
    "  ArrowFiles = ArrowFiles, \n",
    "  outputDirectory = paste(dir.output,\"ArchR\",sep=\"/\"),\n",
    "  copyArrows = TRUE \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f9795403-545f-4385-aa38-2c70b1b2fb21",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: ArchRProject \n",
       "outputDirectory: /maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/Jurkat/1.Genome_wide_prediction/ArchR/ArchR/ArchR \n",
       "samples(1): Jurkat\n",
       "sampleColData names(1): ArrowFiles\n",
       "cellColData names(13): Sample TSSEnrichment ... nDiFrags BlacklistRatio\n",
       "numberOfCells(1): 5276\n",
       "medianTSS(1): 15.406\n",
       "medianFrags(1): 9551"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obj.ArchR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ee876dd5-0b22-46fd-9015-7e22144c5024",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking Inputs...\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addIterativeLSI-6788d5666bbbb-Date-2025-03-22_Time-16-59-36.00687.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-03-22 16:59:36.684237 : Computing Total Across All Features, 0 mins elapsed.\n",
      "\n",
      "2025-03-22 16:59:37.522429 : Computing Top Features, 0.014 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-03-22 16:59:39.834307 : Running LSI (1 of 2) on Top Features, 0.053 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-03-22 16:59:39.849399 : Creating Partial Matrix, 0.053 mins elapsed.\n",
      "\n",
      "2025-03-22 16:59:49.347251 : Computing LSI, 0.211 mins elapsed.\n",
      "\n",
      "2025-03-22 17:00:38.598938 : Identifying Clusters, 1.032 mins elapsed.\n",
      "\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "2025-03-22 17:00:44.199953 : Identified 6 Clusters, 1.126 mins elapsed.\n",
      "\n",
      "2025-03-22 17:00:44.213906 : Saving LSI Iteration, 1.126 mins elapsed.\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************************************************************\n",
      "2025-03-22 17:00:54.339123 : ERROR Found in .saveIteration for  \n",
      "LogFile = ArchRLogs/ArchR-addIterativeLSI-6788d5666bbbb-Date-2025-03-22_Time-16-59-36.00687.log\n",
      "\n",
      "<simpleError in g$grobs[[legend]]: no such index at level 2\n",
      ">\n",
      "\n",
      "************************************************************\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-03-22 17:00:54.341443 : Creating Cluster Matrix on the total Group Features, 1.295 mins elapsed.\n",
      "\n",
      "2025-03-22 17:01:06.56513 : Computing Variable Features, 1.498 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-03-22 17:01:06.751856 : Running LSI (2 of 2) on Variable Features, 1.501 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-03-22 17:01:06.767986 : Creating Partial Matrix, 1.502 mins elapsed.\n",
      "\n",
      "2025-03-22 17:01:18.118703 : Computing LSI, 1.691 mins elapsed.\n",
      "\n",
      "2025-03-22 17:02:07.274138 : Finished Running IterativeLSI, 2.51 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addIterativeLSI(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    useMatrix = \"TileMatrix\", \n",
    "    name = \"IterativeLSI\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2c1d334b-7fd7-425a-bb8d-7be4cbedd3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.ArchR = addPeakSet(obj.ArchR,\n",
    "                       peakSet = reduce(pairs.E2G))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e25c14db-eb4e-4b94-8862-eb34246a365c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeakMatrix-6788d19ca54a3-Date-2025-03-22_Time-17-02-09.765385.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-03-22 17:02:10.240696 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-03-22 17:02:10.821989 : Adding Jurkat to PeakMatrix for Chr (1 of 23)!, 0.009 mins elapsed.\n",
      "\n",
      "2025-03-22 17:02:17.822808 : Adding Jurkat to PeakMatrix for Chr (2 of 23)!, 0.126 mins elapsed.\n",
      "\n",
      "2025-03-22 17:02:23.649179 : Adding Jurkat to PeakMatrix for Chr (3 of 23)!, 0.223 mins elapsed.\n",
      "\n",
      "2025-03-22 17:02:29.03354 : Adding Jurkat to PeakMatrix for Chr (4 of 23)!, 0.313 mins elapsed.\n",
      "\n",
      "2025-03-22 17:02:33.936394 : Adding Jurkat to PeakMatrix for Chr (5 of 23)!, 0.394 mins elapsed.\n",
      "\n",
      "2025-03-22 17:02:38.993649 : Adding Jurkat to PeakMatrix for Chr (6 of 23)!, 0.479 mins elapsed.\n",
      "\n",
      "2025-03-22 17:02:44.170674 : Adding Jurkat to PeakMatrix for Chr (7 of 23)!, 0.565 mins elapsed.\n",
      "\n",
      "2025-03-22 17:02:49.339962 : Adding Jurkat to PeakMatrix for Chr (8 of 23)!, 0.651 mins elapsed.\n",
      "\n",
      "2025-03-22 17:02:54.208525 : Adding Jurkat to PeakMatrix for Chr (9 of 23)!, 0.732 mins elapsed.\n",
      "\n",
      "2025-03-22 17:02:59.09037 : Adding Jurkat to PeakMatrix for Chr (10 of 23)!, 0.814 mins elapsed.\n",
      "\n",
      "2025-03-22 17:03:04.163214 : Adding Jurkat to PeakMatrix for Chr (11 of 23)!, 0.898 mins elapsed.\n",
      "\n",
      "2025-03-22 17:03:09.489104 : Adding Jurkat to PeakMatrix for Chr (12 of 23)!, 0.987 mins elapsed.\n",
      "\n",
      "2025-03-22 17:03:14.568597 : Adding Jurkat to PeakMatrix for Chr (13 of 23)!, 1.072 mins elapsed.\n",
      "\n",
      "2025-03-22 17:03:18.927418 : Adding Jurkat to PeakMatrix for Chr (14 of 23)!, 1.144 mins elapsed.\n",
      "\n",
      "2025-03-22 17:03:23.602628 : Adding Jurkat to PeakMatrix for Chr (15 of 23)!, 1.222 mins elapsed.\n",
      "\n",
      "2025-03-22 17:03:28.362295 : Adding Jurkat to PeakMatrix for Chr (16 of 23)!, 1.301 mins elapsed.\n",
      "\n",
      "2025-03-22 17:03:33.344296 : Adding Jurkat to PeakMatrix for Chr (17 of 23)!, 1.384 mins elapsed.\n",
      "\n",
      "2025-03-22 17:03:38.580989 : Adding Jurkat to PeakMatrix for Chr (18 of 23)!, 1.472 mins elapsed.\n",
      "\n",
      "2025-03-22 17:03:42.779502 : Adding Jurkat to PeakMatrix for Chr (19 of 23)!, 1.542 mins elapsed.\n",
      "\n",
      "2025-03-22 17:03:48.176021 : Adding Jurkat to PeakMatrix for Chr (20 of 23)!, 1.632 mins elapsed.\n",
      "\n",
      "2025-03-22 17:03:52.851174 : Adding Jurkat to PeakMatrix for Chr (21 of 23)!, 1.71 mins elapsed.\n",
      "\n",
      "2025-03-22 17:03:56.959206 : Adding Jurkat to PeakMatrix for Chr (22 of 23)!, 1.778 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:01.521095 : Adding Jurkat to PeakMatrix for Chr (23 of 23)!, 1.854 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeakMatrix-6788d19ca54a3-Date-2025-03-22_Time-17-02-09.765385.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addPeakMatrix(obj.ArchR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "550747c0-ded3-4a82-bc47-9d259a41ca2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna.rename2 = matrix.rna.rename\n",
    "colnames(matrix.rna.rename2) = paste(celltype,colnames(matrix.rna.rename),sep=\"#\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "92b20885-7b28-419c-a66d-74640e850733",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA <- SummarizedExperiment(\n",
    "  assays = SimpleList(counts = matrix.rna.rename2),\n",
    "  rowRanges = gr.tss\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "834ae1ff-160a-4e09-b201-ffe7ab9919eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA = sort.GenomicRanges(sortSeqlevels(seRNA), \n",
    "                           ignore.strand = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4b4e7086-362b-4c79-8e58-327684d0b787",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 17571 5277 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(17571): OR4F5 FAM87B ... TTTY10 EIF1AY\n",
       "rowData names(4): name score Ensembl_ID gene_type\n",
       "colnames(5277): Jurkat#e10l1_AAACCAGGTACTTATG-1\n",
       "  Jurkat#e10l1_AAACCAGGTCACACTG-1 ... Jurkat#e10l2_TTTGTGAGTCCTGAGC-1\n",
       "  Jurkat#e10l2_TTTGTGAGTTTATGTG-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "65cb2097-9075-4ee4-9e54-dd288dc50736",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addGeneExpressionMatrix-6788d10b221ad-Date-2025-03-22_Time-17-04-06.821096.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Overlap w/ scATAC = 1\n",
      "\n",
      "2025-03-22 17:04:07.454319 : \n",
      "\n",
      "Overlap Per Sample w/ scATAC : Jurkat=5276\n",
      "\n",
      "2025-03-22 17:04:07.463572 : \n",
      "\n",
      "2025-03-22 17:04:08.490168 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:09.420528 : Adding Jurkat to GeneExpressionMatrix for Chr (1 of 23)!, 0.016 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:13.320135 : Adding Jurkat to GeneExpressionMatrix for Chr (2 of 23)!, 0.08 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:16.997553 : Adding Jurkat to GeneExpressionMatrix for Chr (3 of 23)!, 0.142 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:22.166369 : Adding Jurkat to GeneExpressionMatrix for Chr (4 of 23)!, 0.228 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:25.761657 : Adding Jurkat to GeneExpressionMatrix for Chr (5 of 23)!, 0.288 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:29.384099 : Adding Jurkat to GeneExpressionMatrix for Chr (6 of 23)!, 0.348 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:34.498576 : Adding Jurkat to GeneExpressionMatrix for Chr (7 of 23)!, 0.433 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:38.102947 : Adding Jurkat to GeneExpressionMatrix for Chr (8 of 23)!, 0.494 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:41.686728 : Adding Jurkat to GeneExpressionMatrix for Chr (9 of 23)!, 0.553 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:46.764408 : Adding Jurkat to GeneExpressionMatrix for Chr (10 of 23)!, 0.638 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:50.39114 : Adding Jurkat to GeneExpressionMatrix for Chr (11 of 23)!, 0.698 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:54.045455 : Adding Jurkat to GeneExpressionMatrix for Chr (12 of 23)!, 0.759 mins elapsed.\n",
      "\n",
      "2025-03-22 17:04:59.293247 : Adding Jurkat to GeneExpressionMatrix for Chr (13 of 23)!, 0.847 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:02.927997 : Adding Jurkat to GeneExpressionMatrix for Chr (14 of 23)!, 0.907 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:06.564755 : Adding Jurkat to GeneExpressionMatrix for Chr (15 of 23)!, 0.968 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:11.643595 : Adding Jurkat to GeneExpressionMatrix for Chr (16 of 23)!, 1.053 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:15.854717 : Adding Jurkat to GeneExpressionMatrix for Chr (17 of 23)!, 1.123 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:19.80947 : Adding Jurkat to GeneExpressionMatrix for Chr (18 of 23)!, 1.189 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:25.239836 : Adding Jurkat to GeneExpressionMatrix for Chr (19 of 23)!, 1.279 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:28.937101 : Adding Jurkat to GeneExpressionMatrix for Chr (20 of 23)!, 1.341 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:32.578024 : Adding Jurkat to GeneExpressionMatrix for Chr (21 of 23)!, 1.401 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:37.907655 : Adding Jurkat to GeneExpressionMatrix for Chr (22 of 23)!, 1.49 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:41.52315 : Adding Jurkat to GeneExpressionMatrix for Chr (23 of 23)!, 1.551 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addGeneExpressionMatrix-6788d10b221ad-Date-2025-03-22_Time-17-04-06.821096.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addGeneExpressionMatrix(obj.ArchR,\n",
    "                                    seRNA = seRNA,\n",
    "                                    strictMatch = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4318703-a143-40c0-b159-274d05033bbc",
   "metadata": {},
   "source": [
    "Run ArchR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "47c8fd5f-19ec-46e6-af87-57502c1e1ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeak2GeneLinks-6788d4d422b55-Date-2025-03-22_Time-17-05-46.82779.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-03-22 17:05:47.398093 : Getting Available Matrices, 0.009 mins elapsed.\n",
      "\n",
      "No predictionScore found. Continuing without predictionScore!\n",
      "\n",
      "2025-03-22 17:05:48.12356 : Filtered Low Prediction Score Cells (0 of 5276, 0), 0 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:48.198036 : Computing KNN, 0.001 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:48.430971 : Identifying Non-Overlapping KNN pairs, 0.005 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:49.933534 : Identified 492 Groupings!, 0.03 mins elapsed.\n",
      "\n",
      "2025-03-22 17:05:49.96273 : Getting Group RNA Matrix, 0.031 mins elapsed.\n",
      "\n",
      "2025-03-22 17:06:02.290691 : Getting Group ATAC Matrix, 0.236 mins elapsed.\n",
      "\n",
      "2025-03-22 17:06:28.564878 : Normalizing Group Matrices, 0.674 mins elapsed.\n",
      "\n",
      "2025-03-22 17:06:33.81362 : Finding Peak Gene Pairings, 0.762 mins elapsed.\n",
      "\n",
      "2025-03-22 17:06:34.551157 : Computing Correlations, 0.774 mins elapsed.\n",
      "\n",
      "2025-03-22 17:06:48.284784 : Completed Peak2Gene Correlations!, 1.003 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeak2GeneLinks-6788d4d422b55-Date-2025-03-22_Time-17-05-46.82779.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "obj.ArchR = addPeak2GeneLinks(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    reducedDims = \"IterativeLSI\",\n",
    "    useMatrix = \"GeneExpressionMatrix\",\n",
    "    maxDist = maxDist,\n",
    "    threads = n.cores\n",
    ")\n",
    "end_time <- Sys.time()\n",
    "execution_time <- end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "128d0065-5f22-495b-b00d-743fc9843337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 1.024715 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "execution_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "36bd2df2-6b32-4b81-bd4f-39ef8913b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g = metadata(obj.ArchR@peakSet)$Peak2GeneLinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e7730413-8fd4-4952-bb13-0df7a3b15b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata(p2g)[[1]]$PeakName = paste(seqnames(metadata(p2g)[[1]]),\n",
    "                                    ranges(metadata(p2g)[[1]]),sep=\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "aeea3810-6735-4aff-9bd7-1697bef979d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g$PairName = paste(metadata(p2g)[[1]][p2g$idxATAC]$PeakName,\n",
    "                     metadata(p2g)[[2]][p2g$idxRNA]$name,\n",
    "                     sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "87f975ec-6ba4-4696-ad6c-2af65da7aa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 3125860 rows and 7 columns\n",
       "                                idxATAC    idxRNA Correlation       FDR\n",
       "                              <integer> <integer>   <numeric> <numeric>\n",
       "chr1-10010-10640_OR4F5                1         1   0.0598673  0.399143\n",
       "chr1-180669-181787_OR4F5              2         1   0.0298197  0.714636\n",
       "chr1-191241-191629_OR4F5              3         1  -0.0701763  0.305854\n",
       "chr1-267890-268113_OR4F5              4         1   0.0879553  0.176898\n",
       "chr1-586090-586306_OR4F5              5         1  -0.0533506  0.463661\n",
       "...                                 ...       ...         ...       ...\n",
       "chrX-156001485-156001953_IL9R    152883     17554  0.06406394  0.359740\n",
       "chrX-156008847-156009445_IL9R    152884     17554 -0.09324993  0.146873\n",
       "chrX-156019836-156020186_IL9R    152885     17554 -0.04345845  0.567586\n",
       "chrX-156029774-156030373_IL9R    152886     17554 -0.04818530  0.517289\n",
       "chrX-156030497-156030826_IL9R    152887     17554  0.00467496  0.962017\n",
       "                               VarQATAC   VarQRNA               PairName\n",
       "                              <numeric> <numeric>            <character>\n",
       "chr1-10010-10640_OR4F5        0.8472924 0.0756523 chr1-10010-10640_OR4F5\n",
       "chr1-180669-181787_OR4F5      0.9081217 0.0756523 chr1-180669-181787_O..\n",
       "chr1-191241-191629_OR4F5      0.0315004 0.0756523 chr1-191241-191629_O..\n",
       "chr1-267890-268113_OR4F5      0.4780851 0.0756523 chr1-267890-268113_O..\n",
       "chr1-586090-586306_OR4F5      0.6430109 0.0756523 chr1-586090-586306_O..\n",
       "...                                 ...       ...                    ...\n",
       "chrX-156001485-156001953_IL9R  0.756323  0.392047 chrX-156001485-15600..\n",
       "chrX-156008847-156009445_IL9R  0.617005  0.392047 chrX-156008847-15600..\n",
       "chrX-156019836-156020186_IL9R  0.130691  0.392047 chrX-156019836-15602..\n",
       "chrX-156029774-156030373_IL9R  0.864449  0.392047 chrX-156029774-15603..\n",
       "chrX-156030497-156030826_IL9R  0.860269  0.392047 chrX-156030497-15603.."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(p2g) = p2g$PairName\n",
    "p2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c7209a32-3904-4734-93af-9c22049f652a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = pairs.E2G[pairs.E2G$PairName %in% p2g$PairName]\n",
    "mcols(pairs.E2G.res)[,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")] = p2g[pairs.E2G.res$PairName,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "1ab4ed70-ce9e-4857-bd1b-63a81d27a401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2590340 ranges and 7 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10001055-10001570      * | APITD1-CORT\n",
       "        [2]     chr1 10001055-10001570      * |       CASZ1\n",
       "        [3]     chr1 10001055-10001570      * |      CLSTN1\n",
       "        [4]     chr1 10001055-10001570      * |    CTNNBIP1\n",
       "        [5]     chr1 10001055-10001570      * |        DFFA\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [2590336]     chrX   9941883-9942743      * |        WWC3\n",
       "  [2590337]     chrX 99603024-99603776      * |      PCDH19\n",
       "  [2590338]     chrX 99632013-99632219      * |      PCDH19\n",
       "  [2590339]     chrX   9995829-9996207      * |       TBL1X\n",
       "  [2590340]     chrX   9995829-9996207      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-10001055-10001570 chr1-10001055-100015..  -0.0654033\n",
       "        [2] chr1-10001055-10001570 chr1-10001055-100015..   0.0852744\n",
       "        [3] chr1-10001055-10001570 chr1-10001055-100015..   0.1068870\n",
       "        [4] chr1-10001055-10001570 chr1-10001055-100015..   0.1667059\n",
       "        [5] chr1-10001055-10001570 chr1-10001055-100015..  -0.0537374\n",
       "        ...                    ...                    ...         ...\n",
       "  [2590336]   chrX-9941883-9942743 chrX-9941883-9942743.. -0.00339905\n",
       "  [2590337] chrX-99603024-99603776 chrX-99603024-996037..  0.11014386\n",
       "  [2590338] chrX-99632013-99632219 chrX-99632013-996322.. -0.00256255\n",
       "  [2590339]   chrX-9995829-9996207 chrX-9995829-9996207.. -0.02969542\n",
       "  [2590340]   chrX-9995829-9996207 chrX-9995829-9996207.. -0.02875369\n",
       "                   FDR  VarQATAC   VarQRNA\n",
       "             <numeric> <numeric> <numeric>\n",
       "        [1] 0.34756989  0.245279  0.317876\n",
       "        [2] 0.19350047  0.245279  0.405264\n",
       "        [3] 0.08652336  0.245279  0.847271\n",
       "        [4] 0.00334923  0.245279  0.855645\n",
       "        [5] 0.45968029  0.245279  0.655748\n",
       "        ...        ...       ...       ...\n",
       "  [2590336]  0.9726253 0.8063537  0.625385\n",
       "  [2590337]  0.0754152 0.4963666  0.366013\n",
       "  [2590338]  0.9795295 0.0199298  0.366013\n",
       "  [2590339]  0.7159998 0.4723554  0.923664\n",
       "  [2590340]  0.7261407 0.4723554  0.625385\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46485351-50d7-461a-bf09-d795d5bdab84",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6d424e21-6a41-4422-b76c-6f1c0bda2222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2590340 ranges and 7 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10001055-10001570      * | APITD1-CORT\n",
       "        [2]     chr1 10001055-10001570      * |       CASZ1\n",
       "        [3]     chr1 10001055-10001570      * |      CLSTN1\n",
       "        [4]     chr1 10001055-10001570      * |    CTNNBIP1\n",
       "        [5]     chr1 10001055-10001570      * |        DFFA\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [2590336]     chrX   9941883-9942743      * |        WWC3\n",
       "  [2590337]     chrX 99603024-99603776      * |      PCDH19\n",
       "  [2590338]     chrX 99632013-99632219      * |      PCDH19\n",
       "  [2590339]     chrX   9995829-9996207      * |       TBL1X\n",
       "  [2590340]     chrX   9995829-9996207      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-10001055-10001570 chr1-10001055-100015..  -0.0654033\n",
       "        [2] chr1-10001055-10001570 chr1-10001055-100015..   0.0852744\n",
       "        [3] chr1-10001055-10001570 chr1-10001055-100015..   0.1068870\n",
       "        [4] chr1-10001055-10001570 chr1-10001055-100015..   0.1667059\n",
       "        [5] chr1-10001055-10001570 chr1-10001055-100015..  -0.0537374\n",
       "        ...                    ...                    ...         ...\n",
       "  [2590336]   chrX-9941883-9942743 chrX-9941883-9942743.. -0.00339905\n",
       "  [2590337] chrX-99603024-99603776 chrX-99603024-996037..  0.11014386\n",
       "  [2590338] chrX-99632013-99632219 chrX-99632013-996322.. -0.00256255\n",
       "  [2590339]   chrX-9995829-9996207 chrX-9995829-9996207.. -0.02969542\n",
       "  [2590340]   chrX-9995829-9996207 chrX-9995829-9996207.. -0.02875369\n",
       "                   FDR  VarQATAC   VarQRNA\n",
       "             <numeric> <numeric> <numeric>\n",
       "        [1] 0.34756989  0.245279  0.317876\n",
       "        [2] 0.19350047  0.245279  0.405264\n",
       "        [3] 0.08652336  0.245279  0.847271\n",
       "        [4] 0.00334923  0.245279  0.855645\n",
       "        [5] 0.45968029  0.245279  0.655748\n",
       "        ...        ...       ...       ...\n",
       "  [2590336]  0.9726253 0.8063537  0.625385\n",
       "  [2590337]  0.0754152 0.4963666  0.366013\n",
       "  [2590338]  0.9795295 0.0199298  0.366013\n",
       "  [2590339]  0.7159998 0.4723554  0.923664\n",
       "  [2590340]  0.7261407 0.4723554  0.625385\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2590340 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>Correlation</th><th scope=col>FDR</th><th scope=col>VarQATAC</th><th scope=col>VarQRNA</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>APITD1-CORT</td><td>Jurkat</td><td>-0.0654033111</td><td>0.347569890</td><td>0.2452792</td><td>0.31787627</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>CASZ1      </td><td>Jurkat</td><td> 0.0852744357</td><td>0.193500470</td><td>0.2452792</td><td>0.40526376</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>CLSTN1     </td><td>Jurkat</td><td> 0.1068869545</td><td>0.086523359</td><td>0.2452792</td><td>0.84727128</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>CTNNBIP1   </td><td>Jurkat</td><td> 0.1667059261</td><td>0.003349231</td><td>0.2452792</td><td>0.85564544</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>DFFA       </td><td>Jurkat</td><td>-0.0537373593</td><td>0.459680288</td><td>0.2452792</td><td>0.65574798</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>GPR157     </td><td>Jurkat</td><td> 0.0249170647</td><td>0.766729796</td><td>0.2452792</td><td>0.45909764</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>H6PD       </td><td>Jurkat</td><td>-0.0666460691</td><td>0.336484757</td><td>0.2452792</td><td>0.48484676</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>KIF1B      </td><td>Jurkat</td><td> 0.0204891917</td><td>0.812859646</td><td>0.2452792</td><td>0.85558847</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>LINC01759  </td><td>Jurkat</td><td> 0.0471842099</td><td>0.527837922</td><td>0.2452792</td><td>0.07855759</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>LZIC       </td><td>Jurkat</td><td> 0.0963173413</td><td>0.131237723</td><td>0.2452792</td><td>0.71425316</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>MIR34AHG   </td><td>Jurkat</td><td> 0.0710284573</td><td>0.298724608</td><td>0.2452792</td><td>0.10755383</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>NMNAT1     </td><td>Jurkat</td><td>-0.0998270702</td><td>0.114880217</td><td>0.2452792</td><td>0.43294976</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>PEX14      </td><td>Jurkat</td><td> 0.0695721481</td><td>0.310952302</td><td>0.2452792</td><td>0.77093540</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>PGD        </td><td>Jurkat</td><td> 0.0748075719</td><td>0.268079900</td><td>0.2452792</td><td>0.94371653</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>PIK3CD     </td><td>Jurkat</td><td>-0.0221642985</td><td>0.795517474</td><td>0.2452792</td><td>0.63011279</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>SLC25A33   </td><td>Jurkat</td><td> 0.0890276702</td><td>0.170520707</td><td>0.2452792</td><td>0.66759713</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>SPSB1      </td><td>Jurkat</td><td> 0.0004905532</td><td>0.996168050</td><td>0.2452792</td><td>0.38874331</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>TMEM201    </td><td>Jurkat</td><td> 0.0456741285</td><td>0.543999597</td><td>0.2452792</td><td>0.49652501</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>UBE4B      </td><td>Jurkat</td><td>-0.0059055346</td><td>0.951501606</td><td>0.2452792</td><td>0.88486955</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>AGL        </td><td>Jurkat</td><td> 0.0926134261</td><td>0.150284403</td><td>0.9522850</td><td>0.73071665</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>CDC14A     </td><td>Jurkat</td><td>-0.0711974189</td><td>0.297326639</td><td>0.9522850</td><td>0.86003190</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>DBT        </td><td>Jurkat</td><td> 0.0163183519</td><td>0.854955447</td><td>0.9522850</td><td>0.72877977</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>DPH5       </td><td>Jurkat</td><td>-0.0103547298</td><td>0.911621564</td><td>0.9522850</td><td>0.72336789</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>EXTL2      </td><td>Jurkat</td><td> 0.0778734593</td><td>0.244724054</td><td>0.9522850</td><td>0.90777031</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>FRRS1      </td><td>Jurkat</td><td>-0.0188514558</td><td>0.829553184</td><td>0.9522850</td><td>0.56431583</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>MFSD14A    </td><td>Jurkat</td><td> 0.1086712923</td><td>0.080277974</td><td>0.9522850</td><td>0.04016179</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>RTCA       </td><td>Jurkat</td><td>-0.0152314939</td><td>0.865631635</td><td>0.9522850</td><td>0.65039307</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>RTCA-AS1   </td><td>Jurkat</td><td>-0.0050515866</td><td>0.958832018</td><td>0.9522850</td><td>0.51834340</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>SASS6      </td><td>Jurkat</td><td>-0.0412282960</td><td>0.591352088</td><td>0.9522850</td><td>0.88395807</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>SLC30A7    </td><td>Jurkat</td><td> 0.1015604294</td><td>0.107326550</td><td>0.9522850</td><td>0.91021989</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>IL3RA    </td><td>Jurkat</td><td> 0.077058907</td><td>2.507832e-01</td><td>0.08815007</td><td>0.4126125</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>LINC00106</td><td>Jurkat</td><td> 0.015647190</td><td>8.616048e-01</td><td>0.08815007</td><td>0.3766093</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>P2RY8    </td><td>Jurkat</td><td> 0.011162050</td><td>9.041405e-01</td><td>0.08815007</td><td>0.8463598</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>PLCXD1   </td><td>Jurkat</td><td>-0.030797294</td><td>7.042059e-01</td><td>0.08815007</td><td>0.5504728</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>PPP2R3B  </td><td>Jurkat</td><td>-0.087598973</td><td>1.790380e-01</td><td>0.08815007</td><td>0.5432950</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>SHOX     </td><td>Jurkat</td><td>-0.015774162</td><td>8.603733e-01</td><td>0.08815007</td><td>0.1614447</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>SLC25A6  </td><td>Jurkat</td><td>-0.103218806</td><td>1.004718e-01</td><td>0.08815007</td><td>0.9649653</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9842372</td><td> 9843073</td><td>TBL1X    </td><td>Jurkat</td><td>-0.185228384</td><td>8.751534e-04</td><td>0.42109532</td><td>0.9236641</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9842372</td><td> 9843073</td><td>WWC3     </td><td>Jurkat</td><td>-0.008494610</td><td>9.284790e-01</td><td>0.42109532</td><td>0.6253845</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9873846</td><td> 9874297</td><td>TBL1X    </td><td>Jurkat</td><td>-0.076225881</td><td>2.570903e-01</td><td>0.14523145</td><td>0.9236641</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9873846</td><td> 9874297</td><td>WWC3     </td><td>Jurkat</td><td>-0.071940118</td><td>2.911850e-01</td><td>0.14523145</td><td>0.6253845</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9885186</td><td> 9885487</td><td>TBL1X    </td><td>Jurkat</td><td> 0.113045017</td><td>6.645980e-02</td><td>0.17157770</td><td>0.9236641</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9885186</td><td> 9885487</td><td>WWC3     </td><td>Jurkat</td><td>-0.017925654</td><td>8.388762e-01</td><td>0.17157770</td><td>0.6253845</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>AKAP17A  </td><td>Jurkat</td><td>-0.121657316</td><td>4.477509e-02</td><td>0.93286545</td><td>0.6066424</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>ASMT     </td><td>Jurkat</td><td>-0.183437574</td><td>1.003947e-03</td><td>0.93286545</td><td>0.1983024</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>ASMTL    </td><td>Jurkat</td><td>-0.083042940</td><td>2.081521e-01</td><td>0.93286545</td><td>0.5086590</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>GTPBP6   </td><td>Jurkat</td><td> 0.122791536</td><td>4.240690e-02</td><td>0.93286545</td><td>0.5461433</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>IL3RA    </td><td>Jurkat</td><td>-0.095782453</td><td>1.338724e-01</td><td>0.93286545</td><td>0.4126125</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>LINC00106</td><td>Jurkat</td><td>-0.098133681</td><td>1.225733e-01</td><td>0.93286545</td><td>0.3766093</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>P2RY8    </td><td>Jurkat</td><td> 0.099039766</td><td>1.184062e-01</td><td>0.93286545</td><td>0.8463598</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>PLCXD1   </td><td>Jurkat</td><td> 0.181000824</td><td>1.206355e-03</td><td>0.93286545</td><td>0.5504728</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>PPP2R3B  </td><td>Jurkat</td><td>-0.241265309</td><td>5.232516e-06</td><td>0.93286545</td><td>0.5432950</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>SHOX     </td><td>Jurkat</td><td> 0.068078130</td><td>3.237836e-01</td><td>0.93286545</td><td>0.1614447</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>SLC25A6  </td><td>Jurkat</td><td>-0.215293892</td><td>6.904075e-05</td><td>0.93286545</td><td>0.9649653</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9941883</td><td> 9942743</td><td>TBL1X    </td><td>Jurkat</td><td> 0.212400696</td><td>8.989957e-05</td><td>0.80635371</td><td>0.9236641</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9941883</td><td> 9942743</td><td>WWC3     </td><td>Jurkat</td><td>-0.003399047</td><td>9.726253e-01</td><td>0.80635371</td><td>0.6253845</td></tr>\n",
       "\t<tr><td>chrX</td><td>99603024</td><td>99603776</td><td>PCDH19   </td><td>Jurkat</td><td> 0.110143858</td><td>7.541516e-02</td><td>0.49636660</td><td>0.3660134</td></tr>\n",
       "\t<tr><td>chrX</td><td>99632013</td><td>99632219</td><td>PCDH19   </td><td>Jurkat</td><td>-0.002562549</td><td>9.795295e-01</td><td>0.01992975</td><td>0.3660134</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995829</td><td> 9996207</td><td>TBL1X    </td><td>Jurkat</td><td>-0.029695418</td><td>7.159998e-01</td><td>0.47235540</td><td>0.9236641</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995829</td><td> 9996207</td><td>WWC3     </td><td>Jurkat</td><td>-0.028753695</td><td>7.261407e-01</td><td>0.47235540</td><td>0.6253845</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2590340 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & Correlation & FDR & VarQATAC & VarQRNA\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  10001055 &  10001570 & APITD1-CORT & Jurkat & -0.0654033111 & 0.347569890 & 0.2452792 & 0.31787627\\\\\n",
       "\t chr1 &  10001055 &  10001570 & CASZ1       & Jurkat &  0.0852744357 & 0.193500470 & 0.2452792 & 0.40526376\\\\\n",
       "\t chr1 &  10001055 &  10001570 & CLSTN1      & Jurkat &  0.1068869545 & 0.086523359 & 0.2452792 & 0.84727128\\\\\n",
       "\t chr1 &  10001055 &  10001570 & CTNNBIP1    & Jurkat &  0.1667059261 & 0.003349231 & 0.2452792 & 0.85564544\\\\\n",
       "\t chr1 &  10001055 &  10001570 & DFFA        & Jurkat & -0.0537373593 & 0.459680288 & 0.2452792 & 0.65574798\\\\\n",
       "\t chr1 &  10001055 &  10001570 & GPR157      & Jurkat &  0.0249170647 & 0.766729796 & 0.2452792 & 0.45909764\\\\\n",
       "\t chr1 &  10001055 &  10001570 & H6PD        & Jurkat & -0.0666460691 & 0.336484757 & 0.2452792 & 0.48484676\\\\\n",
       "\t chr1 &  10001055 &  10001570 & KIF1B       & Jurkat &  0.0204891917 & 0.812859646 & 0.2452792 & 0.85558847\\\\\n",
       "\t chr1 &  10001055 &  10001570 & LINC01759   & Jurkat &  0.0471842099 & 0.527837922 & 0.2452792 & 0.07855759\\\\\n",
       "\t chr1 &  10001055 &  10001570 & LZIC        & Jurkat &  0.0963173413 & 0.131237723 & 0.2452792 & 0.71425316\\\\\n",
       "\t chr1 &  10001055 &  10001570 & MIR34AHG    & Jurkat &  0.0710284573 & 0.298724608 & 0.2452792 & 0.10755383\\\\\n",
       "\t chr1 &  10001055 &  10001570 & NMNAT1      & Jurkat & -0.0998270702 & 0.114880217 & 0.2452792 & 0.43294976\\\\\n",
       "\t chr1 &  10001055 &  10001570 & PEX14       & Jurkat &  0.0695721481 & 0.310952302 & 0.2452792 & 0.77093540\\\\\n",
       "\t chr1 &  10001055 &  10001570 & PGD         & Jurkat &  0.0748075719 & 0.268079900 & 0.2452792 & 0.94371653\\\\\n",
       "\t chr1 &  10001055 &  10001570 & PIK3CD      & Jurkat & -0.0221642985 & 0.795517474 & 0.2452792 & 0.63011279\\\\\n",
       "\t chr1 &  10001055 &  10001570 & SLC25A33    & Jurkat &  0.0890276702 & 0.170520707 & 0.2452792 & 0.66759713\\\\\n",
       "\t chr1 &  10001055 &  10001570 & SPSB1       & Jurkat &  0.0004905532 & 0.996168050 & 0.2452792 & 0.38874331\\\\\n",
       "\t chr1 &  10001055 &  10001570 & TMEM201     & Jurkat &  0.0456741285 & 0.543999597 & 0.2452792 & 0.49652501\\\\\n",
       "\t chr1 &  10001055 &  10001570 & UBE4B       & Jurkat & -0.0059055346 & 0.951501606 & 0.2452792 & 0.88486955\\\\\n",
       "\t chr1 & 100037799 & 100039086 & AGL         & Jurkat &  0.0926134261 & 0.150284403 & 0.9522850 & 0.73071665\\\\\n",
       "\t chr1 & 100037799 & 100039086 & CDC14A      & Jurkat & -0.0711974189 & 0.297326639 & 0.9522850 & 0.86003190\\\\\n",
       "\t chr1 & 100037799 & 100039086 & DBT         & Jurkat &  0.0163183519 & 0.854955447 & 0.9522850 & 0.72877977\\\\\n",
       "\t chr1 & 100037799 & 100039086 & DPH5        & Jurkat & -0.0103547298 & 0.911621564 & 0.9522850 & 0.72336789\\\\\n",
       "\t chr1 & 100037799 & 100039086 & EXTL2       & Jurkat &  0.0778734593 & 0.244724054 & 0.9522850 & 0.90777031\\\\\n",
       "\t chr1 & 100037799 & 100039086 & FRRS1       & Jurkat & -0.0188514558 & 0.829553184 & 0.9522850 & 0.56431583\\\\\n",
       "\t chr1 & 100037799 & 100039086 & MFSD14A     & Jurkat &  0.1086712923 & 0.080277974 & 0.9522850 & 0.04016179\\\\\n",
       "\t chr1 & 100037799 & 100039086 & RTCA        & Jurkat & -0.0152314939 & 0.865631635 & 0.9522850 & 0.65039307\\\\\n",
       "\t chr1 & 100037799 & 100039086 & RTCA-AS1    & Jurkat & -0.0050515866 & 0.958832018 & 0.9522850 & 0.51834340\\\\\n",
       "\t chr1 & 100037799 & 100039086 & SASS6       & Jurkat & -0.0412282960 & 0.591352088 & 0.9522850 & 0.88395807\\\\\n",
       "\t chr1 & 100037799 & 100039086 & SLC30A7     & Jurkat &  0.1015604294 & 0.107326550 & 0.9522850 & 0.91021989\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &   978694 &   979118 & IL3RA     & Jurkat &  0.077058907 & 2.507832e-01 & 0.08815007 & 0.4126125\\\\\n",
       "\t chrX &   978694 &   979118 & LINC00106 & Jurkat &  0.015647190 & 8.616048e-01 & 0.08815007 & 0.3766093\\\\\n",
       "\t chrX &   978694 &   979118 & P2RY8     & Jurkat &  0.011162050 & 9.041405e-01 & 0.08815007 & 0.8463598\\\\\n",
       "\t chrX &   978694 &   979118 & PLCXD1    & Jurkat & -0.030797294 & 7.042059e-01 & 0.08815007 & 0.5504728\\\\\n",
       "\t chrX &   978694 &   979118 & PPP2R3B   & Jurkat & -0.087598973 & 1.790380e-01 & 0.08815007 & 0.5432950\\\\\n",
       "\t chrX &   978694 &   979118 & SHOX      & Jurkat & -0.015774162 & 8.603733e-01 & 0.08815007 & 0.1614447\\\\\n",
       "\t chrX &   978694 &   979118 & SLC25A6   & Jurkat & -0.103218806 & 1.004718e-01 & 0.08815007 & 0.9649653\\\\\n",
       "\t chrX &  9842372 &  9843073 & TBL1X     & Jurkat & -0.185228384 & 8.751534e-04 & 0.42109532 & 0.9236641\\\\\n",
       "\t chrX &  9842372 &  9843073 & WWC3      & Jurkat & -0.008494610 & 9.284790e-01 & 0.42109532 & 0.6253845\\\\\n",
       "\t chrX &  9873846 &  9874297 & TBL1X     & Jurkat & -0.076225881 & 2.570903e-01 & 0.14523145 & 0.9236641\\\\\n",
       "\t chrX &  9873846 &  9874297 & WWC3      & Jurkat & -0.071940118 & 2.911850e-01 & 0.14523145 & 0.6253845\\\\\n",
       "\t chrX &  9885186 &  9885487 & TBL1X     & Jurkat &  0.113045017 & 6.645980e-02 & 0.17157770 & 0.9236641\\\\\n",
       "\t chrX &  9885186 &  9885487 & WWC3      & Jurkat & -0.017925654 & 8.388762e-01 & 0.17157770 & 0.6253845\\\\\n",
       "\t chrX &   989869 &   990628 & AKAP17A   & Jurkat & -0.121657316 & 4.477509e-02 & 0.93286545 & 0.6066424\\\\\n",
       "\t chrX &   989869 &   990628 & ASMT      & Jurkat & -0.183437574 & 1.003947e-03 & 0.93286545 & 0.1983024\\\\\n",
       "\t chrX &   989869 &   990628 & ASMTL     & Jurkat & -0.083042940 & 2.081521e-01 & 0.93286545 & 0.5086590\\\\\n",
       "\t chrX &   989869 &   990628 & GTPBP6    & Jurkat &  0.122791536 & 4.240690e-02 & 0.93286545 & 0.5461433\\\\\n",
       "\t chrX &   989869 &   990628 & IL3RA     & Jurkat & -0.095782453 & 1.338724e-01 & 0.93286545 & 0.4126125\\\\\n",
       "\t chrX &   989869 &   990628 & LINC00106 & Jurkat & -0.098133681 & 1.225733e-01 & 0.93286545 & 0.3766093\\\\\n",
       "\t chrX &   989869 &   990628 & P2RY8     & Jurkat &  0.099039766 & 1.184062e-01 & 0.93286545 & 0.8463598\\\\\n",
       "\t chrX &   989869 &   990628 & PLCXD1    & Jurkat &  0.181000824 & 1.206355e-03 & 0.93286545 & 0.5504728\\\\\n",
       "\t chrX &   989869 &   990628 & PPP2R3B   & Jurkat & -0.241265309 & 5.232516e-06 & 0.93286545 & 0.5432950\\\\\n",
       "\t chrX &   989869 &   990628 & SHOX      & Jurkat &  0.068078130 & 3.237836e-01 & 0.93286545 & 0.1614447\\\\\n",
       "\t chrX &   989869 &   990628 & SLC25A6   & Jurkat & -0.215293892 & 6.904075e-05 & 0.93286545 & 0.9649653\\\\\n",
       "\t chrX &  9941883 &  9942743 & TBL1X     & Jurkat &  0.212400696 & 8.989957e-05 & 0.80635371 & 0.9236641\\\\\n",
       "\t chrX &  9941883 &  9942743 & WWC3      & Jurkat & -0.003399047 & 9.726253e-01 & 0.80635371 & 0.6253845\\\\\n",
       "\t chrX & 99603024 & 99603776 & PCDH19    & Jurkat &  0.110143858 & 7.541516e-02 & 0.49636660 & 0.3660134\\\\\n",
       "\t chrX & 99632013 & 99632219 & PCDH19    & Jurkat & -0.002562549 & 9.795295e-01 & 0.01992975 & 0.3660134\\\\\n",
       "\t chrX &  9995829 &  9996207 & TBL1X     & Jurkat & -0.029695418 & 7.159998e-01 & 0.47235540 & 0.9236641\\\\\n",
       "\t chrX &  9995829 &  9996207 & WWC3      & Jurkat & -0.028753695 & 7.261407e-01 & 0.47235540 & 0.6253845\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2590340 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | Correlation &lt;dbl&gt; | FDR &lt;dbl&gt; | VarQATAC &lt;dbl&gt; | VarQRNA &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  10001055 |  10001570 | APITD1-CORT | Jurkat | -0.0654033111 | 0.347569890 | 0.2452792 | 0.31787627 |\n",
       "| chr1 |  10001055 |  10001570 | CASZ1       | Jurkat |  0.0852744357 | 0.193500470 | 0.2452792 | 0.40526376 |\n",
       "| chr1 |  10001055 |  10001570 | CLSTN1      | Jurkat |  0.1068869545 | 0.086523359 | 0.2452792 | 0.84727128 |\n",
       "| chr1 |  10001055 |  10001570 | CTNNBIP1    | Jurkat |  0.1667059261 | 0.003349231 | 0.2452792 | 0.85564544 |\n",
       "| chr1 |  10001055 |  10001570 | DFFA        | Jurkat | -0.0537373593 | 0.459680288 | 0.2452792 | 0.65574798 |\n",
       "| chr1 |  10001055 |  10001570 | GPR157      | Jurkat |  0.0249170647 | 0.766729796 | 0.2452792 | 0.45909764 |\n",
       "| chr1 |  10001055 |  10001570 | H6PD        | Jurkat | -0.0666460691 | 0.336484757 | 0.2452792 | 0.48484676 |\n",
       "| chr1 |  10001055 |  10001570 | KIF1B       | Jurkat |  0.0204891917 | 0.812859646 | 0.2452792 | 0.85558847 |\n",
       "| chr1 |  10001055 |  10001570 | LINC01759   | Jurkat |  0.0471842099 | 0.527837922 | 0.2452792 | 0.07855759 |\n",
       "| chr1 |  10001055 |  10001570 | LZIC        | Jurkat |  0.0963173413 | 0.131237723 | 0.2452792 | 0.71425316 |\n",
       "| chr1 |  10001055 |  10001570 | MIR34AHG    | Jurkat |  0.0710284573 | 0.298724608 | 0.2452792 | 0.10755383 |\n",
       "| chr1 |  10001055 |  10001570 | NMNAT1      | Jurkat | -0.0998270702 | 0.114880217 | 0.2452792 | 0.43294976 |\n",
       "| chr1 |  10001055 |  10001570 | PEX14       | Jurkat |  0.0695721481 | 0.310952302 | 0.2452792 | 0.77093540 |\n",
       "| chr1 |  10001055 |  10001570 | PGD         | Jurkat |  0.0748075719 | 0.268079900 | 0.2452792 | 0.94371653 |\n",
       "| chr1 |  10001055 |  10001570 | PIK3CD      | Jurkat | -0.0221642985 | 0.795517474 | 0.2452792 | 0.63011279 |\n",
       "| chr1 |  10001055 |  10001570 | SLC25A33    | Jurkat |  0.0890276702 | 0.170520707 | 0.2452792 | 0.66759713 |\n",
       "| chr1 |  10001055 |  10001570 | SPSB1       | Jurkat |  0.0004905532 | 0.996168050 | 0.2452792 | 0.38874331 |\n",
       "| chr1 |  10001055 |  10001570 | TMEM201     | Jurkat |  0.0456741285 | 0.543999597 | 0.2452792 | 0.49652501 |\n",
       "| chr1 |  10001055 |  10001570 | UBE4B       | Jurkat | -0.0059055346 | 0.951501606 | 0.2452792 | 0.88486955 |\n",
       "| chr1 | 100037799 | 100039086 | AGL         | Jurkat |  0.0926134261 | 0.150284403 | 0.9522850 | 0.73071665 |\n",
       "| chr1 | 100037799 | 100039086 | CDC14A      | Jurkat | -0.0711974189 | 0.297326639 | 0.9522850 | 0.86003190 |\n",
       "| chr1 | 100037799 | 100039086 | DBT         | Jurkat |  0.0163183519 | 0.854955447 | 0.9522850 | 0.72877977 |\n",
       "| chr1 | 100037799 | 100039086 | DPH5        | Jurkat | -0.0103547298 | 0.911621564 | 0.9522850 | 0.72336789 |\n",
       "| chr1 | 100037799 | 100039086 | EXTL2       | Jurkat |  0.0778734593 | 0.244724054 | 0.9522850 | 0.90777031 |\n",
       "| chr1 | 100037799 | 100039086 | FRRS1       | Jurkat | -0.0188514558 | 0.829553184 | 0.9522850 | 0.56431583 |\n",
       "| chr1 | 100037799 | 100039086 | MFSD14A     | Jurkat |  0.1086712923 | 0.080277974 | 0.9522850 | 0.04016179 |\n",
       "| chr1 | 100037799 | 100039086 | RTCA        | Jurkat | -0.0152314939 | 0.865631635 | 0.9522850 | 0.65039307 |\n",
       "| chr1 | 100037799 | 100039086 | RTCA-AS1    | Jurkat | -0.0050515866 | 0.958832018 | 0.9522850 | 0.51834340 |\n",
       "| chr1 | 100037799 | 100039086 | SASS6       | Jurkat | -0.0412282960 | 0.591352088 | 0.9522850 | 0.88395807 |\n",
       "| chr1 | 100037799 | 100039086 | SLC30A7     | Jurkat |  0.1015604294 | 0.107326550 | 0.9522850 | 0.91021989 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |   978694 |   979118 | IL3RA     | Jurkat |  0.077058907 | 2.507832e-01 | 0.08815007 | 0.4126125 |\n",
       "| chrX |   978694 |   979118 | LINC00106 | Jurkat |  0.015647190 | 8.616048e-01 | 0.08815007 | 0.3766093 |\n",
       "| chrX |   978694 |   979118 | P2RY8     | Jurkat |  0.011162050 | 9.041405e-01 | 0.08815007 | 0.8463598 |\n",
       "| chrX |   978694 |   979118 | PLCXD1    | Jurkat | -0.030797294 | 7.042059e-01 | 0.08815007 | 0.5504728 |\n",
       "| chrX |   978694 |   979118 | PPP2R3B   | Jurkat | -0.087598973 | 1.790380e-01 | 0.08815007 | 0.5432950 |\n",
       "| chrX |   978694 |   979118 | SHOX      | Jurkat | -0.015774162 | 8.603733e-01 | 0.08815007 | 0.1614447 |\n",
       "| chrX |   978694 |   979118 | SLC25A6   | Jurkat | -0.103218806 | 1.004718e-01 | 0.08815007 | 0.9649653 |\n",
       "| chrX |  9842372 |  9843073 | TBL1X     | Jurkat | -0.185228384 | 8.751534e-04 | 0.42109532 | 0.9236641 |\n",
       "| chrX |  9842372 |  9843073 | WWC3      | Jurkat | -0.008494610 | 9.284790e-01 | 0.42109532 | 0.6253845 |\n",
       "| chrX |  9873846 |  9874297 | TBL1X     | Jurkat | -0.076225881 | 2.570903e-01 | 0.14523145 | 0.9236641 |\n",
       "| chrX |  9873846 |  9874297 | WWC3      | Jurkat | -0.071940118 | 2.911850e-01 | 0.14523145 | 0.6253845 |\n",
       "| chrX |  9885186 |  9885487 | TBL1X     | Jurkat |  0.113045017 | 6.645980e-02 | 0.17157770 | 0.9236641 |\n",
       "| chrX |  9885186 |  9885487 | WWC3      | Jurkat | -0.017925654 | 8.388762e-01 | 0.17157770 | 0.6253845 |\n",
       "| chrX |   989869 |   990628 | AKAP17A   | Jurkat | -0.121657316 | 4.477509e-02 | 0.93286545 | 0.6066424 |\n",
       "| chrX |   989869 |   990628 | ASMT      | Jurkat | -0.183437574 | 1.003947e-03 | 0.93286545 | 0.1983024 |\n",
       "| chrX |   989869 |   990628 | ASMTL     | Jurkat | -0.083042940 | 2.081521e-01 | 0.93286545 | 0.5086590 |\n",
       "| chrX |   989869 |   990628 | GTPBP6    | Jurkat |  0.122791536 | 4.240690e-02 | 0.93286545 | 0.5461433 |\n",
       "| chrX |   989869 |   990628 | IL3RA     | Jurkat | -0.095782453 | 1.338724e-01 | 0.93286545 | 0.4126125 |\n",
       "| chrX |   989869 |   990628 | LINC00106 | Jurkat | -0.098133681 | 1.225733e-01 | 0.93286545 | 0.3766093 |\n",
       "| chrX |   989869 |   990628 | P2RY8     | Jurkat |  0.099039766 | 1.184062e-01 | 0.93286545 | 0.8463598 |\n",
       "| chrX |   989869 |   990628 | PLCXD1    | Jurkat |  0.181000824 | 1.206355e-03 | 0.93286545 | 0.5504728 |\n",
       "| chrX |   989869 |   990628 | PPP2R3B   | Jurkat | -0.241265309 | 5.232516e-06 | 0.93286545 | 0.5432950 |\n",
       "| chrX |   989869 |   990628 | SHOX      | Jurkat |  0.068078130 | 3.237836e-01 | 0.93286545 | 0.1614447 |\n",
       "| chrX |   989869 |   990628 | SLC25A6   | Jurkat | -0.215293892 | 6.904075e-05 | 0.93286545 | 0.9649653 |\n",
       "| chrX |  9941883 |  9942743 | TBL1X     | Jurkat |  0.212400696 | 8.989957e-05 | 0.80635371 | 0.9236641 |\n",
       "| chrX |  9941883 |  9942743 | WWC3      | Jurkat | -0.003399047 | 9.726253e-01 | 0.80635371 | 0.6253845 |\n",
       "| chrX | 99603024 | 99603776 | PCDH19    | Jurkat |  0.110143858 | 7.541516e-02 | 0.49636660 | 0.3660134 |\n",
       "| chrX | 99632013 | 99632219 | PCDH19    | Jurkat | -0.002562549 | 9.795295e-01 | 0.01992975 | 0.3660134 |\n",
       "| chrX |  9995829 |  9996207 | TBL1X     | Jurkat | -0.029695418 | 7.159998e-01 | 0.47235540 | 0.9236641 |\n",
       "| chrX |  9995829 |  9996207 | WWC3      | Jurkat | -0.028753695 | 7.261407e-01 | 0.47235540 | 0.6253845 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene  CellType Correlation  \n",
       "1       chr1  10001055  10001570 APITD1-CORT Jurkat   -0.0654033111\n",
       "2       chr1  10001055  10001570 CASZ1       Jurkat    0.0852744357\n",
       "3       chr1  10001055  10001570 CLSTN1      Jurkat    0.1068869545\n",
       "4       chr1  10001055  10001570 CTNNBIP1    Jurkat    0.1667059261\n",
       "5       chr1  10001055  10001570 DFFA        Jurkat   -0.0537373593\n",
       "6       chr1  10001055  10001570 GPR157      Jurkat    0.0249170647\n",
       "7       chr1  10001055  10001570 H6PD        Jurkat   -0.0666460691\n",
       "8       chr1  10001055  10001570 KIF1B       Jurkat    0.0204891917\n",
       "9       chr1  10001055  10001570 LINC01759   Jurkat    0.0471842099\n",
       "10      chr1  10001055  10001570 LZIC        Jurkat    0.0963173413\n",
       "11      chr1  10001055  10001570 MIR34AHG    Jurkat    0.0710284573\n",
       "12      chr1  10001055  10001570 NMNAT1      Jurkat   -0.0998270702\n",
       "13      chr1  10001055  10001570 PEX14       Jurkat    0.0695721481\n",
       "14      chr1  10001055  10001570 PGD         Jurkat    0.0748075719\n",
       "15      chr1  10001055  10001570 PIK3CD      Jurkat   -0.0221642985\n",
       "16      chr1  10001055  10001570 SLC25A33    Jurkat    0.0890276702\n",
       "17      chr1  10001055  10001570 SPSB1       Jurkat    0.0004905532\n",
       "18      chr1  10001055  10001570 TMEM201     Jurkat    0.0456741285\n",
       "19      chr1  10001055  10001570 UBE4B       Jurkat   -0.0059055346\n",
       "20      chr1 100037799 100039086 AGL         Jurkat    0.0926134261\n",
       "21      chr1 100037799 100039086 CDC14A      Jurkat   -0.0711974189\n",
       "22      chr1 100037799 100039086 DBT         Jurkat    0.0163183519\n",
       "23      chr1 100037799 100039086 DPH5        Jurkat   -0.0103547298\n",
       "24      chr1 100037799 100039086 EXTL2       Jurkat    0.0778734593\n",
       "25      chr1 100037799 100039086 FRRS1       Jurkat   -0.0188514558\n",
       "26      chr1 100037799 100039086 MFSD14A     Jurkat    0.1086712923\n",
       "27      chr1 100037799 100039086 RTCA        Jurkat   -0.0152314939\n",
       "28      chr1 100037799 100039086 RTCA-AS1    Jurkat   -0.0050515866\n",
       "29      chr1 100037799 100039086 SASS6       Jurkat   -0.0412282960\n",
       "30      chr1 100037799 100039086 SLC30A7     Jurkat    0.1015604294\n",
       "⋮       ⋮    ⋮         ⋮         ⋮           ⋮        ⋮            \n",
       "2590311 chrX   978694    979118  IL3RA       Jurkat    0.077058907 \n",
       "2590312 chrX   978694    979118  LINC00106   Jurkat    0.015647190 \n",
       "2590313 chrX   978694    979118  P2RY8       Jurkat    0.011162050 \n",
       "2590314 chrX   978694    979118  PLCXD1      Jurkat   -0.030797294 \n",
       "2590315 chrX   978694    979118  PPP2R3B     Jurkat   -0.087598973 \n",
       "2590316 chrX   978694    979118  SHOX        Jurkat   -0.015774162 \n",
       "2590317 chrX   978694    979118  SLC25A6     Jurkat   -0.103218806 \n",
       "2590318 chrX  9842372   9843073  TBL1X       Jurkat   -0.185228384 \n",
       "2590319 chrX  9842372   9843073  WWC3        Jurkat   -0.008494610 \n",
       "2590320 chrX  9873846   9874297  TBL1X       Jurkat   -0.076225881 \n",
       "2590321 chrX  9873846   9874297  WWC3        Jurkat   -0.071940118 \n",
       "2590322 chrX  9885186   9885487  TBL1X       Jurkat    0.113045017 \n",
       "2590323 chrX  9885186   9885487  WWC3        Jurkat   -0.017925654 \n",
       "2590324 chrX   989869    990628  AKAP17A     Jurkat   -0.121657316 \n",
       "2590325 chrX   989869    990628  ASMT        Jurkat   -0.183437574 \n",
       "2590326 chrX   989869    990628  ASMTL       Jurkat   -0.083042940 \n",
       "2590327 chrX   989869    990628  GTPBP6      Jurkat    0.122791536 \n",
       "2590328 chrX   989869    990628  IL3RA       Jurkat   -0.095782453 \n",
       "2590329 chrX   989869    990628  LINC00106   Jurkat   -0.098133681 \n",
       "2590330 chrX   989869    990628  P2RY8       Jurkat    0.099039766 \n",
       "2590331 chrX   989869    990628  PLCXD1      Jurkat    0.181000824 \n",
       "2590332 chrX   989869    990628  PPP2R3B     Jurkat   -0.241265309 \n",
       "2590333 chrX   989869    990628  SHOX        Jurkat    0.068078130 \n",
       "2590334 chrX   989869    990628  SLC25A6     Jurkat   -0.215293892 \n",
       "2590335 chrX  9941883   9942743  TBL1X       Jurkat    0.212400696 \n",
       "2590336 chrX  9941883   9942743  WWC3        Jurkat   -0.003399047 \n",
       "2590337 chrX 99603024  99603776  PCDH19      Jurkat    0.110143858 \n",
       "2590338 chrX 99632013  99632219  PCDH19      Jurkat   -0.002562549 \n",
       "2590339 chrX  9995829   9996207  TBL1X       Jurkat   -0.029695418 \n",
       "2590340 chrX  9995829   9996207  WWC3        Jurkat   -0.028753695 \n",
       "        FDR          VarQATAC   VarQRNA   \n",
       "1       0.347569890  0.2452792  0.31787627\n",
       "2       0.193500470  0.2452792  0.40526376\n",
       "3       0.086523359  0.2452792  0.84727128\n",
       "4       0.003349231  0.2452792  0.85564544\n",
       "5       0.459680288  0.2452792  0.65574798\n",
       "6       0.766729796  0.2452792  0.45909764\n",
       "7       0.336484757  0.2452792  0.48484676\n",
       "8       0.812859646  0.2452792  0.85558847\n",
       "9       0.527837922  0.2452792  0.07855759\n",
       "10      0.131237723  0.2452792  0.71425316\n",
       "11      0.298724608  0.2452792  0.10755383\n",
       "12      0.114880217  0.2452792  0.43294976\n",
       "13      0.310952302  0.2452792  0.77093540\n",
       "14      0.268079900  0.2452792  0.94371653\n",
       "15      0.795517474  0.2452792  0.63011279\n",
       "16      0.170520707  0.2452792  0.66759713\n",
       "17      0.996168050  0.2452792  0.38874331\n",
       "18      0.543999597  0.2452792  0.49652501\n",
       "19      0.951501606  0.2452792  0.88486955\n",
       "20      0.150284403  0.9522850  0.73071665\n",
       "21      0.297326639  0.9522850  0.86003190\n",
       "22      0.854955447  0.9522850  0.72877977\n",
       "23      0.911621564  0.9522850  0.72336789\n",
       "24      0.244724054  0.9522850  0.90777031\n",
       "25      0.829553184  0.9522850  0.56431583\n",
       "26      0.080277974  0.9522850  0.04016179\n",
       "27      0.865631635  0.9522850  0.65039307\n",
       "28      0.958832018  0.9522850  0.51834340\n",
       "29      0.591352088  0.9522850  0.88395807\n",
       "30      0.107326550  0.9522850  0.91021989\n",
       "⋮       ⋮            ⋮          ⋮         \n",
       "2590311 2.507832e-01 0.08815007 0.4126125 \n",
       "2590312 8.616048e-01 0.08815007 0.3766093 \n",
       "2590313 9.041405e-01 0.08815007 0.8463598 \n",
       "2590314 7.042059e-01 0.08815007 0.5504728 \n",
       "2590315 1.790380e-01 0.08815007 0.5432950 \n",
       "2590316 8.603733e-01 0.08815007 0.1614447 \n",
       "2590317 1.004718e-01 0.08815007 0.9649653 \n",
       "2590318 8.751534e-04 0.42109532 0.9236641 \n",
       "2590319 9.284790e-01 0.42109532 0.6253845 \n",
       "2590320 2.570903e-01 0.14523145 0.9236641 \n",
       "2590321 2.911850e-01 0.14523145 0.6253845 \n",
       "2590322 6.645980e-02 0.17157770 0.9236641 \n",
       "2590323 8.388762e-01 0.17157770 0.6253845 \n",
       "2590324 4.477509e-02 0.93286545 0.6066424 \n",
       "2590325 1.003947e-03 0.93286545 0.1983024 \n",
       "2590326 2.081521e-01 0.93286545 0.5086590 \n",
       "2590327 4.240690e-02 0.93286545 0.5461433 \n",
       "2590328 1.338724e-01 0.93286545 0.4126125 \n",
       "2590329 1.225733e-01 0.93286545 0.3766093 \n",
       "2590330 1.184062e-01 0.93286545 0.8463598 \n",
       "2590331 1.206355e-03 0.93286545 0.5504728 \n",
       "2590332 5.232516e-06 0.93286545 0.5432950 \n",
       "2590333 3.237836e-01 0.93286545 0.1614447 \n",
       "2590334 6.904075e-05 0.93286545 0.9649653 \n",
       "2590335 8.989957e-05 0.80635371 0.9236641 \n",
       "2590336 9.726253e-01 0.80635371 0.6253845 \n",
       "2590337 7.541516e-02 0.49636660 0.3660134 \n",
       "2590338 9.795295e-01 0.01992975 0.3660134 \n",
       "2590339 7.159998e-01 0.47235540 0.9236641 \n",
       "2590340 7.261407e-01 0.47235540 0.6253845 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(obj.ArchR,\n",
    "        paste(dir.output,\"obj.ArchR.rds\",sep = \"/\"))\n",
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = celltype\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"Correlation\",\n",
    "                         \"FDR\",\n",
    "                         \"VarQATAC\",\n",
    "                         \"VarQRNA\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "pairs.E2G.res\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "4c909ed5-7321-4e69-8208-1544b371da6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] nabor_0.5.0                       uwot_0.1.16                      \n",
       " [3] Rsamtools_2.18.0                  BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] dplyr_1.1.4                       genomation_1.34.0                \n",
       "[11] Signac_1.13.0                     Seurat_5.0.3                     \n",
       "[13] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[15] rhdf5_2.44.0                      SummarizedExperiment_1.30.2      \n",
       "[17] Biobase_2.60.0                    MatrixGenerics_1.12.3            \n",
       "[19] Rcpp_1.0.12                       Matrix_1.6-5                     \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               matrixStats_1.3.0                \n",
       "[27] data.table_1.15.2                 stringr_1.5.1                    \n",
       "[29] plyr_1.8.9                        magrittr_2.0.3                   \n",
       "[31] ggplot2_3.5.1                     gtable_0.3.5                     \n",
       "[33] gtools_3.9.5                      gridExtra_2.3                    \n",
       "[35] ArchR_1.0.2                      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22         splines_4.3.3            later_1.3.2             \n",
       "  [4] pbdZMQ_0.3-11            BiocIO_1.10.0            bitops_1.0-7            \n",
       "  [7] tibble_3.2.1             polyclip_1.10-6          XML_3.99-0.16.1         \n",
       " [10] fastDummies_1.7.3        lifecycle_1.0.4          vroom_1.6.5             \n",
       " [13] globals_0.16.3           lattice_0.22-6           MASS_7.3-60             \n",
       " [16] plotly_4.10.4            plotrix_3.8-4            yaml_2.3.8              \n",
       " [19] httpuv_1.6.15            sctransform_0.4.1        spam_2.10-0             \n",
       " [22] spatstat.sparse_3.0-3    reticulate_1.36.1        cowplot_1.1.3           \n",
       " [25] pbapply_1.7-2            RColorBrewer_1.1-3       abind_1.4-5             \n",
       " [28] zlibbioc_1.48.0          Rtsne_0.17               purrr_1.0.2             \n",
       " [31] RCurl_1.98-1.14          GenomeInfoDbData_1.2.11  ggrepel_0.9.5           \n",
       " [34] irlba_2.3.5.1            listenv_0.9.1            spatstat.utils_3.0-4    \n",
       " [37] goftest_1.2-3            RSpectra_0.16-1          spatstat.random_3.2-3   \n",
       " [40] fitdistrplus_1.1-11      parallelly_1.37.1        leiden_0.4.3.1          \n",
       " [43] codetools_0.2-20         DelayedArray_0.26.7      RcppRoll_0.3.0          \n",
       " [46] tidyselect_1.2.1         farver_2.1.1             base64enc_0.1-3         \n",
       " [49] spatstat.explore_3.2-6   GenomicAlignments_1.36.0 jsonlite_1.8.8          \n",
       " [52] progressr_0.14.0         ggridges_0.5.6           survival_3.6-4          \n",
       " [55] tools_4.3.3              ica_1.0-3                glue_1.7.0              \n",
       " [58] IRdisplay_1.1            withr_3.0.0              fastmap_1.1.1           \n",
       " [61] rhdf5filters_1.12.1      fansi_1.0.6              digest_0.6.35           \n",
       " [64] R6_2.5.1                 mime_0.12                seqPattern_1.34.0       \n",
       " [67] colorspace_2.1-0         Cairo_1.6-2              scattermore_1.2         \n",
       " [70] tensor_1.5               spatstat.data_3.0-4      utf8_1.2.4              \n",
       " [73] tidyr_1.3.1              generics_0.1.3           httr_1.4.7              \n",
       " [76] htmlwidgets_1.6.4        S4Arrays_1.2.1           pkgconfig_2.0.3         \n",
       " [79] impute_1.76.0            lmtest_0.9-40            htmltools_0.5.8.1       \n",
       " [82] dotCall64_1.1-1          scales_1.3.0             png_0.1-8               \n",
       " [85] tzdb_0.4.0               reshape2_1.4.4           rjson_0.2.21            \n",
       " [88] uuid_1.2-0               nlme_3.1-164             repr_1.1.7              \n",
       " [91] zoo_1.8-12               KernSmooth_2.23-22       miniUI_0.1.1.1          \n",
       " [94] restfulr_0.0.15          pillar_1.9.0             vctrs_0.6.5             \n",
       " [97] RANN_2.6.1               promises_1.3.0           xtable_1.8-4            \n",
       "[100] cluster_2.1.6            evaluate_0.23            readr_2.1.5             \n",
       "[103] cli_3.6.2                compiler_4.3.3           rlang_1.1.3             \n",
       "[106] crayon_1.5.2             future.apply_1.11.2      labeling_0.4.3          \n",
       "[109] stringi_1.8.3            gridBase_0.4-7           viridisLite_0.4.2       \n",
       "[112] deldir_2.0-4             BiocParallel_1.36.0      munsell_0.5.1           \n",
       "[115] lazyeval_0.2.2           spatstat.geom_3.2-9      IRkernel_1.3.2          \n",
       "[118] RcppHNSW_0.6.0           hms_1.1.3                patchwork_1.2.0         \n",
       "[121] bit64_4.0.5              future_1.33.2            Rhdf5lib_1.22.1         \n",
       "[124] shiny_1.8.1.1            ROCR_1.0-11              igraph_2.0.3            \n",
       "[127] fastmatch_1.1-4          bit_4.0.5               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7690cee-49f1-4886-ae66-0b63a43ce524",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
