{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8466b35-cae5-4f62-8138-37b2de484a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.2\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : grid v4.3.3\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "\tLoading Package : gtools v3.9.5\n",
      "\n",
      "\tLoading Package : gtable v0.3.5\n",
      "\n",
      "\tLoading Package : ggplot2 v3.5.1\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "\tLoading Package : plyr v1.8.9\n",
      "\n",
      "\tLoading Package : stringr v1.5.1\n",
      "\n",
      "\tLoading Package : data.table v1.15.2\n",
      "\n",
      "\tLoading Package : matrixStats v1.3.0\n",
      "\n",
      "\tLoading Package : S4Vectors v0.40.2\n",
      "\n",
      "\tLoading Package : GenomicRanges v1.54.1\n",
      "\n",
      "\tLoading Package : BiocGenerics v0.48.1\n",
      "\n",
      "\tLoading Package : Matrix v1.6.5\n",
      "\n",
      "\tLoading Package : Rcpp v1.0.12\n",
      "\n",
      "\tLoading Package : SummarizedExperiment v1.30.2\n",
      "\n",
      "\tLoading Package : rhdf5 v2.44.0\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(magrittr)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e85020d7-d358-492d-9358-c81270f4b624",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1bec6690-52ec-450d-bba7-7d4528a185cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.seurat = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_IGVF/1.prepare_data/1.import_IGVF_processed_data.240508/obj.seurat.rds\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_IGVF/K562/Kendall/Pairs.tsv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_IGVF/3.Genome_wide_prediction/ArchR_250226/ArchR.250226/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "340dac32-44ac-4231-b964-b5f3b9e2dbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16\n",
    "maxDist = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da973385-4c81-4b1c-9669-cf5933b23a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 16.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRThreads(threads = n.cores) \n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeb26496-d3b9-4778-92ed-9c63a5018aef",
   "metadata": {},
   "source": [
    "Import seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c9b19a09-be95-4d9a-bdc3-0530980d4757",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.seurat = readRDS(path.seurat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86be3e66-c954-4620-9e64-c6a1c0fcb84c",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6f363969-bf0b-45ce-858c-a8cb64cc210a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3538ffdc-4fed-48eb-9c9d-86805a91cf06",
   "metadata": {},
   "source": [
    "Create an ArchRProject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0300f485-fafa-4cef-b9d0-534108ec6f95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 2\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/IGVF_240503/230707_IGVF10_McGinnis/ATAC/CEL\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"c69fb4ea81ca5efb0834a597ffe6e12c\" \"abb247b297226d16f586a20ae0771d0a\"\n",
      "  .. ..@ cells: Named chr [1:2723] \"AAACCAGGTCACAGGG-1\" \"AAACCGCCAAGGAATA-1\" \"AAACCGCCAGCCGGTA-1\" \"AAACCGGTCAAGGAAG-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:2723] \"e10l1_AAACCAGGTCACAGGG-1\" \"e10l1_AAACCGCCAAGGAATA-1\" \"e10l1_AAACCGCCAGCCGGTA-1\" \"e10l1_AAACCGGTCAAGGAAG-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/IGVF_240503/230707_IGVF10_McGinnis/ATAC/CEL\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"2f8b867830dcf7eb020d0be82d029441\" \"dcdf4989f54c370141b644b44292d116\"\n",
      "  .. ..@ cells: Named chr [1:3543] \"AAACAAGCATGAGGGA-1\" \"AAACGAACAACGTGGT-1\" \"AAACGAACATCAATCC-1\" \"AAACGGGCATTTCGCG-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:3543] \"e10l2_AAACAAGCATGAGGGA-1\" \"e10l2_AAACGAACAACGTGGT-1\" \"e10l2_AAACGAACATCAATCC-1\" \"e10l2_AAACGGGCATTTCGCG-1\" ...\n"
     ]
    }
   ],
   "source": [
    "str(Fragments(obj.seurat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1753ffaa-89e2-43bf-bcf6-19937782d63a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-createArrows-2a155ea21d2a1-Date-2025-02-26_Time-14-44-26.02189.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Cleaning Temporary Files\n",
      "\n",
      "2025-02-26 14:44:26.577383 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-createArrows-2a155ea21d2a1-Date-2025-02-26_Time-14-44-26.02189.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ArrowFiles <- createArrowFiles(\n",
    "  inputFiles = c(Fragments(obj.seurat)[[1]]@path,\n",
    "                Fragments(obj.seurat)[[2]]@path),\n",
    "  sampleNames = c(\"e10l1\",\n",
    "                  \"e10l2\"),\n",
    "  validBarcodes = list(e10l1 = unname(Fragments(obj.seurat)[[1]]@cells),\n",
    "                       e10l2 = unname(Fragments(obj.seurat)[[2]]@cells)),\n",
    "  addTileMat = T,\n",
    "  addGeneScoreMat = F,\n",
    "  force = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ada3db45-d050-4659-a7cf-2891e04b1249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'e10l2.arrow'</li><li>'e10l1.arrow'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'e10l2.arrow'\n",
       "\\item 'e10l1.arrow'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'e10l2.arrow'\n",
       "2. 'e10l1.arrow'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"e10l2.arrow\" \"e10l1.arrow\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ArrowFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5ba28283-d488-4982-8d4c-745fc3f59365",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"ArchR\",sep=\"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aa217507-8323-43bf-b23c-0ec25ee33c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Validating Arrows...\n",
      "\n",
      "Getting SampleNames...\n",
      "\n",
      "\n",
      "\n",
      "Copying ArrowFiles to Ouptut Directory! If you want to save disk space set copyArrows = FALSE\n",
      "\n",
      "1 \n",
      "2 \n",
      "\n",
      "\n",
      "Getting Cell Metadata...\n",
      "\n",
      "\n",
      "\n",
      "Merging Cell Metadata...\n",
      "\n",
      "Initializing ArchRProject...\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR <- ArchRProject(\n",
    "  ArrowFiles = ArrowFiles, \n",
    "  outputDirectory = paste(dir.output,\"ArchR\",sep=\"/\"),\n",
    "  copyArrows = TRUE \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f9795403-545f-4385-aa38-2c70b1b2fb21",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: ArchRProject \n",
       "outputDirectory: /maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_IGVF/3.Genome_wide_prediction/ArchR_250226/ArchR.250226/ArchR \n",
       "samples(2): e10l2 e10l1\n",
       "sampleColData names(1): ArrowFiles\n",
       "cellColData names(13): Sample TSSEnrichment ... nDiFrags BlacklistRatio\n",
       "numberOfCells(1): 6258\n",
       "medianTSS(1): 13.809\n",
       "medianFrags(1): 11558.5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obj.ArchR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ee876dd5-0b22-46fd-9015-7e22144c5024",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking Inputs...\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addIterativeLSI-2a155e24ef83ad-Date-2025-02-26_Time-14-59-23.574368.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 14:59:24.574919 : Computing Total Across All Features, 0.006 mins elapsed.\n",
      "\n",
      "2025-02-26 14:59:27.509902 : Computing Top Features, 0.055 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-26 14:59:31.318895 : Running LSI (1 of 2) on Top Features, 0.118 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-02-26 14:59:31.333432 : Creating Partial Matrix, 0.119 mins elapsed.\n",
      "\n",
      "2025-02-26 14:59:47.704025 : Computing LSI, 0.391 mins elapsed.\n",
      "\n",
      "2025-02-26 15:00:37.550515 : Identifying Clusters, 1.222 mins elapsed.\n",
      "\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "2025-02-26 15:00:45.558342 : Identified 6 Clusters, 1.356 mins elapsed.\n",
      "\n",
      "2025-02-26 15:00:45.624111 : Saving LSI Iteration, 1.357 mins elapsed.\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************************************************************\n",
      "2025-02-26 15:01:02.44543 : ERROR Found in .saveIteration for  \n",
      "LogFile = ArchRLogs/ArchR-addIterativeLSI-2a155e24ef83ad-Date-2025-02-26_Time-14-59-23.574368.log\n",
      "\n",
      "<simpleError in g$grobs[[legend]]: no such index at level 2\n",
      ">\n",
      "\n",
      "************************************************************\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-26 15:01:02.447441 : Creating Cluster Matrix on the total Group Features, 1.637 mins elapsed.\n",
      "\n",
      "2025-02-26 15:01:19.468551 : Computing Variable Features, 1.921 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-26 15:01:19.620102 : Running LSI (2 of 2) on Variable Features, 1.923 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-02-26 15:01:19.646132 : Creating Partial Matrix, 1.924 mins elapsed.\n",
      "\n",
      "2025-02-26 15:01:33.603912 : Computing LSI, 2.156 mins elapsed.\n",
      "\n",
      "2025-02-26 15:02:08.298721 : Finished Running IterativeLSI, 2.735 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addIterativeLSI(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    useMatrix = \"TileMatrix\", \n",
    "    name = \"IterativeLSI\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2c1d334b-7fd7-425a-bb8d-7be4cbedd3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.ArchR = addPeakSet(obj.ArchR,\n",
    "                       peakSet = reduce(pairs.E2G))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e25c14db-eb4e-4b94-8862-eb34246a365c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeakMatrix-2a155e72be8352-Date-2025-02-26_Time-15-02-11.887715.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 15:02:12.257862 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeakMatrix-2a155e72be8352-Date-2025-02-26_Time-15-02-11.887715.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addPeakMatrix(obj.ArchR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "aa6d686e-4205-4eda-8d76-ed4d9e3e30e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA.Xu = ArchR:::.importFM(featureMatrix = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_230720/ArrayExpress/cellranger_res/EMTAB11264_01/outs/filtered_feature_bc_matrix.h5\",\n",
    "                             name = \"K562_Xu\",\n",
    "                             featureType = \"Gene Expression\")\n",
    "rowdata.Xu = rowData(seRNA.Xu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d843eb16-9fed-4146-b76e-0fb053f1557f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene.intersect = intersect(rownames(obj.seurat@assays$RNA$counts),\n",
    "                           rowdata.Xu$name)\n",
    "rowdata.intersect = rowdata.Xu[gene.intersect,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fdf69211-b595-4599-ac8d-1c832875c062",
   "metadata": {},
   "outputs": [],
   "source": [
    "rr = GRanges(paste0(rowdata.intersect$interval))\n",
    "mcols(rr) <- rowdata.intersect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1e8f8d64-f6be-43c2-8e76-ea1cf4ad0453",
   "metadata": {},
   "outputs": [],
   "source": [
    "count.rna = obj.seurat@assays$RNA$counts[gene.intersect,]\n",
    "colnames(count.rna) = gsub(\"e10l1_\",\"e10l1#\",colnames(count.rna))\n",
    "colnames(count.rna) = gsub(\"e10l2_\",\"e10l2#\",colnames(count.rna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "174a97a2-ab68-478a-9085-8e8fbc3bac29",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA <- SummarizedExperiment(assays = SimpleList(counts = count.rna), \n",
    "                              rowRanges = rr)\n",
    "seRNA = sort.GenomicRanges(sortSeqlevels(seRNA), \n",
    "                           ignore.strand = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4656d98c-a9cc-486a-afaa-fe59ca6ef055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 39032 6266 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(39032): MIR1302-2HG WASH7P ... CTBP2P1 PRODH\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(6266): e10l1#AAACCAGGTCACAGGG-1 e10l1#AAACCGCCAAGGAATA-1 ...\n",
       "  e10l2#TTTGTGAGTTAGCGTT-1 e10l2#TTTGTGAGTTCCGGAC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7422efe-875c-4694-8efb-c9f0aee359bc",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "34e9e7c2-76b1-45f1-8eef-55881d74844c",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "43713237-7b10-45ae-b074-d826cdd8d555",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biostrings’:\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seRNA.rename = map_gene_names(seRNA,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "65cb2097-9075-4ee4-9e54-dd288dc50736",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addGeneExpressionMatrix-2a155ee821ac8-Date-2025-02-26_Time-15-05-24.811861.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Overlap w/ scATAC = 1\n",
      "\n",
      "2025-02-26 15:05:25.313453 : \n",
      "\n",
      "Overlap Per Sample w/ scATAC : e10l1=2719,e10l2=3539\n",
      "\n",
      "2025-02-26 15:05:25.328288 : \n",
      "\n",
      "2025-02-26 15:05:26.85532 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addGeneExpressionMatrix-2a155ee821ac8-Date-2025-02-26_Time-15-05-24.811861.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addGeneExpressionMatrix(obj.ArchR,\n",
    "                                    seRNA = seRNA.rename,\n",
    "                                    strictMatch = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4318703-a143-40c0-b159-274d05033bbc",
   "metadata": {},
   "source": [
    "Run ArchR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "47c8fd5f-19ec-46e6-af87-57502c1e1ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeak2GeneLinks-2a155eab85b92-Date-2025-02-26_Time-15-07-38.554256.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 15:07:39.02847 : Getting Available Matrices, 0.008 mins elapsed.\n",
      "\n",
      "2025-02-26 15:07:40.088628 : Filtered Low Prediction Score Cells (0 of 6258, 0), 0.007 mins elapsed.\n",
      "\n",
      "2025-02-26 15:07:40.388388 : Computing KNN, 0.012 mins elapsed.\n",
      "\n",
      "2025-02-26 15:07:40.788683 : Identifying Non-Overlapping KNN pairs, 0.019 mins elapsed.\n",
      "\n",
      "2025-02-26 15:07:42.482666 : Identified 495 Groupings!, 0.047 mins elapsed.\n",
      "\n",
      "2025-02-26 15:07:42.535873 : Getting Group RNA Matrix, 0.048 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:07.739378 : Getting Group ATAC Matrix, 0.468 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:33.699848 : Normalizing Group Matrices, 0.9 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:40.407774 : Finding Peak Gene Pairings, 1.012 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:41.181772 : Computing Correlations, 1.025 mins elapsed.\n",
      "\n",
      "2025-02-26 15:08:56.80782 : Completed Peak2Gene Correlations!, 1.286 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeak2GeneLinks-2a155eab85b92-Date-2025-02-26_Time-15-07-38.554256.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "obj.ArchR = addPeak2GeneLinks(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    reducedDims = \"IterativeLSI\",\n",
    "    useMatrix = \"GeneExpressionMatrix\",\n",
    "    maxDist = maxDist,\n",
    "    threads = n.cores\n",
    ")\n",
    "end_time <- Sys.time()\n",
    "execution_time <- end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "128d0065-5f22-495b-b00d-743fc9843337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 1.304804 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "execution_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "36bd2df2-6b32-4b81-bd4f-39ef8913b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g = metadata(obj.ArchR@peakSet)$Peak2GeneLinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e7730413-8fd4-4952-bb13-0df7a3b15b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata(p2g)[[1]]$PeakName = paste(seqnames(metadata(p2g)[[1]]),\n",
    "                                    ranges(metadata(p2g)[[1]]),sep=\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "aeea3810-6735-4aff-9bd7-1697bef979d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g$PairName = paste(metadata(p2g)[[1]][p2g$idxATAC]$PeakName,\n",
    "                     metadata(p2g)[[2]][p2g$idxRNA]$name,\n",
    "                     sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "87f975ec-6ba4-4696-ad6c-2af65da7aa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 3445938 rows and 7 columns\n",
       "                                idxATAC    idxRNA Correlation       FDR\n",
       "                              <integer> <integer>   <numeric> <numeric>\n",
       "chr1-10028-10426_OR4F5                1         1  -0.0311087 0.6704738\n",
       "chr1-115520-115920_OR4F5              2         1   0.0749592 0.2312145\n",
       "chr1-138922-139504_OR4F5              3         1   0.0713042 0.2591966\n",
       "chr1-180739-181862_OR4F5              4         1   0.0485509 0.4744507\n",
       "chr1-267886-268131_OR4F5              5         1   0.1228924 0.0312106\n",
       "...                                 ...       ...         ...       ...\n",
       "chrX-156002453-156003206_IL9R    153518     18572  0.13993306 0.0120360\n",
       "chrX-156016531-156016738_IL9R    153519     18572  0.01721796 0.8277481\n",
       "chrX-156019803-156020136_IL9R    153520     18572 -0.03089001 0.6729810\n",
       "chrX-156029672-156030380_IL9R    153521     18572 -0.00462169 0.9572665\n",
       "chrX-156030503-156030833_IL9R    153522     18572 -0.12875263 0.0228307\n",
       "                               VarQATAC   VarQRNA               PairName\n",
       "                              <numeric> <numeric>            <character>\n",
       "chr1-10028-10426_OR4F5         0.497740  0.237454 chr1-10028-10426_OR4F5\n",
       "chr1-115520-115920_OR4F5       0.714634  0.237454 chr1-115520-115920_O..\n",
       "chr1-138922-139504_OR4F5       0.411107  0.237454 chr1-138922-139504_O..\n",
       "chr1-180739-181862_OR4F5       0.759513  0.237454 chr1-180739-181862_O..\n",
       "chr1-267886-268131_OR4F5       0.637466  0.237454 chr1-267886-268131_O..\n",
       "...                                 ...       ...                    ...\n",
       "chrX-156002453-156003206_IL9R 0.7800836  0.203909 chrX-156002453-15600..\n",
       "chrX-156016531-156016738_IL9R 0.0709345  0.203909 chrX-156016531-15601..\n",
       "chrX-156019803-156020136_IL9R 0.1405987  0.203909 chrX-156019803-15602..\n",
       "chrX-156029672-156030380_IL9R 0.5796238  0.203909 chrX-156029672-15603..\n",
       "chrX-156030503-156030833_IL9R 0.3307995  0.203909 chrX-156030503-15603.."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(p2g) = p2g$PairName\n",
    "p2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c7209a32-3904-4734-93af-9c22049f652a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = pairs.E2G[pairs.E2G$PairName %in% p2g$PairName]\n",
    "mcols(pairs.E2G.res)[,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")] = p2g[pairs.E2G.res$PairName,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1ab4ed70-ce9e-4857-bd1b-63a81d27a401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2821838 ranges and 7 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10002583-10003163      * | APITD1-CORT\n",
       "        [2]     chr1 10002583-10003163      * |       CASZ1\n",
       "        [3]     chr1 10002583-10003163      * |      CLSTN1\n",
       "        [4]     chr1 10002583-10003163      * |    CTNNBIP1\n",
       "        [5]     chr1 10002583-10003163      * |        DFFA\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [2821834]     chrX   9940388-9940697      * |        WWC3\n",
       "  [2821835]     chrX   9995829-9996183      * |       CLCN4\n",
       "  [2821836]     chrX   9995829-9996183      * |       FAM9B\n",
       "  [2821837]     chrX   9995829-9996183      * |       TBL1X\n",
       "  [2821838]     chrX   9995829-9996183      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-10002583-10003163 chr1-10002583-100031..  0.01701014\n",
       "        [2] chr1-10002583-10003163 chr1-10002583-100031..  0.01680164\n",
       "        [3] chr1-10002583-10003163 chr1-10002583-100031.. -0.02066524\n",
       "        [4] chr1-10002583-10003163 chr1-10002583-100031.. -0.00684081\n",
       "        [5] chr1-10002583-10003163 chr1-10002583-100031.. -0.00312169\n",
       "        ...                    ...                    ...         ...\n",
       "  [2821834]   chrX-9940388-9940697 chrX-9940388-9940697..  -0.2150580\n",
       "  [2821835]   chrX-9995829-9996183 chrX-9995829-9996183..   0.0630113\n",
       "  [2821836]   chrX-9995829-9996183 chrX-9995829-9996183..  -0.0419168\n",
       "  [2821837]   chrX-9995829-9996183 chrX-9995829-9996183..  -0.0222846\n",
       "  [2821838]   chrX-9995829-9996183 chrX-9995829-9996183..  -0.0163114\n",
       "                    FDR  VarQATAC   VarQRNA\n",
       "              <numeric> <numeric> <numeric>\n",
       "        [1]    0.830043  0.260582  0.294422\n",
       "        [2]    0.832309  0.260582  0.277407\n",
       "        [3]    0.789622  0.260582  0.850474\n",
       "        [4]    0.935920  0.260582  0.653618\n",
       "        [5]    0.971442  0.260582  0.809067\n",
       "        ...         ...       ...       ...\n",
       "  [2821834] 3.25486e-05 0.0124998  0.498331\n",
       "  [2821835] 3.30070e-01 0.4861127  0.793506\n",
       "  [2821836] 5.47389e-01 0.4861127  0.492623\n",
       "  [2821837] 7.71454e-01 0.4861127  0.966240\n",
       "  [2821838] 8.37664e-01 0.4861127  0.498331\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46485351-50d7-461a-bf09-d795d5bdab84",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6d424e21-6a41-4422-b76c-6f1c0bda2222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2821838 ranges and 7 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10002583-10003163      * | APITD1-CORT\n",
       "        [2]     chr1 10002583-10003163      * |       CASZ1\n",
       "        [3]     chr1 10002583-10003163      * |      CLSTN1\n",
       "        [4]     chr1 10002583-10003163      * |    CTNNBIP1\n",
       "        [5]     chr1 10002583-10003163      * |        DFFA\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [2821834]     chrX   9940388-9940697      * |        WWC3\n",
       "  [2821835]     chrX   9995829-9996183      * |       CLCN4\n",
       "  [2821836]     chrX   9995829-9996183      * |       FAM9B\n",
       "  [2821837]     chrX   9995829-9996183      * |       TBL1X\n",
       "  [2821838]     chrX   9995829-9996183      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-10002583-10003163 chr1-10002583-100031..  0.01701014\n",
       "        [2] chr1-10002583-10003163 chr1-10002583-100031..  0.01680164\n",
       "        [3] chr1-10002583-10003163 chr1-10002583-100031.. -0.02066524\n",
       "        [4] chr1-10002583-10003163 chr1-10002583-100031.. -0.00684081\n",
       "        [5] chr1-10002583-10003163 chr1-10002583-100031.. -0.00312169\n",
       "        ...                    ...                    ...         ...\n",
       "  [2821834]   chrX-9940388-9940697 chrX-9940388-9940697..  -0.2150580\n",
       "  [2821835]   chrX-9995829-9996183 chrX-9995829-9996183..   0.0630113\n",
       "  [2821836]   chrX-9995829-9996183 chrX-9995829-9996183..  -0.0419168\n",
       "  [2821837]   chrX-9995829-9996183 chrX-9995829-9996183..  -0.0222846\n",
       "  [2821838]   chrX-9995829-9996183 chrX-9995829-9996183..  -0.0163114\n",
       "                    FDR  VarQATAC   VarQRNA\n",
       "              <numeric> <numeric> <numeric>\n",
       "        [1]    0.830043  0.260582  0.294422\n",
       "        [2]    0.832309  0.260582  0.277407\n",
       "        [3]    0.789622  0.260582  0.850474\n",
       "        [4]    0.935920  0.260582  0.653618\n",
       "        [5]    0.971442  0.260582  0.809067\n",
       "        ...         ...       ...       ...\n",
       "  [2821834] 3.25486e-05 0.0124998  0.498331\n",
       "  [2821835] 3.30070e-01 0.4861127  0.793506\n",
       "  [2821836] 5.47389e-01 0.4861127  0.492623\n",
       "  [2821837] 7.71454e-01 0.4861127  0.966240\n",
       "  [2821838] 8.37664e-01 0.4861127  0.498331\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2821838 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>Correlation</th><th scope=col>FDR</th><th scope=col>VarQATAC</th><th scope=col>VarQRNA</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>APITD1-CORT</td><td>K562</td><td> 0.017010137</td><td>0.830042645</td><td>0.2605815</td><td>0.29442171</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>CASZ1      </td><td>K562</td><td> 0.016801643</td><td>0.832309354</td><td>0.2605815</td><td>0.27740685</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>CLSTN1     </td><td>K562</td><td>-0.020665242</td><td>0.789621852</td><td>0.2605815</td><td>0.85047383</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>CTNNBIP1   </td><td>K562</td><td>-0.006840807</td><td>0.935920148</td><td>0.2605815</td><td>0.65361835</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>DFFA       </td><td>K562</td><td>-0.003121688</td><td>0.971441733</td><td>0.2605815</td><td>0.80906741</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>GPR157     </td><td>K562</td><td> 0.036729313</td><td>0.606002854</td><td>0.2605815</td><td>0.35655826</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>H6PD       </td><td>K562</td><td> 0.112138720</td><td>0.053289189</td><td>0.2605815</td><td>0.58997415</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>KIF1B      </td><td>K562</td><td> 0.005909615</td><td>0.945024393</td><td>0.2605815</td><td>0.83227439</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>LINC01759  </td><td>K562</td><td>-0.132520934</td><td>0.018524241</td><td>0.2605815</td><td>0.43727116</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>LZIC       </td><td>K562</td><td>-0.036094554</td><td>0.613229614</td><td>0.2605815</td><td>0.64936463</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>MIR34AHG   </td><td>K562</td><td> 0.152116008</td><td>0.005600112</td><td>0.2605815</td><td>0.24230024</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>NMNAT1     </td><td>K562</td><td> 0.060646005</td><td>0.352007481</td><td>0.2605815</td><td>0.68840190</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>PEX14      </td><td>K562</td><td> 0.011833752</td><td>0.885466295</td><td>0.2605815</td><td>0.70848589</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>PGD        </td><td>K562</td><td> 0.002841041</td><td>0.974104442</td><td>0.2605815</td><td>0.94550937</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>PIK3CD     </td><td>K562</td><td> 0.068508028</td><td>0.282054760</td><td>0.2605815</td><td>0.59708163</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>SLC25A33   </td><td>K562</td><td>-0.155489298</td><td>0.004474502</td><td>0.2605815</td><td>0.88202671</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>SLC2A5     </td><td>K562</td><td> 0.049715363</td><td>0.462017332</td><td>0.2605815</td><td>0.30998277</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>SPSB1      </td><td>K562</td><td>-0.044097656</td><td>0.523075062</td><td>0.2605815</td><td>0.30513677</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>TMEM201    </td><td>K562</td><td>-0.007256146</td><td>0.931834230</td><td>0.2605815</td><td>0.52019169</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>UBE4B      </td><td>K562</td><td>-0.024771615</td><td>0.743084945</td><td>0.2605815</td><td>0.78537583</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>AGL        </td><td>K562</td><td> 0.060478796</td><td>0.353584122</td><td>0.5666289</td><td>0.79625242</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>CDC14A     </td><td>K562</td><td> 0.032225890</td><td>0.657653381</td><td>0.5666289</td><td>0.86188886</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>DBT        </td><td>K562</td><td> 0.049697656</td><td>0.462203044</td><td>0.5666289</td><td>0.79022184</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>DPH5       </td><td>K562</td><td> 0.029092205</td><td>0.693670655</td><td>0.5666289</td><td>0.79948309</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>EXTL2      </td><td>K562</td><td>-0.023751619</td><td>0.754742667</td><td>0.5666289</td><td>0.82188240</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>FRRS1      </td><td>K562</td><td>-0.106765924</td><td>0.068233816</td><td>0.5666289</td><td>0.98093905</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>LRRC39     </td><td>K562</td><td> 0.066930203</td><td>0.295346572</td><td>0.5666289</td><td>0.76389188</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>MFSD14A    </td><td>K562</td><td>-0.012000586</td><td>0.883719409</td><td>0.5666289</td><td>0.09945079</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>PALMD      </td><td>K562</td><td>-0.037325987</td><td>0.599236084</td><td>0.5666289</td><td>0.77353005</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>PLPPR4     </td><td>K562</td><td> 0.029738446</td><td>0.686290396</td><td>0.5666289</td><td>0.16767176</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882537</td><td>9883461</td><td>FAM9B    </td><td>K562</td><td>-0.01361413</td><td>8.665948e-01</td><td>0.76744701</td><td>0.4926233</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882537</td><td>9883461</td><td>TBL1X    </td><td>K562</td><td> 0.07595770</td><td>2.238654e-01</td><td>0.76744701</td><td>0.9662395</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882537</td><td>9883461</td><td>WWC3     </td><td>K562</td><td>-0.03804509</td><td>5.910865e-01</td><td>0.76744701</td><td>0.4983308</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885193</td><td>9885478</td><td>CLCN4    </td><td>K562</td><td> 0.03261987</td><td>6.531249e-01</td><td>0.38963145</td><td>0.7935064</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885193</td><td>9885478</td><td>FAM9B    </td><td>K562</td><td> 0.05781315</td><td>3.792489e-01</td><td>0.38963145</td><td>0.4926233</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885193</td><td>9885478</td><td>TBL1X    </td><td>K562</td><td> 0.11334697</td><td>5.029517e-02</td><td>0.38963145</td><td>0.9662395</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885193</td><td>9885478</td><td>WWC3     </td><td>K562</td><td>-0.12932860</td><td>2.211699e-02</td><td>0.38963145</td><td>0.4983308</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895197</td><td>CLCN4    </td><td>K562</td><td> 0.12601040</td><td>2.647670e-02</td><td>0.11462852</td><td>0.7935064</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895197</td><td>FAM9B    </td><td>K562</td><td>-0.03388436</td><td>6.385784e-01</td><td>0.11462852</td><td>0.4926233</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895197</td><td>TBL1X    </td><td>K562</td><td>-0.11737230</td><td>4.132221e-02</td><td>0.11462852</td><td>0.9662395</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895197</td><td>WWC3     </td><td>K562</td><td> 0.12575662</td><td>2.683861e-02</td><td>0.11462852</td><td>0.4983308</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>AKAP17A  </td><td>K562</td><td> 0.11460279</td><td>4.734087e-02</td><td>0.90799364</td><td>0.8655503</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>ASMTL    </td><td>K562</td><td> 0.03863736</td><td>5.843763e-01</td><td>0.90799364</td><td>0.5412987</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>CSF2RA   </td><td>K562</td><td>-0.11941881</td><td>3.730055e-02</td><td>0.90799364</td><td>0.2627611</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>GTPBP6   </td><td>K562</td><td> 0.17912446</td><td>7.909311e-04</td><td>0.90799364</td><td>0.5426448</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>IL3RA    </td><td>K562</td><td> 0.01717281</td><td>8.282507e-01</td><td>0.90799364</td><td>0.3873573</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>LINC00106</td><td>K562</td><td>-0.06192200</td><td>3.400441e-01</td><td>0.90799364</td><td>0.4632781</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>P2RY8    </td><td>K562</td><td>-0.03020252</td><td>6.809207e-01</td><td>0.90799364</td><td>0.3972647</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>PLCXD1   </td><td>K562</td><td> 0.02128416</td><td>7.827100e-01</td><td>0.90799364</td><td>0.6565260</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>PPP2R3B  </td><td>K562</td><td> 0.07577722</td><td>2.251865e-01</td><td>0.90799364</td><td>0.5677902</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>SHOX     </td><td>K562</td><td>-0.15222922</td><td>5.558175e-03</td><td>0.90799364</td><td>0.3061060</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>SLC25A6  </td><td>K562</td><td>-0.08303932</td><td>1.762341e-01</td><td>0.90799364</td><td>0.9011953</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940697</td><td>CLCN4    </td><td>K562</td><td>-0.27306047</td><td>4.086437e-08</td><td>0.01249984</td><td>0.7935064</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940697</td><td>FAM9B    </td><td>K562</td><td> 0.04978357</td><td>4.612842e-01</td><td>0.01249984</td><td>0.4926233</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940697</td><td>TBL1X    </td><td>K562</td><td> 0.30111040</td><td>7.756537e-10</td><td>0.01249984</td><td>0.9662395</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940697</td><td>WWC3     </td><td>K562</td><td>-0.21505795</td><td>3.254855e-05</td><td>0.01249984</td><td>0.4983308</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>CLCN4    </td><td>K562</td><td> 0.06301131</td><td>3.300701e-01</td><td>0.48611274</td><td>0.7935064</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>FAM9B    </td><td>K562</td><td>-0.04191677</td><td>5.473889e-01</td><td>0.48611274</td><td>0.4926233</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>TBL1X    </td><td>K562</td><td>-0.02228461</td><td>7.714537e-01</td><td>0.48611274</td><td>0.9662395</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>WWC3     </td><td>K562</td><td>-0.01631136</td><td>8.376636e-01</td><td>0.48611274</td><td>0.4983308</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2821838 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & Correlation & FDR & VarQATAC & VarQRNA\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  10002583 &  10003163 & APITD1-CORT & K562 &  0.017010137 & 0.830042645 & 0.2605815 & 0.29442171\\\\\n",
       "\t chr1 &  10002583 &  10003163 & CASZ1       & K562 &  0.016801643 & 0.832309354 & 0.2605815 & 0.27740685\\\\\n",
       "\t chr1 &  10002583 &  10003163 & CLSTN1      & K562 & -0.020665242 & 0.789621852 & 0.2605815 & 0.85047383\\\\\n",
       "\t chr1 &  10002583 &  10003163 & CTNNBIP1    & K562 & -0.006840807 & 0.935920148 & 0.2605815 & 0.65361835\\\\\n",
       "\t chr1 &  10002583 &  10003163 & DFFA        & K562 & -0.003121688 & 0.971441733 & 0.2605815 & 0.80906741\\\\\n",
       "\t chr1 &  10002583 &  10003163 & GPR157      & K562 &  0.036729313 & 0.606002854 & 0.2605815 & 0.35655826\\\\\n",
       "\t chr1 &  10002583 &  10003163 & H6PD        & K562 &  0.112138720 & 0.053289189 & 0.2605815 & 0.58997415\\\\\n",
       "\t chr1 &  10002583 &  10003163 & KIF1B       & K562 &  0.005909615 & 0.945024393 & 0.2605815 & 0.83227439\\\\\n",
       "\t chr1 &  10002583 &  10003163 & LINC01759   & K562 & -0.132520934 & 0.018524241 & 0.2605815 & 0.43727116\\\\\n",
       "\t chr1 &  10002583 &  10003163 & LZIC        & K562 & -0.036094554 & 0.613229614 & 0.2605815 & 0.64936463\\\\\n",
       "\t chr1 &  10002583 &  10003163 & MIR34AHG    & K562 &  0.152116008 & 0.005600112 & 0.2605815 & 0.24230024\\\\\n",
       "\t chr1 &  10002583 &  10003163 & NMNAT1      & K562 &  0.060646005 & 0.352007481 & 0.2605815 & 0.68840190\\\\\n",
       "\t chr1 &  10002583 &  10003163 & PEX14       & K562 &  0.011833752 & 0.885466295 & 0.2605815 & 0.70848589\\\\\n",
       "\t chr1 &  10002583 &  10003163 & PGD         & K562 &  0.002841041 & 0.974104442 & 0.2605815 & 0.94550937\\\\\n",
       "\t chr1 &  10002583 &  10003163 & PIK3CD      & K562 &  0.068508028 & 0.282054760 & 0.2605815 & 0.59708163\\\\\n",
       "\t chr1 &  10002583 &  10003163 & SLC25A33    & K562 & -0.155489298 & 0.004474502 & 0.2605815 & 0.88202671\\\\\n",
       "\t chr1 &  10002583 &  10003163 & SLC2A5      & K562 &  0.049715363 & 0.462017332 & 0.2605815 & 0.30998277\\\\\n",
       "\t chr1 &  10002583 &  10003163 & SPSB1       & K562 & -0.044097656 & 0.523075062 & 0.2605815 & 0.30513677\\\\\n",
       "\t chr1 &  10002583 &  10003163 & TMEM201     & K562 & -0.007256146 & 0.931834230 & 0.2605815 & 0.52019169\\\\\n",
       "\t chr1 &  10002583 &  10003163 & UBE4B       & K562 & -0.024771615 & 0.743084945 & 0.2605815 & 0.78537583\\\\\n",
       "\t chr1 & 100028213 & 100029660 & AGL         & K562 &  0.060478796 & 0.353584122 & 0.5666289 & 0.79625242\\\\\n",
       "\t chr1 & 100028213 & 100029660 & CDC14A      & K562 &  0.032225890 & 0.657653381 & 0.5666289 & 0.86188886\\\\\n",
       "\t chr1 & 100028213 & 100029660 & DBT         & K562 &  0.049697656 & 0.462203044 & 0.5666289 & 0.79022184\\\\\n",
       "\t chr1 & 100028213 & 100029660 & DPH5        & K562 &  0.029092205 & 0.693670655 & 0.5666289 & 0.79948309\\\\\n",
       "\t chr1 & 100028213 & 100029660 & EXTL2       & K562 & -0.023751619 & 0.754742667 & 0.5666289 & 0.82188240\\\\\n",
       "\t chr1 & 100028213 & 100029660 & FRRS1       & K562 & -0.106765924 & 0.068233816 & 0.5666289 & 0.98093905\\\\\n",
       "\t chr1 & 100028213 & 100029660 & LRRC39      & K562 &  0.066930203 & 0.295346572 & 0.5666289 & 0.76389188\\\\\n",
       "\t chr1 & 100028213 & 100029660 & MFSD14A     & K562 & -0.012000586 & 0.883719409 & 0.5666289 & 0.09945079\\\\\n",
       "\t chr1 & 100028213 & 100029660 & PALMD       & K562 & -0.037325987 & 0.599236084 & 0.5666289 & 0.77353005\\\\\n",
       "\t chr1 & 100028213 & 100029660 & PLPPR4      & K562 &  0.029738446 & 0.686290396 & 0.5666289 & 0.16767176\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9882537 & 9883461 & FAM9B     & K562 & -0.01361413 & 8.665948e-01 & 0.76744701 & 0.4926233\\\\\n",
       "\t chrX & 9882537 & 9883461 & TBL1X     & K562 &  0.07595770 & 2.238654e-01 & 0.76744701 & 0.9662395\\\\\n",
       "\t chrX & 9882537 & 9883461 & WWC3      & K562 & -0.03804509 & 5.910865e-01 & 0.76744701 & 0.4983308\\\\\n",
       "\t chrX & 9885193 & 9885478 & CLCN4     & K562 &  0.03261987 & 6.531249e-01 & 0.38963145 & 0.7935064\\\\\n",
       "\t chrX & 9885193 & 9885478 & FAM9B     & K562 &  0.05781315 & 3.792489e-01 & 0.38963145 & 0.4926233\\\\\n",
       "\t chrX & 9885193 & 9885478 & TBL1X     & K562 &  0.11334697 & 5.029517e-02 & 0.38963145 & 0.9662395\\\\\n",
       "\t chrX & 9885193 & 9885478 & WWC3      & K562 & -0.12932860 & 2.211699e-02 & 0.38963145 & 0.4983308\\\\\n",
       "\t chrX & 9894528 & 9895197 & CLCN4     & K562 &  0.12601040 & 2.647670e-02 & 0.11462852 & 0.7935064\\\\\n",
       "\t chrX & 9894528 & 9895197 & FAM9B     & K562 & -0.03388436 & 6.385784e-01 & 0.11462852 & 0.4926233\\\\\n",
       "\t chrX & 9894528 & 9895197 & TBL1X     & K562 & -0.11737230 & 4.132221e-02 & 0.11462852 & 0.9662395\\\\\n",
       "\t chrX & 9894528 & 9895197 & WWC3      & K562 &  0.12575662 & 2.683861e-02 & 0.11462852 & 0.4983308\\\\\n",
       "\t chrX &  989766 &  990624 & AKAP17A   & K562 &  0.11460279 & 4.734087e-02 & 0.90799364 & 0.8655503\\\\\n",
       "\t chrX &  989766 &  990624 & ASMTL     & K562 &  0.03863736 & 5.843763e-01 & 0.90799364 & 0.5412987\\\\\n",
       "\t chrX &  989766 &  990624 & CSF2RA    & K562 & -0.11941881 & 3.730055e-02 & 0.90799364 & 0.2627611\\\\\n",
       "\t chrX &  989766 &  990624 & GTPBP6    & K562 &  0.17912446 & 7.909311e-04 & 0.90799364 & 0.5426448\\\\\n",
       "\t chrX &  989766 &  990624 & IL3RA     & K562 &  0.01717281 & 8.282507e-01 & 0.90799364 & 0.3873573\\\\\n",
       "\t chrX &  989766 &  990624 & LINC00106 & K562 & -0.06192200 & 3.400441e-01 & 0.90799364 & 0.4632781\\\\\n",
       "\t chrX &  989766 &  990624 & P2RY8     & K562 & -0.03020252 & 6.809207e-01 & 0.90799364 & 0.3972647\\\\\n",
       "\t chrX &  989766 &  990624 & PLCXD1    & K562 &  0.02128416 & 7.827100e-01 & 0.90799364 & 0.6565260\\\\\n",
       "\t chrX &  989766 &  990624 & PPP2R3B   & K562 &  0.07577722 & 2.251865e-01 & 0.90799364 & 0.5677902\\\\\n",
       "\t chrX &  989766 &  990624 & SHOX      & K562 & -0.15222922 & 5.558175e-03 & 0.90799364 & 0.3061060\\\\\n",
       "\t chrX &  989766 &  990624 & SLC25A6   & K562 & -0.08303932 & 1.762341e-01 & 0.90799364 & 0.9011953\\\\\n",
       "\t chrX & 9940388 & 9940697 & CLCN4     & K562 & -0.27306047 & 4.086437e-08 & 0.01249984 & 0.7935064\\\\\n",
       "\t chrX & 9940388 & 9940697 & FAM9B     & K562 &  0.04978357 & 4.612842e-01 & 0.01249984 & 0.4926233\\\\\n",
       "\t chrX & 9940388 & 9940697 & TBL1X     & K562 &  0.30111040 & 7.756537e-10 & 0.01249984 & 0.9662395\\\\\n",
       "\t chrX & 9940388 & 9940697 & WWC3      & K562 & -0.21505795 & 3.254855e-05 & 0.01249984 & 0.4983308\\\\\n",
       "\t chrX & 9995829 & 9996183 & CLCN4     & K562 &  0.06301131 & 3.300701e-01 & 0.48611274 & 0.7935064\\\\\n",
       "\t chrX & 9995829 & 9996183 & FAM9B     & K562 & -0.04191677 & 5.473889e-01 & 0.48611274 & 0.4926233\\\\\n",
       "\t chrX & 9995829 & 9996183 & TBL1X     & K562 & -0.02228461 & 7.714537e-01 & 0.48611274 & 0.9662395\\\\\n",
       "\t chrX & 9995829 & 9996183 & WWC3      & K562 & -0.01631136 & 8.376636e-01 & 0.48611274 & 0.4983308\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2821838 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | Correlation &lt;dbl&gt; | FDR &lt;dbl&gt; | VarQATAC &lt;dbl&gt; | VarQRNA &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  10002583 |  10003163 | APITD1-CORT | K562 |  0.017010137 | 0.830042645 | 0.2605815 | 0.29442171 |\n",
       "| chr1 |  10002583 |  10003163 | CASZ1       | K562 |  0.016801643 | 0.832309354 | 0.2605815 | 0.27740685 |\n",
       "| chr1 |  10002583 |  10003163 | CLSTN1      | K562 | -0.020665242 | 0.789621852 | 0.2605815 | 0.85047383 |\n",
       "| chr1 |  10002583 |  10003163 | CTNNBIP1    | K562 | -0.006840807 | 0.935920148 | 0.2605815 | 0.65361835 |\n",
       "| chr1 |  10002583 |  10003163 | DFFA        | K562 | -0.003121688 | 0.971441733 | 0.2605815 | 0.80906741 |\n",
       "| chr1 |  10002583 |  10003163 | GPR157      | K562 |  0.036729313 | 0.606002854 | 0.2605815 | 0.35655826 |\n",
       "| chr1 |  10002583 |  10003163 | H6PD        | K562 |  0.112138720 | 0.053289189 | 0.2605815 | 0.58997415 |\n",
       "| chr1 |  10002583 |  10003163 | KIF1B       | K562 |  0.005909615 | 0.945024393 | 0.2605815 | 0.83227439 |\n",
       "| chr1 |  10002583 |  10003163 | LINC01759   | K562 | -0.132520934 | 0.018524241 | 0.2605815 | 0.43727116 |\n",
       "| chr1 |  10002583 |  10003163 | LZIC        | K562 | -0.036094554 | 0.613229614 | 0.2605815 | 0.64936463 |\n",
       "| chr1 |  10002583 |  10003163 | MIR34AHG    | K562 |  0.152116008 | 0.005600112 | 0.2605815 | 0.24230024 |\n",
       "| chr1 |  10002583 |  10003163 | NMNAT1      | K562 |  0.060646005 | 0.352007481 | 0.2605815 | 0.68840190 |\n",
       "| chr1 |  10002583 |  10003163 | PEX14       | K562 |  0.011833752 | 0.885466295 | 0.2605815 | 0.70848589 |\n",
       "| chr1 |  10002583 |  10003163 | PGD         | K562 |  0.002841041 | 0.974104442 | 0.2605815 | 0.94550937 |\n",
       "| chr1 |  10002583 |  10003163 | PIK3CD      | K562 |  0.068508028 | 0.282054760 | 0.2605815 | 0.59708163 |\n",
       "| chr1 |  10002583 |  10003163 | SLC25A33    | K562 | -0.155489298 | 0.004474502 | 0.2605815 | 0.88202671 |\n",
       "| chr1 |  10002583 |  10003163 | SLC2A5      | K562 |  0.049715363 | 0.462017332 | 0.2605815 | 0.30998277 |\n",
       "| chr1 |  10002583 |  10003163 | SPSB1       | K562 | -0.044097656 | 0.523075062 | 0.2605815 | 0.30513677 |\n",
       "| chr1 |  10002583 |  10003163 | TMEM201     | K562 | -0.007256146 | 0.931834230 | 0.2605815 | 0.52019169 |\n",
       "| chr1 |  10002583 |  10003163 | UBE4B       | K562 | -0.024771615 | 0.743084945 | 0.2605815 | 0.78537583 |\n",
       "| chr1 | 100028213 | 100029660 | AGL         | K562 |  0.060478796 | 0.353584122 | 0.5666289 | 0.79625242 |\n",
       "| chr1 | 100028213 | 100029660 | CDC14A      | K562 |  0.032225890 | 0.657653381 | 0.5666289 | 0.86188886 |\n",
       "| chr1 | 100028213 | 100029660 | DBT         | K562 |  0.049697656 | 0.462203044 | 0.5666289 | 0.79022184 |\n",
       "| chr1 | 100028213 | 100029660 | DPH5        | K562 |  0.029092205 | 0.693670655 | 0.5666289 | 0.79948309 |\n",
       "| chr1 | 100028213 | 100029660 | EXTL2       | K562 | -0.023751619 | 0.754742667 | 0.5666289 | 0.82188240 |\n",
       "| chr1 | 100028213 | 100029660 | FRRS1       | K562 | -0.106765924 | 0.068233816 | 0.5666289 | 0.98093905 |\n",
       "| chr1 | 100028213 | 100029660 | LRRC39      | K562 |  0.066930203 | 0.295346572 | 0.5666289 | 0.76389188 |\n",
       "| chr1 | 100028213 | 100029660 | MFSD14A     | K562 | -0.012000586 | 0.883719409 | 0.5666289 | 0.09945079 |\n",
       "| chr1 | 100028213 | 100029660 | PALMD       | K562 | -0.037325987 | 0.599236084 | 0.5666289 | 0.77353005 |\n",
       "| chr1 | 100028213 | 100029660 | PLPPR4      | K562 |  0.029738446 | 0.686290396 | 0.5666289 | 0.16767176 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9882537 | 9883461 | FAM9B     | K562 | -0.01361413 | 8.665948e-01 | 0.76744701 | 0.4926233 |\n",
       "| chrX | 9882537 | 9883461 | TBL1X     | K562 |  0.07595770 | 2.238654e-01 | 0.76744701 | 0.9662395 |\n",
       "| chrX | 9882537 | 9883461 | WWC3      | K562 | -0.03804509 | 5.910865e-01 | 0.76744701 | 0.4983308 |\n",
       "| chrX | 9885193 | 9885478 | CLCN4     | K562 |  0.03261987 | 6.531249e-01 | 0.38963145 | 0.7935064 |\n",
       "| chrX | 9885193 | 9885478 | FAM9B     | K562 |  0.05781315 | 3.792489e-01 | 0.38963145 | 0.4926233 |\n",
       "| chrX | 9885193 | 9885478 | TBL1X     | K562 |  0.11334697 | 5.029517e-02 | 0.38963145 | 0.9662395 |\n",
       "| chrX | 9885193 | 9885478 | WWC3      | K562 | -0.12932860 | 2.211699e-02 | 0.38963145 | 0.4983308 |\n",
       "| chrX | 9894528 | 9895197 | CLCN4     | K562 |  0.12601040 | 2.647670e-02 | 0.11462852 | 0.7935064 |\n",
       "| chrX | 9894528 | 9895197 | FAM9B     | K562 | -0.03388436 | 6.385784e-01 | 0.11462852 | 0.4926233 |\n",
       "| chrX | 9894528 | 9895197 | TBL1X     | K562 | -0.11737230 | 4.132221e-02 | 0.11462852 | 0.9662395 |\n",
       "| chrX | 9894528 | 9895197 | WWC3      | K562 |  0.12575662 | 2.683861e-02 | 0.11462852 | 0.4983308 |\n",
       "| chrX |  989766 |  990624 | AKAP17A   | K562 |  0.11460279 | 4.734087e-02 | 0.90799364 | 0.8655503 |\n",
       "| chrX |  989766 |  990624 | ASMTL     | K562 |  0.03863736 | 5.843763e-01 | 0.90799364 | 0.5412987 |\n",
       "| chrX |  989766 |  990624 | CSF2RA    | K562 | -0.11941881 | 3.730055e-02 | 0.90799364 | 0.2627611 |\n",
       "| chrX |  989766 |  990624 | GTPBP6    | K562 |  0.17912446 | 7.909311e-04 | 0.90799364 | 0.5426448 |\n",
       "| chrX |  989766 |  990624 | IL3RA     | K562 |  0.01717281 | 8.282507e-01 | 0.90799364 | 0.3873573 |\n",
       "| chrX |  989766 |  990624 | LINC00106 | K562 | -0.06192200 | 3.400441e-01 | 0.90799364 | 0.4632781 |\n",
       "| chrX |  989766 |  990624 | P2RY8     | K562 | -0.03020252 | 6.809207e-01 | 0.90799364 | 0.3972647 |\n",
       "| chrX |  989766 |  990624 | PLCXD1    | K562 |  0.02128416 | 7.827100e-01 | 0.90799364 | 0.6565260 |\n",
       "| chrX |  989766 |  990624 | PPP2R3B   | K562 |  0.07577722 | 2.251865e-01 | 0.90799364 | 0.5677902 |\n",
       "| chrX |  989766 |  990624 | SHOX      | K562 | -0.15222922 | 5.558175e-03 | 0.90799364 | 0.3061060 |\n",
       "| chrX |  989766 |  990624 | SLC25A6   | K562 | -0.08303932 | 1.762341e-01 | 0.90799364 | 0.9011953 |\n",
       "| chrX | 9940388 | 9940697 | CLCN4     | K562 | -0.27306047 | 4.086437e-08 | 0.01249984 | 0.7935064 |\n",
       "| chrX | 9940388 | 9940697 | FAM9B     | K562 |  0.04978357 | 4.612842e-01 | 0.01249984 | 0.4926233 |\n",
       "| chrX | 9940388 | 9940697 | TBL1X     | K562 |  0.30111040 | 7.756537e-10 | 0.01249984 | 0.9662395 |\n",
       "| chrX | 9940388 | 9940697 | WWC3      | K562 | -0.21505795 | 3.254855e-05 | 0.01249984 | 0.4983308 |\n",
       "| chrX | 9995829 | 9996183 | CLCN4     | K562 |  0.06301131 | 3.300701e-01 | 0.48611274 | 0.7935064 |\n",
       "| chrX | 9995829 | 9996183 | FAM9B     | K562 | -0.04191677 | 5.473889e-01 | 0.48611274 | 0.4926233 |\n",
       "| chrX | 9995829 | 9996183 | TBL1X     | K562 | -0.02228461 | 7.714537e-01 | 0.48611274 | 0.9662395 |\n",
       "| chrX | 9995829 | 9996183 | WWC3      | K562 | -0.01631136 | 8.376636e-01 | 0.48611274 | 0.4983308 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene  CellType Correlation  FDR         \n",
       "1       chr1  10002583  10003163 APITD1-CORT K562      0.017010137 0.830042645 \n",
       "2       chr1  10002583  10003163 CASZ1       K562      0.016801643 0.832309354 \n",
       "3       chr1  10002583  10003163 CLSTN1      K562     -0.020665242 0.789621852 \n",
       "4       chr1  10002583  10003163 CTNNBIP1    K562     -0.006840807 0.935920148 \n",
       "5       chr1  10002583  10003163 DFFA        K562     -0.003121688 0.971441733 \n",
       "6       chr1  10002583  10003163 GPR157      K562      0.036729313 0.606002854 \n",
       "7       chr1  10002583  10003163 H6PD        K562      0.112138720 0.053289189 \n",
       "8       chr1  10002583  10003163 KIF1B       K562      0.005909615 0.945024393 \n",
       "9       chr1  10002583  10003163 LINC01759   K562     -0.132520934 0.018524241 \n",
       "10      chr1  10002583  10003163 LZIC        K562     -0.036094554 0.613229614 \n",
       "11      chr1  10002583  10003163 MIR34AHG    K562      0.152116008 0.005600112 \n",
       "12      chr1  10002583  10003163 NMNAT1      K562      0.060646005 0.352007481 \n",
       "13      chr1  10002583  10003163 PEX14       K562      0.011833752 0.885466295 \n",
       "14      chr1  10002583  10003163 PGD         K562      0.002841041 0.974104442 \n",
       "15      chr1  10002583  10003163 PIK3CD      K562      0.068508028 0.282054760 \n",
       "16      chr1  10002583  10003163 SLC25A33    K562     -0.155489298 0.004474502 \n",
       "17      chr1  10002583  10003163 SLC2A5      K562      0.049715363 0.462017332 \n",
       "18      chr1  10002583  10003163 SPSB1       K562     -0.044097656 0.523075062 \n",
       "19      chr1  10002583  10003163 TMEM201     K562     -0.007256146 0.931834230 \n",
       "20      chr1  10002583  10003163 UBE4B       K562     -0.024771615 0.743084945 \n",
       "21      chr1 100028213 100029660 AGL         K562      0.060478796 0.353584122 \n",
       "22      chr1 100028213 100029660 CDC14A      K562      0.032225890 0.657653381 \n",
       "23      chr1 100028213 100029660 DBT         K562      0.049697656 0.462203044 \n",
       "24      chr1 100028213 100029660 DPH5        K562      0.029092205 0.693670655 \n",
       "25      chr1 100028213 100029660 EXTL2       K562     -0.023751619 0.754742667 \n",
       "26      chr1 100028213 100029660 FRRS1       K562     -0.106765924 0.068233816 \n",
       "27      chr1 100028213 100029660 LRRC39      K562      0.066930203 0.295346572 \n",
       "28      chr1 100028213 100029660 MFSD14A     K562     -0.012000586 0.883719409 \n",
       "29      chr1 100028213 100029660 PALMD       K562     -0.037325987 0.599236084 \n",
       "30      chr1 100028213 100029660 PLPPR4      K562      0.029738446 0.686290396 \n",
       "⋮       ⋮    ⋮         ⋮         ⋮           ⋮        ⋮            ⋮           \n",
       "2821809 chrX 9882537   9883461   FAM9B       K562     -0.01361413  8.665948e-01\n",
       "2821810 chrX 9882537   9883461   TBL1X       K562      0.07595770  2.238654e-01\n",
       "2821811 chrX 9882537   9883461   WWC3        K562     -0.03804509  5.910865e-01\n",
       "2821812 chrX 9885193   9885478   CLCN4       K562      0.03261987  6.531249e-01\n",
       "2821813 chrX 9885193   9885478   FAM9B       K562      0.05781315  3.792489e-01\n",
       "2821814 chrX 9885193   9885478   TBL1X       K562      0.11334697  5.029517e-02\n",
       "2821815 chrX 9885193   9885478   WWC3        K562     -0.12932860  2.211699e-02\n",
       "2821816 chrX 9894528   9895197   CLCN4       K562      0.12601040  2.647670e-02\n",
       "2821817 chrX 9894528   9895197   FAM9B       K562     -0.03388436  6.385784e-01\n",
       "2821818 chrX 9894528   9895197   TBL1X       K562     -0.11737230  4.132221e-02\n",
       "2821819 chrX 9894528   9895197   WWC3        K562      0.12575662  2.683861e-02\n",
       "2821820 chrX  989766    990624   AKAP17A     K562      0.11460279  4.734087e-02\n",
       "2821821 chrX  989766    990624   ASMTL       K562      0.03863736  5.843763e-01\n",
       "2821822 chrX  989766    990624   CSF2RA      K562     -0.11941881  3.730055e-02\n",
       "2821823 chrX  989766    990624   GTPBP6      K562      0.17912446  7.909311e-04\n",
       "2821824 chrX  989766    990624   IL3RA       K562      0.01717281  8.282507e-01\n",
       "2821825 chrX  989766    990624   LINC00106   K562     -0.06192200  3.400441e-01\n",
       "2821826 chrX  989766    990624   P2RY8       K562     -0.03020252  6.809207e-01\n",
       "2821827 chrX  989766    990624   PLCXD1      K562      0.02128416  7.827100e-01\n",
       "2821828 chrX  989766    990624   PPP2R3B     K562      0.07577722  2.251865e-01\n",
       "2821829 chrX  989766    990624   SHOX        K562     -0.15222922  5.558175e-03\n",
       "2821830 chrX  989766    990624   SLC25A6     K562     -0.08303932  1.762341e-01\n",
       "2821831 chrX 9940388   9940697   CLCN4       K562     -0.27306047  4.086437e-08\n",
       "2821832 chrX 9940388   9940697   FAM9B       K562      0.04978357  4.612842e-01\n",
       "2821833 chrX 9940388   9940697   TBL1X       K562      0.30111040  7.756537e-10\n",
       "2821834 chrX 9940388   9940697   WWC3        K562     -0.21505795  3.254855e-05\n",
       "2821835 chrX 9995829   9996183   CLCN4       K562      0.06301131  3.300701e-01\n",
       "2821836 chrX 9995829   9996183   FAM9B       K562     -0.04191677  5.473889e-01\n",
       "2821837 chrX 9995829   9996183   TBL1X       K562     -0.02228461  7.714537e-01\n",
       "2821838 chrX 9995829   9996183   WWC3        K562     -0.01631136  8.376636e-01\n",
       "        VarQATAC   VarQRNA   \n",
       "1       0.2605815  0.29442171\n",
       "2       0.2605815  0.27740685\n",
       "3       0.2605815  0.85047383\n",
       "4       0.2605815  0.65361835\n",
       "5       0.2605815  0.80906741\n",
       "6       0.2605815  0.35655826\n",
       "7       0.2605815  0.58997415\n",
       "8       0.2605815  0.83227439\n",
       "9       0.2605815  0.43727116\n",
       "10      0.2605815  0.64936463\n",
       "11      0.2605815  0.24230024\n",
       "12      0.2605815  0.68840190\n",
       "13      0.2605815  0.70848589\n",
       "14      0.2605815  0.94550937\n",
       "15      0.2605815  0.59708163\n",
       "16      0.2605815  0.88202671\n",
       "17      0.2605815  0.30998277\n",
       "18      0.2605815  0.30513677\n",
       "19      0.2605815  0.52019169\n",
       "20      0.2605815  0.78537583\n",
       "21      0.5666289  0.79625242\n",
       "22      0.5666289  0.86188886\n",
       "23      0.5666289  0.79022184\n",
       "24      0.5666289  0.79948309\n",
       "25      0.5666289  0.82188240\n",
       "26      0.5666289  0.98093905\n",
       "27      0.5666289  0.76389188\n",
       "28      0.5666289  0.09945079\n",
       "29      0.5666289  0.77353005\n",
       "30      0.5666289  0.16767176\n",
       "⋮       ⋮          ⋮         \n",
       "2821809 0.76744701 0.4926233 \n",
       "2821810 0.76744701 0.9662395 \n",
       "2821811 0.76744701 0.4983308 \n",
       "2821812 0.38963145 0.7935064 \n",
       "2821813 0.38963145 0.4926233 \n",
       "2821814 0.38963145 0.9662395 \n",
       "2821815 0.38963145 0.4983308 \n",
       "2821816 0.11462852 0.7935064 \n",
       "2821817 0.11462852 0.4926233 \n",
       "2821818 0.11462852 0.9662395 \n",
       "2821819 0.11462852 0.4983308 \n",
       "2821820 0.90799364 0.8655503 \n",
       "2821821 0.90799364 0.5412987 \n",
       "2821822 0.90799364 0.2627611 \n",
       "2821823 0.90799364 0.5426448 \n",
       "2821824 0.90799364 0.3873573 \n",
       "2821825 0.90799364 0.4632781 \n",
       "2821826 0.90799364 0.3972647 \n",
       "2821827 0.90799364 0.6565260 \n",
       "2821828 0.90799364 0.5677902 \n",
       "2821829 0.90799364 0.3061060 \n",
       "2821830 0.90799364 0.9011953 \n",
       "2821831 0.01249984 0.7935064 \n",
       "2821832 0.01249984 0.4926233 \n",
       "2821833 0.01249984 0.9662395 \n",
       "2821834 0.01249984 0.4983308 \n",
       "2821835 0.48611274 0.7935064 \n",
       "2821836 0.48611274 0.4926233 \n",
       "2821837 0.48611274 0.9662395 \n",
       "2821838 0.48611274 0.4983308 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(obj.ArchR,\n",
    "        paste(dir.output,\"obj.ArchR.rds\",sep = \"/\"))\n",
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"Correlation\",\n",
    "                         \"FDR\",\n",
    "                         \"VarQATAC\",\n",
    "                         \"VarQRNA\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "pairs.E2G.res\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4c909ed5-7321-4e69-8208-1544b371da6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] nabor_0.5.0                       dplyr_1.1.4                      \n",
       " [3] uwot_0.1.16                       BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] genomation_1.34.0                 Signac_1.13.0                    \n",
       "[11] Seurat_5.0.3                      SeuratObject_5.0.1               \n",
       "[13] sp_2.1-4                          rhdf5_2.44.0                     \n",
       "[15] SummarizedExperiment_1.30.2       Biobase_2.60.0                   \n",
       "[17] MatrixGenerics_1.12.3             Rcpp_1.0.12                      \n",
       "[19] Matrix_1.6-5                      GenomicRanges_1.54.1             \n",
       "[21] GenomeInfoDb_1.38.1               IRanges_2.36.0                   \n",
       "[23] S4Vectors_0.40.2                  BiocGenerics_0.48.1              \n",
       "[25] matrixStats_1.3.0                 data.table_1.15.2                \n",
       "[27] stringr_1.5.1                     plyr_1.8.9                       \n",
       "[29] magrittr_2.0.3                    ggplot2_3.5.1                    \n",
       "[31] gtable_0.3.5                      gtools_3.9.5                     \n",
       "[33] gridExtra_2.3                     ArchR_1.0.2                      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22         splines_4.3.3            later_1.3.2             \n",
       "  [4] pbdZMQ_0.3-11            BiocIO_1.10.0            bitops_1.0-7            \n",
       "  [7] tibble_3.2.1             polyclip_1.10-6          XML_3.99-0.16.1         \n",
       " [10] fastDummies_1.7.3        lifecycle_1.0.4          vroom_1.6.5             \n",
       " [13] globals_0.16.3           lattice_0.22-6           MASS_7.3-60             \n",
       " [16] plotly_4.10.4            plotrix_3.8-4            yaml_2.3.8              \n",
       " [19] httpuv_1.6.15            sctransform_0.4.1        spam_2.10-0             \n",
       " [22] spatstat.sparse_3.0-3    reticulate_1.36.1        cowplot_1.1.3           \n",
       " [25] pbapply_1.7-2            RColorBrewer_1.1-3       abind_1.4-5             \n",
       " [28] zlibbioc_1.48.0          Rtsne_0.17               purrr_1.0.2             \n",
       " [31] RCurl_1.98-1.14          GenomeInfoDbData_1.2.11  ggrepel_0.9.5           \n",
       " [34] irlba_2.3.5.1            listenv_0.9.1            spatstat.utils_3.0-4    \n",
       " [37] goftest_1.2-3            RSpectra_0.16-1          spatstat.random_3.2-3   \n",
       " [40] fitdistrplus_1.1-11      parallelly_1.37.1        leiden_0.4.3.1          \n",
       " [43] codetools_0.2-20         DelayedArray_0.26.7      RcppRoll_0.3.0          \n",
       " [46] tidyselect_1.2.1         farver_2.1.1             base64enc_0.1-3         \n",
       " [49] spatstat.explore_3.2-6   GenomicAlignments_1.36.0 jsonlite_1.8.8          \n",
       " [52] progressr_0.14.0         ggridges_0.5.6           survival_3.6-4          \n",
       " [55] tools_4.3.3              ica_1.0-3                glue_1.7.0              \n",
       " [58] IRdisplay_1.1            withr_3.0.0              fastmap_1.1.1           \n",
       " [61] rhdf5filters_1.12.1      fansi_1.0.6              digest_0.6.35           \n",
       " [64] R6_2.5.1                 mime_0.12                seqPattern_1.34.0       \n",
       " [67] colorspace_2.1-0         Cairo_1.6-2              scattermore_1.2         \n",
       " [70] tensor_1.5               spatstat.data_3.0-4      utf8_1.2.4              \n",
       " [73] tidyr_1.3.1              generics_0.1.3           httr_1.4.7              \n",
       " [76] htmlwidgets_1.6.4        S4Arrays_1.2.1           pkgconfig_2.0.3         \n",
       " [79] impute_1.76.0            lmtest_0.9-40            htmltools_0.5.8.1       \n",
       " [82] dotCall64_1.1-1          scales_1.3.0             png_0.1-8               \n",
       " [85] tzdb_0.4.0               reshape2_1.4.4           rjson_0.2.21            \n",
       " [88] uuid_1.2-0               nlme_3.1-164             repr_1.1.7              \n",
       " [91] zoo_1.8-12               KernSmooth_2.23-22       miniUI_0.1.1.1          \n",
       " [94] restfulr_0.0.15          pillar_1.9.0             vctrs_0.6.5             \n",
       " [97] RANN_2.6.1               promises_1.3.0           xtable_1.8-4            \n",
       "[100] cluster_2.1.6            evaluate_0.23            readr_2.1.5             \n",
       "[103] cli_3.6.2                compiler_4.3.3           Rsamtools_2.18.0        \n",
       "[106] rlang_1.1.3              crayon_1.5.2             future.apply_1.11.2     \n",
       "[109] labeling_0.4.3           stringi_1.8.3            gridBase_0.4-7          \n",
       "[112] viridisLite_0.4.2        deldir_2.0-4             BiocParallel_1.36.0     \n",
       "[115] munsell_0.5.1            lazyeval_0.2.2           spatstat.geom_3.2-9     \n",
       "[118] IRkernel_1.3.2           RcppHNSW_0.6.0           hms_1.1.3               \n",
       "[121] patchwork_1.2.0          bit64_4.0.5              future_1.33.2           \n",
       "[124] Rhdf5lib_1.22.1          shiny_1.8.1.1            ROCR_1.0-11             \n",
       "[127] igraph_2.0.3             fastmatch_1.1-4          bit_4.0.5               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7690cee-49f1-4886-ae66-0b63a43ce524",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
