{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_IGVF/K562/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_IGVF/K562/Predictions/EnhancerPredictionsAllPutative.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_IGVF/K562/Kendall/atac_matrix.csv.gz\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_IGVF/1.prepare_data/1.import_IGVF_processed_data.240508/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_IGVF/3.Genome_wide_prediction/FigR/FigR.240521/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2991532 ranges and 3 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10002583-10003163      * | APITD1-CORT\n",
       "        [2]     chr1 10002583-10003163      * |       CASZ1\n",
       "        [3]     chr1 10002583-10003163      * |       CENPS\n",
       "        [4]     chr1 10002583-10003163      * |      CLSTN1\n",
       "        [5]     chr1 10002583-10003163      * |    CTNNBIP1\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [2991528]     chrX   9940388-9940697      * |        WWC3\n",
       "  [2991529]     chrX   9995829-9996183      * |       CLCN4\n",
       "  [2991530]     chrX   9995829-9996183      * |       FAM9B\n",
       "  [2991531]     chrX   9995829-9996183      * |       TBL1X\n",
       "  [2991532]     chrX   9995829-9996183      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-10002583-10003163 chr1-10002583-100031..\n",
       "        [2] chr1-10002583-10003163 chr1-10002583-100031..\n",
       "        [3] chr1-10002583-10003163 chr1-10002583-100031..\n",
       "        [4] chr1-10002583-10003163 chr1-10002583-100031..\n",
       "        [5] chr1-10002583-10003163 chr1-10002583-100031..\n",
       "        ...                    ...                    ...\n",
       "  [2991528]   chrX-9940388-9940697 chrX-9940388-9940697..\n",
       "  [2991529]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  [2991530]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  [2991531]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  [2991532]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  4000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  4001  to  5000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  5001  to  6000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  6001  to  6266 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = read.csv(path.matrix.atac_count,\n",
    "                             row.names = 1,\n",
    "                             check.names = F)\n",
    "matrix.atac_count = Matrix(as.matrix(matrix.atac_count), sparse = TRUE)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 153522 6266 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(153522): chr1-10002583-10003163 chr1-100028213-100029660 ...\n",
       "  chrX-9940388-9940697 chrX-9995829-9996183\n",
       "rowData names(2): PeakName bias\n",
       "colnames(6266): e10l1_AAACCAGGTCACAGGG-1 e10l1_AAACCGCCAAGGAATA-1 ...\n",
       "  e10l2_TTTGTGAGTTAGCGTT-1 e10l2_TTTGTGAGTTCCGGAC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna),\n",
    "                       index = 1:nrow(matrix.rna))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(dir.output, \"chr\", sep = \"/\"), recursive = T):\n",
      "“'/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_IGVF/3.Genome_wide_prediction/FigR/FigR.240521//chr' already exists”\n"
     ]
    }
   ],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  17.4629943370819 secs \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5a93ede-8b2c-4646-9a36-b27ff71e2958",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2652977 ranges and 7 metadata columns:\n",
       "       seqnames            ranges strand |  TargetGene               PeakName\n",
       "          <Rle>         <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 10002583-10003163      * |       CASZ1 chr1-10002583-10003163\n",
       "  chr1     chr1 10002583-10003163      * |      CLSTN1 chr1-10002583-10003163\n",
       "  chr1     chr1 10002583-10003163      * |    CTNNBIP1 chr1-10002583-10003163\n",
       "  chr1     chr1 10002583-10003163      * |        DFFA chr1-10002583-10003163\n",
       "  chr1     chr1 10002583-10003163      * |      GPR157 chr1-10002583-10003163\n",
       "   ...      ...               ...    ... .         ...                    ...\n",
       "  chrX     chrX   9940388-9940697      * |        WWC3   chrX-9940388-9940697\n",
       "  chrX     chrX   9995829-9996183      * |       CLCN4   chrX-9995829-9996183\n",
       "  chrX     chrX   9995829-9996183      * |       FAM9B   chrX-9995829-9996183\n",
       "  chrX     chrX   9995829-9996183      * |       TBL1X   chrX-9995829-9996183\n",
       "  chrX     chrX   9995829-9996183      * |        WWC3   chrX-9995829-9996183\n",
       "                     PairName         rObs     rBgSD     rBgMean     pvalZ\n",
       "                  <character>    <numeric> <numeric>   <numeric> <numeric>\n",
       "  chr1 chr1-10002583-100031.. -0.008124274 0.0189605 0.006144801  0.774146\n",
       "  chr1 chr1-10002583-100031..  0.000657789 0.0140815 0.010612508  0.760196\n",
       "  chr1 chr1-10002583-100031.. -0.004844037 0.0129370 0.005985051  0.798721\n",
       "  chr1 chr1-10002583-100031..  0.005749848 0.0138833 0.008509491  0.578780\n",
       "  chr1 chr1-10002583-100031.. -0.007351548 0.0126400 0.000386855  0.729803\n",
       "   ...                    ...          ...       ...         ...       ...\n",
       "  chrX chrX-9940388-9940697.. -0.011425029 0.0154703  0.00310524  0.826195\n",
       "  chrX chrX-9995829-9996183..  0.000443103 0.0145916  0.01197411  0.785310\n",
       "  chrX chrX-9995829-9996183..  0.003983342 0.0158320  0.00761377  0.590686\n",
       "  chrX chrX-9995829-9996183..  0.012241754 0.0130908  0.00612860  0.320256\n",
       "  chrX chrX-9995829-9996183.. -0.010775611 0.0170661  0.01085367  0.897490\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2652977 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>CASZ1   </td><td>K562</td><td>-0.0081242737</td><td>0.01896046</td><td> 0.0061448005</td><td>0.77414583</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>CLSTN1  </td><td>K562</td><td> 0.0006577892</td><td>0.01408155</td><td> 0.0106125084</td><td>0.76019617</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>CTNNBIP1</td><td>K562</td><td>-0.0048440374</td><td>0.01293704</td><td> 0.0059850509</td><td>0.79872077</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>DFFA    </td><td>K562</td><td> 0.0057498484</td><td>0.01388333</td><td> 0.0085094908</td><td>0.57878014</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>GPR157  </td><td>K562</td><td>-0.0073515484</td><td>0.01263999</td><td> 0.0003868547</td><td>0.72980256</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>H6PD    </td><td>K562</td><td> 0.0145440003</td><td>0.01367281</td><td> 0.0060959871</td><td>0.26833059</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>KIF1B   </td><td>K562</td><td> 0.0099476403</td><td>0.01262953</td><td> 0.0030629305</td><td>0.29283281</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>LZIC    </td><td>K562</td><td>-0.0086760439</td><td>0.01471175</td><td> 0.0064147802</td><td>0.84749929</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>MIR34AHG</td><td>K562</td><td>-0.0062829789</td><td>0.01508511</td><td> 0.0029525886</td><td>0.72980739</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>NMNAT1  </td><td>K562</td><td> 0.0004576080</td><td>0.01553630</td><td> 0.0033567967</td><td>0.57401580</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>PEX14   </td><td>K562</td><td>-0.0021044751</td><td>0.01228127</td><td> 0.0061407749</td><td>0.74900681</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>PGD     </td><td>K562</td><td>-0.0097266341</td><td>0.01364105</td><td> 0.0109669256</td><td>0.93536747</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>PIK3CD  </td><td>K562</td><td> 0.0116406203</td><td>0.01159300</td><td> 0.0069426180</td><td>0.34264887</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>SLC25A33</td><td>K562</td><td>-0.0157542699</td><td>0.01327563</td><td> 0.0052311207</td><td>0.94303184</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>SLC2A5  </td><td>K562</td><td>-0.0077900364</td><td>0.01531539</td><td> 0.0020858030</td><td>0.74048171</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>SPSB1   </td><td>K562</td><td> 0.0101172609</td><td>0.01361995</td><td> 0.0016742368</td><td>0.26766147</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>TMEM201 </td><td>K562</td><td>-0.0138773320</td><td>0.01204423</td><td> 0.0050178666</td><td>0.94165470</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002583</td><td> 10003163</td><td>UBE4B   </td><td>K562</td><td> 0.0206362252</td><td>0.01192020</td><td> 0.0026118996</td><td>0.06525638</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>AGL     </td><td>K562</td><td> 0.0196482001</td><td>0.01279492</td><td> 0.0140633530</td><td>0.33124088</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>CDC14A  </td><td>K562</td><td> 0.0151807925</td><td>0.01109560</td><td> 0.0050404750</td><td>0.18038353</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>DBT     </td><td>K562</td><td> 0.0200868099</td><td>0.01393867</td><td> 0.0153017161</td><td>0.36568774</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>DPH5    </td><td>K562</td><td> 0.0352576269</td><td>0.01341538</td><td> 0.0146437083</td><td>0.06219701</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>EXTL2   </td><td>K562</td><td> 0.0113832481</td><td>0.01286317</td><td> 0.0091136428</td><td>0.42997327</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>FRRS1   </td><td>K562</td><td>-0.0086361998</td><td>0.01525691</td><td> 0.0029532273</td><td>0.77625857</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>LRRC39  </td><td>K562</td><td> 0.0183022959</td><td>0.01240822</td><td>-0.0015795927</td><td>0.05454285</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>MFSD14A </td><td>K562</td><td>-0.0023994597</td><td>0.01303874</td><td> 0.0002541471</td><td>0.58063459</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>PALMD   </td><td>K562</td><td> 0.0133023695</td><td>0.01394112</td><td> 0.0043858530</td><td>0.26122161</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>PLPPR4  </td><td>K562</td><td>-0.0041573131</td><td>0.01675531</td><td> 0.0045401385</td><td>0.69814969</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>RTCA    </td><td>K562</td><td> 0.0223314283</td><td>0.01298133</td><td> 0.0091116599</td><td>0.15425158</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028213</td><td>100029660</td><td>RTCA-AS1</td><td>K562</td><td> 0.0206360588</td><td>0.01278954</td><td> 0.0101930946</td><td>0.20710027</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882537</td><td>9883461</td><td>FAM9B    </td><td>K562</td><td> 0.0495612400</td><td>0.01628201</td><td>0.015729547</td><td>0.01886123</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882537</td><td>9883461</td><td>TBL1X    </td><td>K562</td><td> 0.0086979517</td><td>0.01382529</td><td>0.011814008</td><td>0.58916135</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882537</td><td>9883461</td><td>WWC3     </td><td>K562</td><td> 0.0318434754</td><td>0.01303893</td><td>0.013951028</td><td>0.08499545</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885193</td><td>9885478</td><td>CLCN4    </td><td>K562</td><td> 0.0304057499</td><td>0.01461636</td><td>0.008018696</td><td>0.06280515</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885193</td><td>9885478</td><td>FAM9B    </td><td>K562</td><td> 0.0260712046</td><td>0.01419383</td><td>0.006425263</td><td>0.08316111</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885193</td><td>9885478</td><td>TBL1X    </td><td>K562</td><td> 0.0138417233</td><td>0.01209908</td><td>0.005940276</td><td>0.25685817</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885193</td><td>9885478</td><td>WWC3     </td><td>K562</td><td> 0.0286078138</td><td>0.01493548</td><td>0.008122034</td><td>0.08509116</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895197</td><td>CLCN4    </td><td>K562</td><td> 0.0020997138</td><td>0.01403380</td><td>0.004577434</td><td>0.57007054</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895197</td><td>FAM9B    </td><td>K562</td><td>-0.0039155294</td><td>0.01641407</td><td>0.006766755</td><td>0.74241234</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895197</td><td>TBL1X    </td><td>K562</td><td> 0.0221218213</td><td>0.01131808</td><td>0.005799689</td><td>0.07463291</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895197</td><td>WWC3     </td><td>K562</td><td> 0.0194707907</td><td>0.01600771</td><td>0.008696931</td><td>0.25046029</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>AKAP17A  </td><td>K562</td><td> 0.0151072702</td><td>0.01450843</td><td>0.019885247</td><td>0.62904471</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>ASMTL    </td><td>K562</td><td> 0.0258398681</td><td>0.01214658</td><td>0.017759999</td><td>0.25296218</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>CSF2RA   </td><td>K562</td><td>-0.0018385678</td><td>0.01172884</td><td>0.009957915</td><td>0.84273632</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>GTPBP6   </td><td>K562</td><td> 0.0103101238</td><td>0.01257263</td><td>0.018756129</td><td>0.74913721</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>IL3RA    </td><td>K562</td><td> 0.0422719623</td><td>0.01384934</td><td>0.010607540</td><td>0.01111693</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>LINC00106</td><td>K562</td><td> 0.0194724634</td><td>0.01475951</td><td>0.011179045</td><td>0.28709096</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>P2RY8    </td><td>K562</td><td> 0.0162448765</td><td>0.01206066</td><td>0.004982340</td><td>0.17519737</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>PLCXD1   </td><td>K562</td><td> 0.0222317067</td><td>0.01357990</td><td>0.030507948</td><td>0.72888618</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>PPP2R3B  </td><td>K562</td><td> 0.0374172544</td><td>0.01480182</td><td>0.022819064</td><td>0.16200695</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>SHOX     </td><td>K562</td><td>-0.0093762835</td><td>0.01501508</td><td>0.002963468</td><td>0.79441139</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990624</td><td>SLC25A6  </td><td>K562</td><td> 0.0254863612</td><td>0.01467255</td><td>0.040325540</td><td>0.84407702</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940697</td><td>CLCN4    </td><td>K562</td><td>-0.0077134263</td><td>0.01366020</td><td>0.003995483</td><td>0.80432036</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940697</td><td>FAM9B    </td><td>K562</td><td> 0.0084791372</td><td>0.01289502</td><td>0.002140178</td><td>0.31150743</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940697</td><td>TBL1X    </td><td>K562</td><td> 0.0119357733</td><td>0.01079224</td><td>0.003320242</td><td>0.21234587</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940697</td><td>WWC3     </td><td>K562</td><td>-0.0114250295</td><td>0.01547033</td><td>0.003105238</td><td>0.82619482</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>CLCN4    </td><td>K562</td><td> 0.0004431027</td><td>0.01459155</td><td>0.011974109</td><td>0.78530973</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>FAM9B    </td><td>K562</td><td> 0.0039833415</td><td>0.01583196</td><td>0.007613770</td><td>0.59068606</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>TBL1X    </td><td>K562</td><td> 0.0122417538</td><td>0.01309076</td><td>0.006128604</td><td>0.32025643</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>WWC3     </td><td>K562</td><td>-0.0107756109</td><td>0.01706612</td><td>0.010853665</td><td>0.89749045</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2652977 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  10002583 &  10003163 & CASZ1    & K562 & -0.0081242737 & 0.01896046 &  0.0061448005 & 0.77414583\\\\\n",
       "\t chr1 &  10002583 &  10003163 & CLSTN1   & K562 &  0.0006577892 & 0.01408155 &  0.0106125084 & 0.76019617\\\\\n",
       "\t chr1 &  10002583 &  10003163 & CTNNBIP1 & K562 & -0.0048440374 & 0.01293704 &  0.0059850509 & 0.79872077\\\\\n",
       "\t chr1 &  10002583 &  10003163 & DFFA     & K562 &  0.0057498484 & 0.01388333 &  0.0085094908 & 0.57878014\\\\\n",
       "\t chr1 &  10002583 &  10003163 & GPR157   & K562 & -0.0073515484 & 0.01263999 &  0.0003868547 & 0.72980256\\\\\n",
       "\t chr1 &  10002583 &  10003163 & H6PD     & K562 &  0.0145440003 & 0.01367281 &  0.0060959871 & 0.26833059\\\\\n",
       "\t chr1 &  10002583 &  10003163 & KIF1B    & K562 &  0.0099476403 & 0.01262953 &  0.0030629305 & 0.29283281\\\\\n",
       "\t chr1 &  10002583 &  10003163 & LZIC     & K562 & -0.0086760439 & 0.01471175 &  0.0064147802 & 0.84749929\\\\\n",
       "\t chr1 &  10002583 &  10003163 & MIR34AHG & K562 & -0.0062829789 & 0.01508511 &  0.0029525886 & 0.72980739\\\\\n",
       "\t chr1 &  10002583 &  10003163 & NMNAT1   & K562 &  0.0004576080 & 0.01553630 &  0.0033567967 & 0.57401580\\\\\n",
       "\t chr1 &  10002583 &  10003163 & PEX14    & K562 & -0.0021044751 & 0.01228127 &  0.0061407749 & 0.74900681\\\\\n",
       "\t chr1 &  10002583 &  10003163 & PGD      & K562 & -0.0097266341 & 0.01364105 &  0.0109669256 & 0.93536747\\\\\n",
       "\t chr1 &  10002583 &  10003163 & PIK3CD   & K562 &  0.0116406203 & 0.01159300 &  0.0069426180 & 0.34264887\\\\\n",
       "\t chr1 &  10002583 &  10003163 & SLC25A33 & K562 & -0.0157542699 & 0.01327563 &  0.0052311207 & 0.94303184\\\\\n",
       "\t chr1 &  10002583 &  10003163 & SLC2A5   & K562 & -0.0077900364 & 0.01531539 &  0.0020858030 & 0.74048171\\\\\n",
       "\t chr1 &  10002583 &  10003163 & SPSB1    & K562 &  0.0101172609 & 0.01361995 &  0.0016742368 & 0.26766147\\\\\n",
       "\t chr1 &  10002583 &  10003163 & TMEM201  & K562 & -0.0138773320 & 0.01204423 &  0.0050178666 & 0.94165470\\\\\n",
       "\t chr1 &  10002583 &  10003163 & UBE4B    & K562 &  0.0206362252 & 0.01192020 &  0.0026118996 & 0.06525638\\\\\n",
       "\t chr1 & 100028213 & 100029660 & AGL      & K562 &  0.0196482001 & 0.01279492 &  0.0140633530 & 0.33124088\\\\\n",
       "\t chr1 & 100028213 & 100029660 & CDC14A   & K562 &  0.0151807925 & 0.01109560 &  0.0050404750 & 0.18038353\\\\\n",
       "\t chr1 & 100028213 & 100029660 & DBT      & K562 &  0.0200868099 & 0.01393867 &  0.0153017161 & 0.36568774\\\\\n",
       "\t chr1 & 100028213 & 100029660 & DPH5     & K562 &  0.0352576269 & 0.01341538 &  0.0146437083 & 0.06219701\\\\\n",
       "\t chr1 & 100028213 & 100029660 & EXTL2    & K562 &  0.0113832481 & 0.01286317 &  0.0091136428 & 0.42997327\\\\\n",
       "\t chr1 & 100028213 & 100029660 & FRRS1    & K562 & -0.0086361998 & 0.01525691 &  0.0029532273 & 0.77625857\\\\\n",
       "\t chr1 & 100028213 & 100029660 & LRRC39   & K562 &  0.0183022959 & 0.01240822 & -0.0015795927 & 0.05454285\\\\\n",
       "\t chr1 & 100028213 & 100029660 & MFSD14A  & K562 & -0.0023994597 & 0.01303874 &  0.0002541471 & 0.58063459\\\\\n",
       "\t chr1 & 100028213 & 100029660 & PALMD    & K562 &  0.0133023695 & 0.01394112 &  0.0043858530 & 0.26122161\\\\\n",
       "\t chr1 & 100028213 & 100029660 & PLPPR4   & K562 & -0.0041573131 & 0.01675531 &  0.0045401385 & 0.69814969\\\\\n",
       "\t chr1 & 100028213 & 100029660 & RTCA     & K562 &  0.0223314283 & 0.01298133 &  0.0091116599 & 0.15425158\\\\\n",
       "\t chr1 & 100028213 & 100029660 & RTCA-AS1 & K562 &  0.0206360588 & 0.01278954 &  0.0101930946 & 0.20710027\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9882537 & 9883461 & FAM9B     & K562 &  0.0495612400 & 0.01628201 & 0.015729547 & 0.01886123\\\\\n",
       "\t chrX & 9882537 & 9883461 & TBL1X     & K562 &  0.0086979517 & 0.01382529 & 0.011814008 & 0.58916135\\\\\n",
       "\t chrX & 9882537 & 9883461 & WWC3      & K562 &  0.0318434754 & 0.01303893 & 0.013951028 & 0.08499545\\\\\n",
       "\t chrX & 9885193 & 9885478 & CLCN4     & K562 &  0.0304057499 & 0.01461636 & 0.008018696 & 0.06280515\\\\\n",
       "\t chrX & 9885193 & 9885478 & FAM9B     & K562 &  0.0260712046 & 0.01419383 & 0.006425263 & 0.08316111\\\\\n",
       "\t chrX & 9885193 & 9885478 & TBL1X     & K562 &  0.0138417233 & 0.01209908 & 0.005940276 & 0.25685817\\\\\n",
       "\t chrX & 9885193 & 9885478 & WWC3      & K562 &  0.0286078138 & 0.01493548 & 0.008122034 & 0.08509116\\\\\n",
       "\t chrX & 9894528 & 9895197 & CLCN4     & K562 &  0.0020997138 & 0.01403380 & 0.004577434 & 0.57007054\\\\\n",
       "\t chrX & 9894528 & 9895197 & FAM9B     & K562 & -0.0039155294 & 0.01641407 & 0.006766755 & 0.74241234\\\\\n",
       "\t chrX & 9894528 & 9895197 & TBL1X     & K562 &  0.0221218213 & 0.01131808 & 0.005799689 & 0.07463291\\\\\n",
       "\t chrX & 9894528 & 9895197 & WWC3      & K562 &  0.0194707907 & 0.01600771 & 0.008696931 & 0.25046029\\\\\n",
       "\t chrX &  989766 &  990624 & AKAP17A   & K562 &  0.0151072702 & 0.01450843 & 0.019885247 & 0.62904471\\\\\n",
       "\t chrX &  989766 &  990624 & ASMTL     & K562 &  0.0258398681 & 0.01214658 & 0.017759999 & 0.25296218\\\\\n",
       "\t chrX &  989766 &  990624 & CSF2RA    & K562 & -0.0018385678 & 0.01172884 & 0.009957915 & 0.84273632\\\\\n",
       "\t chrX &  989766 &  990624 & GTPBP6    & K562 &  0.0103101238 & 0.01257263 & 0.018756129 & 0.74913721\\\\\n",
       "\t chrX &  989766 &  990624 & IL3RA     & K562 &  0.0422719623 & 0.01384934 & 0.010607540 & 0.01111693\\\\\n",
       "\t chrX &  989766 &  990624 & LINC00106 & K562 &  0.0194724634 & 0.01475951 & 0.011179045 & 0.28709096\\\\\n",
       "\t chrX &  989766 &  990624 & P2RY8     & K562 &  0.0162448765 & 0.01206066 & 0.004982340 & 0.17519737\\\\\n",
       "\t chrX &  989766 &  990624 & PLCXD1    & K562 &  0.0222317067 & 0.01357990 & 0.030507948 & 0.72888618\\\\\n",
       "\t chrX &  989766 &  990624 & PPP2R3B   & K562 &  0.0374172544 & 0.01480182 & 0.022819064 & 0.16200695\\\\\n",
       "\t chrX &  989766 &  990624 & SHOX      & K562 & -0.0093762835 & 0.01501508 & 0.002963468 & 0.79441139\\\\\n",
       "\t chrX &  989766 &  990624 & SLC25A6   & K562 &  0.0254863612 & 0.01467255 & 0.040325540 & 0.84407702\\\\\n",
       "\t chrX & 9940388 & 9940697 & CLCN4     & K562 & -0.0077134263 & 0.01366020 & 0.003995483 & 0.80432036\\\\\n",
       "\t chrX & 9940388 & 9940697 & FAM9B     & K562 &  0.0084791372 & 0.01289502 & 0.002140178 & 0.31150743\\\\\n",
       "\t chrX & 9940388 & 9940697 & TBL1X     & K562 &  0.0119357733 & 0.01079224 & 0.003320242 & 0.21234587\\\\\n",
       "\t chrX & 9940388 & 9940697 & WWC3      & K562 & -0.0114250295 & 0.01547033 & 0.003105238 & 0.82619482\\\\\n",
       "\t chrX & 9995829 & 9996183 & CLCN4     & K562 &  0.0004431027 & 0.01459155 & 0.011974109 & 0.78530973\\\\\n",
       "\t chrX & 9995829 & 9996183 & FAM9B     & K562 &  0.0039833415 & 0.01583196 & 0.007613770 & 0.59068606\\\\\n",
       "\t chrX & 9995829 & 9996183 & TBL1X     & K562 &  0.0122417538 & 0.01309076 & 0.006128604 & 0.32025643\\\\\n",
       "\t chrX & 9995829 & 9996183 & WWC3      & K562 & -0.0107756109 & 0.01706612 & 0.010853665 & 0.89749045\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2652977 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  10002583 |  10003163 | CASZ1    | K562 | -0.0081242737 | 0.01896046 |  0.0061448005 | 0.77414583 |\n",
       "| chr1 |  10002583 |  10003163 | CLSTN1   | K562 |  0.0006577892 | 0.01408155 |  0.0106125084 | 0.76019617 |\n",
       "| chr1 |  10002583 |  10003163 | CTNNBIP1 | K562 | -0.0048440374 | 0.01293704 |  0.0059850509 | 0.79872077 |\n",
       "| chr1 |  10002583 |  10003163 | DFFA     | K562 |  0.0057498484 | 0.01388333 |  0.0085094908 | 0.57878014 |\n",
       "| chr1 |  10002583 |  10003163 | GPR157   | K562 | -0.0073515484 | 0.01263999 |  0.0003868547 | 0.72980256 |\n",
       "| chr1 |  10002583 |  10003163 | H6PD     | K562 |  0.0145440003 | 0.01367281 |  0.0060959871 | 0.26833059 |\n",
       "| chr1 |  10002583 |  10003163 | KIF1B    | K562 |  0.0099476403 | 0.01262953 |  0.0030629305 | 0.29283281 |\n",
       "| chr1 |  10002583 |  10003163 | LZIC     | K562 | -0.0086760439 | 0.01471175 |  0.0064147802 | 0.84749929 |\n",
       "| chr1 |  10002583 |  10003163 | MIR34AHG | K562 | -0.0062829789 | 0.01508511 |  0.0029525886 | 0.72980739 |\n",
       "| chr1 |  10002583 |  10003163 | NMNAT1   | K562 |  0.0004576080 | 0.01553630 |  0.0033567967 | 0.57401580 |\n",
       "| chr1 |  10002583 |  10003163 | PEX14    | K562 | -0.0021044751 | 0.01228127 |  0.0061407749 | 0.74900681 |\n",
       "| chr1 |  10002583 |  10003163 | PGD      | K562 | -0.0097266341 | 0.01364105 |  0.0109669256 | 0.93536747 |\n",
       "| chr1 |  10002583 |  10003163 | PIK3CD   | K562 |  0.0116406203 | 0.01159300 |  0.0069426180 | 0.34264887 |\n",
       "| chr1 |  10002583 |  10003163 | SLC25A33 | K562 | -0.0157542699 | 0.01327563 |  0.0052311207 | 0.94303184 |\n",
       "| chr1 |  10002583 |  10003163 | SLC2A5   | K562 | -0.0077900364 | 0.01531539 |  0.0020858030 | 0.74048171 |\n",
       "| chr1 |  10002583 |  10003163 | SPSB1    | K562 |  0.0101172609 | 0.01361995 |  0.0016742368 | 0.26766147 |\n",
       "| chr1 |  10002583 |  10003163 | TMEM201  | K562 | -0.0138773320 | 0.01204423 |  0.0050178666 | 0.94165470 |\n",
       "| chr1 |  10002583 |  10003163 | UBE4B    | K562 |  0.0206362252 | 0.01192020 |  0.0026118996 | 0.06525638 |\n",
       "| chr1 | 100028213 | 100029660 | AGL      | K562 |  0.0196482001 | 0.01279492 |  0.0140633530 | 0.33124088 |\n",
       "| chr1 | 100028213 | 100029660 | CDC14A   | K562 |  0.0151807925 | 0.01109560 |  0.0050404750 | 0.18038353 |\n",
       "| chr1 | 100028213 | 100029660 | DBT      | K562 |  0.0200868099 | 0.01393867 |  0.0153017161 | 0.36568774 |\n",
       "| chr1 | 100028213 | 100029660 | DPH5     | K562 |  0.0352576269 | 0.01341538 |  0.0146437083 | 0.06219701 |\n",
       "| chr1 | 100028213 | 100029660 | EXTL2    | K562 |  0.0113832481 | 0.01286317 |  0.0091136428 | 0.42997327 |\n",
       "| chr1 | 100028213 | 100029660 | FRRS1    | K562 | -0.0086361998 | 0.01525691 |  0.0029532273 | 0.77625857 |\n",
       "| chr1 | 100028213 | 100029660 | LRRC39   | K562 |  0.0183022959 | 0.01240822 | -0.0015795927 | 0.05454285 |\n",
       "| chr1 | 100028213 | 100029660 | MFSD14A  | K562 | -0.0023994597 | 0.01303874 |  0.0002541471 | 0.58063459 |\n",
       "| chr1 | 100028213 | 100029660 | PALMD    | K562 |  0.0133023695 | 0.01394112 |  0.0043858530 | 0.26122161 |\n",
       "| chr1 | 100028213 | 100029660 | PLPPR4   | K562 | -0.0041573131 | 0.01675531 |  0.0045401385 | 0.69814969 |\n",
       "| chr1 | 100028213 | 100029660 | RTCA     | K562 |  0.0223314283 | 0.01298133 |  0.0091116599 | 0.15425158 |\n",
       "| chr1 | 100028213 | 100029660 | RTCA-AS1 | K562 |  0.0206360588 | 0.01278954 |  0.0101930946 | 0.20710027 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9882537 | 9883461 | FAM9B     | K562 |  0.0495612400 | 0.01628201 | 0.015729547 | 0.01886123 |\n",
       "| chrX | 9882537 | 9883461 | TBL1X     | K562 |  0.0086979517 | 0.01382529 | 0.011814008 | 0.58916135 |\n",
       "| chrX | 9882537 | 9883461 | WWC3      | K562 |  0.0318434754 | 0.01303893 | 0.013951028 | 0.08499545 |\n",
       "| chrX | 9885193 | 9885478 | CLCN4     | K562 |  0.0304057499 | 0.01461636 | 0.008018696 | 0.06280515 |\n",
       "| chrX | 9885193 | 9885478 | FAM9B     | K562 |  0.0260712046 | 0.01419383 | 0.006425263 | 0.08316111 |\n",
       "| chrX | 9885193 | 9885478 | TBL1X     | K562 |  0.0138417233 | 0.01209908 | 0.005940276 | 0.25685817 |\n",
       "| chrX | 9885193 | 9885478 | WWC3      | K562 |  0.0286078138 | 0.01493548 | 0.008122034 | 0.08509116 |\n",
       "| chrX | 9894528 | 9895197 | CLCN4     | K562 |  0.0020997138 | 0.01403380 | 0.004577434 | 0.57007054 |\n",
       "| chrX | 9894528 | 9895197 | FAM9B     | K562 | -0.0039155294 | 0.01641407 | 0.006766755 | 0.74241234 |\n",
       "| chrX | 9894528 | 9895197 | TBL1X     | K562 |  0.0221218213 | 0.01131808 | 0.005799689 | 0.07463291 |\n",
       "| chrX | 9894528 | 9895197 | WWC3      | K562 |  0.0194707907 | 0.01600771 | 0.008696931 | 0.25046029 |\n",
       "| chrX |  989766 |  990624 | AKAP17A   | K562 |  0.0151072702 | 0.01450843 | 0.019885247 | 0.62904471 |\n",
       "| chrX |  989766 |  990624 | ASMTL     | K562 |  0.0258398681 | 0.01214658 | 0.017759999 | 0.25296218 |\n",
       "| chrX |  989766 |  990624 | CSF2RA    | K562 | -0.0018385678 | 0.01172884 | 0.009957915 | 0.84273632 |\n",
       "| chrX |  989766 |  990624 | GTPBP6    | K562 |  0.0103101238 | 0.01257263 | 0.018756129 | 0.74913721 |\n",
       "| chrX |  989766 |  990624 | IL3RA     | K562 |  0.0422719623 | 0.01384934 | 0.010607540 | 0.01111693 |\n",
       "| chrX |  989766 |  990624 | LINC00106 | K562 |  0.0194724634 | 0.01475951 | 0.011179045 | 0.28709096 |\n",
       "| chrX |  989766 |  990624 | P2RY8     | K562 |  0.0162448765 | 0.01206066 | 0.004982340 | 0.17519737 |\n",
       "| chrX |  989766 |  990624 | PLCXD1    | K562 |  0.0222317067 | 0.01357990 | 0.030507948 | 0.72888618 |\n",
       "| chrX |  989766 |  990624 | PPP2R3B   | K562 |  0.0374172544 | 0.01480182 | 0.022819064 | 0.16200695 |\n",
       "| chrX |  989766 |  990624 | SHOX      | K562 | -0.0093762835 | 0.01501508 | 0.002963468 | 0.79441139 |\n",
       "| chrX |  989766 |  990624 | SLC25A6   | K562 |  0.0254863612 | 0.01467255 | 0.040325540 | 0.84407702 |\n",
       "| chrX | 9940388 | 9940697 | CLCN4     | K562 | -0.0077134263 | 0.01366020 | 0.003995483 | 0.80432036 |\n",
       "| chrX | 9940388 | 9940697 | FAM9B     | K562 |  0.0084791372 | 0.01289502 | 0.002140178 | 0.31150743 |\n",
       "| chrX | 9940388 | 9940697 | TBL1X     | K562 |  0.0119357733 | 0.01079224 | 0.003320242 | 0.21234587 |\n",
       "| chrX | 9940388 | 9940697 | WWC3      | K562 | -0.0114250295 | 0.01547033 | 0.003105238 | 0.82619482 |\n",
       "| chrX | 9995829 | 9996183 | CLCN4     | K562 |  0.0004431027 | 0.01459155 | 0.011974109 | 0.78530973 |\n",
       "| chrX | 9995829 | 9996183 | FAM9B     | K562 |  0.0039833415 | 0.01583196 | 0.007613770 | 0.59068606 |\n",
       "| chrX | 9995829 | 9996183 | TBL1X     | K562 |  0.0122417538 | 0.01309076 | 0.006128604 | 0.32025643 |\n",
       "| chrX | 9995829 | 9996183 | WWC3      | K562 | -0.0107756109 | 0.01706612 | 0.010853665 | 0.89749045 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType rObs          rBgSD     \n",
       "1       chr1  10002583  10003163 CASZ1      K562     -0.0081242737 0.01896046\n",
       "2       chr1  10002583  10003163 CLSTN1     K562      0.0006577892 0.01408155\n",
       "3       chr1  10002583  10003163 CTNNBIP1   K562     -0.0048440374 0.01293704\n",
       "4       chr1  10002583  10003163 DFFA       K562      0.0057498484 0.01388333\n",
       "5       chr1  10002583  10003163 GPR157     K562     -0.0073515484 0.01263999\n",
       "6       chr1  10002583  10003163 H6PD       K562      0.0145440003 0.01367281\n",
       "7       chr1  10002583  10003163 KIF1B      K562      0.0099476403 0.01262953\n",
       "8       chr1  10002583  10003163 LZIC       K562     -0.0086760439 0.01471175\n",
       "9       chr1  10002583  10003163 MIR34AHG   K562     -0.0062829789 0.01508511\n",
       "10      chr1  10002583  10003163 NMNAT1     K562      0.0004576080 0.01553630\n",
       "11      chr1  10002583  10003163 PEX14      K562     -0.0021044751 0.01228127\n",
       "12      chr1  10002583  10003163 PGD        K562     -0.0097266341 0.01364105\n",
       "13      chr1  10002583  10003163 PIK3CD     K562      0.0116406203 0.01159300\n",
       "14      chr1  10002583  10003163 SLC25A33   K562     -0.0157542699 0.01327563\n",
       "15      chr1  10002583  10003163 SLC2A5     K562     -0.0077900364 0.01531539\n",
       "16      chr1  10002583  10003163 SPSB1      K562      0.0101172609 0.01361995\n",
       "17      chr1  10002583  10003163 TMEM201    K562     -0.0138773320 0.01204423\n",
       "18      chr1  10002583  10003163 UBE4B      K562      0.0206362252 0.01192020\n",
       "19      chr1 100028213 100029660 AGL        K562      0.0196482001 0.01279492\n",
       "20      chr1 100028213 100029660 CDC14A     K562      0.0151807925 0.01109560\n",
       "21      chr1 100028213 100029660 DBT        K562      0.0200868099 0.01393867\n",
       "22      chr1 100028213 100029660 DPH5       K562      0.0352576269 0.01341538\n",
       "23      chr1 100028213 100029660 EXTL2      K562      0.0113832481 0.01286317\n",
       "24      chr1 100028213 100029660 FRRS1      K562     -0.0086361998 0.01525691\n",
       "25      chr1 100028213 100029660 LRRC39     K562      0.0183022959 0.01240822\n",
       "26      chr1 100028213 100029660 MFSD14A    K562     -0.0023994597 0.01303874\n",
       "27      chr1 100028213 100029660 PALMD      K562      0.0133023695 0.01394112\n",
       "28      chr1 100028213 100029660 PLPPR4     K562     -0.0041573131 0.01675531\n",
       "29      chr1 100028213 100029660 RTCA       K562      0.0223314283 0.01298133\n",
       "30      chr1 100028213 100029660 RTCA-AS1   K562      0.0206360588 0.01278954\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮         \n",
       "2652948 chrX 9882537   9883461   FAM9B      K562      0.0495612400 0.01628201\n",
       "2652949 chrX 9882537   9883461   TBL1X      K562      0.0086979517 0.01382529\n",
       "2652950 chrX 9882537   9883461   WWC3       K562      0.0318434754 0.01303893\n",
       "2652951 chrX 9885193   9885478   CLCN4      K562      0.0304057499 0.01461636\n",
       "2652952 chrX 9885193   9885478   FAM9B      K562      0.0260712046 0.01419383\n",
       "2652953 chrX 9885193   9885478   TBL1X      K562      0.0138417233 0.01209908\n",
       "2652954 chrX 9885193   9885478   WWC3       K562      0.0286078138 0.01493548\n",
       "2652955 chrX 9894528   9895197   CLCN4      K562      0.0020997138 0.01403380\n",
       "2652956 chrX 9894528   9895197   FAM9B      K562     -0.0039155294 0.01641407\n",
       "2652957 chrX 9894528   9895197   TBL1X      K562      0.0221218213 0.01131808\n",
       "2652958 chrX 9894528   9895197   WWC3       K562      0.0194707907 0.01600771\n",
       "2652959 chrX  989766    990624   AKAP17A    K562      0.0151072702 0.01450843\n",
       "2652960 chrX  989766    990624   ASMTL      K562      0.0258398681 0.01214658\n",
       "2652961 chrX  989766    990624   CSF2RA     K562     -0.0018385678 0.01172884\n",
       "2652962 chrX  989766    990624   GTPBP6     K562      0.0103101238 0.01257263\n",
       "2652963 chrX  989766    990624   IL3RA      K562      0.0422719623 0.01384934\n",
       "2652964 chrX  989766    990624   LINC00106  K562      0.0194724634 0.01475951\n",
       "2652965 chrX  989766    990624   P2RY8      K562      0.0162448765 0.01206066\n",
       "2652966 chrX  989766    990624   PLCXD1     K562      0.0222317067 0.01357990\n",
       "2652967 chrX  989766    990624   PPP2R3B    K562      0.0374172544 0.01480182\n",
       "2652968 chrX  989766    990624   SHOX       K562     -0.0093762835 0.01501508\n",
       "2652969 chrX  989766    990624   SLC25A6    K562      0.0254863612 0.01467255\n",
       "2652970 chrX 9940388   9940697   CLCN4      K562     -0.0077134263 0.01366020\n",
       "2652971 chrX 9940388   9940697   FAM9B      K562      0.0084791372 0.01289502\n",
       "2652972 chrX 9940388   9940697   TBL1X      K562      0.0119357733 0.01079224\n",
       "2652973 chrX 9940388   9940697   WWC3       K562     -0.0114250295 0.01547033\n",
       "2652974 chrX 9995829   9996183   CLCN4      K562      0.0004431027 0.01459155\n",
       "2652975 chrX 9995829   9996183   FAM9B      K562      0.0039833415 0.01583196\n",
       "2652976 chrX 9995829   9996183   TBL1X      K562      0.0122417538 0.01309076\n",
       "2652977 chrX 9995829   9996183   WWC3       K562     -0.0107756109 0.01706612\n",
       "        rBgMean       pvalZ     \n",
       "1        0.0061448005 0.77414583\n",
       "2        0.0106125084 0.76019617\n",
       "3        0.0059850509 0.79872077\n",
       "4        0.0085094908 0.57878014\n",
       "5        0.0003868547 0.72980256\n",
       "6        0.0060959871 0.26833059\n",
       "7        0.0030629305 0.29283281\n",
       "8        0.0064147802 0.84749929\n",
       "9        0.0029525886 0.72980739\n",
       "10       0.0033567967 0.57401580\n",
       "11       0.0061407749 0.74900681\n",
       "12       0.0109669256 0.93536747\n",
       "13       0.0069426180 0.34264887\n",
       "14       0.0052311207 0.94303184\n",
       "15       0.0020858030 0.74048171\n",
       "16       0.0016742368 0.26766147\n",
       "17       0.0050178666 0.94165470\n",
       "18       0.0026118996 0.06525638\n",
       "19       0.0140633530 0.33124088\n",
       "20       0.0050404750 0.18038353\n",
       "21       0.0153017161 0.36568774\n",
       "22       0.0146437083 0.06219701\n",
       "23       0.0091136428 0.42997327\n",
       "24       0.0029532273 0.77625857\n",
       "25      -0.0015795927 0.05454285\n",
       "26       0.0002541471 0.58063459\n",
       "27       0.0043858530 0.26122161\n",
       "28       0.0045401385 0.69814969\n",
       "29       0.0091116599 0.15425158\n",
       "30       0.0101930946 0.20710027\n",
       "⋮       ⋮             ⋮         \n",
       "2652948 0.015729547   0.01886123\n",
       "2652949 0.011814008   0.58916135\n",
       "2652950 0.013951028   0.08499545\n",
       "2652951 0.008018696   0.06280515\n",
       "2652952 0.006425263   0.08316111\n",
       "2652953 0.005940276   0.25685817\n",
       "2652954 0.008122034   0.08509116\n",
       "2652955 0.004577434   0.57007054\n",
       "2652956 0.006766755   0.74241234\n",
       "2652957 0.005799689   0.07463291\n",
       "2652958 0.008696931   0.25046029\n",
       "2652959 0.019885247   0.62904471\n",
       "2652960 0.017759999   0.25296218\n",
       "2652961 0.009957915   0.84273632\n",
       "2652962 0.018756129   0.74913721\n",
       "2652963 0.010607540   0.01111693\n",
       "2652964 0.011179045   0.28709096\n",
       "2652965 0.004982340   0.17519737\n",
       "2652966 0.030507948   0.72888618\n",
       "2652967 0.022819064   0.16200695\n",
       "2652968 0.002963468   0.79441139\n",
       "2652969 0.040325540   0.84407702\n",
       "2652970 0.003995483   0.80432036\n",
       "2652971 0.002140178   0.31150743\n",
       "2652972 0.003320242   0.21234587\n",
       "2652973 0.003105238   0.82619482\n",
       "2652974 0.011974109   0.78530973\n",
       "2652975 0.007613770   0.59068606\n",
       "2652976 0.006128604   0.32025643\n",
       "2652977 0.010853665   0.89749045"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.9 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   foreach_1.5.2                    \n",
       " [3] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.68.0                  \n",
       " [5] rtracklayer_1.60.1                Biostrings_2.70.1                \n",
       " [7] XVector_0.42.0                    genomation_1.34.0                \n",
       " [9] Seurat_5.0.3                      SeuratObject_5.0.1               \n",
       "[11] sp_2.1-4                          FigR_0.1.0                       \n",
       "[13] motifmatchr_1.24.0                chromVAR_1.24.0                  \n",
       "[15] cowplot_1.1.3                     ggplot2_3.5.1                    \n",
       "[17] dplyr_1.1.4                       SummarizedExperiment_1.30.2      \n",
       "[19] Biobase_2.60.0                    GenomicRanges_1.54.1             \n",
       "[21] GenomeInfoDb_1.38.1               IRanges_2.36.0                   \n",
       "[23] S4Vectors_0.40.2                  BiocGenerics_0.48.1              \n",
       "[25] MatrixGenerics_1.12.3             matrixStats_1.3.0                \n",
       "[27] Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             vroom_1.6.5                \n",
       " [15] globals_0.16.3              lattice_0.22-6             \n",
       " [17] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [19] plotly_4.10.4               plotrix_3.8-4              \n",
       " [21] yaml_2.3.8                  httpuv_1.6.15              \n",
       " [23] sctransform_0.4.1           spam_2.10-0                \n",
       " [25] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [27] pbapply_1.7-2               DBI_1.2.2                  \n",
       " [29] CNEr_1.38.0                 RColorBrewer_1.1-3         \n",
       " [31] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [33] Rtsne_0.17                  purrr_1.0.2                \n",
       " [35] R.utils_2.12.3              RCurl_1.98-1.14            \n",
       " [37] pracma_2.4.4                GenomeInfoDbData_1.2.11    \n",
       " [39] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [41] spatstat.utils_3.0-4        listenv_0.9.1              \n",
       " [43] seqLogo_1.68.0              goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             spatstat.random_3.2-3      \n",
       " [47] annotate_1.80.0             fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           leiden_0.4.3.1             \n",
       " [51] codetools_0.2-20            DelayedArray_0.26.7        \n",
       " [53] DT_0.33                     tidyselect_1.2.1           \n",
       " [55] spatstat.explore_3.2-6      base64enc_0.1-3            \n",
       " [57] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [59] progressr_0.14.0            iterators_1.0.14           \n",
       " [61] ggridges_0.5.6              survival_3.6-4             \n",
       " [63] tools_4.3.3                 TFMPvalue_0.0.9            \n",
       " [65] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [67] glue_1.7.0                  gridExtra_2.3              \n",
       " [69] IRdisplay_1.1               withr_3.0.0                \n",
       " [71] fastmap_1.1.1               fansi_1.0.6                \n",
       " [73] caTools_1.18.2              digest_0.6.35              \n",
       " [75] R6_2.5.1                    mime_0.12                  \n",
       " [77] nabor_0.5.0                 seqPattern_1.34.0          \n",
       " [79] colorspace_2.1-0            scattermore_1.2            \n",
       " [81] GO.db_3.18.0                tensor_1.5                 \n",
       " [83] gtools_3.9.5                poweRlaw_0.80.0            \n",
       " [85] spatstat.data_3.0-4         RSQLite_2.3.6              \n",
       " [87] R.methodsS3_1.8.2           utf8_1.2.4                 \n",
       " [89] tidyr_1.3.1                 generics_0.1.3             \n",
       " [91] data.table_1.15.2           httr_1.4.7                 \n",
       " [93] htmlwidgets_1.6.4           S4Arrays_1.2.1             \n",
       " [95] TFBSTools_1.40.0            uwot_0.1.16                \n",
       " [97] pkgconfig_2.0.3             gtable_0.3.5               \n",
       " [99] blob_1.2.4                  impute_1.76.0              \n",
       "[101] lmtest_0.9-40               htmltools_0.5.8.1          \n",
       "[103] dotCall64_1.1-1             scales_1.3.0               \n",
       "[105] png_0.1-8                   tzdb_0.4.0                 \n",
       "[107] reshape2_1.4.4              rjson_0.2.21               \n",
       "[109] uuid_1.2-0                  nlme_3.1-164               \n",
       "[111] repr_1.1.7                  cachem_1.0.8               \n",
       "[113] zoo_1.8-12                  stringr_1.5.1              \n",
       "[115] KernSmooth_2.23-22          miniUI_0.1.1.1             \n",
       "[117] AnnotationDbi_1.64.1        restfulr_0.0.15            \n",
       "[119] pillar_1.9.0                vctrs_0.6.5                \n",
       "[121] RANN_2.6.1                  promises_1.3.0             \n",
       "[123] xtable_1.8-4                cluster_2.1.6              \n",
       "[125] evaluate_0.23               readr_2.1.5                \n",
       "[127] cli_3.6.2                   compiler_4.3.3             \n",
       "[129] Rsamtools_2.18.0            rlang_1.1.3                \n",
       "[131] crayon_1.5.2                future.apply_1.11.2        \n",
       "[133] plyr_1.8.9                  stringi_1.8.3              \n",
       "[135] gridBase_0.4-7              deldir_2.0-4               \n",
       "[137] viridisLite_0.4.2           BiocParallel_1.36.0        \n",
       "[139] munsell_0.5.1               lazyeval_0.2.2             \n",
       "[141] spatstat.geom_3.2-9         IRkernel_1.3.2             \n",
       "[143] RcppHNSW_0.6.0              hms_1.1.3                  \n",
       "[145] patchwork_1.2.0             bit64_4.0.5                \n",
       "[147] future_1.33.2               KEGGREST_1.40.1            \n",
       "[149] shiny_1.8.1.1               ROCR_1.0-11                \n",
       "[151] igraph_2.0.3                memoise_2.0.1              \n",
       "[153] bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
