{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240610/sc-E2G/test/results/K562_IGVF_filter/K562/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240610/sc-E2G/test/results/K562_IGVF_filter/K562/Predictions/EnhancerPredictionsAllPutative.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240610/sc-E2G/test/results/K562_IGVF_filter/K562/Kendall/atac_matrix.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_IGVF/1.prepare_data/2.Filter_out_low_quality_cells.240614/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_IGVF/3.Genome_wide_prediction/FigR/FigR.240615/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3001947 ranges and 3 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10002642-10003186      * | APITD1-CORT\n",
       "        [2]     chr1 10002642-10003186      * |       CASZ1\n",
       "        [3]     chr1 10002642-10003186      * |       CENPS\n",
       "        [4]     chr1 10002642-10003186      * |      CLSTN1\n",
       "        [5]     chr1 10002642-10003186      * |    CTNNBIP1\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [3001943]     chrX   9995829-9996183      * |       CLCN4\n",
       "  [3001944]     chrX   9995829-9996183      * |      CLDN34\n",
       "  [3001945]     chrX   9995829-9996183      * |       FAM9B\n",
       "  [3001946]     chrX   9995829-9996183      * |       TBL1X\n",
       "  [3001947]     chrX   9995829-9996183      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-10002642-10003186 chr1-10002642-100031..\n",
       "        [2] chr1-10002642-10003186 chr1-10002642-100031..\n",
       "        [3] chr1-10002642-10003186 chr1-10002642-100031..\n",
       "        [4] chr1-10002642-10003186 chr1-10002642-100031..\n",
       "        [5] chr1-10002642-10003186 chr1-10002642-100031..\n",
       "        ...                    ...                    ...\n",
       "  [3001943]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  [3001944]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  [3001945]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  [3001946]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  [3001947]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  4000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  4001  to  4459 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 153222 4459 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(153222): chr1-10002642-10003186 chr1-100028297-100029654 ...\n",
       "  chrX-9966874-9967060 chrX-9995829-9996183\n",
       "rowData names(2): PeakName bias\n",
       "colnames(4459): e10l1_AAACCAGGTCACAGGG-1 e10l1_AAACCGCCAAGGAATA-1 ...\n",
       "  e10l2_TTTGTGAGTTAGCGTT-1 e10l2_TTTGTGAGTTCCGGAC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna),\n",
    "                       index = 1:nrow(matrix.rna))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  19.6696245670319 secs \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Time Elapsed:  36.5908469438553 mins \n",
      "\n",
      "[1] \"chr10\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.85596561431885 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  13.3825923800468 mins \n",
      "\n",
      "[1] \"chr11\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  15.0050194263458 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  22.0842373967171 mins \n",
      "\n",
      "[1] \"chr12\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  9.18919968605042 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  16.6003611286481 mins \n",
      "\n",
      "[1] \"chr13\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.61435842514038 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  7.89692099094391 mins \n",
      "\n",
      "[1] \"chr14\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.63823127746582 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.67579548756282 mins \n",
      "\n",
      "[1] \"chr15\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.48868322372437 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  10.314786930879 mins \n",
      "\n",
      "[1] \"chr16\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  11.8507189750671 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  22.0904416243235 mins \n",
      "\n",
      "[1] \"chr17\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  15.2019395828247 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  27.4870441595713 mins \n",
      "\n",
      "[1] \"chr18\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.66282486915588 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  7.69378431638082 mins \n",
      "\n",
      "[1] \"chr19\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  23.2831583023071 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  37.9410207708677 mins \n",
      "\n",
      "[1] \"chr2\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.1543688774109 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  17.7023610432943 mins \n",
      "\n",
      "[1] \"chr20\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.40900278091431 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  14.4344625155131 mins \n",
      "\n",
      "[1] \"chr21\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.6095244884491 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.74216038386027 mins \n",
      "\n",
      "[1] \"chr22\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.38961315155029 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  15.4954694827398 mins \n",
      "\n",
      "[1] \"chr3\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  12.2226881980896 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  16.5294797499975 mins \n",
      "\n",
      "[1] \"chr4\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.48353433609009 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.92706807454427 mins \n",
      "\n",
      "[1] \"chr5\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.85180354118347 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  13.7569530407588 mins \n",
      "\n",
      "[1] \"chr6\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  12.4104645252228 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  18.2822884043058 mins \n",
      "\n",
      "[1] \"chr7\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  12.5519304275513 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  17.7540306091309 mins \n",
      "\n",
      "[1] \"chr8\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.69827079772949 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.6421656409899 mins \n",
      "\n",
      "[1] \"chr9\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  9.10752701759338 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  16.2118041753769 mins \n",
      "\n",
      "[1] \"chrX\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.55130362510681 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  12.0438080668449 mins \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5a93ede-8b2c-4646-9a36-b27ff71e2958",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2653513 ranges and 7 metadata columns:\n",
       "       seqnames            ranges strand |  TargetGene               PeakName\n",
       "          <Rle>         <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 10002642-10003186      * |       CASZ1 chr1-10002642-10003186\n",
       "  chr1     chr1 10002642-10003186      * |      CLSTN1 chr1-10002642-10003186\n",
       "  chr1     chr1 10002642-10003186      * |    CTNNBIP1 chr1-10002642-10003186\n",
       "  chr1     chr1 10002642-10003186      * |        DFFA chr1-10002642-10003186\n",
       "  chr1     chr1 10002642-10003186      * |      GPR157 chr1-10002642-10003186\n",
       "   ...      ...               ...    ... .         ...                    ...\n",
       "  chrX     chrX   9966874-9967060      * |        WWC3   chrX-9966874-9967060\n",
       "  chrX     chrX   9995829-9996183      * |       CLCN4   chrX-9995829-9996183\n",
       "  chrX     chrX   9995829-9996183      * |       FAM9B   chrX-9995829-9996183\n",
       "  chrX     chrX   9995829-9996183      * |       TBL1X   chrX-9995829-9996183\n",
       "  chrX     chrX   9995829-9996183      * |        WWC3   chrX-9995829-9996183\n",
       "                     PairName        rObs     rBgSD      rBgMean     pvalZ\n",
       "                  <character>   <numeric> <numeric>    <numeric> <numeric>\n",
       "  chr1 chr1-10002642-100031.. -0.00960605 0.0154133   0.00214243  0.777038\n",
       "  chr1 chr1-10002642-100031.. -0.00080545 0.0142397   0.00697229  0.707536\n",
       "  chr1 chr1-10002642-100031.. -0.01299398 0.0155059   0.00498880  0.876923\n",
       "  chr1 chr1-10002642-100031..  0.00521724 0.0146906   0.00269305  0.431788\n",
       "  chr1 chr1-10002642-100031.. -0.00871826 0.0145066  -0.00212222  0.675334\n",
       "   ...                    ...         ...       ...          ...       ...\n",
       "  chrX chrX-9966874-9967060..  0.01144466 0.0171498  0.006272181  0.381476\n",
       "  chrX chrX-9995829-9996183.. -0.00684313 0.0151717  0.004924915  0.781024\n",
       "  chrX chrX-9995829-9996183..  0.00365796 0.0145557  0.006262907  0.571017\n",
       "  chrX chrX-9995829-9996183..  0.00507682 0.0154460 -0.000116925  0.368340\n",
       "  chrX chrX-9995829-9996183.. -0.01392255 0.0175534  0.006681999  0.879766\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2653513 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>CASZ1   </td><td>K562</td><td>-0.0096060484</td><td>0.01541333</td><td> 2.142434e-03</td><td>0.77703819</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>CLSTN1  </td><td>K562</td><td>-0.0008054497</td><td>0.01423974</td><td> 6.972290e-03</td><td>0.70753565</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>CTNNBIP1</td><td>K562</td><td>-0.0129939829</td><td>0.01550587</td><td> 4.988797e-03</td><td>0.87692265</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>DFFA    </td><td>K562</td><td> 0.0052172383</td><td>0.01469062</td><td> 2.693047e-03</td><td>0.43178824</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>GPR157  </td><td>K562</td><td>-0.0087182622</td><td>0.01450663</td><td>-2.122221e-03</td><td>0.67533437</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>H6PD    </td><td>K562</td><td> 0.0078009778</td><td>0.01440673</td><td> 6.882766e-04</td><td>0.31075659</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>KIF1B   </td><td>K562</td><td> 0.0069304915</td><td>0.01510187</td><td>-2.369062e-03</td><td>0.26901716</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>LZIC    </td><td>K562</td><td>-0.0184650589</td><td>0.01455070</td><td> 6.951889e-04</td><td>0.90604582</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>MIR34AHG</td><td>K562</td><td>-0.0075814544</td><td>0.01924649</td><td> 2.161824e-03</td><td>0.69365473</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>NMNAT1  </td><td>K562</td><td>-0.0010100285</td><td>0.01471110</td><td>-3.126793e-03</td><td>0.44279408</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>PEX14   </td><td>K562</td><td>-0.0121290380</td><td>0.01337946</td><td>-8.347778e-04</td><td>0.80070691</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>PGD     </td><td>K562</td><td>-0.0252161331</td><td>0.01464405</td><td> 2.555215e-03</td><td>0.97104812</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>PIK3CD  </td><td>K562</td><td> 0.0119132205</td><td>0.01432623</td><td> 5.503612e-03</td><td>0.32729182</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>SLC25A33</td><td>K562</td><td>-0.0249527298</td><td>0.01567779</td><td> 1.056457e-03</td><td>0.95144043</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>SLC2A5  </td><td>K562</td><td>-0.0093607122</td><td>0.01444122</td><td> 9.074393e-04</td><td>0.76146741</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>SPSB1   </td><td>K562</td><td>-0.0103082946</td><td>0.01439415</td><td> 1.799800e-03</td><td>0.79987685</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>TMEM201 </td><td>K562</td><td>-0.0172728593</td><td>0.01464909</td><td> 9.494190e-04</td><td>0.89323521</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>UBE4B   </td><td>K562</td><td> 0.0205714412</td><td>0.01307370</td><td>-3.572222e-03</td><td>0.03239273</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>AGL     </td><td>K562</td><td> 0.0181149617</td><td>0.01496825</td><td> 6.893536e-03</td><td>0.22672319</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>CDC14A  </td><td>K562</td><td> 0.0136594070</td><td>0.01628053</td><td>-9.244985e-04</td><td>0.18518303</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>DBT     </td><td>K562</td><td> 0.0113747569</td><td>0.01567396</td><td> 7.882345e-03</td><td>0.41183931</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>DPH5    </td><td>K562</td><td> 0.0199049489</td><td>0.01577747</td><td> 7.841147e-03</td><td>0.22224827</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>EXTL2   </td><td>K562</td><td> 0.0048057414</td><td>0.01598051</td><td> 3.959549e-05</td><td>0.38275717</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>FRRS1   </td><td>K562</td><td>-0.0271973758</td><td>0.01762083</td><td>-2.523108e-03</td><td>0.91928671</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>LRRC39  </td><td>K562</td><td>-0.0027678953</td><td>0.01526162</td><td> 9.347857e-04</td><td>0.59584772</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>MFSD14A </td><td>K562</td><td>-0.0030573297</td><td>0.01631812</td><td> 6.485323e-04</td><td>0.58982739</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>PALMD   </td><td>K562</td><td> 0.0177690722</td><td>0.01578930</td><td> 1.717670e-03</td><td>0.15467181</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>PLPPR4  </td><td>K562</td><td>-0.0052978269</td><td>0.01800101</td><td> 4.455497e-03</td><td>0.70602909</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>RTCA    </td><td>K562</td><td> 0.0098032306</td><td>0.01473568</td><td> 2.920263e-03</td><td>0.32021581</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>RTCA-AS1</td><td>K562</td><td> 0.0232658289</td><td>0.01481171</td><td> 7.521928e-03</td><td>0.14390548</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885199</td><td>9885463</td><td>FAM9B    </td><td>K562</td><td> 0.031850530</td><td>0.01523209</td><td> 0.0052494084</td><td>0.0403718632</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885199</td><td>9885463</td><td>TBL1X    </td><td>K562</td><td> 0.021210152</td><td>0.01396888</td><td>-0.0007315222</td><td>0.0581199159</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885199</td><td>9885463</td><td>WWC3     </td><td>K562</td><td> 0.034233352</td><td>0.01624065</td><td> 0.0024082731</td><td>0.0250216090</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895196</td><td>CLCN4    </td><td>K562</td><td> 0.001865242</td><td>0.01566359</td><td> 0.0053783468</td><td>0.5887321216</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895196</td><td>FAM9B    </td><td>K562</td><td>-0.004652758</td><td>0.01666583</td><td> 0.0047236203</td><td>0.7131500459</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895196</td><td>TBL1X    </td><td>K562</td><td> 0.017698823</td><td>0.01445458</td><td>-0.0017474744</td><td>0.0892579741</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895196</td><td>WWC3     </td><td>K562</td><td> 0.011519642</td><td>0.01630682</td><td> 0.0029702083</td><td>0.3000398689</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>AKAP17A  </td><td>K562</td><td>-0.001055460</td><td>0.01460103</td><td> 0.0024386213</td><td>0.5945650092</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>ASMTL    </td><td>K562</td><td> 0.015289069</td><td>0.01378935</td><td> 0.0065987033</td><td>0.2642744040</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>CSF2RA   </td><td>K562</td><td>-0.004989390</td><td>0.01638307</td><td> 0.0049746719</td><td>0.7284701458</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>GTPBP6   </td><td>K562</td><td> 0.002390020</td><td>0.01679011</td><td> 0.0111667657</td><td>0.6994199824</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>IL3RA    </td><td>K562</td><td> 0.049977526</td><td>0.01422610</td><td> 0.0040645397</td><td>0.0006246544</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>LINC00106</td><td>K562</td><td> 0.006358128</td><td>0.01217576</td><td> 0.0028438796</td><td>0.3864335510</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>P2RY8    </td><td>K562</td><td> 0.014640945</td><td>0.01463580</td><td> 0.0020771930</td><td>0.1953286518</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>PLCXD1   </td><td>K562</td><td> 0.004401955</td><td>0.01364984</td><td> 0.0102000165</td><td>0.6644983489</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>PPP2R3B  </td><td>K562</td><td> 0.022713604</td><td>0.01604802</td><td> 0.0079187067</td><td>0.1782867200</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>SHOX     </td><td>K562</td><td>-0.022202358</td><td>0.01393664</td><td>-0.0021165436</td><td>0.9252392403</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>SLC25A6  </td><td>K562</td><td> 0.013008551</td><td>0.01891164</td><td> 0.0242297976</td><td>0.7235269669</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940708</td><td>CLCN4    </td><td>K562</td><td>-0.001994304</td><td>0.01561964</td><td> 0.0017454670</td><td>0.5946129244</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940708</td><td>FAM9B    </td><td>K562</td><td> 0.007636923</td><td>0.01702616</td><td> 0.0029819040</td><td>0.3922713594</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940708</td><td>TBL1X    </td><td>K562</td><td> 0.010044487</td><td>0.01451973</td><td>-0.0011340263</td><td>0.2206841684</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940708</td><td>WWC3     </td><td>K562</td><td>-0.014825054</td><td>0.01567265</td><td> 0.0015824022</td><td>0.8524236249</td></tr>\n",
       "\t<tr><td>chrX</td><td>9966874</td><td>9967060</td><td>CLCN4    </td><td>K562</td><td> 0.014094775</td><td>0.01478269</td><td> 0.0051527791</td><td>0.2726239544</td></tr>\n",
       "\t<tr><td>chrX</td><td>9966874</td><td>9967060</td><td>FAM9B    </td><td>K562</td><td> 0.002890514</td><td>0.01506628</td><td> 0.0034487762</td><td>0.5147789201</td></tr>\n",
       "\t<tr><td>chrX</td><td>9966874</td><td>9967060</td><td>TBL1X    </td><td>K562</td><td> 0.014597917</td><td>0.01475706</td><td> 0.0036394525</td><td>0.2288646312</td></tr>\n",
       "\t<tr><td>chrX</td><td>9966874</td><td>9967060</td><td>WWC3     </td><td>K562</td><td> 0.011444659</td><td>0.01714979</td><td> 0.0062721813</td><td>0.3814762692</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>CLCN4    </td><td>K562</td><td>-0.006843134</td><td>0.01517172</td><td> 0.0049249149</td><td>0.7810242086</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>FAM9B    </td><td>K562</td><td> 0.003657961</td><td>0.01455570</td><td> 0.0062629070</td><td>0.5710170276</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>TBL1X    </td><td>K562</td><td> 0.005076823</td><td>0.01544599</td><td>-0.0001169249</td><td>0.3683403197</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>WWC3     </td><td>K562</td><td>-0.013922548</td><td>0.01755345</td><td> 0.0066819990</td><td>0.8797659369</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2653513 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  10002642 &  10003186 & CASZ1    & K562 & -0.0096060484 & 0.01541333 &  2.142434e-03 & 0.77703819\\\\\n",
       "\t chr1 &  10002642 &  10003186 & CLSTN1   & K562 & -0.0008054497 & 0.01423974 &  6.972290e-03 & 0.70753565\\\\\n",
       "\t chr1 &  10002642 &  10003186 & CTNNBIP1 & K562 & -0.0129939829 & 0.01550587 &  4.988797e-03 & 0.87692265\\\\\n",
       "\t chr1 &  10002642 &  10003186 & DFFA     & K562 &  0.0052172383 & 0.01469062 &  2.693047e-03 & 0.43178824\\\\\n",
       "\t chr1 &  10002642 &  10003186 & GPR157   & K562 & -0.0087182622 & 0.01450663 & -2.122221e-03 & 0.67533437\\\\\n",
       "\t chr1 &  10002642 &  10003186 & H6PD     & K562 &  0.0078009778 & 0.01440673 &  6.882766e-04 & 0.31075659\\\\\n",
       "\t chr1 &  10002642 &  10003186 & KIF1B    & K562 &  0.0069304915 & 0.01510187 & -2.369062e-03 & 0.26901716\\\\\n",
       "\t chr1 &  10002642 &  10003186 & LZIC     & K562 & -0.0184650589 & 0.01455070 &  6.951889e-04 & 0.90604582\\\\\n",
       "\t chr1 &  10002642 &  10003186 & MIR34AHG & K562 & -0.0075814544 & 0.01924649 &  2.161824e-03 & 0.69365473\\\\\n",
       "\t chr1 &  10002642 &  10003186 & NMNAT1   & K562 & -0.0010100285 & 0.01471110 & -3.126793e-03 & 0.44279408\\\\\n",
       "\t chr1 &  10002642 &  10003186 & PEX14    & K562 & -0.0121290380 & 0.01337946 & -8.347778e-04 & 0.80070691\\\\\n",
       "\t chr1 &  10002642 &  10003186 & PGD      & K562 & -0.0252161331 & 0.01464405 &  2.555215e-03 & 0.97104812\\\\\n",
       "\t chr1 &  10002642 &  10003186 & PIK3CD   & K562 &  0.0119132205 & 0.01432623 &  5.503612e-03 & 0.32729182\\\\\n",
       "\t chr1 &  10002642 &  10003186 & SLC25A33 & K562 & -0.0249527298 & 0.01567779 &  1.056457e-03 & 0.95144043\\\\\n",
       "\t chr1 &  10002642 &  10003186 & SLC2A5   & K562 & -0.0093607122 & 0.01444122 &  9.074393e-04 & 0.76146741\\\\\n",
       "\t chr1 &  10002642 &  10003186 & SPSB1    & K562 & -0.0103082946 & 0.01439415 &  1.799800e-03 & 0.79987685\\\\\n",
       "\t chr1 &  10002642 &  10003186 & TMEM201  & K562 & -0.0172728593 & 0.01464909 &  9.494190e-04 & 0.89323521\\\\\n",
       "\t chr1 &  10002642 &  10003186 & UBE4B    & K562 &  0.0205714412 & 0.01307370 & -3.572222e-03 & 0.03239273\\\\\n",
       "\t chr1 & 100028297 & 100029654 & AGL      & K562 &  0.0181149617 & 0.01496825 &  6.893536e-03 & 0.22672319\\\\\n",
       "\t chr1 & 100028297 & 100029654 & CDC14A   & K562 &  0.0136594070 & 0.01628053 & -9.244985e-04 & 0.18518303\\\\\n",
       "\t chr1 & 100028297 & 100029654 & DBT      & K562 &  0.0113747569 & 0.01567396 &  7.882345e-03 & 0.41183931\\\\\n",
       "\t chr1 & 100028297 & 100029654 & DPH5     & K562 &  0.0199049489 & 0.01577747 &  7.841147e-03 & 0.22224827\\\\\n",
       "\t chr1 & 100028297 & 100029654 & EXTL2    & K562 &  0.0048057414 & 0.01598051 &  3.959549e-05 & 0.38275717\\\\\n",
       "\t chr1 & 100028297 & 100029654 & FRRS1    & K562 & -0.0271973758 & 0.01762083 & -2.523108e-03 & 0.91928671\\\\\n",
       "\t chr1 & 100028297 & 100029654 & LRRC39   & K562 & -0.0027678953 & 0.01526162 &  9.347857e-04 & 0.59584772\\\\\n",
       "\t chr1 & 100028297 & 100029654 & MFSD14A  & K562 & -0.0030573297 & 0.01631812 &  6.485323e-04 & 0.58982739\\\\\n",
       "\t chr1 & 100028297 & 100029654 & PALMD    & K562 &  0.0177690722 & 0.01578930 &  1.717670e-03 & 0.15467181\\\\\n",
       "\t chr1 & 100028297 & 100029654 & PLPPR4   & K562 & -0.0052978269 & 0.01800101 &  4.455497e-03 & 0.70602909\\\\\n",
       "\t chr1 & 100028297 & 100029654 & RTCA     & K562 &  0.0098032306 & 0.01473568 &  2.920263e-03 & 0.32021581\\\\\n",
       "\t chr1 & 100028297 & 100029654 & RTCA-AS1 & K562 &  0.0232658289 & 0.01481171 &  7.521928e-03 & 0.14390548\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9885199 & 9885463 & FAM9B     & K562 &  0.031850530 & 0.01523209 &  0.0052494084 & 0.0403718632\\\\\n",
       "\t chrX & 9885199 & 9885463 & TBL1X     & K562 &  0.021210152 & 0.01396888 & -0.0007315222 & 0.0581199159\\\\\n",
       "\t chrX & 9885199 & 9885463 & WWC3      & K562 &  0.034233352 & 0.01624065 &  0.0024082731 & 0.0250216090\\\\\n",
       "\t chrX & 9894528 & 9895196 & CLCN4     & K562 &  0.001865242 & 0.01566359 &  0.0053783468 & 0.5887321216\\\\\n",
       "\t chrX & 9894528 & 9895196 & FAM9B     & K562 & -0.004652758 & 0.01666583 &  0.0047236203 & 0.7131500459\\\\\n",
       "\t chrX & 9894528 & 9895196 & TBL1X     & K562 &  0.017698823 & 0.01445458 & -0.0017474744 & 0.0892579741\\\\\n",
       "\t chrX & 9894528 & 9895196 & WWC3      & K562 &  0.011519642 & 0.01630682 &  0.0029702083 & 0.3000398689\\\\\n",
       "\t chrX &  989766 &  990603 & AKAP17A   & K562 & -0.001055460 & 0.01460103 &  0.0024386213 & 0.5945650092\\\\\n",
       "\t chrX &  989766 &  990603 & ASMTL     & K562 &  0.015289069 & 0.01378935 &  0.0065987033 & 0.2642744040\\\\\n",
       "\t chrX &  989766 &  990603 & CSF2RA    & K562 & -0.004989390 & 0.01638307 &  0.0049746719 & 0.7284701458\\\\\n",
       "\t chrX &  989766 &  990603 & GTPBP6    & K562 &  0.002390020 & 0.01679011 &  0.0111667657 & 0.6994199824\\\\\n",
       "\t chrX &  989766 &  990603 & IL3RA     & K562 &  0.049977526 & 0.01422610 &  0.0040645397 & 0.0006246544\\\\\n",
       "\t chrX &  989766 &  990603 & LINC00106 & K562 &  0.006358128 & 0.01217576 &  0.0028438796 & 0.3864335510\\\\\n",
       "\t chrX &  989766 &  990603 & P2RY8     & K562 &  0.014640945 & 0.01463580 &  0.0020771930 & 0.1953286518\\\\\n",
       "\t chrX &  989766 &  990603 & PLCXD1    & K562 &  0.004401955 & 0.01364984 &  0.0102000165 & 0.6644983489\\\\\n",
       "\t chrX &  989766 &  990603 & PPP2R3B   & K562 &  0.022713604 & 0.01604802 &  0.0079187067 & 0.1782867200\\\\\n",
       "\t chrX &  989766 &  990603 & SHOX      & K562 & -0.022202358 & 0.01393664 & -0.0021165436 & 0.9252392403\\\\\n",
       "\t chrX &  989766 &  990603 & SLC25A6   & K562 &  0.013008551 & 0.01891164 &  0.0242297976 & 0.7235269669\\\\\n",
       "\t chrX & 9940388 & 9940708 & CLCN4     & K562 & -0.001994304 & 0.01561964 &  0.0017454670 & 0.5946129244\\\\\n",
       "\t chrX & 9940388 & 9940708 & FAM9B     & K562 &  0.007636923 & 0.01702616 &  0.0029819040 & 0.3922713594\\\\\n",
       "\t chrX & 9940388 & 9940708 & TBL1X     & K562 &  0.010044487 & 0.01451973 & -0.0011340263 & 0.2206841684\\\\\n",
       "\t chrX & 9940388 & 9940708 & WWC3      & K562 & -0.014825054 & 0.01567265 &  0.0015824022 & 0.8524236249\\\\\n",
       "\t chrX & 9966874 & 9967060 & CLCN4     & K562 &  0.014094775 & 0.01478269 &  0.0051527791 & 0.2726239544\\\\\n",
       "\t chrX & 9966874 & 9967060 & FAM9B     & K562 &  0.002890514 & 0.01506628 &  0.0034487762 & 0.5147789201\\\\\n",
       "\t chrX & 9966874 & 9967060 & TBL1X     & K562 &  0.014597917 & 0.01475706 &  0.0036394525 & 0.2288646312\\\\\n",
       "\t chrX & 9966874 & 9967060 & WWC3      & K562 &  0.011444659 & 0.01714979 &  0.0062721813 & 0.3814762692\\\\\n",
       "\t chrX & 9995829 & 9996183 & CLCN4     & K562 & -0.006843134 & 0.01517172 &  0.0049249149 & 0.7810242086\\\\\n",
       "\t chrX & 9995829 & 9996183 & FAM9B     & K562 &  0.003657961 & 0.01455570 &  0.0062629070 & 0.5710170276\\\\\n",
       "\t chrX & 9995829 & 9996183 & TBL1X     & K562 &  0.005076823 & 0.01544599 & -0.0001169249 & 0.3683403197\\\\\n",
       "\t chrX & 9995829 & 9996183 & WWC3      & K562 & -0.013922548 & 0.01755345 &  0.0066819990 & 0.8797659369\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2653513 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  10002642 |  10003186 | CASZ1    | K562 | -0.0096060484 | 0.01541333 |  2.142434e-03 | 0.77703819 |\n",
       "| chr1 |  10002642 |  10003186 | CLSTN1   | K562 | -0.0008054497 | 0.01423974 |  6.972290e-03 | 0.70753565 |\n",
       "| chr1 |  10002642 |  10003186 | CTNNBIP1 | K562 | -0.0129939829 | 0.01550587 |  4.988797e-03 | 0.87692265 |\n",
       "| chr1 |  10002642 |  10003186 | DFFA     | K562 |  0.0052172383 | 0.01469062 |  2.693047e-03 | 0.43178824 |\n",
       "| chr1 |  10002642 |  10003186 | GPR157   | K562 | -0.0087182622 | 0.01450663 | -2.122221e-03 | 0.67533437 |\n",
       "| chr1 |  10002642 |  10003186 | H6PD     | K562 |  0.0078009778 | 0.01440673 |  6.882766e-04 | 0.31075659 |\n",
       "| chr1 |  10002642 |  10003186 | KIF1B    | K562 |  0.0069304915 | 0.01510187 | -2.369062e-03 | 0.26901716 |\n",
       "| chr1 |  10002642 |  10003186 | LZIC     | K562 | -0.0184650589 | 0.01455070 |  6.951889e-04 | 0.90604582 |\n",
       "| chr1 |  10002642 |  10003186 | MIR34AHG | K562 | -0.0075814544 | 0.01924649 |  2.161824e-03 | 0.69365473 |\n",
       "| chr1 |  10002642 |  10003186 | NMNAT1   | K562 | -0.0010100285 | 0.01471110 | -3.126793e-03 | 0.44279408 |\n",
       "| chr1 |  10002642 |  10003186 | PEX14    | K562 | -0.0121290380 | 0.01337946 | -8.347778e-04 | 0.80070691 |\n",
       "| chr1 |  10002642 |  10003186 | PGD      | K562 | -0.0252161331 | 0.01464405 |  2.555215e-03 | 0.97104812 |\n",
       "| chr1 |  10002642 |  10003186 | PIK3CD   | K562 |  0.0119132205 | 0.01432623 |  5.503612e-03 | 0.32729182 |\n",
       "| chr1 |  10002642 |  10003186 | SLC25A33 | K562 | -0.0249527298 | 0.01567779 |  1.056457e-03 | 0.95144043 |\n",
       "| chr1 |  10002642 |  10003186 | SLC2A5   | K562 | -0.0093607122 | 0.01444122 |  9.074393e-04 | 0.76146741 |\n",
       "| chr1 |  10002642 |  10003186 | SPSB1    | K562 | -0.0103082946 | 0.01439415 |  1.799800e-03 | 0.79987685 |\n",
       "| chr1 |  10002642 |  10003186 | TMEM201  | K562 | -0.0172728593 | 0.01464909 |  9.494190e-04 | 0.89323521 |\n",
       "| chr1 |  10002642 |  10003186 | UBE4B    | K562 |  0.0205714412 | 0.01307370 | -3.572222e-03 | 0.03239273 |\n",
       "| chr1 | 100028297 | 100029654 | AGL      | K562 |  0.0181149617 | 0.01496825 |  6.893536e-03 | 0.22672319 |\n",
       "| chr1 | 100028297 | 100029654 | CDC14A   | K562 |  0.0136594070 | 0.01628053 | -9.244985e-04 | 0.18518303 |\n",
       "| chr1 | 100028297 | 100029654 | DBT      | K562 |  0.0113747569 | 0.01567396 |  7.882345e-03 | 0.41183931 |\n",
       "| chr1 | 100028297 | 100029654 | DPH5     | K562 |  0.0199049489 | 0.01577747 |  7.841147e-03 | 0.22224827 |\n",
       "| chr1 | 100028297 | 100029654 | EXTL2    | K562 |  0.0048057414 | 0.01598051 |  3.959549e-05 | 0.38275717 |\n",
       "| chr1 | 100028297 | 100029654 | FRRS1    | K562 | -0.0271973758 | 0.01762083 | -2.523108e-03 | 0.91928671 |\n",
       "| chr1 | 100028297 | 100029654 | LRRC39   | K562 | -0.0027678953 | 0.01526162 |  9.347857e-04 | 0.59584772 |\n",
       "| chr1 | 100028297 | 100029654 | MFSD14A  | K562 | -0.0030573297 | 0.01631812 |  6.485323e-04 | 0.58982739 |\n",
       "| chr1 | 100028297 | 100029654 | PALMD    | K562 |  0.0177690722 | 0.01578930 |  1.717670e-03 | 0.15467181 |\n",
       "| chr1 | 100028297 | 100029654 | PLPPR4   | K562 | -0.0052978269 | 0.01800101 |  4.455497e-03 | 0.70602909 |\n",
       "| chr1 | 100028297 | 100029654 | RTCA     | K562 |  0.0098032306 | 0.01473568 |  2.920263e-03 | 0.32021581 |\n",
       "| chr1 | 100028297 | 100029654 | RTCA-AS1 | K562 |  0.0232658289 | 0.01481171 |  7.521928e-03 | 0.14390548 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9885199 | 9885463 | FAM9B     | K562 |  0.031850530 | 0.01523209 |  0.0052494084 | 0.0403718632 |\n",
       "| chrX | 9885199 | 9885463 | TBL1X     | K562 |  0.021210152 | 0.01396888 | -0.0007315222 | 0.0581199159 |\n",
       "| chrX | 9885199 | 9885463 | WWC3      | K562 |  0.034233352 | 0.01624065 |  0.0024082731 | 0.0250216090 |\n",
       "| chrX | 9894528 | 9895196 | CLCN4     | K562 |  0.001865242 | 0.01566359 |  0.0053783468 | 0.5887321216 |\n",
       "| chrX | 9894528 | 9895196 | FAM9B     | K562 | -0.004652758 | 0.01666583 |  0.0047236203 | 0.7131500459 |\n",
       "| chrX | 9894528 | 9895196 | TBL1X     | K562 |  0.017698823 | 0.01445458 | -0.0017474744 | 0.0892579741 |\n",
       "| chrX | 9894528 | 9895196 | WWC3      | K562 |  0.011519642 | 0.01630682 |  0.0029702083 | 0.3000398689 |\n",
       "| chrX |  989766 |  990603 | AKAP17A   | K562 | -0.001055460 | 0.01460103 |  0.0024386213 | 0.5945650092 |\n",
       "| chrX |  989766 |  990603 | ASMTL     | K562 |  0.015289069 | 0.01378935 |  0.0065987033 | 0.2642744040 |\n",
       "| chrX |  989766 |  990603 | CSF2RA    | K562 | -0.004989390 | 0.01638307 |  0.0049746719 | 0.7284701458 |\n",
       "| chrX |  989766 |  990603 | GTPBP6    | K562 |  0.002390020 | 0.01679011 |  0.0111667657 | 0.6994199824 |\n",
       "| chrX |  989766 |  990603 | IL3RA     | K562 |  0.049977526 | 0.01422610 |  0.0040645397 | 0.0006246544 |\n",
       "| chrX |  989766 |  990603 | LINC00106 | K562 |  0.006358128 | 0.01217576 |  0.0028438796 | 0.3864335510 |\n",
       "| chrX |  989766 |  990603 | P2RY8     | K562 |  0.014640945 | 0.01463580 |  0.0020771930 | 0.1953286518 |\n",
       "| chrX |  989766 |  990603 | PLCXD1    | K562 |  0.004401955 | 0.01364984 |  0.0102000165 | 0.6644983489 |\n",
       "| chrX |  989766 |  990603 | PPP2R3B   | K562 |  0.022713604 | 0.01604802 |  0.0079187067 | 0.1782867200 |\n",
       "| chrX |  989766 |  990603 | SHOX      | K562 | -0.022202358 | 0.01393664 | -0.0021165436 | 0.9252392403 |\n",
       "| chrX |  989766 |  990603 | SLC25A6   | K562 |  0.013008551 | 0.01891164 |  0.0242297976 | 0.7235269669 |\n",
       "| chrX | 9940388 | 9940708 | CLCN4     | K562 | -0.001994304 | 0.01561964 |  0.0017454670 | 0.5946129244 |\n",
       "| chrX | 9940388 | 9940708 | FAM9B     | K562 |  0.007636923 | 0.01702616 |  0.0029819040 | 0.3922713594 |\n",
       "| chrX | 9940388 | 9940708 | TBL1X     | K562 |  0.010044487 | 0.01451973 | -0.0011340263 | 0.2206841684 |\n",
       "| chrX | 9940388 | 9940708 | WWC3      | K562 | -0.014825054 | 0.01567265 |  0.0015824022 | 0.8524236249 |\n",
       "| chrX | 9966874 | 9967060 | CLCN4     | K562 |  0.014094775 | 0.01478269 |  0.0051527791 | 0.2726239544 |\n",
       "| chrX | 9966874 | 9967060 | FAM9B     | K562 |  0.002890514 | 0.01506628 |  0.0034487762 | 0.5147789201 |\n",
       "| chrX | 9966874 | 9967060 | TBL1X     | K562 |  0.014597917 | 0.01475706 |  0.0036394525 | 0.2288646312 |\n",
       "| chrX | 9966874 | 9967060 | WWC3      | K562 |  0.011444659 | 0.01714979 |  0.0062721813 | 0.3814762692 |\n",
       "| chrX | 9995829 | 9996183 | CLCN4     | K562 | -0.006843134 | 0.01517172 |  0.0049249149 | 0.7810242086 |\n",
       "| chrX | 9995829 | 9996183 | FAM9B     | K562 |  0.003657961 | 0.01455570 |  0.0062629070 | 0.5710170276 |\n",
       "| chrX | 9995829 | 9996183 | TBL1X     | K562 |  0.005076823 | 0.01544599 | -0.0001169249 | 0.3683403197 |\n",
       "| chrX | 9995829 | 9996183 | WWC3      | K562 | -0.013922548 | 0.01755345 |  0.0066819990 | 0.8797659369 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType rObs          rBgSD     \n",
       "1       chr1  10002642  10003186 CASZ1      K562     -0.0096060484 0.01541333\n",
       "2       chr1  10002642  10003186 CLSTN1     K562     -0.0008054497 0.01423974\n",
       "3       chr1  10002642  10003186 CTNNBIP1   K562     -0.0129939829 0.01550587\n",
       "4       chr1  10002642  10003186 DFFA       K562      0.0052172383 0.01469062\n",
       "5       chr1  10002642  10003186 GPR157     K562     -0.0087182622 0.01450663\n",
       "6       chr1  10002642  10003186 H6PD       K562      0.0078009778 0.01440673\n",
       "7       chr1  10002642  10003186 KIF1B      K562      0.0069304915 0.01510187\n",
       "8       chr1  10002642  10003186 LZIC       K562     -0.0184650589 0.01455070\n",
       "9       chr1  10002642  10003186 MIR34AHG   K562     -0.0075814544 0.01924649\n",
       "10      chr1  10002642  10003186 NMNAT1     K562     -0.0010100285 0.01471110\n",
       "11      chr1  10002642  10003186 PEX14      K562     -0.0121290380 0.01337946\n",
       "12      chr1  10002642  10003186 PGD        K562     -0.0252161331 0.01464405\n",
       "13      chr1  10002642  10003186 PIK3CD     K562      0.0119132205 0.01432623\n",
       "14      chr1  10002642  10003186 SLC25A33   K562     -0.0249527298 0.01567779\n",
       "15      chr1  10002642  10003186 SLC2A5     K562     -0.0093607122 0.01444122\n",
       "16      chr1  10002642  10003186 SPSB1      K562     -0.0103082946 0.01439415\n",
       "17      chr1  10002642  10003186 TMEM201    K562     -0.0172728593 0.01464909\n",
       "18      chr1  10002642  10003186 UBE4B      K562      0.0205714412 0.01307370\n",
       "19      chr1 100028297 100029654 AGL        K562      0.0181149617 0.01496825\n",
       "20      chr1 100028297 100029654 CDC14A     K562      0.0136594070 0.01628053\n",
       "21      chr1 100028297 100029654 DBT        K562      0.0113747569 0.01567396\n",
       "22      chr1 100028297 100029654 DPH5       K562      0.0199049489 0.01577747\n",
       "23      chr1 100028297 100029654 EXTL2      K562      0.0048057414 0.01598051\n",
       "24      chr1 100028297 100029654 FRRS1      K562     -0.0271973758 0.01762083\n",
       "25      chr1 100028297 100029654 LRRC39     K562     -0.0027678953 0.01526162\n",
       "26      chr1 100028297 100029654 MFSD14A    K562     -0.0030573297 0.01631812\n",
       "27      chr1 100028297 100029654 PALMD      K562      0.0177690722 0.01578930\n",
       "28      chr1 100028297 100029654 PLPPR4     K562     -0.0052978269 0.01800101\n",
       "29      chr1 100028297 100029654 RTCA       K562      0.0098032306 0.01473568\n",
       "30      chr1 100028297 100029654 RTCA-AS1   K562      0.0232658289 0.01481171\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮         \n",
       "2653484 chrX 9885199   9885463   FAM9B      K562      0.031850530  0.01523209\n",
       "2653485 chrX 9885199   9885463   TBL1X      K562      0.021210152  0.01396888\n",
       "2653486 chrX 9885199   9885463   WWC3       K562      0.034233352  0.01624065\n",
       "2653487 chrX 9894528   9895196   CLCN4      K562      0.001865242  0.01566359\n",
       "2653488 chrX 9894528   9895196   FAM9B      K562     -0.004652758  0.01666583\n",
       "2653489 chrX 9894528   9895196   TBL1X      K562      0.017698823  0.01445458\n",
       "2653490 chrX 9894528   9895196   WWC3       K562      0.011519642  0.01630682\n",
       "2653491 chrX  989766    990603   AKAP17A    K562     -0.001055460  0.01460103\n",
       "2653492 chrX  989766    990603   ASMTL      K562      0.015289069  0.01378935\n",
       "2653493 chrX  989766    990603   CSF2RA     K562     -0.004989390  0.01638307\n",
       "2653494 chrX  989766    990603   GTPBP6     K562      0.002390020  0.01679011\n",
       "2653495 chrX  989766    990603   IL3RA      K562      0.049977526  0.01422610\n",
       "2653496 chrX  989766    990603   LINC00106  K562      0.006358128  0.01217576\n",
       "2653497 chrX  989766    990603   P2RY8      K562      0.014640945  0.01463580\n",
       "2653498 chrX  989766    990603   PLCXD1     K562      0.004401955  0.01364984\n",
       "2653499 chrX  989766    990603   PPP2R3B    K562      0.022713604  0.01604802\n",
       "2653500 chrX  989766    990603   SHOX       K562     -0.022202358  0.01393664\n",
       "2653501 chrX  989766    990603   SLC25A6    K562      0.013008551  0.01891164\n",
       "2653502 chrX 9940388   9940708   CLCN4      K562     -0.001994304  0.01561964\n",
       "2653503 chrX 9940388   9940708   FAM9B      K562      0.007636923  0.01702616\n",
       "2653504 chrX 9940388   9940708   TBL1X      K562      0.010044487  0.01451973\n",
       "2653505 chrX 9940388   9940708   WWC3       K562     -0.014825054  0.01567265\n",
       "2653506 chrX 9966874   9967060   CLCN4      K562      0.014094775  0.01478269\n",
       "2653507 chrX 9966874   9967060   FAM9B      K562      0.002890514  0.01506628\n",
       "2653508 chrX 9966874   9967060   TBL1X      K562      0.014597917  0.01475706\n",
       "2653509 chrX 9966874   9967060   WWC3       K562      0.011444659  0.01714979\n",
       "2653510 chrX 9995829   9996183   CLCN4      K562     -0.006843134  0.01517172\n",
       "2653511 chrX 9995829   9996183   FAM9B      K562      0.003657961  0.01455570\n",
       "2653512 chrX 9995829   9996183   TBL1X      K562      0.005076823  0.01544599\n",
       "2653513 chrX 9995829   9996183   WWC3       K562     -0.013922548  0.01755345\n",
       "        rBgMean       pvalZ       \n",
       "1        2.142434e-03 0.77703819  \n",
       "2        6.972290e-03 0.70753565  \n",
       "3        4.988797e-03 0.87692265  \n",
       "4        2.693047e-03 0.43178824  \n",
       "5       -2.122221e-03 0.67533437  \n",
       "6        6.882766e-04 0.31075659  \n",
       "7       -2.369062e-03 0.26901716  \n",
       "8        6.951889e-04 0.90604582  \n",
       "9        2.161824e-03 0.69365473  \n",
       "10      -3.126793e-03 0.44279408  \n",
       "11      -8.347778e-04 0.80070691  \n",
       "12       2.555215e-03 0.97104812  \n",
       "13       5.503612e-03 0.32729182  \n",
       "14       1.056457e-03 0.95144043  \n",
       "15       9.074393e-04 0.76146741  \n",
       "16       1.799800e-03 0.79987685  \n",
       "17       9.494190e-04 0.89323521  \n",
       "18      -3.572222e-03 0.03239273  \n",
       "19       6.893536e-03 0.22672319  \n",
       "20      -9.244985e-04 0.18518303  \n",
       "21       7.882345e-03 0.41183931  \n",
       "22       7.841147e-03 0.22224827  \n",
       "23       3.959549e-05 0.38275717  \n",
       "24      -2.523108e-03 0.91928671  \n",
       "25       9.347857e-04 0.59584772  \n",
       "26       6.485323e-04 0.58982739  \n",
       "27       1.717670e-03 0.15467181  \n",
       "28       4.455497e-03 0.70602909  \n",
       "29       2.920263e-03 0.32021581  \n",
       "30       7.521928e-03 0.14390548  \n",
       "⋮       ⋮             ⋮           \n",
       "2653484  0.0052494084 0.0403718632\n",
       "2653485 -0.0007315222 0.0581199159\n",
       "2653486  0.0024082731 0.0250216090\n",
       "2653487  0.0053783468 0.5887321216\n",
       "2653488  0.0047236203 0.7131500459\n",
       "2653489 -0.0017474744 0.0892579741\n",
       "2653490  0.0029702083 0.3000398689\n",
       "2653491  0.0024386213 0.5945650092\n",
       "2653492  0.0065987033 0.2642744040\n",
       "2653493  0.0049746719 0.7284701458\n",
       "2653494  0.0111667657 0.6994199824\n",
       "2653495  0.0040645397 0.0006246544\n",
       "2653496  0.0028438796 0.3864335510\n",
       "2653497  0.0020771930 0.1953286518\n",
       "2653498  0.0102000165 0.6644983489\n",
       "2653499  0.0079187067 0.1782867200\n",
       "2653500 -0.0021165436 0.9252392403\n",
       "2653501  0.0242297976 0.7235269669\n",
       "2653502  0.0017454670 0.5946129244\n",
       "2653503  0.0029819040 0.3922713594\n",
       "2653504 -0.0011340263 0.2206841684\n",
       "2653505  0.0015824022 0.8524236249\n",
       "2653506  0.0051527791 0.2726239544\n",
       "2653507  0.0034487762 0.5147789201\n",
       "2653508  0.0036394525 0.2288646312\n",
       "2653509  0.0062721813 0.3814762692\n",
       "2653510  0.0049249149 0.7810242086\n",
       "2653511  0.0062629070 0.5710170276\n",
       "2653512 -0.0001169249 0.3683403197\n",
       "2653513  0.0066819990 0.8797659369"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.9 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   foreach_1.5.2                    \n",
       " [3] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.68.0                  \n",
       " [5] rtracklayer_1.60.1                Biostrings_2.70.1                \n",
       " [7] XVector_0.42.0                    genomation_1.34.0                \n",
       " [9] Seurat_5.0.3                      SeuratObject_5.0.1               \n",
       "[11] sp_2.1-4                          FigR_0.1.0                       \n",
       "[13] motifmatchr_1.24.0                chromVAR_1.24.0                  \n",
       "[15] cowplot_1.1.3                     ggplot2_3.5.1                    \n",
       "[17] dplyr_1.1.4                       SummarizedExperiment_1.30.2      \n",
       "[19] Biobase_2.60.0                    GenomicRanges_1.54.1             \n",
       "[21] GenomeInfoDb_1.38.1               IRanges_2.36.0                   \n",
       "[23] S4Vectors_0.40.2                  BiocGenerics_0.48.1              \n",
       "[25] MatrixGenerics_1.12.3             matrixStats_1.3.0                \n",
       "[27] Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             vroom_1.6.5                \n",
       " [15] globals_0.16.3              lattice_0.22-6             \n",
       " [17] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [19] plotly_4.10.4               plotrix_3.8-4              \n",
       " [21] yaml_2.3.8                  httpuv_1.6.15              \n",
       " [23] sctransform_0.4.1           spam_2.10-0                \n",
       " [25] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [27] pbapply_1.7-2               DBI_1.2.2                  \n",
       " [29] CNEr_1.38.0                 RColorBrewer_1.1-3         \n",
       " [31] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [33] Rtsne_0.17                  purrr_1.0.2                \n",
       " [35] R.utils_2.12.3              RCurl_1.98-1.14            \n",
       " [37] pracma_2.4.4                GenomeInfoDbData_1.2.11    \n",
       " [39] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [41] spatstat.utils_3.0-4        listenv_0.9.1              \n",
       " [43] seqLogo_1.68.0              goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             spatstat.random_3.2-3      \n",
       " [47] annotate_1.80.0             fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           leiden_0.4.3.1             \n",
       " [51] codetools_0.2-20            DelayedArray_0.26.7        \n",
       " [53] DT_0.33                     tidyselect_1.2.1           \n",
       " [55] spatstat.explore_3.2-6      base64enc_0.1-3            \n",
       " [57] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [59] progressr_0.14.0            iterators_1.0.14           \n",
       " [61] ggridges_0.5.6              survival_3.6-4             \n",
       " [63] tools_4.3.3                 TFMPvalue_0.0.9            \n",
       " [65] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [67] glue_1.7.0                  gridExtra_2.3              \n",
       " [69] IRdisplay_1.1               withr_3.0.0                \n",
       " [71] fastmap_1.1.1               fansi_1.0.6                \n",
       " [73] caTools_1.18.2              digest_0.6.35              \n",
       " [75] R6_2.5.1                    mime_0.12                  \n",
       " [77] nabor_0.5.0                 seqPattern_1.34.0          \n",
       " [79] colorspace_2.1-0            scattermore_1.2            \n",
       " [81] GO.db_3.18.0                tensor_1.5                 \n",
       " [83] gtools_3.9.5                poweRlaw_0.80.0            \n",
       " [85] spatstat.data_3.0-4         RSQLite_2.3.6              \n",
       " [87] R.methodsS3_1.8.2           utf8_1.2.4                 \n",
       " [89] tidyr_1.3.1                 generics_0.1.3             \n",
       " [91] data.table_1.15.2           httr_1.4.7                 \n",
       " [93] htmlwidgets_1.6.4           S4Arrays_1.2.1             \n",
       " [95] TFBSTools_1.40.0            uwot_0.1.16                \n",
       " [97] pkgconfig_2.0.3             gtable_0.3.5               \n",
       " [99] blob_1.2.4                  impute_1.76.0              \n",
       "[101] lmtest_0.9-40               htmltools_0.5.8.1          \n",
       "[103] dotCall64_1.1-1             scales_1.3.0               \n",
       "[105] png_0.1-8                   tzdb_0.4.0                 \n",
       "[107] reshape2_1.4.4              rjson_0.2.21               \n",
       "[109] uuid_1.2-0                  nlme_3.1-164               \n",
       "[111] repr_1.1.7                  cachem_1.0.8               \n",
       "[113] zoo_1.8-12                  stringr_1.5.1              \n",
       "[115] KernSmooth_2.23-22          miniUI_0.1.1.1             \n",
       "[117] AnnotationDbi_1.64.1        restfulr_0.0.15            \n",
       "[119] pillar_1.9.0                vctrs_0.6.5                \n",
       "[121] RANN_2.6.1                  promises_1.3.0             \n",
       "[123] xtable_1.8-4                cluster_2.1.6              \n",
       "[125] evaluate_0.23               readr_2.1.5                \n",
       "[127] cli_3.6.2                   compiler_4.3.3             \n",
       "[129] Rsamtools_2.18.0            rlang_1.1.3                \n",
       "[131] crayon_1.5.2                future.apply_1.11.2        \n",
       "[133] plyr_1.8.9                  stringi_1.8.3              \n",
       "[135] gridBase_0.4-7              deldir_2.0-4               \n",
       "[137] viridisLite_0.4.2           BiocParallel_1.36.0        \n",
       "[139] munsell_0.5.1               lazyeval_0.2.2             \n",
       "[141] spatstat.geom_3.2-9         IRkernel_1.3.2             \n",
       "[143] RcppHNSW_0.6.0              hms_1.1.3                  \n",
       "[145] patchwork_1.2.0             bit64_4.0.5                \n",
       "[147] future_1.33.2               KEGGREST_1.40.1            \n",
       "[149] shiny_1.8.1.1               ROCR_1.0-11                \n",
       "[151] igraph_2.0.3                memoise_2.0.1              \n",
       "[153] bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
