{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240610/sc-E2G/test/results/K562_IGVF_filter/K562/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240610/sc-E2G/test/results/K562_IGVF_filter/K562/Predictions/EnhancerPredictionsAllPutative.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240610/sc-E2G/test/results/K562_IGVF_filter/K562/Kendall/atac_matrix.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_IGVF/1.prepare_data/2.Filter_out_low_quality_cells.240614/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_IGVF/3.Genome_wide_prediction/FigR/FigR.250211/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3001947 ranges and 3 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10002642-10003186      * | APITD1-CORT\n",
       "        [2]     chr1 10002642-10003186      * |       CASZ1\n",
       "        [3]     chr1 10002642-10003186      * |       CENPS\n",
       "        [4]     chr1 10002642-10003186      * |      CLSTN1\n",
       "        [5]     chr1 10002642-10003186      * |    CTNNBIP1\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [3001943]     chrX   9995829-9996183      * |       CLCN4\n",
       "  [3001944]     chrX   9995829-9996183      * |      CLDN34\n",
       "  [3001945]     chrX   9995829-9996183      * |       FAM9B\n",
       "  [3001946]     chrX   9995829-9996183      * |       TBL1X\n",
       "  [3001947]     chrX   9995829-9996183      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-10002642-10003186 chr1-10002642-100031..\n",
       "        [2] chr1-10002642-10003186 chr1-10002642-100031..\n",
       "        [3] chr1-10002642-10003186 chr1-10002642-100031..\n",
       "        [4] chr1-10002642-10003186 chr1-10002642-100031..\n",
       "        [5] chr1-10002642-10003186 chr1-10002642-100031..\n",
       "        ...                    ...                    ...\n",
       "  [3001943]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  [3001944]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  [3001945]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  [3001946]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  [3001947]   chrX-9995829-9996183 chrX-9995829-9996183..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  4000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  4001  to  4459 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dd80eda-2392-4d59-8250-1ae62d7e1d1e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b9e26c70-11fd-47c4-8e05-9c96b17d3f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a76b1603-6b00-4ac0-9399-fd341c5a7b88",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 153222 4459 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(153222): chr1-10002642-10003186 chr1-100028297-100029654 ...\n",
       "  chrX-9966874-9967060 chrX-9995829-9996183\n",
       "rowData names(2): PeakName bias\n",
       "colnames(4459): e10l1_AAACCAGGTCACAGGG-1 e10l1_AAACCGCCAAGGAATA-1 ...\n",
       "  e10l2_TTTGTGAGTTAGCGTT-1 e10l2_TTTGTGAGTTCCGGAC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna.rename),\n",
    "                       index = 1:nrow(matrix.rna.rename))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  42.3999555110931 secs \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Time Elapsed:  56.5846376140912 mins \n",
      "\n",
      "[1] \"chr10\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.1192600727081 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  17.0974353909493 mins \n",
      "\n",
      "[1] \"chr11\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  23.6531155109406 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  32.5164919773738 mins \n",
      "\n",
      "[1] \"chr12\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  13.6358208656311 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  24.4348252733548 mins \n",
      "\n",
      "[1] \"chr13\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  3.87738180160522 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  7.41869362990061 mins \n",
      "\n",
      "[1] \"chr14\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.30488634109497 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  13.0410640041033 mins \n",
      "\n",
      "[1] \"chr15\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.87120866775513 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  12.5532987316449 mins \n",
      "\n",
      "[1] \"chr16\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  22.4691076278687 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  32.7927533507347 mins \n",
      "\n",
      "[1] \"chr17\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  24.5315818786621 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  40.7707592447599 mins \n",
      "\n",
      "[1] \"chr18\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  3.54378747940063 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  6.8354917605718 mins \n",
      "\n",
      "[1] \"chr19\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  31.8284912109375 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  52.6393885374069 mins \n",
      "\n",
      "[1] \"chr2\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  14.8055753707886 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  24.2008984923363 mins \n",
      "\n",
      "[1] \"chr20\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  12.3688580989838 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  17.6372516949972 mins \n",
      "\n",
      "[1] \"chr21\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.45904135704041 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.09651641050974 mins \n",
      "\n",
      "[1] \"chr22\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  11.5387699604034 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  18.0806606332461 mins \n",
      "\n",
      "[1] \"chr3\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  12.7251250743866 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  22.6215415596962 mins \n",
      "\n",
      "[1] \"chr4\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.82409238815308 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  13.0166826486588 mins \n",
      "\n",
      "[1] \"chr5\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.4826641082764 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  18.2736454765002 mins \n",
      "\n",
      "[1] \"chr6\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  16.3821475505829 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  27.7717500368754 mins \n",
      "\n",
      "[1] \"chr7\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  14.2230999469757 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  25.9440119306246 mins \n",
      "\n",
      "[1] \"chr8\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  11.8131518363953 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  17.180622680982 mins \n",
      "\n",
      "[1] \"chr9\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.7932796478271 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  22.3220884998639 mins \n",
      "\n",
      "[1] \"chrX\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.2493236064911 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  16.7308589180311 mins \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna.rename,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2772969 ranges and 7 metadata columns:\n",
       "       seqnames            ranges strand |  TargetGene               PeakName\n",
       "          <Rle>         <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 10002642-10003186      * | APITD1-CORT chr1-10002642-10003186\n",
       "  chr1     chr1 10002642-10003186      * |       CASZ1 chr1-10002642-10003186\n",
       "  chr1     chr1 10002642-10003186      * |      CLSTN1 chr1-10002642-10003186\n",
       "  chr1     chr1 10002642-10003186      * |    CTNNBIP1 chr1-10002642-10003186\n",
       "  chr1     chr1 10002642-10003186      * |        DFFA chr1-10002642-10003186\n",
       "   ...      ...               ...    ... .         ...                    ...\n",
       "  chrX     chrX   9966874-9967060      * |        WWC3   chrX-9966874-9967060\n",
       "  chrX     chrX   9995829-9996183      * |       CLCN4   chrX-9995829-9996183\n",
       "  chrX     chrX   9995829-9996183      * |       FAM9B   chrX-9995829-9996183\n",
       "  chrX     chrX   9995829-9996183      * |       TBL1X   chrX-9995829-9996183\n",
       "  chrX     chrX   9995829-9996183      * |        WWC3   chrX-9995829-9996183\n",
       "                     PairName        rObs     rBgSD    rBgMean     pvalZ\n",
       "                  <character>   <numeric> <numeric>  <numeric> <numeric>\n",
       "  chr1 chr1-10002642-100031.. -0.00710964 0.0169253 0.00135690  0.691543\n",
       "  chr1 chr1-10002642-100031.. -0.00960605 0.0166227 0.00362732  0.787014\n",
       "  chr1 chr1-10002642-100031.. -0.00080545 0.0167260 0.00666498  0.672431\n",
       "  chr1 chr1-10002642-100031.. -0.01299398 0.0148232 0.00290785  0.858312\n",
       "  chr1 chr1-10002642-100031..  0.00521724 0.0160604 0.00116324  0.400358\n",
       "   ...                    ...         ...       ...        ...       ...\n",
       "  chrX chrX-9966874-9967060..  0.01144466 0.0185329 0.00529564  0.370024\n",
       "  chrX chrX-9995829-9996183.. -0.00684313 0.0153658 0.00725052  0.820484\n",
       "  chrX chrX-9995829-9996183..  0.00365796 0.0173102 0.00873203  0.615287\n",
       "  chrX chrX-9995829-9996183..  0.00507682 0.0160785 0.00276087  0.442734\n",
       "  chrX chrX-9995829-9996183.. -0.01392255 0.0167966 0.00623455  0.884944\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2772969 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>APITD1-CORT</td><td>K562</td><td>-0.0071096383</td><td>0.01692533</td><td> 0.0013568974</td><td>0.69154301</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>CASZ1      </td><td>K562</td><td>-0.0096060484</td><td>0.01662268</td><td> 0.0036273219</td><td>0.78701402</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>CLSTN1     </td><td>K562</td><td>-0.0008054497</td><td>0.01672599</td><td> 0.0066649755</td><td>0.67243096</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>CTNNBIP1   </td><td>K562</td><td>-0.0129939829</td><td>0.01482323</td><td> 0.0029078481</td><td>0.85831150</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>DFFA       </td><td>K562</td><td> 0.0052172383</td><td>0.01606043</td><td> 0.0011632385</td><td>0.40035758</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>GPR157     </td><td>K562</td><td>-0.0087182622</td><td>0.01737368</td><td> 0.0008530634</td><td>0.70915214</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>H6PD       </td><td>K562</td><td> 0.0078009778</td><td>0.01499688</td><td> 0.0008884468</td><td>0.32242397</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>KIF1B      </td><td>K562</td><td> 0.0069304915</td><td>0.01541464</td><td>-0.0035702761</td><td>0.24786598</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>LINC01759  </td><td>K562</td><td>-0.0021388303</td><td>0.01791484</td><td> 0.0008765883</td><td>0.56683406</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>LZIC       </td><td>K562</td><td>-0.0184650589</td><td>0.01724992</td><td> 0.0023708795</td><td>0.88645442</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>MIR34AHG   </td><td>K562</td><td>-0.0075814544</td><td>0.01740125</td><td> 0.0017516705</td><td>0.70414098</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>NMNAT1     </td><td>K562</td><td>-0.0010100285</td><td>0.01315259</td><td>-0.0030480695</td><td>0.43842896</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>PEX14      </td><td>K562</td><td>-0.0121290380</td><td>0.01221703</td><td>-0.0007436041</td><td>0.82431405</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>PGD        </td><td>K562</td><td>-0.0252161331</td><td>0.01504693</td><td> 0.0070027324</td><td>0.98387204</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>PIK3CD     </td><td>K562</td><td> 0.0119132205</td><td>0.01366802</td><td> 0.0036980615</td><td>0.27390342</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>SLC25A33   </td><td>K562</td><td>-0.0249527298</td><td>0.01589097</td><td> 0.0023425388</td><td>0.95707054</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>SLC2A5     </td><td>K562</td><td>-0.0093607122</td><td>0.01653240</td><td> 0.0022885724</td><td>0.75948084</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>SPSB1      </td><td>K562</td><td>-0.0103082946</td><td>0.01553774</td><td> 0.0020180914</td><td>0.78620403</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>TMEM201    </td><td>K562</td><td>-0.0172728593</td><td>0.01455992</td><td> 0.0004556541</td><td>0.88831662</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10002642</td><td> 10003186</td><td>UBE4B      </td><td>K562</td><td> 0.0205714412</td><td>0.01581837</td><td>-0.0038581716</td><td>0.06124787</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>AGL        </td><td>K562</td><td> 0.0181149617</td><td>0.01591655</td><td> 0.0066421943</td><td>0.23551407</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>CDC14A     </td><td>K562</td><td> 0.0136594070</td><td>0.01790693</td><td>-0.0011760450</td><td>0.20370061</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>DBT        </td><td>K562</td><td> 0.0113747569</td><td>0.01588014</td><td> 0.0078046269</td><td>0.41106069</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>DPH5       </td><td>K562</td><td> 0.0199049489</td><td>0.01626385</td><td> 0.0058230610</td><td>0.19328897</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>EXTL2      </td><td>K562</td><td> 0.0048057414</td><td>0.01540007</td><td> 0.0002475034</td><td>0.38361954</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>FRRS1      </td><td>K562</td><td>-0.0271973758</td><td>0.01867604</td><td>-0.0042273398</td><td>0.89063650</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>LRRC39     </td><td>K562</td><td>-0.0027678953</td><td>0.01490727</td><td>-0.0016713748</td><td>0.52931819</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>MFSD14A    </td><td>K562</td><td>-0.0030573297</td><td>0.01078384</td><td>-0.0015535268</td><td>0.55545259</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>PALMD      </td><td>K562</td><td> 0.0177690722</td><td>0.01424706</td><td> 0.0020953814</td><td>0.13563661</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028297</td><td>100029654</td><td>PLPPR4     </td><td>K562</td><td>-0.0052978269</td><td>0.01321914</td><td>-0.0009890121</td><td>0.62776997</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885199</td><td>9885463</td><td>FAM9B    </td><td>K562</td><td> 0.031850530</td><td>0.01560865</td><td> 0.0078713916</td><td>0.062235794</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885199</td><td>9885463</td><td>TBL1X    </td><td>K562</td><td> 0.021210152</td><td>0.01620138</td><td>-0.0022734134</td><td>0.073601906</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885199</td><td>9885463</td><td>WWC3     </td><td>K562</td><td> 0.034233352</td><td>0.01564010</td><td> 0.0057975544</td><td>0.034521801</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895196</td><td>CLCN4    </td><td>K562</td><td> 0.001865242</td><td>0.01565018</td><td> 0.0020553891</td><td>0.504846956</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895196</td><td>FAM9B    </td><td>K562</td><td>-0.004652758</td><td>0.01714408</td><td> 0.0046117879</td><td>0.705537079</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895196</td><td>TBL1X    </td><td>K562</td><td> 0.017698823</td><td>0.01463871</td><td> 0.0014924705</td><td>0.134127778</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894528</td><td>9895196</td><td>WWC3     </td><td>K562</td><td> 0.011519642</td><td>0.01444829</td><td> 0.0039659077</td><td>0.300552718</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>AKAP17A  </td><td>K562</td><td>-0.001055460</td><td>0.01576670</td><td>-0.0002775904</td><td>0.519674335</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>ASMTL    </td><td>K562</td><td> 0.015289069</td><td>0.01568637</td><td> 0.0096892770</td><td>0.360551642</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>CSF2RA   </td><td>K562</td><td>-0.004989390</td><td>0.01524794</td><td> 0.0071475634</td><td>0.786976285</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>GTPBP6   </td><td>K562</td><td> 0.002390020</td><td>0.01486042</td><td> 0.0102416456</td><td>0.701374633</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>IL3RA    </td><td>K562</td><td> 0.049977526</td><td>0.01625135</td><td> 0.0045826680</td><td>0.002608684</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>LINC00106</td><td>K562</td><td> 0.006358128</td><td>0.01599122</td><td> 0.0034456309</td><td>0.427739964</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>P2RY8    </td><td>K562</td><td> 0.014640945</td><td>0.01601811</td><td> 0.0014535009</td><td>0.205173414</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>PLCXD1   </td><td>K562</td><td> 0.004401955</td><td>0.01427149</td><td> 0.0103568343</td><td>0.661754804</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>PPP2R3B  </td><td>K562</td><td> 0.022713604</td><td>0.01539855</td><td> 0.0089150582</td><td>0.185101283</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>SHOX     </td><td>K562</td><td>-0.022202358</td><td>0.01253452</td><td>-0.0011807335</td><td>0.953238422</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989766</td><td> 990603</td><td>SLC25A6  </td><td>K562</td><td> 0.013008551</td><td>0.01737777</td><td> 0.0241870144</td><td>0.739973001</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940708</td><td>CLCN4    </td><td>K562</td><td>-0.001994304</td><td>0.01550964</td><td> 0.0044519956</td><td>0.661160248</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940708</td><td>FAM9B    </td><td>K562</td><td> 0.007636923</td><td>0.01449607</td><td> 0.0019191473</td><td>0.346629478</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940708</td><td>TBL1X    </td><td>K562</td><td> 0.010044487</td><td>0.01471595</td><td>-0.0020705456</td><td>0.205180409</td></tr>\n",
       "\t<tr><td>chrX</td><td>9940388</td><td>9940708</td><td>WWC3     </td><td>K562</td><td>-0.014825054</td><td>0.01381555</td><td> 0.0009726881</td><td>0.873579388</td></tr>\n",
       "\t<tr><td>chrX</td><td>9966874</td><td>9967060</td><td>CLCN4    </td><td>K562</td><td> 0.014094775</td><td>0.01687259</td><td> 0.0043709824</td><td>0.282203773</td></tr>\n",
       "\t<tr><td>chrX</td><td>9966874</td><td>9967060</td><td>FAM9B    </td><td>K562</td><td> 0.002890514</td><td>0.01241721</td><td>-0.0004791829</td><td>0.393052017</td></tr>\n",
       "\t<tr><td>chrX</td><td>9966874</td><td>9967060</td><td>TBL1X    </td><td>K562</td><td> 0.014597917</td><td>0.01402102</td><td>-0.0006191393</td><td>0.138893716</td></tr>\n",
       "\t<tr><td>chrX</td><td>9966874</td><td>9967060</td><td>WWC3     </td><td>K562</td><td> 0.011444659</td><td>0.01853288</td><td> 0.0052956398</td><td>0.370023985</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>CLCN4    </td><td>K562</td><td>-0.006843134</td><td>0.01536576</td><td> 0.0072505169</td><td>0.820484034</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>FAM9B    </td><td>K562</td><td> 0.003657961</td><td>0.01731023</td><td> 0.0087320334</td><td>0.615286966</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>TBL1X    </td><td>K562</td><td> 0.005076823</td><td>0.01607847</td><td> 0.0027608704</td><td>0.442734209</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995829</td><td>9996183</td><td>WWC3     </td><td>K562</td><td>-0.013922548</td><td>0.01679656</td><td> 0.0062345480</td><td>0.884944432</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2772969 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  10002642 &  10003186 & APITD1-CORT & K562 & -0.0071096383 & 0.01692533 &  0.0013568974 & 0.69154301\\\\\n",
       "\t chr1 &  10002642 &  10003186 & CASZ1       & K562 & -0.0096060484 & 0.01662268 &  0.0036273219 & 0.78701402\\\\\n",
       "\t chr1 &  10002642 &  10003186 & CLSTN1      & K562 & -0.0008054497 & 0.01672599 &  0.0066649755 & 0.67243096\\\\\n",
       "\t chr1 &  10002642 &  10003186 & CTNNBIP1    & K562 & -0.0129939829 & 0.01482323 &  0.0029078481 & 0.85831150\\\\\n",
       "\t chr1 &  10002642 &  10003186 & DFFA        & K562 &  0.0052172383 & 0.01606043 &  0.0011632385 & 0.40035758\\\\\n",
       "\t chr1 &  10002642 &  10003186 & GPR157      & K562 & -0.0087182622 & 0.01737368 &  0.0008530634 & 0.70915214\\\\\n",
       "\t chr1 &  10002642 &  10003186 & H6PD        & K562 &  0.0078009778 & 0.01499688 &  0.0008884468 & 0.32242397\\\\\n",
       "\t chr1 &  10002642 &  10003186 & KIF1B       & K562 &  0.0069304915 & 0.01541464 & -0.0035702761 & 0.24786598\\\\\n",
       "\t chr1 &  10002642 &  10003186 & LINC01759   & K562 & -0.0021388303 & 0.01791484 &  0.0008765883 & 0.56683406\\\\\n",
       "\t chr1 &  10002642 &  10003186 & LZIC        & K562 & -0.0184650589 & 0.01724992 &  0.0023708795 & 0.88645442\\\\\n",
       "\t chr1 &  10002642 &  10003186 & MIR34AHG    & K562 & -0.0075814544 & 0.01740125 &  0.0017516705 & 0.70414098\\\\\n",
       "\t chr1 &  10002642 &  10003186 & NMNAT1      & K562 & -0.0010100285 & 0.01315259 & -0.0030480695 & 0.43842896\\\\\n",
       "\t chr1 &  10002642 &  10003186 & PEX14       & K562 & -0.0121290380 & 0.01221703 & -0.0007436041 & 0.82431405\\\\\n",
       "\t chr1 &  10002642 &  10003186 & PGD         & K562 & -0.0252161331 & 0.01504693 &  0.0070027324 & 0.98387204\\\\\n",
       "\t chr1 &  10002642 &  10003186 & PIK3CD      & K562 &  0.0119132205 & 0.01366802 &  0.0036980615 & 0.27390342\\\\\n",
       "\t chr1 &  10002642 &  10003186 & SLC25A33    & K562 & -0.0249527298 & 0.01589097 &  0.0023425388 & 0.95707054\\\\\n",
       "\t chr1 &  10002642 &  10003186 & SLC2A5      & K562 & -0.0093607122 & 0.01653240 &  0.0022885724 & 0.75948084\\\\\n",
       "\t chr1 &  10002642 &  10003186 & SPSB1       & K562 & -0.0103082946 & 0.01553774 &  0.0020180914 & 0.78620403\\\\\n",
       "\t chr1 &  10002642 &  10003186 & TMEM201     & K562 & -0.0172728593 & 0.01455992 &  0.0004556541 & 0.88831662\\\\\n",
       "\t chr1 &  10002642 &  10003186 & UBE4B       & K562 &  0.0205714412 & 0.01581837 & -0.0038581716 & 0.06124787\\\\\n",
       "\t chr1 & 100028297 & 100029654 & AGL         & K562 &  0.0181149617 & 0.01591655 &  0.0066421943 & 0.23551407\\\\\n",
       "\t chr1 & 100028297 & 100029654 & CDC14A      & K562 &  0.0136594070 & 0.01790693 & -0.0011760450 & 0.20370061\\\\\n",
       "\t chr1 & 100028297 & 100029654 & DBT         & K562 &  0.0113747569 & 0.01588014 &  0.0078046269 & 0.41106069\\\\\n",
       "\t chr1 & 100028297 & 100029654 & DPH5        & K562 &  0.0199049489 & 0.01626385 &  0.0058230610 & 0.19328897\\\\\n",
       "\t chr1 & 100028297 & 100029654 & EXTL2       & K562 &  0.0048057414 & 0.01540007 &  0.0002475034 & 0.38361954\\\\\n",
       "\t chr1 & 100028297 & 100029654 & FRRS1       & K562 & -0.0271973758 & 0.01867604 & -0.0042273398 & 0.89063650\\\\\n",
       "\t chr1 & 100028297 & 100029654 & LRRC39      & K562 & -0.0027678953 & 0.01490727 & -0.0016713748 & 0.52931819\\\\\n",
       "\t chr1 & 100028297 & 100029654 & MFSD14A     & K562 & -0.0030573297 & 0.01078384 & -0.0015535268 & 0.55545259\\\\\n",
       "\t chr1 & 100028297 & 100029654 & PALMD       & K562 &  0.0177690722 & 0.01424706 &  0.0020953814 & 0.13563661\\\\\n",
       "\t chr1 & 100028297 & 100029654 & PLPPR4      & K562 & -0.0052978269 & 0.01321914 & -0.0009890121 & 0.62776997\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9885199 & 9885463 & FAM9B     & K562 &  0.031850530 & 0.01560865 &  0.0078713916 & 0.062235794\\\\\n",
       "\t chrX & 9885199 & 9885463 & TBL1X     & K562 &  0.021210152 & 0.01620138 & -0.0022734134 & 0.073601906\\\\\n",
       "\t chrX & 9885199 & 9885463 & WWC3      & K562 &  0.034233352 & 0.01564010 &  0.0057975544 & 0.034521801\\\\\n",
       "\t chrX & 9894528 & 9895196 & CLCN4     & K562 &  0.001865242 & 0.01565018 &  0.0020553891 & 0.504846956\\\\\n",
       "\t chrX & 9894528 & 9895196 & FAM9B     & K562 & -0.004652758 & 0.01714408 &  0.0046117879 & 0.705537079\\\\\n",
       "\t chrX & 9894528 & 9895196 & TBL1X     & K562 &  0.017698823 & 0.01463871 &  0.0014924705 & 0.134127778\\\\\n",
       "\t chrX & 9894528 & 9895196 & WWC3      & K562 &  0.011519642 & 0.01444829 &  0.0039659077 & 0.300552718\\\\\n",
       "\t chrX &  989766 &  990603 & AKAP17A   & K562 & -0.001055460 & 0.01576670 & -0.0002775904 & 0.519674335\\\\\n",
       "\t chrX &  989766 &  990603 & ASMTL     & K562 &  0.015289069 & 0.01568637 &  0.0096892770 & 0.360551642\\\\\n",
       "\t chrX &  989766 &  990603 & CSF2RA    & K562 & -0.004989390 & 0.01524794 &  0.0071475634 & 0.786976285\\\\\n",
       "\t chrX &  989766 &  990603 & GTPBP6    & K562 &  0.002390020 & 0.01486042 &  0.0102416456 & 0.701374633\\\\\n",
       "\t chrX &  989766 &  990603 & IL3RA     & K562 &  0.049977526 & 0.01625135 &  0.0045826680 & 0.002608684\\\\\n",
       "\t chrX &  989766 &  990603 & LINC00106 & K562 &  0.006358128 & 0.01599122 &  0.0034456309 & 0.427739964\\\\\n",
       "\t chrX &  989766 &  990603 & P2RY8     & K562 &  0.014640945 & 0.01601811 &  0.0014535009 & 0.205173414\\\\\n",
       "\t chrX &  989766 &  990603 & PLCXD1    & K562 &  0.004401955 & 0.01427149 &  0.0103568343 & 0.661754804\\\\\n",
       "\t chrX &  989766 &  990603 & PPP2R3B   & K562 &  0.022713604 & 0.01539855 &  0.0089150582 & 0.185101283\\\\\n",
       "\t chrX &  989766 &  990603 & SHOX      & K562 & -0.022202358 & 0.01253452 & -0.0011807335 & 0.953238422\\\\\n",
       "\t chrX &  989766 &  990603 & SLC25A6   & K562 &  0.013008551 & 0.01737777 &  0.0241870144 & 0.739973001\\\\\n",
       "\t chrX & 9940388 & 9940708 & CLCN4     & K562 & -0.001994304 & 0.01550964 &  0.0044519956 & 0.661160248\\\\\n",
       "\t chrX & 9940388 & 9940708 & FAM9B     & K562 &  0.007636923 & 0.01449607 &  0.0019191473 & 0.346629478\\\\\n",
       "\t chrX & 9940388 & 9940708 & TBL1X     & K562 &  0.010044487 & 0.01471595 & -0.0020705456 & 0.205180409\\\\\n",
       "\t chrX & 9940388 & 9940708 & WWC3      & K562 & -0.014825054 & 0.01381555 &  0.0009726881 & 0.873579388\\\\\n",
       "\t chrX & 9966874 & 9967060 & CLCN4     & K562 &  0.014094775 & 0.01687259 &  0.0043709824 & 0.282203773\\\\\n",
       "\t chrX & 9966874 & 9967060 & FAM9B     & K562 &  0.002890514 & 0.01241721 & -0.0004791829 & 0.393052017\\\\\n",
       "\t chrX & 9966874 & 9967060 & TBL1X     & K562 &  0.014597917 & 0.01402102 & -0.0006191393 & 0.138893716\\\\\n",
       "\t chrX & 9966874 & 9967060 & WWC3      & K562 &  0.011444659 & 0.01853288 &  0.0052956398 & 0.370023985\\\\\n",
       "\t chrX & 9995829 & 9996183 & CLCN4     & K562 & -0.006843134 & 0.01536576 &  0.0072505169 & 0.820484034\\\\\n",
       "\t chrX & 9995829 & 9996183 & FAM9B     & K562 &  0.003657961 & 0.01731023 &  0.0087320334 & 0.615286966\\\\\n",
       "\t chrX & 9995829 & 9996183 & TBL1X     & K562 &  0.005076823 & 0.01607847 &  0.0027608704 & 0.442734209\\\\\n",
       "\t chrX & 9995829 & 9996183 & WWC3      & K562 & -0.013922548 & 0.01679656 &  0.0062345480 & 0.884944432\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2772969 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  10002642 |  10003186 | APITD1-CORT | K562 | -0.0071096383 | 0.01692533 |  0.0013568974 | 0.69154301 |\n",
       "| chr1 |  10002642 |  10003186 | CASZ1       | K562 | -0.0096060484 | 0.01662268 |  0.0036273219 | 0.78701402 |\n",
       "| chr1 |  10002642 |  10003186 | CLSTN1      | K562 | -0.0008054497 | 0.01672599 |  0.0066649755 | 0.67243096 |\n",
       "| chr1 |  10002642 |  10003186 | CTNNBIP1    | K562 | -0.0129939829 | 0.01482323 |  0.0029078481 | 0.85831150 |\n",
       "| chr1 |  10002642 |  10003186 | DFFA        | K562 |  0.0052172383 | 0.01606043 |  0.0011632385 | 0.40035758 |\n",
       "| chr1 |  10002642 |  10003186 | GPR157      | K562 | -0.0087182622 | 0.01737368 |  0.0008530634 | 0.70915214 |\n",
       "| chr1 |  10002642 |  10003186 | H6PD        | K562 |  0.0078009778 | 0.01499688 |  0.0008884468 | 0.32242397 |\n",
       "| chr1 |  10002642 |  10003186 | KIF1B       | K562 |  0.0069304915 | 0.01541464 | -0.0035702761 | 0.24786598 |\n",
       "| chr1 |  10002642 |  10003186 | LINC01759   | K562 | -0.0021388303 | 0.01791484 |  0.0008765883 | 0.56683406 |\n",
       "| chr1 |  10002642 |  10003186 | LZIC        | K562 | -0.0184650589 | 0.01724992 |  0.0023708795 | 0.88645442 |\n",
       "| chr1 |  10002642 |  10003186 | MIR34AHG    | K562 | -0.0075814544 | 0.01740125 |  0.0017516705 | 0.70414098 |\n",
       "| chr1 |  10002642 |  10003186 | NMNAT1      | K562 | -0.0010100285 | 0.01315259 | -0.0030480695 | 0.43842896 |\n",
       "| chr1 |  10002642 |  10003186 | PEX14       | K562 | -0.0121290380 | 0.01221703 | -0.0007436041 | 0.82431405 |\n",
       "| chr1 |  10002642 |  10003186 | PGD         | K562 | -0.0252161331 | 0.01504693 |  0.0070027324 | 0.98387204 |\n",
       "| chr1 |  10002642 |  10003186 | PIK3CD      | K562 |  0.0119132205 | 0.01366802 |  0.0036980615 | 0.27390342 |\n",
       "| chr1 |  10002642 |  10003186 | SLC25A33    | K562 | -0.0249527298 | 0.01589097 |  0.0023425388 | 0.95707054 |\n",
       "| chr1 |  10002642 |  10003186 | SLC2A5      | K562 | -0.0093607122 | 0.01653240 |  0.0022885724 | 0.75948084 |\n",
       "| chr1 |  10002642 |  10003186 | SPSB1       | K562 | -0.0103082946 | 0.01553774 |  0.0020180914 | 0.78620403 |\n",
       "| chr1 |  10002642 |  10003186 | TMEM201     | K562 | -0.0172728593 | 0.01455992 |  0.0004556541 | 0.88831662 |\n",
       "| chr1 |  10002642 |  10003186 | UBE4B       | K562 |  0.0205714412 | 0.01581837 | -0.0038581716 | 0.06124787 |\n",
       "| chr1 | 100028297 | 100029654 | AGL         | K562 |  0.0181149617 | 0.01591655 |  0.0066421943 | 0.23551407 |\n",
       "| chr1 | 100028297 | 100029654 | CDC14A      | K562 |  0.0136594070 | 0.01790693 | -0.0011760450 | 0.20370061 |\n",
       "| chr1 | 100028297 | 100029654 | DBT         | K562 |  0.0113747569 | 0.01588014 |  0.0078046269 | 0.41106069 |\n",
       "| chr1 | 100028297 | 100029654 | DPH5        | K562 |  0.0199049489 | 0.01626385 |  0.0058230610 | 0.19328897 |\n",
       "| chr1 | 100028297 | 100029654 | EXTL2       | K562 |  0.0048057414 | 0.01540007 |  0.0002475034 | 0.38361954 |\n",
       "| chr1 | 100028297 | 100029654 | FRRS1       | K562 | -0.0271973758 | 0.01867604 | -0.0042273398 | 0.89063650 |\n",
       "| chr1 | 100028297 | 100029654 | LRRC39      | K562 | -0.0027678953 | 0.01490727 | -0.0016713748 | 0.52931819 |\n",
       "| chr1 | 100028297 | 100029654 | MFSD14A     | K562 | -0.0030573297 | 0.01078384 | -0.0015535268 | 0.55545259 |\n",
       "| chr1 | 100028297 | 100029654 | PALMD       | K562 |  0.0177690722 | 0.01424706 |  0.0020953814 | 0.13563661 |\n",
       "| chr1 | 100028297 | 100029654 | PLPPR4      | K562 | -0.0052978269 | 0.01321914 | -0.0009890121 | 0.62776997 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9885199 | 9885463 | FAM9B     | K562 |  0.031850530 | 0.01560865 |  0.0078713916 | 0.062235794 |\n",
       "| chrX | 9885199 | 9885463 | TBL1X     | K562 |  0.021210152 | 0.01620138 | -0.0022734134 | 0.073601906 |\n",
       "| chrX | 9885199 | 9885463 | WWC3      | K562 |  0.034233352 | 0.01564010 |  0.0057975544 | 0.034521801 |\n",
       "| chrX | 9894528 | 9895196 | CLCN4     | K562 |  0.001865242 | 0.01565018 |  0.0020553891 | 0.504846956 |\n",
       "| chrX | 9894528 | 9895196 | FAM9B     | K562 | -0.004652758 | 0.01714408 |  0.0046117879 | 0.705537079 |\n",
       "| chrX | 9894528 | 9895196 | TBL1X     | K562 |  0.017698823 | 0.01463871 |  0.0014924705 | 0.134127778 |\n",
       "| chrX | 9894528 | 9895196 | WWC3      | K562 |  0.011519642 | 0.01444829 |  0.0039659077 | 0.300552718 |\n",
       "| chrX |  989766 |  990603 | AKAP17A   | K562 | -0.001055460 | 0.01576670 | -0.0002775904 | 0.519674335 |\n",
       "| chrX |  989766 |  990603 | ASMTL     | K562 |  0.015289069 | 0.01568637 |  0.0096892770 | 0.360551642 |\n",
       "| chrX |  989766 |  990603 | CSF2RA    | K562 | -0.004989390 | 0.01524794 |  0.0071475634 | 0.786976285 |\n",
       "| chrX |  989766 |  990603 | GTPBP6    | K562 |  0.002390020 | 0.01486042 |  0.0102416456 | 0.701374633 |\n",
       "| chrX |  989766 |  990603 | IL3RA     | K562 |  0.049977526 | 0.01625135 |  0.0045826680 | 0.002608684 |\n",
       "| chrX |  989766 |  990603 | LINC00106 | K562 |  0.006358128 | 0.01599122 |  0.0034456309 | 0.427739964 |\n",
       "| chrX |  989766 |  990603 | P2RY8     | K562 |  0.014640945 | 0.01601811 |  0.0014535009 | 0.205173414 |\n",
       "| chrX |  989766 |  990603 | PLCXD1    | K562 |  0.004401955 | 0.01427149 |  0.0103568343 | 0.661754804 |\n",
       "| chrX |  989766 |  990603 | PPP2R3B   | K562 |  0.022713604 | 0.01539855 |  0.0089150582 | 0.185101283 |\n",
       "| chrX |  989766 |  990603 | SHOX      | K562 | -0.022202358 | 0.01253452 | -0.0011807335 | 0.953238422 |\n",
       "| chrX |  989766 |  990603 | SLC25A6   | K562 |  0.013008551 | 0.01737777 |  0.0241870144 | 0.739973001 |\n",
       "| chrX | 9940388 | 9940708 | CLCN4     | K562 | -0.001994304 | 0.01550964 |  0.0044519956 | 0.661160248 |\n",
       "| chrX | 9940388 | 9940708 | FAM9B     | K562 |  0.007636923 | 0.01449607 |  0.0019191473 | 0.346629478 |\n",
       "| chrX | 9940388 | 9940708 | TBL1X     | K562 |  0.010044487 | 0.01471595 | -0.0020705456 | 0.205180409 |\n",
       "| chrX | 9940388 | 9940708 | WWC3      | K562 | -0.014825054 | 0.01381555 |  0.0009726881 | 0.873579388 |\n",
       "| chrX | 9966874 | 9967060 | CLCN4     | K562 |  0.014094775 | 0.01687259 |  0.0043709824 | 0.282203773 |\n",
       "| chrX | 9966874 | 9967060 | FAM9B     | K562 |  0.002890514 | 0.01241721 | -0.0004791829 | 0.393052017 |\n",
       "| chrX | 9966874 | 9967060 | TBL1X     | K562 |  0.014597917 | 0.01402102 | -0.0006191393 | 0.138893716 |\n",
       "| chrX | 9966874 | 9967060 | WWC3      | K562 |  0.011444659 | 0.01853288 |  0.0052956398 | 0.370023985 |\n",
       "| chrX | 9995829 | 9996183 | CLCN4     | K562 | -0.006843134 | 0.01536576 |  0.0072505169 | 0.820484034 |\n",
       "| chrX | 9995829 | 9996183 | FAM9B     | K562 |  0.003657961 | 0.01731023 |  0.0087320334 | 0.615286966 |\n",
       "| chrX | 9995829 | 9996183 | TBL1X     | K562 |  0.005076823 | 0.01607847 |  0.0027608704 | 0.442734209 |\n",
       "| chrX | 9995829 | 9996183 | WWC3      | K562 | -0.013922548 | 0.01679656 |  0.0062345480 | 0.884944432 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene  CellType rObs          rBgSD     \n",
       "1       chr1  10002642  10003186 APITD1-CORT K562     -0.0071096383 0.01692533\n",
       "2       chr1  10002642  10003186 CASZ1       K562     -0.0096060484 0.01662268\n",
       "3       chr1  10002642  10003186 CLSTN1      K562     -0.0008054497 0.01672599\n",
       "4       chr1  10002642  10003186 CTNNBIP1    K562     -0.0129939829 0.01482323\n",
       "5       chr1  10002642  10003186 DFFA        K562      0.0052172383 0.01606043\n",
       "6       chr1  10002642  10003186 GPR157      K562     -0.0087182622 0.01737368\n",
       "7       chr1  10002642  10003186 H6PD        K562      0.0078009778 0.01499688\n",
       "8       chr1  10002642  10003186 KIF1B       K562      0.0069304915 0.01541464\n",
       "9       chr1  10002642  10003186 LINC01759   K562     -0.0021388303 0.01791484\n",
       "10      chr1  10002642  10003186 LZIC        K562     -0.0184650589 0.01724992\n",
       "11      chr1  10002642  10003186 MIR34AHG    K562     -0.0075814544 0.01740125\n",
       "12      chr1  10002642  10003186 NMNAT1      K562     -0.0010100285 0.01315259\n",
       "13      chr1  10002642  10003186 PEX14       K562     -0.0121290380 0.01221703\n",
       "14      chr1  10002642  10003186 PGD         K562     -0.0252161331 0.01504693\n",
       "15      chr1  10002642  10003186 PIK3CD      K562      0.0119132205 0.01366802\n",
       "16      chr1  10002642  10003186 SLC25A33    K562     -0.0249527298 0.01589097\n",
       "17      chr1  10002642  10003186 SLC2A5      K562     -0.0093607122 0.01653240\n",
       "18      chr1  10002642  10003186 SPSB1       K562     -0.0103082946 0.01553774\n",
       "19      chr1  10002642  10003186 TMEM201     K562     -0.0172728593 0.01455992\n",
       "20      chr1  10002642  10003186 UBE4B       K562      0.0205714412 0.01581837\n",
       "21      chr1 100028297 100029654 AGL         K562      0.0181149617 0.01591655\n",
       "22      chr1 100028297 100029654 CDC14A      K562      0.0136594070 0.01790693\n",
       "23      chr1 100028297 100029654 DBT         K562      0.0113747569 0.01588014\n",
       "24      chr1 100028297 100029654 DPH5        K562      0.0199049489 0.01626385\n",
       "25      chr1 100028297 100029654 EXTL2       K562      0.0048057414 0.01540007\n",
       "26      chr1 100028297 100029654 FRRS1       K562     -0.0271973758 0.01867604\n",
       "27      chr1 100028297 100029654 LRRC39      K562     -0.0027678953 0.01490727\n",
       "28      chr1 100028297 100029654 MFSD14A     K562     -0.0030573297 0.01078384\n",
       "29      chr1 100028297 100029654 PALMD       K562      0.0177690722 0.01424706\n",
       "30      chr1 100028297 100029654 PLPPR4      K562     -0.0052978269 0.01321914\n",
       "⋮       ⋮    ⋮         ⋮         ⋮           ⋮        ⋮             ⋮         \n",
       "2772940 chrX 9885199   9885463   FAM9B       K562      0.031850530  0.01560865\n",
       "2772941 chrX 9885199   9885463   TBL1X       K562      0.021210152  0.01620138\n",
       "2772942 chrX 9885199   9885463   WWC3        K562      0.034233352  0.01564010\n",
       "2772943 chrX 9894528   9895196   CLCN4       K562      0.001865242  0.01565018\n",
       "2772944 chrX 9894528   9895196   FAM9B       K562     -0.004652758  0.01714408\n",
       "2772945 chrX 9894528   9895196   TBL1X       K562      0.017698823  0.01463871\n",
       "2772946 chrX 9894528   9895196   WWC3        K562      0.011519642  0.01444829\n",
       "2772947 chrX  989766    990603   AKAP17A     K562     -0.001055460  0.01576670\n",
       "2772948 chrX  989766    990603   ASMTL       K562      0.015289069  0.01568637\n",
       "2772949 chrX  989766    990603   CSF2RA      K562     -0.004989390  0.01524794\n",
       "2772950 chrX  989766    990603   GTPBP6      K562      0.002390020  0.01486042\n",
       "2772951 chrX  989766    990603   IL3RA       K562      0.049977526  0.01625135\n",
       "2772952 chrX  989766    990603   LINC00106   K562      0.006358128  0.01599122\n",
       "2772953 chrX  989766    990603   P2RY8       K562      0.014640945  0.01601811\n",
       "2772954 chrX  989766    990603   PLCXD1      K562      0.004401955  0.01427149\n",
       "2772955 chrX  989766    990603   PPP2R3B     K562      0.022713604  0.01539855\n",
       "2772956 chrX  989766    990603   SHOX        K562     -0.022202358  0.01253452\n",
       "2772957 chrX  989766    990603   SLC25A6     K562      0.013008551  0.01737777\n",
       "2772958 chrX 9940388   9940708   CLCN4       K562     -0.001994304  0.01550964\n",
       "2772959 chrX 9940388   9940708   FAM9B       K562      0.007636923  0.01449607\n",
       "2772960 chrX 9940388   9940708   TBL1X       K562      0.010044487  0.01471595\n",
       "2772961 chrX 9940388   9940708   WWC3        K562     -0.014825054  0.01381555\n",
       "2772962 chrX 9966874   9967060   CLCN4       K562      0.014094775  0.01687259\n",
       "2772963 chrX 9966874   9967060   FAM9B       K562      0.002890514  0.01241721\n",
       "2772964 chrX 9966874   9967060   TBL1X       K562      0.014597917  0.01402102\n",
       "2772965 chrX 9966874   9967060   WWC3        K562      0.011444659  0.01853288\n",
       "2772966 chrX 9995829   9996183   CLCN4       K562     -0.006843134  0.01536576\n",
       "2772967 chrX 9995829   9996183   FAM9B       K562      0.003657961  0.01731023\n",
       "2772968 chrX 9995829   9996183   TBL1X       K562      0.005076823  0.01607847\n",
       "2772969 chrX 9995829   9996183   WWC3        K562     -0.013922548  0.01679656\n",
       "        rBgMean       pvalZ      \n",
       "1        0.0013568974 0.69154301 \n",
       "2        0.0036273219 0.78701402 \n",
       "3        0.0066649755 0.67243096 \n",
       "4        0.0029078481 0.85831150 \n",
       "5        0.0011632385 0.40035758 \n",
       "6        0.0008530634 0.70915214 \n",
       "7        0.0008884468 0.32242397 \n",
       "8       -0.0035702761 0.24786598 \n",
       "9        0.0008765883 0.56683406 \n",
       "10       0.0023708795 0.88645442 \n",
       "11       0.0017516705 0.70414098 \n",
       "12      -0.0030480695 0.43842896 \n",
       "13      -0.0007436041 0.82431405 \n",
       "14       0.0070027324 0.98387204 \n",
       "15       0.0036980615 0.27390342 \n",
       "16       0.0023425388 0.95707054 \n",
       "17       0.0022885724 0.75948084 \n",
       "18       0.0020180914 0.78620403 \n",
       "19       0.0004556541 0.88831662 \n",
       "20      -0.0038581716 0.06124787 \n",
       "21       0.0066421943 0.23551407 \n",
       "22      -0.0011760450 0.20370061 \n",
       "23       0.0078046269 0.41106069 \n",
       "24       0.0058230610 0.19328897 \n",
       "25       0.0002475034 0.38361954 \n",
       "26      -0.0042273398 0.89063650 \n",
       "27      -0.0016713748 0.52931819 \n",
       "28      -0.0015535268 0.55545259 \n",
       "29       0.0020953814 0.13563661 \n",
       "30      -0.0009890121 0.62776997 \n",
       "⋮       ⋮             ⋮          \n",
       "2772940  0.0078713916 0.062235794\n",
       "2772941 -0.0022734134 0.073601906\n",
       "2772942  0.0057975544 0.034521801\n",
       "2772943  0.0020553891 0.504846956\n",
       "2772944  0.0046117879 0.705537079\n",
       "2772945  0.0014924705 0.134127778\n",
       "2772946  0.0039659077 0.300552718\n",
       "2772947 -0.0002775904 0.519674335\n",
       "2772948  0.0096892770 0.360551642\n",
       "2772949  0.0071475634 0.786976285\n",
       "2772950  0.0102416456 0.701374633\n",
       "2772951  0.0045826680 0.002608684\n",
       "2772952  0.0034456309 0.427739964\n",
       "2772953  0.0014535009 0.205173414\n",
       "2772954  0.0103568343 0.661754804\n",
       "2772955  0.0089150582 0.185101283\n",
       "2772956 -0.0011807335 0.953238422\n",
       "2772957  0.0241870144 0.739973001\n",
       "2772958  0.0044519956 0.661160248\n",
       "2772959  0.0019191473 0.346629478\n",
       "2772960 -0.0020705456 0.205180409\n",
       "2772961  0.0009726881 0.873579388\n",
       "2772962  0.0043709824 0.282203773\n",
       "2772963 -0.0004791829 0.393052017\n",
       "2772964 -0.0006191393 0.138893716\n",
       "2772965  0.0052956398 0.370023985\n",
       "2772966  0.0072505169 0.820484034\n",
       "2772967  0.0087320334 0.615286966\n",
       "2772968  0.0027608704 0.442734209\n",
       "2772969  0.0062345480 0.884944432"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   data.table_1.15.2                \n",
       " [3] foreach_1.5.2                     BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] genomation_1.34.0                 Seurat_5.0.3                     \n",
       "[11] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[13] FigR_0.1.0                        motifmatchr_1.24.0               \n",
       "[15] chromVAR_1.24.0                   cowplot_1.1.3                    \n",
       "[17] ggplot2_3.5.1                     dplyr_1.1.4                      \n",
       "[19] SummarizedExperiment_1.30.2       Biobase_2.60.0                   \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               MatrixGenerics_1.12.3            \n",
       "[27] matrixStats_1.3.0                 Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             vroom_1.6.5                \n",
       " [15] globals_0.16.3              lattice_0.22-6             \n",
       " [17] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [19] plotly_4.10.4               plotrix_3.8-4              \n",
       " [21] yaml_2.3.8                  httpuv_1.6.15              \n",
       " [23] sctransform_0.4.1           spam_2.10-0                \n",
       " [25] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [27] pbapply_1.7-2               DBI_1.2.2                  \n",
       " [29] CNEr_1.38.0                 RColorBrewer_1.1-3         \n",
       " [31] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [33] Rtsne_0.17                  purrr_1.0.2                \n",
       " [35] R.utils_2.12.3              RCurl_1.98-1.14            \n",
       " [37] pracma_2.4.4                GenomeInfoDbData_1.2.11    \n",
       " [39] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [41] spatstat.utils_3.0-4        listenv_0.9.1              \n",
       " [43] seqLogo_1.68.0              goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             spatstat.random_3.2-3      \n",
       " [47] annotate_1.80.0             fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           leiden_0.4.3.1             \n",
       " [51] codetools_0.2-20            DelayedArray_0.26.7        \n",
       " [53] DT_0.33                     tidyselect_1.2.1           \n",
       " [55] spatstat.explore_3.2-6      base64enc_0.1-3            \n",
       " [57] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [59] progressr_0.14.0            iterators_1.0.14           \n",
       " [61] ggridges_0.5.6              survival_3.6-4             \n",
       " [63] tools_4.3.3                 TFMPvalue_0.0.9            \n",
       " [65] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [67] glue_1.7.0                  gridExtra_2.3              \n",
       " [69] IRdisplay_1.1               withr_3.0.0                \n",
       " [71] fastmap_1.1.1               fansi_1.0.6                \n",
       " [73] caTools_1.18.2              digest_0.6.35              \n",
       " [75] R6_2.5.1                    mime_0.12                  \n",
       " [77] nabor_0.5.0                 seqPattern_1.34.0          \n",
       " [79] colorspace_2.1-0            scattermore_1.2            \n",
       " [81] GO.db_3.18.0                tensor_1.5                 \n",
       " [83] gtools_3.9.5                poweRlaw_0.80.0            \n",
       " [85] spatstat.data_3.0-4         RSQLite_2.3.6              \n",
       " [87] R.methodsS3_1.8.2           utf8_1.2.4                 \n",
       " [89] tidyr_1.3.1                 generics_0.1.3             \n",
       " [91] httr_1.4.7                  htmlwidgets_1.6.4          \n",
       " [93] S4Arrays_1.2.1              TFBSTools_1.40.0           \n",
       " [95] uwot_0.1.16                 pkgconfig_2.0.3            \n",
       " [97] gtable_0.3.5                blob_1.2.4                 \n",
       " [99] impute_1.76.0               lmtest_0.9-40              \n",
       "[101] htmltools_0.5.8.1           dotCall64_1.1-1            \n",
       "[103] scales_1.3.0                png_0.1-8                  \n",
       "[105] tzdb_0.4.0                  reshape2_1.4.4             \n",
       "[107] rjson_0.2.21                uuid_1.2-0                 \n",
       "[109] nlme_3.1-164                repr_1.1.7                 \n",
       "[111] cachem_1.0.8                zoo_1.8-12                 \n",
       "[113] stringr_1.5.1               KernSmooth_2.23-22         \n",
       "[115] miniUI_0.1.1.1              AnnotationDbi_1.64.1       \n",
       "[117] restfulr_0.0.15             pillar_1.9.0               \n",
       "[119] vctrs_0.6.5                 RANN_2.6.1                 \n",
       "[121] promises_1.3.0              xtable_1.8-4               \n",
       "[123] cluster_2.1.6               evaluate_0.23              \n",
       "[125] readr_2.1.5                 cli_3.6.2                  \n",
       "[127] compiler_4.3.3              Rsamtools_2.18.0           \n",
       "[129] rlang_1.1.3                 crayon_1.5.2               \n",
       "[131] future.apply_1.11.2         plyr_1.8.9                 \n",
       "[133] stringi_1.8.3               gridBase_0.4-7             \n",
       "[135] deldir_2.0-4                viridisLite_0.4.2          \n",
       "[137] BiocParallel_1.36.0         munsell_0.5.1              \n",
       "[139] lazyeval_0.2.2              spatstat.geom_3.2-9        \n",
       "[141] IRkernel_1.3.2              RcppHNSW_0.6.0             \n",
       "[143] hms_1.1.3                   patchwork_1.2.0            \n",
       "[145] bit64_4.0.5                 future_1.33.2              \n",
       "[147] KEGGREST_1.40.1             shiny_1.8.1.1              \n",
       "[149] ROCR_1.0-11                 igraph_2.0.3               \n",
       "[151] memoise_2.0.1               bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
