{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Xu/K562/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Xu/K562/Predictions/EnhancerPredictionsAllPutative.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Xu/K562/Kendall/atac_matrix.csv.gz\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/1.prepare_data/1.seurat_pipeline.240507/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/3.Genome_wide_prediction/FigR/FigR.240521/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 3150487 ranges and 3 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10001196-10001745      * | APITD1-CORT\n",
       "        [2]     chr1 10001196-10001745      * |       CASZ1\n",
       "        [3]     chr1 10001196-10001745      * |       CENPS\n",
       "        [4]     chr1 10001196-10001745      * |      CLSTN1\n",
       "        [5]     chr1 10001196-10001745      * |        CORT\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [3150483]     chrX     989754-991253      * |     SLC25A6\n",
       "  [3150484]     chrX   9995827-9996208      * |       CLCN4\n",
       "  [3150485]     chrX   9995827-9996208      * |       FAM9B\n",
       "  [3150486]     chrX   9995827-9996208      * |       TBL1X\n",
       "  [3150487]     chrX   9995827-9996208      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-10001196-10001745 chr1-10001196-100017..\n",
       "        [2] chr1-10001196-10001745 chr1-10001196-100017..\n",
       "        [3] chr1-10001196-10001745 chr1-10001196-100017..\n",
       "        [4] chr1-10001196-10001745 chr1-10001196-100017..\n",
       "        [5] chr1-10001196-10001745 chr1-10001196-100017..\n",
       "        ...                    ...                    ...\n",
       "  [3150483]     chrX-989754-991253 chrX-989754-991253_S..\n",
       "  [3150484]   chrX-9995827-9996208 chrX-9995827-9996208..\n",
       "  [3150485]   chrX-9995827-9996208 chrX-9995827-9996208..\n",
       "  [3150486]   chrX-9995827-9996208 chrX-9995827-9996208..\n",
       "  [3150487]   chrX-9995827-9996208 chrX-9995827-9996208..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  4000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  4001  to  5000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  5001  to  6000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  6001  to  7000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  7001  to  7821 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = read.csv(path.matrix.atac_count,\n",
    "                             row.names = 1,\n",
    "                             check.names = F)\n",
    "matrix.atac_count = Matrix(as.matrix(matrix.atac_count), sparse = TRUE)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 157600 7821 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(157600): chr1-10001196-10001745 chr1-100027917-100029695 ...\n",
       "  chrX-989754-991253 chrX-9995827-9996208\n",
       "rowData names(2): PeakName bias\n",
       "colnames(7821): K562_Xu_AAACAGCCAAGCGATG-1 K562_Xu_AAACAGCCACATGCTA-1\n",
       "  ... K562_Xu_TTTGTTGGTTTATGGG-1 K562_Xu_TTTGTTGGTTTGTGGA-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna),\n",
    "                       index = 1:nrow(matrix.rna))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(dir.output, \"chr\", sep = \"/\"), recursive = T):\n",
      "“'/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/3.Genome_wide_prediction/FigR/FigR.240521//chr' already exists”\n"
     ]
    }
   ],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr10\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  11.0219037532806 secs \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2914790 ranges and 7 metadata columns:\n",
       "       seqnames            ranges strand |  TargetGene               PeakName\n",
       "          <Rle>         <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 10001196-10001745      * |       CASZ1 chr1-10001196-10001745\n",
       "  chr1     chr1 10001196-10001745      * |       CENPS chr1-10001196-10001745\n",
       "  chr1     chr1 10001196-10001745      * |      CLSTN1 chr1-10001196-10001745\n",
       "  chr1     chr1 10001196-10001745      * |        CORT chr1-10001196-10001745\n",
       "  chr1     chr1 10001196-10001745      * |    CTNNBIP1 chr1-10001196-10001745\n",
       "   ...      ...               ...    ... .         ...                    ...\n",
       "  chrX     chrX     989754-991253      * |     SLC25A6     chrX-989754-991253\n",
       "  chrX     chrX   9995827-9996208      * |       CLCN4   chrX-9995827-9996208\n",
       "  chrX     chrX   9995827-9996208      * |       FAM9B   chrX-9995827-9996208\n",
       "  chrX     chrX   9995827-9996208      * |       TBL1X   chrX-9995827-9996208\n",
       "  chrX     chrX   9995827-9996208      * |        WWC3   chrX-9995827-9996208\n",
       "                     PairName        rObs      rBgSD    rBgMean     pvalZ\n",
       "                  <character>   <numeric>  <numeric>  <numeric> <numeric>\n",
       "  chr1 chr1-10001196-100017..  0.02373545 0.01245971 0.00397643 0.0563887\n",
       "  chr1 chr1-10001196-100017..  0.01594539 0.00981424 0.00614042 0.1588838\n",
       "  chr1 chr1-10001196-100017..  0.01164222 0.01070784 0.01002446 0.4399557\n",
       "  chr1 chr1-10001196-100017.. -0.00396555 0.01519252 0.00204995 0.6539296\n",
       "  chr1 chr1-10001196-100017..  0.01418203 0.01105881 0.00841524 0.3010210\n",
       "   ...                    ...         ...        ...        ...       ...\n",
       "  chrX chrX-989754-991253_S.. -0.01042524  0.0151401 0.00522230  0.849319\n",
       "  chrX chrX-9995827-9996208..  0.00400341  0.0120582 0.00798961  0.629519\n",
       "  chrX chrX-9995827-9996208..  0.00736103  0.0120227 0.00258489  0.345588\n",
       "  chrX chrX-9995827-9996208..  0.00938996  0.0101517 0.00944960  0.502344\n",
       "  chrX chrX-9995827-9996208..  0.00495331  0.0124520 0.00595176  0.531955\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2914790 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CASZ1   </td><td>K562</td><td> 2.373545e-02</td><td>0.012459711</td><td> 3.976434e-03</td><td>0.056388671</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CENPS   </td><td>K562</td><td> 1.594539e-02</td><td>0.009814236</td><td> 6.140422e-03</td><td>0.158883787</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CLSTN1  </td><td>K562</td><td> 1.164222e-02</td><td>0.010707841</td><td> 1.002446e-02</td><td>0.439955671</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CORT    </td><td>K562</td><td>-3.965550e-03</td><td>0.015192518</td><td> 2.049952e-03</td><td>0.653929629</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CTNNBIP1</td><td>K562</td><td> 1.418203e-02</td><td>0.011058809</td><td> 8.415236e-03</td><td>0.301020957</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>DFFA    </td><td>K562</td><td> 6.978517e-03</td><td>0.011281599</td><td> 1.798258e-03</td><td>0.323053266</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>GPR157  </td><td>K562</td><td>-1.038945e-02</td><td>0.012452020</td><td> 1.296185e-03</td><td>0.825994104</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>H6PD    </td><td>K562</td><td>-1.595742e-03</td><td>0.013186113</td><td> 4.040827e-03</td><td>0.665478754</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>KIF1B   </td><td>K562</td><td> 2.809136e-02</td><td>0.011970877</td><td> 5.099747e-03</td><td>0.027389267</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>LZIC    </td><td>K562</td><td> 2.520966e-03</td><td>0.010951603</td><td> 1.770884e-03</td><td>0.472697533</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>MIR34AHG</td><td>K562</td><td>-3.619804e-03</td><td>0.012189632</td><td> 8.564796e-04</td><td>0.643272738</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>NMNAT1  </td><td>K562</td><td> 1.289296e-02</td><td>0.011081800</td><td> 7.657213e-03</td><td>0.318297863</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>PEX14   </td><td>K562</td><td>-2.889875e-05</td><td>0.010171680</td><td> 5.089914e-03</td><td>0.692602795</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>PGD     </td><td>K562</td><td> 1.932427e-02</td><td>0.011505577</td><td> 2.936272e-03</td><td>0.077172276</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>PIK3CD  </td><td>K562</td><td> 8.370059e-03</td><td>0.011481796</td><td> 9.661978e-03</td><td>0.544794009</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>RBP7    </td><td>K562</td><td> 2.865984e-02</td><td>0.010501881</td><td>-8.728314e-05</td><td>0.003097000</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>SLC25A33</td><td>K562</td><td> 1.426118e-02</td><td>0.011428111</td><td> 3.542587e-03</td><td>0.174144098</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>SPSB1   </td><td>K562</td><td> 2.290241e-02</td><td>0.012223231</td><td>-2.884131e-04</td><td>0.028895871</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>TMEM201 </td><td>K562</td><td> 1.616036e-02</td><td>0.011704654</td><td> 4.605808e-03</td><td>0.161778287</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>UBE4B   </td><td>K562</td><td> 2.469338e-02</td><td>0.010455244</td><td>-1.298238e-03</td><td>0.006459610</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>AGL     </td><td>K562</td><td> 3.426649e-02</td><td>0.012183675</td><td> 6.876781e-03</td><td>0.012285988</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>CDC14A  </td><td>K562</td><td> 3.003275e-02</td><td>0.013179488</td><td> 7.757991e-03</td><td>0.045503671</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>DBT     </td><td>K562</td><td> 4.251608e-02</td><td>0.010802897</td><td> 1.524219e-02</td><td>0.005790143</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>DPH5    </td><td>K562</td><td> 4.069875e-02</td><td>0.012485810</td><td> 9.412662e-03</td><td>0.006109918</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>EXTL2   </td><td>K562</td><td> 1.327720e-02</td><td>0.011202587</td><td> 9.953417e-03</td><td>0.383348753</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>FRRS1   </td><td>K562</td><td> 7.008389e-03</td><td>0.013718161</td><td> 2.996208e-03</td><td>0.384962798</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>MFSD14A </td><td>K562</td><td> 1.332009e-02</td><td>0.013399283</td><td> 1.232827e-02</td><td>0.470497259</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>PALMD   </td><td>K562</td><td> 9.294320e-03</td><td>0.012127027</td><td> 6.780437e-03</td><td>0.417889388</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>RTCA    </td><td>K562</td><td> 4.944637e-03</td><td>0.010487785</td><td> 7.361271e-03</td><td>0.591118716</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>RTCA-AS1</td><td>K562</td><td> 1.181748e-02</td><td>0.013982719</td><td> 1.672369e-02</td><td>0.637159569</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>CLCN4    </td><td>K562</td><td>-0.0051374098</td><td>0.014699103</td><td> 0.0091326389</td><td>0.834178767</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>FAM9B    </td><td>K562</td><td> 0.0130287285</td><td>0.013710062</td><td> 0.0039890571</td><td>0.254836886</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>TBL1X    </td><td>K562</td><td>-0.0015979203</td><td>0.011089781</td><td> 0.0106163251</td><td>0.864637934</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>WWC3     </td><td>K562</td><td>-0.0005950215</td><td>0.011010227</td><td> 0.0058402515</td><td>0.720551768</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>CLCN4    </td><td>K562</td><td> 0.0131380399</td><td>0.012619797</td><td> 0.0051295616</td><td>0.262845823</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>FAM9B    </td><td>K562</td><td> 0.0081721127</td><td>0.012714557</td><td> 0.0060027287</td><td>0.432260472</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>TBL1X    </td><td>K562</td><td>-0.0004361237</td><td>0.011821801</td><td> 0.0098067019</td><td>0.806874244</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>WWC3     </td><td>K562</td><td>-0.0117169728</td><td>0.012745833</td><td> 0.0063099199</td><td>0.921368385</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>CLCN4    </td><td>K562</td><td>-0.0073263795</td><td>0.013692948</td><td> 0.0067406944</td><td>0.847865693</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>FAM9B    </td><td>K562</td><td> 0.0163467680</td><td>0.011938149</td><td> 0.0037657621</td><td>0.145976029</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>TBL1X    </td><td>K562</td><td>-0.0042601726</td><td>0.009579401</td><td> 0.0093703566</td><td>0.922617423</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>WWC3     </td><td>K562</td><td>-0.0018792908</td><td>0.013527142</td><td> 0.0072123988</td><td>0.749242276</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>CLCN4    </td><td>K562</td><td> 0.0189453440</td><td>0.011605712</td><td> 0.0056802853</td><td>0.126524132</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>FAM9B    </td><td>K562</td><td> 0.0121673228</td><td>0.012443440</td><td> 0.0009215641</td><td>0.183063992</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>TBL1X    </td><td>K562</td><td> 0.0193937461</td><td>0.013165201</td><td> 0.0110601228</td><td>0.263365548</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>WWC3     </td><td>K562</td><td> 0.0029598921</td><td>0.012439112</td><td> 0.0050124685</td><td>0.565531895</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>AKAP17A  </td><td>K562</td><td> 0.0061596812</td><td>0.012948778</td><td> 0.0107104723</td><td>0.637372967</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>ASMT     </td><td>K562</td><td>-0.0022939125</td><td>0.010662312</td><td> 0.0016068427</td><td>0.642759693</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>ASMTL    </td><td>K562</td><td> 0.0439144995</td><td>0.012199586</td><td> 0.0150632279</td><td>0.009016532</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>GTPBP6   </td><td>K562</td><td> 0.0305578173</td><td>0.011430865</td><td> 0.0239597817</td><td>0.281898029</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>IL3RA    </td><td>K562</td><td> 0.0125659674</td><td>0.010855956</td><td>-0.0024098551</td><td>0.083869885</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>LINC00106</td><td>K562</td><td> 0.0229243072</td><td>0.012028640</td><td> 0.0283747272</td><td>0.674768909</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>P2RY8    </td><td>K562</td><td>-0.0041985963</td><td>0.013039770</td><td>-0.0024558172</td><td>0.553160746</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>PLCXD1   </td><td>K562</td><td> 0.0149861580</td><td>0.014968166</td><td> 0.0132007179</td><td>0.452525779</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>PPP2R3B  </td><td>K562</td><td> 0.0492611216</td><td>0.011158934</td><td> 0.0228027464</td><td>0.008868848</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989754</td><td> 991253</td><td>SLC25A6  </td><td>K562</td><td>-0.0104252429</td><td>0.015140078</td><td> 0.0052223026</td><td>0.849319261</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>CLCN4    </td><td>K562</td><td> 0.0040034126</td><td>0.012058187</td><td> 0.0079896140</td><td>0.629519308</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>FAM9B    </td><td>K562</td><td> 0.0073610288</td><td>0.012022706</td><td> 0.0025848947</td><td>0.345588059</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>TBL1X    </td><td>K562</td><td> 0.0093899593</td><td>0.010151704</td><td> 0.0094496039</td><td>0.502343903</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>WWC3     </td><td>K562</td><td> 0.0049533099</td><td>0.012451989</td><td> 0.0059517592</td><td>0.531954509</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2914790 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  10001196 &  10001745 & CASZ1    & K562 &  2.373545e-02 & 0.012459711 &  3.976434e-03 & 0.056388671\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CENPS    & K562 &  1.594539e-02 & 0.009814236 &  6.140422e-03 & 0.158883787\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CLSTN1   & K562 &  1.164222e-02 & 0.010707841 &  1.002446e-02 & 0.439955671\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CORT     & K562 & -3.965550e-03 & 0.015192518 &  2.049952e-03 & 0.653929629\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CTNNBIP1 & K562 &  1.418203e-02 & 0.011058809 &  8.415236e-03 & 0.301020957\\\\\n",
       "\t chr1 &  10001196 &  10001745 & DFFA     & K562 &  6.978517e-03 & 0.011281599 &  1.798258e-03 & 0.323053266\\\\\n",
       "\t chr1 &  10001196 &  10001745 & GPR157   & K562 & -1.038945e-02 & 0.012452020 &  1.296185e-03 & 0.825994104\\\\\n",
       "\t chr1 &  10001196 &  10001745 & H6PD     & K562 & -1.595742e-03 & 0.013186113 &  4.040827e-03 & 0.665478754\\\\\n",
       "\t chr1 &  10001196 &  10001745 & KIF1B    & K562 &  2.809136e-02 & 0.011970877 &  5.099747e-03 & 0.027389267\\\\\n",
       "\t chr1 &  10001196 &  10001745 & LZIC     & K562 &  2.520966e-03 & 0.010951603 &  1.770884e-03 & 0.472697533\\\\\n",
       "\t chr1 &  10001196 &  10001745 & MIR34AHG & K562 & -3.619804e-03 & 0.012189632 &  8.564796e-04 & 0.643272738\\\\\n",
       "\t chr1 &  10001196 &  10001745 & NMNAT1   & K562 &  1.289296e-02 & 0.011081800 &  7.657213e-03 & 0.318297863\\\\\n",
       "\t chr1 &  10001196 &  10001745 & PEX14    & K562 & -2.889875e-05 & 0.010171680 &  5.089914e-03 & 0.692602795\\\\\n",
       "\t chr1 &  10001196 &  10001745 & PGD      & K562 &  1.932427e-02 & 0.011505577 &  2.936272e-03 & 0.077172276\\\\\n",
       "\t chr1 &  10001196 &  10001745 & PIK3CD   & K562 &  8.370059e-03 & 0.011481796 &  9.661978e-03 & 0.544794009\\\\\n",
       "\t chr1 &  10001196 &  10001745 & RBP7     & K562 &  2.865984e-02 & 0.010501881 & -8.728314e-05 & 0.003097000\\\\\n",
       "\t chr1 &  10001196 &  10001745 & SLC25A33 & K562 &  1.426118e-02 & 0.011428111 &  3.542587e-03 & 0.174144098\\\\\n",
       "\t chr1 &  10001196 &  10001745 & SPSB1    & K562 &  2.290241e-02 & 0.012223231 & -2.884131e-04 & 0.028895871\\\\\n",
       "\t chr1 &  10001196 &  10001745 & TMEM201  & K562 &  1.616036e-02 & 0.011704654 &  4.605808e-03 & 0.161778287\\\\\n",
       "\t chr1 &  10001196 &  10001745 & UBE4B    & K562 &  2.469338e-02 & 0.010455244 & -1.298238e-03 & 0.006459610\\\\\n",
       "\t chr1 & 100027917 & 100029695 & AGL      & K562 &  3.426649e-02 & 0.012183675 &  6.876781e-03 & 0.012285988\\\\\n",
       "\t chr1 & 100027917 & 100029695 & CDC14A   & K562 &  3.003275e-02 & 0.013179488 &  7.757991e-03 & 0.045503671\\\\\n",
       "\t chr1 & 100027917 & 100029695 & DBT      & K562 &  4.251608e-02 & 0.010802897 &  1.524219e-02 & 0.005790143\\\\\n",
       "\t chr1 & 100027917 & 100029695 & DPH5     & K562 &  4.069875e-02 & 0.012485810 &  9.412662e-03 & 0.006109918\\\\\n",
       "\t chr1 & 100027917 & 100029695 & EXTL2    & K562 &  1.327720e-02 & 0.011202587 &  9.953417e-03 & 0.383348753\\\\\n",
       "\t chr1 & 100027917 & 100029695 & FRRS1    & K562 &  7.008389e-03 & 0.013718161 &  2.996208e-03 & 0.384962798\\\\\n",
       "\t chr1 & 100027917 & 100029695 & MFSD14A  & K562 &  1.332009e-02 & 0.013399283 &  1.232827e-02 & 0.470497259\\\\\n",
       "\t chr1 & 100027917 & 100029695 & PALMD    & K562 &  9.294320e-03 & 0.012127027 &  6.780437e-03 & 0.417889388\\\\\n",
       "\t chr1 & 100027917 & 100029695 & RTCA     & K562 &  4.944637e-03 & 0.010487785 &  7.361271e-03 & 0.591118716\\\\\n",
       "\t chr1 & 100027917 & 100029695 & RTCA-AS1 & K562 &  1.181748e-02 & 0.013982719 &  1.672369e-02 & 0.637159569\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9873821 & 9874426 & CLCN4     & K562 & -0.0051374098 & 0.014699103 &  0.0091326389 & 0.834178767\\\\\n",
       "\t chrX & 9873821 & 9874426 & FAM9B     & K562 &  0.0130287285 & 0.013710062 &  0.0039890571 & 0.254836886\\\\\n",
       "\t chrX & 9873821 & 9874426 & TBL1X     & K562 & -0.0015979203 & 0.011089781 &  0.0106163251 & 0.864637934\\\\\n",
       "\t chrX & 9873821 & 9874426 & WWC3      & K562 & -0.0005950215 & 0.011010227 &  0.0058402515 & 0.720551768\\\\\n",
       "\t chrX & 9882635 & 9883318 & CLCN4     & K562 &  0.0131380399 & 0.012619797 &  0.0051295616 & 0.262845823\\\\\n",
       "\t chrX & 9882635 & 9883318 & FAM9B     & K562 &  0.0081721127 & 0.012714557 &  0.0060027287 & 0.432260472\\\\\n",
       "\t chrX & 9882635 & 9883318 & TBL1X     & K562 & -0.0004361237 & 0.011821801 &  0.0098067019 & 0.806874244\\\\\n",
       "\t chrX & 9882635 & 9883318 & WWC3      & K562 & -0.0117169728 & 0.012745833 &  0.0063099199 & 0.921368385\\\\\n",
       "\t chrX & 9885160 & 9885475 & CLCN4     & K562 & -0.0073263795 & 0.013692948 &  0.0067406944 & 0.847865693\\\\\n",
       "\t chrX & 9885160 & 9885475 & FAM9B     & K562 &  0.0163467680 & 0.011938149 &  0.0037657621 & 0.145976029\\\\\n",
       "\t chrX & 9885160 & 9885475 & TBL1X     & K562 & -0.0042601726 & 0.009579401 &  0.0093703566 & 0.922617423\\\\\n",
       "\t chrX & 9885160 & 9885475 & WWC3      & K562 & -0.0018792908 & 0.013527142 &  0.0072123988 & 0.749242276\\\\\n",
       "\t chrX & 9894484 & 9895274 & CLCN4     & K562 &  0.0189453440 & 0.011605712 &  0.0056802853 & 0.126524132\\\\\n",
       "\t chrX & 9894484 & 9895274 & FAM9B     & K562 &  0.0121673228 & 0.012443440 &  0.0009215641 & 0.183063992\\\\\n",
       "\t chrX & 9894484 & 9895274 & TBL1X     & K562 &  0.0193937461 & 0.013165201 &  0.0110601228 & 0.263365548\\\\\n",
       "\t chrX & 9894484 & 9895274 & WWC3      & K562 &  0.0029598921 & 0.012439112 &  0.0050124685 & 0.565531895\\\\\n",
       "\t chrX &  989754 &  991253 & AKAP17A   & K562 &  0.0061596812 & 0.012948778 &  0.0107104723 & 0.637372967\\\\\n",
       "\t chrX &  989754 &  991253 & ASMT      & K562 & -0.0022939125 & 0.010662312 &  0.0016068427 & 0.642759693\\\\\n",
       "\t chrX &  989754 &  991253 & ASMTL     & K562 &  0.0439144995 & 0.012199586 &  0.0150632279 & 0.009016532\\\\\n",
       "\t chrX &  989754 &  991253 & GTPBP6    & K562 &  0.0305578173 & 0.011430865 &  0.0239597817 & 0.281898029\\\\\n",
       "\t chrX &  989754 &  991253 & IL3RA     & K562 &  0.0125659674 & 0.010855956 & -0.0024098551 & 0.083869885\\\\\n",
       "\t chrX &  989754 &  991253 & LINC00106 & K562 &  0.0229243072 & 0.012028640 &  0.0283747272 & 0.674768909\\\\\n",
       "\t chrX &  989754 &  991253 & P2RY8     & K562 & -0.0041985963 & 0.013039770 & -0.0024558172 & 0.553160746\\\\\n",
       "\t chrX &  989754 &  991253 & PLCXD1    & K562 &  0.0149861580 & 0.014968166 &  0.0132007179 & 0.452525779\\\\\n",
       "\t chrX &  989754 &  991253 & PPP2R3B   & K562 &  0.0492611216 & 0.011158934 &  0.0228027464 & 0.008868848\\\\\n",
       "\t chrX &  989754 &  991253 & SLC25A6   & K562 & -0.0104252429 & 0.015140078 &  0.0052223026 & 0.849319261\\\\\n",
       "\t chrX & 9995827 & 9996208 & CLCN4     & K562 &  0.0040034126 & 0.012058187 &  0.0079896140 & 0.629519308\\\\\n",
       "\t chrX & 9995827 & 9996208 & FAM9B     & K562 &  0.0073610288 & 0.012022706 &  0.0025848947 & 0.345588059\\\\\n",
       "\t chrX & 9995827 & 9996208 & TBL1X     & K562 &  0.0093899593 & 0.010151704 &  0.0094496039 & 0.502343903\\\\\n",
       "\t chrX & 9995827 & 9996208 & WWC3      & K562 &  0.0049533099 & 0.012451989 &  0.0059517592 & 0.531954509\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2914790 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  10001196 |  10001745 | CASZ1    | K562 |  2.373545e-02 | 0.012459711 |  3.976434e-03 | 0.056388671 |\n",
       "| chr1 |  10001196 |  10001745 | CENPS    | K562 |  1.594539e-02 | 0.009814236 |  6.140422e-03 | 0.158883787 |\n",
       "| chr1 |  10001196 |  10001745 | CLSTN1   | K562 |  1.164222e-02 | 0.010707841 |  1.002446e-02 | 0.439955671 |\n",
       "| chr1 |  10001196 |  10001745 | CORT     | K562 | -3.965550e-03 | 0.015192518 |  2.049952e-03 | 0.653929629 |\n",
       "| chr1 |  10001196 |  10001745 | CTNNBIP1 | K562 |  1.418203e-02 | 0.011058809 |  8.415236e-03 | 0.301020957 |\n",
       "| chr1 |  10001196 |  10001745 | DFFA     | K562 |  6.978517e-03 | 0.011281599 |  1.798258e-03 | 0.323053266 |\n",
       "| chr1 |  10001196 |  10001745 | GPR157   | K562 | -1.038945e-02 | 0.012452020 |  1.296185e-03 | 0.825994104 |\n",
       "| chr1 |  10001196 |  10001745 | H6PD     | K562 | -1.595742e-03 | 0.013186113 |  4.040827e-03 | 0.665478754 |\n",
       "| chr1 |  10001196 |  10001745 | KIF1B    | K562 |  2.809136e-02 | 0.011970877 |  5.099747e-03 | 0.027389267 |\n",
       "| chr1 |  10001196 |  10001745 | LZIC     | K562 |  2.520966e-03 | 0.010951603 |  1.770884e-03 | 0.472697533 |\n",
       "| chr1 |  10001196 |  10001745 | MIR34AHG | K562 | -3.619804e-03 | 0.012189632 |  8.564796e-04 | 0.643272738 |\n",
       "| chr1 |  10001196 |  10001745 | NMNAT1   | K562 |  1.289296e-02 | 0.011081800 |  7.657213e-03 | 0.318297863 |\n",
       "| chr1 |  10001196 |  10001745 | PEX14    | K562 | -2.889875e-05 | 0.010171680 |  5.089914e-03 | 0.692602795 |\n",
       "| chr1 |  10001196 |  10001745 | PGD      | K562 |  1.932427e-02 | 0.011505577 |  2.936272e-03 | 0.077172276 |\n",
       "| chr1 |  10001196 |  10001745 | PIK3CD   | K562 |  8.370059e-03 | 0.011481796 |  9.661978e-03 | 0.544794009 |\n",
       "| chr1 |  10001196 |  10001745 | RBP7     | K562 |  2.865984e-02 | 0.010501881 | -8.728314e-05 | 0.003097000 |\n",
       "| chr1 |  10001196 |  10001745 | SLC25A33 | K562 |  1.426118e-02 | 0.011428111 |  3.542587e-03 | 0.174144098 |\n",
       "| chr1 |  10001196 |  10001745 | SPSB1    | K562 |  2.290241e-02 | 0.012223231 | -2.884131e-04 | 0.028895871 |\n",
       "| chr1 |  10001196 |  10001745 | TMEM201  | K562 |  1.616036e-02 | 0.011704654 |  4.605808e-03 | 0.161778287 |\n",
       "| chr1 |  10001196 |  10001745 | UBE4B    | K562 |  2.469338e-02 | 0.010455244 | -1.298238e-03 | 0.006459610 |\n",
       "| chr1 | 100027917 | 100029695 | AGL      | K562 |  3.426649e-02 | 0.012183675 |  6.876781e-03 | 0.012285988 |\n",
       "| chr1 | 100027917 | 100029695 | CDC14A   | K562 |  3.003275e-02 | 0.013179488 |  7.757991e-03 | 0.045503671 |\n",
       "| chr1 | 100027917 | 100029695 | DBT      | K562 |  4.251608e-02 | 0.010802897 |  1.524219e-02 | 0.005790143 |\n",
       "| chr1 | 100027917 | 100029695 | DPH5     | K562 |  4.069875e-02 | 0.012485810 |  9.412662e-03 | 0.006109918 |\n",
       "| chr1 | 100027917 | 100029695 | EXTL2    | K562 |  1.327720e-02 | 0.011202587 |  9.953417e-03 | 0.383348753 |\n",
       "| chr1 | 100027917 | 100029695 | FRRS1    | K562 |  7.008389e-03 | 0.013718161 |  2.996208e-03 | 0.384962798 |\n",
       "| chr1 | 100027917 | 100029695 | MFSD14A  | K562 |  1.332009e-02 | 0.013399283 |  1.232827e-02 | 0.470497259 |\n",
       "| chr1 | 100027917 | 100029695 | PALMD    | K562 |  9.294320e-03 | 0.012127027 |  6.780437e-03 | 0.417889388 |\n",
       "| chr1 | 100027917 | 100029695 | RTCA     | K562 |  4.944637e-03 | 0.010487785 |  7.361271e-03 | 0.591118716 |\n",
       "| chr1 | 100027917 | 100029695 | RTCA-AS1 | K562 |  1.181748e-02 | 0.013982719 |  1.672369e-02 | 0.637159569 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9873821 | 9874426 | CLCN4     | K562 | -0.0051374098 | 0.014699103 |  0.0091326389 | 0.834178767 |\n",
       "| chrX | 9873821 | 9874426 | FAM9B     | K562 |  0.0130287285 | 0.013710062 |  0.0039890571 | 0.254836886 |\n",
       "| chrX | 9873821 | 9874426 | TBL1X     | K562 | -0.0015979203 | 0.011089781 |  0.0106163251 | 0.864637934 |\n",
       "| chrX | 9873821 | 9874426 | WWC3      | K562 | -0.0005950215 | 0.011010227 |  0.0058402515 | 0.720551768 |\n",
       "| chrX | 9882635 | 9883318 | CLCN4     | K562 |  0.0131380399 | 0.012619797 |  0.0051295616 | 0.262845823 |\n",
       "| chrX | 9882635 | 9883318 | FAM9B     | K562 |  0.0081721127 | 0.012714557 |  0.0060027287 | 0.432260472 |\n",
       "| chrX | 9882635 | 9883318 | TBL1X     | K562 | -0.0004361237 | 0.011821801 |  0.0098067019 | 0.806874244 |\n",
       "| chrX | 9882635 | 9883318 | WWC3      | K562 | -0.0117169728 | 0.012745833 |  0.0063099199 | 0.921368385 |\n",
       "| chrX | 9885160 | 9885475 | CLCN4     | K562 | -0.0073263795 | 0.013692948 |  0.0067406944 | 0.847865693 |\n",
       "| chrX | 9885160 | 9885475 | FAM9B     | K562 |  0.0163467680 | 0.011938149 |  0.0037657621 | 0.145976029 |\n",
       "| chrX | 9885160 | 9885475 | TBL1X     | K562 | -0.0042601726 | 0.009579401 |  0.0093703566 | 0.922617423 |\n",
       "| chrX | 9885160 | 9885475 | WWC3      | K562 | -0.0018792908 | 0.013527142 |  0.0072123988 | 0.749242276 |\n",
       "| chrX | 9894484 | 9895274 | CLCN4     | K562 |  0.0189453440 | 0.011605712 |  0.0056802853 | 0.126524132 |\n",
       "| chrX | 9894484 | 9895274 | FAM9B     | K562 |  0.0121673228 | 0.012443440 |  0.0009215641 | 0.183063992 |\n",
       "| chrX | 9894484 | 9895274 | TBL1X     | K562 |  0.0193937461 | 0.013165201 |  0.0110601228 | 0.263365548 |\n",
       "| chrX | 9894484 | 9895274 | WWC3      | K562 |  0.0029598921 | 0.012439112 |  0.0050124685 | 0.565531895 |\n",
       "| chrX |  989754 |  991253 | AKAP17A   | K562 |  0.0061596812 | 0.012948778 |  0.0107104723 | 0.637372967 |\n",
       "| chrX |  989754 |  991253 | ASMT      | K562 | -0.0022939125 | 0.010662312 |  0.0016068427 | 0.642759693 |\n",
       "| chrX |  989754 |  991253 | ASMTL     | K562 |  0.0439144995 | 0.012199586 |  0.0150632279 | 0.009016532 |\n",
       "| chrX |  989754 |  991253 | GTPBP6    | K562 |  0.0305578173 | 0.011430865 |  0.0239597817 | 0.281898029 |\n",
       "| chrX |  989754 |  991253 | IL3RA     | K562 |  0.0125659674 | 0.010855956 | -0.0024098551 | 0.083869885 |\n",
       "| chrX |  989754 |  991253 | LINC00106 | K562 |  0.0229243072 | 0.012028640 |  0.0283747272 | 0.674768909 |\n",
       "| chrX |  989754 |  991253 | P2RY8     | K562 | -0.0041985963 | 0.013039770 | -0.0024558172 | 0.553160746 |\n",
       "| chrX |  989754 |  991253 | PLCXD1    | K562 |  0.0149861580 | 0.014968166 |  0.0132007179 | 0.452525779 |\n",
       "| chrX |  989754 |  991253 | PPP2R3B   | K562 |  0.0492611216 | 0.011158934 |  0.0228027464 | 0.008868848 |\n",
       "| chrX |  989754 |  991253 | SLC25A6   | K562 | -0.0104252429 | 0.015140078 |  0.0052223026 | 0.849319261 |\n",
       "| chrX | 9995827 | 9996208 | CLCN4     | K562 |  0.0040034126 | 0.012058187 |  0.0079896140 | 0.629519308 |\n",
       "| chrX | 9995827 | 9996208 | FAM9B     | K562 |  0.0073610288 | 0.012022706 |  0.0025848947 | 0.345588059 |\n",
       "| chrX | 9995827 | 9996208 | TBL1X     | K562 |  0.0093899593 | 0.010151704 |  0.0094496039 | 0.502343903 |\n",
       "| chrX | 9995827 | 9996208 | WWC3      | K562 |  0.0049533099 | 0.012451989 |  0.0059517592 | 0.531954509 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType rObs          rBgSD      \n",
       "1       chr1  10001196  10001745 CASZ1      K562      2.373545e-02 0.012459711\n",
       "2       chr1  10001196  10001745 CENPS      K562      1.594539e-02 0.009814236\n",
       "3       chr1  10001196  10001745 CLSTN1     K562      1.164222e-02 0.010707841\n",
       "4       chr1  10001196  10001745 CORT       K562     -3.965550e-03 0.015192518\n",
       "5       chr1  10001196  10001745 CTNNBIP1   K562      1.418203e-02 0.011058809\n",
       "6       chr1  10001196  10001745 DFFA       K562      6.978517e-03 0.011281599\n",
       "7       chr1  10001196  10001745 GPR157     K562     -1.038945e-02 0.012452020\n",
       "8       chr1  10001196  10001745 H6PD       K562     -1.595742e-03 0.013186113\n",
       "9       chr1  10001196  10001745 KIF1B      K562      2.809136e-02 0.011970877\n",
       "10      chr1  10001196  10001745 LZIC       K562      2.520966e-03 0.010951603\n",
       "11      chr1  10001196  10001745 MIR34AHG   K562     -3.619804e-03 0.012189632\n",
       "12      chr1  10001196  10001745 NMNAT1     K562      1.289296e-02 0.011081800\n",
       "13      chr1  10001196  10001745 PEX14      K562     -2.889875e-05 0.010171680\n",
       "14      chr1  10001196  10001745 PGD        K562      1.932427e-02 0.011505577\n",
       "15      chr1  10001196  10001745 PIK3CD     K562      8.370059e-03 0.011481796\n",
       "16      chr1  10001196  10001745 RBP7       K562      2.865984e-02 0.010501881\n",
       "17      chr1  10001196  10001745 SLC25A33   K562      1.426118e-02 0.011428111\n",
       "18      chr1  10001196  10001745 SPSB1      K562      2.290241e-02 0.012223231\n",
       "19      chr1  10001196  10001745 TMEM201    K562      1.616036e-02 0.011704654\n",
       "20      chr1  10001196  10001745 UBE4B      K562      2.469338e-02 0.010455244\n",
       "21      chr1 100027917 100029695 AGL        K562      3.426649e-02 0.012183675\n",
       "22      chr1 100027917 100029695 CDC14A     K562      3.003275e-02 0.013179488\n",
       "23      chr1 100027917 100029695 DBT        K562      4.251608e-02 0.010802897\n",
       "24      chr1 100027917 100029695 DPH5       K562      4.069875e-02 0.012485810\n",
       "25      chr1 100027917 100029695 EXTL2      K562      1.327720e-02 0.011202587\n",
       "26      chr1 100027917 100029695 FRRS1      K562      7.008389e-03 0.013718161\n",
       "27      chr1 100027917 100029695 MFSD14A    K562      1.332009e-02 0.013399283\n",
       "28      chr1 100027917 100029695 PALMD      K562      9.294320e-03 0.012127027\n",
       "29      chr1 100027917 100029695 RTCA       K562      4.944637e-03 0.010487785\n",
       "30      chr1 100027917 100029695 RTCA-AS1   K562      1.181748e-02 0.013982719\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮          \n",
       "2914761 chrX 9873821   9874426   CLCN4      K562     -0.0051374098 0.014699103\n",
       "2914762 chrX 9873821   9874426   FAM9B      K562      0.0130287285 0.013710062\n",
       "2914763 chrX 9873821   9874426   TBL1X      K562     -0.0015979203 0.011089781\n",
       "2914764 chrX 9873821   9874426   WWC3       K562     -0.0005950215 0.011010227\n",
       "2914765 chrX 9882635   9883318   CLCN4      K562      0.0131380399 0.012619797\n",
       "2914766 chrX 9882635   9883318   FAM9B      K562      0.0081721127 0.012714557\n",
       "2914767 chrX 9882635   9883318   TBL1X      K562     -0.0004361237 0.011821801\n",
       "2914768 chrX 9882635   9883318   WWC3       K562     -0.0117169728 0.012745833\n",
       "2914769 chrX 9885160   9885475   CLCN4      K562     -0.0073263795 0.013692948\n",
       "2914770 chrX 9885160   9885475   FAM9B      K562      0.0163467680 0.011938149\n",
       "2914771 chrX 9885160   9885475   TBL1X      K562     -0.0042601726 0.009579401\n",
       "2914772 chrX 9885160   9885475   WWC3       K562     -0.0018792908 0.013527142\n",
       "2914773 chrX 9894484   9895274   CLCN4      K562      0.0189453440 0.011605712\n",
       "2914774 chrX 9894484   9895274   FAM9B      K562      0.0121673228 0.012443440\n",
       "2914775 chrX 9894484   9895274   TBL1X      K562      0.0193937461 0.013165201\n",
       "2914776 chrX 9894484   9895274   WWC3       K562      0.0029598921 0.012439112\n",
       "2914777 chrX  989754    991253   AKAP17A    K562      0.0061596812 0.012948778\n",
       "2914778 chrX  989754    991253   ASMT       K562     -0.0022939125 0.010662312\n",
       "2914779 chrX  989754    991253   ASMTL      K562      0.0439144995 0.012199586\n",
       "2914780 chrX  989754    991253   GTPBP6     K562      0.0305578173 0.011430865\n",
       "2914781 chrX  989754    991253   IL3RA      K562      0.0125659674 0.010855956\n",
       "2914782 chrX  989754    991253   LINC00106  K562      0.0229243072 0.012028640\n",
       "2914783 chrX  989754    991253   P2RY8      K562     -0.0041985963 0.013039770\n",
       "2914784 chrX  989754    991253   PLCXD1     K562      0.0149861580 0.014968166\n",
       "2914785 chrX  989754    991253   PPP2R3B    K562      0.0492611216 0.011158934\n",
       "2914786 chrX  989754    991253   SLC25A6    K562     -0.0104252429 0.015140078\n",
       "2914787 chrX 9995827   9996208   CLCN4      K562      0.0040034126 0.012058187\n",
       "2914788 chrX 9995827   9996208   FAM9B      K562      0.0073610288 0.012022706\n",
       "2914789 chrX 9995827   9996208   TBL1X      K562      0.0093899593 0.010151704\n",
       "2914790 chrX 9995827   9996208   WWC3       K562      0.0049533099 0.012451989\n",
       "        rBgMean       pvalZ      \n",
       "1        3.976434e-03 0.056388671\n",
       "2        6.140422e-03 0.158883787\n",
       "3        1.002446e-02 0.439955671\n",
       "4        2.049952e-03 0.653929629\n",
       "5        8.415236e-03 0.301020957\n",
       "6        1.798258e-03 0.323053266\n",
       "7        1.296185e-03 0.825994104\n",
       "8        4.040827e-03 0.665478754\n",
       "9        5.099747e-03 0.027389267\n",
       "10       1.770884e-03 0.472697533\n",
       "11       8.564796e-04 0.643272738\n",
       "12       7.657213e-03 0.318297863\n",
       "13       5.089914e-03 0.692602795\n",
       "14       2.936272e-03 0.077172276\n",
       "15       9.661978e-03 0.544794009\n",
       "16      -8.728314e-05 0.003097000\n",
       "17       3.542587e-03 0.174144098\n",
       "18      -2.884131e-04 0.028895871\n",
       "19       4.605808e-03 0.161778287\n",
       "20      -1.298238e-03 0.006459610\n",
       "21       6.876781e-03 0.012285988\n",
       "22       7.757991e-03 0.045503671\n",
       "23       1.524219e-02 0.005790143\n",
       "24       9.412662e-03 0.006109918\n",
       "25       9.953417e-03 0.383348753\n",
       "26       2.996208e-03 0.384962798\n",
       "27       1.232827e-02 0.470497259\n",
       "28       6.780437e-03 0.417889388\n",
       "29       7.361271e-03 0.591118716\n",
       "30       1.672369e-02 0.637159569\n",
       "⋮       ⋮             ⋮          \n",
       "2914761  0.0091326389 0.834178767\n",
       "2914762  0.0039890571 0.254836886\n",
       "2914763  0.0106163251 0.864637934\n",
       "2914764  0.0058402515 0.720551768\n",
       "2914765  0.0051295616 0.262845823\n",
       "2914766  0.0060027287 0.432260472\n",
       "2914767  0.0098067019 0.806874244\n",
       "2914768  0.0063099199 0.921368385\n",
       "2914769  0.0067406944 0.847865693\n",
       "2914770  0.0037657621 0.145976029\n",
       "2914771  0.0093703566 0.922617423\n",
       "2914772  0.0072123988 0.749242276\n",
       "2914773  0.0056802853 0.126524132\n",
       "2914774  0.0009215641 0.183063992\n",
       "2914775  0.0110601228 0.263365548\n",
       "2914776  0.0050124685 0.565531895\n",
       "2914777  0.0107104723 0.637372967\n",
       "2914778  0.0016068427 0.642759693\n",
       "2914779  0.0150632279 0.009016532\n",
       "2914780  0.0239597817 0.281898029\n",
       "2914781 -0.0024098551 0.083869885\n",
       "2914782  0.0283747272 0.674768909\n",
       "2914783 -0.0024558172 0.553160746\n",
       "2914784  0.0132007179 0.452525779\n",
       "2914785  0.0228027464 0.008868848\n",
       "2914786  0.0052223026 0.849319261\n",
       "2914787  0.0079896140 0.629519308\n",
       "2914788  0.0025848947 0.345588059\n",
       "2914789  0.0094496039 0.502343903\n",
       "2914790  0.0059517592 0.531954509"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.9 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   foreach_1.5.2                    \n",
       " [3] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.68.0                  \n",
       " [5] rtracklayer_1.60.1                Biostrings_2.70.1                \n",
       " [7] XVector_0.42.0                    genomation_1.34.0                \n",
       " [9] Seurat_5.0.3                      SeuratObject_5.0.1               \n",
       "[11] sp_2.1-4                          FigR_0.1.0                       \n",
       "[13] motifmatchr_1.24.0                chromVAR_1.24.0                  \n",
       "[15] cowplot_1.1.3                     ggplot2_3.5.1                    \n",
       "[17] dplyr_1.1.4                       SummarizedExperiment_1.30.2      \n",
       "[19] Biobase_2.60.0                    GenomicRanges_1.54.1             \n",
       "[21] GenomeInfoDb_1.38.1               IRanges_2.36.0                   \n",
       "[23] S4Vectors_0.40.2                  BiocGenerics_0.48.1              \n",
       "[25] MatrixGenerics_1.12.3             matrixStats_1.3.0                \n",
       "[27] Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             vroom_1.6.5                \n",
       " [15] globals_0.16.3              lattice_0.22-6             \n",
       " [17] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [19] plotly_4.10.4               plotrix_3.8-4              \n",
       " [21] yaml_2.3.8                  httpuv_1.6.15              \n",
       " [23] sctransform_0.4.1           spam_2.10-0                \n",
       " [25] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [27] pbapply_1.7-2               DBI_1.2.2                  \n",
       " [29] CNEr_1.38.0                 RColorBrewer_1.1-3         \n",
       " [31] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [33] Rtsne_0.17                  purrr_1.0.2                \n",
       " [35] R.utils_2.12.3              RCurl_1.98-1.14            \n",
       " [37] pracma_2.4.4                GenomeInfoDbData_1.2.11    \n",
       " [39] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [41] spatstat.utils_3.0-4        listenv_0.9.1              \n",
       " [43] seqLogo_1.68.0              goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             spatstat.random_3.2-3      \n",
       " [47] annotate_1.80.0             fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           leiden_0.4.3.1             \n",
       " [51] codetools_0.2-20            DelayedArray_0.26.7        \n",
       " [53] DT_0.33                     tidyselect_1.2.1           \n",
       " [55] spatstat.explore_3.2-6      base64enc_0.1-3            \n",
       " [57] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [59] progressr_0.14.0            iterators_1.0.14           \n",
       " [61] ggridges_0.5.6              survival_3.6-4             \n",
       " [63] tools_4.3.3                 TFMPvalue_0.0.9            \n",
       " [65] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [67] glue_1.7.0                  gridExtra_2.3              \n",
       " [69] IRdisplay_1.1               withr_3.0.0                \n",
       " [71] fastmap_1.1.1               fansi_1.0.6                \n",
       " [73] caTools_1.18.2              digest_0.6.35              \n",
       " [75] R6_2.5.1                    mime_0.12                  \n",
       " [77] nabor_0.5.0                 seqPattern_1.34.0          \n",
       " [79] colorspace_2.1-0            scattermore_1.2            \n",
       " [81] GO.db_3.18.0                tensor_1.5                 \n",
       " [83] gtools_3.9.5                poweRlaw_0.80.0            \n",
       " [85] spatstat.data_3.0-4         RSQLite_2.3.6              \n",
       " [87] R.methodsS3_1.8.2           utf8_1.2.4                 \n",
       " [89] tidyr_1.3.1                 generics_0.1.3             \n",
       " [91] data.table_1.15.2           httr_1.4.7                 \n",
       " [93] htmlwidgets_1.6.4           S4Arrays_1.2.1             \n",
       " [95] TFBSTools_1.40.0            uwot_0.1.16                \n",
       " [97] pkgconfig_2.0.3             gtable_0.3.5               \n",
       " [99] blob_1.2.4                  impute_1.76.0              \n",
       "[101] lmtest_0.9-40               htmltools_0.5.8.1          \n",
       "[103] dotCall64_1.1-1             scales_1.3.0               \n",
       "[105] png_0.1-8                   tzdb_0.4.0                 \n",
       "[107] reshape2_1.4.4              rjson_0.2.21               \n",
       "[109] uuid_1.2-0                  nlme_3.1-164               \n",
       "[111] repr_1.1.7                  cachem_1.0.8               \n",
       "[113] zoo_1.8-12                  stringr_1.5.1              \n",
       "[115] KernSmooth_2.23-22          miniUI_0.1.1.1             \n",
       "[117] AnnotationDbi_1.64.1        restfulr_0.0.15            \n",
       "[119] pillar_1.9.0                vctrs_0.6.5                \n",
       "[121] RANN_2.6.1                  promises_1.3.0             \n",
       "[123] xtable_1.8-4                cluster_2.1.6              \n",
       "[125] evaluate_0.23               readr_2.1.5                \n",
       "[127] cli_3.6.2                   compiler_4.3.3             \n",
       "[129] Rsamtools_2.18.0            rlang_1.1.3                \n",
       "[131] crayon_1.5.2                future.apply_1.11.2        \n",
       "[133] plyr_1.8.9                  stringi_1.8.3              \n",
       "[135] gridBase_0.4-7              deldir_2.0-4               \n",
       "[137] viridisLite_0.4.2           BiocParallel_1.36.0        \n",
       "[139] munsell_0.5.1               lazyeval_0.2.2             \n",
       "[141] spatstat.geom_3.2-9         IRkernel_1.3.2             \n",
       "[143] RcppHNSW_0.6.0              hms_1.1.3                  \n",
       "[145] patchwork_1.2.0             bit64_4.0.5                \n",
       "[147] future_1.33.2               KEGGREST_1.40.1            \n",
       "[149] shiny_1.8.1.1               ROCR_1.0-11                \n",
       "[151] igraph_2.0.3                memoise_2.0.1              \n",
       "[153] bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
