{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/5.merge_peaks_from_5_super_groups.240524/pairs.rds\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/5.merge_peaks_from_5_super_groups.240524/matrix.atac.5_super_groups.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/5.merge_peaks_from_5_super_groups.240524/matrix.rna.5_super_groups.rds\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/FigR/BMMC5.merge.250211/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readRDS(path.pairs.E2G)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  4000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  4001  to  5000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  5001  to  6000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  6001  to  7000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  7001  to  8000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  8001  to  9000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  9001  to  10000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  10001  to  11000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  11001  to  12000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  12001  to  13000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  13001  to  14000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  14001  to  15000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  15001  to  16000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  16001  to  17000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  17001  to  18000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  18001  to  19000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  19001  to  20000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  20001  to  21000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  21001  to  22000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  22001  to  23000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  23001  to  24000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  24001  to  25000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  25001  to  26000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  26001  to  27000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  27001  to  28000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  28001  to  29000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  29001  to  30000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  30001  to  31000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  31001  to  32000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  32001  to  33000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  33001  to  34000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  34001  to  35000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  35001  to  36000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  36001  to  37000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  37001  to  38000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  38001  to  39000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  39001  to  40000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  40001  to  41000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  41001  to  42000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  42001  to  43000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  43001  to  44000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  44001  to  45000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  45001  to  46000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  46001  to  47000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  47001  to  48000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  48001  to  49000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  49001  to  50000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  50001  to  51000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  51001  to  52000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  52001  to  53000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  53001  to  54000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  54001  to  55000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  55001  to  56000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  56001  to  57000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  57001  to  57371 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = readRDS(path.matrix.rna_count)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dd80eda-2392-4d59-8250-1ae62d7e1d1e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b9e26c70-11fd-47c4-8e05-9c96b17d3f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a76b1603-6b00-4ac0-9399-fd341c5a7b88",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter = pairs.E2G[pairs.E2G$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                              pairs.E2G$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 303482 57371 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(303482): chr1-10010-10512 chr1-115495-115916 ...\n",
       "  chrX-156029840-156030373 chrX-156030498-156030725\n",
       "rowData names(2): PeakName bias\n",
       "colnames(57371): s1d1_TAGTTGTCACCCTCAC-1 s1d1_CTATGGCCATAACGGG-1 ...\n",
       "  s4d9_ACACTTGCAACTAGAA-1 s4d9_CACTTAAAGTCTGGGC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna.rename),\n",
    "                       index = 1:nrow(matrix.rna.rename))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(dir.output, \"chr\", sep = \"/\"), recursive = T):\n",
      "“'/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/FigR/BMMC5.merge.250211//chr' already exists”\n"
     ]
    }
   ],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr21\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  14.7866826057434 secs \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Time Elapsed:  32.062141708533 mins \n",
      "\n",
      "[1] \"chr22\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  1.15030011336009 mins \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  2.08849605858326 hours \n",
      "\n",
      "[1] \"chr3\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.04071628252665 mins \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  3.15077706886662 hours \n",
      "\n",
      "[1] \"chr4\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  49.1753671169281 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.53947281711631 hours \n",
      "\n",
      "[1] \"chr5\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  1.33399886687597 mins \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  2.16965261201064 hours \n",
      "\n",
      "[1] \"chr6\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.22416047652562 mins \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  3.53671783612834 hours \n",
      "\n",
      "[1] \"chr7\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  1.55734712680181 mins \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  2.58392297857338 hours \n",
      "\n",
      "[1] \"chr8\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  1.04015526771545 mins \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.75225075582663 hours \n",
      "\n",
      "[1] \"chr9\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  1.62597877581914 mins \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  2.6162059047487 hours \n",
      "\n",
      "[1] \"chrX\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  46.1921992301941 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  1.34582013693121 hours \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter[seqnames(pairs.E2G.filter) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 5660858 ranges and 7 metadata columns:\n",
       "       seqnames              ranges strand |  TargetGene               PeakName\n",
       "          <Rle>           <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 100015450-100016026      * |         AGL chr1-100015450-10001..\n",
       "  chr1     chr1 100015450-100016026      * |      CDC14A chr1-100015450-10001..\n",
       "  chr1     chr1 100015450-100016026      * |         DBT chr1-100015450-10001..\n",
       "  chr1     chr1 100015450-100016026      * |       EXTL2 chr1-100015450-10001..\n",
       "  chr1     chr1 100015450-100016026      * |     MFSD14A chr1-100015450-10001..\n",
       "   ...      ...                 ...    ... .         ...                    ...\n",
       "  chrX     chrX     9995615-9996434      * |        WWC3   chrX-9995615-9996434\n",
       "  chrX     chrX     9997010-9998101      * |       CLCN4   chrX-9997010-9998101\n",
       "  chrX     chrX     9997010-9998101      * |     SHROOM2   chrX-9997010-9998101\n",
       "  chrX     chrX     9997010-9998101      * |       TBL1X   chrX-9997010-9998101\n",
       "  chrX     chrX     9997010-9998101      * |        WWC3   chrX-9997010-9998101\n",
       "                     PairName         rObs       rBgSD      rBgMean     pvalZ\n",
       "                  <character>    <numeric>   <numeric>    <numeric> <numeric>\n",
       "  chr1 chr1-100015450-10001.. -0.000661452 5.91973e-03  0.004338985  0.800863\n",
       "  chr1 chr1-100015450-10001.. -0.003983708 6.18173e-03  0.002464266  0.851542\n",
       "  chr1 chr1-100015450-10001..  0.006195893 4.97894e-03  0.002686028  0.240423\n",
       "  chr1 chr1-100015450-10001.. -0.000534022 3.83454e-05 -0.000551329  0.325872\n",
       "  chr1 chr1-100015450-10001.. -0.000436019 5.88565e-03  0.000385271  0.555489\n",
       "   ...                    ...          ...         ...          ...       ...\n",
       "  chrX chrX-9995615-9996434.. -0.001144368  0.00403125 -5.47942e-06  0.611226\n",
       "  chrX chrX-9997010-9998101.. -0.005716089  0.01201793  2.16844e-03  0.744108\n",
       "  chrX chrX-9997010-9998101..  0.008236849  0.00515909  1.66961e-03  0.101519\n",
       "  chrX chrX-9997010-9998101.. -0.000438857  0.00387469 -6.69463e-05  0.538234\n",
       "  chrX chrX-9997010-9998101.. -0.000620642  0.00652713  9.91703e-04  0.597555\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5660858 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>AGL     </td><td>K562</td><td>-0.0006614520</td><td>0.0059197342</td><td> 0.0043389850</td><td>8.008626e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>CDC14A  </td><td>K562</td><td>-0.0039837077</td><td>0.0061817326</td><td> 0.0024642663</td><td>8.515418e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>DBT     </td><td>K562</td><td> 0.0061958932</td><td>0.0049789360</td><td> 0.0026860277</td><td>2.404229e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>EXTL2   </td><td>K562</td><td>-0.0005340217</td><td>0.0000383454</td><td>-0.0005513286</td><td>3.258719e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>MFSD14A </td><td>K562</td><td>-0.0004360193</td><td>0.0058856524</td><td> 0.0003852706</td><td>5.554887e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>RTCA    </td><td>K562</td><td>-0.0006540575</td><td>0.0060965714</td><td> 0.0008532802</td><td>5.976401e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>RTCA-AS1</td><td>K562</td><td>-0.0002091114</td><td>0.0047291358</td><td> 0.0020535302</td><td>6.838340e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>SASS6   </td><td>K562</td><td> 0.0137621741</td><td>0.0036691607</td><td>-0.0002125908</td><td>6.984717e-05</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>SLC30A7 </td><td>K562</td><td> 0.0105676312</td><td>0.0061930899</td><td> 0.0038290119</td><td>1.382784e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>SLC35A3 </td><td>K562</td><td>-0.0011742738</td><td>0.0037260567</td><td>-0.0001972569</td><td>6.034210e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>TRMT13  </td><td>K562</td><td>-0.0025056092</td><td>0.0122275696</td><td>-0.0057694497</td><td>3.947636e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>AGL     </td><td>K562</td><td> 0.0278457386</td><td>0.0121738836</td><td> 0.0111956212</td><td>8.570433e-02</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>CDC14A  </td><td>K562</td><td>-0.0026537264</td><td>0.0079679811</td><td> 0.0025162589</td><td>7.417807e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>DBT     </td><td>K562</td><td> 0.0117954883</td><td>0.0109596796</td><td> 0.0074657244</td><td>3.463982e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>DPH5    </td><td>K562</td><td>-0.0016608076</td><td>0.0048327807</td><td> 0.0025884944</td><td>8.103716e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>EXTL2   </td><td>K562</td><td>-0.0019021608</td><td>0.0031057761</td><td>-0.0010307033</td><td>6.104885e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>FRRS1   </td><td>K562</td><td>-0.0018225207</td><td>0.0045129186</td><td> 0.0054690385</td><td>9.469213e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>MFSD14A </td><td>K562</td><td>-0.0015530807</td><td>0.0052372370</td><td> 0.0014037088</td><td>7.138170e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>RTCA    </td><td>K562</td><td>-0.0023297226</td><td>0.0044975008</td><td> 0.0008471760</td><td>7.600209e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>RTCA-AS1</td><td>K562</td><td> 0.0120304351</td><td>0.0106523666</td><td> 0.0055588379</td><td>2.717507e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>SASS6   </td><td>K562</td><td> 0.0087167034</td><td>0.0036965608</td><td>-0.0005695629</td><td>6.000128e-03</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>SLC30A7 </td><td>K562</td><td>-0.0057156356</td><td>0.0062197158</td><td> 0.0064811663</td><td>9.750599e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>SLC35A3 </td><td>K562</td><td> 0.0000542782</td><td>0.0045307631</td><td> 0.0002894356</td><td>5.206968e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>TRMT13  </td><td>K562</td><td> 0.0846365588</td><td>0.0418850655</td><td>-0.0086291027</td><td>1.298352e-02</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>VCAM1   </td><td>K562</td><td>-0.0010981748</td><td>0.0060224843</td><td> 0.0015115508</td><td>6.676126e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>AGL     </td><td>K562</td><td> 0.0263232501</td><td>0.0059680608</td><td> 0.0220879581</td><td>2.389576e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>CDC14A  </td><td>K562</td><td> 0.0021381014</td><td>0.0046937895</td><td> 0.0023220994</td><td>5.156347e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>DBT     </td><td>K562</td><td> 0.0363325793</td><td>0.0051884612</td><td> 0.0215515451</td><td>2.194029e-03</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>DPH5    </td><td>K562</td><td> 0.0053203018</td><td>0.0036461029</td><td> 0.0037683685</td><td>3.351845e-01</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>EXTL2   </td><td>K562</td><td> 0.0003892575</td><td>0.0043369744</td><td>-0.0006921897</td><td>4.015430e-01</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9986538</td><td>9987289</td><td>CLCN4    </td><td>K562</td><td>-0.0112570135</td><td>2.743366e-02</td><td> 2.804689e-03</td><td>0.69587435</td></tr>\n",
       "\t<tr><td>chrX</td><td>9986538</td><td>9987289</td><td>SHROOM2  </td><td>K562</td><td> 0.0079796480</td><td>5.002890e-03</td><td> 1.134285e-03</td><td>0.08561195</td></tr>\n",
       "\t<tr><td>chrX</td><td>9986538</td><td>9987289</td><td>TBL1X    </td><td>K562</td><td>-0.0005480673</td><td>3.431085e-03</td><td>-2.007719e-04</td><td>0.54031220</td></tr>\n",
       "\t<tr><td>chrX</td><td>9986538</td><td>9987289</td><td>WWC3     </td><td>K562</td><td>-0.0007750909</td><td>5.596625e-03</td><td> 6.367026e-04</td><td>0.59957918</td></tr>\n",
       "\t<tr><td>chrX</td><td>9987345</td><td>9987562</td><td>CLCN4    </td><td>K562</td><td>-0.0053087279</td><td>9.423946e-03</td><td> 3.708658e-03</td><td>0.83068073</td></tr>\n",
       "\t<tr><td>chrX</td><td>9987345</td><td>9987562</td><td>TBL1X    </td><td>K562</td><td>-0.0002027633</td><td>1.430256e-05</td><td>-2.087018e-04</td><td>0.33899585</td></tr>\n",
       "\t<tr><td>chrX</td><td>9987345</td><td>9987562</td><td>WWC3     </td><td>K562</td><td>-0.0002867532</td><td>6.171567e-03</td><td> 3.218846e-04</td><td>0.53927986</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988628</td><td>9989100</td><td>CLCN4    </td><td>K562</td><td> 0.0054868499</td><td>1.445065e-02</td><td> 4.464044e-03</td><td>0.47178672</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988628</td><td>9989100</td><td>SHROOM2  </td><td>K562</td><td> 0.0030273131</td><td>4.384056e-03</td><td> 1.054755e-03</td><td>0.32637720</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988628</td><td>9989100</td><td>TBL1X    </td><td>K562</td><td>-0.0003032040</td><td>8.277900e-03</td><td> 8.853532e-04</td><td>0.55708471</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988628</td><td>9989100</td><td>WWC3     </td><td>K562</td><td>-0.0004287989</td><td>9.491570e-03</td><td> 1.750256e-03</td><td>0.59079013</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>AKAP17A  </td><td>K562</td><td>-0.0015313904</td><td>5.515657e-03</td><td> 1.512487e-03</td><td>0.70947825</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>ASMT     </td><td>K562</td><td> 0.0087693601</td><td>4.903672e-03</td><td> 2.042287e-03</td><td>0.08505603</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>ASMTL    </td><td>K562</td><td>-0.0034319373</td><td>5.639414e-03</td><td> 8.719698e-03</td><td>0.98441005</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>CSF2RA   </td><td>K562</td><td>-0.0014856410</td><td>6.592915e-03</td><td> 1.880887e-03</td><td>0.69519426</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>GTPBP6   </td><td>K562</td><td>-0.0010503603</td><td>5.424612e-03</td><td> 9.037726e-04</td><td>0.64066416</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>IL3RA    </td><td>K562</td><td>-0.0035121158</td><td>4.502553e-03</td><td> 2.242138e-03</td><td>0.89937493</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>LINC00106</td><td>K562</td><td>-0.0015313904</td><td>6.874118e-03</td><td> 3.143991e-03</td><td>0.75179294</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>P2RY8    </td><td>K562</td><td>-0.0006946936</td><td>7.292276e-03</td><td> 1.697270e-03</td><td>0.62854921</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>PLCXD1   </td><td>K562</td><td>-0.0008303393</td><td>5.337704e-03</td><td> 5.947121e-04</td><td>0.60525710</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>PPP2R3B  </td><td>K562</td><td>-0.0017619532</td><td>4.611258e-03</td><td> 9.138654e-04</td><td>0.71913693</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>SLC25A6  </td><td>K562</td><td>-0.0016611130</td><td>4.439692e-03</td><td> 5.123424e-04</td><td>0.68777414</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995615</td><td>9996434</td><td>CLCN4    </td><td>K562</td><td>-0.0023989965</td><td>2.759288e-02</td><td>-3.692285e-03</td><td>0.48130827</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995615</td><td>9996434</td><td>SHROOM2  </td><td>K562</td><td>-0.0009758725</td><td>5.372765e-03</td><td> 3.538505e-03</td><td>0.79961129</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995615</td><td>9996434</td><td>TBL1X    </td><td>K562</td><td>-0.0008091830</td><td>4.515297e-03</td><td> 1.320677e-04</td><td>0.58256439</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995615</td><td>9996434</td><td>WWC3     </td><td>K562</td><td>-0.0011443676</td><td>4.031253e-03</td><td>-5.479417e-06</td><td>0.61122556</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997010</td><td>9998101</td><td>CLCN4    </td><td>K562</td><td>-0.0057160887</td><td>1.201793e-02</td><td> 2.168439e-03</td><td>0.74410833</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997010</td><td>9998101</td><td>SHROOM2  </td><td>K562</td><td> 0.0082368487</td><td>5.159095e-03</td><td> 1.669611e-03</td><td>0.10151899</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997010</td><td>9998101</td><td>TBL1X    </td><td>K562</td><td>-0.0004388566</td><td>3.874692e-03</td><td>-6.694626e-05</td><td>0.53823356</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997010</td><td>9998101</td><td>WWC3     </td><td>K562</td><td>-0.0006206424</td><td>6.527127e-03</td><td> 9.917034e-04</td><td>0.59755450</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5660858 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100015450 & 100016026 & AGL      & K562 & -0.0006614520 & 0.0059197342 &  0.0043389850 & 8.008626e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & CDC14A   & K562 & -0.0039837077 & 0.0061817326 &  0.0024642663 & 8.515418e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & DBT      & K562 &  0.0061958932 & 0.0049789360 &  0.0026860277 & 2.404229e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & EXTL2    & K562 & -0.0005340217 & 0.0000383454 & -0.0005513286 & 3.258719e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & MFSD14A  & K562 & -0.0004360193 & 0.0058856524 &  0.0003852706 & 5.554887e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & RTCA     & K562 & -0.0006540575 & 0.0060965714 &  0.0008532802 & 5.976401e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & RTCA-AS1 & K562 & -0.0002091114 & 0.0047291358 &  0.0020535302 & 6.838340e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & SASS6    & K562 &  0.0137621741 & 0.0036691607 & -0.0002125908 & 6.984717e-05\\\\\n",
       "\t chr1 & 100015450 & 100016026 & SLC30A7  & K562 &  0.0105676312 & 0.0061930899 &  0.0038290119 & 1.382784e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & SLC35A3  & K562 & -0.0011742738 & 0.0037260567 & -0.0001972569 & 6.034210e-01\\\\\n",
       "\t chr1 & 100015450 & 100016026 & TRMT13   & K562 & -0.0025056092 & 0.0122275696 & -0.0057694497 & 3.947636e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & AGL      & K562 &  0.0278457386 & 0.0121738836 &  0.0111956212 & 8.570433e-02\\\\\n",
       "\t chr1 & 100028395 & 100029373 & CDC14A   & K562 & -0.0026537264 & 0.0079679811 &  0.0025162589 & 7.417807e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & DBT      & K562 &  0.0117954883 & 0.0109596796 &  0.0074657244 & 3.463982e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & DPH5     & K562 & -0.0016608076 & 0.0048327807 &  0.0025884944 & 8.103716e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & EXTL2    & K562 & -0.0019021608 & 0.0031057761 & -0.0010307033 & 6.104885e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & FRRS1    & K562 & -0.0018225207 & 0.0045129186 &  0.0054690385 & 9.469213e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & MFSD14A  & K562 & -0.0015530807 & 0.0052372370 &  0.0014037088 & 7.138170e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & RTCA     & K562 & -0.0023297226 & 0.0044975008 &  0.0008471760 & 7.600209e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & RTCA-AS1 & K562 &  0.0120304351 & 0.0106523666 &  0.0055588379 & 2.717507e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & SASS6    & K562 &  0.0087167034 & 0.0036965608 & -0.0005695629 & 6.000128e-03\\\\\n",
       "\t chr1 & 100028395 & 100029373 & SLC30A7  & K562 & -0.0057156356 & 0.0062197158 &  0.0064811663 & 9.750599e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & SLC35A3  & K562 &  0.0000542782 & 0.0045307631 &  0.0002894356 & 5.206968e-01\\\\\n",
       "\t chr1 & 100028395 & 100029373 & TRMT13   & K562 &  0.0846365588 & 0.0418850655 & -0.0086291027 & 1.298352e-02\\\\\n",
       "\t chr1 & 100028395 & 100029373 & VCAM1    & K562 & -0.0010981748 & 0.0060224843 &  0.0015115508 & 6.676126e-01\\\\\n",
       "\t chr1 & 100037560 & 100039288 & AGL      & K562 &  0.0263232501 & 0.0059680608 &  0.0220879581 & 2.389576e-01\\\\\n",
       "\t chr1 & 100037560 & 100039288 & CDC14A   & K562 &  0.0021381014 & 0.0046937895 &  0.0023220994 & 5.156347e-01\\\\\n",
       "\t chr1 & 100037560 & 100039288 & DBT      & K562 &  0.0363325793 & 0.0051884612 &  0.0215515451 & 2.194029e-03\\\\\n",
       "\t chr1 & 100037560 & 100039288 & DPH5     & K562 &  0.0053203018 & 0.0036461029 &  0.0037683685 & 3.351845e-01\\\\\n",
       "\t chr1 & 100037560 & 100039288 & EXTL2    & K562 &  0.0003892575 & 0.0043369744 & -0.0006921897 & 4.015430e-01\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9986538 & 9987289 & CLCN4     & K562 & -0.0112570135 & 2.743366e-02 &  2.804689e-03 & 0.69587435\\\\\n",
       "\t chrX & 9986538 & 9987289 & SHROOM2   & K562 &  0.0079796480 & 5.002890e-03 &  1.134285e-03 & 0.08561195\\\\\n",
       "\t chrX & 9986538 & 9987289 & TBL1X     & K562 & -0.0005480673 & 3.431085e-03 & -2.007719e-04 & 0.54031220\\\\\n",
       "\t chrX & 9986538 & 9987289 & WWC3      & K562 & -0.0007750909 & 5.596625e-03 &  6.367026e-04 & 0.59957918\\\\\n",
       "\t chrX & 9987345 & 9987562 & CLCN4     & K562 & -0.0053087279 & 9.423946e-03 &  3.708658e-03 & 0.83068073\\\\\n",
       "\t chrX & 9987345 & 9987562 & TBL1X     & K562 & -0.0002027633 & 1.430256e-05 & -2.087018e-04 & 0.33899585\\\\\n",
       "\t chrX & 9987345 & 9987562 & WWC3      & K562 & -0.0002867532 & 6.171567e-03 &  3.218846e-04 & 0.53927986\\\\\n",
       "\t chrX & 9988628 & 9989100 & CLCN4     & K562 &  0.0054868499 & 1.445065e-02 &  4.464044e-03 & 0.47178672\\\\\n",
       "\t chrX & 9988628 & 9989100 & SHROOM2   & K562 &  0.0030273131 & 4.384056e-03 &  1.054755e-03 & 0.32637720\\\\\n",
       "\t chrX & 9988628 & 9989100 & TBL1X     & K562 & -0.0003032040 & 8.277900e-03 &  8.853532e-04 & 0.55708471\\\\\n",
       "\t chrX & 9988628 & 9989100 & WWC3      & K562 & -0.0004287989 & 9.491570e-03 &  1.750256e-03 & 0.59079013\\\\\n",
       "\t chrX &  999358 &  999880 & AKAP17A   & K562 & -0.0015313904 & 5.515657e-03 &  1.512487e-03 & 0.70947825\\\\\n",
       "\t chrX &  999358 &  999880 & ASMT      & K562 &  0.0087693601 & 4.903672e-03 &  2.042287e-03 & 0.08505603\\\\\n",
       "\t chrX &  999358 &  999880 & ASMTL     & K562 & -0.0034319373 & 5.639414e-03 &  8.719698e-03 & 0.98441005\\\\\n",
       "\t chrX &  999358 &  999880 & CSF2RA    & K562 & -0.0014856410 & 6.592915e-03 &  1.880887e-03 & 0.69519426\\\\\n",
       "\t chrX &  999358 &  999880 & GTPBP6    & K562 & -0.0010503603 & 5.424612e-03 &  9.037726e-04 & 0.64066416\\\\\n",
       "\t chrX &  999358 &  999880 & IL3RA     & K562 & -0.0035121158 & 4.502553e-03 &  2.242138e-03 & 0.89937493\\\\\n",
       "\t chrX &  999358 &  999880 & LINC00106 & K562 & -0.0015313904 & 6.874118e-03 &  3.143991e-03 & 0.75179294\\\\\n",
       "\t chrX &  999358 &  999880 & P2RY8     & K562 & -0.0006946936 & 7.292276e-03 &  1.697270e-03 & 0.62854921\\\\\n",
       "\t chrX &  999358 &  999880 & PLCXD1    & K562 & -0.0008303393 & 5.337704e-03 &  5.947121e-04 & 0.60525710\\\\\n",
       "\t chrX &  999358 &  999880 & PPP2R3B   & K562 & -0.0017619532 & 4.611258e-03 &  9.138654e-04 & 0.71913693\\\\\n",
       "\t chrX &  999358 &  999880 & SLC25A6   & K562 & -0.0016611130 & 4.439692e-03 &  5.123424e-04 & 0.68777414\\\\\n",
       "\t chrX & 9995615 & 9996434 & CLCN4     & K562 & -0.0023989965 & 2.759288e-02 & -3.692285e-03 & 0.48130827\\\\\n",
       "\t chrX & 9995615 & 9996434 & SHROOM2   & K562 & -0.0009758725 & 5.372765e-03 &  3.538505e-03 & 0.79961129\\\\\n",
       "\t chrX & 9995615 & 9996434 & TBL1X     & K562 & -0.0008091830 & 4.515297e-03 &  1.320677e-04 & 0.58256439\\\\\n",
       "\t chrX & 9995615 & 9996434 & WWC3      & K562 & -0.0011443676 & 4.031253e-03 & -5.479417e-06 & 0.61122556\\\\\n",
       "\t chrX & 9997010 & 9998101 & CLCN4     & K562 & -0.0057160887 & 1.201793e-02 &  2.168439e-03 & 0.74410833\\\\\n",
       "\t chrX & 9997010 & 9998101 & SHROOM2   & K562 &  0.0082368487 & 5.159095e-03 &  1.669611e-03 & 0.10151899\\\\\n",
       "\t chrX & 9997010 & 9998101 & TBL1X     & K562 & -0.0004388566 & 3.874692e-03 & -6.694626e-05 & 0.53823356\\\\\n",
       "\t chrX & 9997010 & 9998101 & WWC3      & K562 & -0.0006206424 & 6.527127e-03 &  9.917034e-04 & 0.59755450\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5660858 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100015450 | 100016026 | AGL      | K562 | -0.0006614520 | 0.0059197342 |  0.0043389850 | 8.008626e-01 |\n",
       "| chr1 | 100015450 | 100016026 | CDC14A   | K562 | -0.0039837077 | 0.0061817326 |  0.0024642663 | 8.515418e-01 |\n",
       "| chr1 | 100015450 | 100016026 | DBT      | K562 |  0.0061958932 | 0.0049789360 |  0.0026860277 | 2.404229e-01 |\n",
       "| chr1 | 100015450 | 100016026 | EXTL2    | K562 | -0.0005340217 | 0.0000383454 | -0.0005513286 | 3.258719e-01 |\n",
       "| chr1 | 100015450 | 100016026 | MFSD14A  | K562 | -0.0004360193 | 0.0058856524 |  0.0003852706 | 5.554887e-01 |\n",
       "| chr1 | 100015450 | 100016026 | RTCA     | K562 | -0.0006540575 | 0.0060965714 |  0.0008532802 | 5.976401e-01 |\n",
       "| chr1 | 100015450 | 100016026 | RTCA-AS1 | K562 | -0.0002091114 | 0.0047291358 |  0.0020535302 | 6.838340e-01 |\n",
       "| chr1 | 100015450 | 100016026 | SASS6    | K562 |  0.0137621741 | 0.0036691607 | -0.0002125908 | 6.984717e-05 |\n",
       "| chr1 | 100015450 | 100016026 | SLC30A7  | K562 |  0.0105676312 | 0.0061930899 |  0.0038290119 | 1.382784e-01 |\n",
       "| chr1 | 100015450 | 100016026 | SLC35A3  | K562 | -0.0011742738 | 0.0037260567 | -0.0001972569 | 6.034210e-01 |\n",
       "| chr1 | 100015450 | 100016026 | TRMT13   | K562 | -0.0025056092 | 0.0122275696 | -0.0057694497 | 3.947636e-01 |\n",
       "| chr1 | 100028395 | 100029373 | AGL      | K562 |  0.0278457386 | 0.0121738836 |  0.0111956212 | 8.570433e-02 |\n",
       "| chr1 | 100028395 | 100029373 | CDC14A   | K562 | -0.0026537264 | 0.0079679811 |  0.0025162589 | 7.417807e-01 |\n",
       "| chr1 | 100028395 | 100029373 | DBT      | K562 |  0.0117954883 | 0.0109596796 |  0.0074657244 | 3.463982e-01 |\n",
       "| chr1 | 100028395 | 100029373 | DPH5     | K562 | -0.0016608076 | 0.0048327807 |  0.0025884944 | 8.103716e-01 |\n",
       "| chr1 | 100028395 | 100029373 | EXTL2    | K562 | -0.0019021608 | 0.0031057761 | -0.0010307033 | 6.104885e-01 |\n",
       "| chr1 | 100028395 | 100029373 | FRRS1    | K562 | -0.0018225207 | 0.0045129186 |  0.0054690385 | 9.469213e-01 |\n",
       "| chr1 | 100028395 | 100029373 | MFSD14A  | K562 | -0.0015530807 | 0.0052372370 |  0.0014037088 | 7.138170e-01 |\n",
       "| chr1 | 100028395 | 100029373 | RTCA     | K562 | -0.0023297226 | 0.0044975008 |  0.0008471760 | 7.600209e-01 |\n",
       "| chr1 | 100028395 | 100029373 | RTCA-AS1 | K562 |  0.0120304351 | 0.0106523666 |  0.0055588379 | 2.717507e-01 |\n",
       "| chr1 | 100028395 | 100029373 | SASS6    | K562 |  0.0087167034 | 0.0036965608 | -0.0005695629 | 6.000128e-03 |\n",
       "| chr1 | 100028395 | 100029373 | SLC30A7  | K562 | -0.0057156356 | 0.0062197158 |  0.0064811663 | 9.750599e-01 |\n",
       "| chr1 | 100028395 | 100029373 | SLC35A3  | K562 |  0.0000542782 | 0.0045307631 |  0.0002894356 | 5.206968e-01 |\n",
       "| chr1 | 100028395 | 100029373 | TRMT13   | K562 |  0.0846365588 | 0.0418850655 | -0.0086291027 | 1.298352e-02 |\n",
       "| chr1 | 100028395 | 100029373 | VCAM1    | K562 | -0.0010981748 | 0.0060224843 |  0.0015115508 | 6.676126e-01 |\n",
       "| chr1 | 100037560 | 100039288 | AGL      | K562 |  0.0263232501 | 0.0059680608 |  0.0220879581 | 2.389576e-01 |\n",
       "| chr1 | 100037560 | 100039288 | CDC14A   | K562 |  0.0021381014 | 0.0046937895 |  0.0023220994 | 5.156347e-01 |\n",
       "| chr1 | 100037560 | 100039288 | DBT      | K562 |  0.0363325793 | 0.0051884612 |  0.0215515451 | 2.194029e-03 |\n",
       "| chr1 | 100037560 | 100039288 | DPH5     | K562 |  0.0053203018 | 0.0036461029 |  0.0037683685 | 3.351845e-01 |\n",
       "| chr1 | 100037560 | 100039288 | EXTL2    | K562 |  0.0003892575 | 0.0043369744 | -0.0006921897 | 4.015430e-01 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9986538 | 9987289 | CLCN4     | K562 | -0.0112570135 | 2.743366e-02 |  2.804689e-03 | 0.69587435 |\n",
       "| chrX | 9986538 | 9987289 | SHROOM2   | K562 |  0.0079796480 | 5.002890e-03 |  1.134285e-03 | 0.08561195 |\n",
       "| chrX | 9986538 | 9987289 | TBL1X     | K562 | -0.0005480673 | 3.431085e-03 | -2.007719e-04 | 0.54031220 |\n",
       "| chrX | 9986538 | 9987289 | WWC3      | K562 | -0.0007750909 | 5.596625e-03 |  6.367026e-04 | 0.59957918 |\n",
       "| chrX | 9987345 | 9987562 | CLCN4     | K562 | -0.0053087279 | 9.423946e-03 |  3.708658e-03 | 0.83068073 |\n",
       "| chrX | 9987345 | 9987562 | TBL1X     | K562 | -0.0002027633 | 1.430256e-05 | -2.087018e-04 | 0.33899585 |\n",
       "| chrX | 9987345 | 9987562 | WWC3      | K562 | -0.0002867532 | 6.171567e-03 |  3.218846e-04 | 0.53927986 |\n",
       "| chrX | 9988628 | 9989100 | CLCN4     | K562 |  0.0054868499 | 1.445065e-02 |  4.464044e-03 | 0.47178672 |\n",
       "| chrX | 9988628 | 9989100 | SHROOM2   | K562 |  0.0030273131 | 4.384056e-03 |  1.054755e-03 | 0.32637720 |\n",
       "| chrX | 9988628 | 9989100 | TBL1X     | K562 | -0.0003032040 | 8.277900e-03 |  8.853532e-04 | 0.55708471 |\n",
       "| chrX | 9988628 | 9989100 | WWC3      | K562 | -0.0004287989 | 9.491570e-03 |  1.750256e-03 | 0.59079013 |\n",
       "| chrX |  999358 |  999880 | AKAP17A   | K562 | -0.0015313904 | 5.515657e-03 |  1.512487e-03 | 0.70947825 |\n",
       "| chrX |  999358 |  999880 | ASMT      | K562 |  0.0087693601 | 4.903672e-03 |  2.042287e-03 | 0.08505603 |\n",
       "| chrX |  999358 |  999880 | ASMTL     | K562 | -0.0034319373 | 5.639414e-03 |  8.719698e-03 | 0.98441005 |\n",
       "| chrX |  999358 |  999880 | CSF2RA    | K562 | -0.0014856410 | 6.592915e-03 |  1.880887e-03 | 0.69519426 |\n",
       "| chrX |  999358 |  999880 | GTPBP6    | K562 | -0.0010503603 | 5.424612e-03 |  9.037726e-04 | 0.64066416 |\n",
       "| chrX |  999358 |  999880 | IL3RA     | K562 | -0.0035121158 | 4.502553e-03 |  2.242138e-03 | 0.89937493 |\n",
       "| chrX |  999358 |  999880 | LINC00106 | K562 | -0.0015313904 | 6.874118e-03 |  3.143991e-03 | 0.75179294 |\n",
       "| chrX |  999358 |  999880 | P2RY8     | K562 | -0.0006946936 | 7.292276e-03 |  1.697270e-03 | 0.62854921 |\n",
       "| chrX |  999358 |  999880 | PLCXD1    | K562 | -0.0008303393 | 5.337704e-03 |  5.947121e-04 | 0.60525710 |\n",
       "| chrX |  999358 |  999880 | PPP2R3B   | K562 | -0.0017619532 | 4.611258e-03 |  9.138654e-04 | 0.71913693 |\n",
       "| chrX |  999358 |  999880 | SLC25A6   | K562 | -0.0016611130 | 4.439692e-03 |  5.123424e-04 | 0.68777414 |\n",
       "| chrX | 9995615 | 9996434 | CLCN4     | K562 | -0.0023989965 | 2.759288e-02 | -3.692285e-03 | 0.48130827 |\n",
       "| chrX | 9995615 | 9996434 | SHROOM2   | K562 | -0.0009758725 | 5.372765e-03 |  3.538505e-03 | 0.79961129 |\n",
       "| chrX | 9995615 | 9996434 | TBL1X     | K562 | -0.0008091830 | 4.515297e-03 |  1.320677e-04 | 0.58256439 |\n",
       "| chrX | 9995615 | 9996434 | WWC3      | K562 | -0.0011443676 | 4.031253e-03 | -5.479417e-06 | 0.61122556 |\n",
       "| chrX | 9997010 | 9998101 | CLCN4     | K562 | -0.0057160887 | 1.201793e-02 |  2.168439e-03 | 0.74410833 |\n",
       "| chrX | 9997010 | 9998101 | SHROOM2   | K562 |  0.0082368487 | 5.159095e-03 |  1.669611e-03 | 0.10151899 |\n",
       "| chrX | 9997010 | 9998101 | TBL1X     | K562 | -0.0004388566 | 3.874692e-03 | -6.694626e-05 | 0.53823356 |\n",
       "| chrX | 9997010 | 9998101 | WWC3      | K562 | -0.0006206424 | 6.527127e-03 |  9.917034e-04 | 0.59755450 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType rObs          rBgSD       \n",
       "1       chr1 100015450 100016026 AGL        K562     -0.0006614520 0.0059197342\n",
       "2       chr1 100015450 100016026 CDC14A     K562     -0.0039837077 0.0061817326\n",
       "3       chr1 100015450 100016026 DBT        K562      0.0061958932 0.0049789360\n",
       "4       chr1 100015450 100016026 EXTL2      K562     -0.0005340217 0.0000383454\n",
       "5       chr1 100015450 100016026 MFSD14A    K562     -0.0004360193 0.0058856524\n",
       "6       chr1 100015450 100016026 RTCA       K562     -0.0006540575 0.0060965714\n",
       "7       chr1 100015450 100016026 RTCA-AS1   K562     -0.0002091114 0.0047291358\n",
       "8       chr1 100015450 100016026 SASS6      K562      0.0137621741 0.0036691607\n",
       "9       chr1 100015450 100016026 SLC30A7    K562      0.0105676312 0.0061930899\n",
       "10      chr1 100015450 100016026 SLC35A3    K562     -0.0011742738 0.0037260567\n",
       "11      chr1 100015450 100016026 TRMT13     K562     -0.0025056092 0.0122275696\n",
       "12      chr1 100028395 100029373 AGL        K562      0.0278457386 0.0121738836\n",
       "13      chr1 100028395 100029373 CDC14A     K562     -0.0026537264 0.0079679811\n",
       "14      chr1 100028395 100029373 DBT        K562      0.0117954883 0.0109596796\n",
       "15      chr1 100028395 100029373 DPH5       K562     -0.0016608076 0.0048327807\n",
       "16      chr1 100028395 100029373 EXTL2      K562     -0.0019021608 0.0031057761\n",
       "17      chr1 100028395 100029373 FRRS1      K562     -0.0018225207 0.0045129186\n",
       "18      chr1 100028395 100029373 MFSD14A    K562     -0.0015530807 0.0052372370\n",
       "19      chr1 100028395 100029373 RTCA       K562     -0.0023297226 0.0044975008\n",
       "20      chr1 100028395 100029373 RTCA-AS1   K562      0.0120304351 0.0106523666\n",
       "21      chr1 100028395 100029373 SASS6      K562      0.0087167034 0.0036965608\n",
       "22      chr1 100028395 100029373 SLC30A7    K562     -0.0057156356 0.0062197158\n",
       "23      chr1 100028395 100029373 SLC35A3    K562      0.0000542782 0.0045307631\n",
       "24      chr1 100028395 100029373 TRMT13     K562      0.0846365588 0.0418850655\n",
       "25      chr1 100028395 100029373 VCAM1      K562     -0.0010981748 0.0060224843\n",
       "26      chr1 100037560 100039288 AGL        K562      0.0263232501 0.0059680608\n",
       "27      chr1 100037560 100039288 CDC14A     K562      0.0021381014 0.0046937895\n",
       "28      chr1 100037560 100039288 DBT        K562      0.0363325793 0.0051884612\n",
       "29      chr1 100037560 100039288 DPH5       K562      0.0053203018 0.0036461029\n",
       "30      chr1 100037560 100039288 EXTL2      K562      0.0003892575 0.0043369744\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮           \n",
       "5660829 chrX 9986538   9987289   CLCN4      K562     -0.0112570135 2.743366e-02\n",
       "5660830 chrX 9986538   9987289   SHROOM2    K562      0.0079796480 5.002890e-03\n",
       "5660831 chrX 9986538   9987289   TBL1X      K562     -0.0005480673 3.431085e-03\n",
       "5660832 chrX 9986538   9987289   WWC3       K562     -0.0007750909 5.596625e-03\n",
       "5660833 chrX 9987345   9987562   CLCN4      K562     -0.0053087279 9.423946e-03\n",
       "5660834 chrX 9987345   9987562   TBL1X      K562     -0.0002027633 1.430256e-05\n",
       "5660835 chrX 9987345   9987562   WWC3       K562     -0.0002867532 6.171567e-03\n",
       "5660836 chrX 9988628   9989100   CLCN4      K562      0.0054868499 1.445065e-02\n",
       "5660837 chrX 9988628   9989100   SHROOM2    K562      0.0030273131 4.384056e-03\n",
       "5660838 chrX 9988628   9989100   TBL1X      K562     -0.0003032040 8.277900e-03\n",
       "5660839 chrX 9988628   9989100   WWC3       K562     -0.0004287989 9.491570e-03\n",
       "5660840 chrX  999358    999880   AKAP17A    K562     -0.0015313904 5.515657e-03\n",
       "5660841 chrX  999358    999880   ASMT       K562      0.0087693601 4.903672e-03\n",
       "5660842 chrX  999358    999880   ASMTL      K562     -0.0034319373 5.639414e-03\n",
       "5660843 chrX  999358    999880   CSF2RA     K562     -0.0014856410 6.592915e-03\n",
       "5660844 chrX  999358    999880   GTPBP6     K562     -0.0010503603 5.424612e-03\n",
       "5660845 chrX  999358    999880   IL3RA      K562     -0.0035121158 4.502553e-03\n",
       "5660846 chrX  999358    999880   LINC00106  K562     -0.0015313904 6.874118e-03\n",
       "5660847 chrX  999358    999880   P2RY8      K562     -0.0006946936 7.292276e-03\n",
       "5660848 chrX  999358    999880   PLCXD1     K562     -0.0008303393 5.337704e-03\n",
       "5660849 chrX  999358    999880   PPP2R3B    K562     -0.0017619532 4.611258e-03\n",
       "5660850 chrX  999358    999880   SLC25A6    K562     -0.0016611130 4.439692e-03\n",
       "5660851 chrX 9995615   9996434   CLCN4      K562     -0.0023989965 2.759288e-02\n",
       "5660852 chrX 9995615   9996434   SHROOM2    K562     -0.0009758725 5.372765e-03\n",
       "5660853 chrX 9995615   9996434   TBL1X      K562     -0.0008091830 4.515297e-03\n",
       "5660854 chrX 9995615   9996434   WWC3       K562     -0.0011443676 4.031253e-03\n",
       "5660855 chrX 9997010   9998101   CLCN4      K562     -0.0057160887 1.201793e-02\n",
       "5660856 chrX 9997010   9998101   SHROOM2    K562      0.0082368487 5.159095e-03\n",
       "5660857 chrX 9997010   9998101   TBL1X      K562     -0.0004388566 3.874692e-03\n",
       "5660858 chrX 9997010   9998101   WWC3       K562     -0.0006206424 6.527127e-03\n",
       "        rBgMean       pvalZ       \n",
       "1        0.0043389850 8.008626e-01\n",
       "2        0.0024642663 8.515418e-01\n",
       "3        0.0026860277 2.404229e-01\n",
       "4       -0.0005513286 3.258719e-01\n",
       "5        0.0003852706 5.554887e-01\n",
       "6        0.0008532802 5.976401e-01\n",
       "7        0.0020535302 6.838340e-01\n",
       "8       -0.0002125908 6.984717e-05\n",
       "9        0.0038290119 1.382784e-01\n",
       "10      -0.0001972569 6.034210e-01\n",
       "11      -0.0057694497 3.947636e-01\n",
       "12       0.0111956212 8.570433e-02\n",
       "13       0.0025162589 7.417807e-01\n",
       "14       0.0074657244 3.463982e-01\n",
       "15       0.0025884944 8.103716e-01\n",
       "16      -0.0010307033 6.104885e-01\n",
       "17       0.0054690385 9.469213e-01\n",
       "18       0.0014037088 7.138170e-01\n",
       "19       0.0008471760 7.600209e-01\n",
       "20       0.0055588379 2.717507e-01\n",
       "21      -0.0005695629 6.000128e-03\n",
       "22       0.0064811663 9.750599e-01\n",
       "23       0.0002894356 5.206968e-01\n",
       "24      -0.0086291027 1.298352e-02\n",
       "25       0.0015115508 6.676126e-01\n",
       "26       0.0220879581 2.389576e-01\n",
       "27       0.0023220994 5.156347e-01\n",
       "28       0.0215515451 2.194029e-03\n",
       "29       0.0037683685 3.351845e-01\n",
       "30      -0.0006921897 4.015430e-01\n",
       "⋮       ⋮             ⋮           \n",
       "5660829  2.804689e-03 0.69587435  \n",
       "5660830  1.134285e-03 0.08561195  \n",
       "5660831 -2.007719e-04 0.54031220  \n",
       "5660832  6.367026e-04 0.59957918  \n",
       "5660833  3.708658e-03 0.83068073  \n",
       "5660834 -2.087018e-04 0.33899585  \n",
       "5660835  3.218846e-04 0.53927986  \n",
       "5660836  4.464044e-03 0.47178672  \n",
       "5660837  1.054755e-03 0.32637720  \n",
       "5660838  8.853532e-04 0.55708471  \n",
       "5660839  1.750256e-03 0.59079013  \n",
       "5660840  1.512487e-03 0.70947825  \n",
       "5660841  2.042287e-03 0.08505603  \n",
       "5660842  8.719698e-03 0.98441005  \n",
       "5660843  1.880887e-03 0.69519426  \n",
       "5660844  9.037726e-04 0.64066416  \n",
       "5660845  2.242138e-03 0.89937493  \n",
       "5660846  3.143991e-03 0.75179294  \n",
       "5660847  1.697270e-03 0.62854921  \n",
       "5660848  5.947121e-04 0.60525710  \n",
       "5660849  9.138654e-04 0.71913693  \n",
       "5660850  5.123424e-04 0.68777414  \n",
       "5660851 -3.692285e-03 0.48130827  \n",
       "5660852  3.538505e-03 0.79961129  \n",
       "5660853  1.320677e-04 0.58256439  \n",
       "5660854 -5.479417e-06 0.61122556  \n",
       "5660855  2.168439e-03 0.74410833  \n",
       "5660856  1.669611e-03 0.10151899  \n",
       "5660857 -6.694626e-05 0.53823356  \n",
       "5660858  9.917034e-04 0.59755450  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   data.table_1.15.2                \n",
       " [3] foreach_1.5.2                     BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] genomation_1.34.0                 Seurat_5.0.3                     \n",
       "[11] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[13] FigR_0.1.0                        motifmatchr_1.24.0               \n",
       "[15] chromVAR_1.24.0                   cowplot_1.1.3                    \n",
       "[17] ggplot2_3.5.1                     dplyr_1.1.4                      \n",
       "[19] SummarizedExperiment_1.30.2       Biobase_2.60.0                   \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               MatrixGenerics_1.12.3            \n",
       "[27] matrixStats_1.3.0                 Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             globals_0.16.3             \n",
       " [15] lattice_0.22-6              MASS_7.3-60                \n",
       " [17] magrittr_2.0.3              plotly_4.10.4              \n",
       " [19] plotrix_3.8-4               yaml_2.3.8                 \n",
       " [21] httpuv_1.6.15               sctransform_0.4.1          \n",
       " [23] spam_2.10-0                 spatstat.sparse_3.0-3      \n",
       " [25] reticulate_1.36.1           pbapply_1.7-2              \n",
       " [27] DBI_1.2.2                   CNEr_1.38.0                \n",
       " [29] RColorBrewer_1.1-3          abind_1.4-5                \n",
       " [31] zlibbioc_1.48.0             Rtsne_0.17                 \n",
       " [33] purrr_1.0.2                 R.utils_2.12.3             \n",
       " [35] RCurl_1.98-1.14             pracma_2.4.4               \n",
       " [37] GenomeInfoDbData_1.2.11     ggrepel_0.9.5              \n",
       " [39] irlba_2.3.5.1               spatstat.utils_3.0-4       \n",
       " [41] listenv_0.9.1               seqLogo_1.68.0             \n",
       " [43] goftest_1.2-3               RSpectra_0.16-1            \n",
       " [45] spatstat.random_3.2-3       annotate_1.80.0            \n",
       " [47] fitdistrplus_1.1-11         parallelly_1.37.1          \n",
       " [49] leiden_0.4.3.1              codetools_0.2-20           \n",
       " [51] DelayedArray_0.26.7         DT_0.33                    \n",
       " [53] tidyselect_1.2.1            spatstat.explore_3.2-6     \n",
       " [55] base64enc_0.1-3             GenomicAlignments_1.36.0   \n",
       " [57] jsonlite_1.8.8              progressr_0.14.0           \n",
       " [59] iterators_1.0.14            ggridges_0.5.6             \n",
       " [61] survival_3.6-4              tools_4.3.3                \n",
       " [63] TFMPvalue_0.0.9             ica_1.0-3                  \n",
       " [65] Rcpp_1.0.12                 glue_1.7.0                 \n",
       " [67] gridExtra_2.3               IRdisplay_1.1              \n",
       " [69] withr_3.0.0                 fastmap_1.1.1              \n",
       " [71] fansi_1.0.6                 caTools_1.18.2             \n",
       " [73] digest_0.6.35               R6_2.5.1                   \n",
       " [75] mime_0.12                   nabor_0.5.0                \n",
       " [77] seqPattern_1.34.0           colorspace_2.1-0           \n",
       " [79] scattermore_1.2             GO.db_3.18.0               \n",
       " [81] tensor_1.5                  gtools_3.9.5               \n",
       " [83] poweRlaw_0.80.0             spatstat.data_3.0-4        \n",
       " [85] RSQLite_2.3.6               R.methodsS3_1.8.2          \n",
       " [87] utf8_1.2.4                  tidyr_1.3.1                \n",
       " [89] generics_0.1.3              httr_1.4.7                 \n",
       " [91] htmlwidgets_1.6.4           S4Arrays_1.2.1             \n",
       " [93] TFBSTools_1.40.0            uwot_0.1.16                \n",
       " [95] pkgconfig_2.0.3             gtable_0.3.5               \n",
       " [97] blob_1.2.4                  impute_1.76.0              \n",
       " [99] lmtest_0.9-40               htmltools_0.5.8.1          \n",
       "[101] dotCall64_1.1-1             scales_1.3.0               \n",
       "[103] png_0.1-8                   tzdb_0.4.0                 \n",
       "[105] reshape2_1.4.4              rjson_0.2.21               \n",
       "[107] uuid_1.2-0                  nlme_3.1-164               \n",
       "[109] repr_1.1.7                  cachem_1.0.8               \n",
       "[111] zoo_1.8-12                  stringr_1.5.1              \n",
       "[113] KernSmooth_2.23-22          miniUI_0.1.1.1             \n",
       "[115] AnnotationDbi_1.64.1        restfulr_0.0.15            \n",
       "[117] pillar_1.9.0                vctrs_0.6.5                \n",
       "[119] RANN_2.6.1                  promises_1.3.0             \n",
       "[121] xtable_1.8-4                cluster_2.1.6              \n",
       "[123] evaluate_0.23               readr_2.1.5                \n",
       "[125] cli_3.6.2                   compiler_4.3.3             \n",
       "[127] Rsamtools_2.18.0            rlang_1.1.3                \n",
       "[129] crayon_1.5.2                future.apply_1.11.2        \n",
       "[131] plyr_1.8.9                  stringi_1.8.3              \n",
       "[133] gridBase_0.4-7              deldir_2.0-4               \n",
       "[135] viridisLite_0.4.2           BiocParallel_1.36.0        \n",
       "[137] munsell_0.5.1               lazyeval_0.2.2             \n",
       "[139] spatstat.geom_3.2-9         IRkernel_1.3.2             \n",
       "[141] RcppHNSW_0.6.0              hms_1.1.3                  \n",
       "[143] patchwork_1.2.0             bit64_4.0.5                \n",
       "[145] future_1.33.2               KEGGREST_1.40.1            \n",
       "[147] shiny_1.8.1.1               ROCR_1.0-11                \n",
       "[149] igraph_2.0.3                memoise_2.0.1              \n",
       "[151] bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
