{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/4.merge_peaks_from_22_cell_types.240523/pairs.rds\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/4.merge_peaks_from_22_cell_types.240523/matrix.atac.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/4.merge_peaks_from_22_cell_types.240523/matrix.rna.rds\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/FigR/BMMC22.merge.240525/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readRDS(path.pairs.E2G)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  4000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  4001  to  5000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  5001  to  6000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  6001  to  7000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  7001  to  8000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  8001  to  9000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  9001  to  10000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  10001  to  11000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  11001  to  12000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  12001  to  13000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  13001  to  14000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  14001  to  15000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  15001  to  16000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  16001  to  17000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  17001  to  18000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  18001  to  19000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  19001  to  20000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  20001  to  21000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  21001  to  22000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  22001  to  23000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  23001  to  24000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  24001  to  25000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  25001  to  26000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  26001  to  27000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  27001  to  28000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  28001  to  29000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  29001  to  30000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  30001  to  31000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  31001  to  32000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  32001  to  33000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  33001  to  34000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  34001  to  35000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  35001  to  36000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  36001  to  37000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  37001  to  38000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  38001  to  39000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  39001  to  40000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  40001  to  41000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  41001  to  42000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  42001  to  43000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  43001  to  44000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  44001  to  45000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  45001  to  46000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  46001  to  47000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  47001  to  48000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  48001  to  49000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  49001  to  50000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  50001  to  51000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  51001  to  52000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  52001  to  53000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  53001  to  54000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  54001  to  55000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  55001  to  56000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  56001  to  57000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  57001  to  58000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  58001  to  59000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  59001  to  60000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  60001  to  61000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  61001  to  62000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  62001  to  63000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  63001  to  64000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  64001  to  65000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  65001  to  66000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  66001  to  67000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  67001  to  68000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  68001  to  69000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  69001  to  69249 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = readRDS(path.matrix.rna_count)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter = pairs.E2G[pairs.E2G$TargetGene %in% rownames(matrix.rna) &\n",
    "                              pairs.E2G$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 540429 69249 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(540429): chr1-10010-10536 chr1-96496-96699 ...\n",
       "  chrX-156029831-156030380 chrX-156030497-156030869\n",
       "rowData names(2): PeakName bias\n",
       "colnames(69249): s1d1_TAGTTGTCACCCTCAC-1 s1d1_CTATGGCCATAACGGG-1 ...\n",
       "  s4d9_ACACTTGCAACTAGAA-1 s4d9_CACTTAAAGTCTGGGC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna),\n",
    "                       index = 1:nrow(matrix.rna))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(dir.output, \"chr\", sep = \"/\"), recursive = T):\n",
      "“'/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/FigR/BMMC22.merge.240525//chr' already exists”\n"
     ]
    }
   ],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr2\"\n",
      "Running in parallel using  16 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.65777132908503 mins \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter[seqnames(pairs.E2G.filter) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 9320865 ranges and 7 metadata columns:\n",
       "       seqnames              ranges strand |  TargetGene               PeakName\n",
       "          <Rle>           <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 100004103-100004301      * |         AGL chr1-100004103-10000..\n",
       "  chr1     chr1 100004103-100004301      * |      CDC14A chr1-100004103-10000..\n",
       "  chr1     chr1 100004103-100004301      * |         DBT chr1-100004103-10000..\n",
       "  chr1     chr1 100004103-100004301      * |       EXTL2 chr1-100004103-10000..\n",
       "  chr1     chr1 100004103-100004301      * |       FRRS1 chr1-100004103-10000..\n",
       "   ...      ...                 ...    ... .         ...                    ...\n",
       "  chrX     chrX   99980528-99980848      * |      PCDH19 chrX-99980528-99980848\n",
       "  chrX     chrX   99980528-99980848      * |      TSPAN6 chrX-99980528-99980848\n",
       "  chrX     chrX   99980528-99980848      * |        XKRX chrX-99980528-99980848\n",
       "  chrX     chrX     9999312-9999592      * |       TBL1X   chrX-9999312-9999592\n",
       "  chrX     chrX     9999312-9999592      * |        WWC3   chrX-9999312-9999592\n",
       "                     PairName         rObs      rBgSD     rBgMean     pvalZ\n",
       "                  <character>    <numeric>  <numeric>   <numeric> <numeric>\n",
       "  chr1 chr1-100004103-10000..  9.35867e-03 0.00423669  0.00293859 0.0648415\n",
       "  chr1 chr1-100004103-10000..  4.23202e-03 0.00416996  0.00148872 0.2553094\n",
       "  chr1 chr1-100004103-10000.. -4.15938e-03 0.00374209  0.00230425 0.9579411\n",
       "  chr1 chr1-100004103-10000..  6.89448e-05 0.00472927  0.00314326 0.7421738\n",
       "  chr1 chr1-100004103-10000.. -1.80730e-03 0.00590399  0.00479550 0.8682940\n",
       "   ...                    ...          ...        ...         ...       ...\n",
       "  chrX chrX-99980528-999808.. -0.000576719 0.00783336 0.002033605  0.630520\n",
       "  chrX chrX-99980528-999808.. -0.000684152 0.00510080 0.000612720  0.600348\n",
       "  chrX chrX-99980528-999808.. -0.000898400 0.00585269 0.000986661  0.626306\n",
       "  chrX chrX-9999312-9999592..  0.006520943 0.00554531 0.006581973  0.504391\n",
       "  chrX chrX-9999312-9999592..  0.005181724 0.00408558 0.003438679  0.334823\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 9320865 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>AGL     </td><td>K562</td><td> 9.358673e-03</td><td>0.004236689</td><td>0.0029385937</td><td>0.06484154</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>CDC14A  </td><td>K562</td><td> 4.232024e-03</td><td>0.004169957</td><td>0.0014887156</td><td>0.25530939</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>DBT     </td><td>K562</td><td>-4.159382e-03</td><td>0.003742090</td><td>0.0023042473</td><td>0.95794114</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>EXTL2   </td><td>K562</td><td> 6.894481e-05</td><td>0.004729265</td><td>0.0031432588</td><td>0.74217381</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>FRRS1   </td><td>K562</td><td>-1.807305e-03</td><td>0.005903987</td><td>0.0047954950</td><td>0.86829396</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>MFSD14A </td><td>K562</td><td> 3.830694e-03</td><td>0.004692147</td><td>0.0011929008</td><td>0.28699946</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>PLPPR5  </td><td>K562</td><td> 9.581224e-03</td><td>0.004553465</td><td>0.0003852251</td><td>0.02171450</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>RTCA    </td><td>K562</td><td> 2.302253e-03</td><td>0.004360907</td><td>0.0027574381</td><td>0.54156554</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>RTCA-AS1</td><td>K562</td><td> 1.512708e-03</td><td>0.004689124</td><td>0.0020641671</td><td>0.54680921</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>SASS6   </td><td>K562</td><td> 9.859255e-03</td><td>0.006346381</td><td>0.0047106148</td><td>0.20860478</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>SLC30A7 </td><td>K562</td><td> 6.497182e-03</td><td>0.004161483</td><td>0.0041820439</td><td>0.28899422</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>SLC35A3 </td><td>K562</td><td>-1.958088e-03</td><td>0.004848018</td><td>0.0028152210</td><td>0.83758719</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>TRMT13  </td><td>K562</td><td>-2.337990e-03</td><td>0.004295456</td><td>0.0025148374</td><td>0.87071096</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>AGL     </td><td>K562</td><td> 6.134480e-03</td><td>0.004764311</td><td>0.0052318741</td><td>0.42486944</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>CDC14A  </td><td>K562</td><td> 5.161858e-03</td><td>0.004296600</td><td>0.0048119070</td><td>0.46754269</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>DBT     </td><td>K562</td><td> 6.448491e-03</td><td>0.004697196</td><td>0.0050944364</td><td>0.38657052</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>EXTL2   </td><td>K562</td><td> 7.033145e-03</td><td>0.005819578</td><td>0.0060990211</td><td>0.43623809</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>FRRS1   </td><td>K562</td><td> 5.847128e-03</td><td>0.006973861</td><td>0.0099751473</td><td>0.72305039</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>GPR88   </td><td>K562</td><td>-7.898643e-04</td><td>0.004958619</td><td>0.0005951681</td><td>0.60999967</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>MFSD14A </td><td>K562</td><td>-1.921321e-03</td><td>0.004463099</td><td>0.0027844781</td><td>0.85414534</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>RTCA    </td><td>K562</td><td> 9.807983e-04</td><td>0.004749742</td><td>0.0074081651</td><td>0.91200466</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>RTCA-AS1</td><td>K562</td><td> 6.974910e-03</td><td>0.004860645</td><td>0.0031090458</td><td>0.21320795</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>SASS6   </td><td>K562</td><td> 2.937326e-04</td><td>0.006978947</td><td>0.0089728086</td><td>0.89317811</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>SLC30A7 </td><td>K562</td><td> 8.314463e-03</td><td>0.004132772</td><td>0.0075431466</td><td>0.42597378</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>SLC35A3 </td><td>K562</td><td> 7.389209e-03</td><td>0.004956981</td><td>0.0047840954</td><td>0.29960236</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>TRMT13  </td><td>K562</td><td> 1.996004e-03</td><td>0.004858661</td><td>0.0057760949</td><td>0.78171869</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001440</td><td> 10001830</td><td>CENPS   </td><td>K562</td><td> 1.289310e-03</td><td>0.005077602</td><td>0.0025163930</td><td>0.59548042</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001440</td><td> 10001830</td><td>CLSTN1  </td><td>K562</td><td> 2.475942e-03</td><td>0.004153976</td><td>0.0028507939</td><td>0.53595148</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001440</td><td> 10001830</td><td>CTNNBIP1</td><td>K562</td><td> 9.529089e-03</td><td>0.004746680</td><td>0.0050551713</td><td>0.17295910</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001440</td><td> 10001830</td><td>DFFA    </td><td>K562</td><td> 1.752186e-03</td><td>0.004712637</td><td>0.0039890025</td><td>0.68247897</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9992130</td><td> 9992304</td><td>WWC3     </td><td>K562</td><td> 0.0157748075</td><td>0.004661213</td><td> 0.0028659841</td><td>0.0028078725</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>AKAP17A  </td><td>K562</td><td> 0.0045777497</td><td>0.004724912</td><td> 0.0079597499</td><td>0.7629365895</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>ASMT     </td><td>K562</td><td> 0.0064767438</td><td>0.004279951</td><td> 0.0019755971</td><td>0.1464727768</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>ASMTL    </td><td>K562</td><td> 0.0068472914</td><td>0.005702023</td><td> 0.0107407103</td><td>0.7526376802</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>CSF2RA   </td><td>K562</td><td>-0.0047577648</td><td>0.017351725</td><td> 0.0229882058</td><td>0.9450932941</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>GTPBP6   </td><td>K562</td><td> 0.0042248817</td><td>0.005181977</td><td> 0.0097327699</td><td>0.8560847947</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>IL3RA    </td><td>K562</td><td> 0.0057608553</td><td>0.010742025</td><td> 0.0140587070</td><td>0.7800808373</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>LINC00106</td><td>K562</td><td> 0.0024121951</td><td>0.006179358</td><td> 0.0082509920</td><td>0.8276418070</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>P2RY8    </td><td>K562</td><td> 0.0014519845</td><td>0.007000607</td><td> 0.0017178756</td><td>0.5151486438</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>PLCXD1   </td><td>K562</td><td> 0.0019468732</td><td>0.007382400</td><td> 0.0124266609</td><td>0.9221326570</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>PPP2R3B  </td><td>K562</td><td> 0.0221433285</td><td>0.004550415</td><td> 0.0059378629</td><td>0.0001845004</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>SHOX     </td><td>K562</td><td>-0.0020159514</td><td>0.005312047</td><td> 0.0018151653</td><td>0.7646107048</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>SLC25A6  </td><td>K562</td><td>-0.0003095650</td><td>0.005695148</td><td> 0.0064220735</td><td>0.8813962226</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>CLCN4    </td><td>K562</td><td> 0.0057503822</td><td>0.032003243</td><td> 0.0120131761</td><td>0.5775745631</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>FAM9B    </td><td>K562</td><td> 0.0029513821</td><td>0.005828314</td><td> 0.0036285268</td><td>0.5462458166</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>GPR143   </td><td>K562</td><td>-0.0034432095</td><td>0.004139223</td><td>-0.0001507983</td><td>0.7868147201</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>SHROOM2  </td><td>K562</td><td> 0.0024026854</td><td>0.004128080</td><td> 0.0027475437</td><td>0.5332887597</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>TBL1X    </td><td>K562</td><td> 0.0087609107</td><td>0.034661704</td><td> 0.0267738479</td><td>0.6983561119</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>WWC3     </td><td>K562</td><td> 0.0138036816</td><td>0.008285236</td><td> 0.0087839226</td><td>0.2723012134</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>CLCN4    </td><td>K562</td><td>-0.0042395499</td><td>0.017054537</td><td> 0.0114908744</td><td>0.8218296124</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>GPR143   </td><td>K562</td><td>-0.0003710198</td><td>0.004202489</td><td> 0.0005607982</td><td>0.5877379695</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>SHROOM2  </td><td>K562</td><td> 0.0021265315</td><td>0.004548762</td><td> 0.0019635492</td><td>0.4857089464</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>TBL1X    </td><td>K562</td><td> 0.0243735568</td><td>0.015059284</td><td> 0.0159309980</td><td>0.2875277933</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>WWC3     </td><td>K562</td><td> 0.0062108511</td><td>0.005488895</td><td> 0.0071986289</td><td>0.5714077491</td></tr>\n",
       "\t<tr><td>chrX</td><td>99980528</td><td>99980848</td><td>CSTF2    </td><td>K562</td><td> 0.0020477071</td><td>0.004930169</td><td> 0.0018141611</td><td>0.4811088551</td></tr>\n",
       "\t<tr><td>chrX</td><td>99980528</td><td>99980848</td><td>PCDH19   </td><td>K562</td><td>-0.0005767188</td><td>0.007833359</td><td> 0.0020336049</td><td>0.6305203186</td></tr>\n",
       "\t<tr><td>chrX</td><td>99980528</td><td>99980848</td><td>TSPAN6   </td><td>K562</td><td>-0.0006841517</td><td>0.005100800</td><td> 0.0006127197</td><td>0.6003482458</td></tr>\n",
       "\t<tr><td>chrX</td><td>99980528</td><td>99980848</td><td>XKRX     </td><td>K562</td><td>-0.0008983998</td><td>0.005852686</td><td> 0.0009866613</td><td>0.6263057708</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9999312</td><td> 9999592</td><td>TBL1X    </td><td>K562</td><td> 0.0065209431</td><td>0.005545312</td><td> 0.0065819725</td><td>0.5043905101</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9999312</td><td> 9999592</td><td>WWC3     </td><td>K562</td><td> 0.0051817244</td><td>0.004085582</td><td> 0.0034386788</td><td>0.3348231885</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 9320865 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100004103 & 100004301 & AGL      & K562 &  9.358673e-03 & 0.004236689 & 0.0029385937 & 0.06484154\\\\\n",
       "\t chr1 & 100004103 & 100004301 & CDC14A   & K562 &  4.232024e-03 & 0.004169957 & 0.0014887156 & 0.25530939\\\\\n",
       "\t chr1 & 100004103 & 100004301 & DBT      & K562 & -4.159382e-03 & 0.003742090 & 0.0023042473 & 0.95794114\\\\\n",
       "\t chr1 & 100004103 & 100004301 & EXTL2    & K562 &  6.894481e-05 & 0.004729265 & 0.0031432588 & 0.74217381\\\\\n",
       "\t chr1 & 100004103 & 100004301 & FRRS1    & K562 & -1.807305e-03 & 0.005903987 & 0.0047954950 & 0.86829396\\\\\n",
       "\t chr1 & 100004103 & 100004301 & MFSD14A  & K562 &  3.830694e-03 & 0.004692147 & 0.0011929008 & 0.28699946\\\\\n",
       "\t chr1 & 100004103 & 100004301 & PLPPR5   & K562 &  9.581224e-03 & 0.004553465 & 0.0003852251 & 0.02171450\\\\\n",
       "\t chr1 & 100004103 & 100004301 & RTCA     & K562 &  2.302253e-03 & 0.004360907 & 0.0027574381 & 0.54156554\\\\\n",
       "\t chr1 & 100004103 & 100004301 & RTCA-AS1 & K562 &  1.512708e-03 & 0.004689124 & 0.0020641671 & 0.54680921\\\\\n",
       "\t chr1 & 100004103 & 100004301 & SASS6    & K562 &  9.859255e-03 & 0.006346381 & 0.0047106148 & 0.20860478\\\\\n",
       "\t chr1 & 100004103 & 100004301 & SLC30A7  & K562 &  6.497182e-03 & 0.004161483 & 0.0041820439 & 0.28899422\\\\\n",
       "\t chr1 & 100004103 & 100004301 & SLC35A3  & K562 & -1.958088e-03 & 0.004848018 & 0.0028152210 & 0.83758719\\\\\n",
       "\t chr1 & 100004103 & 100004301 & TRMT13   & K562 & -2.337990e-03 & 0.004295456 & 0.0025148374 & 0.87071096\\\\\n",
       "\t chr1 & 100010532 & 100010875 & AGL      & K562 &  6.134480e-03 & 0.004764311 & 0.0052318741 & 0.42486944\\\\\n",
       "\t chr1 & 100010532 & 100010875 & CDC14A   & K562 &  5.161858e-03 & 0.004296600 & 0.0048119070 & 0.46754269\\\\\n",
       "\t chr1 & 100010532 & 100010875 & DBT      & K562 &  6.448491e-03 & 0.004697196 & 0.0050944364 & 0.38657052\\\\\n",
       "\t chr1 & 100010532 & 100010875 & EXTL2    & K562 &  7.033145e-03 & 0.005819578 & 0.0060990211 & 0.43623809\\\\\n",
       "\t chr1 & 100010532 & 100010875 & FRRS1    & K562 &  5.847128e-03 & 0.006973861 & 0.0099751473 & 0.72305039\\\\\n",
       "\t chr1 & 100010532 & 100010875 & GPR88    & K562 & -7.898643e-04 & 0.004958619 & 0.0005951681 & 0.60999967\\\\\n",
       "\t chr1 & 100010532 & 100010875 & MFSD14A  & K562 & -1.921321e-03 & 0.004463099 & 0.0027844781 & 0.85414534\\\\\n",
       "\t chr1 & 100010532 & 100010875 & RTCA     & K562 &  9.807983e-04 & 0.004749742 & 0.0074081651 & 0.91200466\\\\\n",
       "\t chr1 & 100010532 & 100010875 & RTCA-AS1 & K562 &  6.974910e-03 & 0.004860645 & 0.0031090458 & 0.21320795\\\\\n",
       "\t chr1 & 100010532 & 100010875 & SASS6    & K562 &  2.937326e-04 & 0.006978947 & 0.0089728086 & 0.89317811\\\\\n",
       "\t chr1 & 100010532 & 100010875 & SLC30A7  & K562 &  8.314463e-03 & 0.004132772 & 0.0075431466 & 0.42597378\\\\\n",
       "\t chr1 & 100010532 & 100010875 & SLC35A3  & K562 &  7.389209e-03 & 0.004956981 & 0.0047840954 & 0.29960236\\\\\n",
       "\t chr1 & 100010532 & 100010875 & TRMT13   & K562 &  1.996004e-03 & 0.004858661 & 0.0057760949 & 0.78171869\\\\\n",
       "\t chr1 &  10001440 &  10001830 & CENPS    & K562 &  1.289310e-03 & 0.005077602 & 0.0025163930 & 0.59548042\\\\\n",
       "\t chr1 &  10001440 &  10001830 & CLSTN1   & K562 &  2.475942e-03 & 0.004153976 & 0.0028507939 & 0.53595148\\\\\n",
       "\t chr1 &  10001440 &  10001830 & CTNNBIP1 & K562 &  9.529089e-03 & 0.004746680 & 0.0050551713 & 0.17295910\\\\\n",
       "\t chr1 &  10001440 &  10001830 & DFFA     & K562 &  1.752186e-03 & 0.004712637 & 0.0039890025 & 0.68247897\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  9992130 &  9992304 & WWC3      & K562 &  0.0157748075 & 0.004661213 &  0.0028659841 & 0.0028078725\\\\\n",
       "\t chrX &   999313 &  1000203 & AKAP17A   & K562 &  0.0045777497 & 0.004724912 &  0.0079597499 & 0.7629365895\\\\\n",
       "\t chrX &   999313 &  1000203 & ASMT      & K562 &  0.0064767438 & 0.004279951 &  0.0019755971 & 0.1464727768\\\\\n",
       "\t chrX &   999313 &  1000203 & ASMTL     & K562 &  0.0068472914 & 0.005702023 &  0.0107407103 & 0.7526376802\\\\\n",
       "\t chrX &   999313 &  1000203 & CSF2RA    & K562 & -0.0047577648 & 0.017351725 &  0.0229882058 & 0.9450932941\\\\\n",
       "\t chrX &   999313 &  1000203 & GTPBP6    & K562 &  0.0042248817 & 0.005181977 &  0.0097327699 & 0.8560847947\\\\\n",
       "\t chrX &   999313 &  1000203 & IL3RA     & K562 &  0.0057608553 & 0.010742025 &  0.0140587070 & 0.7800808373\\\\\n",
       "\t chrX &   999313 &  1000203 & LINC00106 & K562 &  0.0024121951 & 0.006179358 &  0.0082509920 & 0.8276418070\\\\\n",
       "\t chrX &   999313 &  1000203 & P2RY8     & K562 &  0.0014519845 & 0.007000607 &  0.0017178756 & 0.5151486438\\\\\n",
       "\t chrX &   999313 &  1000203 & PLCXD1    & K562 &  0.0019468732 & 0.007382400 &  0.0124266609 & 0.9221326570\\\\\n",
       "\t chrX &   999313 &  1000203 & PPP2R3B   & K562 &  0.0221433285 & 0.004550415 &  0.0059378629 & 0.0001845004\\\\\n",
       "\t chrX &   999313 &  1000203 & SHOX      & K562 & -0.0020159514 & 0.005312047 &  0.0018151653 & 0.7646107048\\\\\n",
       "\t chrX &   999313 &  1000203 & SLC25A6   & K562 & -0.0003095650 & 0.005695148 &  0.0064220735 & 0.8813962226\\\\\n",
       "\t chrX &  9995566 &  9996680 & CLCN4     & K562 &  0.0057503822 & 0.032003243 &  0.0120131761 & 0.5775745631\\\\\n",
       "\t chrX &  9995566 &  9996680 & FAM9B     & K562 &  0.0029513821 & 0.005828314 &  0.0036285268 & 0.5462458166\\\\\n",
       "\t chrX &  9995566 &  9996680 & GPR143    & K562 & -0.0034432095 & 0.004139223 & -0.0001507983 & 0.7868147201\\\\\n",
       "\t chrX &  9995566 &  9996680 & SHROOM2   & K562 &  0.0024026854 & 0.004128080 &  0.0027475437 & 0.5332887597\\\\\n",
       "\t chrX &  9995566 &  9996680 & TBL1X     & K562 &  0.0087609107 & 0.034661704 &  0.0267738479 & 0.6983561119\\\\\n",
       "\t chrX &  9995566 &  9996680 & WWC3      & K562 &  0.0138036816 & 0.008285236 &  0.0087839226 & 0.2723012134\\\\\n",
       "\t chrX &  9996996 &  9998157 & CLCN4     & K562 & -0.0042395499 & 0.017054537 &  0.0114908744 & 0.8218296124\\\\\n",
       "\t chrX &  9996996 &  9998157 & GPR143    & K562 & -0.0003710198 & 0.004202489 &  0.0005607982 & 0.5877379695\\\\\n",
       "\t chrX &  9996996 &  9998157 & SHROOM2   & K562 &  0.0021265315 & 0.004548762 &  0.0019635492 & 0.4857089464\\\\\n",
       "\t chrX &  9996996 &  9998157 & TBL1X     & K562 &  0.0243735568 & 0.015059284 &  0.0159309980 & 0.2875277933\\\\\n",
       "\t chrX &  9996996 &  9998157 & WWC3      & K562 &  0.0062108511 & 0.005488895 &  0.0071986289 & 0.5714077491\\\\\n",
       "\t chrX & 99980528 & 99980848 & CSTF2     & K562 &  0.0020477071 & 0.004930169 &  0.0018141611 & 0.4811088551\\\\\n",
       "\t chrX & 99980528 & 99980848 & PCDH19    & K562 & -0.0005767188 & 0.007833359 &  0.0020336049 & 0.6305203186\\\\\n",
       "\t chrX & 99980528 & 99980848 & TSPAN6    & K562 & -0.0006841517 & 0.005100800 &  0.0006127197 & 0.6003482458\\\\\n",
       "\t chrX & 99980528 & 99980848 & XKRX      & K562 & -0.0008983998 & 0.005852686 &  0.0009866613 & 0.6263057708\\\\\n",
       "\t chrX &  9999312 &  9999592 & TBL1X     & K562 &  0.0065209431 & 0.005545312 &  0.0065819725 & 0.5043905101\\\\\n",
       "\t chrX &  9999312 &  9999592 & WWC3      & K562 &  0.0051817244 & 0.004085582 &  0.0034386788 & 0.3348231885\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 9320865 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100004103 | 100004301 | AGL      | K562 |  9.358673e-03 | 0.004236689 | 0.0029385937 | 0.06484154 |\n",
       "| chr1 | 100004103 | 100004301 | CDC14A   | K562 |  4.232024e-03 | 0.004169957 | 0.0014887156 | 0.25530939 |\n",
       "| chr1 | 100004103 | 100004301 | DBT      | K562 | -4.159382e-03 | 0.003742090 | 0.0023042473 | 0.95794114 |\n",
       "| chr1 | 100004103 | 100004301 | EXTL2    | K562 |  6.894481e-05 | 0.004729265 | 0.0031432588 | 0.74217381 |\n",
       "| chr1 | 100004103 | 100004301 | FRRS1    | K562 | -1.807305e-03 | 0.005903987 | 0.0047954950 | 0.86829396 |\n",
       "| chr1 | 100004103 | 100004301 | MFSD14A  | K562 |  3.830694e-03 | 0.004692147 | 0.0011929008 | 0.28699946 |\n",
       "| chr1 | 100004103 | 100004301 | PLPPR5   | K562 |  9.581224e-03 | 0.004553465 | 0.0003852251 | 0.02171450 |\n",
       "| chr1 | 100004103 | 100004301 | RTCA     | K562 |  2.302253e-03 | 0.004360907 | 0.0027574381 | 0.54156554 |\n",
       "| chr1 | 100004103 | 100004301 | RTCA-AS1 | K562 |  1.512708e-03 | 0.004689124 | 0.0020641671 | 0.54680921 |\n",
       "| chr1 | 100004103 | 100004301 | SASS6    | K562 |  9.859255e-03 | 0.006346381 | 0.0047106148 | 0.20860478 |\n",
       "| chr1 | 100004103 | 100004301 | SLC30A7  | K562 |  6.497182e-03 | 0.004161483 | 0.0041820439 | 0.28899422 |\n",
       "| chr1 | 100004103 | 100004301 | SLC35A3  | K562 | -1.958088e-03 | 0.004848018 | 0.0028152210 | 0.83758719 |\n",
       "| chr1 | 100004103 | 100004301 | TRMT13   | K562 | -2.337990e-03 | 0.004295456 | 0.0025148374 | 0.87071096 |\n",
       "| chr1 | 100010532 | 100010875 | AGL      | K562 |  6.134480e-03 | 0.004764311 | 0.0052318741 | 0.42486944 |\n",
       "| chr1 | 100010532 | 100010875 | CDC14A   | K562 |  5.161858e-03 | 0.004296600 | 0.0048119070 | 0.46754269 |\n",
       "| chr1 | 100010532 | 100010875 | DBT      | K562 |  6.448491e-03 | 0.004697196 | 0.0050944364 | 0.38657052 |\n",
       "| chr1 | 100010532 | 100010875 | EXTL2    | K562 |  7.033145e-03 | 0.005819578 | 0.0060990211 | 0.43623809 |\n",
       "| chr1 | 100010532 | 100010875 | FRRS1    | K562 |  5.847128e-03 | 0.006973861 | 0.0099751473 | 0.72305039 |\n",
       "| chr1 | 100010532 | 100010875 | GPR88    | K562 | -7.898643e-04 | 0.004958619 | 0.0005951681 | 0.60999967 |\n",
       "| chr1 | 100010532 | 100010875 | MFSD14A  | K562 | -1.921321e-03 | 0.004463099 | 0.0027844781 | 0.85414534 |\n",
       "| chr1 | 100010532 | 100010875 | RTCA     | K562 |  9.807983e-04 | 0.004749742 | 0.0074081651 | 0.91200466 |\n",
       "| chr1 | 100010532 | 100010875 | RTCA-AS1 | K562 |  6.974910e-03 | 0.004860645 | 0.0031090458 | 0.21320795 |\n",
       "| chr1 | 100010532 | 100010875 | SASS6    | K562 |  2.937326e-04 | 0.006978947 | 0.0089728086 | 0.89317811 |\n",
       "| chr1 | 100010532 | 100010875 | SLC30A7  | K562 |  8.314463e-03 | 0.004132772 | 0.0075431466 | 0.42597378 |\n",
       "| chr1 | 100010532 | 100010875 | SLC35A3  | K562 |  7.389209e-03 | 0.004956981 | 0.0047840954 | 0.29960236 |\n",
       "| chr1 | 100010532 | 100010875 | TRMT13   | K562 |  1.996004e-03 | 0.004858661 | 0.0057760949 | 0.78171869 |\n",
       "| chr1 |  10001440 |  10001830 | CENPS    | K562 |  1.289310e-03 | 0.005077602 | 0.0025163930 | 0.59548042 |\n",
       "| chr1 |  10001440 |  10001830 | CLSTN1   | K562 |  2.475942e-03 | 0.004153976 | 0.0028507939 | 0.53595148 |\n",
       "| chr1 |  10001440 |  10001830 | CTNNBIP1 | K562 |  9.529089e-03 | 0.004746680 | 0.0050551713 | 0.17295910 |\n",
       "| chr1 |  10001440 |  10001830 | DFFA     | K562 |  1.752186e-03 | 0.004712637 | 0.0039890025 | 0.68247897 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  9992130 |  9992304 | WWC3      | K562 |  0.0157748075 | 0.004661213 |  0.0028659841 | 0.0028078725 |\n",
       "| chrX |   999313 |  1000203 | AKAP17A   | K562 |  0.0045777497 | 0.004724912 |  0.0079597499 | 0.7629365895 |\n",
       "| chrX |   999313 |  1000203 | ASMT      | K562 |  0.0064767438 | 0.004279951 |  0.0019755971 | 0.1464727768 |\n",
       "| chrX |   999313 |  1000203 | ASMTL     | K562 |  0.0068472914 | 0.005702023 |  0.0107407103 | 0.7526376802 |\n",
       "| chrX |   999313 |  1000203 | CSF2RA    | K562 | -0.0047577648 | 0.017351725 |  0.0229882058 | 0.9450932941 |\n",
       "| chrX |   999313 |  1000203 | GTPBP6    | K562 |  0.0042248817 | 0.005181977 |  0.0097327699 | 0.8560847947 |\n",
       "| chrX |   999313 |  1000203 | IL3RA     | K562 |  0.0057608553 | 0.010742025 |  0.0140587070 | 0.7800808373 |\n",
       "| chrX |   999313 |  1000203 | LINC00106 | K562 |  0.0024121951 | 0.006179358 |  0.0082509920 | 0.8276418070 |\n",
       "| chrX |   999313 |  1000203 | P2RY8     | K562 |  0.0014519845 | 0.007000607 |  0.0017178756 | 0.5151486438 |\n",
       "| chrX |   999313 |  1000203 | PLCXD1    | K562 |  0.0019468732 | 0.007382400 |  0.0124266609 | 0.9221326570 |\n",
       "| chrX |   999313 |  1000203 | PPP2R3B   | K562 |  0.0221433285 | 0.004550415 |  0.0059378629 | 0.0001845004 |\n",
       "| chrX |   999313 |  1000203 | SHOX      | K562 | -0.0020159514 | 0.005312047 |  0.0018151653 | 0.7646107048 |\n",
       "| chrX |   999313 |  1000203 | SLC25A6   | K562 | -0.0003095650 | 0.005695148 |  0.0064220735 | 0.8813962226 |\n",
       "| chrX |  9995566 |  9996680 | CLCN4     | K562 |  0.0057503822 | 0.032003243 |  0.0120131761 | 0.5775745631 |\n",
       "| chrX |  9995566 |  9996680 | FAM9B     | K562 |  0.0029513821 | 0.005828314 |  0.0036285268 | 0.5462458166 |\n",
       "| chrX |  9995566 |  9996680 | GPR143    | K562 | -0.0034432095 | 0.004139223 | -0.0001507983 | 0.7868147201 |\n",
       "| chrX |  9995566 |  9996680 | SHROOM2   | K562 |  0.0024026854 | 0.004128080 |  0.0027475437 | 0.5332887597 |\n",
       "| chrX |  9995566 |  9996680 | TBL1X     | K562 |  0.0087609107 | 0.034661704 |  0.0267738479 | 0.6983561119 |\n",
       "| chrX |  9995566 |  9996680 | WWC3      | K562 |  0.0138036816 | 0.008285236 |  0.0087839226 | 0.2723012134 |\n",
       "| chrX |  9996996 |  9998157 | CLCN4     | K562 | -0.0042395499 | 0.017054537 |  0.0114908744 | 0.8218296124 |\n",
       "| chrX |  9996996 |  9998157 | GPR143    | K562 | -0.0003710198 | 0.004202489 |  0.0005607982 | 0.5877379695 |\n",
       "| chrX |  9996996 |  9998157 | SHROOM2   | K562 |  0.0021265315 | 0.004548762 |  0.0019635492 | 0.4857089464 |\n",
       "| chrX |  9996996 |  9998157 | TBL1X     | K562 |  0.0243735568 | 0.015059284 |  0.0159309980 | 0.2875277933 |\n",
       "| chrX |  9996996 |  9998157 | WWC3      | K562 |  0.0062108511 | 0.005488895 |  0.0071986289 | 0.5714077491 |\n",
       "| chrX | 99980528 | 99980848 | CSTF2     | K562 |  0.0020477071 | 0.004930169 |  0.0018141611 | 0.4811088551 |\n",
       "| chrX | 99980528 | 99980848 | PCDH19    | K562 | -0.0005767188 | 0.007833359 |  0.0020336049 | 0.6305203186 |\n",
       "| chrX | 99980528 | 99980848 | TSPAN6    | K562 | -0.0006841517 | 0.005100800 |  0.0006127197 | 0.6003482458 |\n",
       "| chrX | 99980528 | 99980848 | XKRX      | K562 | -0.0008983998 | 0.005852686 |  0.0009866613 | 0.6263057708 |\n",
       "| chrX |  9999312 |  9999592 | TBL1X     | K562 |  0.0065209431 | 0.005545312 |  0.0065819725 | 0.5043905101 |\n",
       "| chrX |  9999312 |  9999592 | WWC3      | K562 |  0.0051817244 | 0.004085582 |  0.0034386788 | 0.3348231885 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType rObs          rBgSD      \n",
       "1       chr1 100004103 100004301 AGL        K562      9.358673e-03 0.004236689\n",
       "2       chr1 100004103 100004301 CDC14A     K562      4.232024e-03 0.004169957\n",
       "3       chr1 100004103 100004301 DBT        K562     -4.159382e-03 0.003742090\n",
       "4       chr1 100004103 100004301 EXTL2      K562      6.894481e-05 0.004729265\n",
       "5       chr1 100004103 100004301 FRRS1      K562     -1.807305e-03 0.005903987\n",
       "6       chr1 100004103 100004301 MFSD14A    K562      3.830694e-03 0.004692147\n",
       "7       chr1 100004103 100004301 PLPPR5     K562      9.581224e-03 0.004553465\n",
       "8       chr1 100004103 100004301 RTCA       K562      2.302253e-03 0.004360907\n",
       "9       chr1 100004103 100004301 RTCA-AS1   K562      1.512708e-03 0.004689124\n",
       "10      chr1 100004103 100004301 SASS6      K562      9.859255e-03 0.006346381\n",
       "11      chr1 100004103 100004301 SLC30A7    K562      6.497182e-03 0.004161483\n",
       "12      chr1 100004103 100004301 SLC35A3    K562     -1.958088e-03 0.004848018\n",
       "13      chr1 100004103 100004301 TRMT13     K562     -2.337990e-03 0.004295456\n",
       "14      chr1 100010532 100010875 AGL        K562      6.134480e-03 0.004764311\n",
       "15      chr1 100010532 100010875 CDC14A     K562      5.161858e-03 0.004296600\n",
       "16      chr1 100010532 100010875 DBT        K562      6.448491e-03 0.004697196\n",
       "17      chr1 100010532 100010875 EXTL2      K562      7.033145e-03 0.005819578\n",
       "18      chr1 100010532 100010875 FRRS1      K562      5.847128e-03 0.006973861\n",
       "19      chr1 100010532 100010875 GPR88      K562     -7.898643e-04 0.004958619\n",
       "20      chr1 100010532 100010875 MFSD14A    K562     -1.921321e-03 0.004463099\n",
       "21      chr1 100010532 100010875 RTCA       K562      9.807983e-04 0.004749742\n",
       "22      chr1 100010532 100010875 RTCA-AS1   K562      6.974910e-03 0.004860645\n",
       "23      chr1 100010532 100010875 SASS6      K562      2.937326e-04 0.006978947\n",
       "24      chr1 100010532 100010875 SLC30A7    K562      8.314463e-03 0.004132772\n",
       "25      chr1 100010532 100010875 SLC35A3    K562      7.389209e-03 0.004956981\n",
       "26      chr1 100010532 100010875 TRMT13     K562      1.996004e-03 0.004858661\n",
       "27      chr1  10001440  10001830 CENPS      K562      1.289310e-03 0.005077602\n",
       "28      chr1  10001440  10001830 CLSTN1     K562      2.475942e-03 0.004153976\n",
       "29      chr1  10001440  10001830 CTNNBIP1   K562      9.529089e-03 0.004746680\n",
       "30      chr1  10001440  10001830 DFFA       K562      1.752186e-03 0.004712637\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮          \n",
       "9320836 chrX  9992130   9992304  WWC3       K562      0.0157748075 0.004661213\n",
       "9320837 chrX   999313   1000203  AKAP17A    K562      0.0045777497 0.004724912\n",
       "9320838 chrX   999313   1000203  ASMT       K562      0.0064767438 0.004279951\n",
       "9320839 chrX   999313   1000203  ASMTL      K562      0.0068472914 0.005702023\n",
       "9320840 chrX   999313   1000203  CSF2RA     K562     -0.0047577648 0.017351725\n",
       "9320841 chrX   999313   1000203  GTPBP6     K562      0.0042248817 0.005181977\n",
       "9320842 chrX   999313   1000203  IL3RA      K562      0.0057608553 0.010742025\n",
       "9320843 chrX   999313   1000203  LINC00106  K562      0.0024121951 0.006179358\n",
       "9320844 chrX   999313   1000203  P2RY8      K562      0.0014519845 0.007000607\n",
       "9320845 chrX   999313   1000203  PLCXD1     K562      0.0019468732 0.007382400\n",
       "9320846 chrX   999313   1000203  PPP2R3B    K562      0.0221433285 0.004550415\n",
       "9320847 chrX   999313   1000203  SHOX       K562     -0.0020159514 0.005312047\n",
       "9320848 chrX   999313   1000203  SLC25A6    K562     -0.0003095650 0.005695148\n",
       "9320849 chrX  9995566   9996680  CLCN4      K562      0.0057503822 0.032003243\n",
       "9320850 chrX  9995566   9996680  FAM9B      K562      0.0029513821 0.005828314\n",
       "9320851 chrX  9995566   9996680  GPR143     K562     -0.0034432095 0.004139223\n",
       "9320852 chrX  9995566   9996680  SHROOM2    K562      0.0024026854 0.004128080\n",
       "9320853 chrX  9995566   9996680  TBL1X      K562      0.0087609107 0.034661704\n",
       "9320854 chrX  9995566   9996680  WWC3       K562      0.0138036816 0.008285236\n",
       "9320855 chrX  9996996   9998157  CLCN4      K562     -0.0042395499 0.017054537\n",
       "9320856 chrX  9996996   9998157  GPR143     K562     -0.0003710198 0.004202489\n",
       "9320857 chrX  9996996   9998157  SHROOM2    K562      0.0021265315 0.004548762\n",
       "9320858 chrX  9996996   9998157  TBL1X      K562      0.0243735568 0.015059284\n",
       "9320859 chrX  9996996   9998157  WWC3       K562      0.0062108511 0.005488895\n",
       "9320860 chrX 99980528  99980848  CSTF2      K562      0.0020477071 0.004930169\n",
       "9320861 chrX 99980528  99980848  PCDH19     K562     -0.0005767188 0.007833359\n",
       "9320862 chrX 99980528  99980848  TSPAN6     K562     -0.0006841517 0.005100800\n",
       "9320863 chrX 99980528  99980848  XKRX       K562     -0.0008983998 0.005852686\n",
       "9320864 chrX  9999312   9999592  TBL1X      K562      0.0065209431 0.005545312\n",
       "9320865 chrX  9999312   9999592  WWC3       K562      0.0051817244 0.004085582\n",
       "        rBgMean       pvalZ       \n",
       "1       0.0029385937  0.06484154  \n",
       "2       0.0014887156  0.25530939  \n",
       "3       0.0023042473  0.95794114  \n",
       "4       0.0031432588  0.74217381  \n",
       "5       0.0047954950  0.86829396  \n",
       "6       0.0011929008  0.28699946  \n",
       "7       0.0003852251  0.02171450  \n",
       "8       0.0027574381  0.54156554  \n",
       "9       0.0020641671  0.54680921  \n",
       "10      0.0047106148  0.20860478  \n",
       "11      0.0041820439  0.28899422  \n",
       "12      0.0028152210  0.83758719  \n",
       "13      0.0025148374  0.87071096  \n",
       "14      0.0052318741  0.42486944  \n",
       "15      0.0048119070  0.46754269  \n",
       "16      0.0050944364  0.38657052  \n",
       "17      0.0060990211  0.43623809  \n",
       "18      0.0099751473  0.72305039  \n",
       "19      0.0005951681  0.60999967  \n",
       "20      0.0027844781  0.85414534  \n",
       "21      0.0074081651  0.91200466  \n",
       "22      0.0031090458  0.21320795  \n",
       "23      0.0089728086  0.89317811  \n",
       "24      0.0075431466  0.42597378  \n",
       "25      0.0047840954  0.29960236  \n",
       "26      0.0057760949  0.78171869  \n",
       "27      0.0025163930  0.59548042  \n",
       "28      0.0028507939  0.53595148  \n",
       "29      0.0050551713  0.17295910  \n",
       "30      0.0039890025  0.68247897  \n",
       "⋮       ⋮             ⋮           \n",
       "9320836  0.0028659841 0.0028078725\n",
       "9320837  0.0079597499 0.7629365895\n",
       "9320838  0.0019755971 0.1464727768\n",
       "9320839  0.0107407103 0.7526376802\n",
       "9320840  0.0229882058 0.9450932941\n",
       "9320841  0.0097327699 0.8560847947\n",
       "9320842  0.0140587070 0.7800808373\n",
       "9320843  0.0082509920 0.8276418070\n",
       "9320844  0.0017178756 0.5151486438\n",
       "9320845  0.0124266609 0.9221326570\n",
       "9320846  0.0059378629 0.0001845004\n",
       "9320847  0.0018151653 0.7646107048\n",
       "9320848  0.0064220735 0.8813962226\n",
       "9320849  0.0120131761 0.5775745631\n",
       "9320850  0.0036285268 0.5462458166\n",
       "9320851 -0.0001507983 0.7868147201\n",
       "9320852  0.0027475437 0.5332887597\n",
       "9320853  0.0267738479 0.6983561119\n",
       "9320854  0.0087839226 0.2723012134\n",
       "9320855  0.0114908744 0.8218296124\n",
       "9320856  0.0005607982 0.5877379695\n",
       "9320857  0.0019635492 0.4857089464\n",
       "9320858  0.0159309980 0.2875277933\n",
       "9320859  0.0071986289 0.5714077491\n",
       "9320860  0.0018141611 0.4811088551\n",
       "9320861  0.0020336049 0.6305203186\n",
       "9320862  0.0006127197 0.6003482458\n",
       "9320863  0.0009866613 0.6263057708\n",
       "9320864  0.0065819725 0.5043905101\n",
       "9320865  0.0034386788 0.3348231885"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.9 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   foreach_1.5.2                    \n",
       " [3] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.68.0                  \n",
       " [5] rtracklayer_1.60.1                Biostrings_2.70.1                \n",
       " [7] XVector_0.42.0                    genomation_1.34.0                \n",
       " [9] Seurat_5.0.3                      SeuratObject_5.0.1               \n",
       "[11] sp_2.1-4                          FigR_0.1.0                       \n",
       "[13] motifmatchr_1.24.0                chromVAR_1.24.0                  \n",
       "[15] cowplot_1.1.3                     ggplot2_3.5.1                    \n",
       "[17] dplyr_1.1.4                       SummarizedExperiment_1.30.2      \n",
       "[19] Biobase_2.60.0                    GenomicRanges_1.54.1             \n",
       "[21] GenomeInfoDb_1.38.1               IRanges_2.36.0                   \n",
       "[23] S4Vectors_0.40.2                  BiocGenerics_0.48.1              \n",
       "[25] MatrixGenerics_1.12.3             matrixStats_1.3.0                \n",
       "[27] Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             globals_0.16.3             \n",
       " [15] lattice_0.22-6              MASS_7.3-60                \n",
       " [17] magrittr_2.0.3              plotly_4.10.4              \n",
       " [19] plotrix_3.8-4               yaml_2.3.8                 \n",
       " [21] httpuv_1.6.15               sctransform_0.4.1          \n",
       " [23] spam_2.10-0                 spatstat.sparse_3.0-3      \n",
       " [25] reticulate_1.36.1           pbapply_1.7-2              \n",
       " [27] DBI_1.2.2                   CNEr_1.38.0                \n",
       " [29] RColorBrewer_1.1-3          abind_1.4-5                \n",
       " [31] zlibbioc_1.48.0             Rtsne_0.17                 \n",
       " [33] purrr_1.0.2                 R.utils_2.12.3             \n",
       " [35] RCurl_1.98-1.14             pracma_2.4.4               \n",
       " [37] GenomeInfoDbData_1.2.11     ggrepel_0.9.5              \n",
       " [39] irlba_2.3.5.1               spatstat.utils_3.0-4       \n",
       " [41] listenv_0.9.1               seqLogo_1.68.0             \n",
       " [43] goftest_1.2-3               RSpectra_0.16-1            \n",
       " [45] spatstat.random_3.2-3       annotate_1.80.0            \n",
       " [47] fitdistrplus_1.1-11         parallelly_1.37.1          \n",
       " [49] leiden_0.4.3.1              codetools_0.2-20           \n",
       " [51] DelayedArray_0.26.7         DT_0.33                    \n",
       " [53] tidyselect_1.2.1            spatstat.explore_3.2-6     \n",
       " [55] base64enc_0.1-3             GenomicAlignments_1.36.0   \n",
       " [57] jsonlite_1.8.8              progressr_0.14.0           \n",
       " [59] iterators_1.0.14            ggridges_0.5.6             \n",
       " [61] survival_3.6-4              tools_4.3.3                \n",
       " [63] TFMPvalue_0.0.9             ica_1.0-3                  \n",
       " [65] Rcpp_1.0.12                 glue_1.7.0                 \n",
       " [67] gridExtra_2.3               IRdisplay_1.1              \n",
       " [69] withr_3.0.0                 fastmap_1.1.1              \n",
       " [71] fansi_1.0.6                 caTools_1.18.2             \n",
       " [73] digest_0.6.35               R6_2.5.1                   \n",
       " [75] mime_0.12                   nabor_0.5.0                \n",
       " [77] seqPattern_1.34.0           colorspace_2.1-0           \n",
       " [79] scattermore_1.2             GO.db_3.18.0               \n",
       " [81] tensor_1.5                  gtools_3.9.5               \n",
       " [83] poweRlaw_0.80.0             spatstat.data_3.0-4        \n",
       " [85] RSQLite_2.3.6               R.methodsS3_1.8.2          \n",
       " [87] utf8_1.2.4                  tidyr_1.3.1                \n",
       " [89] generics_0.1.3              data.table_1.15.2          \n",
       " [91] httr_1.4.7                  htmlwidgets_1.6.4          \n",
       " [93] S4Arrays_1.2.1              TFBSTools_1.40.0           \n",
       " [95] uwot_0.1.16                 pkgconfig_2.0.3            \n",
       " [97] gtable_0.3.5                blob_1.2.4                 \n",
       " [99] impute_1.76.0               lmtest_0.9-40              \n",
       "[101] htmltools_0.5.8.1           dotCall64_1.1-1            \n",
       "[103] scales_1.3.0                png_0.1-8                  \n",
       "[105] tzdb_0.4.0                  reshape2_1.4.4             \n",
       "[107] rjson_0.2.21                uuid_1.2-0                 \n",
       "[109] nlme_3.1-164                repr_1.1.7                 \n",
       "[111] cachem_1.0.8                zoo_1.8-12                 \n",
       "[113] stringr_1.5.1               KernSmooth_2.23-22         \n",
       "[115] miniUI_0.1.1.1              AnnotationDbi_1.64.1       \n",
       "[117] restfulr_0.0.15             pillar_1.9.0               \n",
       "[119] vctrs_0.6.5                 RANN_2.6.1                 \n",
       "[121] promises_1.3.0              xtable_1.8-4               \n",
       "[123] cluster_2.1.6               evaluate_0.23              \n",
       "[125] readr_2.1.5                 cli_3.6.2                  \n",
       "[127] compiler_4.3.3              Rsamtools_2.18.0           \n",
       "[129] rlang_1.1.3                 crayon_1.5.2               \n",
       "[131] future.apply_1.11.2         plyr_1.8.9                 \n",
       "[133] stringi_1.8.3               gridBase_0.4-7             \n",
       "[135] deldir_2.0-4                viridisLite_0.4.2          \n",
       "[137] BiocParallel_1.36.0         munsell_0.5.1              \n",
       "[139] lazyeval_0.2.2              spatstat.geom_3.2-9        \n",
       "[141] IRkernel_1.3.2              RcppHNSW_0.6.0             \n",
       "[143] hms_1.1.3                   patchwork_1.2.0            \n",
       "[145] bit64_4.0.5                 future_1.33.2              \n",
       "[147] KEGGREST_1.40.1             shiny_1.8.1.1              \n",
       "[149] ROCR_1.0-11                 igraph_2.0.3               \n",
       "[151] memoise_2.0.1               bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
