{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/cellranger_arc/refdata-cellranger-arc-GRCh38-2020-A-2.0.0/genes/genes.gtf.gz\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/WTC11_cluster5/WTC11_cluster5/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/WTC11_cluster5/WTC11_cluster5/multiome_powerlaw_v3/encode_e2g_predictions.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/WTC11_cluster5/WTC11_cluster5/Kendall/atac_matrix.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/scE2G_input_Maya/WTC11_HPC_lenient/cluster_5/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/WTC11/1.Genome_wide_prediction/FigR/FigR.250325/\"\n",
    "celltype = \"WTC11_cluster5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2227086 ranges and 3 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100034573-100034728      * |         AGL\n",
       "        [2]     chr1 100034573-100034728      * |      CDC14A\n",
       "        [3]     chr1 100034573-100034728      * |         DBT\n",
       "        [4]     chr1 100034573-100034728      * |        DPH5\n",
       "        [5]     chr1 100034573-100034728      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2227082]     chrX     9995697-9996195      * |       CLCN4\n",
       "  [2227083]     chrX     9995697-9996195      * |      GPR143\n",
       "  [2227084]     chrX     9995697-9996195      * |     SHROOM2\n",
       "  [2227085]     chrX     9995697-9996195      * |       TBL1X\n",
       "  [2227086]     chrX     9995697-9996195      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-100034573-10003.. chr1-100034573-10003..\n",
       "        [2] chr1-100034573-10003.. chr1-100034573-10003..\n",
       "        [3] chr1-100034573-10003.. chr1-100034573-10003..\n",
       "        [4] chr1-100034573-10003.. chr1-100034573-10003..\n",
       "        [5] chr1-100034573-10003.. chr1-100034573-10003..\n",
       "        ...                    ...                    ...\n",
       "  [2227082]   chrX-9995697-9996195 chrX-9995697-9996195..\n",
       "  [2227083]   chrX-9995697-9996195 chrX-9995697-9996195..\n",
       "  [2227084]   chrX-9995697-9996195 chrX-9995697-9996195..\n",
       "  [2227085]   chrX-9995697-9996195 chrX-9995697-9996195..\n",
       "  [2227086]   chrX-9995697-9996195 chrX-9995697-9996195..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  791 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dd80eda-2392-4d59-8250-1ae62d7e1d1e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b9e26c70-11fd-47c4-8e05-9c96b17d3f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a76b1603-6b00-4ac0-9399-fd341c5a7b88",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 152404 791 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(152404): chr1-100034573-100034728 chr1-100037552-100039045 ...\n",
       "  chrX-9981419-9981826 chrX-9995697-9996195\n",
       "rowData names(2): PeakName bias\n",
       "colnames(791): AAACAGCCAAACCCTA-1 AAACCAACAACATAAG-1 ...\n",
       "  TTTGTCTAGATAACCC-1 TTTGTGGCAGGCAAGC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna.rename),\n",
    "                       index = 1:nrow(matrix.rna.rename))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "Sys.setenv(OPENBLAS_NUM_THREADS=1)\n",
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  15.6034159660339 secs \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Time Elapsed:  14.4731779615084 mins \n",
      "\n",
      "[1] \"chr10\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  3.38750123977661 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  6.74984438816706 mins \n",
      "\n",
      "[1] \"chr11\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.32704401016235 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.6970133582751 mins \n",
      "\n",
      "[1] \"chr12\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.80144667625427 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.72554869651794 mins \n",
      "\n",
      "[1] \"chr13\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.83996415138245 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  4.44559415578842 mins \n",
      "\n",
      "[1] \"chr14\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  3.68578147888184 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  6.33455610672633 mins \n",
      "\n",
      "[1] \"chr15\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  3.27528357505798 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  6.56127539873123 mins \n",
      "\n",
      "[1] \"chr16\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.49521803855896 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.43577768007914 mins \n",
      "\n",
      "[1] \"chr17\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.38963794708252 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.548837852478 mins \n",
      "\n",
      "[1] \"chr18\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.55391550064087 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  4.19220817486445 mins \n",
      "\n",
      "[1] \"chr19\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  9.79938530921936 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  14.9941877683004 mins \n",
      "\n",
      "[1] \"chr2\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.07425475120544 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.82339389721553 mins \n",
      "\n",
      "[1] \"chr20\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  3.14512705802917 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  5.9522763967514 mins \n",
      "\n",
      "[1] \"chr21\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.62249541282654 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  3.69209731419881 mins \n",
      "\n",
      "[1] \"chr22\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.03002548217773 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  6.73887941837311 mins \n",
      "\n",
      "[1] \"chr3\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.49310612678528 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.35397816896439 mins \n",
      "\n",
      "[1] \"chr4\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.69790649414062 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  5.58737363020579 mins \n",
      "\n",
      "[1] \"chr5\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.572918176651 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  10.4329528808594 mins \n",
      "\n",
      "[1] \"chr6\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.7689197063446 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.1892667333285 mins \n",
      "\n",
      "[1] \"chr7\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.56503915786743 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.80001393159231 mins \n",
      "\n",
      "[1] \"chr8\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.3337619304657 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  7.14100793202718 mins \n",
      "\n",
      "[1] \"chr9\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.82439422607422 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.28703435659409 mins \n",
      "\n",
      "[1] \"chrX\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.88400101661682 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  6.64748710393906 mins \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna.rename,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2154053 ranges and 7 metadata columns:\n",
       "       seqnames              ranges strand |  TargetGene               PeakName\n",
       "          <Rle>           <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 100034573-100034728      * |         AGL chr1-100034573-10003..\n",
       "  chr1     chr1 100034573-100034728      * |      CDC14A chr1-100034573-10003..\n",
       "  chr1     chr1 100034573-100034728      * |         DBT chr1-100034573-10003..\n",
       "  chr1     chr1 100034573-100034728      * |        DPH5 chr1-100034573-10003..\n",
       "  chr1     chr1 100034573-100034728      * |       EXTL2 chr1-100034573-10003..\n",
       "   ...      ...                 ...    ... .         ...                    ...\n",
       "  chrX     chrX     9995697-9996195      * |       CLCN4   chrX-9995697-9996195\n",
       "  chrX     chrX     9995697-9996195      * |      GPR143   chrX-9995697-9996195\n",
       "  chrX     chrX     9995697-9996195      * |     SHROOM2   chrX-9995697-9996195\n",
       "  chrX     chrX     9995697-9996195      * |       TBL1X   chrX-9995697-9996195\n",
       "  chrX     chrX     9995697-9996195      * |        WWC3   chrX-9995697-9996195\n",
       "                     PairName         rObs     rBgSD      rBgMean     pvalZ\n",
       "                  <character>    <numeric> <numeric>    <numeric> <numeric>\n",
       "  chr1 chr1-100034573-10003..  0.000335479 0.0350077  0.002209057 0.5213408\n",
       "  chr1 chr1-100034573-10003.. -0.018814328 0.0375177 -0.001648926 0.6763542\n",
       "  chr1 chr1-100034573-10003..  0.047814905 0.0344872  0.001903482 0.0915518\n",
       "  chr1 chr1-100034573-10003.. -0.001607700 0.0343355  0.000778737 0.5277055\n",
       "  chr1 chr1-100034573-10003..  0.006819694 0.0351611  0.002194891 0.4476773\n",
       "   ...                    ...          ...       ...          ...       ...\n",
       "  chrX chrX-9995697-9996195..   0.00642881 0.0382631   0.00858879 0.5225087\n",
       "  chrX chrX-9995697-9996195..  -0.04878065 0.0359965  -0.00370152 0.8947732\n",
       "  chrX chrX-9995697-9996195..   0.04668553 0.0341919  -0.00399702 0.0691307\n",
       "  chrX chrX-9995697-9996195..  -0.00117275 0.0354678  -0.00399683 0.4682683\n",
       "  chrX chrX-9995697-9996195..   0.01935969 0.0393863   0.00608929 0.3680850\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2154053 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>AGL     </td><td>WTC11_cluster5</td><td> 0.0003354786</td><td>0.03500772</td><td> 0.0022090568</td><td>0.52134081</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>CDC14A  </td><td>WTC11_cluster5</td><td>-0.0188143279</td><td>0.03751772</td><td>-0.0016489257</td><td>0.67635418</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>DBT     </td><td>WTC11_cluster5</td><td> 0.0478149048</td><td>0.03448720</td><td> 0.0019034824</td><td>0.09155178</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>DPH5    </td><td>WTC11_cluster5</td><td>-0.0016077002</td><td>0.03433555</td><td> 0.0007787370</td><td>0.52770554</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>EXTL2   </td><td>WTC11_cluster5</td><td> 0.0068196943</td><td>0.03516111</td><td> 0.0021948907</td><td>0.44767732</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>MFSD14A </td><td>WTC11_cluster5</td><td>-0.0372838307</td><td>0.03580765</td><td> 0.0021248389</td><td>0.86445713</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>PLPPR4  </td><td>WTC11_cluster5</td><td>-0.0262150922</td><td>0.02994208</td><td>-0.0072910050</td><td>0.73631413</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>RTCA    </td><td>WTC11_cluster5</td><td>-0.0659947687</td><td>0.03795531</td><td> 0.0007213697</td><td>0.96060541</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>RTCA-AS1</td><td>WTC11_cluster5</td><td>-0.0296154188</td><td>0.03994222</td><td> 0.0072616889</td><td>0.82206451</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>SASS6   </td><td>WTC11_cluster5</td><td> 0.0039510270</td><td>0.03485896</td><td>-0.0001721060</td><td>0.45292273</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>SLC30A7 </td><td>WTC11_cluster5</td><td>-0.0183975910</td><td>0.03821186</td><td> 0.0053436148</td><td>0.73280037</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>SLC35A3 </td><td>WTC11_cluster5</td><td>-0.0171093812</td><td>0.03619236</td><td>-0.0014530316</td><td>0.66734264</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034573</td><td>100034728</td><td>TRMT13  </td><td>WTC11_cluster5</td><td> 0.0062569123</td><td>0.03136140</td><td> 0.0061813478</td><td>0.49903876</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>AGL     </td><td>WTC11_cluster5</td><td> 0.0436979107</td><td>0.03161747</td><td>-0.0057465276</td><td>0.05892840</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>CDC14A  </td><td>WTC11_cluster5</td><td>-0.0031480618</td><td>0.03425375</td><td> 0.0015830328</td><td>0.55492682</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>DBT     </td><td>WTC11_cluster5</td><td> 0.0437271229</td><td>0.03076109</td><td> 0.0084762993</td><td>0.12590689</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>DPH5    </td><td>WTC11_cluster5</td><td> 0.0283888473</td><td>0.03958892</td><td>-0.0045085480</td><td>0.20299390</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>EXTL2   </td><td>WTC11_cluster5</td><td> 0.0052479460</td><td>0.03377062</td><td>-0.0038545854</td><td>0.39375711</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>MFSD14A </td><td>WTC11_cluster5</td><td> 0.0267116943</td><td>0.03359450</td><td>-0.0043032403</td><td>0.17794768</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>PLPPR4  </td><td>WTC11_cluster5</td><td> 0.0139751739</td><td>0.03831282</td><td>-0.0006972310</td><td>0.35087349</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>RTCA    </td><td>WTC11_cluster5</td><td>-0.0268706996</td><td>0.03361914</td><td> 0.0069642818</td><td>0.84289323</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>RTCA-AS1</td><td>WTC11_cluster5</td><td> 0.0483917552</td><td>0.03669881</td><td> 0.0105400049</td><td>0.15117285</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>SASS6   </td><td>WTC11_cluster5</td><td> 0.0191934246</td><td>0.03705862</td><td> 0.0104966867</td><td>0.40723044</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>SLC30A7 </td><td>WTC11_cluster5</td><td> 0.0299745112</td><td>0.03147454</td><td> 0.0064535439</td><td>0.22744086</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>SLC35A3 </td><td>WTC11_cluster5</td><td> 0.0048069113</td><td>0.03262103</td><td> 0.0099584246</td><td>0.56274007</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037552</td><td>100039045</td><td>TRMT13  </td><td>WTC11_cluster5</td><td> 0.0385632499</td><td>0.03344443</td><td> 0.0070783637</td><td>0.17324765</td></tr>\n",
       "\t<tr><td>chr1</td><td>100049383</td><td>100049815</td><td>AGL     </td><td>WTC11_cluster5</td><td> 0.0096791735</td><td>0.03416016</td><td>-0.0056006958</td><td>0.32732890</td></tr>\n",
       "\t<tr><td>chr1</td><td>100049383</td><td>100049815</td><td>CDC14A  </td><td>WTC11_cluster5</td><td>-0.0383510517</td><td>0.03182288</td><td>-0.0100945757</td><td>0.81271067</td></tr>\n",
       "\t<tr><td>chr1</td><td>100049383</td><td>100049815</td><td>DBT     </td><td>WTC11_cluster5</td><td>-0.0075967855</td><td>0.03662336</td><td> 0.0031704657</td><td>0.61562086</td></tr>\n",
       "\t<tr><td>chr1</td><td>100049383</td><td>100049815</td><td>DPH5    </td><td>WTC11_cluster5</td><td>-0.0406631346</td><td>0.03366280</td><td>-0.0103589079</td><td>0.81600082</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9959363</td><td> 9959794</td><td>CLCN4  </td><td>WTC11_cluster5</td><td> 0.0200155624</td><td>0.03725265</td><td> 1.557573e-02</td><td>0.45256567</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9959363</td><td> 9959794</td><td>GPR143 </td><td>WTC11_cluster5</td><td>-0.0005338998</td><td>0.03752539</td><td>-1.929640e-03</td><td>0.48516493</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9959363</td><td> 9959794</td><td>SHROOM2</td><td>WTC11_cluster5</td><td>-0.0008966689</td><td>0.03155441</td><td>-7.379180e-03</td><td>0.41861453</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9959363</td><td> 9959794</td><td>TBL1X  </td><td>WTC11_cluster5</td><td> 0.0551660723</td><td>0.03608956</td><td> 3.680218e-03</td><td>0.07684569</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9959363</td><td> 9959794</td><td>WWC3   </td><td>WTC11_cluster5</td><td>-0.0586722240</td><td>0.03694833</td><td> 2.287367e-05</td><td>0.94392152</td></tr>\n",
       "\t<tr><td>chrX</td><td>99603193</td><td>99603678</td><td>PCDH19 </td><td>WTC11_cluster5</td><td> 0.0207685609</td><td>0.03024148</td><td>-7.058040e-03</td><td>0.17874803</td></tr>\n",
       "\t<tr><td>chrX</td><td>99722173</td><td>99722445</td><td>PCDH19 </td><td>WTC11_cluster5</td><td> 0.0039019582</td><td>0.03468627</td><td>-2.759235e-03</td><td>0.42385496</td></tr>\n",
       "\t<tr><td>chrX</td><td>99722173</td><td>99722445</td><td>TSPAN6 </td><td>WTC11_cluster5</td><td>-0.0245582011</td><td>0.03640750</td><td> 2.007551e-03</td><td>0.76720651</td></tr>\n",
       "\t<tr><td>chrX</td><td>  997278</td><td>  998378</td><td>AKAP17A</td><td>WTC11_cluster5</td><td>-0.0268017392</td><td>0.03523213</td><td> 3.026012e-03</td><td>0.80139279</td></tr>\n",
       "\t<tr><td>chrX</td><td>  997278</td><td>  998378</td><td>ASMTL  </td><td>WTC11_cluster5</td><td>-0.0355930402</td><td>0.03516824</td><td>-5.127752e-03</td><td>0.80682970</td></tr>\n",
       "\t<tr><td>chrX</td><td>  997278</td><td>  998378</td><td>PLCXD1 </td><td>WTC11_cluster5</td><td> 0.0147361250</td><td>0.03820177</td><td>-4.742011e-03</td><td>0.30506946</td></tr>\n",
       "\t<tr><td>chrX</td><td>  997278</td><td>  998378</td><td>PPP2R3B</td><td>WTC11_cluster5</td><td> 0.0016139973</td><td>0.03189453</td><td> 5.927069e-03</td><td>0.55378467</td></tr>\n",
       "\t<tr><td>chrX</td><td>  997278</td><td>  998378</td><td>SHOX   </td><td>WTC11_cluster5</td><td>-0.0133403901</td><td>0.01664838</td><td>-1.064262e-02</td><td>0.56436450</td></tr>\n",
       "\t<tr><td>chrX</td><td>  997278</td><td>  998378</td><td>SLC25A6</td><td>WTC11_cluster5</td><td>-0.0056443322</td><td>0.03458086</td><td> 2.742322e-03</td><td>0.59581254</td></tr>\n",
       "\t<tr><td>chrX</td><td>99736932</td><td>99737301</td><td>PCDH19 </td><td>WTC11_cluster5</td><td>-0.0152605944</td><td>0.03692273</td><td>-1.165598e-02</td><td>0.53888534</td></tr>\n",
       "\t<tr><td>chrX</td><td>99736932</td><td>99737301</td><td>SYTL4  </td><td>WTC11_cluster5</td><td>-0.0196086170</td><td>0.03077845</td><td>-1.196892e-03</td><td>0.72514737</td></tr>\n",
       "\t<tr><td>chrX</td><td>99736932</td><td>99737301</td><td>TSPAN6 </td><td>WTC11_cluster5</td><td>-0.0407624938</td><td>0.04694311</td><td> 7.929405e-03</td><td>0.85019111</td></tr>\n",
       "\t<tr><td>chrX</td><td>99777683</td><td>99778218</td><td>PCDH19 </td><td>WTC11_cluster5</td><td>-0.0096141240</td><td>0.03386363</td><td>-4.489820e-03</td><td>0.56013906</td></tr>\n",
       "\t<tr><td>chrX</td><td>99777683</td><td>99778218</td><td>SYTL4  </td><td>WTC11_cluster5</td><td> 0.0476469913</td><td>0.03064946</td><td>-2.746163e-03</td><td>0.05006977</td></tr>\n",
       "\t<tr><td>chrX</td><td>99777683</td><td>99778218</td><td>TSPAN6 </td><td>WTC11_cluster5</td><td> 0.0808838540</td><td>0.03281378</td><td> 1.411632e-03</td><td>0.00771945</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9981419</td><td> 9981826</td><td>CLCN4  </td><td>WTC11_cluster5</td><td>-0.0417834216</td><td>0.03579531</td><td> 2.417928e-03</td><td>0.89155427</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9981419</td><td> 9981826</td><td>GPR143 </td><td>WTC11_cluster5</td><td>-0.0630702662</td><td>0.03655568</td><td> 1.891174e-03</td><td>0.96222036</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9981419</td><td> 9981826</td><td>SHROOM2</td><td>WTC11_cluster5</td><td>-0.0676506335</td><td>0.03411475</td><td>-4.654688e-03</td><td>0.96759667</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9981419</td><td> 9981826</td><td>TBL1X  </td><td>WTC11_cluster5</td><td> 0.0540469303</td><td>0.03513857</td><td> 2.464299e-04</td><td>0.06287291</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9981419</td><td> 9981826</td><td>WWC3   </td><td>WTC11_cluster5</td><td>-0.0118586393</td><td>0.03354974</td><td>-1.789698e-03</td><td>0.61795710</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995697</td><td> 9996195</td><td>CLCN4  </td><td>WTC11_cluster5</td><td> 0.0064288064</td><td>0.03826308</td><td> 8.588788e-03</td><td>0.52250866</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995697</td><td> 9996195</td><td>GPR143 </td><td>WTC11_cluster5</td><td>-0.0487806513</td><td>0.03599652</td><td>-3.701521e-03</td><td>0.89477319</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995697</td><td> 9996195</td><td>SHROOM2</td><td>WTC11_cluster5</td><td> 0.0466855334</td><td>0.03419192</td><td>-3.997019e-03</td><td>0.06913074</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995697</td><td> 9996195</td><td>TBL1X  </td><td>WTC11_cluster5</td><td>-0.0011727521</td><td>0.03546782</td><td>-3.996833e-03</td><td>0.46826825</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995697</td><td> 9996195</td><td>WWC3   </td><td>WTC11_cluster5</td><td> 0.0193596879</td><td>0.03938626</td><td> 6.089293e-03</td><td>0.36808501</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2154053 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100034573 & 100034728 & AGL      & WTC11\\_cluster5 &  0.0003354786 & 0.03500772 &  0.0022090568 & 0.52134081\\\\\n",
       "\t chr1 & 100034573 & 100034728 & CDC14A   & WTC11\\_cluster5 & -0.0188143279 & 0.03751772 & -0.0016489257 & 0.67635418\\\\\n",
       "\t chr1 & 100034573 & 100034728 & DBT      & WTC11\\_cluster5 &  0.0478149048 & 0.03448720 &  0.0019034824 & 0.09155178\\\\\n",
       "\t chr1 & 100034573 & 100034728 & DPH5     & WTC11\\_cluster5 & -0.0016077002 & 0.03433555 &  0.0007787370 & 0.52770554\\\\\n",
       "\t chr1 & 100034573 & 100034728 & EXTL2    & WTC11\\_cluster5 &  0.0068196943 & 0.03516111 &  0.0021948907 & 0.44767732\\\\\n",
       "\t chr1 & 100034573 & 100034728 & MFSD14A  & WTC11\\_cluster5 & -0.0372838307 & 0.03580765 &  0.0021248389 & 0.86445713\\\\\n",
       "\t chr1 & 100034573 & 100034728 & PLPPR4   & WTC11\\_cluster5 & -0.0262150922 & 0.02994208 & -0.0072910050 & 0.73631413\\\\\n",
       "\t chr1 & 100034573 & 100034728 & RTCA     & WTC11\\_cluster5 & -0.0659947687 & 0.03795531 &  0.0007213697 & 0.96060541\\\\\n",
       "\t chr1 & 100034573 & 100034728 & RTCA-AS1 & WTC11\\_cluster5 & -0.0296154188 & 0.03994222 &  0.0072616889 & 0.82206451\\\\\n",
       "\t chr1 & 100034573 & 100034728 & SASS6    & WTC11\\_cluster5 &  0.0039510270 & 0.03485896 & -0.0001721060 & 0.45292273\\\\\n",
       "\t chr1 & 100034573 & 100034728 & SLC30A7  & WTC11\\_cluster5 & -0.0183975910 & 0.03821186 &  0.0053436148 & 0.73280037\\\\\n",
       "\t chr1 & 100034573 & 100034728 & SLC35A3  & WTC11\\_cluster5 & -0.0171093812 & 0.03619236 & -0.0014530316 & 0.66734264\\\\\n",
       "\t chr1 & 100034573 & 100034728 & TRMT13   & WTC11\\_cluster5 &  0.0062569123 & 0.03136140 &  0.0061813478 & 0.49903876\\\\\n",
       "\t chr1 & 100037552 & 100039045 & AGL      & WTC11\\_cluster5 &  0.0436979107 & 0.03161747 & -0.0057465276 & 0.05892840\\\\\n",
       "\t chr1 & 100037552 & 100039045 & CDC14A   & WTC11\\_cluster5 & -0.0031480618 & 0.03425375 &  0.0015830328 & 0.55492682\\\\\n",
       "\t chr1 & 100037552 & 100039045 & DBT      & WTC11\\_cluster5 &  0.0437271229 & 0.03076109 &  0.0084762993 & 0.12590689\\\\\n",
       "\t chr1 & 100037552 & 100039045 & DPH5     & WTC11\\_cluster5 &  0.0283888473 & 0.03958892 & -0.0045085480 & 0.20299390\\\\\n",
       "\t chr1 & 100037552 & 100039045 & EXTL2    & WTC11\\_cluster5 &  0.0052479460 & 0.03377062 & -0.0038545854 & 0.39375711\\\\\n",
       "\t chr1 & 100037552 & 100039045 & MFSD14A  & WTC11\\_cluster5 &  0.0267116943 & 0.03359450 & -0.0043032403 & 0.17794768\\\\\n",
       "\t chr1 & 100037552 & 100039045 & PLPPR4   & WTC11\\_cluster5 &  0.0139751739 & 0.03831282 & -0.0006972310 & 0.35087349\\\\\n",
       "\t chr1 & 100037552 & 100039045 & RTCA     & WTC11\\_cluster5 & -0.0268706996 & 0.03361914 &  0.0069642818 & 0.84289323\\\\\n",
       "\t chr1 & 100037552 & 100039045 & RTCA-AS1 & WTC11\\_cluster5 &  0.0483917552 & 0.03669881 &  0.0105400049 & 0.15117285\\\\\n",
       "\t chr1 & 100037552 & 100039045 & SASS6    & WTC11\\_cluster5 &  0.0191934246 & 0.03705862 &  0.0104966867 & 0.40723044\\\\\n",
       "\t chr1 & 100037552 & 100039045 & SLC30A7  & WTC11\\_cluster5 &  0.0299745112 & 0.03147454 &  0.0064535439 & 0.22744086\\\\\n",
       "\t chr1 & 100037552 & 100039045 & SLC35A3  & WTC11\\_cluster5 &  0.0048069113 & 0.03262103 &  0.0099584246 & 0.56274007\\\\\n",
       "\t chr1 & 100037552 & 100039045 & TRMT13   & WTC11\\_cluster5 &  0.0385632499 & 0.03344443 &  0.0070783637 & 0.17324765\\\\\n",
       "\t chr1 & 100049383 & 100049815 & AGL      & WTC11\\_cluster5 &  0.0096791735 & 0.03416016 & -0.0056006958 & 0.32732890\\\\\n",
       "\t chr1 & 100049383 & 100049815 & CDC14A   & WTC11\\_cluster5 & -0.0383510517 & 0.03182288 & -0.0100945757 & 0.81271067\\\\\n",
       "\t chr1 & 100049383 & 100049815 & DBT      & WTC11\\_cluster5 & -0.0075967855 & 0.03662336 &  0.0031704657 & 0.61562086\\\\\n",
       "\t chr1 & 100049383 & 100049815 & DPH5     & WTC11\\_cluster5 & -0.0406631346 & 0.03366280 & -0.0103589079 & 0.81600082\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  9959363 &  9959794 & CLCN4   & WTC11\\_cluster5 &  0.0200155624 & 0.03725265 &  1.557573e-02 & 0.45256567\\\\\n",
       "\t chrX &  9959363 &  9959794 & GPR143  & WTC11\\_cluster5 & -0.0005338998 & 0.03752539 & -1.929640e-03 & 0.48516493\\\\\n",
       "\t chrX &  9959363 &  9959794 & SHROOM2 & WTC11\\_cluster5 & -0.0008966689 & 0.03155441 & -7.379180e-03 & 0.41861453\\\\\n",
       "\t chrX &  9959363 &  9959794 & TBL1X   & WTC11\\_cluster5 &  0.0551660723 & 0.03608956 &  3.680218e-03 & 0.07684569\\\\\n",
       "\t chrX &  9959363 &  9959794 & WWC3    & WTC11\\_cluster5 & -0.0586722240 & 0.03694833 &  2.287367e-05 & 0.94392152\\\\\n",
       "\t chrX & 99603193 & 99603678 & PCDH19  & WTC11\\_cluster5 &  0.0207685609 & 0.03024148 & -7.058040e-03 & 0.17874803\\\\\n",
       "\t chrX & 99722173 & 99722445 & PCDH19  & WTC11\\_cluster5 &  0.0039019582 & 0.03468627 & -2.759235e-03 & 0.42385496\\\\\n",
       "\t chrX & 99722173 & 99722445 & TSPAN6  & WTC11\\_cluster5 & -0.0245582011 & 0.03640750 &  2.007551e-03 & 0.76720651\\\\\n",
       "\t chrX &   997278 &   998378 & AKAP17A & WTC11\\_cluster5 & -0.0268017392 & 0.03523213 &  3.026012e-03 & 0.80139279\\\\\n",
       "\t chrX &   997278 &   998378 & ASMTL   & WTC11\\_cluster5 & -0.0355930402 & 0.03516824 & -5.127752e-03 & 0.80682970\\\\\n",
       "\t chrX &   997278 &   998378 & PLCXD1  & WTC11\\_cluster5 &  0.0147361250 & 0.03820177 & -4.742011e-03 & 0.30506946\\\\\n",
       "\t chrX &   997278 &   998378 & PPP2R3B & WTC11\\_cluster5 &  0.0016139973 & 0.03189453 &  5.927069e-03 & 0.55378467\\\\\n",
       "\t chrX &   997278 &   998378 & SHOX    & WTC11\\_cluster5 & -0.0133403901 & 0.01664838 & -1.064262e-02 & 0.56436450\\\\\n",
       "\t chrX &   997278 &   998378 & SLC25A6 & WTC11\\_cluster5 & -0.0056443322 & 0.03458086 &  2.742322e-03 & 0.59581254\\\\\n",
       "\t chrX & 99736932 & 99737301 & PCDH19  & WTC11\\_cluster5 & -0.0152605944 & 0.03692273 & -1.165598e-02 & 0.53888534\\\\\n",
       "\t chrX & 99736932 & 99737301 & SYTL4   & WTC11\\_cluster5 & -0.0196086170 & 0.03077845 & -1.196892e-03 & 0.72514737\\\\\n",
       "\t chrX & 99736932 & 99737301 & TSPAN6  & WTC11\\_cluster5 & -0.0407624938 & 0.04694311 &  7.929405e-03 & 0.85019111\\\\\n",
       "\t chrX & 99777683 & 99778218 & PCDH19  & WTC11\\_cluster5 & -0.0096141240 & 0.03386363 & -4.489820e-03 & 0.56013906\\\\\n",
       "\t chrX & 99777683 & 99778218 & SYTL4   & WTC11\\_cluster5 &  0.0476469913 & 0.03064946 & -2.746163e-03 & 0.05006977\\\\\n",
       "\t chrX & 99777683 & 99778218 & TSPAN6  & WTC11\\_cluster5 &  0.0808838540 & 0.03281378 &  1.411632e-03 & 0.00771945\\\\\n",
       "\t chrX &  9981419 &  9981826 & CLCN4   & WTC11\\_cluster5 & -0.0417834216 & 0.03579531 &  2.417928e-03 & 0.89155427\\\\\n",
       "\t chrX &  9981419 &  9981826 & GPR143  & WTC11\\_cluster5 & -0.0630702662 & 0.03655568 &  1.891174e-03 & 0.96222036\\\\\n",
       "\t chrX &  9981419 &  9981826 & SHROOM2 & WTC11\\_cluster5 & -0.0676506335 & 0.03411475 & -4.654688e-03 & 0.96759667\\\\\n",
       "\t chrX &  9981419 &  9981826 & TBL1X   & WTC11\\_cluster5 &  0.0540469303 & 0.03513857 &  2.464299e-04 & 0.06287291\\\\\n",
       "\t chrX &  9981419 &  9981826 & WWC3    & WTC11\\_cluster5 & -0.0118586393 & 0.03354974 & -1.789698e-03 & 0.61795710\\\\\n",
       "\t chrX &  9995697 &  9996195 & CLCN4   & WTC11\\_cluster5 &  0.0064288064 & 0.03826308 &  8.588788e-03 & 0.52250866\\\\\n",
       "\t chrX &  9995697 &  9996195 & GPR143  & WTC11\\_cluster5 & -0.0487806513 & 0.03599652 & -3.701521e-03 & 0.89477319\\\\\n",
       "\t chrX &  9995697 &  9996195 & SHROOM2 & WTC11\\_cluster5 &  0.0466855334 & 0.03419192 & -3.997019e-03 & 0.06913074\\\\\n",
       "\t chrX &  9995697 &  9996195 & TBL1X   & WTC11\\_cluster5 & -0.0011727521 & 0.03546782 & -3.996833e-03 & 0.46826825\\\\\n",
       "\t chrX &  9995697 &  9996195 & WWC3    & WTC11\\_cluster5 &  0.0193596879 & 0.03938626 &  6.089293e-03 & 0.36808501\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2154053 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100034573 | 100034728 | AGL      | WTC11_cluster5 |  0.0003354786 | 0.03500772 |  0.0022090568 | 0.52134081 |\n",
       "| chr1 | 100034573 | 100034728 | CDC14A   | WTC11_cluster5 | -0.0188143279 | 0.03751772 | -0.0016489257 | 0.67635418 |\n",
       "| chr1 | 100034573 | 100034728 | DBT      | WTC11_cluster5 |  0.0478149048 | 0.03448720 |  0.0019034824 | 0.09155178 |\n",
       "| chr1 | 100034573 | 100034728 | DPH5     | WTC11_cluster5 | -0.0016077002 | 0.03433555 |  0.0007787370 | 0.52770554 |\n",
       "| chr1 | 100034573 | 100034728 | EXTL2    | WTC11_cluster5 |  0.0068196943 | 0.03516111 |  0.0021948907 | 0.44767732 |\n",
       "| chr1 | 100034573 | 100034728 | MFSD14A  | WTC11_cluster5 | -0.0372838307 | 0.03580765 |  0.0021248389 | 0.86445713 |\n",
       "| chr1 | 100034573 | 100034728 | PLPPR4   | WTC11_cluster5 | -0.0262150922 | 0.02994208 | -0.0072910050 | 0.73631413 |\n",
       "| chr1 | 100034573 | 100034728 | RTCA     | WTC11_cluster5 | -0.0659947687 | 0.03795531 |  0.0007213697 | 0.96060541 |\n",
       "| chr1 | 100034573 | 100034728 | RTCA-AS1 | WTC11_cluster5 | -0.0296154188 | 0.03994222 |  0.0072616889 | 0.82206451 |\n",
       "| chr1 | 100034573 | 100034728 | SASS6    | WTC11_cluster5 |  0.0039510270 | 0.03485896 | -0.0001721060 | 0.45292273 |\n",
       "| chr1 | 100034573 | 100034728 | SLC30A7  | WTC11_cluster5 | -0.0183975910 | 0.03821186 |  0.0053436148 | 0.73280037 |\n",
       "| chr1 | 100034573 | 100034728 | SLC35A3  | WTC11_cluster5 | -0.0171093812 | 0.03619236 | -0.0014530316 | 0.66734264 |\n",
       "| chr1 | 100034573 | 100034728 | TRMT13   | WTC11_cluster5 |  0.0062569123 | 0.03136140 |  0.0061813478 | 0.49903876 |\n",
       "| chr1 | 100037552 | 100039045 | AGL      | WTC11_cluster5 |  0.0436979107 | 0.03161747 | -0.0057465276 | 0.05892840 |\n",
       "| chr1 | 100037552 | 100039045 | CDC14A   | WTC11_cluster5 | -0.0031480618 | 0.03425375 |  0.0015830328 | 0.55492682 |\n",
       "| chr1 | 100037552 | 100039045 | DBT      | WTC11_cluster5 |  0.0437271229 | 0.03076109 |  0.0084762993 | 0.12590689 |\n",
       "| chr1 | 100037552 | 100039045 | DPH5     | WTC11_cluster5 |  0.0283888473 | 0.03958892 | -0.0045085480 | 0.20299390 |\n",
       "| chr1 | 100037552 | 100039045 | EXTL2    | WTC11_cluster5 |  0.0052479460 | 0.03377062 | -0.0038545854 | 0.39375711 |\n",
       "| chr1 | 100037552 | 100039045 | MFSD14A  | WTC11_cluster5 |  0.0267116943 | 0.03359450 | -0.0043032403 | 0.17794768 |\n",
       "| chr1 | 100037552 | 100039045 | PLPPR4   | WTC11_cluster5 |  0.0139751739 | 0.03831282 | -0.0006972310 | 0.35087349 |\n",
       "| chr1 | 100037552 | 100039045 | RTCA     | WTC11_cluster5 | -0.0268706996 | 0.03361914 |  0.0069642818 | 0.84289323 |\n",
       "| chr1 | 100037552 | 100039045 | RTCA-AS1 | WTC11_cluster5 |  0.0483917552 | 0.03669881 |  0.0105400049 | 0.15117285 |\n",
       "| chr1 | 100037552 | 100039045 | SASS6    | WTC11_cluster5 |  0.0191934246 | 0.03705862 |  0.0104966867 | 0.40723044 |\n",
       "| chr1 | 100037552 | 100039045 | SLC30A7  | WTC11_cluster5 |  0.0299745112 | 0.03147454 |  0.0064535439 | 0.22744086 |\n",
       "| chr1 | 100037552 | 100039045 | SLC35A3  | WTC11_cluster5 |  0.0048069113 | 0.03262103 |  0.0099584246 | 0.56274007 |\n",
       "| chr1 | 100037552 | 100039045 | TRMT13   | WTC11_cluster5 |  0.0385632499 | 0.03344443 |  0.0070783637 | 0.17324765 |\n",
       "| chr1 | 100049383 | 100049815 | AGL      | WTC11_cluster5 |  0.0096791735 | 0.03416016 | -0.0056006958 | 0.32732890 |\n",
       "| chr1 | 100049383 | 100049815 | CDC14A   | WTC11_cluster5 | -0.0383510517 | 0.03182288 | -0.0100945757 | 0.81271067 |\n",
       "| chr1 | 100049383 | 100049815 | DBT      | WTC11_cluster5 | -0.0075967855 | 0.03662336 |  0.0031704657 | 0.61562086 |\n",
       "| chr1 | 100049383 | 100049815 | DPH5     | WTC11_cluster5 | -0.0406631346 | 0.03366280 | -0.0103589079 | 0.81600082 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  9959363 |  9959794 | CLCN4   | WTC11_cluster5 |  0.0200155624 | 0.03725265 |  1.557573e-02 | 0.45256567 |\n",
       "| chrX |  9959363 |  9959794 | GPR143  | WTC11_cluster5 | -0.0005338998 | 0.03752539 | -1.929640e-03 | 0.48516493 |\n",
       "| chrX |  9959363 |  9959794 | SHROOM2 | WTC11_cluster5 | -0.0008966689 | 0.03155441 | -7.379180e-03 | 0.41861453 |\n",
       "| chrX |  9959363 |  9959794 | TBL1X   | WTC11_cluster5 |  0.0551660723 | 0.03608956 |  3.680218e-03 | 0.07684569 |\n",
       "| chrX |  9959363 |  9959794 | WWC3    | WTC11_cluster5 | -0.0586722240 | 0.03694833 |  2.287367e-05 | 0.94392152 |\n",
       "| chrX | 99603193 | 99603678 | PCDH19  | WTC11_cluster5 |  0.0207685609 | 0.03024148 | -7.058040e-03 | 0.17874803 |\n",
       "| chrX | 99722173 | 99722445 | PCDH19  | WTC11_cluster5 |  0.0039019582 | 0.03468627 | -2.759235e-03 | 0.42385496 |\n",
       "| chrX | 99722173 | 99722445 | TSPAN6  | WTC11_cluster5 | -0.0245582011 | 0.03640750 |  2.007551e-03 | 0.76720651 |\n",
       "| chrX |   997278 |   998378 | AKAP17A | WTC11_cluster5 | -0.0268017392 | 0.03523213 |  3.026012e-03 | 0.80139279 |\n",
       "| chrX |   997278 |   998378 | ASMTL   | WTC11_cluster5 | -0.0355930402 | 0.03516824 | -5.127752e-03 | 0.80682970 |\n",
       "| chrX |   997278 |   998378 | PLCXD1  | WTC11_cluster5 |  0.0147361250 | 0.03820177 | -4.742011e-03 | 0.30506946 |\n",
       "| chrX |   997278 |   998378 | PPP2R3B | WTC11_cluster5 |  0.0016139973 | 0.03189453 |  5.927069e-03 | 0.55378467 |\n",
       "| chrX |   997278 |   998378 | SHOX    | WTC11_cluster5 | -0.0133403901 | 0.01664838 | -1.064262e-02 | 0.56436450 |\n",
       "| chrX |   997278 |   998378 | SLC25A6 | WTC11_cluster5 | -0.0056443322 | 0.03458086 |  2.742322e-03 | 0.59581254 |\n",
       "| chrX | 99736932 | 99737301 | PCDH19  | WTC11_cluster5 | -0.0152605944 | 0.03692273 | -1.165598e-02 | 0.53888534 |\n",
       "| chrX | 99736932 | 99737301 | SYTL4   | WTC11_cluster5 | -0.0196086170 | 0.03077845 | -1.196892e-03 | 0.72514737 |\n",
       "| chrX | 99736932 | 99737301 | TSPAN6  | WTC11_cluster5 | -0.0407624938 | 0.04694311 |  7.929405e-03 | 0.85019111 |\n",
       "| chrX | 99777683 | 99778218 | PCDH19  | WTC11_cluster5 | -0.0096141240 | 0.03386363 | -4.489820e-03 | 0.56013906 |\n",
       "| chrX | 99777683 | 99778218 | SYTL4   | WTC11_cluster5 |  0.0476469913 | 0.03064946 | -2.746163e-03 | 0.05006977 |\n",
       "| chrX | 99777683 | 99778218 | TSPAN6  | WTC11_cluster5 |  0.0808838540 | 0.03281378 |  1.411632e-03 | 0.00771945 |\n",
       "| chrX |  9981419 |  9981826 | CLCN4   | WTC11_cluster5 | -0.0417834216 | 0.03579531 |  2.417928e-03 | 0.89155427 |\n",
       "| chrX |  9981419 |  9981826 | GPR143  | WTC11_cluster5 | -0.0630702662 | 0.03655568 |  1.891174e-03 | 0.96222036 |\n",
       "| chrX |  9981419 |  9981826 | SHROOM2 | WTC11_cluster5 | -0.0676506335 | 0.03411475 | -4.654688e-03 | 0.96759667 |\n",
       "| chrX |  9981419 |  9981826 | TBL1X   | WTC11_cluster5 |  0.0540469303 | 0.03513857 |  2.464299e-04 | 0.06287291 |\n",
       "| chrX |  9981419 |  9981826 | WWC3    | WTC11_cluster5 | -0.0118586393 | 0.03354974 | -1.789698e-03 | 0.61795710 |\n",
       "| chrX |  9995697 |  9996195 | CLCN4   | WTC11_cluster5 |  0.0064288064 | 0.03826308 |  8.588788e-03 | 0.52250866 |\n",
       "| chrX |  9995697 |  9996195 | GPR143  | WTC11_cluster5 | -0.0487806513 | 0.03599652 | -3.701521e-03 | 0.89477319 |\n",
       "| chrX |  9995697 |  9996195 | SHROOM2 | WTC11_cluster5 |  0.0466855334 | 0.03419192 | -3.997019e-03 | 0.06913074 |\n",
       "| chrX |  9995697 |  9996195 | TBL1X   | WTC11_cluster5 | -0.0011727521 | 0.03546782 | -3.996833e-03 | 0.46826825 |\n",
       "| chrX |  9995697 |  9996195 | WWC3    | WTC11_cluster5 |  0.0193596879 | 0.03938626 |  6.089293e-03 | 0.36808501 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType       rObs         \n",
       "1       chr1 100034573 100034728 AGL        WTC11_cluster5  0.0003354786\n",
       "2       chr1 100034573 100034728 CDC14A     WTC11_cluster5 -0.0188143279\n",
       "3       chr1 100034573 100034728 DBT        WTC11_cluster5  0.0478149048\n",
       "4       chr1 100034573 100034728 DPH5       WTC11_cluster5 -0.0016077002\n",
       "5       chr1 100034573 100034728 EXTL2      WTC11_cluster5  0.0068196943\n",
       "6       chr1 100034573 100034728 MFSD14A    WTC11_cluster5 -0.0372838307\n",
       "7       chr1 100034573 100034728 PLPPR4     WTC11_cluster5 -0.0262150922\n",
       "8       chr1 100034573 100034728 RTCA       WTC11_cluster5 -0.0659947687\n",
       "9       chr1 100034573 100034728 RTCA-AS1   WTC11_cluster5 -0.0296154188\n",
       "10      chr1 100034573 100034728 SASS6      WTC11_cluster5  0.0039510270\n",
       "11      chr1 100034573 100034728 SLC30A7    WTC11_cluster5 -0.0183975910\n",
       "12      chr1 100034573 100034728 SLC35A3    WTC11_cluster5 -0.0171093812\n",
       "13      chr1 100034573 100034728 TRMT13     WTC11_cluster5  0.0062569123\n",
       "14      chr1 100037552 100039045 AGL        WTC11_cluster5  0.0436979107\n",
       "15      chr1 100037552 100039045 CDC14A     WTC11_cluster5 -0.0031480618\n",
       "16      chr1 100037552 100039045 DBT        WTC11_cluster5  0.0437271229\n",
       "17      chr1 100037552 100039045 DPH5       WTC11_cluster5  0.0283888473\n",
       "18      chr1 100037552 100039045 EXTL2      WTC11_cluster5  0.0052479460\n",
       "19      chr1 100037552 100039045 MFSD14A    WTC11_cluster5  0.0267116943\n",
       "20      chr1 100037552 100039045 PLPPR4     WTC11_cluster5  0.0139751739\n",
       "21      chr1 100037552 100039045 RTCA       WTC11_cluster5 -0.0268706996\n",
       "22      chr1 100037552 100039045 RTCA-AS1   WTC11_cluster5  0.0483917552\n",
       "23      chr1 100037552 100039045 SASS6      WTC11_cluster5  0.0191934246\n",
       "24      chr1 100037552 100039045 SLC30A7    WTC11_cluster5  0.0299745112\n",
       "25      chr1 100037552 100039045 SLC35A3    WTC11_cluster5  0.0048069113\n",
       "26      chr1 100037552 100039045 TRMT13     WTC11_cluster5  0.0385632499\n",
       "27      chr1 100049383 100049815 AGL        WTC11_cluster5  0.0096791735\n",
       "28      chr1 100049383 100049815 CDC14A     WTC11_cluster5 -0.0383510517\n",
       "29      chr1 100049383 100049815 DBT        WTC11_cluster5 -0.0075967855\n",
       "30      chr1 100049383 100049815 DPH5       WTC11_cluster5 -0.0406631346\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮              ⋮            \n",
       "2154024 chrX  9959363   9959794  CLCN4      WTC11_cluster5  0.0200155624\n",
       "2154025 chrX  9959363   9959794  GPR143     WTC11_cluster5 -0.0005338998\n",
       "2154026 chrX  9959363   9959794  SHROOM2    WTC11_cluster5 -0.0008966689\n",
       "2154027 chrX  9959363   9959794  TBL1X      WTC11_cluster5  0.0551660723\n",
       "2154028 chrX  9959363   9959794  WWC3       WTC11_cluster5 -0.0586722240\n",
       "2154029 chrX 99603193  99603678  PCDH19     WTC11_cluster5  0.0207685609\n",
       "2154030 chrX 99722173  99722445  PCDH19     WTC11_cluster5  0.0039019582\n",
       "2154031 chrX 99722173  99722445  TSPAN6     WTC11_cluster5 -0.0245582011\n",
       "2154032 chrX   997278    998378  AKAP17A    WTC11_cluster5 -0.0268017392\n",
       "2154033 chrX   997278    998378  ASMTL      WTC11_cluster5 -0.0355930402\n",
       "2154034 chrX   997278    998378  PLCXD1     WTC11_cluster5  0.0147361250\n",
       "2154035 chrX   997278    998378  PPP2R3B    WTC11_cluster5  0.0016139973\n",
       "2154036 chrX   997278    998378  SHOX       WTC11_cluster5 -0.0133403901\n",
       "2154037 chrX   997278    998378  SLC25A6    WTC11_cluster5 -0.0056443322\n",
       "2154038 chrX 99736932  99737301  PCDH19     WTC11_cluster5 -0.0152605944\n",
       "2154039 chrX 99736932  99737301  SYTL4      WTC11_cluster5 -0.0196086170\n",
       "2154040 chrX 99736932  99737301  TSPAN6     WTC11_cluster5 -0.0407624938\n",
       "2154041 chrX 99777683  99778218  PCDH19     WTC11_cluster5 -0.0096141240\n",
       "2154042 chrX 99777683  99778218  SYTL4      WTC11_cluster5  0.0476469913\n",
       "2154043 chrX 99777683  99778218  TSPAN6     WTC11_cluster5  0.0808838540\n",
       "2154044 chrX  9981419   9981826  CLCN4      WTC11_cluster5 -0.0417834216\n",
       "2154045 chrX  9981419   9981826  GPR143     WTC11_cluster5 -0.0630702662\n",
       "2154046 chrX  9981419   9981826  SHROOM2    WTC11_cluster5 -0.0676506335\n",
       "2154047 chrX  9981419   9981826  TBL1X      WTC11_cluster5  0.0540469303\n",
       "2154048 chrX  9981419   9981826  WWC3       WTC11_cluster5 -0.0118586393\n",
       "2154049 chrX  9995697   9996195  CLCN4      WTC11_cluster5  0.0064288064\n",
       "2154050 chrX  9995697   9996195  GPR143     WTC11_cluster5 -0.0487806513\n",
       "2154051 chrX  9995697   9996195  SHROOM2    WTC11_cluster5  0.0466855334\n",
       "2154052 chrX  9995697   9996195  TBL1X      WTC11_cluster5 -0.0011727521\n",
       "2154053 chrX  9995697   9996195  WWC3       WTC11_cluster5  0.0193596879\n",
       "        rBgSD      rBgMean       pvalZ     \n",
       "1       0.03500772  0.0022090568 0.52134081\n",
       "2       0.03751772 -0.0016489257 0.67635418\n",
       "3       0.03448720  0.0019034824 0.09155178\n",
       "4       0.03433555  0.0007787370 0.52770554\n",
       "5       0.03516111  0.0021948907 0.44767732\n",
       "6       0.03580765  0.0021248389 0.86445713\n",
       "7       0.02994208 -0.0072910050 0.73631413\n",
       "8       0.03795531  0.0007213697 0.96060541\n",
       "9       0.03994222  0.0072616889 0.82206451\n",
       "10      0.03485896 -0.0001721060 0.45292273\n",
       "11      0.03821186  0.0053436148 0.73280037\n",
       "12      0.03619236 -0.0014530316 0.66734264\n",
       "13      0.03136140  0.0061813478 0.49903876\n",
       "14      0.03161747 -0.0057465276 0.05892840\n",
       "15      0.03425375  0.0015830328 0.55492682\n",
       "16      0.03076109  0.0084762993 0.12590689\n",
       "17      0.03958892 -0.0045085480 0.20299390\n",
       "18      0.03377062 -0.0038545854 0.39375711\n",
       "19      0.03359450 -0.0043032403 0.17794768\n",
       "20      0.03831282 -0.0006972310 0.35087349\n",
       "21      0.03361914  0.0069642818 0.84289323\n",
       "22      0.03669881  0.0105400049 0.15117285\n",
       "23      0.03705862  0.0104966867 0.40723044\n",
       "24      0.03147454  0.0064535439 0.22744086\n",
       "25      0.03262103  0.0099584246 0.56274007\n",
       "26      0.03344443  0.0070783637 0.17324765\n",
       "27      0.03416016 -0.0056006958 0.32732890\n",
       "28      0.03182288 -0.0100945757 0.81271067\n",
       "29      0.03662336  0.0031704657 0.61562086\n",
       "30      0.03366280 -0.0103589079 0.81600082\n",
       "⋮       ⋮          ⋮             ⋮         \n",
       "2154024 0.03725265  1.557573e-02 0.45256567\n",
       "2154025 0.03752539 -1.929640e-03 0.48516493\n",
       "2154026 0.03155441 -7.379180e-03 0.41861453\n",
       "2154027 0.03608956  3.680218e-03 0.07684569\n",
       "2154028 0.03694833  2.287367e-05 0.94392152\n",
       "2154029 0.03024148 -7.058040e-03 0.17874803\n",
       "2154030 0.03468627 -2.759235e-03 0.42385496\n",
       "2154031 0.03640750  2.007551e-03 0.76720651\n",
       "2154032 0.03523213  3.026012e-03 0.80139279\n",
       "2154033 0.03516824 -5.127752e-03 0.80682970\n",
       "2154034 0.03820177 -4.742011e-03 0.30506946\n",
       "2154035 0.03189453  5.927069e-03 0.55378467\n",
       "2154036 0.01664838 -1.064262e-02 0.56436450\n",
       "2154037 0.03458086  2.742322e-03 0.59581254\n",
       "2154038 0.03692273 -1.165598e-02 0.53888534\n",
       "2154039 0.03077845 -1.196892e-03 0.72514737\n",
       "2154040 0.04694311  7.929405e-03 0.85019111\n",
       "2154041 0.03386363 -4.489820e-03 0.56013906\n",
       "2154042 0.03064946 -2.746163e-03 0.05006977\n",
       "2154043 0.03281378  1.411632e-03 0.00771945\n",
       "2154044 0.03579531  2.417928e-03 0.89155427\n",
       "2154045 0.03655568  1.891174e-03 0.96222036\n",
       "2154046 0.03411475 -4.654688e-03 0.96759667\n",
       "2154047 0.03513857  2.464299e-04 0.06287291\n",
       "2154048 0.03354974 -1.789698e-03 0.61795710\n",
       "2154049 0.03826308  8.588788e-03 0.52250866\n",
       "2154050 0.03599652 -3.701521e-03 0.89477319\n",
       "2154051 0.03419192 -3.997019e-03 0.06913074\n",
       "2154052 0.03546782 -3.996833e-03 0.46826825\n",
       "2154053 0.03938626  6.089293e-03 0.36808501"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = celltype\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   data.table_1.15.2                \n",
       " [3] foreach_1.5.2                     BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] genomation_1.34.0                 Seurat_5.0.3                     \n",
       "[11] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[13] FigR_0.1.0                        motifmatchr_1.24.0               \n",
       "[15] chromVAR_1.24.0                   cowplot_1.1.3                    \n",
       "[17] ggplot2_3.5.1                     dplyr_1.1.4                      \n",
       "[19] SummarizedExperiment_1.30.2       Biobase_2.60.0                   \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               MatrixGenerics_1.12.3            \n",
       "[27] matrixStats_1.3.0                 Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             vroom_1.6.5                \n",
       " [15] globals_0.16.3              lattice_0.22-6             \n",
       " [17] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [19] plotly_4.10.4               plotrix_3.8-4              \n",
       " [21] yaml_2.3.8                  httpuv_1.6.15              \n",
       " [23] sctransform_0.4.1           spam_2.10-0                \n",
       " [25] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [27] pbapply_1.7-2               DBI_1.2.2                  \n",
       " [29] CNEr_1.38.0                 RColorBrewer_1.1-3         \n",
       " [31] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [33] Rtsne_0.17                  purrr_1.0.2                \n",
       " [35] R.utils_2.12.3              RCurl_1.98-1.14            \n",
       " [37] pracma_2.4.4                GenomeInfoDbData_1.2.11    \n",
       " [39] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [41] spatstat.utils_3.0-4        listenv_0.9.1              \n",
       " [43] seqLogo_1.68.0              goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             spatstat.random_3.2-3      \n",
       " [47] annotate_1.80.0             fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           leiden_0.4.3.1             \n",
       " [51] codetools_0.2-20            DelayedArray_0.26.7        \n",
       " [53] DT_0.33                     tidyselect_1.2.1           \n",
       " [55] spatstat.explore_3.2-6      base64enc_0.1-3            \n",
       " [57] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [59] progressr_0.14.0            iterators_1.0.14           \n",
       " [61] ggridges_0.5.6              survival_3.6-4             \n",
       " [63] tools_4.3.3                 TFMPvalue_0.0.9            \n",
       " [65] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [67] glue_1.7.0                  gridExtra_2.3              \n",
       " [69] IRdisplay_1.1               withr_3.0.0                \n",
       " [71] fastmap_1.1.1               fansi_1.0.6                \n",
       " [73] caTools_1.18.2              digest_0.6.35              \n",
       " [75] R6_2.5.1                    mime_0.12                  \n",
       " [77] nabor_0.5.0                 seqPattern_1.34.0          \n",
       " [79] colorspace_2.1-0            scattermore_1.2            \n",
       " [81] GO.db_3.18.0                tensor_1.5                 \n",
       " [83] gtools_3.9.5                poweRlaw_0.80.0            \n",
       " [85] spatstat.data_3.0-4         RSQLite_2.3.6              \n",
       " [87] R.methodsS3_1.8.2           utf8_1.2.4                 \n",
       " [89] tidyr_1.3.1                 generics_0.1.3             \n",
       " [91] httr_1.4.7                  htmlwidgets_1.6.4          \n",
       " [93] S4Arrays_1.2.1              TFBSTools_1.40.0           \n",
       " [95] uwot_0.1.16                 pkgconfig_2.0.3            \n",
       " [97] gtable_0.3.5                blob_1.2.4                 \n",
       " [99] impute_1.76.0               lmtest_0.9-40              \n",
       "[101] htmltools_0.5.8.1           dotCall64_1.1-1            \n",
       "[103] scales_1.3.0                png_0.1-8                  \n",
       "[105] tzdb_0.4.0                  reshape2_1.4.4             \n",
       "[107] rjson_0.2.21                uuid_1.2-0                 \n",
       "[109] nlme_3.1-164                repr_1.1.7                 \n",
       "[111] cachem_1.0.8                zoo_1.8-12                 \n",
       "[113] stringr_1.5.1               KernSmooth_2.23-22         \n",
       "[115] miniUI_0.1.1.1              AnnotationDbi_1.64.1       \n",
       "[117] restfulr_0.0.15             pillar_1.9.0               \n",
       "[119] vctrs_0.6.5                 RANN_2.6.1                 \n",
       "[121] promises_1.3.0              xtable_1.8-4               \n",
       "[123] cluster_2.1.6               evaluate_0.23              \n",
       "[125] readr_2.1.5                 cli_3.6.2                  \n",
       "[127] compiler_4.3.3              Rsamtools_2.18.0           \n",
       "[129] rlang_1.1.3                 crayon_1.5.2               \n",
       "[131] future.apply_1.11.2         plyr_1.8.9                 \n",
       "[133] stringi_1.8.3               gridBase_0.4-7             \n",
       "[135] deldir_2.0-4                viridisLite_0.4.2          \n",
       "[137] BiocParallel_1.36.0         munsell_0.5.1              \n",
       "[139] lazyeval_0.2.2              spatstat.geom_3.2-9        \n",
       "[141] IRkernel_1.3.2              RcppHNSW_0.6.0             \n",
       "[143] hms_1.1.3                   patchwork_1.2.0            \n",
       "[145] bit64_4.0.5                 future_1.33.2              \n",
       "[147] KEGGREST_1.40.1             shiny_1.8.1.1              \n",
       "[149] ROCR_1.0-11                 igraph_2.0.3               \n",
       "[151] memoise_2.0.1               bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
