{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a1e63f9-7479-4a30-83ea-381f1ecdbe50",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: cowplot\n",
      "\n",
      "Loading required package: chromVAR\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: motifmatchr\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:grid’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(FigR)\n",
    "library(Seurat)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(foreach)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53f3774-52c3-4baa-b707-5b3a0450b130",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0785f3fc-9913-428a-bea7-5710120c9328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/Jurkat/Jurkat/Kendall/Pairs.tsv.gz\"\n",
    "path.pairs.ABC = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/Jurkat/Jurkat/multiome_powerlaw_v3/encode_e2g_predictions.tsv.gz\"\n",
    "path.matrix.atac_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/Jurkat/Jurkat/Kendall/atac_matrix.rds\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/scE2G_input_Maya/Jurkat/rna_count_matrix.csv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/Jurkat/1.Genome_wide_prediction/FigR/FigR.250321/\"\n",
    "celltype = \"Jurkat\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "500ba317-011f-4cab-834c-de750cb7f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9cdab1-486b-4539-99d3-3cd5d1d99795",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5852d15-7e3f-412b-b137-7d29a6fb47b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79f3220-9bc6-4496-9e9a-a2dcc6e7f751",
   "metadata": {},
   "source": [
    "Import ABC results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6bc3f6ba-5e9a-488c-ba26-e0c84007ef4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC = readGeneric(path.pairs.ABC,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d660489-4656-4b85-a3df-381acf0aa4a1",
   "metadata": {},
   "source": [
    "Filter ABC results which distance < 1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31b4d7f1-0feb-428b-934a-9c32cb2a4f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.ABC.1M = pairs.ABC[pairs.ABC$distance < 10^6]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639699dd-5bf9-4b19-9178-82dbbf2c62d0",
   "metadata": {},
   "source": [
    "Filter E-G pairs overlaping with pairs.ABC.1M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7db8c1b-5245-49e1-862e-979d9f1de142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2793593 ranges and 3 metadata columns:\n",
       "            seqnames            ranges strand |  TargetGene\n",
       "               <Rle>         <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 10001055-10001570      * | APITD1-CORT\n",
       "        [2]     chr1 10001055-10001570      * |       CASZ1\n",
       "        [3]     chr1 10001055-10001570      * |       CENPS\n",
       "        [4]     chr1 10001055-10001570      * |      CLSTN1\n",
       "        [5]     chr1 10001055-10001570      * |    CTNNBIP1\n",
       "        ...      ...               ...    ... .         ...\n",
       "  [2793589]     chrX   9941883-9942743      * |        WWC3\n",
       "  [2793590]     chrX 99603024-99603776      * |      PCDH19\n",
       "  [2793591]     chrX 99632013-99632219      * |      PCDH19\n",
       "  [2793592]     chrX   9995829-9996207      * |       TBL1X\n",
       "  [2793593]     chrX   9995829-9996207      * |        WWC3\n",
       "                          PeakName               PairName\n",
       "                       <character>            <character>\n",
       "        [1] chr1-10001055-10001570 chr1-10001055-100015..\n",
       "        [2] chr1-10001055-10001570 chr1-10001055-100015..\n",
       "        [3] chr1-10001055-10001570 chr1-10001055-100015..\n",
       "        [4] chr1-10001055-10001570 chr1-10001055-100015..\n",
       "        [5] chr1-10001055-10001570 chr1-10001055-100015..\n",
       "        ...                    ...                    ...\n",
       "  [2793589]   chrX-9941883-9942743 chrX-9941883-9942743..\n",
       "  [2793590] chrX-99603024-99603776 chrX-99603024-996037..\n",
       "  [2793591] chrX-99632013-99632219 chrX-99632013-996322..\n",
       "  [2793592]   chrX-9995829-9996207 chrX-9995829-9996207..\n",
       "  [2793593]   chrX-9995829-9996207 chrX-9995829-9996207..\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.pairs.E2G.chr_rename = as.data.frame(pairs.E2G)[,1:3]\n",
    "df.pairs.E2G.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.E2G),\n",
    "                                             mcols(pairs.E2G)[,\"TargetGene\"],\n",
    "                                             sep = \"_\")\n",
    "pairs.E2G.chr_rename = GRanges(df.pairs.E2G.chr_rename)\n",
    "rm(df.pairs.E2G.chr_rename)\n",
    "\n",
    "df.pairs.ABC.1M.chr_rename = as.data.frame(pairs.ABC.1M)[,1:3]\n",
    "df.pairs.ABC.1M.chr_rename[,\"seqnames\"] = paste(seqnames(pairs.ABC.1M),\n",
    "                                                mcols(pairs.ABC.1M)[,\"TargetGene\"],\n",
    "                                                sep = \"_\")\n",
    "pairs.ABC.1M.chr_rename = GRanges(df.pairs.ABC.1M.chr_rename)\n",
    "rm(df.pairs.ABC.1M.chr_rename)\n",
    "\n",
    "pairs.E2G.filter = pairs.E2G[countOverlaps(pairs.E2G.chr_rename,\n",
    "                                           pairs.ABC.1M.chr_rename) > 0]\n",
    "rm(pairs.E2G.chr_rename)\n",
    "rm(pairs.ABC.1M.chr_rename)\n",
    "pairs.E2G.filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcdfac67-ce32-4d9a-8f4d-65dab644ec63",
   "metadata": {},
   "source": [
    "Import ATAC matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd59feb4-3e33-4ebe-a107-f3b6d347a68e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix object input detectedCentering counts for cells sequentially in groups of size  1000  ..\n",
      "\n",
      "Computing centered counts for cells:  1  to  1000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  1001  to  2000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  2001  to  3000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  3001  to  4000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  4001  to  5000 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Computing centered counts for cells:  5001  to  5277 ..\n",
      "Computing centered counts per cell using mean reads in features ..\n",
      "\n",
      "Merging results..\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "matrix.atac_count = readRDS(path.matrix.atac_count)\n",
    "matrix.atac = centerCounts(matrix.atac_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c5cb96-2fc4-42c4-a421-fe9a5ca537da",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "96e10e90-9400-4cfe-bc7b-46a7eaee1fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna_count = read.csv(path.matrix.rna_count,\n",
    "                            row.names = 1,\n",
    "                            check.names = F)\n",
    "matrix.rna_count = Matrix(as.matrix(matrix.rna_count), sparse = TRUE)\n",
    "matrix.rna_count = matrix.rna_count[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna_count[rowSums(matrix.rna_count) > 0,]\n",
    "matrix.rna = NormalizeData(matrix.rna)\n",
    "rm(matrix.rna_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dd80eda-2392-4d59-8250-1ae62d7e1d1e",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b9e26c70-11fd-47c4-8e05-9c96b17d3f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a76b1603-6b00-4ac0-9399-fd341c5a7b88",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ad0f51bc-6e09-4b83-be71-560130dd0fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.filter2 = pairs.E2G.filter[pairs.E2G.filter$TargetGene %in% rownames(matrix.rna.rename) &\n",
    "                                     pairs.E2G.filter$PeakName %in% rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16f7d5b-eabd-4e93-b8cd-f19a54df21f2",
   "metadata": {},
   "source": [
    "Prepare FigR input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "49c6d711-b870-4bab-869e-fd88f2f58100",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.peak = pairs.E2G\n",
    "mcols(bed.peak) = NULL\n",
    "bed.peak = unique(bed.peak)\n",
    "bed.peak$PeakName = paste(seqnames(bed.peak),\n",
    "                          start(bed.peak),\n",
    "                          end(bed.peak),\n",
    "                          sep = \"-\")\n",
    "names(bed.peak) = bed.peak$PeakName\n",
    "bed.peak = bed.peak[rownames(matrix.atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c828f792-f8d7-4054-930d-4495d04a6c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 152887 5277 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(152887): chr1-10001055-10001570 chr1-100037799-100039086 ...\n",
       "  chrX-99632013-99632219 chrX-9995829-9996207\n",
       "rowData names(2): PeakName bias\n",
       "colnames(5277): e10l1_AAACCAGGTACTTATG-1 e10l1_AAACCAGGTCACACTG-1 ...\n",
       "  e10l2_TTTGTGAGTCCTGAGC-1 e10l2_TTTGTGAGTTTATGTG-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ATAC.se = SummarizedExperiment(assays = SimpleList(counts = matrix.atac_count),\n",
    "                               rowRanges = bed.peak)\n",
    "rm(matrix.atac_count)\n",
    "ATAC.se <- chromVAR::addGCBias(ATAC.se, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "ATAC.se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "79c8b86b-7706-475a-a501-c52c7d98beb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg <- chromVAR::getBackgroundPeaks(ATAC.se, niterations = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ff93bd06-be5a-41cc-a9bb-3372df4270b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "index.gene = data.frame(gene_name = rownames(matrix.rna.rename),\n",
    "                       index = 1:nrow(matrix.rna.rename))\n",
    "rownames(index.gene) = index.gene$gene_name\n",
    "\n",
    "index.peak = data.frame(peak_name = rownames(matrix.atac),\n",
    "                       index = 1:nrow(matrix.atac))\n",
    "rownames(index.peak) = index.peak$peak_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126b2dcf-5fe7-42a4-8d71-bff93d330cc2",
   "metadata": {},
   "source": [
    "Run FigR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f6909129-c4bd-48a5-8d0d-cea5d8306efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "Sys.setenv(OPENBLAS_NUM_THREADS=1)\n",
    "PeakGeneCor.modified = function (ATAC, RNA, OV, ncores = 4, chunkSize = 200, metric = \"spearman\", \n",
    "    bg = NULL) \n",
    "{\n",
    "    stopifnot(ncol(ATAC) == ncol(RNA))\n",
    "    if (chunkSize > 1000) \n",
    "        stop(\"Do not specify very large chunk sizes. Please use chunkSize < 1000\")\n",
    "    # n <- length(OV)\n",
    "    n <- nrow(OV) # import OV as a dataframe instead of Hits object\n",
    "    starts <- seq(1, n, chunkSize)\n",
    "    ends <- starts + chunkSize - 1\n",
    "    ends[length(ends)] <- n\n",
    "    # OVd <- OV %>% as.data.frame() %>% dplyr::rename(Gene = \"queryHits\", \n",
    "    #     Peak = \"subjectHits\")\n",
    "    OVd <- OV %>% dplyr::rename(Gene = \"queryHits\", Peak = \"subjectHits\")\n",
    "    chunkList <- mapply(c, starts, ends, SIMPLIFY = FALSE)\n",
    "    time_elapsed <- Sys.time()\n",
    "    cat(\"Running in parallel using \", ncores, \"cores ..\\n\")\n",
    "    cat(\"Computing observed correlations ..\\n\")\n",
    "    corList <- pbmcapply::pbmclapply(X = chunkList, FUN = function(x) {\n",
    "        FigR::chunkCore(chunk = x, A = ATAC, R = RNA, O = OVd, \n",
    "            met = metric)\n",
    "    }, mc.cores = ncores)\n",
    "    if (any(unlist(sapply(corList, is.null)))) {\n",
    "        message(\"One or more of the chunk processes failed unexpectedly (returned NULL) ..\")\n",
    "        message(\"Please check to see you have enough cores/memory allocated\")\n",
    "        message(\"Also make sure you have filtered down to non-zero peaks/genes\")\n",
    "    }\n",
    "    OVd$rObs <- unlist(corList)\n",
    "    cat(\"Finished!\\n\")\n",
    "    time_elapsed <- Sys.time() - time_elapsed\n",
    "    cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "        \"\\n\\n\")\n",
    "    if (!is.null(bg)) {\n",
    "        n_iter <- ncol(bg)\n",
    "        cat(\"Computing background correlations ..\\n\")\n",
    "        time_elapsed <- Sys.time()\n",
    "        bgCor <- foreach(i = 1:n_iter, .combine = \"cbind\", .export = c(\"chunkCore\", \n",
    "            \"t\"), .packages = c(\"pbmcapply\", \"FigR\", \"Matrix\")) %do% \n",
    "            {\n",
    "                OVdBg <- OVd[, 1:2]\n",
    "                OVdBg$Peak <- bg[OVdBg$Peak, i]\n",
    "                bgCorList <- pbmcapply::pbmclapply(X = chunkList, \n",
    "                  FUN = function(x) {\n",
    "                    chunkCore(chunk = x, A = ATAC, R = RNA, O = OVdBg, \n",
    "                      met = metric)\n",
    "                  }, mc.cores = ncores)\n",
    "                unlist(bgCorList)\n",
    "            }\n",
    "        if (sum(is.null(bgCor)) != 0 | sum(is.na(bgCor)) != 0) \n",
    "            stop(\"One or more of the chunk processes failed unexpectedly (returned NULL) .. Please check to see you have enough cores/m\\n           emory allocated\")\n",
    "        time_elapsed <- Sys.time() - time_elapsed\n",
    "        cat(paste(\"\\nTime Elapsed: \", time_elapsed, units(time_elapsed)), \n",
    "            \"\\n\\n\")\n",
    "        colnames(bgCor) <- paste0(\"rBg\", 1:ncol(bgCor))\n",
    "        OVd <- cbind(OVd, bgCor)\n",
    "    }\n",
    "    return(OVd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2024aa79-81fb-480d-8523-36bc72fbc65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"chr\",sep = \"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "926aa663-820b-4a81-b814-58945f5aa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.done = dir(paste(dir.output,\"chr\",sep = \"/\"))\n",
    "chr.run = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "chr.run = chr.run[!chr.run %in% chr.done]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "17822bf5-e813-4429-9527-30ed50efbc2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"chr1\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  16.1432342529297 secs \n",
      "\n",
      "Computing background correlations ..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: parallel\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Time Elapsed:  30.0909704764684 mins \n",
      "\n",
      "[1] \"chr10\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.121141910553 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.9289192756017 mins \n",
      "\n",
      "[1] \"chr11\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  13.3429152965546 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  20.7545115788778 mins \n",
      "\n",
      "[1] \"chr12\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.3885886669159 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  15.1394369324048 mins \n",
      "\n",
      "[1] \"chr13\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.86807775497437 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  5.57543821334839 mins \n",
      "\n",
      "[1] \"chr14\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.41060042381287 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.23266266981761 mins \n",
      "\n",
      "[1] \"chr15\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.26974368095398 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  9.06810755332311 mins \n",
      "\n",
      "[1] \"chr16\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  14.9288682937622 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  21.6074665347735 mins \n",
      "\n",
      "[1] \"chr17\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  15.8725936412811 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  26.438124871254 mins \n",
      "\n",
      "[1] \"chr18\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.6236617565155 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  4.04457355340322 mins \n",
      "\n",
      "[1] \"chr19\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  19.3255090713501 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  36.697988152504 mins \n",
      "\n",
      "[1] \"chr2\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  10.9960496425629 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  18.7515698512395 mins \n",
      "\n",
      "[1] \"chr20\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.36860632896423 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  11.3327712972959 mins \n",
      "\n",
      "[1] \"chr21\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  2.67120027542114 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  4.5242839217186 mins \n",
      "\n",
      "[1] \"chr22\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.6109459400177 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  10.1519068837166 mins \n",
      "\n",
      "[1] \"chr3\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  8.48781108856201 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  13.9451219439507 mins \n",
      "\n",
      "[1] \"chr4\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.89036750793457 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.18260622024536 mins \n",
      "\n",
      "[1] \"chr5\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.56172442436218 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  10.9124864737193 mins \n",
      "\n",
      "[1] \"chr6\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  9.56671333312988 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  15.3082725008329 mins \n",
      "\n",
      "[1] \"chr7\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  7.75353980064392 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  12.634365773201 mins \n",
      "\n",
      "[1] \"chr8\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  5.43592596054077 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.48530897299449 mins \n",
      "\n",
      "[1] \"chr9\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  6.45518493652344 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  12.7037832657496 mins \n",
      "\n",
      "[1] \"chrX\"\n",
      "Running in parallel using  8 cores ..\n",
      "Computing observed correlations ..\n",
      "Finished!\n",
      "\n",
      "Time Elapsed:  4.40184640884399 secs \n",
      "\n",
      "Computing background correlations ..\n",
      "\n",
      "Time Elapsed:  8.02499696413676 mins \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(chr.tmp in chr.run){\n",
    "  print(chr.tmp)\n",
    "  pairs.E2G.chr.res = pairs.E2G.filter2[seqnames(pairs.E2G.filter2) == chr.tmp]\n",
    "  genePeakOv.chr <- data.frame(queryHits = index.gene[pairs.E2G.chr.res$TargetGene, \"index\"],\n",
    "                               subjectHits = index.peak[pairs.E2G.chr.res$PeakName, \"index\"])\n",
    "  ObsCor.chr = PeakGeneCor.modified(ATAC = matrix.atac, \n",
    "                                    RNA = matrix.rna.rename,\n",
    "                                    OV = genePeakOv.chr, \n",
    "                                    chunkSize = 500,\n",
    "                                    ncores = n.cores, \n",
    "                                    bg = bg)\n",
    "  pairs.E2G.chr.res$rObs = ObsCor.chr[,\"rObs\"]\n",
    "  pairs.E2G.chr.res$rBgSD <- matrixStats::rowSds(as.matrix(ObsCor.chr[, 4:103]))\n",
    "  pairs.E2G.chr.res$rBgMean <- rowMeans(ObsCor.chr[, 4:103])\n",
    "  pairs.E2G.chr.res$pvalZ <- 1 - stats::pnorm(q = pairs.E2G.chr.res$rObs, \n",
    "                                              mean = pairs.E2G.chr.res$rBgMean,\n",
    "                                              sd = pairs.E2G.chr.res$rBgSD)\n",
    "  saveRDS(pairs.E2G.chr.res,\n",
    "          paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "  rm (pairs.E2G.chr.res)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da9d1451-54e0-49e0-ae8b-cd0158115820",
   "metadata": {},
   "source": [
    "Merge results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6b4278f1-7d9d-402f-95c6-f274a6befea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr.merge = as.character(unique(seqnames(pairs.E2G.filter2)))\n",
    "list.res = list()\n",
    "for(chr.tmp in chr.merge){\n",
    "    list.res[[chr.tmp]] = readRDS(paste(dir.output,\"chr\",chr.tmp,sep = \"/\"))\n",
    "}\n",
    "pairs.E2G.res = unlist(as(list.res, \"GRangesList\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b387943f-773e-4f6a-8938-b6e7d91181ac",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "48824f05-aedd-476c-88d6-ae58d1833fc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2590409 ranges and 7 metadata columns:\n",
       "       seqnames            ranges strand |  TargetGene               PeakName\n",
       "          <Rle>         <IRanges>  <Rle> | <character>            <character>\n",
       "  chr1     chr1 10001055-10001570      * | APITD1-CORT chr1-10001055-10001570\n",
       "  chr1     chr1 10001055-10001570      * |       CASZ1 chr1-10001055-10001570\n",
       "  chr1     chr1 10001055-10001570      * |      CLSTN1 chr1-10001055-10001570\n",
       "  chr1     chr1 10001055-10001570      * |    CTNNBIP1 chr1-10001055-10001570\n",
       "  chr1     chr1 10001055-10001570      * |        DFFA chr1-10001055-10001570\n",
       "   ...      ...               ...    ... .         ...                    ...\n",
       "  chrX     chrX   9941883-9942743      * |        WWC3   chrX-9941883-9942743\n",
       "  chrX     chrX 99603024-99603776      * |      PCDH19 chrX-99603024-99603776\n",
       "  chrX     chrX 99632013-99632219      * |      PCDH19 chrX-99632013-99632219\n",
       "  chrX     chrX   9995829-9996207      * |       TBL1X   chrX-9995829-9996207\n",
       "  chrX     chrX   9995829-9996207      * |        WWC3   chrX-9995829-9996207\n",
       "                     PairName        rObs     rBgSD      rBgMean     pvalZ\n",
       "                  <character>   <numeric> <numeric>    <numeric> <numeric>\n",
       "  chr1 chr1-10001055-100015.. -0.00714113 0.0141306 -0.000317594 0.6854129\n",
       "  chr1 chr1-10001055-100015.. -0.00225330 0.0153861  0.003392746 0.6431751\n",
       "  chr1 chr1-10001055-100015..  0.01976536 0.0137973  0.000469834 0.0809821\n",
       "  chr1 chr1-10001055-100015.. -0.02771419 0.0145250  0.002252927 0.9804503\n",
       "  chr1 chr1-10001055-100015.. -0.01853964 0.0142589 -0.000524187 0.8967874\n",
       "   ...                    ...         ...       ...          ...       ...\n",
       "  chrX chrX-9941883-9942743..  0.00980701 0.0149033  0.004247553  0.354561\n",
       "  chrX chrX-99603024-996037.. -0.01207779 0.0153561  0.002882824  0.835033\n",
       "  chrX chrX-99632013-996322.. -0.00703253 0.0154541  0.000510751  0.687264\n",
       "  chrX chrX-9995829-9996207.. -0.00747444 0.0155100 -0.000374349  0.676443\n",
       "  chrX chrX-9995829-9996207..  0.01146317 0.0155526  0.002561411  0.283537\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2297128a-9ed6-42b8-90c7-459969fb5afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2590409 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>rObs</th><th scope=col>rBgSD</th><th scope=col>rBgMean</th><th scope=col>pvalZ</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>APITD1-CORT</td><td>Jurkat</td><td>-0.0071411282</td><td>0.014130634</td><td>-3.175940e-04</td><td>0.68541289</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>CASZ1      </td><td>Jurkat</td><td>-0.0022533046</td><td>0.015386062</td><td> 3.392746e-03</td><td>0.64317512</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>CLSTN1     </td><td>Jurkat</td><td> 0.0197653646</td><td>0.013797346</td><td> 4.698342e-04</td><td>0.08098212</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>CTNNBIP1   </td><td>Jurkat</td><td>-0.0277141855</td><td>0.014525003</td><td> 2.252927e-03</td><td>0.98045033</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>DFFA       </td><td>Jurkat</td><td>-0.0185396358</td><td>0.014258860</td><td>-5.241866e-04</td><td>0.89678741</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>GPR157     </td><td>Jurkat</td><td> 0.0148254118</td><td>0.013936212</td><td> 1.939612e-03</td><td>0.17757994</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>H6PD       </td><td>Jurkat</td><td>-0.0050693443</td><td>0.014255930</td><td> 8.466787e-04</td><td>0.66092423</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>KIF1B      </td><td>Jurkat</td><td> 0.0107859814</td><td>0.013129322</td><td>-1.285664e-05</td><td>0.20539682</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>LINC01759  </td><td>Jurkat</td><td>-0.0019488251</td><td>0.008548509</td><td>-1.129132e-03</td><td>0.53819492</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>LZIC       </td><td>Jurkat</td><td>-0.0236758420</td><td>0.014126920</td><td> 1.161149e-03</td><td>0.96063748</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>MIR34AHG   </td><td>Jurkat</td><td>-0.0033773843</td><td>0.014595356</td><td> 5.302928e-04</td><td>0.60554807</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>NMNAT1     </td><td>Jurkat</td><td>-0.0064528480</td><td>0.014527124</td><td>-5.517122e-05</td><td>0.67017458</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>PEX14      </td><td>Jurkat</td><td> 0.0162888946</td><td>0.013443455</td><td> 2.504272e-03</td><td>0.15259244</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>PGD        </td><td>Jurkat</td><td>-0.0005149352</td><td>0.013118354</td><td> 2.411489e-03</td><td>0.58826284</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>PIK3CD     </td><td>Jurkat</td><td> 0.0151808859</td><td>0.013401102</td><td> 3.841117e-03</td><td>0.19872566</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>SLC25A33   </td><td>Jurkat</td><td>-0.0021572634</td><td>0.014036426</td><td> 3.328312e-03</td><td>0.65203114</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>SPSB1      </td><td>Jurkat</td><td>-0.0008457267</td><td>0.017602230</td><td> 2.974110e-03</td><td>0.58589921</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>TMEM201    </td><td>Jurkat</td><td> 0.0149095049</td><td>0.013324155</td><td> 1.553319e-03</td><td>0.15807425</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001055</td><td> 10001570</td><td>UBE4B      </td><td>Jurkat</td><td> 0.0169501125</td><td>0.012241596</td><td>-5.240791e-04</td><td>0.07672600</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>AGL        </td><td>Jurkat</td><td> 0.0075230643</td><td>0.013984123</td><td> 6.665222e-03</td><td>0.47554262</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>CDC14A     </td><td>Jurkat</td><td> 0.0061538299</td><td>0.012993107</td><td> 8.899091e-03</td><td>0.58366792</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>DBT        </td><td>Jurkat</td><td> 0.0103637288</td><td>0.014309521</td><td> 7.772240e-03</td><td>0.42814360</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>DPH5       </td><td>Jurkat</td><td> 0.0230590370</td><td>0.014420749</td><td>-1.051256e-03</td><td>0.04727036</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>EXTL2      </td><td>Jurkat</td><td> 0.0223085057</td><td>0.014093499</td><td> 1.356398e-02</td><td>0.26747587</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>FRRS1      </td><td>Jurkat</td><td> 0.0192187427</td><td>0.012409335</td><td> 6.481489e-03</td><td>0.15234561</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>MFSD14A    </td><td>Jurkat</td><td>-0.0081440005</td><td>0.014368065</td><td> 9.300673e-04</td><td>0.73615762</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>RTCA       </td><td>Jurkat</td><td> 0.0155712767</td><td>0.015053319</td><td> 6.233542e-03</td><td>0.26752662</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>RTCA-AS1   </td><td>Jurkat</td><td> 0.0088131761</td><td>0.013010994</td><td> 7.623465e-03</td><td>0.46357192</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>SASS6      </td><td>Jurkat</td><td> 0.0062294533</td><td>0.016029459</td><td> 6.523191e-03</td><td>0.50731015</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037799</td><td>100039086</td><td>SLC30A7    </td><td>Jurkat</td><td> 0.0029814707</td><td>0.013460286</td><td> 7.207882e-03</td><td>0.62323615</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>IL3RA    </td><td>Jurkat</td><td>-0.0114866278</td><td>0.01180815</td><td>-1.632231e-03</td><td>0.79801226</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>LINC00106</td><td>Jurkat</td><td> 0.0055581009</td><td>0.01510983</td><td> 1.448196e-03</td><td>0.39281023</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>P2RY8    </td><td>Jurkat</td><td> 0.0131535710</td><td>0.01301975</td><td> 1.549212e-03</td><td>0.18638707</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>PLCXD1   </td><td>Jurkat</td><td> 0.0083694509</td><td>0.01523631</td><td> 4.252237e-03</td><td>0.39349404</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>PPP2R3B  </td><td>Jurkat</td><td>-0.0092444213</td><td>0.01373220</td><td> 2.057085e-03</td><td>0.79474412</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>SHOX     </td><td>Jurkat</td><td>-0.0029084531</td><td>0.01644049</td><td> 1.525765e-03</td><td>0.60630956</td></tr>\n",
       "\t<tr><td>chrX</td><td>  978694</td><td>  979118</td><td>SLC25A6  </td><td>Jurkat</td><td>-0.0101946609</td><td>0.01486111</td><td> 2.546487e-03</td><td>0.80437383</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9842372</td><td> 9843073</td><td>TBL1X    </td><td>Jurkat</td><td>-0.0068704126</td><td>0.01525434</td><td> 8.456191e-04</td><td>0.69351041</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9842372</td><td> 9843073</td><td>WWC3     </td><td>Jurkat</td><td>-0.0269497814</td><td>0.01507101</td><td> 3.371021e-03</td><td>0.97788277</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9873846</td><td> 9874297</td><td>TBL1X    </td><td>Jurkat</td><td>-0.0207520878</td><td>0.01379675</td><td> 2.788075e-03</td><td>0.95601552</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9873846</td><td> 9874297</td><td>WWC3     </td><td>Jurkat</td><td>-0.0018099022</td><td>0.01317112</td><td> 1.838355e-03</td><td>0.60910581</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9885186</td><td> 9885487</td><td>TBL1X    </td><td>Jurkat</td><td> 0.0162982433</td><td>0.01313418</td><td>-1.695516e-04</td><td>0.10495515</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9885186</td><td> 9885487</td><td>WWC3     </td><td>Jurkat</td><td> 0.0032578932</td><td>0.01416270</td><td> 2.684010e-03</td><td>0.48383899</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>AKAP17A  </td><td>Jurkat</td><td> 0.0009846862</td><td>0.01358493</td><td> 1.016017e-02</td><td>0.75029448</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>ASMT     </td><td>Jurkat</td><td> 0.0235034382</td><td>0.01244424</td><td> 6.322977e-05</td><td>0.02980831</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>ASMTL    </td><td>Jurkat</td><td> 0.0019765367</td><td>0.01372576</td><td> 1.847555e-03</td><td>0.49625117</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>GTPBP6   </td><td>Jurkat</td><td> 0.0213985047</td><td>0.01314494</td><td> 6.783700e-03</td><td>0.13310782</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>IL3RA    </td><td>Jurkat</td><td> 0.0252265799</td><td>0.01450229</td><td> 4.717340e-03</td><td>0.07865053</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>LINC00106</td><td>Jurkat</td><td> 0.0089426480</td><td>0.01196052</td><td> 2.099447e-03</td><td>0.28361044</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>P2RY8    </td><td>Jurkat</td><td>-0.0090354619</td><td>0.01333771</td><td> 7.675060e-03</td><td>0.89487488</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>PLCXD1   </td><td>Jurkat</td><td> 0.0255621263</td><td>0.01511793</td><td> 1.085956e-02</td><td>0.16539481</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>PPP2R3B  </td><td>Jurkat</td><td> 0.0066262678</td><td>0.01454721</td><td> 6.960481e-03</td><td>0.50916465</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>SHOX     </td><td>Jurkat</td><td>-0.0004223613</td><td>0.01454376</td><td> 1.260592e-03</td><td>0.54606140</td></tr>\n",
       "\t<tr><td>chrX</td><td>  989869</td><td>  990628</td><td>SLC25A6  </td><td>Jurkat</td><td>-0.0098100685</td><td>0.01206209</td><td> 3.043785e-03</td><td>0.85670698</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9941883</td><td> 9942743</td><td>TBL1X    </td><td>Jurkat</td><td> 0.0105512284</td><td>0.01400607</td><td> 2.866355e-05</td><td>0.22624024</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9941883</td><td> 9942743</td><td>WWC3     </td><td>Jurkat</td><td> 0.0098070060</td><td>0.01490330</td><td> 4.247553e-03</td><td>0.35456119</td></tr>\n",
       "\t<tr><td>chrX</td><td>99603024</td><td>99603776</td><td>PCDH19   </td><td>Jurkat</td><td>-0.0120777939</td><td>0.01535607</td><td> 2.882824e-03</td><td>0.83503325</td></tr>\n",
       "\t<tr><td>chrX</td><td>99632013</td><td>99632219</td><td>PCDH19   </td><td>Jurkat</td><td>-0.0070325259</td><td>0.01545406</td><td> 5.107510e-04</td><td>0.68726393</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995829</td><td> 9996207</td><td>TBL1X    </td><td>Jurkat</td><td>-0.0074744363</td><td>0.01551003</td><td>-3.743486e-04</td><td>0.67644261</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995829</td><td> 9996207</td><td>WWC3     </td><td>Jurkat</td><td> 0.0114631748</td><td>0.01555261</td><td> 2.561411e-03</td><td>0.28353750</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2590409 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & rObs & rBgSD & rBgMean & pvalZ\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  10001055 &  10001570 & APITD1-CORT & Jurkat & -0.0071411282 & 0.014130634 & -3.175940e-04 & 0.68541289\\\\\n",
       "\t chr1 &  10001055 &  10001570 & CASZ1       & Jurkat & -0.0022533046 & 0.015386062 &  3.392746e-03 & 0.64317512\\\\\n",
       "\t chr1 &  10001055 &  10001570 & CLSTN1      & Jurkat &  0.0197653646 & 0.013797346 &  4.698342e-04 & 0.08098212\\\\\n",
       "\t chr1 &  10001055 &  10001570 & CTNNBIP1    & Jurkat & -0.0277141855 & 0.014525003 &  2.252927e-03 & 0.98045033\\\\\n",
       "\t chr1 &  10001055 &  10001570 & DFFA        & Jurkat & -0.0185396358 & 0.014258860 & -5.241866e-04 & 0.89678741\\\\\n",
       "\t chr1 &  10001055 &  10001570 & GPR157      & Jurkat &  0.0148254118 & 0.013936212 &  1.939612e-03 & 0.17757994\\\\\n",
       "\t chr1 &  10001055 &  10001570 & H6PD        & Jurkat & -0.0050693443 & 0.014255930 &  8.466787e-04 & 0.66092423\\\\\n",
       "\t chr1 &  10001055 &  10001570 & KIF1B       & Jurkat &  0.0107859814 & 0.013129322 & -1.285664e-05 & 0.20539682\\\\\n",
       "\t chr1 &  10001055 &  10001570 & LINC01759   & Jurkat & -0.0019488251 & 0.008548509 & -1.129132e-03 & 0.53819492\\\\\n",
       "\t chr1 &  10001055 &  10001570 & LZIC        & Jurkat & -0.0236758420 & 0.014126920 &  1.161149e-03 & 0.96063748\\\\\n",
       "\t chr1 &  10001055 &  10001570 & MIR34AHG    & Jurkat & -0.0033773843 & 0.014595356 &  5.302928e-04 & 0.60554807\\\\\n",
       "\t chr1 &  10001055 &  10001570 & NMNAT1      & Jurkat & -0.0064528480 & 0.014527124 & -5.517122e-05 & 0.67017458\\\\\n",
       "\t chr1 &  10001055 &  10001570 & PEX14       & Jurkat &  0.0162888946 & 0.013443455 &  2.504272e-03 & 0.15259244\\\\\n",
       "\t chr1 &  10001055 &  10001570 & PGD         & Jurkat & -0.0005149352 & 0.013118354 &  2.411489e-03 & 0.58826284\\\\\n",
       "\t chr1 &  10001055 &  10001570 & PIK3CD      & Jurkat &  0.0151808859 & 0.013401102 &  3.841117e-03 & 0.19872566\\\\\n",
       "\t chr1 &  10001055 &  10001570 & SLC25A33    & Jurkat & -0.0021572634 & 0.014036426 &  3.328312e-03 & 0.65203114\\\\\n",
       "\t chr1 &  10001055 &  10001570 & SPSB1       & Jurkat & -0.0008457267 & 0.017602230 &  2.974110e-03 & 0.58589921\\\\\n",
       "\t chr1 &  10001055 &  10001570 & TMEM201     & Jurkat &  0.0149095049 & 0.013324155 &  1.553319e-03 & 0.15807425\\\\\n",
       "\t chr1 &  10001055 &  10001570 & UBE4B       & Jurkat &  0.0169501125 & 0.012241596 & -5.240791e-04 & 0.07672600\\\\\n",
       "\t chr1 & 100037799 & 100039086 & AGL         & Jurkat &  0.0075230643 & 0.013984123 &  6.665222e-03 & 0.47554262\\\\\n",
       "\t chr1 & 100037799 & 100039086 & CDC14A      & Jurkat &  0.0061538299 & 0.012993107 &  8.899091e-03 & 0.58366792\\\\\n",
       "\t chr1 & 100037799 & 100039086 & DBT         & Jurkat &  0.0103637288 & 0.014309521 &  7.772240e-03 & 0.42814360\\\\\n",
       "\t chr1 & 100037799 & 100039086 & DPH5        & Jurkat &  0.0230590370 & 0.014420749 & -1.051256e-03 & 0.04727036\\\\\n",
       "\t chr1 & 100037799 & 100039086 & EXTL2       & Jurkat &  0.0223085057 & 0.014093499 &  1.356398e-02 & 0.26747587\\\\\n",
       "\t chr1 & 100037799 & 100039086 & FRRS1       & Jurkat &  0.0192187427 & 0.012409335 &  6.481489e-03 & 0.15234561\\\\\n",
       "\t chr1 & 100037799 & 100039086 & MFSD14A     & Jurkat & -0.0081440005 & 0.014368065 &  9.300673e-04 & 0.73615762\\\\\n",
       "\t chr1 & 100037799 & 100039086 & RTCA        & Jurkat &  0.0155712767 & 0.015053319 &  6.233542e-03 & 0.26752662\\\\\n",
       "\t chr1 & 100037799 & 100039086 & RTCA-AS1    & Jurkat &  0.0088131761 & 0.013010994 &  7.623465e-03 & 0.46357192\\\\\n",
       "\t chr1 & 100037799 & 100039086 & SASS6       & Jurkat &  0.0062294533 & 0.016029459 &  6.523191e-03 & 0.50731015\\\\\n",
       "\t chr1 & 100037799 & 100039086 & SLC30A7     & Jurkat &  0.0029814707 & 0.013460286 &  7.207882e-03 & 0.62323615\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &   978694 &   979118 & IL3RA     & Jurkat & -0.0114866278 & 0.01180815 & -1.632231e-03 & 0.79801226\\\\\n",
       "\t chrX &   978694 &   979118 & LINC00106 & Jurkat &  0.0055581009 & 0.01510983 &  1.448196e-03 & 0.39281023\\\\\n",
       "\t chrX &   978694 &   979118 & P2RY8     & Jurkat &  0.0131535710 & 0.01301975 &  1.549212e-03 & 0.18638707\\\\\n",
       "\t chrX &   978694 &   979118 & PLCXD1    & Jurkat &  0.0083694509 & 0.01523631 &  4.252237e-03 & 0.39349404\\\\\n",
       "\t chrX &   978694 &   979118 & PPP2R3B   & Jurkat & -0.0092444213 & 0.01373220 &  2.057085e-03 & 0.79474412\\\\\n",
       "\t chrX &   978694 &   979118 & SHOX      & Jurkat & -0.0029084531 & 0.01644049 &  1.525765e-03 & 0.60630956\\\\\n",
       "\t chrX &   978694 &   979118 & SLC25A6   & Jurkat & -0.0101946609 & 0.01486111 &  2.546487e-03 & 0.80437383\\\\\n",
       "\t chrX &  9842372 &  9843073 & TBL1X     & Jurkat & -0.0068704126 & 0.01525434 &  8.456191e-04 & 0.69351041\\\\\n",
       "\t chrX &  9842372 &  9843073 & WWC3      & Jurkat & -0.0269497814 & 0.01507101 &  3.371021e-03 & 0.97788277\\\\\n",
       "\t chrX &  9873846 &  9874297 & TBL1X     & Jurkat & -0.0207520878 & 0.01379675 &  2.788075e-03 & 0.95601552\\\\\n",
       "\t chrX &  9873846 &  9874297 & WWC3      & Jurkat & -0.0018099022 & 0.01317112 &  1.838355e-03 & 0.60910581\\\\\n",
       "\t chrX &  9885186 &  9885487 & TBL1X     & Jurkat &  0.0162982433 & 0.01313418 & -1.695516e-04 & 0.10495515\\\\\n",
       "\t chrX &  9885186 &  9885487 & WWC3      & Jurkat &  0.0032578932 & 0.01416270 &  2.684010e-03 & 0.48383899\\\\\n",
       "\t chrX &   989869 &   990628 & AKAP17A   & Jurkat &  0.0009846862 & 0.01358493 &  1.016017e-02 & 0.75029448\\\\\n",
       "\t chrX &   989869 &   990628 & ASMT      & Jurkat &  0.0235034382 & 0.01244424 &  6.322977e-05 & 0.02980831\\\\\n",
       "\t chrX &   989869 &   990628 & ASMTL     & Jurkat &  0.0019765367 & 0.01372576 &  1.847555e-03 & 0.49625117\\\\\n",
       "\t chrX &   989869 &   990628 & GTPBP6    & Jurkat &  0.0213985047 & 0.01314494 &  6.783700e-03 & 0.13310782\\\\\n",
       "\t chrX &   989869 &   990628 & IL3RA     & Jurkat &  0.0252265799 & 0.01450229 &  4.717340e-03 & 0.07865053\\\\\n",
       "\t chrX &   989869 &   990628 & LINC00106 & Jurkat &  0.0089426480 & 0.01196052 &  2.099447e-03 & 0.28361044\\\\\n",
       "\t chrX &   989869 &   990628 & P2RY8     & Jurkat & -0.0090354619 & 0.01333771 &  7.675060e-03 & 0.89487488\\\\\n",
       "\t chrX &   989869 &   990628 & PLCXD1    & Jurkat &  0.0255621263 & 0.01511793 &  1.085956e-02 & 0.16539481\\\\\n",
       "\t chrX &   989869 &   990628 & PPP2R3B   & Jurkat &  0.0066262678 & 0.01454721 &  6.960481e-03 & 0.50916465\\\\\n",
       "\t chrX &   989869 &   990628 & SHOX      & Jurkat & -0.0004223613 & 0.01454376 &  1.260592e-03 & 0.54606140\\\\\n",
       "\t chrX &   989869 &   990628 & SLC25A6   & Jurkat & -0.0098100685 & 0.01206209 &  3.043785e-03 & 0.85670698\\\\\n",
       "\t chrX &  9941883 &  9942743 & TBL1X     & Jurkat &  0.0105512284 & 0.01400607 &  2.866355e-05 & 0.22624024\\\\\n",
       "\t chrX &  9941883 &  9942743 & WWC3      & Jurkat &  0.0098070060 & 0.01490330 &  4.247553e-03 & 0.35456119\\\\\n",
       "\t chrX & 99603024 & 99603776 & PCDH19    & Jurkat & -0.0120777939 & 0.01535607 &  2.882824e-03 & 0.83503325\\\\\n",
       "\t chrX & 99632013 & 99632219 & PCDH19    & Jurkat & -0.0070325259 & 0.01545406 &  5.107510e-04 & 0.68726393\\\\\n",
       "\t chrX &  9995829 &  9996207 & TBL1X     & Jurkat & -0.0074744363 & 0.01551003 & -3.743486e-04 & 0.67644261\\\\\n",
       "\t chrX &  9995829 &  9996207 & WWC3      & Jurkat &  0.0114631748 & 0.01555261 &  2.561411e-03 & 0.28353750\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2590409 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | rObs &lt;dbl&gt; | rBgSD &lt;dbl&gt; | rBgMean &lt;dbl&gt; | pvalZ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  10001055 |  10001570 | APITD1-CORT | Jurkat | -0.0071411282 | 0.014130634 | -3.175940e-04 | 0.68541289 |\n",
       "| chr1 |  10001055 |  10001570 | CASZ1       | Jurkat | -0.0022533046 | 0.015386062 |  3.392746e-03 | 0.64317512 |\n",
       "| chr1 |  10001055 |  10001570 | CLSTN1      | Jurkat |  0.0197653646 | 0.013797346 |  4.698342e-04 | 0.08098212 |\n",
       "| chr1 |  10001055 |  10001570 | CTNNBIP1    | Jurkat | -0.0277141855 | 0.014525003 |  2.252927e-03 | 0.98045033 |\n",
       "| chr1 |  10001055 |  10001570 | DFFA        | Jurkat | -0.0185396358 | 0.014258860 | -5.241866e-04 | 0.89678741 |\n",
       "| chr1 |  10001055 |  10001570 | GPR157      | Jurkat |  0.0148254118 | 0.013936212 |  1.939612e-03 | 0.17757994 |\n",
       "| chr1 |  10001055 |  10001570 | H6PD        | Jurkat | -0.0050693443 | 0.014255930 |  8.466787e-04 | 0.66092423 |\n",
       "| chr1 |  10001055 |  10001570 | KIF1B       | Jurkat |  0.0107859814 | 0.013129322 | -1.285664e-05 | 0.20539682 |\n",
       "| chr1 |  10001055 |  10001570 | LINC01759   | Jurkat | -0.0019488251 | 0.008548509 | -1.129132e-03 | 0.53819492 |\n",
       "| chr1 |  10001055 |  10001570 | LZIC        | Jurkat | -0.0236758420 | 0.014126920 |  1.161149e-03 | 0.96063748 |\n",
       "| chr1 |  10001055 |  10001570 | MIR34AHG    | Jurkat | -0.0033773843 | 0.014595356 |  5.302928e-04 | 0.60554807 |\n",
       "| chr1 |  10001055 |  10001570 | NMNAT1      | Jurkat | -0.0064528480 | 0.014527124 | -5.517122e-05 | 0.67017458 |\n",
       "| chr1 |  10001055 |  10001570 | PEX14       | Jurkat |  0.0162888946 | 0.013443455 |  2.504272e-03 | 0.15259244 |\n",
       "| chr1 |  10001055 |  10001570 | PGD         | Jurkat | -0.0005149352 | 0.013118354 |  2.411489e-03 | 0.58826284 |\n",
       "| chr1 |  10001055 |  10001570 | PIK3CD      | Jurkat |  0.0151808859 | 0.013401102 |  3.841117e-03 | 0.19872566 |\n",
       "| chr1 |  10001055 |  10001570 | SLC25A33    | Jurkat | -0.0021572634 | 0.014036426 |  3.328312e-03 | 0.65203114 |\n",
       "| chr1 |  10001055 |  10001570 | SPSB1       | Jurkat | -0.0008457267 | 0.017602230 |  2.974110e-03 | 0.58589921 |\n",
       "| chr1 |  10001055 |  10001570 | TMEM201     | Jurkat |  0.0149095049 | 0.013324155 |  1.553319e-03 | 0.15807425 |\n",
       "| chr1 |  10001055 |  10001570 | UBE4B       | Jurkat |  0.0169501125 | 0.012241596 | -5.240791e-04 | 0.07672600 |\n",
       "| chr1 | 100037799 | 100039086 | AGL         | Jurkat |  0.0075230643 | 0.013984123 |  6.665222e-03 | 0.47554262 |\n",
       "| chr1 | 100037799 | 100039086 | CDC14A      | Jurkat |  0.0061538299 | 0.012993107 |  8.899091e-03 | 0.58366792 |\n",
       "| chr1 | 100037799 | 100039086 | DBT         | Jurkat |  0.0103637288 | 0.014309521 |  7.772240e-03 | 0.42814360 |\n",
       "| chr1 | 100037799 | 100039086 | DPH5        | Jurkat |  0.0230590370 | 0.014420749 | -1.051256e-03 | 0.04727036 |\n",
       "| chr1 | 100037799 | 100039086 | EXTL2       | Jurkat |  0.0223085057 | 0.014093499 |  1.356398e-02 | 0.26747587 |\n",
       "| chr1 | 100037799 | 100039086 | FRRS1       | Jurkat |  0.0192187427 | 0.012409335 |  6.481489e-03 | 0.15234561 |\n",
       "| chr1 | 100037799 | 100039086 | MFSD14A     | Jurkat | -0.0081440005 | 0.014368065 |  9.300673e-04 | 0.73615762 |\n",
       "| chr1 | 100037799 | 100039086 | RTCA        | Jurkat |  0.0155712767 | 0.015053319 |  6.233542e-03 | 0.26752662 |\n",
       "| chr1 | 100037799 | 100039086 | RTCA-AS1    | Jurkat |  0.0088131761 | 0.013010994 |  7.623465e-03 | 0.46357192 |\n",
       "| chr1 | 100037799 | 100039086 | SASS6       | Jurkat |  0.0062294533 | 0.016029459 |  6.523191e-03 | 0.50731015 |\n",
       "| chr1 | 100037799 | 100039086 | SLC30A7     | Jurkat |  0.0029814707 | 0.013460286 |  7.207882e-03 | 0.62323615 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |   978694 |   979118 | IL3RA     | Jurkat | -0.0114866278 | 0.01180815 | -1.632231e-03 | 0.79801226 |\n",
       "| chrX |   978694 |   979118 | LINC00106 | Jurkat |  0.0055581009 | 0.01510983 |  1.448196e-03 | 0.39281023 |\n",
       "| chrX |   978694 |   979118 | P2RY8     | Jurkat |  0.0131535710 | 0.01301975 |  1.549212e-03 | 0.18638707 |\n",
       "| chrX |   978694 |   979118 | PLCXD1    | Jurkat |  0.0083694509 | 0.01523631 |  4.252237e-03 | 0.39349404 |\n",
       "| chrX |   978694 |   979118 | PPP2R3B   | Jurkat | -0.0092444213 | 0.01373220 |  2.057085e-03 | 0.79474412 |\n",
       "| chrX |   978694 |   979118 | SHOX      | Jurkat | -0.0029084531 | 0.01644049 |  1.525765e-03 | 0.60630956 |\n",
       "| chrX |   978694 |   979118 | SLC25A6   | Jurkat | -0.0101946609 | 0.01486111 |  2.546487e-03 | 0.80437383 |\n",
       "| chrX |  9842372 |  9843073 | TBL1X     | Jurkat | -0.0068704126 | 0.01525434 |  8.456191e-04 | 0.69351041 |\n",
       "| chrX |  9842372 |  9843073 | WWC3      | Jurkat | -0.0269497814 | 0.01507101 |  3.371021e-03 | 0.97788277 |\n",
       "| chrX |  9873846 |  9874297 | TBL1X     | Jurkat | -0.0207520878 | 0.01379675 |  2.788075e-03 | 0.95601552 |\n",
       "| chrX |  9873846 |  9874297 | WWC3      | Jurkat | -0.0018099022 | 0.01317112 |  1.838355e-03 | 0.60910581 |\n",
       "| chrX |  9885186 |  9885487 | TBL1X     | Jurkat |  0.0162982433 | 0.01313418 | -1.695516e-04 | 0.10495515 |\n",
       "| chrX |  9885186 |  9885487 | WWC3      | Jurkat |  0.0032578932 | 0.01416270 |  2.684010e-03 | 0.48383899 |\n",
       "| chrX |   989869 |   990628 | AKAP17A   | Jurkat |  0.0009846862 | 0.01358493 |  1.016017e-02 | 0.75029448 |\n",
       "| chrX |   989869 |   990628 | ASMT      | Jurkat |  0.0235034382 | 0.01244424 |  6.322977e-05 | 0.02980831 |\n",
       "| chrX |   989869 |   990628 | ASMTL     | Jurkat |  0.0019765367 | 0.01372576 |  1.847555e-03 | 0.49625117 |\n",
       "| chrX |   989869 |   990628 | GTPBP6    | Jurkat |  0.0213985047 | 0.01314494 |  6.783700e-03 | 0.13310782 |\n",
       "| chrX |   989869 |   990628 | IL3RA     | Jurkat |  0.0252265799 | 0.01450229 |  4.717340e-03 | 0.07865053 |\n",
       "| chrX |   989869 |   990628 | LINC00106 | Jurkat |  0.0089426480 | 0.01196052 |  2.099447e-03 | 0.28361044 |\n",
       "| chrX |   989869 |   990628 | P2RY8     | Jurkat | -0.0090354619 | 0.01333771 |  7.675060e-03 | 0.89487488 |\n",
       "| chrX |   989869 |   990628 | PLCXD1    | Jurkat |  0.0255621263 | 0.01511793 |  1.085956e-02 | 0.16539481 |\n",
       "| chrX |   989869 |   990628 | PPP2R3B   | Jurkat |  0.0066262678 | 0.01454721 |  6.960481e-03 | 0.50916465 |\n",
       "| chrX |   989869 |   990628 | SHOX      | Jurkat | -0.0004223613 | 0.01454376 |  1.260592e-03 | 0.54606140 |\n",
       "| chrX |   989869 |   990628 | SLC25A6   | Jurkat | -0.0098100685 | 0.01206209 |  3.043785e-03 | 0.85670698 |\n",
       "| chrX |  9941883 |  9942743 | TBL1X     | Jurkat |  0.0105512284 | 0.01400607 |  2.866355e-05 | 0.22624024 |\n",
       "| chrX |  9941883 |  9942743 | WWC3      | Jurkat |  0.0098070060 | 0.01490330 |  4.247553e-03 | 0.35456119 |\n",
       "| chrX | 99603024 | 99603776 | PCDH19    | Jurkat | -0.0120777939 | 0.01535607 |  2.882824e-03 | 0.83503325 |\n",
       "| chrX | 99632013 | 99632219 | PCDH19    | Jurkat | -0.0070325259 | 0.01545406 |  5.107510e-04 | 0.68726393 |\n",
       "| chrX |  9995829 |  9996207 | TBL1X     | Jurkat | -0.0074744363 | 0.01551003 | -3.743486e-04 | 0.67644261 |\n",
       "| chrX |  9995829 |  9996207 | WWC3      | Jurkat |  0.0114631748 | 0.01555261 |  2.561411e-03 | 0.28353750 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene  CellType rObs          rBgSD      \n",
       "1       chr1  10001055  10001570 APITD1-CORT Jurkat   -0.0071411282 0.014130634\n",
       "2       chr1  10001055  10001570 CASZ1       Jurkat   -0.0022533046 0.015386062\n",
       "3       chr1  10001055  10001570 CLSTN1      Jurkat    0.0197653646 0.013797346\n",
       "4       chr1  10001055  10001570 CTNNBIP1    Jurkat   -0.0277141855 0.014525003\n",
       "5       chr1  10001055  10001570 DFFA        Jurkat   -0.0185396358 0.014258860\n",
       "6       chr1  10001055  10001570 GPR157      Jurkat    0.0148254118 0.013936212\n",
       "7       chr1  10001055  10001570 H6PD        Jurkat   -0.0050693443 0.014255930\n",
       "8       chr1  10001055  10001570 KIF1B       Jurkat    0.0107859814 0.013129322\n",
       "9       chr1  10001055  10001570 LINC01759   Jurkat   -0.0019488251 0.008548509\n",
       "10      chr1  10001055  10001570 LZIC        Jurkat   -0.0236758420 0.014126920\n",
       "11      chr1  10001055  10001570 MIR34AHG    Jurkat   -0.0033773843 0.014595356\n",
       "12      chr1  10001055  10001570 NMNAT1      Jurkat   -0.0064528480 0.014527124\n",
       "13      chr1  10001055  10001570 PEX14       Jurkat    0.0162888946 0.013443455\n",
       "14      chr1  10001055  10001570 PGD         Jurkat   -0.0005149352 0.013118354\n",
       "15      chr1  10001055  10001570 PIK3CD      Jurkat    0.0151808859 0.013401102\n",
       "16      chr1  10001055  10001570 SLC25A33    Jurkat   -0.0021572634 0.014036426\n",
       "17      chr1  10001055  10001570 SPSB1       Jurkat   -0.0008457267 0.017602230\n",
       "18      chr1  10001055  10001570 TMEM201     Jurkat    0.0149095049 0.013324155\n",
       "19      chr1  10001055  10001570 UBE4B       Jurkat    0.0169501125 0.012241596\n",
       "20      chr1 100037799 100039086 AGL         Jurkat    0.0075230643 0.013984123\n",
       "21      chr1 100037799 100039086 CDC14A      Jurkat    0.0061538299 0.012993107\n",
       "22      chr1 100037799 100039086 DBT         Jurkat    0.0103637288 0.014309521\n",
       "23      chr1 100037799 100039086 DPH5        Jurkat    0.0230590370 0.014420749\n",
       "24      chr1 100037799 100039086 EXTL2       Jurkat    0.0223085057 0.014093499\n",
       "25      chr1 100037799 100039086 FRRS1       Jurkat    0.0192187427 0.012409335\n",
       "26      chr1 100037799 100039086 MFSD14A     Jurkat   -0.0081440005 0.014368065\n",
       "27      chr1 100037799 100039086 RTCA        Jurkat    0.0155712767 0.015053319\n",
       "28      chr1 100037799 100039086 RTCA-AS1    Jurkat    0.0088131761 0.013010994\n",
       "29      chr1 100037799 100039086 SASS6       Jurkat    0.0062294533 0.016029459\n",
       "30      chr1 100037799 100039086 SLC30A7     Jurkat    0.0029814707 0.013460286\n",
       "⋮       ⋮    ⋮         ⋮         ⋮           ⋮        ⋮             ⋮          \n",
       "2590380 chrX   978694    979118  IL3RA       Jurkat   -0.0114866278 0.01180815 \n",
       "2590381 chrX   978694    979118  LINC00106   Jurkat    0.0055581009 0.01510983 \n",
       "2590382 chrX   978694    979118  P2RY8       Jurkat    0.0131535710 0.01301975 \n",
       "2590383 chrX   978694    979118  PLCXD1      Jurkat    0.0083694509 0.01523631 \n",
       "2590384 chrX   978694    979118  PPP2R3B     Jurkat   -0.0092444213 0.01373220 \n",
       "2590385 chrX   978694    979118  SHOX        Jurkat   -0.0029084531 0.01644049 \n",
       "2590386 chrX   978694    979118  SLC25A6     Jurkat   -0.0101946609 0.01486111 \n",
       "2590387 chrX  9842372   9843073  TBL1X       Jurkat   -0.0068704126 0.01525434 \n",
       "2590388 chrX  9842372   9843073  WWC3        Jurkat   -0.0269497814 0.01507101 \n",
       "2590389 chrX  9873846   9874297  TBL1X       Jurkat   -0.0207520878 0.01379675 \n",
       "2590390 chrX  9873846   9874297  WWC3        Jurkat   -0.0018099022 0.01317112 \n",
       "2590391 chrX  9885186   9885487  TBL1X       Jurkat    0.0162982433 0.01313418 \n",
       "2590392 chrX  9885186   9885487  WWC3        Jurkat    0.0032578932 0.01416270 \n",
       "2590393 chrX   989869    990628  AKAP17A     Jurkat    0.0009846862 0.01358493 \n",
       "2590394 chrX   989869    990628  ASMT        Jurkat    0.0235034382 0.01244424 \n",
       "2590395 chrX   989869    990628  ASMTL       Jurkat    0.0019765367 0.01372576 \n",
       "2590396 chrX   989869    990628  GTPBP6      Jurkat    0.0213985047 0.01314494 \n",
       "2590397 chrX   989869    990628  IL3RA       Jurkat    0.0252265799 0.01450229 \n",
       "2590398 chrX   989869    990628  LINC00106   Jurkat    0.0089426480 0.01196052 \n",
       "2590399 chrX   989869    990628  P2RY8       Jurkat   -0.0090354619 0.01333771 \n",
       "2590400 chrX   989869    990628  PLCXD1      Jurkat    0.0255621263 0.01511793 \n",
       "2590401 chrX   989869    990628  PPP2R3B     Jurkat    0.0066262678 0.01454721 \n",
       "2590402 chrX   989869    990628  SHOX        Jurkat   -0.0004223613 0.01454376 \n",
       "2590403 chrX   989869    990628  SLC25A6     Jurkat   -0.0098100685 0.01206209 \n",
       "2590404 chrX  9941883   9942743  TBL1X       Jurkat    0.0105512284 0.01400607 \n",
       "2590405 chrX  9941883   9942743  WWC3        Jurkat    0.0098070060 0.01490330 \n",
       "2590406 chrX 99603024  99603776  PCDH19      Jurkat   -0.0120777939 0.01535607 \n",
       "2590407 chrX 99632013  99632219  PCDH19      Jurkat   -0.0070325259 0.01545406 \n",
       "2590408 chrX  9995829   9996207  TBL1X       Jurkat   -0.0074744363 0.01551003 \n",
       "2590409 chrX  9995829   9996207  WWC3        Jurkat    0.0114631748 0.01555261 \n",
       "        rBgMean       pvalZ     \n",
       "1       -3.175940e-04 0.68541289\n",
       "2        3.392746e-03 0.64317512\n",
       "3        4.698342e-04 0.08098212\n",
       "4        2.252927e-03 0.98045033\n",
       "5       -5.241866e-04 0.89678741\n",
       "6        1.939612e-03 0.17757994\n",
       "7        8.466787e-04 0.66092423\n",
       "8       -1.285664e-05 0.20539682\n",
       "9       -1.129132e-03 0.53819492\n",
       "10       1.161149e-03 0.96063748\n",
       "11       5.302928e-04 0.60554807\n",
       "12      -5.517122e-05 0.67017458\n",
       "13       2.504272e-03 0.15259244\n",
       "14       2.411489e-03 0.58826284\n",
       "15       3.841117e-03 0.19872566\n",
       "16       3.328312e-03 0.65203114\n",
       "17       2.974110e-03 0.58589921\n",
       "18       1.553319e-03 0.15807425\n",
       "19      -5.240791e-04 0.07672600\n",
       "20       6.665222e-03 0.47554262\n",
       "21       8.899091e-03 0.58366792\n",
       "22       7.772240e-03 0.42814360\n",
       "23      -1.051256e-03 0.04727036\n",
       "24       1.356398e-02 0.26747587\n",
       "25       6.481489e-03 0.15234561\n",
       "26       9.300673e-04 0.73615762\n",
       "27       6.233542e-03 0.26752662\n",
       "28       7.623465e-03 0.46357192\n",
       "29       6.523191e-03 0.50731015\n",
       "30       7.207882e-03 0.62323615\n",
       "⋮       ⋮             ⋮         \n",
       "2590380 -1.632231e-03 0.79801226\n",
       "2590381  1.448196e-03 0.39281023\n",
       "2590382  1.549212e-03 0.18638707\n",
       "2590383  4.252237e-03 0.39349404\n",
       "2590384  2.057085e-03 0.79474412\n",
       "2590385  1.525765e-03 0.60630956\n",
       "2590386  2.546487e-03 0.80437383\n",
       "2590387  8.456191e-04 0.69351041\n",
       "2590388  3.371021e-03 0.97788277\n",
       "2590389  2.788075e-03 0.95601552\n",
       "2590390  1.838355e-03 0.60910581\n",
       "2590391 -1.695516e-04 0.10495515\n",
       "2590392  2.684010e-03 0.48383899\n",
       "2590393  1.016017e-02 0.75029448\n",
       "2590394  6.322977e-05 0.02980831\n",
       "2590395  1.847555e-03 0.49625117\n",
       "2590396  6.783700e-03 0.13310782\n",
       "2590397  4.717340e-03 0.07865053\n",
       "2590398  2.099447e-03 0.28361044\n",
       "2590399  7.675060e-03 0.89487488\n",
       "2590400  1.085956e-02 0.16539481\n",
       "2590401  6.960481e-03 0.50916465\n",
       "2590402  1.260592e-03 0.54606140\n",
       "2590403  3.043785e-03 0.85670698\n",
       "2590404  2.866355e-05 0.22624024\n",
       "2590405  4.247553e-03 0.35456119\n",
       "2590406  2.882824e-03 0.83503325\n",
       "2590407  5.107510e-04 0.68726393\n",
       "2590408 -3.743486e-04 0.67644261\n",
       "2590409  2.561411e-03 0.28353750"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = celltype\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"rObs\",\n",
    "                         \"rBgSD\",\n",
    "                         \"rBgMean\",\n",
    "                         \"pvalZ\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4d77c58e-215a-408d-8495-c68c72f559c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  grid      stats4    stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbmcapply_1.5.1                   data.table_1.15.2                \n",
       " [3] foreach_1.5.2                     BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] genomation_1.34.0                 Seurat_5.0.3                     \n",
       "[11] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[13] FigR_0.1.0                        motifmatchr_1.24.0               \n",
       "[15] chromVAR_1.24.0                   cowplot_1.1.3                    \n",
       "[17] ggplot2_3.5.1                     dplyr_1.1.4                      \n",
       "[19] SummarizedExperiment_1.30.2       Biobase_2.60.0                   \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               MatrixGenerics_1.12.3            \n",
       "[27] matrixStats_1.3.0                 Matrix_1.6-5                     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.3              \n",
       "  [3] later_1.3.2                 pbdZMQ_0.3-11              \n",
       "  [5] BiocIO_1.10.0               bitops_1.0-7               \n",
       "  [7] tibble_3.2.1                R.oo_1.26.0                \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16.1            \n",
       " [11] DirichletMultinomial_1.44.0 fastDummies_1.7.3          \n",
       " [13] lifecycle_1.0.4             vroom_1.6.5                \n",
       " [15] globals_0.16.3              lattice_0.22-6             \n",
       " [17] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [19] plotly_4.10.4               plotrix_3.8-4              \n",
       " [21] yaml_2.3.8                  httpuv_1.6.15              \n",
       " [23] sctransform_0.4.1           spam_2.10-0                \n",
       " [25] spatstat.sparse_3.0-3       reticulate_1.36.1          \n",
       " [27] pbapply_1.7-2               DBI_1.2.2                  \n",
       " [29] CNEr_1.38.0                 RColorBrewer_1.1-3         \n",
       " [31] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [33] Rtsne_0.17                  purrr_1.0.2                \n",
       " [35] R.utils_2.12.3              RCurl_1.98-1.14            \n",
       " [37] pracma_2.4.4                GenomeInfoDbData_1.2.11    \n",
       " [39] ggrepel_0.9.5               irlba_2.3.5.1              \n",
       " [41] spatstat.utils_3.0-4        listenv_0.9.1              \n",
       " [43] seqLogo_1.68.0              goftest_1.2-3              \n",
       " [45] RSpectra_0.16-1             spatstat.random_3.2-3      \n",
       " [47] annotate_1.80.0             fitdistrplus_1.1-11        \n",
       " [49] parallelly_1.37.1           leiden_0.4.3.1             \n",
       " [51] codetools_0.2-20            DelayedArray_0.26.7        \n",
       " [53] DT_0.33                     tidyselect_1.2.1           \n",
       " [55] spatstat.explore_3.2-6      base64enc_0.1-3            \n",
       " [57] GenomicAlignments_1.36.0    jsonlite_1.8.8             \n",
       " [59] progressr_0.14.0            iterators_1.0.14           \n",
       " [61] ggridges_0.5.6              survival_3.6-4             \n",
       " [63] tools_4.3.3                 TFMPvalue_0.0.9            \n",
       " [65] ica_1.0-3                   Rcpp_1.0.12                \n",
       " [67] glue_1.7.0                  gridExtra_2.3              \n",
       " [69] IRdisplay_1.1               withr_3.0.0                \n",
       " [71] fastmap_1.1.1               fansi_1.0.6                \n",
       " [73] caTools_1.18.2              digest_0.6.35              \n",
       " [75] R6_2.5.1                    mime_0.12                  \n",
       " [77] nabor_0.5.0                 seqPattern_1.34.0          \n",
       " [79] colorspace_2.1-0            scattermore_1.2            \n",
       " [81] GO.db_3.18.0                tensor_1.5                 \n",
       " [83] gtools_3.9.5                poweRlaw_0.80.0            \n",
       " [85] spatstat.data_3.0-4         RSQLite_2.3.6              \n",
       " [87] R.methodsS3_1.8.2           utf8_1.2.4                 \n",
       " [89] tidyr_1.3.1                 generics_0.1.3             \n",
       " [91] httr_1.4.7                  htmlwidgets_1.6.4          \n",
       " [93] S4Arrays_1.2.1              TFBSTools_1.40.0           \n",
       " [95] uwot_0.1.16                 pkgconfig_2.0.3            \n",
       " [97] gtable_0.3.5                blob_1.2.4                 \n",
       " [99] impute_1.76.0               lmtest_0.9-40              \n",
       "[101] htmltools_0.5.8.1           dotCall64_1.1-1            \n",
       "[103] scales_1.3.0                png_0.1-8                  \n",
       "[105] tzdb_0.4.0                  reshape2_1.4.4             \n",
       "[107] rjson_0.2.21                uuid_1.2-0                 \n",
       "[109] nlme_3.1-164                repr_1.1.7                 \n",
       "[111] cachem_1.0.8                zoo_1.8-12                 \n",
       "[113] stringr_1.5.1               KernSmooth_2.23-22         \n",
       "[115] miniUI_0.1.1.1              AnnotationDbi_1.64.1       \n",
       "[117] restfulr_0.0.15             pillar_1.9.0               \n",
       "[119] vctrs_0.6.5                 RANN_2.6.1                 \n",
       "[121] promises_1.3.0              xtable_1.8-4               \n",
       "[123] cluster_2.1.6               evaluate_0.23              \n",
       "[125] readr_2.1.5                 cli_3.6.2                  \n",
       "[127] compiler_4.3.3              Rsamtools_2.18.0           \n",
       "[129] rlang_1.1.3                 crayon_1.5.2               \n",
       "[131] future.apply_1.11.2         plyr_1.8.9                 \n",
       "[133] stringi_1.8.3               gridBase_0.4-7             \n",
       "[135] deldir_2.0-4                viridisLite_0.4.2          \n",
       "[137] BiocParallel_1.36.0         munsell_0.5.1              \n",
       "[139] lazyeval_0.2.2              spatstat.geom_3.2-9        \n",
       "[141] IRkernel_1.3.2              RcppHNSW_0.6.0             \n",
       "[143] hms_1.1.3                   patchwork_1.2.0            \n",
       "[145] bit64_4.0.5                 future_1.33.2              \n",
       "[147] KEGGREST_1.40.1             shiny_1.8.1.1              \n",
       "[149] ROCR_1.0-11                 igraph_2.0.3               \n",
       "[151] memoise_2.0.1               bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1a0283-11d6-4df1-bbd8-7b47e9e8ead8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
