{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8466b35-cae5-4f62-8138-37b2de484a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.2\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : grid v4.3.3\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "\tLoading Package : gtools v3.9.5\n",
      "\n",
      "\tLoading Package : gtable v0.3.5\n",
      "\n",
      "\tLoading Package : ggplot2 v3.5.1\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "\tLoading Package : plyr v1.8.9\n",
      "\n",
      "\tLoading Package : stringr v1.5.1\n",
      "\n",
      "\tLoading Package : data.table v1.15.2\n",
      "\n",
      "\tLoading Package : matrixStats v1.3.0\n",
      "\n",
      "\tLoading Package : S4Vectors v0.40.2\n",
      "\n",
      "\tLoading Package : GenomicRanges v1.54.1\n",
      "\n",
      "\tLoading Package : BiocGenerics v0.48.1\n",
      "\n",
      "\tLoading Package : Matrix v1.6.5\n",
      "\n",
      "\tLoading Package : Rcpp v1.0.12\n",
      "\n",
      "\tLoading Package : SummarizedExperiment v1.30.2\n",
      "\n",
      "\tLoading Package : rhdf5 v2.44.0\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(magrittr)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(ggplot2)\n",
    "library(dplyr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e85020d7-d358-492d-9358-c81270f4b624",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1bec6690-52ec-450d-bba7-7d4528a185cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "TSS_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.bed\"\n",
    "path.atac_frag = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/scE2G_input_Maya/HCT116/atac_fragments.tsv.gz\"\n",
    "path.matrix.rna_count = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/scE2G_input_Maya/HCT116/rna_count_matrix.csv.gz\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/250319/scE2G/test/results/HCT116/HCT116/Kendall/Pairs.tsv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/HCT116/1.Genome_wide_prediction/ArchR/ArchR/\"\n",
    "celltype = \"HCT116\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "340dac32-44ac-4231-b964-b5f3b9e2dbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16\n",
    "maxDist = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da973385-4c81-4b1c-9669-cf5933b23a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 16.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRThreads(threads = n.cores) \n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86be3e66-c954-4620-9e64-c6a1c0fcb84c",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6f363969-bf0b-45ce-858c-a8cb64cc210a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1ff4f07-c695-45a1-bf15-dc523d20497f",
   "metadata": {},
   "source": [
    "Import RNA matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2a104a7a-b6b9-4315-8b15-6a22c48e142a",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna = read.csv(path.matrix.rna_count,\n",
    "                      row.names = 1,\n",
    "                      check.names = F)\n",
    "matrix.rna = Matrix(as.matrix(matrix.rna), sparse = TRUE)\n",
    "# matrix.rna = matrix.rna[,colnames(matrix.atac)]\n",
    "matrix.rna = matrix.rna[rowSums(matrix.rna) > 0,]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "287f8d61-9ad5-4f60-97b6-9e7b71795042",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ad6e79b8-2429-425a-ac78-6773abe4dbc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6ddc5a2b-371a-435d-9250-751ecffa65c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna.rename = map_gene_names(matrix.rna,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65a5d8fc-a99f-481c-9eec-8e88c8875dd3",
   "metadata": {},
   "source": [
    "make TSS GRanges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "815b37df-d305-4ba7-bd48-2270290ca3b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tss <- fread(TSS_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\"))\n",
    "df.tss = as.data.frame(df.tss)\n",
    "rownames(df.tss) = df.tss$name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fce150b6-ee55-46d7-9035-cd9698b5ee9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tss[df.tss$strand == \"+\",\"end\"] = df.tss[df.tss$strand == \"+\",\"start\"]\n",
    "df.tss[df.tss$strand == \"-\",\"start\"] = df.tss[df.tss$strand == \"-\",\"end\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7d73721a-8ec9-4b65-91b5-29cca6872927",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tss = df.tss[unname(rownames(matrix.rna.rename)),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a5a5da89-e7fd-4ffb-b477-48beff459909",
   "metadata": {},
   "outputs": [],
   "source": [
    "gr.tss <- makeGRangesFromDataFrame(\n",
    "  df.tss,\n",
    "  seqnames.field = \"chr\",\n",
    "  start.field = \"start\",\n",
    "  end.field = \"end\",\n",
    "  strand.field = \"strand\",    \n",
    "  keep.extra.columns = TRUE   \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a2259f88-32ce-4299-9592-b39c2273f03e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 18507 ranges and 4 metadata columns:\n",
       "            seqnames    ranges strand |        name     score      Ensembl_ID\n",
       "               <Rle> <IRanges>  <Rle> | <character> <integer>     <character>\n",
       "     FAM87B     chr1    817370      + |      FAM87B         0 ENSG00000177757\n",
       "  LINC00115     chr1    827522      - |   LINC00115         0 ENSG00000225880\n",
       "  LINC01128     chr1    827590      + |   LINC01128         0 ENSG00000228794\n",
       "     FAM41C     chr1    876802      - |      FAM41C         0 ENSG00000230368\n",
       "     SAMD11     chr1    925740      + |      SAMD11         0 ENSG00000187634\n",
       "        ...      ...       ...    ... .         ...       ...             ...\n",
       "     NLGN4Y     chrY  14523745      + |      NLGN4Y         0 ENSG00000165246\n",
       "     TTTY14     chrY  19077547      - |      TTTY14         0 ENSG00000176728\n",
       "      KDM5D     chrY  19744939      - |       KDM5D         0 ENSG00000012817\n",
       "     TTTY10     chrY  20519228      - |      TTTY10         0 ENSG00000229236\n",
       "     EIF1AY     chrY  20575710      + |      EIF1AY         0 ENSG00000198692\n",
       "                       gene_type\n",
       "                     <character>\n",
       "     FAM87B              lincRNA\n",
       "  LINC00115              lincRNA\n",
       "  LINC01128 processed_transcript\n",
       "     FAM41C              lincRNA\n",
       "     SAMD11       protein_coding\n",
       "        ...                  ...\n",
       "     NLGN4Y       protein_coding\n",
       "     TTTY14              lincRNA\n",
       "      KDM5D       protein_coding\n",
       "     TTTY10              lincRNA\n",
       "     EIF1AY       protein_coding\n",
       "  -------\n",
       "  seqinfo: 24 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gr.tss"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3538ffdc-4fed-48eb-9c9d-86805a91cf06",
   "metadata": {},
   "source": [
    "Create an ArchRProject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1753ffaa-89e2-43bf-bcf6-19937782d63a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-createArrows-273ebf78442f20-Date-2025-03-22_Time-16-25-08.50193.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Cleaning Temporary Files\n",
      "\n",
      "2025-03-22 16:25:09.042546 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "(HCT116 : 1 of 1) Checking if completed file exists!\n",
      "\n",
      "2025-03-22 16:25:09.107715 : (HCT116 : 1 of 1) Arrow Exists! Overriding since force = TRUE!, 0.001 mins elapsed.\n",
      "\n",
      "(HCT116 : 1 of 1) Determining Arrow Method to use!\n",
      "\n",
      "2025-03-22 16:25:09.262686 : (HCT116 : 1 of 1) Reading In Fragments from inputFiles (readMethod = tabix), 0.004 mins elapsed.\n",
      "\n",
      "2025-03-22 16:25:09.272751 : (HCT116 : 1 of 1) Tabix Bed To Temporary File, 0.004 mins elapsed.\n",
      "\n",
      "2025-03-22 16:26:24.357439 : (HCT116 : 1 of 1) Successful creation of Temporary File, 1.255 mins elapsed.\n",
      "\n",
      "2025-03-22 16:26:24.367574 : (HCT116 : 1 of 1) Creating ArrowFile From Temporary File, 1.255 mins elapsed.\n",
      "\n",
      "2025-03-22 16:26:52.594409 : (HCT116 : 1 of 1) Successful creation of Arrow File, 1.726 mins elapsed.\n",
      "\n",
      "2025-03-22 16:27:06.379524 : Continuing through after error ggplot for Fragment Size Distribution, 1.956 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-03-22 16:27:35.691829 : (HCT116 : 1 of 1) CellStats : Number of Cells Pass Filter = 4820 , 2.444 mins elapsed.\n",
      "\n",
      "2025-03-22 16:27:35.696581 : (HCT116 : 1 of 1) CellStats : Median Frags = 7478.5 , 2.444 mins elapsed.\n",
      "\n",
      "2025-03-22 16:27:35.702266 : (HCT116 : 1 of 1) CellStats : Median TSS Enrichment = 13.9535 , 2.444 mins elapsed.\n",
      "\n",
      "2025-03-22 16:27:36.722388 : Continuing through after error ggplot for TSS by Frags, 2.461 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-03-22 16:27:40.032605 : (HCT116 : 1 of 1) Adding Additional Feature Counts!, 2.517 mins elapsed.\n",
      "\n",
      "2025-03-22 16:28:02.351323 : (HCT116 : 1 of 1) Removing Fragments from Filtered Cells, 2.888 mins elapsed.\n",
      "\n",
      "2025-03-22 16:28:02.374692 : (HCT116 : 1 of 1) Creating Filtered Arrow File, 2.889 mins elapsed.\n",
      "\n",
      "2025-03-22 16:29:28.089848 : (HCT116 : 1 of 1) Finished Constructing Filtered Arrow File!, 4.317 mins elapsed.\n",
      "\n",
      "2025-03-22 16:29:28.245448 : (HCT116 : 1 of 1) Adding TileMatrix!, 4.32 mins elapsed.\n",
      "\n",
      "2025-03-22 16:33:24.397411 : (HCT116 : 1 of 1) Finished Creating Arrow File, 8.256 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-createArrows-273ebf78442f20-Date-2025-03-22_Time-16-25-08.50193.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ArrowFiles <- createArrowFiles(\n",
    "  inputFiles = c(path.atac_frag),\n",
    "  sampleNames = c(celltype),\n",
    "  validBarcodes = list(HCT116 = colnames(matrix.rna.rename)),\n",
    "  addTileMat = T,\n",
    "  addGeneScoreMat = F,\n",
    "  force = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ada3db45-d050-4659-a7cf-2891e04b1249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'HCT116.arrow'"
      ],
      "text/latex": [
       "'HCT116.arrow'"
      ],
      "text/markdown": [
       "'HCT116.arrow'"
      ],
      "text/plain": [
       "[1] \"HCT116.arrow\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ArrowFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5ba28283-d488-4982-8d4c-745fc3f59365",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(paste(dir.output, \"ArchR\", sep = \"/\"), recursive = T):\n",
      "“'/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/HCT116/1.Genome_wide_prediction/ArchR/ArchR//ArchR' already exists”\n"
     ]
    }
   ],
   "source": [
    "dir.create(paste(dir.output,\"ArchR\",sep=\"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "aa217507-8323-43bf-b23c-0ec25ee33c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Validating Arrows...\n",
      "\n",
      "Getting SampleNames...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Copying ArrowFiles to Ouptut Directory! If you want to save disk space set copyArrows = FALSE\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Getting Cell Metadata...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Merging Cell Metadata...\n",
      "\n",
      "Initializing ArchRProject...\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR <- ArchRProject(\n",
    "  ArrowFiles = ArrowFiles, \n",
    "  outputDirectory = paste(dir.output,\"ArchR\",sep=\"/\"),\n",
    "  copyArrows = TRUE \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f9795403-545f-4385-aa38-2c70b1b2fb21",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: ArchRProject \n",
       "outputDirectory: /maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/HCT116/1.Genome_wide_prediction/ArchR/ArchR/ArchR \n",
       "samples(1): HCT116\n",
       "sampleColData names(1): ArrowFiles\n",
       "cellColData names(13): Sample TSSEnrichment ... nDiFrags BlacklistRatio\n",
       "numberOfCells(1): 4820\n",
       "medianTSS(1): 13.9535\n",
       "medianFrags(1): 7478.5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obj.ArchR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ee876dd5-0b22-46fd-9015-7e22144c5024",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking Inputs...\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addIterativeLSI-273ebf9b4b9e-Date-2025-03-22_Time-16-33-35.300146.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-03-22 16:33:35.853881 : Computing Total Across All Features, 0 mins elapsed.\n",
      "\n",
      "2025-03-22 16:33:36.945175 : Computing Top Features, 0.019 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-03-22 16:33:40.300735 : Running LSI (1 of 2) on Top Features, 0.075 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-03-22 16:33:40.31784 : Creating Partial Matrix, 0.075 mins elapsed.\n",
      "\n",
      "2025-03-22 16:33:49.30453 : Computing LSI, 0.225 mins elapsed.\n",
      "\n",
      "2025-03-22 16:34:30.701454 : Identifying Clusters, 0.915 mins elapsed.\n",
      "\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "2025-03-22 16:34:36.869205 : Identified 6 Clusters, 1.017 mins elapsed.\n",
      "\n",
      "2025-03-22 16:34:36.88252 : Saving LSI Iteration, 1.018 mins elapsed.\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************************************************************\n",
      "2025-03-22 16:34:53.311017 : ERROR Found in .saveIteration for  \n",
      "LogFile = ArchRLogs/ArchR-addIterativeLSI-273ebf9b4b9e-Date-2025-03-22_Time-16-33-35.300146.log\n",
      "\n",
      "<simpleError in g$grobs[[legend]]: no such index at level 2\n",
      ">\n",
      "\n",
      "************************************************************\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-03-22 16:34:53.313029 : Creating Cluster Matrix on the total Group Features, 1.291 mins elapsed.\n",
      "\n",
      "2025-03-22 16:35:10.03345 : Computing Variable Features, 1.57 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-03-22 16:35:10.210411 : Running LSI (2 of 2) on Variable Features, 1.573 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-03-22 16:35:10.227274 : Creating Partial Matrix, 1.573 mins elapsed.\n",
      "\n",
      "2025-03-22 16:35:20.364643 : Computing LSI, 1.742 mins elapsed.\n",
      "\n",
      "2025-03-22 16:36:04.260698 : Finished Running IterativeLSI, 2.474 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addIterativeLSI(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    useMatrix = \"TileMatrix\", \n",
    "    name = \"IterativeLSI\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2c1d334b-7fd7-425a-bb8d-7be4cbedd3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.ArchR = addPeakSet(obj.ArchR,\n",
    "                       peakSet = reduce(pairs.E2G))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e25c14db-eb4e-4b94-8862-eb34246a365c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeakMatrix-273ebf27b41c52-Date-2025-03-22_Time-16-36-07.793911.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-03-22 16:36:08.108703 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-03-22 16:36:09.043859 : Adding HCT116 to PeakMatrix for Chr (1 of 23)!, 0.015 mins elapsed.\n",
      "\n",
      "2025-03-22 16:36:17.52651 : Adding HCT116 to PeakMatrix for Chr (2 of 23)!, 0.156 mins elapsed.\n",
      "\n",
      "2025-03-22 16:36:24.748023 : Adding HCT116 to PeakMatrix for Chr (3 of 23)!, 0.277 mins elapsed.\n",
      "\n",
      "2025-03-22 16:36:31.990789 : Adding HCT116 to PeakMatrix for Chr (4 of 23)!, 0.397 mins elapsed.\n",
      "\n",
      "2025-03-22 16:36:38.219588 : Adding HCT116 to PeakMatrix for Chr (5 of 23)!, 0.501 mins elapsed.\n",
      "\n",
      "2025-03-22 16:36:44.735247 : Adding HCT116 to PeakMatrix for Chr (6 of 23)!, 0.61 mins elapsed.\n",
      "\n",
      "2025-03-22 16:36:51.058805 : Adding HCT116 to PeakMatrix for Chr (7 of 23)!, 0.715 mins elapsed.\n",
      "\n",
      "2025-03-22 16:36:57.669287 : Adding HCT116 to PeakMatrix for Chr (8 of 23)!, 0.825 mins elapsed.\n",
      "\n",
      "2025-03-22 16:37:04.275297 : Adding HCT116 to PeakMatrix for Chr (9 of 23)!, 0.935 mins elapsed.\n",
      "\n",
      "2025-03-22 16:37:10.906224 : Adding HCT116 to PeakMatrix for Chr (10 of 23)!, 1.046 mins elapsed.\n",
      "\n",
      "2025-03-22 16:37:16.978745 : Adding HCT116 to PeakMatrix for Chr (11 of 23)!, 1.147 mins elapsed.\n",
      "\n",
      "2025-03-22 16:37:23.312689 : Adding HCT116 to PeakMatrix for Chr (12 of 23)!, 1.253 mins elapsed.\n",
      "\n",
      "2025-03-22 16:37:29.94605 : Adding HCT116 to PeakMatrix for Chr (13 of 23)!, 1.363 mins elapsed.\n",
      "\n",
      "2025-03-22 16:37:35.960537 : Adding HCT116 to PeakMatrix for Chr (14 of 23)!, 1.463 mins elapsed.\n",
      "\n",
      "2025-03-22 16:37:42.056038 : Adding HCT116 to PeakMatrix for Chr (15 of 23)!, 1.565 mins elapsed.\n",
      "\n",
      "2025-03-22 16:37:47.749692 : Adding HCT116 to PeakMatrix for Chr (16 of 23)!, 1.66 mins elapsed.\n",
      "\n",
      "2025-03-22 16:37:54.093922 : Adding HCT116 to PeakMatrix for Chr (17 of 23)!, 1.766 mins elapsed.\n",
      "\n",
      "2025-03-22 16:38:00.881098 : Adding HCT116 to PeakMatrix for Chr (18 of 23)!, 1.879 mins elapsed.\n",
      "\n",
      "2025-03-22 16:38:06.377882 : Adding HCT116 to PeakMatrix for Chr (19 of 23)!, 1.97 mins elapsed.\n",
      "\n",
      "2025-03-22 16:38:12.782398 : Adding HCT116 to PeakMatrix for Chr (20 of 23)!, 2.077 mins elapsed.\n",
      "\n",
      "2025-03-22 16:38:18.788114 : Adding HCT116 to PeakMatrix for Chr (21 of 23)!, 2.177 mins elapsed.\n",
      "\n",
      "2025-03-22 16:38:24.24694 : Adding HCT116 to PeakMatrix for Chr (22 of 23)!, 2.268 mins elapsed.\n",
      "\n",
      "2025-03-22 16:38:29.923114 : Adding HCT116 to PeakMatrix for Chr (23 of 23)!, 2.363 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeakMatrix-273ebf27b41c52-Date-2025-03-22_Time-16-36-07.793911.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addPeakMatrix(obj.ArchR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "550747c0-ded3-4a82-bc47-9d259a41ca2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.rna.rename2 = matrix.rna.rename\n",
    "colnames(matrix.rna.rename2) = paste(celltype,colnames(matrix.rna.rename),sep=\"#\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "92b20885-7b28-419c-a66d-74640e850733",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA <- SummarizedExperiment(\n",
    "  assays = SimpleList(counts = matrix.rna.rename2),\n",
    "  rowRanges = gr.tss\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "834ae1ff-160a-4e09-b201-ffe7ab9919eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA = sort.GenomicRanges(sortSeqlevels(seRNA), \n",
    "                           ignore.strand = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4b4e7086-362b-4c79-8e58-327684d0b787",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 18507 4825 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(18507): FAM87B LINC00115 ... TTTY10 EIF1AY\n",
       "rowData names(4): name score Ensembl_ID gene_type\n",
       "colnames(4825): HCT116#e18l1_AAACCGGTCGCCTGTG-1\n",
       "  HCT116#e18l1_AAACCGGTCGTTGTCA-1 ... HCT116#e18l4_TTTCAGTAGTGAGGCC-1\n",
       "  HCT116#e18l4_TTTGCTCTCCGCCATG-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "65cb2097-9075-4ee4-9e54-dd288dc50736",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addGeneExpressionMatrix-273ebf43bb5e3c-Date-2025-03-22_Time-16-38-36.197716.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Overlap w/ scATAC = 1\n",
      "\n",
      "2025-03-22 16:38:36.65379 : \n",
      "\n",
      "Overlap Per Sample w/ scATAC : HCT116=4820\n",
      "\n",
      "2025-03-22 16:38:36.665195 : \n",
      "\n",
      "2025-03-22 16:38:37.974174 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-03-22 16:38:38.646524 : Adding HCT116 to GeneExpressionMatrix for Chr (1 of 23)!, 0.011 mins elapsed.\n",
      "\n",
      "2025-03-22 16:38:44.244678 : Adding HCT116 to GeneExpressionMatrix for Chr (2 of 23)!, 0.105 mins elapsed.\n",
      "\n",
      "2025-03-22 16:38:49.506628 : Adding HCT116 to GeneExpressionMatrix for Chr (3 of 23)!, 0.192 mins elapsed.\n",
      "\n",
      "2025-03-22 16:38:56.908768 : Adding HCT116 to GeneExpressionMatrix for Chr (4 of 23)!, 0.316 mins elapsed.\n",
      "\n",
      "2025-03-22 16:39:02.065362 : Adding HCT116 to GeneExpressionMatrix for Chr (5 of 23)!, 0.402 mins elapsed.\n",
      "\n",
      "2025-03-22 16:39:07.575245 : Adding HCT116 to GeneExpressionMatrix for Chr (6 of 23)!, 0.493 mins elapsed.\n",
      "\n",
      "2025-03-22 16:39:15.03322 : Adding HCT116 to GeneExpressionMatrix for Chr (7 of 23)!, 0.618 mins elapsed.\n",
      "\n",
      "2025-03-22 16:39:20.227997 : Adding HCT116 to GeneExpressionMatrix for Chr (8 of 23)!, 0.704 mins elapsed.\n",
      "\n",
      "2025-03-22 16:39:26.040163 : Adding HCT116 to GeneExpressionMatrix for Chr (9 of 23)!, 0.801 mins elapsed.\n",
      "\n",
      "2025-03-22 16:39:33.801141 : Adding HCT116 to GeneExpressionMatrix for Chr (10 of 23)!, 0.93 mins elapsed.\n",
      "\n",
      "2025-03-22 16:39:38.892623 : Adding HCT116 to GeneExpressionMatrix for Chr (11 of 23)!, 1.015 mins elapsed.\n",
      "\n",
      "2025-03-22 16:39:44.905201 : Adding HCT116 to GeneExpressionMatrix for Chr (12 of 23)!, 1.116 mins elapsed.\n",
      "\n",
      "2025-03-22 16:39:52.468469 : Adding HCT116 to GeneExpressionMatrix for Chr (13 of 23)!, 1.242 mins elapsed.\n",
      "\n",
      "2025-03-22 16:39:57.79852 : Adding HCT116 to GeneExpressionMatrix for Chr (14 of 23)!, 1.33 mins elapsed.\n",
      "\n",
      "2025-03-22 16:40:02.857861 : Adding HCT116 to GeneExpressionMatrix for Chr (15 of 23)!, 1.415 mins elapsed.\n",
      "\n",
      "2025-03-22 16:40:10.242967 : Adding HCT116 to GeneExpressionMatrix for Chr (16 of 23)!, 1.538 mins elapsed.\n",
      "\n",
      "2025-03-22 16:40:15.674055 : Adding HCT116 to GeneExpressionMatrix for Chr (17 of 23)!, 1.628 mins elapsed.\n",
      "\n",
      "2025-03-22 16:40:20.900499 : Adding HCT116 to GeneExpressionMatrix for Chr (18 of 23)!, 1.715 mins elapsed.\n",
      "\n",
      "2025-03-22 16:40:28.851834 : Adding HCT116 to GeneExpressionMatrix for Chr (19 of 23)!, 1.848 mins elapsed.\n",
      "\n",
      "2025-03-22 16:40:33.756501 : Adding HCT116 to GeneExpressionMatrix for Chr (20 of 23)!, 1.93 mins elapsed.\n",
      "\n",
      "2025-03-22 16:40:39.270412 : Adding HCT116 to GeneExpressionMatrix for Chr (21 of 23)!, 2.022 mins elapsed.\n",
      "\n",
      "2025-03-22 16:40:46.732675 : Adding HCT116 to GeneExpressionMatrix for Chr (22 of 23)!, 2.146 mins elapsed.\n",
      "\n",
      "2025-03-22 16:40:51.846711 : Adding HCT116 to GeneExpressionMatrix for Chr (23 of 23)!, 2.231 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addGeneExpressionMatrix-273ebf43bb5e3c-Date-2025-03-22_Time-16-38-36.197716.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addGeneExpressionMatrix(obj.ArchR,\n",
    "                                    seRNA = seRNA,\n",
    "                                    strictMatch = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4318703-a143-40c0-b159-274d05033bbc",
   "metadata": {},
   "source": [
    "Run ArchR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "47c8fd5f-19ec-46e6-af87-57502c1e1ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeak2GeneLinks-273ebfdc093e-Date-2025-03-22_Time-16-40-59.44763.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-03-22 16:40:59.852361 : Getting Available Matrices, 0.007 mins elapsed.\n",
      "\n",
      "No predictionScore found. Continuing without predictionScore!\n",
      "\n",
      "2025-03-22 16:41:00.221672 : Filtered Low Prediction Score Cells (0 of 4820, 0), 0 mins elapsed.\n",
      "\n",
      "2025-03-22 16:41:00.314559 : Computing KNN, 0.002 mins elapsed.\n",
      "\n",
      "2025-03-22 16:41:00.442999 : Identifying Non-Overlapping KNN pairs, 0.004 mins elapsed.\n",
      "\n",
      "2025-03-22 16:41:02.225569 : Identified 490 Groupings!, 0.034 mins elapsed.\n",
      "\n",
      "2025-03-22 16:41:02.263244 : Getting Group RNA Matrix, 0.034 mins elapsed.\n",
      "\n",
      "2025-03-22 16:41:19.778969 : Getting Group ATAC Matrix, 0.326 mins elapsed.\n",
      "\n",
      "2025-03-22 16:41:56.07273 : Normalizing Group Matrices, 0.931 mins elapsed.\n",
      "\n",
      "2025-03-22 16:42:01.985246 : Finding Peak Gene Pairings, 1.03 mins elapsed.\n",
      "\n",
      "2025-03-22 16:42:02.785524 : Computing Correlations, 1.043 mins elapsed.\n",
      "\n",
      "2025-03-22 16:42:17.56801 : Completed Peak2Gene Correlations!, 1.289 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeak2GeneLinks-273ebfdc093e-Date-2025-03-22_Time-16-40-59.44763.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "obj.ArchR = addPeak2GeneLinks(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    reducedDims = \"IterativeLSI\",\n",
    "    useMatrix = \"GeneExpressionMatrix\",\n",
    "    maxDist = maxDist,\n",
    "    threads = n.cores\n",
    ")\n",
    "end_time <- Sys.time()\n",
    "execution_time <- end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "128d0065-5f22-495b-b00d-743fc9843337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 1.302415 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "execution_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "36bd2df2-6b32-4b81-bd4f-39ef8913b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g = metadata(obj.ArchR@peakSet)$Peak2GeneLinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e7730413-8fd4-4952-bb13-0df7a3b15b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata(p2g)[[1]]$PeakName = paste(seqnames(metadata(p2g)[[1]]),\n",
    "                                    ranges(metadata(p2g)[[1]]),sep=\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "aeea3810-6735-4aff-9bd7-1697bef979d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g$PairName = paste(metadata(p2g)[[1]][p2g$idxATAC]$PeakName,\n",
    "                     metadata(p2g)[[2]][p2g$idxRNA]$name,\n",
    "                     sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "87f975ec-6ba4-4696-ad6c-2af65da7aa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 3432522 rows and 7 columns\n",
       "                                idxATAC    idxRNA Correlation       FDR\n",
       "                              <integer> <integer>   <numeric> <numeric>\n",
       "chr1-10081-10420_FAM87B               1         1  0.14751150 0.0174716\n",
       "chr1-91097-91445_FAM87B               2         1 -0.07298123 0.3247146\n",
       "chr1-180701-181904_FAM87B             3         1  0.00180292 0.9874658\n",
       "chr1-267895-268118_FAM87B             4         1 -0.01362302 0.8941518\n",
       "chr1-586092-586306_FAM87B             5         1 -0.03564690 0.6851630\n",
       "...                                 ...       ...         ...       ...\n",
       "chrX-156003477-156004102_IL9R    153350     18493 -0.03705767  0.670750\n",
       "chrX-156005417-156005618_IL9R    153351     18493 -0.05091176  0.530522\n",
       "chrX-156019909-156020179_IL9R    153352     18493 -0.03818377  0.659411\n",
       "chrX-156029903-156030297_IL9R    153353     18493  0.04229849  0.617300\n",
       "chrX-156030522-156030719_IL9R    153354     18493 -0.00732478  0.945949\n",
       "                               VarQATAC   VarQRNA               PairName\n",
       "                              <numeric> <numeric>            <character>\n",
       "chr1-10081-10420_FAM87B        0.334448 0.0466122 chr1-10081-10420_FAM..\n",
       "chr1-91097-91445_FAM87B        0.365514 0.0466122 chr1-91097-91445_FAM..\n",
       "chr1-180701-181904_FAM87B      0.819626 0.0466122 chr1-180701-181904_F..\n",
       "chr1-267895-268118_FAM87B      0.378901 0.0466122 chr1-267895-268118_F..\n",
       "chr1-586092-586306_FAM87B      0.560494 0.0466122 chr1-586092-586306_F..\n",
       "...                                 ...       ...                    ...\n",
       "chrX-156003477-156004102_IL9R 0.5109029  0.023252 chrX-156003477-15600..\n",
       "chrX-156005417-156005618_IL9R 0.0169803  0.023252 chrX-156005417-15600..\n",
       "chrX-156019909-156020179_IL9R 0.0376645  0.023252 chrX-156019909-15602..\n",
       "chrX-156029903-156030297_IL9R 0.4007199  0.023252 chrX-156029903-15603..\n",
       "chrX-156030522-156030719_IL9R 0.1587895  0.023252 chrX-156030522-15603.."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(p2g) = p2g$PairName\n",
    "p2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c7209a32-3904-4734-93af-9c22049f652a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = pairs.E2G[pairs.E2G$PairName %in% p2g$PairName]\n",
    "mcols(pairs.E2G.res)[,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")] = p2g[pairs.E2G.res$PairName,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "1ab4ed70-ce9e-4857-bd1b-63a81d27a401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2843272 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100034595-100034763      * |         AGL\n",
       "        [2]     chr1 100034595-100034763      * |      CDC14A\n",
       "        [3]     chr1 100034595-100034763      * |         DBT\n",
       "        [4]     chr1 100034595-100034763      * |        DPH5\n",
       "        [5]     chr1 100034595-100034763      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2843268]     chrX     9911462-9911841      * |        WWC3\n",
       "  [2843269]     chrX     9995795-9996095      * |       CLCN4\n",
       "  [2843270]     chrX     9995795-9996095      * |       FAM9B\n",
       "  [2843271]     chrX     9995795-9996095      * |       TBL1X\n",
       "  [2843272]     chrX     9995795-9996095      * |        WWC3\n",
       "                          PeakName               PairName Correlation       FDR\n",
       "                       <character>            <character>   <numeric> <numeric>\n",
       "        [1] chr1-100034595-10003.. chr1-100034595-10003..  -0.0886901  0.208552\n",
       "        [2] chr1-100034595-10003.. chr1-100034595-10003..   0.0704272  0.346398\n",
       "        [3] chr1-100034595-10003.. chr1-100034595-10003..   0.0310577  0.730996\n",
       "        [4] chr1-100034595-10003.. chr1-100034595-10003..   0.0783784  0.281611\n",
       "        [5] chr1-100034595-10003.. chr1-100034595-10003..  -0.0433381  0.606808\n",
       "        ...                    ...                    ...         ...       ...\n",
       "  [2843268]   chrX-9911462-9911841 chrX-9911462-9911841.. -0.00227393  0.984079\n",
       "  [2843269]   chrX-9995795-9996095 chrX-9995795-9996095..  0.02948741  0.746463\n",
       "  [2843270]   chrX-9995795-9996095 chrX-9995795-9996095..  0.01553558  0.877422\n",
       "  [2843271]   chrX-9995795-9996095 chrX-9995795-9996095.. -0.03031286  0.738278\n",
       "  [2843272]   chrX-9995795-9996095 chrX-9995795-9996095..  0.00132809  0.990777\n",
       "             VarQATAC   VarQRNA\n",
       "            <numeric> <numeric>\n",
       "        [1]  0.339293  0.777105\n",
       "        [2]  0.339293  0.735954\n",
       "        [3]  0.339293  0.566647\n",
       "        [4]  0.339293  0.743254\n",
       "        [5]  0.339293  0.751690\n",
       "        ...       ...       ...\n",
       "  [2843268]  0.383655  0.471800\n",
       "  [2843269]  0.627352  0.363381\n",
       "  [2843270]  0.627352  0.230574\n",
       "  [2843271]  0.627352  0.964040\n",
       "  [2843272]  0.627352  0.471800\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46485351-50d7-461a-bf09-d795d5bdab84",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6d424e21-6a41-4422-b76c-6f1c0bda2222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2843272 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100034595-100034763      * |         AGL\n",
       "        [2]     chr1 100034595-100034763      * |      CDC14A\n",
       "        [3]     chr1 100034595-100034763      * |         DBT\n",
       "        [4]     chr1 100034595-100034763      * |        DPH5\n",
       "        [5]     chr1 100034595-100034763      * |       EXTL2\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [2843268]     chrX     9911462-9911841      * |        WWC3\n",
       "  [2843269]     chrX     9995795-9996095      * |       CLCN4\n",
       "  [2843270]     chrX     9995795-9996095      * |       FAM9B\n",
       "  [2843271]     chrX     9995795-9996095      * |       TBL1X\n",
       "  [2843272]     chrX     9995795-9996095      * |        WWC3\n",
       "                          PeakName               PairName Correlation       FDR\n",
       "                       <character>            <character>   <numeric> <numeric>\n",
       "        [1] chr1-100034595-10003.. chr1-100034595-10003..  -0.0886901  0.208552\n",
       "        [2] chr1-100034595-10003.. chr1-100034595-10003..   0.0704272  0.346398\n",
       "        [3] chr1-100034595-10003.. chr1-100034595-10003..   0.0310577  0.730996\n",
       "        [4] chr1-100034595-10003.. chr1-100034595-10003..   0.0783784  0.281611\n",
       "        [5] chr1-100034595-10003.. chr1-100034595-10003..  -0.0433381  0.606808\n",
       "        ...                    ...                    ...         ...       ...\n",
       "  [2843268]   chrX-9911462-9911841 chrX-9911462-9911841.. -0.00227393  0.984079\n",
       "  [2843269]   chrX-9995795-9996095 chrX-9995795-9996095..  0.02948741  0.746463\n",
       "  [2843270]   chrX-9995795-9996095 chrX-9995795-9996095..  0.01553558  0.877422\n",
       "  [2843271]   chrX-9995795-9996095 chrX-9995795-9996095.. -0.03031286  0.738278\n",
       "  [2843272]   chrX-9995795-9996095 chrX-9995795-9996095..  0.00132809  0.990777\n",
       "             VarQATAC   VarQRNA\n",
       "            <numeric> <numeric>\n",
       "        [1]  0.339293  0.777105\n",
       "        [2]  0.339293  0.735954\n",
       "        [3]  0.339293  0.566647\n",
       "        [4]  0.339293  0.743254\n",
       "        [5]  0.339293  0.751690\n",
       "        ...       ...       ...\n",
       "  [2843268]  0.383655  0.471800\n",
       "  [2843269]  0.627352  0.363381\n",
       "  [2843270]  0.627352  0.230574\n",
       "  [2843271]  0.627352  0.964040\n",
       "  [2843272]  0.627352  0.471800\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2843272 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>Correlation</th><th scope=col>FDR</th><th scope=col>VarQATAC</th><th scope=col>VarQRNA</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>AGL     </td><td>HCT116</td><td>-0.0886901403</td><td>0.208551811</td><td>0.3392934</td><td>0.7771049</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>CDC14A  </td><td>HCT116</td><td> 0.0704271925</td><td>0.346397921</td><td>0.3392934</td><td>0.7359541</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>DBT     </td><td>HCT116</td><td> 0.0310577472</td><td>0.730995946</td><td>0.3392934</td><td>0.5666468</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>DPH5    </td><td>HCT116</td><td> 0.0783784441</td><td>0.281610661</td><td>0.3392934</td><td>0.7432542</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>EXTL2   </td><td>HCT116</td><td>-0.0433380562</td><td>0.606807519</td><td>0.3392934</td><td>0.7516898</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>FRRS1   </td><td>HCT116</td><td>-0.0440303083</td><td>0.599852960</td><td>0.3392934</td><td>0.6593306</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>MFSD14A </td><td>HCT116</td><td>-0.0267752615</td><td>0.773023884</td><td>0.3392934</td><td>0.7187584</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>PLPPR4  </td><td>HCT116</td><td>-0.0446559226</td><td>0.593473464</td><td>0.3392934</td><td>0.2750230</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>RTCA    </td><td>HCT116</td><td> 0.1656950857</td><td>0.006051716</td><td>0.3392934</td><td>0.6582491</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>RTCA-AS1</td><td>HCT116</td><td> 0.0866128169</td><td>0.222146813</td><td>0.3392934</td><td>0.3840372</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>SASS6   </td><td>HCT116</td><td>-0.0434165946</td><td>0.606024750</td><td>0.3392934</td><td>0.9644730</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>SLC30A7 </td><td>HCT116</td><td>-0.1261787498</td><td>0.050401322</td><td>0.3392934</td><td>0.8625967</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>SLC35A3 </td><td>HCT116</td><td>-0.0928974219</td><td>0.182483419</td><td>0.3392934</td><td>0.8550803</td></tr>\n",
       "\t<tr><td>chr1</td><td>100034595</td><td>100034763</td><td>TRMT13  </td><td>HCT116</td><td> 0.0809073331</td><td>0.262443923</td><td>0.3392934</td><td>0.6971827</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>AGL     </td><td>HCT116</td><td> 0.0480132223</td><td>0.559449654</td><td>0.9450748</td><td>0.7771049</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>CDC14A  </td><td>HCT116</td><td> 0.0857924057</td><td>0.227736003</td><td>0.9450748</td><td>0.7359541</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>DBT     </td><td>HCT116</td><td> 0.0563508124</td><td>0.476463767</td><td>0.9450748</td><td>0.5666468</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>DPH5    </td><td>HCT116</td><td>-0.0563447091</td><td>0.476522614</td><td>0.9450748</td><td>0.7432542</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>EXTL2   </td><td>HCT116</td><td>-0.0001555414</td><td>0.998912633</td><td>0.9450748</td><td>0.7516898</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>FRRS1   </td><td>HCT116</td><td>-0.1169496106</td><td>0.075115315</td><td>0.9450748</td><td>0.6593306</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>MFSD14A </td><td>HCT116</td><td>-0.0423920783</td><td>0.616379155</td><td>0.9450748</td><td>0.7187584</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>PLPPR4  </td><td>HCT116</td><td>-0.1116924115</td><td>0.092907328</td><td>0.9450748</td><td>0.2750230</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>RTCA    </td><td>HCT116</td><td>-0.0764297740</td><td>0.296791791</td><td>0.9450748</td><td>0.6582491</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>RTCA-AS1</td><td>HCT116</td><td>-0.0008550988</td><td>0.994144352</td><td>0.9450748</td><td>0.3840372</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>SASS6   </td><td>HCT116</td><td> 0.1110469905</td><td>0.095304821</td><td>0.9450748</td><td>0.9644730</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>SLC30A7 </td><td>HCT116</td><td>-0.0082582349</td><td>0.938575787</td><td>0.9450748</td><td>0.8625967</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>SLC35A3 </td><td>HCT116</td><td>-0.0354639776</td><td>0.687013383</td><td>0.9450748</td><td>0.8550803</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037727</td><td>100038876</td><td>TRMT13  </td><td>HCT116</td><td> 0.1441419463</td><td>0.020935051</td><td>0.9450748</td><td>0.6971827</td></tr>\n",
       "\t<tr><td>chr1</td><td>100046025</td><td>100046641</td><td>AGL     </td><td>HCT116</td><td> 0.0417944635</td><td>0.622436241</td><td>0.4954289</td><td>0.7771049</td></tr>\n",
       "\t<tr><td>chr1</td><td>100046025</td><td>100046641</td><td>CDC14A  </td><td>HCT116</td><td> 0.0360638239</td><td>0.680894157</td><td>0.4954289</td><td>0.7359541</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>ASMTL    </td><td>HCT116</td><td> 0.081595477</td><td>2.573726e-01</td><td>0.8313314</td><td>0.72995187</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>CSF2RA   </td><td>HCT116</td><td> 0.002350350</td><td>9.835086e-01</td><td>0.8313314</td><td>0.48142540</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>GTPBP6   </td><td>HCT116</td><td> 0.008604201</td><td>9.358693e-01</td><td>0.8313314</td><td>0.46158006</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>IL3RA    </td><td>HCT116</td><td> 0.072006376</td><td>3.329167e-01</td><td>0.8313314</td><td>0.36419186</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>LINC00106</td><td>HCT116</td><td>-0.025367214</td><td>7.866829e-01</td><td>0.8313314</td><td>0.44173471</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>P2RY8    </td><td>HCT116</td><td> 0.019598716</td><td>8.409217e-01</td><td>0.8313314</td><td>0.15854648</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>PLCXD1   </td><td>HCT116</td><td> 0.008309264</td><td>9.381773e-01</td><td>0.8313314</td><td>0.41042557</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>PPP2R3B  </td><td>HCT116</td><td>-0.068946916</td><td>3.591959e-01</td><td>0.8313314</td><td>0.39355432</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>SHOX     </td><td>HCT116</td><td>-0.074102506</td><td>3.154506e-01</td><td>0.8313314</td><td>0.06786352</td></tr>\n",
       "\t<tr><td>chrX</td><td> 989884</td><td> 990608</td><td>SLC25A6  </td><td>HCT116</td><td> 0.045813453</td><td>5.817672e-01</td><td>0.8313314</td><td>0.93467799</td></tr>\n",
       "\t<tr><td>chrX</td><td>9902739</td><td>9902992</td><td>CLCN4    </td><td>HCT116</td><td>-0.009229020</td><td>9.308115e-01</td><td>0.1164691</td><td>0.36338074</td></tr>\n",
       "\t<tr><td>chrX</td><td>9902739</td><td>9902992</td><td>FAM9B    </td><td>HCT116</td><td>-0.042768501</td><td>6.125822e-01</td><td>0.1164691</td><td>0.23057373</td></tr>\n",
       "\t<tr><td>chrX</td><td>9902739</td><td>9902992</td><td>TBL1X    </td><td>HCT116</td><td> 0.054178092</td><td>4.979246e-01</td><td>0.1164691</td><td>0.96404045</td></tr>\n",
       "\t<tr><td>chrX</td><td>9902739</td><td>9902992</td><td>WWC3     </td><td>HCT116</td><td>-0.035251446</td><td>6.891605e-01</td><td>0.1164691</td><td>0.47180014</td></tr>\n",
       "\t<tr><td>chrX</td><td>9903251</td><td>9903473</td><td>CLCN4    </td><td>HCT116</td><td> 0.004175567</td><td>9.700268e-01</td><td>0.0574227</td><td>0.36338074</td></tr>\n",
       "\t<tr><td>chrX</td><td>9903251</td><td>9903473</td><td>FAM9B    </td><td>HCT116</td><td>-0.047956688</td><td>5.600324e-01</td><td>0.0574227</td><td>0.23057373</td></tr>\n",
       "\t<tr><td>chrX</td><td>9903251</td><td>9903473</td><td>TBL1X    </td><td>HCT116</td><td> 0.013077286</td><td>8.988093e-01</td><td>0.0574227</td><td>0.96404045</td></tr>\n",
       "\t<tr><td>chrX</td><td>9903251</td><td>9903473</td><td>WWC3     </td><td>HCT116</td><td>-0.019766476</td><td>8.393776e-01</td><td>0.0574227</td><td>0.47180014</td></tr>\n",
       "\t<tr><td>chrX</td><td>9906862</td><td>9908314</td><td>CLCN4    </td><td>HCT116</td><td>-0.012810840</td><td>9.010722e-01</td><td>0.8139468</td><td>0.36338074</td></tr>\n",
       "\t<tr><td>chrX</td><td>9906862</td><td>9908314</td><td>FAM9B    </td><td>HCT116</td><td>-0.078322039</td><td>2.820293e-01</td><td>0.8139468</td><td>0.23057373</td></tr>\n",
       "\t<tr><td>chrX</td><td>9906862</td><td>9908314</td><td>TBL1X    </td><td>HCT116</td><td> 0.158013527</td><td>9.649729e-03</td><td>0.8139468</td><td>0.96404045</td></tr>\n",
       "\t<tr><td>chrX</td><td>9906862</td><td>9908314</td><td>WWC3     </td><td>HCT116</td><td> 0.017699494</td><td>8.582497e-01</td><td>0.8139468</td><td>0.47180014</td></tr>\n",
       "\t<tr><td>chrX</td><td>9911462</td><td>9911841</td><td>CLCN4    </td><td>HCT116</td><td>-0.045770739</td><td>5.821938e-01</td><td>0.3836548</td><td>0.36338074</td></tr>\n",
       "\t<tr><td>chrX</td><td>9911462</td><td>9911841</td><td>FAM9B    </td><td>HCT116</td><td> 0.266343131</td><td>7.939026e-07</td><td>0.3836548</td><td>0.23057373</td></tr>\n",
       "\t<tr><td>chrX</td><td>9911462</td><td>9911841</td><td>TBL1X    </td><td>HCT116</td><td>-0.026788843</td><td>7.728932e-01</td><td>0.3836548</td><td>0.96404045</td></tr>\n",
       "\t<tr><td>chrX</td><td>9911462</td><td>9911841</td><td>WWC3     </td><td>HCT116</td><td>-0.002273930</td><td>9.840794e-01</td><td>0.3836548</td><td>0.47180014</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995795</td><td>9996095</td><td>CLCN4    </td><td>HCT116</td><td> 0.029487413</td><td>7.464629e-01</td><td>0.6273524</td><td>0.36338074</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995795</td><td>9996095</td><td>FAM9B    </td><td>HCT116</td><td> 0.015535582</td><td>8.774221e-01</td><td>0.6273524</td><td>0.23057373</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995795</td><td>9996095</td><td>TBL1X    </td><td>HCT116</td><td>-0.030312857</td><td>7.382784e-01</td><td>0.6273524</td><td>0.96404045</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995795</td><td>9996095</td><td>WWC3     </td><td>HCT116</td><td> 0.001328088</td><td>9.907773e-01</td><td>0.6273524</td><td>0.47180014</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2843272 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & Correlation & FDR & VarQATAC & VarQRNA\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100034595 & 100034763 & AGL      & HCT116 & -0.0886901403 & 0.208551811 & 0.3392934 & 0.7771049\\\\\n",
       "\t chr1 & 100034595 & 100034763 & CDC14A   & HCT116 &  0.0704271925 & 0.346397921 & 0.3392934 & 0.7359541\\\\\n",
       "\t chr1 & 100034595 & 100034763 & DBT      & HCT116 &  0.0310577472 & 0.730995946 & 0.3392934 & 0.5666468\\\\\n",
       "\t chr1 & 100034595 & 100034763 & DPH5     & HCT116 &  0.0783784441 & 0.281610661 & 0.3392934 & 0.7432542\\\\\n",
       "\t chr1 & 100034595 & 100034763 & EXTL2    & HCT116 & -0.0433380562 & 0.606807519 & 0.3392934 & 0.7516898\\\\\n",
       "\t chr1 & 100034595 & 100034763 & FRRS1    & HCT116 & -0.0440303083 & 0.599852960 & 0.3392934 & 0.6593306\\\\\n",
       "\t chr1 & 100034595 & 100034763 & MFSD14A  & HCT116 & -0.0267752615 & 0.773023884 & 0.3392934 & 0.7187584\\\\\n",
       "\t chr1 & 100034595 & 100034763 & PLPPR4   & HCT116 & -0.0446559226 & 0.593473464 & 0.3392934 & 0.2750230\\\\\n",
       "\t chr1 & 100034595 & 100034763 & RTCA     & HCT116 &  0.1656950857 & 0.006051716 & 0.3392934 & 0.6582491\\\\\n",
       "\t chr1 & 100034595 & 100034763 & RTCA-AS1 & HCT116 &  0.0866128169 & 0.222146813 & 0.3392934 & 0.3840372\\\\\n",
       "\t chr1 & 100034595 & 100034763 & SASS6    & HCT116 & -0.0434165946 & 0.606024750 & 0.3392934 & 0.9644730\\\\\n",
       "\t chr1 & 100034595 & 100034763 & SLC30A7  & HCT116 & -0.1261787498 & 0.050401322 & 0.3392934 & 0.8625967\\\\\n",
       "\t chr1 & 100034595 & 100034763 & SLC35A3  & HCT116 & -0.0928974219 & 0.182483419 & 0.3392934 & 0.8550803\\\\\n",
       "\t chr1 & 100034595 & 100034763 & TRMT13   & HCT116 &  0.0809073331 & 0.262443923 & 0.3392934 & 0.6971827\\\\\n",
       "\t chr1 & 100037727 & 100038876 & AGL      & HCT116 &  0.0480132223 & 0.559449654 & 0.9450748 & 0.7771049\\\\\n",
       "\t chr1 & 100037727 & 100038876 & CDC14A   & HCT116 &  0.0857924057 & 0.227736003 & 0.9450748 & 0.7359541\\\\\n",
       "\t chr1 & 100037727 & 100038876 & DBT      & HCT116 &  0.0563508124 & 0.476463767 & 0.9450748 & 0.5666468\\\\\n",
       "\t chr1 & 100037727 & 100038876 & DPH5     & HCT116 & -0.0563447091 & 0.476522614 & 0.9450748 & 0.7432542\\\\\n",
       "\t chr1 & 100037727 & 100038876 & EXTL2    & HCT116 & -0.0001555414 & 0.998912633 & 0.9450748 & 0.7516898\\\\\n",
       "\t chr1 & 100037727 & 100038876 & FRRS1    & HCT116 & -0.1169496106 & 0.075115315 & 0.9450748 & 0.6593306\\\\\n",
       "\t chr1 & 100037727 & 100038876 & MFSD14A  & HCT116 & -0.0423920783 & 0.616379155 & 0.9450748 & 0.7187584\\\\\n",
       "\t chr1 & 100037727 & 100038876 & PLPPR4   & HCT116 & -0.1116924115 & 0.092907328 & 0.9450748 & 0.2750230\\\\\n",
       "\t chr1 & 100037727 & 100038876 & RTCA     & HCT116 & -0.0764297740 & 0.296791791 & 0.9450748 & 0.6582491\\\\\n",
       "\t chr1 & 100037727 & 100038876 & RTCA-AS1 & HCT116 & -0.0008550988 & 0.994144352 & 0.9450748 & 0.3840372\\\\\n",
       "\t chr1 & 100037727 & 100038876 & SASS6    & HCT116 &  0.1110469905 & 0.095304821 & 0.9450748 & 0.9644730\\\\\n",
       "\t chr1 & 100037727 & 100038876 & SLC30A7  & HCT116 & -0.0082582349 & 0.938575787 & 0.9450748 & 0.8625967\\\\\n",
       "\t chr1 & 100037727 & 100038876 & SLC35A3  & HCT116 & -0.0354639776 & 0.687013383 & 0.9450748 & 0.8550803\\\\\n",
       "\t chr1 & 100037727 & 100038876 & TRMT13   & HCT116 &  0.1441419463 & 0.020935051 & 0.9450748 & 0.6971827\\\\\n",
       "\t chr1 & 100046025 & 100046641 & AGL      & HCT116 &  0.0417944635 & 0.622436241 & 0.4954289 & 0.7771049\\\\\n",
       "\t chr1 & 100046025 & 100046641 & CDC14A   & HCT116 &  0.0360638239 & 0.680894157 & 0.4954289 & 0.7359541\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  989884 &  990608 & ASMTL     & HCT116 &  0.081595477 & 2.573726e-01 & 0.8313314 & 0.72995187\\\\\n",
       "\t chrX &  989884 &  990608 & CSF2RA    & HCT116 &  0.002350350 & 9.835086e-01 & 0.8313314 & 0.48142540\\\\\n",
       "\t chrX &  989884 &  990608 & GTPBP6    & HCT116 &  0.008604201 & 9.358693e-01 & 0.8313314 & 0.46158006\\\\\n",
       "\t chrX &  989884 &  990608 & IL3RA     & HCT116 &  0.072006376 & 3.329167e-01 & 0.8313314 & 0.36419186\\\\\n",
       "\t chrX &  989884 &  990608 & LINC00106 & HCT116 & -0.025367214 & 7.866829e-01 & 0.8313314 & 0.44173471\\\\\n",
       "\t chrX &  989884 &  990608 & P2RY8     & HCT116 &  0.019598716 & 8.409217e-01 & 0.8313314 & 0.15854648\\\\\n",
       "\t chrX &  989884 &  990608 & PLCXD1    & HCT116 &  0.008309264 & 9.381773e-01 & 0.8313314 & 0.41042557\\\\\n",
       "\t chrX &  989884 &  990608 & PPP2R3B   & HCT116 & -0.068946916 & 3.591959e-01 & 0.8313314 & 0.39355432\\\\\n",
       "\t chrX &  989884 &  990608 & SHOX      & HCT116 & -0.074102506 & 3.154506e-01 & 0.8313314 & 0.06786352\\\\\n",
       "\t chrX &  989884 &  990608 & SLC25A6   & HCT116 &  0.045813453 & 5.817672e-01 & 0.8313314 & 0.93467799\\\\\n",
       "\t chrX & 9902739 & 9902992 & CLCN4     & HCT116 & -0.009229020 & 9.308115e-01 & 0.1164691 & 0.36338074\\\\\n",
       "\t chrX & 9902739 & 9902992 & FAM9B     & HCT116 & -0.042768501 & 6.125822e-01 & 0.1164691 & 0.23057373\\\\\n",
       "\t chrX & 9902739 & 9902992 & TBL1X     & HCT116 &  0.054178092 & 4.979246e-01 & 0.1164691 & 0.96404045\\\\\n",
       "\t chrX & 9902739 & 9902992 & WWC3      & HCT116 & -0.035251446 & 6.891605e-01 & 0.1164691 & 0.47180014\\\\\n",
       "\t chrX & 9903251 & 9903473 & CLCN4     & HCT116 &  0.004175567 & 9.700268e-01 & 0.0574227 & 0.36338074\\\\\n",
       "\t chrX & 9903251 & 9903473 & FAM9B     & HCT116 & -0.047956688 & 5.600324e-01 & 0.0574227 & 0.23057373\\\\\n",
       "\t chrX & 9903251 & 9903473 & TBL1X     & HCT116 &  0.013077286 & 8.988093e-01 & 0.0574227 & 0.96404045\\\\\n",
       "\t chrX & 9903251 & 9903473 & WWC3      & HCT116 & -0.019766476 & 8.393776e-01 & 0.0574227 & 0.47180014\\\\\n",
       "\t chrX & 9906862 & 9908314 & CLCN4     & HCT116 & -0.012810840 & 9.010722e-01 & 0.8139468 & 0.36338074\\\\\n",
       "\t chrX & 9906862 & 9908314 & FAM9B     & HCT116 & -0.078322039 & 2.820293e-01 & 0.8139468 & 0.23057373\\\\\n",
       "\t chrX & 9906862 & 9908314 & TBL1X     & HCT116 &  0.158013527 & 9.649729e-03 & 0.8139468 & 0.96404045\\\\\n",
       "\t chrX & 9906862 & 9908314 & WWC3      & HCT116 &  0.017699494 & 8.582497e-01 & 0.8139468 & 0.47180014\\\\\n",
       "\t chrX & 9911462 & 9911841 & CLCN4     & HCT116 & -0.045770739 & 5.821938e-01 & 0.3836548 & 0.36338074\\\\\n",
       "\t chrX & 9911462 & 9911841 & FAM9B     & HCT116 &  0.266343131 & 7.939026e-07 & 0.3836548 & 0.23057373\\\\\n",
       "\t chrX & 9911462 & 9911841 & TBL1X     & HCT116 & -0.026788843 & 7.728932e-01 & 0.3836548 & 0.96404045\\\\\n",
       "\t chrX & 9911462 & 9911841 & WWC3      & HCT116 & -0.002273930 & 9.840794e-01 & 0.3836548 & 0.47180014\\\\\n",
       "\t chrX & 9995795 & 9996095 & CLCN4     & HCT116 &  0.029487413 & 7.464629e-01 & 0.6273524 & 0.36338074\\\\\n",
       "\t chrX & 9995795 & 9996095 & FAM9B     & HCT116 &  0.015535582 & 8.774221e-01 & 0.6273524 & 0.23057373\\\\\n",
       "\t chrX & 9995795 & 9996095 & TBL1X     & HCT116 & -0.030312857 & 7.382784e-01 & 0.6273524 & 0.96404045\\\\\n",
       "\t chrX & 9995795 & 9996095 & WWC3      & HCT116 &  0.001328088 & 9.907773e-01 & 0.6273524 & 0.47180014\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2843272 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | Correlation &lt;dbl&gt; | FDR &lt;dbl&gt; | VarQATAC &lt;dbl&gt; | VarQRNA &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100034595 | 100034763 | AGL      | HCT116 | -0.0886901403 | 0.208551811 | 0.3392934 | 0.7771049 |\n",
       "| chr1 | 100034595 | 100034763 | CDC14A   | HCT116 |  0.0704271925 | 0.346397921 | 0.3392934 | 0.7359541 |\n",
       "| chr1 | 100034595 | 100034763 | DBT      | HCT116 |  0.0310577472 | 0.730995946 | 0.3392934 | 0.5666468 |\n",
       "| chr1 | 100034595 | 100034763 | DPH5     | HCT116 |  0.0783784441 | 0.281610661 | 0.3392934 | 0.7432542 |\n",
       "| chr1 | 100034595 | 100034763 | EXTL2    | HCT116 | -0.0433380562 | 0.606807519 | 0.3392934 | 0.7516898 |\n",
       "| chr1 | 100034595 | 100034763 | FRRS1    | HCT116 | -0.0440303083 | 0.599852960 | 0.3392934 | 0.6593306 |\n",
       "| chr1 | 100034595 | 100034763 | MFSD14A  | HCT116 | -0.0267752615 | 0.773023884 | 0.3392934 | 0.7187584 |\n",
       "| chr1 | 100034595 | 100034763 | PLPPR4   | HCT116 | -0.0446559226 | 0.593473464 | 0.3392934 | 0.2750230 |\n",
       "| chr1 | 100034595 | 100034763 | RTCA     | HCT116 |  0.1656950857 | 0.006051716 | 0.3392934 | 0.6582491 |\n",
       "| chr1 | 100034595 | 100034763 | RTCA-AS1 | HCT116 |  0.0866128169 | 0.222146813 | 0.3392934 | 0.3840372 |\n",
       "| chr1 | 100034595 | 100034763 | SASS6    | HCT116 | -0.0434165946 | 0.606024750 | 0.3392934 | 0.9644730 |\n",
       "| chr1 | 100034595 | 100034763 | SLC30A7  | HCT116 | -0.1261787498 | 0.050401322 | 0.3392934 | 0.8625967 |\n",
       "| chr1 | 100034595 | 100034763 | SLC35A3  | HCT116 | -0.0928974219 | 0.182483419 | 0.3392934 | 0.8550803 |\n",
       "| chr1 | 100034595 | 100034763 | TRMT13   | HCT116 |  0.0809073331 | 0.262443923 | 0.3392934 | 0.6971827 |\n",
       "| chr1 | 100037727 | 100038876 | AGL      | HCT116 |  0.0480132223 | 0.559449654 | 0.9450748 | 0.7771049 |\n",
       "| chr1 | 100037727 | 100038876 | CDC14A   | HCT116 |  0.0857924057 | 0.227736003 | 0.9450748 | 0.7359541 |\n",
       "| chr1 | 100037727 | 100038876 | DBT      | HCT116 |  0.0563508124 | 0.476463767 | 0.9450748 | 0.5666468 |\n",
       "| chr1 | 100037727 | 100038876 | DPH5     | HCT116 | -0.0563447091 | 0.476522614 | 0.9450748 | 0.7432542 |\n",
       "| chr1 | 100037727 | 100038876 | EXTL2    | HCT116 | -0.0001555414 | 0.998912633 | 0.9450748 | 0.7516898 |\n",
       "| chr1 | 100037727 | 100038876 | FRRS1    | HCT116 | -0.1169496106 | 0.075115315 | 0.9450748 | 0.6593306 |\n",
       "| chr1 | 100037727 | 100038876 | MFSD14A  | HCT116 | -0.0423920783 | 0.616379155 | 0.9450748 | 0.7187584 |\n",
       "| chr1 | 100037727 | 100038876 | PLPPR4   | HCT116 | -0.1116924115 | 0.092907328 | 0.9450748 | 0.2750230 |\n",
       "| chr1 | 100037727 | 100038876 | RTCA     | HCT116 | -0.0764297740 | 0.296791791 | 0.9450748 | 0.6582491 |\n",
       "| chr1 | 100037727 | 100038876 | RTCA-AS1 | HCT116 | -0.0008550988 | 0.994144352 | 0.9450748 | 0.3840372 |\n",
       "| chr1 | 100037727 | 100038876 | SASS6    | HCT116 |  0.1110469905 | 0.095304821 | 0.9450748 | 0.9644730 |\n",
       "| chr1 | 100037727 | 100038876 | SLC30A7  | HCT116 | -0.0082582349 | 0.938575787 | 0.9450748 | 0.8625967 |\n",
       "| chr1 | 100037727 | 100038876 | SLC35A3  | HCT116 | -0.0354639776 | 0.687013383 | 0.9450748 | 0.8550803 |\n",
       "| chr1 | 100037727 | 100038876 | TRMT13   | HCT116 |  0.1441419463 | 0.020935051 | 0.9450748 | 0.6971827 |\n",
       "| chr1 | 100046025 | 100046641 | AGL      | HCT116 |  0.0417944635 | 0.622436241 | 0.4954289 | 0.7771049 |\n",
       "| chr1 | 100046025 | 100046641 | CDC14A   | HCT116 |  0.0360638239 | 0.680894157 | 0.4954289 | 0.7359541 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  989884 |  990608 | ASMTL     | HCT116 |  0.081595477 | 2.573726e-01 | 0.8313314 | 0.72995187 |\n",
       "| chrX |  989884 |  990608 | CSF2RA    | HCT116 |  0.002350350 | 9.835086e-01 | 0.8313314 | 0.48142540 |\n",
       "| chrX |  989884 |  990608 | GTPBP6    | HCT116 |  0.008604201 | 9.358693e-01 | 0.8313314 | 0.46158006 |\n",
       "| chrX |  989884 |  990608 | IL3RA     | HCT116 |  0.072006376 | 3.329167e-01 | 0.8313314 | 0.36419186 |\n",
       "| chrX |  989884 |  990608 | LINC00106 | HCT116 | -0.025367214 | 7.866829e-01 | 0.8313314 | 0.44173471 |\n",
       "| chrX |  989884 |  990608 | P2RY8     | HCT116 |  0.019598716 | 8.409217e-01 | 0.8313314 | 0.15854648 |\n",
       "| chrX |  989884 |  990608 | PLCXD1    | HCT116 |  0.008309264 | 9.381773e-01 | 0.8313314 | 0.41042557 |\n",
       "| chrX |  989884 |  990608 | PPP2R3B   | HCT116 | -0.068946916 | 3.591959e-01 | 0.8313314 | 0.39355432 |\n",
       "| chrX |  989884 |  990608 | SHOX      | HCT116 | -0.074102506 | 3.154506e-01 | 0.8313314 | 0.06786352 |\n",
       "| chrX |  989884 |  990608 | SLC25A6   | HCT116 |  0.045813453 | 5.817672e-01 | 0.8313314 | 0.93467799 |\n",
       "| chrX | 9902739 | 9902992 | CLCN4     | HCT116 | -0.009229020 | 9.308115e-01 | 0.1164691 | 0.36338074 |\n",
       "| chrX | 9902739 | 9902992 | FAM9B     | HCT116 | -0.042768501 | 6.125822e-01 | 0.1164691 | 0.23057373 |\n",
       "| chrX | 9902739 | 9902992 | TBL1X     | HCT116 |  0.054178092 | 4.979246e-01 | 0.1164691 | 0.96404045 |\n",
       "| chrX | 9902739 | 9902992 | WWC3      | HCT116 | -0.035251446 | 6.891605e-01 | 0.1164691 | 0.47180014 |\n",
       "| chrX | 9903251 | 9903473 | CLCN4     | HCT116 |  0.004175567 | 9.700268e-01 | 0.0574227 | 0.36338074 |\n",
       "| chrX | 9903251 | 9903473 | FAM9B     | HCT116 | -0.047956688 | 5.600324e-01 | 0.0574227 | 0.23057373 |\n",
       "| chrX | 9903251 | 9903473 | TBL1X     | HCT116 |  0.013077286 | 8.988093e-01 | 0.0574227 | 0.96404045 |\n",
       "| chrX | 9903251 | 9903473 | WWC3      | HCT116 | -0.019766476 | 8.393776e-01 | 0.0574227 | 0.47180014 |\n",
       "| chrX | 9906862 | 9908314 | CLCN4     | HCT116 | -0.012810840 | 9.010722e-01 | 0.8139468 | 0.36338074 |\n",
       "| chrX | 9906862 | 9908314 | FAM9B     | HCT116 | -0.078322039 | 2.820293e-01 | 0.8139468 | 0.23057373 |\n",
       "| chrX | 9906862 | 9908314 | TBL1X     | HCT116 |  0.158013527 | 9.649729e-03 | 0.8139468 | 0.96404045 |\n",
       "| chrX | 9906862 | 9908314 | WWC3      | HCT116 |  0.017699494 | 8.582497e-01 | 0.8139468 | 0.47180014 |\n",
       "| chrX | 9911462 | 9911841 | CLCN4     | HCT116 | -0.045770739 | 5.821938e-01 | 0.3836548 | 0.36338074 |\n",
       "| chrX | 9911462 | 9911841 | FAM9B     | HCT116 |  0.266343131 | 7.939026e-07 | 0.3836548 | 0.23057373 |\n",
       "| chrX | 9911462 | 9911841 | TBL1X     | HCT116 | -0.026788843 | 7.728932e-01 | 0.3836548 | 0.96404045 |\n",
       "| chrX | 9911462 | 9911841 | WWC3      | HCT116 | -0.002273930 | 9.840794e-01 | 0.3836548 | 0.47180014 |\n",
       "| chrX | 9995795 | 9996095 | CLCN4     | HCT116 |  0.029487413 | 7.464629e-01 | 0.6273524 | 0.36338074 |\n",
       "| chrX | 9995795 | 9996095 | FAM9B     | HCT116 |  0.015535582 | 8.774221e-01 | 0.6273524 | 0.23057373 |\n",
       "| chrX | 9995795 | 9996095 | TBL1X     | HCT116 | -0.030312857 | 7.382784e-01 | 0.6273524 | 0.96404045 |\n",
       "| chrX | 9995795 | 9996095 | WWC3      | HCT116 |  0.001328088 | 9.907773e-01 | 0.6273524 | 0.47180014 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType Correlation   FDR         \n",
       "1       chr1 100034595 100034763 AGL        HCT116   -0.0886901403 0.208551811 \n",
       "2       chr1 100034595 100034763 CDC14A     HCT116    0.0704271925 0.346397921 \n",
       "3       chr1 100034595 100034763 DBT        HCT116    0.0310577472 0.730995946 \n",
       "4       chr1 100034595 100034763 DPH5       HCT116    0.0783784441 0.281610661 \n",
       "5       chr1 100034595 100034763 EXTL2      HCT116   -0.0433380562 0.606807519 \n",
       "6       chr1 100034595 100034763 FRRS1      HCT116   -0.0440303083 0.599852960 \n",
       "7       chr1 100034595 100034763 MFSD14A    HCT116   -0.0267752615 0.773023884 \n",
       "8       chr1 100034595 100034763 PLPPR4     HCT116   -0.0446559226 0.593473464 \n",
       "9       chr1 100034595 100034763 RTCA       HCT116    0.1656950857 0.006051716 \n",
       "10      chr1 100034595 100034763 RTCA-AS1   HCT116    0.0866128169 0.222146813 \n",
       "11      chr1 100034595 100034763 SASS6      HCT116   -0.0434165946 0.606024750 \n",
       "12      chr1 100034595 100034763 SLC30A7    HCT116   -0.1261787498 0.050401322 \n",
       "13      chr1 100034595 100034763 SLC35A3    HCT116   -0.0928974219 0.182483419 \n",
       "14      chr1 100034595 100034763 TRMT13     HCT116    0.0809073331 0.262443923 \n",
       "15      chr1 100037727 100038876 AGL        HCT116    0.0480132223 0.559449654 \n",
       "16      chr1 100037727 100038876 CDC14A     HCT116    0.0857924057 0.227736003 \n",
       "17      chr1 100037727 100038876 DBT        HCT116    0.0563508124 0.476463767 \n",
       "18      chr1 100037727 100038876 DPH5       HCT116   -0.0563447091 0.476522614 \n",
       "19      chr1 100037727 100038876 EXTL2      HCT116   -0.0001555414 0.998912633 \n",
       "20      chr1 100037727 100038876 FRRS1      HCT116   -0.1169496106 0.075115315 \n",
       "21      chr1 100037727 100038876 MFSD14A    HCT116   -0.0423920783 0.616379155 \n",
       "22      chr1 100037727 100038876 PLPPR4     HCT116   -0.1116924115 0.092907328 \n",
       "23      chr1 100037727 100038876 RTCA       HCT116   -0.0764297740 0.296791791 \n",
       "24      chr1 100037727 100038876 RTCA-AS1   HCT116   -0.0008550988 0.994144352 \n",
       "25      chr1 100037727 100038876 SASS6      HCT116    0.1110469905 0.095304821 \n",
       "26      chr1 100037727 100038876 SLC30A7    HCT116   -0.0082582349 0.938575787 \n",
       "27      chr1 100037727 100038876 SLC35A3    HCT116   -0.0354639776 0.687013383 \n",
       "28      chr1 100037727 100038876 TRMT13     HCT116    0.1441419463 0.020935051 \n",
       "29      chr1 100046025 100046641 AGL        HCT116    0.0417944635 0.622436241 \n",
       "30      chr1 100046025 100046641 CDC14A     HCT116    0.0360638239 0.680894157 \n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮             ⋮           \n",
       "2843243 chrX  989884    990608   ASMTL      HCT116    0.081595477  2.573726e-01\n",
       "2843244 chrX  989884    990608   CSF2RA     HCT116    0.002350350  9.835086e-01\n",
       "2843245 chrX  989884    990608   GTPBP6     HCT116    0.008604201  9.358693e-01\n",
       "2843246 chrX  989884    990608   IL3RA      HCT116    0.072006376  3.329167e-01\n",
       "2843247 chrX  989884    990608   LINC00106  HCT116   -0.025367214  7.866829e-01\n",
       "2843248 chrX  989884    990608   P2RY8      HCT116    0.019598716  8.409217e-01\n",
       "2843249 chrX  989884    990608   PLCXD1     HCT116    0.008309264  9.381773e-01\n",
       "2843250 chrX  989884    990608   PPP2R3B    HCT116   -0.068946916  3.591959e-01\n",
       "2843251 chrX  989884    990608   SHOX       HCT116   -0.074102506  3.154506e-01\n",
       "2843252 chrX  989884    990608   SLC25A6    HCT116    0.045813453  5.817672e-01\n",
       "2843253 chrX 9902739   9902992   CLCN4      HCT116   -0.009229020  9.308115e-01\n",
       "2843254 chrX 9902739   9902992   FAM9B      HCT116   -0.042768501  6.125822e-01\n",
       "2843255 chrX 9902739   9902992   TBL1X      HCT116    0.054178092  4.979246e-01\n",
       "2843256 chrX 9902739   9902992   WWC3       HCT116   -0.035251446  6.891605e-01\n",
       "2843257 chrX 9903251   9903473   CLCN4      HCT116    0.004175567  9.700268e-01\n",
       "2843258 chrX 9903251   9903473   FAM9B      HCT116   -0.047956688  5.600324e-01\n",
       "2843259 chrX 9903251   9903473   TBL1X      HCT116    0.013077286  8.988093e-01\n",
       "2843260 chrX 9903251   9903473   WWC3       HCT116   -0.019766476  8.393776e-01\n",
       "2843261 chrX 9906862   9908314   CLCN4      HCT116   -0.012810840  9.010722e-01\n",
       "2843262 chrX 9906862   9908314   FAM9B      HCT116   -0.078322039  2.820293e-01\n",
       "2843263 chrX 9906862   9908314   TBL1X      HCT116    0.158013527  9.649729e-03\n",
       "2843264 chrX 9906862   9908314   WWC3       HCT116    0.017699494  8.582497e-01\n",
       "2843265 chrX 9911462   9911841   CLCN4      HCT116   -0.045770739  5.821938e-01\n",
       "2843266 chrX 9911462   9911841   FAM9B      HCT116    0.266343131  7.939026e-07\n",
       "2843267 chrX 9911462   9911841   TBL1X      HCT116   -0.026788843  7.728932e-01\n",
       "2843268 chrX 9911462   9911841   WWC3       HCT116   -0.002273930  9.840794e-01\n",
       "2843269 chrX 9995795   9996095   CLCN4      HCT116    0.029487413  7.464629e-01\n",
       "2843270 chrX 9995795   9996095   FAM9B      HCT116    0.015535582  8.774221e-01\n",
       "2843271 chrX 9995795   9996095   TBL1X      HCT116   -0.030312857  7.382784e-01\n",
       "2843272 chrX 9995795   9996095   WWC3       HCT116    0.001328088  9.907773e-01\n",
       "        VarQATAC  VarQRNA   \n",
       "1       0.3392934 0.7771049 \n",
       "2       0.3392934 0.7359541 \n",
       "3       0.3392934 0.5666468 \n",
       "4       0.3392934 0.7432542 \n",
       "5       0.3392934 0.7516898 \n",
       "6       0.3392934 0.6593306 \n",
       "7       0.3392934 0.7187584 \n",
       "8       0.3392934 0.2750230 \n",
       "9       0.3392934 0.6582491 \n",
       "10      0.3392934 0.3840372 \n",
       "11      0.3392934 0.9644730 \n",
       "12      0.3392934 0.8625967 \n",
       "13      0.3392934 0.8550803 \n",
       "14      0.3392934 0.6971827 \n",
       "15      0.9450748 0.7771049 \n",
       "16      0.9450748 0.7359541 \n",
       "17      0.9450748 0.5666468 \n",
       "18      0.9450748 0.7432542 \n",
       "19      0.9450748 0.7516898 \n",
       "20      0.9450748 0.6593306 \n",
       "21      0.9450748 0.7187584 \n",
       "22      0.9450748 0.2750230 \n",
       "23      0.9450748 0.6582491 \n",
       "24      0.9450748 0.3840372 \n",
       "25      0.9450748 0.9644730 \n",
       "26      0.9450748 0.8625967 \n",
       "27      0.9450748 0.8550803 \n",
       "28      0.9450748 0.6971827 \n",
       "29      0.4954289 0.7771049 \n",
       "30      0.4954289 0.7359541 \n",
       "⋮       ⋮         ⋮         \n",
       "2843243 0.8313314 0.72995187\n",
       "2843244 0.8313314 0.48142540\n",
       "2843245 0.8313314 0.46158006\n",
       "2843246 0.8313314 0.36419186\n",
       "2843247 0.8313314 0.44173471\n",
       "2843248 0.8313314 0.15854648\n",
       "2843249 0.8313314 0.41042557\n",
       "2843250 0.8313314 0.39355432\n",
       "2843251 0.8313314 0.06786352\n",
       "2843252 0.8313314 0.93467799\n",
       "2843253 0.1164691 0.36338074\n",
       "2843254 0.1164691 0.23057373\n",
       "2843255 0.1164691 0.96404045\n",
       "2843256 0.1164691 0.47180014\n",
       "2843257 0.0574227 0.36338074\n",
       "2843258 0.0574227 0.23057373\n",
       "2843259 0.0574227 0.96404045\n",
       "2843260 0.0574227 0.47180014\n",
       "2843261 0.8139468 0.36338074\n",
       "2843262 0.8139468 0.23057373\n",
       "2843263 0.8139468 0.96404045\n",
       "2843264 0.8139468 0.47180014\n",
       "2843265 0.3836548 0.36338074\n",
       "2843266 0.3836548 0.23057373\n",
       "2843267 0.3836548 0.96404045\n",
       "2843268 0.3836548 0.47180014\n",
       "2843269 0.6273524 0.36338074\n",
       "2843270 0.6273524 0.23057373\n",
       "2843271 0.6273524 0.96404045\n",
       "2843272 0.6273524 0.47180014"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(obj.ArchR,\n",
    "        paste(dir.output,\"obj.ArchR.rds\",sep = \"/\"))\n",
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = celltype\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"Correlation\",\n",
    "                         \"FDR\",\n",
    "                         \"VarQATAC\",\n",
    "                         \"VarQRNA\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "pairs.E2G.res\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "4c909ed5-7321-4e69-8208-1544b371da6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] uwot_0.1.16                       nabor_0.5.0                      \n",
       " [3] Rsamtools_2.18.0                  BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] dplyr_1.1.4                       genomation_1.34.0                \n",
       "[11] Signac_1.13.0                     Seurat_5.0.3                     \n",
       "[13] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[15] rhdf5_2.44.0                      SummarizedExperiment_1.30.2      \n",
       "[17] Biobase_2.60.0                    MatrixGenerics_1.12.3            \n",
       "[19] Rcpp_1.0.12                       Matrix_1.6-5                     \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               matrixStats_1.3.0                \n",
       "[27] data.table_1.15.2                 stringr_1.5.1                    \n",
       "[29] plyr_1.8.9                        magrittr_2.0.3                   \n",
       "[31] ggplot2_3.5.1                     gtable_0.3.5                     \n",
       "[33] gtools_3.9.5                      gridExtra_2.3                    \n",
       "[35] ArchR_1.0.2                      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22         splines_4.3.3            later_1.3.2             \n",
       "  [4] pbdZMQ_0.3-11            BiocIO_1.10.0            bitops_1.0-7            \n",
       "  [7] tibble_3.2.1             polyclip_1.10-6          XML_3.99-0.16.1         \n",
       " [10] fastDummies_1.7.3        lifecycle_1.0.4          vroom_1.6.5             \n",
       " [13] globals_0.16.3           lattice_0.22-6           MASS_7.3-60             \n",
       " [16] plotly_4.10.4            plotrix_3.8-4            yaml_2.3.8              \n",
       " [19] httpuv_1.6.15            sctransform_0.4.1        spam_2.10-0             \n",
       " [22] spatstat.sparse_3.0-3    reticulate_1.36.1        cowplot_1.1.3           \n",
       " [25] pbapply_1.7-2            RColorBrewer_1.1-3       abind_1.4-5             \n",
       " [28] zlibbioc_1.48.0          Rtsne_0.17               purrr_1.0.2             \n",
       " [31] RCurl_1.98-1.14          GenomeInfoDbData_1.2.11  ggrepel_0.9.5           \n",
       " [34] irlba_2.3.5.1            listenv_0.9.1            spatstat.utils_3.0-4    \n",
       " [37] goftest_1.2-3            RSpectra_0.16-1          spatstat.random_3.2-3   \n",
       " [40] fitdistrplus_1.1-11      parallelly_1.37.1        leiden_0.4.3.1          \n",
       " [43] codetools_0.2-20         DelayedArray_0.26.7      RcppRoll_0.3.0          \n",
       " [46] tidyselect_1.2.1         farver_2.1.1             base64enc_0.1-3         \n",
       " [49] spatstat.explore_3.2-6   GenomicAlignments_1.36.0 jsonlite_1.8.8          \n",
       " [52] progressr_0.14.0         ggridges_0.5.6           survival_3.6-4          \n",
       " [55] tools_4.3.3              ica_1.0-3                glue_1.7.0              \n",
       " [58] IRdisplay_1.1            withr_3.0.0              fastmap_1.1.1           \n",
       " [61] rhdf5filters_1.12.1      fansi_1.0.6              digest_0.6.35           \n",
       " [64] R6_2.5.1                 mime_0.12                seqPattern_1.34.0       \n",
       " [67] colorspace_2.1-0         Cairo_1.6-2              scattermore_1.2         \n",
       " [70] tensor_1.5               spatstat.data_3.0-4      utf8_1.2.4              \n",
       " [73] tidyr_1.3.1              generics_0.1.3           httr_1.4.7              \n",
       " [76] htmlwidgets_1.6.4        S4Arrays_1.2.1           pkgconfig_2.0.3         \n",
       " [79] impute_1.76.0            lmtest_0.9-40            htmltools_0.5.8.1       \n",
       " [82] dotCall64_1.1-1          scales_1.3.0             png_0.1-8               \n",
       " [85] tzdb_0.4.0               reshape2_1.4.4           rjson_0.2.21            \n",
       " [88] uuid_1.2-0               nlme_3.1-164             repr_1.1.7              \n",
       " [91] zoo_1.8-12               KernSmooth_2.23-22       miniUI_0.1.1.1          \n",
       " [94] restfulr_0.0.15          pillar_1.9.0             vctrs_0.6.5             \n",
       " [97] RANN_2.6.1               promises_1.3.0           xtable_1.8-4            \n",
       "[100] cluster_2.1.6            evaluate_0.23            readr_2.1.5             \n",
       "[103] cli_3.6.2                compiler_4.3.3           rlang_1.1.3             \n",
       "[106] crayon_1.5.2             future.apply_1.11.2      labeling_0.4.3          \n",
       "[109] stringi_1.8.3            gridBase_0.4-7           viridisLite_0.4.2       \n",
       "[112] deldir_2.0-4             BiocParallel_1.36.0      munsell_0.5.1           \n",
       "[115] lazyeval_0.2.2           spatstat.geom_3.2-9      IRkernel_1.3.2          \n",
       "[118] RcppHNSW_0.6.0           hms_1.1.3                patchwork_1.2.0         \n",
       "[121] bit64_4.0.5              future_1.33.2            Rhdf5lib_1.22.1         \n",
       "[124] shiny_1.8.1.1            ROCR_1.0-11              igraph_2.0.3            \n",
       "[127] fastmatch_1.1-4          bit_4.0.5               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7690cee-49f1-4886-ae66-0b63a43ce524",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
