{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8466b35-cae5-4f62-8138-37b2de484a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.2\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : grid v4.3.3\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "\tLoading Package : gtools v3.9.5\n",
      "\n",
      "\tLoading Package : gtable v0.3.5\n",
      "\n",
      "\tLoading Package : ggplot2 v3.5.1\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "\tLoading Package : plyr v1.8.9\n",
      "\n",
      "\tLoading Package : stringr v1.5.1\n",
      "\n",
      "\tLoading Package : data.table v1.15.2\n",
      "\n",
      "\tLoading Package : matrixStats v1.3.0\n",
      "\n",
      "\tLoading Package : S4Vectors v0.40.2\n",
      "\n",
      "\tLoading Package : GenomicRanges v1.54.1\n",
      "\n",
      "\tLoading Package : BiocGenerics v0.48.1\n",
      "\n",
      "\tLoading Package : Matrix v1.6.5\n",
      "\n",
      "\tLoading Package : Rcpp v1.0.12\n",
      "\n",
      "\tLoading Package : SummarizedExperiment v1.30.2\n",
      "\n",
      "\tLoading Package : rhdf5 v2.44.0\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(magrittr)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e85020d7-d358-492d-9358-c81270f4b624",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1bec6690-52ec-450d-bba7-7d4528a185cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.seurat = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/1.prepare_data/1.seurat_pipeline.240507/obj.seurat.qc.rds\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/240508/sc-E2G/test/results_K562_Xu/K562/Kendall/Pairs.tsv.gz\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/3.Genome_wide_prediction/ArchR.default_250226/ArchR.250226/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "60b75d89-aa5d-485e-b1a3-ec1988eb7a3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "inputFiles = c(\"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_230720/ArrayExpress/cellranger_res/EMTAB11264_01/outs/atac_fragments.tsv.gz\")\n",
    "names(inputFiles) = c(\"K562_Xu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da973385-4c81-4b1c-9669-cf5933b23a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 16.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRThreads(threads = 16) \n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeb26496-d3b9-4778-92ed-9c63a5018aef",
   "metadata": {},
   "source": [
    "Import seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c9b19a09-be95-4d9a-bdc3-0530980d4757",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.seurat = readRDS(path.seurat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86be3e66-c954-4620-9e64-c6a1c0fcb84c",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6f363969-bf0b-45ce-858c-a8cb64cc210a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readGeneric(path.pairs.E2G,\n",
    "                        header = T,\n",
    "                        keep.all.metadata = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3538ffdc-4fed-48eb-9c9d-86805a91cf06",
   "metadata": {},
   "source": [
    "Create an ArchRProject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0300f485-fafa-4cef-b9d0-534108ec6f95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 1\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_230720/ArrayExpress/cellranger\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"c8366071805c37835af9f47f43ae7435\" \"af5268737bf82cc3e7978e6407c893ea\"\n",
      "  .. ..@ cells: Named chr [1:7821] \"AAACAGCCAAGCGATG-1\" \"AAACAGCCACATGCTA-1\" \"AAACAGCCACGGTACT-1\" \"AAACAGCCATAAAGCA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:7821] \"K562_Xu_AAACAGCCAAGCGATG-1\" \"K562_Xu_AAACAGCCACATGCTA-1\" \"K562_Xu_AAACAGCCACGGTACT-1\" \"K562_Xu_AAACAGCCATAAAGCA-1\" ...\n"
     ]
    }
   ],
   "source": [
    "str(Fragments(obj.seurat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1753ffaa-89e2-43bf-bcf6-19937782d63a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-createArrows-2a42fe7164eacd-Date-2025-02-26_Time-14-31-16.56323.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Cleaning Temporary Files\n",
      "\n",
      "2025-02-26 14:31:16.988444 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "(K562_Xu : 1 of 1) Determining Arrow Method to use!\n",
      "\n",
      "2025-02-26 14:31:17.057429 : (K562_Xu : 1 of 1) Reading In Fragments from inputFiles (readMethod = tabix), 0.001 mins elapsed.\n",
      "\n",
      "2025-02-26 14:31:17.067606 : (K562_Xu : 1 of 1) Tabix Bed To Temporary File, 0.001 mins elapsed.\n",
      "\n",
      "2025-02-26 14:33:21.372063 : (K562_Xu : 1 of 1) Successful creation of Temporary File, 2.073 mins elapsed.\n",
      "\n",
      "2025-02-26 14:33:21.382462 : (K562_Xu : 1 of 1) Creating ArrowFile From Temporary File, 2.073 mins elapsed.\n",
      "\n",
      "2025-02-26 14:34:05.06408 : (K562_Xu : 1 of 1) Successful creation of Arrow File, 2.801 mins elapsed.\n",
      "\n",
      "2025-02-26 14:34:51.917733 : Continuing through after error ggplot for Fragment Size Distribution, 3.582 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-02-26 14:35:26.36141 : (K562_Xu : 1 of 1) CellStats : Number of Cells Pass Filter = 7699 , 4.156 mins elapsed.\n",
      "\n",
      "2025-02-26 14:35:26.372603 : (K562_Xu : 1 of 1) CellStats : Median Frags = 31472 , 4.156 mins elapsed.\n",
      "\n",
      "2025-02-26 14:35:26.378604 : (K562_Xu : 1 of 1) CellStats : Median TSS Enrichment = 9.17 , 4.157 mins elapsed.\n",
      "\n",
      "2025-02-26 14:35:27.267224 : Continuing through after error ggplot for TSS by Frags, 4.171 mins elapsed.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2025-02-26 14:35:30.283615 : (K562_Xu : 1 of 1) Adding Additional Feature Counts!, 4.222 mins elapsed.\n",
      "\n",
      "2025-02-26 14:35:56.743878 : (K562_Xu : 1 of 1) Removing Fragments from Filtered Cells, 4.663 mins elapsed.\n",
      "\n",
      "2025-02-26 14:35:56.753065 : (K562_Xu : 1 of 1) Creating Filtered Arrow File, 4.663 mins elapsed.\n",
      "\n",
      "2025-02-26 14:37:51.165371 : (K562_Xu : 1 of 1) Finished Constructing Filtered Arrow File!, 6.57 mins elapsed.\n",
      "\n",
      "2025-02-26 14:37:51.399408 : (K562_Xu : 1 of 1) Adding TileMatrix!, 6.574 mins elapsed.\n",
      "\n",
      "2025-02-26 14:47:00.026733 : (K562_Xu : 1 of 1) Finished Creating Arrow File, 15.717 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-createArrows-2a42fe7164eacd-Date-2025-02-26_Time-14-31-16.56323.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ArrowFiles <- createArrowFiles(\n",
    "  inputFiles = c(Fragments(obj.seurat)[[1]]@path),\n",
    "  sampleNames = c(\"K562_Xu\"),\n",
    "  validBarcodes = list(K562_Xu = unname(Fragments(obj.seurat)[[1]]@cells)),\n",
    "  addTileMat = T,\n",
    "  addGeneScoreMat = F,\n",
    "  force = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ada3db45-d050-4659-a7cf-2891e04b1249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'K562_Xu.arrow'"
      ],
      "text/latex": [
       "'K562\\_Xu.arrow'"
      ],
      "text/markdown": [
       "'K562_Xu.arrow'"
      ],
      "text/plain": [
       "[1] \"K562_Xu.arrow\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ArrowFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5ba28283-d488-4982-8d4c-745fc3f59365",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"ArchR\",sep=\"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aa217507-8323-43bf-b23c-0ec25ee33c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Validating Arrows...\n",
      "\n",
      "Getting SampleNames...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Copying ArrowFiles to Ouptut Directory! If you want to save disk space set copyArrows = FALSE\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Getting Cell Metadata...\n",
      "\n",
      "1 \n",
      "\n",
      "\n",
      "Merging Cell Metadata...\n",
      "\n",
      "Initializing ArchRProject...\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR <- ArchRProject(\n",
    "  ArrowFiles = ArrowFiles, \n",
    "  outputDirectory = paste(dir.output,\"ArchR\",sep=\"/\"),\n",
    "  copyArrows = TRUE \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f9795403-545f-4385-aa38-2c70b1b2fb21",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: ArchRProject \n",
       "outputDirectory: /maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/K562_Xu/3.Genome_wide_prediction/ArchR.default_250226/ArchR.250226/ArchR \n",
       "samples(1): K562_Xu\n",
       "sampleColData names(1): ArrowFiles\n",
       "cellColData names(13): Sample TSSEnrichment ... nDiFrags BlacklistRatio\n",
       "numberOfCells(1): 7699\n",
       "medianTSS(1): 9.17\n",
       "medianFrags(1): 31472"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obj.ArchR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ee876dd5-0b22-46fd-9015-7e22144c5024",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking Inputs...\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addIterativeLSI-2a42fe179ee1e6-Date-2025-02-26_Time-14-47-13.251495.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 14:47:13.805352 : Computing Total Across All Features, 0 mins elapsed.\n",
      "\n",
      "2025-02-26 14:47:14.776077 : Computing Top Features, 0.017 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-26 14:47:18.441121 : Running LSI (1 of 2) on Top Features, 0.078 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-02-26 14:47:18.459857 : Creating Partial Matrix, 0.078 mins elapsed.\n",
      "\n",
      "2025-02-26 14:47:58.966594 : Computing LSI, 0.753 mins elapsed.\n",
      "\n",
      "2025-02-26 14:49:09.498015 : Identifying Clusters, 1.929 mins elapsed.\n",
      "\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "2025-02-26 14:49:17.458746 : Identified 6 Clusters, 2.061 mins elapsed.\n",
      "\n",
      "2025-02-26 14:49:17.476054 : Saving LSI Iteration, 2.062 mins elapsed.\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************************************************************\n",
      "2025-02-26 14:49:34.885082 : ERROR Found in .saveIteration for  \n",
      "LogFile = ArchRLogs/ArchR-addIterativeLSI-2a42fe179ee1e6-Date-2025-02-26_Time-14-47-13.251495.log\n",
      "\n",
      "<simpleError in g$grobs[[legend]]: no such index at level 2\n",
      ">\n",
      "\n",
      "************************************************************\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-26 14:49:34.887272 : Creating Cluster Matrix on the total Group Features, 2.352 mins elapsed.\n",
      "\n",
      "2025-02-26 14:49:55.456003 : Computing Variable Features, 2.695 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-26 14:49:55.636307 : Running LSI (2 of 2) on Variable Features, 2.698 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-02-26 14:49:55.653011 : Creating Partial Matrix, 2.698 mins elapsed.\n",
      "\n",
      "2025-02-26 14:50:33.607482 : Computing LSI, 3.33 mins elapsed.\n",
      "\n",
      "2025-02-26 14:51:17.665385 : Finished Running IterativeLSI, 4.065 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addIterativeLSI(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    useMatrix = \"TileMatrix\", \n",
    "    name = \"IterativeLSI\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2c1d334b-7fd7-425a-bb8d-7be4cbedd3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.ArchR = addPeakSet(obj.ArchR,\n",
    "                       peakSet = reduce(pairs.E2G))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e25c14db-eb4e-4b94-8862-eb34246a365c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeakMatrix-2a42fe367fdd1d-Date-2025-02-26_Time-14-51-20.671696.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 14:51:20.946418 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-02-26 14:51:22.124906 : Adding K562_Xu to PeakMatrix for Chr (1 of 23)!, 0.019 mins elapsed.\n",
      "\n",
      "2025-02-26 14:51:45.504779 : Adding K562_Xu to PeakMatrix for Chr (2 of 23)!, 0.409 mins elapsed.\n",
      "\n",
      "2025-02-26 14:51:59.962287 : Adding K562_Xu to PeakMatrix for Chr (3 of 23)!, 0.65 mins elapsed.\n",
      "\n",
      "2025-02-26 14:52:13.080742 : Adding K562_Xu to PeakMatrix for Chr (4 of 23)!, 0.868 mins elapsed.\n",
      "\n",
      "2025-02-26 14:52:22.20224 : Adding K562_Xu to PeakMatrix for Chr (5 of 23)!, 1.02 mins elapsed.\n",
      "\n",
      "2025-02-26 14:52:32.892949 : Adding K562_Xu to PeakMatrix for Chr (6 of 23)!, 1.198 mins elapsed.\n",
      "\n",
      "2025-02-26 14:52:46.066713 : Adding K562_Xu to PeakMatrix for Chr (7 of 23)!, 1.418 mins elapsed.\n",
      "\n",
      "2025-02-26 14:52:59.28871 : Adding K562_Xu to PeakMatrix for Chr (8 of 23)!, 1.638 mins elapsed.\n",
      "\n",
      "2025-02-26 14:53:09.227687 : Adding K562_Xu to PeakMatrix for Chr (9 of 23)!, 1.804 mins elapsed.\n",
      "\n",
      "2025-02-26 14:53:18.622673 : Adding K562_Xu to PeakMatrix for Chr (10 of 23)!, 1.961 mins elapsed.\n",
      "\n",
      "2025-02-26 14:53:28.392098 : Adding K562_Xu to PeakMatrix for Chr (11 of 23)!, 2.123 mins elapsed.\n",
      "\n",
      "2025-02-26 14:53:40.699184 : Adding K562_Xu to PeakMatrix for Chr (12 of 23)!, 2.328 mins elapsed.\n",
      "\n",
      "2025-02-26 14:53:51.231041 : Adding K562_Xu to PeakMatrix for Chr (13 of 23)!, 2.504 mins elapsed.\n",
      "\n",
      "2025-02-26 14:53:58.012377 : Adding K562_Xu to PeakMatrix for Chr (14 of 23)!, 2.617 mins elapsed.\n",
      "\n",
      "2025-02-26 14:54:05.799448 : Adding K562_Xu to PeakMatrix for Chr (15 of 23)!, 2.747 mins elapsed.\n",
      "\n",
      "2025-02-26 14:54:14.594033 : Adding K562_Xu to PeakMatrix for Chr (16 of 23)!, 2.893 mins elapsed.\n",
      "\n",
      "2025-02-26 14:54:24.744309 : Adding K562_Xu to PeakMatrix for Chr (17 of 23)!, 3.063 mins elapsed.\n",
      "\n",
      "2025-02-26 14:54:35.79237 : Adding K562_Xu to PeakMatrix for Chr (18 of 23)!, 3.247 mins elapsed.\n",
      "\n",
      "2025-02-26 14:54:42.850349 : Adding K562_Xu to PeakMatrix for Chr (19 of 23)!, 3.364 mins elapsed.\n",
      "\n",
      "2025-02-26 14:54:57.324894 : Adding K562_Xu to PeakMatrix for Chr (20 of 23)!, 3.606 mins elapsed.\n",
      "\n",
      "2025-02-26 14:55:04.955812 : Adding K562_Xu to PeakMatrix for Chr (21 of 23)!, 3.733 mins elapsed.\n",
      "\n",
      "2025-02-26 14:55:12.033195 : Adding K562_Xu to PeakMatrix for Chr (22 of 23)!, 3.851 mins elapsed.\n",
      "\n",
      "2025-02-26 14:55:21.156521 : Adding K562_Xu to PeakMatrix for Chr (23 of 23)!, 4.003 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeakMatrix-2a42fe367fdd1d-Date-2025-02-26_Time-14-51-20.671696.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addPeakMatrix(obj.ArchR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3b913333-8823-435e-a677-42fa6d8777ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seRNA = import10xFeatureMatrix(c(\"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_230720/ArrayExpress/cellranger_res/EMTAB11264_01/outs/filtered_feature_bc_matrix.h5\"),\n",
    "                               c(\"K562_Xu\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "895234f3-e668-4e83-adf0-562024a6711b",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA = seRNA[,rownames(obj.ArchR@cellColData)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4656d98c-a9cc-486a-afaa-fe59ca6ef055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 7699 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(7699): K562_Xu#GTCATGCCAGGATGGC-1 K562_Xu#TATCGCGAGGTTAGAG-1\n",
       "  ... K562_Xu#CGTGTTACAGTCTATG-1 K562_Xu#AAGTGAAGTTGCCTCA-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b39610de-ce74-419c-8767-7e537dbdbd39",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c37bedbf-7457-4bad-8093-8e8465cdd7f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "40436ee4-39d8-4306-a297-064a88c61738",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biostrings’:\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seRNA.rename = map_gene_names(seRNA,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "65cb2097-9075-4ee4-9e54-dd288dc50736",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addGeneExpressionMatrix-2a42fe1ab7da03-Date-2025-02-26_Time-14-56-20.678465.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Overlap w/ scATAC = 1\n",
      "\n",
      "2025-02-26 14:56:21.189126 : \n",
      "\n",
      "Overlap Per Sample w/ scATAC : K562_Xu=7699\n",
      "\n",
      "2025-02-26 14:56:21.204276 : \n",
      "\n",
      "2025-02-26 14:56:25.181516 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-02-26 14:56:26.355954 : Adding K562_Xu to GeneExpressionMatrix for Chr (1 of 23)!, 0.02 mins elapsed.\n",
      "\n",
      "2025-02-26 14:56:32.295981 : Adding K562_Xu to GeneExpressionMatrix for Chr (2 of 23)!, 0.119 mins elapsed.\n",
      "\n",
      "2025-02-26 14:56:38.094322 : Adding K562_Xu to GeneExpressionMatrix for Chr (3 of 23)!, 0.215 mins elapsed.\n",
      "\n",
      "2025-02-26 14:56:45.98266 : Adding K562_Xu to GeneExpressionMatrix for Chr (4 of 23)!, 0.347 mins elapsed.\n",
      "\n",
      "2025-02-26 14:56:51.2469 : Adding K562_Xu to GeneExpressionMatrix for Chr (5 of 23)!, 0.434 mins elapsed.\n",
      "\n",
      "2025-02-26 14:56:57.002897 : Adding K562_Xu to GeneExpressionMatrix for Chr (6 of 23)!, 0.53 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:04.704926 : Adding K562_Xu to GeneExpressionMatrix for Chr (7 of 23)!, 0.659 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:10.007952 : Adding K562_Xu to GeneExpressionMatrix for Chr (8 of 23)!, 0.747 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:15.437362 : Adding K562_Xu to GeneExpressionMatrix for Chr (9 of 23)!, 0.838 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:23.475986 : Adding K562_Xu to GeneExpressionMatrix for Chr (10 of 23)!, 0.972 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:28.700785 : Adding K562_Xu to GeneExpressionMatrix for Chr (11 of 23)!, 1.059 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:34.231045 : Adding K562_Xu to GeneExpressionMatrix for Chr (12 of 23)!, 1.151 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:42.061548 : Adding K562_Xu to GeneExpressionMatrix for Chr (13 of 23)!, 1.281 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:47.418448 : Adding K562_Xu to GeneExpressionMatrix for Chr (14 of 23)!, 1.371 mins elapsed.\n",
      "\n",
      "2025-02-26 14:57:52.70455 : Adding K562_Xu to GeneExpressionMatrix for Chr (15 of 23)!, 1.459 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:00.46584 : Adding K562_Xu to GeneExpressionMatrix for Chr (16 of 23)!, 1.588 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:06.10817 : Adding K562_Xu to GeneExpressionMatrix for Chr (17 of 23)!, 1.682 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:11.711151 : Adding K562_Xu to GeneExpressionMatrix for Chr (18 of 23)!, 1.775 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:19.474494 : Adding K562_Xu to GeneExpressionMatrix for Chr (19 of 23)!, 1.905 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:24.988593 : Adding K562_Xu to GeneExpressionMatrix for Chr (20 of 23)!, 1.997 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:30.428879 : Adding K562_Xu to GeneExpressionMatrix for Chr (21 of 23)!, 2.087 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:38.023578 : Adding K562_Xu to GeneExpressionMatrix for Chr (22 of 23)!, 2.214 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:43.230867 : Adding K562_Xu to GeneExpressionMatrix for Chr (23 of 23)!, 2.301 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addGeneExpressionMatrix-2a42fe1ab7da03-Date-2025-02-26_Time-14-56-20.678465.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addGeneExpressionMatrix(obj.ArchR,\n",
    "                                    seRNA = seRNA.rename,\n",
    "                                    strictMatch = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4318703-a143-40c0-b159-274d05033bbc",
   "metadata": {},
   "source": [
    "Run ArchR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "47c8fd5f-19ec-46e6-af87-57502c1e1ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeak2GeneLinks-2a42fe4f1f3eff-Date-2025-02-26_Time-14-58-50.970028.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 14:58:51.341342 : Getting Available Matrices, 0.006 mins elapsed.\n",
      "\n",
      "No predictionScore found. Continuing without predictionScore!\n",
      "\n",
      "2025-02-26 14:58:51.688836 : Filtered Low Prediction Score Cells (0 of 7699, 0), 0 mins elapsed.\n",
      "\n",
      "Filtering 1 dims correlated > 0.75 to log10(depth + 1)\n",
      "\n",
      "2025-02-26 14:58:51.786633 : Computing KNN, 0.002 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:52.130786 : Identifying Non-Overlapping KNN pairs, 0.008 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:53.797663 : Identified 494 Groupings!, 0.036 mins elapsed.\n",
      "\n",
      "2025-02-26 14:58:53.847038 : Getting Group RNA Matrix, 0.036 mins elapsed.\n",
      "\n",
      "2025-02-26 14:59:15.326424 : Getting Group ATAC Matrix, 0.394 mins elapsed.\n",
      "\n",
      "2025-02-26 14:59:48.08897 : Normalizing Group Matrices, 0.94 mins elapsed.\n",
      "\n",
      "2025-02-26 14:59:54.352564 : Finding Peak Gene Pairings, 1.045 mins elapsed.\n",
      "\n",
      "2025-02-26 14:59:54.81347 : Computing Correlations, 1.053 mins elapsed.\n",
      "\n",
      "2025-02-26 15:00:05.611723 : Completed Peak2Gene Correlations!, 1.232 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeak2GeneLinks-2a42fe4f1f3eff-Date-2025-02-26_Time-14-58-50.970028.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "obj.ArchR = addPeak2GeneLinks(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    reducedDims = \"IterativeLSI\",\n",
    "    useMatrix = \"GeneExpressionMatrix\",\n",
    "    maxDist = 250000,\n",
    "    threads = 16\n",
    ")\n",
    "end_time <- Sys.time()\n",
    "execution_time <- end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "128d0065-5f22-495b-b00d-743fc9843337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 1.244431 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "execution_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "36bd2df2-6b32-4b81-bd4f-39ef8913b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g = metadata(obj.ArchR@peakSet)$Peak2GeneLinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e7730413-8fd4-4952-bb13-0df7a3b15b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata(p2g)[[1]]$PeakName = paste(seqnames(metadata(p2g)[[1]]),\n",
    "                                    ranges(metadata(p2g)[[1]]),sep=\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "aeea3810-6735-4aff-9bd7-1697bef979d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g$PairName = paste(metadata(p2g)[[1]][p2g$idxATAC]$PeakName,\n",
    "                     metadata(p2g)[[2]][p2g$idxRNA]$name,\n",
    "                     sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "87f975ec-6ba4-4696-ad6c-2af65da7aa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 1113182 rows and 7 columns\n",
       "                                idxATAC    idxRNA Correlation         FDR\n",
       "                              <integer> <integer>   <numeric>   <numeric>\n",
       "chr1-9926-10618_FAM138A               1         1         NaN         NaN\n",
       "chr1-15894-16509_FAM138A              2         1         NaN         NaN\n",
       "chr1-17271-17713_FAM138A              3         1         NaN         NaN\n",
       "chr1-29230-29495_FAM138A              4         1         NaN         NaN\n",
       "chr1-115493-115961_FAM138A            5         1         NaN         NaN\n",
       "...                                 ...       ...         ...         ...\n",
       "chrX-156009722-156010137_IL9R    157596     20493   0.0787222 0.139159667\n",
       "chrX-156016191-156016823_IL9R    157597     20493   0.0747140 0.162911793\n",
       "chrX-156019811-156020191_IL9R    157598     20493   0.0323247 0.581744450\n",
       "chrX-156025028-156025591_IL9R    157599     20493  -0.1003445 0.052806442\n",
       "chrX-156030021-156030865_IL9R    157600     20493   0.1683037 0.000617572\n",
       "                               VarQATAC   VarQRNA               PairName\n",
       "                              <numeric> <numeric>            <character>\n",
       "chr1-9926-10618_FAM138A        0.780032 0.0597765 chr1-9926-10618_FAM1..\n",
       "chr1-15894-16509_FAM138A       0.274651 0.0597765 chr1-15894-16509_FAM..\n",
       "chr1-17271-17713_FAM138A       0.338503 0.0597765 chr1-17271-17713_FAM..\n",
       "chr1-29230-29495_FAM138A       0.222716 0.0597765 chr1-29230-29495_FAM..\n",
       "chr1-115493-115961_FAM138A     0.843997 0.0597765 chr1-115493-115961_F..\n",
       "...                                 ...       ...                    ...\n",
       "chrX-156009722-156010137_IL9R 0.0590863  0.349485 chrX-156009722-15601..\n",
       "chrX-156016191-156016823_IL9R 0.5508439  0.349485 chrX-156016191-15601..\n",
       "chrX-156019811-156020191_IL9R 0.1883376  0.349485 chrX-156019811-15602..\n",
       "chrX-156025028-156025591_IL9R 0.4855330  0.349485 chrX-156025028-15602..\n",
       "chrX-156030021-156030865_IL9R 0.8127475  0.349485 chrX-156030021-15603.."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(p2g) = p2g$PairName\n",
    "p2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c7209a32-3904-4734-93af-9c22049f652a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = pairs.E2G[pairs.E2G$PairName %in% p2g$PairName]\n",
    "mcols(pairs.E2G.res)[,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")] = p2g[pairs.E2G.res$PairName,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1ab4ed70-ce9e-4857-bd1b-63a81d27a401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 933499 ranges and 7 metadata columns:\n",
       "           seqnames            ranges strand |  TargetGene\n",
       "              <Rle>         <IRanges>  <Rle> | <character>\n",
       "       [1]     chr1 10001196-10001745      * |      CLSTN1\n",
       "       [2]     chr1 10001196-10001745      * |    CTNNBIP1\n",
       "       [3]     chr1 10001196-10001745      * |       KIF1B\n",
       "       [4]     chr1 10001196-10001745      * |        LZIC\n",
       "       [5]     chr1 10001196-10001745      * |      NMNAT1\n",
       "       ...      ...               ...    ... .         ...\n",
       "  [933495]     chrX   9882635-9883318      * |        WWC3\n",
       "  [933496]     chrX   9885160-9885475      * |        WWC3\n",
       "  [933497]     chrX   9894484-9895274      * |        WWC3\n",
       "  [933498]     chrX   9995827-9996208      * |       CLCN4\n",
       "  [933499]     chrX   9995827-9996208      * |        WWC3\n",
       "                         PeakName               PairName Correlation        FDR\n",
       "                      <character>            <character>   <numeric>  <numeric>\n",
       "       [1] chr1-10001196-10001745 chr1-10001196-100017.. -0.00856967 0.89392649\n",
       "       [2] chr1-10001196-10001745 chr1-10001196-100017..  0.10927651 0.03331330\n",
       "       [3] chr1-10001196-10001745 chr1-10001196-100017.. -0.08571135 0.10397671\n",
       "       [4] chr1-10001196-10001745 chr1-10001196-100017.. -0.15389096 0.00191243\n",
       "       [5] chr1-10001196-10001745 chr1-10001196-100017..  0.00880569 0.89091627\n",
       "       ...                    ...                    ...         ...        ...\n",
       "  [933495]   chrX-9882635-9883318 chrX-9882635-9883318..  0.01862477   0.762186\n",
       "  [933496]   chrX-9885160-9885475 chrX-9885160-9885475..  0.00286508   0.965457\n",
       "  [933497]   chrX-9894484-9895274 chrX-9894484-9895274..  0.00195320   0.976592\n",
       "  [933498]   chrX-9995827-9996208 chrX-9995827-9996208.. -0.02470933   0.681096\n",
       "  [933499]   chrX-9995827-9996208 chrX-9995827-9996208.. -0.03551510   0.541243\n",
       "            VarQATAC   VarQRNA\n",
       "           <numeric> <numeric>\n",
       "       [1]  0.121796  0.664959\n",
       "       [2]  0.121796  0.695311\n",
       "       [3]  0.121796  0.859171\n",
       "       [4]  0.121796  0.706876\n",
       "       [5]  0.121796  0.713512\n",
       "       ...       ...       ...\n",
       "  [933495]  0.471237  0.461231\n",
       "  [933496]  0.101599  0.461231\n",
       "  [933497]  0.317132  0.461231\n",
       "  [933498]  0.339911  0.425365\n",
       "  [933499]  0.339911  0.461231\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46485351-50d7-461a-bf09-d795d5bdab84",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6d424e21-6a41-4422-b76c-6f1c0bda2222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 933499 ranges and 7 metadata columns:\n",
       "           seqnames            ranges strand |  TargetGene\n",
       "              <Rle>         <IRanges>  <Rle> | <character>\n",
       "       [1]     chr1 10001196-10001745      * |      CLSTN1\n",
       "       [2]     chr1 10001196-10001745      * |    CTNNBIP1\n",
       "       [3]     chr1 10001196-10001745      * |       KIF1B\n",
       "       [4]     chr1 10001196-10001745      * |        LZIC\n",
       "       [5]     chr1 10001196-10001745      * |      NMNAT1\n",
       "       ...      ...               ...    ... .         ...\n",
       "  [933495]     chrX   9882635-9883318      * |        WWC3\n",
       "  [933496]     chrX   9885160-9885475      * |        WWC3\n",
       "  [933497]     chrX   9894484-9895274      * |        WWC3\n",
       "  [933498]     chrX   9995827-9996208      * |       CLCN4\n",
       "  [933499]     chrX   9995827-9996208      * |        WWC3\n",
       "                         PeakName               PairName Correlation        FDR\n",
       "                      <character>            <character>   <numeric>  <numeric>\n",
       "       [1] chr1-10001196-10001745 chr1-10001196-100017.. -0.00856967 0.89392649\n",
       "       [2] chr1-10001196-10001745 chr1-10001196-100017..  0.10927651 0.03331330\n",
       "       [3] chr1-10001196-10001745 chr1-10001196-100017.. -0.08571135 0.10397671\n",
       "       [4] chr1-10001196-10001745 chr1-10001196-100017.. -0.15389096 0.00191243\n",
       "       [5] chr1-10001196-10001745 chr1-10001196-100017..  0.00880569 0.89091627\n",
       "       ...                    ...                    ...         ...        ...\n",
       "  [933495]   chrX-9882635-9883318 chrX-9882635-9883318..  0.01862477   0.762186\n",
       "  [933496]   chrX-9885160-9885475 chrX-9885160-9885475..  0.00286508   0.965457\n",
       "  [933497]   chrX-9894484-9895274 chrX-9894484-9895274..  0.00195320   0.976592\n",
       "  [933498]   chrX-9995827-9996208 chrX-9995827-9996208.. -0.02470933   0.681096\n",
       "  [933499]   chrX-9995827-9996208 chrX-9995827-9996208.. -0.03551510   0.541243\n",
       "            VarQATAC   VarQRNA\n",
       "           <numeric> <numeric>\n",
       "       [1]  0.121796  0.664959\n",
       "       [2]  0.121796  0.695311\n",
       "       [3]  0.121796  0.859171\n",
       "       [4]  0.121796  0.706876\n",
       "       [5]  0.121796  0.713512\n",
       "       ...       ...       ...\n",
       "  [933495]  0.471237  0.461231\n",
       "  [933496]  0.101599  0.461231\n",
       "  [933497]  0.317132  0.461231\n",
       "  [933498]  0.339911  0.425365\n",
       "  [933499]  0.339911  0.461231\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 933499 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>Correlation</th><th scope=col>FDR</th><th scope=col>VarQATAC</th><th scope=col>VarQRNA</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CLSTN1  </td><td>K562</td><td>-0.008569671</td><td>8.939265e-01</td><td>0.12179569</td><td>0.6649588</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>CTNNBIP1</td><td>K562</td><td> 0.109276509</td><td>3.331330e-02</td><td>0.12179569</td><td>0.6953106</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>KIF1B   </td><td>K562</td><td>-0.085711347</td><td>1.039767e-01</td><td>0.12179569</td><td>0.8591714</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>LZIC    </td><td>K562</td><td>-0.153890963</td><td>1.912432e-03</td><td>0.12179569</td><td>0.7068755</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>NMNAT1  </td><td>K562</td><td> 0.008805695</td><td>8.909163e-01</td><td>0.12179569</td><td>0.7135119</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>RBP7    </td><td>K562</td><td> 0.055430425</td><td>3.168859e-01</td><td>0.12179569</td><td>0.2445713</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001196</td><td> 10001745</td><td>UBE4B   </td><td>K562</td><td> 0.139513416</td><td>5.324126e-03</td><td>0.12179569</td><td>0.8348216</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>AGL     </td><td>K562</td><td>-0.060843072</td><td>2.669373e-01</td><td>0.66776015</td><td>0.7150734</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>DBT     </td><td>K562</td><td> 0.113399517</td><td>2.660451e-02</td><td>0.66776015</td><td>0.6860391</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>MFSD14A </td><td>K562</td><td>-0.033765151</td><td>5.633475e-01</td><td>0.66776015</td><td>0.4949983</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>RTCA    </td><td>K562</td><td>-0.296069972</td><td>2.140068e-10</td><td>0.66776015</td><td>0.7483531</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>RTCA-AS1</td><td>K562</td><td>-0.078206426</td><td>1.420580e-01</td><td>0.66776015</td><td>0.6117211</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>SASS6   </td><td>K562</td><td>-0.005903034</td><td>9.277100e-01</td><td>0.66776015</td><td>0.9037232</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>SLC35A3 </td><td>K562</td><td>-0.031298666</td><td>5.947948e-01</td><td>0.66776015</td><td>0.6436344</td></tr>\n",
       "\t<tr><td>chr1</td><td>100027917</td><td>100029695</td><td>TRMT13  </td><td>K562</td><td> 0.051259410</td><td>3.588414e-01</td><td>0.66776015</td><td>0.7907090</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10003541</td><td> 10003870</td><td>CLSTN1  </td><td>K562</td><td>-0.059352630</td><td>2.801764e-01</td><td>0.03770939</td><td>0.6649588</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10003541</td><td> 10003870</td><td>CTNNBIP1</td><td>K562</td><td>-0.082603548</td><td>1.186765e-01</td><td>0.03770939</td><td>0.6953106</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10003541</td><td> 10003870</td><td>KIF1B   </td><td>K562</td><td>-0.057018137</td><td>3.017411e-01</td><td>0.03770939</td><td>0.8591714</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10003541</td><td> 10003870</td><td>LZIC    </td><td>K562</td><td>-0.011783743</td><td>8.524567e-01</td><td>0.03770939</td><td>0.7068755</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10003541</td><td> 10003870</td><td>NMNAT1  </td><td>K562</td><td>-0.039432846</td><td>4.929231e-01</td><td>0.03770939</td><td>0.7135119</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10003541</td><td> 10003870</td><td>RBP7    </td><td>K562</td><td>-0.052121018</td><td>3.499480e-01</td><td>0.03770939</td><td>0.2445713</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10003541</td><td> 10003870</td><td>UBE4B   </td><td>K562</td><td> 0.056684970</td><td>3.048841e-01</td><td>0.03770939</td><td>0.8348216</td></tr>\n",
       "\t<tr><td>chr1</td><td>100036824</td><td>100037384</td><td>AGL     </td><td>K562</td><td>-0.174757022</td><td>3.599259e-04</td><td>0.73982234</td><td>0.7150734</td></tr>\n",
       "\t<tr><td>chr1</td><td>100036824</td><td>100037384</td><td>DBT     </td><td>K562</td><td>-0.084248849</td><td>1.106875e-01</td><td>0.73982234</td><td>0.6860391</td></tr>\n",
       "\t<tr><td>chr1</td><td>100036824</td><td>100037384</td><td>MFSD14A </td><td>K562</td><td>-0.108289741</td><td>3.512104e-02</td><td>0.73982234</td><td>0.4949983</td></tr>\n",
       "\t<tr><td>chr1</td><td>100036824</td><td>100037384</td><td>RTCA    </td><td>K562</td><td> 0.062065170</td><td>2.563674e-01</td><td>0.73982234</td><td>0.7483531</td></tr>\n",
       "\t<tr><td>chr1</td><td>100036824</td><td>100037384</td><td>RTCA-AS1</td><td>K562</td><td>-0.159306953</td><td>1.265751e-03</td><td>0.73982234</td><td>0.6117211</td></tr>\n",
       "\t<tr><td>chr1</td><td>100036824</td><td>100037384</td><td>SASS6   </td><td>K562</td><td> 0.277445564</td><td>3.357707e-09</td><td>0.73982234</td><td>0.9037232</td></tr>\n",
       "\t<tr><td>chr1</td><td>100036824</td><td>100037384</td><td>SLC35A3 </td><td>K562</td><td> 0.001882806</td><td>9.774065e-01</td><td>0.73982234</td><td>0.6436344</td></tr>\n",
       "\t<tr><td>chr1</td><td>100036824</td><td>100037384</td><td>TRMT13  </td><td>K562</td><td> 0.331370468</td><td>6.330091e-13</td><td>0.73982234</td><td>0.7907090</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9515603</td><td>9516754</td><td>TBL1X</td><td>K562</td><td> 0.513764666</td><td>1.218433e-32</td><td>0.993857868</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9519003</td><td>9519166</td><td>TBL1X</td><td>K562</td><td> 0.192456958</td><td>7.353360e-05</td><td>0.076395939</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9523396</td><td>9523614</td><td>TBL1X</td><td>K562</td><td> 0.139923602</td><td>5.178057e-03</td><td>0.019790609</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9526743</td><td>9527004</td><td>TBL1X</td><td>K562</td><td> 0.154020902</td><td>1.893822e-03</td><td>0.036770305</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9527097</td><td>9527256</td><td>TBL1X</td><td>K562</td><td> 0.222023623</td><td>3.578773e-06</td><td>0.022461929</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9527473</td><td>9527741</td><td>TBL1X</td><td>K562</td><td> 0.160792514</td><td>1.127579e-03</td><td>0.063705584</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9530658</td><td>9531443</td><td>TBL1X</td><td>K562</td><td> 0.133005904</td><td>8.196805e-03</td><td>0.763521574</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9533720</td><td>9534013</td><td>TBL1X</td><td>K562</td><td> 0.059816442</td><td>2.759620e-01</td><td>0.035831218</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9538535</td><td>9539381</td><td>TBL1X</td><td>K562</td><td> 0.266441620</td><td>1.543532e-08</td><td>0.433432741</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9540026</td><td>9540401</td><td>TBL1X</td><td>K562</td><td> 0.112173135</td><td>2.847067e-02</td><td>0.036357868</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9542740</td><td>9543112</td><td>TBL1X</td><td>K562</td><td> 0.047767576</td><td>3.962190e-01</td><td>0.029961929</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9550938</td><td>9551425</td><td>TBL1X</td><td>K562</td><td> 0.195457064</td><td>5.530382e-05</td><td>0.037906091</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9554809</td><td>9554981</td><td>TBL1X</td><td>K562</td><td>-0.015526458</td><td>8.033311e-01</td><td>0.007722081</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9555057</td><td>9556694</td><td>TBL1X</td><td>K562</td><td> 0.123634939</td><td>1.471748e-02</td><td>0.584714467</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9569158</td><td>9569613</td><td>TBL1X</td><td>K562</td><td> 0.089513403</td><td>8.796009e-02</td><td>0.032595178</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9586788</td><td>9586944</td><td>TBL1X</td><td>K562</td><td>-0.246656984</td><td>2.008704e-07</td><td>0.003826142</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9587015</td><td>9587399</td><td>TBL1X</td><td>K562</td><td> 0.102109587</td><td>4.836585e-02</td><td>0.037290609</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9642694</td><td>9643219</td><td>TBL1X</td><td>K562</td><td>-0.062598490</td><td>2.518334e-01</td><td>0.048115482</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9646431</td><td>9647138</td><td>TBL1X</td><td>K562</td><td>-0.013798935</td><td>8.260575e-01</td><td>0.536230964</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9670192</td><td>9670610</td><td>TBL1X</td><td>K562</td><td> 0.142847783</td><td>4.237276e-03</td><td>0.089638325</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9707468</td><td>9708164</td><td>TBL1X</td><td>K562</td><td> 0.055507472</td><td>3.161291e-01</td><td>0.149505076</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9708493</td><td>9709582</td><td>TBL1X</td><td>K562</td><td> 0.046906423</td><td>4.058372e-01</td><td>0.659003807</td><td>0.9171424</td></tr>\n",
       "\t<tr><td>chrX</td><td>9786114</td><td>9786869</td><td>WWC3 </td><td>K562</td><td> 0.145894223</td><td>3.421467e-03</td><td>0.377214467</td><td>0.4612307</td></tr>\n",
       "\t<tr><td>chrX</td><td>9864063</td><td>9864567</td><td>WWC3 </td><td>K562</td><td> 0.187272821</td><td>1.190708e-04</td><td>0.270717005</td><td>0.4612307</td></tr>\n",
       "\t<tr><td>chrX</td><td>9873821</td><td>9874426</td><td>WWC3 </td><td>K562</td><td> 0.075693801</td><td>1.568375e-01</td><td>0.274917513</td><td>0.4612307</td></tr>\n",
       "\t<tr><td>chrX</td><td>9882635</td><td>9883318</td><td>WWC3 </td><td>K562</td><td> 0.018624767</td><td>7.621862e-01</td><td>0.471237310</td><td>0.4612307</td></tr>\n",
       "\t<tr><td>chrX</td><td>9885160</td><td>9885475</td><td>WWC3 </td><td>K562</td><td> 0.002865084</td><td>9.654569e-01</td><td>0.101598985</td><td>0.4612307</td></tr>\n",
       "\t<tr><td>chrX</td><td>9894484</td><td>9895274</td><td>WWC3 </td><td>K562</td><td> 0.001953204</td><td>9.765916e-01</td><td>0.317131980</td><td>0.4612307</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>CLCN4</td><td>K562</td><td>-0.024709327</td><td>6.810957e-01</td><td>0.339911168</td><td>0.4253648</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995827</td><td>9996208</td><td>WWC3 </td><td>K562</td><td>-0.035515096</td><td>5.412425e-01</td><td>0.339911168</td><td>0.4612307</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 933499 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & Correlation & FDR & VarQATAC & VarQRNA\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 &  10001196 &  10001745 & CLSTN1   & K562 & -0.008569671 & 8.939265e-01 & 0.12179569 & 0.6649588\\\\\n",
       "\t chr1 &  10001196 &  10001745 & CTNNBIP1 & K562 &  0.109276509 & 3.331330e-02 & 0.12179569 & 0.6953106\\\\\n",
       "\t chr1 &  10001196 &  10001745 & KIF1B    & K562 & -0.085711347 & 1.039767e-01 & 0.12179569 & 0.8591714\\\\\n",
       "\t chr1 &  10001196 &  10001745 & LZIC     & K562 & -0.153890963 & 1.912432e-03 & 0.12179569 & 0.7068755\\\\\n",
       "\t chr1 &  10001196 &  10001745 & NMNAT1   & K562 &  0.008805695 & 8.909163e-01 & 0.12179569 & 0.7135119\\\\\n",
       "\t chr1 &  10001196 &  10001745 & RBP7     & K562 &  0.055430425 & 3.168859e-01 & 0.12179569 & 0.2445713\\\\\n",
       "\t chr1 &  10001196 &  10001745 & UBE4B    & K562 &  0.139513416 & 5.324126e-03 & 0.12179569 & 0.8348216\\\\\n",
       "\t chr1 & 100027917 & 100029695 & AGL      & K562 & -0.060843072 & 2.669373e-01 & 0.66776015 & 0.7150734\\\\\n",
       "\t chr1 & 100027917 & 100029695 & DBT      & K562 &  0.113399517 & 2.660451e-02 & 0.66776015 & 0.6860391\\\\\n",
       "\t chr1 & 100027917 & 100029695 & MFSD14A  & K562 & -0.033765151 & 5.633475e-01 & 0.66776015 & 0.4949983\\\\\n",
       "\t chr1 & 100027917 & 100029695 & RTCA     & K562 & -0.296069972 & 2.140068e-10 & 0.66776015 & 0.7483531\\\\\n",
       "\t chr1 & 100027917 & 100029695 & RTCA-AS1 & K562 & -0.078206426 & 1.420580e-01 & 0.66776015 & 0.6117211\\\\\n",
       "\t chr1 & 100027917 & 100029695 & SASS6    & K562 & -0.005903034 & 9.277100e-01 & 0.66776015 & 0.9037232\\\\\n",
       "\t chr1 & 100027917 & 100029695 & SLC35A3  & K562 & -0.031298666 & 5.947948e-01 & 0.66776015 & 0.6436344\\\\\n",
       "\t chr1 & 100027917 & 100029695 & TRMT13   & K562 &  0.051259410 & 3.588414e-01 & 0.66776015 & 0.7907090\\\\\n",
       "\t chr1 &  10003541 &  10003870 & CLSTN1   & K562 & -0.059352630 & 2.801764e-01 & 0.03770939 & 0.6649588\\\\\n",
       "\t chr1 &  10003541 &  10003870 & CTNNBIP1 & K562 & -0.082603548 & 1.186765e-01 & 0.03770939 & 0.6953106\\\\\n",
       "\t chr1 &  10003541 &  10003870 & KIF1B    & K562 & -0.057018137 & 3.017411e-01 & 0.03770939 & 0.8591714\\\\\n",
       "\t chr1 &  10003541 &  10003870 & LZIC     & K562 & -0.011783743 & 8.524567e-01 & 0.03770939 & 0.7068755\\\\\n",
       "\t chr1 &  10003541 &  10003870 & NMNAT1   & K562 & -0.039432846 & 4.929231e-01 & 0.03770939 & 0.7135119\\\\\n",
       "\t chr1 &  10003541 &  10003870 & RBP7     & K562 & -0.052121018 & 3.499480e-01 & 0.03770939 & 0.2445713\\\\\n",
       "\t chr1 &  10003541 &  10003870 & UBE4B    & K562 &  0.056684970 & 3.048841e-01 & 0.03770939 & 0.8348216\\\\\n",
       "\t chr1 & 100036824 & 100037384 & AGL      & K562 & -0.174757022 & 3.599259e-04 & 0.73982234 & 0.7150734\\\\\n",
       "\t chr1 & 100036824 & 100037384 & DBT      & K562 & -0.084248849 & 1.106875e-01 & 0.73982234 & 0.6860391\\\\\n",
       "\t chr1 & 100036824 & 100037384 & MFSD14A  & K562 & -0.108289741 & 3.512104e-02 & 0.73982234 & 0.4949983\\\\\n",
       "\t chr1 & 100036824 & 100037384 & RTCA     & K562 &  0.062065170 & 2.563674e-01 & 0.73982234 & 0.7483531\\\\\n",
       "\t chr1 & 100036824 & 100037384 & RTCA-AS1 & K562 & -0.159306953 & 1.265751e-03 & 0.73982234 & 0.6117211\\\\\n",
       "\t chr1 & 100036824 & 100037384 & SASS6    & K562 &  0.277445564 & 3.357707e-09 & 0.73982234 & 0.9037232\\\\\n",
       "\t chr1 & 100036824 & 100037384 & SLC35A3  & K562 &  0.001882806 & 9.774065e-01 & 0.73982234 & 0.6436344\\\\\n",
       "\t chr1 & 100036824 & 100037384 & TRMT13   & K562 &  0.331370468 & 6.330091e-13 & 0.73982234 & 0.7907090\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9515603 & 9516754 & TBL1X & K562 &  0.513764666 & 1.218433e-32 & 0.993857868 & 0.9171424\\\\\n",
       "\t chrX & 9519003 & 9519166 & TBL1X & K562 &  0.192456958 & 7.353360e-05 & 0.076395939 & 0.9171424\\\\\n",
       "\t chrX & 9523396 & 9523614 & TBL1X & K562 &  0.139923602 & 5.178057e-03 & 0.019790609 & 0.9171424\\\\\n",
       "\t chrX & 9526743 & 9527004 & TBL1X & K562 &  0.154020902 & 1.893822e-03 & 0.036770305 & 0.9171424\\\\\n",
       "\t chrX & 9527097 & 9527256 & TBL1X & K562 &  0.222023623 & 3.578773e-06 & 0.022461929 & 0.9171424\\\\\n",
       "\t chrX & 9527473 & 9527741 & TBL1X & K562 &  0.160792514 & 1.127579e-03 & 0.063705584 & 0.9171424\\\\\n",
       "\t chrX & 9530658 & 9531443 & TBL1X & K562 &  0.133005904 & 8.196805e-03 & 0.763521574 & 0.9171424\\\\\n",
       "\t chrX & 9533720 & 9534013 & TBL1X & K562 &  0.059816442 & 2.759620e-01 & 0.035831218 & 0.9171424\\\\\n",
       "\t chrX & 9538535 & 9539381 & TBL1X & K562 &  0.266441620 & 1.543532e-08 & 0.433432741 & 0.9171424\\\\\n",
       "\t chrX & 9540026 & 9540401 & TBL1X & K562 &  0.112173135 & 2.847067e-02 & 0.036357868 & 0.9171424\\\\\n",
       "\t chrX & 9542740 & 9543112 & TBL1X & K562 &  0.047767576 & 3.962190e-01 & 0.029961929 & 0.9171424\\\\\n",
       "\t chrX & 9550938 & 9551425 & TBL1X & K562 &  0.195457064 & 5.530382e-05 & 0.037906091 & 0.9171424\\\\\n",
       "\t chrX & 9554809 & 9554981 & TBL1X & K562 & -0.015526458 & 8.033311e-01 & 0.007722081 & 0.9171424\\\\\n",
       "\t chrX & 9555057 & 9556694 & TBL1X & K562 &  0.123634939 & 1.471748e-02 & 0.584714467 & 0.9171424\\\\\n",
       "\t chrX & 9569158 & 9569613 & TBL1X & K562 &  0.089513403 & 8.796009e-02 & 0.032595178 & 0.9171424\\\\\n",
       "\t chrX & 9586788 & 9586944 & TBL1X & K562 & -0.246656984 & 2.008704e-07 & 0.003826142 & 0.9171424\\\\\n",
       "\t chrX & 9587015 & 9587399 & TBL1X & K562 &  0.102109587 & 4.836585e-02 & 0.037290609 & 0.9171424\\\\\n",
       "\t chrX & 9642694 & 9643219 & TBL1X & K562 & -0.062598490 & 2.518334e-01 & 0.048115482 & 0.9171424\\\\\n",
       "\t chrX & 9646431 & 9647138 & TBL1X & K562 & -0.013798935 & 8.260575e-01 & 0.536230964 & 0.9171424\\\\\n",
       "\t chrX & 9670192 & 9670610 & TBL1X & K562 &  0.142847783 & 4.237276e-03 & 0.089638325 & 0.9171424\\\\\n",
       "\t chrX & 9707468 & 9708164 & TBL1X & K562 &  0.055507472 & 3.161291e-01 & 0.149505076 & 0.9171424\\\\\n",
       "\t chrX & 9708493 & 9709582 & TBL1X & K562 &  0.046906423 & 4.058372e-01 & 0.659003807 & 0.9171424\\\\\n",
       "\t chrX & 9786114 & 9786869 & WWC3  & K562 &  0.145894223 & 3.421467e-03 & 0.377214467 & 0.4612307\\\\\n",
       "\t chrX & 9864063 & 9864567 & WWC3  & K562 &  0.187272821 & 1.190708e-04 & 0.270717005 & 0.4612307\\\\\n",
       "\t chrX & 9873821 & 9874426 & WWC3  & K562 &  0.075693801 & 1.568375e-01 & 0.274917513 & 0.4612307\\\\\n",
       "\t chrX & 9882635 & 9883318 & WWC3  & K562 &  0.018624767 & 7.621862e-01 & 0.471237310 & 0.4612307\\\\\n",
       "\t chrX & 9885160 & 9885475 & WWC3  & K562 &  0.002865084 & 9.654569e-01 & 0.101598985 & 0.4612307\\\\\n",
       "\t chrX & 9894484 & 9895274 & WWC3  & K562 &  0.001953204 & 9.765916e-01 & 0.317131980 & 0.4612307\\\\\n",
       "\t chrX & 9995827 & 9996208 & CLCN4 & K562 & -0.024709327 & 6.810957e-01 & 0.339911168 & 0.4253648\\\\\n",
       "\t chrX & 9995827 & 9996208 & WWC3  & K562 & -0.035515096 & 5.412425e-01 & 0.339911168 & 0.4612307\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 933499 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | Correlation &lt;dbl&gt; | FDR &lt;dbl&gt; | VarQATAC &lt;dbl&gt; | VarQRNA &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 |  10001196 |  10001745 | CLSTN1   | K562 | -0.008569671 | 8.939265e-01 | 0.12179569 | 0.6649588 |\n",
       "| chr1 |  10001196 |  10001745 | CTNNBIP1 | K562 |  0.109276509 | 3.331330e-02 | 0.12179569 | 0.6953106 |\n",
       "| chr1 |  10001196 |  10001745 | KIF1B    | K562 | -0.085711347 | 1.039767e-01 | 0.12179569 | 0.8591714 |\n",
       "| chr1 |  10001196 |  10001745 | LZIC     | K562 | -0.153890963 | 1.912432e-03 | 0.12179569 | 0.7068755 |\n",
       "| chr1 |  10001196 |  10001745 | NMNAT1   | K562 |  0.008805695 | 8.909163e-01 | 0.12179569 | 0.7135119 |\n",
       "| chr1 |  10001196 |  10001745 | RBP7     | K562 |  0.055430425 | 3.168859e-01 | 0.12179569 | 0.2445713 |\n",
       "| chr1 |  10001196 |  10001745 | UBE4B    | K562 |  0.139513416 | 5.324126e-03 | 0.12179569 | 0.8348216 |\n",
       "| chr1 | 100027917 | 100029695 | AGL      | K562 | -0.060843072 | 2.669373e-01 | 0.66776015 | 0.7150734 |\n",
       "| chr1 | 100027917 | 100029695 | DBT      | K562 |  0.113399517 | 2.660451e-02 | 0.66776015 | 0.6860391 |\n",
       "| chr1 | 100027917 | 100029695 | MFSD14A  | K562 | -0.033765151 | 5.633475e-01 | 0.66776015 | 0.4949983 |\n",
       "| chr1 | 100027917 | 100029695 | RTCA     | K562 | -0.296069972 | 2.140068e-10 | 0.66776015 | 0.7483531 |\n",
       "| chr1 | 100027917 | 100029695 | RTCA-AS1 | K562 | -0.078206426 | 1.420580e-01 | 0.66776015 | 0.6117211 |\n",
       "| chr1 | 100027917 | 100029695 | SASS6    | K562 | -0.005903034 | 9.277100e-01 | 0.66776015 | 0.9037232 |\n",
       "| chr1 | 100027917 | 100029695 | SLC35A3  | K562 | -0.031298666 | 5.947948e-01 | 0.66776015 | 0.6436344 |\n",
       "| chr1 | 100027917 | 100029695 | TRMT13   | K562 |  0.051259410 | 3.588414e-01 | 0.66776015 | 0.7907090 |\n",
       "| chr1 |  10003541 |  10003870 | CLSTN1   | K562 | -0.059352630 | 2.801764e-01 | 0.03770939 | 0.6649588 |\n",
       "| chr1 |  10003541 |  10003870 | CTNNBIP1 | K562 | -0.082603548 | 1.186765e-01 | 0.03770939 | 0.6953106 |\n",
       "| chr1 |  10003541 |  10003870 | KIF1B    | K562 | -0.057018137 | 3.017411e-01 | 0.03770939 | 0.8591714 |\n",
       "| chr1 |  10003541 |  10003870 | LZIC     | K562 | -0.011783743 | 8.524567e-01 | 0.03770939 | 0.7068755 |\n",
       "| chr1 |  10003541 |  10003870 | NMNAT1   | K562 | -0.039432846 | 4.929231e-01 | 0.03770939 | 0.7135119 |\n",
       "| chr1 |  10003541 |  10003870 | RBP7     | K562 | -0.052121018 | 3.499480e-01 | 0.03770939 | 0.2445713 |\n",
       "| chr1 |  10003541 |  10003870 | UBE4B    | K562 |  0.056684970 | 3.048841e-01 | 0.03770939 | 0.8348216 |\n",
       "| chr1 | 100036824 | 100037384 | AGL      | K562 | -0.174757022 | 3.599259e-04 | 0.73982234 | 0.7150734 |\n",
       "| chr1 | 100036824 | 100037384 | DBT      | K562 | -0.084248849 | 1.106875e-01 | 0.73982234 | 0.6860391 |\n",
       "| chr1 | 100036824 | 100037384 | MFSD14A  | K562 | -0.108289741 | 3.512104e-02 | 0.73982234 | 0.4949983 |\n",
       "| chr1 | 100036824 | 100037384 | RTCA     | K562 |  0.062065170 | 2.563674e-01 | 0.73982234 | 0.7483531 |\n",
       "| chr1 | 100036824 | 100037384 | RTCA-AS1 | K562 | -0.159306953 | 1.265751e-03 | 0.73982234 | 0.6117211 |\n",
       "| chr1 | 100036824 | 100037384 | SASS6    | K562 |  0.277445564 | 3.357707e-09 | 0.73982234 | 0.9037232 |\n",
       "| chr1 | 100036824 | 100037384 | SLC35A3  | K562 |  0.001882806 | 9.774065e-01 | 0.73982234 | 0.6436344 |\n",
       "| chr1 | 100036824 | 100037384 | TRMT13   | K562 |  0.331370468 | 6.330091e-13 | 0.73982234 | 0.7907090 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9515603 | 9516754 | TBL1X | K562 |  0.513764666 | 1.218433e-32 | 0.993857868 | 0.9171424 |\n",
       "| chrX | 9519003 | 9519166 | TBL1X | K562 |  0.192456958 | 7.353360e-05 | 0.076395939 | 0.9171424 |\n",
       "| chrX | 9523396 | 9523614 | TBL1X | K562 |  0.139923602 | 5.178057e-03 | 0.019790609 | 0.9171424 |\n",
       "| chrX | 9526743 | 9527004 | TBL1X | K562 |  0.154020902 | 1.893822e-03 | 0.036770305 | 0.9171424 |\n",
       "| chrX | 9527097 | 9527256 | TBL1X | K562 |  0.222023623 | 3.578773e-06 | 0.022461929 | 0.9171424 |\n",
       "| chrX | 9527473 | 9527741 | TBL1X | K562 |  0.160792514 | 1.127579e-03 | 0.063705584 | 0.9171424 |\n",
       "| chrX | 9530658 | 9531443 | TBL1X | K562 |  0.133005904 | 8.196805e-03 | 0.763521574 | 0.9171424 |\n",
       "| chrX | 9533720 | 9534013 | TBL1X | K562 |  0.059816442 | 2.759620e-01 | 0.035831218 | 0.9171424 |\n",
       "| chrX | 9538535 | 9539381 | TBL1X | K562 |  0.266441620 | 1.543532e-08 | 0.433432741 | 0.9171424 |\n",
       "| chrX | 9540026 | 9540401 | TBL1X | K562 |  0.112173135 | 2.847067e-02 | 0.036357868 | 0.9171424 |\n",
       "| chrX | 9542740 | 9543112 | TBL1X | K562 |  0.047767576 | 3.962190e-01 | 0.029961929 | 0.9171424 |\n",
       "| chrX | 9550938 | 9551425 | TBL1X | K562 |  0.195457064 | 5.530382e-05 | 0.037906091 | 0.9171424 |\n",
       "| chrX | 9554809 | 9554981 | TBL1X | K562 | -0.015526458 | 8.033311e-01 | 0.007722081 | 0.9171424 |\n",
       "| chrX | 9555057 | 9556694 | TBL1X | K562 |  0.123634939 | 1.471748e-02 | 0.584714467 | 0.9171424 |\n",
       "| chrX | 9569158 | 9569613 | TBL1X | K562 |  0.089513403 | 8.796009e-02 | 0.032595178 | 0.9171424 |\n",
       "| chrX | 9586788 | 9586944 | TBL1X | K562 | -0.246656984 | 2.008704e-07 | 0.003826142 | 0.9171424 |\n",
       "| chrX | 9587015 | 9587399 | TBL1X | K562 |  0.102109587 | 4.836585e-02 | 0.037290609 | 0.9171424 |\n",
       "| chrX | 9642694 | 9643219 | TBL1X | K562 | -0.062598490 | 2.518334e-01 | 0.048115482 | 0.9171424 |\n",
       "| chrX | 9646431 | 9647138 | TBL1X | K562 | -0.013798935 | 8.260575e-01 | 0.536230964 | 0.9171424 |\n",
       "| chrX | 9670192 | 9670610 | TBL1X | K562 |  0.142847783 | 4.237276e-03 | 0.089638325 | 0.9171424 |\n",
       "| chrX | 9707468 | 9708164 | TBL1X | K562 |  0.055507472 | 3.161291e-01 | 0.149505076 | 0.9171424 |\n",
       "| chrX | 9708493 | 9709582 | TBL1X | K562 |  0.046906423 | 4.058372e-01 | 0.659003807 | 0.9171424 |\n",
       "| chrX | 9786114 | 9786869 | WWC3  | K562 |  0.145894223 | 3.421467e-03 | 0.377214467 | 0.4612307 |\n",
       "| chrX | 9864063 | 9864567 | WWC3  | K562 |  0.187272821 | 1.190708e-04 | 0.270717005 | 0.4612307 |\n",
       "| chrX | 9873821 | 9874426 | WWC3  | K562 |  0.075693801 | 1.568375e-01 | 0.274917513 | 0.4612307 |\n",
       "| chrX | 9882635 | 9883318 | WWC3  | K562 |  0.018624767 | 7.621862e-01 | 0.471237310 | 0.4612307 |\n",
       "| chrX | 9885160 | 9885475 | WWC3  | K562 |  0.002865084 | 9.654569e-01 | 0.101598985 | 0.4612307 |\n",
       "| chrX | 9894484 | 9895274 | WWC3  | K562 |  0.001953204 | 9.765916e-01 | 0.317131980 | 0.4612307 |\n",
       "| chrX | 9995827 | 9996208 | CLCN4 | K562 | -0.024709327 | 6.810957e-01 | 0.339911168 | 0.4253648 |\n",
       "| chrX | 9995827 | 9996208 | WWC3  | K562 | -0.035515096 | 5.412425e-01 | 0.339911168 | 0.4612307 |\n",
       "\n"
      ],
      "text/plain": [
       "       chr  start     end       TargetGene CellType Correlation  FDR         \n",
       "1      chr1  10001196  10001745 CLSTN1     K562     -0.008569671 8.939265e-01\n",
       "2      chr1  10001196  10001745 CTNNBIP1   K562      0.109276509 3.331330e-02\n",
       "3      chr1  10001196  10001745 KIF1B      K562     -0.085711347 1.039767e-01\n",
       "4      chr1  10001196  10001745 LZIC       K562     -0.153890963 1.912432e-03\n",
       "5      chr1  10001196  10001745 NMNAT1     K562      0.008805695 8.909163e-01\n",
       "6      chr1  10001196  10001745 RBP7       K562      0.055430425 3.168859e-01\n",
       "7      chr1  10001196  10001745 UBE4B      K562      0.139513416 5.324126e-03\n",
       "8      chr1 100027917 100029695 AGL        K562     -0.060843072 2.669373e-01\n",
       "9      chr1 100027917 100029695 DBT        K562      0.113399517 2.660451e-02\n",
       "10     chr1 100027917 100029695 MFSD14A    K562     -0.033765151 5.633475e-01\n",
       "11     chr1 100027917 100029695 RTCA       K562     -0.296069972 2.140068e-10\n",
       "12     chr1 100027917 100029695 RTCA-AS1   K562     -0.078206426 1.420580e-01\n",
       "13     chr1 100027917 100029695 SASS6      K562     -0.005903034 9.277100e-01\n",
       "14     chr1 100027917 100029695 SLC35A3    K562     -0.031298666 5.947948e-01\n",
       "15     chr1 100027917 100029695 TRMT13     K562      0.051259410 3.588414e-01\n",
       "16     chr1  10003541  10003870 CLSTN1     K562     -0.059352630 2.801764e-01\n",
       "17     chr1  10003541  10003870 CTNNBIP1   K562     -0.082603548 1.186765e-01\n",
       "18     chr1  10003541  10003870 KIF1B      K562     -0.057018137 3.017411e-01\n",
       "19     chr1  10003541  10003870 LZIC       K562     -0.011783743 8.524567e-01\n",
       "20     chr1  10003541  10003870 NMNAT1     K562     -0.039432846 4.929231e-01\n",
       "21     chr1  10003541  10003870 RBP7       K562     -0.052121018 3.499480e-01\n",
       "22     chr1  10003541  10003870 UBE4B      K562      0.056684970 3.048841e-01\n",
       "23     chr1 100036824 100037384 AGL        K562     -0.174757022 3.599259e-04\n",
       "24     chr1 100036824 100037384 DBT        K562     -0.084248849 1.106875e-01\n",
       "25     chr1 100036824 100037384 MFSD14A    K562     -0.108289741 3.512104e-02\n",
       "26     chr1 100036824 100037384 RTCA       K562      0.062065170 2.563674e-01\n",
       "27     chr1 100036824 100037384 RTCA-AS1   K562     -0.159306953 1.265751e-03\n",
       "28     chr1 100036824 100037384 SASS6      K562      0.277445564 3.357707e-09\n",
       "29     chr1 100036824 100037384 SLC35A3    K562      0.001882806 9.774065e-01\n",
       "30     chr1 100036824 100037384 TRMT13     K562      0.331370468 6.330091e-13\n",
       "⋮      ⋮    ⋮         ⋮         ⋮          ⋮        ⋮            ⋮           \n",
       "933470 chrX 9515603   9516754   TBL1X      K562      0.513764666 1.218433e-32\n",
       "933471 chrX 9519003   9519166   TBL1X      K562      0.192456958 7.353360e-05\n",
       "933472 chrX 9523396   9523614   TBL1X      K562      0.139923602 5.178057e-03\n",
       "933473 chrX 9526743   9527004   TBL1X      K562      0.154020902 1.893822e-03\n",
       "933474 chrX 9527097   9527256   TBL1X      K562      0.222023623 3.578773e-06\n",
       "933475 chrX 9527473   9527741   TBL1X      K562      0.160792514 1.127579e-03\n",
       "933476 chrX 9530658   9531443   TBL1X      K562      0.133005904 8.196805e-03\n",
       "933477 chrX 9533720   9534013   TBL1X      K562      0.059816442 2.759620e-01\n",
       "933478 chrX 9538535   9539381   TBL1X      K562      0.266441620 1.543532e-08\n",
       "933479 chrX 9540026   9540401   TBL1X      K562      0.112173135 2.847067e-02\n",
       "933480 chrX 9542740   9543112   TBL1X      K562      0.047767576 3.962190e-01\n",
       "933481 chrX 9550938   9551425   TBL1X      K562      0.195457064 5.530382e-05\n",
       "933482 chrX 9554809   9554981   TBL1X      K562     -0.015526458 8.033311e-01\n",
       "933483 chrX 9555057   9556694   TBL1X      K562      0.123634939 1.471748e-02\n",
       "933484 chrX 9569158   9569613   TBL1X      K562      0.089513403 8.796009e-02\n",
       "933485 chrX 9586788   9586944   TBL1X      K562     -0.246656984 2.008704e-07\n",
       "933486 chrX 9587015   9587399   TBL1X      K562      0.102109587 4.836585e-02\n",
       "933487 chrX 9642694   9643219   TBL1X      K562     -0.062598490 2.518334e-01\n",
       "933488 chrX 9646431   9647138   TBL1X      K562     -0.013798935 8.260575e-01\n",
       "933489 chrX 9670192   9670610   TBL1X      K562      0.142847783 4.237276e-03\n",
       "933490 chrX 9707468   9708164   TBL1X      K562      0.055507472 3.161291e-01\n",
       "933491 chrX 9708493   9709582   TBL1X      K562      0.046906423 4.058372e-01\n",
       "933492 chrX 9786114   9786869   WWC3       K562      0.145894223 3.421467e-03\n",
       "933493 chrX 9864063   9864567   WWC3       K562      0.187272821 1.190708e-04\n",
       "933494 chrX 9873821   9874426   WWC3       K562      0.075693801 1.568375e-01\n",
       "933495 chrX 9882635   9883318   WWC3       K562      0.018624767 7.621862e-01\n",
       "933496 chrX 9885160   9885475   WWC3       K562      0.002865084 9.654569e-01\n",
       "933497 chrX 9894484   9895274   WWC3       K562      0.001953204 9.765916e-01\n",
       "933498 chrX 9995827   9996208   CLCN4      K562     -0.024709327 6.810957e-01\n",
       "933499 chrX 9995827   9996208   WWC3       K562     -0.035515096 5.412425e-01\n",
       "       VarQATAC    VarQRNA  \n",
       "1      0.12179569  0.6649588\n",
       "2      0.12179569  0.6953106\n",
       "3      0.12179569  0.8591714\n",
       "4      0.12179569  0.7068755\n",
       "5      0.12179569  0.7135119\n",
       "6      0.12179569  0.2445713\n",
       "7      0.12179569  0.8348216\n",
       "8      0.66776015  0.7150734\n",
       "9      0.66776015  0.6860391\n",
       "10     0.66776015  0.4949983\n",
       "11     0.66776015  0.7483531\n",
       "12     0.66776015  0.6117211\n",
       "13     0.66776015  0.9037232\n",
       "14     0.66776015  0.6436344\n",
       "15     0.66776015  0.7907090\n",
       "16     0.03770939  0.6649588\n",
       "17     0.03770939  0.6953106\n",
       "18     0.03770939  0.8591714\n",
       "19     0.03770939  0.7068755\n",
       "20     0.03770939  0.7135119\n",
       "21     0.03770939  0.2445713\n",
       "22     0.03770939  0.8348216\n",
       "23     0.73982234  0.7150734\n",
       "24     0.73982234  0.6860391\n",
       "25     0.73982234  0.4949983\n",
       "26     0.73982234  0.7483531\n",
       "27     0.73982234  0.6117211\n",
       "28     0.73982234  0.9037232\n",
       "29     0.73982234  0.6436344\n",
       "30     0.73982234  0.7907090\n",
       "⋮      ⋮           ⋮        \n",
       "933470 0.993857868 0.9171424\n",
       "933471 0.076395939 0.9171424\n",
       "933472 0.019790609 0.9171424\n",
       "933473 0.036770305 0.9171424\n",
       "933474 0.022461929 0.9171424\n",
       "933475 0.063705584 0.9171424\n",
       "933476 0.763521574 0.9171424\n",
       "933477 0.035831218 0.9171424\n",
       "933478 0.433432741 0.9171424\n",
       "933479 0.036357868 0.9171424\n",
       "933480 0.029961929 0.9171424\n",
       "933481 0.037906091 0.9171424\n",
       "933482 0.007722081 0.9171424\n",
       "933483 0.584714467 0.9171424\n",
       "933484 0.032595178 0.9171424\n",
       "933485 0.003826142 0.9171424\n",
       "933486 0.037290609 0.9171424\n",
       "933487 0.048115482 0.9171424\n",
       "933488 0.536230964 0.9171424\n",
       "933489 0.089638325 0.9171424\n",
       "933490 0.149505076 0.9171424\n",
       "933491 0.659003807 0.9171424\n",
       "933492 0.377214467 0.4612307\n",
       "933493 0.270717005 0.4612307\n",
       "933494 0.274917513 0.4612307\n",
       "933495 0.471237310 0.4612307\n",
       "933496 0.101598985 0.4612307\n",
       "933497 0.317131980 0.4612307\n",
       "933498 0.339911168 0.4253648\n",
       "933499 0.339911168 0.4612307"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(obj.ArchR,\n",
    "        paste(dir.output,\"obj.ArchR.rds\",sep = \"/\"))\n",
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"Correlation\",\n",
    "                         \"FDR\",\n",
    "                         \"VarQATAC\",\n",
    "                         \"VarQRNA\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "pairs.E2G.res\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4c909ed5-7321-4e69-8208-1544b371da6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] nabor_0.5.0                       dplyr_1.1.4                      \n",
       " [3] uwot_0.1.16                       Rsamtools_2.18.0                 \n",
       " [5] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.68.0                  \n",
       " [7] rtracklayer_1.60.1                Biostrings_2.70.1                \n",
       " [9] XVector_0.42.0                    genomation_1.34.0                \n",
       "[11] Signac_1.13.0                     Seurat_5.0.3                     \n",
       "[13] SeuratObject_5.0.1                sp_2.1-4                         \n",
       "[15] rhdf5_2.44.0                      SummarizedExperiment_1.30.2      \n",
       "[17] Biobase_2.60.0                    MatrixGenerics_1.12.3            \n",
       "[19] Rcpp_1.0.12                       Matrix_1.6-5                     \n",
       "[21] GenomicRanges_1.54.1              GenomeInfoDb_1.38.1              \n",
       "[23] IRanges_2.36.0                    S4Vectors_0.40.2                 \n",
       "[25] BiocGenerics_0.48.1               matrixStats_1.3.0                \n",
       "[27] data.table_1.15.2                 stringr_1.5.1                    \n",
       "[29] plyr_1.8.9                        magrittr_2.0.3                   \n",
       "[31] ggplot2_3.5.1                     gtable_0.3.5                     \n",
       "[33] gtools_3.9.5                      gridExtra_2.3                    \n",
       "[35] ArchR_1.0.2                      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22         splines_4.3.3            later_1.3.2             \n",
       "  [4] pbdZMQ_0.3-11            BiocIO_1.10.0            bitops_1.0-7            \n",
       "  [7] tibble_3.2.1             polyclip_1.10-6          XML_3.99-0.16.1         \n",
       " [10] fastDummies_1.7.3        lifecycle_1.0.4          vroom_1.6.5             \n",
       " [13] globals_0.16.3           lattice_0.22-6           MASS_7.3-60             \n",
       " [16] plotly_4.10.4            plotrix_3.8-4            yaml_2.3.8              \n",
       " [19] httpuv_1.6.15            sctransform_0.4.1        spam_2.10-0             \n",
       " [22] spatstat.sparse_3.0-3    reticulate_1.36.1        cowplot_1.1.3           \n",
       " [25] pbapply_1.7-2            RColorBrewer_1.1-3       abind_1.4-5             \n",
       " [28] zlibbioc_1.48.0          Rtsne_0.17               purrr_1.0.2             \n",
       " [31] RCurl_1.98-1.14          GenomeInfoDbData_1.2.11  ggrepel_0.9.5           \n",
       " [34] irlba_2.3.5.1            listenv_0.9.1            spatstat.utils_3.0-4    \n",
       " [37] goftest_1.2-3            RSpectra_0.16-1          spatstat.random_3.2-3   \n",
       " [40] fitdistrplus_1.1-11      parallelly_1.37.1        leiden_0.4.3.1          \n",
       " [43] codetools_0.2-20         DelayedArray_0.26.7      RcppRoll_0.3.0          \n",
       " [46] tidyselect_1.2.1         farver_2.1.1             base64enc_0.1-3         \n",
       " [49] spatstat.explore_3.2-6   GenomicAlignments_1.36.0 jsonlite_1.8.8          \n",
       " [52] progressr_0.14.0         ggridges_0.5.6           survival_3.6-4          \n",
       " [55] tools_4.3.3              ica_1.0-3                glue_1.7.0              \n",
       " [58] IRdisplay_1.1            withr_3.0.0              fastmap_1.1.1           \n",
       " [61] rhdf5filters_1.12.1      fansi_1.0.6              digest_0.6.35           \n",
       " [64] R6_2.5.1                 mime_0.12                seqPattern_1.34.0       \n",
       " [67] colorspace_2.1-0         Cairo_1.6-2              scattermore_1.2         \n",
       " [70] tensor_1.5               spatstat.data_3.0-4      utf8_1.2.4              \n",
       " [73] tidyr_1.3.1              generics_0.1.3           httr_1.4.7              \n",
       " [76] htmlwidgets_1.6.4        S4Arrays_1.2.1           pkgconfig_2.0.3         \n",
       " [79] impute_1.76.0            lmtest_0.9-40            htmltools_0.5.8.1       \n",
       " [82] dotCall64_1.1-1          scales_1.3.0             png_0.1-8               \n",
       " [85] tzdb_0.4.0               reshape2_1.4.4           rjson_0.2.21            \n",
       " [88] uuid_1.2-0               nlme_3.1-164             repr_1.1.7              \n",
       " [91] zoo_1.8-12               KernSmooth_2.23-22       miniUI_0.1.1.1          \n",
       " [94] restfulr_0.0.15          pillar_1.9.0             vctrs_0.6.5             \n",
       " [97] RANN_2.6.1               promises_1.3.0           xtable_1.8-4            \n",
       "[100] cluster_2.1.6            evaluate_0.23            readr_2.1.5             \n",
       "[103] cli_3.6.2                compiler_4.3.3           rlang_1.1.3             \n",
       "[106] crayon_1.5.2             future.apply_1.11.2      labeling_0.4.3          \n",
       "[109] stringi_1.8.3            gridBase_0.4-7           viridisLite_0.4.2       \n",
       "[112] deldir_2.0-4             BiocParallel_1.36.0      munsell_0.5.1           \n",
       "[115] lazyeval_0.2.2           spatstat.geom_3.2-9      IRkernel_1.3.2          \n",
       "[118] RcppHNSW_0.6.0           hms_1.1.3                patchwork_1.2.0         \n",
       "[121] bit64_4.0.5              future_1.33.2            Rhdf5lib_1.22.1         \n",
       "[124] shiny_1.8.1.1            ROCR_1.0-11              igraph_2.0.3            \n",
       "[127] fastmatch_1.1-4          bit_4.0.5               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b54012ba-20d1-4787-8d0d-d851ef1744c4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
