{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8466b35-cae5-4f62-8138-37b2de484a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.2\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : grid v4.3.3\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "\tLoading Package : gtools v3.9.5\n",
      "\n",
      "\tLoading Package : gtable v0.3.5\n",
      "\n",
      "\tLoading Package : ggplot2 v3.5.1\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "\tLoading Package : plyr v1.8.9\n",
      "\n",
      "\tLoading Package : stringr v1.5.1\n",
      "\n",
      "\tLoading Package : data.table v1.15.2\n",
      "\n",
      "\tLoading Package : matrixStats v1.3.0\n",
      "\n",
      "\tLoading Package : S4Vectors v0.40.2\n",
      "\n",
      "\tLoading Package : GenomicRanges v1.54.1\n",
      "\n",
      "\tLoading Package : BiocGenerics v0.48.1\n",
      "\n",
      "\tLoading Package : Matrix v1.6.5\n",
      "\n",
      "\tLoading Package : Rcpp v1.0.12\n",
      "\n",
      "\tLoading Package : SummarizedExperiment v1.30.2\n",
      "\n",
      "\tLoading Package : rhdf5 v2.44.0\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(magrittr)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e85020d7-d358-492d-9358-c81270f4b624",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1bec6690-52ec-450d-bba7-7d4528a185cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.seurat = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/5.merge_peaks_from_5_super_groups.240524/obj.seurat.rds\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/5.merge_peaks_from_5_super_groups.240524/pairs.rds\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/ArchR_250226/BMMC5.merge.250226/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "340dac32-44ac-4231-b964-b5f3b9e2dbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16\n",
    "maxDist = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da973385-4c81-4b1c-9669-cf5933b23a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 16.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRThreads(threads = n.cores) \n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeb26496-d3b9-4778-92ed-9c63a5018aef",
   "metadata": {},
   "source": [
    "Import seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c9b19a09-be95-4d9a-bdc3-0530980d4757",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.seurat = readRDS(path.seurat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9ed4b927-8c9e-4270-a6a2-f7cfca6fc95c",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.seurat = subset(obj.seurat,\n",
    "                    cells = colnames(obj.seurat)[!is.na(obj.seurat$cell_type.merged)])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86be3e66-c954-4620-9e64-c6a1c0fcb84c",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6f363969-bf0b-45ce-858c-a8cb64cc210a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readRDS(path.pairs.E2G)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3538ffdc-4fed-48eb-9c9d-86805a91cf06",
   "metadata": {},
   "source": [
    "Create an ArchRProject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0300f485-fafa-4cef-b9d0-534108ec6f95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 13\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"86350e844e0a3a1991e9c650ccac7f43\" \"555000121b665f095f13712669597ac8\"\n",
      "  .. ..@ cells: Named chr [1:5315] \"TAGTTGTCACCCTCAC-1\" \"CTATGGCCATAACGGG-1\" \"CCGCACACAGGTTAAA-1\" \"TCATTTGGTAATGGAA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5315] \"s1d1_TAGTTGTCACCCTCAC-1\" \"s1d1_CTATGGCCATAACGGG-1\" \"s1d1_CCGCACACAGGTTAAA-1\" \"s1d1_TCATTTGGTAATGGAA-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"3ee058ae058031e3f233c68b0db9e6b9\" \"1039e761a9da92608ebabb952d11b3c8\"\n",
      "  .. ..@ cells: Named chr [1:5246] \"GATGCAGCAGCAACAG-1\" \"TGAGTGAAGAGGAAGG-1\" \"CCGCTAAAGGACCGCT-1\" \"CTAACCACAGTTTGGC-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5246] \"s1d2_GATGCAGCAGCAACAG-1\" \"s1d2_TGAGTGAAGAGGAAGG-1\" \"s1d2_CCGCTAAAGGACCGCT-1\" \"s1d2_CTAACCACAGTTTGGC-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"2276352420919cf89b8bb75ac819824c\" \"75537e8fd45162781c38e6a7db2c86cf\"\n",
      "  .. ..@ cells: Named chr [1:3354] \"TCACCTGGTTAGGTTG-1\" \"CGTTAACAGGTGTCCA-1\" \"GGACCGAAGTGAGGTA-1\" \"AGTGCGGAGTAAGGGC-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:3354] \"s1d3_TCACCTGGTTAGGTTG-1\" \"s1d3_CGTTAACAGGTGTCCA-1\" \"s1d3_GGACCGAAGTGAGGTA-1\" \"s1d3_AGTGCGGAGTAAGGGC-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"1fc4f6dc8e1c53094db34b5c7a766a87\" \"b8c37340d7a1d3941d55772be354bd9a\"\n",
      "  .. ..@ cells: Named chr [1:3665] \"ACGTTACAGGCATTAC-1\" \"GGTGATTTCGCTAGAT-1\" \"ACAGGATCACTAAGAA-1\" \"CGCTACTTCATCCACC-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:3665] \"s2d1_ACGTTACAGGCATTAC-1\" \"s2d1_GGTGATTTCGCTAGAT-1\" \"s2d1_ACAGGATCACTAAGAA-1\" \"s2d1_CGCTACTTCATCCACC-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"a2acfe30c69e24ee4fd35cd8b067c630\" \"9ffc0908f69ebc6daf04984782449572\"\n",
      "  .. ..@ cells: Named chr [1:5556] \"TAATGGTGTTAGCGTA-1\" \"TCTAACTTCGTTTCGC-1\" \"GGCGCTAGTTTAGCTC-1\" \"AAAGCACCACTAAGCC-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5556] \"s2d4_TAATGGTGTTAGCGTA-1\" \"s2d4_TCTAACTTCGTTTCGC-1\" \"s2d4_GGCGCTAGTTTAGCTC-1\" \"s2d4_AAAGCACCACTAAGCC-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"2176bbdb8bb8100786137c9f2a51b3ca\" \"90506bed3c58854bc99c88adb8d34e91\"\n",
      "  .. ..@ cells: Named chr [1:3916] \"GCTTGACCATCTAGCA-1\" \"GATTCAATCTAATCTC-1\" \"TAAGGACCAAAGCGCA-1\" \"AGCGTGCTCATTGCAA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:3916] \"s2d5_GCTTGACCATCTAGCA-1\" \"s2d5_GATTCAATCTAATCTC-1\" \"s2d5_TAAGGACCAAAGCGCA-1\" \"s2d5_AGCGTGCTCATTGCAA-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"2a5bab58fd0b28bc8ef3723ed57e6c08\" \"238b44ce8ad4cb929abf90c361c5d9f1\"\n",
      "  .. ..@ cells: Named chr [1:5447] \"AGGAGCTAGACAAGTG-1\" \"GCGTAACGTCATAGAT-1\" \"CGATCCTTCGAGGTGG-1\" \"GATTCGCCAAATACCT-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:5447] \"s3d10_AGGAGCTAGACAAGTG-1\" \"s3d10_GCGTAACGTCATAGAT-1\" \"s3d10_CGATCCTTCGAGGTGG-1\" \"s3d10_GATTCGCCAAATACCT-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"983f45aef00cf12883d18d626d839a36\" \"72f5dc7b926b9533af5db0f68b5b69f6\"\n",
      "  .. ..@ cells: Named chr [1:3553] \"AGTCGCATCTCACTAT-1\" \"ATGTCATCAATCTCTC-1\" \"ATTGACTCAGGAATCG-1\" \"CTAATGTCACAACAGG-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:3553] \"s3d3_AGTCGCATCTCACTAT-1\" \"s3d3_ATGTCATCAATCTCTC-1\" \"s3d3_ATTGACTCAGGAATCG-1\" \"s3d3_CTAATGTCACAACAGG-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"5d837fe9f6aafc00db7add9a9fe3ef29\" \"b296ae5676ca1413959c46f7df802834\"\n",
      "  .. ..@ cells: Named chr [1:1452] \"GTACAATGTTCCATTA-1\" \"TGTGGCGGTGTTGCAA-1\" \"GCGGTTATCGACCTGA-1\" \"CCTGTAACAGGCTTCG-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:1452] \"s3d6_GTACAATGTTCCATTA-1\" \"s3d6_TGTGGCGGTGTTGCAA-1\" \"s3d6_GCGGTTATCGACCTGA-1\" \"s3d6_CCTGTAACAGGCTTCG-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"4f4da183ea8734759ddc8987b29aac25\" \"d9b91a39dbf76399e9f01458cf21e627\"\n",
      "  .. ..@ cells: Named chr [1:1450] \"CCAAACCCACTAGCGT-1\" \"TTAACTGAGAAACTGG-1\" \"GGAGCGATCACAGCCA-1\" \"CGCTGTGCAGGCATGA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:1450] \"s3d7_CCAAACCCACTAGCGT-1\" \"s3d7_TTAACTGAGAAACTGG-1\" \"s3d7_GGAGCGATCACAGCCA-1\" \"s3d7_CGCTGTGCAGGCATGA-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"c1bcab73870a86531668a7e4c5fd9f82\" \"fb36f28f535f141bf0a4af04c17d5bd9\"\n",
      "  .. ..@ cells: Named chr [1:6117] \"AGCGATTTCATCCTGC-1\" \"TTGACTAAGGTCAAAG-1\" \"AGGAACGGTGAGGTGA-1\" \"TAGGGTTTCCGTGACA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:6117] \"s4d1_AGCGATTTCATCCTGC-1\" \"s4d1_TTGACTAAGGTCAAAG-1\" \"s4d1_AGGAACGGTGAGGTGA-1\" \"s4d1_TAGGGTTTCCGTGACA-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"65a18b88a6a8f2e1b7282c2c08da1067\" \"cc03488d12a1f1e49ae43613e0c61606\"\n",
      "  .. ..@ cells: Named chr [1:8874] \"TATCCAGCAATTGAGA-1\" \"ACTCGCGCAAACTGTT-1\" \"GACTATTCATGTCGCG-1\" \"CTAATGTCATTGTTGG-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:8874] \"s4d8_TATCCAGCAATTGAGA-1\" \"s4d8_ACTCGCGCAAACTGTT-1\" \"s4d8_GACTATTCATGTCGCG-1\" \"s4d8_CTAATGTCATTGTTGG-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"10d377a58a8fbc07927e8d3a4ec79e73\" \"e3dbabb04370289ef70c67216288b530\"\n",
      "  .. ..@ cells: Named chr [1:3426] \"TGTCCTTCAGGCTAGA-1\" \"AATACCGGTAAGGTCC-1\" \"TGTAAAGCACTTGTTC-1\" \"CAGTATGGTTCGCTCA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:3426] \"s4d9_TGTCCTTCAGGCTAGA-1\" \"s4d9_AATACCGGTAAGGTCC-1\" \"s4d9_TGTAAAGCACTTGTTC-1\" \"s4d9_CAGTATGGTTCGCTCA-1\" ...\n"
     ]
    }
   ],
   "source": [
    "str(Fragments(obj.seurat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1753ffaa-89e2-43bf-bcf6-19937782d63a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-createArrows-cb5e9429f6bd2-Date-2025-02-26_Time-15-00-52.852547.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Cleaning Temporary Files\n",
      "\n",
      "2025-02-26 15:00:53.561106 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-createArrows-cb5e9429f6bd2-Date-2025-02-26_Time-15-00-52.852547.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ArrowFiles <- createArrowFiles(\n",
    "  inputFiles = c(Fragments(obj.seurat)[[1]]@path,\n",
    "                 Fragments(obj.seurat)[[2]]@path,\n",
    "                 Fragments(obj.seurat)[[3]]@path,\n",
    "                 Fragments(obj.seurat)[[4]]@path,\n",
    "                 Fragments(obj.seurat)[[5]]@path,\n",
    "                 Fragments(obj.seurat)[[6]]@path,\n",
    "                 Fragments(obj.seurat)[[7]]@path,\n",
    "                 Fragments(obj.seurat)[[8]]@path,\n",
    "                 Fragments(obj.seurat)[[9]]@path,\n",
    "                 Fragments(obj.seurat)[[10]]@path,\n",
    "                 Fragments(obj.seurat)[[11]]@path,\n",
    "                 Fragments(obj.seurat)[[12]]@path,\n",
    "                 Fragments(obj.seurat)[[13]]@path),\n",
    "  sampleNames = c(\"s1d1\",\n",
    "                  \"s1d2\",\n",
    "                  \"s1d3\",\n",
    "                  \"s2d1\",\n",
    "                  \"s2d4\",\n",
    "                  \"s2d5\",\n",
    "                  \"s3d10\",\n",
    "                  \"s3d3\",\n",
    "                  \"s3d6\",\n",
    "                  \"s3d7\",\n",
    "                  \"s4d1\",\n",
    "                  \"s4d8\",\n",
    "                  \"s4d9\"),\n",
    "  validBarcodes = list(s1d1 = unname(Fragments(obj.seurat)[[1]]@cells),\n",
    "                       s1d2 = unname(Fragments(obj.seurat)[[2]]@cells),\n",
    "                       s1d3 = unname(Fragments(obj.seurat)[[3]]@cells),\n",
    "                       s2d1 = unname(Fragments(obj.seurat)[[4]]@cells),\n",
    "                       s2d4 = unname(Fragments(obj.seurat)[[5]]@cells),\n",
    "                       s2d5 = unname(Fragments(obj.seurat)[[6]]@cells),\n",
    "                       s3d10 = unname(Fragments(obj.seurat)[[7]]@cells),\n",
    "                       s3d3 = unname(Fragments(obj.seurat)[[8]]@cells),\n",
    "                       s3d6 = unname(Fragments(obj.seurat)[[9]]@cells),\n",
    "                       s3d7 = unname(Fragments(obj.seurat)[[10]]@cells),\n",
    "                       s4d1 = unname(Fragments(obj.seurat)[[11]]@cells),\n",
    "                       s4d8 = unname(Fragments(obj.seurat)[[12]]@cells),\n",
    "                       s4d9 = unname(Fragments(obj.seurat)[[13]]@cells)),\n",
    "  addTileMat = T,\n",
    "  addGeneScoreMat = F,\n",
    "  force = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ada3db45-d050-4659-a7cf-2891e04b1249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'s3d10.arrow'</li><li>'s3d3.arrow'</li><li>'s3d7.arrow'</li><li>'s4d1.arrow'</li><li>'s1d1.arrow'</li><li>'s4d8.arrow'</li><li>'s2d4.arrow'</li><li>'s2d1.arrow'</li><li>'s2d5.arrow'</li><li>'s1d2.arrow'</li><li>'s1d3.arrow'</li><li>'s4d9.arrow'</li><li>'s3d6.arrow'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 's3d10.arrow'\n",
       "\\item 's3d3.arrow'\n",
       "\\item 's3d7.arrow'\n",
       "\\item 's4d1.arrow'\n",
       "\\item 's1d1.arrow'\n",
       "\\item 's4d8.arrow'\n",
       "\\item 's2d4.arrow'\n",
       "\\item 's2d1.arrow'\n",
       "\\item 's2d5.arrow'\n",
       "\\item 's1d2.arrow'\n",
       "\\item 's1d3.arrow'\n",
       "\\item 's4d9.arrow'\n",
       "\\item 's3d6.arrow'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 's3d10.arrow'\n",
       "2. 's3d3.arrow'\n",
       "3. 's3d7.arrow'\n",
       "4. 's4d1.arrow'\n",
       "5. 's1d1.arrow'\n",
       "6. 's4d8.arrow'\n",
       "7. 's2d4.arrow'\n",
       "8. 's2d1.arrow'\n",
       "9. 's2d5.arrow'\n",
       "10. 's1d2.arrow'\n",
       "11. 's1d3.arrow'\n",
       "12. 's4d9.arrow'\n",
       "13. 's3d6.arrow'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"s3d10.arrow\" \"s3d3.arrow\"  \"s3d7.arrow\"  \"s4d1.arrow\"  \"s1d1.arrow\" \n",
       " [6] \"s4d8.arrow\"  \"s2d4.arrow\"  \"s2d1.arrow\"  \"s2d5.arrow\"  \"s1d2.arrow\" \n",
       "[11] \"s1d3.arrow\"  \"s4d9.arrow\"  \"s3d6.arrow\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ArrowFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5ba28283-d488-4982-8d4c-745fc3f59365",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"ArchR\",sep=\"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "aa217507-8323-43bf-b23c-0ec25ee33c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Validating Arrows...\n",
      "\n",
      "Getting SampleNames...\n",
      "\n",
      "\n",
      "\n",
      "Copying ArrowFiles to Ouptut Directory! If you want to save disk space set copyArrows = FALSE\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "9 \n",
      "10 \n",
      "11 \n",
      "12 \n",
      "13 \n",
      "\n",
      "\n",
      "Getting Cell Metadata...\n",
      "\n",
      "\n",
      "\n",
      "Merging Cell Metadata...\n",
      "\n",
      "Initializing ArchRProject...\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR <- ArchRProject(\n",
    "  ArrowFiles = ArrowFiles, \n",
    "  outputDirectory = paste(dir.output,\"ArchR\",sep=\"/\"),\n",
    "  copyArrows = TRUE \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f9795403-545f-4385-aa38-2c70b1b2fb21",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: ArchRProject \n",
       "outputDirectory: /maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/ArchR_250226/BMMC5.merge.250226/ArchR \n",
       "samples(13): s3d10 s3d3 ... s4d9 s3d6\n",
       "sampleColData names(1): ArrowFiles\n",
       "cellColData names(13): Sample TSSEnrichment ... nDiFrags BlacklistRatio\n",
       "numberOfCells(1): 55479\n",
       "medianTSS(1): 16.031\n",
       "medianFrags(1): 9711"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obj.ArchR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ee876dd5-0b22-46fd-9015-7e22144c5024",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking Inputs...\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addIterativeLSI-cb5e94f40bcc7-Date-2025-02-26_Time-15-34-43.648311.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 15:34:45.28732 : Computing Total Across All Features, 0.01 mins elapsed.\n",
      "\n",
      "2025-02-26 15:34:52.796609 : Computing Top Features, 0.135 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-26 15:34:55.10371 : Running LSI (1 of 2) on Top Features, 0.174 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-02-26 15:34:55.158746 : Sampling Cells (N = 10007) for Estimated LSI, 0.175 mins elapsed.\n",
      "\n",
      "2025-02-26 15:34:55.169883 : Creating Sampled Partial Matrix, 0.175 mins elapsed.\n",
      "\n",
      "2025-02-26 15:35:05.112386 : Computing Estimated LSI (projectAll = FALSE), 0.341 mins elapsed.\n",
      "\n",
      "2025-02-26 15:36:09.731648 : Identifying Clusters, 1.418 mins elapsed.\n",
      "\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "2025-02-26 15:36:21.320922 : Identified 6 Clusters, 1.611 mins elapsed.\n",
      "\n",
      "2025-02-26 15:36:21.346459 : Saving LSI Iteration, 1.611 mins elapsed.\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************************************************************\n",
      "2025-02-26 15:36:31.40406 : ERROR Found in .saveIteration for  \n",
      "LogFile = ArchRLogs/ArchR-addIterativeLSI-cb5e94f40bcc7-Date-2025-02-26_Time-15-34-43.648311.log\n",
      "\n",
      "<simpleError in g$grobs[[legend]]: no such index at level 2\n",
      ">\n",
      "\n",
      "************************************************************\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-26 15:36:31.40662 : Creating Cluster Matrix on the total Group Features, 1.779 mins elapsed.\n",
      "\n",
      "2025-02-26 15:36:46.424224 : Computing Variable Features, 2.029 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-26 15:36:46.573563 : Running LSI (2 of 2) on Variable Features, 2.032 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-02-26 15:36:46.601951 : Creating Partial Matrix, 2.032 mins elapsed.\n",
      "\n",
      "2025-02-26 15:37:07.781082 : Computing LSI, 2.385 mins elapsed.\n",
      "\n",
      "2025-02-26 15:38:24.472785 : Finished Running IterativeLSI, 3.663 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addIterativeLSI(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    useMatrix = \"TileMatrix\", \n",
    "    name = \"IterativeLSI\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2c1d334b-7fd7-425a-bb8d-7be4cbedd3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.ArchR = addPeakSet(obj.ArchR,\n",
    "                       peakSet = reduce(pairs.E2G))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e25c14db-eb4e-4b94-8862-eb34246a365c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeakMatrix-cb5e941737fff-Date-2025-02-26_Time-15-38-27.586949.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 15:38:27.967899 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeakMatrix-cb5e941737fff-Date-2025-02-26_Time-15-38-27.586949.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addPeakMatrix(obj.ArchR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8887890f-a614-412f-83fb-9a44ecbd0e74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "paths.h5 = c(\"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site1_donor1/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site1_donor2/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site1_donor3/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site2_donor1/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site2_donor4/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site2_donor5/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site3_donor10/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site3_donor3/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site3_donor6/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site3_donor7/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site4_donor1/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site4_donor8/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site4_donor9/outs/raw_feature_bc_matrix.h5\")\n",
    "names.sample = c(\"s1d1\",\n",
    "                 \"s1d2\",\n",
    "                 \"s1d3\",\n",
    "                 \"s2d1\",\n",
    "                 \"s2d4\",\n",
    "                 \"s2d5\",\n",
    "                 \"s3d10\",\n",
    "                 \"s3d3\",\n",
    "                 \"s3d6\",\n",
    "                 \"s3d7\",\n",
    "                 \"s4d1\",\n",
    "                 \"s4d8\",\n",
    "                 \"s4d9\")\n",
    "list.seRNA = lapply(1:length(names.sample),function(x){\n",
    "    import10xFeatureMatrix(paths.h5[x],\n",
    "                           names.sample[x])\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "93d78aef-1337-4876-9bc8-38bfd7d6e112",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "61218"
      ],
      "text/latex": [
       "61218"
      ],
      "text/markdown": [
       "61218"
      ],
      "text/plain": [
       "[1] 61218"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "61165"
      ],
      "text/latex": [
       "61165"
      ],
      "text/markdown": [
       "61165"
      ],
      "text/plain": [
       "[1] 61165"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "id.intersect = rowData(list.seRNA[[1]])[,\"id\"]\n",
    "length(id.intersect)\n",
    "for(n.tmp in 2:13){\n",
    "    id.intersect = intersect(id.intersect,rowData(list.seRNA[[n.tmp]])[,\"id\"])\n",
    "}\n",
    "length(id.intersect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "504f6e06-183d-4ecf-a745-32719a2972d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 61165 9222766 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61165): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(9222766): s1d1#AAACAGCCAAACAACA-1 s1d1#AAACAGCCAAACATAG-1 ...\n",
       "  s4d9#TTTGTTGGTTTGTGGA-1 s4d9#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA = list.seRNA[[1]][rowData(list.seRNA[[1]])[,\"id\"] %in% id.intersect]\n",
    "for(n.tmp in 2:13){\n",
    "    seRNA = cbind(seRNA,list.seRNA[[n.tmp]][rowData(list.seRNA[[n.tmp]])[,\"id\"] %in% id.intersect])\n",
    "}\n",
    "seRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "895234f3-e668-4e83-adf0-562024a6711b",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA = seRNA[,rownames(obj.ArchR@cellColData)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4656d98c-a9cc-486a-afaa-fe59ca6ef055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 61165 55479 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61165): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(55479): s3d10#GTAGCTGTCAGCAAAG-1 s3d10#AACTAGTGTTACCTGT-1 ...\n",
       "  s3d6#TCACCGGCAGCCTTGG-1 s3d6#GAACCTGTCGTTACAA-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a879efc-14cf-4e16-bed3-15cb2e82a2dc",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "fa7f01b3-ff5a-459f-bddd-ad388d554826",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0a713a77-2752-4f1e-bbfa-6e262e8e6f19",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biostrings’:\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seRNA.rename = map_gene_names(seRNA,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "65cb2097-9075-4ee4-9e54-dd288dc50736",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addGeneExpressionMatrix-cb5e97e6d45fb-Date-2025-02-26_Time-15-44-21.646315.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Overlap w/ scATAC = 1\n",
      "\n",
      "2025-02-26 15:44:22.66762 : \n",
      "\n",
      "Overlap Per Sample w/ scATAC : s1d1=5255,s1d2=5205,s1d3=3342,s2d1=3446,s2d4=4835,s2d5=3660,s3d10=5404,s3d3=3500,s3d6=1421,s3d7=1319,s4d1=6035,s4d8=8730,s4d9=3327\n",
      "\n",
      "2025-02-26 15:44:22.733185 : \n",
      "\n",
      "2025-02-26 15:44:25.516919 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addGeneExpressionMatrix-cb5e97e6d45fb-Date-2025-02-26_Time-15-44-21.646315.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addGeneExpressionMatrix(obj.ArchR,\n",
    "                                    seRNA = seRNA.rename,\n",
    "                                    strictMatch = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4318703-a143-40c0-b159-274d05033bbc",
   "metadata": {},
   "source": [
    "Run ArchR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "47c8fd5f-19ec-46e6-af87-57502c1e1ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeak2GeneLinks-cb5e96a960a1f-Date-2025-02-26_Time-15-47-02.520915.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-26 15:47:03.025618 : Getting Available Matrices, 0.008 mins elapsed.\n",
      "\n",
      "2025-02-26 15:47:05.567197 : Filtered Low Prediction Score Cells (0 of 55479, 0), 0.019 mins elapsed.\n",
      "\n",
      "2025-02-26 15:47:06.75913 : Computing KNN, 0.039 mins elapsed.\n",
      "\n",
      "2025-02-26 15:47:07.006213 : Identifying Non-Overlapping KNN pairs, 0.043 mins elapsed.\n",
      "\n",
      "2025-02-26 15:47:08.606753 : Identified 500 Groupings!, 0.069 mins elapsed.\n",
      "\n",
      "2025-02-26 15:47:08.652627 : Getting Group RNA Matrix, 0.07 mins elapsed.\n",
      "\n",
      "2025-02-26 15:47:39.774134 : Getting Group ATAC Matrix, 0.589 mins elapsed.\n",
      "\n",
      "2025-02-26 15:48:30.138413 : Normalizing Group Matrices, 1.428 mins elapsed.\n",
      "\n",
      "2025-02-26 15:48:38.513787 : Finding Peak Gene Pairings, 1.568 mins elapsed.\n",
      "\n",
      "2025-02-26 15:48:39.951799 : Computing Correlations, 1.592 mins elapsed.\n",
      "\n",
      "2025-02-26 15:49:03.984368 : Completed Peak2Gene Correlations!, 1.992 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeak2GeneLinks-cb5e96a960a1f-Date-2025-02-26_Time-15-47-02.520915.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "obj.ArchR = addPeak2GeneLinks(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    reducedDims = \"IterativeLSI\",\n",
    "    useMatrix = \"GeneExpressionMatrix\",\n",
    "    maxDist = maxDist,\n",
    "    threads = n.cores\n",
    ")\n",
    "end_time <- Sys.time()\n",
    "execution_time <- end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "128d0065-5f22-495b-b00d-743fc9843337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 2.024864 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "execution_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "36bd2df2-6b32-4b81-bd4f-39ef8913b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g = metadata(obj.ArchR@peakSet)$Peak2GeneLinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e7730413-8fd4-4952-bb13-0df7a3b15b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata(p2g)[[1]]$PeakName = paste(seqnames(metadata(p2g)[[1]]),\n",
    "                                    ranges(metadata(p2g)[[1]]),sep=\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "aeea3810-6735-4aff-9bd7-1697bef979d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g$PairName = paste(metadata(p2g)[[1]][p2g$idxATAC]$PeakName,\n",
    "                     metadata(p2g)[[2]][p2g$idxRNA]$name,\n",
    "                     sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "87f975ec-6ba4-4696-ad6c-2af65da7aa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 7057138 rows and 7 columns\n",
       "                                idxATAC    idxRNA Correlation        FDR\n",
       "                              <integer> <integer>   <numeric>  <numeric>\n",
       "chr1-10010-10512_FAM138A              1         1         NaN        NaN\n",
       "chr1-115495-115916_FAM138A            2         1         NaN        NaN\n",
       "chr1-180477-181904_FAM138A            3         1         NaN        NaN\n",
       "chr1-191243-191960_FAM138A            4         1         NaN        NaN\n",
       "chr1-267886-268118_FAM138A            5         1         NaN        NaN\n",
       "...                                 ...       ...         ...        ...\n",
       "chrX-156008487-156008683_IL9R    303478     20486  -0.0286262 0.67903951\n",
       "chrX-156008873-156009316_IL9R    303479     20486   0.0977911 0.07583251\n",
       "chrX-156019835-156020122_IL9R    303480     20486  -0.0272367 0.69545735\n",
       "chrX-156029840-156030373_IL9R    303481     20486   0.1436141 0.00516523\n",
       "chrX-156030498-156030725_IL9R    303482     20486   0.0305847 0.65587454\n",
       "                               VarQATAC   VarQRNA               PairName\n",
       "                              <numeric> <numeric>            <character>\n",
       "chr1-10010-10512_FAM138A       0.638614 0.0279703 chr1-10010-10512_FAM..\n",
       "chr1-115495-115916_FAM138A     0.629873 0.0279703 chr1-115495-115916_F..\n",
       "chr1-180477-181904_FAM138A     0.808045 0.0279703 chr1-180477-181904_F..\n",
       "chr1-191243-191960_FAM138A     0.627523 0.0279703 chr1-191243-191960_F..\n",
       "chr1-267886-268118_FAM138A     0.580249 0.0279703 chr1-267886-268118_F..\n",
       "...                                 ...       ...                    ...\n",
       "chrX-156008487-156008683_IL9R 0.0490144  0.240701 chrX-156008487-15600..\n",
       "chrX-156008873-156009316_IL9R 0.2521863  0.240701 chrX-156008873-15600..\n",
       "chrX-156019835-156020122_IL9R 0.1540520  0.240701 chrX-156019835-15602..\n",
       "chrX-156029840-156030373_IL9R 0.5811119  0.240701 chrX-156029840-15603..\n",
       "chrX-156030498-156030725_IL9R 0.6106458  0.240701 chrX-156030498-15603.."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(p2g) = p2g$PairName\n",
    "p2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c7209a32-3904-4734-93af-9c22049f652a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = pairs.E2G[pairs.E2G$PairName %in% p2g$PairName]\n",
    "mcols(pairs.E2G.res)[,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")] = p2g[pairs.E2G.res$PairName,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1ab4ed70-ce9e-4857-bd1b-63a81d27a401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 5682007 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100015450-100016026      * |         AGL\n",
       "        [2]     chr1 100015450-100016026      * |      CDC14A\n",
       "        [3]     chr1 100015450-100016026      * |         DBT\n",
       "        [4]     chr1 100015450-100016026      * |       EXTL2\n",
       "        [5]     chr1 100015450-100016026      * |     MFSD14A\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [5682003]     chrX     9995615-9996434      * |        WWC3\n",
       "  [5682004]     chrX     9997010-9998101      * |       CLCN4\n",
       "  [5682005]     chrX     9997010-9998101      * |     SHROOM2\n",
       "  [5682006]     chrX     9997010-9998101      * |       TBL1X\n",
       "  [5682007]     chrX     9997010-9998101      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-100015450-10001.. chr1-100015450-10001..  -0.0607661\n",
       "        [2] chr1-100015450-10001.. chr1-100015450-10001..  -0.0105788\n",
       "        [3] chr1-100015450-10001.. chr1-100015450-10001..   0.0253900\n",
       "        [4] chr1-100015450-10001.. chr1-100015450-10001..   0.0590558\n",
       "        [5] chr1-100015450-10001.. chr1-100015450-10001..   0.0230463\n",
       "        ...                    ...                    ...         ...\n",
       "  [5682003]   chrX-9995615-9996434 chrX-9995615-9996434.. -0.00150565\n",
       "  [5682004]   chrX-9997010-9998101 chrX-9997010-9998101.. -0.15726283\n",
       "  [5682005]   chrX-9997010-9998101 chrX-9997010-9998101.. -0.04683889\n",
       "  [5682006]   chrX-9997010-9998101 chrX-9997010-9998101..  0.22534651\n",
       "  [5682007]   chrX-9997010-9998101 chrX-9997010-9998101.. -0.00424860\n",
       "                    FDR  VarQATAC   VarQRNA\n",
       "              <numeric> <numeric> <numeric>\n",
       "        [1]    0.318213  0.173809  0.799912\n",
       "        [2]    0.887344  0.173809  0.970663\n",
       "        [3]    0.717145  0.173809  0.620912\n",
       "        [4]    0.334935  0.173809  0.485551\n",
       "        [5]    0.744541  0.173809  0.383774\n",
       "        ...         ...       ...       ...\n",
       "  [5682003] 9.84893e-01  0.797649  0.770233\n",
       "  [5682004] 1.89154e-03  0.649254  0.848091\n",
       "  [5682005] 4.65163e-01  0.649254  0.296056\n",
       "  [5682006] 2.85341e-06  0.649254  0.968710\n",
       "  [5682007] 9.56459e-01  0.649254  0.770233\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46485351-50d7-461a-bf09-d795d5bdab84",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6d424e21-6a41-4422-b76c-6f1c0bda2222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 5682007 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100015450-100016026      * |         AGL\n",
       "        [2]     chr1 100015450-100016026      * |      CDC14A\n",
       "        [3]     chr1 100015450-100016026      * |         DBT\n",
       "        [4]     chr1 100015450-100016026      * |       EXTL2\n",
       "        [5]     chr1 100015450-100016026      * |     MFSD14A\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [5682003]     chrX     9995615-9996434      * |        WWC3\n",
       "  [5682004]     chrX     9997010-9998101      * |       CLCN4\n",
       "  [5682005]     chrX     9997010-9998101      * |     SHROOM2\n",
       "  [5682006]     chrX     9997010-9998101      * |       TBL1X\n",
       "  [5682007]     chrX     9997010-9998101      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-100015450-10001.. chr1-100015450-10001..  -0.0607661\n",
       "        [2] chr1-100015450-10001.. chr1-100015450-10001..  -0.0105788\n",
       "        [3] chr1-100015450-10001.. chr1-100015450-10001..   0.0253900\n",
       "        [4] chr1-100015450-10001.. chr1-100015450-10001..   0.0590558\n",
       "        [5] chr1-100015450-10001.. chr1-100015450-10001..   0.0230463\n",
       "        ...                    ...                    ...         ...\n",
       "  [5682003]   chrX-9995615-9996434 chrX-9995615-9996434.. -0.00150565\n",
       "  [5682004]   chrX-9997010-9998101 chrX-9997010-9998101.. -0.15726283\n",
       "  [5682005]   chrX-9997010-9998101 chrX-9997010-9998101.. -0.04683889\n",
       "  [5682006]   chrX-9997010-9998101 chrX-9997010-9998101..  0.22534651\n",
       "  [5682007]   chrX-9997010-9998101 chrX-9997010-9998101.. -0.00424860\n",
       "                    FDR  VarQATAC   VarQRNA\n",
       "              <numeric> <numeric> <numeric>\n",
       "        [1]    0.318213  0.173809  0.799912\n",
       "        [2]    0.887344  0.173809  0.970663\n",
       "        [3]    0.717145  0.173809  0.620912\n",
       "        [4]    0.334935  0.173809  0.485551\n",
       "        [5]    0.744541  0.173809  0.383774\n",
       "        ...         ...       ...       ...\n",
       "  [5682003] 9.84893e-01  0.797649  0.770233\n",
       "  [5682004] 1.89154e-03  0.649254  0.848091\n",
       "  [5682005] 4.65163e-01  0.649254  0.296056\n",
       "  [5682006] 2.85341e-06  0.649254  0.968710\n",
       "  [5682007] 9.56459e-01  0.649254  0.770233\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5682007 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>Correlation</th><th scope=col>FDR</th><th scope=col>VarQATAC</th><th scope=col>VarQRNA</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>AGL     </td><td>K562</td><td>-0.06076611</td><td>3.182129e-01</td><td>0.1738093</td><td>0.7999121</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>CDC14A  </td><td>K562</td><td>-0.01057877</td><td>8.873440e-01</td><td>0.1738093</td><td>0.9706629</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>DBT     </td><td>K562</td><td> 0.02538999</td><td>7.171450e-01</td><td>0.1738093</td><td>0.6209118</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>EXTL2   </td><td>K562</td><td> 0.05905580</td><td>3.349352e-01</td><td>0.1738093</td><td>0.4855511</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>MFSD14A </td><td>K562</td><td> 0.02304631</td><td>7.445408e-01</td><td>0.1738093</td><td>0.3837743</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>RTCA    </td><td>K562</td><td> 0.02181795</td><td>7.588357e-01</td><td>0.1738093</td><td>0.6939373</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>RTCA-AS1</td><td>K562</td><td>-0.04240575</td><td>5.160080e-01</td><td>0.1738093</td><td>0.4477692</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>SASS6   </td><td>K562</td><td> 0.02975432</td><td>6.656602e-01</td><td>0.1738093</td><td>0.8313971</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>SLC30A7 </td><td>K562</td><td> 0.04653812</td><td>4.685653e-01</td><td>0.1738093</td><td>0.8036708</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>SLC35A3 </td><td>K562</td><td> 0.06122999</td><td>3.137617e-01</td><td>0.1738093</td><td>0.6069999</td></tr>\n",
       "\t<tr><td>chr1</td><td>100015450</td><td>100016026</td><td>TRMT13  </td><td>K562</td><td>-0.01626609</td><td>8.228318e-01</td><td>0.1738093</td><td>0.7434834</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>AGL     </td><td>K562</td><td> 0.31184231</td><td>1.527218e-11</td><td>0.9533943</td><td>0.7999121</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>CDC14A  </td><td>K562</td><td> 0.18052732</td><td>2.724585e-04</td><td>0.9533943</td><td>0.9706629</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>DBT     </td><td>K562</td><td>-0.05083314</td><td>4.206896e-01</td><td>0.9533943</td><td>0.6209118</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>DPH5    </td><td>K562</td><td> 0.13719738</td><td>8.014335e-03</td><td>0.9533943</td><td>0.6874939</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>EXTL2   </td><td>K562</td><td> 0.14929456</td><td>3.439412e-03</td><td>0.9533943</td><td>0.4855511</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>FRRS1   </td><td>K562</td><td>-0.29352990</td><td>2.948981e-10</td><td>0.9533943</td><td>0.7172215</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>MFSD14A </td><td>K562</td><td>-0.07116979</td><td>2.266227e-01</td><td>0.9533943</td><td>0.3837743</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>RTCA    </td><td>K562</td><td> 0.29246523</td><td>3.479333e-10</td><td>0.9533943</td><td>0.6939373</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>RTCA-AS1</td><td>K562</td><td>-0.08207998</td><td>1.506122e-01</td><td>0.9533943</td><td>0.4477692</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>SASS6   </td><td>K562</td><td>-0.07922408</td><td>1.685085e-01</td><td>0.9533943</td><td>0.8313971</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>SLC30A7 </td><td>K562</td><td> 0.12574940</td><td>1.666043e-02</td><td>0.9533943</td><td>0.8036708</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>SLC35A3 </td><td>K562</td><td> 0.13399883</td><td>9.898855e-03</td><td>0.9533943</td><td>0.6069999</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>TRMT13  </td><td>K562</td><td> 0.21231999</td><td>1.203807e-05</td><td>0.9533943</td><td>0.7434834</td></tr>\n",
       "\t<tr><td>chr1</td><td>100028395</td><td>100029373</td><td>VCAM1   </td><td>K562</td><td> 0.16235249</td><td>1.268392e-03</td><td>0.9533943</td><td>0.4038368</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>AGL     </td><td>K562</td><td> 0.13433221</td><td>9.685408e-03</td><td>0.9728287</td><td>0.7999121</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>CDC14A  </td><td>K562</td><td> 0.29301833</td><td>3.192920e-10</td><td>0.9728287</td><td>0.9706629</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>DBT     </td><td>K562</td><td>-0.01598537</td><td>8.260568e-01</td><td>0.9728287</td><td>0.6209118</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>DPH5    </td><td>K562</td><td> 0.11088114</td><td>3.911413e-02</td><td>0.9728287</td><td>0.6874939</td></tr>\n",
       "\t<tr><td>chr1</td><td>100037560</td><td>100039288</td><td>EXTL2   </td><td>K562</td><td>-0.16232868</td><td>1.270819e-03</td><td>0.9728287</td><td>0.4855511</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td>9986538</td><td>9987289</td><td>CLCN4    </td><td>K562</td><td> 0.037187800</td><td>5.774532e-01</td><td>0.78926922</td><td>0.8480914</td></tr>\n",
       "\t<tr><td>chrX</td><td>9986538</td><td>9987289</td><td>SHROOM2  </td><td>K562</td><td> 0.142248219</td><td>5.681758e-03</td><td>0.78926922</td><td>0.2960558</td></tr>\n",
       "\t<tr><td>chrX</td><td>9986538</td><td>9987289</td><td>TBL1X    </td><td>K562</td><td> 0.575399333</td><td>2.266292e-43</td><td>0.78926922</td><td>0.9687103</td></tr>\n",
       "\t<tr><td>chrX</td><td>9986538</td><td>9987289</td><td>WWC3     </td><td>K562</td><td> 0.032353533</td><td>6.348530e-01</td><td>0.78926922</td><td>0.7702333</td></tr>\n",
       "\t<tr><td>chrX</td><td>9987345</td><td>9987562</td><td>CLCN4    </td><td>K562</td><td>-0.021014990</td><td>7.681128e-01</td><td>0.06792166</td><td>0.8480914</td></tr>\n",
       "\t<tr><td>chrX</td><td>9987345</td><td>9987562</td><td>TBL1X    </td><td>K562</td><td> 0.197579839</td><td>5.478092e-05</td><td>0.06792166</td><td>0.9687103</td></tr>\n",
       "\t<tr><td>chrX</td><td>9987345</td><td>9987562</td><td>WWC3     </td><td>K562</td><td>-0.047640308</td><td>4.561077e-01</td><td>0.06792166</td><td>0.7702333</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988628</td><td>9989100</td><td>CLCN4    </td><td>K562</td><td> 0.095054755</td><td>8.618485e-02</td><td>0.37045031</td><td>0.8480914</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988628</td><td>9989100</td><td>SHROOM2  </td><td>K562</td><td> 0.014561063</td><td>8.422837e-01</td><td>0.37045031</td><td>0.2960558</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988628</td><td>9989100</td><td>TBL1X    </td><td>K562</td><td> 0.006381679</td><td>9.336685e-01</td><td>0.37045031</td><td>0.9687103</td></tr>\n",
       "\t<tr><td>chrX</td><td>9988628</td><td>9989100</td><td>WWC3     </td><td>K562</td><td> 0.026255144</td><td>7.069812e-01</td><td>0.37045031</td><td>0.7702333</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>AKAP17A  </td><td>K562</td><td>-0.119012803</td><td>2.485925e-02</td><td>0.49845131</td><td>0.6820756</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>ASMT     </td><td>K562</td><td> 0.164070473</td><td>1.105059e-03</td><td>0.49845131</td><td>0.2794591</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>ASMTL    </td><td>K562</td><td> 0.103219212</td><td>5.819556e-02</td><td>0.49845131</td><td>0.6633311</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>CSF2RA   </td><td>K562</td><td>-0.107129755</td><td>4.768766e-02</td><td>0.49845131</td><td>0.9671483</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>GTPBP6   </td><td>K562</td><td>-0.087276876</td><td>1.215645e-01</td><td>0.49845131</td><td>0.6100264</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>IL3RA    </td><td>K562</td><td> 0.198514710</td><td>4.994479e-05</td><td>0.49845131</td><td>0.8448697</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>LINC00106</td><td>K562</td><td> 0.002582176</td><td>9.738480e-01</td><td>0.49845131</td><td>0.5631163</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>P2RY8    </td><td>K562</td><td> 0.033599599</td><td>6.200585e-01</td><td>0.49845131</td><td>0.9627551</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>PLCXD1   </td><td>K562</td><td>-0.089603732</td><td>1.100183e-01</td><td>0.49845131</td><td>0.6014351</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>PPP2R3B  </td><td>K562</td><td> 0.400044699</td><td>3.673303e-19</td><td>0.49845131</td><td>0.5802499</td></tr>\n",
       "\t<tr><td>chrX</td><td> 999358</td><td> 999880</td><td>SLC25A6  </td><td>K562</td><td>-0.067419655</td><td>2.575750e-01</td><td>0.49845131</td><td>0.8953920</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995615</td><td>9996434</td><td>CLCN4    </td><td>K562</td><td>-0.159547585</td><td>1.583567e-03</td><td>0.79764862</td><td>0.8480914</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995615</td><td>9996434</td><td>SHROOM2  </td><td>K562</td><td>-0.015861376</td><td>8.274849e-01</td><td>0.79764862</td><td>0.2960558</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995615</td><td>9996434</td><td>TBL1X    </td><td>K562</td><td>-0.246847482</td><td>2.135975e-07</td><td>0.79764862</td><td>0.9687103</td></tr>\n",
       "\t<tr><td>chrX</td><td>9995615</td><td>9996434</td><td>WWC3     </td><td>K562</td><td>-0.001505648</td><td>9.848932e-01</td><td>0.79764862</td><td>0.7702333</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997010</td><td>9998101</td><td>CLCN4    </td><td>K562</td><td>-0.157262832</td><td>1.891540e-03</td><td>0.64925432</td><td>0.8480914</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997010</td><td>9998101</td><td>SHROOM2  </td><td>K562</td><td>-0.046838890</td><td>4.651629e-01</td><td>0.64925432</td><td>0.2960558</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997010</td><td>9998101</td><td>TBL1X    </td><td>K562</td><td> 0.225346512</td><td>2.853413e-06</td><td>0.64925432</td><td>0.9687103</td></tr>\n",
       "\t<tr><td>chrX</td><td>9997010</td><td>9998101</td><td>WWC3     </td><td>K562</td><td>-0.004248595</td><td>9.564588e-01</td><td>0.64925432</td><td>0.7702333</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5682007 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & Correlation & FDR & VarQATAC & VarQRNA\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100015450 & 100016026 & AGL      & K562 & -0.06076611 & 3.182129e-01 & 0.1738093 & 0.7999121\\\\\n",
       "\t chr1 & 100015450 & 100016026 & CDC14A   & K562 & -0.01057877 & 8.873440e-01 & 0.1738093 & 0.9706629\\\\\n",
       "\t chr1 & 100015450 & 100016026 & DBT      & K562 &  0.02538999 & 7.171450e-01 & 0.1738093 & 0.6209118\\\\\n",
       "\t chr1 & 100015450 & 100016026 & EXTL2    & K562 &  0.05905580 & 3.349352e-01 & 0.1738093 & 0.4855511\\\\\n",
       "\t chr1 & 100015450 & 100016026 & MFSD14A  & K562 &  0.02304631 & 7.445408e-01 & 0.1738093 & 0.3837743\\\\\n",
       "\t chr1 & 100015450 & 100016026 & RTCA     & K562 &  0.02181795 & 7.588357e-01 & 0.1738093 & 0.6939373\\\\\n",
       "\t chr1 & 100015450 & 100016026 & RTCA-AS1 & K562 & -0.04240575 & 5.160080e-01 & 0.1738093 & 0.4477692\\\\\n",
       "\t chr1 & 100015450 & 100016026 & SASS6    & K562 &  0.02975432 & 6.656602e-01 & 0.1738093 & 0.8313971\\\\\n",
       "\t chr1 & 100015450 & 100016026 & SLC30A7  & K562 &  0.04653812 & 4.685653e-01 & 0.1738093 & 0.8036708\\\\\n",
       "\t chr1 & 100015450 & 100016026 & SLC35A3  & K562 &  0.06122999 & 3.137617e-01 & 0.1738093 & 0.6069999\\\\\n",
       "\t chr1 & 100015450 & 100016026 & TRMT13   & K562 & -0.01626609 & 8.228318e-01 & 0.1738093 & 0.7434834\\\\\n",
       "\t chr1 & 100028395 & 100029373 & AGL      & K562 &  0.31184231 & 1.527218e-11 & 0.9533943 & 0.7999121\\\\\n",
       "\t chr1 & 100028395 & 100029373 & CDC14A   & K562 &  0.18052732 & 2.724585e-04 & 0.9533943 & 0.9706629\\\\\n",
       "\t chr1 & 100028395 & 100029373 & DBT      & K562 & -0.05083314 & 4.206896e-01 & 0.9533943 & 0.6209118\\\\\n",
       "\t chr1 & 100028395 & 100029373 & DPH5     & K562 &  0.13719738 & 8.014335e-03 & 0.9533943 & 0.6874939\\\\\n",
       "\t chr1 & 100028395 & 100029373 & EXTL2    & K562 &  0.14929456 & 3.439412e-03 & 0.9533943 & 0.4855511\\\\\n",
       "\t chr1 & 100028395 & 100029373 & FRRS1    & K562 & -0.29352990 & 2.948981e-10 & 0.9533943 & 0.7172215\\\\\n",
       "\t chr1 & 100028395 & 100029373 & MFSD14A  & K562 & -0.07116979 & 2.266227e-01 & 0.9533943 & 0.3837743\\\\\n",
       "\t chr1 & 100028395 & 100029373 & RTCA     & K562 &  0.29246523 & 3.479333e-10 & 0.9533943 & 0.6939373\\\\\n",
       "\t chr1 & 100028395 & 100029373 & RTCA-AS1 & K562 & -0.08207998 & 1.506122e-01 & 0.9533943 & 0.4477692\\\\\n",
       "\t chr1 & 100028395 & 100029373 & SASS6    & K562 & -0.07922408 & 1.685085e-01 & 0.9533943 & 0.8313971\\\\\n",
       "\t chr1 & 100028395 & 100029373 & SLC30A7  & K562 &  0.12574940 & 1.666043e-02 & 0.9533943 & 0.8036708\\\\\n",
       "\t chr1 & 100028395 & 100029373 & SLC35A3  & K562 &  0.13399883 & 9.898855e-03 & 0.9533943 & 0.6069999\\\\\n",
       "\t chr1 & 100028395 & 100029373 & TRMT13   & K562 &  0.21231999 & 1.203807e-05 & 0.9533943 & 0.7434834\\\\\n",
       "\t chr1 & 100028395 & 100029373 & VCAM1    & K562 &  0.16235249 & 1.268392e-03 & 0.9533943 & 0.4038368\\\\\n",
       "\t chr1 & 100037560 & 100039288 & AGL      & K562 &  0.13433221 & 9.685408e-03 & 0.9728287 & 0.7999121\\\\\n",
       "\t chr1 & 100037560 & 100039288 & CDC14A   & K562 &  0.29301833 & 3.192920e-10 & 0.9728287 & 0.9706629\\\\\n",
       "\t chr1 & 100037560 & 100039288 & DBT      & K562 & -0.01598537 & 8.260568e-01 & 0.9728287 & 0.6209118\\\\\n",
       "\t chr1 & 100037560 & 100039288 & DPH5     & K562 &  0.11088114 & 3.911413e-02 & 0.9728287 & 0.6874939\\\\\n",
       "\t chr1 & 100037560 & 100039288 & EXTL2    & K562 & -0.16232868 & 1.270819e-03 & 0.9728287 & 0.4855511\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX & 9986538 & 9987289 & CLCN4     & K562 &  0.037187800 & 5.774532e-01 & 0.78926922 & 0.8480914\\\\\n",
       "\t chrX & 9986538 & 9987289 & SHROOM2   & K562 &  0.142248219 & 5.681758e-03 & 0.78926922 & 0.2960558\\\\\n",
       "\t chrX & 9986538 & 9987289 & TBL1X     & K562 &  0.575399333 & 2.266292e-43 & 0.78926922 & 0.9687103\\\\\n",
       "\t chrX & 9986538 & 9987289 & WWC3      & K562 &  0.032353533 & 6.348530e-01 & 0.78926922 & 0.7702333\\\\\n",
       "\t chrX & 9987345 & 9987562 & CLCN4     & K562 & -0.021014990 & 7.681128e-01 & 0.06792166 & 0.8480914\\\\\n",
       "\t chrX & 9987345 & 9987562 & TBL1X     & K562 &  0.197579839 & 5.478092e-05 & 0.06792166 & 0.9687103\\\\\n",
       "\t chrX & 9987345 & 9987562 & WWC3      & K562 & -0.047640308 & 4.561077e-01 & 0.06792166 & 0.7702333\\\\\n",
       "\t chrX & 9988628 & 9989100 & CLCN4     & K562 &  0.095054755 & 8.618485e-02 & 0.37045031 & 0.8480914\\\\\n",
       "\t chrX & 9988628 & 9989100 & SHROOM2   & K562 &  0.014561063 & 8.422837e-01 & 0.37045031 & 0.2960558\\\\\n",
       "\t chrX & 9988628 & 9989100 & TBL1X     & K562 &  0.006381679 & 9.336685e-01 & 0.37045031 & 0.9687103\\\\\n",
       "\t chrX & 9988628 & 9989100 & WWC3      & K562 &  0.026255144 & 7.069812e-01 & 0.37045031 & 0.7702333\\\\\n",
       "\t chrX &  999358 &  999880 & AKAP17A   & K562 & -0.119012803 & 2.485925e-02 & 0.49845131 & 0.6820756\\\\\n",
       "\t chrX &  999358 &  999880 & ASMT      & K562 &  0.164070473 & 1.105059e-03 & 0.49845131 & 0.2794591\\\\\n",
       "\t chrX &  999358 &  999880 & ASMTL     & K562 &  0.103219212 & 5.819556e-02 & 0.49845131 & 0.6633311\\\\\n",
       "\t chrX &  999358 &  999880 & CSF2RA    & K562 & -0.107129755 & 4.768766e-02 & 0.49845131 & 0.9671483\\\\\n",
       "\t chrX &  999358 &  999880 & GTPBP6    & K562 & -0.087276876 & 1.215645e-01 & 0.49845131 & 0.6100264\\\\\n",
       "\t chrX &  999358 &  999880 & IL3RA     & K562 &  0.198514710 & 4.994479e-05 & 0.49845131 & 0.8448697\\\\\n",
       "\t chrX &  999358 &  999880 & LINC00106 & K562 &  0.002582176 & 9.738480e-01 & 0.49845131 & 0.5631163\\\\\n",
       "\t chrX &  999358 &  999880 & P2RY8     & K562 &  0.033599599 & 6.200585e-01 & 0.49845131 & 0.9627551\\\\\n",
       "\t chrX &  999358 &  999880 & PLCXD1    & K562 & -0.089603732 & 1.100183e-01 & 0.49845131 & 0.6014351\\\\\n",
       "\t chrX &  999358 &  999880 & PPP2R3B   & K562 &  0.400044699 & 3.673303e-19 & 0.49845131 & 0.5802499\\\\\n",
       "\t chrX &  999358 &  999880 & SLC25A6   & K562 & -0.067419655 & 2.575750e-01 & 0.49845131 & 0.8953920\\\\\n",
       "\t chrX & 9995615 & 9996434 & CLCN4     & K562 & -0.159547585 & 1.583567e-03 & 0.79764862 & 0.8480914\\\\\n",
       "\t chrX & 9995615 & 9996434 & SHROOM2   & K562 & -0.015861376 & 8.274849e-01 & 0.79764862 & 0.2960558\\\\\n",
       "\t chrX & 9995615 & 9996434 & TBL1X     & K562 & -0.246847482 & 2.135975e-07 & 0.79764862 & 0.9687103\\\\\n",
       "\t chrX & 9995615 & 9996434 & WWC3      & K562 & -0.001505648 & 9.848932e-01 & 0.79764862 & 0.7702333\\\\\n",
       "\t chrX & 9997010 & 9998101 & CLCN4     & K562 & -0.157262832 & 1.891540e-03 & 0.64925432 & 0.8480914\\\\\n",
       "\t chrX & 9997010 & 9998101 & SHROOM2   & K562 & -0.046838890 & 4.651629e-01 & 0.64925432 & 0.2960558\\\\\n",
       "\t chrX & 9997010 & 9998101 & TBL1X     & K562 &  0.225346512 & 2.853413e-06 & 0.64925432 & 0.9687103\\\\\n",
       "\t chrX & 9997010 & 9998101 & WWC3      & K562 & -0.004248595 & 9.564588e-01 & 0.64925432 & 0.7702333\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5682007 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | Correlation &lt;dbl&gt; | FDR &lt;dbl&gt; | VarQATAC &lt;dbl&gt; | VarQRNA &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100015450 | 100016026 | AGL      | K562 | -0.06076611 | 3.182129e-01 | 0.1738093 | 0.7999121 |\n",
       "| chr1 | 100015450 | 100016026 | CDC14A   | K562 | -0.01057877 | 8.873440e-01 | 0.1738093 | 0.9706629 |\n",
       "| chr1 | 100015450 | 100016026 | DBT      | K562 |  0.02538999 | 7.171450e-01 | 0.1738093 | 0.6209118 |\n",
       "| chr1 | 100015450 | 100016026 | EXTL2    | K562 |  0.05905580 | 3.349352e-01 | 0.1738093 | 0.4855511 |\n",
       "| chr1 | 100015450 | 100016026 | MFSD14A  | K562 |  0.02304631 | 7.445408e-01 | 0.1738093 | 0.3837743 |\n",
       "| chr1 | 100015450 | 100016026 | RTCA     | K562 |  0.02181795 | 7.588357e-01 | 0.1738093 | 0.6939373 |\n",
       "| chr1 | 100015450 | 100016026 | RTCA-AS1 | K562 | -0.04240575 | 5.160080e-01 | 0.1738093 | 0.4477692 |\n",
       "| chr1 | 100015450 | 100016026 | SASS6    | K562 |  0.02975432 | 6.656602e-01 | 0.1738093 | 0.8313971 |\n",
       "| chr1 | 100015450 | 100016026 | SLC30A7  | K562 |  0.04653812 | 4.685653e-01 | 0.1738093 | 0.8036708 |\n",
       "| chr1 | 100015450 | 100016026 | SLC35A3  | K562 |  0.06122999 | 3.137617e-01 | 0.1738093 | 0.6069999 |\n",
       "| chr1 | 100015450 | 100016026 | TRMT13   | K562 | -0.01626609 | 8.228318e-01 | 0.1738093 | 0.7434834 |\n",
       "| chr1 | 100028395 | 100029373 | AGL      | K562 |  0.31184231 | 1.527218e-11 | 0.9533943 | 0.7999121 |\n",
       "| chr1 | 100028395 | 100029373 | CDC14A   | K562 |  0.18052732 | 2.724585e-04 | 0.9533943 | 0.9706629 |\n",
       "| chr1 | 100028395 | 100029373 | DBT      | K562 | -0.05083314 | 4.206896e-01 | 0.9533943 | 0.6209118 |\n",
       "| chr1 | 100028395 | 100029373 | DPH5     | K562 |  0.13719738 | 8.014335e-03 | 0.9533943 | 0.6874939 |\n",
       "| chr1 | 100028395 | 100029373 | EXTL2    | K562 |  0.14929456 | 3.439412e-03 | 0.9533943 | 0.4855511 |\n",
       "| chr1 | 100028395 | 100029373 | FRRS1    | K562 | -0.29352990 | 2.948981e-10 | 0.9533943 | 0.7172215 |\n",
       "| chr1 | 100028395 | 100029373 | MFSD14A  | K562 | -0.07116979 | 2.266227e-01 | 0.9533943 | 0.3837743 |\n",
       "| chr1 | 100028395 | 100029373 | RTCA     | K562 |  0.29246523 | 3.479333e-10 | 0.9533943 | 0.6939373 |\n",
       "| chr1 | 100028395 | 100029373 | RTCA-AS1 | K562 | -0.08207998 | 1.506122e-01 | 0.9533943 | 0.4477692 |\n",
       "| chr1 | 100028395 | 100029373 | SASS6    | K562 | -0.07922408 | 1.685085e-01 | 0.9533943 | 0.8313971 |\n",
       "| chr1 | 100028395 | 100029373 | SLC30A7  | K562 |  0.12574940 | 1.666043e-02 | 0.9533943 | 0.8036708 |\n",
       "| chr1 | 100028395 | 100029373 | SLC35A3  | K562 |  0.13399883 | 9.898855e-03 | 0.9533943 | 0.6069999 |\n",
       "| chr1 | 100028395 | 100029373 | TRMT13   | K562 |  0.21231999 | 1.203807e-05 | 0.9533943 | 0.7434834 |\n",
       "| chr1 | 100028395 | 100029373 | VCAM1    | K562 |  0.16235249 | 1.268392e-03 | 0.9533943 | 0.4038368 |\n",
       "| chr1 | 100037560 | 100039288 | AGL      | K562 |  0.13433221 | 9.685408e-03 | 0.9728287 | 0.7999121 |\n",
       "| chr1 | 100037560 | 100039288 | CDC14A   | K562 |  0.29301833 | 3.192920e-10 | 0.9728287 | 0.9706629 |\n",
       "| chr1 | 100037560 | 100039288 | DBT      | K562 | -0.01598537 | 8.260568e-01 | 0.9728287 | 0.6209118 |\n",
       "| chr1 | 100037560 | 100039288 | DPH5     | K562 |  0.11088114 | 3.911413e-02 | 0.9728287 | 0.6874939 |\n",
       "| chr1 | 100037560 | 100039288 | EXTL2    | K562 | -0.16232868 | 1.270819e-03 | 0.9728287 | 0.4855511 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX | 9986538 | 9987289 | CLCN4     | K562 |  0.037187800 | 5.774532e-01 | 0.78926922 | 0.8480914 |\n",
       "| chrX | 9986538 | 9987289 | SHROOM2   | K562 |  0.142248219 | 5.681758e-03 | 0.78926922 | 0.2960558 |\n",
       "| chrX | 9986538 | 9987289 | TBL1X     | K562 |  0.575399333 | 2.266292e-43 | 0.78926922 | 0.9687103 |\n",
       "| chrX | 9986538 | 9987289 | WWC3      | K562 |  0.032353533 | 6.348530e-01 | 0.78926922 | 0.7702333 |\n",
       "| chrX | 9987345 | 9987562 | CLCN4     | K562 | -0.021014990 | 7.681128e-01 | 0.06792166 | 0.8480914 |\n",
       "| chrX | 9987345 | 9987562 | TBL1X     | K562 |  0.197579839 | 5.478092e-05 | 0.06792166 | 0.9687103 |\n",
       "| chrX | 9987345 | 9987562 | WWC3      | K562 | -0.047640308 | 4.561077e-01 | 0.06792166 | 0.7702333 |\n",
       "| chrX | 9988628 | 9989100 | CLCN4     | K562 |  0.095054755 | 8.618485e-02 | 0.37045031 | 0.8480914 |\n",
       "| chrX | 9988628 | 9989100 | SHROOM2   | K562 |  0.014561063 | 8.422837e-01 | 0.37045031 | 0.2960558 |\n",
       "| chrX | 9988628 | 9989100 | TBL1X     | K562 |  0.006381679 | 9.336685e-01 | 0.37045031 | 0.9687103 |\n",
       "| chrX | 9988628 | 9989100 | WWC3      | K562 |  0.026255144 | 7.069812e-01 | 0.37045031 | 0.7702333 |\n",
       "| chrX |  999358 |  999880 | AKAP17A   | K562 | -0.119012803 | 2.485925e-02 | 0.49845131 | 0.6820756 |\n",
       "| chrX |  999358 |  999880 | ASMT      | K562 |  0.164070473 | 1.105059e-03 | 0.49845131 | 0.2794591 |\n",
       "| chrX |  999358 |  999880 | ASMTL     | K562 |  0.103219212 | 5.819556e-02 | 0.49845131 | 0.6633311 |\n",
       "| chrX |  999358 |  999880 | CSF2RA    | K562 | -0.107129755 | 4.768766e-02 | 0.49845131 | 0.9671483 |\n",
       "| chrX |  999358 |  999880 | GTPBP6    | K562 | -0.087276876 | 1.215645e-01 | 0.49845131 | 0.6100264 |\n",
       "| chrX |  999358 |  999880 | IL3RA     | K562 |  0.198514710 | 4.994479e-05 | 0.49845131 | 0.8448697 |\n",
       "| chrX |  999358 |  999880 | LINC00106 | K562 |  0.002582176 | 9.738480e-01 | 0.49845131 | 0.5631163 |\n",
       "| chrX |  999358 |  999880 | P2RY8     | K562 |  0.033599599 | 6.200585e-01 | 0.49845131 | 0.9627551 |\n",
       "| chrX |  999358 |  999880 | PLCXD1    | K562 | -0.089603732 | 1.100183e-01 | 0.49845131 | 0.6014351 |\n",
       "| chrX |  999358 |  999880 | PPP2R3B   | K562 |  0.400044699 | 3.673303e-19 | 0.49845131 | 0.5802499 |\n",
       "| chrX |  999358 |  999880 | SLC25A6   | K562 | -0.067419655 | 2.575750e-01 | 0.49845131 | 0.8953920 |\n",
       "| chrX | 9995615 | 9996434 | CLCN4     | K562 | -0.159547585 | 1.583567e-03 | 0.79764862 | 0.8480914 |\n",
       "| chrX | 9995615 | 9996434 | SHROOM2   | K562 | -0.015861376 | 8.274849e-01 | 0.79764862 | 0.2960558 |\n",
       "| chrX | 9995615 | 9996434 | TBL1X     | K562 | -0.246847482 | 2.135975e-07 | 0.79764862 | 0.9687103 |\n",
       "| chrX | 9995615 | 9996434 | WWC3      | K562 | -0.001505648 | 9.848932e-01 | 0.79764862 | 0.7702333 |\n",
       "| chrX | 9997010 | 9998101 | CLCN4     | K562 | -0.157262832 | 1.891540e-03 | 0.64925432 | 0.8480914 |\n",
       "| chrX | 9997010 | 9998101 | SHROOM2   | K562 | -0.046838890 | 4.651629e-01 | 0.64925432 | 0.2960558 |\n",
       "| chrX | 9997010 | 9998101 | TBL1X     | K562 |  0.225346512 | 2.853413e-06 | 0.64925432 | 0.9687103 |\n",
       "| chrX | 9997010 | 9998101 | WWC3      | K562 | -0.004248595 | 9.564588e-01 | 0.64925432 | 0.7702333 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene CellType Correlation  FDR         \n",
       "1       chr1 100015450 100016026 AGL        K562     -0.06076611  3.182129e-01\n",
       "2       chr1 100015450 100016026 CDC14A     K562     -0.01057877  8.873440e-01\n",
       "3       chr1 100015450 100016026 DBT        K562      0.02538999  7.171450e-01\n",
       "4       chr1 100015450 100016026 EXTL2      K562      0.05905580  3.349352e-01\n",
       "5       chr1 100015450 100016026 MFSD14A    K562      0.02304631  7.445408e-01\n",
       "6       chr1 100015450 100016026 RTCA       K562      0.02181795  7.588357e-01\n",
       "7       chr1 100015450 100016026 RTCA-AS1   K562     -0.04240575  5.160080e-01\n",
       "8       chr1 100015450 100016026 SASS6      K562      0.02975432  6.656602e-01\n",
       "9       chr1 100015450 100016026 SLC30A7    K562      0.04653812  4.685653e-01\n",
       "10      chr1 100015450 100016026 SLC35A3    K562      0.06122999  3.137617e-01\n",
       "11      chr1 100015450 100016026 TRMT13     K562     -0.01626609  8.228318e-01\n",
       "12      chr1 100028395 100029373 AGL        K562      0.31184231  1.527218e-11\n",
       "13      chr1 100028395 100029373 CDC14A     K562      0.18052732  2.724585e-04\n",
       "14      chr1 100028395 100029373 DBT        K562     -0.05083314  4.206896e-01\n",
       "15      chr1 100028395 100029373 DPH5       K562      0.13719738  8.014335e-03\n",
       "16      chr1 100028395 100029373 EXTL2      K562      0.14929456  3.439412e-03\n",
       "17      chr1 100028395 100029373 FRRS1      K562     -0.29352990  2.948981e-10\n",
       "18      chr1 100028395 100029373 MFSD14A    K562     -0.07116979  2.266227e-01\n",
       "19      chr1 100028395 100029373 RTCA       K562      0.29246523  3.479333e-10\n",
       "20      chr1 100028395 100029373 RTCA-AS1   K562     -0.08207998  1.506122e-01\n",
       "21      chr1 100028395 100029373 SASS6      K562     -0.07922408  1.685085e-01\n",
       "22      chr1 100028395 100029373 SLC30A7    K562      0.12574940  1.666043e-02\n",
       "23      chr1 100028395 100029373 SLC35A3    K562      0.13399883  9.898855e-03\n",
       "24      chr1 100028395 100029373 TRMT13     K562      0.21231999  1.203807e-05\n",
       "25      chr1 100028395 100029373 VCAM1      K562      0.16235249  1.268392e-03\n",
       "26      chr1 100037560 100039288 AGL        K562      0.13433221  9.685408e-03\n",
       "27      chr1 100037560 100039288 CDC14A     K562      0.29301833  3.192920e-10\n",
       "28      chr1 100037560 100039288 DBT        K562     -0.01598537  8.260568e-01\n",
       "29      chr1 100037560 100039288 DPH5       K562      0.11088114  3.911413e-02\n",
       "30      chr1 100037560 100039288 EXTL2      K562     -0.16232868  1.270819e-03\n",
       "⋮       ⋮    ⋮         ⋮         ⋮          ⋮        ⋮            ⋮           \n",
       "5681978 chrX 9986538   9987289   CLCN4      K562      0.037187800 5.774532e-01\n",
       "5681979 chrX 9986538   9987289   SHROOM2    K562      0.142248219 5.681758e-03\n",
       "5681980 chrX 9986538   9987289   TBL1X      K562      0.575399333 2.266292e-43\n",
       "5681981 chrX 9986538   9987289   WWC3       K562      0.032353533 6.348530e-01\n",
       "5681982 chrX 9987345   9987562   CLCN4      K562     -0.021014990 7.681128e-01\n",
       "5681983 chrX 9987345   9987562   TBL1X      K562      0.197579839 5.478092e-05\n",
       "5681984 chrX 9987345   9987562   WWC3       K562     -0.047640308 4.561077e-01\n",
       "5681985 chrX 9988628   9989100   CLCN4      K562      0.095054755 8.618485e-02\n",
       "5681986 chrX 9988628   9989100   SHROOM2    K562      0.014561063 8.422837e-01\n",
       "5681987 chrX 9988628   9989100   TBL1X      K562      0.006381679 9.336685e-01\n",
       "5681988 chrX 9988628   9989100   WWC3       K562      0.026255144 7.069812e-01\n",
       "5681989 chrX  999358    999880   AKAP17A    K562     -0.119012803 2.485925e-02\n",
       "5681990 chrX  999358    999880   ASMT       K562      0.164070473 1.105059e-03\n",
       "5681991 chrX  999358    999880   ASMTL      K562      0.103219212 5.819556e-02\n",
       "5681992 chrX  999358    999880   CSF2RA     K562     -0.107129755 4.768766e-02\n",
       "5681993 chrX  999358    999880   GTPBP6     K562     -0.087276876 1.215645e-01\n",
       "5681994 chrX  999358    999880   IL3RA      K562      0.198514710 4.994479e-05\n",
       "5681995 chrX  999358    999880   LINC00106  K562      0.002582176 9.738480e-01\n",
       "5681996 chrX  999358    999880   P2RY8      K562      0.033599599 6.200585e-01\n",
       "5681997 chrX  999358    999880   PLCXD1     K562     -0.089603732 1.100183e-01\n",
       "5681998 chrX  999358    999880   PPP2R3B    K562      0.400044699 3.673303e-19\n",
       "5681999 chrX  999358    999880   SLC25A6    K562     -0.067419655 2.575750e-01\n",
       "5682000 chrX 9995615   9996434   CLCN4      K562     -0.159547585 1.583567e-03\n",
       "5682001 chrX 9995615   9996434   SHROOM2    K562     -0.015861376 8.274849e-01\n",
       "5682002 chrX 9995615   9996434   TBL1X      K562     -0.246847482 2.135975e-07\n",
       "5682003 chrX 9995615   9996434   WWC3       K562     -0.001505648 9.848932e-01\n",
       "5682004 chrX 9997010   9998101   CLCN4      K562     -0.157262832 1.891540e-03\n",
       "5682005 chrX 9997010   9998101   SHROOM2    K562     -0.046838890 4.651629e-01\n",
       "5682006 chrX 9997010   9998101   TBL1X      K562      0.225346512 2.853413e-06\n",
       "5682007 chrX 9997010   9998101   WWC3       K562     -0.004248595 9.564588e-01\n",
       "        VarQATAC   VarQRNA  \n",
       "1       0.1738093  0.7999121\n",
       "2       0.1738093  0.9706629\n",
       "3       0.1738093  0.6209118\n",
       "4       0.1738093  0.4855511\n",
       "5       0.1738093  0.3837743\n",
       "6       0.1738093  0.6939373\n",
       "7       0.1738093  0.4477692\n",
       "8       0.1738093  0.8313971\n",
       "9       0.1738093  0.8036708\n",
       "10      0.1738093  0.6069999\n",
       "11      0.1738093  0.7434834\n",
       "12      0.9533943  0.7999121\n",
       "13      0.9533943  0.9706629\n",
       "14      0.9533943  0.6209118\n",
       "15      0.9533943  0.6874939\n",
       "16      0.9533943  0.4855511\n",
       "17      0.9533943  0.7172215\n",
       "18      0.9533943  0.3837743\n",
       "19      0.9533943  0.6939373\n",
       "20      0.9533943  0.4477692\n",
       "21      0.9533943  0.8313971\n",
       "22      0.9533943  0.8036708\n",
       "23      0.9533943  0.6069999\n",
       "24      0.9533943  0.7434834\n",
       "25      0.9533943  0.4038368\n",
       "26      0.9728287  0.7999121\n",
       "27      0.9728287  0.9706629\n",
       "28      0.9728287  0.6209118\n",
       "29      0.9728287  0.6874939\n",
       "30      0.9728287  0.4855511\n",
       "⋮       ⋮          ⋮        \n",
       "5681978 0.78926922 0.8480914\n",
       "5681979 0.78926922 0.2960558\n",
       "5681980 0.78926922 0.9687103\n",
       "5681981 0.78926922 0.7702333\n",
       "5681982 0.06792166 0.8480914\n",
       "5681983 0.06792166 0.9687103\n",
       "5681984 0.06792166 0.7702333\n",
       "5681985 0.37045031 0.8480914\n",
       "5681986 0.37045031 0.2960558\n",
       "5681987 0.37045031 0.9687103\n",
       "5681988 0.37045031 0.7702333\n",
       "5681989 0.49845131 0.6820756\n",
       "5681990 0.49845131 0.2794591\n",
       "5681991 0.49845131 0.6633311\n",
       "5681992 0.49845131 0.9671483\n",
       "5681993 0.49845131 0.6100264\n",
       "5681994 0.49845131 0.8448697\n",
       "5681995 0.49845131 0.5631163\n",
       "5681996 0.49845131 0.9627551\n",
       "5681997 0.49845131 0.6014351\n",
       "5681998 0.49845131 0.5802499\n",
       "5681999 0.49845131 0.8953920\n",
       "5682000 0.79764862 0.8480914\n",
       "5682001 0.79764862 0.2960558\n",
       "5682002 0.79764862 0.9687103\n",
       "5682003 0.79764862 0.7702333\n",
       "5682004 0.64925432 0.8480914\n",
       "5682005 0.64925432 0.2960558\n",
       "5682006 0.64925432 0.9687103\n",
       "5682007 0.64925432 0.7702333"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(obj.ArchR,\n",
    "        paste(dir.output,\"obj.ArchR.rds\",sep = \"/\"))\n",
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"Correlation\",\n",
    "                         \"FDR\",\n",
    "                         \"VarQATAC\",\n",
    "                         \"VarQRNA\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "pairs.E2G.res\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4c909ed5-7321-4e69-8208-1544b371da6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] dplyr_1.1.4                       uwot_0.1.16                      \n",
       " [3] nabor_0.5.0                       BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] genomation_1.34.0                 Signac_1.13.0                    \n",
       "[11] Seurat_5.0.3                      SeuratObject_5.0.1               \n",
       "[13] sp_2.1-4                          rhdf5_2.44.0                     \n",
       "[15] SummarizedExperiment_1.30.2       Biobase_2.60.0                   \n",
       "[17] MatrixGenerics_1.12.3             Rcpp_1.0.12                      \n",
       "[19] Matrix_1.6-5                      GenomicRanges_1.54.1             \n",
       "[21] GenomeInfoDb_1.38.1               IRanges_2.36.0                   \n",
       "[23] S4Vectors_0.40.2                  BiocGenerics_0.48.1              \n",
       "[25] matrixStats_1.3.0                 data.table_1.15.2                \n",
       "[27] stringr_1.5.1                     plyr_1.8.9                       \n",
       "[29] magrittr_2.0.3                    ggplot2_3.5.1                    \n",
       "[31] gtable_0.3.5                      gtools_3.9.5                     \n",
       "[33] gridExtra_2.3                     ArchR_1.0.2                      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22         splines_4.3.3            later_1.3.2             \n",
       "  [4] pbdZMQ_0.3-11            BiocIO_1.10.0            bitops_1.0-7            \n",
       "  [7] tibble_3.2.1             polyclip_1.10-6          XML_3.99-0.16.1         \n",
       " [10] fastDummies_1.7.3        lifecycle_1.0.4          globals_0.16.3          \n",
       " [13] lattice_0.22-6           MASS_7.3-60              plotly_4.10.4           \n",
       " [16] plotrix_3.8-4            yaml_2.3.8               httpuv_1.6.15           \n",
       " [19] sctransform_0.4.1        spam_2.10-0              spatstat.sparse_3.0-3   \n",
       " [22] reticulate_1.36.1        cowplot_1.1.3            pbapply_1.7-2           \n",
       " [25] RColorBrewer_1.1-3       abind_1.4-5              zlibbioc_1.48.0         \n",
       " [28] Rtsne_0.17               purrr_1.0.2              RCurl_1.98-1.14         \n",
       " [31] GenomeInfoDbData_1.2.11  ggrepel_0.9.5            irlba_2.3.5.1           \n",
       " [34] listenv_0.9.1            spatstat.utils_3.0-4     goftest_1.2-3           \n",
       " [37] RSpectra_0.16-1          spatstat.random_3.2-3    fitdistrplus_1.1-11     \n",
       " [40] parallelly_1.37.1        leiden_0.4.3.1           codetools_0.2-20        \n",
       " [43] DelayedArray_0.26.7      RcppRoll_0.3.0           tidyselect_1.2.1        \n",
       " [46] farver_2.1.1             base64enc_0.1-3          spatstat.explore_3.2-6  \n",
       " [49] GenomicAlignments_1.36.0 jsonlite_1.8.8           progressr_0.14.0        \n",
       " [52] ggridges_0.5.6           survival_3.6-4           tools_4.3.3             \n",
       " [55] ica_1.0-3                glue_1.7.0               IRdisplay_1.1           \n",
       " [58] withr_3.0.0              fastmap_1.1.1            rhdf5filters_1.12.1     \n",
       " [61] fansi_1.0.6              digest_0.6.35            R6_2.5.1                \n",
       " [64] mime_0.12                seqPattern_1.34.0        colorspace_2.1-0        \n",
       " [67] Cairo_1.6-2              scattermore_1.2          tensor_1.5              \n",
       " [70] spatstat.data_3.0-4      utf8_1.2.4               tidyr_1.3.1             \n",
       " [73] generics_0.1.3           httr_1.4.7               htmlwidgets_1.6.4       \n",
       " [76] S4Arrays_1.2.1           pkgconfig_2.0.3          impute_1.76.0           \n",
       " [79] lmtest_0.9-40            htmltools_0.5.8.1        dotCall64_1.1-1         \n",
       " [82] scales_1.3.0             png_0.1-8                tzdb_0.4.0              \n",
       " [85] reshape2_1.4.4           rjson_0.2.21             uuid_1.2-0              \n",
       " [88] nlme_3.1-164             repr_1.1.7               zoo_1.8-12              \n",
       " [91] KernSmooth_2.23-22       miniUI_0.1.1.1           restfulr_0.0.15         \n",
       " [94] pillar_1.9.0             vctrs_0.6.5              RANN_2.6.1              \n",
       " [97] promises_1.3.0           xtable_1.8-4             cluster_2.1.6           \n",
       "[100] evaluate_0.23            readr_2.1.5              cli_3.6.2               \n",
       "[103] compiler_4.3.3           Rsamtools_2.18.0         rlang_1.1.3             \n",
       "[106] crayon_1.5.2             future.apply_1.11.2      labeling_0.4.3          \n",
       "[109] stringi_1.8.3            gridBase_0.4-7           viridisLite_0.4.2       \n",
       "[112] deldir_2.0-4             BiocParallel_1.36.0      munsell_0.5.1           \n",
       "[115] lazyeval_0.2.2           spatstat.geom_3.2-9      IRkernel_1.3.2          \n",
       "[118] RcppHNSW_0.6.0           hms_1.1.3                patchwork_1.2.0         \n",
       "[121] future_1.33.2            Rhdf5lib_1.22.1          shiny_1.8.1.1           \n",
       "[124] ROCR_1.0-11              igraph_2.0.3             fastmatch_1.1-4         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b54012ba-20d1-4787-8d0d-d851ef1744c4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
