{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8466b35-cae5-4f62-8138-37b2de484a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.2\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : grid v4.3.3\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "\tLoading Package : gtools v3.9.5\n",
      "\n",
      "\tLoading Package : gtable v0.3.5\n",
      "\n",
      "\tLoading Package : ggplot2 v3.5.1\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "\tLoading Package : plyr v1.8.9\n",
      "\n",
      "\tLoading Package : stringr v1.5.1\n",
      "\n",
      "\tLoading Package : data.table v1.15.2\n",
      "\n",
      "\tLoading Package : matrixStats v1.3.0\n",
      "\n",
      "\tLoading Package : S4Vectors v0.40.2\n",
      "\n",
      "\tLoading Package : GenomicRanges v1.54.1\n",
      "\n",
      "\tLoading Package : BiocGenerics v0.48.1\n",
      "\n",
      "\tLoading Package : Matrix v1.6.5\n",
      "\n",
      "\tLoading Package : Rcpp v1.0.12\n",
      "\n",
      "\tLoading Package : SummarizedExperiment v1.30.2\n",
      "\n",
      "\tLoading Package : rhdf5 v2.44.0\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘sp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::pattern’ by ‘grid::pattern’ when loading ‘genomation’”\n"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(magrittr)\n",
    "library(genomation)\n",
    "library(GenomicRanges)\n",
    "library(Matrix)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e85020d7-d358-492d-9358-c81270f4b624",
   "metadata": {},
   "source": [
    "Specify file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1bec6690-52ec-450d-bba7-7d4528a185cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_gtf_path = \"/maps/projects/ralab/data/genome/hg38/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "abc_genes_path = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/software/scE2G_pipeline/241203/scE2G/ENCODE_rE2G/ABC/reference/hg38/CollapsedGeneBounds.hg38.TSS500bp.bed\"\n",
    "path.seurat = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/4.merge_peaks_from_22_cell_types.240523/obj.seurat.rds\"\n",
    "path.pairs.E2G = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/1.prepare_data/4.merge_peaks_from_22_cell_types.240523/pairs.rds\"\n",
    "dir.output = \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/ArchR_250226/BMMC22.merge.250317/BMMC22.merge.250317/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "340dac32-44ac-4231-b964-b5f3b9e2dbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "n.cores = 16\n",
    "maxDist = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da973385-4c81-4b1c-9669-cf5933b23a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 16.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRThreads(threads = n.cores) \n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeb26496-d3b9-4778-92ed-9c63a5018aef",
   "metadata": {},
   "source": [
    "Import seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c9b19a09-be95-4d9a-bdc3-0530980d4757",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.seurat = readRDS(path.seurat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86be3e66-c954-4620-9e64-c6a1c0fcb84c",
   "metadata": {},
   "source": [
    "Import candidate E-G pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6f363969-bf0b-45ce-858c-a8cb64cc210a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G = readRDS(path.pairs.E2G)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3538ffdc-4fed-48eb-9c9d-86805a91cf06",
   "metadata": {},
   "source": [
    "Create an ArchRProject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0300f485-fafa-4cef-b9d0-534108ec6f95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 13\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"86350e844e0a3a1991e9c650ccac7f43\" \"555000121b665f095f13712669597ac8\"\n",
      "  .. ..@ cells: Named chr [1:6224] \"TAGTTGTCACCCTCAC-1\" \"CTATGGCCATAACGGG-1\" \"CCGCACACAGGTTAAA-1\" \"TCATTTGGTAATGGAA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:6224] \"s1d1_TAGTTGTCACCCTCAC-1\" \"s1d1_CTATGGCCATAACGGG-1\" \"s1d1_CCGCACACAGGTTAAA-1\" \"s1d1_TCATTTGGTAATGGAA-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"3ee058ae058031e3f233c68b0db9e6b9\" \"1039e761a9da92608ebabb952d11b3c8\"\n",
      "  .. ..@ cells: Named chr [1:6740] \"CTACCGGCAGTTGCGT-1\" \"GATGCAGCAGCAACAG-1\" \"TGAGTGAAGAGGAAGG-1\" \"CCGCTAAAGGACCGCT-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:6740] \"s1d2_CTACCGGCAGTTGCGT-1\" \"s1d2_GATGCAGCAGCAACAG-1\" \"s1d2_TGAGTGAAGAGGAAGG-1\" \"s1d2_CCGCTAAAGGACCGCT-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"2276352420919cf89b8bb75ac819824c\" \"75537e8fd45162781c38e6a7db2c86cf\"\n",
      "  .. ..@ cells: Named chr [1:4279] \"TCACCTGGTTAGGTTG-1\" \"CGTTAACAGGTGTCCA-1\" \"ATTCGTTTCAGTATTG-1\" \"GGACCGAAGTGAGGTA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:4279] \"s1d3_TCACCTGGTTAGGTTG-1\" \"s1d3_CGTTAACAGGTGTCCA-1\" \"s1d3_ATTCGTTTCAGTATTG-1\" \"s1d3_GGACCGAAGTGAGGTA-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"1fc4f6dc8e1c53094db34b5c7a766a87\" \"b8c37340d7a1d3941d55772be354bd9a\"\n",
      "  .. ..@ cells: Named chr [1:4220] \"ACGTTACAGGCATTAC-1\" \"GGTGATTTCGCTAGAT-1\" \"ACAGGATCACTAAGAA-1\" \"CGCTACTTCATCCACC-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:4220] \"s2d1_ACGTTACAGGCATTAC-1\" \"s2d1_GGTGATTTCGCTAGAT-1\" \"s2d1_ACAGGATCACTAAGAA-1\" \"s2d1_CGCTACTTCATCCACC-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"a2acfe30c69e24ee4fd35cd8b067c630\" \"9ffc0908f69ebc6daf04984782449572\"\n",
      "  .. ..@ cells: Named chr [1:6111] \"TAATGGTGTTAGCGTA-1\" \"TCTAACTTCGTTTCGC-1\" \"GGCGCTAGTTTAGCTC-1\" \"AAAGCACCACTAAGCC-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:6111] \"s2d4_TAATGGTGTTAGCGTA-1\" \"s2d4_TCTAACTTCGTTTCGC-1\" \"s2d4_GGCGCTAGTTTAGCTC-1\" \"s2d4_AAAGCACCACTAAGCC-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"2176bbdb8bb8100786137c9f2a51b3ca\" \"90506bed3c58854bc99c88adb8d34e91\"\n",
      "  .. ..@ cells: Named chr [1:4895] \"GCTTGACCATCTAGCA-1\" \"GATTCAATCTAATCTC-1\" \"TAAGGACCAAAGCGCA-1\" \"AGCGTGCTCATTGCAA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:4895] \"s2d5_GCTTGACCATCTAGCA-1\" \"s2d5_GATTCAATCTAATCTC-1\" \"s2d5_TAAGGACCAAAGCGCA-1\" \"s2d5_AGCGTGCTCATTGCAA-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"2a5bab58fd0b28bc8ef3723ed57e6c08\" \"238b44ce8ad4cb929abf90c361c5d9f1\"\n",
      "  .. ..@ cells: Named chr [1:6781] \"GAGGACTAGCTTTGTT-1\" \"TAGGTTGGTTACCTGT-1\" \"AGGAGCTAGACAAGTG-1\" \"GCGTAACGTCATAGAT-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:6781] \"s3d10_GAGGACTAGCTTTGTT-1\" \"s3d10_TAGGTTGGTTACCTGT-1\" \"s3d10_AGGAGCTAGACAAGTG-1\" \"s3d10_GCGTAACGTCATAGAT-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"983f45aef00cf12883d18d626d839a36\" \"72f5dc7b926b9533af5db0f68b5b69f6\"\n",
      "  .. ..@ cells: Named chr [1:4325] \"AGTCGCATCTCACTAT-1\" \"ATGTCATCAATCTCTC-1\" \"ATTGACTCAGGAATCG-1\" \"CTAATGTCACAACAGG-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:4325] \"s3d3_AGTCGCATCTCACTAT-1\" \"s3d3_ATGTCATCAATCTCTC-1\" \"s3d3_ATTGACTCAGGAATCG-1\" \"s3d3_CTAATGTCACAACAGG-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"5d837fe9f6aafc00db7add9a9fe3ef29\" \"b296ae5676ca1413959c46f7df802834\"\n",
      "  .. ..@ cells: Named chr [1:1679] \"GTACAATGTTCCATTA-1\" \"TGTGGCGGTGTTGCAA-1\" \"GCGGTTATCGACCTGA-1\" \"CCTGTAACAGGCTTCG-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:1679] \"s3d6_GTACAATGTTCCATTA-1\" \"s3d6_TGTGGCGGTGTTGCAA-1\" \"s3d6_GCGGTTATCGACCTGA-1\" \"s3d6_CCTGTAACAGGCTTCG-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"4f4da183ea8734759ddc8987b29aac25\" \"d9b91a39dbf76399e9f01458cf21e627\"\n",
      "  .. ..@ cells: Named chr [1:1771] \"CCAAACCCACTAGCGT-1\" \"TTAACTGAGAAACTGG-1\" \"GGAGCGATCACAGCCA-1\" \"TTTGTGAAGCTAATCA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:1771] \"s3d7_CCAAACCCACTAGCGT-1\" \"s3d7_TTAACTGAGAAACTGG-1\" \"s3d7_GGAGCGATCACAGCCA-1\" \"s3d7_TTTGTGAAGCTAATCA-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"c1bcab73870a86531668a7e4c5fd9f82\" \"fb36f28f535f141bf0a4af04c17d5bd9\"\n",
      "  .. ..@ cells: Named chr [1:8023] \"AGCGATTTCATCCTGC-1\" \"TTGACTAAGGTCAAAG-1\" \"AGGAACGGTGAGGTGA-1\" \"TAGGGTTTCCGTGACA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:8023] \"s4d1_AGCGATTTCATCCTGC-1\" \"s4d1_TTGACTAAGGTCAAAG-1\" \"s4d1_AGGAACGGTGAGGTGA-1\" \"s4d1_TAGGGTTTCCGTGACA-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"65a18b88a6a8f2e1b7282c2c08da1067\" \"cc03488d12a1f1e49ae43613e0c61606\"\n",
      "  .. ..@ cells: Named chr [1:9876] \"TTCGCAACAATAATGG-1\" \"TATCCAGCAATTGAGA-1\" \"ACTCGCGCAAACTGTT-1\" \"GACTATTCATGTCGCG-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:9876] \"s4d8_TTCGCAACAATAATGG-1\" \"s4d8_TATCCAGCAATTGAGA-1\" \"s4d8_ACTCGCGCAAACTGTT-1\" \"s4d8_GACTATTCATGTCGCG-1\" ...\n",
      " $ :Formal class 'Fragment' [package \"Signac\"] with 3 slots\n",
      "  .. ..@ path : chr \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellra\"| __truncated__\n",
      "  .. ..@ hash : chr [1:2] \"10d377a58a8fbc07927e8d3a4ec79e73\" \"e3dbabb04370289ef70c67216288b530\"\n",
      "  .. ..@ cells: Named chr [1:4325] \"TGTCCTTCAGGCTAGA-1\" \"AATACCGGTAAGGTCC-1\" \"TGTAAAGCACTTGTTC-1\" \"CAGTATGGTTCGCTCA-1\" ...\n",
      "  .. .. ..- attr(*, \"names\")= chr [1:4325] \"s4d9_TGTCCTTCAGGCTAGA-1\" \"s4d9_AATACCGGTAAGGTCC-1\" \"s4d9_TGTAAAGCACTTGTTC-1\" \"s4d9_CAGTATGGTTCGCTCA-1\" ...\n"
     ]
    }
   ],
   "source": [
    "str(Fragments(obj.seurat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1753ffaa-89e2-43bf-bcf6-19937782d63a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-createArrows-183fb6441ec1b3-Date-2025-03-17_Time-17-34-20.284561.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Cleaning Temporary Files\n",
      "\n",
      "2025-03-17 17:34:20.655211 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-createArrows-183fb6441ec1b3-Date-2025-03-17_Time-17-34-20.284561.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ArrowFiles <- createArrowFiles(\n",
    "  inputFiles = c(Fragments(obj.seurat)[[1]]@path,\n",
    "                 Fragments(obj.seurat)[[2]]@path,\n",
    "                 Fragments(obj.seurat)[[3]]@path,\n",
    "                 Fragments(obj.seurat)[[4]]@path,\n",
    "                 Fragments(obj.seurat)[[5]]@path,\n",
    "                 Fragments(obj.seurat)[[6]]@path,\n",
    "                 Fragments(obj.seurat)[[7]]@path,\n",
    "                 Fragments(obj.seurat)[[8]]@path,\n",
    "                 Fragments(obj.seurat)[[9]]@path,\n",
    "                 Fragments(obj.seurat)[[10]]@path,\n",
    "                 Fragments(obj.seurat)[[11]]@path,\n",
    "                 Fragments(obj.seurat)[[12]]@path,\n",
    "                 Fragments(obj.seurat)[[13]]@path),\n",
    "  sampleNames = c(\"s1d1\",\n",
    "                  \"s1d2\",\n",
    "                  \"s1d3\",\n",
    "                  \"s2d1\",\n",
    "                  \"s2d4\",\n",
    "                  \"s2d5\",\n",
    "                  \"s3d10\",\n",
    "                  \"s3d3\",\n",
    "                  \"s3d6\",\n",
    "                  \"s3d7\",\n",
    "                  \"s4d1\",\n",
    "                  \"s4d8\",\n",
    "                  \"s4d9\"),\n",
    "  validBarcodes = list(s1d1 = unname(Fragments(obj.seurat)[[1]]@cells),\n",
    "                       s1d2 = unname(Fragments(obj.seurat)[[2]]@cells),\n",
    "                       s1d3 = unname(Fragments(obj.seurat)[[3]]@cells),\n",
    "                       s2d1 = unname(Fragments(obj.seurat)[[4]]@cells),\n",
    "                       s2d4 = unname(Fragments(obj.seurat)[[5]]@cells),\n",
    "                       s2d5 = unname(Fragments(obj.seurat)[[6]]@cells),\n",
    "                       s3d10 = unname(Fragments(obj.seurat)[[7]]@cells),\n",
    "                       s3d3 = unname(Fragments(obj.seurat)[[8]]@cells),\n",
    "                       s3d6 = unname(Fragments(obj.seurat)[[9]]@cells),\n",
    "                       s3d7 = unname(Fragments(obj.seurat)[[10]]@cells),\n",
    "                       s4d1 = unname(Fragments(obj.seurat)[[11]]@cells),\n",
    "                       s4d8 = unname(Fragments(obj.seurat)[[12]]@cells),\n",
    "                       s4d9 = unname(Fragments(obj.seurat)[[13]]@cells)),\n",
    "  addTileMat = T,\n",
    "  addGeneScoreMat = F,\n",
    "  force = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ada3db45-d050-4659-a7cf-2891e04b1249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'s3d10.arrow'</li><li>'s3d3.arrow'</li><li>'s3d7.arrow'</li><li>'s4d1.arrow'</li><li>'s1d1.arrow'</li><li>'s4d8.arrow'</li><li>'s2d4.arrow'</li><li>'s2d1.arrow'</li><li>'s2d5.arrow'</li><li>'s1d2.arrow'</li><li>'s1d3.arrow'</li><li>'s4d9.arrow'</li><li>'s3d6.arrow'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 's3d10.arrow'\n",
       "\\item 's3d3.arrow'\n",
       "\\item 's3d7.arrow'\n",
       "\\item 's4d1.arrow'\n",
       "\\item 's1d1.arrow'\n",
       "\\item 's4d8.arrow'\n",
       "\\item 's2d4.arrow'\n",
       "\\item 's2d1.arrow'\n",
       "\\item 's2d5.arrow'\n",
       "\\item 's1d2.arrow'\n",
       "\\item 's1d3.arrow'\n",
       "\\item 's4d9.arrow'\n",
       "\\item 's3d6.arrow'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 's3d10.arrow'\n",
       "2. 's3d3.arrow'\n",
       "3. 's3d7.arrow'\n",
       "4. 's4d1.arrow'\n",
       "5. 's1d1.arrow'\n",
       "6. 's4d8.arrow'\n",
       "7. 's2d4.arrow'\n",
       "8. 's2d1.arrow'\n",
       "9. 's2d5.arrow'\n",
       "10. 's1d2.arrow'\n",
       "11. 's1d3.arrow'\n",
       "12. 's4d9.arrow'\n",
       "13. 's3d6.arrow'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"s3d10.arrow\" \"s3d3.arrow\"  \"s3d7.arrow\"  \"s4d1.arrow\"  \"s1d1.arrow\" \n",
       " [6] \"s4d8.arrow\"  \"s2d4.arrow\"  \"s2d1.arrow\"  \"s2d5.arrow\"  \"s1d2.arrow\" \n",
       "[11] \"s1d3.arrow\"  \"s4d9.arrow\"  \"s3d6.arrow\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ArrowFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5ba28283-d488-4982-8d4c-745fc3f59365",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(dir.output,\"ArchR\",sep=\"/\"),recursive = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aa217507-8323-43bf-b23c-0ec25ee33c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Using GeneAnnotation set by addArchRGenome(Hg38)!\n",
      "\n",
      "Validating Arrows...\n",
      "\n",
      "Getting SampleNames...\n",
      "\n",
      "\n",
      "\n",
      "Copying ArrowFiles to Ouptut Directory! If you want to save disk space set copyArrows = FALSE\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "9 \n",
      "10 \n",
      "11 \n",
      "12 \n",
      "13 \n",
      "\n",
      "\n",
      "Getting Cell Metadata...\n",
      "\n",
      "\n",
      "\n",
      "Merging Cell Metadata...\n",
      "\n",
      "Initializing ArchRProject...\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR <- ArchRProject(\n",
    "  ArrowFiles = ArrowFiles, \n",
    "  outputDirectory = paste(dir.output,\"ArchR\",sep=\"/\"),\n",
    "  copyArrows = TRUE \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f9795403-545f-4385-aa38-2c70b1b2fb21",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: ArchRProject \n",
       "outputDirectory: /maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/analysis/E2G_240503/data/BMMC/2.Genome_wide_prediction/ArchR_250226/BMMC22.merge.250317/BMMC22.merge.250317/ArchR \n",
       "samples(13): s3d10 s3d3 ... s4d9 s3d6\n",
       "sampleColData names(1): ArrowFiles\n",
       "cellColData names(13): Sample TSSEnrichment ... nDiFrags BlacklistRatio\n",
       "numberOfCells(1): 67013\n",
       "medianTSS(1): 16.107\n",
       "medianFrags(1): 9747"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obj.ArchR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ee876dd5-0b22-46fd-9015-7e22144c5024",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking Inputs...\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addIterativeLSI-183fb62d34030a-Date-2025-03-17_Time-18-15-51.513029.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-03-17 18:15:53.578401 : Computing Total Across All Features, 0.019 mins elapsed.\n",
      "\n",
      "2025-03-17 18:16:00.968106 : Computing Top Features, 0.143 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-03-17 18:16:06.095866 : Running LSI (1 of 2) on Top Features, 0.228 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-03-17 18:16:06.167298 : Sampling Cells (N = 10007) for Estimated LSI, 0.229 mins elapsed.\n",
      "\n",
      "2025-03-17 18:16:06.172835 : Creating Sampled Partial Matrix, 0.229 mins elapsed.\n",
      "\n",
      "2025-03-17 18:16:18.980752 : Computing Estimated LSI (projectAll = FALSE), 0.443 mins elapsed.\n",
      "\n",
      "2025-03-17 18:17:09.384171 : Identifying Clusters, 1.283 mins elapsed.\n",
      "\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "2025-03-17 18:17:25.041169 : Identified 6 Clusters, 1.544 mins elapsed.\n",
      "\n",
      "2025-03-17 18:17:25.068293 : Saving LSI Iteration, 1.544 mins elapsed.\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************************************************************\n",
      "2025-03-17 18:17:42.556368 : ERROR Found in .saveIteration for  \n",
      "LogFile = ArchRLogs/ArchR-addIterativeLSI-183fb62d34030a-Date-2025-03-17_Time-18-15-51.513029.log\n",
      "\n",
      "<simpleError in g$grobs[[legend]]: no such index at level 2\n",
      ">\n",
      "\n",
      "************************************************************\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-03-17 18:17:42.560317 : Creating Cluster Matrix on the total Group Features, 1.836 mins elapsed.\n",
      "\n",
      "2025-03-17 18:18:04.990031 : Computing Variable Features, 2.21 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-03-17 18:18:05.150654 : Running LSI (2 of 2) on Variable Features, 2.212 mins elapsed.\n",
      "###########\n",
      "\n",
      "2025-03-17 18:18:05.194739 : Creating Partial Matrix, 2.213 mins elapsed.\n",
      "\n",
      "2025-03-17 18:18:32.040475 : Computing LSI, 2.661 mins elapsed.\n",
      "\n",
      "2025-03-17 18:19:52.699161 : Finished Running IterativeLSI, 4.005 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addIterativeLSI(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    useMatrix = \"TileMatrix\", \n",
    "    name = \"IterativeLSI\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2c1d334b-7fd7-425a-bb8d-7be4cbedd3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj.ArchR = addPeakSet(obj.ArchR,\n",
    "                       peakSet = reduce(pairs.E2G))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e25c14db-eb4e-4b94-8862-eb34246a365c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeakMatrix-183fb64d68f7b4-Date-2025-03-17_Time-18-19-58.45656.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-03-17 18:19:58.750583 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeakMatrix-183fb64d68f7b4-Date-2025-03-17_Time-18-19-58.45656.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addPeakMatrix(obj.ArchR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8887890f-a614-412f-83fb-9a44ecbd0e74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n",
      "Importing Feature Matrix 1 of 1\n",
      "\n",
      "Re-ordering RNA matricies for consistency.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "paths.h5 = c(\"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site1_donor1/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site1_donor2/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site1_donor3/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site2_donor1/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site2_donor4/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site2_donor5/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site3_donor10/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site3_donor3/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site3_donor6/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site3_donor7/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site4_donor1/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site4_donor8/outs/raw_feature_bc_matrix.h5\",\n",
    "             \"/maps/projects/ralab_nnfc-AUDIT/people/lpm537/project/E2G/processed/10x_multiome_240430/neurips2021_BMMC/cellranger_res/site4_donor9/outs/raw_feature_bc_matrix.h5\")\n",
    "names.sample = c(\"s1d1\",\n",
    "                 \"s1d2\",\n",
    "                 \"s1d3\",\n",
    "                 \"s2d1\",\n",
    "                 \"s2d4\",\n",
    "                 \"s2d5\",\n",
    "                 \"s3d10\",\n",
    "                 \"s3d3\",\n",
    "                 \"s3d6\",\n",
    "                 \"s3d7\",\n",
    "                 \"s4d1\",\n",
    "                 \"s4d8\",\n",
    "                 \"s4d9\")\n",
    "list.seRNA = lapply(1:length(names.sample),function(x){\n",
    "    import10xFeatureMatrix(paths.h5[x],\n",
    "                           names.sample[x])\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f52ba95e-a547-496d-9296-2c4ea4445ca4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[1]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 727198 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(727198): s1d1#AAACAGCCAAACAACA-1 s1d1#AAACAGCCAAACATAG-1 ...\n",
       "  s1d1#TTTGTTGGTTTGTCTA-1 s1d1#TTTGTTGGTTTGTGGA-1\n",
       "colData names(0):\n",
       "\n",
       "[[2]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 726937 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L2 DDX11L1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(726937): s1d2#AAACAGCCAAACAACA-1 s1d2#AAACAGCCAAACATAG-1 ...\n",
       "  s1d2#TTTGTTGGTTTGTGGA-1 s1d2#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):\n",
       "\n",
       "[[3]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 724480 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(724480): s1d3#AAACAGCCAAACAACA-1 s1d3#AAACAGCCAAACATAG-1 ...\n",
       "  s1d3#TTTGTTGGTTTGTGGA-1 s1d3#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):\n",
       "\n",
       "[[4]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 721194 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(721194): s2d1#AAACAGCCAAACAACA-1 s2d1#AAACAGCCAAACATAG-1 ...\n",
       "  s2d1#TTTGTTGGTTTGTGGA-1 s2d1#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):\n",
       "\n",
       "[[5]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 722665 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(722665): s2d4#AAACAGCCAAACAACA-1 s2d4#AAACAGCCAAACATAG-1 ...\n",
       "  s2d4#TTTGTTGGTTTGTGGA-1 s2d4#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):\n",
       "\n",
       "[[6]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 691458 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(691458): s2d5#AAACAGCCAAACAACA-1 s2d5#AAACAGCCAAACATAG-1 ...\n",
       "  s2d5#TTTGTTGGTTTGTGGA-1 s2d5#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):\n",
       "\n",
       "[[7]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 729180 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(729180): s3d10#AAACAGCCAAACAACA-1 s3d10#AAACAGCCAAACATAG-1 ...\n",
       "  s3d10#TTTGTTGGTTTGTGGA-1 s3d10#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):\n",
       "\n",
       "[[8]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 726129 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(726129): s3d3#AAACAGCCAAACAACA-1 s3d3#AAACAGCCAAACATAG-1 ...\n",
       "  s3d3#TTTGTTGGTTTGTGGA-1 s3d3#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):\n",
       "\n",
       "[[9]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 732427 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(732427): s3d6#AAACAGCCAAACAACA-1 s3d6#AAACAGCCAAACATAG-1 ...\n",
       "  s3d6#TTTGTTGGTTTGTGGA-1 s3d6#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):\n",
       "\n",
       "[[10]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 736031 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(736031): s3d7#AAACAGCCAAACAACA-1 s3d7#AAACAGCCAAACATAG-1 ...\n",
       "  s3d7#TTTGTTGGTTTGTGGA-1 s3d7#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):\n",
       "\n",
       "[[11]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 694554 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(694554): s4d1#AAACAGCCAAACAACA-1 s4d1#AAACAGCCAAACATAG-1 ...\n",
       "  s4d1#TTTGTTGGTTTGTGGA-1 s4d1#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):\n",
       "\n",
       "[[12]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 692673 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(692673): s4d8#AAACAGCCAAACAACA-1 s4d8#AAACAGCCAAACATAG-1 ...\n",
       "  s4d8#TTTGTTGGTTTGTGGA-1 s4d8#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):\n",
       "\n",
       "[[13]]\n",
       "class: RangedSummarizedExperiment \n",
       "dim: 61218 597840 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61218): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(597840): s4d9#AAACAGCCAAACAACA-1 s4d9#AAACAGCCAAACATAG-1 ...\n",
       "  s4d9#TTTGTTGGTTTGTGGA-1 s4d9#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "list.seRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "93d78aef-1337-4876-9bc8-38bfd7d6e112",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "61218"
      ],
      "text/latex": [
       "61218"
      ],
      "text/markdown": [
       "61218"
      ],
      "text/plain": [
       "[1] 61218"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "61165"
      ],
      "text/latex": [
       "61165"
      ],
      "text/markdown": [
       "61165"
      ],
      "text/plain": [
       "[1] 61165"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "id.intersect = rowData(list.seRNA[[1]])[,\"id\"]\n",
    "length(id.intersect)\n",
    "for(n.tmp in 2:13){\n",
    "    id.intersect = intersect(id.intersect,rowData(list.seRNA[[n.tmp]])[,\"id\"])\n",
    "}\n",
    "length(id.intersect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "504f6e06-183d-4ecf-a745-32719a2972d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 61165 9222766 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61165): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(9222766): s1d1#AAACAGCCAAACAACA-1 s1d1#AAACAGCCAAACATAG-1 ...\n",
       "  s4d9#TTTGTTGGTTTGTGGA-1 s4d9#TTTGTTGGTTTGTTGC-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA = list.seRNA[[1]][rowData(list.seRNA[[1]])[,\"id\"] %in% id.intersect]\n",
    "for(n.tmp in 2:13){\n",
    "    seRNA = cbind(seRNA,list.seRNA[[n.tmp]][rowData(list.seRNA[[n.tmp]])[,\"id\"] %in% id.intersect])\n",
    "}\n",
    "seRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "895234f3-e668-4e83-adf0-562024a6711b",
   "metadata": {},
   "outputs": [],
   "source": [
    "seRNA = seRNA[,rownames(obj.ArchR@cellColData)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4656d98c-a9cc-486a-afaa-fe59ca6ef055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 61165 67013 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(61165): DDX11L1 MIR6859-1 ... RNA5-8SN5 ENSG00000273937\n",
       "rowData names(5): feature_type genome id interval name\n",
       "colnames(67013): s3d10#GTAGCTGTCAGCAAAG-1 s3d10#AACTAGTGTTACCTGT-1 ...\n",
       "  s3d6#TCACCGGCAGCCTTGG-1 s3d6#GAACCTGTCGTTACAA-1\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seRNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a879efc-14cf-4e16-bed3-15cb2e82a2dc",
   "metadata": {},
   "source": [
    "Map gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "fa7f01b3-ff5a-459f-bddd-ad388d554826",
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_attributes <- function(gtf_attributes, att_of_interest){\n",
    "  att <- unlist(strsplit(gtf_attributes, \" \"))\n",
    "  if(att_of_interest %in% att){\n",
    "    return(gsub(\"\\\"|;\",\"\", att[which(att %in% att_of_interest)+1]))\n",
    "  } else {\n",
    "    return(NA)}\n",
    "}\n",
    "map_gene_names <- function(rna_matrix, gene_gtf_path, abc_genes_path){\n",
    "    library(dplyr)\n",
    "    library(data.table)\n",
    "    \n",
    "\tgene_ref <- fread(gene_gtf_path, header = FALSE, sep = \"\\t\") %>%\n",
    "\t\tsetNames(c(\"chr\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\")) %>%\n",
    "\t\tdplyr::filter(type == \"gene\")\n",
    "\tgene_ref$gene_ref_name <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_name\"))\n",
    "\tgene_ref$Ensembl_ID <- unlist(lapply(gene_ref$attributes, extract_attributes, \"gene_id\"))\n",
    "\tgene_ref <- dplyr::select(gene_ref, gene_ref_name, Ensembl_ID) %>%\n",
    "\t\tmutate(Ensembl_ID = sub(\"\\\\.\\\\d+$\", \"\", Ensembl_ID)) %>% # remove decimal digits \n",
    "\t\tdistinct()\n",
    "\t\n",
    "\tabc_genes <- fread(abc_genes_path, col.names = c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"Ensembl_ID\", \"gene_type\")) %>%\n",
    "\t\tdplyr::select(name, Ensembl_ID) %>%\n",
    "\t\trename(abc_name = name) %>%\n",
    "\t\tleft_join(gene_ref, by = \"Ensembl_ID\") %>%\n",
    "\t\tgroup_by(Ensembl_ID) %>% # remove cases where multiple genes map to one ensembl ID\n",
    "\t\tfilter(n() == 1) %>%\n",
    "\t\tungroup()\n",
    "\n",
    "\tgene_key <- abc_genes$abc_name\n",
    "\tnames(gene_key) <- abc_genes$gene_ref_name\n",
    "\n",
    "\t# remove genes not in our gene universe\t\n",
    "\trow_sub <- intersect(rownames(rna_matrix), names(gene_key)) # gene ref names\n",
    "\trna_matrix_filt <- rna_matrix[row_sub,] # still gene ref names\n",
    "\trownames(rna_matrix_filt) <- gene_key[row_sub] # converted to abc names\n",
    "\n",
    "\treturn(rna_matrix_filt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0a713a77-2752-4f1e-bbfa-6e262e8e6f19",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biostrings’:\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seRNA.rename = map_gene_names(seRNA,gene_gtf_path, abc_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "65cb2097-9075-4ee4-9e54-dd288dc50736",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addGeneExpressionMatrix-183fb641efd635-Date-2025-03-17_Time-18-28-36.836904.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Overlap w/ scATAC = 1\n",
      "\n",
      "2025-03-17 18:28:38.010964 : \n",
      "\n",
      "Overlap Per Sample w/ scATAC : s1d1=6149,s1d2=6685,s1d3=4260,s2d1=3966,s2d4=5321,s2d5=4558,s3d10=6724,s3d3=4269,s3d6=1645,s3d7=1627,s4d1=7905,s4d8=9705,s4d9=4199\n",
      "\n",
      "2025-03-17 18:28:38.117947 : \n",
      "\n",
      "2025-03-17 18:28:43.951024 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addGeneExpressionMatrix-183fb641efd635-Date-2025-03-17_Time-18-28-36.836904.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj.ArchR = addGeneExpressionMatrix(obj.ArchR,\n",
    "                                    seRNA = seRNA.rename,\n",
    "                                    strictMatch = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4318703-a143-40c0-b159-274d05033bbc",
   "metadata": {},
   "source": [
    "Run ArchR prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "47c8fd5f-19ec-46e6-af87-57502c1e1ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addPeak2GeneLinks-183fb61724756-Date-2025-03-17_Time-18-32-48.144047.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-03-17 18:32:50.670992 : Getting Available Matrices, 0.042 mins elapsed.\n",
      "\n",
      "2025-03-17 18:33:07.476118 : Filtered Low Prediction Score Cells (0 of 67013, 0), 0.062 mins elapsed.\n",
      "\n",
      "2025-03-17 18:33:11.194274 : Computing KNN, 0.124 mins elapsed.\n",
      "\n",
      "2025-03-17 18:33:11.667298 : Identifying Non-Overlapping KNN pairs, 0.132 mins elapsed.\n",
      "\n",
      "2025-03-17 18:33:13.868196 : Identified 500 Groupings!, 0.169 mins elapsed.\n",
      "\n",
      "2025-03-17 18:33:13.96766 : Getting Group RNA Matrix, 0.171 mins elapsed.\n",
      "\n",
      "2025-03-17 18:34:25.291775 : Getting Group ATAC Matrix, 1.359 mins elapsed.\n",
      "\n",
      "2025-03-17 18:36:14.262678 : Normalizing Group Matrices, 3.176 mins elapsed.\n",
      "\n",
      "2025-03-17 18:36:28.574838 : Finding Peak Gene Pairings, 3.414 mins elapsed.\n",
      "\n",
      "2025-03-17 18:36:30.719868 : Computing Correlations, 3.45 mins elapsed.\n",
      "\n",
      "2025-03-17 18:37:13.665362 : Completed Peak2Gene Correlations!, 4.166 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeak2GeneLinks-183fb61724756-Date-2025-03-17_Time-18-32-48.144047.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "obj.ArchR = addPeak2GeneLinks(\n",
    "    ArchRProj = obj.ArchR,\n",
    "    reducedDims = \"IterativeLSI\",\n",
    "    useMatrix = \"GeneExpressionMatrix\",\n",
    "    maxDist = maxDist,\n",
    "    threads = n.cores\n",
    ")\n",
    "end_time <- Sys.time()\n",
    "execution_time <- end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "128d0065-5f22-495b-b00d-743fc9843337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 4.426522 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "execution_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "36bd2df2-6b32-4b81-bd4f-39ef8913b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g = metadata(obj.ArchR@peakSet)$Peak2GeneLinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e7730413-8fd4-4952-bb13-0df7a3b15b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata(p2g)[[1]]$PeakName = paste(seqnames(metadata(p2g)[[1]]),\n",
    "                                    ranges(metadata(p2g)[[1]]),sep=\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "aeea3810-6735-4aff-9bd7-1697bef979d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2g$PairName = paste(metadata(p2g)[[1]][p2g$idxATAC]$PeakName,\n",
    "                     metadata(p2g)[[2]][p2g$idxRNA]$name,\n",
    "                     sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "87f975ec-6ba4-4696-ad6c-2af65da7aa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 11910478 rows and 7 columns\n",
       "                                idxATAC    idxRNA Correlation       FDR\n",
       "                              <integer> <integer>   <numeric> <numeric>\n",
       "chr1-10010-10536_FAM138A              1         1         NaN       NaN\n",
       "chr1-96496-96699_FAM138A              2         1         NaN       NaN\n",
       "chr1-115495-115930_FAM138A            3         1         NaN       NaN\n",
       "chr1-180470-181899_FAM138A            4         1         NaN       NaN\n",
       "chr1-184379-184607_FAM138A            5         1         NaN       NaN\n",
       "...                                 ...       ...         ...       ...\n",
       "chrX-156008869-156009386_IL9R    540425     20486  -0.0467826  0.532075\n",
       "chrX-156019809-156020162_IL9R    540426     20486   0.0524075  0.469659\n",
       "chrX-156025098-156025397_IL9R    540427     20486   0.0828137  0.187923\n",
       "chrX-156029831-156030380_IL9R    540428     20486   0.0439540  0.563653\n",
       "chrX-156030497-156030869_IL9R    540429     20486   0.0870693  0.159768\n",
       "                               VarQATAC   VarQRNA               PairName\n",
       "                              <numeric> <numeric>            <character>\n",
       "chr1-10010-10536_FAM138A       0.798627 0.0263595 chr1-10010-10536_FAM..\n",
       "chr1-96496-96699_FAM138A       0.178284 0.0263595 chr1-96496-96699_FAM..\n",
       "chr1-115495-115930_FAM138A     0.743835 0.0263595 chr1-115495-115930_F..\n",
       "chr1-180470-181899_FAM138A     0.904913 0.0263595 chr1-180470-181899_F..\n",
       "chr1-184379-184607_FAM138A     0.164902 0.0263595 chr1-184379-184607_F..\n",
       "...                                 ...       ...                    ...\n",
       "chrX-156008869-156009386_IL9R  0.596981  0.266816 chrX-156008869-15600..\n",
       "chrX-156019809-156020162_IL9R  0.488364  0.266816 chrX-156019809-15602..\n",
       "chrX-156025098-156025397_IL9R  0.329416  0.266816 chrX-156025098-15602..\n",
       "chrX-156029831-156030380_IL9R  0.756283  0.266816 chrX-156029831-15603..\n",
       "chrX-156030497-156030869_IL9R  0.796954  0.266816 chrX-156030497-15603.."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(p2g) = p2g$PairName\n",
    "p2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c7209a32-3904-4734-93af-9c22049f652a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs.E2G.res = pairs.E2G[pairs.E2G$PairName %in% p2g$PairName]\n",
    "mcols(pairs.E2G.res)[,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")] = p2g[pairs.E2G.res$PairName,c(\"Correlation\",\"FDR\",\"VarQATAC\",\"VarQRNA\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1ab4ed70-ce9e-4857-bd1b-63a81d27a401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 9765178 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100004103-100004301      * |         AGL\n",
       "        [2]     chr1 100004103-100004301      * |      CDC14A\n",
       "        [3]     chr1 100004103-100004301      * |         DBT\n",
       "        [4]     chr1 100004103-100004301      * |       EXTL2\n",
       "        [5]     chr1 100004103-100004301      * |       FRRS1\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [9765174]     chrX   99980528-99980848      * |      PCDH19\n",
       "  [9765175]     chrX   99980528-99980848      * |      TSPAN6\n",
       "  [9765176]     chrX   99980528-99980848      * |        XKRX\n",
       "  [9765177]     chrX     9999312-9999592      * |       TBL1X\n",
       "  [9765178]     chrX     9999312-9999592      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-100004103-10000.. chr1-100004103-10000.. -0.00771067\n",
       "        [2] chr1-100004103-10000.. chr1-100004103-10000.. -0.14538288\n",
       "        [3] chr1-100004103-10000.. chr1-100004103-10000..  0.03498581\n",
       "        [4] chr1-100004103-10000.. chr1-100004103-10000.. -0.05517044\n",
       "        [5] chr1-100004103-10000.. chr1-100004103-10000..  0.10853104\n",
       "        ...                    ...                    ...         ...\n",
       "  [9765174] chrX-99980528-99980848 chrX-99980528-999808..  -0.0169547\n",
       "  [9765175] chrX-99980528-99980848 chrX-99980528-999808..  -0.0333675\n",
       "  [9765176] chrX-99980528-99980848 chrX-99980528-999808..  -0.0255120\n",
       "  [9765177]   chrX-9999312-9999592 chrX-9999312-9999592..  -0.0904746\n",
       "  [9765178]   chrX-9999312-9999592 chrX-9999312-9999592..   0.0684920\n",
       "                   FDR  VarQATAC   VarQRNA\n",
       "             <numeric> <numeric> <numeric>\n",
       "        [1] 0.93245621 0.0305868  0.785219\n",
       "        [2] 0.00695079 0.0305868  0.978815\n",
       "        [3] 0.66323816 0.0305868  0.632139\n",
       "        [4] 0.43963581 0.0305868  0.510202\n",
       "        [5] 0.06157165 0.0305868  0.662599\n",
       "        ...        ...       ...       ...\n",
       "  [9765174]   0.846692 0.0149881  0.205018\n",
       "  [9765175]   0.680799 0.0149881  0.248853\n",
       "  [9765176]   0.763234 0.0149881  0.339793\n",
       "  [9765177]   0.139417 0.1619713  0.963292\n",
       "  [9765178]   0.304900 0.1619713  0.776823\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairs.E2G.res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46485351-50d7-461a-bf09-d795d5bdab84",
   "metadata": {},
   "source": [
    "Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6d424e21-6a41-4422-b76c-6f1c0bda2222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 9765178 ranges and 7 metadata columns:\n",
       "            seqnames              ranges strand |  TargetGene\n",
       "               <Rle>           <IRanges>  <Rle> | <character>\n",
       "        [1]     chr1 100004103-100004301      * |         AGL\n",
       "        [2]     chr1 100004103-100004301      * |      CDC14A\n",
       "        [3]     chr1 100004103-100004301      * |         DBT\n",
       "        [4]     chr1 100004103-100004301      * |       EXTL2\n",
       "        [5]     chr1 100004103-100004301      * |       FRRS1\n",
       "        ...      ...                 ...    ... .         ...\n",
       "  [9765174]     chrX   99980528-99980848      * |      PCDH19\n",
       "  [9765175]     chrX   99980528-99980848      * |      TSPAN6\n",
       "  [9765176]     chrX   99980528-99980848      * |        XKRX\n",
       "  [9765177]     chrX     9999312-9999592      * |       TBL1X\n",
       "  [9765178]     chrX     9999312-9999592      * |        WWC3\n",
       "                          PeakName               PairName Correlation\n",
       "                       <character>            <character>   <numeric>\n",
       "        [1] chr1-100004103-10000.. chr1-100004103-10000.. -0.00771067\n",
       "        [2] chr1-100004103-10000.. chr1-100004103-10000.. -0.14538288\n",
       "        [3] chr1-100004103-10000.. chr1-100004103-10000..  0.03498581\n",
       "        [4] chr1-100004103-10000.. chr1-100004103-10000.. -0.05517044\n",
       "        [5] chr1-100004103-10000.. chr1-100004103-10000..  0.10853104\n",
       "        ...                    ...                    ...         ...\n",
       "  [9765174] chrX-99980528-99980848 chrX-99980528-999808..  -0.0169547\n",
       "  [9765175] chrX-99980528-99980848 chrX-99980528-999808..  -0.0333675\n",
       "  [9765176] chrX-99980528-99980848 chrX-99980528-999808..  -0.0255120\n",
       "  [9765177]   chrX-9999312-9999592 chrX-9999312-9999592..  -0.0904746\n",
       "  [9765178]   chrX-9999312-9999592 chrX-9999312-9999592..   0.0684920\n",
       "                   FDR  VarQATAC   VarQRNA\n",
       "             <numeric> <numeric> <numeric>\n",
       "        [1] 0.93245621 0.0305868  0.785219\n",
       "        [2] 0.00695079 0.0305868  0.978815\n",
       "        [3] 0.66323816 0.0305868  0.632139\n",
       "        [4] 0.43963581 0.0305868  0.510202\n",
       "        [5] 0.06157165 0.0305868  0.662599\n",
       "        ...        ...       ...       ...\n",
       "  [9765174]   0.846692 0.0149881  0.205018\n",
       "  [9765175]   0.680799 0.0149881  0.248853\n",
       "  [9765176]   0.763234 0.0149881  0.339793\n",
       "  [9765177]   0.139417 0.1619713  0.963292\n",
       "  [9765178]   0.304900 0.1619713  0.776823\n",
       "  -------\n",
       "  seqinfo: 23 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 9765178 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>TargetGene</th><th scope=col>CellType</th><th scope=col>Correlation</th><th scope=col>FDR</th><th scope=col>VarQATAC</th><th scope=col>VarQRNA</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>AGL        </td><td>K562</td><td>-0.007710668</td><td>9.324562e-01</td><td>0.03058681</td><td>0.78521917</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>CDC14A     </td><td>K562</td><td>-0.145382884</td><td>6.950791e-03</td><td>0.03058681</td><td>0.97881480</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>DBT        </td><td>K562</td><td> 0.034985805</td><td>6.632382e-01</td><td>0.03058681</td><td>0.63213902</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>EXTL2      </td><td>K562</td><td>-0.055170442</td><td>4.396358e-01</td><td>0.03058681</td><td>0.51020209</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>FRRS1      </td><td>K562</td><td> 0.108531036</td><td>6.157165e-02</td><td>0.03058681</td><td>0.66259885</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>MFSD14A    </td><td>K562</td><td> 0.034960156</td><td>6.635091e-01</td><td>0.03058681</td><td>0.39387875</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>PLPPR5     </td><td>K562</td><td>-0.035251693</td><td>6.603494e-01</td><td>0.03058681</td><td>0.33876794</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>RTCA       </td><td>K562</td><td> 0.063551546</td><td>3.524594e-01</td><td>0.03058681</td><td>0.68192912</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>RTCA-AS1   </td><td>K562</td><td> 0.123300150</td><td>2.794338e-02</td><td>0.03058681</td><td>0.46983306</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>SASS6      </td><td>K562</td><td> 0.219797819</td><td>9.197888e-06</td><td>0.03058681</td><td>0.82192717</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>SLC30A7    </td><td>K562</td><td>-0.064463623</td><td>3.434446e-01</td><td>0.03058681</td><td>0.79371278</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>SLC35A3    </td><td>K562</td><td>-0.052098762</td><td>4.730377e-01</td><td>0.03058681</td><td>0.62935663</td></tr>\n",
       "\t<tr><td>chr1</td><td>100004103</td><td>100004301</td><td>TRMT13     </td><td>K562</td><td>-0.005524069</td><td>9.522898e-01</td><td>0.03058681</td><td>0.71936933</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>AGL        </td><td>K562</td><td>-0.011651441</td><td>8.960816e-01</td><td>0.37993890</td><td>0.78521917</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>CDC14A     </td><td>K562</td><td>-0.005387489</td><td>9.534942e-01</td><td>0.37993890</td><td>0.97881480</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>DBT        </td><td>K562</td><td> 0.026903933</td><td>7.489431e-01</td><td>0.37993890</td><td>0.63213902</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>EXTL2      </td><td>K562</td><td> 0.077011962</td><td>2.312588e-01</td><td>0.37993890</td><td>0.51020209</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>FRRS1      </td><td>K562</td><td> 0.046533983</td><td>5.348459e-01</td><td>0.37993890</td><td>0.66259885</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>GPR88      </td><td>K562</td><td>-0.036878412</td><td>6.424240e-01</td><td>0.37993890</td><td>0.08259299</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>MFSD14A    </td><td>K562</td><td>-0.078276519</td><td>2.213276e-01</td><td>0.37993890</td><td>0.39387875</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>RTCA       </td><td>K562</td><td> 0.095204521</td><td>1.143032e-01</td><td>0.37993890</td><td>0.68192912</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>RTCA-AS1   </td><td>K562</td><td> 0.088975803</td><td>1.481363e-01</td><td>0.37993890</td><td>0.46983306</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>SASS6      </td><td>K562</td><td> 0.044605623</td><td>5.563739e-01</td><td>0.37993890</td><td>0.82192717</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>SLC30A7    </td><td>K562</td><td> 0.038767280</td><td>6.214780e-01</td><td>0.37993890</td><td>0.79371278</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>SLC35A3    </td><td>K562</td><td> 0.030231768</td><td>7.142782e-01</td><td>0.37993890</td><td>0.62935663</td></tr>\n",
       "\t<tr><td>chr1</td><td>100010532</td><td>100010875</td><td>TRMT13     </td><td>K562</td><td>-0.013272558</td><td>8.810266e-01</td><td>0.37993890</td><td>0.71936933</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001440</td><td> 10001830</td><td>APITD1-CORT</td><td>K562</td><td> 0.022895609</td><td>7.894790e-01</td><td>0.24131570</td><td>0.27804354</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001440</td><td> 10001830</td><td>CENPS      </td><td>K562</td><td> 0.016956156</td><td>8.466783e-01</td><td>0.24131570</td><td>0.31777799</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001440</td><td> 10001830</td><td>CLSTN1     </td><td>K562</td><td>-0.073728949</td><td>2.582968e-01</td><td>0.24131570</td><td>0.86278434</td></tr>\n",
       "\t<tr><td>chr1</td><td> 10001440</td><td> 10001830</td><td>CTNNBIP1   </td><td>K562</td><td> 0.009326234</td><td>9.176315e-01</td><td>0.24131570</td><td>0.69496241</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9992130</td><td> 9992304</td><td>WWC3     </td><td>K562</td><td>-0.045365230</td><td>5.478820e-01</td><td>0.09242842</td><td>0.7768232</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>AKAP17A  </td><td>K562</td><td>-0.196275013</td><td>1.042572e-04</td><td>0.77965098</td><td>0.6925217</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>ASMT     </td><td>K562</td><td> 0.051036502</td><td>4.847480e-01</td><td>0.77965098</td><td>0.2818022</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>ASMTL    </td><td>K562</td><td> 0.187692499</td><td>2.337261e-04</td><td>0.77965098</td><td>0.6730450</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>CSF2RA   </td><td>K562</td><td>-0.110197868</td><td>5.663202e-02</td><td>0.77965098</td><td>0.9560676</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>GTPBP6   </td><td>K562</td><td>-0.108812752</td><td>6.071156e-02</td><td>0.77965098</td><td>0.6292590</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>IL3RA    </td><td>K562</td><td> 0.168152813</td><td>1.263510e-03</td><td>0.77965098</td><td>0.7737967</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>LINC00106</td><td>K562</td><td> 0.047051105</td><td>5.290504e-01</td><td>0.77965098</td><td>0.5317778</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>P2RY8    </td><td>K562</td><td> 0.017929530</td><td>8.375119e-01</td><td>0.77965098</td><td>0.9605096</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>PLCXD1   </td><td>K562</td><td> 0.050131514</td><td>4.947555e-01</td><td>0.77965098</td><td>0.6108074</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>PPP2R3B  </td><td>K562</td><td> 0.543813217</td><td>1.448679e-37</td><td>0.77965098</td><td>0.5742946</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>SHOX     </td><td>K562</td><td> 0.104018319</td><td>7.666967e-02</td><td>0.77965098</td><td>0.2214683</td></tr>\n",
       "\t<tr><td>chrX</td><td>  999313</td><td> 1000203</td><td>SLC25A6  </td><td>K562</td><td> 0.027911947</td><td>7.385331e-01</td><td>0.77965098</td><td>0.9036415</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>CLCN4    </td><td>K562</td><td>-0.163316293</td><td>1.857445e-03</td><td>0.87934770</td><td>0.8246119</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>FAM9B    </td><td>K562</td><td> 0.057544579</td><td>4.142582e-01</td><td>0.87934770</td><td>0.3804549</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>GPR143   </td><td>K562</td><td>-0.002817197</td><td>9.761571e-01</td><td>0.87934770</td><td>0.2968857</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>SHROOM2  </td><td>K562</td><td> 0.000200184</td><td>9.983300e-01</td><td>0.87934770</td><td>0.3440886</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>TBL1X    </td><td>K562</td><td>-0.172609478</td><td>8.757065e-04</td><td>0.87934770</td><td>0.9632920</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9995566</td><td> 9996680</td><td>WWC3     </td><td>K562</td><td>-0.038295659</td><td>6.267039e-01</td><td>0.87934770</td><td>0.7768232</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>CLCN4    </td><td>K562</td><td>-0.117502719</td><td>3.860398e-02</td><td>0.77259177</td><td>0.8246119</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>GPR143   </td><td>K562</td><td>-0.011763331</td><td>8.950476e-01</td><td>0.77259177</td><td>0.2968857</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>SHROOM2  </td><td>K562</td><td> 0.037036534</td><td>6.406602e-01</td><td>0.77259177</td><td>0.3440886</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>TBL1X    </td><td>K562</td><td> 0.136142029</td><td>1.283083e-02</td><td>0.77259177</td><td>0.9632920</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9996996</td><td> 9998157</td><td>WWC3     </td><td>K562</td><td> 0.014737462</td><td>8.673974e-01</td><td>0.77259177</td><td>0.7768232</td></tr>\n",
       "\t<tr><td>chrX</td><td>99980528</td><td>99980848</td><td>CSTF2    </td><td>K562</td><td>-0.006521322</td><td>9.433252e-01</td><td>0.01498809</td><td>0.4762277</td></tr>\n",
       "\t<tr><td>chrX</td><td>99980528</td><td>99980848</td><td>PCDH19   </td><td>K562</td><td>-0.016954735</td><td>8.466919e-01</td><td>0.01498809</td><td>0.2050181</td></tr>\n",
       "\t<tr><td>chrX</td><td>99980528</td><td>99980848</td><td>TSPAN6   </td><td>K562</td><td>-0.033367486</td><td>6.807988e-01</td><td>0.01498809</td><td>0.2488529</td></tr>\n",
       "\t<tr><td>chrX</td><td>99980528</td><td>99980848</td><td>XKRX     </td><td>K562</td><td>-0.025512047</td><td>7.632338e-01</td><td>0.01498809</td><td>0.3397930</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9999312</td><td> 9999592</td><td>TBL1X    </td><td>K562</td><td>-0.090474615</td><td>1.394174e-01</td><td>0.16197132</td><td>0.9632920</td></tr>\n",
       "\t<tr><td>chrX</td><td> 9999312</td><td> 9999592</td><td>WWC3     </td><td>K562</td><td> 0.068492027</td><td>3.049001e-01</td><td>0.16197132</td><td>0.7768232</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 9765178 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " chr & start & end & TargetGene & CellType & Correlation & FDR & VarQATAC & VarQRNA\\\\\n",
       " <fct> & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr1 & 100004103 & 100004301 & AGL         & K562 & -0.007710668 & 9.324562e-01 & 0.03058681 & 0.78521917\\\\\n",
       "\t chr1 & 100004103 & 100004301 & CDC14A      & K562 & -0.145382884 & 6.950791e-03 & 0.03058681 & 0.97881480\\\\\n",
       "\t chr1 & 100004103 & 100004301 & DBT         & K562 &  0.034985805 & 6.632382e-01 & 0.03058681 & 0.63213902\\\\\n",
       "\t chr1 & 100004103 & 100004301 & EXTL2       & K562 & -0.055170442 & 4.396358e-01 & 0.03058681 & 0.51020209\\\\\n",
       "\t chr1 & 100004103 & 100004301 & FRRS1       & K562 &  0.108531036 & 6.157165e-02 & 0.03058681 & 0.66259885\\\\\n",
       "\t chr1 & 100004103 & 100004301 & MFSD14A     & K562 &  0.034960156 & 6.635091e-01 & 0.03058681 & 0.39387875\\\\\n",
       "\t chr1 & 100004103 & 100004301 & PLPPR5      & K562 & -0.035251693 & 6.603494e-01 & 0.03058681 & 0.33876794\\\\\n",
       "\t chr1 & 100004103 & 100004301 & RTCA        & K562 &  0.063551546 & 3.524594e-01 & 0.03058681 & 0.68192912\\\\\n",
       "\t chr1 & 100004103 & 100004301 & RTCA-AS1    & K562 &  0.123300150 & 2.794338e-02 & 0.03058681 & 0.46983306\\\\\n",
       "\t chr1 & 100004103 & 100004301 & SASS6       & K562 &  0.219797819 & 9.197888e-06 & 0.03058681 & 0.82192717\\\\\n",
       "\t chr1 & 100004103 & 100004301 & SLC30A7     & K562 & -0.064463623 & 3.434446e-01 & 0.03058681 & 0.79371278\\\\\n",
       "\t chr1 & 100004103 & 100004301 & SLC35A3     & K562 & -0.052098762 & 4.730377e-01 & 0.03058681 & 0.62935663\\\\\n",
       "\t chr1 & 100004103 & 100004301 & TRMT13      & K562 & -0.005524069 & 9.522898e-01 & 0.03058681 & 0.71936933\\\\\n",
       "\t chr1 & 100010532 & 100010875 & AGL         & K562 & -0.011651441 & 8.960816e-01 & 0.37993890 & 0.78521917\\\\\n",
       "\t chr1 & 100010532 & 100010875 & CDC14A      & K562 & -0.005387489 & 9.534942e-01 & 0.37993890 & 0.97881480\\\\\n",
       "\t chr1 & 100010532 & 100010875 & DBT         & K562 &  0.026903933 & 7.489431e-01 & 0.37993890 & 0.63213902\\\\\n",
       "\t chr1 & 100010532 & 100010875 & EXTL2       & K562 &  0.077011962 & 2.312588e-01 & 0.37993890 & 0.51020209\\\\\n",
       "\t chr1 & 100010532 & 100010875 & FRRS1       & K562 &  0.046533983 & 5.348459e-01 & 0.37993890 & 0.66259885\\\\\n",
       "\t chr1 & 100010532 & 100010875 & GPR88       & K562 & -0.036878412 & 6.424240e-01 & 0.37993890 & 0.08259299\\\\\n",
       "\t chr1 & 100010532 & 100010875 & MFSD14A     & K562 & -0.078276519 & 2.213276e-01 & 0.37993890 & 0.39387875\\\\\n",
       "\t chr1 & 100010532 & 100010875 & RTCA        & K562 &  0.095204521 & 1.143032e-01 & 0.37993890 & 0.68192912\\\\\n",
       "\t chr1 & 100010532 & 100010875 & RTCA-AS1    & K562 &  0.088975803 & 1.481363e-01 & 0.37993890 & 0.46983306\\\\\n",
       "\t chr1 & 100010532 & 100010875 & SASS6       & K562 &  0.044605623 & 5.563739e-01 & 0.37993890 & 0.82192717\\\\\n",
       "\t chr1 & 100010532 & 100010875 & SLC30A7     & K562 &  0.038767280 & 6.214780e-01 & 0.37993890 & 0.79371278\\\\\n",
       "\t chr1 & 100010532 & 100010875 & SLC35A3     & K562 &  0.030231768 & 7.142782e-01 & 0.37993890 & 0.62935663\\\\\n",
       "\t chr1 & 100010532 & 100010875 & TRMT13      & K562 & -0.013272558 & 8.810266e-01 & 0.37993890 & 0.71936933\\\\\n",
       "\t chr1 &  10001440 &  10001830 & APITD1-CORT & K562 &  0.022895609 & 7.894790e-01 & 0.24131570 & 0.27804354\\\\\n",
       "\t chr1 &  10001440 &  10001830 & CENPS       & K562 &  0.016956156 & 8.466783e-01 & 0.24131570 & 0.31777799\\\\\n",
       "\t chr1 &  10001440 &  10001830 & CLSTN1      & K562 & -0.073728949 & 2.582968e-01 & 0.24131570 & 0.86278434\\\\\n",
       "\t chr1 &  10001440 &  10001830 & CTNNBIP1    & K562 &  0.009326234 & 9.176315e-01 & 0.24131570 & 0.69496241\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t chrX &  9992130 &  9992304 & WWC3      & K562 & -0.045365230 & 5.478820e-01 & 0.09242842 & 0.7768232\\\\\n",
       "\t chrX &   999313 &  1000203 & AKAP17A   & K562 & -0.196275013 & 1.042572e-04 & 0.77965098 & 0.6925217\\\\\n",
       "\t chrX &   999313 &  1000203 & ASMT      & K562 &  0.051036502 & 4.847480e-01 & 0.77965098 & 0.2818022\\\\\n",
       "\t chrX &   999313 &  1000203 & ASMTL     & K562 &  0.187692499 & 2.337261e-04 & 0.77965098 & 0.6730450\\\\\n",
       "\t chrX &   999313 &  1000203 & CSF2RA    & K562 & -0.110197868 & 5.663202e-02 & 0.77965098 & 0.9560676\\\\\n",
       "\t chrX &   999313 &  1000203 & GTPBP6    & K562 & -0.108812752 & 6.071156e-02 & 0.77965098 & 0.6292590\\\\\n",
       "\t chrX &   999313 &  1000203 & IL3RA     & K562 &  0.168152813 & 1.263510e-03 & 0.77965098 & 0.7737967\\\\\n",
       "\t chrX &   999313 &  1000203 & LINC00106 & K562 &  0.047051105 & 5.290504e-01 & 0.77965098 & 0.5317778\\\\\n",
       "\t chrX &   999313 &  1000203 & P2RY8     & K562 &  0.017929530 & 8.375119e-01 & 0.77965098 & 0.9605096\\\\\n",
       "\t chrX &   999313 &  1000203 & PLCXD1    & K562 &  0.050131514 & 4.947555e-01 & 0.77965098 & 0.6108074\\\\\n",
       "\t chrX &   999313 &  1000203 & PPP2R3B   & K562 &  0.543813217 & 1.448679e-37 & 0.77965098 & 0.5742946\\\\\n",
       "\t chrX &   999313 &  1000203 & SHOX      & K562 &  0.104018319 & 7.666967e-02 & 0.77965098 & 0.2214683\\\\\n",
       "\t chrX &   999313 &  1000203 & SLC25A6   & K562 &  0.027911947 & 7.385331e-01 & 0.77965098 & 0.9036415\\\\\n",
       "\t chrX &  9995566 &  9996680 & CLCN4     & K562 & -0.163316293 & 1.857445e-03 & 0.87934770 & 0.8246119\\\\\n",
       "\t chrX &  9995566 &  9996680 & FAM9B     & K562 &  0.057544579 & 4.142582e-01 & 0.87934770 & 0.3804549\\\\\n",
       "\t chrX &  9995566 &  9996680 & GPR143    & K562 & -0.002817197 & 9.761571e-01 & 0.87934770 & 0.2968857\\\\\n",
       "\t chrX &  9995566 &  9996680 & SHROOM2   & K562 &  0.000200184 & 9.983300e-01 & 0.87934770 & 0.3440886\\\\\n",
       "\t chrX &  9995566 &  9996680 & TBL1X     & K562 & -0.172609478 & 8.757065e-04 & 0.87934770 & 0.9632920\\\\\n",
       "\t chrX &  9995566 &  9996680 & WWC3      & K562 & -0.038295659 & 6.267039e-01 & 0.87934770 & 0.7768232\\\\\n",
       "\t chrX &  9996996 &  9998157 & CLCN4     & K562 & -0.117502719 & 3.860398e-02 & 0.77259177 & 0.8246119\\\\\n",
       "\t chrX &  9996996 &  9998157 & GPR143    & K562 & -0.011763331 & 8.950476e-01 & 0.77259177 & 0.2968857\\\\\n",
       "\t chrX &  9996996 &  9998157 & SHROOM2   & K562 &  0.037036534 & 6.406602e-01 & 0.77259177 & 0.3440886\\\\\n",
       "\t chrX &  9996996 &  9998157 & TBL1X     & K562 &  0.136142029 & 1.283083e-02 & 0.77259177 & 0.9632920\\\\\n",
       "\t chrX &  9996996 &  9998157 & WWC3      & K562 &  0.014737462 & 8.673974e-01 & 0.77259177 & 0.7768232\\\\\n",
       "\t chrX & 99980528 & 99980848 & CSTF2     & K562 & -0.006521322 & 9.433252e-01 & 0.01498809 & 0.4762277\\\\\n",
       "\t chrX & 99980528 & 99980848 & PCDH19    & K562 & -0.016954735 & 8.466919e-01 & 0.01498809 & 0.2050181\\\\\n",
       "\t chrX & 99980528 & 99980848 & TSPAN6    & K562 & -0.033367486 & 6.807988e-01 & 0.01498809 & 0.2488529\\\\\n",
       "\t chrX & 99980528 & 99980848 & XKRX      & K562 & -0.025512047 & 7.632338e-01 & 0.01498809 & 0.3397930\\\\\n",
       "\t chrX &  9999312 &  9999592 & TBL1X     & K562 & -0.090474615 & 1.394174e-01 & 0.16197132 & 0.9632920\\\\\n",
       "\t chrX &  9999312 &  9999592 & WWC3      & K562 &  0.068492027 & 3.049001e-01 & 0.16197132 & 0.7768232\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 9765178 × 9\n",
       "\n",
       "| chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | TargetGene &lt;chr&gt; | CellType &lt;chr&gt; | Correlation &lt;dbl&gt; | FDR &lt;dbl&gt; | VarQATAC &lt;dbl&gt; | VarQRNA &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| chr1 | 100004103 | 100004301 | AGL         | K562 | -0.007710668 | 9.324562e-01 | 0.03058681 | 0.78521917 |\n",
       "| chr1 | 100004103 | 100004301 | CDC14A      | K562 | -0.145382884 | 6.950791e-03 | 0.03058681 | 0.97881480 |\n",
       "| chr1 | 100004103 | 100004301 | DBT         | K562 |  0.034985805 | 6.632382e-01 | 0.03058681 | 0.63213902 |\n",
       "| chr1 | 100004103 | 100004301 | EXTL2       | K562 | -0.055170442 | 4.396358e-01 | 0.03058681 | 0.51020209 |\n",
       "| chr1 | 100004103 | 100004301 | FRRS1       | K562 |  0.108531036 | 6.157165e-02 | 0.03058681 | 0.66259885 |\n",
       "| chr1 | 100004103 | 100004301 | MFSD14A     | K562 |  0.034960156 | 6.635091e-01 | 0.03058681 | 0.39387875 |\n",
       "| chr1 | 100004103 | 100004301 | PLPPR5      | K562 | -0.035251693 | 6.603494e-01 | 0.03058681 | 0.33876794 |\n",
       "| chr1 | 100004103 | 100004301 | RTCA        | K562 |  0.063551546 | 3.524594e-01 | 0.03058681 | 0.68192912 |\n",
       "| chr1 | 100004103 | 100004301 | RTCA-AS1    | K562 |  0.123300150 | 2.794338e-02 | 0.03058681 | 0.46983306 |\n",
       "| chr1 | 100004103 | 100004301 | SASS6       | K562 |  0.219797819 | 9.197888e-06 | 0.03058681 | 0.82192717 |\n",
       "| chr1 | 100004103 | 100004301 | SLC30A7     | K562 | -0.064463623 | 3.434446e-01 | 0.03058681 | 0.79371278 |\n",
       "| chr1 | 100004103 | 100004301 | SLC35A3     | K562 | -0.052098762 | 4.730377e-01 | 0.03058681 | 0.62935663 |\n",
       "| chr1 | 100004103 | 100004301 | TRMT13      | K562 | -0.005524069 | 9.522898e-01 | 0.03058681 | 0.71936933 |\n",
       "| chr1 | 100010532 | 100010875 | AGL         | K562 | -0.011651441 | 8.960816e-01 | 0.37993890 | 0.78521917 |\n",
       "| chr1 | 100010532 | 100010875 | CDC14A      | K562 | -0.005387489 | 9.534942e-01 | 0.37993890 | 0.97881480 |\n",
       "| chr1 | 100010532 | 100010875 | DBT         | K562 |  0.026903933 | 7.489431e-01 | 0.37993890 | 0.63213902 |\n",
       "| chr1 | 100010532 | 100010875 | EXTL2       | K562 |  0.077011962 | 2.312588e-01 | 0.37993890 | 0.51020209 |\n",
       "| chr1 | 100010532 | 100010875 | FRRS1       | K562 |  0.046533983 | 5.348459e-01 | 0.37993890 | 0.66259885 |\n",
       "| chr1 | 100010532 | 100010875 | GPR88       | K562 | -0.036878412 | 6.424240e-01 | 0.37993890 | 0.08259299 |\n",
       "| chr1 | 100010532 | 100010875 | MFSD14A     | K562 | -0.078276519 | 2.213276e-01 | 0.37993890 | 0.39387875 |\n",
       "| chr1 | 100010532 | 100010875 | RTCA        | K562 |  0.095204521 | 1.143032e-01 | 0.37993890 | 0.68192912 |\n",
       "| chr1 | 100010532 | 100010875 | RTCA-AS1    | K562 |  0.088975803 | 1.481363e-01 | 0.37993890 | 0.46983306 |\n",
       "| chr1 | 100010532 | 100010875 | SASS6       | K562 |  0.044605623 | 5.563739e-01 | 0.37993890 | 0.82192717 |\n",
       "| chr1 | 100010532 | 100010875 | SLC30A7     | K562 |  0.038767280 | 6.214780e-01 | 0.37993890 | 0.79371278 |\n",
       "| chr1 | 100010532 | 100010875 | SLC35A3     | K562 |  0.030231768 | 7.142782e-01 | 0.37993890 | 0.62935663 |\n",
       "| chr1 | 100010532 | 100010875 | TRMT13      | K562 | -0.013272558 | 8.810266e-01 | 0.37993890 | 0.71936933 |\n",
       "| chr1 |  10001440 |  10001830 | APITD1-CORT | K562 |  0.022895609 | 7.894790e-01 | 0.24131570 | 0.27804354 |\n",
       "| chr1 |  10001440 |  10001830 | CENPS       | K562 |  0.016956156 | 8.466783e-01 | 0.24131570 | 0.31777799 |\n",
       "| chr1 |  10001440 |  10001830 | CLSTN1      | K562 | -0.073728949 | 2.582968e-01 | 0.24131570 | 0.86278434 |\n",
       "| chr1 |  10001440 |  10001830 | CTNNBIP1    | K562 |  0.009326234 | 9.176315e-01 | 0.24131570 | 0.69496241 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| chrX |  9992130 |  9992304 | WWC3      | K562 | -0.045365230 | 5.478820e-01 | 0.09242842 | 0.7768232 |\n",
       "| chrX |   999313 |  1000203 | AKAP17A   | K562 | -0.196275013 | 1.042572e-04 | 0.77965098 | 0.6925217 |\n",
       "| chrX |   999313 |  1000203 | ASMT      | K562 |  0.051036502 | 4.847480e-01 | 0.77965098 | 0.2818022 |\n",
       "| chrX |   999313 |  1000203 | ASMTL     | K562 |  0.187692499 | 2.337261e-04 | 0.77965098 | 0.6730450 |\n",
       "| chrX |   999313 |  1000203 | CSF2RA    | K562 | -0.110197868 | 5.663202e-02 | 0.77965098 | 0.9560676 |\n",
       "| chrX |   999313 |  1000203 | GTPBP6    | K562 | -0.108812752 | 6.071156e-02 | 0.77965098 | 0.6292590 |\n",
       "| chrX |   999313 |  1000203 | IL3RA     | K562 |  0.168152813 | 1.263510e-03 | 0.77965098 | 0.7737967 |\n",
       "| chrX |   999313 |  1000203 | LINC00106 | K562 |  0.047051105 | 5.290504e-01 | 0.77965098 | 0.5317778 |\n",
       "| chrX |   999313 |  1000203 | P2RY8     | K562 |  0.017929530 | 8.375119e-01 | 0.77965098 | 0.9605096 |\n",
       "| chrX |   999313 |  1000203 | PLCXD1    | K562 |  0.050131514 | 4.947555e-01 | 0.77965098 | 0.6108074 |\n",
       "| chrX |   999313 |  1000203 | PPP2R3B   | K562 |  0.543813217 | 1.448679e-37 | 0.77965098 | 0.5742946 |\n",
       "| chrX |   999313 |  1000203 | SHOX      | K562 |  0.104018319 | 7.666967e-02 | 0.77965098 | 0.2214683 |\n",
       "| chrX |   999313 |  1000203 | SLC25A6   | K562 |  0.027911947 | 7.385331e-01 | 0.77965098 | 0.9036415 |\n",
       "| chrX |  9995566 |  9996680 | CLCN4     | K562 | -0.163316293 | 1.857445e-03 | 0.87934770 | 0.8246119 |\n",
       "| chrX |  9995566 |  9996680 | FAM9B     | K562 |  0.057544579 | 4.142582e-01 | 0.87934770 | 0.3804549 |\n",
       "| chrX |  9995566 |  9996680 | GPR143    | K562 | -0.002817197 | 9.761571e-01 | 0.87934770 | 0.2968857 |\n",
       "| chrX |  9995566 |  9996680 | SHROOM2   | K562 |  0.000200184 | 9.983300e-01 | 0.87934770 | 0.3440886 |\n",
       "| chrX |  9995566 |  9996680 | TBL1X     | K562 | -0.172609478 | 8.757065e-04 | 0.87934770 | 0.9632920 |\n",
       "| chrX |  9995566 |  9996680 | WWC3      | K562 | -0.038295659 | 6.267039e-01 | 0.87934770 | 0.7768232 |\n",
       "| chrX |  9996996 |  9998157 | CLCN4     | K562 | -0.117502719 | 3.860398e-02 | 0.77259177 | 0.8246119 |\n",
       "| chrX |  9996996 |  9998157 | GPR143    | K562 | -0.011763331 | 8.950476e-01 | 0.77259177 | 0.2968857 |\n",
       "| chrX |  9996996 |  9998157 | SHROOM2   | K562 |  0.037036534 | 6.406602e-01 | 0.77259177 | 0.3440886 |\n",
       "| chrX |  9996996 |  9998157 | TBL1X     | K562 |  0.136142029 | 1.283083e-02 | 0.77259177 | 0.9632920 |\n",
       "| chrX |  9996996 |  9998157 | WWC3      | K562 |  0.014737462 | 8.673974e-01 | 0.77259177 | 0.7768232 |\n",
       "| chrX | 99980528 | 99980848 | CSTF2     | K562 | -0.006521322 | 9.433252e-01 | 0.01498809 | 0.4762277 |\n",
       "| chrX | 99980528 | 99980848 | PCDH19    | K562 | -0.016954735 | 8.466919e-01 | 0.01498809 | 0.2050181 |\n",
       "| chrX | 99980528 | 99980848 | TSPAN6    | K562 | -0.033367486 | 6.807988e-01 | 0.01498809 | 0.2488529 |\n",
       "| chrX | 99980528 | 99980848 | XKRX      | K562 | -0.025512047 | 7.632338e-01 | 0.01498809 | 0.3397930 |\n",
       "| chrX |  9999312 |  9999592 | TBL1X     | K562 | -0.090474615 | 1.394174e-01 | 0.16197132 | 0.9632920 |\n",
       "| chrX |  9999312 |  9999592 | WWC3      | K562 |  0.068492027 | 3.049001e-01 | 0.16197132 | 0.7768232 |\n",
       "\n"
      ],
      "text/plain": [
       "        chr  start     end       TargetGene  CellType Correlation  FDR         \n",
       "1       chr1 100004103 100004301 AGL         K562     -0.007710668 9.324562e-01\n",
       "2       chr1 100004103 100004301 CDC14A      K562     -0.145382884 6.950791e-03\n",
       "3       chr1 100004103 100004301 DBT         K562      0.034985805 6.632382e-01\n",
       "4       chr1 100004103 100004301 EXTL2       K562     -0.055170442 4.396358e-01\n",
       "5       chr1 100004103 100004301 FRRS1       K562      0.108531036 6.157165e-02\n",
       "6       chr1 100004103 100004301 MFSD14A     K562      0.034960156 6.635091e-01\n",
       "7       chr1 100004103 100004301 PLPPR5      K562     -0.035251693 6.603494e-01\n",
       "8       chr1 100004103 100004301 RTCA        K562      0.063551546 3.524594e-01\n",
       "9       chr1 100004103 100004301 RTCA-AS1    K562      0.123300150 2.794338e-02\n",
       "10      chr1 100004103 100004301 SASS6       K562      0.219797819 9.197888e-06\n",
       "11      chr1 100004103 100004301 SLC30A7     K562     -0.064463623 3.434446e-01\n",
       "12      chr1 100004103 100004301 SLC35A3     K562     -0.052098762 4.730377e-01\n",
       "13      chr1 100004103 100004301 TRMT13      K562     -0.005524069 9.522898e-01\n",
       "14      chr1 100010532 100010875 AGL         K562     -0.011651441 8.960816e-01\n",
       "15      chr1 100010532 100010875 CDC14A      K562     -0.005387489 9.534942e-01\n",
       "16      chr1 100010532 100010875 DBT         K562      0.026903933 7.489431e-01\n",
       "17      chr1 100010532 100010875 EXTL2       K562      0.077011962 2.312588e-01\n",
       "18      chr1 100010532 100010875 FRRS1       K562      0.046533983 5.348459e-01\n",
       "19      chr1 100010532 100010875 GPR88       K562     -0.036878412 6.424240e-01\n",
       "20      chr1 100010532 100010875 MFSD14A     K562     -0.078276519 2.213276e-01\n",
       "21      chr1 100010532 100010875 RTCA        K562      0.095204521 1.143032e-01\n",
       "22      chr1 100010532 100010875 RTCA-AS1    K562      0.088975803 1.481363e-01\n",
       "23      chr1 100010532 100010875 SASS6       K562      0.044605623 5.563739e-01\n",
       "24      chr1 100010532 100010875 SLC30A7     K562      0.038767280 6.214780e-01\n",
       "25      chr1 100010532 100010875 SLC35A3     K562      0.030231768 7.142782e-01\n",
       "26      chr1 100010532 100010875 TRMT13      K562     -0.013272558 8.810266e-01\n",
       "27      chr1  10001440  10001830 APITD1-CORT K562      0.022895609 7.894790e-01\n",
       "28      chr1  10001440  10001830 CENPS       K562      0.016956156 8.466783e-01\n",
       "29      chr1  10001440  10001830 CLSTN1      K562     -0.073728949 2.582968e-01\n",
       "30      chr1  10001440  10001830 CTNNBIP1    K562      0.009326234 9.176315e-01\n",
       "⋮       ⋮    ⋮         ⋮         ⋮           ⋮        ⋮            ⋮           \n",
       "9765149 chrX  9992130   9992304  WWC3        K562     -0.045365230 5.478820e-01\n",
       "9765150 chrX   999313   1000203  AKAP17A     K562     -0.196275013 1.042572e-04\n",
       "9765151 chrX   999313   1000203  ASMT        K562      0.051036502 4.847480e-01\n",
       "9765152 chrX   999313   1000203  ASMTL       K562      0.187692499 2.337261e-04\n",
       "9765153 chrX   999313   1000203  CSF2RA      K562     -0.110197868 5.663202e-02\n",
       "9765154 chrX   999313   1000203  GTPBP6      K562     -0.108812752 6.071156e-02\n",
       "9765155 chrX   999313   1000203  IL3RA       K562      0.168152813 1.263510e-03\n",
       "9765156 chrX   999313   1000203  LINC00106   K562      0.047051105 5.290504e-01\n",
       "9765157 chrX   999313   1000203  P2RY8       K562      0.017929530 8.375119e-01\n",
       "9765158 chrX   999313   1000203  PLCXD1      K562      0.050131514 4.947555e-01\n",
       "9765159 chrX   999313   1000203  PPP2R3B     K562      0.543813217 1.448679e-37\n",
       "9765160 chrX   999313   1000203  SHOX        K562      0.104018319 7.666967e-02\n",
       "9765161 chrX   999313   1000203  SLC25A6     K562      0.027911947 7.385331e-01\n",
       "9765162 chrX  9995566   9996680  CLCN4       K562     -0.163316293 1.857445e-03\n",
       "9765163 chrX  9995566   9996680  FAM9B       K562      0.057544579 4.142582e-01\n",
       "9765164 chrX  9995566   9996680  GPR143      K562     -0.002817197 9.761571e-01\n",
       "9765165 chrX  9995566   9996680  SHROOM2     K562      0.000200184 9.983300e-01\n",
       "9765166 chrX  9995566   9996680  TBL1X       K562     -0.172609478 8.757065e-04\n",
       "9765167 chrX  9995566   9996680  WWC3        K562     -0.038295659 6.267039e-01\n",
       "9765168 chrX  9996996   9998157  CLCN4       K562     -0.117502719 3.860398e-02\n",
       "9765169 chrX  9996996   9998157  GPR143      K562     -0.011763331 8.950476e-01\n",
       "9765170 chrX  9996996   9998157  SHROOM2     K562      0.037036534 6.406602e-01\n",
       "9765171 chrX  9996996   9998157  TBL1X       K562      0.136142029 1.283083e-02\n",
       "9765172 chrX  9996996   9998157  WWC3        K562      0.014737462 8.673974e-01\n",
       "9765173 chrX 99980528  99980848  CSTF2       K562     -0.006521322 9.433252e-01\n",
       "9765174 chrX 99980528  99980848  PCDH19      K562     -0.016954735 8.466919e-01\n",
       "9765175 chrX 99980528  99980848  TSPAN6      K562     -0.033367486 6.807988e-01\n",
       "9765176 chrX 99980528  99980848  XKRX        K562     -0.025512047 7.632338e-01\n",
       "9765177 chrX  9999312   9999592  TBL1X       K562     -0.090474615 1.394174e-01\n",
       "9765178 chrX  9999312   9999592  WWC3        K562      0.068492027 3.049001e-01\n",
       "        VarQATAC   VarQRNA   \n",
       "1       0.03058681 0.78521917\n",
       "2       0.03058681 0.97881480\n",
       "3       0.03058681 0.63213902\n",
       "4       0.03058681 0.51020209\n",
       "5       0.03058681 0.66259885\n",
       "6       0.03058681 0.39387875\n",
       "7       0.03058681 0.33876794\n",
       "8       0.03058681 0.68192912\n",
       "9       0.03058681 0.46983306\n",
       "10      0.03058681 0.82192717\n",
       "11      0.03058681 0.79371278\n",
       "12      0.03058681 0.62935663\n",
       "13      0.03058681 0.71936933\n",
       "14      0.37993890 0.78521917\n",
       "15      0.37993890 0.97881480\n",
       "16      0.37993890 0.63213902\n",
       "17      0.37993890 0.51020209\n",
       "18      0.37993890 0.66259885\n",
       "19      0.37993890 0.08259299\n",
       "20      0.37993890 0.39387875\n",
       "21      0.37993890 0.68192912\n",
       "22      0.37993890 0.46983306\n",
       "23      0.37993890 0.82192717\n",
       "24      0.37993890 0.79371278\n",
       "25      0.37993890 0.62935663\n",
       "26      0.37993890 0.71936933\n",
       "27      0.24131570 0.27804354\n",
       "28      0.24131570 0.31777799\n",
       "29      0.24131570 0.86278434\n",
       "30      0.24131570 0.69496241\n",
       "⋮       ⋮          ⋮         \n",
       "9765149 0.09242842 0.7768232 \n",
       "9765150 0.77965098 0.6925217 \n",
       "9765151 0.77965098 0.2818022 \n",
       "9765152 0.77965098 0.6730450 \n",
       "9765153 0.77965098 0.9560676 \n",
       "9765154 0.77965098 0.6292590 \n",
       "9765155 0.77965098 0.7737967 \n",
       "9765156 0.77965098 0.5317778 \n",
       "9765157 0.77965098 0.9605096 \n",
       "9765158 0.77965098 0.6108074 \n",
       "9765159 0.77965098 0.5742946 \n",
       "9765160 0.77965098 0.2214683 \n",
       "9765161 0.77965098 0.9036415 \n",
       "9765162 0.87934770 0.8246119 \n",
       "9765163 0.87934770 0.3804549 \n",
       "9765164 0.87934770 0.2968857 \n",
       "9765165 0.87934770 0.3440886 \n",
       "9765166 0.87934770 0.9632920 \n",
       "9765167 0.87934770 0.7768232 \n",
       "9765168 0.77259177 0.8246119 \n",
       "9765169 0.77259177 0.2968857 \n",
       "9765170 0.77259177 0.3440886 \n",
       "9765171 0.77259177 0.9632920 \n",
       "9765172 0.77259177 0.7768232 \n",
       "9765173 0.01498809 0.4762277 \n",
       "9765174 0.01498809 0.2050181 \n",
       "9765175 0.01498809 0.2488529 \n",
       "9765176 0.01498809 0.3397930 \n",
       "9765177 0.16197132 0.9632920 \n",
       "9765178 0.16197132 0.7768232 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveRDS(obj.ArchR,\n",
    "        paste(dir.output,\"obj.ArchR.rds\",sep = \"/\"))\n",
    "saveRDS(pairs.E2G.res,\n",
    "        paste(dir.output,\"pairs.E2G.res.rds\",sep = \"/\"))\n",
    "df.output = as.data.frame(pairs.E2G.res,row.names = NULL)\n",
    "colnames(df.output)[1] = \"chr\"\n",
    "df.output[,\"CellType\"] = \"K562\"\n",
    "df.output = df.output[,c(\"chr\",\n",
    "                         \"start\",\n",
    "                         \"end\",\n",
    "                         \"TargetGene\",\n",
    "                         \"CellType\",\n",
    "                         \"Correlation\",\n",
    "                         \"FDR\",\n",
    "                         \"VarQATAC\",\n",
    "                         \"VarQRNA\")]\n",
    "data.table::fwrite(df.output,\n",
    "                   file = paste(dir.output,\"pairs.E2G.res.tsv.gz\",sep = \"/\"),\n",
    "                   row.names = F,\n",
    "                   quote = F,\n",
    "                   sep = \"\\t\")\n",
    "pairs.E2G.res\n",
    "df.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4c909ed5-7321-4e69-8208-1544b371da6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Red Hat Enterprise Linux 8.10 (Ootpa)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /maps/projects/ralab/people/lpm537/software/anaconda3/envs/Notebook_E2G_240505/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Copenhagen\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] dplyr_1.1.4                       uwot_0.1.16                      \n",
       " [3] nabor_0.5.0                       BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.70.1                 XVector_0.42.0                   \n",
       " [9] genomation_1.34.0                 Signac_1.13.0                    \n",
       "[11] Seurat_5.0.3                      SeuratObject_5.0.1               \n",
       "[13] sp_2.1-4                          rhdf5_2.44.0                     \n",
       "[15] SummarizedExperiment_1.30.2       Biobase_2.60.0                   \n",
       "[17] MatrixGenerics_1.12.3             Rcpp_1.0.12                      \n",
       "[19] Matrix_1.6-5                      GenomicRanges_1.54.1             \n",
       "[21] GenomeInfoDb_1.38.1               IRanges_2.36.0                   \n",
       "[23] S4Vectors_0.40.2                  BiocGenerics_0.48.1              \n",
       "[25] matrixStats_1.3.0                 data.table_1.15.2                \n",
       "[27] stringr_1.5.1                     plyr_1.8.9                       \n",
       "[29] magrittr_2.0.3                    ggplot2_3.5.1                    \n",
       "[31] gtable_0.3.5                      gtools_3.9.5                     \n",
       "[33] gridExtra_2.3                     ArchR_1.0.2                      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22         splines_4.3.3            later_1.3.2             \n",
       "  [4] pbdZMQ_0.3-11            BiocIO_1.10.0            bitops_1.0-7            \n",
       "  [7] tibble_3.2.1             polyclip_1.10-6          XML_3.99-0.16.1         \n",
       " [10] fastDummies_1.7.3        lifecycle_1.0.4          globals_0.16.3          \n",
       " [13] lattice_0.22-6           MASS_7.3-60              plotly_4.10.4           \n",
       " [16] plotrix_3.8-4            yaml_2.3.8               httpuv_1.6.15           \n",
       " [19] sctransform_0.4.1        spam_2.10-0              spatstat.sparse_3.0-3   \n",
       " [22] reticulate_1.36.1        cowplot_1.1.3            pbapply_1.7-2           \n",
       " [25] RColorBrewer_1.1-3       abind_1.4-5              zlibbioc_1.48.0         \n",
       " [28] Rtsne_0.17               purrr_1.0.2              RCurl_1.98-1.14         \n",
       " [31] GenomeInfoDbData_1.2.11  ggrepel_0.9.5            irlba_2.3.5.1           \n",
       " [34] listenv_0.9.1            spatstat.utils_3.0-4     goftest_1.2-3           \n",
       " [37] RSpectra_0.16-1          spatstat.random_3.2-3    fitdistrplus_1.1-11     \n",
       " [40] parallelly_1.37.1        leiden_0.4.3.1           codetools_0.2-20        \n",
       " [43] DelayedArray_0.26.7      RcppRoll_0.3.0           tidyselect_1.2.1        \n",
       " [46] farver_2.1.1             base64enc_0.1-3          spatstat.explore_3.2-6  \n",
       " [49] GenomicAlignments_1.36.0 jsonlite_1.8.8           progressr_0.14.0        \n",
       " [52] ggridges_0.5.6           survival_3.6-4           tools_4.3.3             \n",
       " [55] ica_1.0-3                glue_1.7.0               IRdisplay_1.1           \n",
       " [58] withr_3.0.0              fastmap_1.1.1            rhdf5filters_1.12.1     \n",
       " [61] fansi_1.0.6              digest_0.6.35            R6_2.5.1                \n",
       " [64] mime_0.12                seqPattern_1.34.0        colorspace_2.1-0        \n",
       " [67] Cairo_1.6-2              scattermore_1.2          tensor_1.5              \n",
       " [70] spatstat.data_3.0-4      utf8_1.2.4               tidyr_1.3.1             \n",
       " [73] generics_0.1.3           httr_1.4.7               htmlwidgets_1.6.4       \n",
       " [76] S4Arrays_1.2.1           pkgconfig_2.0.3          impute_1.76.0           \n",
       " [79] lmtest_0.9-40            htmltools_0.5.8.1        dotCall64_1.1-1         \n",
       " [82] scales_1.3.0             png_0.1-8                tzdb_0.4.0              \n",
       " [85] reshape2_1.4.4           rjson_0.2.21             uuid_1.2-0              \n",
       " [88] nlme_3.1-164             repr_1.1.7               zoo_1.8-12              \n",
       " [91] KernSmooth_2.23-22       miniUI_0.1.1.1           restfulr_0.0.15         \n",
       " [94] pillar_1.9.0             vctrs_0.6.5              RANN_2.6.1              \n",
       " [97] promises_1.3.0           xtable_1.8-4             cluster_2.1.6           \n",
       "[100] evaluate_0.23            readr_2.1.5              cli_3.6.2               \n",
       "[103] compiler_4.3.3           Rsamtools_2.18.0         rlang_1.1.3             \n",
       "[106] crayon_1.5.2             future.apply_1.11.2      labeling_0.4.3          \n",
       "[109] stringi_1.8.3            gridBase_0.4-7           viridisLite_0.4.2       \n",
       "[112] deldir_2.0-4             BiocParallel_1.36.0      munsell_0.5.1           \n",
       "[115] lazyeval_0.2.2           spatstat.geom_3.2-9      IRkernel_1.3.2          \n",
       "[118] RcppHNSW_0.6.0           hms_1.1.3                patchwork_1.2.0         \n",
       "[121] future_1.33.2            Rhdf5lib_1.22.1          shiny_1.8.1.1           \n",
       "[124] ROCR_1.0-11              igraph_2.0.3             fastmatch_1.1-4         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b54012ba-20d1-4787-8d0d-d851ef1744c4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
